<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
  <title>GeRnika: An R Package for the Simulation, Visualization and Comparison of Tumor Phylogenies</title>

  <meta property="description" itemprop="description" content="The development of methods to study intratumoral heterogeneity and&#10;tumor phylogenies is a highly active area of research. However, the&#10;advancement of these approaches often necessitates access to&#10;substantial amounts of data, which can be challenging and expensive to&#10;acquire. Moreover, the assessment of results requires tools for&#10;visualizing and comparing tumor phylogenies. In this paper, we&#10;introduce GeRnika, an R package designed to address these needs by&#10;enabling the simulation, visualization, and comparison of tumor&#10;evolution data. In summary, GeRnika provides researchers with a&#10;user-friendly tool that facilitates the analysis of their approaches&#10;aimed at studying tumor composition and evolutionary history."/>

  <link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>

  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2026-01-07"/>
  <meta property="article:created" itemprop="dateCreated" content="2026-01-07"/>
  <meta name="article:author" content="Aitor Sánchez-Ferrera"/>
  <meta name="article:author" content="Maitena Tellaetxe-Abete"/>
  <meta name="article:author" content="Borja Calvo-Molinos"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="GeRnika: An R Package for the Simulation, Visualization and Comparison of Tumor Phylogenies"/>
  <meta property="og:type" content="article"/>
  <meta property="og:description" content="The development of methods to study intratumoral heterogeneity and&#10;tumor phylogenies is a highly active area of research. However, the&#10;advancement of these approaches often necessitates access to&#10;substantial amounts of data, which can be challenging and expensive to&#10;acquire. Moreover, the assessment of results requires tools for&#10;visualizing and comparing tumor phylogenies. In this paper, we&#10;introduce GeRnika, an R package designed to address these needs by&#10;enabling the simulation, visualization, and comparison of tumor&#10;evolution data. In summary, GeRnika provides researchers with a&#10;user-friendly tool that facilitates the analysis of their approaches&#10;aimed at studying tumor composition and evolutionary history."/>
  <meta property="og:locale" content="en_US"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="GeRnika: An R Package for the Simulation, Visualization and Comparison of Tumor Phylogenies"/>
  <meta property="twitter:description" content="The development of methods to study intratumoral heterogeneity and&#10;tumor phylogenies is a highly active area of research. However, the&#10;advancement of these approaches often necessitates access to&#10;substantial amounts of data, which can be challenging and expensive to&#10;acquire. Moreover, the assessment of results requires tools for&#10;visualizing and comparing tumor phylogenies. In this paper, we&#10;introduce GeRnika, an R package designed to address these needs by&#10;enabling the simulation, visualization, and comparison of tumor&#10;evolution data. In summary, GeRnika provides researchers with a&#10;user-friendly tool that facilitates the analysis of their approaches&#10;aimed at studying tumor composition and evolutionary history."/>

  <!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
  <meta name="citation_title" content="GeRnika: An R Package for the Simulation, Visualization and Comparison of Tumor Phylogenies"/>
  <meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2025-042"/>
  <meta name="citation_pdf_url" content="RJ-2025-042.pdf"/>
  <meta name="citation_volume" content="17"/>
  <meta name="citation_issue" content="4"/>
  <meta name="citation_doi" content="10.32614/RJ-2025-042"/>
  <meta name="citation_journal_title" content="The R Journal"/>
  <meta name="citation_issn" content="2073-4859"/>
  <meta name="citation_firstpage" content="1"/>
  <meta name="citation_fulltext_world_readable" content=""/>
  <meta name="citation_online_date" content="2026/01/07"/>
  <meta name="citation_publication_date" content="2026/01/07"/>
  <meta name="citation_author" content="Aitor Sánchez-Ferrera"/>
  <meta name="citation_author_institution" content="Intelligent Systems Group, Computer Science Faculty, University of the&#10;Basque Country"/>
  <meta name="citation_author" content="Maitena Tellaetxe-Abete"/>
  <meta name="citation_author_institution" content="Intelligent Systems Group, Computer Science Faculty, University of the&#10;Basque Country"/>
  <meta name="citation_author" content="Borja Calvo-Molinos"/>
  <meta name="citation_author_institution" content="Intelligent Systems Group, Computer Science Faculty, University of the&#10;Basque Country"/>
  <!--/radix_placeholder_meta_tags-->
  
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","date_received","journal","volume","issue","slug","citation_url","packages","preview","bibliography","CTV","legacy_pdf","legacy_converted","output","draft","pdf_url","doi","creative_commons","csl"]}},"value":[{"type":"character","attributes":{},"value":["GeRnika: An R Package for the Simulation, Visualization and Comparison of Tumor Phylogenies"]},{"type":"character","attributes":{},"value":["The development of methods to study intratumoral heterogeneity and\ntumor phylogenies is a highly active area of research. However, the\nadvancement of these approaches often necessitates access to\nsubstantial amounts of data, which can be challenging and expensive to\nacquire. Moreover, the assessment of results requires tools for\nvisualizing and comparing tumor phylogenies. In this paper, we\nintroduce GeRnika, an R package designed to address these needs by\nenabling the simulation, visualization, and comparison of tumor\nevolution data. In summary, GeRnika provides researchers with a\nuser-friendly tool that facilitates the analysis of their approaches\naimed at studying tumor composition and evolutionary history.\n"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","orcid_id","address"]}},"value":[{"type":"character","attributes":{},"value":["Aitor Sánchez-Ferrera"]},{"type":"character","attributes":{},"value":["Intelligent Systems Group, Computer Science Faculty, University of the\nBasque Country"]},{"type":"character","attributes":{},"value":["0000-0001-6127-0686"]},{"type":"character","attributes":{},"value":["Paseo Manuel Lardizabal, Donostia/San Sebastian, 20018","Spain","[aitor.sanchezf@ehu.eus](aitor.sanchezf@ehu.eus){.uri}\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","orcid_id","address"]}},"value":[{"type":"character","attributes":{},"value":["Maitena Tellaetxe-Abete"]},{"type":"character","attributes":{},"value":["Intelligent Systems Group, Computer Science Faculty, University of the\nBasque Country"]},{"type":"character","attributes":{},"value":["0000-0003-1894-4547"]},{"type":"character","attributes":{},"value":["Paseo Manuel Lardizabal, Donostia/San Sebastian, 20018","Spain","[maitena.tellaetxe@ehu.eus](maitena.tellaetxe@ehu.eus){.uri}\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","orcid_id","address"]}},"value":[{"type":"character","attributes":{},"value":["Borja Calvo-Molinos"]},{"type":"character","attributes":{},"value":["Intelligent Systems Group, Computer Science Faculty, University of the\nBasque Country"]},{"type":"character","attributes":{},"value":["0000-0001-9969-9664"]},{"type":"character","attributes":{},"value":["Paseo Manuel Lardizabal, Donostia/San Sebastian, 20018","Spain","[borja.calvo@ehu.eus](borja.calvo@ehu.eus){.uri}\n"]}]}]},{"type":"character","attributes":{},"value":["2026-01-07"]},{"type":"character","attributes":{},"value":["2025-03-31"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"double","attributes":{},"value":[1]},{"type":"NULL"}]},{"type":"integer","attributes":{},"value":[17]},{"type":"integer","attributes":{},"value":[4]},{"type":"character","attributes":{},"value":["RJ-2025-042"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2025-042"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"character","attributes":{},"value":["GeRnika","data.tree"]},{"type":"character","attributes":{},"value":["clevRvis"]}]},{"type":"character","attributes":{},"value":["preview.png"]},{"type":"character","attributes":{},"value":["GeRnika.bib"]},{"type":"list","attributes":{},"value":[]},{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[true]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc","mathjax","md_extension"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js"]},{"type":"character","attributes":{},"value":["-tex_math_single_backslash"]}]}]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["RJ-2025-042.pdf"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2025-042"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"character","attributes":{},"value":["/home/hyndman/R/x86_64-pc-linux-gnu-library/4.5/rjtools/rjournal.csl"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["figs/commonsubtrees.png","figs/comparisonphylotrees.png","figs/consensus.png","figs/noisy_F_error.pdf","figs/noisy_F_error.png","figs/phylotree.png","figs/proportions.png","figs/ternary_plots_cropped.pdf","figs/ternary_plots_cropped.png","GeRnika.bib","GeRnika.tex","RJ-2025-042_files/anchor-4.2.2/anchor.min.js","RJ-2025-042_files/bowser-1.9.3/bowser.min.js","RJ-2025-042_files/distill-2.2.21/template.v2.js","RJ-2025-042_files/header-attrs-2.30/header-attrs.js","RJ-2025-042_files/jquery-3.6.0/jquery-3.6.0.js","RJ-2025-042_files/jquery-3.6.0/jquery-3.6.0.min.js","RJ-2025-042_files/jquery-3.6.0/jquery-3.6.0.min.map","RJ-2025-042_files/popper-2.6.0/popper.min.js","RJ-2025-042_files/tippy-6.2.7/tippy-bundle.umd.min.js","RJ-2025-042_files/tippy-6.2.7/tippy-light-border.css","RJ-2025-042_files/tippy-6.2.7/tippy.css","RJ-2025-042_files/tippy-6.2.7/tippy.umd.min.js","RJ-2025-042_files/webcomponents-2.0.0/webcomponents.js","RJ-2025-042.zip","RJournal.sty","RJwrapper.aux","RJwrapper.bbl","RJwrapper.blg","RJwrapper.brf","RJwrapper.fdb_latexmk","RJwrapper.fls","RJwrapper.log","RJwrapper.out","RJwrapper.synctex.gz","RJwrapper.tex"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
    font-size: 100%;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  hr.section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    margin: 0px;
  }


  d-byline {
    border-top: none;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
    border-top: none;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  /* tweak for Pandoc numbered line within distill */
  d-article pre.numberSource code > span {
      left: -2em;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // separator
    var separator = '<hr class="section-separator" style="clear: both"/>';
    // prepend separator above appendix
    $('.d-byline').before(separator);
    $('.d-article').before(separator);

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme, except when numbering line
    // in code chunk
    $('pre:not(.numberLines) code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        var author_name = front_matter.authors[i].author
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        const citeChild = $(this).children()[0]
        // Do not process if @xyz has been used without escaping and without bibliography activated
        // https://github.com/rstudio/distill/issues/466
        if (citeChild === undefined) return true

        if (citeChild.nodeName == "D-FOOTNOTE") {
          var fn = citeChild
          $(this).html(fn.shadowRoot.querySelector("sup"))
          $(this).id = fn.id
          fn.remove()
        }
        var refs = $(this).attr('data-cites').split(" ");
        var refHtml = refs.map(function(ref) {
          // Could use CSS.escape too here, we insure backward compatibility in navigator
          return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
        }).join("\n");
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // fix footnotes in tables (#411)
      // replacing broken distill.pub feature
      $('table d-footnote').each(function() {
        // we replace internal showAtNode methode which is triggered when hovering a footnote
        this.hoverBox.showAtNode = function(node) {
          // ported from https://github.com/distillpub/template/pull/105/files
          calcOffset = function(elem) {
              let x = elem.offsetLeft;
              let y = elem.offsetTop;
              // Traverse upwards until an `absolute` element is found or `elem`
              // becomes null.
              while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                  x += elem.offsetLeft;
                  y += elem.offsetTop;
              }

              return { left: x, top: y };
          }
          // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
          const bbox = node.getBoundingClientRect();
          const offset = calcOffset(node);
          this.show([offset.left + bbox.width, offset.top + bbox.height]);
        }
      })

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      // ignore leaflet img layers (#106)
      figures = figures.filter(':not(img[class*="leaflet"])')
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="RJ-2025-042_files/header-attrs-2.30/header-attrs.js"></script>
  <script src="RJ-2025-042_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="RJ-2025-042_files/popper-2.6.0/popper.min.js"></script>
  <link href="RJ-2025-042_files/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="RJ-2025-042_files/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="RJ-2025-042_files/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="RJ-2025-042_files/anchor-4.2.2/anchor.min.js"></script>
  <script src="RJ-2025-042_files/bowser-1.9.3/bowser.min.js"></script>
  <script src="RJ-2025-042_files/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="RJ-2025-042_files/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Mathjax config (add automatic linebreaks when supported)
      // MathJax = {
      //    tex: {
      //        inlineMath: [['$', '$'], ['\\(', '\\)']],
      //        displayMath: [['$$', '$$'], ['\\[', '\\]']],
      //        tags: 'ams',
      //        multline: true,
      //    },
      //    options: {
      //        linebreaks: { automatic: true },
      //    },
      // };

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      
      // Conditionally add supplementary files button
      if (document.getElementById('supplementary-materials')) {
        // create element safely
        const btn_suppl = document.createElement('a');
        btn_suppl.href = slug + '.zip';
        btn_suppl.className = 'dt-tag';
        btn_suppl.innerHTML = '<i class="fas fa-file-zipper"></i> Supplement';
        buttons.append(btn_suppl);
      }

      // adds Abstract: in front of the first <p> in the title section --
      // unless it happens to be the subtitle (FIXME: this is a bad hack - can't distill do this?)
      var tpar = $("d-title p:not(:empty)").filter(function() {
        return !$(this).hasClass("subtitle");
      }).first();
      if (tpar) {
        const abstract = $('<d-abstract>')
        abstract.append('<b>Abstract:</b><br>')
        abstract.append(tpar) // Move description to d-abstract
        $("d-title p:empty").remove() // Remove empty paragraphs after title
        abstract.append(buttons)
        abstract.insertAfter($('d-title')) // Add abstract section after title */
      }

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
    
    d-appendix .citation-appendix,
    .d-appendix .citation-appendix {
      color: rgb(60, 60, 60);
    }

    d-article h2 {
      border-bottom: 0px solid rgba(0, 0, 0, 0.1);
      padding-bottom: 0rem;
    }
    d-article h3 {
      font-size: 20px;
    }
    d-article h4 {
      font-size: 18px;
      text-transform: none;
    }

    @media (min-width: 1024px) {
      d-article h2 {
        font-size: 32px;
      }
      d-article h3 {
        font-size: 24px;
      }
      d-article h4 {
        font-size: 20px;
      }
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"GeRnika: An R Package for the Simulation, Visualization and Comparison of Tumor Phylogenies","description":"The development of methods to study intratumoral heterogeneity and\ntumor phylogenies is a highly active area of research. However, the\nadvancement of these approaches often necessitates access to\nsubstantial amounts of data, which can be challenging and expensive to\nacquire. Moreover, the assessment of results requires tools for\nvisualizing and comparing tumor phylogenies. In this paper, we\nintroduce GeRnika, an R package designed to address these needs by\nenabling the simulation, visualization, and comparison of tumor\nevolution data. In summary, GeRnika provides researchers with a\nuser-friendly tool that facilitates the analysis of their approaches\naimed at studying tumor composition and evolutionary history.","doi":"10.32614/RJ-2025-042","authors":[{"author":"Aitor Sánchez-Ferrera","authorURL":"#","affiliation":"Intelligent Systems Group, Computer Science Faculty, University of the\nBasque Country","affiliationURL":"#","orcidID":"0000-0001-6127-0686"},{"author":"Maitena Tellaetxe-Abete","authorURL":"#","affiliation":"Intelligent Systems Group, Computer Science Faculty, University of the\nBasque Country","affiliationURL":"#","orcidID":"0000-0003-1894-4547"},{"author":"Borja Calvo-Molinos","authorURL":"#","affiliation":"Intelligent Systems Group, Computer Science Faculty, University of the\nBasque Country","affiliationURL":"#","orcidID":"0000-0001-9969-9664"}],"publishedDate":"2026-01-07T00:00:00.000+11:00","citationText":"Sánchez-Ferrera, et al., 2026"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>GeRnika: An R Package for the Simulation, Visualization and Comparison of Tumor Phylogenies</h1>

<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>The development of methods to study intratumoral heterogeneity and
tumor phylogenies is a highly active area of research. However, the
advancement of these approaches often necessitates access to
substantial amounts of data, which can be challenging and expensive to
acquire. Moreover, the assessment of results requires tools for
visualizing and comparing tumor phylogenies. In this paper, we
introduce GeRnika, an R package designed to address these needs by
enabling the simulation, visualization, and comparison of tumor
evolution data. In summary, GeRnika provides researchers with a
user-friendly tool that facilitates the analysis of their approaches
aimed at studying tumor composition and evolutionary history.</p></p>
</div>

<div class="d-byline">
  Aitor Sánchez-Ferrera  (Intelligent Systems Group, Computer Science Faculty, University of the
Basque Country)
  
,   Maitena Tellaetxe-Abete  (Intelligent Systems Group, Computer Science Faculty, University of the
Basque Country)
  
,   Borja Calvo-Molinos  (Intelligent Systems Group, Computer Science Faculty, University of the
Basque Country)
  
<br/>2026-01-07
</div>

<div class="d-article">
<div class="article">
<p><span class="math display">\[1\]</span>These authors contributed equally to this work.</p>
<h3 data-number="1" id="introduction"><span class="header-section-number">1</span> Introduction</h3>
<p>The development of tumors is a complex and dynamic process characterized
by a succession of events where DNA mutations accumulate over time.
These mutations give rise to genetic diversity within the tumor, leading
to the emergence of distinct clonal subpopulations or, simply, clones
(Nowell 1976). Each of these subpopulations exhibits unique mutational
profiles, resulting in varied phenotypic and behavioral characteristics
among the cancer cells (Marass et al. 2016). This phenomenon, known as
intratumoral heterogeneity (ITH), significantly hinders the design of
effective medical therapies, since different clones within the same
tumor may respond differently to treatments, ultimately leading to
therapy resistance and disease recurrence (Burrell et al. 2013). To
address this challenge, several innovative approaches are being
developed to study the tumor composition in greater detail and to
reconstruct its evolutionary history.</p>
<p>One common approach to studying tumor composition and phylogeny involves
using bulk DNA sequencing data from multiple tumor biopsies. This data
is relatively straightforward to obtain and provides a broad overview of
the genetic alterations within the tumor. However, using bulk sequencing
data in the study of ITH faces the challenge that each sample
potentially contains a mixture of different clonal populations rather
than just one clone. Consequently, the observed mutation
frequencies—measured as variant allele frequencies (VAFs)—do not
directly estimate the fraction of individual clones. Instead, the VAF
values represent a composite signal: the sum of the fractions of all
clones that harbor each mutation in a given sample.</p>
<p>This complexity implies that reconstructing the tumor’s evolutionary
history requires deconvolving these clonal admixtures within the
samples. This task is precisely the focus of the Clonal Deconvolution
and Evolution Problem (CDEP) (Strino et al. 2013; El-Kebir et al. 2015),
which can be summarized as determining the tumor’s clonal
structure—that is, identifying the number, proportion, and mutational
composition of clones in each sample—as well as reconstructing the
clonal phylogenetic tree that leads to the observed clonal mosaic. In
this context, one of the most prominent approaches to addressing the
CDEP is the Variant Allele Frequency Factorization Problem (VAFFP)
(El-Kebir et al. 2015).</p>
<p>Given <span class="math inline">\(s\)</span> tumor samples and <span class="math inline">\(n\)</span> mutations identified across these
samples, we define a matrix <span class="math inline">\(\boldsymbol{F} \in [0, 1]^{s \times n}\)</span>,
where each element <span class="math inline">\(f_{ij}\)</span> represents the variant VAF value, or
equivalently, the fraction of cells that carry mutation <span class="math inline">\(j\)</span> in sample
<span class="math inline">\(i\)</span>. The VAFFP seeks to decompose this input matrix <span class="math inline">\(\boldsymbol{F}\)</span>
into two matrices: a matrix <span class="math inline">\(\boldsymbol{B} \in \{0, 1\}^{n \times n}\)</span>
that represents the clonal phylogeny, and a matrix
<span class="math inline">\(\boldsymbol{U} \in [0, 1]^{s \times n}\)</span> that captures the clone
proportions in each tumor sample:</p>
<p><span class="math display" id="eq:eqF">\[\begin{equation}
\boldsymbol{F} = \boldsymbol{U} \cdot \boldsymbol{B}
\label{eqF}
\end{equation}   \tag{1}\]</span></p>
<p>The <span class="math inline">\(\boldsymbol{B}\)</span> matrix is a binary square matrix of size <span class="math inline">\(n\)</span>, where
<span class="math inline">\(b_{ij} = 1\)</span> iff clone <span class="math inline">\(i\)</span> contains the mutation <span class="math inline">\(j\)</span> (Gusfield 1991).
The matrix <span class="math inline">\(\boldsymbol{U}\)</span> is an <span class="math inline">\(s \times n\)</span> matrix where <span class="math inline">\(u_{ij}\)</span> is
the fraction of clone <span class="math inline">\(j\)</span> in sample <span class="math inline">\(i\)</span>.</p>
<p>The VAFFP operates under two key assumptions: tumors have a monoclonal
origin, meaning they arise from a single abnormal cell, and the infinite
sites assumption (ISA), which states that mutations occur at most once
and cannot disappear over time (Kimura 1969). Under these assumptions,
the tumor’s clonal structure can be modeled as a perfect phylogeny
(Gusfield 1991). This model imposes two key constraints: (1) if two
clones share a mutation, they must either be identical or ancestrally
related, and (2) once a clone acquires a mutation, that mutation is
inherited by all its descendants.</p>
<p>Over the years, numerous methods have been developed to solve the CDEP
(Sengupta et al. 2014; Yuan et al. 2015; Deshwar et al. 2015; El-Kebir
et al. 2015, 2016; Malikic et al. 2015; Popic et al. 2015; Jiang et al.
2016; Husić et al. 2019; Fu et al. 2022; <span class="nocase">Grigoriadis et al.</span>
2024), with recent advancements addressing reformulations of the problem
that incorporate single-cell sequencing-derived information or variants
in metastases, and account for temporal resolution (Ross and Markowetz
2016; El-Kebir et al. 2018; Malikic et al. 2019; Satas et al. 2020;
Sollier et al. 2023). These methods primarily focus on reconstructing
tumor phylogenies and clonal compositions using both real and simulated
data. However, the simulation tools they employ often have limitations.
For instance, while MiPUP uses simulated data, it allows only basic
parameter adjustments—such as the number of mutations, samples, and
reads—and lacks the flexibility to fine-tune biological parameters.
BitPhylogeny creates highly realistic and complex simulated datasets
representing different modes of evolution, but these simulations are
manually crafted and limited in number, posing scalability issues.</p>
<p>Several tools specifically designed for simulating data for the CDEP
have also been introduced. Pearsim, written in Python, allows control
over parameters like read depth, number of subclones, samples, and
mutations (Kulman et al. 2022). OncoLib, a C++ library, facilitates the
simulation of tumor heterogeneity and the reconstruction of NGS
sequencing data of metastatic tumors, offering control over parameters
such as driver mutation probability, per-base sequencing error rate,
migration rate, and mutation rate (<strong>oncolib?</strong>). Machina, also in C++,
provides a framework for simulating metastatic tumors and visualizing
their phylogenetic trees and migration graphs (El-Kebir et al. 2018).
HeteroGenesis, implemented in Python, simulates heterogeneous tumors at
the level of clone genomes, but it is not specifically tailored for CDEP
instances and requires additional processing to generate suitable
datasets (Tanner et al. 2019).</p>
<p>Visualization of tumor phylogenies is another crucial aspect of studying
tumor evolution. Tools like CALDER, Clonevol, SPRUCE, <strong>fishplot</strong>,
ClonArch, and
<a href="https://www.bioconductor.org/packages/release/bioc/html/clevRvis.html"><strong>clevRvis</strong></a>
offer solutions for visualizing clonal structures and evolutionary
trajectories (Myers et al. 2019; Dang et al. 2017; El-Kebir et al. 2016;
Miller et al. 2016; Wu and El-Kebir 2020; Sandmann et al. 2023). Among
these, <strong>fishplot</strong> and <strong>clevRvis</strong> are available as R packages.
Additionally, methods for generating consensus trees, such as TuELIP and
ConTreeDP, have been developed to summarize multiple phylogenetic trees
into a single representative tree (Guang et al. 2023; Fu and Schwartz
2021).</p>
<p>Despite the availability of existing tools, there remains a lack of
options in the R programming environment for realistically simulating
tumor evolution in a way that is both flexible and user-friendly, while
also enabling effective visualization and comparison.</p>
<p>In this paper, we introduce
<a href="https://CRAN.R-project.org/package=GeRnika"><strong>GeRnika</strong></a>, an R package
that provides a comprehensive solution for simulating, visualizing, and
comparing tumor evolution data. Although <strong>GeRnika</strong>’s data simulation
functionality was primarily devised to create instances for solving the
CDEP, the simulated data are not restricted to this purpose and can also
be used for exploring evolutionary dynamics in broader contexts. To
accommodate diverse research needs, we have implemented the procedures
to be highly customizable, allowing users to adjust a wide range of
parameters such as the number of clones, selective pressures, mutation
rates, and sequencing noise levels. Unlike existing tools that may offer
limited customization or are implemented in other programming languages,
<strong>GeRnika</strong> is fully integrated into the R environment, making it easy
to use alongside other bioinformatics packages. By combining simulation
capabilities with visualization and comparison tools in a user-friendly
interface, <strong>GeRnika</strong> offers an accessible and flexible option within
the R ecosystem for researchers studying tumor evolution. It is
important to note that <strong>GeRnika</strong> does not implement algorithms for
inferring clonal composition or reconstructing phylogenies from
experimental datasets. Instead, it provides a controlled framework for
generating, visualizing, and comparing data that can be used to
benchmark such methods.</p>
<h3 data-number="2" id="sec:simulation"><span class="header-section-number">2</span> Simulation of tumor evolution</h3>
<p>The main contribution of this work is the introduction of a novel
approach for simulating biologically plausible instances of the VAFFP
that accounts for several key factors, including the number of clones,
selective pressures, and sequencing noise. In this section, we provide a
detailed description of the approach.</p>
<p>Broadly speaking, each problem instance consists of a matrix
<span class="math inline">\(\boldsymbol{F}\)</span> containing the VAF values of a set of mutations in a
set of samples, as described previously. This matrix is built from a
pair of matrices <span class="math inline">\(\boldsymbol{B}\)</span> and <span class="math inline">\(\boldsymbol{U}\)</span> that represent a
tumor phylogeny fulfilling the ISA and the proportions of the clones in
the samples, respectively, following Equation <a href="#eq:eqF">(1)</a>.</p>
<p>In order to simulate the <span class="math inline">\(\boldsymbol{B}\)</span> and <span class="math inline">\(\boldsymbol{U}\)</span> matrices,
we have devised two models: a tumor model that simulates the
evolutionary history and current state of the tumor, and a sampling
model that represents the tumor sampling process. A third model, namely
the sequencing noise model, has been devised to optionally introduce
sequencing noise to the VAF values in the <span class="math inline">\(\boldsymbol{F}\)</span> matrix, if
noisy data is desired. The following subsections describe these models
in detail.</p>
<h4 class="unnumbered" data-number="2.1" id="tumor-model">Tumor model</h4>
<p>The tumor model generates a clonal tree <span class="math inline">\(T\)</span> and an associated matrix
<span class="math inline">\(\boldsymbol{B}\)</span>, together with the clone proportions <span class="math inline">\(\boldsymbol{c}\)</span>
and tumor blend at the moment of sampling. Briefly, for a tumor with a
set of <span class="math inline">\(n\)</span> mutations denoted by <span class="math inline">\(M\)</span>, <span class="math inline">\(T\)</span> is a rooted tree on an
<span class="math inline">\(n\)</span>-sized vertex set <span class="math inline">\(V_n = \{v_{1}, \dots, v_{n} \}\)</span>, where <span class="math inline">\(v_{i}\)</span>
represents clone <span class="math inline">\(i\)</span> and simultaneously corresponds to the first clone
containing mutation <span class="math inline">\(M_i\)</span>. This one-to-one correspondence between clones
and mutations allows us to refer to them interchangeably. The tree is
further defined by an <span class="math inline">\((n-1)\)</span>-sized edge set <span class="math inline">\(E_T\)</span>, where each edge
<span class="math inline">\(e_{ij} \in E_T\)</span> represents a direct ancestral relationship from vertex
<span class="math inline">\(v_{i}\)</span> to vertex <span class="math inline">\(v_{j}\)</span>.</p>
<p>In our tumor model, <span class="math inline">\(T\)</span> is iteratively generated with a random topology,
as follows. First, the root node of <span class="math inline">\(T\)</span>, <span class="math inline">\(\mathcal R(T)\)</span>, is set, and a
random mutation <span class="math inline">\(M_i \in M\)</span> is assigned to it. For each of the remaining
<span class="math inline">\(M_j \in M - \{M_i\}\)</span> mutations, a new node <span class="math inline">\(v_j\)</span> is created and the
mutation <span class="math inline">\(M_j\)</span> is assigned to this node. The node <span class="math inline">\(v_j\)</span> is then attached
as a child to one of the nodes already included in <span class="math inline">\(T\)</span>. To adhere to the
ISA model, each newly added node inherits all the mutations present in
its parent node.</p>
<p>The attachment of nodes to the tree is not uniformly random. Instead,
the nodes in the growing tree <span class="math inline">\(T\)</span> have different probabilities of being
selected as parents for the new nodes, depending on the number of
ascendants, <span class="math inline">\(\mathcal A(v_i)\)</span>, they have. Specifically,
<span class="math inline">\(\forall v_j \neq \mathcal R(T)\)</span>, the parent node of <span class="math inline">\(v_j\)</span> is sampled
from a multinomial distribution where the probabilities are calculated
as:</p>
<p><span class="math display">\[\begin{equation}
\boldsymbol{p}(v_i; k) = \frac{k^{\frac{|\mathcal{A}(v_i)| + 1}{\delta}}}{\sum_{v_l \in V&#39;} k^{\frac{|\mathcal{A}(v_l)| + 1}{\delta}}}; \quad v_i \in V&#39;
\end{equation}\]</span></p>
<p>Here, <span class="math inline">\(\delta\)</span> represents the depth of the growing tree, i.e., the
number of levels or layers in the tree structure. <span class="math inline">\(k \in (0, +\infty)\)</span>
is the topology parameter that determines whether the topology tends to
be branched, with a decreasing probability for increasing numbers of
ascendants (<span class="math inline">\(k\)</span> <span class="math inline">\(&lt;\)</span> 1), or linear, with an increasing probability for
increasing numbers of ascendants (<span class="math inline">\(k\)</span> <span class="math inline">\(&gt;\)</span> 1).</p>
<p>Once <span class="math inline">\(T\)</span> has been generated, it is represented in the form of a
<span class="math inline">\(\boldsymbol{B}\)</span> matrix, constructed by initializing an identity matrix
<span class="math inline">\(B_{n}\)</span> and setting <span class="math inline">\(b_{ji}\)</span> to 1 for each pair of nodes <span class="math inline">\(v_i\)</span> and <span class="math inline">\(v_j\)</span>
where node <span class="math inline">\(v_j\)</span> is a descendant of node <span class="math inline">\(v_i\)</span> in <span class="math inline">\(T\)</span>.</p>
<p>After obtaining <span class="math inline">\(\boldsymbol{B}\)</span>, the proportions of the clones in the
whole tumor, denoted as <span class="math inline">\(\boldsymbol{c} = \{c_{1}, \dots, c_{n} \}\)</span>, are
simulated. It is important to note that these proportions are not the
same as those appearing in the <span class="math inline">\(\boldsymbol{U}\)</span> matrix, which represent
the <em>sampled</em> clone proportions and depend not only on the global clone
proportions but also on the spatial distribution of the clones and the
sampling sites.</p>
<p>These clone proportions <span class="math inline">\(\boldsymbol{c}\)</span> are calculated by sequentially
sampling a Dirichlet distribution at each multifurcation in <span class="math inline">\(T\)</span>,
starting from the root. For instance, for a node <span class="math inline">\(v_i\)</span> with children
<span class="math inline">\(\mathcal K(v_i)\)</span> = {<span class="math inline">\(v_j\)</span>, <span class="math inline">\(v_k\)</span>}, we draw a sample <span class="math inline">\((x_i, x_j, x_k)\)</span>
that represents the proportions of the parent clone and its two
children, respectively, from a Dirichlet distribution
<span class="math inline">\(Dir(\alpha_i, \alpha_j, \alpha_k)\)</span>. When this sampling is performed at
a node <span class="math inline">\(v_i \neq \mathcal R(T)\)</span>, these proportions are scaled relative
to the original proportion of the parent clone. This ensures that the
sum rule is met, and that once all multifurcations have been visited,
the proportions of all clones in <span class="math inline">\(T\)</span> sum up to one. While several
approaches can exist to determine these proportions, this method
provides a natural approximation to the problem that can be interpreted
as the distribution of the mass or proportion of each clone between
itself and its descendants.</p>
<p>The parameters of the Dirichlet distribution depend on the tumor’s
evolution model. In this work, we consider two fundamental cases:
positive selection-driven evolution and neutral evolution. In positive
selection-driven evolution, certain mutations confer a growth advantage,
while most mutations do not. As a result, the clones carrying these
advantageous mutations outcompete other clones and dominate the tumor.
Consequently, tumors are predominantly composed of a few dominant
clones, with the remaining clones present in too small proportions.
Under neutral evolution, instead, there is no significant number of
mutations that provide a fitness advantage, and clones accumulate solely
due to tumor progression. As a result, all clones are present in similar
proportions (Davis et al. 2017).</p>
<p>Based on this, all the parameters for the Dirichlet distribution for
positive selection-driven evolution are set to 0.3. For neutral
evolution, the parameter corresponding to the parent node (<span class="math inline">\(\alpha_{p}\)</span>)
is set to 5, and the parameters corresponding to the children node(s)
(<span class="math inline">\(\alpha_{c}\)</span>) are set to 10. Different alpha values are used for parent
and children nodes in neutral evolution to ensure that clones arising
late in the evolution do not end up with proportions that are too small
solely due to their position in the topology, preventing the deviation
from the expected clone proportion distribution for this type of
evolution model. These values have been chosen empirically, and their
effect is illustrated in Figure <a href="#ternary_plots" data-reference-type="ref" data-reference="ternary_plots">1</a>, which shows how 5,000 random samples from
the mentioned Dirichlet distributions (for the particular case of 3
dimensions, i.e., one parent and two children nodes) are distributed. As
observed, in the case of positive selection
(<span class="math inline">\(\alpha_{i} = \alpha_{j} = \alpha_{k} = 0.3\)</span>), the <span class="math inline">\((x_i, x_j, x_k)\)</span>
values are equally pushed towards the three corners of the simplex. In
other words, the samples tend to be sparse, with typically one component
having a large value and the rest close to 0. Instead, when neutral
selection is adopted (<span class="math inline">\(\boldsymbol{\alpha}\)</span> = (5, 10, 10)), the
<span class="math inline">\((x_i, x_j, x_k)\)</span> values concentrate close to the center of the simplex,
but with a tendency to deviate towards those components with larger
<span class="math inline">\(\alpha\)</span> value. This means that samples <span class="math inline">\((x_i, x_j, x_k)\)</span> are less
sparse in neutral evolution, with larger values for <span class="math inline">\(x_2\)</span> and <span class="math inline">\(x_3\)</span> in
this case, which represent the children nodes.</p>
<figure id="ternary_plots" data-latex-placement="t">
<img src="figs/ternary_plots_cropped.png" style="width:100.0%"
alt="graphic without alt text" />
<figcaption>
Figure 1: Ternary density plots of 5,000 samples drawn from
two 3-dimensional Dirichlet distributions. The parameters of the
Dirichlet distribution on the left are <span class="math inline"><em>α</em></span> = (0.3, 0.3, 0.3) and the
distribution is used to represent positive selection-driven evolution.
The distribution on the right has parameters <span class="math inline"><em>α</em></span> = (5, 10, 10) and is used to
represent neutral evolution. Samples drawn from these distributions (or
their generalization to higher spaces) are used to calculate clone
proportions in each tree multifurcation.
</figcaption>
</figure>
<p>Taking into account that marginalizing the Dirichlet distribution
results in a Beta distribution, the proportion of the clone
<span class="math inline">\(v_i \in V_n \; | \; v_i \neq \mathcal R(T)\)</span> in the tumor, denoted as
<span class="math inline">\(C_i\)</span>, follows the distribution:</p>
<p><span class="math display">\[\begin{equation}
    C_i \sim C_{\mathcal P(v_i)} \cdot \Gamma_i \cdot \Gamma_i^\prime
\end{equation}\]</span></p>
<p>where</p>
<p><span class="math display">\[\begin{equation}
\Gamma_i \sim Beta(\alpha_{c}, \alpha_{p} + \alpha_{c} \cdot (|\mathcal K(\mathcal P(v_i))| - 1))
\end{equation}\]</span></p>
<p>and</p>
<p><span class="math display">\[\begin{align}
\Gamma_i^\prime = 1 \quad \text{if } |\mathcal K(v_i)| = 0 \\
\Gamma_i^\prime \sim Beta(\alpha_{p}, \alpha_{c} \cdot |\mathcal K(v_i)|)  \quad \text{if } |\mathcal K(v_i)| \neq 0
\end{align}\]</span></p>
<p>For the case where <span class="math inline">\(v_i = \mathcal R(T)\)</span>, the root node, <span class="math inline">\(C_i\)</span>, follows:</p>
<p><span class="math display">\[\begin{equation}
C_i \sim Beta(\alpha_{p}, \alpha_{c} \cdot  |\mathcal K(v_i)|)
\end{equation}\]</span></p>
<p>Here, <span class="math inline">\(\alpha_{p}\)</span> and <span class="math inline">\(\alpha_{c}\)</span> are the parameters of the Dirichlet
distribution assigned to parent and child nodes, respectively.</p>
<p>To complete the tumor model, the tumor blend is simulated, which
represents the degree of physical mixing between the tumor clones. In
order to do this, we simplify the spatial distribution to one dimension
and model the tumor as a Gaussian mixture model with <span class="math inline">\(n\)</span> components,
where each component <span class="math inline">\(G_i\)</span> represents a tumor clone, and the mixture
weights are given by <span class="math inline">\(\boldsymbol{c}\)</span>. The variance for all components
is set to 1, while the mean values are random variables.</p>
<p>Specifically, we start by selecting a random clone, and its component’s
mean value is set to 0. Then, the mean values of the remaining <span class="math inline">\(n - 1\)</span>
components are calculated sequentially by adding <span class="math inline">\(d\)</span> units to the mean
value of the previous component. To introduce variability in the tumor
blend, the value of <span class="math inline">\(d\)</span> is chosen from the set <span class="math inline">\(\{0, 0.1, \ldots, 4\}\)</span>.
For <span class="math inline">\(d\)</span> = 0, the two clones are completely mixed, while for <span class="math inline">\(d\)</span> = 4,
they are physically far apart from each other. The choice of the upper
limit for <span class="math inline">\(d\)</span> has been determined empirically, considering that with
this value, the overlapping area between the two clones becomes
negligible.</p>
<p>To ensure the separation between the clones is random and that most of
the time the separation is small, we use an exponential-like
distribution with the form <span class="math inline">\(Beta(\alpha=1, \beta=X)\)</span> to sample the
values of <span class="math inline">\(d\)</span>. Specifically, we set <span class="math inline">\(\beta = 5\)</span> to ensure that the
samples obtained from the mixture are not excessively sparse. We can
express this mathematically as:</p>
<p><span class="math display">\[\begin{equation}
D \sim 4 \cdot Beta(\alpha=1, \beta=5)
\end{equation}\]</span></p>
<h4 class="unnumbered" data-number="2.2" id="sampling-simulation">Sampling simulation</h4>
<p>So far, we have described how the clones of a tumor are modelled by the
tumor model. However, in real practice, there is no easy way of
observing these global properties of a tumor. Instead, we typically have
access to information provided by samples or biopsies. This means that
certain tumor characteristics, such as the real clone proportions
<span class="math inline">\(\boldsymbol{c}\)</span>, cannot be directly obtained. Instead, we can only
determine the <em>sampled</em> clone proportions, which depend on the specific
sampling procedure employed. Unless there is a perfectly uniform mixture
of the clone cells, their sampled proportions will not match the global
proportions. These sampled clone proportions are, in fact, the
<span class="math inline">\(\boldsymbol{u}_{i}.\)</span> elements in the <span class="math inline">\(\boldsymbol{U}\)</span> matrix.</p>
<p>The sampling simulation we have devised simulates the physical sampling
of the tumor and allows us to construct the <span class="math inline">\(\boldsymbol{U}\)</span> matrix of
the problem. This procedure operates on the data simulated using the
tumor model. Specifically, it simulates a sampling procedure carried out
in a grid manner over the tumor Gaussian mixture model described in the
previous section. Let <span class="math inline">\(G_1\)</span> and <span class="math inline">\(G_n\)</span> be the components with the lowest
and largest mean values, respectively, in the Gaussian mixture model.
The 1<sup>st</sup> and <span class="math inline">\(m\)</span><sup>th</sup> sampling points in the grid are always set to
<span class="math inline">\(\mu_{G_1} - 2.8 \cdot \sigma_{G_1}\)</span> and
<span class="math inline">\(\mu_{G_n} + 2.8 \cdot \sigma_{G_n}\)</span>, respectively, and the remaining
<span class="math inline">\(m\)</span>-2 sampling points are determined by dividing the range between these
two endpoints into <span class="math inline">\(m-1\)</span> equal intervals.</p>
<p>The densities of the Gaussian distributions at each sampling point are
multiplied by the global proportion of the clones sampled from the
Dirichlet distributions, so that for each sampling point <span class="math inline">\(i\)</span>, the
fraction of clone <span class="math inline">\(j\)</span>, <span class="math inline">\(p_{ij}\)</span>, is proportional to their product:</p>
<p><span class="math display" id="eq:pij-2">\[\begin{equation}
p_{ij} \propto c_j \cdot \phi_{ij}
\label{eq:pij_2}
\end{equation}   \tag{2}\]</span></p>
<p>where <span class="math inline">\(c_j\)</span> is the global proportion of clone <span class="math inline">\(j\)</span> and <span class="math inline">\(\phi_{ij}\)</span> is the
density of the Gaussian component associated with clone <span class="math inline">\(j\)</span> at sampling
point <span class="math inline">\(i\)</span>.</p>
<p>Finally, to account for the effect of cell count in the samples, a
multinomial distribution is used to sample a given number of cells
<span class="math inline">\(n_{c}\)</span> for each tumor sample. In that distribution, the probability of
selecting each clone at sampling site <span class="math inline">\(i\)</span> is given by
<span class="math inline">\((p_{i1}, \ldots, p_{in})\)</span>. The resulting values determine the final
tumor clone composition in sample <span class="math inline">\(i\)</span>, which are represented in the
matrix <span class="math inline">\(\boldsymbol{U}\)</span>:</p>
<p><span class="math display">\[\begin{equation}
U_{i.} \sim \frac{M(n = n_{c}, p = (p_{i1}, \ldots, p_{in}))
}{100}
\end{equation}\]</span></p>
<p>Note that selecting a relatively low value for <span class="math inline">\(n\)</span> in the multinomial
distribution can lead to clones with very low frequencies being modeled
as absent in the sample, with composition values equal to 0. This is
indeed more realistic than truly observing them with such low
frequencies.</p>
<h4 class="unnumbered" data-number="2.3" id="sequencing-noise-simulation">Sequencing noise simulation</h4>
<p>Up to this point, the <span class="math inline">\(\boldsymbol{B}\)</span> and <span class="math inline">\(\boldsymbol{U}\)</span> matrices of
an instance have been simulated. In case we are simulating noise-free
data, the simulation is complete once Equation <a href="#eq:eqF">(1)</a> is applied
to obtain the <span class="math inline">\(\boldsymbol{F}\)</span> matrix.</p>
<p>As a brief reminder, each element <span class="math inline">\(f_{ij}\)</span> in <span class="math inline">\(\boldsymbol{F}\)</span> denotes
the frequency or VAF of the mutation <span class="math inline">\(M_j\)</span> in sample <span class="math inline">\(i\)</span> or, in other
words, the proportion of sequencing reads that carry the mutation <span class="math inline">\(M_j\)</span>
in that particular sample. This also means that the proportion of reads
in that sample that do not observe the mutation but instead contain the
reference nucleotide is 1 - <span class="math inline">\(f_{ij}\)</span>.</p>
<p>However, empirical factors can artificially alter the VAF value, leading
it to deviate from the true ratio between the variant and total allele
molecule counts. One of these factors is the noise introduced during the
DNA sequencing process itself, which can arise in two main ways. First,
limitations of the sequencing instrument can lead to incorrect
nucleotide readings of DNA fragments. For example, a position that
actually contains nucleotide A may be read as a T. Second, there can be
a biased number of reads produced for a particular site, which can
result from chemical reaction peculiarities or simply because not all
fragments are sequenced. These limitations can, however, be mitigated to
some extent. For instance, it has been shown that a high depth of
coverage, which refers to the average number of reads that cover each
position, can lead to more accurate VAF values (Petrackova et al. 2019).</p>
<p>In order to incorporate the effect of sequencing noise in the data
instances, we have developed a procedure to simulate sequencing noise.
This procedure introduces noise to the <span class="math inline">\(\boldsymbol{F}\)</span> matrix and
generates a noisy matrix <span class="math inline">\(\boldsymbol{F^{(n)}}\)</span>, where
<span class="math inline">\(\boldsymbol{F^{(n)}} \neq \boldsymbol{U} \cdot \boldsymbol{B}\)</span>. The
procedure simulates noise at the level of the sequencing reads and
recalculates the new <span class="math inline">\(f^{(n)}_{ij}\)</span> values, as follows.</p>
<p>The sequencing depth <span class="math inline">\(r\)</span> at the genomic position where <span class="math inline">\(M_j\)</span> occurs in
sample <span class="math inline">\(i\)</span> is distributed according to a negative binomial distribution:</p>
<p><span class="math display" id="eq:r">\[\begin{equation}
r_{ij} \sim NB(\mu = \mu_{sd}, \alpha = 5)
\label{eq:r}
\end{equation}   \tag{3}\]</span></p>
<p>where <span class="math inline">\(\mu_{sd}\)</span> represents the mean sequencing depth, which is the
average number of reads covering the genomic position of mutation <span class="math inline">\(M_j\)</span>
in the sample, and <span class="math inline">\(\alpha\)</span> is the dispersion parameter, which controls
the variability of the sequencing depth around the mean and is fixed at
5.</p>
<p>The number of reads supporting the alternate allele <span class="math inline">\(r^{a}_{ij}\)</span> is then
modeled by a binomial distribution:</p>
<p><span class="math display" id="eq:ra">\[\begin{equation}
r^{a}_{ij} \sim B(n = r_{ij}, p = f_{ij})
\label{eq:ra}
\end{equation}   \tag{4}\]</span></p>
<p>In sequencing data, errors can occur due to limitations inherent to the
sequencing methodology. These errors vary depending on the technology
used.</p>
<p>To simulate the effect of these errors on the VAF values, the number of
reads <span class="math inline">\(r^{a\prime}_{ij}\)</span> that, despite originally supporting the
alternate allele, contain a different allele as a result of a sequencing
error, is modeled using a binomial distribution:</p>
<p><span class="math display">\[\begin{equation}
r^{a\prime}_{ij} \sim B(n = r^{a}_{ij}, p = \varepsilon),
\end{equation}\]</span></p>
<p>where <span class="math inline">\(\varepsilon\)</span> represents the sequencing error rate.</p>
<p>We also need to consider the situation where the reads contain the
reference nucleotide but are read with the alternate allele as a result
of this error. This can be better understood with an example. Let’s
imagine that at a certain genomic position, the normal cells have a T,
but in some cells, there is a mutation where the T has changed to an A.
In this case, for the normal cells, with a rate of <span class="math inline">\(\varepsilon\)</span>, a
sequencing error may occur, resulting in a read of C, G, or A instead of
T, each with an equal chance. Therefore, in approximately
<span class="math inline">\(\frac{\varepsilon}{3}\)</span> of the cases, reads with the mutation of
interest will arise from normal reads:</p>
<p><span class="math display" id="eq:ramr">\[\begin{equation}
r^{r\prime}_{ij} \sim B(n = r_{ij} - r^{a}_{ij}, p = \frac{\varepsilon}{3})
\label{eq:ramr}
\end{equation}   \tag{5}\]</span></p>
<p>Thus, taking all these into consideration, the final noisy VAF values
<span class="math inline">\(f^{(n)}_{ij}\)</span> are simulated as:</p>
<p><span class="math display" id="eq:noisyVAF">\[\begin{equation}
f^{(n)}_{ij} = \frac{r^{a}_{ij} - r^{a\prime}_{ij} + r^{r\prime}_{ij}}{r_{ij}}
\label{eq:noisyVAF}
\end{equation}   \tag{6}\]</span></p>
<p>By default, the sequencing error rate <span class="math inline">\(\varepsilon\)</span> is set to 0.001,
following commonly reported values for Illumina data (Loman et al.
2012).</p>
<p>As an illustration of the effect of the noise model, in Figure
<a href="#F_error_plot" data-reference-type="ref" data-reference="F_error_plot">2</a>, we
have depicted the density of the mean absolute error between the
<span class="math inline">\(\boldsymbol{F^{(n)}}\)</span> matrix and its corresponding noise-free
<span class="math inline">\(\boldsymbol{F}\)</span> matrix for a collection of noisy instances. As can be
seen, as <span class="math inline">\(\mu_{sd}\)</span> increases, the error introduced to the
<span class="math inline">\(\boldsymbol{F^{(n)}}\)</span> matrix decreases. This is expected because the
<span class="math inline">\(r^{a}_{ij}\)</span> values follow a binomial distribution as described in
Equation <a href="#eq:ra">(4)</a>, where the number of trials is determined by
<span class="math inline">\(\mu_{sd}\)</span> as shown in Equation <a href="#eq:r">(3)</a>, and the event probability
corresponds to the <span class="math inline">\(f^{(n)}_{ij}\)</span> value. Therefore, the larger the
number of trials, the closer the noisy VAF value is to the noise-free
VAF value.</p>
<figure id="F_error_plot" data-latex-placement="t">
<img src="figs/noisy_F_error.png" style="width:100.0%"
alt="graphic without alt text" />
<figcaption>
Figure 2: Density of the mean absolute error in noisy <span class="math inline"><strong>F</strong></span> matrices for different
<span class="math inline"><em>μ</em><sub><em>s</em><em>d</em></sub></span>
values that correspond to different noise levels.
</figcaption>
</figure>
<p>As a final remark, it is important to note that although our data
simulation procedure follows the ISA, the addition of noise may cause
the resulting data to break this assumption.</p>
<h3 data-number="3" id="sec:package"><span class="header-section-number">3</span> The package</h3>
<p><strong>GeRnika</strong> provides three main functionalities for studying tumor
evolution data: (I) simulating artificial tumor evolution, (II)
visualizing tumor phylogenies, and (III) comparing tumor phylogenies.
This section explains the functions that support these features.
Additionally, we describe extra data provided by <strong>GeRnika</strong> that users
can use to try the methods in the package.</p>
<h4 class="unnumbered" data-number="3.1" id="simulation-methods">Simulation methods</h4>
<p>To enable users to simulate tumor evolution data, <strong>GeRnika</strong> provides
various functions inspired by the methods described in
<a href="#sec:simulation" data-reference-type="ref" data-reference="sec:simulation">2</a>.
GeRnika offers two options: a single method for streamlined simulations
and separate methods for performing each step individually, allowing
users to customize or replace specific parts of the process.</p>
<section id="create_instance" class="unnumbered">
<h5 class="unnumbered">create_instance</h5>
<p>The main function for streamlined tumor data simulation is
<code>create_instance</code>. This function provides a convenient way to perform
the entire simulation process in a single step. The following command
demonstrates how to use it to generate the artificial data:</p>
<div class="center">
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>    <span class="fu">create_instance</span>(n, m, k, selection, <span class="at">noisy =</span> <span class="cn">TRUE</span>, <span class="at">depth =</span> <span class="dv">30</span>, <span class="at">seed =</span> <span class="fu">Sys.time</span>())</span></code></pre></div>
</div>
<p>where each argument of the method is described as follows:</p>
<ul>
<li><p><code>n</code>: An integer representing the number of clones.</p></li>
<li><p><code>m</code>: An integer representing the number of samples.</p></li>
<li><p><code>k</code>: A numeric value that determines the linearity of the tree
topology. Also referred to as the topology parameter. Increasing
values of this parameter increase the linearity of the topology. When
<code>k</code> is set to 1, all nodes have equal probabilities of being chosen as
parents, resulting in a uniformly random topology.</p></li>
<li><p><code>selection</code>: A character string representing the evolutionary mode the
tumor follows. This should be either "positive" or "neutral".</p></li>
<li><p><code>noisy</code>: A logical value (<code>TRUE</code> by default) indicating whether to add
noise to the frequency matrix. If <code>TRUE</code>, noise is added to the
frequency matrix. If <code>FALSE</code>, no noise is added.</p></li>
<li><p><code>depth</code>: A numeric value (30 by default) representing the mean depth
of sequencing.</p></li>
<li><p><code>seed</code>: A numeric value (<code>Sys.time()</code> by default) used to set the seed
for the random number generator.</p></li>
</ul>
<p>The <code>create_instance</code> function returns a list containing the following
components:</p>
<ul>
<li><p><code>F_noisy</code>: A matrix representing the noisy frequencies of each
mutation across samples. If the <code>noisy</code> parameter is set to <code>FALSE</code>,
this matrix is equal to <code>F_true</code>.</p></li>
<li><p><code>B</code>: A matrix representing the relationships between mutations and
clones in the tumor.</p></li>
<li><p><code>U</code>: A matrix representing the frequencies of the clones across the
set of samples.</p></li>
<li><p><code>F_true</code>: A matrix representing the noise-free frequencies of each
mutation across the samples.</p></li>
</ul>
<p>As explained, the <code>create_instance</code> function generates all matrices
representing frequencies, proportions, and the phylogeny of the
simulated tumor data in a single step. However, <strong>GeRnika</strong> also
provides individual functions for simulating each of these elements
independently, providing users with greater control over the
characteristics of the simulated tumor data.</p>
</section>
<section id="create_b-create_u-create_f-and-add_noise" class="unnumbered">
<h5 class="unnumbered">create_B, create_U, create_F and add_noise</h5>
<p>These methods provide specialized functions to generate each matrix
involved in representing tumor evolution data. These functions include
the following:</p>
<ul>
<li><p><code>create_B</code>: This function generates a mutation matrix
(<span class="math inline">\(\boldsymbol{B}\)</span> matrix) for a tumor phylogenetic tree with a given
number of nodes and a value <code>k</code> determining the linearity of the tree
topology.</p></li>
<li><p><code>create_U</code>: This function calculates the <span class="math inline">\(\boldsymbol{U}\)</span> matrix,
containing the frequencies of each clone in a set of samples, based on
a <span class="math inline">\(\boldsymbol{B}\)</span> matrix, the number of samples considered, the
number of cells in each sample, and the evolutionary mode of the
tumor.</p></li>
<li><p><code>create_F</code>: This function generates the <span class="math inline">\(\boldsymbol{F}\)</span> matrix, which
contains mutation frequency values for a series of mutations across a
collection of tumor biopsies or samples. The matrix is computed based
on a pair of matrices, <span class="math inline">\(\boldsymbol{U}\)</span> and <span class="math inline">\(\boldsymbol{B}\)</span>, and
considers whether the mutations are heterozygous.</p></li>
<li><p><code>add_noise</code>: This function introduces sequencing noise into the
noise-free <span class="math inline">\(\boldsymbol{F}\)</span> matrix generated by the <code>create_F</code> method.
Users can specify the mean sequencing depth and the overdispersion
parameter, which are used to simulate sequencing depth based on a
negative binomial distribution.</p></li>
</ul>
<p>The reader is encouraged to refer to the package documentation for more
information about these functions and their parameters.</p>
</section>
<h4 class="unnumbered" data-number="3.2" id="visualization-methods">Visualization methods</h4>
<p>The following functions enable the visualization of tumor evolution data
by generating phylogenetic trees based on the data under analysis.</p>
<section id="phylotree-s4-class" class="unnumbered">
<h5 class="unnumbered">Phylotree S4 class</h5>
<p>To simplify the execution of its functionalities, <strong>GeRnika</strong> utilizes
the <code>"Phylotree"</code> class. The <code>"Phylotree"</code> S4 class is a data structure
specifically designed to represent phylogenetic trees, facilitating the
use of the package’s methods and ensuring their computational
efficiency. The attributes of the <code>"Phylotree"</code> class are as follows:</p>
<ul>
<li><p><code>B</code>: A data.frame containing the square matrix that represents the
ancestral relationships among the clones in the phylogenetic tree
(<span class="math inline">\(\boldsymbol{B}\)</span> matrix).</p></li>
<li><p><code>clones</code>: A vector representing the indices of the clones in the
<span class="math inline">\(\boldsymbol{B}\)</span> matrix.</p></li>
<li><p><code>genes</code>: A vector indicating the index of the gene that firstly
mutated in each clone within the <span class="math inline">\(\boldsymbol{B}\)</span> matrix.</p></li>
<li><p><code>parents</code>: A vector indicating the parent clones for each clone in the
phylogenetic tree.</p></li>
<li><p><code>tree</code>: A <code>"Node"</code> class object representing the phylogenetic tree
(this class is inherited from the
<a href="https://CRAN.R-project.org/package=data.tree"><strong>data.tree</strong></a>
package).</p></li>
<li><p><code>labels</code>: A vector containing the gene tags associated with the nodes
in the phylogenetic tree.</p></li>
</ul>
<p>A customized <code>"Phylotree"</code> class object can be instantiated with custom
attributes using the <code>create_instance</code> method. This method takes all the
attributes according to the <code>"Phylotree"</code> class as arguments.
Alternatively, <strong>GeRnika</strong> provides a function that automatically
generates a <code>"Phylotree"</code> class object on the basis of a given
<span class="math inline">\(\boldsymbol{B}\)</span>.</p>
</section>
<section id="b_to_phylotree" class="unnumbered">
<h5 class="unnumbered">B_to_phylotree</h5>
<p>In order to instantiate an object of the “<code>Phylotree</code>” class, the
following command can be used:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>    <span class="fu">B_to_phylotree</span>(B, <span class="at">labels =</span> <span class="cn">NA</span>)</span></code></pre></div>
<p>where each argument of the method is described as follows:</p>
<ul>
<li><p><code>B</code>: A square <span class="math inline">\(\boldsymbol{B}\)</span> matrix that represents the phylogenetic
tree.</p></li>
<li><p><code>labels</code>: An optional vector containing the tags of the genes in the
phylogenetic tree. <code>NA</code> by default.</p></li>
</ul>
<p>This function returns an object of the <code>"Phylotree"</code> class,
automatically generating its attributes based on <code>B</code>, which represents
the phylogenetic tree of the tumor under analysis.</p>
<p>Once instantiated, the phylogenetic tree in a <code>"Phylotree"</code> class object
can be visualized using the generic <code>plot</code> function, which takes the
<code>"Phylotree"</code> object as its argument. The <code>plot</code> function also includes
a <code>labels</code> argument that can be set to <code>TRUE</code> to display node labels on
the phylogenetic tree, using the gene tags stored within the
<code>"Phylotree"</code> object.</p>
<p>The <strong>GeRnika</strong> package provides the <code>plot_proportions</code> function for
visualizing phylogenetic trees, with node sizes and colors reflecting
the proportions of each clone. This function requires two inputs: a
<code>"Phylotree"</code> class object representing the phylogenetic tree and a
numeric vector or matrix specifying clone proportions. If a vector is
provided, a single tree is plotted, with the node sizes and colors
determined by the values in the vector. Instead, if a matrix is
provided, such as the <span class="math inline">\(\boldsymbol{U}\)</span> matrix that represents the
frequencies of clones across samples, the function plots one tree for
each row of the matrix. Each tree is generated based on the clone
proportions specified in the corresponding row. Additionally, users can
enable node labeling by setting the <code>labels</code> argument to <code>TRUE</code>, which
annotates the tree nodes with gene tags from the <code>"Phylotree"</code> object.</p>
</section>
<h4 class="unnumbered" data-number="3.3" id="comparison-methods">Comparison methods</h4>
<p>This section describes the methods included in <strong>GeRnika</strong> that
facilitate the comparison of tumor phylogenies.</p>
<p>A fundamental approach for comparing two phylogenetic trees is to
determine if their evolutionary histories are equivalent. The <code>equals</code>
function performs this comparison by accepting two <code>"Phylotree"</code> class
objects as arguments. This function returns a boolean value indicating
whether the provided phylogenetic trees are equivalent.</p>
<p>To analyze similarities and differences between two phylogenetic trees,
the <code>find_common_subtrees</code> function identifies and plots all maximal
common subtrees between them. In addition to visualizing these subtrees,
the function outputs the number of shared and unique edges (those
present in only one of the trees) and calculates the distance between
the trees, defined as the sum of their unique edges. This method also
includes an option to label the maximal common subtrees with gene tags
by setting <code>labels = TRUE</code>.</p>
<p>The <code>combine_trees</code> function generates a consensus tree by combining the
nodes and edges of two <code>"Phylotree"</code> class objects. The consensus tree
highlights the nodes and edges that form common subtrees between the
original trees, as well as the independent edges unique to each tree,
which are displayed with reduced opacity. This method also allows
labeling the nodes with gene tags by setting <code>labels = TRUE</code> and
customizing the colors of the consensus tree by passing a 3-element
hexadecimal vector to the <code>palette</code> argument.</p>
<h4 class="unnumbered" data-number="3.4" id="exported-data">Exported data</h4>
<p><strong>GeRnika</strong> provides various exported data instances to help users
easily explore the package’s functionalities. These are as follows:</p>
<ul>
<li><p><code>B_mats</code>: A list of 10 trios of <span class="math inline">\(\boldsymbol{B}\)</span> matrices. Each trio
includes a real <span class="math inline">\(\boldsymbol{B}\)</span> matrix and two <span class="math inline">\(\boldsymbol{B}\)</span>
matrices generated using different algorithms that infer evolutionary
relationships from a given <span class="math inline">\(\boldsymbol{F}\)</span> matrix. These matrices
serve as illustrative examples for testing and exploring the package’s
functionalities.</p></li>
<li><p><code>palettes</code>: A data frame containing three predefined palettes for use
with methods in <strong>GeRnika</strong> that require color palettes.</p></li>
</ul>
<h3 data-number="4" id="examples"><span class="header-section-number">4</span> Examples</h3>
<p>In this section we show examples of the use of the methods explained in
<a href="#sec:package" data-reference-type="ref" data-reference="sec:package">3</a>. Please
note that in the following examples, we will set the seeds of
non-deterministic methods to a predefined value to ensure
reproducibility.</p>
<h4 class="unnumbered" data-number="4.1" id="simulating-tumor-evolution-data">Simulating tumor evolution data</h4>
<p>The simulation of tumor clonal data involves generating the matrices
<span class="math inline">\(\boldsymbol{B}\)</span>, <span class="math inline">\(\boldsymbol{U}\)</span>, <span class="math inline">\(\boldsymbol{F}\)</span>, and
<span class="math inline">\(\boldsymbol{F^{(n)}}\)</span> associated with a specific instance. For example,
we can simulate a noisy instance of a tumor composed of 5
clones/mutations, which has evolved under neutral evolution with a <span class="math inline">\(k\)</span>
value of 0.5, and from which 3 samples have been taken in a single line
of code, as follows:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> I <span class="ot">&lt;-</span> <span class="fu">create_instance</span>(<span class="at">n =</span> <span class="dv">5</span>, <span class="at">m =</span> <span class="dv">3</span>, <span class="at">k =</span> <span class="fl">0.5</span>, <span class="at">selection =</span> <span class="st">&quot;neutral&quot;</span>, <span class="at">seed =</span> <span class="dv">1</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> I</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>F_noisy</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>         mut1       mut2      mut3      mut4      mut5</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>sample1 <span class="fl">1.000</span> <span class="fl">0.09090909</span> <span class="fl">0.0000000</span> <span class="fl">0.2777778</span> <span class="fl">0.3548387</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>sample2 <span class="fl">1.000</span> <span class="fl">0.20000000</span> <span class="fl">0.2631579</span> <span class="fl">0.8536585</span> <span class="fl">0.2000000</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>sample3 <span class="fl">0.975</span> <span class="fl">0.03846154</span> <span class="fl">1.0000000</span> <span class="fl">1.0000000</span> <span class="fl">0.0000000</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>B</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>       mut1 mut2 mut3 mut4 mut5</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>clone1    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">0</span>    <span class="dv">0</span>    <span class="dv">0</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>clone2    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">0</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>clone3    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">0</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>clone4    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">0</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>clone5    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">1</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>U</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>        clone1 clone2 clone3 clone4 clone5</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>sample1   <span class="fl">0.59</span>   <span class="fl">0.13</span>   <span class="fl">0.00</span>   <span class="fl">0.01</span>   <span class="fl">0.27</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>sample2   <span class="fl">0.13</span>   <span class="fl">0.27</span>   <span class="fl">0.24</span>   <span class="fl">0.19</span>   <span class="fl">0.17</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>sample3   <span class="fl">0.00</span>   <span class="fl">0.04</span>   <span class="fl">0.89</span>   <span class="fl">0.07</span>   <span class="fl">0.00</span></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>F_true</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>        mut1 mut2 mut3 mut4 mut5</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>sample1    <span class="dv">1</span> <span class="fl">0.13</span> <span class="fl">0.00</span> <span class="fl">0.41</span> <span class="fl">0.27</span></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>sample2    <span class="dv">1</span> <span class="fl">0.27</span> <span class="fl">0.24</span> <span class="fl">0.87</span> <span class="fl">0.17</span></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>sample3    <span class="dv">1</span> <span class="fl">0.04</span> <span class="fl">0.89</span> <span class="fl">1.00</span> <span class="fl">0.00</span></span></code></pre></div>
<p>Using this approach, the previously mentioned four matrices are
simulated. Note that the noise-free <span class="math inline">\(\boldsymbol{F}\)</span> matrix is referred
to as <code>F_true</code> in the package’s code, while the noisy
<span class="math inline">\(\boldsymbol{F^{(n)}}\)</span> is denoted as <code>F_noisy</code>.</p>
<p>The previous method allows users to generate instances easily and
quickly. However, some users may require more precise control over the
data, which can be achieved using the <code>create_B</code>, <code>create_U</code>,
<code>create_F</code>, and <code>add_noise</code> methods. For examples on how to use these
methods, please refer to the package documentation.</p>
<h4 class="unnumbered" data-number="4.2" id="visualizing-tumor-phylogenies">Visualizing tumor phylogenies</h4>
<p>Once the matrices associated with our tumor instance have been
generated, we can create a <code>"Phylotree"</code> class object, as follows:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> phylotree <span class="ot">&lt;-</span> <span class="fu">B_to_phylotree</span>(<span class="at">B =</span> I<span class="sc">$</span>B)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> phylotree</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>An object of class <span class="st">&quot;Phylotree&quot;</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>Slot <span class="st">&quot;B&quot;</span><span class="sc">:</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>       mut1 mut2 mut3 mut4 mut5</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>clone1    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">0</span>    <span class="dv">0</span>    <span class="dv">0</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>clone2    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">0</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>clone3    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">0</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>clone4    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">0</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>clone5    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">1</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>Slot <span class="st">&quot;clones&quot;</span><span class="sc">:</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>mut1 mut2 mut3 mut4 mut5 </span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>   <span class="dv">1</span>    <span class="dv">2</span>    <span class="dv">3</span>    <span class="dv">4</span>    <span class="dv">5</span> </span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>Slot <span class="st">&quot;genes&quot;</span><span class="sc">:</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>mut1 mut2 mut3 mut4 mut5 </span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>   <span class="dv">1</span>    <span class="dv">2</span>    <span class="dv">3</span>    <span class="dv">4</span>    <span class="dv">5</span> </span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>Slot <span class="st">&quot;parents&quot;</span><span class="sc">:</span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="sc">-</span><span class="dv">1</span>  <span class="dv">4</span>  <span class="dv">4</span>  <span class="dv">1</span>  <span class="dv">4</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>Slot <span class="st">&quot;tree&quot;</span><span class="sc">:</span></span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>  levelName</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> <span class="dv">1</span>        </span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>  °<span class="sc">--</span><span class="dv">4</span>    </span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>      ¦<span class="sc">--</span><span class="dv">2</span></span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>      ¦<span class="sc">--</span><span class="dv">3</span></span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>      °<span class="sc">--</span><span class="dv">5</span></span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>Slot <span class="st">&quot;labels&quot;</span><span class="sc">:</span></span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">&quot;mut1&quot;</span> <span class="st">&quot;mut4&quot;</span> <span class="st">&quot;mut2&quot;</span> <span class="st">&quot;mut3&quot;</span> <span class="st">&quot;mut5&quot;</span></span></code></pre></div>
<p>Since no list of tags is provided to the <code>labels</code> parameter, a default
set of labels is automatically assigned to the instantiated
<code>"Phylotree"</code> class object.</p>
<p>Afterwards, we can visualize the tumor phylogeny associated to the
simulated <span class="math inline">\(\boldsymbol{B}\)</span> matrix by using the generic <code>plot</code> method, as
follows:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot</span>(phylotree)</span></code></pre></div>
<figure id="fig:phylotree" data-latex-placement="t">
<img src="figs/phylotree.png" style="width:50.0%"
alt="graphic without alt text" />
<figcaption>
Figure 3: Phylogenetic tree associated to the generated
<code>“Phylotree”</code> class object.
</figcaption>
</figure>
<p>The resulting plot is shown in Figure
<a href="#fig:phylotree" data-reference-type="ref" data-reference="fig:phylotree">3</a>.
Instead of clone numbers, the user can utilize the predefined tags in
the <code>"Phylotree"</code> class object to label the nodes in the tree by setting
the <code>labels = TRUE</code> parameter in the <code>plot</code> function.</p>
<p>When plotting a <code>"Phylotree"</code> class object, its nodes can be resized
according to the proportions of the clones that compose the tumor
samples. To achieve this, we can use the <span class="math inline">\(\boldsymbol{U}\)</span> matrix from
the previously generated instance to determine the proportions of the
clones, as shown below:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot_proportions</span>(phylotree, I<span class="sc">$</span>U, <span class="at">labels =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<figure id="fig:proportions" data-latex-placement="t">
<img src="figs/proportions.png" style="width:100.0%"
alt="graphic without alt text" />
<figcaption>
Figure 4: Phylogenetic trees associated to the generated
<code>“Phylotree”</code> class object, using the proportions associated
to the previously generated <span class="math inline"><strong>U</strong></span> matrix.
</figcaption>
</figure>
<p>The resulting plot is shown in Figure
<a href="#fig:proportions" data-reference-type="ref" data-reference="fig:proportions">4</a>.
This method plots the proportions of the <span class="math inline">\(\boldsymbol{U}\)</span> matrix, where
each tree represents a sample from the <span class="math inline">\(\boldsymbol{U}\)</span> matrix,
illustrating the proportion of each clone within that specific sample.
In this case, we have set the <code>labels</code> parameter to <code>TRUE</code> to label the
nodes in the tree using the predefined tags "mut1", "mut4",
"mut2", "mut3", and "mut5".</p>
<h4 class="unnumbered" data-number="4.3" id="comparing-tumor-phylogenies">Comparing tumor phylogenies</h4>
<p>Now, we present examples showing the use of <strong>GeRnika</strong>’s
functionalities for comparing tumor phylogenies. For this purpose, we
will use the <code>B_mats</code> object included in the <strong>GeRnika</strong> package, which
contains 10 <span class="math inline">\(\boldsymbol{B}\)</span> matrix trios. Specifically, we will use the
first trio of matrices, and set a predefined set of tags for the clones
in the trees:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> B_mats <span class="ot">&lt;-</span> GeRnika<span class="sc">::</span>B_mats</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> B_real <span class="ot">&lt;-</span> B_mats[[<span class="dv">1</span>]]<span class="sc">$</span>B_real</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> B_alg1 <span class="ot">&lt;-</span> B_mats[[<span class="dv">1</span>]]<span class="sc">$</span>B_alg1</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> B_alg2 <span class="ot">&lt;-</span> B_mats[[<span class="dv">1</span>]]<span class="sc">$</span>B_alg2</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> tags <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;TP53&quot;</span>, <span class="st">&quot;KRAS&quot;</span>, <span class="st">&quot;PIK3CA&quot;</span>, <span class="st">&quot;APC&quot;</span>, <span class="st">&quot;EGFR&quot;</span>, <span class="st">&quot;BRCA1&quot;</span>, <span class="st">&quot;PTEN&quot;</span>, <span class="st">&quot;BRAF&quot;</span>, </span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>            <span class="st">&quot;MYC&quot;</span>, <span class="st">&quot;CDKN2A&quot;</span>)</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> phylotree_real <span class="ot">&lt;-</span> <span class="fu">B_to_phylotree</span>(<span class="at">B =</span> B_real, <span class="at">labels =</span> tags)</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> phylotree_alg2 <span class="ot">&lt;-</span> <span class="fu">B_to_phylotree</span>(<span class="at">B =</span> B_alg2, <span class="at">labels =</span> tags)</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> phylotree_alg1 <span class="ot">&lt;-</span> <span class="fu">B_to_phylotree</span>(<span class="at">B =</span> B_alg1, <span class="at">labels =</span> tags)</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot</span>(phylotree_real, <span class="at">labels=</span><span class="cn">TRUE</span>)</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot</span>(phylotree_alg1, <span class="at">labels=</span><span class="cn">TRUE</span>)</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">plot</span>(phylotree_alg2, <span class="at">labels=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>The plots of the three instantiated <code>Phylotree</code> class objects are
depicted in Figure <a href="#fig:phylocomparison" data-reference-type="ref" data-reference="fig:phylocomparison">5</a>.</p>
<figure id="fig:phylocomparison" data-latex-placement="t">
<img src="figs/comparisonphylotrees.png" style="width:100.0%"
alt="graphic without alt text" />
<figcaption>
Figure 5: <code>phylotree_real</code>,
<code>phylotree_alg2</code> and <code>phylotree_alg1</code>, from the
left to the right.
</figcaption>
</figure>
<p>We can check if the phylogenies of two tumors are equivalent using the
<code>equals</code> method:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">equals</span>(<span class="at">phylotree_1 =</span> phylotree_real, <span class="at">phylotree_2 =</span> phylotree_alg1)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">FALSE</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">equals</span>(<span class="at">phylotree_1 =</span> phylotree_real, <span class="at">phylotree_2 =</span> phylotree_real)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="cn">TRUE</span></span></code></pre></div>
<p>In this case, <code>phylotree_real</code> and <code>phylotree_alg1</code> are not identical,
as some edges present in <code>phylotree_real</code> are absent in
<code>phylotree_alg1</code>, and vice versa. However, a phylogenetic tree will
always be identical to itself, as shown when comparing <code>phylotree_real</code>
to itself.</p>
<p>To find the maximal common subtrees between two phylogenetic trees, we
do it using the following command:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">find_common_subtrees</span>(<span class="at">phylotree_1 =</span> phylotree_real, <span class="at">phylotree_2 =</span> phylotree_alg2, </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">labels =</span> <span class="cn">TRUE</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>Independent edges of tree1<span class="sc">:</span> <span class="dv">2</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>Independent edges of tree2<span class="sc">:</span> <span class="dv">2</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>Common edges<span class="sc">:</span> <span class="dv">7</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>Distance<span class="sc">:</span> <span class="dv">4</span></span></code></pre></div>
<figure id="fig:commonsubtrees" data-latex-placement="t">
<img src="figs/commonsubtrees.png" style="width:100.0%"
alt="graphic without alt text" />
<figcaption>
Figure 6: Maximal common subtrees between
<code>phylotree_real</code> and <code>phylotree_alg2</code> using
predefined tags. In this case, there exists a single common subtree, but
there may exist more in other cases.
</figcaption>
</figure>
<figure id="fig:consensus" data-latex-placement="t!">
<img src="figs/consensus.png" style="width:40.0%"
alt="graphic without alt text" />
<figcaption>
Figure 7: Consensus tree between <code>phylotree_real</code>
and <code>phylotree_alg1</code> using the <em>Lancet</em> palette and
predefined tags.
</figcaption>
</figure>
<p>The maximal common subtrees (in this case, one subtree) between
<code>phylotree_real</code> and<br />
<code>phylotree_alg2</code> are shown in Figure
<a href="#fig:commonsubtrees" data-reference-type="ref" data-reference="fig:commonsubtrees">6</a>. Note that the clones in the maximal
common subtree are represented by the predefined tags in the <code>Phylotree</code>
class objects as we have set <code>labels = TRUE</code>. Additionally, this method
prints the number of common and independent edges of the trees, along
with the distance between them.</p>
<p>Finally, we generate the consensus tree between two phylogenetic trees
using one of the custom palettes offered by <strong>GeRnika</strong>, specifically
the <em>Lancet</em> palette, and the predefined tags for the clones as follows:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> palette <span class="ot">&lt;-</span> GeRnika<span class="sc">::</span>palettes<span class="sc">$</span>Lancet</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> consensus_real_alg1 <span class="ot">&lt;-</span> <span class="fu">combine_trees</span>(<span class="at">phylotree_1 =</span> phylotree_real, </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">phylotree_2 =</span> phylotree_alg1,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">labels =</span> <span class="cn">TRUE</span>, </span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>                                      <span class="at">palette =</span> palette)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> DiagrammeR<span class="sc">::</span><span class="fu">render_graph</span>(consensus_real_alg1)</span></code></pre></div>
<p>The consensus tree between <code>phylotree_real</code> and <code>phylotree_alg1</code> is
depicted in Figure <a href="#fig:consensus" data-reference-type="ref" data-reference="fig:consensus">7</a>. Here, the nodes and the edges that compose
the common subtrees between the original trees are green. In addition,
pink edges denote the independent edges of the tree passed as the first
parameter of the method, while blue edges represent the independent
edges of the second tree. Note that the independent edges of both trees
are presented with translucent colors.</p>
<h3 data-number="5" id="conclusions"><span class="header-section-number">5</span> Conclusions</h3>
<p><strong>GeRnika</strong> is a comprehensive R package designed to address a critical
gap in the tools available for studying tumor evolution within the R
environment. To this end, it provides researchers with an integrated
suite for simulating, visualizing, and comparing tumor phylogenies.
Unlike many existing tools, <strong>GeRnika</strong> is fully implemented in R,
making it particularly accessible for the bioinformatics community,
which widely relies on R for data analysis and visualization.</p>
<p>One of <strong>GeRnika</strong>’s key contributions is providing tools to generate
biologically plausible datasets for studying intratumoral heterogeneity
and clonal dynamics. With varied, easily customizable
parameters—controlling features such as clonal tree topology,
selective pressures, and sequencing noise—<strong>GeRnika</strong> enables
exploration of a wide range of evolutionary patterns and complexities
that provide a valuable resource for testing new methods and hypotheses
in tumor heterogeneity research.</p>
<p>Beyond its core simulation features, <strong>GeRnika</strong> includes tools for
visualizing and comparing tumor phylogenies, offering a unified solution
that eliminates the need for multiple packages or complex data
processing workflows. Future work will focus on enhancing <strong>GeRnika</strong>’s
compatibility with other R packages related to tumor evolution, allowing
for easier integration with existing resources.</p>
<p>Overall, <strong>GeRnika</strong> provides an accessible, user-friendly tool that
supports research into intratumoral heterogeneity, with the potential to
substantially advance tumor phylogeny research. While it does not
perform clonal deconvolution or phylogeny inference from real sequencing
data, <strong>GeRnika</strong> serves as a valuable platform for simulation,
visualization, and benchmarking, supporting the development and
evaluation of such algorithms.</p>
<h3 data-number="6" id="r-software"><span class="header-section-number">6</span> R software</h3>
<p>The R package <a href="https://CRAN.R-project.org/package=GeRnika"><strong>GeRnika</strong></a>
is now available on CRAN.</p>
</div>
<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-articleeffect" class="csl-entry" role="listitem">
Burrell, Rebecca, Nicholas Mcgranahan, Jiri Bartek, and Charles Swanton.
2013. “The Causes and Consequences of Genetic Heterogeneity in Cancer
Evolution.” <em>Nature</em> 501 (September): 338–45.
<a href="https://doi.org/10.1038/nature12625" class="uri">https://doi.org/10.1038/nature12625</a>.
</div>
<div id="ref-clonevol" class="csl-entry" role="listitem">
Dang, HX, BS White, SM Foltz, et al. 2017. “ClonEvol: Clonal Ordering
and Visualization in Cancer Sequencing.” <em>Annals of Oncology</em> 28 (12):
3076–82. <a href="https://doi.org/10.1093/annonc/mdx517" class="uri">https://doi.org/10.1093/annonc/mdx517</a>.
</div>
<div id="ref-davis2017tumor" class="csl-entry" role="listitem">
Davis, Alexander, Ruli Gao, and Nicholas Navin. 2017. “Tumor Evolution:
Linear, Branching, Neutral or Punctuated?” <em>Biochimica Et Biophysica
Acta (BBA)-Reviews on Cancer</em> 1867 (2): 151–61.
<a href="https://doi.org/10.1016/j.bbcan.2017.01.003" class="uri">https://doi.org/10.1016/j.bbcan.2017.01.003</a>.
</div>
<div id="ref-phylowgs" class="csl-entry" role="listitem">
Deshwar, Amit G, Shankar Vembu, Christina K Yung, Gun Ho Jang, Lincoln
Stein, and Quaid Morris. 2015. “PhyloWGS: Reconstructing Subclonal
Composition and Evolution from Whole-Genome Sequencing of Tumors.”
<em>Genome Biology</em> 16 (1): 1–20.
<a href="https://doi.org/10.1186/s13059-015-0602-8" class="uri">https://doi.org/10.1186/s13059-015-0602-8</a>.
</div>
<div id="ref-ancestree" class="csl-entry" role="listitem">
El-Kebir, Mohammed, Layla Oesper, Hannah Acheson-Field, and Benjamin J.
Raphael. 2015. “<span class="nocase">Reconstruction of clonal trees and tumor composition
from multi-sample sequencing data</span>.” <em>Bioinformatics</em> 31 (12):
62–70. <a href="https://doi.org/10.1093/bioinformatics/btv261" class="uri">https://doi.org/10.1093/bioinformatics/btv261</a>.
</div>
<div id="ref-spruce" class="csl-entry" role="listitem">
El-Kebir, Mohammed, Gryte Satas, Layla Oesper, and Benjamin J Raphael.
2016. “Inferring the Mutational History of a Tumor Using Multi-State
Perfect Phylogeny Mixtures.” <em>Cell Systems</em> 3 (1): 43–53.
<a href="https://doi.org/10.1016/j.cels.2016.07.004" class="uri">https://doi.org/10.1016/j.cels.2016.07.004</a>.
</div>
<div id="ref-machina" class="csl-entry" role="listitem">
El-Kebir, Mohammed, Gryte Satas, and Benjamin J Raphael. 2018.
“Inferring Parsimonious Migration Histories for Metastatic Cancers.”
<em>Nature Genetics</em> 50 (5): 718–26.
<a href="https://doi.org/10.1038/s41588-018-0106-z" class="uri">https://doi.org/10.1038/s41588-018-0106-z</a>.
</div>
<div id="ref-fu2022reconstructing" class="csl-entry" role="listitem">
Fu, Xuecong, Haoyun Lei, Yifeng Tao, and Russell Schwartz. 2022.
“Reconstructing Tumor Clonal Lineage Trees Incorporating
Single-Nucleotide Variants, Copy Number Alterations and Structural
Variations.” <em>Bioinformatics</em> 38 (Supplement_1): i125–33.
<a href="https://doi.org/10.1093/bioinformatics/btac253" class="uri">https://doi.org/10.1093/bioinformatics/btac253</a>.
</div>
<div id="ref-contreedp" class="csl-entry" role="listitem">
Fu, Xuecong, and Russell Schwartz. 2021. “ConTreeDP: A Consensus Method
of Tumor Trees Based on Maximum Directed Partition Support Problem.”
<em>2021 IEEE International Conference on Bioinformatics and Biomedicine
(BIBM)</em>, 125–30. <a href="https://doi.org/10.1101/2021.10.13.463978" class="uri">https://doi.org/10.1101/2021.10.13.463978</a>.
</div>
<div id="ref-grigoriadis2024conipher" class="csl-entry" role="listitem">
<span class="nocase">Grigoriadis, Kristiana, Ariana Huebner, Abigail Bunkum, et
al.</span> 2024. “CONIPHER: A Computational Framework for Scalable
Phylogenetic Reconstruction with Error Correction.” <em>Nature Protocols</em>
19 (1): 159–83. <a href="https://doi.org/10.1038/s41596-023-00913-9" class="uri">https://doi.org/10.1038/s41596-023-00913-9</a>.
</div>
<div id="ref-tuelip" class="csl-entry" role="listitem">
Guang, Ziyun, Matthew Smith-Erb, and Layla Oesper. 2023. “A Weighted
Distance-Based Approach for Deriving Consensus Tumor Evolutionary
Trees.” <em>Bioinformatics</em> 39 (Supplement_1): i204–12.
<a href="https://doi.org/10.1093/bioinformatics/btad230" class="uri">https://doi.org/10.1093/bioinformatics/btad230</a>.
</div>
<div id="ref-gusfield" class="csl-entry" role="listitem">
Gusfield, Dan. 1991. “Efficient Algorithms for Inferring Evolutionary
Trees.” <em>Networks</em> 21 (1): 19–28.
<a href="https://doi.org/10.1002/net.3230210104" class="uri">https://doi.org/10.1002/net.3230210104</a>.
</div>
<div id="ref-mipup" class="csl-entry" role="listitem">
Husić, Edin, Xinyue Li, Ademir Hujdurović, et al. 2019. “MIPUP: Minimum
Perfect Unmixed Phylogenies for Multi-Sampled Tumors via Branchings and
ILP.” <em>Bioinformatics</em> 35 (5): 769–77.
<a href="https://doi.org/10.1093/bioinformatics/bty683" class="uri">https://doi.org/10.1093/bioinformatics/bty683</a>.
</div>
<div id="ref-canopy" class="csl-entry" role="listitem">
Jiang, Yuchao, Yu Qiu, Andy J Minn, and Nancy R Zhang. 2016. “Assessing
Intratumor Heterogeneity and Tracking Longitudinal and Spatial Clonal
Evolutionary History by Next-Generation Sequencing.” <em>Proceedings of the
National Academy of Sciences</em> 113 (37): E5528–37.
<a href="https://doi.org/10.1073/pnas.1522203113" class="uri">https://doi.org/10.1073/pnas.1522203113</a>.
</div>
<div id="ref-kimuraISA" class="csl-entry" role="listitem">
Kimura, Motoo. 1969. “The Number of Heterozygous Nucleotide Sites
Maintained in a Finite Population Due to Steady Flux of Mutations.”
<em>Genetics</em> 61 (4): 893. <a href="https://doi.org/10.1093/genetics/61.4.893" class="uri">https://doi.org/10.1093/genetics/61.4.893</a>.
</div>
<div id="ref-pairtree" class="csl-entry" role="listitem">
Kulman, Ethan, Jeff Wintersinger, and Quaid Morris. 2022.
“Reconstructing Cancer Phylogenies Using Pairtree, a Clone Tree
Reconstruction Algorithm.” <em>STAR Protocols</em> 3 (4): 101706.
<a href="https://doi.org/10.1016/j.xpro.2022.101706" class="uri">https://doi.org/10.1016/j.xpro.2022.101706</a>.
</div>
<div id="ref-loman2012performance" class="csl-entry" role="listitem">
Loman, Nicholas J, Raju V Misra, Timothy J Dallman, et al. 2012.
“Performance Comparison of Benchtop High-Throughput Sequencing
Platforms.” <em>Nature Biotechnology</em> 30 (5): 434–39.
<a href="https://doi.org/10.1038/nbt.2198" class="uri">https://doi.org/10.1038/nbt.2198</a>.
</div>
<div id="ref-citup" class="csl-entry" role="listitem">
Malikic, Salem, Andrew W McPherson, Nilgun Donmez, and Cenk S Sahinalp.
2015. “Clonality Inference in Multiple Tumor Samples Using Phylogeny.”
<em>Bioinformatics</em> 31 (9): 1349–56.
<a href="https://doi.org/10.1093/bioinformatics/btv003" class="uri">https://doi.org/10.1093/bioinformatics/btv003</a>.
</div>
<div id="ref-malikic2019phiscs" class="csl-entry" role="listitem">
Malikic, Salem, Farid Rashidi Mehrabadi, Simone Ciccolella, et al. 2019.
“PhISCS: A Combinatorial Approach for Subperfect Tumor Phylogeny
Reconstruction via Integrative Use of Single-Cell and Bulk Sequencing
Data.” <em>Genome Research</em> 29 (11): 1860–77.
<a href="https://doi.org/10.1101/gr.234435.118" class="uri">https://doi.org/10.1101/gr.234435.118</a>.
</div>
<div id="ref-1" class="csl-entry" role="listitem">
Marass, Francesco, Florent Mouliere, Ke Yuan, Nitzan Rosenfeld, and
Florian Markowetz. 2016. “A Phylogenetic Latent Feature Model for Clonal
Deconvolution.” <em>The Annals of Applied Statistics</em> 10: 2377–404.
<a href="https://doi.org/10.1214/16-AOAS986" class="uri">https://doi.org/10.1214/16-AOAS986</a>.
</div>
<div id="ref-fishplot" class="csl-entry" role="listitem">
Miller, Christopher A, Joshua McMichael, Ha X Dang, et al. 2016.
“Visualizing Tumor Evolution with the Fishplot Package for R.” <em>BMC
Genomics</em> 17: 1–3. <a href="https://doi.org/10.1186/s12864-016-3195-z" class="uri">https://doi.org/10.1186/s12864-016-3195-z</a>.
</div>
<div id="ref-calder" class="csl-entry" role="listitem">
Myers, Matthew A, Gryte Satas, and Benjamin J Raphael. 2019. “CALDER:
Inferring Phylogenetic Trees from Longitudinal Tumor Samples.” <em>Cell
Systems</em> 8 (6): 514–22. <a href="https://doi.org/10.1016/j.cels.2019.05.010" class="uri">https://doi.org/10.1016/j.cels.2019.05.010</a>.
</div>
<div id="ref-nowell1976clonal" class="csl-entry" role="listitem">
Nowell, Peter C. 1976. “The Clonal Evolution of Tumor Cell Populations.”
<em>Science</em> 194 (4260): 23–28. <a href="https://doi.org/10.1126/science.959840" class="uri">https://doi.org/10.1126/science.959840</a>.
</div>
<div id="ref-depth_error" class="csl-entry" role="listitem">
Petrackova, Anna, Michal Vasinek, Lenka Sedlarikova, et al. 2019.
“Standardization of Sequencing Coverage Depth in NGS: Recommendation for
Detection of Clonal and Subclonal Mutations in Cancer Diagnostics.”
<em>Frontiers in Oncology</em> 9: 851.
<a href="https://doi.org/10.3389/fonc.2019.00851" class="uri">https://doi.org/10.3389/fonc.2019.00851</a>.
</div>
<div id="ref-lichee" class="csl-entry" role="listitem">
Popic, Victoria, Raheleh Salari, Iman Hajirasouliha, Dorna
Kashef-Haghighi, Robert B West, and Serafim Batzoglou. 2015. “Fast and
Scalable Inference of Multi-Sample Cancer Lineages.” <em>Genome Biology</em> 16
(1): 91. <a href="https://doi.org/10.1186/s13059-015-0647-8" class="uri">https://doi.org/10.1186/s13059-015-0647-8</a>.
</div>
<div id="ref-ross2016onconem" class="csl-entry" role="listitem">
Ross, Edith M, and Florian Markowetz. 2016. “OncoNEM: Inferring Tumor
Evolution from Single-Cell Sequencing Data.” <em>Genome Biology</em> 17: 1–14.
<a href="https://doi.org/10.1186/s13059-016-0929-9" class="uri">https://doi.org/10.1186/s13059-016-0929-9</a>.
</div>
<div id="ref-clevrvis" class="csl-entry" role="listitem">
Sandmann, Sarah, Clara Inserte, and Julian Varghese. 2023. “ClevRvis:
Visualization Techniques for Clonal Evolution.” <em>GigaScience</em> 12:
giad020. <a href="https://doi.org/10.1093/gigascience/giad020" class="uri">https://doi.org/10.1093/gigascience/giad020</a>.
</div>
<div id="ref-satas2020scarlet" class="csl-entry" role="listitem">
Satas, Gryte, Simone Zaccaria, Geoffrey Mon, and Benjamin J Raphael.
2020. “SCARLET: Single-Cell Tumor Phylogeny Inference with Copy-Number
Constrained Mutation Losses.” <em>Cell Systems</em> 10 (4): 323–32.
<a href="https://doi.org/10.1016/j.cels.2020.04.001" class="uri">https://doi.org/10.1016/j.cels.2020.04.001</a>.
</div>
<div id="ref-bayclone" class="csl-entry" role="listitem">
Sengupta, Subhajit, Jin Wang, Juhee Lee, et al. 2014. “Bayclone:
Bayesian Nonparametric Inference of Tumor Subclones Using NGS Data.”
<em>Pacific Symposium on Biocomputing Co-Chairs</em>, 467–78.
<a href="https://doi.org/10.1142/9789814644730_0044" class="uri">https://doi.org/10.1142/9789814644730_0044</a>.
</div>
<div id="ref-sollier2023compass" class="csl-entry" role="listitem">
Sollier, Etienne, Jack Kuipers, Koichi Takahashi, Niko Beerenwinkel, and
Katharina Jahn. 2023. “COMPASS: Joint Copy Number and Mutation Phylogeny
Reconstruction from Amplicon Single-Cell Sequencing Data.” <em>Nature
Communications</em> 14 (1): 4921.
<a href="https://doi.org/10.1038/s41467-023-40378-8" class="uri">https://doi.org/10.1038/s41467-023-40378-8</a>.
</div>
<div id="ref-trap" class="csl-entry" role="listitem">
Strino, Francesco, Fabio Parisi, Mariann Micsinai, and Yuval Kluger.
2013. “TrAp: A Tree Approach for Fingerprinting Subclonal Tumor
Composition.” <em>Nucleic Acids Research</em> 41: e165–65.
<a href="https://doi.org/10.1093/nar/gkt641" class="uri">https://doi.org/10.1093/nar/gkt641</a>.
</div>
<div id="ref-tanner2019simulation" class="csl-entry" role="listitem">
Tanner, Georgette, David R Westhead, Alastair Droop, and Lucy F Stead.
2019. “Simulation of Heterogeneous Tumour Genomes with HeteroGenesis and
in Silico Whole Exome Sequencing.” <em>Bioinformatics</em> 35 (16): 2850–52.
<a href="https://doi.org/10.1093/bioinformatics/bty1063" class="uri">https://doi.org/10.1093/bioinformatics/bty1063</a>.
</div>
<div id="ref-clonarch" class="csl-entry" role="listitem">
Wu, Jiaqi, and Mohammed El-Kebir. 2020. “ClonArch: Visualizing the
Spatial Clonal Architecture of Tumors.” <em>Bioinformatics</em> 36
(Supplement_1): i161–68.
<a href="https://doi.org/10.1093/bioinformatics/btaa471" class="uri">https://doi.org/10.1093/bioinformatics/btaa471</a>.
</div>
<div id="ref-bitphylogeny" class="csl-entry" role="listitem">
Yuan, Ke, Thomas Sakoparnig, Florian Markowetz, and Niko Beerenwinkel.
2015. “BitPhylogeny: A Probabilistic Framework for Reconstructing
Intra-Tumor Phylogenies.” <em>Genome Biology</em> 16 (1): 1–16.
<a href="https://doi.org/10.1186/s13059-015-0592-6" class="uri">https://doi.org/10.1186/s13059-015-0592-6</a>.
</div>
</div>
<div class="sourceCode" id="cb11"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h3 class="appendix" data-number="7" id="supplementary-materials"><span class="header-section-number">7</span> Supplementary materials</h3>
<p>Supplementary materials are available in addition to this article. It can be downloaded at
<a href="RJ-2025-042.zip">RJ-2025-042.zip</a></p>
<h3 class="appendix" data-number="8" id="cran-packages-used"><span class="header-section-number">8</span> CRAN packages used</h3>
<p><a href="https://cran.r-project.org/package=GeRnika">GeRnika</a>, <a href="https://cran.r-project.org/package=data.tree">data.tree</a></p>
<h3 class="appendix" data-number="9" id="cran-task-views-implied-by-cited-packages"><span class="header-section-number">9</span> CRAN Task Views implied by cited packages</h3>
<h3 class="appendix" data-number="10" id="bioconductor-packages-used"><span class="header-section-number">10</span> Bioconductor packages used</h3>
<p><a href="https://www.bioconductor.org/packages/clevRvis">clevRvis</a></p>
<h3 class="appendix" data-number="11" id="note"><span class="header-section-number">11</span> Note</h3>
<p>This article is converted from a Legacy LaTeX article using the
<a href="https://cran.r-project.org/package=texor">texor</a> package.
The pdf version is the official version. To report a problem with the html,
refer to CONTRIBUTE on the R Journal homepage.</p>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="references">References</h3>
<div id="references-listing"></div>
<h3 id="reuse">Reuse</h3>
<p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
<h3 id="citation">Citation</h3>
<p>For attribution, please cite this work as</p>
<pre class="citation-appendix short">Sánchez-Ferrera, et al., "GeRnika: An R Package for the Simulation, Visualization and Comparison of Tumor Phylogenies", The R Journal, 2026</pre>
<p>BibTeX citation</p>
<pre class="citation-appendix long">@article{RJ-2025-042,
  author = {Sánchez-Ferrera, Aitor and Tellaetxe-Abete, Maitena and Calvo-Molinos, Borja},
  title = {GeRnika: An R Package for the Simulation, Visualization and Comparison of Tumor Phylogenies},
  journal = {The R Journal},
  year = {2026},
  note = {https://doi.org/10.32614/RJ-2025-042},
  doi = {10.32614/RJ-2025-042},
  volume = {17},
  issue = {4},
  issn = {2073-4859},
  pages = {1}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
