---
title: News from the Bioconductor Project
abstract: The 'News from the Bioconductor Project' article from the 2011-1 issue.
author:
- name: Bioconductor Team
  affiliation: |
    Program in Computational Biology, Fred Hutchinson Cancer Research
    Center
date: '2011-06-01'
date_received: ~
journal:
  firstpage: 101
  lastpage: 101
volume: 3
issue: 1
slug: RJ-2011-1-bioconductor
citation_url: https://rjournal.github.io/
packages:
  cran: []
  bioc:
  - ArrayExpressHTS
  - mosaics
  - Rsubread
  - qrqc
  - seqbias
  - TEQC
  - clst
  - clstutils
  - inveRsion
  - chopsticks
  - snpStats
  - flowPhyto
  - flowPlots
  - IPPD
  - MSnbase
  - procoil
  - a4
  - ExiMiR
  - pvac
  - snm
  - TurboNorm
  - cn.farms
  - genoset
  - Vega
  - anota
  - Clonality
  - clusterProfiler
  - gaia
  - genefu
  - GSVA
  - ibh
  - joda
  - lol
  - mgsa
  - MLP
  - phenoDist
  - phenoTest
  - RNAinteract
  - survcomp
  - TDARACNE
  - AnnotationFuncs
  - NCIgraph
  - ENVISIONQuery
  - mcaGUI
  - GenomicFeatures
preview: preview.png
bibliography: ~
CTV: ~
output:
  rjtools::rjournal_web_article:
    self_contained: yes
    toc: no
    legacy_pdf: yes
    web_only: yes
    mathjax: https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js
    md_extension: -tex_math_single_backslash

---

::: article
We are pleased to announce Bioconductor 2.8, released on 14 April 2011.
Bioconductor 2.8 is compatible with R 2.13.0, and consists of 467
software packages and more than 500 up-to-date annotation packages.
There are 49 new software packages, and enhancements to many others.
Explore Bioconductor at <http://bioconductor.org>, and install packages
with

``` r
> source("http://bioconductor.org/biocLite.R")
> biocLite() # install standard packages...
> biocLite("IRanges") # ...or only IRanges
```

A Bioconductor machine instance for use with Amazon's cloud service is
available; see <http://bioconductor.org/help/bioconductor-cloud-ami>.

# New and revised packages

This release includes new packages for diverse areas of high-throughput
analysis. Highlights include:

Next-generation sequencing:

:   [*ArrayExpressHTS*](https://www.bioconductor.org/packages/release/bioc/html/ArrayExpressHTS.html)
    (RNA-seq),
    [*mosaics*](https://www.bioconductor.org/packages/release/bioc/html/mosaics.html)
    (ChIP-seq),
    [*Rsubread*](https://www.bioconductor.org/packages/release/bioc/html/Rsubread.html)
    (alignment),
    [*qrqc*](https://www.bioconductor.org/packages/release/bioc/html/qrqc.html),
    [*seqbias*](https://www.bioconductor.org/packages/release/bioc/html/seqbias.html),
    [*TEQC*](https://www.bioconductor.org/packages/release/bioc/html/TEQC.html)
    (quality assessment),
    [*clst*](https://www.bioconductor.org/packages/release/bioc/html/clst.html),
    [*clstutils*](https://www.bioconductor.org/packages/release/bioc/html/clstutils.html)
    (taxonomic placement).

SNPs and small variants:

:   [*inveRsion*](https://www.bioconductor.org/packages/release/bioc/html/inveRsion.html),
    [*chopsticks*](https://www.bioconductor.org/packages/release/bioc/html/chopsticks.html),
    [*snpStats*](https://www.bioconductor.org/packages/release/bioc/html/snpStats.html).

Flow cytometry and proteomics:

:   [*flowPhyto*](https://www.bioconductor.org/packages/release/bioc/html/flowPhyto.html),
    [*flowPlots*](https://www.bioconductor.org/packages/release/bioc/html/flowPlots.html).
    [*IPPD*](https://www.bioconductor.org/packages/release/bioc/html/IPPD.html),
    [*MSnbase*](https://www.bioconductor.org/packages/release/bioc/html/MSnbase.html),
    [*procoil*](https://www.bioconductor.org/packages/release/bioc/html/procoil.html).

Microarray:

:   The
    [*a4*](https://www.bioconductor.org/packages/release/bioc/html/a4.html)
    collection of packages,
    [*ExiMiR*](https://www.bioconductor.org/packages/release/bioc/html/ExiMiR.html),
    [*pvac*](https://www.bioconductor.org/packages/release/bioc/html/pvac.html),
    [*snm*](https://www.bioconductor.org/packages/release/bioc/html/snm.html),
    [*TurboNorm*](https://www.bioconductor.org/packages/release/bioc/html/TurboNorm.html).

Copy number variation:

:   [*cn.farms*](https://www.bioconductor.org/packages/release/bioc/html/cn.farms.html),
    [*genoset*](https://www.bioconductor.org/packages/release/bioc/html/genoset.html),
    [*Vega*](https://www.bioconductor.org/packages/release/bioc/html/Vega.html).

Advanced statistical implementations:

:   [*anota*](https://www.bioconductor.org/packages/release/bioc/html/anota.html),
    [*Clonality*](https://www.bioconductor.org/packages/release/bioc/html/Clonality.html),
    [*clusterProfiler*](https://www.bioconductor.org/packages/release/bioc/html/clusterProfiler.html),
    [*gaia*](https://www.bioconductor.org/packages/release/bioc/html/gaia.html),
    [*genefu*](https://www.bioconductor.org/packages/release/bioc/html/genefu.html),
    [*GSVA*](https://www.bioconductor.org/packages/release/bioc/html/GSVA.html),
    [*ibh*](https://www.bioconductor.org/packages/release/bioc/html/ibh.html),
    [*joda*](https://www.bioconductor.org/packages/release/bioc/html/joda.html),
    [*lol*](https://www.bioconductor.org/packages/release/bioc/html/lol.html),
    [*mgsa*](https://www.bioconductor.org/packages/release/bioc/html/mgsa.html),
    [*MLP*](https://www.bioconductor.org/packages/release/bioc/html/MLP.html),
    [*phenoDist*](https://www.bioconductor.org/packages/release/bioc/html/phenoDist.html),
    [*phenoTest*](https://www.bioconductor.org/packages/release/bioc/html/phenoTest.html),
    [*RNAinteract*](https://www.bioconductor.org/packages/release/bioc/html/RNAinteract.html),
    [*survcomp*](https://www.bioconductor.org/packages/release/bioc/html/survcomp.html),
    [*TDARACNE*](https://www.bioconductor.org/packages/release/bioc/html/TDARACNE.html).

Annotation and visualization:

:   [*AnnotationFuncs*](https://www.bioconductor.org/packages/release/bioc/html/AnnotationFuncs.html),
    [*NCIgraph*](https://www.bioconductor.org/packages/release/bioc/html/NCIgraph.html),
    [*ENVISIONQuery*](https://www.bioconductor.org/packages/release/bioc/html/ENVISIONQuery.html),
    [*mcaGUI*](https://www.bioconductor.org/packages/release/bioc/html/mcaGUI.html).

Our large collection of microarray- and organism-specific annotation
packages have, as usual, been updated to include current information.
Developments in the
[*GenomicFeatures*](https://www.bioconductor.org/packages/release/bioc/html/GenomicFeatures.html)
package allow users to retrieve and easily save annotation tracks found
at the UCSC genome browser. This provides easy access to diverse
information about, e.g., binding factor locations or nucleosome
positions, and complements known gene annotations already available as
*TranscriptDb* objects.

Further information on new and existing packages can be found on the
Bioconductor web site; 'BiocViews' identify coherent groups of packages,
with links to package descriptions, vignettes, reference manuals, and
use statistics.

# Other activities

The Bioconductor community meets on July 27-29 at our annual conference
(<https://secure.bioconductor.org/BioC2011>) in Seattle for a
combination of scientific talks and hands-on tutorials; a developer
meeting in the late fall will highlight contributions from the European
developer community. The active Bioconductor mailing lists
(<http://bioconductor.org/help/mailing-list/>) connect users with each
other, to domain experts, and to maintainers eager to ensure that their
packages satisfy the needs of leading edge approaches. Bioconductor
package maintainers and the Bioconductor team invest considerable effort
in producing high-quality software. The Bioconductor team continues to
ensure quality software through technical and scientific reviews of new
packages, and daily builds of released packages on Linux, Windows, and
MacOS platforms.

# Looking forward

Contributions from the Bioconductor community shape each release; we are
seeing a tremendous number of new contributed packages addressing
sequence, advanced microarray, and other diverse areas of
high-throughput genomic analysis.

Sequence analysis continues to pose significant statistical and
computational challenges, although remarkable progress is being made
both in the Bioconductor community and more generally. The next release
will see contributions enabling common data management (e.g.,
efficiently querying very large sequence alignments), representation
(e.g., of gapped alignments and small variants), and work flow (e.g.,
RNA-seq read counts and estimation) tasks.

Important developments in the annotation domain include efforts to
enhance reproducibility by providing transcript annotation packages, and
the ability to easily generate packages for non-model organisms. New
efforts are also under way to better use existing and new annotation
resources to help users with common tasks in next-generation sequence
work flows, for instance immediately identifying whether single
nucleotide and other small variants disrupt protein production.
:::
