<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
      margin-bottom: 0em;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
<title>ebmstate: An R Package For Disease Progression Analysis Under Empirical Bayes Cox Models</title>

<meta property="description" itemprop="description" content="The new R package ebmstate is a package for multi-state survival&#10;analysis. It is suitable for high-dimensional data and allows point&#10;and interval estimation of relative transition hazards, cumulative&#10;transition hazards and state occupation probabilities, under&#10;clock-forward and clock-reset Cox models. Our package extends the&#10;package mstate in a threefold manner: it transforms the Cox regression&#10;model into an empirical Bayes model that can handle high-dimensional&#10;data; it introduces an analytical, Fourier transform-based estimator&#10;of state occupation probabilities for clock-reset models that is much&#10;faster than the corresponding, simulation-based estimator in mstate;&#10;and it replaces asymptotic confidence intervals meant for the&#10;low-dimensional setting by non-parametric bootstrap confidence&#10;intervals. Our package supports multi-state models of arbitrary&#10;structure, but the estimators of state occupation probabilities are&#10;valid for transition structures without cycles only. Once the input&#10;data is in the required format, estimation is handled automatically.&#10;The present paper includes a tutorial on how to use ebmstate to&#10;estimate transition hazards and state occupation probabilities, as&#10;well as a simulation study showing how it outperforms mstate in&#10;higher-dimensional settings."/>

<link rel="canonical" href="https://doi.org/10.32614/RJ-2024-002/"/>
<link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>
<link rel="icon" type="image/vnd.microsoft.icon" href="../../resources/favicon.ico"/>

<!--  https://schema.org/Article -->
<meta property="article:published" itemprop="datePublished" content="2025-01-11"/>
<meta property="article:created" itemprop="dateCreated" content="2025-01-11"/>
<meta name="article:author" content="Rui J. Costa"/>
<meta name="article:author" content="Moritz Gerstung"/>

<!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
<meta property="og:title" content="ebmstate: An R Package For Disease Progression Analysis Under Empirical Bayes Cox Models"/>
<meta property="og:type" content="article"/>
<meta property="og:description" content="The new R package ebmstate is a package for multi-state survival&#10;analysis. It is suitable for high-dimensional data and allows point&#10;and interval estimation of relative transition hazards, cumulative&#10;transition hazards and state occupation probabilities, under&#10;clock-forward and clock-reset Cox models. Our package extends the&#10;package mstate in a threefold manner: it transforms the Cox regression&#10;model into an empirical Bayes model that can handle high-dimensional&#10;data; it introduces an analytical, Fourier transform-based estimator&#10;of state occupation probabilities for clock-reset models that is much&#10;faster than the corresponding, simulation-based estimator in mstate;&#10;and it replaces asymptotic confidence intervals meant for the&#10;low-dimensional setting by non-parametric bootstrap confidence&#10;intervals. Our package supports multi-state models of arbitrary&#10;structure, but the estimators of state occupation probabilities are&#10;valid for transition structures without cycles only. Once the input&#10;data is in the required format, estimation is handled automatically.&#10;The present paper includes a tutorial on how to use ebmstate to&#10;estimate transition hazards and state occupation probabilities, as&#10;well as a simulation study showing how it outperforms mstate in&#10;higher-dimensional settings."/>
<meta property="og:url" content="https://doi.org/10.32614/RJ-2024-002/"/>
<meta property="og:image" content="https://journal.r-project.org/articles/RJ-2024-002/figures/package_summary_figure.png"/>
<meta property="og:image:width" content="1943"/>
<meta property="og:image:height" content="1130"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="The R Journal"/>

<!--  https://dev.twitter.com/cards/types/summary -->
<meta property="twitter:card" content="summary_large_image"/>
<meta property="twitter:title" content="ebmstate: An R Package For Disease Progression Analysis Under Empirical Bayes Cox Models"/>
<meta property="twitter:description" content="The new R package ebmstate is a package for multi-state survival&#10;analysis. It is suitable for high-dimensional data and allows point&#10;and interval estimation of relative transition hazards, cumulative&#10;transition hazards and state occupation probabilities, under&#10;clock-forward and clock-reset Cox models. Our package extends the&#10;package mstate in a threefold manner: it transforms the Cox regression&#10;model into an empirical Bayes model that can handle high-dimensional&#10;data; it introduces an analytical, Fourier transform-based estimator&#10;of state occupation probabilities for clock-reset models that is much&#10;faster than the corresponding, simulation-based estimator in mstate;&#10;and it replaces asymptotic confidence intervals meant for the&#10;low-dimensional setting by non-parametric bootstrap confidence&#10;intervals. Our package supports multi-state models of arbitrary&#10;structure, but the estimators of state occupation probabilities are&#10;valid for transition structures without cycles only. Once the input&#10;data is in the required format, estimation is handled automatically.&#10;The present paper includes a tutorial on how to use ebmstate to&#10;estimate transition hazards and state occupation probabilities, as&#10;well as a simulation study showing how it outperforms mstate in&#10;higher-dimensional settings."/>
<meta property="twitter:url" content="https://doi.org/10.32614/RJ-2024-002/"/>
<meta property="twitter:image" content="https://journal.r-project.org/articles/RJ-2024-002/figures/package_summary_figure.png"/>
<meta property="twitter:image:width" content="1943"/>
<meta property="twitter:image:height" content="1130"/>

<!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
<meta name="citation_title" content="ebmstate: An R Package For Disease Progression Analysis Under Empirical Bayes Cox Models"/>
<meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2024-002"/>
<meta name="citation_pdf_url" content="RJ-2024-002.pdf"/>
<meta name="citation_volume" content="16"/>
<meta name="citation_issue" content="1"/>
<meta name="citation_doi" content="10.32614/RJ-2024-002"/>
<meta name="citation_journal_title" content="The R Journal"/>
<meta name="citation_issn" content="2073-4859"/>
<meta name="citation_firstpage" content="15"/>
<meta name="citation_lastpage" content="38"/>
<meta name="citation_fulltext_world_readable" content=""/>
<meta name="citation_online_date" content="2025/01/11"/>
<meta name="citation_publication_date" content="2025/01/11"/>
<meta name="citation_author" content="Rui J. Costa"/>
<meta name="citation_author_institution" content="European Molecular Biology Laboratory"/>
<meta name="citation_author" content="Moritz Gerstung"/>
<meta name="citation_author_institution" content="aff. 1: European Molecular Biology Laboratory"/>
<!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=A linear regression model for the analysis of life times;citation_publisher=Wiley Online Library;citation_volume=8;citation_author=Odd O Aalen"/>
  <meta name="citation_reference" content="citation_title=Survival and event history analysis;citation_publisher=Springer;citation_author=O Aalen;citation_author=O Borgan;citation_author=H Gjessing"/>
  <meta name="citation_reference" content="citation_title=Statistical models based on counting processes;citation_publisher=Springer;citation_author=PK Andersen;citation_author=O Borgan;citation_author=RD Gill;citation_author=N Keiding"/>
  <meta name="citation_reference" content="citation_title=Competing risks and time-dependent covariates;citation_publisher=Wiley Online Library;citation_volume=52;citation_author=Giuliana Cortese;citation_author=Per K Andersen"/>
  <meta name="citation_reference" content="citation_title=Bayesian methods for data analysis;citation_publisher=CRC Press;citation_author=BP Carlin;citation_author=TA Louis"/>
  <meta name="citation_reference" content="citation_title=flexsurv: A platform for parametric survival modeling in R;citation_volume=70;citation_doi=10.18637/jss.v070.i08;citation_author=Christopher Jackson"/>
  <meta name="citation_reference" content="citation_title=gamboostMSM;citation_author=Holger Reulen"/>
  <meta name="citation_reference" content="citation_title=Bayesian data analysis;citation_publisher=CRC Press;citation_author=A Gelman;citation_author=JB Carlin;citation_author=HS Stern;citation_author=DB Dunson;citation_author=A Vehtari;citation_author=DB Rubin"/>
  <meta name="citation_reference" content="citation_title=Evaluating the Yield of Medical Tests;citation_volume=247;citation_doi=10.1001/jama.1982.03320430047030;citation_issn=0098-7484;citation_author=Jr Harrell;citation_author=Robert M. Califf;citation_author=David B. Pryor;citation_author=Kerry L. Lee;citation_author=Robert A. Rosati"/>
  <meta name="citation_reference" content="citation_title=The elements of statistical learning: Data mining, inference, and prediction;citation_publisher=Springer;citation_volume=2;citation_author=Trevor Hastie;citation_author=Robert Tibshirani;citation_author=Jerome H Friedman;citation_author=Jerome H Friedman"/>
  <meta name="citation_reference" content="citation_title=Multi-state models: A review;citation_publisher=Springer;citation_volume=5;citation_author=Philip Hougaard"/>
  <meta name="citation_reference" content="citation_title=Multi-state models for panel data: The msm package for R;citation_volume=38;citation_author=Christopher H. Jackson"/>
  <meta name="citation_reference" content="citation_title=SemiMarkov: An R Package for Parametric Estimation in Multi-State Semi-Markov Models;citation_publisher=University of California, Los Angeles;citation_volume=66;citation_doi=10.18637/jss.v066.i06;citation_author=Agnieszka Listwon;citation_author=Philippe Saint-Pierre"/>
  <meta name="citation_reference" content="citation_title=Mboost: Model-based boosting;citation_author=Torsten Hothorn;citation_author=Peter Buehlmann;citation_author=Thomas Kneib;citation_author=Matthias Schmid;citation_author=Benjamin Hofner"/>
  <meta name="citation_reference" content="citation_title=Clinical and biological implications of driver mutations in myelodysplastic syndromes;citation_publisher=Am Soc Hematology;citation_volume=122;citation_author=Elli Papaemmanuil;citation_author=Moritz Gerstung;citation_author=Luca Malcovati;citation_author=Sudhir Tauro;citation_author=Gunes Gundem;citation_author=Peter Van Loo;citation_author=Chris J Yoon;citation_author=Peter Ellis;citation_author=David C Wedge;citation_author=Andrea Pellagatti;citation_author=Clinical and biological implications of driver mutations in myelodysplastic syndromes"/>
  <meta name="citation_reference" content="citation_title=penMSM;citation_author=Holger Reulen"/>
  <meta name="citation_reference" content="citation_title=Cox models with dynamic ridge penalties on time-varying effects of the covariates;citation_publisher=Wiley Online Library;citation_volume=33;citation_author=Aris Perperoglou"/>
  <meta name="citation_reference" content="citation_title=mstate: An R package for the analysis of competing risks and multi-state models;citation_volume=38;citation_author=Liesbeth C. Wreede;citation_author=Marta Fiocco;citation_author=Hein Putter"/>
  <meta name="citation_reference" content="citation_title=Tutorial in biostatistics: Competing risks and multi-state models;citation_publisher=Wiley Online Library;citation_volume=26;citation_author=Hein Putter;citation_author=Marta Fiocco;citation_author=Ronald B Geskus"/>
  <meta name="citation_reference" content="citation_title=Tutorial in biostatistics: Competing risks and multi-state models analyses using the mstate package;citation_author=Hein Putter"/>
  <meta name="citation_reference" content="citation_title=Steinâ€™s paradox;citation_publisher=The Archimedeans;citation_volume=62;citation_author=Richard J Samworth"/>
  <meta name="citation_reference" content="citation_title=Estimation in generalized linear models with random effects;citation_volume=78;citation_doi=10.1093/biomet/78.4.719;citation_author=Robert Schall"/>
  <meta name="citation_reference" content="citation_title=Estimating the dimension of a model;citation_publisher=JSTOR;citation_author=Gideon Schwarz"/>
  <meta name="citation_reference" content="citation_title=A package for survival analysis in s;citation_author=Terry M Therneau"/>
  <meta name="citation_reference" content="citation_title=Estimation and asymptotic theory for transition probabilities in markov renewal multi-state models;citation_volume=8;citation_doi=doi:10.1515/1557-4679.1375;citation_author=Cristian Spitoni;citation_author=Marion Verduijn;citation_author=Hein Putter"/>
  <meta name="citation_reference" content="citation_title=The mstate package for estimation and prediction in non- and semi-parametric multi-state and competing risks models;citation_volume=99;citation_doi=https://doi.org/10.1016/j.cmpb.2010.01.001;citation_issn=0169-2607;citation_author=Liesbeth C. Wreede;citation_author=Marta Fiocco;citation_author=Hein Putter"/>
  <meta name="citation_reference" content="citation_title=Regularization and variable selection via the elastic net;citation_volume=67;citation_doi=https://doi.org/10.1111/j.1467-9868.2005.00503.x;citation_author=Hui Zou;citation_author=Trevor Hastie"/>
  <!--radix_placeholder_rmarkdown_metadata-->

<script type="text/json" id="radix-rmarkdown-metadata">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","date_received","journal","volume","issue","slug","citation_url","packages","preview","bibliography","CTV","legacy_pdf","legacy_converted","output","draft","pdf_url","doi","creative_commons","csl","canonical_url"]}},"value":[{"type":"character","attributes":{},"value":["ebmstate: An R Package For Disease Progression Analysis Under Empirical Bayes Cox Models"]},{"type":"character","attributes":{},"value":["The new R package ebmstate is a package for multi-state survival\nanalysis. It is suitable for high-dimensional data and allows point\nand interval estimation of relative transition hazards, cumulative\ntransition hazards and state occupation probabilities, under\nclock-forward and clock-reset Cox models. Our package extends the\npackage mstate in a threefold manner: it transforms the Cox regression\nmodel into an empirical Bayes model that can handle high-dimensional\ndata; it introduces an analytical, Fourier transform-based estimator\nof state occupation probabilities for clock-reset models that is much\nfaster than the corresponding, simulation-based estimator in mstate;\nand it replaces asymptotic confidence intervals meant for the\nlow-dimensional setting by non-parametric bootstrap confidence\nintervals. Our package supports multi-state models of arbitrary\nstructure, but the estimators of state occupation probabilities are\nvalid for transition structures without cycles only. Once the input\ndata is in the required format, estimation is handled automatically.\nThe present paper includes a tutorial on how to use ebmstate to\nestimate transition hazards and state occupation probabilities, as\nwell as a simulation study showing how it outperforms mstate in\nhigher-dimensional settings."]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Rui J. Costa"]},{"type":"character","attributes":{},"value":["European Molecular Biology Laboratory"]},{"type":"character","attributes":{},"value":["European Bioinformatics Institute (EMBL-EBI)","Hinxton, CB10 1SD","United Kingdom","[ruibarrigana@hotmail.com](ruibarrigana@hotmail.com){.uri}\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Moritz Gerstung"]},{"type":"character","attributes":{},"value":["aff. 1: European Molecular Biology Laboratory"]},{"type":"character","attributes":{},"value":["European Bioinformatics Institute (EMBL-EBI)","Hinxton, CB10 1SD","United Kindom","aff. 2: German Cancer Research Center (DKFZ)","Im Neuenheimer Feld 280","69120 Heidelberg","Germany","[moritz.gerstung@dkfz.de](moritz.gerstung@dkfz.de){.uri}\n"]}]}]},{"type":"character","attributes":{},"value":["2025-01-11"]},{"type":"character","attributes":{},"value":["2022-06-27"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"integer","attributes":{},"value":[15]},{"type":"integer","attributes":{},"value":[38]}]},{"type":"integer","attributes":{},"value":[16]},{"type":"integer","attributes":{},"value":[1]},{"type":"character","attributes":{},"value":["RJ-2024-002"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2024-002"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"character","attributes":{},"value":["msm","SemiMarkov","survival","mstate","mboost","gamboostMSM","penMSM","ebmstate"]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["preview.png"]},{"type":"character","attributes":{},"value":["costa-gerstung.bib"]},{"type":"character","attributes":{},"value":["ClinicalTrials","Distributions","Econometrics","Epidemiology","MachineLearning","Survival"]},{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[true]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc","mathjax","md_extension"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js"]},{"type":"character","attributes":{},"value":["-tex_math_single_backslash"]}]}]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["RJ-2024-002.pdf"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2024-002"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"character","attributes":{},"value":["/home/mitchell/R/x86_64-pc-linux-gnu-library/4.4/rjtools/rjournal.csl"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2024-002/"]}]}
</script>
<!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["costa-gerstung.bib","costa-gerstung.tex","data/mds.paper.clin.txt","data/MDS.TPD.20Nov2012.csv","figures/coef_plots.pdf","figures/coef_plots.png","figures/data_summary_figs2.pdf","figures/data_summary_figs2.png","figures/estimator_performance_boxplots_1000patients.pdf","figures/estimator_performance_boxplots_1000patients.png","figures/estimator_performance_boxplots_100patients.pdf","figures/estimator_performance_boxplots_100patients.png","figures/mssample_and_probtrans_fft.pdf","figures/mssample_and_probtrans_fft.png","figures/na_props_1000patients_coxph.pdf","figures/na_props_1000patients_coxph.png","figures/na_props_100patients_coxph.pdf","figures/na_props_100patients_coxph.png","figures/package_summary_figure.pdf","figures/package_summary_figure.png","figures/patient78_cumhaz_final.png","figures/patient78_transProbs_final.png","figures/Rlogo-5.png","figures/transition_structures.pdf","figures/transition_structures.png","figures/workflow0.pdf","figures/workflow0.png","RJ-2024-002_files/anchor-4.2.2/anchor.min.js","RJ-2024-002_files/bowser-1.9.3/bowser.min.js","RJ-2024-002_files/distill-2.2.21/template.v2.js","RJ-2024-002_files/header-attrs-2.29/header-attrs.js","RJ-2024-002_files/jquery-3.6.0/jquery-3.6.0.js","RJ-2024-002_files/jquery-3.6.0/jquery-3.6.0.min.js","RJ-2024-002_files/jquery-3.6.0/jquery-3.6.0.min.map","RJ-2024-002_files/popper-2.6.0/popper.min.js","RJ-2024-002_files/tippy-6.2.7/tippy-bundle.umd.min.js","RJ-2024-002_files/tippy-6.2.7/tippy-light-border.css","RJ-2024-002_files/tippy-6.2.7/tippy.css","RJ-2024-002_files/tippy-6.2.7/tippy.umd.min.js","RJ-2024-002_files/webcomponents-2.0.0/webcomponents.js","RJ-2024-002.zip","RJournal.sty","RJwrapper.log","RJwrapper.tex","Rlogo-5.png","scripts/ESM_1.html","scripts/ESM_1.Rmd","scripts/ESM_2.html","scripts/ESM_2.Rmd","scripts/ESM_3.R","web/costa-gerstung.bib","web/costa-gerstung.R","web/figures/coef_plots.png","web/figures/data_summary_figs2.png","web/figures/estimator_performance_boxplots_1000patients.png","web/figures/estimator_performance_boxplots_100patients.png","web/figures/mssample_and_probtrans_fft.png","web/figures/na_props_1000patients_coxph.png","web/figures/na_props_100patients_coxph.png","web/figures/package_summary_figure.png","web/figures/patient78_cumhaz_final.png","web/figures/patient78_transProbs_final.png","web/figures/transition_structures.png","web/figures/workflow0.png","web/RJwrapper.Rmd"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
<meta name="distill:offset" content="../.."/>

<script type="application/javascript">

  window.headroom_prevent_pin = false;

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // initialize headroom for banner
    var header = $('header').get(0);
    var headerHeight = header.offsetHeight;
    var headroom = new Headroom(header, {
      tolerance: 5,
      onPin : function() {
        if (window.headroom_prevent_pin) {
          window.headroom_prevent_pin = false;
          headroom.unpin();
        }
      }
    });
    headroom.init();
    if(window.location.hash)
      headroom.unpin();
    $(header).addClass('headroom--transition');

    // offset scroll location for banner on hash change
    // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
    window.addEventListener("hashchange", function(event) {
      window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
    });

    // responsive menu
    $('.distill-site-header').each(function(i, val) {
      var topnav = $(this);
      var toggle = topnav.find('.nav-toggle');
      toggle.on('click', function() {
        topnav.toggleClass('responsive');
      });
    });

    // nav dropdowns
    $('.nav-dropbtn').click(function(e) {
      $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
      $(this).parent().siblings('.nav-dropdown')
         .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $("body").click(function(e){
      $('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $(".nav-dropdown").click(function(e){
      e.stopPropagation();
    });
  });
</script>

<style type="text/css">

/* Theme (user-documented overrideables for nav appearance) */

.distill-site-nav {
  color: rgba(255, 255, 255, 0.8);
  background-color: #0F2E3D;
  font-size: 15px;
  font-weight: 300;
}

.distill-site-nav a {
  color: inherit;
  text-decoration: none;
}

.distill-site-nav a:hover {
  color: white;
}

@media print {
  .distill-site-nav {
    display: none;
  }
}

.distill-site-header {

}

.distill-site-footer {

}


/* Site Header */

.distill-site-header {
  width: 100%;
  box-sizing: border-box;
  z-index: 3;
}

.distill-site-header .nav-left {
  display: inline-block;
  margin-left: 8px;
}

@media screen and (max-width: 768px) {
  .distill-site-header .nav-left {
    margin-left: 0;
  }
}


.distill-site-header .nav-right {
  float: right;
  margin-right: 8px;
}

.distill-site-header a,
.distill-site-header .title {
  display: inline-block;
  text-align: center;
  padding: 14px 10px 14px 10px;
}

.distill-site-header .title {
  font-size: 18px;
  min-width: 150px;
}

.distill-site-header .logo {
  padding: 0;
}

.distill-site-header .logo img {
  display: none;
  max-height: 20px;
  width: auto;
  margin-bottom: -4px;
}

.distill-site-header .nav-image img {
  max-height: 18px;
  width: auto;
  display: inline-block;
  margin-bottom: -3px;
}



@media screen and (min-width: 1000px) {
  .distill-site-header .logo img {
    display: inline-block;
  }
  .distill-site-header .nav-left {
    margin-left: 20px;
  }
  .distill-site-header .nav-right {
    margin-right: 20px;
  }
  .distill-site-header .title {
    padding-left: 12px;
  }
}


.distill-site-header .nav-toggle {
  display: none;
}

.nav-dropdown {
  display: inline-block;
  position: relative;
}

.nav-dropdown .nav-dropbtn {
  border: none;
  outline: none;
  color: rgba(255, 255, 255, 0.8);
  padding: 16px 10px;
  background-color: transparent;
  font-family: inherit;
  font-size: inherit;
  font-weight: inherit;
  margin: 0;
  margin-top: 1px;
  z-index: 2;
}

.nav-dropdown-content {
  display: none;
  position: absolute;
  background-color: white;
  min-width: 200px;
  border: 1px solid rgba(0,0,0,0.15);
  border-radius: 4px;
  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
  z-index: 1;
  margin-top: 2px;
  white-space: nowrap;
  padding-top: 4px;
  padding-bottom: 4px;
}

.nav-dropdown-content hr {
  margin-top: 4px;
  margin-bottom: 4px;
  border: none;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.nav-dropdown-active {
  display: block;
}

.nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
  color: black;
  padding: 6px 24px;
  text-decoration: none;
  display: block;
  text-align: left;
}

.nav-dropdown-content .nav-dropdown-header {
  display: block;
  padding: 5px 24px;
  padding-bottom: 0;
  text-transform: uppercase;
  font-size: 14px;
  color: #999999;
  white-space: nowrap;
}

.nav-dropdown:hover .nav-dropbtn {
  color: white;
}

.nav-dropdown-content a:hover {
  background-color: #ddd;
  color: black;
}

.nav-right .nav-dropdown-content {
  margin-left: -45%;
  right: 0;
}

@media screen and (max-width: 768px) {
  .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
  .distill-site-header a.nav-toggle {
    float: right;
    display: block;
  }
  .distill-site-header .title {
    margin-left: 0;
  }
  .distill-site-header .nav-right {
    margin-right: 0;
  }
  .distill-site-header {
    overflow: hidden;
  }
  .nav-right .nav-dropdown-content {
    margin-left: 0;
  }
}


@media screen and (max-width: 768px) {
  .distill-site-header.responsive {position: relative; min-height: 500px; }
  .distill-site-header.responsive a.nav-toggle {
    position: absolute;
    right: 0;
    top: 0;
  }
  .distill-site-header.responsive a,
  .distill-site-header.responsive .nav-dropdown {
    display: block;
    text-align: left;
  }
  .distill-site-header.responsive .nav-left,
  .distill-site-header.responsive .nav-right {
    width: 100%;
  }
  .distill-site-header.responsive .nav-dropdown {float: none;}
  .distill-site-header.responsive .nav-dropdown-content {position: relative;}
  .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
    display: block;
    width: 100%;
    text-align: left;
  }
}

/* Site Footer */

.distill-site-footer {
  width: 100%;
  overflow: hidden;
  box-sizing: border-box;
  z-index: 3;
  margin-top: 30px;
  padding-top: 30px;
  padding-bottom: 30px;
  text-align: center;
}

/* Headroom */

d-title {
  padding-top: 6rem;
}

@media print {
  d-title {
    padding-top: 4rem;
  }
}

.headroom {
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
}

.headroom--transition {
  transition: all .4s ease-in-out;
}

.headroom--unpinned {
  top: -100px;
}

.headroom--pinned {
  top: 0;
}

/* adjust viewport for navbar height */
/* helps vertically center bootstrap (non-distill) content */
.min-vh-100 {
  min-height: calc(100vh - 100px) !important;
}

</style>

<script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="../../site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"/>
<link href="../../site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"/>
<script src="../../site_libs/headroom-0.9.4/headroom.min.js"></script>
<!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

<style type="text/css">

body {
  background-color: white;
}

.pandoc-table {
  width: 100%;
}

.pandoc-table>caption {
  margin-bottom: 10px;
}

.pandoc-table th:not([align]) {
  text-align: left;
}

.pagedtable-footer {
  font-size: 15px;
}

d-byline .byline {
  grid-template-columns: 2fr 2fr;
}

d-byline .byline h3 {
  margin-block-start: 1.5em;
}

d-byline .byline .authors-affiliations h3 {
  margin-block-start: 0.5em;
}

.authors-affiliations .orcid-id {
  width: 16px;
  height:16px;
  margin-left: 4px;
  margin-right: 4px;
  vertical-align: middle;
  padding-bottom: 2px;
}

d-title .dt-tags {
  margin-top: 1em;
  grid-column: text;
}

.dt-tags .dt-tag {
  text-decoration: none;
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0em 0.4em;
  margin-right: 0.5em;
  margin-bottom: 0.4em;
  font-size: 70%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

d-article table.gt_table td,
d-article table.gt_table th {
  border-bottom: none;
  font-size: 100%;
}

.html-widget {
  margin-bottom: 2.0em;
}

.l-screen-inset {
  padding-right: 16px;
}

.l-screen .caption {
  margin-left: 10px;
}

.shaded {
  background: rgb(247, 247, 247);
  padding-top: 20px;
  padding-bottom: 20px;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .html-widget {
  margin-bottom: 0;
  border: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .shaded-content {
  background: white;
}

.text-output {
  margin-top: 0;
  line-height: 1.5em;
}

.hidden {
  display: none !important;
}

hr.section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  margin: 0px;
}


d-byline {
  border-top: none;
}

d-article {
  padding-top: 2.5rem;
  padding-bottom: 30px;
  border-top: none;
}

d-appendix {
  padding-top: 30px;
}

d-article>p>img {
  width: 100%;
}

d-article h2 {
  margin: 1rem 0 1.5rem 0;
}

d-article h3 {
  margin-top: 1.5rem;
}

d-article iframe {
  border: 1px solid rgba(0, 0, 0, 0.1);
  margin-bottom: 2.0em;
  width: 100%;
}

/* Tweak code blocks */

d-article div.sourceCode code,
d-article pre code {
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
}

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: auto;
}

d-article div.sourceCode {
  background-color: white;
}

d-article div.sourceCode pre {
  padding-left: 10px;
  font-size: 12px;
  border-left: 2px solid rgba(0,0,0,0.1);
}

d-article pre {
  font-size: 12px;
  color: black;
  background: none;
  margin-top: 0;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

d-article pre a {
  border-bottom: none;
}

d-article pre a:hover {
  border-bottom: none;
  text-decoration: underline;
}

d-article details {
  grid-column: text;
  margin-bottom: 0.8em;
}

@media(min-width: 768px) {

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: visible !important;
}

d-article div.sourceCode pre {
  padding-left: 18px;
  font-size: 14px;
}

/* tweak for Pandoc numbered line within distill */
d-article pre.numberSource code > span {
    left: -2em;
}

d-article pre {
  font-size: 14px;
}

}

figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

/* CSS for d-contents */

.d-contents {
  grid-column: text;
  color: rgba(0,0,0,0.8);
  font-size: 0.9em;
  padding-bottom: 1em;
  margin-bottom: 1em;
  padding-bottom: 0.5em;
  margin-bottom: 1em;
  padding-left: 0.25em;
  justify-self: start;
}

@media(min-width: 1000px) {
  .d-contents.d-contents-float {
    height: 0;
    grid-column-start: 1;
    grid-column-end: 4;
    justify-self: center;
    padding-right: 3em;
    padding-left: 2em;
  }
}

.d-contents nav h3 {
  font-size: 18px;
  margin-top: 0;
  margin-bottom: 1em;
}

.d-contents li {
  list-style-type: none
}

.d-contents nav > ul {
  padding-left: 0;
}

.d-contents ul {
  padding-left: 1em
}

.d-contents nav ul li {
  margin-top: 0.6em;
  margin-bottom: 0.2em;
}

.d-contents nav a {
  font-size: 13px;
  border-bottom: none;
  text-decoration: none
  color: rgba(0, 0, 0, 0.8);
}

.d-contents nav a:hover {
  text-decoration: underline solid rgba(0, 0, 0, 0.6)
}

.d-contents nav > ul > li > a {
  font-weight: 600;
}

.d-contents nav > ul > li > ul {
  font-weight: inherit;
}

.d-contents nav > ul > li > ul > li {
  margin-top: 0.2em;
}


.d-contents nav ul {
  margin-top: 0;
  margin-bottom: 0.25em;
}

.d-article-with-toc h2:nth-child(2) {
  margin-top: 0;
}


/* Figure */

.figure {
  position: relative;
  margin-bottom: 2.5em;
  margin-top: 1.5em;
}

.figure .caption {
  color: rgba(0, 0, 0, 0.6);
  font-size: 12px;
  line-height: 1.5em;
}

.figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

.figure .caption a {
  color: rgba(0, 0, 0, 0.6);
}

.figure .caption b,
.figure .caption strong, {
  font-weight: 600;
  color: rgba(0, 0, 0, 1.0);
}

/* Citations */

d-article .citation {
  color: inherit;
  cursor: inherit;
}

div.hanging-indent{
  margin-left: 1em; text-indent: -1em;
}

/* Citation hover box */

.tippy-box[data-theme~=light-border] {
  background-color: rgba(250, 250, 250, 0.95);
}

.tippy-content > p {
  margin-bottom: 0;
  padding: 2px;
}


/* Tweak 1000px media break to show more text */

@media(min-width: 1000px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 16px;
  }

  .grid {
    grid-column-gap: 16px;
  }

  d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
  }
  figure .caption, .figure .caption, figure figcaption {
    font-size: 13px;
  }
}

@media(min-width: 1180px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 32px;
  }

  .grid {
    grid-column-gap: 32px;
  }
}


/* Get the citation styles for the appendix (not auto-injected on render since
   we do our own rendering of the citation appendix) */

d-appendix .citation-appendix,
.d-appendix .citation-appendix {
  font-size: 11px;
  line-height: 15px;
  border-left: 1px solid rgba(0, 0, 0, 0.1);
  padding-left: 18px;
  border: 1px solid rgba(0,0,0,0.1);
  background: rgba(0, 0, 0, 0.02);
  padding: 10px 18px;
  border-radius: 3px;
  color: rgba(150, 150, 150, 1);
  overflow: hidden;
  margin-top: -12px;
  white-space: pre-wrap;
  word-wrap: break-word;
}

/* Include appendix styles here so they can be overridden */

d-appendix {
  contain: layout style;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-top: 60px;
  margin-bottom: 0;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  color: rgba(0,0,0,0.5);
  padding-top: 60px;
  padding-bottom: 48px;
}

d-appendix h3 {
  grid-column: page-start / text-start;
  font-size: 15px;
  font-weight: 500;
  margin-top: 1em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.65);
}

d-appendix h3 + * {
  margin-top: 1em;
}

d-appendix ol {
  padding: 0 0 0 15px;
}

@media (min-width: 768px) {
  d-appendix ol {
    padding: 0 0 0 30px;
    margin-left: -30px;
  }
}

d-appendix li {
  margin-bottom: 1em;
}

d-appendix a {
  color: rgba(0, 0, 0, 0.6);
}

d-appendix > * {
  grid-column: text;
}

d-appendix > d-footnote-list,
d-appendix > d-citation-list,
d-appendix > distill-appendix {
  grid-column: screen;
}

/* Include footnote styles here so they can be overridden */

d-footnote-list {
  contain: layout style;
}

d-footnote-list > * {
  grid-column: text;
}

d-footnote-list a.footnote-backlink {
  color: rgba(0,0,0,0.3);
  padding-left: 0.5em;
}



/* Anchor.js */

.anchorjs-link {
  /*transition: all .25s linear; */
  text-decoration: none;
  border-bottom: none;
}
*:hover > .anchorjs-link {
  margin-left: -1.125em !important;
  text-decoration: none;
  border-bottom: none;
}

/* Social footer */

.social_footer {
  margin-top: 30px;
  margin-bottom: 0;
  color: rgba(0,0,0,0.67);
}

.disqus-comments {
  margin-right: 30px;
}

.disqus-comment-count {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  cursor: pointer;
}

#disqus_thread {
  margin-top: 30px;
}

.article-sharing a {
  border-bottom: none;
  margin-right: 8px;
}

.article-sharing a:hover {
  border-bottom: none;
}

.sidebar-section.subscribe {
  font-size: 12px;
  line-height: 1.6em;
}

.subscribe p {
  margin-bottom: 0.5em;
}


.article-footer .subscribe {
  font-size: 15px;
  margin-top: 45px;
}


.sidebar-section.custom {
  font-size: 12px;
  line-height: 1.6em;
}

.custom p {
  margin-bottom: 0.5em;
}

/* Styles for listing layout (hide title) */
.layout-listing d-title, .layout-listing .d-title {
  display: none;
}

/* Styles for posts lists (not auto-injected) */


.posts-with-sidebar {
  padding-left: 45px;
  padding-right: 45px;
}

.posts-list .description h2,
.posts-list .description p {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

.posts-list .description h2 {
  font-weight: 700;
  border-bottom: none;
  padding-bottom: 0;
}

.posts-list h2.post-tag {
  border-bottom: 1px solid rgba(0, 0, 0, 0.2);
  padding-bottom: 12px;
}
.posts-list {
  margin-top: 60px;
  margin-bottom: 24px;
}

.posts-list .post-preview {
  text-decoration: none;
  overflow: hidden;
  display: block;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  padding: 24px 0;
}

.post-preview-last {
  border-bottom: none !important;
}

.posts-list .posts-list-caption {
  grid-column: screen;
  font-weight: 400;
}

.posts-list .post-preview h2 {
  margin: 0 0 6px 0;
  line-height: 1.2em;
  font-style: normal;
  font-size: 24px;
}

.posts-list .post-preview p {
  margin: 0 0 12px 0;
  line-height: 1.4em;
  font-size: 16px;
}

.posts-list .post-preview .thumbnail {
  box-sizing: border-box;
  margin-bottom: 24px;
  position: relative;
  max-width: 500px;
}
.posts-list .post-preview img {
  width: 100%;
  display: block;
}

.posts-list .metadata {
  font-size: 12px;
  line-height: 1.4em;
  margin-bottom: 18px;
}

.posts-list .metadata > * {
  display: inline-block;
}

.posts-list .metadata .publishedDate {
  margin-right: 2em;
}

.posts-list .metadata .dt-authors {
  display: block;
  margin-top: 0.3em;
  margin-right: 2em;
}

.posts-list .dt-tags {
  display: block;
  line-height: 1em;
}

.posts-list .dt-tags .dt-tag {
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0.3em 0.4em;
  margin-right: 0.2em;
  margin-bottom: 0.4em;
  font-size: 60%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

.posts-list img {
  opacity: 1;
}

.posts-list img[data-src] {
  opacity: 0;
}

.posts-more {
  clear: both;
}


.posts-sidebar {
  font-size: 16px;
}

.posts-sidebar h3 {
  font-size: 16px;
  margin-top: 0;
  margin-bottom: 0.5em;
  font-weight: 400;
  text-transform: uppercase;
}

.sidebar-section {
  margin-bottom: 30px;
}

.categories ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.categories li {
  color: rgba(0, 0, 0, 0.8);
  margin-bottom: 0;
}

.categories li>a {
  border-bottom: none;
}

.categories li>a:hover {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
}

.categories .active {
  font-weight: 600;
}

.categories .category-count {
  color: rgba(0, 0, 0, 0.4);
}


@media(min-width: 768px) {
  .posts-list .post-preview h2 {
    font-size: 26px;
  }
  .posts-list .post-preview .thumbnail {
    float: right;
    width: 30%;
    margin-bottom: 0;
  }
  .posts-list .post-preview .description {
    float: left;
    width: 45%;
  }
  .posts-list .post-preview .metadata {
    float: left;
    width: 20%;
    margin-top: 8px;
  }
  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.5em;
    font-size: 16px;
  }
  .posts-with-sidebar .posts-list {
    float: left;
    width: 75%;
  }
  .posts-with-sidebar .posts-sidebar {
    float: right;
    width: 20%;
    margin-top: 60px;
    padding-top: 24px;
    padding-bottom: 24px;
  }
}


/* Improve display for browsers without grid (IE/Edge <= 15) */

.downlevel {
  line-height: 1.6em;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  margin: 0;
}

.downlevel .d-title {
  padding-top: 6rem;
  padding-bottom: 1.5rem;
}

.downlevel .d-title h1 {
  font-size: 50px;
  font-weight: 700;
  line-height: 1.1em;
  margin: 0 0 0.5rem;
}

.downlevel .d-title p {
  font-weight: 300;
  font-size: 1.2rem;
  line-height: 1.55em;
  margin-top: 0;
}

.downlevel .d-byline {
  padding-top: 0.8em;
  padding-bottom: 0.8em;
  font-size: 0.8rem;
  line-height: 1.8em;
}

.downlevel .section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
}

.downlevel .d-article {
  font-size: 1.06rem;
  line-height: 1.7em;
  padding-top: 1rem;
  padding-bottom: 2rem;
}


.downlevel .d-appendix {
  padding-left: 0;
  padding-right: 0;
  max-width: none;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.5);
  padding-top: 40px;
  padding-bottom: 48px;
}

.downlevel .footnotes ol {
  padding-left: 13px;
}

.downlevel .base-grid,
.downlevel .distill-header,
.downlevel .d-title,
.downlevel .d-abstract,
.downlevel .d-article,
.downlevel .d-appendix,
.downlevel .distill-appendix,
.downlevel .d-byline,
.downlevel .d-footnote-list,
.downlevel .d-citation-list,
.downlevel .distill-footer,
.downlevel .appendix-bottom,
.downlevel .posts-container {
  padding-left: 40px;
  padding-right: 40px;
}

@media(min-width: 768px) {
  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
  padding-left: 150px;
  padding-right: 150px;
  max-width: 900px;
}
}

.downlevel pre code {
  display: block;
  border-left: 2px solid rgba(0, 0, 0, .1);
  padding: 0 0 0 20px;
  font-size: 14px;
}

.downlevel code, .downlevel pre {
  color: black;
  background: none;
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

.downlevel .posts-list .post-preview {
  color: inherit;
}



</style>

<script type="application/javascript">

function is_downlevel_browser() {
  if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                 window.navigator.userAgent)) {
    return true;
  } else {
    return window.load_distill_framework === undefined;
  }
}

// show body when load is complete
function on_load_complete() {

  // add anchors
  if (window.anchors) {
    window.anchors.options.placement = 'left';
    window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
  }


  // set body to visible
  document.body.style.visibility = 'visible';

  // force redraw for leaflet widgets
  if (window.HTMLWidgets) {
    var maps = window.HTMLWidgets.findAll(".leaflet");
    $.each(maps, function(i, el) {
      var map = this.getMap();
      map.invalidateSize();
      map.eachLayer(function(layer) {
        if (layer instanceof L.TileLayer)
          layer.redraw();
      });
    });
  }

  // trigger 'shown' so htmlwidgets resize
  $('d-article').trigger('shown');
}

function init_distill() {

  init_common();

  // create front matter
  var front_matter = $('<d-front-matter></d-front-matter>');
  $('#distill-front-matter').wrap(front_matter);

  // create d-title
  $('.d-title').changeElementType('d-title');

  // separator
  var separator = '<hr class="section-separator" style="clear: both"/>';
  // prepend separator above appendix
  $('.d-byline').before(separator);
  $('.d-article').before(separator);

  // create d-byline
  var byline = $('<d-byline></d-byline>');
  $('.d-byline').replaceWith(byline);

  // create d-article
  var article = $('<d-article></d-article>');
  $('.d-article').wrap(article).children().unwrap();

  // move posts container into article
  $('.posts-container').appendTo($('d-article'));

  // create d-appendix
  $('.d-appendix').changeElementType('d-appendix');

  // flag indicating that we have appendix items
  var appendix = $('.appendix-bottom').children('h3').length > 0;

  // replace footnotes with <d-footnote>
  $('.footnote-ref').each(function(i, val) {
    appendix = true;
    var href = $(this).attr('href');
    var id = href.replace('#', '');
    var fn = $('#' + id);
    var fn_p = $('#' + id + '>p');
    fn_p.find('.footnote-back').remove();
    var text = fn_p.html();
    var dtfn = $('<d-footnote></d-footnote>');
    dtfn.html(text);
    $(this).replaceWith(dtfn);
  });
  // remove footnotes
  $('.footnotes').remove();

  // move refs into #references-listing
  $('#references-listing').replaceWith($('#refs'));

  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    var id = $(this).attr('id');
    $('.d-contents a[href="#' + id + '"]').parent().remove();
    appendix = true;
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
  });

  // show d-appendix if we have appendix content
  $("d-appendix").css('display', appendix ? 'grid' : 'none');

  // localize layout chunks to just output
  $('.layout-chunk').each(function(i, val) {

    // capture layout
    var layout = $(this).attr('data-layout');

    // apply layout to markdown level block elements
    var elements = $(this).children().not('details, div.sourceCode, pre, script');
    elements.each(function(i, el) {
      var layout_div = $('<div class="' + layout + '"></div>');
      if (layout_div.hasClass('shaded')) {
        var shaded_content = $('<div class="shaded-content"></div>');
        $(this).wrap(shaded_content);
        $(this).parent().wrap(layout_div);
      } else {
        $(this).wrap(layout_div);
      }
    });


    // unwrap the layout-chunk div
    $(this).children().unwrap();
  });

  // remove code block used to force  highlighting css
  $('.distill-force-highlighting-css').parent().remove();

  // remove empty line numbers inserted by pandoc when using a
  // custom syntax highlighting theme, except when numbering line
  // in code chunk
  $('pre:not(.numberLines) code.sourceCode a:empty').remove();

  // load distill framework
  load_distill_framework();

  // wait for window.distillRunlevel == 4 to do post processing
  function distill_post_process() {

    if (!window.distillRunlevel || window.distillRunlevel < 4)
      return;

    // hide author/affiliations entirely if we have no authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;
    if (!have_authors)
      $('d-byline').addClass('hidden');

    // article with toc class
    $('.d-contents').parent().addClass('d-article-with-toc');

    // strip links that point to #
    $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

    // add orcid ids
    $('.authors-affiliations').find('.author').each(function(i, el) {
      var orcid_id = front_matter.authors[i].orcidID;
      var author_name = front_matter.authors[i].author
      if (orcid_id) {
        var a = $('<a></a>');
        a.attr('href', 'https://orcid.org/' + orcid_id);
        var img = $('<img></img>');
        img.addClass('orcid-id');
        img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
        img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
        a.append(img);
        $(this).append(a);
      }
    });

    // hide elements of author/affiliations grid that have no value
    function hide_byline_column(caption) {
      $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
    }

    // affiliations
    var have_affiliations = false;
    for (var i = 0; i<front_matter.authors.length; ++i) {
      var author = front_matter.authors[i];
      if (author.affiliation !== "&nbsp;") {
        have_affiliations = true;
        break;
      }
    }
    if (!have_affiliations)
      $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

    // published date
    if (!front_matter.publishedDate)
      hide_byline_column("Published");

    // document object identifier
    var doi = $('d-byline').find('h3:contains("DOI")');
    var doi_p = doi.next().empty();
    if (!front_matter.doi) {
      // if we have a citation and valid citationText then link to that
      if ($('#citation').length > 0 && front_matter.citationText) {
        doi.html('Citation');
        $('<a href="#citation"></a>')
          .text(front_matter.citationText)
          .appendTo(doi_p);
      } else {
        hide_byline_column("DOI");
      }
    } else {
      $('<a></a>')
         .attr('href', "https://doi.org/" + front_matter.doi)
         .html(front_matter.doi)
         .appendTo(doi_p);
    }

     // change plural form of authors/affiliations
    if (front_matter.authors.length === 1) {
      var grid = $('.authors-affiliations');
      grid.children('h3:contains("Authors")').text('Author');
      grid.children('h3:contains("Affiliations")').text('Affiliation');
    }

    // remove d-appendix and d-footnote-list local styles
    $('d-appendix > style:first-child').remove();
    $('d-footnote-list > style:first-child').remove();

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // hoverable references
    $('span.citation[data-cites]').each(function() {
      const citeChild = $(this).children()[0]
      // Do not process if @xyz has been used without escaping and without bibliography activated
      // https://github.com/rstudio/distill/issues/466
      if (citeChild === undefined) return true

      if (citeChild.nodeName == "D-FOOTNOTE") {
        var fn = citeChild
        $(this).html(fn.shadowRoot.querySelector("sup"))
        $(this).id = fn.id
        fn.remove()
      }
      var refs = $(this).attr('data-cites').split(" ");
      var refHtml = refs.map(function(ref) {
        // Could use CSS.escape too here, we insure backward compatibility in navigator
        return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
      }).join("\n");
      window.tippy(this, {
        allowHTML: true,
        content: refHtml,
        maxWidth: 500,
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start'
      });
    });

    // fix footnotes in tables (#411)
    // replacing broken distill.pub feature
    $('table d-footnote').each(function() {
      // we replace internal showAtNode methode which is triggered when hovering a footnote
      this.hoverBox.showAtNode = function(node) {
        // ported from https://github.com/distillpub/template/pull/105/files
        calcOffset = function(elem) {
            let x = elem.offsetLeft;
            let y = elem.offsetTop;
            // Traverse upwards until an `absolute` element is found or `elem`
            // becomes null.
            while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                x += elem.offsetLeft;
                y += elem.offsetTop;
            }

            return { left: x, top: y };
        }
        // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
        const bbox = node.getBoundingClientRect();
        const offset = calcOffset(node);
        this.show([offset.left + bbox.width, offset.top + bbox.height]);
      }
    })

    // clear polling timer
    clearInterval(tid);

    // show body now that everything is ready
    on_load_complete();
  }

  var tid = setInterval(distill_post_process, 50);
  distill_post_process();

}

function init_downlevel() {

  init_common();

   // insert hr after d-title
  $('.d-title').after($('<hr class="section-separator"/>'));

  // check if we have authors
  var front_matter = JSON.parse($("#distill-front-matter").html());
  var have_authors = front_matter.authors && front_matter.authors.length > 0;

  // manage byline/border
  if (!have_authors)
    $('.d-byline').remove();
  $('.d-byline').after($('<hr class="section-separator"/>'));
  $('.d-byline a').remove();

  // remove toc
  $('.d-contents').remove();

  // move appendix elements
  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
  });


  // inject headers into references and footnotes
  var refs_header = $('<h3></h3>');
  refs_header.text('References');
  $('#refs').prepend(refs_header);

  var footnotes_header = $('<h3></h3');
  footnotes_header.text('Footnotes');
  $('.footnotes').children('hr').first().replaceWith(footnotes_header);

  // move appendix-bottom entries to the bottom
  $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
  $('.appendix-bottom').remove();

  // remove appendix if it's empty
  if ($('.d-appendix').children().length === 0)
    $('.d-appendix').remove();

  // prepend separator above appendix
  $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

  // trim code
  $('pre>code').each(function(i, val) {
    $(this).html($.trim($(this).html()));
  });

  // move posts-container right before article
  $('.posts-container').insertBefore($('.d-article'));

  $('body').addClass('downlevel');

  on_load_complete();
}


function init_common() {

  // jquery plugin to change element types
  (function($) {
    $.fn.changeElementType = function(newType) {
      var attrs = {};

      $.each(this[0].attributes, function(idx, attr) {
        attrs[attr.nodeName] = attr.nodeValue;
      });

      this.replaceWith(function() {
        return $("<" + newType + "/>", attrs).append($(this).contents());
      });
    };
  })(jQuery);

  // prevent underline for linked images
  $('a > img').parent().css({'border-bottom' : 'none'});

  // mark non-body figures created by knitr chunks as 100% width
  $('.layout-chunk').each(function(i, val) {
    var figures = $(this).find('img, .html-widget');
    // ignore leaflet img layers (#106)
    figures = figures.filter(':not(img[class*="leaflet"])')
    if ($(this).attr('data-layout') !== "l-body") {
      figures.css('width', '100%');
    } else {
      figures.css('max-width', '100%');
      figures.filter("[width]").each(function(i, val) {
        var fig = $(this);
        fig.css('width', fig.attr('width') + 'px');
      });

    }
  });

  // auto-append index.html to post-preview links in file: protocol
  // and in rstudio ide preview
  $('.post-preview').each(function(i, val) {
    if (window.location.protocol === "file:")
      $(this).attr('href', $(this).attr('href') + "index.html");
  });

  // get rid of index.html references in header
  if (window.location.protocol !== "file:") {
    $('.distill-site-header a[href]').each(function(i,val) {
      $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
    });
  }

  // add class to pandoc style tables
  $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
  $('.kable-table').children('table').addClass('pandoc-table');

  // add figcaption style to table captions
  $('caption').parent('table').addClass("figcaption");

  // initialize posts list
  if (window.init_posts_list)
    window.init_posts_list();

  // implmement disqus comment link
  $('.disqus-comment-count').click(function() {
    window.headroom_prevent_pin = true;
    $('#disqus_thread').toggleClass('hidden');
    if (!$('#disqus_thread').hasClass('hidden')) {
      var offset = $(this).offset();
      $(window).resize();
      $('html, body').animate({
        scrollTop: offset.top - 35
      });
    }
  });
}

document.addEventListener('DOMContentLoaded', function() {
  if (is_downlevel_browser())
    init_downlevel();
  else
    window.addEventListener('WebComponentsReady', init_distill);
});

</script>

<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.5);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

.distill-site-footer {
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */
</style>
<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.7);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #ffffff;
}

.distill-site-footer {
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #fafafa;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */

.nav-right > a {
  text-transform: uppercase;
}

d-title h1, d-title p, d-title figure,
d-abstract p, d-abstract b {
  grid-column: page;
}

.rj-blue {
  color: #2467bb;
}

ul li {
  line-height: 1.6;
  margin-top: 0em;
  margin-bottom: 0em;
}</style>
<style type="text/css">
/* base style */

/* FONT FAMILIES */

:root {
  --heading-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  --mono-default: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  --body-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

body,
.posts-list .post-preview p,
.posts-list .description p {
  font-family: var(--body-font), var(--body-default);
}

h1, h2, h3, h4, h5, h6,
.posts-list .post-preview h2,
.posts-list .description h2 {
  font-family: var(--heading-font), var(--heading-default);
}

d-article div.sourceCode code,
d-article pre code {
  font-family: var(--mono-font), var(--mono-default);
}


/*-- TITLE --*/
d-title h1,
.posts-list > h1 {
  color: var(--title-color, black);
}

d-title h1 {
  font-size: var(--title-size, 50px);
}

/*-- HEADERS --*/
d-article h1,
d-article h2,
d-article h3,
d-article h4,
d-article h5,
d-article h6 {
  color: var(--header-color, rgba(0, 0, 0, 0.8));
}

/*-- BODY --*/
d-article > p,  /* only text inside of <p> tags */
d-article > ul, /* lists */
d-article > ol {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
  font-size: var(--body-size, 1.06rem);
}


/*-- CODE --*/
d-article div.sourceCode code,
d-article pre code {
  font-size: var(--code-size, 14px);
}

/*-- ASIDE --*/
d-article aside {
  font-size: var(--aside-size, 12px);
  color: var(--aside-color, rgba(0, 0, 0, 0.6));
}

/*-- FIGURE CAPTIONS --*/
figure .caption,
figure figcaption,
.figure .caption {
  font-size: var(--fig-cap-size, 13px);
  color: var(--fig-cap-color, rgba(0, 0, 0, 0.6));
}

/*-- METADATA --*/
d-byline h3 {
  font-size: var(--heading-size, 0.6rem);
  color: var(--heading-color, rgba(0, 0, 0, 0.5));
}

d-byline {
  font-size: var(--body-size, 0.8rem);
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

d-byline a,
d-article d-byline a {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

/*-- TABLE OF CONTENTS --*/
.d-contents nav h3 {
  font-size: var(--heading-size, 18px);
}

.d-contents nav a {
  font-size: var(--contents-size, 13px);
}

/*-- APPENDIX --*/
d-appendix h3 {
  font-size: var(--heading-size, 15px);
  color: var(--heading-color, rgba(0, 0, 0, 0.65));
}

d-appendix {
  font-size: var(--text-size, 0.8em);
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

d-appendix d-footnote-list a.footnote-backlink {
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

/*-- WEBSITE HEADER + FOOTER --*/
.distill-site-header .title {
  font-size: var(--title-size, 18px);
  font-family: var(--navbar-font), var(--heading-default);
}

.distill-site-header a,
.nav-dropdown .nav-dropbtn {
  font-family: var(--navbar-font), var(--heading-default);
}

.nav-dropdown .nav-dropbtn {
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  font-size: var(--text-size, 15px);
}

.distill-site-header a:hover,
.nav-dropdown:hover .nav-dropbtn {
  color: var(--hover-color, white);
}

.distill-site-header {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer a:hover {
  color: var(--hover-color, white);
}</style>
<!--/radix_placeholder_distill-->
  <script src="../../site_libs/header-attrs-2.29/header-attrs.js"></script>
  <script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="../../site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="../../site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="../../site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="../../site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="../../site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="../../site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="../../site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="../../site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
<!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Mathjax config (add automatic linebreaks when supported)
      // MathJax = {
      //    tex: {
      //        inlineMath: [['$', '$'], ['\\(', '\\)']],
      //        displayMath: [['$$', '$$'], ['\\[', '\\]']],
      //        tags: 'ams',
      //        multline: true,
      //    },
      //    options: {
      //        linebreaks: { automatic: true },
      //    },
      // };

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      buttons.append('<a href="' + slug + '.zip" class="dt-tag"><i class="fas fa-file-zipper"></i> Supplement</a>')

      // adds Abstract: in front of the first <p> in the title section --
      // unless it happens to be the subtitle (FIXME: this is a bad hack - can't distill do this?)
      var tpar = $("d-title p:not(:empty)").filter(function() {
        return !$(this).hasClass("subtitle");
      }).first();
      if (tpar) {
        const abstract = $('<d-abstract>')
        abstract.append('<b>Abstract:</b><br>')
        abstract.append(tpar) // Move description to d-abstract
        $("d-title p:empty").remove() // Remove empty paragraphs after title
        abstract.append(buttons)
        abstract.insertAfter($('d-title')) // Add abstract section after title */
      }

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
    
    d-appendix .citation-appendix,
    .d-appendix .citation-appendix {
      color: rgb(60, 60, 60);
    }

    d-article h2 {
      border-bottom: 0px solid rgba(0, 0, 0, 0.1);
      padding-bottom: 0rem;
    }
    d-article h3 {
      font-size: 20px;
    }
    d-article h4 {
      font-size: 18px;
      text-transform: none;
    }

    @media (min-width: 1024px) {
      d-article h2 {
        font-size: 32px;
      }
      d-article h3 {
        font-size: 24px;
      }
      d-article h4 {
        font-size: 20px;
      }
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"ebmstate: An R Package For Disease Progression Analysis Under Empirical Bayes Cox Models","description":"The new R package ebmstate is a package for multi-state survival\nanalysis. It is suitable for high-dimensional data and allows point\nand interval estimation of relative transition hazards, cumulative\ntransition hazards and state occupation probabilities, under\nclock-forward and clock-reset Cox models. Our package extends the\npackage mstate in a threefold manner: it transforms the Cox regression\nmodel into an empirical Bayes model that can handle high-dimensional\ndata; it introduces an analytical, Fourier transform-based estimator\nof state occupation probabilities for clock-reset models that is much\nfaster than the corresponding, simulation-based estimator in mstate;\nand it replaces asymptotic confidence intervals meant for the\nlow-dimensional setting by non-parametric bootstrap confidence\nintervals. Our package supports multi-state models of arbitrary\nstructure, but the estimators of state occupation probabilities are\nvalid for transition structures without cycles only. Once the input\ndata is in the required format, estimation is handled automatically.\nThe present paper includes a tutorial on how to use ebmstate to\nestimate transition hazards and state occupation probabilities, as\nwell as a simulation study showing how it outperforms mstate in\nhigher-dimensional settings.","doi":"10.32614/RJ-2024-002","authors":[{"author":"Rui J. Costa","authorURL":"#","affiliation":"European Molecular Biology Laboratory","affiliationURL":"#","orcidID":""},{"author":"Moritz Gerstung","authorURL":"#","affiliation":"aff. 1: European Molecular Biology Laboratory","affiliationURL":"#","orcidID":""}],"publishedDate":"2025-01-11T00:00:00.000+00:00","citationText":"Costa & Gerstung, 2025"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a class="logo" href="../../index.html">
<img src="../../resources/rlogo.png" alt="Logo"/>
</a>
<a href="../../index.html" class="title">The R Journal</a>
</div>
<div class="nav-right">
<a href="../../index.html">Home</a>
<a href="../../issues/2024-2">Current</a>
<a href="../../issues.html">Issues</a>
<a href="../../news.html">News</a>
<a href="../../submissions.html">Submit</a>
<a href="../../editors.html">Editorial board</a>
<a href="../../articles.xml">
<i class="fa fa-rss" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>ebmstate: An R Package For Disease Progression Analysis Under Empirical Bayes Cox Models</h1>

<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>The new R package ebmstate is a package for multi-state survival
analysis. It is suitable for high-dimensional data and allows point
and interval estimation of relative transition hazards, cumulative
transition hazards and state occupation probabilities, under
clock-forward and clock-reset Cox models. Our package extends the
package mstate in a threefold manner: it transforms the Cox regression
model into an empirical Bayes model that can handle high-dimensional
data; it introduces an analytical, Fourier transform-based estimator
of state occupation probabilities for clock-reset models that is much
faster than the corresponding, simulation-based estimator in mstate;
and it replaces asymptotic confidence intervals meant for the
low-dimensional setting by non-parametric bootstrap confidence
intervals. Our package supports multi-state models of arbitrary
structure, but the estimators of state occupation probabilities are
valid for transition structures without cycles only. Once the input
data is in the required format, estimation is handled automatically.
The present paper includes a tutorial on how to use ebmstate to
estimate transition hazards and state occupation probabilities, as
well as a simulation study showing how it outperforms mstate in
higher-dimensional settings.</p></p>
</div>

<div class="d-byline">
  Rui J. Costa  (European Molecular Biology Laboratory)
  
,   Moritz Gerstung  (aff. 1: European Molecular Biology Laboratory)
  
<br/>2025-01-11
</div>

<div class="d-article">
<div class="article">
<h3 data-number="1" id="introduction"><span class="header-section-number">1</span> Introduction</h3>
<p>Multi-state models based on transition hazard functions are often used
in the statistical analysis of longitudinal data, in particular disease
progression data <span class="citation" data-cites="Hougaard1999">(<a href="#ref-Hougaard1999" role="doc-biblioref">Hougaard 1999</a>)</span>. The multi-state model framework is
particularly suitable to accommodate the growing level of detail of
modern clinical data: as long as a clinical history can be framed as a
random process which, at any moment in time, occupies one of a few
states, a multi-state model is applicable. Another strong point of this
framework is that it can incorporate a <em>regression model</em>, i.e., a set
of assumptions on how covariates, possibly time-dependent ones, affect
the risk of transitioning between any two states of the disease. Once
estimated, multi-state models with regression features allow the
stratification of patients according to their transition hazards. In
addition, it is possible, under some models, to generate disease outcome
predictions. These come in the form of <em>state occupation probability</em>
estimates, meaning estimates of the probability of being in each state
of the disease over a given time frame.</p>
<p>The survival analysis â€˜task viewâ€™ of the Comprehensive R Archive Network
lists seven R packages that are able to fit <em>general</em> multi-state models
and, at the same time, feature some kind of regression model or
algorithm: <code>flexsurv</code> <span class="citation" data-cites="flexsurv_package">(<a href="#ref-flexsurv_package" role="doc-biblioref">Jackson 2016</a>)</span>,
<a href="https://CRAN.R-project.org/package=msm"><strong>msm</strong></a> <span class="citation" data-cites="Jackson2011">(<a href="#ref-Jackson2011" role="doc-biblioref">Jackson 2011</a>)</span>,
<a href="https://CRAN.R-project.org/package=SemiMarkov"><strong>SemiMarkov</strong></a>
<span class="citation" data-cites="Listwon2015">(<a href="#ref-Listwon2015" role="doc-biblioref">Listwon and Saint-Pierre 2015</a>)</span>,
<a href="https://CRAN.R-project.org/package=survival"><strong>survival</strong></a>
<span class="citation" data-cites="survival_package">(<a href="#ref-survival_package" role="doc-biblioref">Therneau 2015</a>)</span>,
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> <span class="citation" data-cites="Wreede2010">(<a href="#ref-Wreede2010" role="doc-biblioref">Wreede et al. 2010</a>)</span>,
<a href="https://CRAN.R-project.org/package=mboost"><strong>mboost</strong></a>
<span class="citation" data-cites="mboost_package">(<a href="#ref-mboost_package" role="doc-biblioref">Hothorn et al. 2020</a>)</span> â€“ as extended by
<a href="https://CRAN.R-project.org/package=gamboostMSM"><strong>gamboostMSM</strong></a>
<span class="citation" data-cites="gamboostMSM_package">(<a href="#ref-gamboostMSM_package" role="doc-biblioref">Reulen 2014</a>)</span> â€“ and
<a href="https://CRAN.R-project.org/package=penMSM"><strong>penMSM</strong></a>
<span class="citation" data-cites="penMSM_package">(<a href="#ref-penMSM_package" role="doc-biblioref">Reulen 2015</a>)</span>. All of them implement relative risk regression models
<span class="citation" data-cites="Aalen2008">(as defined in <a href="#ref-Aalen2008" role="doc-biblioref">Aalen et al. 2008 133</a>)</span>. The only exceptions are
<a href="https://CRAN.R-project.org/package=survival"><strong>survival</strong></a>, which also
fits Aalenâ€™s additive regression model <span class="citation" data-cites="Aalen1989">(<a href="#ref-Aalen1989" role="doc-biblioref">Aalen 1989</a>)</span>, and <code>flexsurv</code>,
which also implements accelerated failure time models .</p>
<p>Recall that a Cox regression model is a semi-parametric model in which
every transition hazard is assumed to be the product of a baseline
hazard function of unspecified form (the non-parametric component) and
an exponential relative risk function (the parametric component)
<span class="citation" data-cites="Aalen2008">(<a href="#ref-Aalen2008" role="doc-biblioref">Aalen et al. 2008 133</a>)</span>. Generally, the relative risk regression models
implemented in these packages are Cox regression models. However, some
models in <code>flexsurv</code>, as well as those in
<a href="https://CRAN.R-project.org/package=msm"><strong>msm</strong></a> and
<a href="https://CRAN.R-project.org/package=SemiMarkov"><strong>SemiMarkov</strong></a>, also
restrict the baseline hazards to specific parametric families, i.e.Â they
are fully parametric. In
<a href="https://CRAN.R-project.org/package=msm"><strong>msm</strong></a> and
<a href="https://CRAN.R-project.org/package=SemiMarkov"><strong>SemiMarkov</strong></a>, the
stronger assumptions regarding the functional form of the hazard are
leveraged to do away with other common assumptions:
<a href="https://CRAN.R-project.org/package=SemiMarkov"><strong>SemiMarkov</strong></a> drops
the usual Markov property to implement homogeneous semi-Markov models;
<a href="https://CRAN.R-project.org/package=msm"><strong>msm</strong></a> is suitable for <em>panel
data</em>, i.e., data in which the state of each individual is known only at
a finite series of times.</p>
<p>Packages <a href="https://CRAN.R-project.org/package=penMSM"><strong>penMSM</strong></a> and
<a href="https://CRAN.R-project.org/package=gamboostMSM"><strong>gamboostMSM</strong></a> are
the best suited to deal with higher-dimensional covariate data. The
first of these packages relies on a structured fusion lasso method,
while the second implements (jointly with
<a href="https://CRAN.R-project.org/package=mboost"><strong>mboost</strong></a>) a boosting
algorithm. Both methods induce sparsity in the number of non-zero
covariate effects, as well as equality among the different transition
effects of each covariate, and are thus especially useful to reduce
complicated multi-state models to more interpretable ones. The remaining
packages assume standard, fixed effects relative risk regression models
and do not include regularisation or variable selection features.</p>
<p>It is also illustrative to order the seven packages mentioned according
to how extensive their analysis workflow is. Packages
<a href="https://CRAN.R-project.org/package=SemiMarkov"><strong>SemiMarkov</strong></a> and
<a href="https://CRAN.R-project.org/package=penMSM"><strong>penMSM</strong></a> are intended for
the estimation of relative transition hazards only (i.e., for estimating
the impact of covariates on each transition hazard). With the package
<a href="https://CRAN.R-project.org/package=mboost"><strong>mboost</strong></a> (as extended by
<a href="https://CRAN.R-project.org/package=gamboostMSM"><strong>gamboostMSM</strong></a>) it is
also possible to estimate the baseline transition hazards. Finally, a
more complete workflow including estimates of both relative and
cumulative transition hazards, as well as state occupation
probabilities, is implemented in <code>flexsurv</code>,
<a href="https://CRAN.R-project.org/package=msm"><strong>msm</strong></a> and
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a>, and has been
under implementation in
<a href="https://CRAN.R-project.org/package=survival"><strong>survival</strong></a> (version 3.0
or later).</p>
<p>The present paper provides an introduction to
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a>, a new R
package for multi-state survival analysis available for download on the
Comprehensive R Archive Network (CRAN). The main goal of
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> is to
provide an analysis framework for the Cox model that performs better
with higher-dimensional covariate data and is also complete, in the
sense of being able to generate point and interval estimates of relative
transition hazards, cumulative transition hazards and state occupation
probabilities, both under clock-forward and clock-reset models. A
fundamental characteristic of
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> is that it
re-implements and extends the analysis framework of
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a>, which is
complete in the sense just mentioned. In fact, to a large extent, our
package was built by importing, adapting and replacing functions from
the <a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> package.
This not only eliminates redundancies, but also makes our package more
accessible to the numerous users of
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> (the three
papers associated with
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> have jointly
over 2000 citations).</p>
<p>To improve the performance of
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a>â€™s multi-state
Cox model when dealing with higher-dimensional covariate data, a
ridge-type regularisation feature was added. We allow the regression
coefficients of the model to be partitioned into groups, with each group
having its own Gaussian prior. A group can gather, for example, all the
regression coefficients for a given transition. Or, within a given
transition, coefficients can be grouped according to the covariate type
they refer to (for example, demographic, clinical or genomic type). The
resulting hierarchical Bayes model is <em>empirical</em> in that a full prior
elicitation is not required (the mean and variance hyper-parameters of
the Gaussian are estimated from the data). Model fitting relies on the
iterative algorithm introduced by <span class="citation" data-cites="Schall1991">Schall (<a href="#ref-Schall1991" role="doc-biblioref">1991</a>)</span>, which typically converges
after a small number of steps. A simulation study showing that Schallâ€™s
algorithm performance compares well with that of other algorithms for
ridge penalty optimisation, including one based on cross-validation, can
be found in <span class="citation" data-cites="Perperoglou2014">Perperoglou (<a href="#ref-Perperoglou2014" role="doc-biblioref">2014</a>)</span>.</p>
<p>The asymptotic confidence intervals generated by
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> are applicable
when the number of observations is much larger than the number of
parameters to be estimated (see section <a href="#sec:interval_estimation">3.3</a>
below). To preserve the completeness of
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a>â€™s framework in
higher-dimensional settings, we therefore implemented non-parametric
bootstrap intervals of regression coefficients, cumulative transition
hazards and state occupation probabilities.</p>
<p>The high computational cost implied by the non-parametric bootstrap
motivated a third extension to
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a>. We developed an
estimator of state occupation probabilities under clock-reset Cox models
that is based on a convolution argument <span class="citation" data-cites="Spitoni2012">(as in <a href="#ref-Spitoni2012" role="doc-biblioref">Spitoni et al. 2012</a>)</span> and the
Fast Fourier transform (FFT). At present, the estimation of such
probabilities for clock-forward Cox models can be carried out using the
efficient, product-limit based algorithm available in
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a>. However, for
clock-reset Cox models, only a simulation-based estimator is available
in this package (see also the <code>flexsurv</code> package for a similar,
simulation-based estimator). The FFT estimator in
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> was
conceived as a faster alternative to this simulation-based estimator,
but its scope is currently restricted to multi-state models with
transition structures that have no cycles, i.e.Â in which a transition
between two states is either not possible or follows a unique sequence
of states. Figure <a href="#fig:figpackage-summary-figure">1</a> provides a short
graphical summary of
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a>, with the
main inputs â€“ a genomic-clinical data set and an empirical Bayes
multi-state Cox model â€“ and the main outputs â€“ the estimates of
relative hazards and state occupation probabilities (cumulative
transition hazards are omitted).</p>
<p>As already mentioned, our empirical Bayes method improves estimator
performance in models with larger numbers of covariates (see section
<a href="#sec:estimator_performance">4</a> on estimator performance). Also, as a
ridge-type regression method, it can be used as an alternative to the
lasso method of <a href="https://CRAN.R-project.org/package=penMSM"><strong>penMSM</strong></a>
in two particular cases: when the levels of correlation between
covariates are high enough to compromise the stability of lasso-based
covariate selection; or simply to improve prediction accuracy when
interpretability is not essential and the number of covariates is not
greater than the number of observations <span class="citation" data-cites="Zou2005">(<a href="#ref-Zou2005" role="doc-biblioref">Zou and Hastie 2005</a>)</span>. In addition, and
perhaps more importantly,
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> goes beyond
the regularised estimation of transition hazards offered by
<a href="https://CRAN.R-project.org/package=penMSM"><strong>penMSM</strong></a> and
<a href="https://CRAN.R-project.org/package=gamboostMSM"><strong>gamboostMSM</strong></a>: point
and interval estimates of state occupation probabilities under the
regularised Cox model can also be computed.</p>
<h3 data-number="2" id="models"><span class="header-section-number">2</span> Models</h3>
<p>A multi-state Cox model is a continuous-time stochastic process with a
finite (and usually small) state space <span class="math inline">\(\mathcal{S}\)</span>. To better describe
the models implemented in
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a>, we define
the following notation. We let <span class="math inline">\(t\)</span> denote the time since some initiating
event (usually diagnosis or disease onset). For
<span class="math inline">\(t \in \left[0, \infty\right)\)</span>, we define the following random
variables: <span class="math inline">\(X(t)\)</span> represents the disease state of the patient, <span class="math inline">\(S(t)\)</span>
the time spent in the current state, and <span class="math inline">\(\vec{Z}\left(t\right)\)</span> the
value of a covariate vector. The realisation of each component of the
process <span class="math inline">\(\lbrace\vec{Z}\left(t\right)\rbrace\)</span> is a step function,
possibly approximating the evolution in time of a continuous covariate.
In addition, <span class="math inline">\(\lbrace\vec{Z}\left(t\right)\rbrace\)</span> is assumed
not-adapted to the filtration generated by
<span class="math inline">\(\lbrace X\left(t\right)\rbrace\)</span> (an adapted covariate is one whose path
until <span class="math inline">\(t\)</span> is known once <span class="math inline">\(\lbrace X \left(u\right)\rbrace\)</span>, <span class="math inline">\(u \leq t\)</span>,
is known). The transition hazard rate of a patient from state <span class="math inline">\(i\)</span> to
state <span class="math inline">\(j\)</span> (<span class="math inline">\(i\neq j\)</span>) at time <span class="math inline">\(t\)</span>, conditional on the sojourn time and
the covariate vector, is defined as
<span class="math display">\[\begin{aligned}
&amp;\alpha_{ij}\left(t|\mathbf{z},s \right):=\lim_{h \downarrow 0}\frac{1}{h}\mathrm{P}\left[X(t+h)=j\,|\,X(t)=i,S(t)=s,\vec{Z}(t)=\mathbf{z} \right]\;, \;s\in \left[0,\infty\right)\;,\;t\in \left[s,\infty\right)\;.
\end{aligned}\]</span>
Independent right-censoring and left-truncation are assumed throughout
<span class="citation" data-cites="Aalen2008">(<a href="#ref-Aalen2008" role="doc-biblioref">Aalen et al. 2008 57</a>)</span>. The purpose of the present section is to give a (not
necessarily exhaustive) description of the scope of
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> and
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> with respect
to the multi-state Cox model. Using the terminology in <span class="citation" data-cites="Putter2011">de Wreede et al. (<a href="#ref-Putter2011" role="doc-biblioref">2011</a>)</span>, a
Cox model is termed a â€˜clock-resetâ€™ model when
<span class="math display" id="eq:clock-reset-Cox">\[\begin{aligned}
\label{eq:clock_reset_Cox}
\alpha_{ij}\left(t\,|\,\mathbf{z}, s\right)&amp;=\lambda_{ij}^{(0)}\left(s\right)\exp\left[ \boldsymbol{\beta}^{\intercal}_{ij}\,\mathbf{z}\right] \quad,
\end{aligned}   \tag{1}\]</span>
and it is termed a â€˜clock-forwardâ€™ model when
<span class="math display" id="eq:clock-forward-Cox">\[\begin{aligned}
\label{eq:clock_forward_Cox}
\alpha_{ij}\left(t\,|\,\mathbf{z}\right)&amp;=\alpha_{ij}^{(0)}\left(t\right)\exp\left[ \boldsymbol{\beta}^{\intercal}_{ij}\,\mathbf{z}\right] \quad.
\end{aligned}   \tag{2}\]</span>
In both cases, <span class="math inline">\(i,j \in \mathcal{S}\)</span>, with <span class="math inline">\(i\neq j\)</span>;
<span class="math inline">\(\boldsymbol{\beta}_{\scriptscriptstyle ij}\)</span> is an unknown vector of
regression coefficient parameters, and both
<span class="math inline">\(\lambda^{\scriptscriptstyle (0)}_{ij}(\cdot)\)</span> and
<span class="math inline">\(\alpha^{\scriptscriptstyle (0)}_{ij}(\cdot)\)</span> are unknown (baseline
hazard) functions, non-negative on <span class="math inline">\(\mathbb{R}^{+}\)</span>. When, as in
equation <a href="#eq:clock-reset-Cox">(1)</a>,
<span class="math inline">\(\alpha_{ij}\left(t|\mathbf{z},s\right)\)</span> is the same for all <span class="math inline">\(t\geq s\)</span>,
we simplify its notation to <span class="math inline">\(\lambda_{ij}\left(s|\mathbf{z}\right)\)</span>. As
can be seen from equations <a href="#eq:clock-reset-Cox">(1)</a> and
<a href="#eq:clock-forward-Cox">(2)</a>, the â€˜clock-resetâ€™ and â€˜clock-forwardâ€™
models are models for how the transition hazard rates are affected by
time. In the former case, the only relevant time scale is the time <span class="math inline">\(s\)</span>
spent in the current state, whereas in the latter only the time <span class="math inline">\(t\)</span>
since the initiating event matters. While the â€˜clock-forwardâ€™ model is
arguably the default one in multi-state survival analysis
<span class="citation" data-cites="Andersen1993 Aalen2008">(<a href="#ref-Andersen1993" role="doc-biblioref">Andersen et al. 1993</a>; <a href="#ref-Aalen2008" role="doc-biblioref">Aalen et al. 2008</a>)</span>, in some cases the â€˜clock-resetâ€™ model is
more appropriate. For example, in some forms of cancer, it can be
sensible to assume that the transition hazards from the state of
complete remission depend on the sojourn time, rather than on the time
since the initial diagnosis.</p>
<h4 class="unnumbered" data-number="2.1" id="sec:models_relative_hazards">Relative transition hazards</h4>
<p>The parametric component of the transition hazard from <span class="math inline">\(i\)</span> to <span class="math inline">\(j\)</span>,
written
<span class="math inline">\(\exp\left[\boldsymbol{\beta}^{\intercal}_{ij} \,\mathbf{z}\right]\)</span>, is
termed the relative transition hazard. In
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> and
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a>, estimating
the relative transition hazard amounts to estimating the regression
coefficient vector <span class="math inline">\(\boldsymbol{\beta}_{ij}\,\)</span>. In
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a>, these
parameters are assumed to be non-random. With
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a>, the
following prior distributions can be imposed.</p>
<p>Define <span class="math inline">\(\mathcal{P}\)</span> as the set of all pairs of states between which a
direct transition is possible. Let
<span class="math inline">\(\lbrace \boldsymbol{\beta}_{\scriptscriptstyle ij} \rbrace\)</span>, for all
<span class="math inline">\((i, j) \in \mathcal{P}\)</span>, be a partition of <span class="math inline">\(\boldsymbol \beta\)</span>, a
vector containing the regression coefficients for all direct transitions
allowed. Each <span class="math inline">\(\boldsymbol{\beta}_{\scriptscriptstyle ij}\)</span> is further
partitioned into
<span class="math inline">\(\lbrace \boldsymbol{\beta}_{\scriptscriptstyle ijk} \rbrace\)</span>, for
<span class="math inline">\(k \in \left\lbrace 1,2,...,n_{\scriptscriptstyle ij} \right\rbrace\)</span>. In
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a>, the most
general model regarding the prior distribution of <span class="math inline">\(\boldsymbol{\beta}\)</span>
makes two assumptions: a) the scalar components of <span class="math inline">\(\boldsymbol{\beta}\)</span>
are independent and normally distributed; b) the scalar components of
<span class="math inline">\(\boldsymbol{\beta}_{\scriptscriptstyle i j k}\)</span> have a common (and
undetermined) mean <span class="math inline">\(\mu_{\scriptscriptstyle ijk}\)</span> and a common (and also
undetermined) variance <span class="math inline">\(\sigma^{2}_{\scriptscriptstyle ijk}\;\)</span>.</p>
<p>The purpose of the framework just described is to allow the clustering
of covariate effects according to their prior distribution. If there is
no prior knowledge about how this clustering should be done, a single
Gaussian prior can be imposed on all regression coefficients at once. If
prior knowledge allows the grouping of effects according to the
transition they refer to, a different Gaussian prior can be assigned to
the coefficients of each transition. Even within each transition,
different groups of coefficients can be assigned different prior
distributions. In the analysis of biomedical data, for example, there
can be a split between genes which are known to affect the transition
hazard, and other genes whose effect is unknown.</p>
<h4 class="unnumbered" data-number="2.2" id="cumulative-transition-hazard-functions">Cumulative transition hazard functions</h4>
<p>Our package imports from
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> a Breslow
estimator of two types of cumulative transition hazard: one on a global
time scale, defined as
<span class="math display">\[\begin{aligned}
\mathrm{A}_{ij}\left(t\,|\,\mathbf{z}\right)&amp;:=\int_{0}^{t}\alpha_{ij}^{(0)}\left(u\right)\exp\left[ \boldsymbol{\beta}^{\intercal}_{ij}\,\mathbf{z}\right]\mathrm{d}u\quad,
\end{aligned}\]</span>
and another on a sojourn time scale, defined as
<span class="math display">\[\begin{aligned}
&amp;\Lambda_{ij}(s\,|\,\mathbf{z}):=\int_{0}^{s}\lambda_{ij}^{(0)}\left(u\right)\exp\left[ \boldsymbol{\beta}^{\intercal}_{ij}\,\mathbf{z}\right]\mathrm{d}u\quad.
\end{aligned}\]</span>
Note that, in either case, the covariate vector is assumed to remain
constant.</p>
<h4 class="unnumbered" data-number="2.3" id="state-occupation-probabilities">State occupation probabilities</h4>
<p>By state occupation probability, we mean the probability that a patient
in state <span class="math inline">\(i\)</span> at time <span class="math inline">\(0\)</span> finds herself in state <span class="math inline">\(j\)</span> at time <span class="math inline">\(t\)</span>. The
estimates of these probabilities can be seen as functionals of the
estimated cumulative transition hazard functions. For this reason, the
restriction to models with time-fixed covariates, which was just seen to
be applicable to the estimators of cumulative transition hazards,
carries over to the estimation of state occupation probabilities.</p>
<p>When conditioning on a given covariate path (time-fixed or not), state
occupation probability estimates are not valid unless the covariates are
<em>external</em> <span class="citation" data-cites="Cortese2010 Aalen2008">(<a href="#ref-Aalen2008" role="doc-biblioref">Aalen et al. 2008 142</a>; <a href="#ref-Cortese2010" role="doc-biblioref">Cortese and Andersen 2010</a>)</span>. Note that a vector of
covariates <span class="math inline">\(\lbrace \vec{Z}(u)\rbrace_{u\geq 0}\)</span> is said to be
<em>external</em> if, for all <span class="math inline">\(t \in \left[0,\infty\right)\)</span>, each transition
hazard at <span class="math inline">\(t\)</span>, conditional on <span class="math inline">\(\vec{Z}(t)\)</span>, is independent of
<span class="math inline">\(\lbrace \vec{Z}(u)\rbrace_{u&gt;t}\)</span> (i.e.Â independent of the future path
of the covariate). Otherwise, it is said to be <em>internal</em> <span class="citation" data-cites="Kalbfleisch2002">(for more
details on the distinction between internal and external covariates, see <a href="#ref-Kalbfleisch2002" role="doc-biblioref">Kalbfleisch and Prentice 2002 6</a>)</span>. When one does not wish (or is not possible
due to <span class="math inline">\(\vec{Z}\)</span> being <em>internal</em>) to condition on a future covariate
path of the covariate process, the uncertainty introduced by this
process needs to be accounted for. This can be done by extending the
state space of the disease process, so that it includes information on
the disease <em>and</em> the covariate process <span class="citation" data-cites="Andersen1993">(<a href="#ref-Andersen1993" role="doc-biblioref">Andersen et al. 1993 170</a>)</span>. For
example, to include a dichotomous transplant covariate (an internal
covariate) in a simple survival model with two states, the state space
is expanded from <span class="math inline">\(\lbrace\)</span>alive, deceased<span class="math inline">\(\rbrace\)</span> to <span class="math inline">\(\lbrace\)</span>alive
without transplant, alive with transplant, deceased<span class="math inline">\(\rbrace\)</span>. One can
then either assume that transplanted patients have a different baseline
death hazard or, more simply, that transplantation scales the death
hazard by some constant <span class="math inline">\(\exp \left( \gamma\right)\)</span>. A similar but more
detailed example can be found in <span class="citation" data-cites="Wreede2010">Wreede et al. (<a href="#ref-Wreede2010" role="doc-biblioref">2010 2.3.2, â€˜model 3â€™</a>)</span>.</p>
<h3 data-number="3" id="estimation"><span class="header-section-number">3</span> Estimation</h3>
<p>In the current section, we present the estimation methods underlying the
extensions of <a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a>
implemented in
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a>.
<span id="sec:estimation" label="sec:estimation"></span></p>
<h4 class="unnumbered" data-number="3.1" id="relative-and-cumulative-hazard-functions">Relative and cumulative hazard functions</h4>
<p>Let <span class="math inline">\(\boldsymbol{\mu}_{\scriptscriptstyle ij}\)</span>, with
<span class="math inline">\(\left(i,j\right) \in \mathcal{P}\)</span> (the set of direct transitions
allowed), denote a vector whose scalar components are the parameters
<span class="math inline">\(\mu_{\scriptscriptstyle ijk}\)</span>,
<span class="math inline">\(k \in \left\lbrace 1,2,...,n_{\scriptscriptstyle ij} \right\rbrace\)</span>.
Similarly, let <span class="math inline">\(\boldsymbol{\sigma}^{2}_{\scriptscriptstyle ij}\)</span> be
composed of the parameters
<span class="math inline">\(\left\lbrace \sigma^{2}_{\scriptscriptstyle ijk}\right\rbrace_{k}\)</span>. The
estimation of <span class="math inline">\(\boldsymbol{\beta}\)</span>,
<span class="math inline">\(\boldsymbol{\mu}:=\lbrace\boldsymbol{\mu}_{\scriptscriptstyle{ij}}\rbrace\)</span>
and
<span class="math inline">\(\boldsymbol{\sigma}^2:=\lbrace\boldsymbol{\sigma}^2_{\scriptscriptstyle ij }\rbrace\)</span>
relies on the restricted maximum-likelihood (REML) type algorithm
described in <span class="citation" data-cites="Perperoglou2014">(<a href="#ref-Perperoglou2014" role="doc-biblioref">Perperoglou 2014</a>)</span>, and introduced by <span class="citation" data-cites="Schall1991">(<a href="#ref-Schall1991" role="doc-biblioref">Schall 1991</a>)</span>. The
resulting estimate of <span class="math inline">\(\boldsymbol{\beta}\)</span> is a maximum <em>a posteriori</em>
estimate; the estimates of <span class="math inline">\(\boldsymbol{\mu}\)</span> and
<span class="math inline">\(\boldsymbol{\sigma}^{2}\)</span> are empirical Bayes estimates. In
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a>, the
estimator based on this algorithm is implemented in the function
<code>CoxRFX</code> . The results of a simulation study showing its consistency are
included in the Supporting Scripts and Data (file ESM_1.html, section
1).</p>
<p>The computation of cumulative hazard rates for given covariate values
and an estimated regression coefficient vector relies on the function
<code>msfit_generic</code>, which is essentially a wrapper for the function
<code>mstate::msfit</code> (see section <a href="#sec:computing_cumulative_hazards">5.3</a>).
For the mathematical details of this computation, we refer therefore the
reader to <span class="citation" data-cites="Wreede2010">Wreede et al. (<a href="#ref-Wreede2010" role="doc-biblioref">2010</a>)</span>.</p>
<h4 class="unnumbered" data-number="3.2" id="sec:trans_probs">State occupation probabilities</h4>
<p>The package <a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a>
includes a simulation-based estimator that can take as input either
<span class="math inline">\(\hat{\mathrm{A}}_{ij}\left(\cdot\,|\,\mathbf{z}\right)\)</span> or
<span class="math inline">\(\hat{\Lambda}_{ij}\left(\cdot\,|\,\mathbf{z}\right)\)</span> to generate
estimates of state occupation probabilities under the clock-forward or
the clock-reset model respectively. Another available estimator, an
Aalen-Johansen-type estimator based on product integration, is far more
efficient computationally and takes as input
<span class="math inline">\(\hat{\mathrm{A}}_{ij}\left(\cdot\,|\,\mathbf{z}\right)\)</span> only. As the
scope of this estimator has been restricted to clock-forward Cox models
<span class="citation" data-cites="Andersen1993 Aalen2008">(<a href="#ref-Andersen1993" role="doc-biblioref">Andersen et al. 1993</a>; <a href="#ref-Aalen2008" role="doc-biblioref">Aalen et al. 2008</a>)</span>, in our package we implemented a
convolution-based estimator as a computationally efficient alternative
(for models with a transition structure that has no cycles).</p>
<p>For convenience, let the sequence of states from <span class="math inline">\(0\)</span> to <span class="math inline">\(n\)</span> have the
labels <span class="math inline">\(0,1,2,...,n\,\)</span>, where <span class="math inline">\(0\)</span> is the initial state by definition,
and <span class="math inline">\(n\)</span> is some state that might (eventually) be reached by the process.
In addition, define <span class="math inline">\(X_{0}:=X(0)\)</span> and <span class="math inline">\(T_{0}:=0\)</span>, and let
<span class="math inline">\(\left(X_{i},T_{i}\right)\)</span>, <span class="math inline">\(i \in \left\lbrace 1,2,... \right\rbrace\)</span>,
denote the marked point process associated with
<span class="math inline">\(\left\lbrace X(t)\right\rbrace\)</span>, so that <span class="math inline">\(T_{i}\)</span> is the time of the
<span class="math inline">\(i^{th}\)</span> transition and <span class="math inline">\(X_{i}\)</span> is the state the process jumps to at
time <span class="math inline">\(T_{i}\)</span>. The inter-transition times are denoted by
<span class="math inline">\(\tau_{ij}:=T_{j}-T_{i}\)</span>, for <span class="math inline">\(j&gt;i\)</span>. We can write the probability that a
patient in state <span class="math inline">\(0\)</span> at time <span class="math inline">\(0\)</span> finds herself in state <span class="math inline">\(n\)</span> at time <span class="math inline">\(t\)</span>,
conditional on <span class="math inline">\(\vec{Z}(u)=\mathbf{z}\)</span> for all <span class="math inline">\(u \geq 0\)</span>, as
<span class="math display">\[\begin{aligned}
&amp;\mathrm{P}\left[X(t)=n\,|\,X(0)=0\,, \vec{Z}(u)=\mathbf{z},\,u \geq 0 \right]\\
&amp;\,=\mathrm{P}\left[X_{n}=n,\tau_{0,n} &lt; t,\tau_{n,n+1}\geq t- \tau_{0,n} |X_{0}=0\,,  \vec{Z}(u)=\mathbf{z},\,u \geq 0 \right] \,.\nonumber
\end{aligned}\]</span></p>
<p>Recall that <span class="math inline">\(\lambda_{i,i+1}\left(s\,|\, \mathbf{z}\right)\)</span> denotes the
hazard rate of a transition to state <span class="math inline">\(i+1\)</span> at time <span class="math inline">\(s\)</span> since arrival in
state <span class="math inline">\(i\)</span>, for a patient that has covariate vector <span class="math inline">\(\mathbf{z}\)</span>. The
cumulative hazard for the same transition between sojourn times <span class="math inline">\(0\)</span> and
<span class="math inline">\(s\)</span>, if the patientâ€™s covariate vector remains constant at <span class="math inline">\(\mathbf{z}\)</span>,
is represented by
<span class="math inline">\(\Lambda_{i,i+1}\left(s \,|\, \mathbf{z}\right):=\int_{0}^{s}\lambda_{i,i+1}\left(x\,|\, \mathbf{z}\right)\mathrm{d}x\)</span>.
Similarly, we let <span class="math inline">\(\lambda_{i}\left(s\,|\, \mathbf{z}\right)\)</span> represent
the hazard rate of going to any state that can be reached directly from
<span class="math inline">\(i\)</span>, at time <span class="math inline">\(s\)</span> since arrival in state <span class="math inline">\(i\)</span>, for a patient with
covariate vector <span class="math inline">\(\mathbf{z}\)</span>. The cumulative hazard for the same event
between sojourn times <span class="math inline">\(0\)</span> and <span class="math inline">\(s\)</span>, if the patientâ€™s covariate vector
remains constant at <span class="math inline">\(\mathbf{z}\)</span>, is represented by
<span class="math inline">\(\Lambda_{i}\left(s \,|\, \mathbf{z}\right)\)</span>. The expressions
<span class="math inline">\(\hat{\Lambda}_{i}\left(s \,|\, \mathbf{z}\right)\)</span> and
<span class="math inline">\(\hat{\Lambda}_{i,i+1}\left(s \,|\, \mathbf{z}\right)\)</span> denote the
Breslow estimators of the cumulative hazards just defined. In what
follows, all references to probabilities, hazard rates and cumulative
hazards are to be understood as conditional on
<span class="math inline">\(\vec{Z}(u)=\mathbf{z}\,\)</span>, for <span class="math inline">\(u\geq 0\)</span>: this condition is omitted to
simplify the notation.</p>
<p>In <a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a>, the
function <code>probtrans_ebmstate</code> generates a set of state occupation
probability estimates at equally spaced time points:
<span class="math display">\[\begin{aligned}
&amp;\left\lbrace \hat{p}_{0n}\left(k\right)\right\rbrace_{k} :=\left\lbrace \hat{\mathrm{P}}\left[X_{n}=n,\tau_{0,n} &lt; t_{k},\tau_{n,n+1}\geq t_{k}- \tau_{0,n}\,|\, X_{0}=0 \right] \right\rbrace_{k}\;,\; k=0,1,2,...,K\,;\, t_{k}=k\times \Delta t \;.
\end{aligned}\]</span>
The number <span class="math inline">\(K\)</span> of time intervals is <span class="math inline">\(10,000\)</span> by default and <span class="math inline">\(t_{K}\)</span> is a
parameter set by the user. Defining the functions
<span class="math display">\[\begin{aligned}
q_{ij}\left(k\right):=\mathrm{P}\left[X_{j}=j, \tau_{ij}\in \left[t_{k},t_{k+1}\right)\,|\,X_{i}=i\right]
\end{aligned}\]</span>
and
<span class="math display">\[\begin{aligned}
r_{i}\left(k\right):=\mathrm{P}\left[\tau_{i,i+1} &gt; t_{k} \,|\,X_{i}=i\right]\;,
\end{aligned}\]</span>
and the finite difference
<span class="math display">\[\begin{aligned}
    \Delta \hat{\Lambda}_{i,i+1}\left(t_{k}\right):=\hat{\Lambda}_{i,i+1}\left(t_{k+1}\right)-\hat{\Lambda}_{i,i+1}\left(t_{k}\right)\;,
\end{aligned}\]</span>
the algorithm behind <code>probtrans_ebmstate</code> can be described as follows:</p>
<ol type="1">
<li><p>For <span class="math inline">\(j=1,2,...,n\)</span>, compute
<span class="math display" id="eq:est1">\[\begin{aligned}
\label{eq:est1}
\hat{q}_{j-1,j}\left(k\right)&amp;:=\exp \left[-\hat{\Lambda}_{j-1}\left(t_{k}\right)\right]\Delta \hat{\Lambda}_{j-1,j}\left(t_{k}\right)&amp;&amp;
\end{aligned}   \tag{3}\]</span>
for <span class="math inline">\(k=0,1,...,K-1\)</span>.</p></li>
<li><p>For <span class="math inline">\(j=2,3,...,n\)</span>, compute (iteratively)
<span class="math display" id="eq:est2">\[\begin{aligned}
\label{eq:est2}
\hat{q}_{0j}\left(k\right):=&amp;\sum_{l=0}^{k-1} \hat{q}_{j-1,j}\left(k-l-1\right) \hat{q}_{0,j-1} \left(l\right) &amp;&amp;
\end{aligned}   \tag{4}\]</span>
for <span class="math inline">\(k=0,1,...,K-1\)</span>.</p></li>
<li><p>Finally, use the estimates obtained in the last iteration of step 2
to compute
<span class="math display" id="eq:est4">\[\begin{aligned}
\label{eq:est4}
\hat{p}_{0n}\left(k\right):=&amp;\sum_{l=0}^{k-1} \hat{r}_{n}\left(k-l-1\right) \hat{q}_{0,n}\left(l\right)&amp;&amp;
\end{aligned}   \tag{5}\]</span>
for <span class="math inline">\(k=0,1,...,K\)</span>, where
<span class="math inline">\(\hat{r}_{n}\left(\cdot\right):=\exp \left[-\hat{\Lambda}_{n}\left(t_{\scriptscriptstyle\left(\cdot\right)}\right)\right]\,\)</span>.</p></li>
</ol>
<p>Substituting <span class="math inline">\(:=\)</span> for <span class="math inline">\(\approx\)</span> and removing the â€˜hatsâ€™ in definitions
<a href="#eq:est1">(3)</a> to <a href="#eq:est4">(5)</a>, we get the approximate equalities that
justify the algorithm. These approximate equalities are derived in the
Supporting Scripts and Data (file ESM_1.html, section 2).</p>
<p>Apart from <code>probtrans_ebmstate</code>, the function <code>probtrans_fft</code> is also
based on the convolution argument just shown. However, this function
makes use of the convolution theorem, i.e., of the fact that the
convolution of two (vectorized) functions in the time domain is
equivalent to a pointwise product of the same functions in the frequency
domain. The estimation of state occupation probabilities is thus
simplified to
<span class="math display">\[\begin{aligned}
\hat{p}_{0n}:=&amp;\mathcal{F}^{\scriptscriptstyle -1}\left\lbrace \hat{\mathrm q}_{0,1} \boldsymbol{\cdot} \hat{\mathrm q}_{1,2}\boldsymbol{\cdot} \mathrm{...}\boldsymbol{\cdot}\hat{\mathrm q}_{n-1,n}\boldsymbol \cdot \hat{\mathrm r}_{n}\right\rbrace\;,
\end{aligned}\]</span>
where <span class="math inline">\(\mathcal{F}\)</span> denotes the discrete Fourier transform,
<span class="math inline">\(\hat{\mathrm{q}}_{j-1,j}:=\mathcal{F}(\hat{q}_{j-1,j})\)</span> and
<span class="math inline">\(\hat{\mathrm{r}}_{n}:=\mathcal{F}(\hat{r}_{n})\)</span>. Conversion to and from
the frequency domain is carried out using the fast Fourier transform
algorithm implemented in the <code>fft</code> function of the base package <code>stats</code>.
The Supporting Scripts and Data contain a short simulation study
checking that state occupation probabilities can be accurately estimated
with <code>probtrans_ebmstate</code> and <code>probtrans_fft</code> (see file ESM_1.html,
sections 3 and 4).</p>
<p>Figure <a href="#fig:figmssample">2</a> consists of a grid of plots with estimated
curves of state occupation probabilities. It compares, in terms of speed
and accuracy, the estimator in <code>probtrans_fft</code> with an estimator in
<code>mstate::mssample</code> that has the same target, but is simulation-based.
Each plot contains a black curve and a superimposed red curve. The red
curves in any given column of the grid are all based on the same run of
a function: columns 1 to 3 are based on runs of <code>mssample</code> with the
number of samples <span class="math inline">\(n\)</span> equal to <span class="math inline">\(100\)</span>, <span class="math inline">\(1000\)</span> and <span class="math inline">\(10.000\)</span> respectively,
while column 4 is based on a run of <code>probtrans_fft</code>. Each column in the
grid reproduces the same 4 black curves. These are based on a single run
of <code>mssample</code> with <span class="math inline">\(n=100.000\)</span> and serve as benchmark. All function runs
are based on the same input: a set of cumulative transition hazard
estimates for a multi-state model with the â€˜linearâ€™ transition structure
given in the leftmost diagram of figure
<a href="#fig:figtransition-structures">3</a>. Plots in a given row refer to the
same state of the model. The running times on top of each column refer
to the estimation of red curves. The main conclusion suggested by this
analysis of simulated data is that <code>probtrans_fft</code> is as accurate as
<code>mssample</code> with <span class="math inline">\(n=10.000\)</span>, but it is almost 100 times faster (columns 3
and 4). With <span class="math inline">\(n=1000\)</span>, <code>mssample</code> achieves a good approximation to the
true state occupation probabilities, but is still roughly 9 times
slower. The details on how figure <a href="#fig:figmssample">2</a> and its
underlying data were generated are given in the Supporting Scripts and
Data (file ESM_1.html, section 5).</p>
<h4 class="unnumbered" data-number="3.3" id="sec:interval_estimation">Interval estimation</h4>
<p>Under any model estimated by
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> â€“ as in
general under a Bayesian model â€“, one can, if the sample size is large
enough, approximate the posterior by a normal distribution with mean
equal to the maximum <em>a posteriori</em> estimate and covariance matrix equal
to the inverse of the generalised observed Fisher information <span class="citation" data-cites="Gelman2014">(see, for
example, <a href="#ref-Gelman2014" role="doc-biblioref">Gelman et al. 2014 83â€“84</a>)</span>. This approximation has first-order
accuracy and is thus outperformed by Laplaceâ€™s method, which has
second-order accuracy <span class="citation" data-cites="Carlin2009">(<a href="#ref-Carlin2009" role="doc-biblioref">Carlin and Louis 2009 110â€“111</a>)</span>. However, as <span class="citation" data-cites="Carlin2009">Carlin and Louis (<a href="#ref-Carlin2009" role="doc-biblioref">2009 112</a>)</span> observe, â€œfor moderate- to high-dimensional <span class="math inline">\(\boldsymbol\theta\)</span>
(say, bigger than 10), Laplaces method will rarely be of sufficient
accuracy<span class="math display">\[\...\]</span>â€. <span class="citation" data-cites="Carlin2009">Carlin and Louis (<a href="#ref-Carlin2009" role="doc-biblioref">2009 244â€“251</a>)</span> also describe three methods
of interval estimation in empirical Bayes settings, but all of them are
designed for fully parametric models. These reasons, along with the fact
that regularised methods such as the one implemented
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> are
typically used to fit models with more than a dozen covariates, led us
to choose the non-parametric bootstrap as the interval estimation method
in <a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a>. Note
that the non-parametric bootstrap can be given a Bayesian
interpretation. Its interval estimates are approximately the same as
those of a Bayesian model that assumes: a) a multinomial distribution
for the data; and b) a non-informative Dirichlet prior distribution for
the probability assigned to each category in the multinomial
distribution. This is a specific case of the so-called Bayesian
bootstrap <span class="citation" data-cites="Hastie2009">(<a href="#ref-Hastie2009" role="doc-biblioref">Hastie et al. 2009 272</a>)</span>. Further research is needed to determine
the theoretical properties of the non-parametric bootstrap in the
present setting, but this falls beyond the scope of the present paper.
Interval estimates of regression coefficients, cumulative hazards and
state occupation probabilities are implemented in the function
<code>boot_ebmstate</code>.</p>
<h3 data-number="4" id="sec:estimator_performance"><span class="header-section-number">4</span> Estimator performance</h3>
<p>It is a well-documented fact in the statistical literature that standard
least-squares or maximum-likelihood estimators can often be improved by
regularisation or shrinkage <span class="citation" data-cites="Samworth2012">(see, for example, <a href="#ref-Samworth2012" role="doc-biblioref">Samworth 2012</a>)</span>. This
improvement comes about when the model dimensionality is high enough
that the bias introduced by regularisation is outweighed by the
reduction in the estimator variance. In the current setting, one might
therefore ask: what kind of dimensionality does a semi-parametric,
multi-state Cox model need to have to be outperformed by its empirical
Bayes counterpart? A simulation study we carried out offers a tentative
answer to this question, by comparing estimators under both Cox models
for an increasing number of covariates. The study also features a third
method, based on a fully non-parametric model, as a null model method.
This was included to give an idea of how many covariates the empirical
Bayes model can deal with before it becomes no better than a simple
non-regressive model.</p>
<h4 class="unnumbered" data-number="4.1" id="simulation-setup">Simulation setup</h4>
<p>We assessed the performance of all estimators defined by the tuple
<span class="math inline">\(\left[a,m, G, n,p(n)\right]\)</span>, where <span class="math inline">\(a\in \lbrace\)</span>regression
coefficients, relative hazards, state occupation probabilities<span class="math inline">\(\rbrace\)</span>
is the target of estimation, <span class="math inline">\(m\in \lbrace\)</span>standard Cox, empirical Bayes
Cox, null<span class="math inline">\(\rbrace\)</span> is the assumed hazard model, <span class="math inline">\(G \in \lbrace\)</span>linear,
competing risks, â€˜mâ€™ structure<span class="math inline">\(\rbrace\)</span> is the transition structure of
the model (illustrated in figure <a href="#fig:figtransition-structures">3</a>)
and <span class="math inline">\(n\in \lbrace 100,1000\rbrace\)</span> is the number of patients/disease
histories in the training data set; the variable <span class="math inline">\(p\)</span> denotes the number
of coefficients/covariates per transition in the true model and its
range depends on <span class="math inline">\(n\)</span>:
<span class="math inline">\(p\left(100\right) \in \lbrace 10,40,70,100 \rbrace\)</span> whereas
<span class="math inline">\(p\left(100\right) \in \lbrace 10,100,200,300 ,400,500\rbrace\)</span>. By
â€˜relative hazardsâ€™ and â€˜state occupation probabilitiesâ€™, we mean here
the relative transition hazards of an out-of-sample patient, and her
state occupation probabilities at 7 chosen time points. We generated a
batch of 300 independent absolute error observations (â€˜NAâ€™ estimates
included) for each estimator, where each observation is recorded after
training the estimator on a newly simulated data set. Each boxplot in
figures <a href="#fig:figestimator-performance-boxplots-100patients">6</a>
(<span class="math inline">\(n=100\)</span>) and <a href="#fig:figestimator-performance-boxplots-1000patients">7</a>
(<span class="math inline">\(n=1000\)</span>) is based on one of these batches. As all estimators are
<em>vector</em> estimators, each absolute error is actually an <em>average</em>
absolute error, where the average is taken over the components of the
vector.</p>
<p>All training data sets were simulated from clock-reset Cox models. Apart
from <span class="math inline">\(G\)</span> (the model transition structure), <span class="math inline">\(n\)</span> and <span class="math inline">\(p\)</span>, also the true
baseline hazards are held fixed within each batch of 300 training data
sets. The coefficient vectors used in the simulation are always
non-sparse and are scaled by <span class="math inline">\(\sqrt{\frac{10}{p}}\)</span> to keep the
log-hazard variance constant when the dimensionality grows. All
covariates are dichotomous and mutually independent. To compute the
coefficient errors for the non-parametric (null) model method, we think
of it as a degenerate Cox model in which all regression coefficient
estimates are fixed at zero. The estimation of regression coefficients
under the standard Cox and the empirical Bayes Cox models was performed
with <code>survival::coxph</code> and <code>ebmstate::CoxRFX</code> respectively; the
estimation of state occupation probabilities is based on
<code>mstate::probtrans</code> for the null model and on <code>ebmstate::probtrans_fft</code>
for both the standard Cox and the empirical Bayes Cox models.</p>
<p>The reason we did not consider simulation scenarios with more than 500
covariates per transition, in data sets of 1000 patients, was simply
computational cost. For example, generating the data and error
observations for the scenario with <span class="math inline">\(n=1000\)</span>, <span class="math inline">\(p=100\)</span> and <span class="math inline">\(G=\)</span>â€™mâ€™
structure took less than one hour to generate using 20 CPU cores in
parallel; the same scenario but with <span class="math inline">\(p=500\)</span> took 6.5 days using 25 CPU
cores. More details about the simulation setup can be found in the
Supporting Scripts and Data (file ESM_1.html, section 6, subsection
â€˜sample scriptâ€™).</p>
<h4 class="unnumbered" data-number="4.2" id="missing-values">Missing values</h4>
<p>Whenever an estimator was able to compute a valid estimate of its target
for each training data set, i.e., when it did not return any â€˜NAâ€™
estimates, its boxplots are based on 300 valid error observations. This
was always the case with non-parametric estimators: the estimates of
regression coefficients and relative hazards of this type of estimators
are trivial (fixed at zero and one respectively) and hence it is also
straightforward to compute absolute errors. It also happened that
non-parametric estimators of state occupation probabilities had no â€˜NAâ€™
estimates (see file ESM_1.html, section 6, figure 6.3, in the Supporting
Scripts and Data). The situation was similar for the empirical Bayes Cox
model estimators, which showed no more than 5<span class="math inline">\(\%\)</span> missing estimates in
any of the simulation scenarios studied (ibid., figures 6.1 and 6.2).
However, for the standard Cox model ones, the number of â€˜NAâ€™ estimates
depends to a large extent on the number of patients in the data set, as
well as on the dimensionality and transition structure of the model
(figures <a href="#fig:figna-props-100patients-coxph">4</a> and
<a href="#fig:figna-props-1000patients-coxph">5</a>). In data sets of 100
patients, it fares well in models with fewer than 10 covariates per
transition, or in models with up to 40 covariates, if the transition
structure is linear. Otherwise its failure rates range from roughly
25<span class="math inline">\(\%\)</span> to nearly 100<span class="math inline">\(\%\)</span>. In data sets of 1000 patients, the proportion
of â€˜NAâ€™ estimates is never above 10<span class="math inline">\(\%\)</span>, if the transition structure is
linear, but it can climb above 60<span class="math inline">\(\%\)</span> for other transition structures.</p>
<h4 class="unnumbered" data-number="4.3" id="comparison-of-estimators">Comparison of estimators</h4>
<p>With respect to the performance of the three methods studied, the
boxplots in figures
<a href="#fig:figestimator-performance-boxplots-100patients">6</a> and
<a href="#fig:figestimator-performance-boxplots-1000patients">7</a> suggest the
following conclusions:</p>
<ul>
<li><p>As <span class="math inline">\(p/n\)</span> grows, the empirical Bayes estimators quickly outperform
the standard Cox model ones. They already fare substantially better
at <span class="math inline">\(p/n=0.1\)</span> for both <span class="math inline">\(n=100\)</span> and <span class="math inline">\(n=1000\)</span> and for all estimation
targets. At the same time, the relative performance of the empirical
Bayes method with respect to the null model one decreases. At
<span class="math inline">\(p/n=0.5\)</span>, the difference between these two methods is already
rather small for all simulation scenarios.</p></li>
<li><p>The relative performance of the empirical Bayes method with respect
to the null method decreases as the number of co-occurring
transition hazards in the model grows. All other things equal, the
empirical Bayes method has the best performance under the â€˜linearâ€™
structure model, which has no competing transitions; it performs
less well under the â€˜mâ€™ structure transition model, where two
transition hazards can co-occur; and has the worse relative
performances under the â€˜competing risksâ€™ model, where three
transition hazards co-occur. This trend is clearer for <span class="math inline">\(n=100\)</span>
(figure <a href="#fig:figestimator-performance-boxplots-100patients">6</a>)
but can also be detected in the relative hazard errors for <span class="math inline">\(n=1000\)</span>
(figure <a href="#fig:figestimator-performance-boxplots-1000patients">7</a>).
In any case, the empirical Bayes method seems to be far more robust
than the standard Cox model against increases in the number of
co-occurring transition hazards.</p></li>
<li><p>Having as target the regression coefficients or the state occupation
probabilities, instead of relative hazards, makes the empirical
Bayes method better in comparison to the null method. In fact, as
<span class="math inline">\(p/n\)</span> grows, the empirical Bayes method is never outperformed by the
null method except in the estimation of relative hazards.</p></li>
</ul>
<h3 data-number="5" id="survival-analysis-workflow"><span class="header-section-number">5</span> Survival analysis workflow</h3>
<p>The features of <code>mstate</code> were illustrated in <span class="citation" data-cites="Wreede2010">Wreede et al. (<a href="#ref-Wreede2010" role="doc-biblioref">2010</a>)</span> using a simple
workflow. The starting point of this workflow is a data set in â€˜long
formatâ€™. Such data set can be fed into <code>survival::coxph</code> to obtain
estimates of the regression coefficients of a multi-state Cox model. The
resulting model fit object can be passed on to <code>mstate::msfit</code>, along
with a vector of covariates of a particular patient, to get personalised
estimates of the cumulative hazard functions. Finally, state occupation
probabilities for the same patient can be estimated if the object
created by <code>mstate::msfit</code> is fed into <code>mstate::probtrans</code>. In this
section, we describe how
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> extends the
scope of this workflow, i.e., how it uses the packages
<a href="https://CRAN.R-project.org/package=survival"><strong>survival</strong></a> and
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> to generate
estimates under a multi-state empirical Bayes Cox model. A diagram
summarising the extension is shown in figure <a href="#fig:figworkflow">8</a>. In
the <a href="#sec:model_assessment">5.5</a> subsection, we give some
recommendations on how to assess and compare models, but for more
detailed tutorials on how to analyse multi-state data using models
defined by transition hazards, we refer the reader to
<span class="citation" data-cites="Putter2007tutorial">Putter et al. (<a href="#ref-Putter2007tutorial" role="doc-biblioref">2007</a>)</span> and <span class="citation" data-cites="Putter2011tutorial">Putter (<a href="#ref-Putter2011tutorial" role="doc-biblioref">2011</a>)</span>.</p>
<p>The main steps of the
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> workflow are
here illustrated using a data set of patients with myelodysplastic
syndromes (MDS) which has been described and studied in
<span class="citation" data-cites="Papaemmanuil2013">Papaemmanuil et al. (<a href="#ref-Papaemmanuil2013" role="doc-biblioref">2013</a>)</span>. A myelodysplastic syndrome is a form of leukemia in
which the bone marrow is not able to produce enough mature blood cells,
and which sometimes develops into a cancer of white blood cells with a
quick and aggressive progression, i.e., into acute myeloid leukemia
(AML). Figure <a href="#fig:figtrans-diagrams">9</a>a illustrates an illness-death
type model for MDS patients and also gives a breakdown of the number of
transition events. The conversion to a model with a transition structure
that has no cycles (i.e., that can be handled by our convolution-based
estimators) is shown in figure <a href="#fig:figtrans-diagrams">9</a>b. The data
set used for model estimation, obtained after a number of pre-processing
steps, contains the disease history of 576 patients, as well as
measurements on 30 covariates. Of these 30 covariates, 11 are mutation
covariates and the remaining are clinical or demographic (see figure
<a href="#fig:figtrans-diagrams">9</a>c). The running time for the estimation of
relative transition hazards does not exceed 10 seconds in a standard
laptop computer. The same holds for the estimation of cumulative
transition hazards or state occupation probabilities for a given
patient. The complete R code underlying the data analysis in the current
section can be found in the Supporting Scripts and Data (file
ESM_2.html). For running only the R snippets shown below and reproduce
their results, the best option is to use the R script in file ESM_3.R of
the Supporting Scripts and Data.</p>
<h4 class="unnumbered" data-number="5.1" id="input-data">Input data</h4>
<p>Table <a href="#tab:long_format_data">1</a> shows a fragment of the MDS data
set. The data is in â€˜long formatâ€™, which means that each row refers to a
period of risk for a given transition and patient. For example, row <span class="math inline">\(i\)</span>
tells us that, at time <code>Tstart[i]</code>, patient <code>id[i]</code> entered state
<code>from[i]</code>, and thereby began to be at risk for transition <code>trans[i]</code>,
i.e., at risk of going from state <code>from[i]</code> to state <code>to[i]</code>. If the
first transition of patient <code>id[i]</code> after time <code>Tstart[i]</code> occurs before
the last follow-up time for this patient, <code>Tstop[i]</code> records the time of
this transition (regardless of whether the patient moved to state
<code>to[i]</code> or not). Otherwise, <code>Tstop[i]</code> is set to the last follow-up
time. The value of <code>status[i]</code> is set to 1 if and only if the first
transition of patient <code>id[i]</code> after <code>Tstart[i]</code> is to state <code>to[i]</code> and
occurs before the last follow-up (otherwise it is set to 0). The value
of <code>time[i]</code> is defined simply as <code>Tstop[i]</code><span class="math inline">\(-\)</span><code>Tstart[i]</code>, and
<code>strata[i]</code> is the stratum of the baseline hazard for transition
<code>trans[i]</code> (more about this variable in the following section). For <code>x</code>
<span class="math inline">\(\in \left\lbrace \right.\)</span> <code>ASXL1</code>, <code>DNMT3A</code>,
<span class="math inline">\(\dots \left. \right \rbrace\)</span>, <code>x[i]</code> denotes the level of covariate <code>x</code>
between <code>Tstart[i]</code> and <code>Tstop[i]</code> in patient <code>id[i]</code>. (In the MDS data
set, we assume that the relative hazard of a patient is determined by
her covariate vector at <span class="math inline">\(t=0\)</span>, i.e., we assume all covariates to be
time-fixed.) If a patient enters a new state, and this state
communicates directly with <span class="math inline">\(n\)</span> other states, then, as long as the
patient actually spends time in the new state (i.e.Â the time of
transition is not the same as the last follow-up time), <span class="math inline">\(n\)</span> rows must be
added to the data set, with each row corresponding to a different
possible transition.</p>
<p>From table <a href="#tab:long_format_data">1</a>, we know that patient 77
entered state 1 (â€˜MDSâ€™) at time 0 and remained in this state until time
2029, when she moved to state 3 (â€˜death before AMLâ€™). There are no rows
to describe the evolution of patient 77 after entering state 3, as this
state is an absorbing state. As to patient 78, she remained in state 1
until time 332, and moved from there to state 2 (â€˜AMLâ€™). She lived with
AML for 1117 days and moved to state 4 (â€˜death after AMLâ€™) at time 1449.</p>
<div id="tab:long_format_data">
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>id  from to trans Tstart Tstop time status  strata ASXL1 DNMT3A [...]  </span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="dv">77</span>     <span class="dv">1</span>  <span class="dv">2</span>     <span class="dv">1</span>      <span class="dv">0</span>  <span class="dv">2029</span> <span class="dv">2029</span>      <span class="dv">0</span>       <span class="dv">1</span>     <span class="dv">0</span>      <span class="dv">0</span>    .</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="dv">77</span>     <span class="dv">1</span>  <span class="dv">3</span>     <span class="dv">2</span>      <span class="dv">0</span>  <span class="dv">2029</span> <span class="dv">2029</span>      <span class="dv">1</span>       <span class="dv">2</span>     <span class="dv">0</span>      <span class="dv">0</span>    .</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="dv">78</span>     <span class="dv">1</span>  <span class="dv">2</span>     <span class="dv">1</span>      <span class="dv">0</span>   <span class="dv">332</span>  <span class="dv">332</span>      <span class="dv">1</span>       <span class="dv">1</span>     <span class="dv">1</span>      <span class="dv">0</span>    .</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="dv">78</span>     <span class="dv">1</span>  <span class="dv">3</span>     <span class="dv">2</span>      <span class="dv">0</span>   <span class="dv">332</span>  <span class="dv">332</span>      <span class="dv">0</span>       <span class="dv">2</span>     <span class="dv">1</span>      <span class="dv">0</span>    .</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="dv">78</span>     <span class="dv">2</span>  <span class="dv">4</span>     <span class="dv">3</span>    <span class="dv">332</span>  <span class="dv">1449</span> <span class="dv">1117</span>      <span class="dv">1</span>       <span class="dv">3</span>     <span class="dv">1</span>      <span class="dv">0</span>    .</span></code></pre></div>
<p>Table 1: A 5-row fragment of the MDS data set (in long format)</p>
</div>
<h4 class="unnumbered" data-number="5.2" id="sec:fit_bayes_cox_model">Fitting an empirical Bayes Cox model</h4>
<p>Once the data is in â€˜long formatâ€™, the estimation of an empirical Bayes
model can be carried out using the function <code>CoxRFX</code>. A simple example
of the first argument of <code>CoxRFX</code>, denoted â€˜<code>Z</code>â€™, is a data frame
gathering the <code>trans</code>, <code>strata</code> and covariate columns of the data in
long format:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>outcome_covs <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;id&quot;</span>,<span class="st">&quot;from&quot;</span>,<span class="st">&quot;to&quot;</span>,<span class="st">&quot;trans&quot;</span>,<span class="st">&quot;Tstart&quot;</span>,<span class="st">&quot;Tstop&quot;</span>,<span class="st">&quot;time&quot;</span>,<span class="st">&quot;status&quot;</span>,</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;strata&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>Z <span class="ot">&lt;-</span> mstate_data[<span class="sc">!</span><span class="fu">names</span>(mstate_data) <span class="sc">%in%</span> outcome_covs]</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co">#(`mstate_data&#39; has the data in long format) </span></span></code></pre></div>
<p>The <code>strata</code> column determines which baseline hazard functions are
assumed to be equal. In table <a href="#tab:long_format_data">1</a>, each
transition is assumed to have a (potentially) different baseline hazard.
The modelâ€™s assumptions regarding how covariates affect the hazard are
reflected on the format of the covariate columns of <code>Z</code>. When the <code>Z</code>
argument is the one created in the previous block of code, <code>CoxRFX</code>
returns a single regression coefficient estimate for each covariate. In
other words, the impact of any covariate is assumed to be the same for
every transition.</p>
<p>There are however ways of relaxing this assumption. One can replace the
<code>ASXL1</code> column in Z (or any other covariate column) by several
â€˜type-specificâ€™ <code>ASXL1</code> columns: the <code>ASXL1</code> column specific for type
<span class="math inline">\(i\)</span> would show the mutation status of <code>ASXL1</code> in rows belonging to
transition of type <span class="math inline">\(i\)</span>, and show zero in all other rows. This would
force <code>CoxRFX</code> to estimate a (potentially) different <code>ASXL1</code> coefficient
for each transition type. This process of covariate expansion by type
can be based on any partition of the set of transitions. When each type
corresponds to a single transition, we refer to it simply as â€˜covariate
expansion by transitionâ€™. The output shown below illustrates the effect
of expanding the covariates in â€˜mstate_dataâ€™ by transition.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Columns `id&#39; and `trans&#39; from `mstate_data&#39; together with the first</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># two expanded covariates (patients 77 and 78):</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    id trans ASXL1<span class="fl">.1</span> ASXL1<span class="fl">.2</span> ASXL1<span class="fl">.3</span> DNMT3A<span class="fl">.1</span> DNMT3A<span class="fl">.2</span> DNMT3A<span class="fl">.3</span>  [...]  </span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="dv">77</span>     <span class="dv">1</span>       <span class="dv">0</span>       <span class="dv">0</span>       <span class="dv">0</span>        <span class="dv">0</span>        <span class="dv">0</span>        <span class="dv">0</span>     .</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="dv">77</span>     <span class="dv">2</span>       <span class="dv">0</span>       <span class="dv">0</span>       <span class="dv">0</span>        <span class="dv">0</span>        <span class="dv">0</span>        <span class="dv">0</span>     .</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>    <span class="dv">78</span>     <span class="dv">1</span>       <span class="dv">1</span>       <span class="dv">0</span>       <span class="dv">0</span>        <span class="dv">0</span>        <span class="dv">0</span>        <span class="dv">0</span>     .</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>    <span class="dv">78</span>     <span class="dv">2</span>       <span class="dv">0</span>       <span class="dv">1</span>       <span class="dv">0</span>        <span class="dv">0</span>        <span class="dv">0</span>        <span class="dv">0</span>     .</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>    <span class="dv">78</span>     <span class="dv">3</span>       <span class="dv">0</span>       <span class="dv">0</span>       <span class="dv">1</span>        <span class="dv">0</span>        <span class="dv">0</span>        <span class="dv">0</span>     .</span></code></pre></div>
<p>The example code given below shows how to use
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> to expand
covariates by transition and how to create a <code>Z</code> argument that makes
<code>CoxRFX</code> estimate a regression coefficient for each covariate for
transitions 1 and 2, and assume a fully non-parametric hazard for
transition 3.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># To expand covariates by transition using mstate::expand.covs, </span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># first set the class of `mstate_data&#39; as</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(mstate_data) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;data.frame&quot;</span>,<span class="st">&quot;msdata&quot;</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># then add the transition matrix as attribute:</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(mstate_data,<span class="st">&quot;trans&quot;</span>) <span class="ot">&lt;-</span> tmat </span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="co">#(`tmat&#39; is the output of mstate::transMat)</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Expand covariates by transition:</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>covariates_expanded_123 <span class="ot">&lt;-</span> mstate<span class="sc">::</span><span class="fu">expand.covs</span>(</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    mstate_data,</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">covs =</span> <span class="fu">names</span>(mstate_data)[<span class="sc">!</span> <span class="fu">names</span>(mstate_data) <span class="sc">%in%</span> outcome_covs],</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">append =</span> F</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="co"># remove all covariates for transition 3 from `covariates_expanded_123&#39;</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="co"># to fit a fully non-parametric model on this transition:</span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>covariates_expanded_12 <span class="ot">&lt;-</span> covariates_expanded_123[</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>    <span class="sc">!</span><span class="fu">grepl</span>(<span class="st">&quot;.3&quot;</span>,<span class="fu">names</span>(covariates_expanded_123),<span class="at">fixed =</span> T)</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a><span class="co">#argument `Z&#39; of coxrfx</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>Z_12 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(covariates_expanded_12,<span class="at">strata =</span> mstate_data<span class="sc">$</span>trans,</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>                   <span class="at">trans =</span> mstate_data<span class="sc">$</span>trans)</span></code></pre></div>
<p>The second argument of <code>CoxRFX</code> (â€˜<code>surv</code>â€™) is a survival object that can
easily be built by feeding the outcome variable columns of the data to
the function <code>Surv</code> (from the package
<a href="https://CRAN.R-project.org/package=survival"><strong>survival</strong></a>). Whether
<code>CoxRFX</code> fits a clock-forward model or a clock-reset model depends on
the kind of survival object:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#argument `surv&#39; for a  clock-forward model</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>surv <span class="ot">&lt;-</span> <span class="fu">Surv</span>(mstate_data<span class="sc">$</span>Tstart,mstate_data<span class="sc">$</span>Tstop,mstate_data<span class="sc">$</span>status)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co">#argument `surv&#39; for a clock-reset model</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>surv <span class="ot">&lt;-</span> <span class="fu">Surv</span>(mstate_data<span class="sc">$</span>time,mstate_data<span class="sc">$</span>status)</span></code></pre></div>
<p>The argument <code>groups</code> of <code>CoxRFX</code> is a vector whose length equals the
number of covariates in the data. In other words, the length of <code>groups</code>
is <code>ncol(Z)-2</code>, since the argument <code>Z</code> must include both the covariate
data and the <code>strata</code> and <code>trans</code> columns. If, for <span class="math inline">\(i \neq j\)</span>,
<code>groups[i]</code>=<code>groups[j]</code> <span class="math inline">\(=\text{`foo&#39;}\)</span>, this means that the regression
coefficients of the <span class="math inline">\(i^{th}\)</span> and <span class="math inline">\(j^{th}\)</span> covariates of <code>Z</code> both belong
to a group named â€˜fooâ€™ of coefficients with the same prior. For the <code>Z</code>
object built above, the <code>groups</code> argument created in the following block
of code embodies the assumption that all coefficients associated with a
given transition have the same prior distribution. The final line of
code fits the empirical Bayes model.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co">#argument `groups&#39; of coxrfx</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>groups_12 <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">rep</span>(<span class="st">&quot;group&quot;</span>,<span class="fu">ncol</span>(Z)<span class="sc">-</span><span class="dv">2</span>),<span class="fu">c</span>(<span class="st">&quot;_1&quot;</span>,<span class="st">&quot;_2&quot;</span>))</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co">#fit random effects model</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>model_12 <span class="ot">&lt;-</span> <span class="fu">CoxRFX</span>(Z_12,surv,groups_12,tmat)</span></code></pre></div>
<p>Figure <a href="#fig:figcoef-plots">10</a> shows regression coefficient point
estimates for a clock-reset, empirical Bayes model fitted with the code
above. Also shown are 95% non-parametric bootstrap confidence intervals
computed using <code>ebmstate::boot_ebmstate</code>. The <span class="math inline">\(x\)</span>-axis scale is
logarithmic to allow estimates to be read as relative hazards more
easily. For example, a mutation in <em>RUNX1</em> is associated with a twofold
increase in the hazard of progression from MDS to AML, and treatment
centre 4 is associated with a 3-fold increase in the hazard of dying
before progressing to AML, when compared to the baseline value of
â€˜treatment centreâ€™ (treatment centre = 2 or 5). In covariates that have
been log-transformed (age, platelet count and neutrophil count) or
logit-transformed (proportions of myeloblasts and ring sideroblasts in
the bone marrow), the interpretation of estimates is different. For
example, an increase in age by a factor of <span class="math inline">\(e\)</span> (<span class="math inline">\(\approx 2.72\)</span>) almost
triples the hazard of dying before AML; the same increase in the ratio
<span class="math inline">\(bm\_blasts/(1-bm\_blasts)\)</span> (where <em>bm_blasts</em> is the proportion of
myeloblasts in the bone marrow) is associated with an increment in the
hazard of dying before AML of approximately <span class="math inline">\(16\%\)</span>.</p>
<h4 class="unnumbered" data-number="5.3" id="sec:computing_cumulative_hazards">Computing cumulative transition hazard estimates</h4>
<p>The function <code>msfit_generic</code> is the generic function in
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> that
computes cumulative transition hazards for a given set of covariate
values and an estimated Cox model. It calls a different method according
to the class of its <code>object</code> argument. The default method corresponds to
the original <code>msfit</code> function of the
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> package and is
appropriate for objects of class <code>coxph</code>, i.e., objects that contain the
fit of a Cox model with fixed effects. The other available method for
<code>msfit_generic</code>, <code>msfit_generic.coxrfx</code>, is just the original <code>msfit</code>
function, (slightly) adapted to deal with objects generated by <code>CoxRFX</code>.
Quite importantly, <code>msfit_generic.coxrfx</code> does not allow the variance of
the cumulative hazards to be computed, as this computation relies on
asymptotic results which may not be valid for an empirical Bayes model.
As a result, it only has two other arguments apart from the object of
class <code>coxrfx</code>: a data frame with the covariate values of the patient
whose cumulative hazards we want to compute; and a transition matrix
describing the states and transitions in the model (such as the one that
can be generated using <code>transMat</code> from the package
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a>). The following
block of code exemplifies how these objects can be built and generates
the <code>msfit</code> object containing the cumulative transition hazard estimates
for a sample patient. Note that the object with the patient data must
include a row for each transition, as well as a column specifying the
transition stratum of each row of covariates.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Build `patient_data&#39; data frame with the covariate values for which </span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># cumulative hazards are to be computed (covariate values of patient 78):</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>patient_data <span class="ot">&lt;-</span> mstate.data[mstate.data<span class="sc">$</span>id <span class="sc">==</span> <span class="dv">78</span>,,drop <span class="ot">=</span> F][<span class="fu">rep</span>(<span class="dv">1</span>,<span class="dv">3</span>),]</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>patient_data<span class="sc">$</span>strata <span class="ot">&lt;-</span> patient_data<span class="sc">$</span>trans <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>patient_data <span class="ot">&lt;-</span> mstate<span class="sc">::</span><span class="fu">expand.covs</span>(</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>    patient_data,</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">covs =</span> <span class="fu">names</span>(patient_data)[ <span class="sc">!</span> <span class="fu">names</span>(patient_data) <span class="sc">%in%</span> outcome_covs],</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">append =</span> T</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>patient_data <span class="ot">&lt;-</span> patient_data[ <span class="sc">!</span> <span class="fu">grepl</span>(<span class="st">&quot;.3&quot;</span>,<span class="fu">names</span>(patient_data),<span class="at">fixed =</span> T)]</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="co"># The `patient_data&#39; data frame has only 3 rows (one for each transition).</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="co"># The output below shows its `id&#39; and `trans&#39; columns</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="co"># and expanded covariates ASXL1 and DNMT3A:</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>    id trans ASXL1<span class="fl">.1</span> ASXL1<span class="fl">.2</span> DNMT3A<span class="fl">.1</span> DNMT3A<span class="fl">.2</span>  [...]  </span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>    <span class="dv">78</span>     <span class="dv">1</span>       <span class="dv">1</span>       <span class="dv">0</span>        <span class="dv">0</span>        <span class="dv">0</span>     .</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>    <span class="dv">78</span>     <span class="dv">2</span>       <span class="dv">0</span>       <span class="dv">1</span>        <span class="dv">0</span>        <span class="dv">0</span>     .</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>    <span class="dv">78</span>     <span class="dv">3</span>       <span class="dv">0</span>       <span class="dv">0</span>        <span class="dv">0</span>        <span class="dv">0</span>     .</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a><span class="co"># compute cumulative hazards</span></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>msfit_object_12 <span class="ot">&lt;-</span> <span class="fu">msfit_generic</span>(model_12,patient_data,tmat)</span></code></pre></div>
<p>Figure <a href="#fig:figpatient78-cumhaz">11</a> shows three plots of estimated
cumulative transition hazards for the sampled patient, one for each
transition in the model, along with <span class="math inline">\(95\%\)</span> non-parametric bootstrap
confidence intervals (computed with <code>ebmstate::boot_ebmstate</code>).
Throughout the plotted period, the â€˜slopeâ€™ of the cumulative hazard
(i.e., the hazard rate) for the MDS to AML transition is lower than the
one for the MDS to death transition, and this in turn is lower than the
one for the AML to death transition. It should be recalled that the
cumulative hazard estimate is strictly non-parametric for this last
transition, i.e., it is the same for all patients. The central plot of
figure <a href="#fig:figpatient78-cumhaz">11</a> suggests that, as time since
diagnosis goes by, the hazard of dying in MDS increases (possibly an
effect of age). On the other hand, the hazard of dying in AML seems to
decrease (slightly) with time (rightmost plot). Conclusions regarding
the evolution of the AML hazard are hard to draw, since the confidence
intervals for the corresponding cumulative hazard curve are very wide
(leftmost plot).</p>
<p>If an object generated by <code>msfit_generic</code> is fed to <code>plot</code>, and the
package <a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> is
loaded, the method <code>mstate:::plot.msfit</code> will be called. This is an
efficient way of automatically plotting the cumulative hazard estimates
for all transitions, but confidence interval lines (separately
estimated) cannot be added.</p>
<h4 class="unnumbered" data-number="5.4" id="sec:computing_transition_probs">Computing state occupation probability estimates</h4>
<p>The functions <code>probtrans_mstate</code>, <code>probtrans_ebmstate</code> and
<code>probtrans_fft</code> compute estimates of state occupation probabilities for
a given <code>msfit</code> object. All three functions generate objects of class
<code>probtrans</code> that can be fed to the <code>plot.probtrans</code> method from the
package <a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a>. The
first of these functions should only be used for clock-forward models,
as it relies on product-limit calculations. It calls the method
<code>probtrans_mstate.default</code>, if the <code>msfit</code> object was generated by
<code>msfit_generic.default</code>, or the method <code>probtrans_mstate.coxrfx</code>, if it
was generated by <code>msfit_generic.coxrfx</code>. Both methods are identical to
the function <code>probtrans</code> in the
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> package, with
the reserve that <code>probtrans_mstate.coxrfx</code> does not allow the
computation of the variances or covariances of the state occupation
probability estimator.</p>
<p>The functions <code>probtrans_ebmstate</code> and <code>probtrans_fft</code> are the functions
in <a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> for the
computation of state occupation probability estimates under clock-reset
models with a transition structure that has no cycles. When using
<code>probtrans_fft</code> (the faster, but somewhat less stable, of these two
functions), three arguments must be supplied: the initial state of the
process whose state occupation probabilities one wishes to compute, the
<code>msfit</code> object, and the upper time limit for the generation of estimates
(<code>max_time</code>). Both functions are based on a discrete-time approximation
to a series of convolutions. The default argument <code>nr_steps</code> controls
the number of (equally spaced) time steps used in this approximation.
The arguments <code>max_time</code> and <code>nr_steps</code> should be increased until the
estimated curves become stable.</p>
<p>The following line of code computes point estimates of state occupation
probabilities for the sample patient.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>probtrans_object_12 <span class="ot">&lt;-</span> <span class="fu">probtrans_fft</span>(<span class="st">&quot;MDS&quot;</span>,msfit_object_12, <span class="at">max_time =</span> <span class="dv">4000</span>)</span></code></pre></div>
<p>Estimates are shown in figure <a href="#fig:figpatient78-transProbs">12</a>, along
with <span class="math inline">\(95\%\)</span> non-parametric, bootstrap confidence intervals. For this
particular patient, the estimated probability of being dead after AML
remains below 0.4 throughout a period of 10 years from the MDS
diagnosis; if the patient does reach AML, death is expected to happen
quickly thereafter, as reflected in the very low estimates for the
probability of being in AML at any point in time. The following block of
code shows how to compute confidence intervals with <code>boot_ebmstate</code>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Creating the object arguments for boot_ebmstate()</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># `groups&#39; arguments was already created, but we need to add names to it</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(groups_12) <span class="ot">&lt;-</span> <span class="fu">names</span>(covariates_expanded_12)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="co"># `mstate_data_expanded&#39; argument (similar to `covariates_expanded&#39; but</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co"># including outcome variables)</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>mstate_data_expanded <span class="ot">&lt;-</span> <span class="fu">cbind</span>(</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  mstate_data[<span class="fu">names</span>(mstate_data) <span class="sc">%in%</span> outcome_covs],</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  covariates_expanded_12</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co"># create the non-parametric bootstrap confidence intervals</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>boot_ebmstate_object <span class="ot">&lt;-</span> <span class="fu">boot_ebmstate</span>(</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">mstate_data =</span> mstate_data_expanded,</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">which_group =</span> groups_12,</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">min_nr_samples =</span> <span class="dv">100</span>,</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">patient_data =</span> patient_data,</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">tmat =</span> tmat,</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">initial_state =</span> <span class="st">&quot;MDS&quot;</span>,</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">time_model =</span> <span class="st">&quot;clockreset&quot;</span>,</span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>  <span class="at">input_file =</span> <span class="cn">NULL</span>,</span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">coxrfx_args =</span> <span class="fu">list</span>(<span class="at">max.iter =</span> <span class="dv">200</span>),</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">probtrans_args =</span> <span class="fu">list</span>(<span class="at">max_time =</span> <span class="dv">4000</span>)</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<h4 class="unnumbered" data-number="5.5" id="sec:model_assessment">Model assessment</h4>
<p>For any model fitted with
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a>, two
performance metrics can be easily computed: the <em>concordance</em> statistic
(<span class="citation" data-cites="harrell1982evaluating">(<a href="#ref-harrell1982evaluating" role="doc-biblioref">Harrell et al. 1982</a>)</span>; see also the help page of
<code>survival::concordance</code> for the definition of concordance) and the
<em>Bayesian Information Criterion</em> (BIC) score <span class="citation" data-cites="schwarz1978estimating">(<a href="#ref-schwarz1978estimating" role="doc-biblioref">Schwarz 1978</a>)</span>.
As an example of how these two metrics can be obtained and used for
model comparison, suppose we wish to compare â€˜model_12â€™ fitted above â€“
which consists of a Cox regression including all covariates for
transitions 1 and 2 and a fully non-parametric model for transition 3 â€“
with a model that combines Cox regressions of all covariates for each of
the three transitions (denoted â€˜model_123â€™ below). The following code
snippet shows how to fit this second model.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># arguments `groups&#39; and `Z&#39; for fitting a Cox regression model on all transitions</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>Z_123 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    covariates_expanded_123,</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">strata =</span> mstate_data<span class="sc">$</span>trans,</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">trans =</span> mstate_data<span class="sc">$</span>trans</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>groups_123 <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="fu">rep</span>(<span class="st">&quot;group&quot;</span>, <span class="fu">ncol</span>(Z_123) <span class="sc">-</span> <span class="dv">2</span>), <span class="fu">c</span>(<span class="st">&quot;_1&quot;</span>, <span class="st">&quot;_2&quot;</span>, <span class="st">&quot;_3&quot;</span>))</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit a Cox regression model for all transitions</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>model_123 <span class="ot">&lt;-</span> <span class="fu">CoxRFX</span>(<span class="at">Z =</span> Z_123, <span class="at">surv =</span> surv, <span class="at">groups =</span> groups_123)</span></code></pre></div>
<p>Running the <code>concordance</code> function in the
<a href="https://CRAN.R-project.org/package=survival"><strong>survival</strong></a> package for
each model yields the following output:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">concordance</span>(model_12)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    Call<span class="sc">:</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">concordance.coxph</span>(<span class="at">object =</span> model_12)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    n<span class="ot">=</span> <span class="dv">1210</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    Concordance<span class="ot">=</span> <span class="fl">0.8131</span> se<span class="ot">=</span> <span class="fl">0.01314</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>                concordant discordant tied.x tied.y tied.xy</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    strata<span class="ot">=</span><span class="dv">1</span>      <span class="dv">18040</span>       <span class="dv">2783</span>      <span class="dv">0</span>      <span class="dv">1</span>       <span class="dv">0</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>    strata<span class="ot">=</span><span class="dv">2</span>      <span class="dv">37919</span>       <span class="dv">9678</span>      <span class="dv">0</span>      <span class="dv">7</span>       <span class="dv">0</span></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    strata<span class="ot">=</span><span class="dv">3</span>          <span class="dv">0</span>          <span class="dv">0</span>   <span class="dv">1052</span>      <span class="dv">0</span>       <span class="dv">4</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">concordance</span>(model_123)</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>    Call<span class="sc">:</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">concordance.coxph</span>(<span class="at">object =</span> model_123)</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>    n<span class="ot">=</span> <span class="dv">1210</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>    Concordance<span class="ot">=</span> <span class="fl">0.8168</span> se<span class="ot">=</span> <span class="fl">0.01312</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>                concordant discordant tied.x tied.y tied.xy</span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>    strata<span class="ot">=</span><span class="dv">1</span>      <span class="dv">18041</span>       <span class="dv">2782</span>      <span class="dv">0</span>      <span class="dv">1</span>       <span class="dv">0</span></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>    strata<span class="ot">=</span><span class="dv">2</span>      <span class="dv">37920</span>       <span class="dv">9677</span>      <span class="dv">0</span>      <span class="dv">7</span>       <span class="dv">0</span></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>    strata<span class="ot">=</span><span class="dv">3</span>        <span class="dv">784</span>        <span class="dv">268</span>      <span class="dv">0</span>      <span class="dv">4</span>       <span class="dv">0</span></span></code></pre></div>
<p>The output shows that modelling transition 3 with a Cox model, instead
of a fully parametric one, has a negligible impact on the overall
concordance. However, this is due to the fact that there are far fewer
observations for this transition. The concordance for transition 3 only,
which corresponds to strata 3, is 0.5 under the fully parametric model
(i.e., all patients are assigned the same transition hazard) and
considerably higher under the Cox regression (<span class="math inline">\(784/(784+268)=0.75\)</span>).
Ideally, the comparison of models of different complexity should be
carried out on a test sample rather than on the training data. For this
purpose, the test data can be input into to the <code>concordance</code> function
(argument <code>newdata</code>). However, in the present case, only 61 patients
were ever at risk of dying with AML (i.e.Â of undergoing transition 3),
and of these only 41 actually died, so we might prefer to keep all
patients in the training data, rather than saving a fraction of them for
testing purposes. Such an option will yield more accurate coefficient
estimates, at the expense of not allowing the computation of unbiased
estimates of model performance. If the goal is only to compare models,
we can make do without test data, by using an information score that
penalises model complexity, such as the BIC. To facilitate model
comparison, the BIC score is one of the attributes of the model fit
object:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> model_12<span class="sc">$</span>BIC</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>    [<span class="dv">1</span>] <span class="fl">2508.37</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> model_123<span class="sc">$</span>BIC</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    [<span class="dv">1</span>] <span class="fl">2483.49</span></span></code></pre></div>
<p>The best model is the one with the lowest score, so the choice of
â€˜model_123â€™ is confirmed.</p>
<h3 data-number="6" id="discussion"><span class="header-section-number">6</span> Discussion</h3>
<p>We have shown that
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> is suitable
for higher-dimensional, multi-state survival analysis, and that it is
both efficient and easy-to-use. To a significant extent, the
user-friendliness of
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> stems from
the fact that it was not built â€˜from the ground upâ€™. Instead, we
produced a package that is more easily accessible to the many users of
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> by taking
advantage of whichever features of this package were useful to our
method and by eliminating redundancies. The connection between
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> and
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> is based on the
fact that the function <code>CoxRFX</code> takes the same type of input and
produces the same type of output as <code>coxph</code> from the package <code>survival</code>,
and the function <code>probtrans_fft</code> (or <code>probtrans_ebmstate</code>) has the same
type of input and output as <code>probtrans</code> from
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> (as shown in
figure <a href="#fig:figworkflow">8</a>).</p>
<p>We also sought to improve our packageâ€™s user-friendliness by making it
as efficient as possible. The reduction of computational cost is based
on two features. First, our empirical Bayes method relies on an
expectation-maximisation algorithm that estimates both the parameters
and the hyper-parameters of the model, i.e., no further tuning of the
model is required. Second, in
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a>, the
computation of state occupation probability estimates relies on
analytical results rather than on simulation: not only for clock-forward
models, where we import from
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> a product-limit
estimator, but also for clock-reset models, where we implement our own
estimator based on a convolution argument and the fast Fourier
transform.</p>
<p>To our knowledge,
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> is the first
R package to put together a framework for multi-state model estimation
that is complete and suitable for higher-dimensional data. It does so by
implementing point and interval estimators of regression coefficients,
cumulative transition hazards and state occupation probabilities, under
regularised multi-state Cox models. In section
<a href="#sec:estimator_performance">4</a>, the results of the simulation study
suggest that for data sets with 100 patients or more and a ratio of <span class="math inline">\(p\)</span>
(patients) to <span class="math inline">\(n\)</span> (coefficients per transition) greater than 0.1, the
standard Cox model estimator is clearly outperformed by the empirical
Bayes one when it comes to the estimation of relative hazards and state
occupation probabilities of an out-of-sample patient, or the regression
coefficients of the model. However, the same study suggests that using
an empirical Bayes method instead of a fully non-parametric one is of
limited or no value in settings where <span class="math inline">\(p/n \geq 1\)</span>. This loss of
usefulness can already happen for <span class="math inline">\(p/n\leq 1/2\)</span> when it comes to the
estimation of the relative hazards of an out-of-sample patient,
especially for transition structures with multiple competing
transitions.</p>
<p>As mentioned in previous sections,
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> imports a
product-limit estimator from
<a href="https://CRAN.R-project.org/package=mstate"><strong>mstate</strong></a> that targets the
state occupation probabilities of patients with <em>time-fixed</em> covariate
vectors. However, these estimators are extendible to models with
time-dependent covariates, as long as these are external and the
estimates are conditional on specific covariate paths <span class="citation" data-cites="Aalen2008">(<a href="#ref-Aalen2008" role="doc-biblioref">Aalen et al. 2008 142</a>)</span>. For piecewise constant covariates, it is likely that such an
adaptation could be obtained by combining transition probability
estimates obtained for each period in which the covariates are fixed.
While no significant theoretical obstacles are foreseen in this matter,
the computer implementation for more than a single piecewise constant
covariate is likely to be a laborious task. We have left it therefore
for future work.</p>
<h3 class="unnumbered" id="acknowledgements">Acknowledgements</h3>
<p>The authors are supported by grant NNF17OC0027594 from the Novo Nordisk
Foundation. We thank an anonymous reviewer for their constructive
comments and helpful suggestions which led to a much-improved
manuscript.</p>
<h3 class="unnumbered" id="supporting-scripts-and-data">Supporting Scripts and Data</h3>
<p>In the supporting Scripts and Data, the file <code>ESM_1.html</code> contains
additional simulation results and theoretical demonstrations. Additional
details on the analysis of the MDS data set are given in the file
<code>ESM_2.html</code>. The MDS data set is in files <code>MDS.TPD.20Nov2012.csv</code> and
<code>mds.paper.clin.txt</code>. The file <code>ESM_3.R</code> contains a simplified R script
to run the code snippets in the present paper. The
<a href="https://CRAN.R-project.org/package=ebmstate"><strong>ebmstate</strong></a> package is
available on CRAN.</p>
<h3 data-number="7" id="conflict-of-interest"><span class="header-section-number">7</span> Conflict of interest</h3>
<p>The authors have declared no conflict of interest.</p>
<p><strong>Figures</strong></p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figpackage-summary-figure"></span>
<img src="figures/package_summary_figure.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 1: Summary of inputs and outputs of the package ebmstate. The input data set should be one that violates the assumption â€“ commonly used in survival analysis â€“ that the number of observations is much larger than the number of parameters to be estimated (a genomic-clinical data set is shown as a typical example). The input model is a multi-state Cox model defined by a transition structure and a prior distribution on the regression coefficients. This prior distribution is defined by partitioning the vector of regression coefficients into groups of regression coefficients, with each group having its own Gaussian prior with undetermined mean and variance. The outputs of ebmstate include estimates of the relative transition hazards associated with each covariate, as well as estimates of the probability that a specific patient (with specific covariate measurements) has of occupying each state of the model over some time period. Estimates of cumulative transition hazards are omitted from the figure.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figmssample"></span>
<img src="figures/mssample_and_probtrans_fft.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 2: Comparison of running times and estimation accuracy of mssample and probtrans_fft. Each plot in the grid shows two estimated curves of state occupation probabilities. The black curves are based on a single run of mstate::mssample with n=100.000 observations (approximately 17 minutes of running time) and are the same across columns. They serve as benchmark for precision assessment. In columns 1 to 3 of the grid, the superimposed red curves are based on a run of mssample with respectively 100, 1000, and 10.000 observations. In the rightmost column, the red curves are based on a run of probtrans_fft. All functions have as input the same set of cumulative transition hazards. These were estimated using a non-parametric multi-state model and a data set of 1000 patients generated according to a clock-reset Cox model with a â€˜linearâ€™ transition structure (leftmost diagram of figure 3). Plots in the same row refer to the same state of the model, while those in the same column refer to the same run of a function. Running times and, where appropriate, number of simulations (n) are given on top of each column.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figtransition-structures"></span>
<img src="figures/transition_structures.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 3: Model transition structures. We studied the performance of Cox model estimators, empirical Bayes Cox model estimators and fully non-parametric estimators with respect to these 3 transition structures.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figna-props-100patients-coxph"></span>
<img src="figures/na_props_100patients_coxph.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 4: Proportions of valid, infinite and missing (â€˜NAâ€™) estimates for the standard Cox model estimators in the simulation study of figure 6 (100 patients per simulated data set).
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figna-props-1000patients-coxph"></span>
<img src="figures/na_props_1000patients_coxph.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 5: Proportions of valid, infinite and missing (â€˜NAâ€™) estimates for the standard Cox model estimators in the simulation study of figure 7 (1000 patients per simulated data set).
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figestimator-performance-boxplots-100patients"></span>
<img src="figures/estimator_performance_boxplots_100patients.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 6: Performance comparison of standard Cox, empirical Bayes Cox, and fully non-parametric (null) estimators using training data sets with 100 observations each. In the figure grid there is a boxplot corresponding to every tuple <span class="math inline">\((a,m, G, p)\)</span> such that <span class="math inline">\(a\in \lbrace\)</span>regression coefficients, relative hazards, state occupation probabilities<span class="math inline">\(\rbrace\)</span> is the target of estimation, <span class="math inline">\(m\in \lbrace\)</span>standard Cox, empirical Bayes Cox, null<span class="math inline">\(\rbrace\)</span> is the hazard model, <span class="math inline">\(G \in \lbrace\)</span>linear, competing risks, â€˜mâ€™ structure<span class="math inline">\(\rbrace\)</span> is the transition structure of the model, and <span class="math inline">\(p \in \lbrace 10,40,70,100 \rbrace\)</span> is the number of coefficients/covariates per transition. Each boxplot is based on at most 300 average absolute error observations. Figure 4, together with figures 6.1 and 6.3 in file ESM_1.html of the Supporting Scripts and Data, show the proportion of valid, missing and infinite estimates for each estimator. In each simulation scenario, the upper limit of the plotâ€™s y-axis defines a threshold above which observations are considered very large. Very large observations were replaced by the y-axis upper limit before the boxplots were built.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figestimator-performance-boxplots-1000patients"></span>
<img src="figures/estimator_performance_boxplots_1000patients.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 7:  Performance comparison of standard Cox, empirical Bayes Cox, and fully non-parametric (null) estimators using training data sets with 1000 observations each. In the figure grid there is a boxplot corresponding to every tuple <span class="math inline">\((a,m, G, p)\)</span> such that <span class="math inline">\(a\in \lbrace\)</span>regression coefficients, relative hazards, state occupation probabilities<span class="math inline">\(\rbrace\)</span> is the target of estimation, <span class="math inline">\(m\in \lbrace\)</span>standard Cox, empirical Bayes Cox, null<span class="math inline">\(\rbrace\)</span> is the hazard model, <span class="math inline">\(G \in \lbrace\)</span>linear, competing risks, â€˜mâ€™ structure<span class="math inline">\(\rbrace\)</span> is the transition structure of the model, and <span class="math inline">\(p \in \lbrace 10,100,200,300,400,500 \rbrace\)</span> is the number of coefficients/covariates per transition. Each boxplot is based on at most 300 average absolute error observations. Figure 5, together with figures 6.2 and 6.3 in file ESM_1.html of the Supporting Scripts and Data, show the proportion of valid, missing and infinite estimates for each estimator. In each simulation scenario, the upper limit of the plotâ€™s y-axis defines a threshold above which observations are considered very large. Very large observations were replaced by the y-axis upper limit before the boxplots were built.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figworkflow"></span>
<img src="figures/workflow0.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 8: Extension of the mstate analysis framework by ebmstate. Arrows correspond to functions. Boxes correspond to inputs or outputs of functions. Functions CoxRFX and probtrans_fft from ebmstate compute point estimates only. Interval estimates can be obtained using the non-parametric bootstrap algorithm implemented in the function ebmstate::boot_ebmstate.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figtrans-diagrams"></span>
<img src="figures/data_summary_figs2.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 9: a: transition model implied by the data set of patients with myelodysplastic syndromes, together with transition event numbers; b: conversion to a transition structure without cycles; c: transformations applied to the MDS covariate data and summary statistics for the data before transformation. MDS stands for myelodysplastic syndromes; AML stands for acute myeloid leukemia.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figcoef-plots"></span>
<img src="figures/coef_plots.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 10: Point estimates of regression coefficients for the Cox model fitted to the MDS data, along with 95% non-parametric bootstrap confidence intervals. The x-axis scale is logarithmic so that coefficient estimates can be read as relative hazard estimates. If <span class="math inline">\(\gamma_{ij}\)</span> is the element of <span class="math inline">\(\hat{\boldsymbol{\beta}}_{ij}\)</span> associated with a given covariate, <span class="math inline">\(\exp\left(\gamma_{ij}\right)\)</span> is the estimated relative hazard for this covariate in transition <span class="math inline">\(\left(i,j\right)\)</span>. In general, a relative hazard estimate <span class="math inline">\(r\)</span> for a covariate <span class="math inline">\(z\)</span> in transition <span class="math inline">\(\left(i,j\right)\)</span> means that a one-unit increase in z is associated with an r-fold increase in the hazard of this transition. If z was obtained by log-transformation (as in age, platelet counts and neutrophil counts), a one-unit increase in z corresponds to scaling the original covariate by <span class="math inline">\(e\approx 2.72\)</span>. In case z was obtained by logit-transformation (as in bone marrow blasts and sideroblasts proportions), the same one-unit increase corresponds to scaling the odds of the original covariate by e.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figpatient78-cumhaz"></span>
<img src="figures/patient78_cumhaz_final.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 11: Point estimates of cumulative transition hazards for a sample patient with MDS (black curve), along with <span class="math inline">\(95\\%\)</span> non-parametric confidence intervals (dashed red lines).
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figpatient78-transProbs"></span>
<img src="figures/patient78_transProbs_final.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 12: Point estimates of state occupation probabilities for a sample patient with MDS (black curve), along with <span class="math inline">\(95\\%\)</span> non-parametric confidence intervals (dashed red lines).
</p>
</div>
</div>
</div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h3 class="appendix" data-number="8" id="supplementary-materials"><span class="header-section-number">8</span> Supplementary materials</h3>
<p>Supplementary materials are available in addition to this article. It can be downloaded at
<a href="RJ-2024-002.zip">RJ-2024-002.zip</a></p>
<h3 class="appendix" data-number="9" id="cran-packages-used"><span class="header-section-number">9</span> CRAN packages used</h3>
<p><a href="https://cran.r-project.org/package=msm">msm</a>, <a href="https://cran.r-project.org/package=SemiMarkov">SemiMarkov</a>, <a href="https://cran.r-project.org/package=survival">survival</a>, <a href="https://cran.r-project.org/package=mstate">mstate</a>, <a href="https://cran.r-project.org/package=mboost">mboost</a>, <a href="https://cran.r-project.org/package=gamboostMSM">gamboostMSM</a>, <a href="https://cran.r-project.org/package=penMSM">penMSM</a>, <a href="https://cran.r-project.org/package=ebmstate">ebmstate</a></p>
<h3 class="appendix" data-number="10" id="cran-task-views-implied-by-cited-packages"><span class="header-section-number">10</span> CRAN Task Views implied by cited packages</h3>
<p><a href="https://cran.r-project.org/view=ClinicalTrials">ClinicalTrials</a>, <a href="https://cran.r-project.org/view=Distributions">Distributions</a>, <a href="https://cran.r-project.org/view=Econometrics">Econometrics</a>, <a href="https://cran.r-project.org/view=Epidemiology">Epidemiology</a>, <a href="https://cran.r-project.org/view=MachineLearning">MachineLearning</a>, <a href="https://cran.r-project.org/view=Survival">Survival</a></p>
<h3 class="appendix" data-number="11" id="note"><span class="header-section-number">11</span> Note</h3>
<p>This article is converted from a Legacy LaTeX article using the
<a href="https://cran.r-project.org/package=texor">texor</a> package.
The pdf version is the official version. To report a problem with the html,
refer to CONTRIBUTE on the R Journal homepage.</p>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Aalen1989" class="csl-entry" role="listitem">
O. O. Aalen. A linear regression model for the analysis of life times. <em>Statistics in Medicine</em>, 8(8): 907â€“925, 1989. URL <a href="https://doi.org/10.1002/sim.4780080803">https://doi.org/10.1002/sim.4780080803</a>.
</div>
<div id="ref-Aalen2008" class="csl-entry" role="listitem">
O. Aalen, O. Borgan and H. Gjessing. <em>Survival and event history analysis.</em> Springer, 2008. URL <a href="https://link.springer.com/book/10.1007/978-0-387-68560-1">https://link.springer.com/book/10.1007/978-0-387-68560-1</a>.
</div>
<div id="ref-Andersen1993" class="csl-entry" role="listitem">
P. Andersen, O. Borgan, R. Gill and N. Keiding. <em>Statistical models based on counting processes.</em> Springer, 1993. URL <a href="https://link.springer.com/book/10.1007/978-1-4612-4348-9">https://link.springer.com/book/10.1007/978-1-4612-4348-9</a>.
</div>
<div id="ref-Carlin2009" class="csl-entry" role="listitem">
B. Carlin and T. Louis. <em>Bayesian methods for data analysis.</em> CRC Press, 2009. URL <a href="https://doi.org/10.1201/b14884">https://doi.org/10.1201/b14884</a>.
</div>
<div id="ref-Cortese2010" class="csl-entry" role="listitem">
G. Cortese and P. K. Andersen. Competing risks and time-dependent covariates. <em>Biometrical Journal</em>, 52(1): 138â€“158, 2010. URL <a href="https://doi.org/10.1002/bimj.200900076">https://doi.org/10.1002/bimj.200900076</a>.
</div>
<div id="ref-Putter2011" class="csl-entry" role="listitem">
L. C. de Wreede, M. Fiocco and H. Putter. <span class="nocase">mstate</span>: An <span>R</span> package for the analysis of competing risks and multi-state models. <em>Journal of Statistical Software</em>, 38(7): 1â€“30, 2011. URL <a href="http://www.jstatsoft.org/v38/i07/">http://www.jstatsoft.org/v38/i07/</a>.
</div>
<div id="ref-Gelman2014" class="csl-entry" role="listitem">
A. Gelman, J. Carlin, H. Stern, D. Dunson, A. Vehtari and D. Rubin. <em>Bayesian data analysis.</em> CRC Press, 2014. URL <a href="https://doi.org/10.1201/b16018">https://doi.org/10.1201/b16018</a>.
</div>
<div id="ref-harrell1982evaluating" class="csl-entry" role="listitem">
J. Harrell Frank E., R. M. Califf, D. B. Pryor, K. L. Lee and R. A. Rosati. <span class="nocase">Evaluating the Yield of Medical Tests</span>. <em>JAMA</em>, 247(18): 2543â€“2546, 1982. URL <a href="https://doi.org/10.1001/jama.1982.03320430047030">https://doi.org/10.1001/jama.1982.03320430047030</a>.
</div>
<div id="ref-Hastie2009" class="csl-entry" role="listitem">
T. Hastie, R. Tibshirani, J. H. Friedman and J. H. Friedman. <em>The elements of statistical learning: Data mining, inference, and prediction.</em> Springer, 2009. URL <a href="https://link.springer.com/book/10.1007/978-0-387-84858-7">https://link.springer.com/book/10.1007/978-0-387-84858-7</a>.
</div>
<div id="ref-mboost_package" class="csl-entry" role="listitem">
T. Hothorn, P. Buehlmann, T. Kneib, M. Schmid and B. Hofner. Mboost: Model-based boosting. <em>R package version</em>, 2.9â€“3, 2020. URL <a href="https://CRAN.R-project.org/package=mboost">https://CRAN.R-project.org/package=mboost</a>.
</div>
<div id="ref-Hougaard1999" class="csl-entry" role="listitem">
P. Hougaard. Multi-state models: A review. <em>Lifetime data analysis</em>, 5(3): 239â€“264, 1999. URL <a href="https://doi.org/10.1023/A:1009672031531">https://doi.org/10.1023/A:1009672031531</a>.
</div>
<div id="ref-flexsurv_package" class="csl-entry" role="listitem">
C. Jackson. <span class="nocase">flexsurv</span>: A platform for parametric survival modeling in <span>R</span>. <em>Journal of Statistical Software</em>, 70(8): 1â€“33, 2016. DOI <a href="https://doi.org/10.18637/jss.v070.i08">10.18637/jss.v070.i08</a>.
</div>
<div id="ref-Jackson2011" class="csl-entry" role="listitem">
C. H. Jackson. Multi-state models for panel data: The <span class="nocase">msm</span> package for <span>R</span>. <em>Journal of Statistical Software</em>, 38(8): 1â€“29, 2011. URL <a href="http://www.jstatsoft.org/v38/i08/">http://www.jstatsoft.org/v38/i08/</a>.
</div>
<div id="ref-Kalbfleisch2002" class="csl-entry" role="listitem">
J. D. Kalbfleisch and R. L. Prentice. <em>The statistical analysis of failure time data.</em> John Wiley &amp; Sons, 2002. DOI <a href="https://doi.org/10.1002/9781118032985">10.1002/9781118032985</a>.
</div>
<div id="ref-Listwon2015" class="csl-entry" role="listitem">
A. Listwon and P. Saint-Pierre. <span class="nocase">SemiMarkov: An R Package for Parametric Estimation in Multi-State Semi-Markov Models</span>. <em><span>Journal of Statistical Software</span></em>, 66(6): 784, 2015. URL <a href="https://hal.archives-ouvertes.fr/hal-00860244">https://hal.archives-ouvertes.fr/hal-00860244</a>.
</div>
<div id="ref-Papaemmanuil2013" class="csl-entry" role="listitem">
E. Papaemmanuil, M. Gerstung, L. Malcovati, S. Tauro, G. Gundem, P. Van Loo, C. J. Yoon, P. Ellis, D. C. Wedge, A. Pellagatti, et al. Clinical and biological implications of driver mutations in myelodysplastic syndromes. <em>Blood</em>, 122(22): 3616â€“3627, 2013. URL <a href="https://doi.org/10.1182/blood-2013-08-518886">https://doi.org/10.1182/blood-2013-08-518886</a>.
</div>
<div id="ref-Perperoglou2014" class="csl-entry" role="listitem">
A. Perperoglou. Cox models with dynamic ridge penalties on time-varying effects of the covariates. <em>Statistics in Medicine</em>, 33(1): 170â€“180, 2014. URL <a href="https://doi.org/10.1002/sim.5921">https://doi.org/10.1002/sim.5921</a>.
</div>
<div id="ref-Putter2011tutorial" class="csl-entry" role="listitem">
H. Putter. Tutorial in biostatistics: Competing risks and multi-state models analyses using the mstate package. <em>Companion file for the mstate package</em>, 2011. URL <a href="https://mirror.las.iastate.edu/CRAN/web/packages/mstate/vignettes/Tutorial.pdf">https://mirror.las.iastate.edu/CRAN/web/packages/mstate/vignettes/Tutorial.pdf</a>.
</div>
<div id="ref-Putter2007tutorial" class="csl-entry" role="listitem">
H. Putter, M. Fiocco and R. B. Geskus. Tutorial in biostatistics: Competing risks and multi-state models. <em>Statistics in Medicine</em>, 26(11): 2389â€“2430, 2007. URL <a href="https://doi.org/10.1002/sim.2712">https://doi.org/10.1002/sim.2712</a>.
</div>
<div id="ref-gamboostMSM_package" class="csl-entry" role="listitem">
H. Reulen. gamboostMSM. <em>R package version</em>, 1.1.87, 2014. URL <a href="https://CRAN.R-project.org/package=gamboostMSM">https://CRAN.R-project.org/package=gamboostMSM</a>.
</div>
<div id="ref-penMSM_package" class="csl-entry" role="listitem">
H. Reulen. penMSM. <em>R package version</em>, 0.99, 2015. URL <a href="https://CRAN.R-project.org/package=penMSM">https://CRAN.R-project.org/package=penMSM</a>.
</div>
<div id="ref-Samworth2012" class="csl-entry" role="listitem">
R. J. Samworth. Steinâ€™s paradox. <em>Eureka</em>, 62: 38â€“41, 2012. URL <a href="https://citeseerx.ist.psu.edu/document?repid=rep1&amp;type=pdf&amp;doi=7eebd55f569395544f2b5d367d6aee614901d2c1">https://citeseerx.ist.psu.edu/document?repid=rep1&amp;type=pdf&amp;doi=7eebd55f569395544f2b5d367d6aee614901d2c1</a>.
</div>
<div id="ref-Schall1991" class="csl-entry" role="listitem">
R. Schall. Estimation in generalized linear models with random effects. <em>Biometrika</em>, 78(4): 719â€“727, 1991. URL <a href="http://dx.doi.org/10.1093/biomet/78.4.719">http://dx.doi.org/10.1093/biomet/78.4.719</a>.
</div>
<div id="ref-schwarz1978estimating" class="csl-entry" role="listitem">
G. Schwarz. Estimating the dimension of a model. <em>The annals of statistics</em>, 461â€“464, 1978. URL <a href="https://www.jstor.org/stable/2958889">https://www.jstor.org/stable/2958889</a>.
</div>
<div id="ref-Spitoni2012" class="csl-entry" role="listitem">
C. Spitoni, M. Verduijn and H. Putter. Estimation and asymptotic theory for transition probabilities in markov renewal multi-state models. <em>The International Journal of Biostatistics</em>, 8(1): 2012. URL <a href="https://doi.org/10.1515/1557-4679.1375">https://doi.org/10.1515/1557-4679.1375</a>.
</div>
<div id="ref-survival_package" class="csl-entry" role="listitem">
T. M. Therneau. <em>A package for survival analysis in s.</em> 2015. URL <a href="https://CRAN.R-project.org/package=survival">https://CRAN.R-project.org/package=survival</a>. version 2.38.
</div>
<div id="ref-Wreede2010" class="csl-entry" role="listitem">
L. C. de Wreede, M. Fiocco and H. Putter. The mstate package for estimation and prediction in non- and semi-parametric multi-state and competing risks models. <em>Computer Methods and Programs in Biomedicine</em>, 99(3): 261â€“274, 2010. URL <a href="http://www.sciencedirect.com/science/article/pii/S0169260710000027">http://www.sciencedirect.com/science/article/pii/S0169260710000027</a>.
</div>
<div id="ref-Zou2005" class="csl-entry" role="listitem">
H. Zou and T. Hastie. Regularization and variable selection via the elastic net. <em>Journal of the Royal Statistical Society: Series B (Statistical Methodology)</em>, 67(2): 301â€“320, 2005. URL <a href="https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/j.1467-9868.2005.00503.x">https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/j.1467-9868.2005.00503.x</a>.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
  <h3 id="references">References</h3>
  <div id="references-listing"></div>
  <h3 id="reuse">Reuse</h3>
  <p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
  <h3 id="citation">Citation</h3>
  <p>For attribution, please cite this work as</p>
  <pre class="citation-appendix short">Costa &amp; Gerstung, "ebmstate: An R Package For Disease Progression Analysis Under Empirical Bayes Cox Models", The R Journal, 2025</pre>
  <p>BibTeX citation</p>
  <pre class="citation-appendix long">@article{RJ-2024-002,
  author = {Costa, Rui J. and Gerstung, Moritz},
  title = {ebmstate: An R Package For Disease Progression Analysis Under Empirical Bayes Cox Models},
  journal = {The R Journal},
  year = {2025},
  note = {https://doi.org/10.32614/RJ-2024-002},
  doi = {10.32614/RJ-2024-002},
  volume = {16},
  issue = {1},
  issn = {2073-4859},
  pages = {15-38}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<!--radix_placeholder_navigation_after_body--><html><body>
<div class="distill-site-nav distill-site-footer">
<p>Â© The R Foundation, <a href="mailto:r-journal@r-project.org">web page
contact</a>.</p>
</div>
<!--/radix_placeholder_navigation_after_body-->
</body></html>


</body>

</html>
