<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
      margin-bottom: 0em;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
  <title>BoundaryStats: An R Package to Calculate Boundary Overlap Statistics</title>

  <meta property="description" itemprop="description" content="Ecologists and epidemiologists frequently rely on spatially distributed data. Studies in these fields may concern geographic boundaries, as environmental variation can determine the spatial distribution of organismal traits or diseases. In such cases, environmental boundaries produce coincident geographic boundaries in, for example, disease prevalence. Boundary analysis can be used to investigate the co-occurrence of organismal trait or disease boundaries and underlying environmental boundaries. Within boundary analysis, boundary overlap statistics test for the presence of significant geographic boundaries and spatial associations between the boundaries of two variables. There is one pre-existing implementation of boundary overlap statistics, though it is only on Windows and ESRI ArcView, limiting the availability of boundary overlap statistics to researchers. I have created BoundaryStats—an R package available on CRAN—that implements boundary overlap statistics. BoundaryStats is the first open-source, cross-platform implementation of these statistical methods, making the statistics more widely accessible to researchers."/>

  <link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>

  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2025-10-21"/>
  <meta property="article:created" itemprop="dateCreated" content="2025-10-21"/>
  <meta name="article:author" content="Amy Luo"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="BoundaryStats: An R Package to Calculate Boundary Overlap Statistics"/>
  <meta property="og:type" content="article"/>
  <meta property="og:description" content="Ecologists and epidemiologists frequently rely on spatially distributed data. Studies in these fields may concern geographic boundaries, as environmental variation can determine the spatial distribution of organismal traits or diseases. In such cases, environmental boundaries produce coincident geographic boundaries in, for example, disease prevalence. Boundary analysis can be used to investigate the co-occurrence of organismal trait or disease boundaries and underlying environmental boundaries. Within boundary analysis, boundary overlap statistics test for the presence of significant geographic boundaries and spatial associations between the boundaries of two variables. There is one pre-existing implementation of boundary overlap statistics, though it is only on Windows and ESRI ArcView, limiting the availability of boundary overlap statistics to researchers. I have created BoundaryStats—an R package available on CRAN—that implements boundary overlap statistics. BoundaryStats is the first open-source, cross-platform implementation of these statistical methods, making the statistics more widely accessible to researchers."/>
  <meta property="og:locale" content="en_US"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="BoundaryStats: An R Package to Calculate Boundary Overlap Statistics"/>
  <meta property="twitter:description" content="Ecologists and epidemiologists frequently rely on spatially distributed data. Studies in these fields may concern geographic boundaries, as environmental variation can determine the spatial distribution of organismal traits or diseases. In such cases, environmental boundaries produce coincident geographic boundaries in, for example, disease prevalence. Boundary analysis can be used to investigate the co-occurrence of organismal trait or disease boundaries and underlying environmental boundaries. Within boundary analysis, boundary overlap statistics test for the presence of significant geographic boundaries and spatial associations between the boundaries of two variables. There is one pre-existing implementation of boundary overlap statistics, though it is only on Windows and ESRI ArcView, limiting the availability of boundary overlap statistics to researchers. I have created BoundaryStats—an R package available on CRAN—that implements boundary overlap statistics. BoundaryStats is the first open-source, cross-platform implementation of these statistical methods, making the statistics more widely accessible to researchers."/>

  <!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
  <meta name="citation_title" content="BoundaryStats: An R Package to Calculate Boundary Overlap Statistics"/>
  <meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2025-025"/>
  <meta name="citation_pdf_url" content="RJ-2025-025.pdf"/>
  <meta name="citation_volume" content="17"/>
  <meta name="citation_issue" content="3"/>
  <meta name="citation_doi" content="10.32614/RJ-2025-025"/>
  <meta name="citation_journal_title" content="The R Journal"/>
  <meta name="citation_issn" content="2073-4859"/>
  <meta name="citation_firstpage" content="1"/>
  <meta name="citation_fulltext_world_readable" content=""/>
  <meta name="citation_online_date" content="2025/10/21"/>
  <meta name="citation_publication_date" content="2025/10/21"/>
  <meta name="citation_author" content="Amy Luo"/>
  <meta name="citation_author_institution" content="University of Tennessee, Knoxville"/>
  <!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Forest-sedgeland boundaries are historically stable and resilient to wildfire at Blakes Opening in the Tasmanian Wilderness World Heritage Area, Australia;citation_volume=38;citation_doi=10.1007/s10980-022-01558-x;citation_author=David M. J. S. Bowman;citation_author=Stefania Ondei;citation_author=Arko Lucieer;citation_author=Scott Foyster;citation_author=Lynda D. Prior"/>
  <meta name="citation_reference" content="citation_title=Hybridization at an ecotone: ecological and genetic barriers between three Iberian vipers;citation_volume=23;citation_doi=10.1111/mec.12671;citation_author=Pedro Tarroso;citation_author=Ricardo J. Pereira;citation_author=Fernando Martínez-Freiría;citation_author=Raquel Godinho;citation_author=José C. Brito"/>
  <meta name="citation_reference" content="citation_title=Geographic boundary analysis in spatial and spatio-temporal epidemiology: Perspective and prospects;citation_volume=1;citation_doi=10.1016/j.sste.2010.09.003;citation_author=Geoffrey M. Jacquez"/>
  <meta name="citation_reference" content="citation_title=From fields to objects: A review of geographic boundary analysis;citation_volume=2;citation_doi=10.1007/PL00011456;citation_author=Geoffrey M. Jacquez;citation_author=S. Maruca;citation_author=Marie-Josée. Fortin"/>
  <meta name="citation_reference" content="citation_title=The map comparison problem: Tests for the overlap of geographic boundaries;citation_volume=14;citation_doi=10.1002/sim.4780142107;citation_author=Geoffrey M. Jacquez"/>
  <meta name="citation_reference" content="citation_title=Quantification of the spatial co-occurrences of ecological boundaries;citation_volume=77;citation_doi=10.2307/3545584;citation_author=Marie-Josée Fortin;citation_author=Pierre Drapeau;citation_author=Geoffrey M. Jacquez"/>
  <meta name="citation_reference" content="citation_title=Spatial analysis of landscapes: Concepts and statistics;citation_volume=86;citation_doi=10.1890/04-0914;citation_author=Helene H. Wagner;citation_author=Marie-Josée Fortin"/>
  <meta name="citation_reference" content="citation_title=A conceptual framework for the spatial analysis of landscape genetic data;citation_volume=14;citation_doi=10.1007/s10592-012-0391-5;citation_author=Helene H. Wagner;citation_author=Marie-Josée Fortin"/>
  <meta name="citation_reference" content="citation_title=Quantifying the spatial relationship between bird species’ distributions and landscape feature boundaries in southern Ontario, Canada;citation_volume=27;citation_doi=10.1007/s10980-012-9804-6;citation_author=Aleksandra E. Polakowska;citation_author=Marie-Josée Fortin;citation_author=Andrew Couturier"/>
  <meta name="citation_reference" content="citation_title=Exposure density and neighborhood disparities in COVID-19 infection risk;citation_volume=118;citation_doi=10.1073/pnas.2021258118;citation_author=Boyeong Hong;citation_author=Bartosz J. Bonczak;citation_author=Arpit Gupta;citation_author=Lorna E. Thorpe;citation_author=Constantine E. Kontokosta"/>
  <meta name="citation_reference" content="citation_title=Spatial characteristics of heavy metal contamination and potential human health risk assessment of urban soils: A case study from an urban region of South India;citation_volume=194;citation_doi=10.1016/j.ecoenv.2020.110406;citation_author=Narsimha Adimalla;citation_author=Jie Chen;citation_author=Hui Qian"/>
  <meta name="citation_reference" content="citation_title=Chronic disease surveillance and testing of clustering of disease and exposure: Application to leukemia incidence and TCE-contaminated dumpsites in upstate New York;citation_volume=3;citation_doi=10.1002/env.3170030303;citation_author=Lance A. Waller;citation_author=Bruce W. Turnbull;citation_author=Larry C. Clark;citation_author=Philip Nasca"/>
  <meta name="citation_reference" content="citation_title=Neighbourhood effects research: New perspectives;citation_publisher=Springer Netherlands;citation_doi=10.1007/978-94-007-2309-2"/>
  <meta name="citation_reference" content="citation_title=Spatial association between forest heterogeneity and breeding territory boundaries of two forest songbirds;citation_volume=19;citation_doi=10.1023/B:LAND.0000042849.63040.a9;citation_author=Véronique St-Louis;citation_author=Marie-Josée Fortin;citation_author=André Desrochers"/>
  <meta name="citation_reference" content="citation_title=SpatialBoundaries.jl: Edge detection using spatial wombling;citation_volume=2023;citation_doi=10.1111/ecog.06609;citation_author=Tanya Strydom;citation_author=Timothée Poisot"/>
  <meta name="citation_reference" content="citation_title=Geographic patterns of (genetic, morphological, linguistic) variation: How barriers can be detected with Monmonier’salgorithm;citation_volume=76;citation_doi=10.1353/hub.2004.0034;citation_author=Franz Manni;citation_author=Etienne Guérard;citation_author=Evelyne Heyer"/>
  <meta name="citation_reference" content="citation_title=TESS3: Fast inference of spatial population structure and genome scans for selection;citation_volume=16;citation_doi=10.1111/1755-0998.12471;citation_author=Kevin Caye;citation_author=Timo M. Deist;citation_author=Helena Martins;citation_author=Olivier Michel;citation_author=Olivier François"/>
  <meta name="citation_reference" content="citation_title=Comparison of Bayesian clustering and edge detection methods for inferring boundaries in landscape genetics;citation_volume=12;citation_doi=10.3390/ijms12020865;citation_author=Toni Safner;citation_author=Mark P. Miller;citation_author=Brad H. McRae;citation_author=Marie-Josée Fortin;citation_author=Stéphanie Manel"/>
  <meta name="citation_reference" content="citation_title=Methods to detect edge effected reductions in fire frequency in simulated forest landscapes;citation_volume=8;citation_doi=10.3390/ijgi8060277;citation_author=Xinyuan Wei;citation_author=Chris P. S. Larsen"/>
  <meta name="citation_reference" content="citation_title=Landscape patterns simulation with a modified random clusters method;citation_volume=15;citation_doi=10.1023/A:1008107902848;citation_author=Santiago Saura;citation_author=Javier Martínez-Millan"/>
  <meta name="citation_reference" content="citation_title=Song and genetic divergence within a subspecies of white-crowned sparrow (&lt;i&gt;Zonotrichia leucophrys nuttalli&lt;/i&gt;);citation_volume=19;citation_doi=10.1371/journal.pone.0304348;citation_author=Amy Rongyan Luo;citation_author=Sara Lipshutz;citation_author=Jennifer Phillips;citation_author=Robb T Brumfield;citation_author=Elizabeth Perrault Derryberry"/>
  <meta name="citation_reference" content="citation_title=geoorigins: A new method and R package for trait mapping and geographic provenancing of specimens without categorical constraints;citation_volume=11;citation_doi=10.1111/2041-210X.13444;citation_author=Ardern Hulme-Beaman;citation_author=Anna Rudzinski;citation_author=Joseph E. J. Cooper;citation_author=Robert F. Lachlan;citation_author=Keith Dobney;citation_author=Mark G. Thomas"/>
  <meta name="citation_reference" content="citation_title=fastSTRUCTURE: Variational inference of population structure in large SNP data sets;citation_volume=197;citation_doi=10.1534/genetics.114.164350;citation_author=Anil Raj;citation_author=Matthew Stephens;citation_author=Jonathan K Pritchard"/>
  <meta name="citation_reference" content="citation_title=Terra: Spatial data analysis;citation_author=Robert J. Hijmans"/>
  <meta name="citation_reference" content="citation_title=Multivariable geostatistics in S: The gstat package;citation_volume=30;citation_doi=10.1016/j.cageo.2004.03.012;citation_author=Edzer J Pebesma"/>
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","date","description","draft","author","type","output","bibliography","link-citations","date_received","volume","issue","slug","journal","pdf_url","citation_url","doi","creative_commons","packages","CTV","csl"]}},"value":[{"type":"character","attributes":{},"value":["BoundaryStats: An R Package to Calculate Boundary Overlap Statistics"]},{"type":"character","attributes":{},"value":["2025-10-21"]},{"type":"character","attributes":{},"value":["Ecologists and epidemiologists frequently rely on spatially distributed data. Studies in these fields may concern geographic boundaries, as environmental variation can determine the spatial distribution of organismal traits or diseases. In such cases, environmental boundaries produce coincident geographic boundaries in, for example, disease prevalence. Boundary analysis can be used to investigate the co-occurrence of organismal trait or disease boundaries and underlying environmental boundaries. Within boundary analysis, boundary overlap statistics test for the presence of significant geographic boundaries and spatial associations between the boundaries of two variables. There is one pre-existing implementation of boundary overlap statistics, though it is only on Windows and ESRI ArcView, limiting the availability of boundary overlap statistics to researchers. I have created BoundaryStats—an R package available on CRAN—that implements boundary overlap statistics. BoundaryStats is the first open-source, cross-platform implementation of these statistical methods, making the statistics more widely accessible to researchers.\n"]},{"type":"logical","attributes":{},"value":[false]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","orcid_id","email"]}},"value":[{"type":"character","attributes":{},"value":["Amy Luo"]},{"type":"character","attributes":{},"value":["University of Tennessee, Knoxville"]},{"type":"character","attributes":{},"value":["Department of Ecology and Evolutionary Biology","Knoxville, Tennessee"]},{"type":"character","attributes":{},"value":["0000-0002-9197-4673"]},{"type":"character","attributes":{},"value":["amy.luo.15@gmail.com"]}]}]},{"type":"character","attributes":{},"value":["package"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["rjtools::rjournal_pdf_article","distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["toc"]}},"value":[{"type":"logical","attributes":{},"value":[false]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[false]}]}]},{"type":"character","attributes":{},"value":["BoundaryStats-ms.bib"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2024-12-05"]},{"type":"integer","attributes":{},"value":[17]},{"type":"integer","attributes":{},"value":[3]},{"type":"character","attributes":{},"value":["RJ-2025-025"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"double","attributes":{},"value":[1]},{"type":"NULL"}]},{"type":"character","attributes":{},"value":["RJ-2025-025.pdf"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2025-025"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2025-025"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"character","attributes":{},"value":["BoundaryStats","gstat","fields","terra","ggplot2"]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["ChemPhys","NetworkAnalysis","Phylogenetics","Spatial","SpatioTemporal","TeachingStatistics"]},{"type":"character","attributes":{},"value":["/home/hyndman/R/x86_64-pc-linux-gnu-library/4.5/rjtools/rjournal.csl"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["BoundaryStats-ms_files/figure-html5/fig-5-1.png","BoundaryStats-ms_files/figure-latex/fig-5-1.pdf","BoundaryStats-ms.bib","BoundaryStats-ms.tex","data/2010_2022_song_boundaries.asc","data/genetic_interpolation.asc","figures/fig1.png","figures/fig2.png","figures/fig3.png","figures/fig4.png","RJ-2025-025_files/figure-html5/fig-5-1.png","RJ-2025-025.pdf","RJ-2025-025.tex","RJ-2025-025.zip","RJournal.sty","RJwrapper.bbl","RJwrapper.blg","RJwrapper.brf","RJwrapper.fdb_latexmk","RJwrapper.fls","RJwrapper.out","RJwrapper.tex","Rlogo-5.png","scripts/BoundaryStats-implementation-example.R"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
    font-size: 100%;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  hr.section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    margin: 0px;
  }


  d-byline {
    border-top: none;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
    border-top: none;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  /* tweak for Pandoc numbered line within distill */
  d-article pre.numberSource code > span {
      left: -2em;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // separator
    var separator = '<hr class="section-separator" style="clear: both"/>';
    // prepend separator above appendix
    $('.d-byline').before(separator);
    $('.d-article').before(separator);

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme, except when numbering line
    // in code chunk
    $('pre:not(.numberLines) code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        var author_name = front_matter.authors[i].author
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        const citeChild = $(this).children()[0]
        // Do not process if @xyz has been used without escaping and without bibliography activated
        // https://github.com/rstudio/distill/issues/466
        if (citeChild === undefined) return true

        if (citeChild.nodeName == "D-FOOTNOTE") {
          var fn = citeChild
          $(this).html(fn.shadowRoot.querySelector("sup"))
          $(this).id = fn.id
          fn.remove()
        }
        var refs = $(this).attr('data-cites').split(" ");
        var refHtml = refs.map(function(ref) {
          // Could use CSS.escape too here, we insure backward compatibility in navigator
          return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
        }).join("\n");
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // fix footnotes in tables (#411)
      // replacing broken distill.pub feature
      $('table d-footnote').each(function() {
        // we replace internal showAtNode methode which is triggered when hovering a footnote
        this.hoverBox.showAtNode = function(node) {
          // ported from https://github.com/distillpub/template/pull/105/files
          calcOffset = function(elem) {
              let x = elem.offsetLeft;
              let y = elem.offsetTop;
              // Traverse upwards until an `absolute` element is found or `elem`
              // becomes null.
              while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                  x += elem.offsetLeft;
                  y += elem.offsetTop;
              }

              return { left: x, top: y };
          }
          // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
          const bbox = node.getBoundingClientRect();
          const offset = calcOffset(node);
          this.show([offset.left + bbox.width, offset.top + bbox.height]);
        }
      })

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      // ignore leaflet img layers (#106)
      figures = figures.filter(':not(img[class*="leaflet"])')
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="RJ-2025-025_files/header-attrs-2.29/header-attrs.js"></script>
  <script src="RJ-2025-025_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="RJ-2025-025_files/popper-2.6.0/popper.min.js"></script>
  <link href="RJ-2025-025_files/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="RJ-2025-025_files/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="RJ-2025-025_files/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="RJ-2025-025_files/anchor-4.2.2/anchor.min.js"></script>
  <script src="RJ-2025-025_files/bowser-1.9.3/bowser.min.js"></script>
  <script src="RJ-2025-025_files/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="RJ-2025-025_files/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Mathjax config (add automatic linebreaks when supported)
      // MathJax = {
      //    tex: {
      //        inlineMath: [['$', '$'], ['\\(', '\\)']],
      //        displayMath: [['$$', '$$'], ['\\[', '\\]']],
      //        tags: 'ams',
      //        multline: true,
      //    },
      //    options: {
      //        linebreaks: { automatic: true },
      //    },
      // };

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      buttons.append('<a href="' + slug + '.zip" class="dt-tag"><i class="fas fa-file-zipper"></i> Supplement</a>')

      // adds Abstract: in front of the first <p> in the title section --
      // unless it happens to be the subtitle (FIXME: this is a bad hack - can't distill do this?)
      var tpar = $("d-title p:not(:empty)").filter(function() {
        return !$(this).hasClass("subtitle");
      }).first();
      if (tpar) {
        const abstract = $('<d-abstract>')
        abstract.append('<b>Abstract:</b><br>')
        abstract.append(tpar) // Move description to d-abstract
        $("d-title p:empty").remove() // Remove empty paragraphs after title
        abstract.append(buttons)
        abstract.insertAfter($('d-title')) // Add abstract section after title */
      }

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
    
    d-appendix .citation-appendix,
    .d-appendix .citation-appendix {
      color: rgb(60, 60, 60);
    }

    d-article h2 {
      border-bottom: 0px solid rgba(0, 0, 0, 0.1);
      padding-bottom: 0rem;
    }
    d-article h3 {
      font-size: 20px;
    }
    d-article h4 {
      font-size: 18px;
      text-transform: none;
    }

    @media (min-width: 1024px) {
      d-article h2 {
        font-size: 32px;
      }
      d-article h3 {
        font-size: 24px;
      }
      d-article h4 {
        font-size: 20px;
      }
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"BoundaryStats: An R Package to Calculate Boundary Overlap Statistics","description":"Ecologists and epidemiologists frequently rely on spatially distributed data. Studies in these fields may concern geographic boundaries, as environmental variation can determine the spatial distribution of organismal traits or diseases. In such cases, environmental boundaries produce coincident geographic boundaries in, for example, disease prevalence. Boundary analysis can be used to investigate the co-occurrence of organismal trait or disease boundaries and underlying environmental boundaries. Within boundary analysis, boundary overlap statistics test for the presence of significant geographic boundaries and spatial associations between the boundaries of two variables. There is one pre-existing implementation of boundary overlap statistics, though it is only on Windows and ESRI ArcView, limiting the availability of boundary overlap statistics to researchers. I have created BoundaryStats—an R package available on CRAN—that implements boundary overlap statistics. BoundaryStats is the first open-source, cross-platform implementation of these statistical methods, making the statistics more widely accessible to researchers.","doi":"10.32614/RJ-2025-025","authors":[{"author":"Amy Luo","authorURL":"#","affiliation":"University of Tennessee, Knoxville","affiliationURL":"#","orcidID":"0000-0002-9197-4673"}],"publishedDate":"2025-10-21T00:00:00.000+11:00","citationText":"Luo, 2025"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>BoundaryStats: An R Package to Calculate Boundary Overlap Statistics</h1>

<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>Ecologists and epidemiologists frequently rely on spatially distributed data. Studies in these fields may concern geographic boundaries, as environmental variation can determine the spatial distribution of organismal traits or diseases. In such cases, environmental boundaries produce coincident geographic boundaries in, for example, disease prevalence. Boundary analysis can be used to investigate the co-occurrence of organismal trait or disease boundaries and underlying environmental boundaries. Within boundary analysis, boundary overlap statistics test for the presence of significant geographic boundaries and spatial associations between the boundaries of two variables. There is one pre-existing implementation of boundary overlap statistics, though it is only on Windows and ESRI ArcView, limiting the availability of boundary overlap statistics to researchers. I have created BoundaryStats—an R package available on CRAN—that implements boundary overlap statistics. BoundaryStats is the first open-source, cross-platform implementation of these statistical methods, making the statistics more widely accessible to researchers.</p></p>
</div>

<div class="d-byline">
  Amy Luo  (University of Tennessee, Knoxville)
  
<br/>2025-10-21
</div>

<div class="d-article">
<h2 data-number="1" id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>Geographic boundaries are an intrinsic feature of spatial ecology and epidemiology, as the relationships between an underlying environmental variable and organismal traits or disease prevalence often produce coincident geographic boundaries. Boundaries are areas in which spatially distributed variables (e.g., bird plumage coloration, disease prevalence, annual rainfall) rapidly change over a narrow geographic space. They can also represent edges or discontinuities (e.g., neighborhood edges, ecotype boundaries). Boundary zones themselves may be of interest; for example, the temporal boundary dynamics between ecotypes can provide insight into the factors that produce mosaic landscapes <span class="citation" data-cites="bowman2023">(<a href="#ref-bowman2023" role="doc-biblioref">Bowman et al. 2023</a>)</span>.</p>
<p>Boundary analysis involves the analysis of spatial boundaries to answer questions about values within a bounded area, patterns of change across a landscape, and associations between the spatial patterns of multiple variables <span class="citation" data-cites="jacquez2010">(<a href="#ref-jacquez2010" role="doc-biblioref">Jacquez 2010</a>)</span>. Within boundary analysis, boundary overlap statistics can be used to test the association between the boundaries of two spatially distributed variables <span class="citation" data-cites="jacquez2000">(<a href="#ref-jacquez2000" role="doc-biblioref">Jacquez et al. 2000</a>)</span>. These statistics fall within two categories: boundary statistics (i.e., tests for the presence of cohesive boundaries) and boundary overlap statistics (i.e., tests for spatial association between boundaries). BoundaryStats runs two boundary statistics and three boundary overlap statistics, as initially described in Jacquez <span class="citation" data-cites="jacquez1995">(<a href="#ref-jacquez1995" role="doc-biblioref">1995</a>)</span> and Fortin et al. <span class="citation" data-cites="fortin1996">(<a href="#ref-fortin1996" role="doc-biblioref">1996</a>)</span>. The boundary statistics are (1) the length of the longest boundary and (2) the number of cohesive boundaries on the landscape <span class="citation" data-cites="fortin1996">(<a href="#ref-fortin1996" role="doc-biblioref">Fortin et al. 1996</a>)</span>. The boundary overlap statistics are (1) the amount of direct overlap between boundaries in variables A and B, (2) the mean minimum distance between boundaries in A and B, and (3) the mean minimum distance from boundaries in A to boundaries in B <span class="citation" data-cites="fortin1996 jacquez1995">(<a href="#ref-jacquez1995" role="doc-biblioref">Jacquez 1995</a>; <a href="#ref-fortin1996" role="doc-biblioref">Fortin et al. 1996</a>)</span>.</p>
<p>While other spatial statistics account for complications like spatial autocorrelation and environmental heterogeneity <span class="citation" data-cites="wagner2005">(<a href="#ref-wagner2005" role="doc-biblioref">Wagner and Fortin 2005</a>)</span>, boundary overlap statistics can uniquely leverage geographic discontinuities to answer spatial questions. By identifying significant cohesive boundaries, researchers can delineate relevant geographic sampling units (e.g., populations as conservation units for a species or human communities with increased disease risk) <span class="citation" data-cites="jacquez2010">(<a href="#ref-jacquez2010" role="doc-biblioref">Jacquez 2010</a>)</span>. Associations between the spatial boundaries of two variables can be useful in assessing the extent to which an underlying landscape variable drives the spatial distribution of a dependent variable. For example, ecologists are often interested in whether landscape-level ecological boundaries limit gene flow, thereby producing population structure; if the putative ecological boundary is limiting gene flow, one would expect concordant geographic boundaries in the ecological variable and population structure <span class="citation" data-cites="tarroso2014 wagner2013">(<a href="#ref-wagner2013" role="doc-biblioref">Wagner and Fortin 2013</a>; <a href="#ref-tarroso2014" role="doc-biblioref">Tarroso et al. 2014</a>)</span>. The presence of boundaries can similarly limit the distribution of taxonomically similar species <span class="citation" data-cites="polakowska2012">(<a href="#ref-polakowska2012" role="doc-biblioref">Polakowska et al. 2012</a>)</span>. In an epidemiological context, this may look like neighborhood effects on public health outcomes, including COVID-19 infection risk <span class="citation" data-cites="hong2021 neighbou2012">(<a href="#ref-neighbou2012" role="doc-biblioref">Ham et al. 2012</a>; <a href="#ref-hong2021" role="doc-biblioref">Hong et al. 2021</a>)</span> or spatial relationships between high pollutant density and increased disease risk <span class="citation" data-cites="adimalla2020 waller1992">(<a href="#ref-waller1992" role="doc-biblioref">Waller et al. 1992</a>; <a href="#ref-adimalla2020" role="doc-biblioref">Adimalla et al. 2020</a>)</span>.</p>
<p>Currently, there is at least one tool that has implemented boundary overlap statistics: GEM, which was released as an extension of ESRI ArcView and a standalone Windows package. GEM is not available as a cross-platform, free, and open-source software, thereby limiting its accessibility to researchers. <a href="https://cran.r-project.org/package=BoundaryStats">BoundaryStats</a> implements boundary and boundary overlap statistics in R. It is available to download on CRAN, making the tools more accessible for researchers, especially in epidemiology and spatial ecology.</p>
<h2 data-number="2" id="boundary-definitions"><span class="header-section-number">2</span> Boundary definitions</h2>
<p>In this framework, we classify raster cells into a pseudobinary: boundary elements (1), non-boundary cells (0), or missing data (NA). For categorical variables, the algorithm for identifying boundary elements is simple: if any of a cell’s neighbors—based on the queen criterion (i.e., eight neighboring cells, including diagonal cells)—belongs to a different category, the cell is classified as a boundary element. For quantitative variables, boundaries exist where two different but internally homogeneous areas neighbor one another (e.g., an area with very low values meets an area with very high values), but these boundaries are generally fuzzy; they represent steep transitions that can still occur over the width of multiple cells. Therefore, we define boundary elements as the cells with the highest boundary intensity values, with the threshold set by the user. Boundary intensity values can be calculated through several different methods, including the magnitude of change across cells or the probability that cells belong to each neighboring spatial group (see details below). Boundaries are defined here as subgraphs of boundary elements, or contiguous cells that are all marked as boundary elements (Figure <a href="#fig:fig-1">1</a>).</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-1"></span>
<img src="figures/fig1.png" alt="10 by 10 matrix of squares, outlined in black. Most squares are white, with a few light grat squares in the top right and bottom left. There are three sets of purple squares, each making curved lines 1 to 2 squares wide. The sets pf purple squares are internally connected by blue lines." width="50%" />
<p class="caption">
Figure 1: Example boundary subgraphs. Gray cells are missing values, white cells are non-boundary cells, and purple cells are boundary elements. Subgraphs are each represented using a line connecting all the boundary element cells that comprise them.
</p>
</div>
</div>
<p>Boundary intensities for variables with landscape-level patterns can be calculated in a number of ways, including: lattice- and triangulation-wombling <span class="citation" data-cites="fortin1996 jacquez1995 st-louis2004 strydom2023">(<a href="#ref-jacquez1995" role="doc-biblioref">Jacquez 1995</a>; <a href="#ref-fortin1996" role="doc-biblioref">Fortin et al. 1996</a>; <a href="#ref-st-louis2004" role="doc-biblioref">St-Louis et al. 2004</a>; <a href="#ref-strydom2023" role="doc-biblioref">Strydom and Poisot 2023</a>)</span>, fuzzy set modeling <span class="citation" data-cites="jacquez1995">(<a href="#ref-jacquez1995" role="doc-biblioref">Jacquez 1995</a>)</span>, Monmonier’s algorithm <span class="citation" data-cites="manni2004">(<a href="#ref-manni2004" role="doc-biblioref">Manni et al. 2004</a>)</span>, spatial Bayesian clustering <span class="citation" data-cites="caye2016 safner2011">(<a href="#ref-safner2011" role="doc-biblioref">Safner et al. 2011</a>; <a href="#ref-caye2016" role="doc-biblioref">Caye et al. 2016</a>)</span>, agglomeration of inner lines <span class="citation" data-cites="wei2019">(<a href="#ref-wei2019" role="doc-biblioref">Wei and Larsen 2019</a>)</span>, and removal of outer lines <span class="citation" data-cites="wei2019">(<a href="#ref-wei2019" role="doc-biblioref">Wei and Larsen 2019</a>)</span>. For quantitative variables, BoundaryStats will accept raster objects with the spatial variable directly or boundary intensity values calculated from these or other methods. If given boundary intensity values, boundary elements will be classified directly using the top percent of values. The default proportion of values is 0.2, though this threshold can be changed by the user. When given the variable directly, BoundaryStats will use the Sobel-Feldman operator to calculate the boundary intensity. In accepting either the variables or boundary intensities, there is flexibility for users to define boundaries using relevant metrics for their data.</p>
<p>The Sobel-Feldman operator is commonly used for edge detection in computer vision applications. It approximates the magnitude of the partial derivative (i.e., rate of change) across each cell using the following kernels:
<span class="math display">\[\begin{align}
  G_x &amp;= \begin{pmatrix}
  1 &amp; 0 &amp; -1 \\
  2 &amp; 0 &amp; -2 \\
  1 &amp; 0 &amp; -1
  \end{pmatrix} * A \\
  G_y &amp;= \begin{pmatrix}
  1 &amp; 2 &amp; 1 \\
  0 &amp; 0 &amp; 0 \\
  -1 &amp; -2 &amp; -1
  \end{pmatrix} * A
  \end{align}\]</span>
where <span class="math inline">\(A\)</span> is the input raster cell and its queen neighbors, and <span class="math inline">\(G_{x}\)</span> and <span class="math inline">\(G_{y}\)</span> are the rates of variable change in the horizontal or vertical directions, respectively. The boundary intensity value is the overall rate of change:
<span class="math display">\[\begin{equation}
  G = \sqrt{G_x^2 + G_y^2}
\end{equation}\]</span></p>
<h2 data-number="3" id="statistics"><span class="header-section-number">3</span> Statistics</h2>
<p>BoundaryStats runs two boundary statistics and three boundary overlap statistics, as initially described in Jacquez <span class="citation" data-cites="jacquez1995">(<a href="#ref-jacquez1995" role="doc-biblioref">1995</a>)</span> and Fortin et al. <span class="citation" data-cites="fortin1996">(<a href="#ref-fortin1996" role="doc-biblioref">1996</a>)</span>. Below, I describe these five statistics.</p>
<h3 data-number="3.1" id="number-of-subgraphs"><span class="header-section-number">3.1</span> Number of subgraphs</h3>
<p>The first boundary statistic is the number of subgraphs, which describes the number of boundaries on the landscape for a variable. In a raster of boundary elements, it is the number of unique subgraphs, or sets of contiguous boundary element cells (three subgraphs each in Figure <a href="#fig:fig-2">2</a>, Panels A and B).</p>
<h3 data-number="3.2" id="longest-subgraph"><span class="header-section-number">3.2</span> Longest subgraph</h3>
<p>The other boundary statistic included here is the length of the longest subgraph, or boundary. The function calculates the longest length across each subgraph, then converts the length to distance based on the cell resolution and the projection of the raster. The length of the longest subgraph is then retained (Figure <a href="#fig:fig-2">2</a>, Panel C).</p>
<h3 data-number="3.3" id="direct-overlap"><span class="header-section-number">3.3</span> Direct overlap</h3>
<p>The direct overlap statistic is a count of the number of overlapping boundary elements of two variables, when the two raster objects are overlaid (Figure <a href="#fig:fig-2">2</a>, Panel D).</p>
<h3 data-number="3.4" id="mean-minimum-distance-between-boundaries"><span class="header-section-number">3.4</span> Mean minimum distance between boundaries</h3>
<p>This statistic describes the spatial proximity between boundaries of variables <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span>, as defined by the mean distance to the nearest boundary element of the other variable. Spatial relationships between boundaries may not result in direct overlap, so this statistic accounts for potential correlations in non-overlapping boundaries. For each boundary element in <span class="math inline">\(x\)</span>, the function calculates the distance to the nearest boundary element in <span class="math inline">\(y\)</span>, then repeats the inverse for each boundary element in <span class="math inline">\(y\)</span> (Figure <a href="#fig:fig-2">2</a>, Panels E and F). It then takes the mean of these minimum distances across all boundary elements in both raster objects:
<span class="math display">\[\begin{equation}
  O_{xy} = \frac{{\sum_{i=1}^{N_x}\text{min}(d_i)}+{\sum_{j=1}^{N_y}\text{min}(d_j)}}{N_x+N_y}
\end{equation}\]</span>
where <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> are boundary elements for variables <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span>, respectively; <span class="math inline">\(\text{min}(d_{i})\)</span> is the minimum distance between boundary element <span class="math inline">\(i\)</span> to a boundary element for <span class="math inline">\(y\)</span>; <span class="math inline">\(\text{min}(d_{j})\)</span> is the minimum distance between boundary element <span class="math inline">\(j\)</span> to a boundary element in <span class="math inline">\(x\)</span>; and <span class="math inline">\(N_{x}\)</span> and <span class="math inline">\(N_{y}\)</span> are the number of boundary elements for <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span>, respectively.</p>
<h3 data-number="3.5" id="mean-minimum-distance-from-boundary-x-to-boundary-y"><span class="header-section-number">3.5</span> Mean minimum distance from boundary x to boundary y</h3>
<p>This statistic describes the mean distance from boundary elements in <span class="math inline">\(x\)</span> to the nearest boundary element in <span class="math inline">\(y\)</span>. It is an indicator for whether the boundaries in <span class="math inline">\(x\)</span> depend on <span class="math inline">\(y\)</span>. The reciprocal nature of the previous statistic implies some reciprocity of effect, as opposed to the unidirectionality implicit here. For each boundary element in the raster for <span class="math inline">\(x\)</span>, the function calculates the distance to the nearest boundary element of <span class="math inline">\(y\)</span>, then takes the mean across all boundary elements in <span class="math inline">\(x\)</span> (Figure <a href="#fig:fig-2">2</a>, Panel E):
<span class="math display">\[\begin{equation}
  O_{x} = \frac{\sum_{i=1}^{N_x}\text{min}(d_i)}{N_x}
\end{equation}\]</span></p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-2"></span>
<img src="figures/fig2.png" alt="Six squares labeled A, B, and C on the top row and D, E, and F on the bottom row. Each square is a 10 by 10 matrix of squares with black outlines. All the matrices have 4 light gray squares in the top right and 5 light gray squares in the bottom left. A is the same as figure 1, but without the blue lines. B is like A, except that there are light blue squares instead of purple, and the blue squares are configured differently. There are still three sets of touching blue squares. C is identical to A, except it shows a black line indicating the longest straight line that can be drawn along a set of touching purple squares. D shows what A and B look like overlayed one another. The squares where the purple light blue overlap are now dark blue. Each dark blue square has a white dot in the center. E is the same as D, except without the white dots, and with arrows on them. The arrows start from each purple square and points to the nearest light blue or dark blue square. Each dark blue square has a white circular arrow that points to itself. F is like E, except that the arrows point from each light blue square to the nearest purple or dark blue square." width="100%" />
<p class="caption">
Figure 2: Example boundaries and statistics. (A) and (B) are boundary elements for hypothetical variables A and B. White cells are non-boundaries, gray are missing values, and purple or teal are boundary elements. (C) Length of the longest subgraph. (D) Produced by overlaying cells in A and B. Dark blue cells, highlighted by white dots, are where the boundary elements overlap. (E) For every boundary element for variable A, the nearest boundary element for variable B. Circular arrows indicate distance to self. (F) For every boundary element for B, the nearest boundary element for A.
</p>
</div>
</div>
<h2 data-number="4" id="neutral-models"><span class="header-section-number">4</span> Neutral models</h2>
<p>In addition to calculating each statistic, BoundaryStats uses iterations of a neutral landscape model to determine whether the boundaries in the input landscape differ from a random landscape. Users select a neutral landscape model and number of iterations of that model to produce a null distribution of each statistic, based on the selected model and the structure of the input landscape. BoundaryStats implements three neutral landscape models: stochastic landscapes, Gaussian random fields, and modified random clusters. All three neutral landscape models draw some parameters from the original raster and simulate landscapes with similar parameters. Cells with missing values (i.e., NA values) will be ignored in all models.</p>
<p>The simplest neutral landscape model is complete stochasticity. While this model is not realistic—a complete lack of spatial autocorrelation is unlikely and may inflate the statistical significance of the observed data—we include it here as a complete null for users who are interested in a lack of spatial autocorrelation. This method takes all the cell values from the input raster and assigns each value to a random cell. Each cell in the simulated raster is assigned a value from the original dataset, with no replacement of values. The simulated raster has the exact same values as the original raster, but values are randomly placed with no spatial autocorrelation.</p>
<p>The next neutral landscape model simulates a Gaussian random field with the same degree of spatial autocorrelation as the input raster. It is suited for continuous or discrete quantitative variables. This method calculates the range of autocorrelation in the original raster by fitting a variogram using functions from <a href="https://cran.r-project.org/package=gstat">gstat</a> <span class="citation" data-cites="pebesma2004">(<a href="#ref-pebesma2004" role="doc-biblioref">Pebesma 2004</a>)</span>. The function then simulates a Gaussian random field with the same range of spatial autocorrelation, extent, and resolution as the input raster using methods from the <a href="https://cran.r-project.org/package=fields">fields</a> package.</p>
<p>The modified random cluster model is an implementation of the method described by Saura and Martínez-Millan for simulating neutral landscapes for categorical variables <span class="citation" data-cites="saura2000">(<a href="#ref-saura2000" role="doc-biblioref">Saura and Martínez-Millan 2000</a>)</span>. The first step is a percolated raster (Figure <a href="#fig:fig-3">3</a>, Panel A). Each cell is assigned a value <span class="math inline">\(0 \le x \le 1\)</span> from a uniform distribution, and cells with values above a threshold probability <span class="math inline">\(p\)</span> are marked. <span class="math inline">\(p\)</span> is defined by the user, and higher values of <span class="math inline">\(p\)</span> result in larger cluster sizes in the final simulated raster. Next, contiguous sets of marked cells are grouped into clusters, using the rook criterion (i.e., neighbors are the four edge-touching neighbors) (Figure <a href="#fig:fig-3">3</a>, Panel B). Clusters are then assigned a category (Figure <a href="#fig:fig-3">3</a>, Panel C). Categories from the input raster are chosen one at a time, and random clusters are assigned to that category. When the proportion of that category in the simulated raster reaches the proportion in the original raster, clusters are then assigned to the next category, until all the clusters are assigned. In the last step, the unmarked cells are categorized based on the most frequent category among their neighbors using the queen criterion (Figure <a href="#fig:fig-3">3</a>, Panel D). If there is a tie between two categories, one of the tied categories is picked at random. If all neighbors are unassigned, a random category is picked; probabilities for each category are based on their proportions in the input raster.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-3"></span>
<img src="figures/fig3.png" alt="Four squares in a horizontal row labeled A through D. A is a 10 by 10 matrix of squares outlined in black. Half the squares are white and half are dark blue. B is the same, except that neighboring blue cells have been merged, so there are no longer black boundaries separating them. In C, the formerly dark blue polygons have changed color to light purple, dark purple, light blue, and medium blue. D has the same colors as C, but the white squares have been merged into neighboring polygons, so that there are only polygons and no individual squares." width="100%" />
<p class="caption">
Figure 3: Modified random cluster procedure, adapted from Saura and Martínez-Millan 2000. (A) Percolated raster with p = 0.5. (B) Marked cells merged into clusters. (C) Clusters assigned a category. (D) Unmarked cells filled based on neighbors.
</p>
</div>
</div>
<h2 data-number="5" id="implementation-and-example"><span class="header-section-number">5</span> Implementation and example</h2>
<p>Data in this example are from Luo et al. <span class="citation" data-cites="luo2024">(<a href="#ref-luo2024" role="doc-biblioref">2024</a>)</span>, in which the authors hypothesized that song divergence is facilitating genetic divergence in white-crowned sparrows through speciation by sexual selection. The data below are song boundaries and genetic admixture interpolations from the study. White-crowned sparrows sing different songs that vary across the landscape, and song boundaries are the spatial transitions between two song groups. The boundary intensities between song groups were estimated using GeoOrigins <span class="citation" data-cites="hulme-beaman2020">(<a href="#ref-hulme-beaman2020" role="doc-biblioref">Hulme-Beaman et al. 2020</a>)</span>, based on the acoustic dissimilarity and spatial relationship between recorded songs. Genetic admixture is the estimated proportion of an individual’s genetic material from different populations. In this case, there are two populations (north and south), so the admixture coefficients are the estimated proportions of genetic material from the northern population. The admixture coefficients of individual birds were estimated in fastSTRUCTURE <span class="citation" data-cites="raj2014">(<a href="#ref-raj2014" role="doc-biblioref">Raj et al. 2014</a>)</span>, and the values were interpolated using local kriging in <a href="https://cran.r-project.org/package=gstat">gstat</a>.</p>
<h3 data-number="5.1" id="read-in-data"><span class="header-section-number">5.1</span> Read in data</h3>
<p>Read in raster data to <a href="https://cran.r-project.org/package=terra">terra</a> <span class="citation" data-cites="hijmans2023">(<a href="#ref-hijmans2023" role="doc-biblioref">Hijmans 2023</a>)</span> <code>SpatRaster</code> objects (Figure <a href="#fig:fig-4">4</a>). The two objects need the same projection, extent, and resolution.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://rspatial.org/'>terra</a></span><span class='op'>)</span></span>
<span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://magrittr.tidyverse.org'>magrittr</a></span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>songs</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rspatial.github.io/terra/reference/rast.html'>rast</a></span><span class='op'>(</span><span class='st'>'data/2010_2022_song_boundaries.asc'</span><span class='op'>)</span></span>
<span><span class='va'>genetic</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rspatial.github.io/terra/reference/rast.html'>rast</a></span><span class='op'>(</span><span class='st'>'data/genetic_interpolation.asc'</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span></span>
<span>  <span class='fu'><a href='https://rspatial.github.io/terra/reference/resample.html'>resample</a></span><span class='op'>(</span><span class='va'>.</span>, <span class='va'>songs</span><span class='op'>)</span></span>
<span><span class='va'>songs</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rspatial.github.io/terra/reference/crop.html'>crop</a></span><span class='op'>(</span><span class='va'>songs</span>, <span class='va'>genetic</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span></span>
<span>  <span class='fu'><a href='https://rspatial.github.io/terra/reference/mask.html'>mask</a></span><span class='op'>(</span><span class='va'>.</span>, <span class='va'>genetic</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-4"></span>
<img src="figures/fig4.png" alt="Two maps of the coast of California, centered near the San Francisco Bay. The land is gray and the ocean is white with gridded lines for latitude and longitude. A strip of coast is colored on both maps. On the left map, dark blue lines run from northeast to southwest, with some lines darker than others. On the right map, there is a gradient of color. The top of the strip of color is purple, then around Monterey Bay (two-thirds of the way south in the colored area), the color becomes light blue, and in the south it is dark blue." width="75%" />
<p class="caption">
Figure 4: Maps of (A) song boundary intensity and (B) genetic admixture between two populations.
</p>
</div>
</div>
<h3 data-number="5.2" id="calculate-spatial-boundaries-for-variables"><span class="header-section-number">5.2</span> Calculate spatial boundaries for variables</h3>
<p>The first step of the analysis is to define which cells are boundary elements (i.e., part of a boundary) using the <code>define_boundary</code> function. By default, the function takes quantitative variables (<code>cat = FALSE</code>). For quantitative variables, boundary intensity can be calculated however the user chooses; if the input raster already contains boundary intensities, the argument <code>calculate_intensity</code> should be set to <code>FALSE</code> (default). Users can also set calculate_intensity to <code>TRUE</code> to use the Sobel-Feldman operator to calculate the boundary intensities.</p>
<p>The song raster already contains boundary intensity values, from which boundary elements can be directly determined. But the values in the genetic raster are the trait data, so boundary intensity needs to be calculated from the genetic admixture coefficient values (<code>calculate_intensity = TRUE</code>).</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>BoundaryStats</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>song_boundaries</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/BoundaryStats/man/define_boundary.html'>define_boundary</a></span><span class='op'>(</span><span class='va'>songs</span><span class='op'>)</span></span>
<span><span class='va'>genetic_boundaries</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/BoundaryStats/man/define_boundary.html'>define_boundary</a></span><span class='op'>(</span><span class='va'>genetic</span>, calculate_intensity <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<h3 data-number="5.3" id="plot-boundary-overlap"><span class="header-section-number">5.3</span> Plot boundary overlap</h3>
<p>This optional step is to visualize where the boundaries of the two variables are overlapping using <code>plot_boundary</code> (Figure <a href="#fig:fig-5">5</a>). The function is a wrapper function for <code>ggplot</code> from <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> <span class="citation" data-cites="wickham2016">(<a href="#ref-wickham2016" role="doc-biblioref">Wickham 2016</a>)</span>, and the colors and trait names can optionally be customized. If <code>output_raster</code> is <code>TRUE</code> (default is <code>FALSE</code>), then the function will return a <code>SpatRaster</code> object with one layer that includes boundary elements for each trait and where the boundary elements overlap.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/BoundaryStats/man/plot_boundary.html'>plot_boundary</a></span><span class='op'>(</span><span class='va'>genetic_boundaries</span>, <span class='va'>song_boundaries</span>, trait_names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Genetic"</span>,</span>
<span>    <span class='st'>"Song"</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:fig-5"></span>
<img src="RJ-2025-025_files/figure-html5/fig-5-1.png" alt="Output of the plot\_boundary function." width="624" />
<p class="caption">
Figure 5: Output of the plot_boundary function.
</p>
</div>
</div>
<h3 data-number="5.4" id="create-null-distributions-for-statistics"><span class="header-section-number">5.4</span> Create null distributions for statistics</h3>
<p>For both boundary statistics, use the function <code>boundary_null_distrib.</code> For the three overlap statistics, use the function <code>overlap_null_distrib.</code> Both functions simulate random iterations of a raster based on the specified neutral landscape model and input data. Statistics are calculated for each iteration, and custom null probability distributions are calculated based on the iterations. The resulting objects will be used for the statistical tests described in the section below.</p>
<p>Both functions take the <code>SpatRaster</code> object(s), a neutral landscape model, and the number of iterations. Further arguments may be required, depending on these arguments. For <code>overlap_null_distrib</code>, separate models can be specified for the two variables, and the variable in the first argument is assumed to depend on the variable in the second argument. The argument <code>rand_both</code> specifies whether the function should simulate random landscapes for the second <code>SpatRaster</code> object; since some hypotheses assume <span class="math inline">\(x\)</span> depends on a specific underlying distribution of boundaries in <span class="math inline">\(y\)</span>, users can choose to keep boundaries for <span class="math inline">\(y\)</span> static for each iteration. For this example, the genetic boundary is hypothesized to depend on song boundaries. Therefore, the <code>SpatRaster</code> object containing the genetic admixture interpolation is the first argument, and I keep the song boundaries static (<code>rand_both = FALSE</code>).</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>song_boundary_null</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/BoundaryStats/man/boundary_null_distrib.html'>boundary_null_distrib</a></span><span class='op'>(</span><span class='va'>songs</span>, calculate_intensity <span class='op'>=</span> <span class='cn'>FALSE</span>, cat <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span>    n_iterations <span class='op'>=</span> <span class='fl'>100</span>, threshold <span class='op'>=</span> <span class='fl'>0.2</span>, model <span class='op'>=</span> <span class='st'>"gaussian"</span><span class='op'>)</span></span>
<span><span class='va'>genetic_boundary_null</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/BoundaryStats/man/boundary_null_distrib.html'>boundary_null_distrib</a></span><span class='op'>(</span><span class='va'>genetic</span>, calculate_intensity <span class='op'>=</span> <span class='cn'>TRUE</span>,</span>
<span>    cat <span class='op'>=</span> <span class='cn'>FALSE</span>, n_iterations <span class='op'>=</span> <span class='fl'>100</span>, threshold <span class='op'>=</span> <span class='fl'>0.2</span>, model <span class='op'>=</span> <span class='st'>"gaussian"</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>boundary_overlap_null</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/BoundaryStats/man/overlap_null_distrib.html'>overlap_null_distrib</a></span><span class='op'>(</span><span class='va'>genetic</span>, <span class='va'>songs</span>, rand_both <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span>    n_iterations <span class='op'>=</span> <span class='fl'>100</span>, x_calculate_intensity <span class='op'>=</span> <span class='cn'>TRUE</span>, threshold <span class='op'>=</span> <span class='fl'>0.2</span>, x_model <span class='op'>=</span> <span class='st'>"gaussian"</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<h3 data-number="5.5" id="run-statistical-tests"><span class="header-section-number">5.5</span> Run statistical tests</h3>
<p>The two functions for boundary statistics require only the raster with boundary elements and the matching null distribution object, produced by <code>boundary_null_distrib</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/BoundaryStats/man/n_boundaries.html'>n_boundaries</a></span><span class='op'>(</span><span class='va'>song_boundaries</span>, <span class='va'>song_boundary_null</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>n_boundary    p-value 
        13          0 </code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/BoundaryStats/man/longest_boundary.html'>longest_boundary</a></span><span class='op'>(</span><span class='va'>song_boundaries</span>, <span class='va'>song_boundary_null</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>longest_boundary          p-value 
        45260.75             0.03 </code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/BoundaryStats/man/n_boundaries.html'>n_boundaries</a></span><span class='op'>(</span><span class='va'>genetic_boundaries</span>, <span class='va'>genetic_boundary_null</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>n_boundary    p-value 
         1          0 </code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/BoundaryStats/man/longest_boundary.html'>longest_boundary</a></span><span class='op'>(</span><span class='va'>genetic_boundaries</span>, <span class='va'>genetic_boundary_null</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>longest_boundary          p-value 
        61454.64             0.01 </code></pre>
</div>
<p>The functions for boundary overlap statistics also take the boundary element rasters and null distribution as arguments. In this case, it requires two boundary element <code>SpatRaster</code> objects, one for each variable. The order of the variables should match the order used in <code>overlap_null_distrib</code>. I am interested in whether the genetic boundary depends on song boundaries (i.e., a unidirectional trend), so I am using the <code>n_overlap_boundaries</code> (direct overlap) and <code>average_min_x_to_y</code> (<span class="math inline">\(O_{x}\)</span>) tests but not the <code>average_min_distance</code> (<span class="math inline">\(O_{xy}\)</span>) test. The genetic boundary raster is the first argument, and the song boundary raster is the second argument.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/BoundaryStats/man/n_overlap_boundaries.html'>n_overlap_boundaries</a></span><span class='op'>(</span><span class='va'>genetic_boundaries</span>, <span class='va'>song_boundaries</span>, <span class='va'>boundary_overlap_null</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>n_overlapping       p-value 
           44             0 </code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/BoundaryStats/man/average_min_x_to_y.html'>average_min_x_to_y</a></span><span class='op'>(</span><span class='va'>genetic_boundaries</span>, <span class='va'>song_boundaries</span>, <span class='va'>boundary_overlap_null</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>avg_min_x_to_y        p-value 
      1911.751          0.000 </code></pre>
</div>
<h3 data-number="5.6" id="interpretation-of-example-data-output"><span class="header-section-number">5.6</span> Interpretation of example data output</h3>
<p>When analyzing the data from Luo et al. <span class="citation" data-cites="luo2024">(<a href="#ref-luo2024" role="doc-biblioref">2024</a>)</span>, the boundary statistics for the genetic data were significant, suggesting the presence of one cohesive genetic boundary. The boundary statistics for the song data were less clear, as the p-values were at or around 0.05. If the analysis was repeated with more iterations of the neutral landscape model, it may suggest multiple cohesive song boundaries. Results from the boundary overlap statistics show significant direct overlap between the genetic and song boundaries and spatial proximity from the genetic boundary to song boundaries, suggesting a spatial correlation between boundaries of the two variables. While boundary overlap statistics can only demonstrate a correlation between boundaries, the results are generally consistent with the hypothesis that song boundaries are facilitating a coincident genetic boundary.</p>
<h2 data-number="6" id="summary"><span class="header-section-number">6</span> Summary</h2>
<p>BoundaryStats implements five boundary overlap statistics. Boundary analyses like the boundary overlap statistics implemented here can be used across many contexts that make use of spatially distributed data. For example, spatial ecologists and epidemiologists can use boundary overlap statistics to assess whether environmental variables are influencing the distribution of organismal traits or disease occurrences. Environmental influences can, in some cases, be detected through the co-occurrence and coincidence of geographic boundaries; environmental boundaries may produce boundaries in the variables of interest. As such, this new open-source, cross-platform implementation will make boundary statistical methods more widely accessible to researchers.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h3 class="appendix" data-number="6.1" id="supplementary-materials"><span class="header-section-number">6.1</span> Supplementary materials</h3>
<p>Supplementary materials are available in addition to this article. It can be downloaded at
<a href="RJ-2025-025.zip">RJ-2025-025.zip</a></p>
<h3 class="appendix" data-number="6.2" id="cran-packages-used"><span class="header-section-number">6.2</span> CRAN packages used</h3>
<p><a href="https://cran.r-project.org/package=BoundaryStats">BoundaryStats</a>, <a href="https://cran.r-project.org/package=gstat">gstat</a>, <a href="https://cran.r-project.org/package=fields">fields</a>, <a href="https://cran.r-project.org/package=terra">terra</a>, <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a></p>
<h3 class="appendix" data-number="6.3" id="cran-task-views-implied-by-cited-packages"><span class="header-section-number">6.3</span> CRAN Task Views implied by cited packages</h3>
<p><a href="https://cran.r-project.org/view=ChemPhys">ChemPhys</a>, <a href="https://cran.r-project.org/view=NetworkAnalysis">NetworkAnalysis</a>, <a href="https://cran.r-project.org/view=Phylogenetics">Phylogenetics</a>, <a href="https://cran.r-project.org/view=Spatial">Spatial</a>, <a href="https://cran.r-project.org/view=SpatioTemporal">SpatioTemporal</a>, <a href="https://cran.r-project.org/view=TeachingStatistics">TeachingStatistics</a></p>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-adimalla2020" class="csl-entry" role="listitem">
N. Adimalla, J. Chen and H. Qian. Spatial characteristics of heavy metal contamination and potential human health risk assessment of urban soils: A case study from an urban region of <span>South India</span>. <em>Ecotoxicology and Environmental Safety</em>, 194(110406): 2020. URL <a href="https://linkinghub.elsevier.com/retrieve/pii/S0147651320302451">https://linkinghub.elsevier.com/retrieve/pii/S0147651320302451</a>.
</div>
<div id="ref-bowman2023" class="csl-entry" role="listitem">
D. M. J. S. Bowman, S. Ondei, A. Lucieer, S. Foyster and L. D. Prior. Forest-sedgeland boundaries are historically stable and resilient to wildfire at <span>Blakes Opening</span> in the <span>Tasmanian Wilderness World Heritage Area</span>, <span>Australia</span>. <em>Landscape Ecology</em>, 38: 205–222, 2023. DOI <a href="https://doi.org/10.1007/s10980-022-01558-x">10.1007/s10980-022-01558-x</a>.
</div>
<div id="ref-caye2016" class="csl-entry" role="listitem">
K. Caye, T. M. Deist, H. Martins, O. Michel and O. François. <span>TESS3</span>: Fast inference of spatial population structure and genome scans for selection. <em>Molecular Ecology Resources</em>, 16(2): 540–548, 2016. DOI <a href="https://doi.org/10.1111/1755-0998.12471">10.1111/1755-0998.12471</a>.
</div>
<div id="ref-fortin1996" class="csl-entry" role="listitem">
M.-J. Fortin, P. Drapeau and G. M. Jacquez. Quantification of the spatial co-occurrences of ecological boundaries. <em>Oikos</em>, 77(1): 51–60, 1996. URL <a href="https://www.jstor.org/stable/3545584?origin=crossref">https://www.jstor.org/stable/3545584?origin=crossref</a>.
</div>
<div id="ref-neighbou2012" class="csl-entry" role="listitem">
<span class="smallcaps">M. van Ham, D. Manley, N. Bailey, L. Simpson and D. Maclennan, eds</span>. <em>Neighbourhood effects research: New perspectives.</em> Springer Netherlands, 2012. URL <a href="http://dx.doi.org/10.1007/978-94-007-2309-2">http://dx.doi.org/10.1007/978-94-007-2309-2</a>.
</div>
<div id="ref-hijmans2023" class="csl-entry" role="listitem">
R. J. Hijmans. <em>Terra: Spatial data analysis.</em> 2023. URL <a href="https://CRAN.R-project.org/package=terra">https://CRAN.R-project.org/package=terra</a>.
</div>
<div id="ref-hong2021" class="csl-entry" role="listitem">
B. Hong, B. J. Bonczak, A. Gupta, L. E. Thorpe and C. E. Kontokosta. Exposure density and neighborhood disparities in <span>COVID-19</span> infection risk. <em>Proceedings of the National Academy of Sciences</em>, 118(13): e2021258118, 2021. DOI <a href="https://doi.org/10.1073/pnas.2021258118">10.1073/pnas.2021258118</a>.
</div>
<div id="ref-hulme-beaman2020" class="csl-entry" role="listitem">
A. Hulme-Beaman, A. Rudzinski, J. E. J. Cooper, R. F. Lachlan, K. Dobney and M. G. Thomas. <span>geoorigins</span>: A new method and <span>R</span> package for trait mapping and geographic provenancing of specimens without categorical constraints. <em>Methods in Ecology and Evolution</em>, 11(10): 1247–1257, 2020. DOI <a href="https://doi.org/10.1111/2041-210X.13444">10.1111/2041-210X.13444</a>.
</div>
<div id="ref-jacquez2010" class="csl-entry" role="listitem">
G. M. Jacquez. Geographic boundary analysis in spatial and spatio-temporal epidemiology: Perspective and prospects. <em>Spatial and Spatio-temporal Epidemiology</em>, 1(4): 207–218, 2010. DOI <a href="https://doi.org/10.1016/j.sste.2010.09.003">10.1016/j.sste.2010.09.003</a>.
</div>
<div id="ref-jacquez1995" class="csl-entry" role="listitem">
G. M. Jacquez. The map comparison problem: Tests for the overlap of geographic boundaries. <em>Statistics in Medicine</em>, 14(21-22): 2343–2361, 1995. DOI <a href="https://doi.org/10.1002/sim.4780142107">10.1002/sim.4780142107</a>.
</div>
<div id="ref-jacquez2000" class="csl-entry" role="listitem">
G. M. Jacquez, S. Maruca and M.-Josée. Fortin. From fields to objects: A review of geographic boundary analysis. <em>Journal of Geographical Systems</em>, 2: 221–241, 2000. DOI <a href="https://doi.org/10.1007/PL00011456">10.1007/PL00011456</a>.
</div>
<div id="ref-luo2024" class="csl-entry" role="listitem">
A. R. Luo, S. Lipshutz, J. Phillips, R. T. Brumfield and E. P. Derryberry. Song and genetic divergence within a subspecies of white-crowned sparrow (<span><em>Zonotrichia leucophrys nuttalli</em></span>). <em>PLoS ONE</em>, 19(5): e0304348, 2024. DOI <a href="https://doi.org/10.1371/journal.pone.0304348">10.1371/journal.pone.0304348</a>.
</div>
<div id="ref-manni2004" class="csl-entry" role="listitem">
F. Manni, E. Guérard and E. Heyer. Geographic patterns of (genetic, morphological, linguistic) variation: How barriers can be detected with <span class="nocase">Monmonier’s</span>algorithm. <em>Human Biology</em>, 76(2): 173–190, 2004. DOI <a href="https://doi.org/10.1353/hub.2004.0034">10.1353/hub.2004.0034</a>.
</div>
<div id="ref-pebesma2004" class="csl-entry" role="listitem">
E. J. Pebesma. Multivariable geostatistics in <span>S</span>: The gstat package. <em>Computers &amp; Geosciences</em>, 30(7): 683–691, 2004. DOI <a href="https://doi.org/10.1016/j.cageo.2004.03.012">10.1016/j.cageo.2004.03.012</a>.
</div>
<div id="ref-polakowska2012" class="csl-entry" role="listitem">
A. E. Polakowska, M.-J. Fortin and A. Couturier. Quantifying the spatial relationship between bird species<span>’</span> distributions and landscape feature boundaries in southern <span>Ontario, Canada</span>. <em>Landscape Ecology</em>, 27(10): 1481–1493, 2012. DOI <a href="https://doi.org/10.1007/s10980-012-9804-6">10.1007/s10980-012-9804-6</a>.
</div>
<div id="ref-raj2014" class="csl-entry" role="listitem">
A. Raj, M. Stephens and J. K. Pritchard. <span>fastSTRUCTURE</span>: Variational inference of population structure in large <span>SNP</span> data sets. <em>Genetics</em>, 197(2): 573–589, 2014. URL <a href="https://academic.oup.com/genetics/article/197/2/573/6074271">https://academic.oup.com/genetics/article/197/2/573/6074271</a>.
</div>
<div id="ref-safner2011" class="csl-entry" role="listitem">
T. Safner, M. P. Miller, B. H. McRae, M.-J. Fortin and S. Manel. Comparison of <span>Bayesian</span> clustering and edge detection methods for inferring boundaries in landscape genetics. <em>International Journal of Molecular Sciences</em>, 12(2): 865–889, 2011. URL <a href="http://www.mdpi.com/1422-0067/12/2/865">http://www.mdpi.com/1422-0067/12/2/865</a>.
</div>
<div id="ref-saura2000" class="csl-entry" role="listitem">
S. Saura and J. Martínez-Millan. Landscape patterns simulation with a modified random clusters method. <em>Landscape Ecology</em>, 15: 661–678, 2000. DOI <a href="https://doi.org/10.1023/A:1008107902848">10.1023/A:1008107902848</a>.
</div>
<div id="ref-st-louis2004" class="csl-entry" role="listitem">
V. St-Louis, M.-J. Fortin and A. Desrochers. Spatial association between forest heterogeneity and breeding territory boundaries of two forest songbirds. <em>Landscape Ecology</em>, 19(6): 591–601, 2004. URL <a href="http://link.springer.com/10.1023/B:LAND.0000042849.63040.a9">http://link.springer.com/10.1023/B:LAND.0000042849.63040.a9</a>.
</div>
<div id="ref-strydom2023" class="csl-entry" role="listitem">
T. Strydom and T. Poisot. SpatialBoundaries.jl: Edge detection using spatial wombling. <em>Ecography</em>, 2023(5): e06609, 2023. URL <a href="https://onlinelibrary.wiley.com/doi/10.1111/ecog.06609">https://onlinelibrary.wiley.com/doi/10.1111/ecog.06609</a>.
</div>
<div id="ref-tarroso2014" class="csl-entry" role="listitem">
P. Tarroso, R. J. Pereira, F. Martínez-Freiría, R. Godinho and J. C. Brito. Hybridization at an ecotone: ecological and genetic barriers between three <span>Iberian</span> vipers. <em>Molecular Ecology</em>, 23(5): 1108–1123, 2014. URL <a href="https://onlinelibrary.wiley.com/doi/10.1111/mec.12671">https://onlinelibrary.wiley.com/doi/10.1111/mec.12671</a>.
</div>
<div id="ref-wagner2013" class="csl-entry" role="listitem">
H. H. Wagner and M.-J. Fortin. A conceptual framework for the spatial analysis of landscape genetic data. <em>Conservation Genetics</em>, 14(2): 253–261, 2013. URL <a href="http://link.springer.com/10.1007/s10592-012-0391-5">http://link.springer.com/10.1007/s10592-012-0391-5</a>.
</div>
<div id="ref-wagner2005" class="csl-entry" role="listitem">
H. H. Wagner and M.-J. Fortin. Spatial analysis of landscapes: Concepts and statistics. <em>Ecology</em>, 86(8): 1975–1987, 2005. DOI <a href="https://doi.org/10.1890/04-0914">10.1890/04-0914</a>.
</div>
<div id="ref-waller1992" class="csl-entry" role="listitem">
L. A. Waller, B. W. Turnbull, L. C. Clark and P. Nasca. Chronic disease surveillance and testing of clustering of disease and exposure: Application to leukemia incidence and <span>TCE-contaminated</span> dumpsites in upstate <span>New York</span>. <em>Environmetrics</em>, 3(3): 281–300, 1992. DOI <a href="https://doi.org/10.1002/env.3170030303">10.1002/env.3170030303</a>.
</div>
<div id="ref-wei2019" class="csl-entry" role="listitem">
X. Wei and C. P. S. Larsen. Methods to detect edge effected reductions in fire frequency in simulated forest landscapes. <em>ISPRS International Journal of Geo-Information</em>, 8(6): 277, 2019. URL <a href="https://www.mdpi.com/2220-9964/8/6/277">https://www.mdpi.com/2220-9964/8/6/277</a>.
</div>
<div id="ref-wickham2016" class="csl-entry" role="listitem">
H. Wickham. <em>ggplot2: Elegant graphics for data analysis.</em> 2016. URL <a href="https://ggplot2.tidyverse.org">https://ggplot2.tidyverse.org</a>.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="references">References</h3>
<div id="references-listing"></div>
<h3 id="reuse">Reuse</h3>
<p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
<h3 id="citation">Citation</h3>
<p>For attribution, please cite this work as</p>
<pre class="citation-appendix short">Luo, "BoundaryStats: An R Package to Calculate Boundary Overlap Statistics", The R Journal, 2025</pre>
<p>BibTeX citation</p>
<pre class="citation-appendix long">@article{RJ-2025-025,
  author = {Luo, Amy},
  title = {BoundaryStats: An R Package to Calculate Boundary Overlap Statistics},
  journal = {The R Journal},
  year = {2025},
  note = {https://doi.org/10.32614/RJ-2025-025},
  doi = {10.32614/RJ-2025-025},
  volume = {17},
  issue = {3},
  issn = {2073-4859},
  pages = {1}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
