<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->



  <!--radix_placeholder_meta_tags-->
  <title>netgwas: An R Package for Network-Based Genome Wide Association Studies</title>

  <meta property="description" itemprop="description" content="Graphical models are a powerful tool in modelling and analysing complex biological associations in high-dimensional data. The R-package netgwas implements the recent methodological development on copula graphical models to (i) construct linkage maps, (ii) infer linkage disequilibrium networks from genotype data, and (iii) detect high-dimensional genotype-phenotype networks. The netgwas learns the structure of networks from ordinal data and mixed ordinal-and-continuous data. Here, we apply the functionality in netgwas to various multivariate example datasets taken from the literature to demonstrate the kind of insight that can be obtained from the package. We show that our package offers a more realistic association analysis than the classical approaches, as it discriminates between direct and induced correlations by adjusting for the effect of all other variables while performing pairwise associations. This feature controls for spurious interactions between variables that can arise from conventional approaches in a biological sense. The netgwas package uses a parallelization strategy on multi-core processors to speed-up computations."/>

  <link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>

  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2023-02-10"/>
  <meta property="article:created" itemprop="dateCreated" content="2023-02-10"/>
  <meta name="article:author" content="Pariya Behrouzi"/>
  <meta name="article:author" content="Danny Arends"/>
  <meta name="article:author" content="Ernst C. Wit"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="netgwas: An R Package for Network-Based Genome Wide Association Studies"/>
  <meta property="og:type" content="article"/>
  <meta property="og:description" content="Graphical models are a powerful tool in modelling and analysing complex biological associations in high-dimensional data. The R-package netgwas implements the recent methodological development on copula graphical models to (i) construct linkage maps, (ii) infer linkage disequilibrium networks from genotype data, and (iii) detect high-dimensional genotype-phenotype networks. The netgwas learns the structure of networks from ordinal data and mixed ordinal-and-continuous data. Here, we apply the functionality in netgwas to various multivariate example datasets taken from the literature to demonstrate the kind of insight that can be obtained from the package. We show that our package offers a more realistic association analysis than the classical approaches, as it discriminates between direct and induced correlations by adjusting for the effect of all other variables while performing pairwise associations. This feature controls for spurious interactions between variables that can arise from conventional approaches in a biological sense. The netgwas package uses a parallelization strategy on multi-core processors to speed-up computations."/>
  <meta property="og:locale" content="en_US"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="netgwas: An R Package for Network-Based Genome Wide Association Studies"/>
  <meta property="twitter:description" content="Graphical models are a powerful tool in modelling and analysing complex biological associations in high-dimensional data. The R-package netgwas implements the recent methodological development on copula graphical models to (i) construct linkage maps, (ii) infer linkage disequilibrium networks from genotype data, and (iii) detect high-dimensional genotype-phenotype networks. The netgwas learns the structure of networks from ordinal data and mixed ordinal-and-continuous data. Here, we apply the functionality in netgwas to various multivariate example datasets taken from the literature to demonstrate the kind of insight that can be obtained from the package. We show that our package offers a more realistic association analysis than the classical approaches, as it discriminates between direct and induced correlations by adjusting for the effect of all other variables while performing pairwise associations. This feature controls for spurious interactions between variables that can arise from conventional approaches in a biological sense. The netgwas package uses a parallelization strategy on multi-core processors to speed-up computations."/>

  <!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
  <meta name="citation_title" content="netgwas: An R Package for Network-Based Genome Wide Association Studies"/>
  <meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2023-011"/>
  <meta name="citation_pdf_url" content="RJ-2023-011.pdf"/>
  <meta name="citation_volume" content="14"/>
  <meta name="citation_issue" content="4"/>
  <meta name="citation_doi" content="10.32614/RJ-2023-011"/>
  <meta name="citation_journal_title" content="The R Journal"/>
  <meta name="citation_issn" content="2073-4859"/>
  <meta name="citation_firstpage" content="18"/>
  <meta name="citation_lastpage" content="37"/>
  <meta name="citation_fulltext_world_readable" content=""/>
  <meta name="citation_online_date" content="2023/02/10"/>
  <meta name="citation_publication_date" content="2023/02/10"/>
  <meta name="citation_author" content="Pariya Behrouzi"/>
  <meta name="citation_author_institution" content="Applied Mathematics and Statistics - Biometris"/>
  <meta name="citation_author" content="Danny Arends"/>
  <meta name="citation_author_institution" content="Albrecht Daniel Thaer-Institut für Agrar und Gartenbauwissenschaften"/>
  <meta name="citation_author" content="Ernst C. Wit"/>
  <meta name="citation_author_institution" content="Faculty of Science &amp; Informatics"/>
  <!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Detecting epistatic selection with partially observed genotype data by using copula graphical models;citation_publisher=Wiley Online Library;citation_volume=68;citation_author=Pariya Behrouzi;citation_author=Ernst C Wit"/>
  <meta name="citation_reference" content="citation_title=De novo construction of polyploid linkage maps using discrete graphical models;citation_publisher=Oxford University Press;citation_volume=35;citation_author=Pariya Behrouzi;citation_author=Ernst C Wit"/>
  <meta name="citation_reference" content="citation_title=R package ASMap: Efficient genetic linkage map construction and diagnosis;citation_volume=79;citation_author=Julian Taylor;citation_author=David Butler"/>
  <meta name="citation_reference" content="citation_title=Graphical models;citation_publisher=Oxford University Press, USA;citation_volume=17;citation_author=S. L. Lauritzen"/>
  <meta name="citation_reference" content="citation_title=OneMap: Software for genetic mapping in outcrossing species;citation_publisher=Wiley Online Library;citation_volume=144;citation_author=GRA Margarido;citation_author=AP Souza;citation_author=AAF Garcia"/>
  <meta name="citation_reference" content="citation_title=PERGOLA: Fast and deterministic linkage mapping of polyploids;citation_publisher=BioMed Central;citation_volume=18;citation_author=Fabian Grandke;citation_author=Soumya Ranganathan;citation_author=Nikkie Bers;citation_author=Jorn R Haan;citation_author=Dirk Metzler"/>
  <meta name="citation_reference" content="citation_title=R/qtl: QTL mapping in experimental crosses;citation_publisher=Oxford Univ Press;citation_volume=19;citation_author=Karl W Broman;citation_author=Hao Wu;citation_author=Saunak Sen;citation_author=Gary A Churchill"/>
  <meta name="citation_reference" content="citation_title=The potential of probabilistic graphical models in linkage map construction;citation_publisher=Springer;citation_author=Huange Wang;citation_author=Fred A Eeuwijk;citation_author=Johannes Jansen"/>
  <meta name="citation_reference" content="citation_title=On use of the EM for penalized likelihood estimation;citation_publisher=JSTOR;citation_author=Peter J Green"/>
  <meta name="citation_reference" content="citation_title=High-dimensional semiparametric gaussian copula graphical models;citation_publisher=Institute of Mathematical Statistics;citation_volume=40;citation_author=Han Liu;citation_author=Fang Han;citation_author=Ming Yuan;citation_author=John Lafferty;citation_author=Larry Wasserman;citation_author=High-dimensional semiparametric gaussian copula graphical models"/>
  <meta name="citation_reference" content="citation_title=Sparse inverse covariance estimation with the graphical lasso;citation_publisher=Biometrika Trust;citation_volume=9;citation_author=Jerome Friedman;citation_author=Trevor Hastie;citation_author=Robert Tibshirani"/>
  <meta name="citation_reference" content="citation_title=Sparse inverse covariance matrix estimation using quadratic approximation;citation_author=Cho-Jui Hsieh;citation_author=Inderjit S Dhillon;citation_author=Pradeep K Ravikumar;citation_author=Mátyás A Sustik"/>
  <meta name="citation_reference" content="citation_title=Divergent evolution of duplicate genes leads to genetic incompatibilities within a. thaliana;citation_publisher=American Association for the Advancement of Science;citation_volume=323;citation_author=David Bikard;citation_author=Dhaval Patel;citation_author=Claire Le Mette;citation_author=Veronica Giorgi;citation_author=Christine Camilleri;citation_author=Malcolm J Bennett;citation_author=Olivier Loudet"/>
  <meta name="citation_reference" content="citation_title=Simultaneous maximum likelihood estimation of linkage and linkage phases in outcrossing species;citation_publisher=Elsevier;citation_volume=61;citation_author=Rongling Wu;citation_author=Chang-Xing Ma;citation_author=Ian Painter;citation_author=Zhao-Bang Zeng"/>
  <meta name="citation_reference" content="citation_title=QTL mapping in five new large RIL populations of arabidopsis thaliana genotyped with consensus SNP markers;citation_volume=178;citation_author=M Simon;citation_author=O Loudet;citation_author=S Durand;citation_author=A Berard;citation_author=D Brunel;citation_author=FX Sennesal;citation_author=M Durand-Tardif;citation_author=G Pelletier;citation_author=C Camilleri"/>
  <meta name="citation_reference" content="citation_title=Reducing the bandwidth of sparse symmetric matrices;citation_publisher=ACM;citation_author=Elizabeth Cuthill;citation_author=James McKee"/>
  <meta name="citation_reference" content="citation_title=QTL mapping in new arabidopsis thaliana advanced intercross-recombinant inbred lines;citation_publisher=Public Library of Science;citation_volume=4;citation_author=Sureshkumar Balasubramanian;citation_author=Christopher Schwartz;citation_author=Anandita Singh;citation_author=Norman Warthmann;citation_author=Min Chul Kim;citation_author=Julin N Maloof;citation_author=Olivier Loudet;citation_author=Gabriel T Trainer;citation_author=Tsegaye Dabi;citation_author=Justin O Borevitz;citation_author=QTL mapping in new arabidopsis thaliana advanced intercross-recombinant inbred lines"/>
  <meta name="citation_reference" content="citation_title=Pheno2Geno-high-throughput generation of genetic markers and maps from molecular phenotypes for crosses between inbred strains;citation_publisher=BioMed Central;citation_volume=16;citation_author=Konrad Zych;citation_author=Yang Li;citation_author=Joeri K Velde;citation_author=Ronny VL Joosen;citation_author=Wilco Ligterink;citation_author=Ritsert C Jansen;citation_author=Danny Arends"/>
  <meta name="citation_reference" content="citation_title=Genetic linkage mapping of economically important traits in cultivated tetraploid potato (solanum tuberosum l.);citation_publisher=G3: Genes, Genomes, Genetics;citation_volume=5;citation_author=Alicia N Massa;citation_author=Norma C Manrique-Carpintero;citation_author=Joseph J Coombs;citation_author=Daniel G Zarka;citation_author=Anne E Boone;citation_author=William W Kirk;citation_author=Christine A Hackett;citation_author=Glenn J Bryan;citation_author=David S Douches"/>
  <meta name="citation_reference" content="citation_title=The NHGRI GWAS catalog, a curated resource of SNP-trait associations;citation_publisher=Oxford University Press;citation_volume=42;citation_author=Danielle Welter;citation_author=Jacqueline MacArthur;citation_author=Joannella Morales;citation_author=Tony Burdett;citation_author=Peggy Hall;citation_author=Heather Junkins;citation_author=Alan Klemm;citation_author=Paul Flicek;citation_author=Teri Manolio;citation_author=Lucia Hindorff;citation_author=The NHGRI GWAS catalog, a curated resource of SNP-trait associations"/>
  <meta name="citation_reference" content="citation_title=What should the genome-wide significance threshold be? Empirical replication of borderline genetic associations;citation_publisher=Oxford University Press;citation_volume=41;citation_author=Orestis A Panagiotou;citation_author=John PA Ioannidis;citation_author=Genome-Wide Significance Project"/>
  <meta name="citation_reference" content="citation_title=A variable selection method for genome-wide association studies;citation_publisher=Oxford University Press;citation_volume=27;citation_author=Qianchuan He;citation_author=Dan-Yu Lin"/>
  <meta name="citation_reference" content="citation_title=Simultaneous analysis of all SNPs in genome-wide and re-sequencing association studies;citation_publisher=Public Library of Science;citation_volume=4;citation_author=Clive J Hoggart;citation_author=John C Whittaker;citation_author=Maria De Iorio;citation_author=David J Balding"/>
  <meta name="citation_reference" content="citation_title=A lasso multi-marker mixed model for association mapping with population structure correction;citation_publisher=Oxford University Press;citation_volume=29;citation_author=Barbara Rakitsch;citation_author=Christoph Lippert;citation_author=Oliver Stegle;citation_author=Karsten Borgwardt"/>
  <meta name="citation_reference" content="citation_title=Assessing statistical significance in multivariable genome wide association analysis;citation_publisher=Oxford University Press;citation_volume=32;citation_author=Laura Buzdugan;citation_author=Markus Kalisch;citation_author=Arcadi Navarro;citation_author=Daniel Schunk;citation_author=Ernst Fehr;citation_author=Peter Bühlmann"/>
  <meta name="citation_reference" content="citation_title=A brief tour of r/qtl;citation_author=Karl W Broman"/>
  <meta name="citation_reference" content="citation_title=Systematic detection of errors in genetic linkage data;citation_publisher=Elsevier;citation_volume=14;citation_author=Stephen E Lincoln;citation_author=Eric S Lander"/>
  <meta name="citation_reference" content="citation_title=TetraploidSNPMap: Software for linkage analysis and QTL mapping in autotetraploid populations using SNP dosage data;citation_publisher=Oxford University Press US;citation_volume=108;citation_author=Christine A Hackett;citation_author=Bram Boskamp;citation_author=Athanasios Vogogias;citation_author=Katharine F Preedy;citation_author=Iain Milne"/>
  <meta name="citation_reference" content="citation_title=polymapR–linkage analysis and genetic map construction from F1 populations of outcrossing polyploids;citation_publisher=Oxford University Press;citation_volume=1;citation_author=Peter M Bourke;citation_author=Geert Geest;citation_author=Roeland E Voorrips;citation_author=Johannes Jansen;citation_author=Twan Kranenburg;citation_author=Arwa Shahin;citation_author=Richard GF Visser;citation_author=Paul Arens;citation_author=Marinus JM Smulders;citation_author=Chris Maliepaard"/>
  <meta name="citation_reference" content="citation_title=MAPMAKER: An interactive computer package for constructing primary genetic linkage maps of experimental and natural populations;citation_publisher=Elsevier;citation_volume=1;citation_author=Eric S Lander;citation_author=Philip Green;citation_author=Jeff Abrahamson;citation_author=Aaron Barlow;citation_author=Mark J Daly;citation_author=Stephen E Lincoln;citation_author=Lee Newburg"/>
  <meta name="citation_reference" content="citation_title=Novel measures of linkage disequilibrium that correct the bias due to population structure and relatedness;citation_publisher=Nature Publishing Group;citation_volume=108;citation_author=Brigitte Mangin;citation_author=Aurelie Siberchicot;citation_author=S Nicolas;citation_author=Agnes Doligez;citation_author=Patrice This;citation_author=Christine Cierco-Ayrolles"/>
  <meta name="citation_reference" content="citation_title=Gametic disequilibrium measures: Proceed with caution;citation_publisher=Genetics Soc America;citation_volume=117;citation_author=Philip W Hedrick"/>
  <meta name="citation_reference" content="citation_title=A multi-marker association method for genome-wide association studies without the need for population structure correction;citation_publisher=Nature Publishing Group;citation_volume=7;citation_author=Jonas R Klasen;citation_author=Elke Barbez;citation_author=Lukas Meier;citation_author=Nicolai Meinshausen;citation_author=Peter Bühlmann;citation_author=Maarten Koornneef;citation_author=Wolfgang Busch;citation_author=Korbinian Schneeberger"/>
  <meta name="citation_reference" content="citation_title=The genetic architecture of maize flowering time;citation_publisher=American Association for the Advancement of Science;citation_volume=325;citation_author=Edward S Buckler;citation_author=James B Holland;citation_author=Peter J Bradbury;citation_author=Charlotte B Acharya;citation_author=Patrick J Brown;citation_author=Chris Browne;citation_author=Elhan Ersoz;citation_author=Sherry Flint-Garcia;citation_author=Arturo Garcia;citation_author=Jeffrey C Glaubitz;citation_author=The genetic architecture of maize flowering time"/>
  <meta name="citation_reference" content="citation_title=Genome-wide association study of leaf architecture in the maize nested association mapping population;citation_publisher=Nature Publishing Group;citation_volume=43;citation_author=Feng Tian;citation_author=Peter J Bradbury;citation_author=Patrick J Brown;citation_author=Hsiaoyi Hung;citation_author=Qi Sun;citation_author=Sherry Flint-Garcia;citation_author=Torbert R Rocheford;citation_author=Michael D McMullen;citation_author=James B Holland;citation_author=Edward S Buckler"/>
  <meta name="citation_reference" content="citation_title=R package wgaim: QTL analysis in bi-parental populations using linear mixed models;citation_publisher=Foundation for Open Access Statistics;citation_volume=40;citation_author=Julian Taylor;citation_author=Arunas Verbyla;citation_author=R package wgaim: QTL analysis in bi-parental populations using linear mixed models"/>
  <meta name="citation_reference" content="citation_title=Dlmap: An r package for mixed model QTL and association analysis;citation_volume=50;citation_author=B Emma Huang;citation_author=Rohan Shah;citation_author=Andrew W George;citation_author=Dlmap: An r package for mixed model QTL and association analysis"/>
  <meta name="citation_reference" content="citation_title=Reconstruction of networks with direct and indirect genetic effects;citation_publisher=Genetics Soc America;citation_volume=214;citation_author=Willem Kruijer;citation_author=Pariya Behrouzi;citation_author=Daniela Bustos-Korts;citation_author=Marı́a Xosé Rodrı́guez-Álvarez;citation_author=Seyed Mahdi Mahmoudi;citation_author=Brian Yandell;citation_author=Ernst Wit;citation_author=Fred A Eeuwijk"/>
  <meta name="citation_reference" content="citation_title=Equation of state calculations by fast computing machines;citation_publisher=American Institute of Physics;citation_volume=21;citation_author=Nicholas Metropolis;citation_author=Arianna W Rosenbluth;citation_author=Marshall N Rosenbluth;citation_author=Augusta H Teller;citation_author=Edward Teller"/>
  <meta name="citation_reference" content="citation_title=Monte carlo sampling methods using markov chains and their applications;citation_publisher=Oxford University Press;citation_author=W Keith Hastings"/>
  <meta name="citation_reference" content="citation_title=Stochastic relaxation, gibbs distributions, and the bayesian restoration of images;citation_publisher=IEEE;citation_author=Stuart Geman;citation_author=Donald Geman"/>
  <meta name="citation_reference" content="citation_title=Extending the rank likelihood for semiparametric copula estimation;citation_publisher=Institute of Mathematical Statistics;citation_volume=1;citation_author=Peter D Hoff"/>
  <meta name="citation_reference" content="citation_title=Efficient estimation in the bivariate normal copula model: Normal margins are least favourable;citation_publisher=JSTOR;citation_author=Chris AJ Klaassen;citation_author=Jon A Wellner"/>
  <meta name="citation_reference" content="citation_title=Dynamic chain graph models for ordinal time series data;citation_author=Pariya Behrouzi;citation_author=Fentaw Abegaz;citation_author=Ernst C Wit"/>
  <meta name="citation_reference" content="citation_title=Inferring cellular networks using probabilistic graphical models;citation_publisher=American Association for the Advancement of Science;citation_volume=303;citation_author=Nir Friedman"/>
  <meta name="citation_reference" content="citation_title=Graphical models;citation_publisher=Institute of Mathematical Statistics;citation_volume=19;citation_author=Michael I Jordan"/>
  <meta name="citation_reference" content="citation_title=Graphical models for genetic analyses;citation_publisher=JSTOR;citation_author=Steffen L Lauritzen;citation_author=Nuala A Sheehan"/>
  <meta name="citation_reference" content="citation_title=Sparse graphical models for exploring gene expression data;citation_publisher=Elsevier;citation_volume=90;citation_author=Adrian Dobra;citation_author=Chris Hans;citation_author=Beatrix Jones;citation_author=Joseph R Nevins;citation_author=Guang Yao;citation_author=Mike West"/>
  <meta name="citation_reference" content="citation_title=Using graphical models and genomic expression data to statistically validate models of genetic regulatory networks;citation_publisher=World Scientific;citation_author=Alexander J Hartemink;citation_author=David K Gifford;citation_author=Tommi S Jaakkola;citation_author=Richard A Young"/>
  <meta name="citation_reference" content="citation_title=Selecting high-dimensional mixed graphical models using minimal AIC or BIC forests;citation_publisher=BioMed Central;citation_volume=11;citation_author=David Edwards;citation_author=Gabriel CG De Abreu;citation_author=Rodrigo Labouriau"/>
  <meta name="citation_reference" content="citation_title=QTLNetwork: Mapping and visualizing genetic architecture of complex traits in experimental populations;citation_publisher=Oxford University Press;citation_volume=24;citation_author=Jian Yang;citation_author=Chengcheng Hu;citation_author=Han Hu;citation_author=Rongdong Yu;citation_author=Zhen Xia;citation_author=Xiuzi Ye;citation_author=Jun Zhu"/>
  <meta name="citation_reference" content="citation_title=Basic statistical analysis in genetic case-control studies;citation_publisher=Nature Publishing Group;citation_volume=6;citation_author=Geraldine M Clarke;citation_author=Carl A Anderson;citation_author=Fredrik H Pettersson;citation_author=Lon R Cardon;citation_author=Andrew P Morris;citation_author=Krina T Zondervan"/>
  <meta name="citation_reference" content="citation_title=Chapter 11: Genome-wide association studies;citation_publisher=Public Library of Science San Francisco, USA;citation_volume=8;citation_author=William S Bush;citation_author=Jason H Moore"/>
  <meta name="citation_reference" content="citation_title=Comparing different statistical models and multiple testing corrections for association mapping in soybean and maize;citation_publisher=Frontiers;citation_volume=10;citation_author=Avjinder S Kaler;citation_author=Jason D Gillman;citation_author=Timothy Beissinger;citation_author=Larry C Purcell"/>
  <meta name="citation_reference" content="citation_title=A unified mixed-model method for association mapping that accounts for multiple levels of relatedness;citation_publisher=Nature Publishing Group;citation_volume=38;citation_author=Jianming Yu;citation_author=Gael Pressoir;citation_author=William H Briggs;citation_author=Irie Vroh Bi;citation_author=Masanori Yamasaki;citation_author=John F Doebley;citation_author=Michael D McMullen;citation_author=Brandon S Gaut;citation_author=Dahlia M Nielsen;citation_author=James B Holland;citation_author=A unified mixed-model method for association mapping that accounts for multiple levels of relatedness"/>
  <meta name="citation_reference" content="citation_title=Efficient control of population structure in model organism association mapping;citation_publisher=Oxford University Press;citation_volume=178;citation_author=Hyun Min Kang;citation_author=Noah A Zaitlen;citation_author=Claire M Wade;citation_author=Andrew Kirby;citation_author=David Heckerman;citation_author=Mark J Daly;citation_author=Eleazar Eskin"/>
  <meta name="citation_reference" content="citation_title=Variance component model to account for sample structure in genome-wide association studies;citation_publisher=Nature Publishing Group;citation_volume=42;citation_author=Hyun Min Kang;citation_author=Jae Hoon Sul;citation_author=Susan K Service;citation_author=Noah A Zaitlen;citation_author=Sit-yee Kong;citation_author=Nelson B Freimer;citation_author=Chiara Sabatti;citation_author=Eleazar Eskin"/>
  <meta name="citation_reference" content="citation_title=FaST linear mixed models for genome-wide association studies;citation_publisher=Nature Publishing Group;citation_volume=8;citation_author=Christoph Lippert;citation_author=Jennifer Listgarten;citation_author=Ying Liu;citation_author=Carl M Kadie;citation_author=Robert I Davidson;citation_author=David Heckerman"/>
  <meta name="citation_reference" content="citation_title=Discovering functional relationships between RNA expression and chemotherapeutic susceptibility using relevance networks;citation_publisher=National Acad Sciences;citation_volume=97;citation_author=Atul J Butte;citation_author=Pablo Tamayo;citation_author=Donna Slonim;citation_author=Todd R Golub;citation_author=Isaac S Kohane"/>
  <meta name="citation_reference" content="citation_title=R/qtl2: Software for mapping quantitative trait loci with high-dimensional data and multiparent populations;citation_publisher=Oxford University Press;citation_volume=211;citation_author=Karl W Broman;citation_author=Daniel M Gatti;citation_author=Petr Simecek;citation_author=Nicholas A Furlotte;citation_author=Pjotr Prins;citation_author=Śaunak Sen;citation_author=Brian S Yandell;citation_author=Gary A Churchill"/>
  <meta name="citation_reference" content="citation_title=Accurate genotype imputation in multiparental populations from low-coverage sequence;citation_publisher=Oxford University Press;citation_volume=210;citation_author=Chaozhi Zheng;citation_author=Martin P Boer;citation_author=Fred A Eeuwijk"/>
  <meta name="citation_reference" content="citation_title=A guide to QTL mapping with r/qtl;citation_publisher=Springer;citation_volume=46;citation_author=Karl W Broman;citation_author=Saunak Sen"/>
  <meta name="citation_reference" content="citation_title=Efficient and accurate construction of genetic linkage maps from the minimum spanning tree of a graph;citation_publisher=Public Library of Science San Francisco, USA;citation_volume=4;citation_author=Yonghui Wu;citation_author=Prasanna R Bhat;citation_author=Timothy J Close;citation_author=Stefano Lonardi"/>
  <meta name="citation_reference" content="citation_title=A nonlinear mapping for data structure analysis;citation_publisher=Ieee;citation_volume=100;citation_author=John W Sammon"/>
  <meta name="citation_reference" content="citation_title=Fast algorithm for detecting community structure in networks;citation_publisher=APS;citation_volume=69;citation_author=Mark EJ Newman"/>
  <meta name="citation_reference" content="citation_title=Bayesian structural learning of microbiota systems from count metagenomic data;citation_author=Veronica Vinciotti;citation_author=Pariya Behrouzi;citation_author=Reza Mohammadi"/>
  <meta name="citation_reference" content="citation_title=Matrix: Sparse and dense matrix classes and methods;citation_author=Douglas Bates;citation_author=Martin Maechler;citation_author=Mikael Jagan"/>
  <meta name="citation_reference" content="citation_title=Extended bayesian information criteria for gaussian graphical models;citation_volume=23;citation_author=Rina Foygel;citation_author=Mathias Drton"/>
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","date_received","journal","volume","issue","slug","packages","draft","preview","bibliography","CTV","output","pdf_url","citation_url","doi","creative_commons","csl"]}},"value":[{"type":"character","attributes":{},"value":["netgwas: An R Package for Network-Based Genome Wide Association Studies"]},{"type":"character","attributes":{},"value":["Graphical models are a powerful tool in modelling and analysing complex biological associations in high-dimensional data. The R-package netgwas implements the recent methodological development on copula graphical models to (i) construct linkage maps, (ii) infer linkage disequilibrium networks from genotype data, and (iii) detect high-dimensional genotype-phenotype networks. The netgwas learns the structure of networks from ordinal data and mixed ordinal-and-continuous data. Here, we apply the functionality in netgwas to various multivariate example datasets taken from the literature to demonstrate the kind of insight that can be obtained from the package. We show that our package offers a more realistic association analysis than the classical approaches, as it discriminates between direct and induced correlations by adjusting for the effect of all other variables while performing pairwise associations. This feature controls for spurious interactions between variables that can arise from conventional approaches in a biological sense. The netgwas package uses a parallelization strategy on multi-core processors to speed-up computations."]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","email","address"]}},"value":[{"type":"character","attributes":{},"value":["Pariya Behrouzi"]},{"type":"character","attributes":{},"value":["Applied Mathematics and Statistics - Biometris"]},{"type":"character","attributes":{},"value":["pariya.behrouzi@wur.nl\n"]},{"type":"character","attributes":{},"value":["Wageningen University and Research","6708PB, Wageningen","The Netherlands","(https://orcid.org/0000-0001-6762-5433)"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","email","address"]}},"value":[{"type":"character","attributes":{},"value":["Danny Arends"]},{"type":"character","attributes":{},"value":["Albrecht Daniel Thaer-Institut für Agrar und Gartenbauwissenschaften"]},{"type":"character","attributes":{},"value":["Danny.Arends@Gmail.com\n"]},{"type":"character","attributes":{},"value":["Humboldt-Universität zu Berlin","Invalidenstraße 42, 10115 Berlin","Germany","(https://orcid.org/0000-0001-8738-0162)"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","email","address"]}},"value":[{"type":"character","attributes":{},"value":["Ernst C. Wit"]},{"type":"character","attributes":{},"value":["Faculty of Science & Informatics"]},{"type":"character","attributes":{},"value":["wite@usi.ch\n"]},{"type":"character","attributes":{},"value":["Universita della Svizzera Italiana (USI)","Via Buffi 13","6900 Lugano","Switzerland","(https://orcid.org/0000-0002-3671-9610)"]}]}]},{"type":"character","attributes":{},"value":["2023-02-10"]},{"type":"character","attributes":{},"value":["2021-01-15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"integer","attributes":{},"value":[18]},{"type":"integer","attributes":{},"value":[37]}]},{"type":"integer","attributes":{},"value":[14]},{"type":"integer","attributes":{},"value":[4]},{"type":"character","attributes":{},"value":["RJ-2023-011"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"NULL"},{"type":"NULL"}]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["preview.png"]},{"type":"character","attributes":{},"value":["netgwasRef.bib"]},{"type":"list","attributes":{},"value":[]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc","legacy_pdf"]}},"value":[{"type":"logical","attributes":{},"value":[false]},{"type":"logical","attributes":{},"value":[false]},{"type":"logical","attributes":{},"value":[true]}]}]},{"type":"character","attributes":{},"value":["RJ-2023-011.pdf"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2023-011"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2023-011"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"character","attributes":{},"value":["/Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library/rjtools/rjournal.csl"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["Behrouzi_etal_netgwas.tex","Cover letter-netgwas-Behrouzi-etal.pdf","Fig1_workflow.png","Fig2.pdf","Fig3a.pdf","Fig3b.pdf","Fig4.pdf","Fig5a.pdf","Fig5b.pdf","Fig6.pdf","Fig7.pdf","Fig9.pdf","netgwasRef.bib","preamble.tex","response2AEcomments_PB.pdf","RJ-2023-011_files/anchor-4.2.2/anchor.min.js","RJ-2023-011_files/bowser-1.9.3/bowser.min.js","RJ-2023-011_files/distill-2.2.21/template.v2.js","RJ-2023-011_files/header-attrs-2.20/header-attrs.js","RJ-2023-011_files/jquery-3.6.0/jquery-3.6.0.js","RJ-2023-011_files/jquery-3.6.0/jquery-3.6.0.min.js","RJ-2023-011_files/jquery-3.6.0/jquery-3.6.0.min.map","RJ-2023-011_files/popper-2.6.0/popper.min.js","RJ-2023-011_files/tippy-6.2.7/tippy-bundle.umd.min.js","RJ-2023-011_files/tippy-6.2.7/tippy-light-border.css","RJ-2023-011_files/tippy-6.2.7/tippy.css","RJ-2023-011_files/tippy-6.2.7/tippy.umd.min.js","RJ-2023-011_files/webcomponents-2.0.0/webcomponents.js","RJ-2023-011.pdf","RJ-2023-011.zip","RJournal.sty","RJwrapper.log","RJwrapper.tex"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
    font-size: 100%;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  /* tweak for Pandoc numbered line within distill */
  d-article pre.numberSource code > span {
      left: -2em;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme, except when numbering line
    // in code chunk
    $('pre:not(.numberLines) code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        const citeChild = $(this).children()[0]
        // Do not process if @xyz has been used without escaping and without bibliography activated
        // https://github.com/rstudio/distill/issues/466
        if (citeChild === undefined) return true

        if (citeChild.nodeName == "D-FOOTNOTE") {
          var fn = citeChild
          $(this).html(fn.shadowRoot.querySelector("sup"))
          $(this).id = fn.id
          fn.remove()
        }
        var refs = $(this).attr('data-cites').split(" ");
        var refHtml = refs.map(function(ref) {
          // Could use CSS.escape too here, we insure backward compatibility in navigator
          return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
        }).join("\n");
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // fix footnotes in tables (#411)
      // replacing broken distill.pub feature
      $('table d-footnote').each(function() {
        // we replace internal showAtNode methode which is triggered when hovering a footnote
        this.hoverBox.showAtNode = function(node) {
          // ported from https://github.com/distillpub/template/pull/105/files
          calcOffset = function(elem) {
              let x = elem.offsetLeft;
              let y = elem.offsetTop;
              // Traverse upwards until an `absolute` element is found or `elem`
              // becomes null.
              while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                  x += elem.offsetLeft;
                  y += elem.offsetTop;
              }

              return { left: x, top: y };
          }
          // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
          const bbox = node.getBoundingClientRect();
          const offset = calcOffset(node);
          this.show([offset.left + bbox.width, offset.top + bbox.height]);
        }
      })

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      // ignore leaflet img layers (#106)
      figures = figures.filter(':not(img[class*="leaflet"])')
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="RJ-2023-011_files/header-attrs-2.20/header-attrs.js"></script>
  <script src="RJ-2023-011_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="RJ-2023-011_files/popper-2.6.0/popper.min.js"></script>
  <link href="RJ-2023-011_files/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="RJ-2023-011_files/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="RJ-2023-011_files/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="RJ-2023-011_files/anchor-4.2.2/anchor.min.js"></script>
  <script src="RJ-2023-011_files/bowser-1.9.3/bowser.min.js"></script>
  <script src="RJ-2023-011_files/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="RJ-2023-011_files/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      buttons.append('<a href="https://twitter.com/intent/tweet?text='+title+'&url=' + cite_url + '" class="dt-tag"><i class="fab fa-twitter"></i> Tweet</a>')

      const abstract = $('<d-abstract>')
      abstract.append('<b>Abstract:</b><br>')
      abstract.append($("d-title p:not(:empty)").first()) // Move description to d-abstract
      $("d-title p:empty").remove() // Remove empty paragraphs after title
      abstract.append(buttons)
      abstract.insertAfter($('d-title')) // Add abstract section after title

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"netgwas: An R Package for Network-Based Genome Wide Association Studies","description":"Graphical models are a powerful tool in modelling and analysing complex biological associations in high-dimensional data. The R-package netgwas implements the recent methodological development on copula graphical models to (i) construct linkage maps, (ii) infer linkage disequilibrium networks from genotype data, and (iii) detect high-dimensional genotype-phenotype networks. The netgwas learns the structure of networks from ordinal data and mixed ordinal-and-continuous data. Here, we apply the functionality in netgwas to various multivariate example datasets taken from the literature to demonstrate the kind of insight that can be obtained from the package. We show that our package offers a more realistic association analysis than the classical approaches, as it discriminates between direct and induced correlations by adjusting for the effect of all other variables while performing pairwise associations. This feature controls for spurious interactions between variables that can arise from conventional approaches in a biological sense. The netgwas package uses a parallelization strategy on multi-core processors to speed-up computations.","doi":"10.32614/RJ-2023-011","authors":[{"author":"Pariya Behrouzi","authorURL":"#","affiliation":"Applied Mathematics and Statistics - Biometris","affiliationURL":"#","orcidID":""},{"author":"Danny Arends","authorURL":"#","affiliation":"Albrecht Daniel Thaer-Institut für Agrar und Gartenbauwissenschaften","affiliationURL":"#","orcidID":""},{"author":"Ernst C. Wit","authorURL":"#","affiliation":"Faculty of Science & Informatics","affiliationURL":"#","orcidID":""}],"publishedDate":"2023-02-10T00:00:00.000+00:00","citationText":"Behrouzi, et al., 2023"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>netgwas: An R Package for Network-Based Genome Wide Association Studies</h1>
<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p>Graphical models are a powerful tool in modelling and analysing complex biological associations in high-dimensional data. The R-package netgwas implements the recent methodological development on copula graphical models to (i) construct linkage maps, (ii) infer linkage disequilibrium networks from genotype data, and (iii) detect high-dimensional genotype-phenotype networks. The netgwas learns the structure of networks from ordinal data and mixed ordinal-and-continuous data. Here, we apply the functionality in netgwas to various multivariate example datasets taken from the literature to demonstrate the kind of insight that can be obtained from the package. We show that our package offers a more realistic association analysis than the classical approaches, as it discriminates between direct and induced correlations by adjusting for the effect of all other variables while performing pairwise associations. This feature controls for spurious interactions between variables that can arise from conventional approaches in a biological sense. The netgwas package uses a parallelization strategy on multi-core processors to speed-up computations.</p>
</div>

<div class="d-byline">
  Pariya Behrouzi  (Applied Mathematics and Statistics - Biometris)
  
,   Danny Arends  (Albrecht Daniel Thaer-Institut für Agrar und Gartenbauwissenschaften)
  
,   Ernst C. Wit  (Faculty of Science &amp; Informatics)
  
<br/>2023-02-10
</div>

<div class="d-article">
<div class="l-page">
<p><embed src="RJ-2023-011.pdf" type="application/pdf" height="955px" width="100%"></p>
</div>
<h2 class="appendix unnumbered" id="supplementary-materials">Supplementary materials</h2>
<p>Supplementary materials are available in addition to this article. It can be downloaded at
<a href="RJ-2023-011.zip">RJ-2023-011.zip</a></p>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>


<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="references">References</h3>
<div id="references-listing"></div>
<h3 id="reuse">Reuse</h3>
<p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
<h3 id="citation">Citation</h3>
<p>For attribution, please cite this work as</p>
<pre class="citation-appendix short">Behrouzi, et al., "netgwas: An R Package for Network-Based Genome Wide Association Studies", The R Journal, 2023</pre>
<p>BibTeX citation</p>
<pre class="citation-appendix long">@article{RJ-2023-011,
  author = {Behrouzi, Pariya and Arends, Danny and Wit, Ernst C.},
  title = {netgwas: An R Package for Network-Based Genome Wide Association Studies},
  journal = {The R Journal},
  year = {2023},
  note = {https://doi.org/10.32614/RJ-2023-011},
  doi = {10.32614/RJ-2023-011},
  volume = {14},
  issue = {4},
  issn = {2073-4859},
  pages = {18-37}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
