<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
  <title>vivid: An R package for Variable Importance and Variable Interactions Displays for Machine Learning Models</title>

  <meta property="description" itemprop="description" content="We present vivid, an R package for visualizing variable importance and variable interactions in machine learning models. The package provides  heatmap and graph-based displays for viewing variable importance and interaction jointly, and partial dependence plots in both a matrix layout and an alternative layout emphasizing important variable subsets. With the intention of increasing machine learning models&#39; interpretability and making the work applicable to a wider readership, we discuss the design choices behind our implementation by focusing on the package structure and providing an in-depth look at the package functions and key features. We also provide a practical illustration of the software in use on a data set."/>

  <link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>

  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2023-11-01"/>
  <meta property="article:created" itemprop="dateCreated" content="2023-11-01"/>
  <meta name="article:author" content="Alan Inglis"/>
  <meta name="article:author" content="Andrew Parnell"/>
  <meta name="article:author" content="Catherine Hurley"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="vivid: An R package for Variable Importance and Variable Interactions Displays for Machine Learning Models"/>
  <meta property="og:type" content="article"/>
  <meta property="og:description" content="We present vivid, an R package for visualizing variable importance and variable interactions in machine learning models. The package provides  heatmap and graph-based displays for viewing variable importance and interaction jointly, and partial dependence plots in both a matrix layout and an alternative layout emphasizing important variable subsets. With the intention of increasing machine learning models&#39; interpretability and making the work applicable to a wider readership, we discuss the design choices behind our implementation by focusing on the package structure and providing an in-depth look at the package functions and key features. We also provide a practical illustration of the software in use on a data set."/>
  <meta property="og:locale" content="en_US"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="vivid: An R package for Variable Importance and Variable Interactions Displays for Machine Learning Models"/>
  <meta property="twitter:description" content="We present vivid, an R package for visualizing variable importance and variable interactions in machine learning models. The package provides  heatmap and graph-based displays for viewing variable importance and interaction jointly, and partial dependence plots in both a matrix layout and an alternative layout emphasizing important variable subsets. With the intention of increasing machine learning models&#39; interpretability and making the work applicable to a wider readership, we discuss the design choices behind our implementation by focusing on the package structure and providing an in-depth look at the package functions and key features. We also provide a practical illustration of the software in use on a data set."/>

  <!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
  <meta name="citation_title" content="vivid: An R package for Variable Importance and Variable Interactions Displays for Machine Learning Models"/>
  <meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2023-054"/>
  <meta name="citation_pdf_url" content="RJ-2023-054.pdf"/>
  <meta name="citation_volume" content="15"/>
  <meta name="citation_issue" content="2"/>
  <meta name="citation_doi" content="10.32614/RJ-2023-054"/>
  <meta name="citation_journal_title" content="The R Journal"/>
  <meta name="citation_issn" content="2073-4859"/>
  <meta name="citation_firstpage" content="344"/>
  <meta name="citation_lastpage" content="361"/>
  <meta name="citation_fulltext_world_readable" content=""/>
  <meta name="citation_online_date" content="2023/11/01"/>
  <meta name="citation_publication_date" content="2023/11/01"/>
  <meta name="citation_author" content="Alan Inglis"/>
  <meta name="citation_author_institution" content="Maynooth University"/>
  <meta name="citation_author" content="Andrew Parnell"/>
  <meta name="citation_author_institution" content="Maynooth University"/>
  <meta name="citation_author" content="Catherine Hurley"/>
  <meta name="citation_author_institution" content="Maynooth University"/>
  <!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=A simple and effective model-based variable importance measure;citation_author=Brandon M. Greenwell;citation_author=Bradley C. Boehmke;citation_author=Andrew J. McCarthy"/>
  <meta name="citation_reference" content="citation_title=Eulerian tour algorithms for data visualization and the PairViz package;citation_volume=26;citation_doi=10.1007/s00180-011-0229-5;citation_author=Catherine B. Hurley;citation_author=R. W. Oldford"/>
  <meta name="citation_reference" content="citation_title=PairViz: Visualization using graph traversal;citation_author=Catherine B. Hurley;citation_author=R. W. Oldford"/>
  <meta name="citation_reference" content="citation_title=Explanatory model analysis: Explore, explain and examine predictive models;citation_publisher=Chapman; Hall/CRC;citation_author=Przemyslaw Biecek;citation_author=Tomasz Burzykowski"/>
  <meta name="citation_reference" content="citation_title=e1071: Misc functions of the department of statistics, probability theory group (formerly: E1071), TU wien;citation_author=David Meyer;citation_author=Evgenia Dimitriadou;citation_author=Kurt Hornik;citation_author=Andreas Weingessel;citation_author=Friedrich Leisch"/>
  <meta name="citation_reference" content="citation_title=High-dimensional variable selection for survival data;citation_publisher=Taylor &amp; Francis;citation_volume=105;citation_author=Hemant Ishwaran;citation_author=Udaya B Kogalur;citation_author=Eiran Z Gorodeski;citation_author=Andy J Minn;citation_author=Michael S Lauer"/>
  <meta name="citation_reference" content="citation_title=Predictive learning via rule ensembles;citation_author=J. H. Friedman;citation_author=B. E. Popescu"/>
  <meta name="citation_reference" content="citation_title=randomForestExplainer: Explaining and visualizing random forests in terms of variable importance;citation_author=Aleksandra Paluszynska;citation_author=Przemyslaw Biecek;citation_author=Yue Jiang"/>
  <meta name="citation_reference" content="citation_title=EIX: Explain interactions in ’XGBoost’;citation_author=Szymon Maksymiuk;citation_author=Ewelina Karbowiak;citation_author=Przemyslaw Biecek"/>
  <meta name="citation_reference" content="citation_title=Peeking inside the black box: Visualizing statistical learning with plots of individual conditional expectation;citation_volume=24;citation_doi=10.1080/10618600.2014.907095;citation_author=Alex Goldstein;citation_author=Adam Kapelner;citation_author=Justin Bleich;citation_author=Emil Pitkin"/>
  <meta name="citation_reference" content="citation_title=iml: An R package for interpretable machine learning;citation_publisher=Journal of Open Source Software;citation_volume=3;citation_doi=10.21105/joss.00786;citation_author=Christoph Molnar;citation_author=Bernd Bischl;citation_author=Giuseppe Casalicchio"/>
  <meta name="citation_reference" content="citation_title=Greedy function approximation: A gradient boosting machine;citation_volume=29;citation_doi=10.1214/aos/1013203451;citation_author=J. H. Friedman"/>
  <meta name="citation_reference" content="citation_title=Random forests;citation_volume=45;citation_doi=10.1023/A:1010933404324;citation_author=Leo Breiman"/>
  <meta name="citation_reference" content="citation_title=ranger: A fast implementation of random forests for high dimensional data in C++ and R;citation_volume=77;citation_doi=10.18637/jss.v077.i01;citation_author=Marvin N. Wright;citation_author=Andreas Ziegler"/>
  <meta name="citation_reference" content="citation_title=All models are wrong, but many are useful: Learning a variable’s importance by studying an entire class of prediction models simultaneously.;citation_volume=20;citation_author=Aaron Fisher;citation_author=Cynthia Rudin;citation_author=Francesca Dominici"/>
  <meta name="citation_reference" content="citation_title=mlr: Machine learning in R;citation_volume=17;citation_author=Bernd Bischl;citation_author=Michel Lang;citation_author=Lars Kotthoff;citation_author=Julia Schiffner;citation_author=Jakob Richter;citation_author=Erich Studerus;citation_author=Giuseppe Casalicchio;citation_author=Zachary M. Jones"/>
  <meta name="citation_reference" content="citation_title=mlr3: A modern object-oriented machine learning framework in R;citation_doi=10.21105/joss.01903;citation_author=Michel Lang;citation_author=Martin Binder;citation_author=Jakob Richter;citation_author=Patrick Schratz;citation_author=Florian Pfisterer;citation_author=Stefan Coors;citation_author=Quay Au;citation_author=Giuseppe Casalicchio;citation_author=Lars Kotthoff;citation_author=Bernd Bischl"/>
  <meta name="citation_reference" content="citation_title=condvis2: Interactive conditional visualization for supervised and unsupervised models in shiny;citation_author=Catherine Hurley;citation_author=Mark OConnell;citation_author=Katarina Domijan"/>
  <meta name="citation_reference" content="citation_title=Visualizing variable importance and variable interaction effects in machine learning models;citation_publisher=Taylor &amp; Francis;citation_volume=31;citation_doi=10.1080/10618600.2021.2007935;citation_author=Alan Inglis;citation_author=Andrew Parnell;citation_author=Catherine B Hurley"/>
  <meta name="citation_reference" content="citation_title=Fairness and transparency of machine learning for trustworthy cloud services;citation_publisher=IEEE;citation_author=Nuno Antunes;citation_author=Leandro Balby;citation_author=Flavio Figueiredo;citation_author=Nuno Lourenco;citation_author=Wagner Meira;citation_author=Walter Santos"/>
  <meta name="citation_reference" content="citation_title=Transparency you can trust: Transparency requirements for artificial intelligence between legal norms and contextual concerns;citation_publisher=SAGE Publications Sage UK: London, England;citation_volume=6;citation_author=Heike Felzmann;citation_author=Eduard Fosch Villaronga;citation_author=Christoph Lutz;citation_author=Aurelia Tamò-Larrieux"/>
  <meta name="citation_reference" content="citation_title=DALEX: Explainers for complex predictive models in R;citation_volume=19;citation_author=Przemyslaw Biecek"/>
  <meta name="citation_reference" content="citation_title=Lime: Local interpretable model-agnostic explanations;citation_author=Emil Hvitfeldt;citation_author=Thomas Lin Pedersen;citation_author=Michaël Benesty"/>
  <meta name="citation_reference" content="citation_title=Interpretable machine learning: A guide for making black box models explainable;citation_author=Christoph Molnar"/>
  <meta name="citation_reference" content="citation_title=Methods to quantify variable importance: Implications for the analysis of noisy ecological data;citation_publisher=Wiley Online Library;citation_volume=90;citation_author=Kim Murray;citation_author=Mary M Conner"/>
  <meta name="citation_reference" content="citation_title=Random forest meteorological normalisation models for swiss PM 10 trend analysis;citation_publisher=Copernicus GmbH;citation_volume=18;citation_author=Stuart K Grange;citation_author=David C Carslaw;citation_author=Alastair C Lewis;citation_author=Eirini Boleti;citation_author=Christoph Hueglin"/>
  <meta name="citation_reference" content="citation_title=Random forests for genomic data analysis;citation_publisher=Elsevier;citation_volume=99;citation_author=Xi Chen;citation_author=Hemant Ishwaran"/>
  <meta name="citation_reference" content="citation_title=BART: Bayesian additive regression trees;citation_publisher=Institute of Mathematical Statistics;citation_volume=4;citation_author=Hugh A Chipman;citation_author=Edward I George;citation_author=Robert E McCulloch"/>
  <meta name="citation_reference" content="citation_title=The elements of statistical learning: Data mining, inference, and prediction.;citation_publisher=Springer Series in Statistics. Springer-Verlag.;citation_author=T. Hastie;citation_author=R. Tibshirani;citation_author=J. Friedman"/>
  <meta name="citation_reference" content="citation_title=Modern applied statistics with s;citation_publisher=Springer;citation_author=W. N. Venables;citation_author=B. D. Ripley"/>
  <meta name="citation_reference" content="citation_title=Hedonic housing prices and the demand for clean air;citation_publisher=Elsevier;citation_volume=5;citation_author=David Harrison Jr;citation_author=Daniel L Rubinfeld"/>
  <meta name="citation_reference" content="citation_title=Visualizing the effects of predictor variables in black box supervised learning models;citation_publisher=Wiley Online Library;citation_volume=82;citation_author=Daniel W Apley;citation_author=Jingyu Zhu"/>
  <meta name="citation_reference" content="citation_title=varImp: RF Variable Importance for Arbitrary Measures;citation_author=Philipp Probst"/>
  <meta name="citation_reference" content="citation_title=Conditional variable importance for random forests;citation_volume=9;citation_doi=10.1186/1471-2105-9-307;citation_author=Carolin Strobl;citation_author=Anne-Laure Boulesteix;citation_author=Thomas Kneib;citation_author=Thomas Augustin;citation_author=Achim Zeileis"/>
  <meta name="citation_reference" content="citation_title=kknn: Weighted k-nearest neighbors;citation_author=Klaus Schliep;citation_author=Klaus Hechenbichler"/>
  <meta name="citation_reference" content="citation_title=&quot; why should i trust you?&quot; explaining the predictions of any classifier;citation_author=Marco Tulio Ribeiro;citation_author=Sameer Singh;citation_author=Carlos Guestrin"/>
  <meta name="citation_reference" content="citation_title=R package ’corrplot’: Visualization of a correlation matrix;citation_author=Taiyun Wei;citation_author=Viliam Simko"/>
  <meta name="citation_reference" content="citation_title=Corrgrapher: Explore correlations between variables in a machine learning model;citation_author=Pawel Morgen;citation_author=Przemyslaw Biecek"/>
  <meta name="citation_reference" content="citation_title=Parsnip: A common API to modeling and analysis functions;citation_author=Max Kuhn;citation_author=Davis Vaughan"/>
  <meta name="citation_reference" content="citation_title=Visualizations for bayesian additive regression trees;citation_author=Alan Inglis;citation_author=Andrew Parnell;citation_author=Cathrine Hurley"/>
  <meta name="citation_reference" content="citation_title=Value-Suppressing Uncertainty Palettes;citation_author=Michael Correll;citation_author=Dominik Moritz;citation_author=Jeffrey Heer"/>
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","preamble","draft","author","type","output","bibliography","date","date_received","volume","issue","slug","journal","pdf_url","citation_url","doi","creative_commons","packages","CTV","csl"]}},"value":[{"type":"character","attributes":{},"value":["vivid: An R package for Variable Importance and Variable Interactions Displays for Machine Learning Models"]},{"type":"character","attributes":{},"value":["We present vivid, an R package for visualizing variable importance and variable interactions in machine learning models. The package provides  heatmap and graph-based displays for viewing variable importance and interaction jointly, and partial dependence plots in both a matrix layout and an alternative layout emphasizing important variable subsets. With the intention of increasing machine learning models' interpretability and making the work applicable to a wider readership, we discuss the design choices behind our implementation by focusing on the package structure and providing an in-depth look at the package functions and key features. We also provide a practical illustration of the software in use on a data set.\n"]},{"type":"character","attributes":{},"value":["\\usepackage{subfig} \\usepackage{tabu} \\usepackage{longtable} \\usepackage{colortbl} \\usepackage{xcolor}"]},{"type":"logical","attributes":{},"value":[false]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","orcid_id","email"]}},"value":[{"type":"character","attributes":{},"value":["Alan Inglis"]},{"type":"character","attributes":{},"value":["Maynooth University"]},{"type":"character","attributes":{},"value":["Department of Mathematics and Statistics","Maynooth, Ireland"]},{"type":"character","attributes":{},"value":["0000-0002-1151-6657"]},{"type":"character","attributes":{},"value":["alan.n.inglis@gmail.com"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","email","orcid_id","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Andrew Parnell"]},{"type":"character","attributes":{},"value":["andrew.parnell@mu.ie"]},{"type":"character","attributes":{},"value":["0000-0001-7956-7939"]},{"type":"character","attributes":{},"value":["Maynooth University"]},{"type":"character","attributes":{},"value":["Hamilton Institute","Maynooth, Ireland"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","email","orcid_id","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Catherine Hurley"]},{"type":"character","attributes":{},"value":["catherine.hurley@mu.ie"]},{"type":"character","attributes":{},"value":["0000-0003-2758-5531"]},{"type":"character","attributes":{},"value":["Maynooth University"]},{"type":"character","attributes":{},"value":["Department of Mathematics and Statistics","Maynooth, Ireland"]}]}]},{"type":"character","attributes":{},"value":["package"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc","fig_caption"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[false]},{"type":"logical","attributes":{},"value":[true]}]}]},{"type":"character","attributes":{},"value":["vivid.bib"]},{"type":"character","attributes":{},"value":["2023-11-01"]},{"type":"character","attributes":{},"value":["2022-12-09"]},{"type":"integer","attributes":{},"value":[15]},{"type":"integer","attributes":{},"value":[2]},{"type":"character","attributes":{},"value":["RJ-2023-054"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"integer","attributes":{},"value":[344]},{"type":"integer","attributes":{},"value":[361]}]},{"type":"character","attributes":{},"value":["RJ-2023-054.pdf"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2023-054"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2023-054"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"character","attributes":{},"value":["vivid","vip","ggplot2","iml","flashlight","DALEX","lime","pdp","ICEbox","randomForestExplainer","randomForest","EIX","varImp","party","bartMachine","ingredients","MASS","xgboost","condvis2","ranger","mlr","mlr3","parsnip","e1071","nnet","kknn","igraph","zenplots","zenplot","corrplot","corrgrapher"]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["Bayesian","Cluster","Distributions","Econometrics","Environmetrics","GraphicalModels","HighPerformanceComputing","MachineLearning","MissingData","MixedModels","ModelDeployment","NumericalMathematics","Optimization","Phylogenetics","Psychometrics","Robust","Spatial","Survival","TeachingStatistics"]},{"type":"character","attributes":{},"value":["/Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library/rjtools/rjournal.csl"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["00render-pdf.log","00render-web.log","DESCRIPTION","distill-preview.png","initial_checks.log","pdf-OK","RJ-2023-054_files/anchor-4.2.2/anchor.min.js","RJ-2023-054_files/bowser-1.9.3/bowser.min.js","RJ-2023-054_files/crosstalk-1.2.0/css/crosstalk.min.css","RJ-2023-054_files/crosstalk-1.2.0/js/crosstalk.js","RJ-2023-054_files/crosstalk-1.2.0/js/crosstalk.js.map","RJ-2023-054_files/crosstalk-1.2.0/js/crosstalk.min.js","RJ-2023-054_files/crosstalk-1.2.0/js/crosstalk.min.js.map","RJ-2023-054_files/crosstalk-1.2.0/scss/crosstalk.scss","RJ-2023-054_files/datatables-binding-0.30/datatables.js","RJ-2023-054_files/datatables-css-0.0.0/datatables-crosstalk.css","RJ-2023-054_files/distill-2.2.21/template.v2.js","RJ-2023-054_files/dt-core-1.13.4/css/jquery.dataTables.extra.css","RJ-2023-054_files/dt-core-1.13.4/css/jquery.dataTables.min.css","RJ-2023-054_files/dt-core-1.13.4/js/jquery.dataTables.min.js","RJ-2023-054_files/figure-html5/gpdp-1.png","RJ-2023-054_files/figure-html5/heatmaps-1.png","RJ-2023-054_files/figure-html5/heatmaps-2.png","RJ-2023-054_files/figure-html5/networks1-1.png","RJ-2023-054_files/figure-html5/networks1-2.png","RJ-2023-054_files/figure-html5/pdpRf-1.png","RJ-2023-054_files/figure-html5/zpdpGBM-1.png","RJ-2023-054_files/figure-html5/zpdpGBM-2.png","RJ-2023-054_files/figure-html5/zpdpGbm1-1.png","RJ-2023-054_files/header-attrs-2.25/header-attrs.js","RJ-2023-054_files/htmlwidgets-1.6.2/htmlwidgets.js","RJ-2023-054_files/jquery-3.6.0/jquery-3.6.0.js","RJ-2023-054_files/jquery-3.6.0/jquery-3.6.0.min.js","RJ-2023-054_files/jquery-3.6.0/jquery-3.6.0.min.map","RJ-2023-054_files/popper-2.6.0/popper.min.js","RJ-2023-054_files/tippy-6.2.7/tippy-bundle.umd.min.js","RJ-2023-054_files/tippy-6.2.7/tippy-light-border.css","RJ-2023-054_files/tippy-6.2.7/tippy.css","RJ-2023-054_files/tippy-6.2.7/tippy.umd.min.js","RJ-2023-054_files/webcomponents-2.0.0/webcomponents.js","RJ-2023-054.log","RJ-2023-054.pdf","RJ-2023-054.tex","RJ-2023-054.zip","RJournal.sty","RJwrapper.aux","RJwrapper.log","RJwrapper.out","RJwrapper.tex","Rlogo-5.png","speedTestsPlot.png","vivid_files/figure-html5/gpdp-1.png","vivid_files/figure-html5/heatmaps-1.png","vivid_files/figure-html5/heatmaps-2.png","vivid_files/figure-html5/networks1-1.png","vivid_files/figure-html5/networks1-2.png","vivid_files/figure-html5/pdpRf-1.png","vivid_files/figure-html5/zpdpGBM-1.png","vivid_files/figure-html5/zpdpGBM-2.png","vivid_files/figure-html5/zpdpGbm1-1.png","vivid_files/figure-latex/gpdp-1.pdf","vivid_files/figure-latex/heatmaps-1.pdf","vivid_files/figure-latex/heatmaps-2.pdf","vivid_files/figure-latex/networks1-1.pdf","vivid_files/figure-latex/networks1-2.pdf","vivid_files/figure-latex/pdpRf-1.pdf","vivid_files/figure-latex/zpdpGBM-1.pdf","vivid_files/figure-latex/zpdpGBM-2.pdf","vivid_files/figure-latex/zpdpGbm1-1.pdf","vivid.bib","vivid.log","vivid.tex","web-OK"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
    font-size: 100%;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  hr.section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    margin: 0px;
  }


  d-byline {
    border-top: none;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
    border-top: none;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  /* tweak for Pandoc numbered line within distill */
  d-article pre.numberSource code > span {
      left: -2em;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // separator
    var separator = '<hr class="section-separator" style="clear: both"/>';
    // prepend separator above appendix
    $('.d-byline').before(separator);
    $('.d-article').before(separator);

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme, except when numbering line
    // in code chunk
    $('pre:not(.numberLines) code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        var author_name = front_matter.authors[i].author
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        const citeChild = $(this).children()[0]
        // Do not process if @xyz has been used without escaping and without bibliography activated
        // https://github.com/rstudio/distill/issues/466
        if (citeChild === undefined) return true

        if (citeChild.nodeName == "D-FOOTNOTE") {
          var fn = citeChild
          $(this).html(fn.shadowRoot.querySelector("sup"))
          $(this).id = fn.id
          fn.remove()
        }
        var refs = $(this).attr('data-cites').split(" ");
        var refHtml = refs.map(function(ref) {
          // Could use CSS.escape too here, we insure backward compatibility in navigator
          return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
        }).join("\n");
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // fix footnotes in tables (#411)
      // replacing broken distill.pub feature
      $('table d-footnote').each(function() {
        // we replace internal showAtNode methode which is triggered when hovering a footnote
        this.hoverBox.showAtNode = function(node) {
          // ported from https://github.com/distillpub/template/pull/105/files
          calcOffset = function(elem) {
              let x = elem.offsetLeft;
              let y = elem.offsetTop;
              // Traverse upwards until an `absolute` element is found or `elem`
              // becomes null.
              while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                  x += elem.offsetLeft;
                  y += elem.offsetTop;
              }

              return { left: x, top: y };
          }
          // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
          const bbox = node.getBoundingClientRect();
          const offset = calcOffset(node);
          this.show([offset.left + bbox.width, offset.top + bbox.height]);
        }
      })

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      // ignore leaflet img layers (#106)
      figures = figures.filter(':not(img[class*="leaflet"])')
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="RJ-2023-054_files/header-attrs-2.25/header-attrs.js"></script>
  <script src="RJ-2023-054_files/htmlwidgets-1.6.2/htmlwidgets.js"></script>
  <link href="RJ-2023-054_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
  <script src="RJ-2023-054_files/datatables-binding-0.30/datatables.js"></script>
  <script src="RJ-2023-054_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <link href="RJ-2023-054_files/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet" />
  <link href="RJ-2023-054_files/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet" />
  <script src="RJ-2023-054_files/dt-core-1.13.4/js/jquery.dataTables.min.js"></script>
  <link href="RJ-2023-054_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
  <script src="RJ-2023-054_files/crosstalk-1.2.0/js/crosstalk.min.js"></script>
  <script src="RJ-2023-054_files/popper-2.6.0/popper.min.js"></script>
  <link href="RJ-2023-054_files/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="RJ-2023-054_files/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="RJ-2023-054_files/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="RJ-2023-054_files/anchor-4.2.2/anchor.min.js"></script>
  <script src="RJ-2023-054_files/bowser-1.9.3/bowser.min.js"></script>
  <script src="RJ-2023-054_files/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="RJ-2023-054_files/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      buttons.append('<a href="' + slug + '.zip" class="dt-tag"><i class="fas fa-file-zipper"></i> Supplement</a>')


      const abstract = $('<d-abstract>')
      abstract.append('<b>Abstract:</b><br>')
      abstract.append($("d-title p:not(:empty)").first()) // Move description to d-abstract
      $("d-title p:empty").remove() // Remove empty paragraphs after title
      abstract.append(buttons)
      abstract.insertAfter($('d-title')) // Add abstract section after title

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
    
    d-appendix .citation-appendix,
    .d-appendix .citation-appendix {
      color: rgb(60, 60, 60);
    }

    d-article h2 {
      border-bottom: 0px solid rgba(0, 0, 0, 0.1);
      padding-bottom: 0rem;
    }
    d-article h3 {
      font-size: 20px;
    }
    d-article h4 {
      font-size: 18px;
      text-transform: none;
    }

    @media (min-width: 1024px) {
      d-article h2 {
        font-size: 32px;
      }
      d-article h3 {
        font-size: 24px;
      }
      d-article h4 {
        font-size: 20px;
      }
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"vivid: An R package for Variable Importance and Variable Interactions Displays for Machine Learning Models","description":"We present vivid, an R package for visualizing variable importance and variable interactions in machine learning models. The package provides  heatmap and graph-based displays for viewing variable importance and interaction jointly, and partial dependence plots in both a matrix layout and an alternative layout emphasizing important variable subsets. With the intention of increasing machine learning models' interpretability and making the work applicable to a wider readership, we discuss the design choices behind our implementation by focusing on the package structure and providing an in-depth look at the package functions and key features. We also provide a practical illustration of the software in use on a data set.","doi":"10.32614/RJ-2023-054","authors":[{"author":"Alan Inglis","authorURL":"#","affiliation":"Maynooth University","affiliationURL":"#","orcidID":"0000-0002-1151-6657"},{"author":"Andrew Parnell","authorURL":"#","affiliation":"Maynooth University","affiliationURL":"#","orcidID":"0000-0001-7956-7939"},{"author":"Catherine Hurley","authorURL":"#","affiliation":"Maynooth University","affiliationURL":"#","orcidID":"0000-0003-2758-5531"}],"publishedDate":"2023-11-01T00:00:00.000+13:00","citationText":"Inglis, et al., 2023"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>vivid: An R package for Variable Importance and Variable Interactions Displays for Machine Learning Models</h1>

<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>We present vivid, an R package for visualizing variable importance and variable interactions in machine learning models. The package provides heatmap and graph-based displays for viewing variable importance and interaction jointly, and partial dependence plots in both a matrix layout and an alternative layout emphasizing important variable subsets. With the intention of increasing machine learning models’ interpretability and making the work applicable to a wider readership, we discuss the design choices behind our implementation by focusing on the package structure and providing an in-depth look at the package functions and key features. We also provide a practical illustration of the software in use on a data set.</p></p>
</div>

<div class="d-byline">
  Alan Inglis  (Maynooth University)
  
,   Andrew Parnell  (Maynooth University)
  
,   Catherine Hurley  (Maynooth University)
  
<br/>2023-11-01
</div>

<div class="d-article">
<h2 data-number="1" id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<p>Our motivation behind the creation of the <a href="https://cran.r-project.org/package=vivid">vivid</a> package is to investigate machine learning
models in a way that is simple to understand while also offering helpful
insights into how variables affect the fit. We do this through the use
of heatmaps, network graphs, and both a generalized pairs plot style
partial dependence plot (PDP) <span class="citation" data-cites="Friedpdp">(<a href="#ref-Friedpdp" role="doc-biblioref">Friedman 2000</a>)</span> and a space saving PDP based
on key variable subsets. While the techniques and fundamental goals of
these visualizations have been discussed in <span class="citation" data-cites="inglis2022visualizing">Inglis et al. (<a href="#ref-inglis2022visualizing" role="doc-biblioref">2022b</a>)</span>, we
focus here on the implementation details of the package by providing a
complete listing of the functions and arguments included in the <a href="https://cran.r-project.org/package=vivid">vivid</a>
package with further examples indicating advanced usage beyond that
previously shown. In this work we examine the decisions made when
designing the package and provide an in-depth look at the package
functions and features with the intention of making the work applicable
to a larger readership. This article outlines the general architectural
principles implemented in <a href="https://cran.r-project.org/package=vivid">vivid</a>, such as the data structures we use and data
formatting, function design, filtering techniques, and more. We
illustrate each function by way of a practical example. Our package is
available on the Comprehensive R Archive Network at
<a href="https://cran.r-project.org/web/packages/vivid" class="uri">https://cran.r-project.org/web/packages/vivid</a> or on GitHub at
<a href="https://github.com/AlanInglis/vivid" class="uri">https://github.com/AlanInglis/vivid</a>.</p>
<p>In recent years machine learning (ML) algorithms have emerged as a
valuable tool for both industry and science. However, due to the
black-box nature of many of these algorithms it can be challenging to
communicate the reasoning behind the algorithm’s decision-making
processes. With the need for transparency in ML growing it is important
to gain understanding and clarity about how these algorithms are making
predictions <span class="citation" data-cites="antunes2018fairness felzmann2019transparency">(<a href="#ref-antunes2018fairness" role="doc-biblioref">Antunes et al. 2018</a>; <a href="#ref-felzmann2019transparency" role="doc-biblioref">Felzmann et al. 2019</a>)</span>. Many R
packages are now available that aid in creating interpretable machine
learning (IML) models such as <a href="https://cran.r-project.org/package=iml">iml</a> <span class="citation" data-cites="iml">(<a href="#ref-iml" role="doc-biblioref">Molnar et al. 2018</a>)</span>, <a href="https://cran.r-project.org/package=DALEX">DALEX</a> <span class="citation" data-cites="dalex">(<a href="#ref-dalex" role="doc-biblioref">Biecek 2018</a>)</span>, and
<a href="https://cran.r-project.org/package=lime">lime</a> <span class="citation" data-cites="lime">(<a href="#ref-lime" role="doc-biblioref">Hvitfeldt et al. 2022</a>)</span>. For a
comprehensive review of IML, see <span class="citation" data-cites="molnar2022book">Molnar (<a href="#ref-molnar2022book" role="doc-biblioref">2022</a>)</span> and
<span class="citation" data-cites="biecek2021explanatory">Biecek and Burzykowski (<a href="#ref-biecek2021explanatory" role="doc-biblioref">2021</a>)</span>.</p>
<p>How we choose to visualize aspects of the model output is of vital
importance to how a researcher can interpret and communicate their
findings. Consequently, model summaries such as variable importance and
variable interactions (VImp and VInt; together we term these VIVI) are
frequently used in various fields to comprehend and explain the hidden
structure in an ML fit. In ecology they are employed to determine the
causes of ecological phenomena <span class="citation" data-cites="murray2009methods">(e.g. <a href="#ref-murray2009methods" role="doc-biblioref">Murray and Conner 2009</a>)</span>; in meteorology
VImp measures and partial dependence plots are used to examine air
quality <span class="citation" data-cites="grange2018random">(e.g. <a href="#ref-grange2018random" role="doc-biblioref">Grange et al. 2018</a>)</span>; in bioinformatics, understanding
gene-environment interactions have made these measures an important tool
for genomic analysis <span class="citation" data-cites="chen2012random">(e.g. <a href="#ref-chen2012random" role="doc-biblioref">Chen and Ishwaran 2012</a>)</span>.</p>
<p>In Table 1 we summarize VIVI measures and visualizations
provided by a selection of R packages. VIVI measures are categorized as global or local, depending on whether they refer to the entire predictor space or a specific sub-region. For instance, Individual Conditional Expectation (ICE) curves <span class="citation" data-cites="iceR">(<a href="#ref-iceR" role="doc-biblioref">Goldstein et al. 2015</a>)</span> and Local Interpretable Model-agnostic Explanations (LIME) <span class="citation" data-cites="limeOG">(<a href="#ref-limeOG" role="doc-biblioref">Ribeiro et al. 2016</a>)</span> are considered local methods (implementations of which can be found in the <a href="https://cran.r-project.org/package=ICEbox">ICEbox</a> package <span class="citation" data-cites="iceR">(<a href="#ref-iceR" role="doc-biblioref">Goldstein et al. 2015</a>)</span> and the <a href="https://cran.r-project.org/package=lime">lime</a> package <span class="citation" data-cites="lime">(<a href="#ref-lime" role="doc-biblioref">Hvitfeldt et al. 2022</a>)</span>), whereas partial dependence and permutation importance represent global methods. Packages that incorporate local or global methods can be found under the column ‘Method’ in Table 1. Additionally, VIVI measures can be broken down into model specific (embedded) methods or model agnostic methods. The packages listed in Table 1 are grouped by whether they can compute model specific or agnostic measures and can be found under the column `Measure’. In embedded methods the variable importance is incorporated into the ML algorithm. For example, random forests (RF)
<span class="citation" data-cites="breiman2001randomforest">(<a href="#ref-breiman2001randomforest" role="doc-biblioref">Breiman 2001</a>)</span> and gradient boosting machines (GBM) <span class="citation" data-cites="Friedpdp">(<a href="#ref-Friedpdp" role="doc-biblioref">Friedman 2000</a>)</span> use the tree structure to evaluate the performance of the model. Bayesian additive regression tree models (BART) <span class="citation" data-cites="chipman2010bart">(<a href="#ref-chipman2010bart" role="doc-biblioref">Chipman et al. 2010</a>)</span> also use an embedded method to obtain VIVI measures by looking at the
proportion of splitting rules for variables or variable pairs used in the trees. The package <a href="https://cran.r-project.org/package=randomForestExplainer">randomForestExplainer</a> <span class="citation" data-cites="RFE">(<a href="#ref-RFE" role="doc-biblioref">Paluszynska et al. 2020</a>)</span> provides a set of tools to understand what is happening inside a random forest and uses the concept of minimal depth <span class="citation" data-cites="ishwaran2010high">(<a href="#ref-ishwaran2010high" role="doc-biblioref">Ishwaran et al. 2010</a>)</span> to assess both importance and interaction strength by examining the position of a variable within the trees. Additionally the <a href="https://cran.r-project.org/package=varImp">varImp</a> <span class="citation" data-cites="varImpPack">(<a href="#ref-varImpPack" role="doc-biblioref">Probst 2020</a>)</span> package can be used to compute importance scores for the conditional inference random forest of the <a href="https://cran.r-project.org/package=party">party</a> package <span class="citation" data-cites="party">(<a href="#ref-party" role="doc-biblioref">Strobl et al. 2008</a>)</span>. For gradient boosted machines, the <a href="https://cran.r-project.org/package=EIX">EIX</a> <span class="citation" data-cites="EIX">(<a href="#ref-EIX" role="doc-biblioref">Maksymiuk et al. 2021</a>)</span> package can be used to measure and identify VIVI and visualize the results.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-d8754c6254795643a295" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-d8754c6254795643a295">{"x":{"filter":"none","vertical":false,"caption":"<caption>Summary of a selection of R packages that can be used to assess the variable importance, variable interactions, or partial dependence and if these metrics are global or local and model-specific or model-agnostic. A brief description of available visualizations for evaluating model behavior is also provided.<\/caption>","autoHideNavigation":true,"data":[["<a href = \"https://cran.r-project.org/web/packages/vip/index.html\">vip<\/a>","<a href = \"https://cran.r-project.org/web/packages/iml/index.html\">iml<\/a>","<a href = \"https://cran.r-project.org/web/packages/flashlight/index.html\">flashlight<\/a>","<a href = \"https://cran.r-project.org/web/packages/DALEX/index.html\">DALEX<\/a>","<a href = \"https://cran.r-project.org/web/packages/lime/index.html\">lime<\/a>","<a href = \"https://cran.r-project.org/web/packages/pdp/index.html\">pdp<\/a>","<a href = \"https://cran.r-project.org/web/packages/ICEbox/index.html\">ICEbox<\/a>","<a href = \"https://cran.r-project.org/web/packages/randomForestExplainer/index.html\">randomForestExplainer<\/a>","<a href = \"https://cran.r-project.org/web/packages/randomForest/index.html\">randomForest<\/a>","<a href = \"https://cran.r-project.org/web/packages/EIX/index.html\">EIX<\/a>","<a href = \"https://cran.r-project.org/web/packages/varImp/index.html\">varImp<\/a>","<a href = \"https://cran.r-project.org/web/packages/bartMachine/index.html\">bartMachine<\/a>"],["A general framework for analyzing the behavior of ML models. Also provides PDP based importance and ability to plot Shapley values. Built with ggplot2","A general framework for analyzing the behavior of ML models. Ability to create lollipop, dot, and barplots. Also includes univariate and bivariate PDPs, ICE curves, LIME, and Shapley visualizations. Built with ggplot2.","A general framework for analyzing the behavior of ML models. Ability to plot VIVI measures using barplots. Includes univariate and bivariate PDPs, ICE curves, Global surrogate, and SHAP visualizations. Built with ggplot2.","A general framework for analyzing the behavior of ML models. Contains a suite of visualizations including Ceteris Paribus, Shapley, PDPs, model performance, and diagnostic plots. Built with ggplot2.","A general framework for fitting a local interpretable model. Ability to create VImp and model visualizations using barplots and heatmaps. Can also create interactive plots. Built with ggplot2.","A general framework for constructing PDPs from various types machine  learning models, bivariate, and  trivariate PDPs and ICE curves.  Built with ggplot2.","Used to create Individual Conditional Expectation (ICE) plots.  Provides univariate and bivariate  PDPs and ICE curves.  Built with ggplot2.","Contains a set of model-specific tools to determine which random forests variables are most important. Ability to create VIVI plots displaying the mean minimal depth distribution and conditional minimal depth. Can also display multi-way importance, pairs plots containing different metrics, and Bivariate PDP. Built with ggplot2."," Used to build random forest models. Offers VImp, error rate, and univariate PDPs. Built using base R.","Contains a set of model-specific tools to determine which GBM variables are most important. Ability to create VIVI plots using lollipops, barplots, and heatmaps. Can also display dot and radar plots. Built with ggplot2. ","Computes random forest VImps for the conditional inference random forest of the party package.","Used to build Bayesian additive regression tree models. Ability to plot VIVI measures with uncertainty included using barplots. Also includes a suite of model diagnostic  plots and univariate PDP. Built using  base R."],["Both","Both","Both","Both","VImp","VInt","VInt","Both","VImp","Both","Vimp","Both"],["Agnostic","Agnostic","Agnostic","Agnostic","Agnostic","Agnostic","Agnostic","Specific","Specific","Specific","Specific","Specific"],["Global","Both","Global","Both","Local","Global","Local","Global","Global","Global","Global","Global"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>Package<\/th>\n      <th>Description<\/th>\n      <th>VIVI<\/th>\n      <th>Measure<\/th>\n      <th>Method<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":12,"columnDefs":[],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[10,12,25,50,100]},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
<p>Model-agnostic methods are techniques that can, in principle, be applied
to any ML algorithm. Agnostic methods not only provide flexibility in
relation to model selection but are also useful for comparing different
fitted ML models. An example of a model agnostic approach for evaluating
VImps is permutation importance <span class="citation" data-cites="breiman2001randomforest">(<a href="#ref-breiman2001randomforest" role="doc-biblioref">Breiman 2001</a>)</span>. This method
calculates the difference in a model’s predictive performance following
a variable’s permutation; implementations are available in the packages <a href="https://cran.r-project.org/package=iml">iml</a>, <a href="https://cran.r-project.org/package=flashlight">flashlight</a> <span class="citation" data-cites="flash">(<a href="#ref-flash" role="doc-biblioref">Mayer 2023</a>)</span>, <a href="https://cran.r-project.org/package=DALEX">DALEX</a> and <a href="https://cran.r-project.org/package=vip">vip</a> <span class="citation" data-cites="vip">(<a href="#ref-vip" role="doc-biblioref">Greenwell and Boehmke 2020</a>)</span>. Each package provides options for specifying the performance metric to use in computing the model performance as well as providing options to select the number of permutations, with <a href="https://cran.r-project.org/package=flashlight">flashlight</a> and <a href="https://cran.r-project.org/package=DALEX">DALEX</a> packages additionally providing an option to select the number of observations that should be sampled for the calculation of variable importance. The <a href="https://cran.r-project.org/package=vip">vip</a> package also provides a PDP/ICE-based variable importance method, which quantifies the variability in PDP and ICE plots. For VInts, Friedman’s <span class="math inline">\(H\)</span>-statistic <span class="citation" data-cites="Friedmans_H">(<a href="#ref-Friedmans_H" role="doc-biblioref">Friedman and Popescu 2008</a>)</span> is an agnostic interaction measure derived from the partial dependence by comparing a pair of variables’ joint effects with the sum of their marginal
effects. Packages <a href="https://cran.r-project.org/package=iml">iml</a> and <a href="https://cran.r-project.org/package=flashlight">flashlight</a> provide implementations.</p>
<p>Partial dependence plots (PDPs) were first introduced by <span class="citation" data-cites="Friedpdp">Friedman (<a href="#ref-Friedpdp" role="doc-biblioref">2000</a>)</span> as
a model agnostic way to visualize the relationship between a specified
predictor variable and the fit, averaging over other predictors’
effects. Similar to PDPs, individual conditional expectation curves (ICE)
<span class="citation" data-cites="iceR">(<a href="#ref-iceR" role="doc-biblioref">Goldstein et al. 2015</a>)</span> show the relationship between a specified predictor and the
fit, fixing the levels of other predictors at those of a particular
observation. PDP curves are then the average of the ICE curves over all
observations in the dataset. R packages offering PDPs include
<a href="https://cran.r-project.org/package=iml">iml</a>, <a href="https://cran.r-project.org/package=DALEX">DALEX</a> (which calls the <a href="https://cran.r-project.org/package=ingredients">ingredients</a> package <span class="citation" data-cites="ingred">(<a href="#ref-ingred" role="doc-biblioref">Biecek and Baniecki 2023</a>)</span>), <a href="https://cran.r-project.org/package=flashlight">flashlight</a>, and <a href="https://cran.r-project.org/package=pdp">pdp</a> <span class="citation" data-cites="pdpR">(<a href="#ref-pdpR" role="doc-biblioref">Greenwell 2017</a>)</span>;
the package <a href="https://cran.r-project.org/package=ICEbox">ICEbox</a> <span class="citation" data-cites="iceR">(<a href="#ref-iceR" role="doc-biblioref">Goldstein et al. 2015</a>)</span> provides ICE curves and variations. Each package provides options for selecting the size of the grid for evaluating the predictions, with <a href="https://cran.r-project.org/package=flashlight">flashlight</a> and <a href="https://cran.r-project.org/package=DALEX">DALEX</a> providing options to select the number of observations to consider in calculating the partial dependence.</p>
<p>In <a href="https://cran.r-project.org/package=vivid">vivid</a> we provide a suite of functions (see Table 2) for calculating and visualizing variable importance, interactions and the partial dependence. Our displays conveniently show (either model specific and agnostic) VImp and VInt jointly using heatmaps and network graphs. Through the use of seriation techniques, we group together variables with the greatest impact on the response. For our network plot, additional filtering to remove less influential variables and clustering options to group highly interacting variables are provided, thus providing a more informative picture identifying relevant features. Additionally, our implementation makes it possible to apply our methods to subsets of data, which leads to locally-based importance measures. Our generalized PDP (GPDP) displays partial dependence plots in a matrix layout combining univariate and bivariate partial dependence plots with variable scatterplots. This layout, coupled with seriation, allows for quick assessment of of how pairs of variables have an impact on the model fit. We furthermore provide a more compact version of the GDPD, the so-called zen-partial dependence plot (ZPDP) consisting only of those bivariate partial dependence plots with high VInt. All of our displays are designed to quickly identify how variables, both singly and jointly, affect the fitted response and can be used for regression or classification fits. As the output of our displays are <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> objects <span class="citation" data-cites="ggplot2">(<a href="#ref-ggplot2" role="doc-biblioref">Wickham 2016</a>)</span>, they are easily customizable and provide the flexibility to create custom VIVI visualizations.</p>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-0b6ec6f6fdf6246683ee" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-0b6ec6f6fdf6246683ee">{"x":{"filter":"none","vertical":false,"caption":"<caption>Summary of functions available in the vivid package. \nThe main construction function is vivi which is used to calculate the VIVI values \nfor subsequent use in the visualizations.<\/caption>","autoHideNavigation":true,"data":[["<tt>vivi<\/tt>","<tt>viviReorder<\/tt>","<tt>viviHeatmap<\/tt>","<tt>viviNetwork<\/tt>","<tt>pdpVars<\/tt>","<tt>pdpPairs<\/tt>","<tt>pdpZen<\/tt>","<tt>CVpredictfun<\/tt>","<tt>zPath<\/tt>","<tt>as.data.frame.vivid<\/tt>","<tt>vip2vivid<\/tt>"],["Create a VIVI matrix of class  <tt>vivid<\/tt>","Reorders a square matrix so high VIVI values are pushed to the top left of the matrix","Heatmap plot of VIVI values","Network plot of VIVI values","Univariate partial dependence plot with ICE curves  displayed as a grid","Pairs plot showing bivariate PDP, ICE/univariate PDP,  and data","A zigzag expanded navigation plot (zenplot) displaying partial dependence values","Predict function","Constructs a zenpath for connecting and displaying pairs  to be used with pdpZen","Takes a matrix of class vivid and turns it into a data frame","Takes measured importance and interactions from the vip  package and turns them into vivid matrix which can be  used for plotting"],["VIVI construction","VIVI construction","Visualization","Visualization","Visualization","Visualization","Visualization","Utility","Utility","Utility","Utility"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th>Function<\/th>\n      <th>Description<\/th>\n      <th>Type<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":11,"columnDefs":[],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[10,11,25,50,100]},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
</div>
<p>This paper is structured as follows. First we introduce a dataset and fit models that will be used as examples throughout this paper. Following this, we describe <a href="https://cran.r-project.org/package=vivid">vivid</a> functionality for calculating VIVI. We then move on to visualizations and focus on the functionality provided by the two functions <code>viviHeatmap</code> and <code>viviNetwork</code> for displaying VIVI, and two functions for displaying PDPs namely, <code>pdpPairs</code> and <code>pdpZen</code>. Finally we provide some concluding discussion.</p>
<h2 data-number="2" id="datamodel"><span class="header-section-number">2</span> Example: Data and Models</h2>
<p>The well-known Boston housing data <span class="citation" data-cites="boston">(<a href="#ref-boston" role="doc-biblioref">Harrison Jr and Rubinfeld 1978</a>)</span> from the R package <a href="https://cran.r-project.org/package=MASS">MASS</a> <span class="citation" data-cites="mass">(<a href="#ref-mass" role="doc-biblioref">Venables and Ripley 2002</a>)</span> concerns prices of 506 houses and 14 predictor variables including property attributes such as number of rooms and social attributes including crime rate and pollution levels. The response is the median value of owner-occupied homes in $1000s (medv).</p>
<p>We first fit a random forest (using the <a href="https://cran.r-project.org/package=randomForest">randomForest</a> package). In order to access all the available embedded variable importance scores, the <code>importance</code> argument must be <code>TRUE</code> when calling the <code>randomForest</code> function. This allows any of the provided importance metrics to be used in <a href="https://cran.r-project.org/package=vivid">vivid</a>. However, in our following examples, we use an agnostic VImp measure supplied by the <a href="https://cran.r-project.org/package=vivid">vivid</a> package, which allows us to directly compare VImp values across different fits.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='st'><a href='https://www.stat.berkeley.edu/~breiman/RandomForests/'>"randomForest"</a></span><span class='op'>)</span> </span>
<span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='st'><a href='http://www.stats.ox.ac.uk/pub/MASS4/'>"MASS"</a></span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>1701</span><span class='op'>)</span>  </span>
<span><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"Boston"</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>rf</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/randomForest/man/randomForest.html'>randomForest</a></span><span class='op'>(</span><span class='va'>medv</span> <span class='op'>~</span><span class='va'>.</span>, data <span class='op'>=</span> <span class='va'>Boston</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>Next we fit a gradient boosted machine (using the <a href="https://cran.r-project.org/package=xgboost">xgboost</a> package). For the GBM we set the maximum number of boosting iterations, <code>nrounds</code>, to 100 as no default is provided in <a href="https://cran.r-project.org/package=xgboost">xgboost</a>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='st'><a href='https://github.com/dmlc/xgboost'>"xgboost"</a></span><span class='op'>)</span> </span>
<span><span class='va'>gbst</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/xgboost/man/xgb.train.html'>xgboost</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='va'>Boston</span><span class='op'>[</span>,<span class='fl'>1</span><span class='op'>:</span><span class='fl'>13</span><span class='op'>]</span><span class='op'>)</span>, </span>
<span>                label <span class='op'>=</span>  <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='va'>Boston</span><span class='op'>[</span>,<span class='fl'>14</span><span class='op'>]</span><span class='op'>)</span>,</span>
<span>                nrounds <span class='op'>=</span> <span class='fl'>100</span>,</span>
<span>                verbose <span class='op'>=</span> <span class='fl'>0</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>In the following sections we will explain how aspects of the two fits can be compared with <a href="https://cran.r-project.org/package=vivid">vivid</a> software. We will also explain aspects of our software design with reference to these fits.</p>
<h2 data-number="3" id="data"><span class="header-section-number">3</span> Calculating VIVI</h2>
<p>The first step in using <a href="https://cran.r-project.org/package=vivid">vivid</a> is to calculate variable importance and interactions for a model fit.
The <code>vivi</code> function calculates both of these, creating a square, symmetric matrix containing variable importance on the diagonal and variable interactions on the off-diagonal. Required inputs are a fitted ML model, a data frame on which the model was trained, and the name of the response variable for the fit. The returned matrix has importance and interaction values for all variables in the supplied data frame, excluding the response. Our visualizations functions <code>viviHeatmap</code> and <code>viviNetwork</code> are designed to show the results of a <code>vivi</code> calculation, but will work equally well for any square matrix with identical row and column names. (Note, the symmetry assumption is not required for <code>viviHeatmap</code> and <code>viviNetwork</code> uses interaction values from the lower-triangular part of the matrix only.)</p>
<p>The code snippet below shows the creation of a <code>vivid</code> matrix for the random forest fit. For clarity, we include all of the <code>vivi</code> function arguments for the random forest fit, though only the first three are required. Other inputs will be described in the section <a href="#sec:vivimatarg"><code>vivi</code> function additional arguments</a>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='st'><a href='https://alaninglis.github.io/vivid/'>"vivid"</a></span><span class='op'>)</span></span>
<span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>1701</span><span class='op'>)</span></span>
<span><span class='va'>viviRf</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://alaninglis.github.io/vivid//reference/vivi.html'>vivi</a></span><span class='op'>(</span>fit <span class='op'>=</span> <span class='va'>rf</span>,</span>
<span>               data <span class='op'>=</span> <span class='va'>Boston</span>,</span>
<span>               response <span class='op'>=</span> <span class='st'>"medv"</span>,</span>
<span>               reorder <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span>               normalized <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span>               importanceType <span class='op'>=</span> <span class='st'>"agnostic"</span>,</span>
<span>               gridSize <span class='op'>=</span> <span class='fl'>50</span>,</span>
<span>               nmax <span class='op'>=</span> <span class='fl'>500</span>,</span>
<span>               class <span class='op'>=</span> <span class='fl'>1</span>,</span>
<span>               predictFun <span class='op'>=</span> <span class='cn'>NULL</span>,</span>
<span>               numPerm <span class='op'>=</span> <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<p>In the absence of any model-specific importance measure we use an agnostic permutation method described by <span class="citation" data-cites="fisher2019all">Fisher et al. (<a href="#ref-fisher2019all" role="doc-biblioref">2019</a>)</span> to obtain the variable importance scores. In this method a model error score (root mean square error) is calculated, then each feature is randomly permuted and the model error is re-calculated. The difference in performance is considered to be the variable importance score for that feature. By default the permutation is set to be replicated four times to account for variability. However, we provide an option to select a desired number of permutations (via the <code>numPerm</code> argument)).</p>
<p>The <code>vivi</code> function calculates both the importance and interactions using S3 methods. By default, the agnostic importance and interaction scores in <a href="https://cran.r-project.org/package=vivid">vivid</a> are computed using the generalized predict function from the <a href="https://cran.r-project.org/package=condvis2">condvis2</a> package <span class="citation" data-cites="condvis2">(<a href="#ref-condvis2" role="doc-biblioref">Hurley et al. 2022</a>)</span>. Consequently, <a href="https://cran.r-project.org/package=vivid">vivid</a> can be used out-of-the-box with any model type that works with <a href="https://cran.r-project.org/package=condvis2">condvis2</a> predict (see <code>CVpredict</code> from <a href="https://cran.r-project.org/package=condvis2">condvis2</a> for a full list of compatible model types). To allow <a href="https://cran.r-project.org/package=vivid">vivid</a> to run with other model fits, a custom predict function must be passed to the <code>predictFun</code> argument (as discussed below).</p>
<p>The S3 method used to obtain the importance is called <code>vividImportance</code>. <code>vivid</code> relies on the <a href="https://cran.r-project.org/package=flashlight">flashlight</a> package to calculate agnostic importance via <code>flashlight::light_importance</code> which currently works for numeric and numeric binary responses only. For model-specific variable importance, we provide individual methods to access importance scores for some of the most popular model fitting <code>R</code> packages, namely; <a href="https://cran.r-project.org/package=ranger">ranger</a> <span class="citation" data-cites="ranger">(<a href="#ref-ranger" role="doc-biblioref">Wright and Ziegler 2017</a>)</span>, <a href="https://cran.r-project.org/package=randomForest">randomForest</a>, <a href="https://cran.r-project.org/package=mlr">mlr</a> <span class="citation" data-cites="mlr">(<a href="#ref-mlr" role="doc-biblioref">Bischl et al. 2016</a>)</span>, <a href="https://cran.r-project.org/package=mlr3">mlr3</a> <span class="citation" data-cites="mlr3">(<a href="#ref-mlr3" role="doc-biblioref">Lang et al. 2019</a>)</span>, and <a href="https://cran.r-project.org/package=parsnip">parsnip</a> <span class="citation" data-cites="parsnip">(<a href="#ref-parsnip" role="doc-biblioref">Kuhn and Vaughan 2022</a>)</span> (however, more could be added via additional methods). To access any available model-specific variable importance from the aforementioned packages, the <code>importanceType</code> argument must be set to equal the selected importance metric. For example, to select the percent increase in mean square error importance score from the <a href="https://cran.r-project.org/package=randomForest">randomForest</a> package, the <code>importanceType</code> argument must be set to equal this measure as it is called within the <a href="https://cran.r-project.org/package=randomForest">randomForest</a> package, that is; <code>importanceType = "%IncMSE"</code>. If the <code>importanceType</code> is not set or is set to equal <code>agnostic</code>, then the agnostic importance is calculated. It should be noted that when comparing different model fits, using model-specific variable importance will result in importance measures that are not directly comparable (however, comparing model specific scores could be useful when comparing the same fitting procedure evaluated using different parameters). In the case that a practitioner may wish to use our visualizations to compare different model fits, we recommend using the agnostic permutation approach supplied by <a href="https://cran.r-project.org/package=vivid">vivid</a> to make a direct comparison of the importance measures.</p>
<p>For variable interactions, we use the model-agnostic Friedman’s <span class="math inline">\(H\)</span>-statistic to identify any pairwise interactions. As discussed in <span class="citation" data-cites="inglis2022visualizing">Inglis et al. (<a href="#ref-inglis2022visualizing" role="doc-biblioref">2022b</a>)</span>, we recommend the unnormalized version of the <span class="math inline">\(H\)</span>-statistic which prevents detection of spurious interactions which can occur when the bivariate partial dependence function (used in the construction of the <span class="math inline">\(H\)</span>-statistic) is flat. In the case of a binary response classification model, we follow <span class="citation" data-cites="elementsStats">Hastie et al. (<a href="#ref-elementsStats" role="doc-biblioref">2009</a>)</span> and compute the <span class="math inline">\(H\)</span>-statistic and partial dependence on the logit scale.</p>
<p>The <code>vivi</code> function calculates interactions using an S3 method called <code>vividInteraction</code>, which again relies on the <a href="https://cran.r-project.org/package=flashlight">flashlight</a> package to calculate Friedman’s <span class="math inline">\(H\)</span>-statistic via
<code>flashlight::light_interaction</code>. Friedman’s <span class="math inline">\(H\)</span>-statistic is the only interaction measure currently available in <a href="https://cran.r-project.org/package=vivid">vivid</a>, though the method of <span class="citation" data-cites="greenwell2018simple">Greenwell et al. (<a href="#ref-greenwell2018simple" role="doc-biblioref">2018</a>)</span> could also be used for this purpose. Embedded interaction measures could easily be incorporated via S3 methods in future.</p>
<p><a href="https://cran.r-project.org/package=flashlight">flashlight</a> simplifies the calculation of VIVI values
as it allows a custom predict function to be supplied for the calculation of agnostic importance and the <span class="math inline">\(H\)</span>-statistic; this flexibility means importance and the <span class="math inline">\(H\)</span>-statistic can be calculated for any ML model. Note that as <a href="https://cran.r-project.org/package=flashlight">flashlight</a> importance and interaction functions act in a model agnostic way, they will give a VIVI of zero for variables in the dataset (except the response) that are not used by the supplied ML fit.
We supply an internal custom predict function called <code>CVpredictfun</code> to both <code>flashlight::light_importance</code> and <code>flashlight::light_interaction</code>. <code>CVpredictfun</code> is a wrapper around <code>CVpredict</code> from the <a href="https://cran.r-project.org/package=condvis2">condvis2</a> package, which adds an option for the classification to select (via the <code>class</code> argument to <code>vivi</code>) the class to be used for prediction and calculates predictions on the logit scale by default.
<code>CVpredict</code> accepts a broad range of fit classes thus streamlining the process of calculating VIVI.</p>
<p>In situations where the fit class is not handled by <code>CVpredict</code> (as is the case for the GBM model created from <a href="https://cran.r-project.org/package=xgboost">xgboost</a>), supplying a custom predict function to the <code>vivi</code> function by way of the <code>predictFun</code> argument allows the agnostic VIVI values to be calculated.
In the code snippet below, we build the <code>vivid</code> matrix for the GBM fit using a custom predict function, which must be of the form given in the code snippet.
For brevity we omit some of the optional <code>vivi</code> function arguments. By default, the agnostic variable importance is used to allow for a direct comparison of the importance measures for both model fits.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># predict function for GBM</span></span>
<span><span class='va'>pFun</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>fit</span>, <span class='va'>data</span>, <span class='va'>...</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/stats/predict.html'>predict</a></span><span class='op'>(</span><span class='va'>fit</span>, <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='va'>data</span><span class='op'>[</span>,<span class='fl'>1</span><span class='op'>:</span><span class='fl'>13</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>1701</span><span class='op'>)</span> </span>
<span><span class='va'>viviGBst</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://alaninglis.github.io/vivid//reference/vivi.html'>vivi</a></span><span class='op'>(</span>fit <span class='op'>=</span> <span class='va'>gbst</span>,</span>
<span>                 data <span class='op'>=</span> <span class='va'>Boston</span>,</span>
<span>                 response <span class='op'>=</span> <span class='st'>"medv"</span>,</span>
<span>                 reorder <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span>                 normalized <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span>                 predictFun <span class='op'>=</span> <span class='va'>pFun</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<h3 data-number="3.1" id="sec:vivimatarg"><span class="header-section-number">3.1</span> <code>vivi</code> function additional arguments</h3>
<p>The <code>vivi</code> function has 11 arguments. Some of these have been discussed above, including <code>fit</code>, <code>data</code>, <code>response</code>, <code>importanceType</code>, and <code>predictFun</code>. Here we provide a summary of the remaining arguments. First, the <code>normalized</code> argument determines if Friedman’s <span class="math inline">\(H\)</span>-statistic should be normalized or not (see <span class="citation" data-cites="inglis2022visualizing">Inglis et al. (<a href="#ref-inglis2022visualizing" role="doc-biblioref">2022b</a>)</span>, for the pros and cons of each version). The arguments <code>gridSize</code> and <code>nmax</code> are used to set the size of the grid for evaluating the predictions and maximum number of data rows to consider, respectively, in the calculation of the <span class="math inline">\(H\)</span>-statistic. Lowering the grid size can provide a significant speed boost, though at the expense of predictive accuracy. Additionally, sampling the data via <code>nmax</code> offers a speed boost. The default values for <code>gridSize</code> and <code>nmax</code> are 50 and 500, respectively.</p>
<h3 data-number="3.2" id="sec:speed"><span class="header-section-number">3.2</span> Speed tests</h3>
<p>A drawback of using Friedman’s <span class="math inline">\(H\)</span>-statistic as a measure of interaction is that it is a computationally expensive calculation, and may be especially time-consuming for models where prediction is slow.
Figure <a href="#fig:speedtest">1</a>
shows the build time (rounded to the nearest second) averaged over five runs for the creation of a <code>vivid</code> matrix with default parameters for different ML algorithms using the Boston Housing data. As the Boston housing data has 13 predictor variables, Friedman’s <span class="math inline">\(H\)</span>-statistic is computed for 91 predictor pairs. The ML algorithms are: GBM, random forest, support vector machine (SVM), neural network (NN), and k-nearest neighbors (KNN). The SVM, NN, and KNN were built using the <a href="https://cran.r-project.org/package=e1071">e1071</a> <span class="citation" data-cites="svmPack">(<a href="#ref-svmPack" role="doc-biblioref">Meyer et al. 2021</a>)</span>, <a href="https://cran.r-project.org/package=nnet">nnet</a> <span class="citation" data-cites="mass">(<a href="#ref-mass" role="doc-biblioref">Venables and Ripley 2002</a>)</span>, and <a href="https://cran.r-project.org/package=kknn">kknn</a> <span class="citation" data-cites="kknn">(<a href="#ref-kknn" role="doc-biblioref">Schliep and Hechenbichler 2016</a>)</span> packages, with the KNN being built through the <a href="https://cran.r-project.org/package=mlr3">mlr3</a> <span class="citation" data-cites="mlr3">(<a href="#ref-mlr3" role="doc-biblioref">Lang et al. 2019</a>)</span> framework. Each of the models were built using their default settings and, for each model fit, the agnostic VIVI was measured. The speed tests were performed on both a 2017 Mac-book Pro 2.3 GHz Dual-Core Intel Core i5 with 8GB of RAM and a 2021 32GB Mac-book M1 Pro. Here we are essentially comparing predict times for the various fits. The NN fit created using the <a href="https://cran.r-project.org/package=nnet">nnet</a> package was the fastest, followed by GBM. Both random forests are the slowest, and surprisingly, the older Mac beats its higher spec cousin for the <a href="https://cran.r-project.org/package=randomForest">randomForest</a> fit.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:speedtest"></span>
<img src="speedTestsPlot.png" alt="Mean time over five runs, on two MacBooks, for the creation of a vivid matrix for different models. Times are highly dependent on the model fit, with NN the fastest and random forests the slowest." width="60%" />
<p class="caption">
Figure 1: Mean time over five runs, on two MacBooks, for the creation of a vivid matrix for different models. Times are highly dependent on the model fit, with NN the fastest and random forests the slowest.
</p>
</div>
</div>
<h3 data-number="3.3" id="sec:vip2vivid"><span class="header-section-number">3.3</span> Alternative construction of a <code>vivid</code> matrix</h3>
<p>A <code>vivid</code> matrix may also be obtained from variable importance and interaction values calculated elsewhere. The package <a href="https://cran.r-project.org/package=vip">vip</a> offers these, and evaluates interactions using a method called the <em>feature importance ranking measure</em> (FIRM, see <span class="citation" data-cites="greenwell2018simple">Greenwell et al. (<a href="#ref-greenwell2018simple" role="doc-biblioref">2018</a>)</span>, for more details). The <code>vip2vivid</code> function we provide in <a href="https://cran.r-project.org/package=vivid">vivid</a> takes VIVI values created in <a href="https://cran.r-project.org/package=vip">vip</a> and turns them into a <code>vivid</code> matrix, that can be subsequently used with our plotting tools. For example, in the code below, model-specific VImp and FIRM VInt scores are calculated for the random forest fit, and subsequently arranged into a <code>vivid</code> matrix with the VImps on the diagonal and VInts on the off-diagonal.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='st'><a href='https://github.com/koalaverse/vip/'>"vip"</a></span><span class='op'>)</span></span>
<span><span class='co'># get model specific VImps using vip package</span></span>
<span><span class='va'>vipVImp</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://koalaverse.github.io/vip/reference/vi.html'>vi</a></span><span class='op'>(</span><span class='va'>rf</span>, method <span class='op'>=</span> <span class='st'>'model'</span><span class='op'>)</span></span>
<span><span class='co'># get VInts using vip package</span></span>
<span><span class='va'>vipVInt</span> <span class='op'>&lt;-</span> <span class='fu'>vint</span><span class='op'>(</span><span class='va'>rf</span>, feature_names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>Boston</span><span class='op'>[</span><span class='op'>-</span><span class='fl'>14</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span></span>
<span></span>
<span><span class='co'># turn into vivi-matrix</span></span>
<span><span class='va'>vipViviMat</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://alaninglis.github.io/vivid//reference/vip2vivid.html'>vip2vivid</a></span><span class='op'>(</span>importance <span class='op'>=</span> <span class='va'>vipVImp</span>, interaction <span class='op'>=</span> <span class='va'>vipVInt</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<h2 data-number="4" id="sec:heatmapSec"><span class="header-section-number">4</span> Heatmap of Variable Importance and Variable Interactions</h2>
<p>The <code>viviHeatmap</code> function constructs a heatmap displaying both importance and interactions, with importance on the diagonal and interactions on the off-diagonals. A <code>vivid</code> matrix is the only required input, which does not necessarily need to be symmetric (for example, the interaction measures from the <a href="https://cran.r-project.org/package=randomForestExplainer">randomForestExplainer</a> package are asymmetric and could be visualized using our heatmap. Color palettes for the importance and interactions are optionally provided via <code>impPal</code> and <code>intPal</code> arguments.
For the default color palette we choose single-hue, color-blind friendly sequential color palettes from <span class="citation" data-cites="colorspaceJSS">Zeileis et al. (<a href="#ref-colorspaceJSS" role="doc-biblioref">2020</a>)</span>, where low and high VIVI values are represented by low and high luminance color values respectively, aiding in highlighting values of interest.</p>
<p>The ordering of the heatmap is taken from the ordering of the input matrix. As <code>reorder</code> was set to <code>FALSE</code> when building both the random forest and GBM fit <code>vivid</code> matrix, the ordering of the heatmaps matches the variable order in the dataset. This is useful for directly comparing multiple heatmaps, however it does not necessarily lend itself for easy identification of the largest VIVI values. If we were to seriate both <code>vivi</code> matrices separately, we would end up with different optimal orderings for each matrix. An alternative is to create a common ordering by averaging over the two <code>vivid</code> matrix objects and applying the <code>vividReorder</code> function to the result.
(This function uses a seriation algorithm based on the techniques of <span class="citation" data-cites="dendserPaper">Earle and Hurley (<a href="#ref-dendserPaper" role="doc-biblioref">2015</a>)</span> designed to place high interaction variables adjacently and to pull high VIVI variables towards the top-left; see <span class="citation" data-cites="inglis2022visualizing">Inglis et al. (<a href="#ref-inglis2022visualizing" role="doc-biblioref">2022b</a>)</span> for details.)
Both VIVI matrices are then re-ordered using the newly obtained variable order.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># average over matrices and seriate to get common ordering</span></span>
<span><span class='va'>viviAvg</span> <span class='op'>&lt;-</span> <span class='op'>(</span><span class='va'>viviRf</span> <span class='op'>+</span> <span class='va'>viviGBst</span><span class='op'>)</span> <span class='op'>/</span> <span class='fl'>2</span></span>
<span><span class='va'>viviAvgReorder</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://alaninglis.github.io/vivid//reference/vividReorder.html'>vividReorder</a></span><span class='op'>(</span><span class='va'>viviAvg</span><span class='op'>)</span></span>
<span></span>
<span><span class='co'># reorder vivi-matrices </span></span>
<span><span class='va'>ord</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>viviAvgReorder</span><span class='op'>)</span></span>
<span><span class='va'>viviRf</span> <span class='op'>&lt;-</span> <span class='va'>viviRf</span><span class='op'>[</span><span class='va'>ord</span>,<span class='va'>ord</span><span class='op'>]</span></span>
<span><span class='va'>viviGBst</span> <span class='op'>&lt;-</span> <span class='va'>viviGBst</span><span class='op'>[</span><span class='va'>ord</span>,<span class='va'>ord</span><span class='op'>]</span></span></code></pre>
</div>
</div>
<p>Arguments <code>impLims</code> and <code>intLims</code> specify the range of importance and interaction values to be mapped to colors. Default values are calculated from the maximum and minimum VIVI values in the <code>vivid</code> matrix. Importance and interaction values falling outside the supplied limits are squished to the closest limit. It can be useful to specify these limits in the situation where there is an extremely large VIVI value that dominates the display, or where we wish two or more plots to have the same limits for comparison purposes, as in the example below. The <code>angle</code> argument is used to rotate the x-axis labels.</p>
<p>Figure <a href="#fig:heatmaps">2</a> shows our improved ordering so that variables with high VIVI values are pushed to the top left of the plots. Filtering can also be applied to the input matrix to display a subset of variables. When compared to the GBM fit in (b), the random forest fit in (a) has weaker interactions and lower importance scores. Both plots identify <span class="math inline">\(lstat\)</span> as being the most important. Both fits also show that <span class="math inline">\(lstat\)</span> (the percentage of lower status of the population) interacts with several other variables, though the interactions are much stronger for the GBM. Notably, the strongest interaction in both fits are different. These are <span class="math inline">\(lstat:crim\)</span> (where <span class="math inline">\(crim\)</span> is the per capita crime rate by town) for the random forest fit and <span class="math inline">\(lstat:nox\)</span> (where <span class="math inline">\(nox\)</span> is parts per 10 million nitrogen oxides concentration) for the GBM fit.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://alaninglis.github.io/vivid//reference/viviHeatmap.html'>viviHeatmap</a></span><span class='op'>(</span><span class='va'>viviRf</span>, angle <span class='op'>=</span> <span class='fl'>45</span>, intLims <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,<span class='fl'>1</span><span class='op'>)</span>, impLims <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,<span class='fl'>8</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://alaninglis.github.io/vivid//reference/viviHeatmap.html'>viviHeatmap</a></span><span class='op'>(</span><span class='va'>viviGBst</span>, angle <span class='op'>=</span> <span class='fl'>45</span>, intLims <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,<span class='fl'>1</span><span class='op'>)</span>, impLims <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,<span class='fl'>8</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:heatmaps"></span>
<img src="RJ-2023-054_files/figure-html5/heatmaps-1.png" alt="Agnostic variable importance and variable interaction scores for a random forest fit in (a) and GBM fit in (b) on the Boston housing data displayed as a heatmap. The random forest fit has weaker interactions and lower importance scores than the GBM fit. Both fits identify $lstat$ as the most important followed by $rm$. In both fits, $lstat$ has numerous interactions with other variables, notably $crim$ in the random forest fit and $nox$ in the GBM fit." width="50%" /><img src="RJ-2023-054_files/figure-html5/heatmaps-2.png" alt="Agnostic variable importance and variable interaction scores for a random forest fit in (a) and GBM fit in (b) on the Boston housing data displayed as a heatmap. The random forest fit has weaker interactions and lower importance scores than the GBM fit. Both fits identify $lstat$ as the most important followed by $rm$. In both fits, $lstat$ has numerous interactions with other variables, notably $crim$ in the random forest fit and $nox$ in the GBM fit." width="50%" />
<p class="caption">
Figure 2: Agnostic variable importance and variable interaction scores for a random forest fit in (a) and GBM fit in (b) on the Boston housing data displayed as a heatmap. The random forest fit has weaker interactions and lower importance scores than the GBM fit. Both fits identify <span class="math inline">\(lstat\)</span> as the most important followed by <span class="math inline">\(rm\)</span>. In both fits, <span class="math inline">\(lstat\)</span> has numerous interactions with other variables, notably <span class="math inline">\(crim\)</span> in the random forest fit and <span class="math inline">\(nox\)</span> in the GBM fit.
</p>
</div>
</div>
<h2 data-number="5" id="networkSec"><span class="header-section-number">5</span> Network of Variable Importance and Variable Interactions</h2>
<p>The <code>viviNetwork</code> function constructs a network graph displaying both importance and interactions.
Similarly to <code>viviHeatmap</code> , this function takes a <code>vivid</code> matrix as the only required input and provides a visual representation of the magnitude of the importance and interaction values through the size of the nodes and edges in the graph, in addition to color. In the plot each variable is represented as a node, with its importance being represented through size and color such that larger, darker nodes indicate a higher importance. Each pairwise interaction is represented by a connecting edge, where larger interaction values get thicker, darker edges; Figure <a href="#fig:networks1">3</a> provides an example. This type of plot enables the user to quickly identify the magnitude of the importance and interactions of the variables that have the most impact on the response. The <code>viviNetwork</code> function optional arguments follows the same conventions as <code>viviHeatmap</code>: custom color palettes for importance and interactions are provided via the <code>impPal</code> and <code>intPal</code>, and the range of VIVI values to be mapped to the colors are specified via the <code>impLims</code> and <code>intLims</code>.</p>
<p>By default, we choose a circular layout to display the graphs, as when coupled with the seriated <code>vivid</code> matrix, variables with high VIVI are grouped in a clock-wise arrangement starting at the top. This arrangement allows for easy identification of variables with high VIVI. Custom layouts are possible by providing a numeric matrix with two columns and one row per node to the <code>layout</code> argument. Additionally, any of the layouts available in the <a href="https://cran.r-project.org/package=igraph">igraph</a> package <span class="citation" data-cites="igraph">(<a href="#ref-igraph" role="doc-biblioref">Csardi and Nepusz 2006</a>)</span> can be specified. The subject of network graph layouts has been extensively studied (for examples see <span class="citation" data-cites="netviz3">Purchase (<a href="#ref-netviz3" role="doc-biblioref">1997</a>)</span>, <span class="citation" data-cites="netviz2">Herman et al. (<a href="#ref-netviz2" role="doc-biblioref">2000</a>)</span>, <span class="citation" data-cites="netviz1">Freeman (<a href="#ref-netviz1" role="doc-biblioref">2000</a>)</span>). It has been shown that certain layouts of network graphs can significantly aid in cognitive interpretation. For example, <span class="citation" data-cites="netviz3">Purchase (<a href="#ref-netviz3" role="doc-biblioref">1997</a>)</span> note that reducing the number of edge crossovers is by far the most important aesthetic (even for small amounts of data), while maximizing symmetry has a lesser effect. Several of the layouts provided by the <a href="https://cran.r-project.org/package=igraph">igraph</a> package can aid interpretation (such as the Sugiyama layout algorithm <span class="citation" data-cites="sugiyama1981methods">(<a href="#ref-sugiyama1981methods" role="doc-biblioref">Sugiyama et al. 1981</a>)</span>, which tries to minimize edge crossover.)</p>
<p>We provide options to filter the graph via the <code>intThreshold</code> and <code>removeNode</code> arguments. This helps to highlight variables with high VIVI scores, which is useful in settings with many predictors. The <code>intThreshold</code> argument filters edges with weight (i.e., VInt value) below a specified value and <code>removeNode</code> removes nodes with no connecting edges after thresholding interaction values. We can optionally cluster similar variables together with respect to their VIVI scores via the <code>cluster</code> argument, thereby aiding in the process of highlighting variables of interest. The <code>cluster</code> argument can take either a vector of cluster memberships for nodes or an appropriate <a href="https://cran.r-project.org/package=igraph">igraph</a> clustering function.</p>
<p>We demonstrate network plots displaying VIVI values for the GBM fit. In Figure <a href="#fig:networks1">3</a>, we show both a default network plot including all variables in (a) and a filtered and clustered network plot in (b). For the filtered plot we select VIVI values in the top decile. This selection allows us to focus only on the variables with the most impact on the response. The variables that remain are <span class="math inline">\(lstat\)</span>, <span class="math inline">\(nox\)</span>, <span class="math inline">\(rm\)</span>, <span class="math inline">\(crim\)</span>, <span class="math inline">\(dis\)</span> (weighted mean of distances to five Boston employment centers), <span class="math inline">\(tax\)</span> (full-value property-tax rate per $10,000), and <span class="math inline">\(ptratio\)</span> (pupil-teacher ratio by town). We then perform a hierarchical clustering treating variable interactions as similarities, with the goal of grouping together high-interaction variables. Here we manually select the number of groups we want to show via the <code>cutree</code> function (which cuts clustered data into a desired number of groups). Finally we rearrange the layout using <a href="https://cran.r-project.org/package=igraph">igraph</a>. Here, <code>igraph::layout_as_star</code> places the first variable (deemed most relevant using the VIVI seriation process above) at the center, which in Figure <a href="#fig:networks1">3</a> (b) emphasizes its key role as the most important predictor which also has the strongest interactions.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># default network plot for GBM</span></span>
<span><span class='fu'><a href='https://alaninglis.github.io/vivid//reference/viviNetwork.html'>viviNetwork</a></span><span class='op'>(</span><span class='va'>viviGBst</span><span class='op'>)</span></span>
<span></span>
<span><span class='co'># clustered and filtered network for GBM</span></span>
<span><span class='va'>intVals</span> <span class='op'>&lt;-</span> <span class='va'>viviGBst</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/diag.html'>diag</a></span><span class='op'>(</span><span class='va'>intVals</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='cn'>NA</span> </span>
<span></span>
<span><span class='co'># select VIVI values in top 10%</span></span>
<span><span class='va'>impTresh</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/quantile.html'>quantile</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/diag.html'>diag</a></span><span class='op'>(</span><span class='va'>viviGBst</span><span class='op'>)</span>,<span class='fl'>.9</span><span class='op'>)</span></span>
<span><span class='va'>intThresh</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/quantile.html'>quantile</a></span><span class='op'>(</span><span class='va'>intVals</span>,<span class='fl'>.9</span>,na.rm<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span></span>
<span><span class='va'>sv</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/which.html'>which</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/diag.html'>diag</a></span><span class='op'>(</span><span class='va'>viviGBst</span><span class='op'>)</span> <span class='op'>&gt;</span> <span class='va'>impTresh</span> <span class='op'>|</span></span>
<span>              <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='va'>intVals</span>, <span class='fl'>1</span>, <span class='va'>max</span>, na.rm<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span> <span class='op'>&gt;</span> <span class='va'>intThresh</span><span class='op'>)</span></span>
<span>              </span>
<span><span class='va'>h</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/hclust.html'>hclust</a></span><span class='op'>(</span><span class='op'>-</span><span class='fu'><a href='https://rdrr.io/r/stats/dist.html'>as.dist</a></span><span class='op'>(</span><span class='va'>viviGBst</span><span class='op'>[</span><span class='va'>sv</span>,<span class='va'>sv</span><span class='op'>]</span><span class='op'>)</span>, method <span class='op'>=</span> <span class='st'>"single"</span><span class='op'>)</span></span>
<span></span>
<span><span class='fu'><a href='https://alaninglis.github.io/vivid//reference/viviNetwork.html'>viviNetwork</a></span><span class='op'>(</span><span class='va'>viviGBst</span><span class='op'>[</span><span class='va'>sv</span>,<span class='va'>sv</span><span class='op'>]</span>,</span>
<span>            intLims <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,<span class='fl'>1</span><span class='op'>)</span>,</span>
<span>            impLims <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,<span class='fl'>8</span><span class='op'>)</span>,</span>
<span>            cluster <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/cutree.html'>cutree</a></span><span class='op'>(</span><span class='va'>h</span>, k <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>, <span class='co'># specify number of groups</span></span>
<span>            layout <span class='op'>=</span> <span class='fu'>igraph</span><span class='fu'>::</span><span class='va'><a href='https://r.igraph.org/reference/layout_as_star.html'>layout_as_star</a></span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:networks1"></span>
<img src="RJ-2023-054_files/figure-html5/networks1-1.png" alt="Network plots showing VIVI scores obtained from a GBM fit on the Boston housing data. In (a) we display the all values in a circle. In (b) we use a hierarchical clustering to group variable with high VIVI together and rearrange the layout using an igraph function." width="50%" /><img src="RJ-2023-054_files/figure-html5/networks1-2.png" alt="Network plots showing VIVI scores obtained from a GBM fit on the Boston housing data. In (a) we display the all values in a circle. In (b) we use a hierarchical clustering to group variable with high VIVI together and rearrange the layout using an igraph function." width="50%" />
<p class="caption">
Figure 3: Network plots showing VIVI scores obtained from a GBM fit on the Boston housing data. In (a) we display the all values in a circle. In (b) we use a hierarchical clustering to group variable with high VIVI together and rearrange the layout using an igraph function.
</p>
</div>
</div>
<p>In Figure <a href="#fig:networks1">3</a>(a), when displaying all the variables, we can clearly identify which variables have the highest VIVI values. The large darker nodes of <span class="math inline">\(lstat\)</span> and <span class="math inline">\(rm\)</span> indicate their importance and the dark, thick connecting edge between <span class="math inline">\(lstat\)</span> and <span class="math inline">\(nox\)</span> tell us that these two variables strongly interact. In (b), after applying a hierarchical clustering, we can see the strongest mutual interactions have been grouped together for the GBM fit. Namely, <span class="math inline">\(lstat\)</span>, <span class="math inline">\(nox\)</span>, <span class="math inline">\(crim\)</span>, <span class="math inline">\(rm\)</span>, and <span class="math inline">\(dis\)</span> are all grouped together. The remaining variables are individually clustered.</p>
<p>We provide a conversion of <code>vivid</code> matrix objects to a data frame via an <code>as.data.frame</code> method, as demonstrated below. This facilitates plotting with base R and <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a>, for example a barplot of either VImp or VInt values. Note that while <code>vivi</code> returns a matrix of class <code>vivid</code>, the class attribute was dropped when the matrix was re-ordered.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/class.html'>class</a></span><span class='op'>(</span><span class='va'>viviRf</span><span class='op'>)</span><span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"vivid"</span>, <span class='fu'><a href='https://rdrr.io/r/base/class.html'>class</a></span><span class='op'>(</span><span class='va'>viviRf</span><span class='op'>)</span><span class='op'>)</span> </span>
<span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/as.data.frame.html'>as.data.frame</a></span><span class='op'>(</span><span class='va'>viviRf</span><span class='op'>)</span>, <span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>  Variable_1 Variable_2      Value Measure Row Col
1      lstat      lstat 4.80970237    Vimp   1   1
2        nox      lstat 0.06387693    Vint   2   1
3         rm      lstat 0.15226649    Vint   3   1
4       crim      lstat 0.44728494    Vint   4   1</code></pre>
</div>
<h2 data-number="6" id="GPDP"><span class="header-section-number">6</span> Partial Dependence and Individual Conditional Expectation Curves</h2>
<h3 data-number="6.1" id="univariate-partial-dependence-plot"><span class="header-section-number">6.1</span> Univariate Partial Dependence Plot</h3>
<p>The <code>pdpVars</code> function constructs a grid of univariate PDPs with ICE curves for selected variables. We use ICE curves to assist in the identification of linear or non-linear effects. The fit, data frame used to train the model, and the name of the response variable are required inputs. In the code below, we show an example of the partial dependence and ICE curves for the first five features from the GBM <code>vivid</code> matrix, with output shown in Figure <a href="#fig:pdpRf">4</a>. We use the custom GBM predict function given previously.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>top5</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>viviGBst</span><span class='op'>)</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>5</span><span class='op'>]</span></span>
<span><span class='fu'><a href='https://alaninglis.github.io/vivid//reference/pdpVars.html'>pdpVars</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>Boston</span>,</span>
<span>        fit <span class='op'>=</span> <span class='va'>gbst</span>,</span>
<span>        response <span class='op'>=</span> <span class='st'>"medv"</span>,</span>
<span>        vars <span class='op'>=</span> <span class='va'>top5</span>,</span>
<span>        predictFun <span class='op'>=</span> <span class='va'>pFun</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:pdpRf"></span>
<img src="RJ-2023-054_files/figure-html5/pdpRf-1.png" alt="Partial dependence plots (black line) with individual conditional expectation curves (colored lines) of a GBM fit on the Boston housing data. The changing partial dependence and ICE curves of $lstat$ and $rm$ indicate that these variables have some impact on the response." width="100%" />
<p class="caption">
Figure 4: Partial dependence plots (black line) with individual conditional expectation curves (colored lines) of a GBM fit on the Boston housing data. The changing partial dependence and ICE curves of <span class="math inline">\(lstat\)</span> and <span class="math inline">\(rm\)</span> indicate that these variables have some impact on the response.
</p>
</div>
</div>
<p>All of our PDP variants handle categorical responses and predictors. The color palette is customized via the <code>pal</code> argument. In all of our PDPs, this defaults to a diverging palette which accentuates fitted values that differ from the average. Dark red and dark blue are used to indicate high and low values of <span class="math inline">\(\hat{y}\)</span> respectively. The middle values are displayed in yellow. The <code>nIce</code> argument specifies the number of ICE curves to be drawn. This is either a single number specifying the number of observations to be sampled for the ICE curves, or a vector of row indices, an option that is useful for example to display ICE curves from particular classes. The default value for <code>nIce</code> is 30, which allows individual curves to be seen.</p>
<p>The ordering of the PDPs is taken from the ordering of variables in the data set, or may be specified via the <code>vars</code> argument. In Figure <a href="#fig:pdpRf">4</a>, the ordering is taken directly from our seriated <code>vivid</code> matrix, thereby showing the top five most influential variables. As with the construction of the <code>vivid</code> matrix, the <code>gridSize</code> and <code>nmax</code> arguments determine the number of predictions.</p>
<p>In Figure <a href="#fig:pdpRf">4</a> we can see from the changing PDP and ICE curves that <span class="math inline">\(lstat\)</span> and <span class="math inline">\(rm\)</span> have the clearest impact on the response, with the predicted median house price being higher for low values of <span class="math inline">\(lstat\)</span> and high values of <span class="math inline">\(rm\)</span>. Additionally, the predicted median house price appears to be higher for low values of <span class="math inline">\(dis\)</span> before leveling off at around 2.5. The remaining variables have generally flat partial dependence and ICE curves.</p>
<h3 data-number="6.2" id="generalized-pairs-partial-dependence-plot"><span class="header-section-number">6.2</span> Generalized Pairs Partial Dependence Plot</h3>
<p>The <code>pdpPairs</code> function creates a generalized pairs partial dependence plot (GPDP). In our GPDP, we use a matrix layout and plot the univariate partial dependence (with ICE curves) on the diagonal, bivariate partial dependence on the upper diagonal and a scatterplot of raw variable values on the lower diagonal, where all colors are assigned to points and ICE curves by the predicted <span class="math inline">\(\hat{y}\)</span> value. In the case of categorical predictors, the partial dependence for each factor level is shown in the upper-diagonal (for an example of this, see <span class="citation" data-cites="inglis2022visualizing">Inglis et al. (<a href="#ref-inglis2022visualizing" role="doc-biblioref">2022b</a>)</span>). As with the univariate PDP, the fit, data frame used to train the model, and the name of the response variable are required inputs.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>1701</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://alaninglis.github.io/vivid//reference/pdpPairs.html'>pdpPairs</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>Boston</span>,</span>
<span>         fit <span class='op'>=</span> <span class='va'>gbst</span>,</span>
<span>         response <span class='op'>=</span> <span class='st'>"medv"</span>,</span>
<span>         gridSize <span class='op'>=</span> <span class='fl'>20</span>,</span>
<span>         nIce <span class='op'>=</span> <span class='fl'>50</span>,</span>
<span>         vars <span class='op'>=</span> <span class='va'>top5</span>,</span>
<span>         convexHull <span class='op'>=</span> <span class='cn'>TRUE</span>,</span>
<span>         fitlims <span class='op'>=</span> <span class='st'>"pdp"</span>,</span>
<span>         predictFun <span class='op'>=</span> <span class='va'>pFun</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:gpdp"></span>
<img src="RJ-2023-054_files/figure-html5/gpdp-1.png" alt="Filtered generalized pairs partial dependence plot for a GBM fit on the Boston housing data. From both the univariate and bivariate PDPs, we can see that $lstat$ and $rm$ have an impact on the response. As $lstat$ decreases and $rm$ increases, predicted median house price value goes up. The bivariate PDP of $lstat:nox$ shows that as $nox$ increases, the predicted value decreases." width="75%" />
<p class="caption">
Figure 5: Filtered generalized pairs partial dependence plot for a GBM fit on the Boston housing data. From both the univariate and bivariate PDPs, we can see that <span class="math inline">\(lstat\)</span> and <span class="math inline">\(rm\)</span> have an impact on the response. As <span class="math inline">\(lstat\)</span> decreases and <span class="math inline">\(rm\)</span> increases, predicted median house price value goes up. The bivariate PDP of <span class="math inline">\(lstat:nox\)</span> shows that as <span class="math inline">\(nox\)</span> increases, the predicted value decreases.
</p>
</div>
</div>
<p>As with the univariate PDP, the ordering can be controlled via the <code>var</code> argument. By default, the ordering is taken from the order of the data. In the code above, we display only the interesting variables seen in previous plots by selecting the first five variables from our seriated <code>vivid</code> matrix. We also chose to display 50 ICE curves. As with <code>pdpVars</code>, additional arguments specify the color palette and number of ice curves, while <code>gridSize</code> and <code>nmax</code> determine the number of predictions.</p>
<p>For our GPDP, we follow the general design choices in <a href="https://cran.r-project.org/package=vivid">vivid</a> and specify the range of predicted values to be mapped to the colors via the <code>fitlims</code> argument. We set the default fit range for the color map for the GPDP to the range of the collection of PDP surfaces with <code>fitlims = "pdp"</code>. The setting of this argument at its default value allows for maximum resolution of the bivariate PDPs. Since predictions for specific observations and ICE curves would likely exceed these bounds, the closest value within the color map’s bounds is used to allocate colors. Alternatively <code>fitlims = 'all"</code> specifies that limits are calculated as the full range of predictions shown.</p>
<p>In the upper diagonals we exclude extrapolated areas from the bivariate PDPs to prevent interpretation of the PDPs in areas where there is no data. The removal of extrapolated areas can be prevented by specifying <code>convexHull = FALSE</code>.</p>
<p>In Figure <a href="#fig:gpdp">5</a>, in addition to the univariate PDPs, we capture the effects of the variables on the response via the bivariate PDP on the upper-diagonal and the distribution of the data in the lower-diagonal. The scatterplots are useful for determining if variables are highly correlated, as highly correlated variables may spuriously affect the partial dependence and give erroneous results <span class="citation" data-cites="apley2020visualizing">(<a href="#ref-apley2020visualizing" role="doc-biblioref">Apley and Zhu 2020</a>)</span>. Of note are the variables <span class="math inline">\(lstat\)</span> and <span class="math inline">\(rm\)</span>. We can clearly see that when the number of rooms (<span class="math inline">\(rm\)</span>) is high and the percentage of lower status of the population (<span class="math inline">\(lstat\)</span>) is low, the predicted <span class="math inline">\(\hat{y}\)</span> median house price value is high. This is exemplified in the changing bivariate PDP.</p>
<h2 data-number="7" id="sec:ZPDP"><span class="header-section-number">7</span> Zen Partial Dependence Plots</h2>
<p>The <code>pdpZen</code> function creates partial dependence plots utilizing a space-saving method based on graph Eulerians <span class="citation" data-cites="hierholzer1873moglichkeit">(<a href="#ref-hierholzer1873moglichkeit" role="doc-biblioref">Hierholzer and Wiener 1873</a>)</span>. An Eulerian path, also known as an Eulerian trail, is a route that traverses each edge of a graph exactly once. When this path forms a closed loop, the traversal is referred to as an Eulerian tour. We call this display zen-partial dependence plots (ZPDP). The display is based on the zigzag expanded navigation plots, known as zenplots , available in the <a href="https://cran.r-project.org/package=zenplots">zenplots</a> package <span class="citation" data-cites="zenplot">(<a href="#ref-zenplot" role="doc-biblioref">Hofert and Oldford 2020</a>)</span>. Zenplots were created to display paired graphs of high-dimensional data focusing on the most important 2D displays.
In our adaptation we show bivariate PDPs that focus on the variables with the largest interaction values in a compact zigzag layout, which is helpful when predictor space is high-dimensional.</p>
<p>The code below illustrates <code>pdpZen</code>, here displaying the first five variables from GBM’s <code>vivid</code> matrix. Later we show an example focusing on high-interacting pairs of variables.
We use the same convention as our previous PDPs with regard to color palette and limits, grid size, and the number of rows considered for evaluation. The ZPDP also has a variable rug plot on each axis to avoid interpretation problems that may occur in the presence of skewness.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://alaninglis.github.io/vivid//reference/pdpZen.html'>pdpZen</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>Boston</span>,</span>
<span>       fit <span class='op'>=</span> <span class='va'>gbst</span>,</span>
<span>       response <span class='op'>=</span> <span class='st'>"medv"</span>,</span>
<span>       convexHull <span class='op'>=</span> <span class='cn'>TRUE</span>,</span>
<span>       zpath <span class='op'>=</span> <span class='va'>top5</span>,</span>
<span>       predictFun <span class='op'>=</span> <span class='va'>pFun</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:zpdpGbm1"></span>
<img src="RJ-2023-054_files/figure-html5/zpdpGbm1-1.png" alt="Zen partial dependence plot for the GBM fit on the Boston data. Here we display first five variables from the GBM's `vivid` matrix. Only plots for consecutive variables are shown." width="50%" />
<p class="caption">
Figure 6: Zen partial dependence plot for the GBM fit on the Boston data. Here we display first five variables from the GBM’s <code>vivid</code> matrix. Only plots for consecutive variables are shown.
</p>
</div>
</div>
<p>The argument <code>zpath</code> specifies the variables to be plotted, defaulting to all dataset variables aside from the response. In the code above, <code>zpath</code> is the vector <span class="math inline">\(lstat\)</span>, <span class="math inline">\(nox\)</span>, <span class="math inline">\(rm\)</span>, <span class="math inline">\(crim\)</span> and <span class="math inline">\(dis\)</span>. The resulting plot shown in Figure <a href="#fig:zpdpGbm1">6</a> presents the bivariate PDP for every consecutive pair of variables in a zigzag layout.</p>
<h3 data-number="7.1" id="sec:zp"><span class="header-section-number">7.1</span> <code>Zen-paths</code></h3>
<p>ZPDP are most useful when the bivariate PDPs plotted are selected to be an interesting subset of all pairwise plots. To obtain this subset, we consider a network graph displaying VIVI values, such as that in Figure <a href="#fig:networks1">3</a> (a). We then filter the edges below a selected interaction value, leaving only highly interacting variable pairs, as in Figure <a href="#fig:networks1">3</a>(b). Our goal is to then build a ZPDP consisting of the bivariate plots represented by each edge of the thresholded graph. The <code>zPath</code> function creates a sequence or sequences of variable paths for use in <code>pdpZen</code>.</p>
<p>The <code>zPath</code> function takes four arguments. These are:
<code>viv</code> - a matrix of interaction values,
<code>cutoff</code> - exclude interaction values below this threshold,
<code>method</code> - a string indicating which method to use to create the path, and
<code>connect</code> - a logical value indicating if separate Eulerians should be connected</p>
<p>Two methods are provided, either <code>"greedy.weighted"</code> or <code>"strictly.weighted"</code>. The first option uses the greedy Eulerian path algorithm <span class="citation" data-cites="PairViz1 pviz">(<a href="#ref-PairViz1" role="doc-biblioref">Hurley and Oldford 2011</a>, <a href="#ref-pviz" role="doc-biblioref">2022</a>)</span> for connected graphs. This visits each edge at least once, beginning at the edge with the highest weight and traversing through the remaining edges, giving priority to the highest-weighted edge. Some edges may be visited more than once or additional edges may be visited if the number of nodes in the graph is not even. The second method <code>"strictly.weighted"</code> (provided by <a href="https://cran.r-project.org/package=zenplot">zenplot</a>) visits edges strictly in decreasing order by weight (here the interaction values). If <code>connect</code> is <code>TRUE</code> the sequences obtained by the strictly weighted method are concatenated to form a single path.</p>
<p>In the code below, we provide two examples of creating zen-paths, from the top 10% of interaction scores in <code>viviGBst</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>intThresh</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/quantile.html'>quantile</a></span><span class='op'>(</span><span class='va'>intVals</span>, <span class='fl'>.9</span>, na.rm<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span></span>
<span><span class='co'># set zpaths with different parameters</span></span>
<span><span class='va'>zpGw</span>  <span class='op'>&lt;-</span> <span class='fu'><a href='https://alaninglis.github.io/vivid//reference/zPath.html'>zPath</a></span><span class='op'>(</span>viv <span class='op'>=</span> <span class='va'>viviGBst</span>, cutoff <span class='op'>=</span> <span class='va'>intThresh</span>, method <span class='op'>=</span> <span class='st'>"greedy.weighted"</span><span class='op'>)</span></span>
<span><span class='va'>zpGw</span></span></code></pre>
</div>
<pre><code> [1] &quot;nox&quot;     &quot;lstat&quot;   &quot;dis&quot;     &quot;ptratio&quot; &quot;lstat&quot;   &quot;rm&quot;     
 [7] &quot;crim&quot;    &quot;lstat&quot;   &quot;tax&quot;     &quot;rm&quot;      &quot;nox&quot;    </code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>zpSw</span>  <span class='op'>&lt;-</span> <span class='fu'><a href='https://alaninglis.github.io/vivid//reference/zPath.html'>zPath</a></span><span class='op'>(</span>viv <span class='op'>=</span> <span class='va'>viviGBst</span>, cutoff <span class='op'>=</span> <span class='va'>intThresh</span>, connect <span class='op'>=</span> <span class='cn'>FALSE</span>, method <span class='op'>=</span> <span class='st'>"strictly.weighted"</span><span class='op'>)</span></span>
<span><span class='va'>zpSw</span></span></code></pre>
</div>
<pre><code>[[1]]
[1] &quot;nox&quot;   &quot;lstat&quot; &quot;dis&quot;  

[[2]]
[1] &quot;lstat&quot; &quot;rm&quot;    &quot;nox&quot;  

[[3]]
[1] &quot;lstat&quot; &quot;crim&quot;  &quot;rm&quot;   

[[4]]
[1] &quot;ptratio&quot; &quot;lstat&quot;   &quot;tax&quot;    </code></pre>
</div>
<p>Our first created zen-path object, <code>zpGw</code>, uses the <code>greedy.weighted</code> method and visits each edge exactly once. The second zen-path, <code>zpSw</code>, uses the <code>strictly.weighted</code> method with <code>connect = FALSE</code>.
<code>zpSw</code> consists of four unconnected paths.
The zenplots for two of these paths are constructed below.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://alaninglis.github.io/vivid//reference/pdpZen.html'>pdpZen</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>Boston</span>,</span>
<span>       fit <span class='op'>=</span> <span class='va'>gbst</span>,</span>
<span>       response <span class='op'>=</span> <span class='st'>"medv"</span>,</span>
<span>       zpath <span class='op'>=</span> <span class='va'>zpGw</span>,</span>
<span>       convexHull <span class='op'>=</span> <span class='cn'>TRUE</span>,</span>
<span>       predictFun <span class='op'>=</span> <span class='va'>pFun</span><span class='op'>)</span> </span>
<span></span>
<span><span class='fu'><a href='https://alaninglis.github.io/vivid//reference/pdpZen.html'>pdpZen</a></span><span class='op'>(</span>data <span class='op'>=</span> <span class='va'>Boston</span>,</span>
<span>       fit <span class='op'>=</span> <span class='va'>gbst</span>,</span>
<span>       response <span class='op'>=</span> <span class='st'>"medv"</span>,</span>
<span>       zpath <span class='op'>=</span> <span class='va'>zpSw</span>,</span>
<span>       convexHull <span class='op'>=</span> <span class='cn'>TRUE</span>,</span>
<span>       predictFun <span class='op'>=</span> <span class='va'>pFun</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:zpdpGBM"></span>
<img src="RJ-2023-054_files/figure-html5/zpdpGBM-1.png" alt="ZPDP for a GBM fit on the Boston data. In (a) the zpath is defined by the `greedy.weighted` sorting method. In (b), the sorting method is defined by the `strictly.weighted` method and is unconnected. For low values of $lstat$ and and high values of $rm$, predicted median house price value increases." width="50%" /><img src="RJ-2023-054_files/figure-html5/zpdpGBM-2.png" alt="ZPDP for a GBM fit on the Boston data. In (a) the zpath is defined by the `greedy.weighted` sorting method. In (b), the sorting method is defined by the `strictly.weighted` method and is unconnected. For low values of $lstat$ and and high values of $rm$, predicted median house price value increases." width="50%" />
<p class="caption">
Figure 7: ZPDP for a GBM fit on the Boston data. In (a) the zpath is defined by the <code>greedy.weighted</code> sorting method. In (b), the sorting method is defined by the <code>strictly.weighted</code> method and is unconnected. For low values of <span class="math inline">\(lstat\)</span> and and high values of <span class="math inline">\(rm\)</span>, predicted median house price value increases.
</p>
</div>
</div>
<p>Note that there are 7 different variables involved in high interactions, which could be displayed in a <span class="math inline">\(7 \times 7\)</span> GPDP, showing a total of 21 bivariate PDPs. But only 8 of these have VInt values above the 90% quantile, and Figure <a href="#fig:zpdpGBM">7</a>(b) using the <code>strictly.weighted</code> path shows just these bivariate PDPs’ compact layout. Using the <code>greedy.weighted</code> sorting method in (a) produces a smaller, neater plot but at the expense of including some plots that are not particularly interesting (for example the pair <span class="math inline">\(dis:ptratio\)</span>). However, it should be noted that any of the arguments from the <code>zenplot</code> function from the <a href="https://cran.r-project.org/package=zenplots">zenplots</a> package can be used with the <code>pdpZen</code> function. These include multiple options for different or custom layouts.</p>
<h2 data-number="8" id="summary"><span class="header-section-number">8</span> Summary</h2>
<p>We have presented a detailed exposition of our R package <a href="https://cran.r-project.org/package=vivid">vivid</a> which contains a suite of integrated functions implementing algorithms and novel visualizations for exploring variable importance and variable interactions in machine learning models. Our techniques are intuitive, adaptable, easy to customize and facilitate model comparison. When building the <code>vivid</code> matrix to use in our heatmap and network visualizations, VIVI metrics that are model specific or model-agnostic may be employed. For measuring interactions we currently only provide the option to use the agnostic Friedman’s <span class="math inline">\(H\)</span>-statistic. However, as outlined in the <a href="#data">Calculating VIVI</a> section, the inclusion of different VIVI measures is easily possible.</p>
<p>Our <a href="https://cran.r-project.org/package=vivid">vivid</a> package is a useful addition to the other packages in the area of model visualization, such as those discussed in the <a href="#introduction">Introduction</a> section. Our heatmap and network plots efficiently determine which variables have the greatest impact on the response. When coupled with the seriation, filtering, and clustering techniques, these visualizations enhance the interpretation of ML predictions. Our GPDP and ZPDP can be used to provide a thorough examination of the behavior of a fitted ML model by examining the individual variable effects and their pairwise interactions. These plots combine the bivariate PDP, ICE curves, and scatterplots of the raw variable values. They further allow focusing on subsets of variables with high VInt, and so allow us to efficiently explore a fitted ML model by focusing attention to only the most crucial aspects.</p>
<p>It has been noted that the presence of correlated variables can lead to biased VIVI measures. For example, <span class="citation" data-cites="Friedmans_H">Friedman and Popescu (<a href="#ref-Friedmans_H" role="doc-biblioref">2008</a>)</span> note that when subsets of variables are highly correlated, it becomes difficult to distinguish between low and higher order interactions among them in a straightforward manner when using the <span class="math inline">\(H\)</span>-statistic. Similarly, <span class="citation" data-cites="fisher2019all">Fisher et al. (<a href="#ref-fisher2019all" role="doc-biblioref">2019</a>)</span> state that for permutation variable importance, the presence of highly correlated variables can affect the measured variable importance and propose using a method of conditional importance, such as that implemented for conditional random forests in the <a href="https://cran.r-project.org/package=party">party</a> package. Our GPDP can be used to identify any evident correlations by observing the plot of the data and the PDP with convex hull. However, careful consideration should be employed when interpreting the VIVI measures from <a href="https://cran.r-project.org/package=vivid">vivid</a> and we recommend evaluating any potential correlation between variables in conjunction with our proposed visualizations. Several <code>R</code> packages are available for assessing and visualising correlations, such as <a href="https://cran.r-project.org/package=corrplot">corrplot</a> <span class="citation" data-cites="corrplot2021">(<a href="#ref-corrplot2021" role="doc-biblioref">Wei and Simko 2021</a>)</span> and <a href="https://cran.r-project.org/package=corrgrapher">corrgrapher</a> <span class="citation" data-cites="corrgrapher">(<a href="#ref-corrgrapher" role="doc-biblioref">Morgen and Biecek 2020</a>)</span>.</p>
<p>For future work, the inclusion of other model summaries could be incorporated into <a href="https://cran.r-project.org/package=vivid">vivid</a>, such as the interaction statistics described in <span class="citation" data-cites="greenwell2018simple">Greenwell et al. (<a href="#ref-greenwell2018simple" role="doc-biblioref">2018</a>)</span> or the use of accumulated local effects (ALE) of <span class="citation" data-cites="apley2020visualizing">Apley and Zhu (<a href="#ref-apley2020visualizing" role="doc-biblioref">2020</a>)</span>. This latter method was created to address bias problems with partial dependency functions and could be used in place of the bivariate PDPs seen in both the GPDP and ZPDP. However the calculation of an agnostic, easily interpretable variable interaction measure that accounts for correlated variables remains an ongoing research goal. Allowing a user to view the variation across replicates when performing permutation importance could also be incorporated. This could be added into the heatmap graphic by way of a value suppressing uncertainty palette <span class="citation" data-cites="vsup">(<a href="#ref-vsup" role="doc-biblioref">Correll et al. 2018</a>)</span>, where the uncertainty in included in the visualization (for example, see <span class="citation" data-cites="bartvis">Inglis et al. (<a href="#ref-bartvis" role="doc-biblioref">2022a</a>)</span>). Additionally, providing alternative metrics such as AUC or accuracy for computation of permutation importance in classification, as well as computations and visualizations of higher order interactions, interactive graphics and incorporating multivariate response variables are interesting areas for future research.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h3 class="appendix" data-number="8.1" id="supplementary-materials"><span class="header-section-number">8.1</span> Supplementary materials</h3>
<p>Supplementary materials are available in addition to this article. It can be downloaded at
<a href="RJ-2023-054.zip">RJ-2023-054.zip</a></p>
<h3 class="appendix" data-number="8.2" id="cran-packages-used"><span class="header-section-number">8.2</span> CRAN packages used</h3>
<p><a href="https://cran.r-project.org/package=vivid">vivid</a>, <a href="https://cran.r-project.org/package=vip">vip</a>, <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a>, <a href="https://cran.r-project.org/package=iml">iml</a>, <a href="https://cran.r-project.org/package=flashlight">flashlight</a>, <a href="https://cran.r-project.org/package=DALEX">DALEX</a>, <a href="https://cran.r-project.org/package=lime">lime</a>, <a href="https://cran.r-project.org/package=pdp">pdp</a>, <a href="https://cran.r-project.org/package=ICEbox">ICEbox</a>, <a href="https://cran.r-project.org/package=randomForestExplainer">randomForestExplainer</a>, <a href="https://cran.r-project.org/package=randomForest">randomForest</a>, <a href="https://cran.r-project.org/package=EIX">EIX</a>, <a href="https://cran.r-project.org/package=varImp">varImp</a>, <a href="https://cran.r-project.org/package=party">party</a>, <a href="https://cran.r-project.org/package=bartMachine">bartMachine</a>, <a href="https://cran.r-project.org/package=ingredients">ingredients</a>, <a href="https://cran.r-project.org/package=MASS">MASS</a>, <a href="https://cran.r-project.org/package=xgboost">xgboost</a>, <a href="https://cran.r-project.org/package=condvis2">condvis2</a>, <a href="https://cran.r-project.org/package=ranger">ranger</a>, <a href="https://cran.r-project.org/package=mlr">mlr</a>, <a href="https://cran.r-project.org/package=mlr3">mlr3</a>, <a href="https://cran.r-project.org/package=parsnip">parsnip</a>, <a href="https://cran.r-project.org/package=e1071">e1071</a>, <a href="https://cran.r-project.org/package=nnet">nnet</a>, <a href="https://cran.r-project.org/package=kknn">kknn</a>, <a href="https://cran.r-project.org/package=igraph">igraph</a>, <a href="https://cran.r-project.org/package=zenplots">zenplots</a>, <a href="https://cran.r-project.org/package=zenplot">zenplot</a>, <a href="https://cran.r-project.org/package=corrplot">corrplot</a>, <a href="https://cran.r-project.org/package=corrgrapher">corrgrapher</a></p>
<h3 class="appendix" data-number="8.3" id="cran-task-views-implied-by-cited-packages"><span class="header-section-number">8.3</span> CRAN Task Views implied by cited packages</h3>
<p><a href="https://cran.r-project.org/view=Bayesian">Bayesian</a>, <a href="https://cran.r-project.org/view=Cluster">Cluster</a>, <a href="https://cran.r-project.org/view=Distributions">Distributions</a>, <a href="https://cran.r-project.org/view=Econometrics">Econometrics</a>, <a href="https://cran.r-project.org/view=Environmetrics">Environmetrics</a>, <a href="https://cran.r-project.org/view=GraphicalModels">GraphicalModels</a>, <a href="https://cran.r-project.org/view=HighPerformanceComputing">HighPerformanceComputing</a>, <a href="https://cran.r-project.org/view=MachineLearning">MachineLearning</a>, <a href="https://cran.r-project.org/view=MissingData">MissingData</a>, <a href="https://cran.r-project.org/view=MixedModels">MixedModels</a>, <a href="https://cran.r-project.org/view=ModelDeployment">ModelDeployment</a>, <a href="https://cran.r-project.org/view=NumericalMathematics">NumericalMathematics</a>, <a href="https://cran.r-project.org/view=Optimization">Optimization</a>, <a href="https://cran.r-project.org/view=Phylogenetics">Phylogenetics</a>, <a href="https://cran.r-project.org/view=Psychometrics">Psychometrics</a>, <a href="https://cran.r-project.org/view=Robust">Robust</a>, <a href="https://cran.r-project.org/view=Spatial">Spatial</a>, <a href="https://cran.r-project.org/view=Survival">Survival</a>, <a href="https://cran.r-project.org/view=TeachingStatistics">TeachingStatistics</a></p>
<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-antunes2018fairness" class="csl-entry" role="listitem">
N. Antunes, L. Balby, F. Figueiredo, N. Lourenco, W. Meira and W. Santos. Fairness and transparency of machine learning for trustworthy cloud services. In <em>2018 48th annual IEEE/IFIP international conference on dependable systems and networks workshops (DSN-w)</em>, pages. 188–193 2018. IEEE.
</div>
<div id="ref-apley2020visualizing" class="csl-entry" role="listitem">
D. W. Apley and J. Zhu. Visualizing the effects of predictor variables in black box supervised learning models. <em>Journal of the Royal Statistical Society: Series B (Statistical Methodology)</em>, 82(4): 1059–1086, 2020.
</div>
<div id="ref-dalex" class="csl-entry" role="listitem">
P. Biecek. DALEX: Explainers for complex predictive models in <span>R</span>. <em>Journal of Machine Learning Research</em>, 19(84): 1–5, 2018. URL <a href="https://jmlr.org/papers/v19/18-416.html">https://jmlr.org/papers/v19/18-416.html</a>.
</div>
<div id="ref-ingred" class="csl-entry" role="listitem">
P. Biecek and H. Baniecki. <em>Ingredients: Effects and importances of model ingredients.</em> 2023. URL <a href="https://CRAN.R-project.org/package=ingredients">https://CRAN.R-project.org/package=ingredients</a>. R package version 2.3.0.
</div>
<div id="ref-biecek2021explanatory" class="csl-entry" role="listitem">
P. Biecek and T. Burzykowski. <em>Explanatory model analysis: Explore, explain and examine predictive models.</em> Chapman; Hall/CRC, 2021.
</div>
<div id="ref-mlr" class="csl-entry" role="listitem">
B. Bischl, M. Lang, L. Kotthoff, J. Schiffner, J. Richter, E. Studerus, G. Casalicchio and Z. M. Jones. <span class="nocase">mlr</span>: Machine learning in <span>R</span>. <em>Journal of Machine Learning Research</em>, 17(170): 1–5, 2016. URL <a href="https://jmlr.org/papers/v17/15-066.html">https://jmlr.org/papers/v17/15-066.html</a>.
</div>
<div id="ref-breiman2001randomforest" class="csl-entry" role="listitem">
L. Breiman. Random forests. <em>Machine Learning</em>, 45(1): 5–32, 2001. DOI <a href="https://doi.org/10.1023/A:1010933404324">10.1023/A:1010933404324</a>.
</div>
<div id="ref-chen2012random" class="csl-entry" role="listitem">
X. Chen and H. Ishwaran. Random forests for genomic data analysis. <em>Genomics</em>, 99(6): 323–329, 2012.
</div>
<div id="ref-chipman2010bart" class="csl-entry" role="listitem">
H. A. Chipman, E. I. George and R. E. McCulloch. BART: Bayesian additive regression trees. <em>The Annals of Applied Statistics</em>, 4(1): 266–298, 2010.
</div>
<div id="ref-vsup" class="csl-entry" role="listitem">
M. Correll, D. Moritz and J. Heer. Value-<span>S</span>uppressing <span>U</span>ncertainty <span>P</span>alettes. In <em>Proceedings of the 2018 CHI conference on human factors in computing systems</em>, pages. 1–11 2018.
</div>
<div id="ref-igraph" class="csl-entry" role="listitem">
G. Csardi and T. Nepusz. The igraph software package for complex network research. <em>InterJournal</em>, Complex Systems: 1695, 2006. URL <a href="https://igraph.org">https://igraph.org</a>.
</div>
<div id="ref-dendserPaper" class="csl-entry" role="listitem">
D. Earle and C. Hurley. Advances in dendrogram seriation for application to visualization. <em>Journal of Computational and Graphical Statistics</em>, 24: 2015. DOI <a href="https://doi.org/10.1080/10618600.2013.874295">10.1080/10618600.2013.874295</a>.
</div>
<div id="ref-felzmann2019transparency" class="csl-entry" role="listitem">
H. Felzmann, E. F. Villaronga, C. Lutz and A. Tamò-Larrieux. Transparency you can trust: Transparency requirements for artificial intelligence between legal norms and contextual concerns. <em>Big Data &amp; Society</em>, 6(1): 2053951719860542, 2019.
</div>
<div id="ref-fisher2019all" class="csl-entry" role="listitem">
A. Fisher, C. Rudin and F. Dominici. All models are wrong, but many are useful: Learning a variable’s importance by studying an entire class of prediction models simultaneously. <em>J. Mach. Learn. Res.</em>, 20(177): 1–81, 2019.
</div>
<div id="ref-netviz1" class="csl-entry" role="listitem">
L. C. Freeman. Visualizing social networks. <em>Journal of social structure</em>, 1(1): 4, 2000.
</div>
<div id="ref-Friedpdp" class="csl-entry" role="listitem">
J. H. Friedman. Greedy function approximation: A gradient boosting machine. <em>The Annals of Statistics</em>, 29: 2000. DOI <a href="https://doi.org/10.1214/aos/1013203451">10.1214/aos/1013203451</a>.
</div>
<div id="ref-Friedmans_H" class="csl-entry" role="listitem">
J. H. Friedman and B. E. Popescu. Predictive learning via rule ensembles. <em>The Annals of Applied Statistics.</em>, 916–954, 2008.
</div>
<div id="ref-iceR" class="csl-entry" role="listitem">
A. Goldstein, A. Kapelner, J. Bleich and E. Pitkin. Peeking inside the black box: Visualizing statistical learning with plots of individual conditional expectation. <em>Journal of Computational and Graphical Statistics</em>, 24(1): 44–65, 2015. DOI <a href="https://doi.org/10.1080/10618600.2014.907095">10.1080/10618600.2014.907095</a>.
</div>
<div id="ref-grange2018random" class="csl-entry" role="listitem">
S. K. Grange, D. C. Carslaw, A. C. Lewis, E. Boleti and C. Hueglin. Random forest meteorological normalisation models for swiss PM 10 trend analysis. <em>Atmospheric Chemistry and Physics</em>, 18(9): 6223–6239, 2018.
</div>
<div id="ref-pdpR" class="csl-entry" role="listitem">
B. M. Greenwell. <span class="nocase">pdp</span>: An <span>R</span> package for constructing partial dependence plots. <em>The R Journal</em>, 9(1): 421–436, 2017. URL <a href="https://journal.r-project.org/archive/2017/RJ-2017-016/index.html">https://journal.r-project.org/archive/2017/RJ-2017-016/index.html</a>.
</div>
<div id="ref-vip" class="csl-entry" role="listitem">
B. M. Greenwell and B. C. Boehmke. Variable importance plots—an introduction to the <span class="nocase">vip</span> package. <em>The R Journal</em>, 12(1): 343–366, 2020. URL <a href="https://doi.org/10.32614/RJ-2020-013">https://doi.org/10.32614/RJ-2020-013</a>.
</div>
<div id="ref-greenwell2018simple" class="csl-entry" role="listitem">
B. M. Greenwell, B. C. Boehmke and A. J. McCarthy. A simple and effective model-based variable importance measure. <em>arXiv preprint arXiv:1805.04755</em>, 2018.
</div>
<div id="ref-boston" class="csl-entry" role="listitem">
D. Harrison Jr and D. L. Rubinfeld. Hedonic housing prices and the demand for clean air. <em>Journal of environmental economics and management</em>, 5(1): 81–102, 1978.
</div>
<div id="ref-elementsStats" class="csl-entry" role="listitem">
T. Hastie, R. Tibshirani and J. Friedman. <em>The elements of statistical learning: Data mining, inference, and prediction.</em> 2nd ed Springer Series in Statistics. Springer-Verlag., 2009.
</div>
<div id="ref-netviz2" class="csl-entry" role="listitem">
I. Herman, G. Melançon and M. S. Marshall. Graph visualization and navigation in information visualization: A survey. <em>IEEE Transactions on visualization and computer graphics</em>, 6(1): 24–43, 2000.
</div>
<div id="ref-hierholzer1873moglichkeit" class="csl-entry" role="listitem">
C. Hierholzer and C. Wiener. <span>Ü</span>ber die m<span>ö</span>glichkeit, einen linienzug ohne wiederholung und ohne unterbrechung zu umfahren. <em>Mathematische Annalen</em>, 6(1): 30–32, 1873.
</div>
<div id="ref-zenplot" class="csl-entry" role="listitem">
M. Hofert and W. Oldford. Zigzag expanded navigation plots in <span>R</span>: The <span>R</span> package <span class="nocase">zenplots</span>. <em>Journal of Statistical Software</em>, 95(4): 1–44, 2020. DOI <a href="https://doi.org/10.18637/jss.v095.i04">10.18637/jss.v095.i04</a>.
</div>
<div id="ref-PairViz1" class="csl-entry" role="listitem">
C. B. Hurley and R. W. Oldford. Eulerian tour algorithms for data visualization and the PairViz package. <em>Computational Statistics</em>, 26: 613–633, 2011. DOI <a href="https://doi.org/10.1007/s00180-011-0229-5">10.1007/s00180-011-0229-5</a>.
</div>
<div id="ref-pviz" class="csl-entry" role="listitem">
C. B. Hurley and R. W. Oldford. <em>PairViz: Visualization using graph traversal.</em> 2022. URL <a href="https://CRAN.R-project.org/package=PairViz">https://CRAN.R-project.org/package=PairViz</a>.
</div>
<div id="ref-condvis2" class="csl-entry" role="listitem">
C. Hurley, M. OConnell and K. Domijan. <em>condvis2: Interactive conditional visualization for supervised and unsupervised models in shiny.</em> 2022. https://cbhurley.github.io/condvis2, https://github.com/cbhurley/condvis2.
</div>
<div id="ref-lime" class="csl-entry" role="listitem">
E. Hvitfeldt, T. L. Pedersen and M. Benesty. <em>Lime: Local interpretable model-agnostic explanations.</em> 2022. URL <a href="https://CRAN.R-project.org/package=lime">https://CRAN.R-project.org/package=lime</a>. R package version 0.5.3.
</div>
<div id="ref-bartvis" class="csl-entry" role="listitem">
A. Inglis, A. Parnell and C. Hurley. Visualizations for bayesian additive regression trees. <em>arXiv preprint arXiv:2208.08966</em>, 2022a.
</div>
<div id="ref-inglis2022visualizing" class="csl-entry" role="listitem">
A. Inglis, A. Parnell and C. B. Hurley. Visualizing variable importance and variable interaction effects in machine learning models. <em>Journal of Computational and Graphical Statistics</em>, 31(3): 766–778, 2022b. DOI <a href="https://doi.org/10.1080/10618600.2021.2007935">10.1080/10618600.2021.2007935</a>.
</div>
<div id="ref-ishwaran2010high" class="csl-entry" role="listitem">
H. Ishwaran, U. B. Kogalur, E. Z. Gorodeski, A. J. Minn and M. S. Lauer. High-dimensional variable selection for survival data. <em>Journal of the American Statistical Association</em>, 105(489): 205–217, 2010.
</div>
<div id="ref-parsnip" class="csl-entry" role="listitem">
M. Kuhn and D. Vaughan. <em>Parsnip: A common API to modeling and analysis functions.</em> 2022. URL <a href="https://CRAN.R-project.org/package=parsnip">https://CRAN.R-project.org/package=parsnip</a>. R package version 1.0.3.
</div>
<div id="ref-mlr3" class="csl-entry" role="listitem">
M. Lang, M. Binder, J. Richter, P. Schratz, F. Pfisterer, S. Coors, Q. Au, G. Casalicchio, L. Kotthoff and B. Bischl. <span class="nocase">mlr3</span>: A modern object-oriented machine learning framework in <span>R</span>. <em>Journal of Open Source Software</em>, 2019. URL <a href="https://joss.theoj.org/papers/10.21105/joss.01903">https://joss.theoj.org/papers/10.21105/joss.01903</a>.
</div>
<div id="ref-EIX" class="csl-entry" role="listitem">
S. Maksymiuk, E. Karbowiak and P. Biecek. <em>EIX: Explain interactions in ’XGBoost’.</em> 2021. URL <a href="https://CRAN.R-project.org/package=EIX">https://CRAN.R-project.org/package=EIX</a>. R package version 1.2.0.
</div>
<div id="ref-flash" class="csl-entry" role="listitem">
M. Mayer. <em><span class="nocase">flashlight</span>: Shed light on black box machine learning models.</em> 2023. URL <a href="https://CRAN.R-project.org/package=flashlight">https://CRAN.R-project.org/package=flashlight</a>. R package version 0.9.0.
</div>
<div id="ref-svmPack" class="csl-entry" role="listitem">
D. Meyer, E. Dimitriadou, K. Hornik, A. Weingessel and F. Leisch. <em>e1071: Misc functions of the department of statistics, probability theory group (formerly: E1071), TU wien.</em> 2021. URL <a href="https://CRAN.R-project.org/package=e1071">https://CRAN.R-project.org/package=e1071</a>. R package version 1.7-8.
</div>
<div id="ref-molnar2022book" class="csl-entry" role="listitem">
C. Molnar. <em>Interpretable machine learning: A guide for making black box models explainable.</em> 2nd ed 2022. URL <a href="https://christophm.github.io/interpretable-ml-book">https://christophm.github.io/interpretable-ml-book</a>.
</div>
<div id="ref-iml" class="csl-entry" role="listitem">
C. Molnar, B. Bischl and G. Casalicchio. <span class="nocase">iml</span>: An <span>R</span> package for interpretable machine learning. <em>JOSS</em>, 3(26): 786, 2018. URL <a href="https://joss.theoj.org/papers/10.21105/joss.00786">https://joss.theoj.org/papers/10.21105/joss.00786</a>.
</div>
<div id="ref-corrgrapher" class="csl-entry" role="listitem">
P. Morgen and P. Biecek. <em>Corrgrapher: Explore correlations between variables in a machine learning model.</em> 2020. URL <a href="https://CRAN.R-project.org/package=corrgrapher">https://CRAN.R-project.org/package=corrgrapher</a>. R package version 1.0.4.
</div>
<div id="ref-murray2009methods" class="csl-entry" role="listitem">
K. Murray and M. M. Conner. Methods to quantify variable importance: Implications for the analysis of noisy ecological data. <em>Ecology</em>, 90(2): 348–355, 2009.
</div>
<div id="ref-RFE" class="csl-entry" role="listitem">
A. Paluszynska, P. Biecek and Y. Jiang. <em>randomForestExplainer: Explaining and visualizing random forests in terms of variable importance.</em> 2020. URL <a href="https://CRAN.R-project.org/package=randomForestExplainer">https://CRAN.R-project.org/package=randomForestExplainer</a>. R package version 0.10.1.
</div>
<div id="ref-varImpPack" class="csl-entry" role="listitem">
P. Probst. <em>varImp: <span>RF</span> <span>V</span>ariable <span>I</span>mportance for <span>A</span>rbitrary <span>M</span>easures.</em> 2020. URL <a href="https://CRAN.R-project.org/package=varImp">https://CRAN.R-project.org/package=varImp</a>. R package version 0.4.
</div>
<div id="ref-netviz3" class="csl-entry" role="listitem">
H. Purchase. Which aesthetic has the greatest effect on human understanding? In <em>Graph drawing</em>, pages. 248–261 1997.
</div>
<div id="ref-limeOG" class="csl-entry" role="listitem">
M. T. Ribeiro, S. Singh and C. Guestrin. " why should i trust you?" explaining the predictions of any classifier. In <em>Proceedings of the 22nd ACM SIGKDD international conference on knowledge discovery and data mining</em>, pages. 1135–1144 2016.
</div>
<div id="ref-kknn" class="csl-entry" role="listitem">
K. Schliep and K. Hechenbichler. <em><span class="nocase">kknn</span>: Weighted k-nearest neighbors.</em> 2016. URL <a href="https://CRAN.R-project.org/package=kknn">https://CRAN.R-project.org/package=kknn</a>. R package version 1.3.1.
</div>
<div id="ref-party" class="csl-entry" role="listitem">
C. Strobl, A.-L. Boulesteix, T. Kneib, T. Augustin and A. Zeileis. Conditional variable importance for random forests. <em>BMC Bioinformatics</em>, 9(307): 2008. DOI <a href="https://doi.org/10.1186/1471-2105-9-307">10.1186/1471-2105-9-307</a>.
</div>
<div id="ref-sugiyama1981methods" class="csl-entry" role="listitem">
K. Sugiyama, S. Tagawa and M. Toda. Methods for visual understanding of hierarchical system structures. <em>IEEE Transactions on Systems, Man, and Cybernetics</em>, 11(2): 109–125, 1981.
</div>
<div id="ref-mass" class="csl-entry" role="listitem">
W. N. Venables and B. D. Ripley. <em>Modern applied statistics with s.</em> Fourth New York: Springer, 2002. URL <a href="https://www.stats.ox.ac.uk/pub/MASS4/">https://www.stats.ox.ac.uk/pub/MASS4/</a>. ISBN 0-387-95457-0.
</div>
<div id="ref-corrplot2021" class="csl-entry" role="listitem">
T. Wei and V. Simko. <em>R package ’<span class="nocase">corrplot</span>’: Visualization of a correlation matrix.</em> 2021. URL <a href="https://github.com/taiyun/corrplot">https://github.com/taiyun/corrplot</a>. (Version 0.92).
</div>
<div id="ref-ggplot2" class="csl-entry" role="listitem">
H. Wickham. <em><span class="nocase">ggplot2</span>: Elegant graphics for data analysis.</em> Springer-Verlag New York, 2016. URL <a href="https://ggplot2.tidyverse.org">https://ggplot2.tidyverse.org</a>.
</div>
<div id="ref-ranger" class="csl-entry" role="listitem">
M. N. Wright and A. Ziegler. <span class="nocase">ranger</span>: A fast implementation of random forests for high dimensional data in <span>C++</span> and <span>R</span>. <em>Journal of Statistical Software</em>, 77(1): 1–17, 2017. DOI <a href="https://doi.org/10.18637/jss.v077.i01">10.18637/jss.v077.i01</a>.
</div>
<div id="ref-colorspaceJSS" class="csl-entry" role="listitem">
A. Zeileis, J. C. Fisher, K. Hornik, R. Ihaka, C. D. McWhite, P. Murrell, R. Stauffer and C. O. Wilke. Colorspace: A toolbox for manipulating and assessing colors and palettes. <em>Journal of Statistical Software, Articles</em>, 96(1): 1–49, 2020. DOI <a href="https://doi.org/10.18637/jss.v096.i01">10.18637/jss.v096.i01</a>.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="references">References</h3>
<div id="references-listing"></div>
<h3 id="reuse">Reuse</h3>
<p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
<h3 id="citation">Citation</h3>
<p>For attribution, please cite this work as</p>
<pre class="citation-appendix short">Inglis, et al., "vivid: An R package for Variable Importance and Variable Interactions Displays for Machine Learning Models", The R Journal, 2023</pre>
<p>BibTeX citation</p>
<pre class="citation-appendix long">@article{RJ-2023-054,
  author = {Inglis, Alan and Parnell, Andrew and Hurley, Catherine},
  title = {vivid: An R package for Variable Importance and Variable Interactions Displays for Machine Learning Models},
  journal = {The R Journal},
  year = {2023},
  note = {https://doi.org/10.32614/RJ-2023-054},
  doi = {10.32614/RJ-2023-054},
  volume = {15},
  issue = {2},
  issn = {2073-4859},
  pages = {344-361}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
