---
title: News from the Bioconductor Project
abstract: The 'News from the Bioconductor Project' article from the 2014-1 issue.
author:
- name: The Bioconductor Team
  affiliation: |
    Program in Computational Biology, Fred Hutchinson Cancer Research
    Center
date: '2014-06-01'
date_received: ~
journal:
  firstpage: '184'
  lastpage: '185'
volume: 6
issue: 1
slug: RJ-2014-1-bioconductor
citation_url: https://rjournal.github.io/
packages:
  cran: []
  bioc:
  - deepSNV
  - Rariant
  - SomaticSignatures
  - VariantFiltering
  - CRIPSRseek
  - CCREPE
  - MIMOSA
  - MLSeq
  - NetPathMiner
  - Rcpi
  - alsace
  - GenomicRanges
  - GenomicAlignments
  - GenomicFeatures
  - GenomeInfoDb
  - AnnotationHub
  - BiocParallel
  - BatchJobs
  - BiocCheck
preview: preview.png
bibliography: bioc.bib
CTV: ~
output:
  rjtools::rjournal_web_article:
    self_contained: yes
    toc: no
    legacy_pdf: yes
    web_only: yes
    mathjax: https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js
    md_extension: -tex_math_single_backslash

---

::: article
The Bioconductor project provides tools for the analysis and
comprehension of high-throughput genomic data. The 824 software packages
available in Bioconductor can be viewed at
<http://bioconductor.org/packages/release/>. Navigate packages using
'biocViews' terms and title search. Each package has an html page with a
description, links to vignettes, reference manuals, and usage
statistics. Start using Bioconductor and R version 3.1.0 with

``` r
  source("http://bioconductor.org/biocLite.R")
  biocLite()
```

Install additional packages and dependencies, e.g.,
[*deepSNV*](https://www.bioconductor.org/packages/release/bioc/html/deepSNV.html),
with

``` r
  source("http://bioconductor.org/biocLite.R")
  biocLite("deepSNV")
```

Upgrade installed packages with

``` r
  source("http://bioconductor.org/biocLite.R")
  biocLite()
```

# Bioconductor 2.14 Release Highlights

Bioconductor 2.14 was released on 14 April 2014. It is compatible with R
3.1.0 and consists of 824 software packages, 200 experiment data
packages, and more than 860 current annotation packages. The release
includes 77 new software packages and many updates and improvements to
existing packages. The [release
announcement](http://bioconductor.org/news/bioc_2_14_release/) includes
descriptions of new packages and updated NEWS files provided by current
package maintainers.

New packages continue to represent a wide variety of research areas.
[*Rariant*](https://www.bioconductor.org/packages/release/bioc/html/Rariant.html)
identifies single nucleotide variants (SNVs) based on the difference of
binomially distributed mismatch rates between matched samples.
[*SomaticSignatures*](https://www.bioconductor.org/packages/release/bioc/html/SomaticSignatures.html)
identifies the mutational signatures of SNVs. Filtering of SNVs based on
inheritance models, amino acid change consequence, and minor allele
frequency is offered through
[*VariantFiltering*](https://www.bioconductor.org/packages/release/bioc/html/VariantFiltering.html).
The *monocle* package, currently in the 'devel' branch, preforms
differential expression and time series analysis for single-cell
expression experiments. CRISPR/Cas is a compelling new molecular biology
technique used for gene editing;
[*CRIPSRseek*](https://www.bioconductor.org/packages/release/bioc/html/CRIPSRseek.html)
helps find potential guide RNAs for input target sequences.
[*CCREPE*](https://www.bioconductor.org/packages/release/bioc/html/CCREPE.html)
assesses the significance of similarity measures in 'compositional' data
sets, as found in microbial abundance studies.
[*MIMOSA*](https://www.bioconductor.org/packages/release/bioc/html/MIMOSA.html)
models count data using Dirichlet-multinomial and beta-binomial mixtures
with applications to single-cell assays. Machine learning methods such
as SVM, Random Forest and CART are applied to RNASeq data in
[*MLSeq*](https://www.bioconductor.org/packages/release/bioc/html/MLSeq.html);
clustering and classification methods are used to summarize active paths
in genome-scale metabolic and reaction networks in
[*NetPathMiner*](https://www.bioconductor.org/packages/release/bioc/html/NetPathMiner.html).
Bioconductor hosts a number of packages relevant to chemical compound
discovery. The latest additions integrate bioinformatics and
chemoinformatics into a molecular informatics platform in
[*Rcpi*](https://www.bioconductor.org/packages/release/bioc/html/Rcpi.html),
and performs alternating least squares analysis on chemical data in
[*alsace*](https://www.bioconductor.org/packages/release/bioc/html/alsace.html).

In addition to these contributed packages, 'Ranges' infrastructure
packages such as
[*GenomicRanges*](https://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html),
[*GenomicAlignments*](https://www.bioconductor.org/packages/release/bioc/html/GenomicAlignments.html),
and
[*GenomicFeatures*](https://www.bioconductor.org/packages/release/bioc/html/GenomicFeatures.html)
provide an extensive, mature and extensible framework for interacting
with high throughput sequence data. One recent addition is the
[*GenomeInfoDb*](https://www.bioconductor.org/packages/release/bioc/html/GenomeInfoDb.html)
package, which contains functions that allow translation between
different chromosome sequence naming conventions (e.g., UCSC versus
NCBI). Many packages rely on the Ranges infrastructure for
interoperable, re-usable analysis; [@10.1371/journal.pcbi.1003118]
provides an introduction.

Our large collection of microarray, transcriptome and organism-specific
*annotation packages* have been updated to include current information.
Most of these packages now provide access to the 'select' interface
(keys, columns, keytypes and select) which enable programmatic access to
the databases they contain. The
[*AnnotationHub*](https://www.bioconductor.org/packages/release/bioc/html/AnnotationHub.html),
with 10,780 entries, complements our traditional offerings with diverse
whole genome annotations from Ensembl, ENCODE, dbSNP, UCSC, and
elsewhere.

# Other activities

The Bioconductor [Git-SVN
Bridge](http://bioconductor.org/developers/how-to/git-svn/) allows
developers to synchronize a GitHub repository with the canonical
Bioconductor SVN package repository. Commits made in SVN are propagated
to GitHub and vice versa. This was driven by developer requests for
access to social coding features, such as issue tracking and pull
requests. The service has been well received, with 73 bridges
established as of June 2014.

The Bioconductor Amazon Machine Instance is now compatible with the
StarCluster toolkit. This enhancement makes it straightforward to
configure a cluster with nodes that communicate via MPI, SSH or Sun Grid
Engine, and to control jobs via the
[*BiocParallel*](https://www.bioconductor.org/packages/release/bioc/html/BiocParallel.html)
and
[*BatchJobs*](https://www.bioconductor.org/packages/release/bioc/html/BatchJobs.html)
packages. Details are available at the AMI page
<http://www.bioconductor.org/help/bioconductor-cloud-ami/#using_cluster>.

New Bioconductor package contributors are encouraged to consult the
[Package
Guidelines](http://bioconductor.org/developers/package-guidelines/) and
[Package
Submission](http://bioconductor.org/developers/package-submission/)
sections of the Bioconductor web site, and use the new
[*BiocCheck*](https://www.bioconductor.org/packages/release/bioc/html/BiocCheck.html)
package, in addition to `R CMD check`, for guidance on conforming to
Bioconductor package standards.

The Bioconductor web site advertises [training and community
events](http://bioconductor.org/help/events/); [mailing
lists](http://bioconductor.org/help/mailing-list/) connect users with
each other, to domain experts, and to maintainers eager to ensure that
their packages satisfy the needs of leading edge approaches. Keep
abreast of packages added to the 'devel' branch and other activities by
following \@Bioconductor on Twitter.
:::
