---
title: News from the Bioconductor Project
abstract: ~
author:
- name: Bioconductor Team
  affiliation: |
    Program in Computational Biology, Fred Hutchinson Cancer Research
    Center
date: '2011-12-01'
date_received: ~
journal:
  firstpage: 86
  lastpage: 86
volume: 3
issue: 2
slug: RJ-2011-2-bioconductor
citation_url: https://rjournal.github.io/
packages:
  cran: []
  bioc:
  - DEXSeq
  - DiffBind
  - EDASeq
  - REDseq
  - ReQON
  - Repitools
  - TSSi
  - VariantAnnotation
  - cummeRbund
  - fastseg
  - htSeqTools
  - nucleR
  - r3Cseq
  - tweeDEseq
  - AGDEX
  - CNAnorm
  - CellNOptR
  - Cormotif
  - DECIPHER
  - GWASTools
  - MmPalateMiRNA
  - cn.mops
  - cqn
  - exomeCopy
  - minfi
  - rqubic
  - stepwiseCM
  - DTA
  - RCASPAR
  - dks
  - randPack
  - sva
  - trigger
  - DOSE
  - GOFunction
  - GRENITS
  - GeneExpressionSignature
  - IdMappingRetrieval
  - PAN
  - PREDA
  - RTopper
  - RamiGO
  - RedeR
  - graphite
  - inSilicoDb
  - predictionet
  - NormqPCR
  - ReadqPCR
  - flowType
  - flowWorkspace
  - iontree
  - isobar
  - mzR
  - ncdfFlow
  - biovizBase
  - ggbio
  - qtbase
  - qtpaint
preview: preview.png
bibliography: ~
CTV: ~
output:
  rjtools::rjournal_web_article:
    self_contained: yes
    toc: no
    legacy_pdf: yes
    web_only: yes
    mathjax: https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js
    md_extension: -tex_math_single_backslash

---

::: article
We are pleased to announce Bioconductor 2.9, released on 14 October
2011. Bioconductor 2.9 is compatible with R 2.14.0, and consists of 516
software packages and more than 500 up-to-date annotation packages.
There are 60 new software packages, and enhancements to many others.
Explore Bioconductor at <http://bioconductor.org>. Install packages with

``` r
> source("http://bioconductor.org/biocLite.R")
> # install standard packages...
> biocLite()
> # ...or only VariantAnnotation
> biocLite("VariantAnnotation")
```

A Bioconductor Amazon Machine Instance is available; see
<http://bioconductor.org/help/bioconductor-cloud-ami>.

# New and revised packages

This release includes new packages for diverse areas of high-throughput
analysis. Highlights include:

::: description
High-throughput sequencing:
[*DEXSeq*](https://www.bioconductor.org/packages/release/bioc/html/DEXSeq.html),
[*DiffBind*](https://www.bioconductor.org/packages/release/bioc/html/DiffBind.html),
[*EDASeq*](https://www.bioconductor.org/packages/release/bioc/html/EDASeq.html),
[*REDseq*](https://www.bioconductor.org/packages/release/bioc/html/REDseq.html),
[*ReQON*](https://www.bioconductor.org/packages/release/bioc/html/ReQON.html),
[*Repitools*](https://www.bioconductor.org/packages/release/bioc/html/Repitools.html),
[*TSSi*](https://www.bioconductor.org/packages/release/bioc/html/TSSi.html),
[*VariantAnnotation*](https://www.bioconductor.org/packages/release/bioc/html/VariantAnnotation.html),
[*cummeRbund*](https://www.bioconductor.org/packages/release/bioc/html/cummeRbund.html),
[*fastseg*](https://www.bioconductor.org/packages/release/bioc/html/fastseg.html),
[*htSeqTools*](https://www.bioconductor.org/packages/release/bioc/html/htSeqTools.html),
[*nucleR*](https://www.bioconductor.org/packages/release/bioc/html/nucleR.html),
[*r3Cseq*](https://www.bioconductor.org/packages/release/bioc/html/r3Cseq.html),
[*tweeDEseq*](https://www.bioconductor.org/packages/release/bioc/html/tweeDEseq.html).

Microarrays:
[*AGDEX*](https://www.bioconductor.org/packages/release/bioc/html/AGDEX.html),
[*CNAnorm*](https://www.bioconductor.org/packages/release/bioc/html/CNAnorm.html),
[*CellNOptR*](https://www.bioconductor.org/packages/release/bioc/html/CellNOptR.html),
[*Cormotif*](https://www.bioconductor.org/packages/release/bioc/html/Cormotif.html),
[*DECIPHER*](https://www.bioconductor.org/packages/release/bioc/html/DECIPHER.html),
[*GWASTools*](https://www.bioconductor.org/packages/release/bioc/html/GWASTools.html),
[*MmPalateMiRNA*](https://www.bioconductor.org/packages/release/bioc/html/MmPalateMiRNA.html),
[*cn.mops*](https://www.bioconductor.org/packages/release/bioc/html/cn.mops.html),
[*cqn*](https://www.bioconductor.org/packages/release/bioc/html/cqn.html),
[*exomeCopy*](https://www.bioconductor.org/packages/release/bioc/html/exomeCopy.html),
[*minfi*](https://www.bioconductor.org/packages/release/bioc/html/minfi.html),
[*rqubic*](https://www.bioconductor.org/packages/release/bioc/html/rqubic.html),
[*stepwiseCM*](https://www.bioconductor.org/packages/release/bioc/html/stepwiseCM.html).

Advanced statistical implementations:
[*DTA*](https://www.bioconductor.org/packages/release/bioc/html/DTA.html),
[*RCASPAR*](https://www.bioconductor.org/packages/release/bioc/html/RCASPAR.html),
[*dks*](https://www.bioconductor.org/packages/release/bioc/html/dks.html),
[*randPack*](https://www.bioconductor.org/packages/release/bioc/html/randPack.html),
[*sva*](https://www.bioconductor.org/packages/release/bioc/html/sva.html),
[*trigger*](https://www.bioconductor.org/packages/release/bioc/html/trigger.html).

Networks and pathways:
[*DOSE*](https://www.bioconductor.org/packages/release/bioc/html/DOSE.html),
[*GOFunction*](https://www.bioconductor.org/packages/release/bioc/html/GOFunction.html),
[*GRENITS*](https://www.bioconductor.org/packages/release/bioc/html/GRENITS.html),
[*GeneExpressionSignature*](https://www.bioconductor.org/packages/release/bioc/html/GeneExpressionSignature.html),
[*IdMappingRetrieval*](https://www.bioconductor.org/packages/release/bioc/html/IdMappingRetrieval.html),
[*PAN*](https://www.bioconductor.org/packages/release/bioc/html/PAN.html),
[*PREDA*](https://www.bioconductor.org/packages/release/bioc/html/PREDA.html),
[*RTopper*](https://www.bioconductor.org/packages/release/bioc/html/RTopper.html),
[*RamiGO*](https://www.bioconductor.org/packages/release/bioc/html/RamiGO.html),
[*RedeR*](https://www.bioconductor.org/packages/release/bioc/html/RedeR.html),
[*graphite*](https://www.bioconductor.org/packages/release/bioc/html/graphite.html),
[*inSilicoDb*](https://www.bioconductor.org/packages/release/bioc/html/inSilicoDb.html),
[*predictionet*](https://www.bioconductor.org/packages/release/bioc/html/predictionet.html).

Mass spectrometry, qPCR, and flow cytometry:
[*NormqPCR*](https://www.bioconductor.org/packages/release/bioc/html/NormqPCR.html),
[*ReadqPCR*](https://www.bioconductor.org/packages/release/bioc/html/ReadqPCR.html),
[*flowType*](https://www.bioconductor.org/packages/release/bioc/html/flowType.html),
[*flowWorkspace*](https://www.bioconductor.org/packages/release/bioc/html/flowWorkspace.html),
[*iontree*](https://www.bioconductor.org/packages/release/bioc/html/iontree.html),
[*isobar*](https://www.bioconductor.org/packages/release/bioc/html/isobar.html),
[*mzR*](https://www.bioconductor.org/packages/release/bioc/html/mzR.html),
[*ncdfFlow*](https://www.bioconductor.org/packages/release/bioc/html/ncdfFlow.html).

Visualization:
[*biovizBase*](https://www.bioconductor.org/packages/release/bioc/html/biovizBase.html),
[*ggbio*](https://www.bioconductor.org/packages/release/bioc/html/ggbio.html),
[*qtbase*](https://www.bioconductor.org/packages/release/bioc/html/qtbase.html),
[*qtpaint*](https://www.bioconductor.org/packages/release/bioc/html/qtpaint.html).
:::

Our large collection of microarray- and organism-specific annotation
packages have, as usual, been updated to include current information.
This release introduces a `select` interface to annotation packages,
simplifying queries for multiple types of annotation within and between
data sources. The new
[*VariantAnnotation*](https://www.bioconductor.org/packages/release/bioc/html/VariantAnnotation.html)
package processes VCF files, including interfaces to SIFT and PolyPhen
data bases for assessing consequences of sequence changes.

Further information on new and existing packages can be found on the
Bioconductor web site; 'BiocViews' identify coherent groups of packages,
with links to package descriptions, vignettes, reference manuals, and
use statistics. Our web site has been enhanced with easier access to
package NEWS and other support files.

# Other activities

A meeting in December highlighted contributions from our European
developer community; we look forward to Seattle's Annual Meeting on
23-25 July 2012. The active Bioconductor mailing lists
(<http://bioconductor.org/help/mailing-list/>) connect users with each
other, to domain experts, and to maintainers eager to ensure that their
packages satisfy the needs of leading edge approaches. Bioconductor
package maintainers and the Bioconductor team invest considerable effort
in producing high-quality software. The Bioconductor team continues to
ensure quality software through technical and scientific reviews of new
packages, and daily builds of released packages on Linux, Windows, and
MacOS platforms.

# Looking forward

Our contributors provide a tremendous diversity of high-quality
packages. These are enhancing areas of existing strength in statistical
analysis of microarrays, while driving project growth in high-throughput
sequence analysis and emerging domains like mass spectrometry and qPCR.
These directions provide an important incentive for infrastructure,
contributed by the Bioconductor core team and others, to support
processing and visualization of very large data sets.

Important directions for the Bioconductor core team include
representation and manipulation of complex sequence alignments and
variants, convenient and integrated approaches to gene and genome
annotation, continued efforts to ease cloud-based use of Bioconductor,
and approaches to ease development and support of increasingly
inter-related packages.
:::
