<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
<title>Statistical Models for Repeated Categorical Ratings: The R Package rater</title>

<meta property="description" itemprop="description" content="A common problem in many disciplines is the need to assign a set of items into categories or classes with known labels.  This is often done by one or more expert raters, or sometimes by an automated process.  If these assignments or &#39;ratings&#39; are difficult to make accurately, a common tactic is to repeat them by different raters, or even by the same rater multiple times on different occasions. We present an R package rater, available on CRAN, that implements Bayesian versions of several statistical models for analysis of repeated categorical rating data. Inference is possible for the true underlying (latent) class of each item, as well as the accuracy of each rater. The models are extensions of, and include, the Dawid--Skene model, and we implemented them using the Stan probabilistic programming language.  We illustrate the use of rater through a few examples.  We also discuss in detail the techniques of marginalisation and conditioning, which are necessary for these models but also apply more generally to other models implemented in Stan."/>

<link rel="canonical" href="https://doi.org/10.32614/RJ-2023-064/"/>
<link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>
<link rel="icon" type="image/vnd.microsoft.icon" href="../../resources/favicon.ico"/>

<!--  https://schema.org/Article -->
<meta property="article:published" itemprop="datePublished" content="2023-12-18"/>
<meta property="article:created" itemprop="dateCreated" content="2023-12-18"/>
<meta name="article:author" content="Jeffrey M. Pullin"/>
<meta name="article:author" content="Lyle C. Gurrin"/>
<meta name="article:author" content="Damjan Vukcevic"/>

<!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
<meta property="og:title" content="Statistical Models for Repeated Categorical Ratings: The R Package rater"/>
<meta property="og:type" content="article"/>
<meta property="og:description" content="A common problem in many disciplines is the need to assign a set of items into categories or classes with known labels.  This is often done by one or more expert raters, or sometimes by an automated process.  If these assignments or &#39;ratings&#39; are difficult to make accurately, a common tactic is to repeat them by different raters, or even by the same rater multiple times on different occasions. We present an R package rater, available on CRAN, that implements Bayesian versions of several statistical models for analysis of repeated categorical rating data. Inference is possible for the true underlying (latent) class of each item, as well as the accuracy of each rater. The models are extensions of, and include, the Dawid--Skene model, and we implemented them using the Stan probabilistic programming language.  We illustrate the use of rater through a few examples.  We also discuss in detail the techniques of marginalisation and conditioning, which are necessary for these models but also apply more generally to other models implemented in Stan."/>
<meta property="og:url" content="https://doi.org/10.32614/RJ-2023-064/"/>
<meta property="og:image" content="https://journal.r-project.org/articles/RJ-2023-064/tikz/figrelationships.png"/>
<meta property="og:image:width" content="975"/>
<meta property="og:image:height" content="355"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="The R Journal"/>

<!--  https://dev.twitter.com/cards/types/summary -->
<meta property="twitter:card" content="summary_large_image"/>
<meta property="twitter:title" content="Statistical Models for Repeated Categorical Ratings: The R Package rater"/>
<meta property="twitter:description" content="A common problem in many disciplines is the need to assign a set of items into categories or classes with known labels.  This is often done by one or more expert raters, or sometimes by an automated process.  If these assignments or &#39;ratings&#39; are difficult to make accurately, a common tactic is to repeat them by different raters, or even by the same rater multiple times on different occasions. We present an R package rater, available on CRAN, that implements Bayesian versions of several statistical models for analysis of repeated categorical rating data. Inference is possible for the true underlying (latent) class of each item, as well as the accuracy of each rater. The models are extensions of, and include, the Dawid--Skene model, and we implemented them using the Stan probabilistic programming language.  We illustrate the use of rater through a few examples.  We also discuss in detail the techniques of marginalisation and conditioning, which are necessary for these models but also apply more generally to other models implemented in Stan."/>
<meta property="twitter:url" content="https://doi.org/10.32614/RJ-2023-064/"/>
<meta property="twitter:image" content="https://journal.r-project.org/articles/RJ-2023-064/tikz/figrelationships.png"/>
<meta property="twitter:image:width" content="975"/>
<meta property="twitter:image:height" content="355"/>

<!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
<meta name="citation_title" content="Statistical Models for Repeated Categorical Ratings: The R Package rater"/>
<meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2023-064"/>
<meta name="citation_pdf_url" content="RJ-2023-064.pdf"/>
<meta name="citation_volume" content="15"/>
<meta name="citation_issue" content="3"/>
<meta name="citation_doi" content="10.32614/RJ-2023-064"/>
<meta name="citation_journal_title" content="The R Journal"/>
<meta name="citation_issn" content="2073-4859"/>
<meta name="citation_firstpage" content="93"/>
<meta name="citation_lastpage" content="118"/>
<meta name="citation_fulltext_world_readable" content=""/>
<meta name="citation_online_date" content="2023/12/18"/>
<meta name="citation_publication_date" content="2023/12/18"/>
<meta name="citation_author" content="Jeffrey M. Pullin"/>
<meta name="citation_author_institution" content="University of Melbourne"/>
<meta name="citation_author" content="Lyle C. Gurrin"/>
<meta name="citation_author_institution" content="University of Melbourne"/>
<meta name="citation_author" content="Damjan Vukcevic"/>
<meta name="citation_author_institution" content="Monash University"/>
<!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Rank-Normalization, Folding, and Localization: An Improved Rˆ for Assessing Convergence of MCMC (with Discussion);citation_volume=16;citation_doi=10.1214/20-BA1221;citation_issn=1936-0975, 1931-6690;citation_author=Aki Vehtari;citation_author=Andrew Gelman;citation_author=Daniel Simpson;citation_author=Bob Carpenter;citation_author=Paul-Christian Bürkner"/>
  <meta name="citation_reference" content="citation_title=Inferring ground truth from subjective labelling of Venus images;citation_publisher=MIT Press;citation_volume=7;citation_author=Padhraic Smyth;citation_author=Usama Fayyad;citation_author=Michael Burl;citation_author=Pietro Perona;citation_author=Pierre Baldi"/>
  <meta name="citation_reference" content="citation_title=Estimating prevalence, false-positive rate, and false-negative rate with use of repeated testing when true responses are unknown;citation_publisher=Elsevier;citation_volume=81;citation_doi=10.1086/521582;citation_author=Johanna Jakobsdottir;citation_author=Daniel E Weeks"/>
  <meta name="citation_reference" content="citation_title=The Benefits of a Model of Annotation;citation_publisher=MIT Press;citation_volume=2;citation_doi=10.1162/tacl_a_00185;citation_issn=2307-387X;citation_author=Rebecca J. Passonneau;citation_author=Bob Carpenter"/>
  <meta name="citation_reference" content="citation_title=Quality management on Amazon Mechanical Turk;citation_author=Panagiotis G Ipeirotis;citation_author=Foster Provost;citation_author=Jing Wang"/>
  <meta name="citation_reference" content="citation_title=R: A language and environment for statistical computing;citation_publisher=R Foundation for Statistical Computing;citation_author=R: A language and environment for statistical computing"/>
  <meta name="citation_reference" content="citation_title=Maximum Likelihood Estimation of Observer Error-Rates Using the EM Algorithm;citation_volume=28;citation_doi=10.2307/2346806;citation_issn=00359254;citation_author=A. P. Dawid;citation_author=A. M. Skene"/>
  <meta name="citation_reference" content="citation_title=Comparing Bayesian Models of Annotation;citation_volume=6;citation_doi=10.1162/tacl_a_00040;citation_issn=2307-387X;citation_author=Silviu Paun;citation_author=Bob Carpenter;citation_author=Jon Chamberlain;citation_author=Dirk Hovy;citation_author=Udo Kruschwitz;citation_author=Massimo Poesio"/>
  <meta name="citation_reference" content="citation_title=;citation_author=Pietro Berkes;citation_author=Bob Carpenter;citation_author=Andrey Rzhetsky;citation_author=James Evans"/>
  <meta name="citation_reference" content="citation_title=How to get the most out of your curation effort;citation_publisher=Public Library of Science;citation_volume=5;citation_doi=10.1371/journal.pcbi.1000391;citation_author=Hagit AND Wilbur Rzhetsky"/>
  <meta name="citation_reference" content="citation_title=Exploratory latent structure analysis using both identifiable and unidentifiable models;citation_publisher=Oxford University Press;citation_volume=61;citation_doi=10.1093/biomet/61.2.215;citation_author=Leo A Goodman"/>
  <meta name="citation_reference" content="citation_title=: An R Package for Polytomous Variable Latent Class Analysis;citation_volume=42;citation_doi=10.18637/jss.v042.i10;citation_issn=1548-7660;citation_author=Drew A. Linzer;citation_author=Jeffrey B. Lewis"/>
  <meta name="citation_reference" content="citation_title=: An R Package for Bayesian Latent Class Analysis;citation_volume=61;citation_doi=10.18637/jss.v061.i13;citation_issn=1548-7660;citation_author=Arthur White;citation_author=Thomas Brendan Murphy"/>
  <meta name="citation_reference" content="citation_title=randomLCA: An R Package for Latent Class with Random Effects Analysis;citation_volume=81;citation_doi=10.18637/jss.v081.i13;citation_issn=1548-7660;citation_author=Ken J. Beath"/>
  <meta name="citation_reference" content="citation_title=Stan User’s Guide;citation_author=Stan User’s Guide"/>
  <meta name="citation_reference" content="citation_title=Error Rate Bounds and Iterative Weighted Majority Voting for Crowdsourcing;citation_author=Hongwei Li;citation_author=Bin Yu"/>
  <meta name="citation_reference" content="citation_title=Estimating the Error Rates of Diagnostic Tests;citation_volume=36;citation_doi=10.2307/2530508;citation_issn=0006341X;citation_author=S. L. Hui;citation_author=S. D. Walter"/>
  <meta name="citation_reference" content="citation_title=Different latent class models were used and evaluated for assessing the accuracy of campylobacter diagnostic tests: Overcoming imperfect reference standards?;citation_volume=146;citation_doi=10.1017/S0950268818001723;citation_issn=0950-2688, 1469-4409;citation_author=J. Asselineau;citation_author=A. Paye;citation_author=E. Bessède;citation_author=P. Perez;citation_author=C. Proust-Lima"/>
  <meta name="citation_reference" content="citation_title=Stan: A Probabilistic Programming Language;citation_volume=76;citation_doi=10.18637/jss.v076.i01;citation_issn=1548-7660;citation_author=Bob Carpenter;citation_author=Andrew Gelman;citation_author=Matthew D. Hoffman;citation_author=Daniel Lee;citation_author=Ben Goodrich;citation_author=Michael Betancourt;citation_author=Marcus Brubaker;citation_author=Jiqiang Guo;citation_author=Peter Li;citation_author=Allen Riddell"/>
  <meta name="citation_reference" content="citation_title=The No-U-Turn Sampler: Adaptively setting path lengths in Hamiltonian Monte Carlo;citation_volume=15;citation_author=Matthew D. Hoffman;citation_author=Andrew Gelman"/>
  <meta name="citation_reference" content="citation_title=Monte Carlo theory, methods and examples;citation_author=Art B. Owen"/>
  <meta name="citation_reference" content="citation_title=Rao-Blackwellization in the MCMC era;citation_author=Christian P. Robert;citation_author=Gareth O. Roberts"/>
  <meta name="citation_reference" content="citation_title=Conditioning in Markov chain Monte Carlo;citation_publisher=Taylor &amp; Francis;citation_volume=4;citation_doi=10.1080/10618600.1995.10474672;citation_author=Charles J. Geyer"/>
  <meta name="citation_reference" content="citation_title=Conditional expectation and unbiased sequential estimation;citation_publisher=The Institute of Mathematical Statistics;citation_volume=18;citation_doi=10.1214/aoms/1177730497;citation_author=David Blackwell"/>
  <meta name="citation_reference" content="citation_title=A need for speed in bayesian population models: A practical guide to marginalizing and recovering discrete latent states;citation_volume=30;citation_doi=10.1002/eap.2112;citation_author=Charles B. Yackulic;citation_author=Michael Dodrill;citation_author=Maria Dzul;citation_author=Jamie S. Sanderlin;citation_author=Janice A. Reid"/>
  <meta name="citation_reference" content="citation_title=Investigating the efficiency of marginalising over discrete parameters in Bayesian computations;citation_doi=10.48550/ARXIV.2204.06313;citation_author=Wen Zhang;citation_author=Jeffrey Pullin;citation_author=Lyle Gurrin;citation_author=Damjan Vukcevic"/>
  <meta name="citation_reference" content="citation_title=Using latent class models to characterize and assess relative error in discrete measurements;citation_publisher=[Wiley, International Biometric Society];citation_volume=45;citation_doi=10.2307/2531499;citation_issn=0006341X, 15410420;citation_author=Mark A. Espeland;citation_author=Stanley L. Handelman"/>
  <meta name="citation_reference" content="citation_title=Clinical radiographic evaluation of sealed carious and sound tooth surfaces;citation_volume=113;citation_doi=10.14219/jada.archive.1986.0269;citation_issn=0002-8177;citation_author=Stanley L. Handelman;citation_author=D. H. Leverett;citation_author=Mark A. Espeland;citation_author=Jennifer A. Curzon"/>
  <meta name="citation_reference" content="citation_title=Practical Bayesian model evaluation using leave-one-out cross-validation and WAIC;citation_publisher=Springer;citation_volume=27;citation_doi=10.1007/s11222-016-9696-4;citation_issn=1573-1375;citation_author=Aki Vehtari;citation_author=Andrew Gelman;citation_author=Jonah Gabry"/>
  <meta name="citation_reference" content="citation_title=A Widely Applicable Bayesian Information Criterion;citation_volume=14;citation_issn=1533-7928;citation_author=Sumio Watanabe"/>
  <meta name="citation_reference" content="citation_title=ggplot2: Elegant graphics for data analysis;citation_publisher=Springer-Verlag New York;citation_author=Hadley Wickham"/>
  <meta name="citation_reference" content="citation_title=knitr: A comprehensive tool for reproducible research in R;citation_publisher=Chapman; Hall/CRC;citation_author=Yihui Xie"/>
  <meta name="citation_reference" content="citation_title=Dynamic documents with R and knitr;citation_publisher=Chapman; Hall/CRC;citation_author=Yihui Xie"/>
  <meta name="citation_reference" content="citation_title=knitr: A general-purpose package for dynamic report generation in r;citation_author=Yihui Xie"/>
  <meta name="citation_reference" content="citation_title=rater: Statistical models of repeated categorical rating data;citation_author=Jeffrey Pullin;citation_author=Damjan Vukcevic"/>
  <meta name="citation_reference" content="citation_title=rjtools: Preparing, checking, and submitting articles to the “R Journal”;citation_author=Mitchell O’Hara-Wild;citation_author=Stephanie Kobakian;citation_author=H. Sherry Zhang;citation_author=Di Cook;citation_author=Simon Urbanek;citation_author=Christophe Dervieux"/>
  <meta name="citation_reference" content="citation_title=R markdown: The definitive guide;citation_publisher=Chapman; Hall/CRC;citation_author=Yihui Xie;citation_author=J. J. Allaire;citation_author=Garrett Grolemund"/>
  <meta name="citation_reference" content="citation_title=R markdown cookbook;citation_publisher=Chapman; Hall/CRC;citation_author=Yihui Xie;citation_author=Christophe Dervieux;citation_author=Emily Riederer"/>
  <meta name="citation_reference" content="citation_title=rmarkdown: Dynamic documents for r;citation_author=JJ Allaire;citation_author=Yihui Xie;citation_author=Christophe Dervieux;citation_author=Jonathan McPherson;citation_author=Javier Luraschi;citation_author=Kevin Ushey;citation_author=Aron Atkins;citation_author=Hadley Wickham;citation_author=Joe Cheng;citation_author=Winston Chang;citation_author=Richard Iannone"/>
  <meta name="citation_reference" content="citation_title=: The R interface to Stan;citation_author=: The R interface to Stan"/>
  <meta name="citation_reference" content="citation_title=Loo: Efficient leave-one-out cross-validation and WAIC for Bayesian models;citation_author=Aki Vehtari;citation_author=Jonah Gabry;citation_author=Mans Magnusson;citation_author=Yuling Yao;citation_author=Paul-Christian Bürkner;citation_author=Topi Paananen;citation_author=Andrew Gelman"/>
  <meta name="citation_reference" content="citation_title=CODA: Convergence diagnosis and output analysis for MCMC;citation_volume=6;citation_author=Martyn Plummer;citation_author=Nicky Best;citation_author=Kate Cowles;citation_author=Karen Vines"/>
  <!--radix_placeholder_rmarkdown_metadata-->

<script type="text/json" id="radix-rmarkdown-metadata">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","draft","author","type","output","bibliography","date","date_received","volume","issue","subid","slug","journal","pdf_url","citation_url","doi","creative_commons","packages","CTV","csl","canonical_url"]}},"value":[{"type":"character","attributes":{},"value":["Statistical Models for Repeated Categorical Ratings: The R Package rater"]},{"type":"character","attributes":{},"value":["A common problem in many disciplines is the need to assign a set of items into categories or classes with known labels.  This is often done by one or more expert raters, or sometimes by an automated process.  If these assignments or 'ratings' are difficult to make accurately, a common tactic is to repeat them by different raters, or even by the same rater multiple times on different occasions. We present an R package rater, available on CRAN, that implements Bayesian versions of several statistical models for analysis of repeated categorical rating data. Inference is possible for the true underlying (latent) class of each item, as well as the accuracy of each rater. The models are extensions of, and include, the Dawid--Skene model, and we implemented them using the Stan probabilistic programming language.  We illustrate the use of rater through a few examples.  We also discuss in detail the techniques of marginalisation and conditioning, which are necessary for these models but also apply more generally to other models implemented in Stan."]},{"type":"logical","attributes":{},"value":[false]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","orcid_id","address"]}},"value":[{"type":"character","attributes":{},"value":["Jeffrey M. Pullin"]},{"type":"character","attributes":{},"value":["University of Melbourne"]},{"type":"character","attributes":{},"value":["0000-0003-3651-5471"]},{"type":"character","attributes":{},"value":["School of Mathematics and Statistics","Melbourne, Australia"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","orcid_id","address"]}},"value":[{"type":"character","attributes":{},"value":["Lyle C. Gurrin"]},{"type":"character","attributes":{},"value":["University of Melbourne"]},{"type":"character","attributes":{},"value":["0000-0001-7052-1969"]},{"type":"character","attributes":{},"value":["School of Population and Global Health","Melbourne, Australia"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","email","orcid_id","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Damjan Vukcevic"]},{"type":"character","attributes":{},"value":["damjan.vukcevic@monash.edu"]},{"type":"character","attributes":{},"value":["0000-0001-7780-9586"]},{"type":"character","attributes":{},"value":["Monash University"]},{"type":"character","attributes":{},"value":["Department of Econometrics and Business Statistics","Melbourne, Australia"]}]}]},{"type":"character","attributes":{},"value":["package"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["toc"]}},"value":[{"type":"logical","attributes":{},"value":[true]}]}]},{"type":"character","attributes":{},"value":["rater.bib"]},{"type":"character","attributes":{},"value":["2023-12-18"]},{"type":"character","attributes":{},"value":["2022-07-25"]},{"type":"integer","attributes":{},"value":[15]},{"type":"integer","attributes":{},"value":[3]},{"type":"character","attributes":{},"value":["2022-136"]},{"type":"character","attributes":{},"value":["RJ-2023-064"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"integer","attributes":{},"value":[93]},{"type":"integer","attributes":{},"value":[118]}]},{"type":"character","attributes":{},"value":["RJ-2023-064.pdf"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2023-064"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2023-064"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"character","attributes":{},"value":["rater","poLCA","BayesLCA","randomLCA","coda","rstan","loo","ggplot2","knitr","rjtools","rmarkdown"]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["Bayesian","Cluster","GraphicalModels","MixedModels","Phylogenetics","Psychometrics","ReproducibleResearch","Spatial","TeachingStatistics"]},{"type":"character","attributes":{},"value":["/Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library/rjtools/rjournal.csl"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2023-064/"]}]}
</script>
<!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["rater_files/anchor-4.2.2/anchor.min.js","rater_files/bowser-1.9.3/bowser.min.js","rater_files/distill-2.2.21/template.v2.js","rater_files/figure-html5/plot-pi-1.png","rater_files/figure-html5/plot-ppcs-1.png","rater_files/figure-html5/plot-rstan-traceplot-1.png","rater_files/figure-html5/plot-theta-1.png","rater_files/figure-html5/plot-theta-caries-1.png","rater_files/figure-html5/plot-theta-diff-model-1.png","rater_files/figure-html5/plot-theta-fit2-1.png","rater_files/figure-html5/plot-z-1.png","rater_files/figure-html5/plot-z-fit2-1.png","rater_files/figure-latex/plot-pi-1.pdf","rater_files/figure-latex/plot-ppcs-1.pdf","rater_files/figure-latex/plot-rstan-traceplot-1.pdf","rater_files/figure-latex/plot-theta-1.pdf","rater_files/figure-latex/plot-theta-caries-1.pdf","rater_files/figure-latex/plot-theta-diff-model-1.pdf","rater_files/figure-latex/plot-theta-fit2-1.pdf","rater_files/figure-latex/plot-z-1.pdf","rater_files/figure-latex/plot-z-fit2-1.pdf","rater_files/header-attrs-2.25/header-attrs.js","rater_files/jquery-3.6.0/jquery-3.6.0.js","rater_files/jquery-3.6.0/jquery-3.6.0.min.js","rater_files/jquery-3.6.0/jquery-3.6.0.min.map","rater_files/kePrint-0.0.1/kePrint.js","rater_files/lightable-0.0.1/lightable.css","rater_files/popper-2.6.0/popper.min.js","rater_files/tippy-6.2.7/tippy-bundle.umd.min.js","rater_files/tippy-6.2.7/tippy-light-border.css","rater_files/tippy-6.2.7/tippy.css","rater_files/tippy-6.2.7/tippy.umd.min.js","rater_files/webcomponents-2.0.0/webcomponents.js","rater.bib","rater.tex","RJ-2023-064_files/anchor-4.2.2/anchor.min.js","RJ-2023-064_files/bowser-1.9.3/bowser.min.js","RJ-2023-064_files/distill-2.2.21/template.v2.js","RJ-2023-064_files/figure-html5/plot-pi-1.png","RJ-2023-064_files/figure-html5/plot-ppcs-1.png","RJ-2023-064_files/figure-html5/plot-rstan-traceplot-1.png","RJ-2023-064_files/figure-html5/plot-theta-1.png","RJ-2023-064_files/figure-html5/plot-theta-caries-1.png","RJ-2023-064_files/figure-html5/plot-theta-diff-model-1.png","RJ-2023-064_files/figure-html5/plot-theta-fit2-1.png","RJ-2023-064_files/figure-html5/plot-z-1.png","RJ-2023-064_files/figure-html5/plot-z-fit2-1.png","RJ-2023-064_files/figure-latex/plot-pi-1.pdf","RJ-2023-064_files/figure-latex/plot-ppcs-1.pdf","RJ-2023-064_files/figure-latex/plot-rstan-traceplot-1.pdf","RJ-2023-064_files/figure-latex/plot-theta-1.pdf","RJ-2023-064_files/figure-latex/plot-theta-caries-1.pdf","RJ-2023-064_files/figure-latex/plot-theta-diff-model-1.pdf","RJ-2023-064_files/figure-latex/plot-theta-fit2-1.pdf","RJ-2023-064_files/figure-latex/plot-z-1.pdf","RJ-2023-064_files/figure-latex/plot-z-fit2-1.pdf","RJ-2023-064_files/header-attrs-2.25/header-attrs.js","RJ-2023-064_files/jquery-3.6.0/jquery-3.6.0.js","RJ-2023-064_files/jquery-3.6.0/jquery-3.6.0.min.js","RJ-2023-064_files/jquery-3.6.0/jquery-3.6.0.min.map","RJ-2023-064_files/kePrint-0.0.1/kePrint.js","RJ-2023-064_files/lightable-0.0.1/lightable.css","RJ-2023-064_files/popper-2.6.0/popper.min.js","RJ-2023-064_files/tippy-6.2.7/tippy-bundle.umd.min.js","RJ-2023-064_files/tippy-6.2.7/tippy-light-border.css","RJ-2023-064_files/tippy-6.2.7/tippy.css","RJ-2023-064_files/tippy-6.2.7/tippy.umd.min.js","RJ-2023-064_files/webcomponents-2.0.0/webcomponents.js","RJ-2023-064.log","RJ-2023-064.pdf","RJ-2023-064.tex","RJ-2023-064.zip","RJournal.sty","RJwrapper.aux","RJwrapper.log","RJwrapper.out","RJwrapper.tex","Rlogo-5.png","tikz/figrelationships.pdf","tikz/figrelationships.png","tikz/figrelationships.tex"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
<meta name="distill:offset" content="../.."/>

<script type="application/javascript">

  window.headroom_prevent_pin = false;

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // initialize headroom for banner
    var header = $('header').get(0);
    var headerHeight = header.offsetHeight;
    var headroom = new Headroom(header, {
      tolerance: 5,
      onPin : function() {
        if (window.headroom_prevent_pin) {
          window.headroom_prevent_pin = false;
          headroom.unpin();
        }
      }
    });
    headroom.init();
    if(window.location.hash)
      headroom.unpin();
    $(header).addClass('headroom--transition');

    // offset scroll location for banner on hash change
    // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
    window.addEventListener("hashchange", function(event) {
      window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
    });

    // responsive menu
    $('.distill-site-header').each(function(i, val) {
      var topnav = $(this);
      var toggle = topnav.find('.nav-toggle');
      toggle.on('click', function() {
        topnav.toggleClass('responsive');
      });
    });

    // nav dropdowns
    $('.nav-dropbtn').click(function(e) {
      $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
      $(this).parent().siblings('.nav-dropdown')
         .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $("body").click(function(e){
      $('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $(".nav-dropdown").click(function(e){
      e.stopPropagation();
    });
  });
</script>

<style type="text/css">

/* Theme (user-documented overrideables for nav appearance) */

.distill-site-nav {
  color: rgba(255, 255, 255, 0.8);
  background-color: #0F2E3D;
  font-size: 15px;
  font-weight: 300;
}

.distill-site-nav a {
  color: inherit;
  text-decoration: none;
}

.distill-site-nav a:hover {
  color: white;
}

@media print {
  .distill-site-nav {
    display: none;
  }
}

.distill-site-header {

}

.distill-site-footer {

}


/* Site Header */

.distill-site-header {
  width: 100%;
  box-sizing: border-box;
  z-index: 3;
}

.distill-site-header .nav-left {
  display: inline-block;
  margin-left: 8px;
}

@media screen and (max-width: 768px) {
  .distill-site-header .nav-left {
    margin-left: 0;
  }
}


.distill-site-header .nav-right {
  float: right;
  margin-right: 8px;
}

.distill-site-header a,
.distill-site-header .title {
  display: inline-block;
  text-align: center;
  padding: 14px 10px 14px 10px;
}

.distill-site-header .title {
  font-size: 18px;
  min-width: 150px;
}

.distill-site-header .logo {
  padding: 0;
}

.distill-site-header .logo img {
  display: none;
  max-height: 20px;
  width: auto;
  margin-bottom: -4px;
}

.distill-site-header .nav-image img {
  max-height: 18px;
  width: auto;
  display: inline-block;
  margin-bottom: -3px;
}



@media screen and (min-width: 1000px) {
  .distill-site-header .logo img {
    display: inline-block;
  }
  .distill-site-header .nav-left {
    margin-left: 20px;
  }
  .distill-site-header .nav-right {
    margin-right: 20px;
  }
  .distill-site-header .title {
    padding-left: 12px;
  }
}


.distill-site-header .nav-toggle {
  display: none;
}

.nav-dropdown {
  display: inline-block;
  position: relative;
}

.nav-dropdown .nav-dropbtn {
  border: none;
  outline: none;
  color: rgba(255, 255, 255, 0.8);
  padding: 16px 10px;
  background-color: transparent;
  font-family: inherit;
  font-size: inherit;
  font-weight: inherit;
  margin: 0;
  margin-top: 1px;
  z-index: 2;
}

.nav-dropdown-content {
  display: none;
  position: absolute;
  background-color: white;
  min-width: 200px;
  border: 1px solid rgba(0,0,0,0.15);
  border-radius: 4px;
  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
  z-index: 1;
  margin-top: 2px;
  white-space: nowrap;
  padding-top: 4px;
  padding-bottom: 4px;
}

.nav-dropdown-content hr {
  margin-top: 4px;
  margin-bottom: 4px;
  border: none;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.nav-dropdown-active {
  display: block;
}

.nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
  color: black;
  padding: 6px 24px;
  text-decoration: none;
  display: block;
  text-align: left;
}

.nav-dropdown-content .nav-dropdown-header {
  display: block;
  padding: 5px 24px;
  padding-bottom: 0;
  text-transform: uppercase;
  font-size: 14px;
  color: #999999;
  white-space: nowrap;
}

.nav-dropdown:hover .nav-dropbtn {
  color: white;
}

.nav-dropdown-content a:hover {
  background-color: #ddd;
  color: black;
}

.nav-right .nav-dropdown-content {
  margin-left: -45%;
  right: 0;
}

@media screen and (max-width: 768px) {
  .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
  .distill-site-header a.nav-toggle {
    float: right;
    display: block;
  }
  .distill-site-header .title {
    margin-left: 0;
  }
  .distill-site-header .nav-right {
    margin-right: 0;
  }
  .distill-site-header {
    overflow: hidden;
  }
  .nav-right .nav-dropdown-content {
    margin-left: 0;
  }
}


@media screen and (max-width: 768px) {
  .distill-site-header.responsive {position: relative; min-height: 500px; }
  .distill-site-header.responsive a.nav-toggle {
    position: absolute;
    right: 0;
    top: 0;
  }
  .distill-site-header.responsive a,
  .distill-site-header.responsive .nav-dropdown {
    display: block;
    text-align: left;
  }
  .distill-site-header.responsive .nav-left,
  .distill-site-header.responsive .nav-right {
    width: 100%;
  }
  .distill-site-header.responsive .nav-dropdown {float: none;}
  .distill-site-header.responsive .nav-dropdown-content {position: relative;}
  .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
    display: block;
    width: 100%;
    text-align: left;
  }
}

/* Site Footer */

.distill-site-footer {
  width: 100%;
  overflow: hidden;
  box-sizing: border-box;
  z-index: 3;
  margin-top: 30px;
  padding-top: 30px;
  padding-bottom: 30px;
  text-align: center;
}

/* Headroom */

d-title {
  padding-top: 6rem;
}

@media print {
  d-title {
    padding-top: 4rem;
  }
}

.headroom {
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
}

.headroom--transition {
  transition: all .4s ease-in-out;
}

.headroom--unpinned {
  top: -100px;
}

.headroom--pinned {
  top: 0;
}

/* adjust viewport for navbar height */
/* helps vertically center bootstrap (non-distill) content */
.min-vh-100 {
  min-height: calc(100vh - 100px) !important;
}

</style>

<script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="../../site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"/>
<link href="../../site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"/>
<script src="../../site_libs/headroom-0.9.4/headroom.min.js"></script>
<!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

<style type="text/css">

body {
  background-color: white;
}

.pandoc-table {
  width: 100%;
}

.pandoc-table>caption {
  margin-bottom: 10px;
}

.pandoc-table th:not([align]) {
  text-align: left;
}

.pagedtable-footer {
  font-size: 15px;
}

d-byline .byline {
  grid-template-columns: 2fr 2fr;
}

d-byline .byline h3 {
  margin-block-start: 1.5em;
}

d-byline .byline .authors-affiliations h3 {
  margin-block-start: 0.5em;
}

.authors-affiliations .orcid-id {
  width: 16px;
  height:16px;
  margin-left: 4px;
  margin-right: 4px;
  vertical-align: middle;
  padding-bottom: 2px;
}

d-title .dt-tags {
  margin-top: 1em;
  grid-column: text;
}

.dt-tags .dt-tag {
  text-decoration: none;
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0em 0.4em;
  margin-right: 0.5em;
  margin-bottom: 0.4em;
  font-size: 70%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

d-article table.gt_table td,
d-article table.gt_table th {
  border-bottom: none;
  font-size: 100%;
}

.html-widget {
  margin-bottom: 2.0em;
}

.l-screen-inset {
  padding-right: 16px;
}

.l-screen .caption {
  margin-left: 10px;
}

.shaded {
  background: rgb(247, 247, 247);
  padding-top: 20px;
  padding-bottom: 20px;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .html-widget {
  margin-bottom: 0;
  border: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .shaded-content {
  background: white;
}

.text-output {
  margin-top: 0;
  line-height: 1.5em;
}

.hidden {
  display: none !important;
}

hr.section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  margin: 0px;
}


d-byline {
  border-top: none;
}

d-article {
  padding-top: 2.5rem;
  padding-bottom: 30px;
  border-top: none;
}

d-appendix {
  padding-top: 30px;
}

d-article>p>img {
  width: 100%;
}

d-article h2 {
  margin: 1rem 0 1.5rem 0;
}

d-article h3 {
  margin-top: 1.5rem;
}

d-article iframe {
  border: 1px solid rgba(0, 0, 0, 0.1);
  margin-bottom: 2.0em;
  width: 100%;
}

/* Tweak code blocks */

d-article div.sourceCode code,
d-article pre code {
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
}

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: auto;
}

d-article div.sourceCode {
  background-color: white;
}

d-article div.sourceCode pre {
  padding-left: 10px;
  font-size: 12px;
  border-left: 2px solid rgba(0,0,0,0.1);
}

d-article pre {
  font-size: 12px;
  color: black;
  background: none;
  margin-top: 0;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

d-article pre a {
  border-bottom: none;
}

d-article pre a:hover {
  border-bottom: none;
  text-decoration: underline;
}

d-article details {
  grid-column: text;
  margin-bottom: 0.8em;
}

@media(min-width: 768px) {

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: visible !important;
}

d-article div.sourceCode pre {
  padding-left: 18px;
  font-size: 14px;
}

/* tweak for Pandoc numbered line within distill */
d-article pre.numberSource code > span {
    left: -2em;
}

d-article pre {
  font-size: 14px;
}

}

figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

/* CSS for d-contents */

.d-contents {
  grid-column: text;
  color: rgba(0,0,0,0.8);
  font-size: 0.9em;
  padding-bottom: 1em;
  margin-bottom: 1em;
  padding-bottom: 0.5em;
  margin-bottom: 1em;
  padding-left: 0.25em;
  justify-self: start;
}

@media(min-width: 1000px) {
  .d-contents.d-contents-float {
    height: 0;
    grid-column-start: 1;
    grid-column-end: 4;
    justify-self: center;
    padding-right: 3em;
    padding-left: 2em;
  }
}

.d-contents nav h3 {
  font-size: 18px;
  margin-top: 0;
  margin-bottom: 1em;
}

.d-contents li {
  list-style-type: none
}

.d-contents nav > ul {
  padding-left: 0;
}

.d-contents ul {
  padding-left: 1em
}

.d-contents nav ul li {
  margin-top: 0.6em;
  margin-bottom: 0.2em;
}

.d-contents nav a {
  font-size: 13px;
  border-bottom: none;
  text-decoration: none
  color: rgba(0, 0, 0, 0.8);
}

.d-contents nav a:hover {
  text-decoration: underline solid rgba(0, 0, 0, 0.6)
}

.d-contents nav > ul > li > a {
  font-weight: 600;
}

.d-contents nav > ul > li > ul {
  font-weight: inherit;
}

.d-contents nav > ul > li > ul > li {
  margin-top: 0.2em;
}


.d-contents nav ul {
  margin-top: 0;
  margin-bottom: 0.25em;
}

.d-article-with-toc h2:nth-child(2) {
  margin-top: 0;
}


/* Figure */

.figure {
  position: relative;
  margin-bottom: 2.5em;
  margin-top: 1.5em;
}

.figure .caption {
  color: rgba(0, 0, 0, 0.6);
  font-size: 12px;
  line-height: 1.5em;
}

.figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

.figure .caption a {
  color: rgba(0, 0, 0, 0.6);
}

.figure .caption b,
.figure .caption strong, {
  font-weight: 600;
  color: rgba(0, 0, 0, 1.0);
}

/* Citations */

d-article .citation {
  color: inherit;
  cursor: inherit;
}

div.hanging-indent{
  margin-left: 1em; text-indent: -1em;
}

/* Citation hover box */

.tippy-box[data-theme~=light-border] {
  background-color: rgba(250, 250, 250, 0.95);
}

.tippy-content > p {
  margin-bottom: 0;
  padding: 2px;
}


/* Tweak 1000px media break to show more text */

@media(min-width: 1000px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 16px;
  }

  .grid {
    grid-column-gap: 16px;
  }

  d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
  }
  figure .caption, .figure .caption, figure figcaption {
    font-size: 13px;
  }
}

@media(min-width: 1180px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 32px;
  }

  .grid {
    grid-column-gap: 32px;
  }
}


/* Get the citation styles for the appendix (not auto-injected on render since
   we do our own rendering of the citation appendix) */

d-appendix .citation-appendix,
.d-appendix .citation-appendix {
  font-size: 11px;
  line-height: 15px;
  border-left: 1px solid rgba(0, 0, 0, 0.1);
  padding-left: 18px;
  border: 1px solid rgba(0,0,0,0.1);
  background: rgba(0, 0, 0, 0.02);
  padding: 10px 18px;
  border-radius: 3px;
  color: rgba(150, 150, 150, 1);
  overflow: hidden;
  margin-top: -12px;
  white-space: pre-wrap;
  word-wrap: break-word;
}

/* Include appendix styles here so they can be overridden */

d-appendix {
  contain: layout style;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-top: 60px;
  margin-bottom: 0;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  color: rgba(0,0,0,0.5);
  padding-top: 60px;
  padding-bottom: 48px;
}

d-appendix h3 {
  grid-column: page-start / text-start;
  font-size: 15px;
  font-weight: 500;
  margin-top: 1em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.65);
}

d-appendix h3 + * {
  margin-top: 1em;
}

d-appendix ol {
  padding: 0 0 0 15px;
}

@media (min-width: 768px) {
  d-appendix ol {
    padding: 0 0 0 30px;
    margin-left: -30px;
  }
}

d-appendix li {
  margin-bottom: 1em;
}

d-appendix a {
  color: rgba(0, 0, 0, 0.6);
}

d-appendix > * {
  grid-column: text;
}

d-appendix > d-footnote-list,
d-appendix > d-citation-list,
d-appendix > distill-appendix {
  grid-column: screen;
}

/* Include footnote styles here so they can be overridden */

d-footnote-list {
  contain: layout style;
}

d-footnote-list > * {
  grid-column: text;
}

d-footnote-list a.footnote-backlink {
  color: rgba(0,0,0,0.3);
  padding-left: 0.5em;
}



/* Anchor.js */

.anchorjs-link {
  /*transition: all .25s linear; */
  text-decoration: none;
  border-bottom: none;
}
*:hover > .anchorjs-link {
  margin-left: -1.125em !important;
  text-decoration: none;
  border-bottom: none;
}

/* Social footer */

.social_footer {
  margin-top: 30px;
  margin-bottom: 0;
  color: rgba(0,0,0,0.67);
}

.disqus-comments {
  margin-right: 30px;
}

.disqus-comment-count {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  cursor: pointer;
}

#disqus_thread {
  margin-top: 30px;
}

.article-sharing a {
  border-bottom: none;
  margin-right: 8px;
}

.article-sharing a:hover {
  border-bottom: none;
}

.sidebar-section.subscribe {
  font-size: 12px;
  line-height: 1.6em;
}

.subscribe p {
  margin-bottom: 0.5em;
}


.article-footer .subscribe {
  font-size: 15px;
  margin-top: 45px;
}


.sidebar-section.custom {
  font-size: 12px;
  line-height: 1.6em;
}

.custom p {
  margin-bottom: 0.5em;
}

/* Styles for listing layout (hide title) */
.layout-listing d-title, .layout-listing .d-title {
  display: none;
}

/* Styles for posts lists (not auto-injected) */


.posts-with-sidebar {
  padding-left: 45px;
  padding-right: 45px;
}

.posts-list .description h2,
.posts-list .description p {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

.posts-list .description h2 {
  font-weight: 700;
  border-bottom: none;
  padding-bottom: 0;
}

.posts-list h2.post-tag {
  border-bottom: 1px solid rgba(0, 0, 0, 0.2);
  padding-bottom: 12px;
}
.posts-list {
  margin-top: 60px;
  margin-bottom: 24px;
}

.posts-list .post-preview {
  text-decoration: none;
  overflow: hidden;
  display: block;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  padding: 24px 0;
}

.post-preview-last {
  border-bottom: none !important;
}

.posts-list .posts-list-caption {
  grid-column: screen;
  font-weight: 400;
}

.posts-list .post-preview h2 {
  margin: 0 0 6px 0;
  line-height: 1.2em;
  font-style: normal;
  font-size: 24px;
}

.posts-list .post-preview p {
  margin: 0 0 12px 0;
  line-height: 1.4em;
  font-size: 16px;
}

.posts-list .post-preview .thumbnail {
  box-sizing: border-box;
  margin-bottom: 24px;
  position: relative;
  max-width: 500px;
}
.posts-list .post-preview img {
  width: 100%;
  display: block;
}

.posts-list .metadata {
  font-size: 12px;
  line-height: 1.4em;
  margin-bottom: 18px;
}

.posts-list .metadata > * {
  display: inline-block;
}

.posts-list .metadata .publishedDate {
  margin-right: 2em;
}

.posts-list .metadata .dt-authors {
  display: block;
  margin-top: 0.3em;
  margin-right: 2em;
}

.posts-list .dt-tags {
  display: block;
  line-height: 1em;
}

.posts-list .dt-tags .dt-tag {
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0.3em 0.4em;
  margin-right: 0.2em;
  margin-bottom: 0.4em;
  font-size: 60%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

.posts-list img {
  opacity: 1;
}

.posts-list img[data-src] {
  opacity: 0;
}

.posts-more {
  clear: both;
}


.posts-sidebar {
  font-size: 16px;
}

.posts-sidebar h3 {
  font-size: 16px;
  margin-top: 0;
  margin-bottom: 0.5em;
  font-weight: 400;
  text-transform: uppercase;
}

.sidebar-section {
  margin-bottom: 30px;
}

.categories ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.categories li {
  color: rgba(0, 0, 0, 0.8);
  margin-bottom: 0;
}

.categories li>a {
  border-bottom: none;
}

.categories li>a:hover {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
}

.categories .active {
  font-weight: 600;
}

.categories .category-count {
  color: rgba(0, 0, 0, 0.4);
}


@media(min-width: 768px) {
  .posts-list .post-preview h2 {
    font-size: 26px;
  }
  .posts-list .post-preview .thumbnail {
    float: right;
    width: 30%;
    margin-bottom: 0;
  }
  .posts-list .post-preview .description {
    float: left;
    width: 45%;
  }
  .posts-list .post-preview .metadata {
    float: left;
    width: 20%;
    margin-top: 8px;
  }
  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.5em;
    font-size: 16px;
  }
  .posts-with-sidebar .posts-list {
    float: left;
    width: 75%;
  }
  .posts-with-sidebar .posts-sidebar {
    float: right;
    width: 20%;
    margin-top: 60px;
    padding-top: 24px;
    padding-bottom: 24px;
  }
}


/* Improve display for browsers without grid (IE/Edge <= 15) */

.downlevel {
  line-height: 1.6em;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  margin: 0;
}

.downlevel .d-title {
  padding-top: 6rem;
  padding-bottom: 1.5rem;
}

.downlevel .d-title h1 {
  font-size: 50px;
  font-weight: 700;
  line-height: 1.1em;
  margin: 0 0 0.5rem;
}

.downlevel .d-title p {
  font-weight: 300;
  font-size: 1.2rem;
  line-height: 1.55em;
  margin-top: 0;
}

.downlevel .d-byline {
  padding-top: 0.8em;
  padding-bottom: 0.8em;
  font-size: 0.8rem;
  line-height: 1.8em;
}

.downlevel .section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
}

.downlevel .d-article {
  font-size: 1.06rem;
  line-height: 1.7em;
  padding-top: 1rem;
  padding-bottom: 2rem;
}


.downlevel .d-appendix {
  padding-left: 0;
  padding-right: 0;
  max-width: none;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.5);
  padding-top: 40px;
  padding-bottom: 48px;
}

.downlevel .footnotes ol {
  padding-left: 13px;
}

.downlevel .base-grid,
.downlevel .distill-header,
.downlevel .d-title,
.downlevel .d-abstract,
.downlevel .d-article,
.downlevel .d-appendix,
.downlevel .distill-appendix,
.downlevel .d-byline,
.downlevel .d-footnote-list,
.downlevel .d-citation-list,
.downlevel .distill-footer,
.downlevel .appendix-bottom,
.downlevel .posts-container {
  padding-left: 40px;
  padding-right: 40px;
}

@media(min-width: 768px) {
  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
  padding-left: 150px;
  padding-right: 150px;
  max-width: 900px;
}
}

.downlevel pre code {
  display: block;
  border-left: 2px solid rgba(0, 0, 0, .1);
  padding: 0 0 0 20px;
  font-size: 14px;
}

.downlevel code, .downlevel pre {
  color: black;
  background: none;
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

.downlevel .posts-list .post-preview {
  color: inherit;
}



</style>

<script type="application/javascript">

function is_downlevel_browser() {
  if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                 window.navigator.userAgent)) {
    return true;
  } else {
    return window.load_distill_framework === undefined;
  }
}

// show body when load is complete
function on_load_complete() {

  // add anchors
  if (window.anchors) {
    window.anchors.options.placement = 'left';
    window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
  }


  // set body to visible
  document.body.style.visibility = 'visible';

  // force redraw for leaflet widgets
  if (window.HTMLWidgets) {
    var maps = window.HTMLWidgets.findAll(".leaflet");
    $.each(maps, function(i, el) {
      var map = this.getMap();
      map.invalidateSize();
      map.eachLayer(function(layer) {
        if (layer instanceof L.TileLayer)
          layer.redraw();
      });
    });
  }

  // trigger 'shown' so htmlwidgets resize
  $('d-article').trigger('shown');
}

function init_distill() {

  init_common();

  // create front matter
  var front_matter = $('<d-front-matter></d-front-matter>');
  $('#distill-front-matter').wrap(front_matter);

  // create d-title
  $('.d-title').changeElementType('d-title');

  // separator
  var separator = '<hr class="section-separator" style="clear: both"/>';
  // prepend separator above appendix
  $('.d-byline').before(separator);
  $('.d-article').before(separator);

  // create d-byline
  var byline = $('<d-byline></d-byline>');
  $('.d-byline').replaceWith(byline);

  // create d-article
  var article = $('<d-article></d-article>');
  $('.d-article').wrap(article).children().unwrap();

  // move posts container into article
  $('.posts-container').appendTo($('d-article'));

  // create d-appendix
  $('.d-appendix').changeElementType('d-appendix');

  // flag indicating that we have appendix items
  var appendix = $('.appendix-bottom').children('h3').length > 0;

  // replace footnotes with <d-footnote>
  $('.footnote-ref').each(function(i, val) {
    appendix = true;
    var href = $(this).attr('href');
    var id = href.replace('#', '');
    var fn = $('#' + id);
    var fn_p = $('#' + id + '>p');
    fn_p.find('.footnote-back').remove();
    var text = fn_p.html();
    var dtfn = $('<d-footnote></d-footnote>');
    dtfn.html(text);
    $(this).replaceWith(dtfn);
  });
  // remove footnotes
  $('.footnotes').remove();

  // move refs into #references-listing
  $('#references-listing').replaceWith($('#refs'));

  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    var id = $(this).attr('id');
    $('.d-contents a[href="#' + id + '"]').parent().remove();
    appendix = true;
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
  });

  // show d-appendix if we have appendix content
  $("d-appendix").css('display', appendix ? 'grid' : 'none');

  // localize layout chunks to just output
  $('.layout-chunk').each(function(i, val) {

    // capture layout
    var layout = $(this).attr('data-layout');

    // apply layout to markdown level block elements
    var elements = $(this).children().not('details, div.sourceCode, pre, script');
    elements.each(function(i, el) {
      var layout_div = $('<div class="' + layout + '"></div>');
      if (layout_div.hasClass('shaded')) {
        var shaded_content = $('<div class="shaded-content"></div>');
        $(this).wrap(shaded_content);
        $(this).parent().wrap(layout_div);
      } else {
        $(this).wrap(layout_div);
      }
    });


    // unwrap the layout-chunk div
    $(this).children().unwrap();
  });

  // remove code block used to force  highlighting css
  $('.distill-force-highlighting-css').parent().remove();

  // remove empty line numbers inserted by pandoc when using a
  // custom syntax highlighting theme, except when numbering line
  // in code chunk
  $('pre:not(.numberLines) code.sourceCode a:empty').remove();

  // load distill framework
  load_distill_framework();

  // wait for window.distillRunlevel == 4 to do post processing
  function distill_post_process() {

    if (!window.distillRunlevel || window.distillRunlevel < 4)
      return;

    // hide author/affiliations entirely if we have no authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;
    if (!have_authors)
      $('d-byline').addClass('hidden');

    // article with toc class
    $('.d-contents').parent().addClass('d-article-with-toc');

    // strip links that point to #
    $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

    // add orcid ids
    $('.authors-affiliations').find('.author').each(function(i, el) {
      var orcid_id = front_matter.authors[i].orcidID;
      var author_name = front_matter.authors[i].author
      if (orcid_id) {
        var a = $('<a></a>');
        a.attr('href', 'https://orcid.org/' + orcid_id);
        var img = $('<img></img>');
        img.addClass('orcid-id');
        img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
        img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
        a.append(img);
        $(this).append(a);
      }
    });

    // hide elements of author/affiliations grid that have no value
    function hide_byline_column(caption) {
      $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
    }

    // affiliations
    var have_affiliations = false;
    for (var i = 0; i<front_matter.authors.length; ++i) {
      var author = front_matter.authors[i];
      if (author.affiliation !== "&nbsp;") {
        have_affiliations = true;
        break;
      }
    }
    if (!have_affiliations)
      $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

    // published date
    if (!front_matter.publishedDate)
      hide_byline_column("Published");

    // document object identifier
    var doi = $('d-byline').find('h3:contains("DOI")');
    var doi_p = doi.next().empty();
    if (!front_matter.doi) {
      // if we have a citation and valid citationText then link to that
      if ($('#citation').length > 0 && front_matter.citationText) {
        doi.html('Citation');
        $('<a href="#citation"></a>')
          .text(front_matter.citationText)
          .appendTo(doi_p);
      } else {
        hide_byline_column("DOI");
      }
    } else {
      $('<a></a>')
         .attr('href', "https://doi.org/" + front_matter.doi)
         .html(front_matter.doi)
         .appendTo(doi_p);
    }

     // change plural form of authors/affiliations
    if (front_matter.authors.length === 1) {
      var grid = $('.authors-affiliations');
      grid.children('h3:contains("Authors")').text('Author');
      grid.children('h3:contains("Affiliations")').text('Affiliation');
    }

    // remove d-appendix and d-footnote-list local styles
    $('d-appendix > style:first-child').remove();
    $('d-footnote-list > style:first-child').remove();

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // hoverable references
    $('span.citation[data-cites]').each(function() {
      const citeChild = $(this).children()[0]
      // Do not process if @xyz has been used without escaping and without bibliography activated
      // https://github.com/rstudio/distill/issues/466
      if (citeChild === undefined) return true

      if (citeChild.nodeName == "D-FOOTNOTE") {
        var fn = citeChild
        $(this).html(fn.shadowRoot.querySelector("sup"))
        $(this).id = fn.id
        fn.remove()
      }
      var refs = $(this).attr('data-cites').split(" ");
      var refHtml = refs.map(function(ref) {
        // Could use CSS.escape too here, we insure backward compatibility in navigator
        return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
      }).join("\n");
      window.tippy(this, {
        allowHTML: true,
        content: refHtml,
        maxWidth: 500,
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start'
      });
    });

    // fix footnotes in tables (#411)
    // replacing broken distill.pub feature
    $('table d-footnote').each(function() {
      // we replace internal showAtNode methode which is triggered when hovering a footnote
      this.hoverBox.showAtNode = function(node) {
        // ported from https://github.com/distillpub/template/pull/105/files
        calcOffset = function(elem) {
            let x = elem.offsetLeft;
            let y = elem.offsetTop;
            // Traverse upwards until an `absolute` element is found or `elem`
            // becomes null.
            while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                x += elem.offsetLeft;
                y += elem.offsetTop;
            }

            return { left: x, top: y };
        }
        // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
        const bbox = node.getBoundingClientRect();
        const offset = calcOffset(node);
        this.show([offset.left + bbox.width, offset.top + bbox.height]);
      }
    })

    // clear polling timer
    clearInterval(tid);

    // show body now that everything is ready
    on_load_complete();
  }

  var tid = setInterval(distill_post_process, 50);
  distill_post_process();

}

function init_downlevel() {

  init_common();

   // insert hr after d-title
  $('.d-title').after($('<hr class="section-separator"/>'));

  // check if we have authors
  var front_matter = JSON.parse($("#distill-front-matter").html());
  var have_authors = front_matter.authors && front_matter.authors.length > 0;

  // manage byline/border
  if (!have_authors)
    $('.d-byline').remove();
  $('.d-byline').after($('<hr class="section-separator"/>'));
  $('.d-byline a').remove();

  // remove toc
  $('.d-contents').remove();

  // move appendix elements
  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
  });


  // inject headers into references and footnotes
  var refs_header = $('<h3></h3>');
  refs_header.text('References');
  $('#refs').prepend(refs_header);

  var footnotes_header = $('<h3></h3');
  footnotes_header.text('Footnotes');
  $('.footnotes').children('hr').first().replaceWith(footnotes_header);

  // move appendix-bottom entries to the bottom
  $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
  $('.appendix-bottom').remove();

  // remove appendix if it's empty
  if ($('.d-appendix').children().length === 0)
    $('.d-appendix').remove();

  // prepend separator above appendix
  $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

  // trim code
  $('pre>code').each(function(i, val) {
    $(this).html($.trim($(this).html()));
  });

  // move posts-container right before article
  $('.posts-container').insertBefore($('.d-article'));

  $('body').addClass('downlevel');

  on_load_complete();
}


function init_common() {

  // jquery plugin to change element types
  (function($) {
    $.fn.changeElementType = function(newType) {
      var attrs = {};

      $.each(this[0].attributes, function(idx, attr) {
        attrs[attr.nodeName] = attr.nodeValue;
      });

      this.replaceWith(function() {
        return $("<" + newType + "/>", attrs).append($(this).contents());
      });
    };
  })(jQuery);

  // prevent underline for linked images
  $('a > img').parent().css({'border-bottom' : 'none'});

  // mark non-body figures created by knitr chunks as 100% width
  $('.layout-chunk').each(function(i, val) {
    var figures = $(this).find('img, .html-widget');
    // ignore leaflet img layers (#106)
    figures = figures.filter(':not(img[class*="leaflet"])')
    if ($(this).attr('data-layout') !== "l-body") {
      figures.css('width', '100%');
    } else {
      figures.css('max-width', '100%');
      figures.filter("[width]").each(function(i, val) {
        var fig = $(this);
        fig.css('width', fig.attr('width') + 'px');
      });

    }
  });

  // auto-append index.html to post-preview links in file: protocol
  // and in rstudio ide preview
  $('.post-preview').each(function(i, val) {
    if (window.location.protocol === "file:")
      $(this).attr('href', $(this).attr('href') + "index.html");
  });

  // get rid of index.html references in header
  if (window.location.protocol !== "file:") {
    $('.distill-site-header a[href]').each(function(i,val) {
      $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
    });
  }

  // add class to pandoc style tables
  $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
  $('.kable-table').children('table').addClass('pandoc-table');

  // add figcaption style to table captions
  $('caption').parent('table').addClass("figcaption");

  // initialize posts list
  if (window.init_posts_list)
    window.init_posts_list();

  // implmement disqus comment link
  $('.disqus-comment-count').click(function() {
    window.headroom_prevent_pin = true;
    $('#disqus_thread').toggleClass('hidden');
    if (!$('#disqus_thread').hasClass('hidden')) {
      var offset = $(this).offset();
      $(window).resize();
      $('html, body').animate({
        scrollTop: offset.top - 35
      });
    }
  });
}

document.addEventListener('DOMContentLoaded', function() {
  if (is_downlevel_browser())
    init_downlevel();
  else
    window.addEventListener('WebComponentsReady', init_distill);
});

</script>

<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.5);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

.distill-site-footer {
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */
</style>
<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.7);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #ffffff;
}

.distill-site-footer {
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #fafafa;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */

.nav-right > a {
  text-transform: uppercase;
}

d-title h1, d-title p, d-title figure,
d-abstract p, d-abstract b {
  grid-column: page;
}

.rj-blue {
  color: #2467bb;
}

ul li {
  line-height: 1.6;
  margin-top: 0em;
  margin-bottom: 0em;
}</style>
<style type="text/css">
/* base style */

/* FONT FAMILIES */

:root {
  --heading-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  --mono-default: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  --body-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

body,
.posts-list .post-preview p,
.posts-list .description p {
  font-family: var(--body-font), var(--body-default);
}

h1, h2, h3, h4, h5, h6,
.posts-list .post-preview h2,
.posts-list .description h2 {
  font-family: var(--heading-font), var(--heading-default);
}

d-article div.sourceCode code,
d-article pre code {
  font-family: var(--mono-font), var(--mono-default);
}


/*-- TITLE --*/
d-title h1,
.posts-list > h1 {
  color: var(--title-color, black);
}

d-title h1 {
  font-size: var(--title-size, 50px);
}

/*-- HEADERS --*/
d-article h1,
d-article h2,
d-article h3,
d-article h4,
d-article h5,
d-article h6 {
  color: var(--header-color, rgba(0, 0, 0, 0.8));
}

/*-- BODY --*/
d-article > p,  /* only text inside of <p> tags */
d-article > ul, /* lists */
d-article > ol {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
  font-size: var(--body-size, 1.06rem);
}


/*-- CODE --*/
d-article div.sourceCode code,
d-article pre code {
  font-size: var(--code-size, 14px);
}

/*-- ASIDE --*/
d-article aside {
  font-size: var(--aside-size, 12px);
  color: var(--aside-color, rgba(0, 0, 0, 0.6));
}

/*-- FIGURE CAPTIONS --*/
figure .caption,
figure figcaption,
.figure .caption {
  font-size: var(--fig-cap-size, 13px);
  color: var(--fig-cap-color, rgba(0, 0, 0, 0.6));
}

/*-- METADATA --*/
d-byline h3 {
  font-size: var(--heading-size, 0.6rem);
  color: var(--heading-color, rgba(0, 0, 0, 0.5));
}

d-byline {
  font-size: var(--body-size, 0.8rem);
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

d-byline a,
d-article d-byline a {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

/*-- TABLE OF CONTENTS --*/
.d-contents nav h3 {
  font-size: var(--heading-size, 18px);
}

.d-contents nav a {
  font-size: var(--contents-size, 13px);
}

/*-- APPENDIX --*/
d-appendix h3 {
  font-size: var(--heading-size, 15px);
  color: var(--heading-color, rgba(0, 0, 0, 0.65));
}

d-appendix {
  font-size: var(--text-size, 0.8em);
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

d-appendix d-footnote-list a.footnote-backlink {
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

/*-- WEBSITE HEADER + FOOTER --*/
.distill-site-header .title {
  font-size: var(--title-size, 18px);
  font-family: var(--navbar-font), var(--heading-default);
}

.distill-site-header a,
.nav-dropdown .nav-dropbtn {
  font-family: var(--navbar-font), var(--heading-default);
}

.nav-dropdown .nav-dropbtn {
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  font-size: var(--text-size, 15px);
}

.distill-site-header a:hover,
.nav-dropdown:hover .nav-dropbtn {
  color: var(--hover-color, white);
}

.distill-site-header {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer a:hover {
  color: var(--hover-color, white);
}</style>
<!--/radix_placeholder_distill-->
  <script src="../../site_libs/header-attrs-2.25/header-attrs.js"></script>
  <script src="../../site_libs/kePrint-0.0.1/kePrint.js"></script>
  <link href="../../site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />
  <script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="../../site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="../../site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="../../site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="../../site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="../../site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="../../site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="../../site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="../../site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
<!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      buttons.append('<a href="' + slug + '.zip" class="dt-tag"><i class="fas fa-file-zipper"></i> Supplement</a>')

      // adds Abstract: in front of the first <p> in the title section --
      // unless it happens to be the subtitle (FIXME: this is a bad hack - can't distill do this?)
      var tpar = $("d-title p:not(:empty)").first();
      if (tpar && ! tpar.hasClass("subtitle")) {
        const abstract = $('<d-abstract>')
        abstract.append('<b>Abstract:</b><br>')
        abstract.append($("d-title p:not(:empty)").first()) // Move description to d-abstract
        $("d-title p:empty").remove() // Remove empty paragraphs after title
        abstract.append(buttons)
        abstract.insertAfter($('d-title')) // Add abstract section after title */
      }

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
    
    d-appendix .citation-appendix,
    .d-appendix .citation-appendix {
      color: rgb(60, 60, 60);
    }

    d-article h2 {
      border-bottom: 0px solid rgba(0, 0, 0, 0.1);
      padding-bottom: 0rem;
    }
    d-article h3 {
      font-size: 20px;
    }
    d-article h4 {
      font-size: 18px;
      text-transform: none;
    }

    @media (min-width: 1024px) {
      d-article h2 {
        font-size: 32px;
      }
      d-article h3 {
        font-size: 24px;
      }
      d-article h4 {
        font-size: 20px;
      }
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Statistical Models for Repeated Categorical Ratings: The R Package rater","description":"A common problem in many disciplines is the need to assign a set of items into categories or classes with known labels.  This is often done by one or more expert raters, or sometimes by an automated process.  If these assignments or 'ratings' are difficult to make accurately, a common tactic is to repeat them by different raters, or even by the same rater multiple times on different occasions. We present an R package rater, available on CRAN, that implements Bayesian versions of several statistical models for analysis of repeated categorical rating data. Inference is possible for the true underlying (latent) class of each item, as well as the accuracy of each rater. The models are extensions of, and include, the Dawid--Skene model, and we implemented them using the Stan probabilistic programming language.  We illustrate the use of rater through a few examples.  We also discuss in detail the techniques of marginalisation and conditioning, which are necessary for these models but also apply more generally to other models implemented in Stan.","doi":"10.32614/RJ-2023-064","authors":[{"author":"Jeffrey M. Pullin","authorURL":"#","affiliation":"University of Melbourne","affiliationURL":"#","orcidID":"0000-0003-3651-5471"},{"author":"Lyle C. Gurrin","authorURL":"#","affiliation":"University of Melbourne","affiliationURL":"#","orcidID":"0000-0001-7052-1969"},{"author":"Damjan Vukcevic","authorURL":"#","affiliation":"Monash University","affiliationURL":"#","orcidID":"0000-0001-7780-9586"}],"publishedDate":"2023-12-18T00:00:00.000+00:00","citationText":"Pullin, et al., 2023"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a class="logo" href="../../index.html">
<img src="../../resources/rlogo.png" alt="Logo"/>
</a>
<a href="../../index.html" class="title">The R Journal</a>
</div>
<div class="nav-right">
<a href="../../index.html">Home</a>
<a href="../../issues/2024-2">Current</a>
<a href="../../issues.html">Issues</a>
<a href="../../news.html">News</a>
<a href="../../submissions.html">Submit</a>
<a href="../../editors.html">Editorial board</a>
<a href="../../articles.xml">
<i class="fa fa-rss" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Statistical Models for Repeated Categorical Ratings: The R Package rater</h1>

<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>A common problem in many disciplines is the need to assign a set of items into categories or classes with known labels. This is often done by one or more expert raters, or sometimes by an automated process. If these assignments or ‘ratings’ are difficult to make accurately, a common tactic is to repeat them by different raters, or even by the same rater multiple times on different occasions. We present an R package rater, available on CRAN, that implements Bayesian versions of several statistical models for analysis of repeated categorical rating data. Inference is possible for the true underlying (latent) class of each item, as well as the accuracy of each rater. The models are extensions of, and include, the Dawid–Skene model, and we implemented them using the Stan probabilistic programming language. We illustrate the use of rater through a few examples. We also discuss in detail the techniques of marginalisation and conditioning, which are necessary for these models but also apply more generally to other models implemented in Stan.</p></p>
</div>

<div class="d-byline">
  Jeffrey M. Pullin  (University of Melbourne)
  
,   Lyle C. Gurrin  (University of Melbourne)
  
,   Damjan Vukcevic  (Monash University)
  
<br/>2023-12-18
</div>

<div class="d-article">
<div class="d-contents d-contents-float">
<nav class="l-text toc figcaption" id="TOC">
<h3>Contents</h3>
<ul>
<li><a href="#sec:intro" id="toc-sec:intro"><span class="toc-section-number">1</span> Introduction</a></li>
<li><a href="#sec:data-formats" id="toc-sec:data-formats"><span class="toc-section-number">2</span> Data formats</a></li>
<li><a href="#sec:existing-approaches" id="toc-sec:existing-approaches"><span class="toc-section-number">3</span> Existing approaches</a></li>
<li><a href="#sec:models" id="toc-sec:models"><span class="toc-section-number">4</span> Models</a>
<ul>
<li><a href="#sec:dawid-skene" id="toc-sec:dawid-skene"><span class="toc-section-number">4.1</span> Dawid–Skene model</a></li>
<li><a href="#sec:hyperparameters" id="toc-sec:hyperparameters"><span class="toc-section-number">4.2</span> Hyper-parameters for the Dawid–Skene model</a></li>
<li><a href="#hierarchical-dawidskene-model" id="toc-hierarchical-dawidskene-model"><span class="toc-section-number">4.3</span> Hierarchical Dawid–Skene model</a></li>
<li><a href="#class-conditional-dawidskene" id="toc-class-conditional-dawidskene"><span class="toc-section-number">4.4</span> Class-conditional Dawid–Skene</a></li>
<li><a href="#sec:homogenous" id="toc-sec:homogenous"><span class="toc-section-number">4.5</span> Homogeneous Dawid–Skene model</a></li>
<li><a href="#relationships-between-the-models" id="toc-relationships-between-the-models"><span class="toc-section-number">4.6</span> Relationships between the models</a></li>
<li><a href="#sec:relationships-existing-models" id="toc-sec:relationships-existing-models"><span class="toc-section-number">4.7</span> Relationships to existing models</a></li>
<li><a href="#sec:relationships-existing-packages" id="toc-sec:relationships-existing-packages"><span class="toc-section-number">4.8</span> Relationships to existing packages</a></li>
</ul></li>
<li><a href="#sec:implementation" id="toc-sec:implementation"><span class="toc-section-number">5</span> Implementation details</a>
<ul>
<li><a href="#sec:marginalisation" id="toc-sec:marginalisation"><span class="toc-section-number">5.1</span> Marginalisation</a></li>
<li><a href="#sec:conditioning" id="toc-sec:conditioning"><span class="toc-section-number">5.2</span> Inference for the true class via conditioning</a></li>
<li><a href="#data-summarisation" id="toc-data-summarisation"><span class="toc-section-number">5.3</span> Data summarisation</a></li>
</ul></li>
<li><a href="#sec:rao-blackwellization" id="toc-sec:rao-blackwellization"><span class="toc-section-number">6</span> Marginalisation and Rao–Blackwellization</a>
<ul>
<li><a href="#sec:rao-blackwell-connection" id="toc-sec:rao-blackwell-connection"><span class="toc-section-number">6.1</span> Connection with the Rao–Blackwell theorem</a></li>
<li><a href="#marginalisation-in-stan" id="toc-marginalisation-in-stan"><span class="toc-section-number">6.2</span> Marginalisation in Stan</a></li>
<li><a href="#sec:marginalisation-effciency" id="toc-sec:marginalisation-effciency"><span class="toc-section-number">6.3</span> Efficiency of marginalisation</a></li>
</ul></li>
<li><a href="#sec:usage" id="toc-sec:usage"><span class="toc-section-number">7</span> Example usage</a>
<ul>
<li><a href="#fitting-the-model" id="toc-fitting-the-model"><span class="toc-section-number">7.1</span> Fitting the model</a></li>
<li><a href="#inspecting-the-fitted-model" id="toc-inspecting-the-fitted-model"><span class="toc-section-number">7.2</span> Inspecting the fitted model</a></li>
<li><a href="#inspecting-posterior-distributions" id="toc-inspecting-posterior-distributions"><span class="toc-section-number">7.3</span> Inspecting posterior distributions</a></li>
<li><a href="#plots" id="toc-plots"><span class="toc-section-number">7.4</span> Plots</a></li>
<li><a href="#sec:different-models-priors" id="toc-sec:different-models-priors"><span class="toc-section-number">7.5</span> Different models and priors</a></li>
<li><a href="#sec:using-grouped-data" id="toc-sec:using-grouped-data"><span class="toc-section-number">7.6</span> Using grouped data</a></li>
<li><a href="#sec:convergence-diagnostics" id="toc-sec:convergence-diagnostics"><span class="toc-section-number">7.7</span> Convergence diagnostics</a></li>
<li><a href="#sec:model-assessment-comparison" id="toc-sec:model-assessment-comparison"><span class="toc-section-number">7.8</span> Model assessment and comparison</a></li>
</ul></li>
<li><a href="#sec:summary" id="toc-sec:summary"><span class="toc-section-number">8</span> Summary and discussion</a></li>
<li><a href="#computational-details" id="toc-computational-details"><span class="toc-section-number">9</span> Computational details</a></li>
<li><a href="#acknowledgments" id="toc-acknowledgments"><span class="toc-section-number">10</span> Acknowledgments</a>
<ul>
<li><a href="#supplementary-materials" id="toc-supplementary-materials"><span class="toc-section-number">10.1</span> Supplementary materials</a></li>
<li><a href="#cran-packages-used" id="toc-cran-packages-used"><span class="toc-section-number">10.2</span> CRAN packages used</a></li>
<li><a href="#cran-task-views-implied-by-cited-packages" id="toc-cran-task-views-implied-by-cited-packages"><span class="toc-section-number">10.3</span> CRAN Task Views implied by cited packages</a></li>
</ul></li>
</ul>
</nav>
</div>
<h2 data-number="1" id="sec:intro"><span class="header-section-number">1</span> Introduction</h2>
<p>The practice of measuring phenomena by having one or more raters assign to items
one of a set of ratings is common across many fields. For example, in medicine
one or more doctors may classify a diagnostic image as being evidence for one of
several diagnoses, such as types of cancer. This process of rating is often done
by one or more expert raters, but may also be performed by a
large group of non-experts (e.g., in natural language processing
(NLP), where a large number of crowd-sourced raters are used to classify
pieces of text; see <span class="citation" data-cites="passonneau2014">Passonneau and Carpenter (<a href="#ref-passonneau2014" role="doc-biblioref">2014</a>)</span> and <span class="citation" data-cites="ipeirotis2010">Ipeirotis et al. (<a href="#ref-ipeirotis2010" role="doc-biblioref">2010</a>)</span>) or by a machine
or other automated process (e.g., a laboratory diagnostic test, where ‘test’
may refer to either the raters and/or the ratings). Other fields where ratings
occur include astronomy, for classifying astronomical images <span class="citation" data-cites="smyth1994">(<a href="#ref-smyth1994" role="doc-biblioref">Smyth et al. 1994</a>)</span>, and
bioinformatics, for inferring error rates for some bioinformatics algorithms
<span class="citation" data-cites="jakobsdottir2007">(<a href="#ref-jakobsdottir2007" role="doc-biblioref">Jakobsdottir and Weeks 2007</a>)</span>. Indeed, both studies apply the Dawid–Skene model that we
present below and implement in our software package.</p>
<p>The assignment of items to categories, which are variously referred to as gradings,
annotations or labelled data, we will call <em>ratings</em>. Our hope is that
these are an accurate reflection of the true characteristics of the items being
rated. However, this is not guaranteed; all raters and rating systems are fallible.
We would expect some disagreement between the raters, especially from non-experts,
and even expert raters may disagree when there are some items that are more difficult
to rate than others. A typical way of dealing with this problem is to obtain multiple
ratings for each item from different raters, some of whom might rate the item
more than once on separate occasions. By using an aggregate index (e.g., by averaging
the ratings) one can hope to reduce bias and variance when estimating both
population frequencies and individual item categories.</p>
<p>Despite the fact that ratings data of this type are common, there are few software
packages that can be used to fit useful statistical models to them.
To address this, we introduce <a href="https://cran.r-project.org/package=rater">rater</a>, a software package for R <span class="citation" data-cites="base">(<a href="#ref-base" role="doc-biblioref">R Core Team 2021</a>)</span>
designed specifically for such data and available on CRAN. Our package provides the
ability to fit the Dawid–Skene model <span class="citation" data-cites="dawid1979">(<a href="#ref-dawid1979" role="doc-biblioref">Dawid and Skene 1979</a>)</span> and several of its extensions.
The goal is to estimate accurately the underlying true class of each item (to the
extent that this is meaningful and well-defined) as well as to quantify and
characterise the accuracy of each rater. Models of these data should account
for the possibility that accuracy differs between the raters and that the errors
they make might be class-specific. The package accepts data in multiple formats
and provides functions for extracting and visualizing parameter estimates.</p>
<p>While our package implements optimisation-based inference, the default mode of
inference is Bayesian, using Markov Chain Monte Carlo (MCMC). We chose a Bayesian
approach for several reasons. The first is because the standard classical approach,
using an EM algorithm, often gives parameter estimates on the boundary of the
parameter space (i.e., probabilities of exactly 0 or 1), which would typically
be implausible. Examples of this can be seen in the original paper by <span class="citation" data-cites="dawid1979">Dawid and Skene (<a href="#ref-dawid1979" role="doc-biblioref">1979</a>)</span>,
and also in our own results below that compare MCMC and optimisation-based
inference (e.g., compare <a href="#fig:plot-theta">Figure 2</a> and
<a href="#fig:plot-theta-fit2">Figure 4</a>). A second reason is to take advantage of the
convenient and powerful Stan software <span class="citation" data-cites="stanguide">(<a href="#ref-stanguide" role="doc-biblioref">Stan Development Team 2021</a>)</span>, which made implementing our
package considerably easier. Stan is a probabilistic programming language that
implements MCMC and optimisation algorithms for fitting Bayesian models. The
primary inference algorithm used by Stan is the No U-Turn Sampler (NUTS) variant
of the Hamiltonian Monte Carlo (HMC) sampler, which uses gradient information
to guide exploration of the posterior distribution. A third reason is that it allows
users to easily incorporate any prior information when it is available. In this
context, prior information might be information about the quality of the raters
or specific ratings mistakes that are more common.</p>
<p>We illustrate our methods and software using two examples where the raters are
healthcare practitioners: one with anaesthetists rating the health of patients,
and another with dentists assessing the health of teeth. In both examples, we
are interested in obtaining more accurate health assessments by combining all
of the experts’ ratings rather than using the ratings of a single expert,
as well as estimating the accuracy of each expert and characterising the types
of errors they make.</p>
<p>Our paper is organised as follows. <a href="#sec:data-formats">Section 2</a> describes the
various data formats that can be used with the function implemented in
the package. <a href="#sec:models">Section 4</a> introduces the Dawid–Skene model and
several of its extensions. <a href="#sec:implementation">Section 5</a> briefly describes
the implementation and goes into more detail about some of the more
interesting aspects, followed by an exposition of the technique called
Rao–Blackwellization in <a href="#sec:rao-blackwellization">Section 6</a>, which is
connected with many aspects of the implementation. <a href="#sec:usage">Section 7</a>
introduces the user interface of <a href="https://cran.r-project.org/package=rater">rater</a> along with some
examples. <a href="#sec:summary">Section 8</a> concludes with a discussion of the potential
uses of the model.</p>
<h2 data-number="2" id="sec:data-formats"><span class="header-section-number">2</span> Data formats</h2>
<p>There are several different formats in which categorical rating data can
be expressed. We make a distinction between <em>balanced</em> and <em>unbalanced</em>
data. Balanced data have each item rated at least once by each rater.
Unbalanced data contain information on some items that were not rated by
every rater; stated alternatively, there is at least one rater who does
not rate every item.</p>
<p><a href="#tab:data-formats-balanced">Table 1</a> and <a href="#tab:data-formats-unbalanced">Table 2</a>
illustrate three different formats for categorical rating data. Format
(a) has one row per item-rater rating episode. It requires variables
(data fields) to identify the item, the rater and the rating. This
presentation of data is often referred to as the ‘long’ format. It is
possible for a rater to rate a given item multiple times; each such
rating would have its own row in this format. Format (b) has one row per
item, and one column per rater, with each cell being a single rating.
Format (c) aggregates rows (items) that have identical rating patterns
and records a frequency tally. Formats (b) and (c) are examples of
‘wide’ formats. They only make sense for data where each rater rates
each item at most once. Note that the long format will never require
any structural missing values for any of the variables, whereas the wide formats
require the use of a missing value indicator unless they represent data
generated by a balanced design, see <a href="#tab:data-formats-unbalanced">Table 2</a>.
One benefit of the grouped format is that it allows a more
computationally efficient implementation of the likelihood function. See
<a href="#sec:implementation">Section 5</a> for details about how this is implemented in
<a href="https://cran.r-project.org/package=rater">rater</a>.</p>
<div class="layout-chunk" data-layout="l-body">
<table class="kable_wrapper table" style="margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:data-formats-balanced">Table 1: </span>Formats for balanced rating data.
</caption>
<tbody>
<tr>
<td>
<table>
<thead>
<tr>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="3">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
<ol type="a">
<li>Long
</div>
</th>
</tr>
<tr>
<th style="text-align:right;">
Item
</th>
<th style="text-align:right;">
Rater
</th>
<th style="text-align:right;">
Rating
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
4
</td>
</tr>
<tr>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
</tr>
</tbody>
</table></li>
</ol>
</td>
<td>
<table>
<thead>
<tr>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="3">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
<ol start="2" type="a">
<li>Wide
</div>
</th>
</tr>
<tr>
<th style="text-align:right;">
Item
</th>
<th style="text-align:right;">
Rater 1
</th>
<th style="text-align:right;">
Rater 2
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
4
</td>
</tr>
<tr>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
</tr>
</tbody>
</table></li>
</ol>
</td>
<td>
<table>
<thead>
<tr>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="3">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
<ol start="3" type="a">
<li>Wide (grouped)
</div>
</th>
</tr>
<tr>
<th style="text-align:right;">
Rater 1
</th>
<th style="text-align:right;">
Rater 2
</th>
<th style="text-align:right;">
Tally
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
</tr>
</tbody>
</table></li>
</ol>
</td>
</tr>
</tbody>
</table>
</div>
<div class="layout-chunk" data-layout="l-body">
<table class="kable_wrapper table" style="margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:data-formats-unbalanced">Table 2: </span>Formats for unbalanced rating data. Missing values are indicated by ‘NA’.
</caption>
<tbody>
<tr>
<td>
<table>
<thead>
<tr>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="3">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
<ol type="a">
<li>Long
</div>
</th>
</tr>
<tr>
<th style="text-align:right;">
Item
</th>
<th style="text-align:right;">
Rater
</th>
<th style="text-align:right;">
Rating
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
4
</td>
</tr>
<tr>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
4
</td>
</tr>
</tbody>
</table></li>
</ol>
</td>
<td>
<table>
<thead>
<tr>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="3">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
<ol start="2" type="a">
<li>Wide
</div>
</th>
</tr>
<tr>
<th style="text-align:right;">
Item
</th>
<th style="text-align:right;">
Rater 1
</th>
<th style="text-align:right;">
Rater 2
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
1
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
4
</td>
</tr>
<tr>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
NA
</td>
</tr>
<tr>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
NA
</td>
</tr>
<tr>
<td style="text-align:right;">
6
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
4
</td>
</tr>
</tbody>
</table></li>
</ol>
</td>
<td>
<table>
<thead>
<tr>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="3">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
<ol start="3" type="a">
<li>Wide (grouped)
</div>
</th>
</tr>
<tr>
<th style="text-align:right;">
Rater 1
</th>
<th style="text-align:right;">
Rater 2
</th>
<th style="text-align:right;">
Tally
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
1
</td>
</tr>
<tr>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:right;">
3
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
2
</td>
</tr>
<tr>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
4
</td>
<td style="text-align:right;">
1
</td>
</tr>
</tbody>
</table></li>
</ol>
</td>
</tr>
</tbody>
</table>
</div>
<h2 data-number="3" id="sec:existing-approaches"><span class="header-section-number">3</span> Existing approaches</h2>
<p>A typical approach for modelling categorical ratings is to posit an unobserved underlying
true class for each item, i.e., we represent the true class as
a latent variable. One of the first such models was developed by
<span class="citation" data-cites="dawid1979">Dawid and Skene (<a href="#ref-dawid1979" role="doc-biblioref">1979</a>)</span>. This model is still studied and used to this day, and has
served as a foundation for several extensions <span class="citation" data-cites="paun2018">(e.g., <a href="#ref-paun2018" role="doc-biblioref">Paun et al. 2018</a>)</span>. Our
R package, <a href="https://cran.r-project.org/package=rater">rater</a>, is the first one specifically designed to provide
inference for the Dawid–Skene model and its variants.</p>
<p>The Python library <a href="#">pyanno</a> <span class="citation" data-cites="pyanno">(<a href="#ref-pyanno" role="doc-biblioref">Berkes et al. 2011</a>)</span> also implements a Bayesian
version of the Dawid–Skene model in addition to the models described by
<span class="citation" data-cites="rzhetsky2009">Rzhetsky (<a href="#ref-rzhetsky2009" role="doc-biblioref">2009</a>)</span>. However, unlike <a href="https://cran.r-project.org/package=rater">rater</a>, <a href="#">pyanno</a> only
supports parameter estimation via optimisation rather than MCMC; i.e., it will
only compute posterior modes rather than provide samples from the full posterior
distribution. In addition, <a href="#">pyanno</a> does not support variants of the
Dawid–Skene model nor the support for grouped data implemented in
<a href="https://cran.r-project.org/package=rater">rater</a>.</p>
<p>More broadly, many different so-called ‘latent class’ models have been
developed and implemented in software, for a wide diversity of applications
<span class="citation" data-cites="goodman1974">(e.g., <a href="#ref-goodman1974" role="doc-biblioref">Goodman 1974</a>)</span>. A key aspect of the categorical ratings context
is that we assume our categories are known ahead of time, i.e., the possible
values for the latent classes are fixed. The Dawid–Skene model has this
constraint built-in, whereas other latent class models are better tailored to
other scenarios. Nevertheless, many of these models are closely related and
are implemented as R packages.</p>
<p>Of most interest for categorical ratings analysis are the packages
<a href="https://cran.r-project.org/package=poLCA">poLCA</a> <span class="citation" data-cites="linzer2011">(<a href="#ref-linzer2011" role="doc-biblioref">Linzer and Lewis 2011</a>)</span>, <a href="https://cran.r-project.org/package=BayesLCA">BayesLCA</a> <span class="citation" data-cites="white2014">(<a href="#ref-white2014" role="doc-biblioref">White and Murphy 2014</a>)</span> and
<a href="https://cran.r-project.org/package=randomLCA">randomLCA</a> <span class="citation" data-cites="beath2017">(<a href="#ref-beath2017" role="doc-biblioref">Beath 2017</a>)</span>, all of which are capable of fitting
limited versions of the Dawid–Skene model. We explore the relationship
between different latent class models in more detail in <a href="#sec:relationships-existing-models">Section
4.7</a>. Briefly: <a href="https://cran.r-project.org/package=randomLCA">randomLCA</a> can
fit the Dawid–Skene model only when the data uses binary categories,
<a href="https://cran.r-project.org/package=BayesLCA">BayesLCA</a> can only fit the homogeneous Dawid–Skene model, and
<a href="https://cran.r-project.org/package=poLCA">poLCA</a> can fit the Dawid–Skene model with an arbitrary number of
categories but only supports wide data (where raters do not make repeated
ratings on items). Neither <a href="https://cran.r-project.org/package=poLCA">poLCA</a> nor <a href="https://cran.r-project.org/package=randomLCA">randomLCA</a> support
fully Bayesian inference with MCMC, and none of the packages support fitting
variants of the Dawid–Skene model (which are available in <a href="https://cran.r-project.org/package=rater">rater</a>).</p>
<h2 data-number="4" id="sec:models"><span class="header-section-number">4</span> Models</h2>
<p>One of the first statistical models proposed for categorical rating data was
that of <span class="citation" data-cites="dawid1979">Dawid and Skene (<a href="#ref-dawid1979" role="doc-biblioref">1979</a>)</span>. We describe this model below, extended to include
prior distributions to allow for Bayesian inference. Recently, a number of
direct extensions to the Dawid–Skene model have been proposed. We describe
several of these below, most of which are implemented in <a href="https://cran.r-project.org/package=rater">rater</a>. For
ease of exposition, unless otherwise stated, all notation in this section will
assume we are working with balanced data and where each item is rated exactly
once by each rater, one exception is that we present the Dawid–Skene model
for arbitrary data at the end of <a href="#sec:dawid-skene">Section 4.1</a>.</p>
<p>Assume we have <span class="math inline">\(I\)</span> items (for example, images, people, etc.) and <span class="math inline">\(J\)</span> raters,
with each item presumed to belong to one of the <span class="math inline">\(K\)</span> categories: we refer to
this as its ‘true’ category and also as its latent class (since it is unobserved).
Let <span class="math inline">\(y_{i, j} \in \{1, \dots, K\}\)</span> be the rating for item <span class="math inline">\(i\)</span> given by
rater <span class="math inline">\(j\)</span>.</p>
<h3 data-number="4.1" id="sec:dawid-skene"><span class="header-section-number">4.1</span> Dawid–Skene model</h3>
<p>The model has two sets of parameters:</p>
<ul>
<li><p><span class="math inline">\(\pi_k\)</span>: the prevalence of category <span class="math inline">\(k\)</span> in the population from which
the items are sampled. <span class="math inline">\(\pi_k \in (0, 1)\)</span> for
<span class="math inline">\(k \in \{1, \dots, K\}\)</span>, with <span class="math inline">\(\sum_{k = 1}^K \pi_i = 1\)</span>. All classes
have some non-zero probability of occurring. We collect
all the <span class="math inline">\(\pi_k\)</span> parameters together into a vector,
<span class="math inline">\(\pi = (\pi_1, \dots, \pi_K)\)</span>.</p></li>
<li><p><span class="math inline">\(\theta_{j, k, k&#39;}\)</span>: the probability that rater <span class="math inline">\(j\)</span> responds with
class <span class="math inline">\(k&#39;\)</span> when rating an item of true class <span class="math inline">\(k\)</span>. Here, <span class="math inline">\(j \in \{1,  \dots, J\}\)</span> and <span class="math inline">\(k, k&#39; \in \{1, 2, ..., K\}\)</span>. We will refer
to the <span class="math inline">\(K  \times K\)</span> matrix <span class="math inline">\(\theta_{j, k, k&#39;}\)</span> for a given <span class="math inline">\(j\)</span> as the
<em>error matrix</em> for rater <span class="math inline">\(j\)</span>. We represent the <span class="math inline">\(k\)</span>th row of this
matrix as the vector
<span class="math inline">\(\theta_{j, k} = \left(\theta_{j, k, 1}, \theta_{j, k, 2},  \dots, \theta_{j, k, K}\right)\)</span>: this shows how rater <span class="math inline">\(j\)</span>
responds to an item with true class <span class="math inline">\(k\)</span>, by rating it as being in
class <span class="math inline">\(1, 2,  \dots, K\)</span>, according to the respective probabilities.</p></li>
</ul>
<p>The model also has the following set of latent variables:</p>
<ul>
<li><span class="math inline">\(z_i\)</span>: the true class of item <span class="math inline">\(i\)</span>. <span class="math inline">\(z_i \in \{1 ..., K\}\)</span> for <span class="math inline">\(i \in  \{1, ..., I\}\)</span>.</li>
</ul>
<p>Under the Bayesian perspective, the model parameters and latent variables are both
unobserved random variables that define the joint probability
distribution of the model. As such, inference on them is conducted in
the same way. However, we make the distinction here to help clarify
aspects of our implementation later on.</p>
<p>The model is defined by the following distributional assumptions:
<span class="math display">\[
\begin{aligned}
z_i &amp;\sim
   \textrm{Categorical}(\pi),
   \quad \forall i \in \{1, \dots, I\}, \\
y_{i, j} \mid z_i &amp;\sim
   \textrm{Categorical}(\theta_{j, z_i}),
   \quad \forall i \in \{1, \dots, I\},
   j \in \{1, \dots, J\}.
\end{aligned}
\]</span>
In words, the rating for item <span class="math inline">\(i\)</span> given by rater <span class="math inline">\(j\)</span> will follow the
distribution specified by: the error matrix for rater <span class="math inline">\(j\)</span>, but taking
only the row of that matrix that corresponds to the <strong>value</strong> of the
latent variable for item <span class="math inline">\(i\)</span> (the <span class="math inline">\(z_i\)</span>th row).</p>
<p>We now have a fully specified model, which allows use of
likelihood-based methods. The likelihood function for the observed data
(the <span class="math inline">\(y_{i,j}\)</span>s) is
<span class="math display" id="eq:eqnmarginalise">\[
  \Pr(y \mid \theta, \pi) =
  \prod_{i = I}^{I} \left(\sum_{k = 1}^{K}
  \left(\pi_{k} \cdot \prod_{j = 1}^{J} \theta_{j, k, y_{i,j}}\right)\right).
  \tag{1}
\]</span></p>
<p>Note that the unobserved latent variables,
<span class="math inline">\(z = ({z}_{1}, {z}_{2}, \ldots, {z}_{I})\)</span>, do not appear because they
are integrated out (via the sum over the categories <span class="math inline">\(k\)</span>). Often it is
useful to work with the complete data likelihood where the <span class="math inline">\(z\)</span> have
specific values, for example if implementing an EM algorithm (such as
described by <span class="citation" data-cites="dawid1979">Dawid and Skene (<a href="#ref-dawid1979" role="doc-biblioref">1979</a>)</span>). The somewhat simpler likelihood function in
that case is
<span class="math display">\[
  \Pr(y, z \mid \theta, \pi) =
  \prod_{i = I}^{I} \left(\pi_{z_{i}} \cdot
  \prod_{j = 1}^{J} \theta_{j, z_{i}, y_{i,j}}\right).
\]</span>
In our implementation we use the first version of the likelihood, which
marginalises over <span class="math inline">\(z\)</span> (see <a href="#sec:marginalisation">Section 5.1</a>). We do this to avoid
needing to sample from the posterior distribution of <span class="math inline">\(z\)</span> using Stan,
as the HMC sampling algorithm used by Stan requires gradients with respect to
all parameters and gradients cannot be calculated for discrete
parameters such as <span class="math inline">\(z\)</span>. Alternative Bayesian implementations, perhaps using
Gibbs sampling approaches, could work with the second version of the likelihood
and sample values of <span class="math inline">\(z\)</span> directly.</p>
<p>To allow for Bayesian inference, we place weakly informative prior
probability distributions on the parameters:
<span class="math display">\[
\begin{aligned}
\pi            &amp;\sim \textrm{Dirichlet}(\alpha), \\
\theta_{j, k}  &amp;\sim \textrm{Dirichlet}(\beta_k).
\end{aligned}
\]</span>
The hyper-parameters defining these prior probability distributions are:</p>
<ul>
<li><p><span class="math inline">\(\alpha\)</span>: a vector of length <span class="math inline">\(K\)</span> with all elements greater than 0</p></li>
<li><p><span class="math inline">\(\beta\)</span>: a <span class="math inline">\(K \times K\)</span> matrix with all elements greater than 0,
with <span class="math inline">\(\beta_k\)</span> referring to the <span class="math inline">\(k\)</span>th row of this matrix.</p></li>
</ul>
<p>We use the following default values for these hyper-parameters in
<a href="https://cran.r-project.org/package=rater">rater</a>: <span class="math inline">\(\alpha_k = 3\)</span> for
<span class="math inline">\(k \in \{1, \dots, K\}\)</span>, and
<span class="math display">\[
    \beta_{k, k&#39;} =
    \begin{cases}
           N      p          &amp; \textrm{if } k = k&#39; \\
    \dfrac{N (1 - p)}{K - 1} &amp; \textrm{otherwise}
    \end{cases}
    \quad \forall k, k&#39; \in \{1, \dots, K\}.
\]</span>
Here, <span class="math inline">\(N\)</span> corresponds to an approximate pseudocount of hypothetical
prior observations and <span class="math inline">\(p\)</span> an approximate probability of a correct
rating (applied uniformly across all raters and categories). This
affords us the flexibility to centre the prior on a particular assumed
accuracy (via <span class="math inline">\(p\)</span>) as well as tune how informative the prior will be
(via <span class="math inline">\(N\)</span>). In <a href="https://cran.r-project.org/package=rater">rater</a> we set the default values to be <span class="math inline">\(N = 8\)</span> and <span class="math inline">\(p = 0.6\)</span>,
reflecting a weakly held belief that the raters are have a better than coin-flip
chance of choosing the correct class. This should be suitable for many datasets,
where the number of categories is small, for example, less than ten.
This default would, however, be optimistic if the number of categories is very
large, for example, one hundred. In that case it would make sense for the user
to specify a more realistic prior (see <a href="#sec:different-models-priors">Section 7.5</a>).
A derivation of the default prior specification
is shown in <a href="#sec:hyper-parameters">Section 4.2</a>.</p>
<p>We can also write the Dawid–Skene model in notation that does not
assume the data are balanced. Let <span class="math inline">\(I\)</span>, <span class="math inline">\(J\)</span>, <span class="math inline">\(K\)</span> be as above. Let <span class="math inline">\(N\)</span> be
the total number of ratings in the dataset (previously we had the
special case <span class="math inline">\(N = I \cdot J\)</span>). Define the following quantities relating
to the <span class="math inline">\(n\)</span>th rating: it was performed by rater <span class="math inline">\(j_n\)</span>, who rated item
<span class="math inline">\(i_n\)</span>, and the rating itself is <span class="math inline">\(y_n\)</span>. We can now write the Dawid–Skene
model as:
<span class="math display">\[
\begin{aligned}
z_i &amp;\sim
    \textrm{Categorical}(\pi),
    \quad \forall i \in \{1, \dots, I\}, \\
y_n \mid z_{i_n} &amp;\sim
    \textrm{Categorical}
    (\theta_{j_n, z_{i_n}}),
    \quad \forall n \in \{1, \dots, N\}.
\end{aligned}
\]</span></p>
<h3 data-number="4.2" id="sec:hyperparameters"><span class="header-section-number">4.2</span> Hyper-parameters for the Dawid–Skene model</h3>
<h4 class="unnumbered" data-number="4.2.1" id="hyper-parameters-for-the-error-matrices">Hyper-parameters for the error matrices</h4>
<p>The values we proposed for the <span class="math inline">\(\beta\)</span> hyper-parameters are designed to
be flexible and have an approximate intuitive interpretation. The
derivation is as follows.</p>
<p>First, consider the variant of the model where the true latent class
(<span class="math inline">\(z\)</span>) of each item is known; this model is described under ‘Case 1. True
responses are available’ by <span class="citation" data-cites="dawid1979">Dawid and Skene (<a href="#ref-dawid1979" role="doc-biblioref">1979</a>)</span>. Under this model, we can ignore
<span class="math inline">\(\pi\)</span> because the distribution of <span class="math inline">\(y\)</span> depends only on <span class="math inline">\(\theta\)</span>. It is a
categorical distribution with parameters given by specific rows of the
error matrices (as determined by <span class="math inline">\(z\)</span>, which is known) rather than being
a sum over all possible rows (according to the latent class
distribution).</p>
<p>Under this model, the Dirichlet prior is conjugate; we obtain a
Dirichlet posterior for each row of each rater’s error matrix. Consider
an arbitrary rater <span class="math inline">\(j\)</span> and an arbitrary latent class <span class="math inline">\(k&#39;\)</span>. Let
<span class="math inline">\(c = (c_1, c_2, \dots, c_K)\)</span> be the number of times this rater rates an
item as being in each of the <span class="math inline">\(K\)</span> categories when it is of true class
<span class="math inline">\(k&#39;\)</span>. Let <span class="math inline">\(n_{jk&#39;} = \sum_{k = 1}^K c_k\)</span> be the total number of such
ratings. Also, recall that <span class="math inline">\(\theta_{j, k&#39;}\)</span> refers to the vector of
probabilities from the <span class="math inline">\(k&#39;\)</span>th row of the error matrix for rater <span class="math inline">\(j\)</span>. We
have that
<span class="math display">\[
c \sim \textrm{Multinomial}\left(n_{jk&#39;}, \theta_{j, k&#39;}\right),
\]</span>
giving the posterior
<span class="math display">\[
\theta_{j, k&#39;} \mid c
\sim \textrm{Dirichlet}(\beta_{k&#39;} + c).
\]</span>
Under this model, the hyper-parameter vector <span class="math inline">\(\beta_{k&#39;}\)</span> has the same
influence on the posterior as the vector of counts <span class="math inline">\(c\)</span>. We can therefore
interpret the hyper-parameters as ‘pseudocounts’. This gives a
convenient way to understand and set the values of the hyper-parameters.
Our choices were as follows:</p>
<ul>
<li><p>We constrain their sum to be a given value
<span class="math inline">\(N = \sum_{k = 1}^K \beta_{k&#39;,  k}\)</span>, which we interpret as a pseudo-sample size (of hypothetical
ratings from rater <span class="math inline">\(j\)</span> of items with true class <span class="math inline">\(k&#39;\)</span>).</p></li>
<li><p>We then set <span class="math inline">\(\beta_{k&#39;, k&#39;}\)</span> to a desired value so that it reflects
a specific assumed accuracy. In particular, the prior mean of
<span class="math inline">\(\theta_{k&#39;,  k&#39;}\)</span>, the probability of a correct rating, will be
<span class="math inline">\(\beta_{k&#39;, k&#39;} / N\)</span>. Let this be equal to <span class="math inline">\(p\)</span>, an assumed average
prior accuracy.</p></li>
<li><p>Finally, in the absence of any information about the different
categories, it is reasonable to treat them as exchangeable and thus
assume that errors in the ratings are on average uniform across
categories. This implies that the values of all of the other
hyper-parameters (<span class="math inline">\(\beta_{k&#39;,  k}\)</span> where <span class="math inline">\(k \neq k&#39;\)</span>) are equal and need to be set to a value
to meet the constraint that the sum of the hyper-parameters in the
vector <span class="math inline">\(\beta_{k&#39;}\)</span> is <span class="math inline">\(N\)</span>.</p></li>
<li><p>These choices imply the hyper-parameter values described in
<a href="#sec:dawid-skene">Section 4.1</a>.</p></li>
</ul>
<p>In the Dawid–Skene model, the latent classes (<span class="math inline">\(z\)</span>) are of course not
known. Therefore, we do not have a direct interpretation of the <span class="math inline">\(\beta\)</span>
hyper-parameters as pseudocounts. However, we can treat them
approximately as such.</p>
<p>We chose <span class="math inline">\(N = 8\)</span> and <span class="math inline">\(p = 0.6\)</span> as default values based on obtaining
reasonably good performance in simulations across a diverse range of
scenarios (data not shown).</p>
<h4 class="unnumbered" data-number="4.2.2" id="relationship-to-previous-work">Relationship to previous work</h4>
<p>The <em>Stan User’s Guide</em> <span class="citation" data-cites="stanguide">(<a href="#ref-stanguide" role="doc-biblioref">Stan Development Team 2021 7.4</a>)</span> suggests the following choice of
values for <span class="math inline">\(\beta\)</span>:
<span class="math display">\[
  \beta_{k&#39;, k} =
    \begin{cases}
      2.5 \times K &amp; \textrm{if } k&#39; = k \\
      1 &amp; \textrm{otherwise}
    \end{cases},
  \quad \forall k&#39;, k \in \{1, \dots, K\}.
\]</span>
It is interesting to compare this to our suggested values, above. By
equating the two, we can re-write the Stan choice in terms of <span class="math inline">\(p\)</span> and
<span class="math inline">\(N\)</span>:
<span class="math display">\[
\begin{aligned}
  p &amp;= 2.5 \times K / N  \\
  N &amp;= 3.5 \times K - 1.
\end{aligned}
\]</span>
We see that the Stan default is to use an increasingly informative prior
(<span class="math inline">\(N\)</span> gets larger) as the number of categories (<span class="math inline">\(K\)</span>) increases. The
assumed average prior accuracy also varies based on <span class="math inline">\(K\)</span>. For example:</p>
<ul>
<li><p>For <span class="math inline">\(K = 2\)</span> categories, <span class="math inline">\(p = 0.83\)</span> and <span class="math inline">\(N = 6\)</span>.</p></li>
<li><p>For <span class="math inline">\(K = 5\)</span> categories, <span class="math inline">\(p = 0.76\)</span> and <span class="math inline">\(N = 16.5\)</span>.</p></li>
<li><p>As <span class="math inline">\(K\)</span> increases, <span class="math inline">\(p\)</span> slowly reduces, with a limiting value of
<span class="math inline">\(2.5 / 3.5 = 0.71\)</span>.</p></li>
</ul>
<p>Our default values of <span class="math inline">\(p = 0.6\)</span> and <span class="math inline">\(N = 8\)</span> give a prior that is less
informative and less optimistic about the accuracy of the raters.</p>
<p>In practice we have experienced issues when fitting models via
optimisation when the non-diagonal entries of <span class="math inline">\(\beta\)</span> are less than 1
(i.e., when <span class="math inline">\(\beta_{k&#39;, k} &lt; 1\)</span> for some <span class="math inline">\(k&#39; \neq k\)</span>). The default
hyper-parameters selected in <a href="https://cran.r-project.org/package=rater">rater</a> ensure that when
there are only a few possible categories—specifically, when
<span class="math inline">\(K \leqslant 4\)</span>—then this does not occur. When <span class="math inline">\(K\)</span> is larger than this
and the default hyper-parameters give rise to non-diagonal entries
smaller than 1, <a href="https://cran.r-project.org/package=rater">rater</a> will report a useful warning message.
This only arises when using optimisation; it does not arise when using MCMC.
Users who wish to use optimisation with a large number of categories can
specify different hyper-parameters to avoid the warning message.</p>
<h4 class="unnumbered" data-number="4.2.3" id="hyper-parameters-for-the-prevalence">Hyper-parameters for the prevalence</h4>
<p>The <span class="math inline">\(\alpha\)</span> hyper-parameters define the prior for the prevalence parameters <span class="math inline">\(\pi\)</span>.
All of these are essentially nuisance parameters and not of direct interest,
and would typically have little influence on the inference for <span class="math inline">\(\theta\)</span> or <span class="math inline">\(z\)</span>.
The inclusion in the model of a large amount of prior information on the values
of the prevalence parameters would clearly influence the posterior distributions
of the corresponding population class frequencies. We would, however, expect
inferences about the accuracy of raters (whether this is assumed to be class-specific
or not) to depend largely on the number of ratings captured by the dataset,
not the prevalence parameters themselves. We have, therefore, not explored varying
<span class="math inline">\(\alpha\)</span> and have simply set them to the same values as suggested in the
<em>Stan User’s Guide</em> <span class="citation" data-cites="stanguide">(<a href="#ref-stanguide" role="doc-biblioref">Stan Development Team 2021</a>)</span>: <span class="math inline">\(\alpha_k = 3\)</span> for all <span class="math inline">\(k\)</span>.</p>
<h3 data-number="4.3" id="hierarchical-dawidskene-model"><span class="header-section-number">4.3</span> Hierarchical Dawid–Skene model</h3>
<p><span class="citation" data-cites="paun2018">Paun et al. (<a href="#ref-paun2018" role="doc-biblioref">2018</a>)</span> introduce a ‘Hierarchical Dawid–Skene’ model that extends the
original one by replacing the Dirichlet prior on <span class="math inline">\(\theta_{j,k}\)</span> (and the
associated hyper-parameter <span class="math inline">\(\beta\)</span>) with a hierarchical specification
that allows for partial pooling, which allows the sharing of information about
raters’ performance across rater-specific parameters. It requires two
<span class="math inline">\(K \times K\)</span> matrices of parameters, <span class="math inline">\(\mu\)</span> and <span class="math inline">\(\sigma\)</span>, with
<span class="math display">\[
\begin{aligned}
\mu_{k, k&#39;} &amp;\sim \begin{cases}
    \textrm{Normal}(2, 1), &amp; k = k&#39; \\
    \textrm{Normal}(0, 1), &amp; k \neq k&#39; \\
    \end{cases}
    \quad \forall k, \ k&#39; \in \{1, \dots, K\} \\
\sigma_{k, k&#39;} &amp;\sim \textrm{Half-Normal}(0, 1)
    \quad \forall k, \ k&#39; \in \{1, \dots, K\}.
\end{aligned}
\]</span></p>
<p>We then have that:
<span class="math display">\[
\gamma_{j, k, k&#39;} \sim \textrm{Normal}(\mu_{k, k&#39;}, \sigma_{k, k&#39;})
    \quad \forall
    j \in \{1, \dots, J\},
    k, k&#39; \in \{1, \dots K\}
\]</span>
which are then normalised into probabilities via the multinomial logit
link function (also known as the <em>softmax</em> function) in order to define
the elements of the error matrices, which are:
<span class="math display">\[
\theta_{j, k, k&#39;}
    = \frac{e^{\gamma_{j, k, k&#39;}}}{\sum_{k&#39;=1}^K e^{\gamma_{j, k, k&#39;}}}.
\]</span>
Other details, such as the distribution of <span class="math inline">\(z\)</span> and <span class="math inline">\(y\)</span>, are defined in
the same way as per the Dawid–Skene model. The implementation in
<a href="https://cran.r-project.org/package=rater">rater</a> differs from the
implementation from <span class="citation" data-cites="paun2018">Paun et al. (<a href="#ref-paun2018" role="doc-biblioref">2018</a>)</span> by modifying the prior parameters to
encode the assumption that the raters are generally accurate.
Specifically, the higher means of the diagonal elements of <span class="math inline">\(\mu\)</span> encode
that the raters are accurate, as after transformation by the softmax
function larger values of <span class="math inline">\(\mu\)</span> will produce higher probabilities in
<span class="math inline">\(\theta\)</span>. The assumption that the raters are accurate is also encoded in
the prior distribution for <span class="math inline">\(\theta\)</span> in the Dawid–Skene model described
above.</p>
<p>One interpretation of the hierarchical model is as a ‘partial pooling’
compromise between the full Dawid–Skene model and the model with only
one rater (see <a href="#sec:homogenous">Section 4.5</a>). This is depicted in
<a href="#fig:relationships">Figure 1</a>. Another interpretation of the model is that it
treats the accuracy of raters given a specific latent class as a random
effect, not a fixed effect as in the Dawid–Skene model. <span class="citation" data-cites="paun2018">Paun et al. (<a href="#ref-paun2018" role="doc-biblioref">2018</a>)</span> show,
via a series of simulations and examples, that this hierarchical model
generally improves the quality of predictions over the original
Dawid–Skene model.</p>
<h3 data-number="4.4" id="class-conditional-dawidskene"><span class="header-section-number">4.4</span> Class-conditional Dawid–Skene</h3>
<p>Another variant of the Dawid–Skene model imposes constraints on the
entries in the error matrices of the raters. Given the presumption of
accurate raters, one natural way to do this is to force all non-diagonal
entries in a given row of the error matrix to take a constant value that
is smaller than the corresponding diagonal entry. Formally the model is
the same as the Dawid–Skene model except that we have:
<span class="math display">\[
    \theta_{j, k, k&#39;} =
        \begin{cases}
            p_{j, k} &amp; \text{if}\ k = k&#39; \\
            \dfrac{1 - p_{j,k}}{(K - 1)} &amp; \text{otherwise.}
        \end{cases}
\]</span>
To make the model fully Bayesian we use the following prior probability
distributions:
<span class="math display">\[
    p_{j, k} \sim \textrm{Beta}(\beta_{1,k}, \beta_{2,k}),
        \quad \forall j \in {1, \dots, J}.
\]</span></p>
<p>In <a href="https://cran.r-project.org/package=rater">rater</a> we set
<span class="math inline">\(\beta_{1, k} = N p\)</span> and <span class="math inline">\(\beta_{2, k} = N(1 - p)\)</span> for all <span class="math inline">\(k\)</span>, where
<span class="math inline">\(N = 8\)</span> and <span class="math inline">\(p = 0.6\)</span> as in <a href="#sec:dawid-skene">Section 4.1</a>. While the
constraints in this model may lead to imprecise or biased estimation of
parameters for any raters that are substantially inaccurate or even
antagonistic, the reduced number of parameters of this model relative to
the full Dawid–Skene model make it much easier to fit. This model was
referred to as the ‘class-conditional Dawid–Skene model’ by <span class="citation" data-cites="li2014">Li and Yu (<a href="#ref-li2014" role="doc-biblioref">2014</a>)</span>.</p>
<h3 data-number="4.5" id="sec:homogenous"><span class="header-section-number">4.5</span> Homogeneous Dawid–Skene model</h3>
<p>Another related model is the ‘multinomial model’ described by <span class="citation" data-cites="paun2018">Paun et al. (<a href="#ref-paun2018" role="doc-biblioref">2018</a>)</span>,
where the full Dawid–Skene model is constrained so that the accuracy of
all raters is assumed to be identical. The constraint can be
equivalently formulated as the assumption that all of the ratings were
done by a single rater, thus the raters and the sets of rating they
produce are exchangeable. This model can be fitted using
<a href="https://cran.r-project.org/package=rater">rater</a> by simply modifying the data so that only one rater rates every
item present and then fitting the usual Dawid–Skene model.</p>
<h3 data-number="4.6" id="relationships-between-the-models"><span class="header-section-number">4.6</span> Relationships between the models</h3>
<p>All of the models presented in this paper can be directly related to the
original Dawid–Skene model. <a href="#fig:relationships">Figure 1</a> shows the
relationships between the models implemented in the package, which are
coloured blue. The hierarchical model can be seen as a ‘partial pooling’
model where information about the performance of the raters is shared.
The Dawid–Skene model can then be seen as the ‘no pooling’ extreme
where all raters’ error matrices are estimated independently, while the
homogeneous model is the other extreme of complete pooling where one
error matrix is fitted for all of the raters. In addition, the
class-conditional model is a direct restriction of the Dawid–Skene
model where extra constraints are placed on the error matrices.</p>
<h3 data-number="4.7" id="sec:relationships-existing-models"><span class="header-section-number">4.7</span> Relationships to existing models</h3>
<p>The Dawid–Skene model is also closely related to existing latent class
models. Usually, in latent class models, the number of latent classes
can be chosen to maximise a suitable model selection criterion, such as
the BIC, or to make the interpretation of the latent classes easier. In
the Dawid–Skene model, however, the number of latent classes must be
fixed to the number of categories to ensure that the error matrices can
be interpreted in terms of the ability of the raters to rate the
categories that appear in the data. Some specific models to which the
Dawid–Skene model is related are depicted in <a href="#fig:relationships">Figure 1</a>
and are coloured white.</p>
<p>The latent class model implemented in
<a href="https://cran.r-project.org/package=BayesLCA">BayesLCA</a> is equivalent
to the Dawid–Skene model fitted to binary data if the number of classes
(<span class="math inline">\(G\)</span> in the notation of <span class="citation" data-cites="white2014">White and Murphy (<a href="#ref-white2014" role="doc-biblioref">2014</a>)</span>) is set to 2. In that case, each of
the <span class="math inline">\(M\)</span> dimensions of the binary response variable can be interpreted as
a rater, so that <span class="math inline">\(J = M\)</span>.</p>
<p>The Dawid–Skene model is also a special case of the model implemented
in <a href="https://cran.r-project.org/package=poLCA">poLCA</a> <span class="citation" data-cites="linzer2011">(<a href="#ref-linzer2011" role="doc-biblioref">Linzer and Lewis 2011</a>)</span>. The models are equivalent if they have the
same number of latent classes (<span class="math inline">\(R = K\)</span>, where <span class="math inline">\(R\)</span> is the number of latent
classes in the <a href="https://cran.r-project.org/package=poLCA">poLCA</a> model) and if all of
the <span class="math inline">\(J\)</span> categorical variables in the <a href="https://cran.r-project.org/package=poLCA">poLCA</a> model take values in
the set <span class="math inline">\(\{1, \dots K\}\)</span> (that is, <span class="math inline">\(K_j = K\)</span> for all <span class="math inline">\(j \in J\)</span>), so that
each of the <span class="math inline">\(J\)</span> variables can be interpreted as the ratings of a
particular rater.</p>
<p>In addition, the Dawid–Skene model is a special case of the model
implemented in <a href="https://cran.r-project.org/package=randomLCA">randomLCA</a> <span class="citation" data-cites="beath2017">(<a href="#ref-beath2017" role="doc-biblioref">Beath 2017</a>)</span>. When no random effects are
specified and <span class="math inline">\(K = 2\)</span>, the model is equivalent to the Dawid–Skene model with
two classes. While <a href="https://cran.r-project.org/package=randomLCA">randomLCA</a> allows
selecting an arbitrary number of latent classes, it only supports binary
data.</p>
<p>Finally, when <span class="math inline">\(K = 2\)</span> the Dawid–Skene model reduces to the so-called
‘traditional’ or ‘standard’ latent class model as first described by
<span class="citation" data-cites="hui1980">Hui and Walter (<a href="#ref-hui1980" role="doc-biblioref">1980</a>)</span>. See <span class="citation" data-cites="asselineau2018">Asselineau et al. (<a href="#ref-asselineau2018" role="doc-biblioref">2018</a>)</span> for a more modern description.</p>

<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:relationships"></span>
<img src="tikz/figrelationships.png" alt="Relationships between models. Models coloured blue are implemented in rater. DS: Dawid–Skene. LCM: latent class model." width="100%" />
<p class="caption">
Figure 1: Relationships between models. Models coloured blue are implemented in <a href="https://cran.r-project.org/package=rater">rater</a>. DS: Dawid–Skene. LCM: latent class model.
</p>
</div>
</div>
<h3 data-number="4.8" id="sec:relationships-existing-packages"><span class="header-section-number">4.8</span> Relationships to existing packages</h3>
<p>These relationships to existing latent class models means the
functionality implemented in <a href="https://cran.r-project.org/package=rater">rater</a> is similar to that
of existing R packages for fitting these models. The features of
<a href="https://cran.r-project.org/package=rater">rater</a> and existing packages are summarised in <a href="#tab:package-features">Table 3</a>.</p>
<p>Some further details about these relationships:</p>
<ul>
<li><p>The two Bayesian packages implement different methods of inference.
<a href="https://cran.r-project.org/package=rater">rater</a> uses Hamiltonian
Monte Carlo and optimisation of the log-posterior. <a href="https://cran.r-project.org/package=BayesLCA">BayesLCA</a>
implements Gibbs sampling, the expectation–maximization (EM) algorithm and a
Variational Bayesian approach.</p></li>
<li><p>Unlike the other packages, the implementation of the wide data
format in <a href="https://cran.r-project.org/package=randomLCA">randomLCA</a> supports missing data. This feature is not
implemented in <a href="https://cran.r-project.org/package=rater">rater</a> because the long
data format allows both missing data and repeated ratings.</p></li>
</ul>
<p>Overall, the main features that distinguish <a href="https://cran.r-project.org/package=rater">rater</a> from the other
packages are its full support for the Dawid–Skene model and extensions,
and its support for repeated ratings.</p>
<div class="layout-chunk" data-layout="l-body">
<table class="table" style="margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:package-features">Table 3: </span>Features of
<a href="https://cran.r-project.org/package=rater">rater</a> and existing R
packages for fitting latent class models. DS: Dawid–Skene.
</caption>
<thead>
<tr>
<th style="text-align:left;">
Package
</th>
<th style="text-align:left;">
DS model
</th>
<th style="text-align:left;">
Fitting method
</th>
<th style="text-align:left;">
Response type
</th>
<th style="text-align:left;">
Repeated ratings
</th>
<th style="text-align:left;">
Data formats
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
<a href="https://cran.r-project.org/package=rater">rater</a>
</td>
<td style="text-align:left;">
Yes
</td>
<td style="text-align:left;">
Bayesian
</td>
<td style="text-align:left;">
Polytomous
</td>
<td style="text-align:left;">
Yes
</td>
<td style="text-align:left;">
Wide, grouped, long
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="https://cran.r-project.org/package=BayesLCA">BayesLCA</a>
</td>
<td style="text-align:left;">
When <span class="math inline">\(K = 2\)</span>
</td>
<td style="text-align:left;">
Bayesian
</td>
<td style="text-align:left;">
Binary
</td>
<td style="text-align:left;">
No
</td>
<td style="text-align:left;">
Wide, grouped
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="https://cran.r-project.org/package=randomLCA">randomLCA</a>
</td>
<td style="text-align:left;">
When <span class="math inline">\(K = 2\)</span>
</td>
<td style="text-align:left;">
Frequentist
</td>
<td style="text-align:left;">
Binary
</td>
<td style="text-align:left;">
No
</td>
<td style="text-align:left;">
Wide, grouped
</td>
</tr>
<tr>
<td style="text-align:left;">
<a href="https://cran.r-project.org/package=poLCA">poLCA</a>
</td>
<td style="text-align:left;">
Yes
</td>
<td style="text-align:left;">
Frequentist
</td>
<td style="text-align:left;">
Polytomous
</td>
<td style="text-align:left;">
No
</td>
<td style="text-align:left;">
Wide
</td>
</tr>
</tbody>
</table>
</div>
<h2 data-number="5" id="sec:implementation"><span class="header-section-number">5</span> Implementation details</h2>
<p><a href="https://cran.r-project.org/package=rater">rater</a> uses Stan <span class="citation" data-cites="carpenter2017">(<a href="#ref-carpenter2017" role="doc-biblioref">Carpenter et al. 2017</a>)</span> to fit the above models to data.
It therefore supports
both optimisation and Markov chain Monte Carlo (MCMC) sampling, in
particular using the No U-Turn Sampler (NUTS) algorithm
<span class="citation" data-cites="JMLR:v15:hoffman14a">(<a href="#ref-JMLR:v15:hoffman14a" role="doc-biblioref">Hoffman and Gelman 2014</a>)</span>. For most datasets we recommend using NUTS, due
to the fact that it will generate realisations of the full posterior
distribution. Optimisation may be useful, however, for particularly
large datasets, especially if <span class="math inline">\(K\)</span> is large, or if there is a need to fit
a large number of models.</p>
<h3 data-number="5.1" id="sec:marginalisation"><span class="header-section-number">5.1</span> Marginalisation</h3>
<p>The NUTS algorithm relies on computing derivatives of the (log)
posterior distribution with respect to all parameters. It cannot,
therefore, be used when the posterior contains discrete parameters, such
as the latent class in the Dawid–Skene model. To overcome this
difficulty the Stan models implemented in
<a href="https://cran.r-project.org/package=rater">rater</a> use marginalised
forms of the posterior to allow NUTS to be used. In other words, for the
Dawid–Skene model we implement the likelihood
<span class="math inline">\(\Pr(y \mid \theta, \pi)\)</span>, as described earlier in
<a href="#sec:dawid-skene">Section 4.1</a>, together with priors for <span class="math inline">\(\theta\)</span> and <span class="math inline">\(\pi\)</span>.</p>
<p>To avoid any confusion, we stress the fact that our choice to
marginalise over the vector <span class="math inline">\(z\)</span> is purely because it is discrete and we
want to use the NUTS algorithm. It is not due to the fact that the
components of the vector <span class="math inline">\(z\)</span> are inherently latent variables.
Alternative Bayesian implementations could avoid marginalisation and
sample the <span class="math inline">\({z}_{i}\)</span>’s directly (although this may be less efficient,
see <a href="#sec:marginalisation-effciency">Section 6.3</a> for discussion). Similarly, if
the components of <span class="math inline">\(z\)</span> were all continuous, we would be able to sample
them using NUTS. If one or more of the other non-latent parameters were
discrete, then we would need to marginalise over them too.</p>
<h3 data-number="5.2" id="sec:conditioning"><span class="header-section-number">5.2</span> Inference for the true class via conditioning</h3>
<p>Marginalising over <span class="math inline">\(z\)</span> allows us to use NUTS and sample efficiently from
the posterior probability distribution of <span class="math inline">\(\theta\)</span> and <span class="math inline">\(\pi\)</span>. We are
still able to obtain the posterior distribution of the components
<span class="math inline">\({z}_{i}\)</span> of <span class="math inline">\(z\)</span> as follows. For each posterior draw of <span class="math inline">\(\theta\)</span> and
<span class="math inline">\(\pi\)</span>, we calculate the conditional posterior for <span class="math inline">\({z}_{i}\)</span>,
<span class="math inline">\(\Pr({z}_{i} \mid \theta^*, \pi^*, y)\)</span>, where the conditioning is on the
<strong>drawn</strong> values, <span class="math inline">\(\theta^*\)</span> and <span class="math inline">\(\pi^*\)</span>. This can be done using Bayes’
theorem, for example in the Dawid–Skene model,
<span class="math display">\[
\begin{aligned}
   \Pr(z_i = k \mid \theta^*, \pi^*, y)
&amp;= \frac{\Pr(y \mid z_i = k, \theta^*, \pi^*)
         \Pr(z_i = k \mid \theta^*, \pi^*)}
        {\Pr(y \mid \theta^*, \pi^*)} \\
&amp;= \frac{\pi^*_k \prod_{j = 1}^{J}
         \theta^*_{j, k, y_{i,j}}}
        {\sum_{m = 1}^{K} \pi^*_m \prod_{j = 1}^{J} \theta^*_{j, m, y_{i,j}}}.
\end{aligned}
\]</span>
To get a Monte Carlo estimate of the marginal posterior,
<span class="math inline">\(\Pr({z}_{i} \mid y)\)</span>, we take the mean of the above conditional
probabilities across the posteriors draws of <span class="math inline">\(\theta\)</span> and <span class="math inline">\(\pi\)</span>. To
formally justify this step we can use the same argument that justifies
the general MCMC technique of Rao–Blackwellization, see
<a href="#sec:rao-blackwellization">Section 6</a>. In practice, this calculation is
straightforward with Stan because we can write the log posterior in
terms of the (log scale) unnormalised versions of the conditional
probabilities and output them alongside the posterior draws. Then it is
simply a matter of renormalising them, followed by taking the mean
across draws.</p>
<h3 data-number="5.3" id="data-summarisation"><span class="header-section-number">5.3</span> Data summarisation</h3>
<p><a href="https://cran.r-project.org/package=rater">rater</a> implements the
ability to use different forms of the likelihood depending on the format
of the data that is available. For example, for a balanced design the
full likelihood for the Dawid–Skene model is
<span class="math display">\[
\Pr(y \mid \theta, \pi)
= \prod_{i = 1}^{I}
  \left(\sum_{k = 1}^{K}
        \left(\pi_{k} \cdot \prod_{j = 1}^{J}
                            \theta_{j, k, y_{i, j}}\right)\right).
\]</span>
Looking closely, we see that the contribution of any two items with the
same pattern of ratings, that is, the same raters giving the same
ratings, is identical. We can therefore rewrite it as a product over
patterns rather than items,
<span class="math display">\[
\Pr(y \mid \theta, \pi)
= \prod_{l = 1}^{L}
  \left(\sum_{k = 1}^{K}
        \left(\pi_{k} \cdot \prod_{j = 1}^{J}
                            \theta_{j, k, y_{l, j}}\right)\right)^{n_l},
\]</span>
where <span class="math inline">\(L\)</span> is the number of distinct rating ‘patterns’, <span class="math inline">\(n_l\)</span> is the
number of times pattern <span class="math inline">\(l\)</span> occurs and <span class="math inline">\(y_{l, j}\)</span> denotes the rating
given by rater <span class="math inline">\(j\)</span> within pattern <span class="math inline">\(l\)</span>. This corresponds directly to
having the data in the ‘grouped’ format, see <a href="#sec:data-formats">Section 2</a>.</p>
<p>This rewriting is useful for balanced designs where there are many
ratings but few distinct patterns, as it removes the need to iterate
over every rating during computation of the posterior, reducing the time
for model fitting. An example of this type of data are the <code>caries</code> data
presented in <a href="#sec:usage">Section 7</a>.</p>
<p>This technique is not always helpful. For example, if the data has many
missing entries, the number of distinct patterns <span class="math inline">\(L\)</span> may be similar or
equal to than the total number of ratings. Consequently, rewriting to
use the grouped format may not save much computation. For this reason,
the implementation of <a href="https://cran.r-project.org/package=rater">rater</a> does not allow
missing values in grouped data. Note that in the long data format we
simply drop rows that correspond to missing data, a result of which is
that missing data is never explicitly represented in <a href="https://cran.r-project.org/package=rater">rater</a>.</p>
<p>Both the <a href="https://cran.r-project.org/package=randomLCA">randomLCA</a> and <a href="https://cran.r-project.org/package=BayesLCA">BayesLCA</a> packages support grouped
format data input. Of these, CRANpkg{randomLCA} supports missing values in
grouped format data.</p>
<h2 data-number="6" id="sec:rao-blackwellization"><span class="header-section-number">6</span> Marginalisation and Rao–Blackwellization</h2>
<p>The technique of marginalisation for sampling discrete random variables
discussed in <a href="#sec:implementation">Section 5</a> is well known in the Stan
community. We feel, however, that the interpretation and use of the
relevant conditional probabilities and expectations has not been clearly
described before, so we attempt to do so here. We begin with a brief
review of the theory of marginalisation, which is a special case of a
more general technique often referred to as ‘Rao–Blackwellization’. See
<span class="citation" data-cites="mcbook">Owen (<a href="#ref-mcbook" role="doc-biblioref">2013</a>)</span> for an introduction and <span class="citation" data-cites="robert2021raoblackwellization">Robert and Roberts (<a href="#ref-robert2021raoblackwellization" role="doc-biblioref">2021</a>)</span> for a
recent survey of the use of Rao–Blackwellization in MCMC.</p>
<h3 data-number="6.1" id="sec:rao-blackwell-connection"><span class="header-section-number">6.1</span> Connection with the Rao–Blackwell theorem</h3>
<p>Suppose we are interested in estimating <span class="math inline">\(\mu = \mathbb{E}(f(X, Y))\)</span> for
some function <span class="math inline">\(f\)</span> of random variables <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span>. An example of such an
expectation is <span class="math inline">\(\mu = \mathbb{E}(Y)\)</span> where we take <span class="math inline">\(f(x, y) = y\)</span>. If we
can sample from the joint distribution of <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span>, <span class="math inline">\(p_{X, Y}(x, y)\)</span>,
the obvious Monte Carlo estimator of this expectation is
<span class="math display">\[
\hat{\mu} = \frac{1}{n} \sum_{i = 1}^{n} f(x_i, y_i)
\]</span>
where <span class="math inline">\((x_i, y_i)\)</span> are samples from the joint distribution. Now let
<span class="math inline">\(h(x) = \mathbb{E}(f(X, Y) \mid X = x)\)</span>. An alternate conditional
estimator—the so-called <em>Rao–Blackwellized estimator</em>—of <span class="math inline">\(\mu\)</span> is
<span class="math display">\[
\hat{\mu}_{\textrm{cond}} = \frac{1}{n} \sum_{i = 1}^{n}
h(x_i)
\]</span>
where the <span class="math inline">\(x_i\)</span> are sampled from the distribution of <span class="math inline">\(X\)</span>. This estimator
is justified by the fact that,
<span class="math display">\[
\mathbb{E}(h(X)) = \mathbb{E}(\mathbb{E}(f(X, Y) \mid X)) = \mathbb{E}(f(X,Y)) = \mu
\]</span>
For this to be effective we need to be able to efficiently compute
(ideally, analytically) the conditional expectation <span class="math inline">\(h(X)\)</span>. The original
motivation for using this technique was to reduce the variance of Monte
Carlo estimators. This follows from noting that
<span class="math display">\[
\begin{aligned}
    \mathop{\mathrm{var}}(f(X, Y))
    &amp;= \mathbb{E}(\mathop{\mathrm{var}}(f(X, Y) \mid X)) + \mathop{\mathrm{var}}(\mathbb{E}(f(X, Y) \mid X)) \\
    &amp;= \mathbb{E}(\mathop{\mathrm{var}}(f(X, Y) \mid X)) + \mathop{\mathrm{var}}(h(X)),
\end{aligned}
\]</span>
and since the first term on the right-hand side is non-negative, we have
<span class="math inline">\(\texttt{(}f(X, Y)) \geqslant \texttt{(}h(X))\)</span>.</p>
<p>From this result it follows that, if the draws <span class="math inline">\(x_i\)</span> are independent
samples, then
<span class="math display">\[
\mathop{\mathrm{var}}(\hat{\mu}_{\textrm{cond}}) &lt; \mathop{\mathrm{var}}(\hat{\mu})
\]</span>
for all functions <span class="math inline">\(f(X, Y)\)</span>. In the applications we consider here,
however, <span class="math inline">\(x_i\)</span> will be drawn using MCMC and therefore draws will not be
independent. In this case it is possible to construct models and
functions <span class="math inline">\(f(\cdot)\)</span> for which conditioning can increase the variance of
the estimator <span class="citation" data-cites="geyer1995">(<a href="#ref-geyer1995" role="doc-biblioref">Geyer 1995</a>)</span>.</p>
<p>The name ‘Rao–Blackwellization’ arose due to the similarity of the
above argument to that used in the proof of the Rao–Blackwell theorem
<span class="citation" data-cites="blackwell1947">(<a href="#ref-blackwell1947" role="doc-biblioref">Blackwell 1947</a>)</span>, which states that given an estimator <span class="math inline">\(\hat{\mu}\)</span> of a
parameter <span class="math inline">\(\mu\)</span>, the conditional expectation of <span class="math inline">\(\hat{\mu}\)</span> given a
sufficient statistic for <span class="math inline">\(\mu\)</span> is potentially more efficient and
certainly no less efficient than <span class="math inline">\(\hat{\mu}\)</span> as an estimator of <span class="math inline">\(\mu\)</span>.
The Rao–Blackwellized estimators presented above do not, however, make
any use of sufficiency and do not have the same optimality guarantees
that the Rao–Blackwell theorem provides, making the name less than apt.
Following <span class="citation" data-cites="geyer1995">Geyer (<a href="#ref-geyer1995" role="doc-biblioref">1995</a>)</span>, we prefer to think of the estimators presented
above as simply being formed from averaging conditional expectations.</p>
<h3 data-number="6.2" id="marginalisation-in-stan"><span class="header-section-number">6.2</span> Marginalisation in Stan</h3>
<p>The motivation for marginalisation in Stan is to enable estimation of
<span class="math inline">\(\mu = \mathbb{E}(f(X, Y))\)</span> without having to sample from <span class="math inline">\((X, Y)\)</span> if
either <span class="math inline">\(X\)</span> or <span class="math inline">\(Y\)</span> is discrete. Suppose that <span class="math inline">\(Y\)</span> is discrete and <span class="math inline">\(X\)</span>
continuous. To compute a Monte Carlo estimate of <span class="math inline">\(\mathbb{E}(f(X, Y))\)</span>
using Stan we carry out four steps.</p>
<p>First, marginalise out <span class="math inline">\(Y\)</span> from <span class="math inline">\(p_{X, Y}(x, y)\)</span> to give <span class="math inline">\(p_{X}(x)\)</span>.
(See Equation <a href="#eq:eqnmarginalise">(1)</a> in <a href="#sec:dawid-skene">Section 4.1</a> for how this is
done in the Dawid–Skene model.) This marginal distribution, which only
contains continuous parameters, should then be implemented as a Stan
model. Second, using Stan as usual, sample from the distribution of the
continuous parameters <span class="math inline">\(X\)</span> to give Monte Carlo samples <span class="math inline">\(\{x_i\}\)</span>. Given
only the samples from the distribution of <span class="math inline">\(X\)</span>, we can estimate
<span class="math inline">\(\mathbb{E}(f(X, Y))\)</span> using the Rao–Blackwellized estimator described
in the previous section. Doing so requires us to evaluate
<span class="math inline">\(h(X) = \mathbb{E}(f(X, Y) \mid X)\)</span> for all functions <span class="math inline">\(f(\cdot)\)</span> in
which we may be interested; this is the third step of the process.
Finally, as the fourth step, we can evaluate <span class="math inline">\(h(x)\)</span> for each of the
Monte Carlo draws <span class="math inline">\(x_i\)</span> and estimate <span class="math inline">\(\mathbb{E}(f(X, Y))\)</span> by
<span class="math display">\[
\frac{1}{n} \sum_{i = 1}^n h(x_i).
\]</span></p>
<h4 class="unnumbered" data-number="6.2.1" id="evaluating-the-conditional-expectation">Evaluating the conditional expectation</h4>
<p>The third step outlined above may appear to require substantial
mathematical manipulation. In practice, however, we can use the discrete
nature of the latent class to simplify the calculation. Specifically,
for any function <span class="math inline">\(f(x, y)\)</span> we have
<span class="math display">\[
h(X)
    = \mathbb{E}(f(X, Y) \mid X)
    = \sum_{k} f(X, k) \Pr(Y = k \mid X),
\]</span>
where we sum over the values in the latent discrete parameters. An
essential part of this formula is the probability of the discrete
parameters conditional on the continuous parameters,
<span class="math inline">\(\Pr(Y = k \mid X)\)</span>. This quantity can be derived easily through Bayes’
theorem or can be encoded as part of the marginalised Stan model; see
<a href="#sec:conditioning">Section 5.2</a> or the next section for how this is done in the
case of the Dawid–Skene model.</p>
<p>In the Dawid–Skene model, and many other models with discrete
variables, the discrete variables only take values in a finite set. Some
models however may have discrete parameters which take countably
infinite values, such as Poisson distributions. In this case,
marginalisation is still possible but will generally involve
approximating infinite sums, both when marginalising out the discrete
random variables and calculating the expectations (as in the equation
above), which may be computationally expensive.</p>
<p>Furthermore, for the cases where <span class="math inline">\(f(x, y)\)</span> is a function of only <span class="math inline">\(x\)</span> or
<span class="math inline">\(y\)</span>, the general formula simplifies further. Firstly, when
<span class="math inline">\(f(x, y) = f(x)\)</span> we have
<span class="math display">\[
h(X)
    = \sum_{k} f(X) \Pr(Y = k \mid X)
    = f(X) \sum_{k} \Pr(Y = k \mid X)
    = f(X).
\]</span>
This means that we can estimate <span class="math inline">\(\mathbb{E}(f(X))\)</span> with the standard,
seemingly unconditional, estimator:
<span class="math display">\[
\sum_{i = 1}^n f(x_i).
\]</span>
Even after marginalisation, computing expectations of functions of the
continuous parameters can be performed as if no marginalisation had
taken place.</p>
<p>Secondly, when <span class="math inline">\(f(x, y) = f(y)\)</span> we have that
<span class="math display">\[
h(X) = \sum_{k} f(k) \Pr(Y = k \mid X).
\]</span></p>
<p>An important special case of this result is when
<span class="math inline">\(f(x, y) = \mathop{\mathrm{\mathbf{1}}}(y = k)\)</span>, where
<span class="math inline">\(\mathop{\mathrm{\mathbf{1}}}\)</span> is the indicator function. This is
important because it allows us to recover the probability mass function
of the discrete random variable <span class="math inline">\(Y\)</span>, since
<span class="math inline">\(\mathbb{E}(f(X, Y)) = \mathbb{E}(\mathop{\mathrm{\mathbf{1}}}(Y = k)) = \Pr(Y = k)\)</span>.
In this case we have
<span class="math display">\[
h(X)
    = \sum_{k} \mathop{\mathrm{\mathbf{1}}}(y = k) \Pr(Y = k \mid X)
    = \Pr(Y = k \mid X).
\]</span>
We therefore estimate <span class="math inline">\(\Pr(Y = k)\)</span> with:
<span class="math display">\[
\frac{1}{n} \sum_{i = 1}^n
    \Pr(Y = k \mid X = x_i).
\]</span>
Again, we stress that our ability to do these calculations relies upon
being able to easily compute <span class="math inline">\(\Pr(Y = k \mid X = x_i)\)</span> for each of the
Monte Carlo draws <span class="math inline">\(x_i\)</span>.</p>
<h4 class="unnumbered" data-number="6.2.2" id="estimating-the-conditional-probability-of-the-true-class">Estimating the conditional probability of the true class</h4>
<p>For the categorical rating problem (using the Dawid–Skene model), the
discrete random variable of interest for each item <span class="math inline">\(i\)</span> is the true class
of the item, <span class="math inline">\({z}_{i}\)</span>. We use the technique from the previous section
to calculate the posterior probability of the true class, as described
in <a href="#sec:conditioning">Section 5.2</a>. In this case, the discrete variable is
<span class="math inline">\(z_i\)</span> (taking on the role that <span class="math inline">\(Y\)</span> played in the previous section), the
continuous variables are <span class="math inline">\(\theta\)</span> and <span class="math inline">\(\pi\)</span> (which together take on the
role that <span class="math inline">\(X\)</span> played in the previous section), and all probability
calculations are always conditional on the data (the ratings, <span class="math inline">\(y\)</span>).</p>
<h3 data-number="6.3" id="sec:marginalisation-effciency"><span class="header-section-number">6.3</span> Efficiency of marginalisation</h3>
<p>It is not immediately obvious whether marginalisation is more or less
efficient than techniques that actually realise a value for the discrete
random variable at each iteration, such as Gibbs sampling. Marginalising
can be viewed as a form of Rao–Blackwellization, a general MCMC
technique designed to reduce the variability of Monte Carlo estimators.
This link strongly suggests that marginalisation is more efficient than
using discrete random variables. Unfortunately, limitations of the
theory of Rao–Blackwellization (see
<a href="#sec:rao-blackwell-connection">Section 6.1</a>) and the difficulty of
theoretically comparing different sampling algorithms, such as Gibbs
sampling and NUTS, means that it is unclear whether marginalisation will
always be computationally superior for a given problem.</p>
<p>However, in practice marginalisation does seem to improve convergence at
least for non-conjugate models. For example, <span class="citation" data-cites="yackulic2020">Yackulic et al. (<a href="#ref-yackulic2020" role="doc-biblioref">2020</a>)</span> show that for
the Cormack–Jolly–Seber model marginalisation greatly speeds up
inference in JAGS, BUGS and WinBUGS. They also demonstrate that the
marginalised model implemented in Stan is orders of magnitude more
efficient than the marginalised models in the other languages. These
results show that marginalisation has the potential to speed up classic
algorithms and also allows the use of more efficient gradient-based
algorithms, such as NUTS, for problems involving discrete random
variables.</p>
<p>A recent similar exploration of marginalisation for some simple mixture
models and the Dawid–Skene model, using JAGS and Stan, suggests that
the software implementation has a greater impact on efficiency than the
choice of whether or not marginalisation is used <span class="citation" data-cites="zhang2022">(<a href="#ref-zhang2022" role="doc-biblioref">Zhang et al. 2022</a>)</span>. In their
comparisons, Stan usually achieved the best performance (and note that
Stan requires marginalisation of discrete parameters).</p>
<p>For these reasons, we recommend that practitioners using models with
discrete parameters consider marginalisation if more efficiency is
desired, and implement their models using Stan rather than JAGS.</p>
<h2 data-number="7" id="sec:usage"><span class="header-section-number">7</span> Example usage</h2>
<p>To demonstrate <a href="https://cran.r-project.org/package=rater">rater</a> we use two example datasets.</p>
<p>The first dataset is taken from the original paper introducing the Dawid–Skene
model <span class="citation" data-cites="dawid1979">(<a href="#ref-dawid1979" role="doc-biblioref">Dawid and Skene 1979</a>)</span>. The data consist of ratings, on a 4-point scale, made
by five anaesthetists of patients’ pre-operative health. The ratings were based
on the anaesthetists assessments of a standard form completed for all of the
patients. There are 45 patients (items) and four anaesthetists (raters) in
total. The first anaesthetist assessed the forms a total of three times, spaced
several weeks apart. The other anaesthetists each assessed the forms once. As
in the Dawid–Skene paper, we will not seek to model the effect of time on the
ratings of the first anaesthetist.</p>
<p>First we load the <a href="https://cran.r-project.org/package=rater">rater</a> package:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://jeffreypullin.github.io/rater/'>rater</a></span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>This will display information about altering options used by Stan to
make best use of computational resources on your machine. We can then load and
look at a snippet of the anaesthesia data, which is included in the package.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"anesthesia"</span>, package <span class='op'>=</span> <span class='st'>"rater"</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>anesthesia</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>  item rater rating
1    1     1      1
2    1     1      1
3    1     1      1
4    1     2      1
5    1     3      1
6    1     4      1</code></pre>
</div>
<p>These data are arranged in ‘long’ format where each row corresponds to a single
rating. The first column gives the index of the item that was rated, the second
the index of the rater that rated the item and the third the actual rating that
was given. This layout of the data supports raters rating the same item
multiple times, as happens in the dataset. It is also the most convenient from
the perspective of fitting models but may not always be the optimal way to
store or represent categorical rating data; see
<a href="#sec:using-grouped-data">Section 7.6</a> for an example using the ‘grouped’ data
format.</p>
<h3 data-number="7.1" id="fitting-the-model"><span class="header-section-number">7.1</span> Fitting the model</h3>
<p>We can fit the Dawid–Skene model using MCMC by running the command:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>fit_1</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/rater.html'>rater</a></span><span class='op'>(</span><span class='va'>anesthesia</span>, <span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/models.html'>dawid_skene</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>This command will print the running output from Stan, providing an
indication of the progress of the sampler (for brevity, we have not shown this
output here).
To fit the model via optimisation, we set the <code>method</code> argument to
<code>"optim"</code>:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>fit_2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/rater.html'>rater</a></span><span class='op'>(</span><span class='va'>anesthesia</span>, <span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/models.html'>dawid_skene</a></span><span class='op'>(</span><span class='op'>)</span>, method <span class='op'>=</span> <span class='st'>"optim"</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>The second argument of the <code>rater()</code> function specifies the model to use.
We have implemented this similarly to the <code>family</code> argument in
<code>glm()</code>, which can be passed as either a function or as a character
string. The above examples pass the model as a function. We could have instead
passed it as a string, for example:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>fit_2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/rater.html'>rater</a></span><span class='op'>(</span><span class='va'>anesthesia</span>, <span class='st'>"dawid_skene"</span>, method <span class='op'>=</span> <span class='st'>"optim"</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>Either version will fit the Dawid–Skene model using the default choice of
prior. The benefit of passing the model as a function is that it allows you to
change the prior, see <a href="#sec:different-models-priors">Section 7.5</a>.</p>
<h3 data-number="7.2" id="inspecting-the-fitted-model"><span class="header-section-number">7.2</span> Inspecting the fitted model</h3>
<p><a href="https://cran.r-project.org/package=rater">rater</a> includes several ways to inspect the output of fitted models.
These are summarised in <a href="#tab:inspecting-fitted-models">Table 4</a>, and we
illustrate many of them here. Firstly, we can generate a text summary:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span><span class='op'>(</span><span class='va'>fit_1</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>Model:
Bayesian Dawid and Skene Model 

Prior parameters:

alpha: default
beta: default

Fitting method: MCMC

pi/theta samples:
               mean   5%  95% Rhat ess_bulk
pi[1]          0.38 0.28 0.48    1  8714.19
pi[2]          0.41 0.30 0.51    1  8357.55
pi[3]          0.14 0.07 0.23    1  8202.91
pi[4]          0.08 0.03 0.14    1  7701.45
theta[1, 1, 1] 0.86 0.79 0.93    1  8332.86
theta[1, 1, 2] 0.10 0.05 0.17    1  8015.78
theta[1, 1, 3] 0.02 0.00 0.05    1  5726.94
theta[1, 1, 4] 0.02 0.00 0.05    1  6230.45
# ... with 76 more rows

z:
     MAP Pr(z = 1) Pr(z = 2) Pr(z = 3) Pr(z = 4)
z[1]   1      1.00      0.00      0.00      0.00
z[2]   3      0.00      0.00      0.98      0.02
z[3]   2      0.40      0.60      0.00      0.00
z[4]   2      0.01      0.99      0.00      0.00
z[5]   2      0.00      1.00      0.00      0.00
z[6]   2      0.00      1.00      0.00      0.00
z[7]   1      1.00      0.00      0.00      0.00
z[8]   3      0.00      0.00      1.00      0.00
# ... with 37 more items</code></pre>
</div>
<p>This function will show information about which model has been fitted and the
values of the prior parameters that were used. The displayed text also contains
information about the parameter estimates and posterior distributions, and the
convergence of the sampler.</p>
<div class="layout-chunk" data-layout="l-body">
<table class="table" style="margin-left: auto; margin-right: auto;">
<caption>
<span id="tab:inspecting-fitted-models">Table 4: </span>Methods to inspect fitted models and what values they return.
</caption>
<thead>
<tr>
<th style="text-align:left;">
Function
</th>
<th style="text-align:left;">
MCMC mode
</th>
<th style="text-align:left;">
Optimisation mode
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
summary()
</td>
<td style="text-align:left;">
Basic information about the fitted model
</td>
<td style="text-align:left;">
Basic information about the fitted model
</td>
</tr>
<tr>
<td style="text-align:left;">
point_estimate()
</td>
<td style="text-align:left;">
Posterior means for <span class="math inline">\(\pi\)</span> and <span class="math inline">\(\theta\)</span>, posterior modes for <span class="math inline">\(z\)</span>
</td>
<td style="text-align:left;">
Posterior modes for all quantities
</td>
</tr>
<tr>
<td style="text-align:left;">
posterior_interval()
</td>
<td style="text-align:left;">
Credible intervals for <span class="math inline">\(\pi\)</span> and <span class="math inline">\(\theta\)</span>
</td>
<td style="text-align:left;">
N/A
</td>
</tr>
<tr>
<td style="text-align:left;">
posterior_samples()
</td>
<td style="text-align:left;">
MCMC draws for <span class="math inline">\(\pi\)</span> and <span class="math inline">\(\theta\)</span>
</td>
<td style="text-align:left;">
N/A
</td>
</tr>
<tr>
<td style="text-align:left;">
mcmc_diagnostics()
</td>
<td style="text-align:left;">
MCMC convergence diagnostics for <span class="math inline">\(\pi\)</span> and <span class="math inline">\(\theta\)</span>
</td>
<td style="text-align:left;">
N/A
</td>
</tr>
<tr>
<td style="text-align:left;">
class_probabilities()
</td>
<td style="text-align:left;">
Posterior distribution for <span class="math inline">\(z\)</span>
</td>
<td style="text-align:left;">
Posterior distribution for <span class="math inline">\(z\)</span> conditional on the
posterior modes for <span class="math inline">\(\pi\)</span> and <span class="math inline">\(\theta\)</span>
</td>
</tr>
</tbody>
</table>
</div>
<p>We can extract point estimates for any of the parameters or the latent classes
via the <code>point_estimate()</code> function. For example, the following will
return the latent class with the highest posterior probability (i.e., the
posterior modes) for each item:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/point_estimate.html'>point_estimate</a></span><span class='op'>(</span><span class='va'>fit_1</span>, <span class='st'>"z"</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>$z
 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 
 1  3  2  2  2  2  1  3  2  2  4  2  1  2  1  1  1  1  2  2  2  2  2 
24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 
 2  1  1  2  1  1  1  1  3  1  2  2  3  2  2  3  1  1  1  2  1  2 </code></pre>
</div>
<p>The following will return the posterior means of the prevalence probabilities:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/point_estimate.html'>point_estimate</a></span><span class='op'>(</span><span class='va'>fit_1</span>, <span class='st'>"pi"</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>$pi
[1] 0.37571033 0.40602060 0.14320458 0.07506449</code></pre>
</div>
<p>From the outputs above, we can see that the model has inferred that most
patients have health that should be classified as category 1 or 2 (roughly 40%
each), and categories 3 and 4 being rarer (about 14% and 7% respectively).
Based on the point estimates, there are examples of patients from each category
in the dataset.</p>
<p>The function call <code>point_estimate(fit1, "theta")</code> will return posterior
means for the parameters in the error matrices (not shown here, for brevity).
When used with optimisation fits, <code>point_estimate()</code> will return
posterior modes rather than posterior means for the parameters.</p>
<h3 data-number="7.3" id="inspecting-posterior-distributions"><span class="header-section-number">7.3</span> Inspecting posterior distributions</h3>
<p>To represent uncertainty, we need to look beyond point estimates. The function
<code>posterior_interval()</code> will return credible intervals for the parameters.
For example, 80% credible intervals for the elements of the error matrices:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'><a href='https://mc-stan.org/rstantools/reference/posterior_interval.html'>posterior_interval</a></span><span class='op'>(</span><span class='va'>fit_1</span>, <span class='fl'>0.8</span>, <span class='st'>"theta"</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>                       10%        90%
theta[1, 1, 1] 0.806247440 0.91694947
theta[1, 1, 2] 0.058041621 0.15236819
theta[1, 1, 3] 0.001711161 0.03843961
theta[1, 1, 4] 0.001826953 0.03809251
theta[1, 2, 1] 0.024090185 0.10372077
theta[1, 2, 2] 0.794646454 0.90858998</code></pre>
</div>
<p>The function <code>posterior_samples()</code> will return the actual MCMC draws. For
example (here just illustrating the draws of the <span class="math inline">\(\pi\)</span> parameter):</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/posterior_samples.html'>posterior_samples</a></span><span class='op'>(</span><span class='va'>fit_1</span>, <span class='st'>"pi"</span><span class='op'>)</span><span class='op'>$</span><span class='va'>pi</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>          
iterations      [,1]      [,2]      [,3]       [,4]
      [1,] 0.2302726 0.5159451 0.1546796 0.09910258
      [2,] 0.2561003 0.4406078 0.2147696 0.08852236
      [3,] 0.3508047 0.3865941 0.2230788 0.03952239
      [4,] 0.4447041 0.4053563 0.1227548 0.02718484
      [5,] 0.4137727 0.3221264 0.1759903 0.08811065
      [6,] 0.3771586 0.3577296 0.1653739 0.09973792</code></pre>
</div>
<p>Neither of the above will work for optimisation fits, which are limited to
point estimates only. For the latent classes, we can produce the probability
distribution as follows:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/class_probabilities.html'>class_probabilities</a></span><span class='op'>(</span><span class='va'>fit_1</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>   
            [,1]         [,2]         [,3]         [,4]
  1 9.999994e-01 1.294502e-07 4.119892e-08 4.517351e-07
  2 1.555216e-07 2.390349e-05 9.756994e-01 2.427657e-02
  3 4.031597e-01 5.961319e-01 1.011515e-04 6.071909e-04
  4 5.336614e-03 9.939651e-01 4.260605e-04 2.722057e-04
  5 2.463205e-07 9.999588e-01 3.799143e-05 3.011338e-06
  6 1.654940e-06 9.991634e-01 8.119895e-04 2.300085e-05</code></pre>
</div>
<p>This works for both MCMC and optimisation fits. For the former, the output is
the posterior distribution on <span class="math inline">\(z\)</span>, while for the latter it is the distribution
conditional on the point estimates of the parameters (<span class="math inline">\(\pi\)</span> and <span class="math inline">\(\theta\)</span>).</p>
<h3 data-number="7.4" id="plots"><span class="header-section-number">7.4</span> Plots</h3>
<p>It is often easier to interpret model outputs visually.
Using <a href="https://cran.r-project.org/package=rater">rater</a>, we can plot the parameter estimates and
distribution of latent classes.</p>
<p>The following command will visualise the posterior means of the error rate
parameters, with the output shown in <a href="#fig:plot-theta">Figure 2</a>:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>fit_1</span>, <span class='st'>"raters"</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:plot-theta"></span>
<img src="RJ-2023-064_files/figure-html5/plot-theta-1.png" alt="Visual representation of the inferred parameters in the error matrices ($\theta$) for the Dawid--Skene model fitted via MCMC to the anaesthesia dataset. The values shown are posterior means, with each cell shaded on a gradient from white (value close to 0) to dark blue (value close to 1)." width="528" />
<p class="caption">
Figure 2: Visual representation of the inferred parameters in the error matrices (<span class="math inline">\(\theta\)</span>) for the Dawid–Skene model fitted via MCMC to the anaesthesia dataset. The values shown are posterior means, with each cell shaded on a gradient from white (value close to 0) to dark blue (value close to 1).
</p>
</div>
</div>
<p>We can see high values along the diagonals of these matrices, which indicates
that each of the 5 anaesthetists is inferred as being fairly accurate at rating
pre-operative health. Looking at the non-diagonal entries, we can see that
typically the most common errors are 1-point differences on the rating scale.</p>
<p>The following command visualises the latent class probabilities, with the
output shown in <a href="#fig:plot-z">Figure 3</a>:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>fit_1</span>, <span class='st'>"latent_class"</span>, item_index <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>2</span>, <span class='fl'>3</span>, <span class='fl'>12</span>, <span class='fl'>36</span>, <span class='fl'>38</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:plot-z"></span>
<img src="RJ-2023-064_files/figure-html5/plot-z-1.png" alt="Visualisation of the inferred probability of each latent class, for a selected subset of items, for the Dawid--Skene model fitted via MCMC to the anaesthesia dataset." width="384" />
<p class="caption">
Figure 3: Visualisation of the inferred probability of each latent class, for a selected subset of items, for the Dawid–Skene model fitted via MCMC to the anaesthesia dataset.
</p>
</div>
</div>
<p>For the purpose of illustration, for this plot we have selected the 5 patients
with the greatest uncertainty in their pre-operative health (latent class). The
other 40 patients all have almost no posterior uncertainty for their
pre-operative health. Thus, suppose we wished to use the model to infer the
pre-operative health by combining all of the anaesthetists’ ratings, we can do
so confidently for all but a handful of patients.</p>
<p>The same output for the models fitted via
optimisation rather than MCMC (using <code>fit_2</code> instead of <code>fit_1</code>) are
shown in <a href="#fig:plot-theta-fit2">Figure 4</a> and <a href="#fig:plot-z-fit2">Figure 5</a>. We can
see that this estimation method leads to the same broad conclusions, however the
optimisation-based estimates have considerably less uncertainty: they are
typically closer to 0 or 1. This behaviour reflects the fact that
optimisation-based inference will usually not capture the full uncertainty and
will lead to overconfident estimates. Thus, we recommend using MCMC (which we
have set as the default).</p>


<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:plot-theta-fit2"></span>
<img src="RJ-2023-064_files/figure-html5/plot-theta-fit2-1.png" alt="Visual representation of the inferred parameters in the error matrices ($\theta$) for the Dawid--Skene model fitted via optimisation to the anaesthesia dataset.  Compare with Figure 2, which used MCMC instead of optimisation." width="528" />
<p class="caption">
Figure 4: Visual representation of the inferred parameters in the error matrices (<span class="math inline">\(\theta\)</span>) for the Dawid–Skene model fitted via optimisation to the anaesthesia dataset. Compare with <a href="#fig:plot-theta">Figure 2</a>, which used MCMC instead of optimisation.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:plot-z-fit2"></span>
<img src="RJ-2023-064_files/figure-html5/plot-z-fit2-1.png" alt="Visualisation of the inferred probability of each latent class, for a selected subset of items, for the Dawid--Skene model fitted via optimisation to the anaesthesia dataset.  Compare with Figure 3, which used MCMC instead of optimisation." width="384" />
<p class="caption">
Figure 5: Visualisation of the inferred probability of each latent class, for a selected subset of items, for the Dawid–Skene model fitted via optimisation to the anaesthesia dataset. Compare with <a href="#fig:plot-z">Figure 3</a>, which used MCMC instead of optimisation.
</p>
</div>
</div>
<p>While it is typically of less interest, it is also possible to visualise the
prevalence estimates, together with credible intervals, using the following
command (see output in <a href="#fig:plot-pi">Figure 6</a>):</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>fit_1</span>, <span class='st'>"prevalence"</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:plot-pi"></span>
<img src="RJ-2023-064_files/figure-html5/plot-pi-1.png" alt="Visualisation of the inferred population prevalence parameters, with 90\% credible intervals, for the Dawid--Skene model fitted to the anaesthesia dataset." width="528" />
<p class="caption">
Figure 6: Visualisation of the inferred population prevalence parameters, with 90% credible intervals, for the Dawid–Skene model fitted to the anaesthesia dataset.
</p>
</div>
</div>
<h3 data-number="7.5" id="sec:different-models-priors"><span class="header-section-number">7.5</span> Different models and priors</h3>
<p>The second argument to the <code>rater()</code> function specifies what model to use,
including details of the prior if not using the default one. This gives a
unified place to specify both the model and prior.</p>
<p>For example, this is how to set a different prior using the Dawid–Skene model:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>diff_alpha_fit</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/rater.html'>rater</a></span><span class='op'>(</span><span class='va'>anesthesia</span>, <span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/models.html'>dawid_skene</a></span><span class='op'>(</span>alpha <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>10</span>, <span class='fl'>4</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>When specifying the <span class="math inline">\(\beta\)</span> hyper-parameters for the Dawid–Skene model, the
user can either specify a single matrix or a 3-dimensional array. If only a matrix
specified, it will be interpreted as the hyper-parameter values for all raters.
This is useful in the common situation where the overall quality of the raters
is known but there is no information on the quality of any specific rater.
When a 3-dimensional array is passed, it is taken to specify the hyper-parameters
for each of the raters (i.e., a separate matrix for each rater). This is
useful when prior information about the quality of specific raters is
available, for example when some ‘raters’ are diagnostic tests with known
performance characteristics.</p>
<p>This is how to use the class-conditional Dawid–Skene model (with default prior):</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>diff_model_fit</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/rater.html'>rater</a></span><span class='op'>(</span><span class='va'>anesthesia</span>, <span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/models.html'>class_conditional_dawid_skene</a></span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>Compared with the Dawid–Skene model, the latter uses error matrices with the
constraint that all off-diagonal entries must be the same. We can visualise
the <span class="math inline">\(\theta\)</span> parameter using the following command (output in
<a href="#fig:plot-theta-diff-model">Figure 7</a>) and see that all off-diagonal elements are
indeed equal:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>diff_model_fit</span>, <span class='st'>"theta"</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:plot-theta-diff-model"></span>
<img src="RJ-2023-064_files/figure-html5/plot-theta-diff-model-1.png" alt="Visual representation of the inferred parameters in the error matrices ($\theta$) for the class-conditional Dawid--Skene model fitted via MCMC to the anaesthesia dataset.  Compare with Figure 2, which used the standard Dawid--Skene model." width="528" />
<p class="caption">
Figure 7: Visual representation of the inferred parameters in the error matrices (<span class="math inline">\(\theta\)</span>) for the class-conditional Dawid–Skene model fitted via MCMC to the anaesthesia dataset. Compare with <a href="#fig:plot-theta">Figure 2</a>, which used the standard Dawid–Skene model.
</p>
</div>
</div>
<h3 data-number="7.6" id="sec:using-grouped-data"><span class="header-section-number">7.6</span> Using grouped data</h3>
<p>The second example dataset is taken from <span class="citation" data-cites="espeland1989">Espeland and Handelman (<a href="#ref-espeland1989" role="doc-biblioref">1989</a>)</span>. It consists of
3,689 binary ratings, made by 5 dentists, of whether a given tooth was healthy
or had caries/cavities. The ratings were performed using X-ray
only, which was thought to be more error-prone than visual/tactile assessment
of each tooth (see <span class="citation" data-cites="handelman1986">Handelman et al. (<a href="#ref-handelman1986" role="doc-biblioref">1986</a>)</span> for more information and a description
of the wider dataset from which these binary ratings were taken).</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"caries"</span>, package <span class='op'>=</span> <span class='st'>"rater"</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>caries</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>  rater_1 rater_2 rater_3 rater_4 rater_5    n
1       1       1       1       1       1 1880
2       1       1       1       1       2  789
3       1       1       1       2       1   43
4       1       1       1       2       2   75
5       1       1       2       1       1   23
6       1       1       2       1       2   63</code></pre>
</div>
<p>This is an example of ‘grouped’ data. Each row represents a particular ratings
‘pattern’, with the final column being a tally that records how many instances
of that pattern appear in the data. <a href="https://cran.r-project.org/package=rater">rater</a> accepts data in either
‘long’, ‘wide’ or ‘grouped’ format. The ‘long’ format is the default because
it can represent data with repeated ratings. When available the ‘grouped’
format can greatly speed up the computation of certain models and is convenient
for datasets that are already
recorded in that format. The ‘wide’ format doesn’t provide any computational
advantages but is a common format for data without repeated ratings and so is
provided for convenience.</p>
<p>Here’s how we fit the model to grouped data (note the <code>data_format</code>
argument):</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>fit3</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/rater.html'>rater</a></span><span class='op'>(</span><span class='va'>caries</span>, <span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/models.html'>dawid_skene</a></span><span class='op'>(</span><span class='op'>)</span>, data_format <span class='op'>=</span> <span class='st'>"grouped"</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>From there, we can inspect the model fit in the same way as before. For
example (output shown in <a href="#fig:plot-theta-caries">Figure 8</a>):</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>fit3</span>, <span class='st'>"raters"</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:plot-theta-caries"></span>
<img src="RJ-2023-064_files/figure-html5/plot-theta-caries-1.png" alt="Visual representation of the inferred parameters in the error matrices ($\theta$) for the Dawid--Skene model fitted via MCMC to the caries dataset." width="528" />
<p class="caption">
Figure 8: Visual representation of the inferred parameters in the error matrices (<span class="math inline">\(\theta\)</span>) for the Dawid–Skene model fitted via MCMC to the caries dataset.
</p>
</div>
</div>
<p>The first 4 dentists are highly accurate at diagnosing healthy teeth (rating <span class="math inline">\(= 1\)</span>), but less so at diagnosing caries (rating <span class="math inline">\(= 2\)</span>). In contrast, the 5th
dentist was much better at diagnosing caries than healthy teeth.</p>
<h3 data-number="7.7" id="sec:convergence-diagnostics"><span class="header-section-number">7.7</span> Convergence diagnostics</h3>
<p>A key part of applied Bayesian statistics is assessing whether the
MCMC sampler has converged on the posterior distribution. To summarise the
convergence of a model fit using MCMC, <a href="https://cran.r-project.org/package=rater">rater</a> provides the
<code>mcmc_diagnostics()</code> function.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/mcmc_diagnostics.html'>mcmc_diagnostics</a></span><span class='op'>(</span><span class='va'>fit_1</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>                   Rhat ess_bulk
pi[1]          1.000309 8714.189
pi[2]          1.000377 8357.551
pi[3]          1.000275 8202.908
pi[4]          1.001472 7701.453
theta[1, 1, 1] 1.001018 8332.856
theta[1, 1, 2] 1.001445 8015.780</code></pre>
</div>
<p>This function calculates and displays the <span class="math inline">\(\hat{R}\)</span> statistic (<code>Rhat</code>) and bulk
effective sample size (<code>ess_bulk</code>) for all of the <span class="math inline">\(\pi\)</span> and <span class="math inline">\(\theta\)</span>
parameters. Users can then check for the convergence of specific parameters by
applying standard rules, such as considering that convergence has been reached
for a specific parameter if it’s <span class="math inline">\(\hat{R} &lt; 1.01\)</span> <span class="citation" data-cites="vehtari2021">(<a href="#ref-vehtari2021" role="doc-biblioref">Vehtari et al. 2021</a>)</span>.</p>
<p>Users wishing to calculate other metrics, or produce convergence visualisations
such as trace plots, can either extract the underlying Stan model from the
<a href="https://cran.r-project.org/package=rater">rater</a> fit using the <code>get_stanmodel()</code> function, or convert the
fit into a <code>mcmc.list</code> from the <a href="https://cran.r-project.org/package=coda">coda</a> package using the
<code>as_mcmc.list()</code> function. Functions in the <a href="https://cran.r-project.org/package=rstan">rstan</a> and
<a href="https://cran.r-project.org/package=coda">coda</a> packages will then allow visualisation based assessment of
convergence and the calculation of other, more advanced, diagnostics.</p>
<p>For example, the following code uses <a href="https://cran.r-project.org/package=rstan">rstan</a> to draw the trace plot
(shown in <a href="#fig:plot-rstan-traceplot">Figure 9</a>) for one of the parameters from
the Dawid–Skene model fitted to the anaesthesia data. We can see that the
four chains show good mixing. The trace plots for the other parameters
(not shown here) are similar, indicating that the MCMC sampling procedure has
converged.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>stan_fit</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://jeffreypullin.github.io/rater/reference/get_stanfit.html'>get_stanfit</a></span><span class='op'>(</span><span class='va'>fit_1</span><span class='op'>)</span></span>
<span><span class='fu'>rstan</span><span class='fu'>::</span><span class='fu'><a href='https://mc-stan.org/rstan/reference/stanfit-method-traceplot.html'>traceplot</a></span><span class='op'>(</span><span class='va'>stan_fit</span>, pars <span class='op'>=</span> <span class='st'>"pi[1]"</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:plot-rstan-traceplot"></span>
<img src="RJ-2023-064_files/figure-html5/plot-rstan-traceplot-1.png" alt="A trace plot for the $\pi_1$ parameter from the Dawid--Skene model fitted via MCMC to the anaesthesia dataset." width="480" />
<p class="caption">
Figure 9: A trace plot for the <span class="math inline">\(\pi_1\)</span> parameter from the Dawid–Skene model fitted via MCMC to the anaesthesia dataset.
</p>
</div>
</div>
<h3 data-number="7.8" id="sec:model-assessment-comparison"><span class="header-section-number">7.8</span> Model assessment and comparison</h3>
<p>Finally, <a href="https://cran.r-project.org/package=rater">rater</a> provides facilities to assess and compare fitted
models. First, it provides the <code>posterior_predict()</code> function to simulate from
the posterior predictive distributions of all the models implemented in
<a href="https://cran.r-project.org/package=rater">rater</a>. The simulated data can then be used, for example, to perform
posterior predictive checks. These checks compare the data
simulated from the fitted model to the observed data. A number of datasets
are simulated from the posterior predictive distribution and a summary statistic
is calculated for each dataset. The same summary statistic is calculated on the
observed dataset and is compared to the distribution of simulated statistics.
A large discrepancy between the simulated statistics and the observed statistic
indicates possible model misspecification.</p>
<p>To illustrate this functionality, we consider assessing the fit of the
Dawid–Skene model to the anaesthesia dataset. In this example, we simulate
1,000 datasets from the posterior predictive distribution, and use the
proportion of the ratings that are class 2 as a statistic to summarise each
dataset.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>class2prop</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='op'>)</span> <span class='op'>{</span></span>
<span>  <span class='va'>simdata</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://mc-stan.org/rstantools/reference/posterior_predict.html'>posterior_predict</a></span><span class='op'>(</span><span class='va'>fit_1</span>, <span class='va'>anesthesia</span><span class='op'>[</span>, <span class='fl'>1</span><span class='op'>:</span><span class='fl'>2</span><span class='op'>]</span><span class='op'>)</span></span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>simdata</span><span class='op'>$</span><span class='va'>rating</span> <span class='op'>==</span> <span class='fl'>2</span><span class='op'>)</span> <span class='op'>/</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>simdata</span><span class='op'>)</span></span>
<span><span class='op'>}</span></span>
<span><span class='va'>ppc_statistics</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>replicate</a></span><span class='op'>(</span><span class='fl'>1000</span>, <span class='fu'>class2prop</span><span class='op'>(</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>ppc_statistics</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>[1] 0.3079365 0.3333333 0.4158730 0.2666667 0.2793651 0.3587302</code></pre>
</div>
<p>We can then graphically compare the distribution of these statistics to the
value of the same statistic applied to the anaesthesia dataset. The following
commands will create such a plot (shown in <a href="#fig:plot-ppcs">Figure 10</a>):</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggplot.html'>ggplot</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>prop_class_two <span class='op'>=</span> <span class='va'>ppc_statistics</span><span class='op'>)</span>, <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/aes.html'>aes</a></span><span class='op'>(</span><span class='va'>prop_class_two</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span></span>
<span>  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_histogram.html'>geom_histogram</a></span><span class='op'>(</span>binwidth <span class='op'>=</span> <span class='fl'>0.01</span>, fill <span class='op'>=</span> <span class='st'>"steelblue"</span>, colour <span class='op'>=</span> <span class='st'>"black"</span><span class='op'>)</span> <span class='op'>+</span></span>
<span>  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/geom_abline.html'>geom_vline</a></span><span class='op'>(</span>xintercept <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>anesthesia</span><span class='op'>$</span><span class='va'>rating</span> <span class='op'>==</span> <span class='fl'>2</span><span class='op'>)</span> <span class='op'>/</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>anesthesia</span><span class='op'>)</span>,</span>
<span>             colour <span class='op'>=</span> <span class='st'>"black"</span>, linewidth <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span> <span class='op'>+</span></span>
<span>  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/ggtheme.html'>theme_bw</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>+</span></span>
<span>  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/coord_cartesian.html'>coord_cartesian</a></span><span class='op'>(</span>xlim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.1</span>, <span class='fl'>0.6</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span></span>
<span>  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>labs</a></span><span class='op'>(</span></span>
<span>    x <span class='op'>=</span> <span class='st'>"Proportion of class 2 ratings"</span>,</span>
<span>    y <span class='op'>=</span> <span class='st'>"Count"</span></span>
<span>  <span class='op'>)</span></span></code></pre>
</div>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:plot-ppcs"></span>
<img src="RJ-2023-064_files/figure-html5/plot-ppcs-1.png" alt="An example of a posterior predictive check for the Dawid--Skene model fitted via MCMC to the anaesthesia dataset.  Shown is a histogram of 1,000 simulated datasets from the posterior predictive distribution, each summarised by the proportion of ratings that are class 2.  The vertical black line shows the proportion of ratings that are class 2 in the original dataset, for comparison against the histogram." width="528" />
<p class="caption">
Figure 10: An example of a posterior predictive check for the Dawid–Skene model fitted via MCMC to the anaesthesia dataset. Shown is a histogram of 1,000 simulated datasets from the posterior predictive distribution, each summarised by the proportion of ratings that are class 2. The vertical black line shows the proportion of ratings that are class 2 in the original dataset, for comparison against the histogram.
</p>
</div>
</div>
<p>The statistic calculated on the anaesthesia dataset lies towards the centre of
the distribution of statistics calculated from datasets drawn from the
posterior predictive distribution. This is consistent with the model fitting
the data well. (This is an illustrative example only. A more comprehensive
set of such comparisons should be conducted before concluding that the model
adequately describes the data.)</p>
<p>Second, <a href="https://cran.r-project.org/package=rater">rater</a> implements the functions <code>loo()</code> and <code>waic()</code>
to provide model comparison metrics using the <a href="https://cran.r-project.org/package=loo">loo</a> package. The
function <code>loo()</code> calculates an approximation to leave-out-one
cross-validation, a measure of model performance using Pareto Smoothed
Importance Sampling <span class="citation" data-cites="vehtari2017">(<a href="#ref-vehtari2017" role="doc-biblioref">Vehtari et al. 2017</a>)</span>. The function <code>waic()</code> calculates
measures related to the Widely Applicable Information Criterion
<span class="citation" data-cites="watanabe2013">(<a href="#ref-watanabe2013" role="doc-biblioref">Watanabe 2013</a>)</span>. Because <a href="https://cran.r-project.org/package=rater">rater</a> fits models in a Bayesian
framework, information criteria such as the AIC and BIC are not implemented.</p>
<p>In the context of <a href="https://cran.r-project.org/package=rater">rater</a>, model comparison is only possible between
the different implemented models, not between the same model with different
parameter values or included covariates, as is more commonly the case. For
this reason, considerations of data size and what is known about the
characteristics of the raters should also be taken into account when making
such choices, in addition to the output of <code>loo()</code> and <code>waic()</code>.</p>
<p>To briefly illustrate some of this functionality, we compare the standard
Dawid–Skene model to the class-conditional one.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://mc-stan.org/loo/reference/loo.html'>loo</a></span><span class='op'>(</span><span class='va'>fit_1</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>
Computed from 4000 by 45 log-likelihood matrix

         Estimate   SE
elpd_loo   -234.0 17.0
p_loo        20.2  2.7
looic       468.0 34.0
------
Monte Carlo SE of elpd_loo is 0.1.

Pareto k diagnostic values:
                         Count Pct.    Min. n_eff
(-Inf, 0.5]   (good)     39    86.7%   677       
 (0.5, 0.7]   (ok)        6    13.3%   407       
   (0.7, 1]   (bad)       0     0.0%   &lt;NA&gt;      
   (1, Inf)   (very bad)  0     0.0%   &lt;NA&gt;      

All Pareto k estimates are ok (k &lt; 0.7).
See help(&#39;pareto-k-diagnostic&#39;) for details.</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://mc-stan.org/loo/reference/loo.html'>loo</a></span><span class='op'>(</span><span class='va'>diff_model_fit</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>
Computed from 4000 by 45 log-likelihood matrix

         Estimate   SE
elpd_loo   -245.7 18.1
p_loo        10.2  1.2
looic       491.3 36.1
------
Monte Carlo SE of elpd_loo is 0.1.

All Pareto k estimates are good (k &lt; 0.5).
See help(&#39;pareto-k-diagnostic&#39;) for details.</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://mc-stan.org/loo/reference/loo_compare.html'>loo_compare</a></span><span class='op'>(</span><span class='fu'><a href='https://mc-stan.org/loo/reference/loo.html'>loo</a></span><span class='op'>(</span><span class='va'>fit_1</span><span class='op'>)</span>, <span class='fu'><a href='https://mc-stan.org/loo/reference/loo.html'>loo</a></span><span class='op'>(</span><span class='va'>diff_model_fit</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>       elpd_diff se_diff
model1   0.0       0.0  
model2 -11.7       3.2  </code></pre>
</div>
<p>From these results, we see that both models fit the data well, with the
class-conditional model being slightly preferred.</p>
<h2 data-number="8" id="sec:summary"><span class="header-section-number">8</span> Summary and discussion</h2>
<p>Rating procedures, in which items are sorted into categories, are subject to both
classification error and uncertainty when the categories themselves are defined
subjectively. Data that results from these types of tasks require a proper
statistical treatment if: (1) the population frequencies of the categories are
to be estimated with low bias, (2) items are to be assigned to these
categories reliably, and (3) the agreement of raters is to be assessed.
To date there have been few options for practitioners seeking software that
implements the required range of statistical models. The R package described in
this paper, <a href="https://cran.r-project.org/package=rater">rater</a>, provides the ability to fit
Bayesian versions of a variety of statistical models to categorical rating
data, with a range of tools to extract and visualise parameter estimates.</p>
<p>The statistical models that we have presented are based on the Dawid–Skene
model and recent modifications of it including extensions, such as the
hierarchical Dawid–Skene model where the rater-specific parameters are assumed
to be drawn from a population distribution, and simplifications that, for
example, assume exchangeable raters with identical performance characteristics
(homogeneous Dawid–Skene), or homogeneity criteria where classification
probabilities are identical regardless of the true, underlying category of an
item (class-conditional Dawid–Skene).</p>
<p>We provided: (1) an explanation of the type of data formats that categorical
ratings are recorded in, (2) a description of the construction and implementation of the
package, (3) a comparison of our package to other existing packages for fitting
latent class models in R, (4) an introduction to the user interface,
and (5) worked examples of real-world data analysis using <a href="https://cran.r-project.org/package=rater">rater</a>.</p>
<p>We devoted an entire section to motivating, deriving and explaining the use of
a marginalised version of the joint posterior probability distribution to
remove dependence on the unknown value of the true, underlying rating category
of an item. This is necessary because the No-U-Turn Sampler, the main MCMC algorithm
used in Stan, relies on computing derivatives with respect to all parameters that
must, therefore, be continuously-valued. The technique involves the use of
conditional expectation and is a special case of a more general technique of
conditioning or ‘Rao–Blackwellization’, the process of transforming an
estimator using the Rao–Blackwell theorem to improve its efficiency.</p>
<p>Our package was developed with the classification of medical images in mind,
where there may be a large number of images but
typically only a small number of possible categories and a limited number of
expert raters. The techniques proposed based on extensions of the Dawid–Skene model
readily extend to scenarios where the datasets are much larger, or where the
raters behave erratically (cannot be relied on to rate correctly more
frequently than they rate incorrectly) or are antagonistic (deliberately
attempt to allocate items to a category not favoured by the majority of
raters).</p>
<p>One possible extension of the Dawid–Skene model is to add item- and
rater-specific covariates. These covariates would encode the difficulty of
items and the expertise (or lack thereof) of the raters. This extension is
particularly attractive as it would partially alleviate the strong assumption
of independence conditional on the latent class, replacing it with the weaker
assumption of independence conditional on the latent class and the covariates.
Unfortunately, these types of models would contain many more parameters than
the original Dawid–Skene model making them difficult to fit, especially when
used with relatively small datasets common in the context of rating data in
medical research. Therefore, future methodological research is needed before
these models can be included in the <a href="https://cran.r-project.org/package=rater">rater</a> package. Latent class
models with covariates can be fitted, although only in a frequentist framework,
using the R package <a href="https://cran.r-project.org/package=poLCA">poLCA</a>.</p>
<h2 data-number="9" id="computational-details"><span class="header-section-number">9</span> Computational details</h2>
<p>The results in this paper were obtained using
R 4.2.2 <span class="citation" data-cites="base">(<a href="#ref-base" role="doc-biblioref">R Core Team 2021</a>)</span>
and the following packages:
<a href="https://cran.r-project.org/package=coda">coda</a> 0.19.4 <span class="citation" data-cites="coda">(<a href="#ref-coda" role="doc-biblioref">Plummer et al. 2006</a>)</span>,
<a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> 3.4.3 <span class="citation" data-cites="ggplot2">(<a href="#ref-ggplot2" role="doc-biblioref">Wickham 2016</a>)</span>,
<a href="https://cran.r-project.org/package=knitr">knitr</a> 1.45 <span class="citation" data-cites="knitr2014 knitr2015 knitr2023">(<a href="#ref-knitr2014" role="doc-biblioref">Xie 2014</a>, <a href="#ref-knitr2015" role="doc-biblioref">2015</a>, <a href="#ref-knitr2023" role="doc-biblioref">2023</a>)</span>,
<a href="https://cran.r-project.org/package=loo">loo</a> 2.6.0 <span class="citation" data-cites="loo">(<a href="#ref-loo" role="doc-biblioref">Vehtari et al. 2023</a>)</span>,
<a href="https://cran.r-project.org/package=rater">rater</a> 1.3.1 <span class="citation" data-cites="rater">(<a href="#ref-rater" role="doc-biblioref">Pullin and Vukcevic 2023</a>)</span>,
<a href="https://cran.r-project.org/package=rjtools">rjtools</a> 1.0.12 <span class="citation" data-cites="rjtools">(<a href="#ref-rjtools" role="doc-biblioref">O’Hara-Wild et al. 2023</a>)</span>,
<a href="https://cran.r-project.org/package=rmarkdown">rmarkdown</a> 1.0.12 <span class="citation" data-cites="rmarkdown2018 rmarkdown2020 rmarkdown2023">(<a href="#ref-rmarkdown2018" role="doc-biblioref">Xie et al. 2018</a>, <a href="#ref-rmarkdown2020" role="doc-biblioref">2020</a>; <a href="#ref-rmarkdown2023" role="doc-biblioref">Allaire et al. 2023</a>)</span>,
<a href="https://cran.r-project.org/package=rstan">rstan</a> 2.26.23 <span class="citation" data-cites="rstan">(<a href="#ref-rstan" role="doc-biblioref">Stan Development Team 2023</a>)</span>.
R itself and all packages used are available from the <a href="https://CRAN.R-project.org/">Comprehensive R Archive
Network (CRAN)</a>.</p>
<h2 data-number="10" id="acknowledgments"><span class="header-section-number">10</span> Acknowledgments</h2>
<p>We would like to thank Bob Carpenter for many helpful suggestions, and David
Whitelaw for his flexibility in allowing the first author to work on this paper
while employed at the Australian Institute of Heath and Welfare. Thanks also
to Lars Mølgaard Saxhaug for his code contributions to <a href="https://cran.r-project.org/package=rater">rater</a>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h3 class="appendix" data-number="10.1" id="supplementary-materials"><span class="header-section-number">10.1</span> Supplementary materials</h3>
<p>Supplementary materials are available in addition to this article. It can be downloaded at
<a href="RJ-2023-064.zip">RJ-2023-064.zip</a></p>
<h3 class="appendix" data-number="10.2" id="cran-packages-used"><span class="header-section-number">10.2</span> CRAN packages used</h3>
<p><a href="https://cran.r-project.org/package=rater">rater</a>, <a href="https://cran.r-project.org/package=poLCA">poLCA</a>, <a href="https://cran.r-project.org/package=BayesLCA">BayesLCA</a>, <a href="https://cran.r-project.org/package=randomLCA">randomLCA</a>, <a href="https://cran.r-project.org/package=coda">coda</a>, <a href="https://cran.r-project.org/package=rstan">rstan</a>, <a href="https://cran.r-project.org/package=loo">loo</a>, <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a>, <a href="https://cran.r-project.org/package=knitr">knitr</a>, <a href="https://cran.r-project.org/package=rjtools">rjtools</a>, <a href="https://cran.r-project.org/package=rmarkdown">rmarkdown</a></p>
<h3 class="appendix" data-number="10.3" id="cran-task-views-implied-by-cited-packages"><span class="header-section-number">10.3</span> CRAN Task Views implied by cited packages</h3>
<p><a href="https://cran.r-project.org/view=Bayesian">Bayesian</a>, <a href="https://cran.r-project.org/view=Cluster">Cluster</a>, <a href="https://cran.r-project.org/view=GraphicalModels">GraphicalModels</a>, <a href="https://cran.r-project.org/view=MixedModels">MixedModels</a>, <a href="https://cran.r-project.org/view=Phylogenetics">Phylogenetics</a>, <a href="https://cran.r-project.org/view=Psychometrics">Psychometrics</a>, <a href="https://cran.r-project.org/view=ReproducibleResearch">ReproducibleResearch</a>, <a href="https://cran.r-project.org/view=Spatial">Spatial</a>, <a href="https://cran.r-project.org/view=TeachingStatistics">TeachingStatistics</a></p>
<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-rmarkdown2023" class="csl-entry" role="listitem">
J. Allaire, Y. Xie, C. Dervieux, J. McPherson, J. Luraschi, K. Ushey, A. Atkins, H. Wickham, J. Cheng, W. Chang, et al. <em><span class="nocase">rmarkdown</span>: Dynamic documents for r.</em> 2023. URL <a href="https://github.com/rstudio/rmarkdown">https://github.com/rstudio/rmarkdown</a>. R package version 2.25.
</div>
<div id="ref-asselineau2018" class="csl-entry" role="listitem">
J. Asselineau, A. Paye, E. Bessède, P. Perez and C. Proust-Lima. Different latent class models were used and evaluated for assessing the accuracy of campylobacter diagnostic tests: Overcoming imperfect reference standards? <em>Epidemiology and Infection</em>, 146(12): 1556–1564, 2018. DOI <a href="https://doi.org/10.1017/S0950268818001723">10.1017/S0950268818001723</a>.
</div>
<div id="ref-beath2017" class="csl-entry" role="listitem">
K. J. Beath. <span class="nocase">randomLCA</span>: <span>An</span> <span>R</span> <span>Package</span> for <span>Latent</span> <span>Class</span> with <span>Random</span> <span>Effects</span> <span>Analysis</span>. <em>Journal of Statistical Software</em>, 81(13): 1–25, 2017. DOI <a href="https://doi.org/10.18637/jss.v081.i13">10.18637/jss.v081.i13</a>. Number: 1.
</div>
<div id="ref-pyanno" class="csl-entry" role="listitem">
P. Berkes, B. Carpenter, A. Rzhetsky and J. Evans. 2011. URL <a href="http://docs.enthought.com/uchicago-pyanno/">http://docs.enthought.com/uchicago-pyanno/</a>.
</div>
<div id="ref-blackwell1947" class="csl-entry" role="listitem">
D. Blackwell. Conditional expectation and unbiased sequential estimation. <em>Ann. Math. Statist.</em>, 18(1): 105–110, 1947. DOI <a href="https://doi.org/10.1214/aoms/1177730497">10.1214/aoms/1177730497</a>.
</div>
<div id="ref-carpenter2017" class="csl-entry" role="listitem">
B. Carpenter, A. Gelman, M. D. Hoffman, D. Lee, B. Goodrich, M. Betancourt, M. Brubaker, J. Guo, P. Li and A. Riddell. Stan: <span>A</span> <span>Probabilistic</span> <span>Programming</span> <span>Language</span>. <em>Journal of Statistical Software</em>, 76(1): 2017. DOI <a href="https://doi.org/10.18637/jss.v076.i01">10.18637/jss.v076.i01</a>.
</div>
<div id="ref-dawid1979" class="csl-entry" role="listitem">
A. P. Dawid and A. M. Skene. Maximum <span>Likelihood</span> <span>Estimation</span> of <span>Observer</span> <span>Error</span>-<span>Rates</span> <span>Using</span> the <span>EM</span> <span>Algorithm</span>. <em>Applied Statistics</em>, 28(1): 20, 1979. DOI <a href="https://doi.org/10.2307/2346806">10.2307/2346806</a>.
</div>
<div id="ref-espeland1989" class="csl-entry" role="listitem">
M. A. Espeland and S. L. Handelman. Using latent class models to characterize and assess relative error in discrete measurements. <em>Biometrics</em>, 45(2): 587–599, 1989. DOI <a href="https://doi.org/10.2307/2531499">10.2307/2531499</a>.
</div>
<div id="ref-geyer1995" class="csl-entry" role="listitem">
C. J. Geyer. Conditioning in <span>M</span>arkov chain <span>M</span>onte <span>C</span>arlo. <em>Journal of Computational and Graphical Statistics</em>, 4(2): 148–154, 1995. DOI <a href="https://doi.org/10.1080/10618600.1995.10474672">10.1080/10618600.1995.10474672</a>.
</div>
<div id="ref-goodman1974" class="csl-entry" role="listitem">
L. A. Goodman. Exploratory latent structure analysis using both identifiable and unidentifiable models. <em>Biometrika</em>, 61(2): 215–231, 1974. DOI <a href="https://doi.org/10.1093/biomet/61.2.215">10.1093/biomet/61.2.215</a>.
</div>
<div id="ref-handelman1986" class="csl-entry" role="listitem">
S. L. Handelman, D. H. Leverett, M. A. Espeland and J. A. Curzon. Clinical radiographic evaluation of sealed carious and sound tooth surfaces. <em>The Journal of the American Dental Association</em>, 113(5): 751–754, 1986. DOI <a href="https://doi.org/10.14219/jada.archive.1986.0269">10.14219/jada.archive.1986.0269</a>.
</div>
<div id="ref-JMLR:v15:hoffman14a" class="csl-entry" role="listitem">
M. D. Hoffman and A. Gelman. The <span>No</span>-<span>U</span>-<span>Turn</span> <span>Sampler</span>: Adaptively setting path lengths in <span>Hamiltonian</span> <span>Monte</span> <span>Carlo</span>. <em>Journal of Machine Learning Research</em>, 15: 1593–1623, 2014. URL <a href="http://jmlr.org/papers/v15/hoffman14a.html">http://jmlr.org/papers/v15/hoffman14a.html</a>.
</div>
<div id="ref-hui1980" class="csl-entry" role="listitem">
S. L. Hui and S. D. Walter. Estimating the <span>Error</span> <span>Rates</span> of <span>Diagnostic</span> <span>Tests</span>. <em>Biometrics</em>, 36(1): 167–171, 1980. DOI <a href="https://doi.org/10.2307/2530508">10.2307/2530508</a>.
</div>
<div id="ref-ipeirotis2010" class="csl-entry" role="listitem">
P. G. Ipeirotis, F. Provost and J. Wang. Quality management on <span>A</span>mazon <span>M</span>echanical <span>T</span>urk. In <em>Proceedings of the ACM SIGKDD workshop on human computation</em>, pages. 64–67 2010.
</div>
<div id="ref-jakobsdottir2007" class="csl-entry" role="listitem">
J. Jakobsdottir and D. E. Weeks. Estimating prevalence, false-positive rate, and false-negative rate with use of repeated testing when true responses are unknown. <em>The American Journal of Human Genetics</em>, 81(5): 1111–1113, 2007. DOI <a href="https://doi.org/10.1086/521582">10.1086/521582</a>.
</div>
<div id="ref-li2014" class="csl-entry" role="listitem">
H. Li and B. Yu. Error <span>Rate</span> <span>Bounds</span> and <span>Iterative</span> <span>Weighted</span> <span>Majority</span> <span>Voting</span> for <span>Crowdsourcing</span>. <em>arXiv preprint arXiv:1411.4086</em>, 2014. URL <a href="https://arxiv.org/abs/1411.4086">https://arxiv.org/abs/1411.4086</a>.
</div>
<div id="ref-linzer2011" class="csl-entry" role="listitem">
D. A. Linzer and J. B. Lewis. : <span>An</span> <span>R</span> <span>Package</span> for <span>Polytomous</span> <span>Variable</span> <span>Latent</span> <span>Class</span> <span>Analysis</span>. <em>Journal of Statistical Software</em>, 42(1): 1–29, 2011. DOI <a href="https://doi.org/10.18637/jss.v042.i10">10.18637/jss.v042.i10</a>.
</div>
<div id="ref-rjtools" class="csl-entry" role="listitem">
M. O’Hara-Wild, S. Kobakian, H. S. Zhang, D. Cook, S. Urbanek and C. Dervieux. <em><span class="nocase">rjtools</span>: Preparing, checking, and submitting articles to the <span>“<span>R Journal</span>.”</span></em> 2023. URL <a href="https://CRAN.R-project.org/package=rjtools">https://CRAN.R-project.org/package=rjtools</a>. R package version 1.0.10.
</div>
<div id="ref-mcbook" class="csl-entry" role="listitem">
A. B. Owen. <em>Monte <span>Carlo</span> theory, methods and examples.</em> 2013. URL <a href="https://statweb.stanford.edu/~owen/mc/">https://statweb.stanford.edu/~owen/mc/</a>.
</div>
<div id="ref-passonneau2014" class="csl-entry" role="listitem">
R. J. Passonneau and B. Carpenter. The <span>Benefits</span> of a <span>Model</span> of <span>Annotation</span>. <em>Transactions of the Association for Computational Linguistics</em>, 2: 311–326, 2014. DOI <a href="https://doi.org/10.1162/tacl_a_00185">10.1162/tacl_a_00185</a>.
</div>
<div id="ref-paun2018" class="csl-entry" role="listitem">
S. Paun, B. Carpenter, J. Chamberlain, D. Hovy, U. Kruschwitz and M. Poesio. Comparing <span>Bayesian</span> <span>Models</span> of <span>Annotation</span>. <em>Transactions of the Association for Computational Linguistics</em>, 6: 571–585, 2018. DOI <a href="https://doi.org/10.1162/tacl_a_00040">10.1162/tacl_a_00040</a>.
</div>
<div id="ref-coda" class="csl-entry" role="listitem">
M. Plummer, N. Best, K. Cowles and K. Vines. CODA: Convergence diagnosis and output analysis for MCMC. <em>R News</em>, 6(1): 7–11, 2006. URL <a href="https://journal.r-project.org/archive/">https://journal.r-project.org/archive/</a>.
</div>
<div id="ref-rater" class="csl-entry" role="listitem">
J. Pullin and D. Vukcevic. <em><span class="nocase">rater</span>: Statistical models of repeated categorical rating data.</em> 2023. URL <a href="https://CRAN.R-project.org/package=rater">https://CRAN.R-project.org/package=rater</a>. R package version 1.3.1.
</div>
<div id="ref-base" class="csl-entry" role="listitem">
R Core Team. <em>R: A language and environment for statistical computing.</em> Vienna, Austria: R Foundation for Statistical Computing, 2021. URL <a href="https://www.R-project.org/">https://www.R-project.org/</a>.
</div>
<div id="ref-robert2021raoblackwellization" class="csl-entry" role="listitem">
C. P. Robert and G. O. Roberts. <span>R</span>ao-<span>B</span>lackwellization in the <span>MCMC</span> era. <em>arXiv:2101.01011 [stat]</em>, 2021. URL <a href="https://arxiv.org/abs/2101.01011">https://arxiv.org/abs/2101.01011</a>. arXiv: 2101.01011.
</div>
<div id="ref-rzhetsky2009" class="csl-entry" role="listitem">
H. A. W. Rzhetsky Andrey AND Shatkay. How to get the most out of your curation effort. <em>PLOS Computational Biology</em>, 5(5): 1–13, 2009. DOI <a href="https://doi.org/10.1371/journal.pcbi.1000391">10.1371/journal.pcbi.1000391</a>.
</div>
<div id="ref-smyth1994" class="csl-entry" role="listitem">
P. Smyth, U. Fayyad, M. Burl, P. Perona and P. Baldi. Inferring ground truth from subjective labelling of <span>V</span>enus images. In <em>Advances in neural information processing systems</em>, Eds G. Tesauro, D. Touretzky and T. Leen 1994. MIT Press. URL <a href="https://proceedings.neurips.cc/paper_files/paper/1994/file/3cef96dcc9b8035d23f69e30bb19218a-Paper.pdf">https://proceedings.neurips.cc/paper_files/paper/1994/file/3cef96dcc9b8035d23f69e30bb19218a-Paper.pdf</a>.
</div>
<div id="ref-rstan" class="csl-entry" role="listitem">
Stan Development Team. : The <span>R</span> interface to <span>Stan</span>. 2023. URL <a href="http://mc-stan.org/">http://mc-stan.org/</a>. R package version 2.26.23.
</div>
<div id="ref-stanguide" class="csl-entry" role="listitem">
Stan Development Team. <em>Stan <span>U</span>ser’s <span>G</span>uide.</em> 2021. URL <a href="https://mc-stan.org/docs/2_27/stan-users-guide/">https://mc-stan.org/docs/2_27/stan-users-guide/</a>. Version 2.27.
</div>
<div id="ref-loo" class="csl-entry" role="listitem">
A. Vehtari, J. Gabry, M. Magnusson, Y. Yao, P.-C. Bürkner, T. Paananen and A. Gelman. Loo: Efficient leave-one-out cross-validation and WAIC for <span>B</span>ayesian models. 2023. URL <a href="https://mc-stan.org/loo/">https://mc-stan.org/loo/</a>. R package version 2.6.0.
</div>
<div id="ref-vehtari2017" class="csl-entry" role="listitem">
A. Vehtari, A. Gelman and J. Gabry. Practical <span>Bayesian</span> model evaluation using leave-one-out cross-validation and <span>WAIC</span>. <em>Statistics and Computing</em>, 27(5): 1413–1432, 2017. DOI <a href="https://doi.org/10.1007/s11222-016-9696-4">10.1007/s11222-016-9696-4</a>.
</div>
<div id="ref-vehtari2021" class="csl-entry" role="listitem">
A. Vehtari, A. Gelman, D. Simpson, B. Carpenter and P.-C. Bürkner. Rank-<span>Normalization</span>, <span>Folding</span>, and <span>Localization</span>: <span>An</span> <span>Improved</span> <span>Rˆ</span> for <span>Assessing</span> <span>Convergence</span> of <span>MCMC</span> (with <span>Discussion</span>). <em>Bayesian Analysis</em>, 16(2): 667–718, 2021. DOI <a href="https://doi.org/10.1214/20-BA1221">10.1214/20-BA1221</a>.
</div>
<div id="ref-watanabe2013" class="csl-entry" role="listitem">
S. Watanabe. A <span>Widely Applicable Bayesian Information Criterion</span>. <em>Journal of Machine Learning Research</em>, 14: 867–897, 2013.
</div>
<div id="ref-white2014" class="csl-entry" role="listitem">
A. White and T. B. Murphy. : <span>An</span> <span>R</span> <span>Package</span> for <span>Bayesian</span> <span>Latent</span> <span>Class</span> <span>Analysis</span>. <em>Journal of Statistical Software</em>, 61(13): 2014. DOI <a href="https://doi.org/10.18637/jss.v061.i13">10.18637/jss.v061.i13</a>.
</div>
<div id="ref-ggplot2" class="csl-entry" role="listitem">
H. Wickham. <em>ggplot2: Elegant graphics for data analysis.</em> Springer-Verlag New York, 2016. URL <a href="https://ggplot2.tidyverse.org">https://ggplot2.tidyverse.org</a>.
</div>
<div id="ref-knitr2015" class="csl-entry" role="listitem">
Y. Xie. <em>Dynamic documents with <span>R</span> and knitr.</em> 2nd ed Boca Raton, Florida: Chapman; Hall/CRC, 2015. URL <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>. ISBN 978-1498716963.
</div>
<div id="ref-knitr2014" class="csl-entry" role="listitem">
Y. Xie. <span class="nocase">knitr</span>: A comprehensive tool for reproducible research in <span>R</span>. In <em>Implementing reproducible computational research</em>, Eds V. Stodden, F. Leisch and R. D. Peng 2014. Chapman; Hall/CRC. ISBN 978-1466561595.
</div>
<div id="ref-knitr2023" class="csl-entry" role="listitem">
Y. Xie. <em><span class="nocase">knitr</span>: A general-purpose package for dynamic report generation in r.</em> 2023. URL <a href="https://yihui.org/knitr/">https://yihui.org/knitr/</a>. R package version 1.44.
</div>
<div id="ref-rmarkdown2018" class="csl-entry" role="listitem">
Y. Xie, J. J. Allaire and G. Grolemund. <em>R markdown: The definitive guide.</em> Boca Raton, Florida: Chapman; Hall/CRC, 2018. URL <a href="https://bookdown.org/yihui/rmarkdown">https://bookdown.org/yihui/rmarkdown</a>.
</div>
<div id="ref-rmarkdown2020" class="csl-entry" role="listitem">
Y. Xie, C. Dervieux and E. Riederer. <em>R markdown cookbook.</em> Boca Raton, Florida: Chapman; Hall/CRC, 2020. URL <a href="https://bookdown.org/yihui/rmarkdown-cookbook">https://bookdown.org/yihui/rmarkdown-cookbook</a>.
</div>
<div id="ref-yackulic2020" class="csl-entry" role="listitem">
C. B. Yackulic, M. Dodrill, M. Dzul, J. S. Sanderlin and J. A. Reid. A need for speed in bayesian population models: A practical guide to marginalizing and recovering discrete latent states. <em>Ecological Applications</em>, 30(5): e02112, 2020. DOI <a href="https://doi.org/10.1002/eap.2112">10.1002/eap.2112</a>.
</div>
<div id="ref-zhang2022" class="csl-entry" role="listitem">
W. Zhang, J. Pullin, L. Gurrin and D. Vukcevic. Investigating the efficiency of marginalising over discrete parameters in <span>B</span>ayesian computations. <em>arXiv:2204.06313 [stat]</em>, 2022. URL <a href="https://arxiv.org/abs/2204.06313">https://arxiv.org/abs/2204.06313</a>.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
  <h3 id="references">References</h3>
  <div id="references-listing"></div>
  <h3 id="reuse">Reuse</h3>
  <p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
  <h3 id="citation">Citation</h3>
  <p>For attribution, please cite this work as</p>
  <pre class="citation-appendix short">Pullin, et al., "Statistical Models for Repeated Categorical Ratings: The R Package rater", The R Journal, 2023</pre>
  <p>BibTeX citation</p>
  <pre class="citation-appendix long">@article{RJ-2023-064,
  author = {Pullin, Jeffrey M. and Gurrin, Lyle C. and Vukcevic, Damjan},
  title = {Statistical Models for Repeated Categorical Ratings: The R Package rater},
  journal = {The R Journal},
  year = {2023},
  note = {https://doi.org/10.32614/RJ-2023-064},
  doi = {10.32614/RJ-2023-064},
  volume = {15},
  issue = {3},
  issn = {2073-4859},
  pages = {93-118}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<!--radix_placeholder_navigation_after_body--><html><body>
<div class="distill-site-nav distill-site-footer">
<p>© The R Foundation, <a href="mailto:r-journal@r-project.org">web page
contact</a>.</p>
</div>
<!--/radix_placeholder_navigation_after_body-->
</body></html>


</body>

</html>
