---
title: Bioconductor Notes, December 2025
date: '2025-12-01'
author:
- name: Maria Doyle
  affiliation: University of Limerick
  address: Bioconductor Community Manager
- name: Bioconductor Core Developer Team
  affiliation: Dana-Farber Cancer Institute
  address: Roswell Park Comprehensive Cancer Center, City University of New York,
    Fred Hutchinson Cancer Research Center, Mass General Brigham
output:
  rjtools::rjournal_article:
    self_contained: yes
    toc: no
volume: 17
issue: 4
slug: RJ-2025-4-bioconductor
draft: no
journal:
  lastpage: 324
  firstpage: 321

---


# Introduction

[Bioconductor](https://bioconductor.org) provides
tools for the analysis and comprehension of high-throughput genomic
data.  The project has entered its twenty-first year, with funding
for core development and infrastructure maintenance secured
through 2025 (NIH NHGRI 2U24HG004059).  Additional support is provided
by NIH NCI, Chan-Zuckerberg Initiative, National Science Foundation,
Microsoft, and Amazon. In this news report, we give some updates on
core team and project activities.

# Software

```{r lkdat,results="hide",echo=FALSE,message=FALSE}
library(BiocPkgTools)
softli = biocPkgList(version="3.22")
expli = biocPkgList(version="3.22", repo="BioCexp")
worli = biocPkgList(version="3.22", repo="BioCworkflows")
annli = biocPkgList(version="3.22", repo="BioCann")
curR = "4.5"
nsoft = nrow(softli)
nexp = nrow(expli)
nanno = nrow(annli)
nwkfl = nrow(worli)
nbooks = 8
```

Bioconductor [3.22](https://bioconductor.org/news/bioc_3_22_release/), released in October 2025, is now available. It is
compatible with R `r curR` and consists of `r nsoft` software packages, `r nexp`
experiment data packages, `r nanno` up-to-date annotation packages, `r nwkfl`
workflows, and `r nbooks` books. [Books](https://bioconductor.org/books/release/) are
built regularly from source, ensuring full reproducibility; an example is the
community-developed [Orchestrating Single-Cell Analysis with Bioconductor](https://bioconductor.org/books/release/OSCA/).


# Core Team and Infrastructure Updates

Bioconductor has introduced new GPU infrastructure, funded by CZI EOSS 6, to support developers building GPU‑accelerated packages, including Nvidia GPU build nodes, GPU‑aware containers, and a new biocViews term. Maintainers can now opt into GPU software builds and mark packages as GPU‑optional or GPU‑required. See [the blog post](https://blog.bioconductor.org/posts/2025-10-10-gpus/) for more details.

NEWS summaries for three contributed packages chosen at random from the 59 new software contributions are:

<!-- set.seed(1234); sample(1:59, size=3) gave indices -->

- **dmGsea**: The R package dmGsea provides efficient gene set enrichment analysis specifically for DNA methylation data. It addresses key biases, including probe dependency and varying probe numbers per gene. The package supports Illumina 450K, EPIC, and mouse methylation arrays. Users can also apply it to other omics data by supplying custom probe-to-gene mapping annotations. dmGsea is flexible, fast, and well-suited for large-scale epigenomic studies.

- **goatea**: Geneset Ordinal Association Test Enrichment Analysis (GOATEA) provides a ‘Shiny’ interface with interactive visualizations and utility functions for performing and exploring automated gene set enrichment analysis using the ‘GOAT’ package. ‘GOATEA’ is designed to support large-scale and user-friendly enrichment workflows across multiple gene lists and comparisons, with flexible plotting and output options. Visualizations pre-enrichment include interactive ‘Volcano’ and ‘UpSet’ (overlap) plots. Visualizations post-enrichment include interactive geneset dotplot, geneset treeplot, gene-effectsize heatmap, gene-geneset heatmap and ‘STRING’ database of protein-protein-interactions network graph. ‘GOAT’ reference: Frank Koopmans (2024) <doi:10.1038/s42003-024-06454-5>.

- **igblastr**: The igblastr package provides functions to conveniently install and use a local IgBLAST installation from within R. IgBLAST is described at https://pubmed.ncbi.nlm.nih.gov/23671333/. Online IgBLAST: https://www.ncbi.nlm.nih.gov/igblast/.

See the NEWS section in the [release announcement](https://bioconductor.org/news/bioc_3_22_release/) for
a complete account of changes throughout the ecosystem.


# Community and Impact

## Community Team Updates
Nicholas Cooley has joined the Bioconductor Community Team as Developer Engagement Lead, based at the University of Limerick in Ireland. Working with the Community Manager, his role focuses on supporting package developers, improving onboarding resources, and strengthening connections across the developer community. Nick is funded through [CZI EOSS 6](https://blog.bioconductor.org/posts/2024-07-12-czi-eoss6-grants/).

Laurah Ondari has also joined the team part‑time, based at the International Institute of Tropical Agriculture (IITA) in Kenya. She works with the Community Manager on communications, social media, and community engagement, including supporting Africa-focused capacity‑building efforts. Laurah’s position is also supported by [CZI EOSS 6](https://blog.bioconductor.org/posts/2024-07-12-czi-eoss6-grants/) funding.

## Outreachy Internships

The June–August 2025 [Outreachy](https://www.outreachy.org/) program concluded successfully, with interns contributing to Bioconductor and sharing their reflections in a [blog post](https://blog.bioconductor.org/posts/2025-12-12-outreachy-june25/).

## Community Updates

The Bioconductor Seminar Series is a new quarterly online event showcasing recent advances in computational biology and their relevance to Bioconductor methods, workflows, and community practice. Conceived by Bioconductor founder Robert Gentleman and organised by Erica Feick, the series began in December 2025 and brings together expert speakers, moderated discussions, and open Q&A. The first session was on *Deep-learning-based Gene Perturbation Effect Prediction Does Not Yet Outperform Simple Linear Baselines* (Nature Methods, 2025) with speakers Constantin Ahlmann-Eltze, Wolfgang Huber, Simon Anders and discussant Davide Risso. It was well attended with engaging discussion. More details can be found on the [Bioconductor website](https://bioconductor.org/help/seminar-series/).

## Publications and Preprints

In November 2025, Crowell et al. released a [preprint](https://www.biorxiv.org/content/10.1101/2025.11.20.688607v1) describing their online book Orchestrating Spatial Transcriptomics Analysis with Bioconductor (OSTA), a collaborative effort supported by leaders across the community and built on two decades of Bioconductor’s foundational work. The project invites feedback, suggestions, and contributions from Bioconductor users and developers.


# Conferences and Workshops

## Recaps

*   **EuroBioC 2025:** The European Bioconductor Conference (GBCC 2025) was held in September 2025 in Barcelona. A blog post recap of the conference can be found in the [Bioconductor blog](https://blog.bioconductor.org/posts/2025-10-24-EuroBioC2025-recap/). Recordings of talks are available in a [YouTube playlist](https://www.youtube.com/playlist?list=PLdl4u5ZRDMQS_qvtLJNdDqHL6z5jj5y_7).
*   **BioCAsia 2025:** BioCAsia 2025 was held as part of the ABACBS conference in Adelaide, November 27-28. The focus was on hands-on workshops and saw >100 participants. For more information, visit the [conference website](https://biocasia2025.bioconductor.org/).
*   **Global Training:** We have had a busy second half of the year of training events, with two in-person courses in Africa, in [Ethiopia](https://blog.bioconductor.org/posts/2025-11-24-ethiopia-course/) and  [Benin](https://blog.bioconductor.org/posts/2025-12-11-benin-course/).

## Announcements

*   **EuroBioC 2026:** The European Bioconductor conference is taking place in Turku, Finland from June 3-5, with pre-conference activities June 1-2. For more information, visit the [conference website](https://eurobioc2026.bioconductor.org/).
*   **BioC 2026:** The North American Bioconductor conference is taking place in Seattle from August 10-12, with post-conference activities August 13-14. For more information, visit the [conference website](https://bioc2026.bioconductor.org/).


# Boards and Working Groups Updates

## New Board Members

- The [Community Advisory Board](https://bioconductor.org/about/community-advisory-board/) (CAB) welcomes new members Fabricio Almeida-Silva, Tuomas Borman, Laurent Gatto, Zuguang Gu, Eliana Ibrahimi, Martha Luka and Izabela Mamede. We extend our gratitude to outgoing members Jasmine Daly, Leo Lahti, Nicole Ortogero, Janani Ravi, Luyi Tian, Hedia Tnani and Jiefei Wang for their service.

- The [Technical Advisory Board](https://bioconductor.org/about/technical-advisory-board/) (TAB) welcomes new members Robert Castelo and Hugo Gruson and Gabriele Sales. Outgoing members Stephanie Hicks, Davide Risso and Charlotte Soneson are also warmly thanked for their contributions.

# Using Bioconductor

Start using
Bioconductor by installing the most recent version of R and evaluating
the commands
```
  if (!requireNamespace("BiocManager", quietly = TRUE))
      install.packages("BiocManager")
  BiocManager::install()
```

Install additional packages and dependencies,
e.g., [SingleCellExperiment](https://bioconductor.org/packages/SingleCellExperiment), with
```
  BiocManager::install("SingleCellExperiment")
```
[Docker](https://bioconductor.org/help/docker/)
images provides a very effective on-ramp for power users to rapidly
obtain access to standardized and scalable computing environments.
Key resources include:


- [bioconductor.org](https://bioconductor.org) to install,
  learn, use, and develop Bioconductor packages.
- A list of [available software](https://bioconductor.org/packages)
  linking to pages describing each package.
- A question-and-answer style
  [user support site](https://support.bioconductor.org) and
  developer-oriented [mailing list](https://stat.ethz.ch/mailman/listinfo/bioc-devel).
- A community Zulip workspace ([sign up](https://chat.bioconductor.org))
   for extended technical discussion.
- The [F1000Research Bioconductor gateway](https://f1000research.com/gateways/bioconductor)
for peer-reviewed Bioconductor workflows as well as conference contributions.
- The [Bioconductor YouTube](https://www.youtube.com/user/bioconductor)
     channel includes recordings of keynote and talks from recent
     conferences, in addition to
     video recordings of training courses.
- Our [package submission](https://github.com/Bioconductor/Contributions)
  repository for open technical review of new packages.

Upcoming and recently completed events are browsable at our
[events page](https://bioconductor.org/help/events/).

The [Technical](https://bioconductor.org/about/technical-advisory-board/) and [Community](https://bioconductor.org/about/community-advisory-board/)
Advisory Boards provide guidance to ensure that the project addresses
leading-edge biological problems with advanced technical approaches,
and adopts practices (such as a
project-wide [Code of Conduct](https://bioconductor.org/about/code-of-conduct/))
that encourages all to participate. We look forward to
welcoming you!

We welcome your feedback on these updates and invite you to connect with us through the [Bioconductor Zulip](https://chat.bioconductor.org) workspace or by emailing community@bioconductor.org.
