# Generated by `rjournal_pdf_article()` using `knitr::purl()`: do not edit by hand
# Please edit RJ-2024-022.Rmd to modify this file

## ----setup, include=FALSE-----------------------------------------------------
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)


## ----fig1, out.width = "100%", fig.cap = "Overview of the main parameters and functions comprising homology search in DECIPHER.", fig.alt=""----
knitr::include_graphics(if(knitr::is_latex_output()) "figures/Fig1.pdf" else "figures/Fig1.svg")


## ----fig2, out.width = "100%", fig.cap = "Matches between a query sequence and each target sequence in a database are chained into high scoring hits. Hits are scored based on the number of positions separating matches and the minimum number of implied gaps between matches.", fig.alt=""----
knitr::include_graphics(if(knitr::is_latex_output()) "figures/Fig2.pdf" else "figures/Fig2.svg")


## ----fig3, out.width = "100%", fig.cap = "The SCOPe database of amino acid sequences assigned to protein families was used to compare search programs. $(A)$ Speed is shown relative to protein BLAST (horizontal gray line) for easier comparison. All programs were configured to use the maximum available processor threads. DECIPHER achieved greater accuracy (average AUC1) with inclusion of subject sequences to allow extension of k-mer matches. The value of k-mer length ($k$) is shown on each DECIPHER point. Lower values of $k$ and step size improved accuracy at the expense of speed. MMseqs2 results are shown across a range of user-specified sensitivities from $-s$ $1$ to $-s$ $10$, with the default of $5.7$ shown as a point. BLAST was run under its default settings, which reports hits with E-values less than $10$. $(B)$ The average fraction of true positives given higher scores than those of their corresponding decoy false positive increases with percent identity between the true positives and query sequences. DECIPHER results are shown at different k-mer lengths for a step size of $1$ when supplying the subject sequences (i.e., the black line in $(A)$). MMseqs2 and BLAST results are shown at their default settings.", fig.alt=""----
knitr::include_graphics(if(knitr::is_latex_output()) "figures/Fig3.pdf" else "figures/Fig3.svg")


## ----fig4, out.width = "100%", fig.cap = "The scores output by \\textit{SearchIndex} and \\textit{AlignPairs} for each hit (point) show similar correlations with percent identity. Lower read (query) coverage often resulted in higher percent identities for the same score, implying the hit corresponded to a smaller aligned region with higher \\textit{PID}.", fig.alt=""----
knitr::include_graphics("figures/Fig4.png")


## ----fig5, out.width = "100%", fig.cap = "More orthologous pairs were predicted when searching protein (amino acid) sequences rather than their corresponding coding (nucleotide) sequences. Lower values of k-mer length and, especially, step size permitted the identification of a greater number of putative orthologs with or without matching protein annotations. The approximate p-value (one-sided binomial test) for observing a given number of matched annotations by chance is shown above each bar.", fig.alt=""----
knitr::include_graphics(if(knitr::is_latex_output()) "figures/Fig5.pdf" else "figures/Fig5.svg")

