<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
      margin-bottom: 0em;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
<title>BMRMM: An R Package for Bayesian Markov (Renewal) Mixed Models</title>

<meta property="description" itemprop="description" content="We introduce the BMRMM&#10;package implementing Bayesian inference for a class of Markov renewal&#10;mixed models which can characterize the stochastic dynamics of a&#10;collection of sequences, each comprising alternative instances of&#10;categorical states and associated continuous duration times, while&#10;being influenced by a set of exogenous factors as well as a &#39;random&#39;&#10;individual. The default setting flexibly models the state transition&#10;probabilities using mixtures of Dirichlet distributions and the&#10;duration times using mixtures of gamma kernels while also allowing&#10;variable selection for both. Modeling such data using simpler Markov&#10;mixed models also remains an option, either by ignoring the duration&#10;times altogether or by replacing them with instances of an additional&#10;category obtained by discretizing them by a user-specified unit. The&#10;option is also useful when data on duration times may not be available&#10;in the first place. We demonstrate the package&#39;s utility using two&#10;data sets."/>

<link rel="canonical" href="https://doi.org/10.32614/RJ-2024-011/"/>
<link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>
<link rel="icon" type="image/vnd.microsoft.icon" href="../../resources/favicon.ico"/>

<!--  https://schema.org/Article -->
<meta property="article:published" itemprop="datePublished" content="2025-01-11"/>
<meta property="article:created" itemprop="dateCreated" content="2025-01-11"/>
<meta name="article:author" content="Yutong Wu"/>
<meta name="article:author" content="Abhra Sarkar"/>

<!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
<meta property="og:title" content="BMRMM: An R Package for Bayesian Markov (Renewal) Mixed Models"/>
<meta property="og:type" content="article"/>
<meta property="og:description" content="We introduce the BMRMM&#10;package implementing Bayesian inference for a class of Markov renewal&#10;mixed models which can characterize the stochastic dynamics of a&#10;collection of sequences, each comprising alternative instances of&#10;categorical states and associated continuous duration times, while&#10;being influenced by a set of exogenous factors as well as a &#39;random&#39;&#10;individual. The default setting flexibly models the state transition&#10;probabilities using mixtures of Dirichlet distributions and the&#10;duration times using mixtures of gamma kernels while also allowing&#10;variable selection for both. Modeling such data using simpler Markov&#10;mixed models also remains an option, either by ignoring the duration&#10;times altogether or by replacing them with instances of an additional&#10;category obtained by discretizing them by a user-specified unit. The&#10;option is also useful when data on duration times may not be available&#10;in the first place. We demonstrate the package&#39;s utility using two&#10;data sets."/>
<meta property="og:url" content="https://doi.org/10.32614/RJ-2024-011/"/>
<meta property="og:image" content="https://journal.r-project.org/articles/RJ-2024-011/figs/data-model.png"/>
<meta property="og:image:width" content="1111"/>
<meta property="og:image:height" content="100"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="The R Journal"/>

<!--  https://dev.twitter.com/cards/types/summary -->
<meta property="twitter:card" content="summary_large_image"/>
<meta property="twitter:title" content="BMRMM: An R Package for Bayesian Markov (Renewal) Mixed Models"/>
<meta property="twitter:description" content="We introduce the BMRMM&#10;package implementing Bayesian inference for a class of Markov renewal&#10;mixed models which can characterize the stochastic dynamics of a&#10;collection of sequences, each comprising alternative instances of&#10;categorical states and associated continuous duration times, while&#10;being influenced by a set of exogenous factors as well as a &#39;random&#39;&#10;individual. The default setting flexibly models the state transition&#10;probabilities using mixtures of Dirichlet distributions and the&#10;duration times using mixtures of gamma kernels while also allowing&#10;variable selection for both. Modeling such data using simpler Markov&#10;mixed models also remains an option, either by ignoring the duration&#10;times altogether or by replacing them with instances of an additional&#10;category obtained by discretizing them by a user-specified unit. The&#10;option is also useful when data on duration times may not be available&#10;in the first place. We demonstrate the package&#39;s utility using two&#10;data sets."/>
<meta property="twitter:url" content="https://doi.org/10.32614/RJ-2024-011/"/>
<meta property="twitter:image" content="https://journal.r-project.org/articles/RJ-2024-011/figs/data-model.png"/>
<meta property="twitter:image:width" content="1111"/>
<meta property="twitter:image:height" content="100"/>

<!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
<meta name="citation_title" content="BMRMM: An R Package for Bayesian Markov (Renewal) Mixed Models"/>
<meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2024-011"/>
<meta name="citation_pdf_url" content="RJ-2024-011.pdf"/>
<meta name="citation_volume" content="16"/>
<meta name="citation_issue" content="1"/>
<meta name="citation_doi" content="10.32614/RJ-2024-011"/>
<meta name="citation_journal_title" content="The R Journal"/>
<meta name="citation_issn" content="2073-4859"/>
<meta name="citation_firstpage" content="192"/>
<meta name="citation_lastpage" content="211"/>
<meta name="citation_fulltext_world_readable" content=""/>
<meta name="citation_online_date" content="2025/01/11"/>
<meta name="citation_publication_date" content="2025/01/11"/>
<meta name="citation_author" content="Yutong Wu"/>
<meta name="citation_author_institution" content="Department of Mechanical Engineering"/>
<meta name="citation_author" content="Abhra Sarkar"/>
<meta name="citation_author_institution" content="Department of Statistics and Data Sciences"/>
<!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Flexible regression models for count data based on renewal processes: The Countr package;citation_volume=90;citation_author=Tarak Kharrat;citation_author=Georgi N Boshnakov;citation_author=Ian McHale;citation_author=Rose Baker"/>
  <meta name="citation_reference" content="citation_title=ARPobservation: Simulating recording procedures for direct observation of behavior;citation_author=James E. Pustejovsky"/>
  <meta name="citation_reference" content="citation_title=SMM: An R package for estimation and simulation of discrete-time semi-Markov models;citation_volume=10;citation_author=Vlad Stefan Barbu;citation_author=Caroline Bérard;citation_author=Dominique Cellier;citation_author=Mathilde Sautreuil;citation_author=Nicolas Vergne"/>
  <meta name="citation_reference" content="citation_title=SemiMarkov: An R package for parametric estimation in multi-state semi-Markov models;citation_volume=66;citation_author=Agnieszka Listwon;citation_author=Philippe Saint-Pierre"/>
  <meta name="citation_reference" content="citation_title=vcd: Visualizing categorical data;citation_author=David Meyer;citation_author=Achim Zeileis;citation_author=Kurt Hornik"/>
  <meta name="citation_reference" content="citation_title=catdap, a categorical data analysis program package;citation_volume=14;citation_author=Kouichi Katsura"/>
  <meta name="citation_reference" content="citation_title=Bayesian semiparametric mixed effects Markov models with application to vocalization syntax;citation_publisher=Taylor &amp; Francis;citation_volume=113;citation_author=Abhra Sarkar;citation_author=Jonathan Chabout;citation_author=Joshua Jones Macopson;citation_author=Erich D Jarvis;citation_author=David B Dunson"/>
  <meta name="citation_reference" content="citation_title=Bayesian semiparametric Markov renewal mixed models for vocalization syntax;citation_author=Yutong Wu;citation_author=Erich D Jarvis;citation_author=Abhra Sarkar"/>
  <meta name="citation_reference" content="citation_title=Bayes estimation from a Markov renewal process;citation_publisher=JSTOR;citation_volume=18;citation_author=Michael J Phelan"/>
  <meta name="citation_reference" content="citation_title=Reinforced random processes in continuous time;citation_publisher=Elsevier;citation_volume=104;citation_author=Pietro Muliere;citation_author=Piercesare Secchi;citation_author=Stephen G Walker"/>
  <meta name="citation_reference" content="citation_title=Bayesian nonparametric estimation for reinforced Markov renewal processes;citation_publisher=Springer;citation_volume=10;citation_author=Paolo Bulla;citation_author=Pietro Muliere"/>
  <meta name="citation_reference" content="citation_title=Estimation in stationary Markov renewal processes, with application to earthquake forecasting in Turkey;citation_publisher=Springer;citation_volume=7;citation_author=Enrique E Alvarez"/>
  <meta name="citation_reference" content="citation_title=Bayesian estimation for a parametric Markov renewal model applied to seismic data;citation_publisher=Institute of Mathematical Statistics; Bernoulli Society;citation_volume=8;citation_author=Ilenia Epifani;citation_author=Lucia Ladelli;citation_author=Antonio Pievatolo"/>
  <meta name="citation_reference" content="citation_title=A Foxp2 mutation implicated in human speech deficits alters sequencing of ultrasonic vocalizations in adult male mice;citation_volume=10;citation_author=J. Chabout;citation_author=A. Sarkar;citation_author=S. Patel;citation_author=T. Raiden;citation_author=D. B. Dunson;citation_author=S. E. Fisher;citation_author=E. D. Jarvis"/>
  <meta name="citation_reference" content="citation_title=A predictive approach to model selection;citation_publisher=Taylor &amp; Francis;citation_volume=74;citation_author=Seymour Geisser;citation_author=William F Eddy"/>
  <meta name="citation_reference" content="citation_title=Asymptotic equivalence of Bayes cross validation and widely applicable information criterion in singular learning theory;citation_volume=11;citation_author=Sumio Watanabe;citation_author=Manfred Opper"/>
  <meta name="citation_reference" content="citation_title=Identification of FOXP2 truncation as a novel cause of developmental speech and language deficits;citation_publisher=Elsevier;citation_volume=76;citation_author=Kay D MacDermot;citation_author=Elena Bonora;citation_author=Nuala Sykes;citation_author=Anne-Marie Coupe;citation_author=Cecilia SL Lai;citation_author=Sonja C Vernes;citation_author=Faraneh Vargha-Khadem;citation_author=Fiona McKenzie;citation_author=Robert L Smith;citation_author=Anthony P Monaco;citation_author=Identification of FOXP2 truncation as a novel cause of developmental speech and language deficits"/>
  <meta name="citation_reference" content="citation_title=Ultrasonic vocalization impairment of Foxp2 (R552H) knockin mice related to speech-language disorder and abnormality of Purkinje cells;citation_volume=105;citation_author=E. Fujita;citation_author=Y. Tanabe;citation_author=A. Shiota;citation_author=M. Ueda;citation_author=K. Suwa;citation_author=M. Y. Momoi;citation_author=T. Momoi"/>
  <meta name="citation_reference" content="citation_title=Ultrasonic vocalizations of adult male Foxp2-mutant mice: Behavioral contexts of arousal and emotion;citation_volume=15;citation_author=S. Gaub;citation_author=S. E. Fisher;citation_author=G. Ehret"/>
  <meta name="citation_reference" content="citation_title=Knockout of Foxp2 disrupts vocal development in mice;citation_volume=6;citation_author=G. A. Castellucci;citation_author=M. J. McGinley;citation_author=D. A. McCormick"/>
  <meta name="citation_reference" content="citation_title=msSurv: An R package for nonparametric estimation of multistate models;citation_volume=50;citation_author=Nicole Ferguson;citation_author=Somnath Datta;citation_author=Guy Brock"/>
  <meta name="citation_reference" content="citation_title=Conjugate classes for gamma distributions;citation_publisher=JSTOR;citation_volume=2;citation_author=Eivind Damsleth"/>
  <meta name="citation_reference" content="citation_title=Fast and accurate approximation of the full conditional for gamma shape parameters;citation_publisher=Taylor &amp; Francis;citation_volume=28;citation_author=Jeffrey W Miller"/>
  <meta name="citation_reference" content="citation_title=Markov models for covariate dependence of binary sequences;citation_publisher=JSTOR;citation_volume=41;citation_author=Larry R Muenz;citation_author=Lawrence V Rubinstein"/>
  <meta name="citation_reference" content="citation_title=The analysis of covariates in multi-fate Markov chain nest-failure models;citation_volume=34;citation_author=Matthew A Etterson;citation_author=Brian Olsen;citation_author=Russell S Greenberg"/>
  <meta name="citation_reference" content="citation_title=Effect of gender, age, transmission category, and antiretroviral therapy on the progression of human immunodeficiency virus infection using multistate Markov models;citation_publisher=JSTOR;citation_volume=9;citation_author=Ahmadou Alioum;citation_author=Valériane Leroy;citation_author=Daniel Commenges;citation_author=François Dabis;citation_author=Roger Salamon;citation_author=Groupe Clinique du SIDA en Aquitaine;citation_author=Effect of gender, age, transmission category, and antiretroviral therapy on the progression of human immunodeficiency virus infection using multistate Markov models"/>
  <meta name="citation_reference" content="citation_title=Analyzing sequential categorical data: Individual variation in Markov chains;citation_publisher=Springer;citation_volume=55;citation_author=William Gradner"/>
  <meta name="citation_reference" content="citation_title=A higher order Markov model for analyzing covariate dependence;citation_publisher=Elsevier;citation_volume=30;citation_author=M Ataharul Islam;citation_author=Rafiqul Islam Chowdhury"/>
  <meta name="citation_reference" content="citation_title=Bayesian inference for Markov chains;citation_publisher=Cambridge University Press;citation_volume=39;citation_author=Peter Eichelsbacher;citation_author=Ayalvadi Ganesh"/>
  <meta name="citation_reference" content="citation_title=Bayesian analysis for reversible Markov chains;citation_publisher=JSTOR;citation_volume=34;citation_author=Persi Diaconis;citation_author=Silke WW Rolles"/>
  <meta name="citation_reference" content="citation_title=Bayesian Markov models consistently outperform PWMs at predicting motifs in nucleotide sequences;citation_publisher=Oxford University Press;citation_volume=44;citation_author=Matthias Siebert;citation_author=Johannes Söding"/>
  <meta name="citation_reference" content="citation_title=Bayesian comparison of Markov models of molecular dynamics with detailed balance constraint;citation_publisher=American Institute of Physics;citation_volume=131;citation_author=Sergio Bacallado;citation_author=John D Chodera;citation_author=Vijay Pande"/>
  <meta name="citation_reference" content="citation_title=Markov models for Bayesian analysis about transit route origin–destination matrices;citation_publisher=Elsevier;citation_volume=43;citation_author=Baibing Li"/>
  <meta name="citation_reference" content="citation_title=MKVPCI: A computer program for Markov models with piecewise constant intensities and covariates;citation_publisher=Elsevier;citation_volume=64;citation_author=Ahmadou Alioum;citation_author=Daniel Commenges"/>
  <meta name="citation_reference" content="citation_title=MARKOV: A computer program for multi-state Markov models with covariables;citation_publisher=Elsevier;citation_volume=47;citation_author=Guillermo Marshall;citation_author=Wensheng Guo;citation_author=Richard H Jones"/>
  <meta name="citation_reference" content="citation_title=Multi-state models for panel data: The msm package for R;citation_volume=38;citation_author=Christopher Jackson"/>
  <meta name="citation_reference" content="citation_title=Discrete time Markov chains with R;citation_volume=9;citation_doi=10.32614/RJ-2017-036;citation_author=Giorgio Alfredo Spedicato"/>
  <meta name="citation_reference" content="citation_title=The analysis of asthma control under a Markov assumption with use of covariates;citation_publisher=Wiley Online Library;citation_volume=22;citation_author=P Saint-Pierre;citation_author=C Combescure;citation_author=JP Daures;citation_author=P Godard"/>
  <meta name="citation_reference" content="citation_title=Assessment of variations in control of asthma over time;citation_publisher=Eur Respiratory Soc;citation_volume=22;citation_author=Christelle Combescure;citation_author=Pascal Chanez;citation_author=Philippe Saint-Pierre;citation_author=Jean Pierre Daures;citation_author=Herve Proudhon;citation_author=Philippe Godard;citation_author=Assessment of variations in control of asthma over time"/>
  <meta name="citation_reference" content="citation_title=Development and validation of a questionnaire to measure asthma control;citation_publisher=Eur Respiratory Soc;citation_volume=14;citation_author=EF Juniper;citation_author=PM O’byrne;citation_author=GH Guyatt;citation_author=PJ Ferrie;citation_author=DR King"/>
  <meta name="citation_reference" content="citation_title=Hidden semi Markov models for multiple observation sequences: The mhsmm package for R;citation_volume=39;citation_author=Jared O’Connell;citation_author=Søren Højsgaard"/>
  <meta name="citation_reference" content="citation_title=Hhsmm: An R package for hidden hybrid Markov/semi-Markov models;citation_volume=1;citation_author=Morteza Amini;citation_author=Afarin Bayat;citation_author=Reza Salehian"/>
  <meta name="citation_reference" content="citation_title=Scenario-based assessments in writing: An experimental study;citation_publisher=Taylor &amp; Francis;citation_volume=24;citation_author=Mo Zhang;citation_author=Peter W Rijn;citation_author=Paul Deane;citation_author=Randy E Bennett"/>
  <meta name="citation_reference" content="citation_title=Bayesian nonhomogeneous markov models via pólya-gamma data augmentation with applications to rainfall modeling;citation_publisher=Institute of Mathematical Statistics;citation_volume=11;citation_author=Tracy Holsclaw;citation_author=Arthur M Greene;citation_author=Andrew W Robertson;citation_author=Padhraic Smyth"/>
  <meta name="citation_reference" content="citation_title=Gene hunting with hidden Markov model knockoffs;citation_publisher=Oxford University Press;citation_volume=106;citation_author=Matteo Sesia;citation_author=Chiara Sabatti;citation_author=Emmanuel J Candès"/>
  <meta name="citation_reference" content="citation_title=Markov renewal processes: Definitions and preliminary properties;citation_publisher=JSTOR;citation_volume=32;citation_author=Ronald Pyke"/>
  <!--radix_placeholder_rmarkdown_metadata-->

<script type="text/json" id="radix-rmarkdown-metadata">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","date_received","journal","volume","issue","slug","citation_url","packages","preview","bibliography","CTV","legacy_pdf","legacy_converted","output","draft","pdf_url","doi","creative_commons","csl","canonical_url"]}},"value":[{"type":"character","attributes":{},"value":["BMRMM: An R Package for Bayesian Markov (Renewal) Mixed Models"]},{"type":"character","attributes":{},"value":["We introduce the BMRMM\npackage implementing Bayesian inference for a class of Markov renewal\nmixed models which can characterize the stochastic dynamics of a\ncollection of sequences, each comprising alternative instances of\ncategorical states and associated continuous duration times, while\nbeing influenced by a set of exogenous factors as well as a 'random'\nindividual. The default setting flexibly models the state transition\nprobabilities using mixtures of Dirichlet distributions and the\nduration times using mixtures of gamma kernels while also allowing\nvariable selection for both. Modeling such data using simpler Markov\nmixed models also remains an option, either by ignoring the duration\ntimes altogether or by replacing them with instances of an additional\ncategory obtained by discretizing them by a user-specified unit. The\noption is also useful when data on duration times may not be available\nin the first place. We demonstrate the package's utility using two\ndata sets."]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Yutong Wu"]},{"type":"character","attributes":{},"value":["Department of Mechanical Engineering"]},{"type":"character","attributes":{},"value":["The University of Texas at Austin","204 E Dean Keeton St C2200, Austin, TX 78712-1591","United States","ORCID: 0000-0001-7828-9981","[yutong.wu@utexas.edu](yutong.wu@utexas.edu){.uri}\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Abhra Sarkar"]},{"type":"character","attributes":{},"value":["Department of Statistics and Data Sciences"]},{"type":"character","attributes":{},"value":["The University of Texas at Austin","2317 Speedway D9800, Austin, TX 78712-1823","United States","ORCID: 0000-0002-6924-8464","[abhra.sarkar@utexas.edu](abhra.sarkar@utexas.edu){.uri}\n"]}]}]},{"type":"character","attributes":{},"value":["2025-01-11"]},{"type":"character","attributes":{},"value":["2023-02-07"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"integer","attributes":{},"value":[192]},{"type":"integer","attributes":{},"value":[211]}]},{"type":"integer","attributes":{},"value":[16]},{"type":"integer","attributes":{},"value":[1]},{"type":"character","attributes":{},"value":["RJ-2024-011"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2024-011"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"character","attributes":{},"value":["BMRMM","markovchain","msm","SemiMarkov","SMM","mhsmm","hhsmm","Countr","ARPobservation","catdap","vcd"]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["preview.png"]},{"type":"character","attributes":{},"value":["BMRMM_Package.bib"]},{"type":"character","attributes":{},"value":["Distributions","Finance","MissingData","Survival","TeachingStatistics"]},{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[true]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc","mathjax","md_extension"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js"]},{"type":"character","attributes":{},"value":["-tex_math_single_backslash"]}]}]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["RJ-2024-011.pdf"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2024-011"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"character","attributes":{},"value":["/home/mitchell/R/x86_64-pc-linux-gnu-library/4.4/rjtools/rjournal.csl"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2024-011/"]}]}
</script>
<!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["BMRMM_Package.bib","BMRMM_Package.tex","Data and Codes/asthma.csv","Data and Codes/BMRMM_Package.R","figs/asthma_diag_comp3.pdf","figs/asthma_diag_comp3.png","figs/asthma_diag_tp.pdf","figs/asthma_diag_tp.png","figs/asthma_mix_param.pdf","figs/asthma_sev_men.pdf","figs/asthma_sev_men.png","figs/asthma_sev_women.pdf","figs/asthma_sev_women.png","figs/compare_sev_m.pdf","figs/compare_sev_m.png","figs/compare_sev_w.pdf","figs/compare_sev_w.png","figs/data-model.pdf","figs/data-model.png","figs/diag_FU_um.pdf","figs/diag_FU_um.png","figs/diag_isi_comp2.pdf","figs/diag_isi_comp2.png","figs/global_isi_asthma.pdf","figs/global_isi_asthma.png","figs/global_isi_foxp2.pdf","figs/global_isi_foxp2.png","figs/global_trans_asthma.pdf","figs/global_trans_asthma.png","figs/global_trans_foxp2.pdf","figs/global_trans_foxp2.png","figs/hist_by_comp_asthma.pdf","figs/hist_by_comp_asthma.png","figs/hist_by_comp.pdf","figs/hist_by_comp.png","figs/hist_isi_asthma.pdf","figs/hist_isi_asthma.png","figs/hist_isi.pdf","figs/hist_isi.png","figs/local_A.pdf","figs/local_A.png","figs/local_L.pdf","figs/local_L.png","figs/local_U.pdf","figs/local_U.png","figs/post_mean_FA.pdf","figs/post_mean_FA.png","figs/post_mean_FL.pdf","figs/post_mean_WL.pdf","figs/post_mean_WL.png","RJ-2024-011.zip","RJournal.sty","RJwrapper.log","RJwrapper.tex","table_data_1.csv"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
<meta name="distill:offset" content="../.."/>

<script type="application/javascript">

  window.headroom_prevent_pin = false;

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // initialize headroom for banner
    var header = $('header').get(0);
    var headerHeight = header.offsetHeight;
    var headroom = new Headroom(header, {
      tolerance: 5,
      onPin : function() {
        if (window.headroom_prevent_pin) {
          window.headroom_prevent_pin = false;
          headroom.unpin();
        }
      }
    });
    headroom.init();
    if(window.location.hash)
      headroom.unpin();
    $(header).addClass('headroom--transition');

    // offset scroll location for banner on hash change
    // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
    window.addEventListener("hashchange", function(event) {
      window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
    });

    // responsive menu
    $('.distill-site-header').each(function(i, val) {
      var topnav = $(this);
      var toggle = topnav.find('.nav-toggle');
      toggle.on('click', function() {
        topnav.toggleClass('responsive');
      });
    });

    // nav dropdowns
    $('.nav-dropbtn').click(function(e) {
      $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
      $(this).parent().siblings('.nav-dropdown')
         .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $("body").click(function(e){
      $('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $(".nav-dropdown").click(function(e){
      e.stopPropagation();
    });
  });
</script>

<style type="text/css">

/* Theme (user-documented overrideables for nav appearance) */

.distill-site-nav {
  color: rgba(255, 255, 255, 0.8);
  background-color: #0F2E3D;
  font-size: 15px;
  font-weight: 300;
}

.distill-site-nav a {
  color: inherit;
  text-decoration: none;
}

.distill-site-nav a:hover {
  color: white;
}

@media print {
  .distill-site-nav {
    display: none;
  }
}

.distill-site-header {

}

.distill-site-footer {

}


/* Site Header */

.distill-site-header {
  width: 100%;
  box-sizing: border-box;
  z-index: 3;
}

.distill-site-header .nav-left {
  display: inline-block;
  margin-left: 8px;
}

@media screen and (max-width: 768px) {
  .distill-site-header .nav-left {
    margin-left: 0;
  }
}


.distill-site-header .nav-right {
  float: right;
  margin-right: 8px;
}

.distill-site-header a,
.distill-site-header .title {
  display: inline-block;
  text-align: center;
  padding: 14px 10px 14px 10px;
}

.distill-site-header .title {
  font-size: 18px;
  min-width: 150px;
}

.distill-site-header .logo {
  padding: 0;
}

.distill-site-header .logo img {
  display: none;
  max-height: 20px;
  width: auto;
  margin-bottom: -4px;
}

.distill-site-header .nav-image img {
  max-height: 18px;
  width: auto;
  display: inline-block;
  margin-bottom: -3px;
}



@media screen and (min-width: 1000px) {
  .distill-site-header .logo img {
    display: inline-block;
  }
  .distill-site-header .nav-left {
    margin-left: 20px;
  }
  .distill-site-header .nav-right {
    margin-right: 20px;
  }
  .distill-site-header .title {
    padding-left: 12px;
  }
}


.distill-site-header .nav-toggle {
  display: none;
}

.nav-dropdown {
  display: inline-block;
  position: relative;
}

.nav-dropdown .nav-dropbtn {
  border: none;
  outline: none;
  color: rgba(255, 255, 255, 0.8);
  padding: 16px 10px;
  background-color: transparent;
  font-family: inherit;
  font-size: inherit;
  font-weight: inherit;
  margin: 0;
  margin-top: 1px;
  z-index: 2;
}

.nav-dropdown-content {
  display: none;
  position: absolute;
  background-color: white;
  min-width: 200px;
  border: 1px solid rgba(0,0,0,0.15);
  border-radius: 4px;
  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
  z-index: 1;
  margin-top: 2px;
  white-space: nowrap;
  padding-top: 4px;
  padding-bottom: 4px;
}

.nav-dropdown-content hr {
  margin-top: 4px;
  margin-bottom: 4px;
  border: none;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.nav-dropdown-active {
  display: block;
}

.nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
  color: black;
  padding: 6px 24px;
  text-decoration: none;
  display: block;
  text-align: left;
}

.nav-dropdown-content .nav-dropdown-header {
  display: block;
  padding: 5px 24px;
  padding-bottom: 0;
  text-transform: uppercase;
  font-size: 14px;
  color: #999999;
  white-space: nowrap;
}

.nav-dropdown:hover .nav-dropbtn {
  color: white;
}

.nav-dropdown-content a:hover {
  background-color: #ddd;
  color: black;
}

.nav-right .nav-dropdown-content {
  margin-left: -45%;
  right: 0;
}

@media screen and (max-width: 768px) {
  .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
  .distill-site-header a.nav-toggle {
    float: right;
    display: block;
  }
  .distill-site-header .title {
    margin-left: 0;
  }
  .distill-site-header .nav-right {
    margin-right: 0;
  }
  .distill-site-header {
    overflow: hidden;
  }
  .nav-right .nav-dropdown-content {
    margin-left: 0;
  }
}


@media screen and (max-width: 768px) {
  .distill-site-header.responsive {position: relative; min-height: 500px; }
  .distill-site-header.responsive a.nav-toggle {
    position: absolute;
    right: 0;
    top: 0;
  }
  .distill-site-header.responsive a,
  .distill-site-header.responsive .nav-dropdown {
    display: block;
    text-align: left;
  }
  .distill-site-header.responsive .nav-left,
  .distill-site-header.responsive .nav-right {
    width: 100%;
  }
  .distill-site-header.responsive .nav-dropdown {float: none;}
  .distill-site-header.responsive .nav-dropdown-content {position: relative;}
  .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
    display: block;
    width: 100%;
    text-align: left;
  }
}

/* Site Footer */

.distill-site-footer {
  width: 100%;
  overflow: hidden;
  box-sizing: border-box;
  z-index: 3;
  margin-top: 30px;
  padding-top: 30px;
  padding-bottom: 30px;
  text-align: center;
}

/* Headroom */

d-title {
  padding-top: 6rem;
}

@media print {
  d-title {
    padding-top: 4rem;
  }
}

.headroom {
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
}

.headroom--transition {
  transition: all .4s ease-in-out;
}

.headroom--unpinned {
  top: -100px;
}

.headroom--pinned {
  top: 0;
}

/* adjust viewport for navbar height */
/* helps vertically center bootstrap (non-distill) content */
.min-vh-100 {
  min-height: calc(100vh - 100px) !important;
}

</style>

<script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="../../site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"/>
<link href="../../site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"/>
<script src="../../site_libs/headroom-0.9.4/headroom.min.js"></script>
<!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

<style type="text/css">

body {
  background-color: white;
}

.pandoc-table {
  width: 100%;
}

.pandoc-table>caption {
  margin-bottom: 10px;
}

.pandoc-table th:not([align]) {
  text-align: left;
}

.pagedtable-footer {
  font-size: 15px;
}

d-byline .byline {
  grid-template-columns: 2fr 2fr;
}

d-byline .byline h3 {
  margin-block-start: 1.5em;
}

d-byline .byline .authors-affiliations h3 {
  margin-block-start: 0.5em;
}

.authors-affiliations .orcid-id {
  width: 16px;
  height:16px;
  margin-left: 4px;
  margin-right: 4px;
  vertical-align: middle;
  padding-bottom: 2px;
}

d-title .dt-tags {
  margin-top: 1em;
  grid-column: text;
}

.dt-tags .dt-tag {
  text-decoration: none;
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0em 0.4em;
  margin-right: 0.5em;
  margin-bottom: 0.4em;
  font-size: 70%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

d-article table.gt_table td,
d-article table.gt_table th {
  border-bottom: none;
  font-size: 100%;
}

.html-widget {
  margin-bottom: 2.0em;
}

.l-screen-inset {
  padding-right: 16px;
}

.l-screen .caption {
  margin-left: 10px;
}

.shaded {
  background: rgb(247, 247, 247);
  padding-top: 20px;
  padding-bottom: 20px;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .html-widget {
  margin-bottom: 0;
  border: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .shaded-content {
  background: white;
}

.text-output {
  margin-top: 0;
  line-height: 1.5em;
}

.hidden {
  display: none !important;
}

hr.section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  margin: 0px;
}


d-byline {
  border-top: none;
}

d-article {
  padding-top: 2.5rem;
  padding-bottom: 30px;
  border-top: none;
}

d-appendix {
  padding-top: 30px;
}

d-article>p>img {
  width: 100%;
}

d-article h2 {
  margin: 1rem 0 1.5rem 0;
}

d-article h3 {
  margin-top: 1.5rem;
}

d-article iframe {
  border: 1px solid rgba(0, 0, 0, 0.1);
  margin-bottom: 2.0em;
  width: 100%;
}

/* Tweak code blocks */

d-article div.sourceCode code,
d-article pre code {
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
}

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: auto;
}

d-article div.sourceCode {
  background-color: white;
}

d-article div.sourceCode pre {
  padding-left: 10px;
  font-size: 12px;
  border-left: 2px solid rgba(0,0,0,0.1);
}

d-article pre {
  font-size: 12px;
  color: black;
  background: none;
  margin-top: 0;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

d-article pre a {
  border-bottom: none;
}

d-article pre a:hover {
  border-bottom: none;
  text-decoration: underline;
}

d-article details {
  grid-column: text;
  margin-bottom: 0.8em;
}

@media(min-width: 768px) {

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: visible !important;
}

d-article div.sourceCode pre {
  padding-left: 18px;
  font-size: 14px;
}

/* tweak for Pandoc numbered line within distill */
d-article pre.numberSource code > span {
    left: -2em;
}

d-article pre {
  font-size: 14px;
}

}

figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

/* CSS for d-contents */

.d-contents {
  grid-column: text;
  color: rgba(0,0,0,0.8);
  font-size: 0.9em;
  padding-bottom: 1em;
  margin-bottom: 1em;
  padding-bottom: 0.5em;
  margin-bottom: 1em;
  padding-left: 0.25em;
  justify-self: start;
}

@media(min-width: 1000px) {
  .d-contents.d-contents-float {
    height: 0;
    grid-column-start: 1;
    grid-column-end: 4;
    justify-self: center;
    padding-right: 3em;
    padding-left: 2em;
  }
}

.d-contents nav h3 {
  font-size: 18px;
  margin-top: 0;
  margin-bottom: 1em;
}

.d-contents li {
  list-style-type: none
}

.d-contents nav > ul {
  padding-left: 0;
}

.d-contents ul {
  padding-left: 1em
}

.d-contents nav ul li {
  margin-top: 0.6em;
  margin-bottom: 0.2em;
}

.d-contents nav a {
  font-size: 13px;
  border-bottom: none;
  text-decoration: none
  color: rgba(0, 0, 0, 0.8);
}

.d-contents nav a:hover {
  text-decoration: underline solid rgba(0, 0, 0, 0.6)
}

.d-contents nav > ul > li > a {
  font-weight: 600;
}

.d-contents nav > ul > li > ul {
  font-weight: inherit;
}

.d-contents nav > ul > li > ul > li {
  margin-top: 0.2em;
}


.d-contents nav ul {
  margin-top: 0;
  margin-bottom: 0.25em;
}

.d-article-with-toc h2:nth-child(2) {
  margin-top: 0;
}


/* Figure */

.figure {
  position: relative;
  margin-bottom: 2.5em;
  margin-top: 1.5em;
}

.figure .caption {
  color: rgba(0, 0, 0, 0.6);
  font-size: 12px;
  line-height: 1.5em;
}

.figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

.figure .caption a {
  color: rgba(0, 0, 0, 0.6);
}

.figure .caption b,
.figure .caption strong, {
  font-weight: 600;
  color: rgba(0, 0, 0, 1.0);
}

/* Citations */

d-article .citation {
  color: inherit;
  cursor: inherit;
}

div.hanging-indent{
  margin-left: 1em; text-indent: -1em;
}

/* Citation hover box */

.tippy-box[data-theme~=light-border] {
  background-color: rgba(250, 250, 250, 0.95);
}

.tippy-content > p {
  margin-bottom: 0;
  padding: 2px;
}


/* Tweak 1000px media break to show more text */

@media(min-width: 1000px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 16px;
  }

  .grid {
    grid-column-gap: 16px;
  }

  d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
  }
  figure .caption, .figure .caption, figure figcaption {
    font-size: 13px;
  }
}

@media(min-width: 1180px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 32px;
  }

  .grid {
    grid-column-gap: 32px;
  }
}


/* Get the citation styles for the appendix (not auto-injected on render since
   we do our own rendering of the citation appendix) */

d-appendix .citation-appendix,
.d-appendix .citation-appendix {
  font-size: 11px;
  line-height: 15px;
  border-left: 1px solid rgba(0, 0, 0, 0.1);
  padding-left: 18px;
  border: 1px solid rgba(0,0,0,0.1);
  background: rgba(0, 0, 0, 0.02);
  padding: 10px 18px;
  border-radius: 3px;
  color: rgba(150, 150, 150, 1);
  overflow: hidden;
  margin-top: -12px;
  white-space: pre-wrap;
  word-wrap: break-word;
}

/* Include appendix styles here so they can be overridden */

d-appendix {
  contain: layout style;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-top: 60px;
  margin-bottom: 0;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  color: rgba(0,0,0,0.5);
  padding-top: 60px;
  padding-bottom: 48px;
}

d-appendix h3 {
  grid-column: page-start / text-start;
  font-size: 15px;
  font-weight: 500;
  margin-top: 1em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.65);
}

d-appendix h3 + * {
  margin-top: 1em;
}

d-appendix ol {
  padding: 0 0 0 15px;
}

@media (min-width: 768px) {
  d-appendix ol {
    padding: 0 0 0 30px;
    margin-left: -30px;
  }
}

d-appendix li {
  margin-bottom: 1em;
}

d-appendix a {
  color: rgba(0, 0, 0, 0.6);
}

d-appendix > * {
  grid-column: text;
}

d-appendix > d-footnote-list,
d-appendix > d-citation-list,
d-appendix > distill-appendix {
  grid-column: screen;
}

/* Include footnote styles here so they can be overridden */

d-footnote-list {
  contain: layout style;
}

d-footnote-list > * {
  grid-column: text;
}

d-footnote-list a.footnote-backlink {
  color: rgba(0,0,0,0.3);
  padding-left: 0.5em;
}



/* Anchor.js */

.anchorjs-link {
  /*transition: all .25s linear; */
  text-decoration: none;
  border-bottom: none;
}
*:hover > .anchorjs-link {
  margin-left: -1.125em !important;
  text-decoration: none;
  border-bottom: none;
}

/* Social footer */

.social_footer {
  margin-top: 30px;
  margin-bottom: 0;
  color: rgba(0,0,0,0.67);
}

.disqus-comments {
  margin-right: 30px;
}

.disqus-comment-count {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  cursor: pointer;
}

#disqus_thread {
  margin-top: 30px;
}

.article-sharing a {
  border-bottom: none;
  margin-right: 8px;
}

.article-sharing a:hover {
  border-bottom: none;
}

.sidebar-section.subscribe {
  font-size: 12px;
  line-height: 1.6em;
}

.subscribe p {
  margin-bottom: 0.5em;
}


.article-footer .subscribe {
  font-size: 15px;
  margin-top: 45px;
}


.sidebar-section.custom {
  font-size: 12px;
  line-height: 1.6em;
}

.custom p {
  margin-bottom: 0.5em;
}

/* Styles for listing layout (hide title) */
.layout-listing d-title, .layout-listing .d-title {
  display: none;
}

/* Styles for posts lists (not auto-injected) */


.posts-with-sidebar {
  padding-left: 45px;
  padding-right: 45px;
}

.posts-list .description h2,
.posts-list .description p {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

.posts-list .description h2 {
  font-weight: 700;
  border-bottom: none;
  padding-bottom: 0;
}

.posts-list h2.post-tag {
  border-bottom: 1px solid rgba(0, 0, 0, 0.2);
  padding-bottom: 12px;
}
.posts-list {
  margin-top: 60px;
  margin-bottom: 24px;
}

.posts-list .post-preview {
  text-decoration: none;
  overflow: hidden;
  display: block;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  padding: 24px 0;
}

.post-preview-last {
  border-bottom: none !important;
}

.posts-list .posts-list-caption {
  grid-column: screen;
  font-weight: 400;
}

.posts-list .post-preview h2 {
  margin: 0 0 6px 0;
  line-height: 1.2em;
  font-style: normal;
  font-size: 24px;
}

.posts-list .post-preview p {
  margin: 0 0 12px 0;
  line-height: 1.4em;
  font-size: 16px;
}

.posts-list .post-preview .thumbnail {
  box-sizing: border-box;
  margin-bottom: 24px;
  position: relative;
  max-width: 500px;
}
.posts-list .post-preview img {
  width: 100%;
  display: block;
}

.posts-list .metadata {
  font-size: 12px;
  line-height: 1.4em;
  margin-bottom: 18px;
}

.posts-list .metadata > * {
  display: inline-block;
}

.posts-list .metadata .publishedDate {
  margin-right: 2em;
}

.posts-list .metadata .dt-authors {
  display: block;
  margin-top: 0.3em;
  margin-right: 2em;
}

.posts-list .dt-tags {
  display: block;
  line-height: 1em;
}

.posts-list .dt-tags .dt-tag {
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0.3em 0.4em;
  margin-right: 0.2em;
  margin-bottom: 0.4em;
  font-size: 60%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

.posts-list img {
  opacity: 1;
}

.posts-list img[data-src] {
  opacity: 0;
}

.posts-more {
  clear: both;
}


.posts-sidebar {
  font-size: 16px;
}

.posts-sidebar h3 {
  font-size: 16px;
  margin-top: 0;
  margin-bottom: 0.5em;
  font-weight: 400;
  text-transform: uppercase;
}

.sidebar-section {
  margin-bottom: 30px;
}

.categories ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.categories li {
  color: rgba(0, 0, 0, 0.8);
  margin-bottom: 0;
}

.categories li>a {
  border-bottom: none;
}

.categories li>a:hover {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
}

.categories .active {
  font-weight: 600;
}

.categories .category-count {
  color: rgba(0, 0, 0, 0.4);
}


@media(min-width: 768px) {
  .posts-list .post-preview h2 {
    font-size: 26px;
  }
  .posts-list .post-preview .thumbnail {
    float: right;
    width: 30%;
    margin-bottom: 0;
  }
  .posts-list .post-preview .description {
    float: left;
    width: 45%;
  }
  .posts-list .post-preview .metadata {
    float: left;
    width: 20%;
    margin-top: 8px;
  }
  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.5em;
    font-size: 16px;
  }
  .posts-with-sidebar .posts-list {
    float: left;
    width: 75%;
  }
  .posts-with-sidebar .posts-sidebar {
    float: right;
    width: 20%;
    margin-top: 60px;
    padding-top: 24px;
    padding-bottom: 24px;
  }
}


/* Improve display for browsers without grid (IE/Edge <= 15) */

.downlevel {
  line-height: 1.6em;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  margin: 0;
}

.downlevel .d-title {
  padding-top: 6rem;
  padding-bottom: 1.5rem;
}

.downlevel .d-title h1 {
  font-size: 50px;
  font-weight: 700;
  line-height: 1.1em;
  margin: 0 0 0.5rem;
}

.downlevel .d-title p {
  font-weight: 300;
  font-size: 1.2rem;
  line-height: 1.55em;
  margin-top: 0;
}

.downlevel .d-byline {
  padding-top: 0.8em;
  padding-bottom: 0.8em;
  font-size: 0.8rem;
  line-height: 1.8em;
}

.downlevel .section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
}

.downlevel .d-article {
  font-size: 1.06rem;
  line-height: 1.7em;
  padding-top: 1rem;
  padding-bottom: 2rem;
}


.downlevel .d-appendix {
  padding-left: 0;
  padding-right: 0;
  max-width: none;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.5);
  padding-top: 40px;
  padding-bottom: 48px;
}

.downlevel .footnotes ol {
  padding-left: 13px;
}

.downlevel .base-grid,
.downlevel .distill-header,
.downlevel .d-title,
.downlevel .d-abstract,
.downlevel .d-article,
.downlevel .d-appendix,
.downlevel .distill-appendix,
.downlevel .d-byline,
.downlevel .d-footnote-list,
.downlevel .d-citation-list,
.downlevel .distill-footer,
.downlevel .appendix-bottom,
.downlevel .posts-container {
  padding-left: 40px;
  padding-right: 40px;
}

@media(min-width: 768px) {
  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
  padding-left: 150px;
  padding-right: 150px;
  max-width: 900px;
}
}

.downlevel pre code {
  display: block;
  border-left: 2px solid rgba(0, 0, 0, .1);
  padding: 0 0 0 20px;
  font-size: 14px;
}

.downlevel code, .downlevel pre {
  color: black;
  background: none;
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

.downlevel .posts-list .post-preview {
  color: inherit;
}



</style>

<script type="application/javascript">

function is_downlevel_browser() {
  if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                 window.navigator.userAgent)) {
    return true;
  } else {
    return window.load_distill_framework === undefined;
  }
}

// show body when load is complete
function on_load_complete() {

  // add anchors
  if (window.anchors) {
    window.anchors.options.placement = 'left';
    window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
  }


  // set body to visible
  document.body.style.visibility = 'visible';

  // force redraw for leaflet widgets
  if (window.HTMLWidgets) {
    var maps = window.HTMLWidgets.findAll(".leaflet");
    $.each(maps, function(i, el) {
      var map = this.getMap();
      map.invalidateSize();
      map.eachLayer(function(layer) {
        if (layer instanceof L.TileLayer)
          layer.redraw();
      });
    });
  }

  // trigger 'shown' so htmlwidgets resize
  $('d-article').trigger('shown');
}

function init_distill() {

  init_common();

  // create front matter
  var front_matter = $('<d-front-matter></d-front-matter>');
  $('#distill-front-matter').wrap(front_matter);

  // create d-title
  $('.d-title').changeElementType('d-title');

  // separator
  var separator = '<hr class="section-separator" style="clear: both"/>';
  // prepend separator above appendix
  $('.d-byline').before(separator);
  $('.d-article').before(separator);

  // create d-byline
  var byline = $('<d-byline></d-byline>');
  $('.d-byline').replaceWith(byline);

  // create d-article
  var article = $('<d-article></d-article>');
  $('.d-article').wrap(article).children().unwrap();

  // move posts container into article
  $('.posts-container').appendTo($('d-article'));

  // create d-appendix
  $('.d-appendix').changeElementType('d-appendix');

  // flag indicating that we have appendix items
  var appendix = $('.appendix-bottom').children('h3').length > 0;

  // replace footnotes with <d-footnote>
  $('.footnote-ref').each(function(i, val) {
    appendix = true;
    var href = $(this).attr('href');
    var id = href.replace('#', '');
    var fn = $('#' + id);
    var fn_p = $('#' + id + '>p');
    fn_p.find('.footnote-back').remove();
    var text = fn_p.html();
    var dtfn = $('<d-footnote></d-footnote>');
    dtfn.html(text);
    $(this).replaceWith(dtfn);
  });
  // remove footnotes
  $('.footnotes').remove();

  // move refs into #references-listing
  $('#references-listing').replaceWith($('#refs'));

  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    var id = $(this).attr('id');
    $('.d-contents a[href="#' + id + '"]').parent().remove();
    appendix = true;
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
  });

  // show d-appendix if we have appendix content
  $("d-appendix").css('display', appendix ? 'grid' : 'none');

  // localize layout chunks to just output
  $('.layout-chunk').each(function(i, val) {

    // capture layout
    var layout = $(this).attr('data-layout');

    // apply layout to markdown level block elements
    var elements = $(this).children().not('details, div.sourceCode, pre, script');
    elements.each(function(i, el) {
      var layout_div = $('<div class="' + layout + '"></div>');
      if (layout_div.hasClass('shaded')) {
        var shaded_content = $('<div class="shaded-content"></div>');
        $(this).wrap(shaded_content);
        $(this).parent().wrap(layout_div);
      } else {
        $(this).wrap(layout_div);
      }
    });


    // unwrap the layout-chunk div
    $(this).children().unwrap();
  });

  // remove code block used to force  highlighting css
  $('.distill-force-highlighting-css').parent().remove();

  // remove empty line numbers inserted by pandoc when using a
  // custom syntax highlighting theme, except when numbering line
  // in code chunk
  $('pre:not(.numberLines) code.sourceCode a:empty').remove();

  // load distill framework
  load_distill_framework();

  // wait for window.distillRunlevel == 4 to do post processing
  function distill_post_process() {

    if (!window.distillRunlevel || window.distillRunlevel < 4)
      return;

    // hide author/affiliations entirely if we have no authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;
    if (!have_authors)
      $('d-byline').addClass('hidden');

    // article with toc class
    $('.d-contents').parent().addClass('d-article-with-toc');

    // strip links that point to #
    $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

    // add orcid ids
    $('.authors-affiliations').find('.author').each(function(i, el) {
      var orcid_id = front_matter.authors[i].orcidID;
      var author_name = front_matter.authors[i].author
      if (orcid_id) {
        var a = $('<a></a>');
        a.attr('href', 'https://orcid.org/' + orcid_id);
        var img = $('<img></img>');
        img.addClass('orcid-id');
        img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
        img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
        a.append(img);
        $(this).append(a);
      }
    });

    // hide elements of author/affiliations grid that have no value
    function hide_byline_column(caption) {
      $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
    }

    // affiliations
    var have_affiliations = false;
    for (var i = 0; i<front_matter.authors.length; ++i) {
      var author = front_matter.authors[i];
      if (author.affiliation !== "&nbsp;") {
        have_affiliations = true;
        break;
      }
    }
    if (!have_affiliations)
      $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

    // published date
    if (!front_matter.publishedDate)
      hide_byline_column("Published");

    // document object identifier
    var doi = $('d-byline').find('h3:contains("DOI")');
    var doi_p = doi.next().empty();
    if (!front_matter.doi) {
      // if we have a citation and valid citationText then link to that
      if ($('#citation').length > 0 && front_matter.citationText) {
        doi.html('Citation');
        $('<a href="#citation"></a>')
          .text(front_matter.citationText)
          .appendTo(doi_p);
      } else {
        hide_byline_column("DOI");
      }
    } else {
      $('<a></a>')
         .attr('href', "https://doi.org/" + front_matter.doi)
         .html(front_matter.doi)
         .appendTo(doi_p);
    }

     // change plural form of authors/affiliations
    if (front_matter.authors.length === 1) {
      var grid = $('.authors-affiliations');
      grid.children('h3:contains("Authors")').text('Author');
      grid.children('h3:contains("Affiliations")').text('Affiliation');
    }

    // remove d-appendix and d-footnote-list local styles
    $('d-appendix > style:first-child').remove();
    $('d-footnote-list > style:first-child').remove();

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // hoverable references
    $('span.citation[data-cites]').each(function() {
      const citeChild = $(this).children()[0]
      // Do not process if @xyz has been used without escaping and without bibliography activated
      // https://github.com/rstudio/distill/issues/466
      if (citeChild === undefined) return true

      if (citeChild.nodeName == "D-FOOTNOTE") {
        var fn = citeChild
        $(this).html(fn.shadowRoot.querySelector("sup"))
        $(this).id = fn.id
        fn.remove()
      }
      var refs = $(this).attr('data-cites').split(" ");
      var refHtml = refs.map(function(ref) {
        // Could use CSS.escape too here, we insure backward compatibility in navigator
        return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
      }).join("\n");
      window.tippy(this, {
        allowHTML: true,
        content: refHtml,
        maxWidth: 500,
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start'
      });
    });

    // fix footnotes in tables (#411)
    // replacing broken distill.pub feature
    $('table d-footnote').each(function() {
      // we replace internal showAtNode methode which is triggered when hovering a footnote
      this.hoverBox.showAtNode = function(node) {
        // ported from https://github.com/distillpub/template/pull/105/files
        calcOffset = function(elem) {
            let x = elem.offsetLeft;
            let y = elem.offsetTop;
            // Traverse upwards until an `absolute` element is found or `elem`
            // becomes null.
            while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                x += elem.offsetLeft;
                y += elem.offsetTop;
            }

            return { left: x, top: y };
        }
        // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
        const bbox = node.getBoundingClientRect();
        const offset = calcOffset(node);
        this.show([offset.left + bbox.width, offset.top + bbox.height]);
      }
    })

    // clear polling timer
    clearInterval(tid);

    // show body now that everything is ready
    on_load_complete();
  }

  var tid = setInterval(distill_post_process, 50);
  distill_post_process();

}

function init_downlevel() {

  init_common();

   // insert hr after d-title
  $('.d-title').after($('<hr class="section-separator"/>'));

  // check if we have authors
  var front_matter = JSON.parse($("#distill-front-matter").html());
  var have_authors = front_matter.authors && front_matter.authors.length > 0;

  // manage byline/border
  if (!have_authors)
    $('.d-byline').remove();
  $('.d-byline').after($('<hr class="section-separator"/>'));
  $('.d-byline a').remove();

  // remove toc
  $('.d-contents').remove();

  // move appendix elements
  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
  });


  // inject headers into references and footnotes
  var refs_header = $('<h3></h3>');
  refs_header.text('References');
  $('#refs').prepend(refs_header);

  var footnotes_header = $('<h3></h3');
  footnotes_header.text('Footnotes');
  $('.footnotes').children('hr').first().replaceWith(footnotes_header);

  // move appendix-bottom entries to the bottom
  $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
  $('.appendix-bottom').remove();

  // remove appendix if it's empty
  if ($('.d-appendix').children().length === 0)
    $('.d-appendix').remove();

  // prepend separator above appendix
  $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

  // trim code
  $('pre>code').each(function(i, val) {
    $(this).html($.trim($(this).html()));
  });

  // move posts-container right before article
  $('.posts-container').insertBefore($('.d-article'));

  $('body').addClass('downlevel');

  on_load_complete();
}


function init_common() {

  // jquery plugin to change element types
  (function($) {
    $.fn.changeElementType = function(newType) {
      var attrs = {};

      $.each(this[0].attributes, function(idx, attr) {
        attrs[attr.nodeName] = attr.nodeValue;
      });

      this.replaceWith(function() {
        return $("<" + newType + "/>", attrs).append($(this).contents());
      });
    };
  })(jQuery);

  // prevent underline for linked images
  $('a > img').parent().css({'border-bottom' : 'none'});

  // mark non-body figures created by knitr chunks as 100% width
  $('.layout-chunk').each(function(i, val) {
    var figures = $(this).find('img, .html-widget');
    // ignore leaflet img layers (#106)
    figures = figures.filter(':not(img[class*="leaflet"])')
    if ($(this).attr('data-layout') !== "l-body") {
      figures.css('width', '100%');
    } else {
      figures.css('max-width', '100%');
      figures.filter("[width]").each(function(i, val) {
        var fig = $(this);
        fig.css('width', fig.attr('width') + 'px');
      });

    }
  });

  // auto-append index.html to post-preview links in file: protocol
  // and in rstudio ide preview
  $('.post-preview').each(function(i, val) {
    if (window.location.protocol === "file:")
      $(this).attr('href', $(this).attr('href') + "index.html");
  });

  // get rid of index.html references in header
  if (window.location.protocol !== "file:") {
    $('.distill-site-header a[href]').each(function(i,val) {
      $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
    });
  }

  // add class to pandoc style tables
  $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
  $('.kable-table').children('table').addClass('pandoc-table');

  // add figcaption style to table captions
  $('caption').parent('table').addClass("figcaption");

  // initialize posts list
  if (window.init_posts_list)
    window.init_posts_list();

  // implmement disqus comment link
  $('.disqus-comment-count').click(function() {
    window.headroom_prevent_pin = true;
    $('#disqus_thread').toggleClass('hidden');
    if (!$('#disqus_thread').hasClass('hidden')) {
      var offset = $(this).offset();
      $(window).resize();
      $('html, body').animate({
        scrollTop: offset.top - 35
      });
    }
  });
}

document.addEventListener('DOMContentLoaded', function() {
  if (is_downlevel_browser())
    init_downlevel();
  else
    window.addEventListener('WebComponentsReady', init_distill);
});

</script>

<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.5);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

.distill-site-footer {
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */
</style>
<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.7);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #ffffff;
}

.distill-site-footer {
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #fafafa;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */

.nav-right > a {
  text-transform: uppercase;
}

d-title h1, d-title p, d-title figure,
d-abstract p, d-abstract b {
  grid-column: page;
}

.rj-blue {
  color: #2467bb;
}

ul li {
  line-height: 1.6;
  margin-top: 0em;
  margin-bottom: 0em;
}</style>
<style type="text/css">
/* base style */

/* FONT FAMILIES */

:root {
  --heading-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  --mono-default: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  --body-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

body,
.posts-list .post-preview p,
.posts-list .description p {
  font-family: var(--body-font), var(--body-default);
}

h1, h2, h3, h4, h5, h6,
.posts-list .post-preview h2,
.posts-list .description h2 {
  font-family: var(--heading-font), var(--heading-default);
}

d-article div.sourceCode code,
d-article pre code {
  font-family: var(--mono-font), var(--mono-default);
}


/*-- TITLE --*/
d-title h1,
.posts-list > h1 {
  color: var(--title-color, black);
}

d-title h1 {
  font-size: var(--title-size, 50px);
}

/*-- HEADERS --*/
d-article h1,
d-article h2,
d-article h3,
d-article h4,
d-article h5,
d-article h6 {
  color: var(--header-color, rgba(0, 0, 0, 0.8));
}

/*-- BODY --*/
d-article > p,  /* only text inside of <p> tags */
d-article > ul, /* lists */
d-article > ol {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
  font-size: var(--body-size, 1.06rem);
}


/*-- CODE --*/
d-article div.sourceCode code,
d-article pre code {
  font-size: var(--code-size, 14px);
}

/*-- ASIDE --*/
d-article aside {
  font-size: var(--aside-size, 12px);
  color: var(--aside-color, rgba(0, 0, 0, 0.6));
}

/*-- FIGURE CAPTIONS --*/
figure .caption,
figure figcaption,
.figure .caption {
  font-size: var(--fig-cap-size, 13px);
  color: var(--fig-cap-color, rgba(0, 0, 0, 0.6));
}

/*-- METADATA --*/
d-byline h3 {
  font-size: var(--heading-size, 0.6rem);
  color: var(--heading-color, rgba(0, 0, 0, 0.5));
}

d-byline {
  font-size: var(--body-size, 0.8rem);
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

d-byline a,
d-article d-byline a {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

/*-- TABLE OF CONTENTS --*/
.d-contents nav h3 {
  font-size: var(--heading-size, 18px);
}

.d-contents nav a {
  font-size: var(--contents-size, 13px);
}

/*-- APPENDIX --*/
d-appendix h3 {
  font-size: var(--heading-size, 15px);
  color: var(--heading-color, rgba(0, 0, 0, 0.65));
}

d-appendix {
  font-size: var(--text-size, 0.8em);
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

d-appendix d-footnote-list a.footnote-backlink {
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

/*-- WEBSITE HEADER + FOOTER --*/
.distill-site-header .title {
  font-size: var(--title-size, 18px);
  font-family: var(--navbar-font), var(--heading-default);
}

.distill-site-header a,
.nav-dropdown .nav-dropbtn {
  font-family: var(--navbar-font), var(--heading-default);
}

.nav-dropdown .nav-dropbtn {
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  font-size: var(--text-size, 15px);
}

.distill-site-header a:hover,
.nav-dropdown:hover .nav-dropbtn {
  color: var(--hover-color, white);
}

.distill-site-header {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer a:hover {
  color: var(--hover-color, white);
}</style>
<!--/radix_placeholder_distill-->
  <script src="../../site_libs/header-attrs-2.29/header-attrs.js"></script>
  <script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="../../site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="../../site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="../../site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="../../site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="../../site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="../../site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="../../site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="../../site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
<!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Mathjax config (add automatic linebreaks when supported)
      // MathJax = {
      //    tex: {
      //        inlineMath: [['$', '$'], ['\\(', '\\)']],
      //        displayMath: [['$$', '$$'], ['\\[', '\\]']],
      //        tags: 'ams',
      //        multline: true,
      //    },
      //    options: {
      //        linebreaks: { automatic: true },
      //    },
      // };

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      buttons.append('<a href="' + slug + '.zip" class="dt-tag"><i class="fas fa-file-zipper"></i> Supplement</a>')

      // adds Abstract: in front of the first <p> in the title section --
      // unless it happens to be the subtitle (FIXME: this is a bad hack - can't distill do this?)
      var tpar = $("d-title p:not(:empty)").filter(function() {
        return !$(this).hasClass("subtitle");
      }).first();
      if (tpar) {
        const abstract = $('<d-abstract>')
        abstract.append('<b>Abstract:</b><br>')
        abstract.append(tpar) // Move description to d-abstract
        $("d-title p:empty").remove() // Remove empty paragraphs after title
        abstract.append(buttons)
        abstract.insertAfter($('d-title')) // Add abstract section after title */
      }

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
    
    d-appendix .citation-appendix,
    .d-appendix .citation-appendix {
      color: rgb(60, 60, 60);
    }

    d-article h2 {
      border-bottom: 0px solid rgba(0, 0, 0, 0.1);
      padding-bottom: 0rem;
    }
    d-article h3 {
      font-size: 20px;
    }
    d-article h4 {
      font-size: 18px;
      text-transform: none;
    }

    @media (min-width: 1024px) {
      d-article h2 {
        font-size: 32px;
      }
      d-article h3 {
        font-size: 24px;
      }
      d-article h4 {
        font-size: 20px;
      }
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"BMRMM: An R Package for Bayesian Markov (Renewal) Mixed Models","description":"We introduce the BMRMM\npackage implementing Bayesian inference for a class of Markov renewal\nmixed models which can characterize the stochastic dynamics of a\ncollection of sequences, each comprising alternative instances of\ncategorical states and associated continuous duration times, while\nbeing influenced by a set of exogenous factors as well as a 'random'\nindividual. The default setting flexibly models the state transition\nprobabilities using mixtures of Dirichlet distributions and the\nduration times using mixtures of gamma kernels while also allowing\nvariable selection for both. Modeling such data using simpler Markov\nmixed models also remains an option, either by ignoring the duration\ntimes altogether or by replacing them with instances of an additional\ncategory obtained by discretizing them by a user-specified unit. The\noption is also useful when data on duration times may not be available\nin the first place. We demonstrate the package's utility using two\ndata sets.","doi":"10.32614/RJ-2024-011","authors":[{"author":"Yutong Wu","authorURL":"#","affiliation":"Department of Mechanical Engineering","affiliationURL":"#","orcidID":""},{"author":"Abhra Sarkar","authorURL":"#","affiliation":"Department of Statistics and Data Sciences","affiliationURL":"#","orcidID":""}],"publishedDate":"2025-01-11T00:00:00.000+00:00","citationText":"Wu & Sarkar, 2025"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a class="logo" href="../../index.html">
<img src="../../resources/rlogo.png" alt="Logo"/>
</a>
<a href="../../index.html" class="title">The R Journal</a>
</div>
<div class="nav-right">
<a href="../../index.html">Home</a>
<a href="../../issues/2024-3">Current</a>
<a href="../../issues.html">Issues</a>
<a href="../../news.html">News</a>
<a href="../../submissions.html">Submit</a>
<a href="../../editors.html">Editorial board</a>
<a href="../../articles.xml">
<i class="fa fa-rss" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>BMRMM: An R Package for Bayesian Markov (Renewal) Mixed Models</h1>

<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>We introduce the BMRMM
package implementing Bayesian inference for a class of Markov renewal
mixed models which can characterize the stochastic dynamics of a
collection of sequences, each comprising alternative instances of
categorical states and associated continuous duration times, while
being influenced by a set of exogenous factors as well as a ‘random’
individual. The default setting flexibly models the state transition
probabilities using mixtures of Dirichlet distributions and the
duration times using mixtures of gamma kernels while also allowing
variable selection for both. Modeling such data using simpler Markov
mixed models also remains an option, either by ignoring the duration
times altogether or by replacing them with instances of an additional
category obtained by discretizing them by a user-specified unit. The
option is also useful when data on duration times may not be available
in the first place. We demonstrate the package’s utility using two
data sets.</p></p>
</div>

<div class="d-byline">
  Yutong Wu  (Department of Mechanical Engineering)
  
,   Abhra Sarkar  (Department of Statistics and Data Sciences)
  
<br/>2025-01-11
</div>

<div class="d-article">
<div class="article">
<h3 data-number="1" id="sec:intro"><span class="header-section-number">1</span> Introduction</h3>
<p>Markov models (MMs) are widely used for modeling the transition dynamics
of categorical state sequences. Classical Markov renewal models (MRMs)
additionally model the state duration times, when available, where the
state transitions follow Markov dynamics and the state duration times
follow a continuous distribution that depends on the immediately
preceding and following states (Figure <a href="#fig:figgraph-model">1</a>).
M(R)Ms have been widely used in different variations
<span class="citation" data-cites="phelan1990bayes eichelsbacher2002bayesian muliere2003reinforced alvarez2005estimation diaconis2006bayesian bulla2007bayesian etterson2007analysis bacallado2009bayesian li2009markov epifani2014bayesian siebert2016bayesian holsclaw2017bayesian sesia2019gene">(<a href="#ref-phelan1990bayes" role="doc-biblioref">Phelan 1990</a>; <a href="#ref-eichelsbacher2002bayesian" role="doc-biblioref">Eichelsbacher and Ganesh 2002</a>; <a href="#ref-muliere2003reinforced" role="doc-biblioref">Muliere et al. 2003</a>; <a href="#ref-alvarez2005estimation" role="doc-biblioref">Alvarez 2005</a>; <a href="#ref-diaconis2006bayesian" role="doc-biblioref">Diaconis and Rolles 2006</a>; <a href="#ref-bulla2007bayesian" role="doc-biblioref">Bulla and Muliere 2007</a>; <a href="#ref-etterson2007analysis" role="doc-biblioref">Etterson et al. 2007</a>; <a href="#ref-bacallado2009bayesian" role="doc-biblioref">Bacallado et al. 2009</a>; <a href="#ref-li2009markov" role="doc-biblioref">Li 2009</a>; <a href="#ref-epifani2014bayesian" role="doc-biblioref">Epifani et al. 2014</a>; <a href="#ref-siebert2016bayesian" role="doc-biblioref">Siebert and Söding 2016</a>; <a href="#ref-holsclaw2017bayesian" role="doc-biblioref">Holsclaw et al. 2017</a>; <a href="#ref-sesia2019gene" role="doc-biblioref">Sesia et al. 2019</a>)</span>.
There are also some sparse works on covariate-dependent Markov models
<span class="citation" data-cites="muenz1985markov gradner1990analyzing alioum1998effect islam2006higher">(<a href="#ref-muenz1985markov" role="doc-biblioref">Muenz and Rubinstein 1985</a>; <a href="#ref-gradner1990analyzing" role="doc-biblioref">Gradner 1990</a>; <a href="#ref-alioum1998effect" role="doc-biblioref">Alioum et al. 1998</a>; <a href="#ref-islam2006higher" role="doc-biblioref">Islam and Chowdhury 2006</a>)</span>.</p>
<p>The existing literature however focuses very heavily on modeling single
sequences. <span class="citation" data-cites="sarkar2018bayesian">(<a href="#ref-sarkar2018bayesian" role="doc-biblioref">Sarkar et al. 2018</a>)</span> developed a highly flexible and
computationally efficient class of Bayesian Markov mixed models (BMMMs)
for jointly modeling a collection of categorical sequences, each one
associated with an individual as well as a set of time-invariant
external covariates (e.g., the sex and genotype of the associated
individual, an experimental condition under which the sequence was
generated, etc.). BMMMs characterize the state transition probabilities
using a convex combination of a fixed covariate-dependent component and
a random individual-level component, both being Dirichlet distributed.
They further allow covariate levels with similar effects to be
probabilistically clustered together, allowing automatic selection of
the significant covariates, and providing a sophisticated framework for
analyzing data sets having the aforementioned structure.</p>
<p>BMMMs however do not model duration times of the states which are often
additionally available in real-world applications. Recently,
<span class="citation" data-cites="wu2021bayesian">(<a href="#ref-wu2021bayesian" role="doc-biblioref">Wu et al. 2023</a>)</span> extended BMMMs to Bayesian Markov renewal mixed models
(BMRMMs), allowing for the additional analysis of continuous duration
times which, depending on the application, can either be the duration
for which a state persists or the duration between two consecutive
states, i.e., inter-state intervals (ISIs). Specifically, they modeled
the duration times using mixtures of gamma kernels with mixture
probabilities being a convex combination a covariate-dependent effect
and an individual-level effect, similar to BMMMs. Covariate levels with
similar influences on mixture probabilities are clustered together in
BMRMMs as well, allowing the selection of the significant covariates.
BMRMMs thus holistically model both state transitions and continuous
duration times, painting a comprehensive picture of the underlying
stochastic dynamics.</p>
<p>In this article, we describe the R package
<a href="https://CRAN.R-project.org/package=BMRMM"><strong>BMRMM</strong></a> which implements
BMMMs and BMRMMs, collectively referred to henceforth as BM(R)MMs. The
<strong>BMRMM</strong> package runs posterior inference for categorical state
transitions and continuous duration times, if available, via a Markov
chain Monte Carlo (MCMC) algorithm, returning an object containing
comprehensive inference results. The package also includes a suite of
plotting functions to display the results graphically. Specifically for
continuous duration times, when available, the package provides users
with three different options: (i) ignore the duration times and model
the state transitions alone as a BMMM; (ii) introduce an additional
category by discretizing the continuous duration times by a
user-specified unit, and analyze the appended state transitions as a
BMMM; (iii) model the duration as a mixture of gamma kernels using a
BMRMM, as proposed in <span class="citation" data-cites="wu2021bayesian">Wu et al. (<a href="#ref-wu2021bayesian" role="doc-biblioref">2023</a>)</span>. Additionally, users can choose to
turn off one or both of the fixed covariate effects and the random
individual effects. Overall, the <strong>BMRMM</strong> package thus gives users a
lot of flexibility in handling their data sets, providing inferences for
both Bayesian Markov renewal or non-renewal models as needed.</p>
<p>The <strong>BMRMM</strong> package conveniently includes a synthetic <code>foxp2</code> data set
that describes the laboratory study on the role of the FoxP2 gene
implicated in speech deficiencies for adult mice, which is also the
motivating application for the methodology of BMMMs and BMRMMs.
Mutations in the FoxP2 gene have long been associated with severe
deficits in vocal communication for mammals
<span class="citation" data-cites="macdermot2005identification">(<a href="#ref-macdermot2005identification" role="doc-biblioref">MacDermot et al. 2005</a>)</span>. Mice with and without the mutation
singing under various "social contexts" have thus been studied in many
experiments
<span class="citation" data-cites="Fujita_etal:2008 Castellucci_etal:2016 Gaub_etal:2016 Chabout_etal:2016">(<a href="#ref-Fujita_etal:2008" role="doc-biblioref">Fujita et al. 2008</a>; <a href="#ref-Castellucci_etal:2016" role="doc-biblioref">Castellucci et al. 2016</a>; <a href="#ref-Chabout_etal:2016" role="doc-biblioref">Chabout et al. 2016</a>; <a href="#ref-Gaub_etal:2016" role="doc-biblioref">Gaub et al. 2016</a>)</span>.
[The FoxP2 data set <span class="citation" data-cites="Chabout_etal:2016">(<a href="#ref-Chabout_etal:2016" role="doc-biblioref">Chabout et al. 2016</a>)</span>, e.g., comprises the sequences
of syllables making up the songs as well as the lengths of
inter-syllable intervals (ISIs). The data set <code>foxp2</code> included in the
<strong>BMRMM</strong> package is taken from the simulation study of
<span class="citation" data-cites="wu2021bayesian">(<a href="#ref-wu2021bayesian" role="doc-biblioref">Wu et al. 2023</a>)</span>. It is much shorter than the real FoxP2 data set but
closely mimics its other aspects and is used] in
this paper to demonstrate how to obtain detailed inferences for both
syllable transitions and ISI dynamics for a comprehensive analysis of
the vocal repertoire in mice with and without the FoxP2 mutation.</p>
<p>The utility of the <strong>BMRMM</strong> package goes well beyond the FoxP2 data
set. As described above, the package is designed for scenarios where the
data set consists of categorical state sequences associated with an
individual as well as a number of observed covariates where additional
data on continuous duration times may or may not be available. Data sets
with such structures are frequently observed in different areas of
scientific research and can potentially benefit from the <strong>BMRMM</strong>
package. For example, <span class="citation" data-cites="islam2006higher">Islam and Chowdhury (<a href="#ref-islam2006higher" role="doc-biblioref">2006</a>)</span> analyzed the transitions of
different rainfall orders in three districts of Bangladesh under three
covariates, wind speed, humidity, and maximum temperature. In an
education assessment study, <span class="citation" data-cites="zhang2019scenario">Zhang et al. (<a href="#ref-zhang2019scenario" role="doc-biblioref">2019</a>)</span> recorded sequences of
writing states, characterized by keystroke logs, for 257 eighth graders
of various genders, races, and socioeconomic statuses.
<span class="citation" data-cites="combescure2003assessment">Combescure et al. (<a href="#ref-combescure2003assessment" role="doc-biblioref">2003</a>)</span> estimated the control states of 371 asthma
patients with different body mass indices (BMIs) and disease severity
over a four-year-long study and produced the asthma control data set,
which we will use as an additional example to demonstrate the utility of
our package in this paper. For such data sets, the <strong>BMRMM</strong> package
provides a flexible, sophisticated, and principled way to model fixed
effects of the covariates and random effects of the individuals in both
the state transition dynamics and the distribution of the ISIs.</p>
<p>Other computer programs for Markov models with covariates include MARKOV
<span class="citation" data-cites="marshall1995markov">(<a href="#ref-marshall1995markov" role="doc-biblioref">Marshall et al. 1995</a>)</span> and MKVPCI <span class="citation" data-cites="alioum2001mkvpci">(<a href="#ref-alioum2001mkvpci" role="doc-biblioref">Alioum and Commenges 2001</a>)</span>. R packages for
analyzing discrete Markov models include
<a href="https://CRAN.R-project.org/package=markovchain"><strong>markovchain</strong></a>
<span class="citation" data-cites="Spedicato2017Discrete">(<a href="#ref-Spedicato2017Discrete" role="doc-biblioref">Spedicato 2017</a>)</span> and
<a href="https://CRAN.R-project.org/package=msm"><strong>msm</strong></a> <span class="citation" data-cites="jackson2011multi">(<a href="#ref-jackson2011multi" role="doc-biblioref">Jackson 2011</a>)</span>.
<a href="https://CRAN.R-project.org/package=SemiMarkov"><strong>SemiMarkov</strong></a>
<span class="citation" data-cites="listwon2015semimarkov">(<a href="#ref-listwon2015semimarkov" role="doc-biblioref">Listwon and Saint-Pierre 2015</a>)</span> and
<a href="https://CRAN.R-project.org/package=SMM"><strong>SMM</strong></a> <span class="citation" data-cites="barbu2018smm">(<a href="#ref-barbu2018smm" role="doc-biblioref">Barbu et al. 2018</a>)</span>
provide functions for the simulation and estimation of traditional
semi-Markov models. Some R packages provide the inference of hidden
semi-Markov models, such as
<a href="https://CRAN.R-project.org/package=mhsmm"><strong>mhsmm</strong></a> <span class="citation" data-cites="o2011hidden">(<a href="#ref-o2011hidden" role="doc-biblioref">O’Connell and Højsgaard 2011</a>)</span> and
<a href="https://CRAN.R-project.org/package=hhsmm"><strong>hhsmm</strong></a> <span class="citation" data-cites="amini2022hhsmm">(<a href="#ref-amini2022hhsmm" role="doc-biblioref">Amini et al. 2022</a>)</span>.
<span class="citation" data-cites="ferguson2012mssurv">Ferguson et al. (<a href="#ref-ferguson2012mssurv" role="doc-biblioref">2012</a>)</span> built the <strong>msSurv</strong> package which provides a
nonparametric estimation of semi-Markov models but does not consider
covariates. There are also R packages implementing MRPs in specific
application areas. For example, <span class="citation" data-cites="kharrat2019flexible">Kharrat et al. (<a href="#ref-kharrat2019flexible" role="doc-biblioref">2019</a>)</span> introduced the
<a href="https://CRAN.R-project.org/package=Countr"><strong>Countr</strong></a> package for
flexible regression models based on MRPs. <span class="citation" data-cites="pustejovsky2021">Pustejovsky (<a href="#ref-pustejovsky2021" role="doc-biblioref">2021</a>)</span> developed the
<a href="https://CRAN.R-project.org/package=ARPobservation"><strong>ARPobservation</strong></a>
for simulating behavior streams based on alternating renewal processes.
Other R packages for categorical data analysis include
<a href="https://CRAN.R-project.org/package=catdap"><strong>catdap</strong></a>
<span class="citation" data-cites="katsura1980catdap">(<a href="#ref-katsura1980catdap" role="doc-biblioref">Katsura 1980</a>)</span> and
<a href="https://CRAN.R-project.org/package=vcd"><strong>vcd</strong></a> <span class="citation" data-cites="meyer2022vcd">(<a href="#ref-meyer2022vcd" role="doc-biblioref">Meyer et al. 2022</a>)</span>. To
our knowledge, there has not been an R package that implements flexible
Bayesian M(R)MMs.</p>
<p>[In the following section, we summarize the technical details of BMRMMs.
Documentation for the functions of the <strong>BMRMM</strong> package is then
provided. Next, we demonstrate the usage of our package in analyzing two
different data sets. The final section contains some concluding remarks.]</p>
<h3 data-number="2" id="sec:models"><span class="header-section-number">2</span> The Bayesian Markov (renewal) mixed models</h3>
<p>We briefly describe the BM(R)MM methodologies here – more details
can be found in <span class="citation" data-cites="sarkar2018bayesian">Sarkar et al. (<a href="#ref-sarkar2018bayesian" role="doc-biblioref">2018</a>)</span> and <span class="citation" data-cites="wu2021bayesian">Wu et al. (<a href="#ref-wu2021bayesian" role="doc-biblioref">2023</a>)</span>. Consider
specifically a sequence <span class="math inline">\(s\)</span> comprising <span class="math inline">\(T_s\)</span> state instances and let
<span class="math inline">\(y_{s,t}\)</span> denote the state at time <span class="math inline">\(t\)</span> in sequence <span class="math inline">\(s\)</span>. The states
<span class="math inline">\(y_{s,t}\)</span>’s come from a set <span class="math inline">\({\cal Y}= \{1,2,\dots,d_0\}\)</span>. Within a
sequence <span class="math inline">\(s\)</span>, there are <span class="math inline">\(T_s-1\)</span> duration times (state persistence times
or inter-state intervals), denoted by
<span class="math inline">\(\{\tau_{s,t}\}_{s=1,t=2}^{s_0,T_s}\)</span>, where <span class="math inline">\(\tau_{s,t}\)</span> is the duration
between the <span class="math inline">\((t-1)^{th}\)</span> and <span class="math inline">\(t^{th}\)</span> states in sequence <span class="math inline">\(s\)</span>, and <span class="math inline">\(s_0\)</span>
represents the total number of sequences.
Figure <a href="#fig:figgraph-model">1</a> presents a graphical summary of the
data structure. Each sequence <span class="math inline">\(s\)</span> is associated with <span class="math inline">\(p\)</span> categorical
covariates or factors, denoted by
<span class="math inline">\(x_{s,j} \in {\cal X}_j = \{1,2,\dots,d_j\}\)</span>, and an individual, denoted
by <span class="math inline">\(i_{s}\)</span>. Without loss of generality, we assume that the analyses of
the transition probabilities and the duration times distributions both
include all <span class="math inline">\(p\)</span> covariates. Moreover, the analysis of duration times
counts the previous state <span class="math inline">\(y_{s,t-1}\)</span> as an additional <span class="math inline">\((p+1)^{th}\)</span>
covariate. In the <strong>BMRMM</strong> package, users have the flexibility to
select particular covariates for each analysis and exclude the previous
state from the analysis of duration times. In their original definition
in <span class="citation" data-cites="pyke1961markov">(<a href="#ref-pyke1961markov" role="doc-biblioref">Pyke 1961</a>)</span>, the duration time <span class="math inline">\(\tau_{s,t}\)</span> in an MRP was
allowed to depend on both the preceding state <span class="math inline">\(y_{s,t-1}\)</span> and the
following state <span class="math inline">\(y_{s,t}\)</span>. To keep the notation simple and the
methodology easy to understand for a broad audience, we however only
include the preceding state <span class="math inline">\(y_{s,t-1}\)</span> as a predictor of <span class="math inline">\(\tau_{s,t}\)</span>
in this paper. This analysis can be easily modified to have the pair
<span class="math inline">\((y_{s,t-1}, y_{s,t})\)</span> as a predictor instead of just <span class="math inline">\(y_{s,t-1}\)</span>, as
was actually done in <span class="citation" data-cites="wu2021bayesian">(<a href="#ref-wu2021bayesian" role="doc-biblioref">Wu et al. 2023</a>)</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figgraph-model"></span>
<img src="figs/data-model.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 1: Graphical model showing the data structure: <span class="math inline">\(y_{s,t}\)</span> denotes the observed state at the <span class="math inline">\(t^{th}\)</span> time location in the <span class="math inline">\(s^{th}\)</span> sequence; <span class="math inline">\(\tau_{s,t}\)</span> denotes the observed duration times (either state persistence times or inter-state intervals) between the states <span class="math inline">\(y_{s,t-1}\)</span> and <span class="math inline">\(y_{s,t}\)</span>; each sequence <span class="math inline">\(s\)</span> is also associated with an individual <span class="math inline">\(i_{s}\)</span> and a set of exogenous time-invariant covariates <span class="math inline">\(x_{s,1},\dots, x_{s,p}\)</span>. The Markov mixed model considered in this article analyzes the state transitions <span class="math inline">\(y_{s,t}\)</span> in a collection of sequences; the Markov renewal mixed model additionally analyzes the duration times <span class="math inline">\(\tau_{s,t}\)</span>; both models accommodate fixed effects of the covariates <span class="math inline">\(x_{s,1},\dots, x_{s,p}\)</span> and random effects of the individuals <span class="math inline">\(i_{s}\)</span>.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="2.1" id="model-for-state-transitions-sec-bmmm">Model for state transitions {#sec: BMMM}</h4>
<p>For a sequence <span class="math inline">\(s\)</span> associated with individual <span class="math inline">\(i\)</span> and covariate levels
<span class="math inline">\(x_1,\dots,x_p\)</span>, the transition probabilities
<span class="math inline">\(\Pr(y_{s,t}=y_{t} \mid  i_{s}=i, x_{s,1}=x_{1}, \dots,x_{s,p}=x_{p}, y_{s,t-1}=y_{t-1})=P_{trans,x_{1},\dots, x_{p}}^{(i)}(y_{t} \mid y_{t-1})\)</span>
are defined as a convex combination of a fixed covariate effect
component <span class="math inline">\(\boldsymbol{\lambda}_{trans,x_1,\dots,x_p}(\cdot\mid y_{t-1})\)</span> and a
random effect component <span class="math inline">\(\boldsymbol\lambda_{trans}^{(i)}\)</span>:
<span class="math display" id="eq:-mixed-Markov-model">\[\begin{aligned}
&amp; \hskip -2ex P_{trans,x_{1},\dots, x_{p}}^{(i)} (y_{t} \mid y_{t-1}) = \pi_{trans,0}^{(i)}(y_{t-1}) \lambda_{trans,x_{1},\dots, x_{p}}(y_{t} \mid y_{t-1}) + \pi_{trans,1}^{(i)}(y_{t-1}) \lambda^{(i)}_{trans}(y_{t} \mid y_{t-1}). ~ \label{eq: mixed Markov model}
\end{aligned}   \tag{1}\]</span>
The coefficients of the convex combination, namely,
<span class="math inline">\(\{\pi_{trans,0}^{(i)}(y_{t-1}),\pi_{trans,1}^{(i)}(y_{t-1})\}\)</span> are
individual-specific and satisfy
<span class="math inline">\(\pi_{trans,1}^{(i)}(y_{t-1})=1-\pi_{trans,0}^{(i)}(y_{t-1})\)</span>.</p>
<p>For each covariate <span class="math inline">\(j=1,\dots,p\)</span>, it is possible that some covariate
levels exert a similar effect on the transition dynamics. [For example,
the components
<span class="math inline">\(\boldsymbol\lambda_{trans,x_1=1,x_2,\dots,x_p}(\cdot\mid y_{t-1})\)</span> and
<span class="math inline">\(\boldsymbol\lambda_{trans,x_1=2,x_2,\dots,x_p}(\cdot\mid y_{t-1})\)</span> would be
equal if levels 1 and 2 of covariate 1 have similar influences on
transition dynamics for fixed levels for covariates
<span class="math inline">\(2, \dots, p\)</span>.] A clustering mechanism for
covariate levels allows the fixed component
<span class="math inline">\(\boldsymbol\lambda_{trans,x_1,\dots,x_p}(\cdot\mid y_{t-1})\)</span> to be the same
for all levels with a similar influence. In particular, for covariate
<span class="math inline">\(j\)</span>, we construct the partition
<span class="math inline">\({\cal C}_{trans}^{(j)}=\{{\cal C}_{trans,h_{j}}^{(j)}\}_{h_{j}=1}^{k_{trans,j}}\)</span>
of its levels, where <span class="math inline">\(k_{trans,j}\)</span> is the number of clusters for
covariate <span class="math inline">\(j\)</span> and <span class="math inline">\(h_j\)</span> represents the cluster index. We introduce
latent variables <span class="math inline">\(\{z_{trans,j,\ell}\}_{j=1,\ell=1}^{p,d_{j}}\)</span> that
indicate the cluster index for the <span class="math inline">\(\ell^{th}\)</span> label of covariate <span class="math inline">\(j\)</span>.
[Two levels of the covariate <span class="math inline">\(j\)</span>,
<span class="math inline">\(\ell_1, \ell_2 \in {\cal X}_j = \{1,\dots,d_j\}\)</span>, are clustered
together if and only if
<span class="math inline">\(z_{trans,j,\ell_1} = z_{trans,j,\ell_2}\)</span>.] For the
fixed effects, we then replace the covariate levels <span class="math inline">\(x_1,\dots,x_p\)</span>’s
with cluster indices <span class="math inline">\(h_1,\dots,h_p\)</span>’s and present the fixed effect as
<span class="math inline">\(\boldsymbol\lambda_{trans,h_{1},\dots,h_{p}}(\cdot\mid y_{t-1})\)</span>.</p>
<p>We set Dirichlet priors for both fixed and individual effect components
and let them center around the same mean vector <span class="math inline">\(\boldsymbol\lambda_{trans,0}\)</span>
to facilitate posterior computation. The probability vector
<span class="math inline">\(\boldsymbol\lambda_{trans,0}\)</span> is also given a Dirichlet prior with mean
<span class="math inline">\(\boldsymbol\lambda_{trans,00}\)</span> to capture the natural preferences of certain
states in <span class="math inline">\({\cal Y}\)</span>:
<span class="math display">\[\begin{aligned}
&amp;&amp; \boldsymbol \lambda_{trans,h_{1},\dots,h_{p}}(\cdot\mid y_{t-1}) \sim \hbox{Dir}\left\{\alpha_{trans,0}\lambda_{trans,0}(1 \mid y_{t-1}),\dots,\alpha_{trans,0}\lambda_{trans,0}(d_0 \mid y_{t-1})\right\}, \nonumber \\
&amp;&amp; \boldsymbol \lambda^{(i)}_{trans}(\cdot \mid y_{t-1}) \sim \hbox{Dir}\left\{\alpha^{(0)}_{trans}\lambda_{trans,0}(1 \mid y_{t-1}),\dots,\alpha^{(0)}_{trans}\lambda_{trans,0}(d_0 \mid y_{t-1})\right\}, \nonumber \\
&amp;&amp; \boldsymbol \lambda_{trans,0}(\cdot\mid y_{t-1}) \sim \hbox{Dir}\left\{\alpha_{trans,00}\lambda_{trans,00}(1),\dots,\alpha_{trans,00}\lambda_{trans,00}(d_0)\right\}. \nonumber
\end{aligned}\]</span></p>
<p>We present the complete Bayesian hierarchical model for the transition
dynamics as
<span class="math display">\[\begin{aligned}
&amp;&amp; (y_{s,t} \mid y_{s,t-1}=y_{t-1}, i_{s}=i, z_{trans,1,x_{s,1}} =h_{1},\dots,z_{trans,p,x_{s,p}} =h_{p})  \sim \nonumber\\
&amp;&amp;  \hbox{Mult}\left\{P_{trans,h_{1},\dots,h_{p}}^{(i)}(1\mid y_{t-1}),\dots,P_{trans,h_{1},\dots,h_{p}}^{(i)}(d_0\mid y_{t-1})\right\}, ~~\text{where} \nonumber\\
&amp;&amp; {\mathbf P}_{trans,h_{1},\dots,h_{p}}^{(i)} (\cdot\mid y_{t-1}) = \pi_{trans,0}^{(i)}(y_{t-1}) \boldsymbol \lambda_{trans,h_{1},\dots,h_{p}}(\cdot\mid y_{t-1}) + \pi_{trans,1}^{(i)}(y_{t-1}) \boldsymbol \lambda^{(i)}_{trans}(\cdot\mid y_{t-1}),\nonumber\\
&amp;&amp; z_{trans,j,\ell} \sim \hbox{Mult}\left\{\mu_{trans,j}(1),\dots,\mu_{trans,j}(d_{j})\right\}, ~~~~~ \boldsymbol \mu_{trans,j} \sim \hbox{Dir}(\alpha_{trans,j},\dots,\alpha_{trans,j}), \nonumber\\
&amp;&amp; \boldsymbol \lambda_{trans,h_{1},\dots,h_{p}}(\cdot\mid y_{t-1}) \sim \hbox{Dir}\left\{\alpha_{trans,0}\lambda_{trans,0}(1 \mid y_{t-1}),\dots,\alpha_{trans,0}\lambda_{trans,0}(d_0 \mid y_{t-1})\right\},\nonumber \\
&amp;&amp; \boldsymbol \lambda_{trans}^{(i)}(\cdot \mid y_{t-1}) \sim \hbox{Dir}\left\{\alpha^{(0)}_{trans}\lambda_{trans,0}(1 \mid y_{t-1}),\dots,\alpha^{(0)}_{trans}\lambda_{trans,0}(d_0 \mid y_{t-1})\right\}, \nonumber\\
&amp;&amp; \boldsymbol \lambda_{trans,0}(\cdot\mid y_{t-1}) \sim \hbox{Dir}\left\{\alpha_{trans,00}\lambda_{trans,00}(1),\dots,\alpha_{trans,00}\lambda_{trans,00}(d_0)\right\}, \nonumber \\
&amp;&amp; \pi_{trans,0}^{(i)}(y_{t-1}) \sim \hbox{Beta}(a_{trans,0},a_{trans,1}), \nonumber \\
&amp;&amp; \alpha_{trans,0}\sim\hbox{Ga}(a_{trans,0},b_{trans,0}),~~~~~~~\alpha^{(0)}_{trans}\sim\hbox{Ga}(a^{(0)}_{trans},b^{(0)}_{trans}).\nonumber
\end{aligned}\]</span></p>
<h4 class="unnumbered" data-number="2.2" id="sec:isi">Model for continuous duration times</h4>
<p>The <strong>BMRMM</strong> package provides three options for analyzing the duration
times: (i) ignore the durations altogether and only model the transition
probabilities of the existing states, (ii) treat the durations as blocks
of a new special category, with a discretization unit specified by
users, (iii) model the durations as a continuous random variable with a
flexible mixture of gamma distributions. For the first two options, we
only need to apply the model described in the previous subsection. For
the third option, we need to conduct a separate analysis of the duration
times as described below.</p>
<p>Let <span class="math inline">\(K\)</span> denote the number of gamma mixture components in the model for
the duration times. We let
<span class="math inline">\({\mathbf P}_{dur}^{(i)}(\cdot \mid x_1,\dots,x_p,y_{s,t-1})\)</span> denote the
mixture probability vector given individual <span class="math inline">\(i\)</span>, covariate levels
<span class="math inline">\(x_1,\dots,x_p\)</span>, and preceding syllable <span class="math inline">\(y_{s,t-1}\)</span>. The preceding state
<span class="math inline">\(y_{s,t-1}\)</span> can be removed from the formula if the users do not wish to
consider its influence in the inference of the duration times. The
distribution of the continuous duration times,
<span class="math inline">\(\{\tau_{s,t}\}_{s=1,t=2}^{s_0,T_s}\)</span>, is then modeled as<br />
</p>
<p><span class="math display">\[\begin{aligned}
&amp; f(\tau_{s,t} \mid  i_{s}=i, x_{s,1}=x_{1}, \dots,x_{s,p}=x_{p}, y_{s,t-1}=y_{t-1}) \nonumber \\
&amp; = \sum_{k=1}^{K}P_{dur}^{(i)}(k \mid x_{1},\dots,x_{p},y_{t-1}) \hbox{Ga}(\tau_{s,t} \mid \alpha_{k},\beta_{k}),  \nonumber
\end{aligned}\]</span>
where <span class="math inline">\(\alpha_k\)</span> and <span class="math inline">\(\beta_k\)</span> denote the shape and rate parameters of
the <span class="math inline">\(k^{th}\)</span> gamma mixture component, respectively. We introduce a set
of latent variables <span class="math inline">\(\{z_{dur,s,t}\}_{s=1,t=2}^{s_0,T_s}\)</span> that
represents the index of the mixture component. If <span class="math inline">\(z_{dur,s,t}\)</span> equals
to <span class="math inline">\(k\)</span>, then <span class="math inline">\(\tau_{s,t}\)</span> follows <span class="math inline">\(\hbox{Ga}(\alpha_k,\beta_k)\)</span>
distribution, i.e.,<br />
</p>
<p><span class="math display">\[\begin{aligned}
&amp; f(\tau_{s,t} \mid  z_{dur,s,t}=k) \sim \hbox{Ga}(\tau_{s,t} \mid \alpha_{k},\beta_{k}), \nonumber\\
&amp; \Pr(z_{dur,s,t}=k \mid  i_{s}=i, x_{s,1}=x_{1}, \dots,x_{s,p}=x_{p}, y_{s,t-1}=y_{t-1})=P_{dur}^{(i)}(k \mid x_{1},\dots,x_{p},y_{t-1}) \nonumber.
\end{aligned}\]</span></p>
<p>Similar to the model for the transition probabilities, the mixture
probabilities are a convex combination of a fixed population-level
effect and a random individual-level effect:<br />
</p>
<p><span class="math display">\[\begin{aligned}
&amp;&amp; {\mathbf P}^{(i)}_{dur}(\cdot \mid x_{1},\dots,x_{p},y_{t-1})=\pi_{dur,0}^{(i)}(\cdot)\boldsymbol \lambda_{dur,x_{1},\dots,x_{p},y_{t-1}}(\cdot)+\pi_{dur,1}^{(i)}(\cdot)\boldsymbol \lambda^{(i)}_{dur}(\cdot),
\end{aligned}\]</span>
where <span class="math inline">\(\boldsymbol\lambda_{dur,x_{1},\dots,x_{p},y_{t-1}}(\cdot)\)</span> is the
baseline component, <span class="math inline">\(\boldsymbol\lambda^{(i)}_{dur}(\cdot)\)</span> is the random
individual effect, and
<span class="math inline">\(\{\pi_{dur,0}^{(i)}(k),\pi_{dur,1}^{(i)}(k)\}_{k=1}^K\)</span> are
individual-specific coefficients such that
<span class="math inline">\(\pi_{dur,1}^{(i)}(k)=1-\pi_{dur,0}^{(i)}(k)\)</span>. Again, for each covariate
<span class="math inline">\(r=1,\dots,p,p+1\)</span> (where the <span class="math inline">\((p+1)^{th}\)</span> covariate is the preceding
state <span class="math inline">\(y_{t-1}\)</span>), we construct the partition
<span class="math inline">\({\cal C}_{dur}^{(r)}=\{{\cal C}_{dur,g_{r}}^{(r)}\}_{g_{r}=1}^{k_{dur,r}}\)</span>
of its levels, where <span class="math inline">\(k_{dur,r}\)</span> is the number of clusters for covariate
<span class="math inline">\(r\)</span> and <span class="math inline">\(g_{r}\)</span> represents the cluster index. We introduce latent
variables <span class="math inline">\(\{z_{dur,r,w}\}_{r=1,w=1}^{p+1,d_{r}}\)</span> that indicate the
cluster index for the <span class="math inline">\(w^{th}\)</span> label of the <span class="math inline">\(r^{th}\)</span> covariate. We now
replace the population-level effect
<span class="math inline">\(\boldsymbol\lambda_{dur,x_{1},\dots,x_{p},y_{t-1}}(\cdot)\)</span> with
<span class="math inline">\(\boldsymbol\lambda_{dur,g_{1},\dots,g_{p},g_{p+1}}(\cdot)\)</span>.</p>
<p>The mixture probability vectors are given Dirichlet priors with the mean
vector <span class="math inline">\(\boldsymbol\lambda_{dur,0}\)</span>, which itself centers around a global
vector <span class="math inline">\(\boldsymbol\lambda_{dur,00}\)</span>:<br />
</p>
<p><span class="math display">\[\begin{aligned}
&amp;&amp; \boldsymbol \lambda_{dur,g_{1},\dots,g_{p+1}}(\cdot) \sim \hbox{Dir}\left\{\alpha_{dur,0}\lambda_{dur,0}(1),\dots,\alpha_{dur,0}\lambda_{dur,0}(K)\right\}, \nonumber  \\
&amp;&amp; \boldsymbol \lambda^{(i)}_{dur}(\cdot) \sim \hbox{Dir}\left\{\alpha^{(0)}_{dur}\lambda_{dur,0}(1),\dots,\alpha^{(0)}_{dur}\lambda_{dur,0}(K)\right\}, \nonumber\\
&amp;&amp;\boldsymbol \lambda_{dur,0}(\cdot) \sim \hbox{Dir}\left\{\alpha_{dur,00}\lambda_{dur,00}(1),\dots,\alpha_{dur,00}\lambda_{dur,00}(K)\right\}. \nonumber
\end{aligned}\]</span></p>
<p>We present the complete Bayesian hierarchical model for the continuous
duration times as
<span class="math display">\[\begin{aligned}
&amp;&amp; (\tau_{s,t} \mid z_{dur,s,t}=k) \sim \hbox{Ga}(\tau_{s,t} \mid \alpha_{k},\beta_{k}), \nonumber\\
&amp;&amp; (z_{dur,s,t} \mid i_{s}=i, z_{dur,1,x_{s,1}} =g_{1}, \dots, z_{dur,p,x_{s,p}}=g_{p}, z_{dur,p+1,y_{s,t-1}}=g_{p+1})  \sim \nonumber\\
&amp;&amp;  \hbox{Mult}\left\{P_{dur,g_{1},\dots,g_{p+1}}^{(i)}(1),\dots,P_{dur,g_{1},\dots,g_{p+1}}^{(i)}(K)\right\}, ~~\text{where} \nonumber\\
&amp;&amp; P^{(i)}_{dur,g_{1},\dots,g_{p+1}}(k)=\pi_{dur,0}^{(i)}(k)\lambda_{dur,g_{1},\dots,g_{p+1}}(k)+\pi_{dur,1}^{(i)}(k)\lambda^{(i)}_{dur}(k), \nonumber \\
&amp;&amp; \boldsymbol \lambda_{dur,g_{1},\dots,g_{p+1}}(\cdot) \sim \hbox{Dir}\left\{\alpha_{dur,0}\lambda_{dur,0}(1),\dots,\alpha_{dur,0}\lambda_{dur,0}(K)\right\}, ~~~\alpha_{dur,0}\sim\hbox{Ga}(a_{dur,0},b_{dur,0}), \nonumber\\
&amp;&amp; \boldsymbol \lambda_{dur}^{(i)}(\cdot) \sim \hbox{Dir}\left\{\alpha_{dur}^{(0)}\lambda_{dur,0}(1),\dots,\alpha_{dur}^{(0)}\lambda_{dur,0}(K)\right\}, ~~~\alpha_{dur}^{(0)}\sim\hbox{Ga}(a_{dur}^{(0)},b_{dur}^{(0)}), \nonumber\\
&amp;&amp;\boldsymbol \lambda_{dur,0}(\cdot) \sim \hbox{Dir}\left\{\alpha_{dur,00}\lambda_{dur,00}(1),\dots,\alpha_{dur,00}\lambda_{dur,00}(K)\right\},\nonumber\\
&amp;&amp; \pi_{dur,0}^{(i)}(k) \sim \hbox{Beta}(a_{dur,0},a_{dur,1}),\nonumber\\
&amp;&amp; \alpha_{k} \sim \hbox{Ga}(a_{dur,0},b_{dur,0}), ~~~\beta_{k} \sim \hbox{Ga}(a_{dur,0},b_{dur,0}). \nonumber
\end{aligned}\]</span></p>
<p>Inference is based on samples drawn from the posterior using a
[Metropolis-Hastings-within-Gibbs MCMC algorithm. Most full conditionals
are available in closed form and can be directly sampled from. A
Metropolis-Hastings step is however used for updating the discrete
valued cluster configurations.] There is, however,
no conjugate prior for gamma distributions with unknown shape parameters
<span class="citation" data-cites="damsleth1975conjugate">(<a href="#ref-damsleth1975conjugate" role="doc-biblioref">Damsleth 1975</a>)</span>. Recently, <span class="citation" data-cites="miller2019fast">Miller (<a href="#ref-miller2019fast" role="doc-biblioref">2019</a>)</span> designed a procedure
that efficiently approximates the posterior full conditionals of gamma
shape parameters under a gamma prior with another gamma density. We
adopt this approximation in our MCMC algorithm.</p>
<h3 data-number="3" id="sec:func"><span class="header-section-number">3</span> The <strong>BMRMM</strong> R package</h3>
<h4 class="unnumbered" data-number="3.1" id="package-description">Package description</h4>
<p>The <strong>BMRMM</strong> package is developed to implement Bayesian Markov
(renewal) mixed models. The main function <code>BMRMM</code> of the package carries
out detailed analyses of the state transitions and their duration times
(if applicable) as described in the previous section. [Moreover, the
package includes a number of supplementary functions that use the
results of the main function to produce numerical summaries,
visualizations, and diagnostics. Table <a href="#tab:fn">1</a> provides a brief
description of all functions. ]</p>
<div id="tab:fn">
<div class="layout-chunk" data-layout="l-body">
<table>
<caption><span id="tab:fn">Table 1: </span> Summary of functions in the BMRMM package.</caption>
<colgroup>
<col style="width: 24%" />
<col style="width: 75%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Function</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">BMRMM</td>
<td style="text-align: left;">Creates a BMRMM object.</td>
</tr>
<tr class="even">
<td style="text-align: left;">summary.BMRMM</td>
<td style="text-align: left;">Summary for an object of class BMRMM and create a BMRMMsummary object.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">plot.BMRMMsummary</td>
<td style="text-align: left;">Visualization of a BMRMMsummary object.</td>
</tr>
<tr class="even">
<td style="text-align: left;">hist.BMRMM</td>
<td style="text-align: left;">Returns histograms of duration times for a BMRMM object.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">diag.BMRMM</td>
<td style="text-align: left;">Provides MCMC diagnostic plots for a BMRMM object.</td>
</tr>
<tr class="even">
<td style="text-align: left;">model.selection.scores</td>
<td style="text-align: left;">Returns the LPML and WAIC scores of the mixture gamma model.</td>
</tr>
</tbody>
</table>
</div>
</div>
<h4 class="unnumbered" data-number="3.2" id="the-main-function-bmrmm">The main function BMRMM</h4>
<p>The main function is <code>BMRMM</code> which implements the inference for both the
state transition probabilities and the duration times. We summarize the
parameters in Table <a href="#tab:bmrmm">3</a> and present the function as
follows.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">BMRMM</span>(data, num.cov, <span class="at">cov.labels =</span> <span class="cn">NULL</span>, <span class="at">state.labels =</span> <span class="cn">NULL</span>, </span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">random.effect =</span> <span class="cn">TRUE</span>, <span class="at">fixed.effect =</span> <span class="cn">TRUE</span>, </span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">trans.cov.index =</span> <span class="dv">1</span><span class="sc">:</span>num.cov, <span class="at">duration.cov.index =</span> <span class="dv">1</span><span class="sc">:</span>num.cov, </span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>      <span class="at">duration.distr =</span> <span class="cn">NULL</span>, <span class="at">duration.incl.prev.state =</span> <span class="cn">TRUE</span>,</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">simsize =</span> <span class="dv">10000</span>, <span class="at">burnin =</span> simsize<span class="sc">/</span><span class="dv">2</span>)</span></code></pre></div>
<p>[The parameter <code>data</code> specifies the target data set and needs to follow
a certain structure. ] The first column should list
the individual IDs <span class="math inline">\(i_{s}\)</span>, followed by <span class="math inline">\(p\)</span> columns for the values of
the <span class="math inline">\(p\)</span> associated covariates <span class="math inline">\(x_{s,j}\)</span>, then two columns for the values
of the previous state <span class="math inline">\(y_{s,t-1}\)</span>, the current state <span class="math inline">\(y_{s,t}\)</span>, and
finally a column for duration times <span class="math inline">\(\tau_{s,t}\)</span>. The package supports
one to five categorical covariates that take on values <span class="math inline">\({1,2,\dots}\)</span>.
The duration times column is optional if the user would like to use BMMM
instead of BMRMM to analyze just the state transitions. This is shown in
Table <a href="#tab:data-cols">2</a>. The users can look at the included
[simulated data set] <code>foxp2</code> as an example.</p>
<div id="tab:data-cols">
<table>
<tbody>
<tr class="odd">
<td style="text-align: center;">Id Covariate 1 <span class="math inline">\(\dots\)</span> Covariate <span class="math inline">\(p\)</span> Previous State Current State State Durations/ISI</td>
</tr>
</tbody>
</table>
<hr />
<p>: Table 2: Columns of the desired input data set.</p>
</div>
<p>[The number of covariates in the data set is specified by the argument
<code>num.cov</code>. The argument <code>cov.labels</code> is a list of vectors giving the
names of covariate levels in the covariate order that is presented in
<code>data</code> while the parameter <code>state.labels</code> is a vector providing the
names of the transition states. The default labels are Arabic numerals.] The [<code>random.effect</code>]
parameter gives users the option to exclude the random individual
effects. If [<code>random.effect</code>] is set to <code>FALSE</code>,
the transition probabilities (and the mixture probabilities for duration
times, if applicable) will only consider the influence of the covariate
levels. Similarly, the [<code>fixed.effect</code>] parameter
allows users to exclude the fixed population effects. [The default
values for <code>random.effect</code> and <code>fixed.effect</code> are both
<code>TRUE</code>.] The covariate indices for the two analyses
can be specified by setting <code>trans.cov.index</code> and <code>duration.cov.index</code>.
[We note that indices specified by <code>trans.cov.index</code> and
<code>duration.cov.index</code> refer to the index of the covariate when the first
covariate is given index 1, thus different from its index in
<code>data</code>.]</p>
<p>[Users can define <code>duration.distr</code> in the following three
ways.]</p>
<ol type="1">
<li><p>[If users set <code>duration.distr</code> to be <code>NULL</code>, which is the default
setting, then the duration times will be ignored and not modeled at
all. ] The BMMM described will be implemented
to analyze the existing state transitions alone.</p></li>
<li><p>[If <code>duration.distr</code> is set as <code>list(‘mixDirichlet’, unit)</code>, the
duration times will be used to construct a new state <code>‘dur.state’</code>,
which will be analyzed along with the original set of states.] The additional argument <code>unit</code> must be
defined and acts both as a threshold and as a block size for
duration times. For example, if the <code>unit</code> is set to <span class="math inline">\(5\)</span>, then for
each duration value greater than <span class="math inline">\(5\)</span> units, each block of <span class="math inline">\(5\)</span> unit
in it will be treated as an instance of a new <code>’dur.state’</code> state.
[If there is a state transition from state <code>‘a’</code> to <code>‘b’</code> with a
duration time of <span class="math inline">\(15\)</span> seconds and the <code>unit</code> is specified at <span class="math inline">\(5\)</span>
seconds, then the updated Markov sequence will contain three
consecutive <code>‘dur.state’</code> states, i.e.,
<code>(‘a’, ‘dur.state’, ‘dur.state’, ‘dur.state’, ‘b’)</code>.] Since we adopt the floor operation, a
duration time of say <span class="math inline">\(17.68\)</span> seconds will also be replaced by three
consecutive instances of <code>’dur.state’</code> states in this example. The
BMMM model will then be implemented to analyze the resulting
appended state transitions.</p>
<p>These first two options may naturally result in loss of information
and is therefore not recommended when a detailed analysis of the
distribution of the duration times is warranted.</p></li>
<li><p>[If <code>duration.distr</code> is set to be <code>list(‘mixgamma’, shape, rate)</code>,
the duration times are modeled as a continuous random variable using
a flexible mixture of gamma kernels, as described for a BMRMM
model.] In this case, users can specify the
prior shape and rate parameters with the <code>shape</code> and <code>rate</code>
arguments within the definition of <code>duration.distr</code>. We note that
<code>shape</code> and <code>rate</code> must be numeric vectors of the same length.</p></li>
</ol>
<p>By default, we consider the previous state <span class="math inline">\(y_{s,t-1}\)</span> as a covariate
when we model the duration times as continuous variables, i.e.,
[<code>duration.incl.prev.state</code>] is set to <code>TRUE</code>.
Users can set this parameter to <code>FALSE</code> if they wish to exclude the
previous state when analyzing the duration times. [The remaining
parameters <code>simsize</code> and <code>burnin</code> denote the total number of MCMC
iterations and the number of burn-ins, respectively.]</p>
<div id="tab:bmrmm">
<table style="width:99%;">
<caption>Table 3: Arguments to the <code>BMRMM</code> function.</caption>
<colgroup>
<col style="width: 31%" />
<col style="width: 47%" />
<col style="width: 19%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Argument</th>
<th style="text-align: left;">Explanation</th>
<th style="text-align: left;">Default value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><code>data</code></td>
<td style="text-align: left;">the data set to be used following the required format</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>num.cov</code></td>
<td style="text-align: left;">an integer giving the number of observed covariates in <code>data</code></td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">[<code>cov.labels</code>] a list of vectors giv</td>
<td style="text-align: left;">ing names of all covariate levels [<code>NULL</code>]</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">[<code>state.labels</code>] a vector giving names</td>
<td style="text-align: left;">of the states [<code>NULL</code>]</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">[<code>random.effect</code>] <code>TRUE</code> if random indi</td>
<td style="text-align: left;">vidual effects are included [<code>TRUE</code>]</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">[<code>fixed.effect</code>] <code>TRUE</code> if fixed popul</td>
<td style="text-align: left;">ation effects are included [<code>TRUE</code>]</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>trans.cov.index</code></td>
<td style="text-align: left;">selects the covariates to analyze for transition probabilities</td>
<td style="text-align: left;"><code>1:num.cov</code></td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>duration.cov.index</code></td>
<td style="text-align: left;">selects the covariates to analyze for duration times</td>
<td style="text-align: left;"><code>1:num.cov</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>duration.distr</code></td>
<td style="text-align: left;">specifies the distribution for duration times</td>
<td style="text-align: left;"><code>NULL</code></td>
</tr>
<tr class="even">
<td style="text-align: left;">[<code>duration.incl.prev.state</code>] <code>TRUE</code> if <span class="math inline">\(y_{t-1}\)</span> a</td>
<td style="text-align: left;">cts as a covariate for the analysis of duration times [<code>TRUE</code>]</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;"><code>simsize</code></td>
<td style="text-align: left;">number of MCMC iterations</td>
<td style="text-align: left;">10000</td>
</tr>
<tr class="even">
<td style="text-align: left;"><code>burnin</code></td>
<td style="text-align: left;">number of burnins of the MCMC algorithm</td>
<td style="text-align: left;"><code>simsize</code>/2</td>
</tr>
</tbody>
</table>
</div>
<p>The <code>BMRMM</code> function returns [an object of class <code>BMRMM</code>, which either
contains only <code>results.trans</code> or both of <code>results.trans</code> and
<code>results.duration</code> if duration times follow a mixture gamma
distribution.] For the state transitions, the
posterior mean transition probability matrices for each combination of
the covariate levels and each individual are given by
<code>results.trans$tp.exgns.post.mean</code> and
<code>results.trans$tp.anmls.post.mean</code>, respectively. Additionally,
<code>results.trans$clusters</code> stores cluster configurations for each
covariate from each MCMC iteration. As for duration times, the fields
<code>results.duration$shape.samples</code> and <code>results.duration$rate.samples</code>
record the shape and rate parameters, for each mixture component in
every MCMC iteration, respectively. Meanwhile,
<code>results.duration$comp.assignment</code> gives the assignment of the mixture
component for each data point in the last MCMC iteration. Similar to
transition probabilities, <code>results.duration$clusters</code> gives the cluster
configurations of the covariates. Other elements of <code>results.trans</code> and
<code>results.duration</code> can be found in the detailed R function description.</p>
<h4 class="unnumbered" data-number="3.3" id="summarizing-bmrmm-results">[Summarizing BMRMM results]</h4>
<p>[The <strong>BMRMM</strong> package provides an S3 method for summarizing results of
a <code>BMRMM</code> object as follows. ]</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary.BMRMM</span>(object, <span class="at">delta =</span> <span class="fl">0.02</span>, <span class="at">digits =</span> <span class="dv">2</span>, ...)</span></code></pre></div>
<p>[The <code>object</code> must be of class <code>BMRMM</code>. The argument <code>delta</code> is
associated with local tests for transition probabilities, which we will
explain further. The <code>digit</code> parameter is an integer used for number
formatting, as in the general <code>summary</code> function. The <code>summary.BMRMM</code>
function returns an object of class <code>BMRMMsummary</code> with the following
fields. ]</p>
<ul>
<li><p><code>trans.global</code> and <code>dur.global</code></p>
<p>[These two fields give the global test results from the inference of
transition probabilities and duration times. Global tests show the
significance of the covariates in affecting the state transitions
and duration times. ] Specifically, for each
covariate, the empirical distribution of the size of the clusters in
the stored MCMC iterations is calculated. The null hypothesis that a
covariate is not important is equivalent to the event that all its
levels are in the same cluster, or, in other words, that the cluster
size for the covariate is just one.</p></li>
<li><p><code>trans.probs.mean</code> and <code>trans.probs.sd</code></p>
<p>The two fields provide the mean and standard deviation for the
posterior mean of each transition type under all combinations of
covariate levels, respectively.</p></li>
<li><p><code>trans.local.mean.diff</code> and <code>trans.local.null.test</code></p>
<p>[The <code>BMRMMsummary</code> object also contains local test results for
transition probabilities. Local tests analyze the differences
between the transition probabilities associated with two different
levels of a covariate <span class="math inline">\(j\)</span>, fixing the levels of the other
covariates.] For every pair of levels of
covariate <span class="math inline">\(j\)</span>, <code>trans.local.mean.diff</code> gives the absolute
differences in transition probabilities for each transition type in
the MCMC iterations. The local null hypothesis we test for each
transition type is that this difference is at least the
pre-specified value <code>delta</code>. [Meanwhile, <code>trans.local.null.test</code>
gives the probability of the null hypothesis that the difference
between two covariate levels is not significant under each
transition type. ]</p></li>
<li><p><code>dur.mix.params</code> and <code>dur.mix.probs</code></p>
<p>For each mixture component, <code>dur.mix.params</code> provides the estimates
of the gamma shape and rate parameters from the last MCMC iteration.
For every covariate level, users can obtain the mixture
probabilities by calling the field <code>dur.mix.probs</code>, which can be
further used to estimate the length of the duration times.</p></li>
</ul>
<h4 class="unnumbered" data-number="3.4" id="visualizing-results-with-bmrmm-plotting-functions">[Visualizing results with BMRMM plotting functions]</h4>
<p>[The main plotting function of the package, <code>plot.BMRMMsummary</code>, is an
S3 method for class <code>BMRMMsummary</code>. It gives the barplots for global
tests as well as heatmaps for the posterior mean and standard deviation
for transition probabilities, local tests for transition probabilities,
mixture parameters and probabilities for duration times. The parameters
of <code>plot.BMRMMsummary</code> include <code>x</code>, which must be an object of class
<code>BMRMMsummary</code> and <code>type</code>, which is a single string representing the
field of <code>x</code> that needs to be plotted. The function also takes general
plotting arguments such as <code>xlab</code>, <code>ylab</code>, etc. ]</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.BMRMMsummary</span>(x, type, <span class="at">xlab =</span> <span class="cn">NULL</span>, <span class="at">ylab =</span> <span class="cn">NULL</span>, <span class="at">main =</span> <span class="cn">NULL</span>, <span class="at">col =</span> <span class="cn">NULL</span>, ...)</span></code></pre></div>
<p>[When duration times are analyzed as continuous variables using mixture
gamma distributions, the users can use the S3 method <code>hist.BMRMM</code> to
generate histograms for duration times along with the estimated
posterior distribution. The parameter <code>x</code> is an object of class <code>BMRMM</code>.
The argument <code>comp</code> gives the specific mixture component that the user
would like to investigate. When <code>comp</code> is <code>NULL</code>, which is the default
setting, the histogram of all observed duration times is plotted and
superimposed with the posterior mean of the fitted mixture gamma
distribution. When <code>comp</code> is a specific integer, we will be looking at
the last MCMC iteration. The histogram for duration times assigned with
component <code>comp</code> will be presented alongside the mixture gamma
distribution with the shape and rate parameters from the last MCMC
iteration. Users can refer to the documentation of the general <code>hist</code>
function to see the interpretation for the rest of the parameters.]</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist.BMRMM</span>(x, <span class="at">comp =</span> <span class="cn">NULL</span>, <span class="at">xlim =</span> <span class="cn">NULL</span>, <span class="at">breaks =</span> <span class="cn">NULL</span>, <span class="at">main =</span> <span class="cn">NULL</span>, </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>           <span class="at">col =</span> <span class="st">&#39;gray&#39;</span>, <span class="at">xlab =</span> <span class="st">&#39;Duration times&#39;</span>, <span class="at">ylab =</span> <span class="st">&#39;Density&#39;</span>, ...)</span></code></pre></div>
<p>[Finally, users can check the MCMC diagnostics with the traceplots and
autocorrelation plots produced by the function <code>diag.BMRMM</code>. The
<code>object</code> parameter should be an object of class <code>BMRMM</code>. For transition
probabilities, users can specify the covariate levels as well as the
state transitions they are interested in by defining <code>cov.combs</code> and
<code>transitions</code>, respectively. For duration times, users can define
<code>components</code>, a numeric vector, to obtain the diagnostic plots for shape
and rate parameters of the specific component
kernels.]</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">diag.BMRMM</span>(object, <span class="at">cov.combs =</span> <span class="cn">NULL</span>, <span class="at">transitions =</span> <span class="cn">NULL</span>, <span class="at">components =</span> <span class="cn">NULL</span>) </span></code></pre></div>
<h4 class="unnumbered" data-number="3.5" id="model-selection-scores-for-continuous-duration-times">[Model selection scores for continuous duration times]</h4>
<p>When the duration times are modeled using mixtures of gamma
distributions, model selection can be performed on the number of mixture
components using the function <code>model.selection.scores</code>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">model.selection.scores</span>(object) </span></code></pre></div>
<p>[The function takes an <code>object</code> as its input, which must be an object of
class <code>BMRMM</code>. It returns a list consisting of] the
log pseudo marginal likelihood (LPML) <span class="citation" data-cites="geisser1979predictive">(<a href="#ref-geisser1979predictive" role="doc-biblioref">Geisser and Eddy 1979</a>)</span> and the
widely applicable information criterion (WAIC) <span class="citation" data-cites="watanabe2010asymptotic">(<a href="#ref-watanabe2010asymptotic" role="doc-biblioref">Watanabe and Opper 2010</a>)</span>
scores of the model. Larger values of LPML and smaller values of WAIC
indicate better model fits. They are particularly suitable for complex
Bayesian hierarchical models as they can be easily computed from the
MCMC samples.</p>
<h3 data-number="4" id="sec:foxp2"><span class="header-section-number">4</span> Illustrations on the synthetic FoxP2 data set</h3>
<p>The FoxP2 data set records the songs sung by adult male mice of two
genotypes, wild type or FoxP2, denoted by <span class="math inline">\(W\)</span> and <span class="math inline">\(F\)</span>, respectively
<span class="citation" data-cites="Chabout_etal:2016">(<a href="#ref-Chabout_etal:2016" role="doc-biblioref">Chabout et al. 2016</a>)</span>. The mice sang under three social contexts, <span class="math inline">\(U\)</span>
(fresh female urine on a cotton tip placed inside the male’s cage), <span class="math inline">\(L\)</span>
(an awake and behaving adult female placed inside the cage), and <span class="math inline">\(A\)</span> (an
anesthetized female placed on the lid of the cage). Each song comprises
a sequence of syllables and continuous inter-syllable intervals (ISIs).
The data set can be used to analyze the effect of the FoxP2 gene on the
vocal syntax of mice, in turn providing insights into the effects of the
gene on human vocal communication abilities and related deficiencies.
[The real FoxP2 data set originates from the study by
<span class="citation" data-cites="Chabout_etal:2016">(<a href="#ref-Chabout_etal:2016" role="doc-biblioref">Chabout et al. 2016</a>)</span> and requires permission to use. <span class="citation" data-cites="wu2021bayesian">(<a href="#ref-wu2021bayesian" role="doc-biblioref">Wu et al. 2023</a>)</span>
simulated a data set that closely mimics the real one. For demonstrating
the <strong>BMRMM</strong> package, we included in it a shortened version of this
synthetic data set which we refer to as the <code>foxp2</code> data set.] The <code>foxp2</code> synthetic data set has <span class="math inline">\(17391\)</span> rows
and <span class="math inline">\(6\)</span> columns, which are Id, Genotype, Context, Prev_State, Cur_State,
and Transformed_ISI. The original FoxP2 data set records ISIs in
seconds. In the simulated data set <code>foxp2</code>, following <span class="citation" data-cites="wu2021bayesian">Wu et al. (<a href="#ref-wu2021bayesian" role="doc-biblioref">2023</a>)</span>,
log(1+ISI) values are used which give a better model fit.</p>
<div id="tab:foxp2">
<table style="width:96%;">
<caption>Table 4: Part of the simulated FoxP2 data set <code>foxp2</code>.</caption>
<colgroup>
<col style="width: 6%" />
<col style="width: 15%" />
<col style="width: 13%" />
<col style="width: 18%" />
<col style="width: 16%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">Id</th>
<th style="text-align: center;">Genotype</th>
<th style="text-align: center;">Context</th>
<th style="text-align: center;">Prev_State</th>
<th style="text-align: center;">Cur_State</th>
<th style="text-align: center;">Transformed_ISI</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">0.20197711</td>
</tr>
<tr class="even">
<td style="text-align: center;">1</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">0.06972753</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">0.07211320</td>
</tr>
<tr class="even">
<td style="text-align: center;">1</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">0.15790932</td>
</tr>
<tr class="odd">
<td style="text-align: center;">1</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">0.06781471</td>
</tr>
<tr class="even">
<td style="text-align: center;">1</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">0.09426236</td>
</tr>
</tbody>
</table>
</div>
<p>If we are only interested in analyzing the transition probabilities with
the covariates genotype and social contexts, we would use the main
function as follows.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> res.fp2 <span class="ot">&lt;-</span> <span class="fu">BMRMM</span>(foxp2, <span class="at">num.cov =</span> <span class="dv">2</span>)</span></code></pre></div>
<p>If we would like to pick specific covariates for our analyses, we can
define <code>trans.cov.index</code> and <code>duration.cov.index</code> accordingly. For
example, if we only want to use context for transition probabilities and
genotype for ISIs, we would run the following.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> res.fp2 <span class="ot">&lt;-</span> <span class="fu">BMRMM</span>(foxp2, <span class="at">num.cov =</span> <span class="dv">2</span>, </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">trans.cov.index =</span> <span class="fu">c</span>(<span class="dv">2</span>), <span class="at">duration.cov.index =</span> <span class="fu">c</span>(<span class="dv">1</span>))</span></code></pre></div>
<p>If we would like to analyze the ISIs as part of the original state
sequence following a mixture Dirichlet distribution, as was done by
<span class="citation" data-cites="sarkar2018bayesian">Sarkar et al. (<a href="#ref-sarkar2018bayesian" role="doc-biblioref">2018</a>)</span>, the ISIs are replaced by (possibly consecutive)
"silent" states by dividing them into blocks of 250 milliseconds. The
<code>BMRMM</code> function can do this by setting <code>duration.distr</code> as a list with
the string <code>’mixDirichlet’</code> and the argument <code>unit</code> as
<span class="math inline">\(\hbox{log}(0.25+1)\)</span>, based on the log transformation.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> res.fp2 <span class="ot">&lt;-</span> <span class="fu">BMRMM</span>(foxp2, <span class="at">num.cov =</span> <span class="dv">2</span>,  </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">duration.distr =</span> <span class="fu">list</span>(<span class="st">&#39;mixDirichlet&#39;</span>, <span class="at">unit =</span> <span class="fu">log</span>(<span class="fl">0.25</span><span class="sc">+</span><span class="dv">1</span>)))</span></code></pre></div>
<p>In the next example, we would like to analyze the ISIs as continuous
variables following a mixture gamma distribution. For syllable
transitions, we use both genotype and context as covariates. For the
ISIs, in addition to these two, we also use the preceding syllable as a
covariate.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> res.fp2 <span class="ot">&lt;-</span> <span class="fu">BMRMM</span>(<span class="at">data =</span> foxp2, <span class="at">num.cov =</span> <span class="dv">2</span>, <span class="at">state.labels =</span> <span class="fu">c</span>(<span class="st">&#39;d&#39;</span>, <span class="st">&#39;m&#39;</span>, <span class="st">&#39;s&#39;</span>, <span class="st">&#39;u&#39;</span>), </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">cov.labels =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="st">&#39;F&#39;</span>, <span class="st">&#39;W&#39;</span>), <span class="fu">c</span>(<span class="st">&#39;U&#39;</span>, <span class="st">&#39;L&#39;</span>, <span class="st">&#39;A&#39;</span>)),</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">duration.distr =</span> <span class="fu">list</span>(<span class="st">&#39;mixgamma&#39;</span>, <span class="at">shape =</span> <span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">4</span>), <span class="at">rate =</span> <span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">4</span>)))</span></code></pre></div>
<p>In what follows, we show the results for the last function call. The
returned <code>res.fp2</code> have two parts, which are named
<code>res.fp2$results.trans</code> and <code>res.fp2$results.duration</code>. Now we
demonstrate how we print and visualize the results.</p>
<p>[First, we obtain a <code>BMRMMsummary</code> object, <code>sm.fp2</code>, by calling the
<code>summary.BMRMM</code> function on the returned results <code>res.fp2</code>. The global
test results for identifying the significant covariates can be found by
calling the fields <code>trans.global</code> and <code>dur.global</code>. The function
<code>plot.BMRMMsummary</code> is called to visualize the global tests using
barplots, as presented in Figure <a href="#fig:figglobal">2</a>.] We recall that a covariate is significant when
its levels formed more than one cluster with very high posterior
probability (the bar heights). Figure <a href="#fig:figglobal">2</a> and the
printed results suggest that every covariate is significant for the ISIs
but only the social context is significant for the transition
probabilities.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> sm.fp2 <span class="ot">&lt;-</span> <span class="fu">summary</span>(res.fp2)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> sm.fp2<span class="sc">$</span>trans.global</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>            label_data</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>cluster_data Context Genotype</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>           <span class="dv">1</span>       <span class="dv">0</span>        <span class="dv">1</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>           <span class="dv">3</span>       <span class="dv">1</span>        <span class="dv">0</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> sm.fp2<span class="sc">$</span>dur.global</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>            label_data</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>cluster_data Context Genotype prev_state</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>           <span class="dv">2</span>    <span class="fl">0.00</span>     <span class="fl">1.00</span>       <span class="fl">0.10</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>           <span class="dv">3</span>    <span class="fl">1.00</span>     <span class="fl">0.00</span>       <span class="fl">0.90</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>           <span class="dv">4</span>    <span class="fl">0.00</span>     <span class="fl">0.00</span>       <span class="fl">0.01</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot</span>(sm.fp2, <span class="st">&#39;trans.global&#39;</span>)</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot</span>(sm.fp2, <span class="st">&#39;dur.global&#39;</span>)</span></code></pre></div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figglobal"></span>
<img src="figs/global_trans_foxp2.png" alt="graphic without alt text" width="169" /><img src="figs/global_isi_foxp2.png" alt="graphic without alt text" width="169" />
<p class="caption">
Figure 2: Results for the simulated foxp2 data set showing the global tests of significance of the covariates for the state transitions (left) and the ISIs (right). The bars represent the estimated posterior probabilities of the number of clusters formed by the levels of each covariate.
</p>
</div>
</div>
<p>[The plotting function can be called to visualize the posterior
transition probabilities under different combinations of the covariate
levels. ] We show in Figure <a href="#fig:figpost-mean">3</a>
the heatmaps for the posterior mean and standard deviation of the
transition probabilities for each transition type for the following
combinations of covariates: <span class="math inline">\((F,A)\)</span> and <span class="math inline">\((W,L)\)</span>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot</span>(sm.fp2, <span class="st">&#39;trans.probs.mean&#39;</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot</span>(sm.fp2, <span class="st">&#39;trans.probs.sd&#39;</span>)</span></code></pre></div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figpost-mean"></span>
<img src="figs/post_mean_FA.png" alt="graphic without alt text" width="375" /><img src="figs/post_mean_WL.png" alt="graphic without alt text" width="375" />
<p class="caption">
Figure 3: Results for the simulated foxp2 data set showing the posterior mean and standard deviation for each transition type for selected covariate combinations, (F,A) (top) and (W,L) (bottom).
</p>
</div>
</div>
<p>We also perform the local test to assess the influence of genotype on
the transition probabilities by computing the absolute difference of the
transition probabilities between <span class="math inline">\(F\)</span> and <span class="math inline">\(W\)</span> among the thinned MCMC
samples after burn-ins, i.e.,
<span class="math inline">\(|\Delta \lambda_{trans,\cdot,x_2}(y_t\mid y_{t-1})|=|\lambda_{trans,1,x_2}(y_t\mid y_{t-1})-\lambda_{trans,2,x_2}(y_t\mid y_{t-1})|\)</span>.
The estimated posterior probability for the null hypothesis is therefore
the proportion of times
<span class="math inline">\(|\Delta \lambda_{trans,\cdot,x_2}(y_t\mid y_{t-1}) \le \delta|\)</span> is
observed in the MCMC samples, where <span class="math inline">\(x_2\)</span> is the social context and
<span class="math inline">\(\delta\)</span> is the user-specific difference threshold <code>delta</code>. [The
plotting function <code>plot.BMRMMsummary</code> gives the plots for all local test
results if we set the <code>type</code> to be <code>‘trans.local.mean.diff’</code> or
<code>‘trans.local.null.test’</code>. Here, we show the results of local tests for
the covariate 1 (i.e., genotype) with <code>delta</code> equaling the default value
of 0.02, and present the plots in
Figure <a href="#fig:figlocal">4</a>.] From the figure, we
see that the posterior probabilities of the null hypotheses are
generally large for most transition types (e.g., transitions to the
syllable <span class="math inline">\(u\)</span>) regardless of the social context, indicating that genotype
does not have a strong influence on transition probabilities with a
fixed context under these transition types.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot</span>(sm.fp2, <span class="st">&#39;trans.local.mean.diff&#39;</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot</span>(sm.fp2, <span class="st">&#39;trans.local.null.test&#39;</span>)</span></code></pre></div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figlocal"></span>
<img src="figs/local_U.png" alt="graphic without alt text" width="375" /><img src="figs/local_L.png" alt="graphic without alt text" width="375" /><img src="figs/local_A.png" alt="graphic without alt text" width="375" />
<p class="caption">
Figure 4: Results for the simulated foxp2 data set showing local test results for genotypes fixing the social context, U (top), L (middle), and A (bottom). The averaged absolute difference in transition probabilities between F and W is presented on the left. The posterior probabilities of the corresponding null hypotheses are on the right.
</p>
</div>
</div>
<p>Next, we turn our attention to the ISIs. We first check the fit of our
estimated mixture gamma distribution presented in
Figure <a href="#fig:fighist">5</a>. We then look further into the shape of each
mixture component in Figure <a href="#fig:fighist">5</a>. From the histogram
for each component, we see that components 2 and 4 represent longer ISIs
while components 1 and 3 represent shorter ISIs.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">hist</span>(res.fp2, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="cf">for</span>(comp <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>) {</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>     <span class="fu">hist</span>(res.fp2, <span class="at">comp =</span> comp)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>   }</span></code></pre></div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:fighist"></span>
<img src="figs/hist_isi.png" alt="graphic without alt text" width="230" /><img src="figs/hist_by_comp.png" alt="graphic without alt text" width="210" />
<p class="caption">
Figure 5: Results for the simulated foxp2 data set showing the histograms of the ISIs with the estimated posterior gamma mixture density (left) and the histograms of the ISIs for each mixture component (right).
</p>
</div>
</div>
<p>We examine the values of mixture parameters and mixture probabilities
for each covariate level in the last MCMC iteration, which provides
insights into the influence of the covariate on ISI lengths.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> sm.fp2<span class="sc">$</span>dur.mix.params</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>       shape.k rate.k</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">1</span>   <span class="fl">29.07</span> <span class="fl">394.30</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">2</span>    <span class="fl">1.23</span>   <span class="fl">1.49</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">3</span>    <span class="fl">8.46</span> <span class="fl">465.66</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">4</span>    <span class="fl">3.03</span>  <span class="fl">20.13</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> sm.fp2<span class="sc">$</span>dur.mix.probs</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>Genotype</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>          F    W</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">1</span> <span class="fl">0.46</span> <span class="fl">0.48</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">2</span> <span class="fl">0.19</span> <span class="fl">0.13</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">3</span> <span class="fl">0.10</span> <span class="fl">0.15</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">4</span> <span class="fl">0.25</span> <span class="fl">0.24</span></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>Context </span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>          U    L    A</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">1</span> <span class="fl">0.58</span> <span class="fl">0.35</span> <span class="fl">0.48</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">2</span> <span class="fl">0.14</span> <span class="fl">0.16</span> <span class="fl">0.18</span></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">3</span> <span class="fl">0.08</span> <span class="fl">0.21</span> <span class="fl">0.08</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">4</span> <span class="fl">0.20</span> <span class="fl">0.28</span> <span class="fl">0.25</span></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>prev_state </span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>          d    m    s    u</span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">1</span> <span class="fl">0.52</span> <span class="fl">0.52</span> <span class="fl">0.42</span> <span class="fl">0.42</span></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">2</span> <span class="fl">0.13</span> <span class="fl">0.13</span> <span class="fl">0.22</span> <span class="fl">0.17</span></span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">3</span> <span class="fl">0.11</span> <span class="fl">0.11</span> <span class="fl">0.10</span> <span class="fl">0.18</span></span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">4</span> <span class="fl">0.24</span> <span class="fl">0.24</span> <span class="fl">0.26</span> <span class="fl">0.23</span></span></code></pre></div>
<p>From the mixture probabilities, we see that mice with genotype <span class="math inline">\(F\)</span> have
a much higher mixture probability in component 2 compared to genotype
<span class="math inline">\(W\)</span>, which indicates mice with the FoxP2 mutation require a longer ISI
between pronouncing two syllables, a reflection of vocal impairment.</p>
<p>Finally, we check the MCMC diagnostic plots and see if we had good
mixing for the parameters. Here we focus on a specific transition type,
<span class="math inline">\(u\rightarrow m\)</span>, for covariate combination <span class="math inline">\(\{F,U\}\)</span> and a specific
mixture component (component 2). We show these plots in
Figure <a href="#fig:figdiag">6</a>.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">diag.BMRMM</span>(res.fp2, <span class="at">cov.combs =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>)), </span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">transitions =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">2</span>)), <span class="at">components =</span> <span class="fu">c</span>(<span class="dv">2</span>))</span></code></pre></div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figdiag"></span>
<img src="figs/diag_FU_um.png" alt="graphic without alt text" width="308" /><img src="figs/diag_isi_comp2.png" alt="graphic without alt text" width="298" />
<p class="caption">
Figure 6: Results for the simulated foxp2 data set showing the MCMC diagnostic plots, including traceplots and autocorrelation plots.
</p>
</div>
</div>
<h3 data-number="5" id="sec:asthma"><span class="header-section-number">5</span> Illustrations on the asthma control data set</h3>
<p>The <strong>BMRMM</strong> package is able to analyze duration times in detail which
could either be the ISIs, as seen in the synthetic <code>foxp2</code> data set, or
the state persistence times, as in a traditional semi-Markov model. To
demonstrate the usage of our package in analyzing the state persistence
times, we use the asthma control data set from the ARIA (Association
pour la Recherche en Intelligence Artificielle) study of severe
asthmatic patients <span class="citation" data-cites="combescure2003assessment">(<a href="#ref-combescure2003assessment" role="doc-biblioref">Combescure et al. 2003</a>)</span> in France between 1997
and 2001. At each visit, a chest physician graded the asthma control
status of the patient using control scores <span class="citation" data-cites="juniper1999development">(<a href="#ref-juniper1999development" role="doc-biblioref">Juniper et al. 1999</a>)</span>.
The data set contains the sojourn time of the control states as well as
three covariates: Asthma severity, sex, and the body mass index (BMI) of
the patients. <span class="citation" data-cites="saint2003analysis">Saint-Pierre et al. (<a href="#ref-saint2003analysis" role="doc-biblioref">2003</a>)</span> used a Markov model with piece-wise
constant intensities to model the asthma control evolution and proposed
a regression model for analyzing the effect of covariates.
<span class="citation" data-cites="combescure2003assessment">Combescure et al. (<a href="#ref-combescure2003assessment" role="doc-biblioref">2003</a>)</span> used the data set to assess the relationship
between asthma severity and control of asthma. <span class="citation" data-cites="listwon2015semimarkov">Listwon and Saint-Pierre (<a href="#ref-listwon2015semimarkov" role="doc-biblioref">2015</a>)</span>
fitted a semi-Markov model for the sojourn times using exponential and
Weibull distributions and analyzed the effect of covariates individually
due to complexity. Our <strong>BMRMM</strong> package is able to analyze the effect
of the three covariates while also incorporating random individual
effects exhibited by different patients on transition dynamics and state
duration times.</p>
<p>[The <code>asthma</code> data set we use here is from the <strong>SemiMarkov</strong> package
<span class="citation" data-cites="listwon2015semimarkov">(<a href="#ref-listwon2015semimarkov" role="doc-biblioref">Listwon and Saint-Pierre 2015</a>)</span>. We have renamed and reordered the columns such
that the data set fits the required format.]
Specifically, the data set has <span class="math inline">\(928\)</span> rows, recording the asthma control
states of <span class="math inline">\(371\)</span> patients, which is one of the following three transient
states: Optimal control (State 1), sub-optimal control (State 2), and
unacceptable control (State 3). Each state can transit to any other two
states and the state duration times are recorded. The data set also
contains three binary covariates of the asthma patients, including the
disease severity (1 if mild-moderate and 2 if severe), BMI (body mass
index, 1 if BMI <span class="math inline">\(&lt;25\)</span> and 2 otherwise), and sex (1 if women and 2 if
men). We display part of the processed data in Table <a href="#tab:asthma">5</a>,
where <code>Duration</code> is the sojourn time in <code>Prev_State</code>.</p>
<div id="tab:asthma">
<table style="width:90%;">
<caption>Table 5: Part of the <code>asthma</code> data set from the ARIA study of severe
asthmatic patients.</caption>
<colgroup>
<col style="width: 6%" />
<col style="width: 15%" />
<col style="width: 8%" />
<col style="width: 8%" />
<col style="width: 18%" />
<col style="width: 16%" />
<col style="width: 16%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;">Id</th>
<th style="text-align: center;">Severity</th>
<th style="text-align: center;">BMI</th>
<th style="text-align: center;">Sex</th>
<th style="text-align: center;">Prev_State</th>
<th style="text-align: center;">Cur_State</th>
<th style="text-align: center;">Duration</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">0.1533</td>
</tr>
<tr class="even">
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">4.1232</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">0.0958</td>
</tr>
<tr class="even">
<td style="text-align: center;">3</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">0.2300</td>
</tr>
<tr class="odd">
<td style="text-align: center;">3</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">3</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">0.2656</td>
</tr>
<tr class="even">
<td style="text-align: center;">3</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">2</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">5.4073</td>
</tr>
</tbody>
</table>
</div>
<p>We investigate the transition dynamics and state persistence times of
the <code>asthma</code> data set using the <code>BMRMM</code> function. We consider <span class="math inline">\(K=4\)</span>
mixture components for state persistence times. The choice of <span class="math inline">\(K\)</span> is
derived from running the BMRMM model several times with different <span class="math inline">\(K\)</span>’s
and comparing the fitness of the models using the LPML and WAIC scores.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> res.asm <span class="ot">&lt;-</span> <span class="fu">BMRMM</span>(<span class="at">data =</span> asthma, <span class="at">num.cov =</span> <span class="dv">3</span>, <span class="at">state.labels =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>), </span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">cov.labels =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="st">&#39;Mild-Moderate&#39;</span>,<span class="st">&#39;Severe&#39;</span>), </span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>                                      <span class="fu">c</span>(<span class="st">&#39;BMI&lt;25&#39;</span>,<span class="st">&#39;BMI&gt;=25&#39;</span>), </span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>                                      <span class="fu">c</span>(<span class="st">&#39;Women&#39;</span>,<span class="st">&#39;Men&#39;</span>)),</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">duration.distr =</span> <span class="fu">list</span>(<span class="st">&#39;mixgamma&#39;</span>, <span class="at">shape =</span> <span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">4</span>), </span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>                                                       <span class="at">rate =</span> <span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">4</span>)))</span></code></pre></div>
<p>We name the returned <code>BMRMM</code> object <code>res.asm</code> and obtain the
<code>BMRMMsummary</code> object <code>sm.asm</code> by calling the <code>summary.BMRMM</code> function.
As in the FoxP2 application, we first plot the global test results for
both transition probabilities and state persistence times in
Figure <a href="#fig:figglobal-asthma">7</a>. For the transition probabilities,
only the severity of asthma is significant while for duration times only
the preceding state is significant. The BMI value and the sex are not
significant for either transition dynamics or state durations.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figglobal-asthma"></span>
<img src="figs/global_trans_asthma.png" alt="graphic without alt text" width="234" /><img src="figs/global_isi_asthma.png" alt="graphic without alt text" width="234" />
<p class="caption">
Figure 7: Results for the asthma data set showing the global tests of significance of the covariates for the state transitions (left) and the state persistence times (right). The bars represent the estimated posterior probabilities of the number of clusters formed by the levels of each covariate.
</p>
</div>
</div>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> sm.asm <span class="ot">&lt;-</span> <span class="fu">summary</span>(res.asm)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> sm.asm<span class="sc">$</span>trans.global</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>            label_data</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>cluster_data  BMI Severity  Sex</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>           <span class="dv">1</span> <span class="fl">0.96</span>     <span class="fl">0.00</span> <span class="fl">1.00</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>           <span class="dv">2</span> <span class="fl">0.04</span>     <span class="fl">1.00</span> <span class="fl">0.00</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> sm.asm<span class="sc">$</span>dur.global</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>                       label_data</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>cluster_data  BMI prev_state Severity  Sex</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>           <span class="dv">1</span> <span class="fl">0.88</span>       <span class="fl">0.00</span>     <span class="fl">0.88</span> <span class="fl">0.99</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>           <span class="dv">2</span> <span class="fl">0.12</span>       <span class="fl">0.14</span>     <span class="fl">0.12</span> <span class="fl">0.01</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>           <span class="dv">3</span> <span class="fl">0.00</span>       <span class="fl">0.86</span>     <span class="fl">0.00</span> <span class="fl">0.00</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot</span>(sm.asm, <span class="st">&#39;trans.global&#39;</span>)</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot</span>(sm.asm, <span class="st">&#39;dur.global&#39;</span>)</span></code></pre></div>
<p>We show the posterior mean and standard deviations of the state
transition probabilities for men and women with severe conditions and
BMI <span class="math inline">\(\ge25\)</span> in Figure <a href="#fig:figpost-mean-asthma">8</a>. We see that for
severe patients with BMI <span class="math inline">\(\ge25\)</span>, the transition probabilities are
similar for men and women. We also take a look at the local test results
for the BMI values fixing the severity of the patients in
Figure <a href="#fig:figlocal-asthma">9</a>. Though the absolute differences
between the two covariate levels for BMI are small, the probabilities
for the null hypotheses are also small, especially for transitions to
state 1 and state 2. This suggests that even though the influence of BMI
on state transitions is not significant globally, it is significant
given the severity of asthma condition regardless of sex.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot</span>(sm.asm, <span class="st">&#39;trans.probs.mean&#39;</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot</span>(sm.asm, <span class="st">&#39;trans.probs.sd&#39;</span>)</span></code></pre></div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figpost-mean-asthma"></span>
<img src="figs/asthma_sev_men.png" alt="graphic without alt text" width="457" /><img src="figs/asthma_sev_women.png" alt="graphic without alt text" width="457" />
<p class="caption">
Figure 8: Results for the asthma data set showing the posterior mean and standard deviation for each transition type for selected covariate combinations, <span class="math inline">\(\{\text{Severe, BMI}\ge25,\text{Men}\}\)</span> (top) and <span class="math inline">\(\{\text{Severe, BMI}\ge25,\text{Women}\}\)</span> (bottom).
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figlocal-asthma"></span>
<img src="figs/compare_sev_m.png" alt="graphic without alt text" width="457" /><img src="figs/compare_sev_w.png" alt="graphic without alt text" width="457" />
<p class="caption">
Figure 9: Results for the asthma data set showing local test results for BMI fixing asthma severity condition and sex, men (top) and women (bottom). The averaged absolute difference in transition probabilities between BMI &lt;25 and BMI <span class="math inline">\(\ge25\)</span> is presented on the left. The posterior probability of the null hypothesis is on the right.
</p>
</div>
</div>
<p>Figure <a href="#fig:fighist-asthma">10</a> presents the histograms of the entire
asthma data set, superimposed with the posterior mean of the mixture
gamma distribution. With four components, the estimated mixture gamma
fits the asthma data well. From the histogram for each component, we see
from Figure <a href="#fig:fighist-asthma">10</a> that component 1 and 2
represents shorter state persistence times while components 3 and 4
represent longer durations.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">hist</span>(res.asm, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="cf">for</span>(comp <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>) {</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>     <span class="fu">hist</span>(res.asm, <span class="at">comp =</span> comp)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>   }</span></code></pre></div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:fighist-asthma"></span>
<img src="figs/hist_isi_asthma.png" alt="graphic without alt text" width="221" /><img src="figs/hist_by_comp_asthma.png" alt="graphic without alt text" width="221" />
<p class="caption">
Figure 10: Results for the asthma data set showing the histograms of ISIs with the estimated posterior gamma mixture density (left) and histograms of ISIs for each mixture component (right).
</p>
</div>
</div>
<p>We investigate the covariates’ influence by examining the mixture
probabilities from the last MCMC iteration. An interesting discovery is
that the mixture probabilities for both sexes, BMI levels, and severity
levels are the same, indicating that the levels of these three
covariates do not strongly influence the distributions of state
durations. This matches the global test results in
Figure <a href="#fig:figglobal-asthma">7</a>. If the preceding state is state 1,
which is optimal control for asthma, the state duration time is longer
than that if the previous state is 2 or 3, as there is a lower weight in
component 1 and higher weight in components 3 and 4. On the other hand,
if the preceding state is 3, which is unacceptable control, then the
state duration time is much shorter, as the mixture probability in
component 1 is much higher when the preceding state is state 3. We
present some examples of the diagnostic plots for the <code>asthma</code> data set
in Figure <a href="#fig:figasthma-diag">11</a>.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> sm.asm<span class="sc">$</span>dur.mix.probs</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>Severity </span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>       Mild<span class="sc">-</span>Moderate Severe</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">1</span>          <span class="fl">0.37</span>   <span class="fl">0.37</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">2</span>          <span class="fl">0.26</span>   <span class="fl">0.26</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">3</span>          <span class="fl">0.20</span>   <span class="fl">0.20</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">4</span>          <span class="fl">0.17</span>   <span class="fl">0.17</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>BMI </span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>       BMI<span class="sc">&lt;</span><span class="dv">25</span> BMI<span class="sc">&gt;=</span><span class="dv">25</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">1</span>   <span class="fl">0.37</span>    <span class="fl">0.37</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">2</span>   <span class="fl">0.26</span>    <span class="fl">0.26</span></span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">3</span>   <span class="fl">0.20</span>    <span class="fl">0.20</span></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">4</span>   <span class="fl">0.17</span>    <span class="fl">0.17</span></span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>Sex </span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>       Women  Men</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">1</span>  <span class="fl">0.37</span> <span class="fl">0.37</span></span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">2</span>  <span class="fl">0.26</span> <span class="fl">0.26</span></span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">3</span>  <span class="fl">0.20</span> <span class="fl">0.20</span></span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">4</span>  <span class="fl">0.17</span> <span class="fl">0.17</span></span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>prev_state</span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a>          <span class="dv">1</span>    <span class="dv">2</span>    <span class="dv">3</span></span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">1</span> <span class="fl">0.27</span> <span class="fl">0.33</span> <span class="fl">0.51</span></span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">2</span> <span class="fl">0.23</span> <span class="fl">0.33</span> <span class="fl">0.23</span></span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">3</span> <span class="fl">0.31</span> <span class="fl">0.20</span> <span class="fl">0.09</span></span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a>Comp <span class="dv">4</span> <span class="fl">0.19</span> <span class="fl">0.15</span> <span class="fl">0.17</span></span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-30"><a href="#cb21-30" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">diag.BMRMM</span>(res.fp2, <span class="at">cov.combs =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">1</span>)), </span>
<span id="cb21-31"><a href="#cb21-31" aria-hidden="true" tabindex="-1"></a>              <span class="at">transitions =</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>)), <span class="at">components =</span> <span class="fu">c</span>(<span class="dv">3</span>))</span></code></pre></div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figasthma-diag"></span>
<img src="figs/asthma_diag_tp.png" alt="graphic without alt text" width="262" /><img src="figs/asthma_diag_comp3.png" alt="graphic without alt text" width="378" />
<p class="caption">
Figure 11: Results for the asthma data set showing the MCMC diagnostic plots, including traceplots and autocorrelation plots.
</p>
</div>
</div>
<h3 data-number="6" id="sec:end"><span class="header-section-number">6</span> Conclusion</h3>
<p>We presented the <strong>BMRMM</strong> package which implements both Bayesian Markov
mixed models (BMMM) for analyzing the state transitions and Bayesian
Markov renewal mixed models (BMRMM) for additionally analyzing the
duration times (being either state persistence times or inter-state
intervals) in a collection of categorical sequences, using flexible
Dirichlet and gamma mixtures, respectively. The BMRMM takes into account
fixed effects of the associated covariates as well as random effects of
the associated individuals while simultaneously selecting the
significant covariates separately for the state transitions and the
duration times. The package includes a synthetic <code>foxp2</code> data set to
demonstrate the data framework and function usages. The package also
provides a series of plotting functions for visualizing the results of
the analyses, including various global and local hypotheses tests, MCMC
diagnostics, etc. We are committed to maintaining and further developing
the package in the future. [Future improvements to the package may
include more options for the distribution types of transition
probabilities and duration times beyond the currently available mixture
Dirichlet and mixture gamma distributions,
respectively.]</p>
</div>
<h3 data-number="7" id="sec:ack"><span class="header-section-number">7</span> Acknowledgements</h3>
<p>We thank two anonymous reviewers very much for their careful review of
our work and their constructive comments that led to significant
improvements to both the package and this paper.
:::
:::::::::</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h3 class="appendix" data-number="8" id="supplementary-materials"><span class="header-section-number">8</span> Supplementary materials</h3>
<p>Supplementary materials are available in addition to this article. It can be downloaded at
<a href="RJ-2024-011.zip">RJ-2024-011.zip</a></p>
<h3 class="appendix" data-number="9" id="cran-packages-used"><span class="header-section-number">9</span> CRAN packages used</h3>
<p><a href="https://cran.r-project.org/package=BMRMM">BMRMM</a>, <a href="https://cran.r-project.org/package=markovchain">markovchain</a>, <a href="https://cran.r-project.org/package=msm">msm</a>, <a href="https://cran.r-project.org/package=SemiMarkov">SemiMarkov</a>, <a href="https://cran.r-project.org/package=SMM">SMM</a>, <a href="https://cran.r-project.org/package=mhsmm">mhsmm</a>, <a href="https://cran.r-project.org/package=hhsmm">hhsmm</a>, <a href="https://cran.r-project.org/package=Countr">Countr</a>, <a href="https://cran.r-project.org/package=ARPobservation">ARPobservation</a>, <a href="https://cran.r-project.org/package=catdap">catdap</a>, <a href="https://cran.r-project.org/package=vcd">vcd</a></p>
<h3 class="appendix" data-number="10" id="cran-task-views-implied-by-cited-packages"><span class="header-section-number">10</span> CRAN Task Views implied by cited packages</h3>
<p><a href="https://cran.r-project.org/view=Distributions">Distributions</a>, <a href="https://cran.r-project.org/view=Finance">Finance</a>, <a href="https://cran.r-project.org/view=MissingData">MissingData</a>, <a href="https://cran.r-project.org/view=Survival">Survival</a>, <a href="https://cran.r-project.org/view=TeachingStatistics">TeachingStatistics</a></p>
<h3 class="appendix" data-number="11" id="note"><span class="header-section-number">11</span> Note</h3>
<p>This article is converted from a Legacy LaTeX article using the
<a href="https://cran.r-project.org/package=texor">texor</a> package.
The pdf version is the official version. To report a problem with the html,
refer to CONTRIBUTE on the R Journal homepage.</p>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-alioum2001mkvpci" class="csl-entry" role="listitem">
A. Alioum and D. Commenges. <span>MKVPCI</span>: A computer program for <span>M</span>arkov models with piecewise constant intensities and covariates. <em>Computer Methods and Programs in Biomedicine</em>, 64: 109–119, 2001. URL <a href="https://doi.org/10.1016/s0169-2607(00)00094-8">https://doi.org/10.1016/s0169-2607(00)00094-8</a>.
</div>
<div id="ref-alioum1998effect" class="csl-entry" role="listitem">
A. Alioum, V. Leroy, D. Commenges, F. Dabis, R. Salamon, G. d’Epidémiologie Clinique du SIDA en Aquitaine, et al. Effect of gender, age, transmission category, and antiretroviral therapy on the progression of human immunodeficiency virus infection using multistate <span>M</span>arkov models. <em>Epidemiology</em>, 9: 605–612, 1998. URL <a href="https://www.jstor.org/stable/3702781">https://www.jstor.org/stable/3702781</a>.
</div>
<div id="ref-alvarez2005estimation" class="csl-entry" role="listitem">
E. E. Alvarez. Estimation in stationary <span>M</span>arkov renewal processes, with application to earthquake forecasting in <span>T</span>urkey. <em>Methodology and Computing in Applied Probability</em>, 7: 119–130, 2005. URL <a href="https://doi.org/10.1007/s11009-005-6658-2">https://doi.org/10.1007/s11009-005-6658-2</a>.
</div>
<div id="ref-amini2022hhsmm" class="csl-entry" role="listitem">
M. Amini, A. Bayat and R. Salehian. Hhsmm: An <span>R</span> package for hidden hybrid <span>M</span>arkov/semi-<span>M</span>arkov models. <em>Computational Statistics</em>, 1: 1–53, 2022. URL <a href="https://doi.org/10.1007/s00180-022-01248-x">https://doi.org/10.1007/s00180-022-01248-x</a>.
</div>
<div id="ref-bacallado2009bayesian" class="csl-entry" role="listitem">
S. Bacallado, J. D. Chodera and V. Pande. Bayesian comparison of <span>M</span>arkov models of molecular dynamics with detailed balance constraint. <em>The Journal of Chemical Physics</em>, 131: 1–10, 2009. URL <a href="https://doi.org/10.1063/1.3192309">https://doi.org/10.1063/1.3192309</a>.
</div>
<div id="ref-barbu2018smm" class="csl-entry" role="listitem">
V. S. Barbu, C. Bérard, D. Cellier, M. Sautreuil and N. Vergne. <span>SMM</span>: An <span>R</span> package for estimation and simulation of discrete-time semi-<span>M</span>arkov models. <em>The R Journal</em>, 10: 226–246, 2018. URL <a href="https://doi.org/10.32614/RJ-2018-050">https://doi.org/10.32614/RJ-2018-050</a>.
</div>
<div id="ref-bulla2007bayesian" class="csl-entry" role="listitem">
P. Bulla and P. Muliere. Bayesian nonparametric estimation for reinforced <span>M</span>arkov renewal processes. <em>Statistical Inference for Stochastic Processes</em>, 10: 283–303, 2007. URL <a href="https://doi.org/10.1007/s11203-006-9000-x">https://doi.org/10.1007/s11203-006-9000-x</a>.
</div>
<div id="ref-Castellucci_etal:2016" class="csl-entry" role="listitem">
G. A. Castellucci, M. J. McGinley and D. A. McCormick. Knockout of <span>F</span>oxp2 disrupts vocal development in mice. <em>Nature Scientific Reports</em>, 6: 1–14, 2016. URL <a href="https://doi.org/10.1038/srep23305">https://doi.org/10.1038/srep23305</a>.
</div>
<div id="ref-Chabout_etal:2016" class="csl-entry" role="listitem">
J. Chabout, A. Sarkar, S. Patel, T. Raiden, D. B. Dunson, S. E. Fisher and E. D. Jarvis. A <span>F</span>oxp2 mutation implicated in human speech deficits alters sequencing of ultrasonic vocalizations in adult male mice. <em>Frontiers in Behavioral Neuroscience</em>, 10: 1–18, 2016. URL <a href="https://doi.org/10.3389/fnbeh.2016.00197">https://doi.org/10.3389/fnbeh.2016.00197</a>.
</div>
<div id="ref-combescure2003assessment" class="csl-entry" role="listitem">
C. Combescure, P. Chanez, P. Saint-Pierre, J. P. Daures, H. Proudhon, P. Godard, et al. Assessment of variations in control of asthma over time. <em>European Respiratory Journal</em>, 22: 298–304, 2003. URL <a href="https://doi.org/10.1183/09031936.03.00081102">https://doi.org/10.1183/09031936.03.00081102</a>.
</div>
<div id="ref-damsleth1975conjugate" class="csl-entry" role="listitem">
E. Damsleth. Conjugate classes for gamma distributions. <em>Scandinavian Journal of Statistics</em>, 2: 80–84, 1975. URL <a href="https://www.jstor.org/stable/4615580">https://www.jstor.org/stable/4615580</a>.
</div>
<div id="ref-diaconis2006bayesian" class="csl-entry" role="listitem">
P. Diaconis and S. W. Rolles. Bayesian analysis for reversible <span>M</span>arkov chains. <em>The Annals of Statistics</em>, 34: 1270–1292, 2006. URL <a href="https://doi.org/10.1214/009053606000000290">https://doi.org/10.1214/009053606000000290</a>.
</div>
<div id="ref-eichelsbacher2002bayesian" class="csl-entry" role="listitem">
P. Eichelsbacher and A. Ganesh. Bayesian inference for <span>M</span>arkov chains. <em>Journal of Applied Probability</em>, 39: 91–99, 2002. URL <a href="https://www.jstor.org/stable/3215920">https://www.jstor.org/stable/3215920</a>.
</div>
<div id="ref-epifani2014bayesian" class="csl-entry" role="listitem">
I. Epifani, L. Ladelli and A. Pievatolo. Bayesian estimation for a parametric <span>M</span>arkov renewal model applied to seismic data. <em>Electronic Journal of Statistics</em>, 8: 2264–2295, 2014. URL <a href="https://doi.org/10.1214/14-EJS952">https://doi.org/10.1214/14-EJS952</a>.
</div>
<div id="ref-etterson2007analysis" class="csl-entry" role="listitem">
M. A. Etterson, B. Olsen and R. S. Greenberg. The analysis of covariates in multi-fate <span>M</span>arkov chain nest-failure models. <em>Studies in Avian Biology</em>, 34: 55–64, 2007. URL <a href="https://sora.unm.edu/node/139709">https://sora.unm.edu/node/139709</a>.
</div>
<div id="ref-ferguson2012mssurv" class="csl-entry" role="listitem">
N. Ferguson, S. Datta and G. Brock. <span class="nocase">msSurv</span>: An <span>R</span> package for nonparametric estimation of multistate models. <em>Journal of Statistical Software</em>, 50: 1–24, 2012. URL <a href="https://doi.org/10.18637/jss.v050.i14">https://doi.org/10.18637/jss.v050.i14</a>.
</div>
<div id="ref-Fujita_etal:2008" class="csl-entry" role="listitem">
E. Fujita, Y. Tanabe, A. Shiota, M. Ueda, K. Suwa, M. Y. Momoi and T. Momoi. Ultrasonic vocalization impairment of <span>F</span>oxp2 (<span>R552H</span>) knockin mice related to speech-language disorder and abnormality of <span>P</span>urkinje cells. <em>Proceedings of the National Academy of Sciences</em>, 105: 3117–3122, 2008. URL <a href="https://doi.org/10.1073/pnas.0712298105">https://doi.org/10.1073/pnas.0712298105</a>.
</div>
<div id="ref-Gaub_etal:2016" class="csl-entry" role="listitem">
S. Gaub, S. E. Fisher and G. Ehret. Ultrasonic vocalizations of adult male <span>F</span>oxp2-mutant mice: Behavioral contexts of arousal and emotion. <em>Genes, Brain and Behavior</em>, 15: 243–259, 2016. URL <a href="https://doi.org/10.1111/gbb.12274">https://doi.org/10.1111/gbb.12274</a>.
</div>
<div id="ref-geisser1979predictive" class="csl-entry" role="listitem">
S. Geisser and W. F. Eddy. A predictive approach to model selection. <em>Journal of the American Statistical Association</em>, 74: 153–160, 1979. URL <a href="https://doi.org/10.2307/2286745">https://doi.org/10.2307/2286745</a>.
</div>
<div id="ref-gradner1990analyzing" class="csl-entry" role="listitem">
W. Gradner. Analyzing sequential categorical data: Individual variation in <span>M</span>arkov chains. <em>Psychometrika</em>, 55: 263–275, 1990. URL <a href="https://doi.org/10.1007/BF02295287">https://doi.org/10.1007/BF02295287</a>.
</div>
<div id="ref-holsclaw2017bayesian" class="csl-entry" role="listitem">
T. Holsclaw, A. M. Greene, A. W. Robertson and P. Smyth. Bayesian nonhomogeneous markov models via p<span>ó</span>lya-gamma data augmentation with applications to rainfall modeling. <em>The Annals of Applied Statistics</em>, 11: 393–426, 2017. URL <a href="https://doi.org/10.1214/16-AOAS1009">https://doi.org/10.1214/16-AOAS1009</a>.
</div>
<div id="ref-islam2006higher" class="csl-entry" role="listitem">
M. A. Islam and R. I. Chowdhury. A higher order <span>M</span>arkov model for analyzing covariate dependence. <em>Applied Mathematical Modelling</em>, 30: 477–488, 2006. URL <a href="https://doi.org/10.1016/j.apm.2005.05.006">https://doi.org/10.1016/j.apm.2005.05.006</a>.
</div>
<div id="ref-jackson2011multi" class="csl-entry" role="listitem">
C. Jackson. Multi-state models for panel data: The msm package for <span>R</span>. <em>Journal of Statistical Software</em>, 38: 1–28, 2011. URL <a href="https://doi.org/10.18637/jss.v038.i08">https://doi.org/10.18637/jss.v038.i08</a>.
</div>
<div id="ref-juniper1999development" class="csl-entry" role="listitem">
E. Juniper, P. O’byrne, G. Guyatt, P. Ferrie and D. King. Development and validation of a questionnaire to measure asthma control. <em>European Respiratory Journal</em>, 14: 902–907, 1999. URL <a href="https://doi.org/10.1034/j.1399-3003.1999.14d29.x">https://doi.org/10.1034/j.1399-3003.1999.14d29.x</a>.
</div>
<div id="ref-katsura1980catdap" class="csl-entry" role="listitem">
K. Katsura. <span class="nocase">catdap</span>, a categorical data analysis program package. <em>Computer Science Monograph</em>, 14: 1980. URL <a href="https://CRAN.R-project.org/package=catdap">https://CRAN.R-project.org/package=catdap</a>.
</div>
<div id="ref-kharrat2019flexible" class="csl-entry" role="listitem">
T. Kharrat, G. N. Boshnakov, I. McHale and R. Baker. Flexible regression models for count data based on renewal processes: The <span>Countr</span> package. <em>Journal of Statistical Software</em>, 90: 1–35, 2019. URL <a href="https://doi.org/10.18637/jss.v090.i13">https://doi.org/10.18637/jss.v090.i13</a>.
</div>
<div id="ref-li2009markov" class="csl-entry" role="listitem">
B. Li. Markov models for <span>B</span>ayesian analysis about transit route origin–destination matrices. <em>Transportation Research Part B: Methodological</em>, 43: 301–310, 2009. URL <a href="https://doi.org/10.1016/j.trb.2008.07.001">https://doi.org/10.1016/j.trb.2008.07.001</a>.
</div>
<div id="ref-listwon2015semimarkov" class="csl-entry" role="listitem">
A. Listwon and P. Saint-Pierre. <span>SemiMarkov</span>: An <span>R</span> package for parametric estimation in multi-state semi-<span>M</span>arkov models. <em>Journal of Statistical Software</em>, 66: 1–16, 2015. URL <a href="https://doi.org/10.18637/jss.v066.i06">https://doi.org/10.18637/jss.v066.i06</a>.
</div>
<div id="ref-macdermot2005identification" class="csl-entry" role="listitem">
K. D. MacDermot, E. Bonora, N. Sykes, A.-M. Coupe, C. S. Lai, S. C. Vernes, F. Vargha-Khadem, F. McKenzie, R. L. Smith, A. P. Monaco, et al. Identification of <span>FOXP2</span> truncation as a novel cause of developmental speech and language deficits. <em>The American Journal of Human Genetics</em>, 76: 1074–1080, 2005. URL <a href="https://doi.org/10.1086/430841">https://doi.org/10.1086/430841</a>.
</div>
<div id="ref-marshall1995markov" class="csl-entry" role="listitem">
G. Marshall, W. Guo and R. H. Jones. <span>MARKOV</span>: A computer program for multi-state <span>M</span>arkov models with covariables. <em>Computer Methods and Programs in Biomedicine</em>, 47: 147–156, 1995. URL <a href="https://doi.org/10.1016/0169-2607(95)01641-6">https://doi.org/10.1016/0169-2607(95)01641-6</a>.
</div>
<div id="ref-meyer2022vcd" class="csl-entry" role="listitem">
D. Meyer, A. Zeileis and K. Hornik. <em><span class="nocase">vcd</span>: Visualizing categorical data.</em> 2022. URL <a href="https://CRAN.R-project.org/package=vcd ">https://CRAN.R-project.org/package=vcd </a>. R package version 1.4-10.
</div>
<div id="ref-miller2019fast" class="csl-entry" role="listitem">
J. W. Miller. Fast and accurate approximation of the full conditional for gamma shape parameters. <em>Journal of Computational and Graphical Statistics</em>, 28: 476–480, 2019. URL <a href="https://doi.org/10.1080/10618600.2018.1537929">https://doi.org/10.1080/10618600.2018.1537929</a>.
</div>
<div id="ref-muenz1985markov" class="csl-entry" role="listitem">
L. R. Muenz and L. V. Rubinstein. Markov models for covariate dependence of binary sequences. <em>Biometrics</em>, 41: 91–101, 1985. URL <a href="https://doi.org/10.2307/2530646">https://doi.org/10.2307/2530646</a>.
</div>
<div id="ref-muliere2003reinforced" class="csl-entry" role="listitem">
P. Muliere, P. Secchi and S. G. Walker. Reinforced random processes in continuous time. <em>Stochastic Processes and their Applications</em>, 104: 117–130, 2003. URL <a href="https://doi.org/10.1016/S0304-4149(02)00234-X">https://doi.org/10.1016/S0304-4149(02)00234-X</a>.
</div>
<div id="ref-o2011hidden" class="csl-entry" role="listitem">
J. O’Connell and S. Højsgaard. Hidden semi <span>M</span>arkov models for multiple observation sequences: The mhsmm package for <span>R</span>. <em>Journal of Statistical Software</em>, 39: 1–22, 2011. URL <a href="https://doi.org/10.18637/jss.v039.i04">https://doi.org/10.18637/jss.v039.i04</a>.
</div>
<div id="ref-phelan1990bayes" class="csl-entry" role="listitem">
M. J. Phelan. Bayes estimation from a <span>M</span>arkov renewal process. <em>The Annals of Statistics</em>, 18: 603–616, 1990. URL <a href="https://doi.org/10.1214/aos/1176347618">https://doi.org/10.1214/aos/1176347618</a>.
</div>
<div id="ref-pustejovsky2021" class="csl-entry" role="listitem">
J. E. Pustejovsky. <em>ARPobservation: Simulating recording procedures for direct observation of behavior.</em> 2021. URL <a href="https://CRAN.R-project.org/package=ARPobservation">https://CRAN.R-project.org/package=ARPobservation</a>. R package version 1.1.
</div>
<div id="ref-pyke1961markov" class="csl-entry" role="listitem">
R. Pyke. Markov renewal processes: Definitions and preliminary properties. <em>The Annals of Mathematical Statistics</em>, 32: 1231–1242, 1961. URL <a href="https://www.jstor.org/stable/2237923">https://www.jstor.org/stable/2237923</a>.
</div>
<div id="ref-saint2003analysis" class="csl-entry" role="listitem">
P. Saint-Pierre, C. Combescure, J. Daures and P. Godard. The analysis of asthma control under a <span>M</span>arkov assumption with use of covariates. <em>Statistics in Medicine</em>, 22: 3755–3770, 2003. URL <a href="https://doi.org/10.1002/sim.1680">https://doi.org/10.1002/sim.1680</a>.
</div>
<div id="ref-sarkar2018bayesian" class="csl-entry" role="listitem">
A. Sarkar, J. Chabout, J. J. Macopson, E. D. Jarvis and D. B. Dunson. Bayesian semiparametric mixed effects <span>M</span>arkov models with application to vocalization syntax. <em>Journal of the American Statistical Association</em>, 113: 1515–1527, 2018. URL <a href="https://doi.org/10.1080/01621459.2018.1423986">https://doi.org/10.1080/01621459.2018.1423986</a>.
</div>
<div id="ref-sesia2019gene" class="csl-entry" role="listitem">
M. Sesia, C. Sabatti and E. J. Candès. Gene hunting with hidden <span>M</span>arkov model knockoffs. <em>Biometrika</em>, 106: 1–18, 2019. URL <a href="https://doi.org/10.1093/biomet/asy033">https://doi.org/10.1093/biomet/asy033</a>.
</div>
<div id="ref-siebert2016bayesian" class="csl-entry" role="listitem">
M. Siebert and J. Söding. Bayesian <span>M</span>arkov models consistently outperform <span>PWMs</span> at predicting motifs in nucleotide sequences. <em>Nucleic Acids Research</em>, 44: 6055–6069, 2016. URL <a href="https://doi.org/10.1093/nar/gkw521">https://doi.org/10.1093/nar/gkw521</a>.
</div>
<div id="ref-Spedicato2017Discrete" class="csl-entry" role="listitem">
G. A. Spedicato. Discrete time <span>M</span>arkov chains with <span>R</span>. <em><span>The R Journal</span></em>, 9: 84–104, 2017. URL <a href="https://doi.org/10.32614/RJ-2017-036">https://doi.org/10.32614/RJ-2017-036</a>.
</div>
<div id="ref-watanabe2010asymptotic" class="csl-entry" role="listitem">
S. Watanabe and M. Opper. Asymptotic equivalence of <span>B</span>ayes cross validation and widely applicable information criterion in singular learning theory. <em>Journal of Machine Learning Research</em>, 11: 3571–3594, 2010. URL <a href="https://doi.org/10.48550/arXiv.1004.2316">https://doi.org/10.48550/arXiv.1004.2316</a>.
</div>
<div id="ref-wu2021bayesian" class="csl-entry" role="listitem">
Y. Wu, E. D. Jarvis and A. Sarkar. Bayesian semiparametric <span>M</span>arkov renewal mixed models for vocalization syntax. <em>Biostatistics</em>, 2023. URL <a href="https://doi.org/10.1093/biostatistics/kxac050">https://doi.org/10.1093/biostatistics/kxac050</a>. To appear.
</div>
<div id="ref-zhang2019scenario" class="csl-entry" role="listitem">
M. Zhang, P. W. van Rijn, P. Deane and R. E. Bennett. Scenario-based assessments in writing: An experimental study. <em>Educational Assessment</em>, 24: 73–90, 2019. URL <a href="https://doi.org/10.1080/10627197.2018.1557515">https://doi.org/10.1080/10627197.2018.1557515</a>.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
  <h3 id="references">References</h3>
  <div id="references-listing"></div>
  <h3 id="reuse">Reuse</h3>
  <p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
  <h3 id="citation">Citation</h3>
  <p>For attribution, please cite this work as</p>
  <pre class="citation-appendix short">Wu &amp; Sarkar, "BMRMM: An R Package for Bayesian Markov (Renewal) Mixed Models", The R Journal, 2025</pre>
  <p>BibTeX citation</p>
  <pre class="citation-appendix long">@article{RJ-2024-011,
  author = {Wu, Yutong and Sarkar, Abhra},
  title = {BMRMM: An R Package for Bayesian Markov (Renewal) Mixed Models},
  journal = {The R Journal},
  year = {2025},
  note = {https://doi.org/10.32614/RJ-2024-011},
  doi = {10.32614/RJ-2024-011},
  volume = {16},
  issue = {1},
  issn = {2073-4859},
  pages = {192-211}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<!--radix_placeholder_navigation_after_body--><html><body>
<div class="distill-site-nav distill-site-footer">
<p>© The R Foundation, <a href="mailto:r-journal@r-project.org">web page
contact</a>.</p>
</div>
<!--/radix_placeholder_navigation_after_body-->
</body></html>


</body>

</html>
