---
title: News from the Bioconductor Project
abstract: The 'News from the Bioconductor Project' article from the 2013-1 issue.
author:
- name: Bioconductor Team
  affiliation: Program in Computational Biology, Fred Hutchinson Cancer Research Center
date: '2013-06-01'
date_received: ~
journal:
  firstpage: '218'
  lastpage: '219'
volume: 5
issue: 1
slug: RJ-2013-1-bioconductor
citation_url: https://rjournal.github.io/
packages:
  cran: shiny
  bioc: []
preview: preview.png
bibliography: ~
CTV: ~
output:
  rjtools::rjournal_web_article:
    self_contained: yes
    toc: no
    legacy_pdf: yes
    web_only: yes
    mathjax: https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js
    md_extension: -tex_math_single_backslash

---

::: article
Bioconductor 2.12 was released on 3 October 2012. It is compatible with
R 3.0.1, and consists of 671 software packages and more than 675
up-to-date annotation packages. The release includes 65 new software
packages, and enhancements to many others. Descriptions of new packages
and updated NEWS files provided by current package maintainers are at
<http://bioconductor.org/news/bioc_2_12_release/>.

Start using Bioconductor and R version 3.0.1 with

``` r
> source("http://bioconductor.org/biocLite.R")
> biocLite()
```

Install additional packages, e.g., *VariantTools*, with

``` r
> source("http://bioconductor.org/biocLite.R")
> biocLite("VariantTools")
```

Upgrade installed packages with

``` r
> source("http://bioconductor.org/biocLite.R")
> biocLite()
```

New this release is `biocValid()`, a companion function to detect R /
Bioconductor version mis-matches.

Explore available Bioconductor packages at
<http://bioconductor.org/packages/release/>. All packages are grouped by
'BiocViews' to identify coherent groups of packages. Each package has an
html page with the descriptions and links to vignettes, reference
manuals, and use statistics.

A Bioconductor Amazon Machine Instance is available and updated; see
<http://bioconductor.org/help/bioconductor-cloud-ami>.

# Core annotation and software packages

This release includes *AnnotationHub*, a resource that enables ready
access to large genome-scale resources (e.g., GTF or FASTA files from
Ensembl; ENCODE tracks from UCSC) in formats (e.g., `GRanges` or `VCF`
instances) that allow smooth integration with R work flows. The
*AnnotationHub* resource can be queried through simple tab completion,
or via metadata about resource provenance. Additional new annotation
resources include *ensemblVEP* to query the Ensembl Variant Effect
Predictor, and *KEGGREST* and *UniProt.ws* packages for on-line
integration of data from corresponding resources. Our large collection
of microarray- and organism-specific annotation packages have been
updated to include current information.

*GenomicRanges* and related packages, e.g., *VariantAnnotation*,
*IRanges*, *Biostrings*, *Rsamtools*, *GenomicFeatures* provide an
extensive, mature and extensible framework for interacting with high
throughput sequence data, either as a user or package developer. Many
contributed packages rely on this infrastructure for interoperable,
re-usable analysis.

# Other activities

Bioconductor's Annual Meeting is in Seattle, 17-19 July 2013, see
<http://bioconductor.org/bioc2013>; our European developer community
meets in December, with final arrangements pending. Additional training
and community activities advertised at
<http://bioconductor.org/help/events/>. The active Bioconductor mailing
lists (<http://bioconductor.org/help/mailing-list/>) connect users with
each other, to domain experts, and to maintainers eager to ensure that
their packages satisfy the needs of leading edge approaches. Keep
abreast of packages added to the 'devel' branch and other activities by
following Bioconductor on Twitter.

Package developers will be interested in ongoing efforts planned for our
next release. Activities include better integration of parallel
evaluation, graphical interfaces to Bioconductor objects (e.g., via
[*shiny*](https://CRAN.R-project.org/package=shiny), elaboration of
*AnnotationHub* to support user-contributed and locally curated data, a
repository for workflow and other packages that change infrequently but
require significant computational resources to build, and efforts to
ease integrate with github and other social coding resources. Our Google
Summer of Code participants are enabling progress on some of these
topics.
:::
