<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />
  <link rel="alternate" type="application/rss+xml" href="https://rjournal-distill.netlify.app/articles.xml" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007ba5; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #007ba5; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #007ba5; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>


  <!--radix_placeholder_meta_tags-->
  <title>The R Journal: All articles</title>

  <meta property="description" itemprop="description" content="All articles accepted into the R Journal."/>

  <link rel="icon" type="image/vnd.microsoft.icon" href="resources/favicon.ico"/>


  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="The R Journal: All articles"/>
  <meta property="og:type" content="article"/>
  <meta property="og:description" content="All articles accepted into the R Journal."/>
  <meta property="og:locale" content="en_US"/>
  <meta property="og:site_name" content="The R Journal"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="The R Journal: All articles"/>
  <meta property="twitter:description" content="All articles accepted into the R Journal."/>

  <!--/radix_placeholder_meta_tags-->
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","listing"]}},"value":[{"type":"character","attributes":{},"value":["All articles"]},{"type":"character","attributes":{},"value":["All articles accepted into the R Journal.\n"]},{"type":"character","attributes":{},"value":["articles"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  <!--radix_placeholder_navigation_in_header-->
  <meta name="distill:offset" content=""/>

  <script type="application/javascript">

    window.headroom_prevent_pin = false;

    window.document.addEventListener("DOMContentLoaded", function (event) {

      // initialize headroom for banner
      var header = $('header').get(0);
      var headerHeight = header.offsetHeight;
      var headroom = new Headroom(header, {
        tolerance: 5,
        onPin : function() {
          if (window.headroom_prevent_pin) {
            window.headroom_prevent_pin = false;
            headroom.unpin();
          }
        }
      });
      headroom.init();
      if(window.location.hash)
        headroom.unpin();
      $(header).addClass('headroom--transition');

      // offset scroll location for banner on hash change
      // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
      window.addEventListener("hashchange", function(event) {
        window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
      });

      // responsive menu
      $('.distill-site-header').each(function(i, val) {
        var topnav = $(this);
        var toggle = topnav.find('.nav-toggle');
        toggle.on('click', function() {
          topnav.toggleClass('responsive');
        });
      });

      // nav dropdowns
      $('.nav-dropbtn').click(function(e) {
        $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
        $(this).parent().siblings('.nav-dropdown')
           .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $("body").click(function(e){
        $('.nav-dropdown-content').removeClass('nav-dropdown-active');
      });
      $(".nav-dropdown").click(function(e){
        e.stopPropagation();
      });
    });
  </script>

  <style type="text/css">

  /* Theme (user-documented overrideables for nav appearance) */

  .distill-site-nav {
    color: rgba(255, 255, 255, 0.8);
    background-color: #0F2E3D;
    font-size: 15px;
    font-weight: 300;
  }

  .distill-site-nav a {
    color: inherit;
    text-decoration: none;
  }

  .distill-site-nav a:hover {
    color: white;
  }

  @media print {
    .distill-site-nav {
      display: none;
    }
  }

  .distill-site-header {

  }

  .distill-site-footer {

  }


  /* Site Header */

  .distill-site-header {
    width: 100%;
    box-sizing: border-box;
    z-index: 3;
  }

  .distill-site-header .nav-left {
    display: inline-block;
    margin-left: 8px;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header .nav-left {
      margin-left: 0;
    }
  }


  .distill-site-header .nav-right {
    float: right;
    margin-right: 8px;
  }

  .distill-site-header a,
  .distill-site-header .title {
    display: inline-block;
    text-align: center;
    padding: 14px 10px 14px 10px;
  }

  .distill-site-header .title {
    font-size: 18px;
    min-width: 150px;
  }

  .distill-site-header .logo {
    padding: 0;
  }

  .distill-site-header .logo img {
    display: none;
    max-height: 20px;
    width: auto;
    margin-bottom: -4px;
  }

  .distill-site-header .nav-image img {
    max-height: 18px;
    width: auto;
    display: inline-block;
    margin-bottom: -3px;
  }



  @media screen and (min-width: 1000px) {
    .distill-site-header .logo img {
      display: inline-block;
    }
    .distill-site-header .nav-left {
      margin-left: 20px;
    }
    .distill-site-header .nav-right {
      margin-right: 20px;
    }
    .distill-site-header .title {
      padding-left: 12px;
    }
  }


  .distill-site-header .nav-toggle {
    display: none;
  }

  .nav-dropdown {
    display: inline-block;
    position: relative;
  }

  .nav-dropdown .nav-dropbtn {
    border: none;
    outline: none;
    color: rgba(255, 255, 255, 0.8);
    padding: 16px 10px;
    background-color: transparent;
    font-family: inherit;
    font-size: inherit;
    font-weight: inherit;
    margin: 0;
    margin-top: 1px;
    z-index: 2;
  }

  .nav-dropdown-content {
    display: none;
    position: absolute;
    background-color: white;
    min-width: 200px;
    border: 1px solid rgba(0,0,0,0.15);
    border-radius: 4px;
    box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
    z-index: 1;
    margin-top: 2px;
    white-space: nowrap;
    padding-top: 4px;
    padding-bottom: 4px;
  }

  .nav-dropdown-content hr {
    margin-top: 4px;
    margin-bottom: 4px;
    border: none;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .nav-dropdown-active {
    display: block;
  }

  .nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
    color: black;
    padding: 6px 24px;
    text-decoration: none;
    display: block;
    text-align: left;
  }

  .nav-dropdown-content .nav-dropdown-header {
    display: block;
    padding: 5px 24px;
    padding-bottom: 0;
    text-transform: uppercase;
    font-size: 14px;
    color: #999999;
    white-space: nowrap;
  }

  .nav-dropdown:hover .nav-dropbtn {
    color: white;
  }

  .nav-dropdown-content a:hover {
    background-color: #ddd;
    color: black;
  }

  .nav-right .nav-dropdown-content {
    margin-left: -45%;
    right: 0;
  }

  @media screen and (max-width: 768px) {
    .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
    .distill-site-header a.nav-toggle {
      float: right;
      display: block;
    }
    .distill-site-header .title {
      margin-left: 0;
    }
    .distill-site-header .nav-right {
      margin-right: 0;
    }
    .distill-site-header {
      overflow: hidden;
    }
    .nav-right .nav-dropdown-content {
      margin-left: 0;
    }
  }


  @media screen and (max-width: 768px) {
    .distill-site-header.responsive {position: relative; min-height: 500px; }
    .distill-site-header.responsive a.nav-toggle {
      position: absolute;
      right: 0;
      top: 0;
    }
    .distill-site-header.responsive a,
    .distill-site-header.responsive .nav-dropdown {
      display: block;
      text-align: left;
    }
    .distill-site-header.responsive .nav-left,
    .distill-site-header.responsive .nav-right {
      width: 100%;
    }
    .distill-site-header.responsive .nav-dropdown {float: none;}
    .distill-site-header.responsive .nav-dropdown-content {position: relative;}
    .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
      display: block;
      width: 100%;
      text-align: left;
    }
  }

  /* Site Footer */

  .distill-site-footer {
    width: 100%;
    overflow: hidden;
    box-sizing: border-box;
    z-index: 3;
    margin-top: 30px;
    padding-top: 30px;
    padding-bottom: 30px;
    text-align: center;
  }

  /* Headroom */

  d-title {
    padding-top: 6rem;
  }

  @media print {
    d-title {
      padding-top: 4rem;
    }
  }

  .headroom {
    z-index: 1000;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
  }

  .headroom--transition {
    transition: all .4s ease-in-out;
  }

  .headroom--unpinned {
    top: -100px;
  }

  .headroom--pinned {
    top: 0;
  }

  /* adjust viewport for navbar height */
  /* helps vertically center bootstrap (non-distill) content */
  .min-vh-100 {
    min-height: calc(100vh - 100px) !important;
  }

  </style>

  <script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet"/>
  <link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet"/>
  <script src="site_libs/headroom-0.9.4/headroom.min.js"></script>
  <script src="site_libs/autocomplete-0.37.1/autocomplete.min.js"></script>
  <script src="site_libs/fuse-6.4.1/fuse.min.js"></script>

  <script type="application/javascript">

  function getMeta(metaName) {
    var metas = document.getElementsByTagName('meta');
    for (let i = 0; i < metas.length; i++) {
      if (metas[i].getAttribute('name') === metaName) {
        return metas[i].getAttribute('content');
      }
    }
    return '';
  }

  function offsetURL(url) {
    var offset = getMeta('distill:offset');
    return offset ? offset + '/' + url : url;
  }

  function createFuseIndex() {

    // create fuse index
    var options = {
      keys: [
        { name: 'title', weight: 20 },
        { name: 'categories', weight: 15 },
        { name: 'description', weight: 10 },
        { name: 'contents', weight: 5 },
      ],
      ignoreLocation: true,
      threshold: 0
    };
    var fuse = new window.Fuse([], options);

    // fetch the main search.json
    return fetch(offsetURL('search.json'))
      .then(function(response) {
        if (response.status == 200) {
          return response.json().then(function(json) {
            // index main articles
            json.articles.forEach(function(article) {
              fuse.add(article);
            });
            // download collections and index their articles
            return Promise.all(json.collections.map(function(collection) {
              return fetch(offsetURL(collection)).then(function(response) {
                if (response.status === 200) {
                  return response.json().then(function(articles) {
                    articles.forEach(function(article) {
                      fuse.add(article);
                    });
                  })
                } else {
                  return Promise.reject(
                    new Error('Unexpected status from search index request: ' +
                              response.status)
                  );
                }
              });
            })).then(function() {
              return fuse;
            });
          });

        } else {
          return Promise.reject(
            new Error('Unexpected status from search index request: ' +
                        response.status)
          );
        }
      });
  }

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // get search element (bail if we don't have one)
    var searchEl = window.document.getElementById('distill-search');
    if (!searchEl)
      return;

    createFuseIndex()
      .then(function(fuse) {

        // make search box visible
        searchEl.classList.remove('hidden');

        // initialize autocomplete
        var options = {
          autoselect: true,
          hint: false,
          minLength: 2,
        };
        window.autocomplete(searchEl, options, [{
          source: function(query, callback) {
            const searchOptions = {
              isCaseSensitive: false,
              shouldSort: true,
              minMatchCharLength: 2,
              limit: 10,
            };
            var results = fuse.search(query, searchOptions);
            callback(results
              .map(function(result) { return result.item; })
            );
          },
          templates: {
            suggestion: function(suggestion) {
              var img = suggestion.preview && Object.keys(suggestion.preview).length > 0
                ? `<img src="${offsetURL(suggestion.preview)}"</img>`
                : '';
              var html = `
                <div class="search-item">
                  <h3>${suggestion.title}</h3>
                  <div class="search-item-description">
                    ${suggestion.description || ''}
                  </div>
                  <div class="search-item-preview">
                    ${img}
                  </div>
                </div>
              `;
              return html;
            }
          }
        }]).on('autocomplete:selected', function(event, suggestion) {
          window.location.href = offsetURL(suggestion.path);
        });
        // remove inline display style on autocompleter (we want to
        // manage responsive display via css)
        $('.algolia-autocomplete').css("display", "");
      })
      .catch(function(error) {
        console.log(error);
      });

  });

  </script>

  <style type="text/css">

  .nav-search {
    font-size: x-small;
  }

  /* Algolioa Autocomplete */

  .algolia-autocomplete {
    display: inline-block;
    margin-left: 10px;
    vertical-align: sub;
    background-color: white;
    color: black;
    padding: 6px;
    padding-top: 8px;
    padding-bottom: 0;
    border-radius: 6px;
    border: 1px #0F2E3D solid;
    width: 180px;
  }


  @media screen and (max-width: 768px) {
    .distill-site-nav .algolia-autocomplete {
      display: none;
      visibility: hidden;
    }
    .distill-site-nav.responsive .algolia-autocomplete {
      display: inline-block;
      visibility: visible;
    }
    .distill-site-nav.responsive .algolia-autocomplete .aa-dropdown-menu {
      margin-left: 0;
      width: 400px;
      max-height: 400px;
    }
  }

  .algolia-autocomplete .aa-input, .algolia-autocomplete .aa-hint {
    width: 90%;
    outline: none;
    border: none;
  }

  .algolia-autocomplete .aa-hint {
    color: #999;
  }
  .algolia-autocomplete .aa-dropdown-menu {
    width: 550px;
    max-height: 70vh;
    overflow-x: visible;
    overflow-y: scroll;
    padding: 5px;
    margin-top: 3px;
    margin-left: -150px;
    background-color: #fff;
    border-radius: 5px;
    border: 1px solid #999;
    border-top: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion {
    cursor: pointer;
    padding: 5px 4px;
    border-bottom: 1px solid #eee;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion:last-of-type {
    border-bottom: none;
    margin-bottom: 2px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item {
    overflow: hidden;
    font-size: 0.8em;
    line-height: 1.4em;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item h3 {
    font-size: 1rem;
    margin-block-start: 0;
    margin-block-end: 5px;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-description {
    display: inline-block;
    overflow: hidden;
    height: 2.8em;
    width: 80%;
    margin-right: 4%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview {
    display: inline-block;
    width: 15%;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img {
    height: 3em;
    width: auto;
    display: none;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion .search-item-preview img[src] {
    display: initial;
  }

  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion.aa-cursor {
    background-color: #eee;
  }
  .algolia-autocomplete .aa-dropdown-menu .aa-suggestion em {
    font-weight: bold;
    font-style: normal;
  }

  </style>


  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure img {
    width: 100%;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme
    $('code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        var refs = $(this).attr('data-cites').split(" ");
        var refHtml = refs.map(function(ref) {
          return "<p>" + $('#ref-' + ref).html() + "</p>";
        }).join("\n");
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace("index.html", "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <style type="text/css">
  /* base variables */

  /* Edit the CSS properties in this file to create a custom
     Distill theme. Only edit values in the right column
     for each row; values shown are the CSS defaults.
     To return any property to the default,
     you may set its value to: unset
     All rows must end with a semi-colon.                      */

  /* Optional: embed custom fonts here with `@import`          */
  /* This must remain at the top of this file.                 */



  html {
    /*-- Main font sizes --*/
    --title-size:      50px;
    --body-size:       1.06rem;
    --code-size:       14px;
    --aside-size:      12px;
    --fig-cap-size:    13px;
    /*-- Main font colors --*/
    --title-color:     #000000;
    --header-color:    rgba(0, 0, 0, 0.8);
    --body-color:      rgba(0, 0, 0, 0.8);
    --aside-color:     rgba(0, 0, 0, 0.6);
    --fig-cap-color:   rgba(0, 0, 0, 0.6);
    /*-- Specify custom fonts ~~~ must be imported above   --*/
    --heading-font:    sans-serif;
    --mono-font:       monospace;
    --body-font:       sans-serif;
    --navbar-font:     sans-serif;  /* websites + blogs only */
  }

  /*-- ARTICLE METADATA --*/
  d-byline {
    --heading-size:    0.6rem;
    --heading-color:   rgba(0, 0, 0, 0.5);
    --body-size:       0.8rem;
    --body-color:      rgba(0, 0, 0, 0.8);
  }

  /*-- ARTICLE TABLE OF CONTENTS --*/
  .d-contents {
    --heading-size:    18px;
    --contents-size:   13px;
  }

  /*-- ARTICLE APPENDIX --*/
  d-appendix {
    --heading-size:    15px;
    --heading-color:   rgba(0, 0, 0, 0.65);
    --text-size:       0.8em;
    --text-color:      rgba(0, 0, 0, 0.5);
  }

  /*-- WEBSITE HEADER + FOOTER --*/
  /* These properties only apply to Distill sites and blogs  */

  .distill-site-header {
    --title-size:       18px;
    --text-color:       rgba(255, 255, 255, 0.8);
    --text-size:        15px;
    --hover-color:      white;
    --bkgd-color:       #0F2E3D;
  }

  .distill-site-footer {
    --text-color:       rgba(255, 255, 255, 0.8);
    --text-size:        15px;
    --hover-color:      white;
    --bkgd-color:       #0F2E3D;
  }

  /*-- Additional custom styles --*/
  /* Add any additional CSS rules below                      */
  </style>
  <style type="text/css">
  /* base variables */

  /* Edit the CSS properties in this file to create a custom
     Distill theme. Only edit values in the right column
     for each row; values shown are the CSS defaults.
     To return any property to the default,
     you may set its value to: unset
     All rows must end with a semi-colon.                      */

  /* Optional: embed custom fonts here with `@import`          */
  /* This must remain at the top of this file.                 */



  html {
    /*-- Main font sizes --*/
    --title-size:      50px;
    --body-size:       1.06rem;
    --code-size:       14px;
    --aside-size:      12px;
    --fig-cap-size:    13px;
    /*-- Main font colors --*/
    --title-color:     #000000;
    --header-color:    rgba(0, 0, 0, 0.8);
    --body-color:      rgba(0, 0, 0, 0.8);
    --aside-color:     rgba(0, 0, 0, 0.6);
    --fig-cap-color:   rgba(0, 0, 0, 0.6);
    /*-- Specify custom fonts ~~~ must be imported above   --*/
    --heading-font:    sans-serif;
    --mono-font:       monospace;
    --body-font:       sans-serif;
    --navbar-font:     sans-serif;  /* websites + blogs only */
  }

  /*-- ARTICLE METADATA --*/
  d-byline {
    --heading-size:    0.6rem;
    --heading-color:   rgba(0, 0, 0, 0.5);
    --body-size:       0.8rem;
    --body-color:      rgba(0, 0, 0, 0.8);
  }

  /*-- ARTICLE TABLE OF CONTENTS --*/
  .d-contents {
    --heading-size:    18px;
    --contents-size:   13px;
  }

  /*-- ARTICLE APPENDIX --*/
  d-appendix {
    --heading-size:    15px;
    --heading-color:   rgba(0, 0, 0, 0.65);
    --text-size:       0.8em;
    --text-color:      rgba(0, 0, 0, 0.7);
  }

  /*-- WEBSITE HEADER + FOOTER --*/
  /* These properties only apply to Distill sites and blogs  */

  .distill-site-header {
    --title-size:       18px;
    --text-color:       rgba(0, 0, 0, 0.8);
    --text-size:        15px;
    --hover-color:      black;
    --bkgd-color:       #ffffff;
  }

  .distill-site-footer {
    --text-color:       rgba(0, 0, 0, 0.8);
    --text-size:        15px;
    --hover-color:      black;
    --bkgd-color:       #fafafa;
  }

  /*-- Additional custom styles --*/
  /* Add any additional CSS rules below                      */

  .nav-right > a {
    text-transform: uppercase;
  }

  d-title h1 {
    grid-column: page;
  }

  .rj-blue {
    color: #2467bb;
  }

  ul li {
    line-height: 1.6;
    margin-top: 0em;
    margin-bottom: 0em;
  }</style>
  <style type="text/css">
  /* base style */

  /* FONT FAMILIES */

  :root {
    --heading-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    --mono-default: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    --body-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  body,
  .posts-list .post-preview p,
  .posts-list .description p {
    font-family: var(--body-font), var(--body-default);
  }

  h1, h2, h3, h4, h5, h6,
  .posts-list .post-preview h2,
  .posts-list .description h2 {
    font-family: var(--heading-font), var(--heading-default);
  }

  d-article div.sourceCode code,
  d-article pre code {
    font-family: var(--mono-font), var(--mono-default);
  }


  /*-- TITLE --*/
  d-title h1,
  .posts-list > h1 {
    color: var(--title-color, black);
  }

  d-title h1 {
    font-size: var(--title-size, 50px);
  }

  /*-- HEADERS --*/
  d-article h1,
  d-article h2,
  d-article h3,
  d-article h4,
  d-article h5,
  d-article h6 {
    color: var(--header-color, rgba(0, 0, 0, 0.8));
  }

  /*-- BODY --*/
  d-article > p,  /* only text inside of <p> tags */
  d-article > ul, /* lists */
  d-article > ol {
    color: var(--body-color, rgba(0, 0, 0, 0.8));
    font-size: var(--body-size, 1.06rem);
  }


  /*-- CODE --*/
  d-article div.sourceCode code,
  d-article pre code {
    font-size: var(--code-size, 14px);
  }

  /*-- ASIDE --*/
  d-article aside {
    font-size: var(--aside-size, 12px);
    color: var(--aside-color, rgba(0, 0, 0, 0.6));
  }

  /*-- FIGURE CAPTIONS --*/
  figure .caption,
  figure figcaption,
  .figure .caption {
    font-size: var(--fig-cap-size, 13px);
    color: var(--fig-cap-color, rgba(0, 0, 0, 0.6));
  }

  /*-- METADATA --*/
  d-byline h3 {
    font-size: var(--heading-size, 0.6rem);
    color: var(--heading-color, rgba(0, 0, 0, 0.5));
  }

  d-byline {
    font-size: var(--body-size, 0.8rem);
    color: var(--body-color, rgba(0, 0, 0, 0.8));
  }

  d-byline a,
  d-article d-byline a {
    color: var(--body-color, rgba(0, 0, 0, 0.8));
  }

  /*-- TABLE OF CONTENTS --*/
  .d-contents nav h3 {
    font-size: var(--heading-size, 18px);
  }

  .d-contents nav a {
    font-size: var(--contents-size, 13px);
  }

  /*-- APPENDIX --*/
  d-appendix h3 {
    font-size: var(--heading-size, 15px);
    color: var(--heading-color, rgba(0, 0, 0, 0.65));
  }

  d-appendix {
    font-size: var(--text-size, 0.8em);
    color: var(--text-color, rgba(0, 0, 0, 0.5));
  }

  d-appendix d-footnote-list a.footnote-backlink {
    color: var(--text-color, rgba(0, 0, 0, 0.5));
  }

  /*-- WEBSITE HEADER + FOOTER --*/
  .distill-site-header .title {
    font-size: var(--title-size, 18px);
    font-family: var(--navbar-font), var(--heading-default);
  }

  .distill-site-header a,
  .nav-dropdown .nav-dropbtn {
    font-family: var(--navbar-font), var(--heading-default);
  }

  .nav-dropdown .nav-dropbtn {
    color: var(--text-color, rgba(255, 255, 255, 0.8));
    font-size: var(--text-size, 15px);
  }

  .distill-site-header a:hover,
  .nav-dropdown:hover .nav-dropbtn {
    color: var(--hover-color, white);
  }

  .distill-site-header {
    font-size: var(--text-size, 15px);
    color: var(--text-color, rgba(255, 255, 255, 0.8));
    background-color: var(--bkgd-color, #0F2E3D);
  }

  .distill-site-footer {
    font-size: var(--text-size, 15px);
    color: var(--text-color, rgba(255, 255, 255, 0.8));
    background-color: var(--bkgd-color, #0F2E3D);
  }

  .distill-site-footer a:hover {
    color: var(--hover-color, white);
  }</style>
  <!--/radix_placeholder_distill-->
  <script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
  <script src="site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->


</head>

<body class="layout-listing">

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"All articles","description":"All articles accepted into the R Journal.","authors":[]}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a class="logo" href="index.html">
<img src="resources/rlogo.png" alt="Logo"/>
</a>
<a href="index.html" class="title">The R Journal</a>
<input id="distill-search" class="nav-search hidden" type="text" placeholder="Search..."/>
</div>
<div class="nav-right">
<a href="index.html">Home</a>
<a href="articles.html">Articles</a>
<a href="issues.html">Issues</a>
<a href="submissions.html">Submit</a>
<a href="articles.xml">
<i class="fa fa-rss" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->
<!--radix_placeholder_article_listing-->


<script type="application/javascript">

function init_posts_list() {

  function load_image(img) {
    var src = $(img).attr('data-src');
    if (src) {
      $(img).attr('src', src);
      $(img).on("load", function() {
        img.removeAttribute('data-src');
      });
    }
  }

  function set_posts_visible(posts, visible) {
    if (visible) {

      // show bottom border by default
      $(posts).removeClass('post-preview-last');

      // apply limits if need be
      var max_posts = 25;
      var apply_limits = $('.posts-container').hasClass('posts-apply-limit');
      if (apply_limits && posts.length > max_posts) {
        posts = $(posts).slice(0, max_posts);
      } else {
        $('.posts-more a').addClass('hidden');
      }

      // apply last style
      $(posts.slice(-1)[0]).addClass('post-preview-last');

      $(posts).removeClass('hidden');
      $(posts).find('img[data-src]').each(function(i, img) {
        load_image(img);
      });
    } else {
      $(posts).addClass('hidden');
    }
  }

  function apply_hash_filter() {

    // clear active state
    $('.categories .active').removeClass('active');

    // mark all posts invisible to start
    set_posts_visible($('.posts-list').children('a'), false);

    // if we have a hash filter
    if (window.location.hash && window.location.hash.startsWith("#category:")) {

      // mark posts that match the category visible
      var page_category = window.location.hash.replace(/^#category:/, "");
      page_category = decodeURIComponent(page_category)
      var posts = $('.post-metadata').map(function(idx, script) {
        var metadata = $.parseJSON($(script).html());
        var post = null;
        $.each(metadata.categories, function(idx, category) {
          category = category.replace(/ /g,"_");
          if ((page_category || '').toLowerCase() === "articles" || category === page_category) {
            post = $(script).parent().get();
            return false;
          }
        });
        return post;
      });
      set_posts_visible(posts, true);

      // mark the hash active
      $('.categories li>a[href="' + decodeURIComponent(window.location.hash) + '"]').addClass('active');

      // update the list_caption
      var list_caption = $('.posts-list-caption');
      var caption = (page_category || '').toLowerCase() === "articles"
        ? list_caption.attr('data-caption')
        : ('Category: ' + page_category.replace(/_/g," "));
      list_caption.text(caption);

    } else {

      // no hash filter, make all posts visible (subject to max display)
      set_posts_visible($('.posts-list').children(), true);

      // reset list caption
      var list_caption = $('.posts-list-caption');
      list_caption.text(list_caption.attr('data-caption'));


    }
  }

  // more articles
  function apply_post_limits(apply) {
    if (apply) {
      $('.posts-container').addClass('posts-apply-limit');
      $('.posts-more a').removeClass('hidden');
    } else {
      $('.posts-container').removeClass('posts-apply-limit');
      $('.posts-more a').addClass('hidden');
    }
  }

  // click handling for tags
  $('.dt-tag').click(function(ev) {
    window.location.hash = '#category:' + $(this).text().replace(/ /g, "_");
    return false;
  })

  // hash filter handling
  apply_hash_filter();
  $(window).on('hashchange',function() {
    apply_post_limits(true);
    apply_hash_filter();
  });

  // more articles link
  $('.posts-more a').click(function(e) {
    e.preventDefault();
    apply_post_limits(false);
    apply_hash_filter();
    return false;
  });

}

</script>



<div class="posts-container posts-apply-limit l-page">
<div class="posts-list">
<h1 class="posts-list-caption" data-caption="All articles">All articles</h1>
<a href="articles/RJ-2021-112/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Dec. 12, 2021</div>
<div class="dt-authors">
<div class="dt-author">Melina Vidoni</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>Rejoinder: Software Engineering and R Programming</h2>
<div class="dt-tags"></div>
<p>It is a pleasure to take part in such fruitful discussion about the relationship between Software Engineering and R programming, and what could be gain by allowing each to look more closely at the other. Several discussants make valuable arguments that ought to be further discussed.</p>
</div>
</a>
<a href="articles/RJ-2021-111/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Nov. 24, 2021</div>
<div class="dt-authors">
<div class="dt-author">Simon Urbanek</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>The R Quest: from Users to Developers</h2>
<div class="dt-tags"></div>
<p>R is not a programming language, and this produces the inherent dichotomy between analytics and software engineering. With the emergence of data science, the opportunity exists to bridge this gap, especially through teaching practices.</p>
</div>
</a>
<a href="articles/RJ-2021-110/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 24, 2021</div>
<div class="dt-authors">
<div class="dt-author">Malle Salmon</div>
<div class="dt-author">Karthik Ram</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>The R Developer Community Does Have a Strong Software Engineering Culture</h2>
<div class="dt-tags"></div>
<p>There is a strong software engineering culture in the R developer community. We recommend creating, updating and vetting packages as well as keeping up with community standards. We invite contributions to the rOpenSci project, where participants can gain experience that will shape their work and that of their peers.</p>
</div>
</a>
<a href="articles/RJ-2021-109/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Aug. 24, 2021</div>
<div class="dt-authors">
<div class="dt-author">William Michael Landau</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>We Need Trustworthy R Packages</h2>
<div class="dt-tags"></div>
<p>There is a need for rigorous software engineering in R packages, and there is a need for new research to bridge scientific computing with more traditional computing. Automated tools, interdisciplinary graduate courses, code reviews, and a welcoming developer community will continue to democratize best practices. Democratized software engineering will improve the quality, correctness, and integrity of scientific software, and by extension, the disciplines that rely on it.</p>
</div>
</a>
<a href="articles/RJ-2021-107/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">July 12, 2021</div>
<div class="dt-authors">
<div class="dt-author">Gamze Gven</div>
<div class="dt-author">Skr Actas</div>
<div class="dt-author">Hatice Samkar</div>
<div class="dt-author">Birdal Senoglu</div>
</div>
</div>
<div class="thumbnail"></div>
<div class="description">
<h2>RobustBF: An R Package for Robust Solution to the Behrens-Fisher Problem</h2>
<div class="dt-tags"></div>
<p>Welchs two-sample t-test based on least squares (LS) estimators is generally used to test the equality of two normal means when the variances are not equal. However, this test loses its power when the underlying distribution is not normal. In this paper, two different tests are proposed to test the equality of two long-tailed symmetric (LTS) means under heterogeneous variances. Adaptive modified maximum likelihood (AMML) estimators are used in developing the proposed tests since they are highly efficient under LTS distribution. An R package called RobustBF is given to show the implementation of these tests. Simulated Type I error rates and powers of the proposed tests are also given and compared with Welchs t-test based on LS estimators via an extensive Monte Carlo simulation study.</p>
</div>
</a>
<a href="articles/RJ-2021-050/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">June 8, 2021</div>
<div class="dt-authors">
<div class="dt-author">Earo Wang</div>
<div class="dt-author">Dianne Cook</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-050/figure/highlight-retail-1.png"/>
</div>
<div class="description">
<h2>Conversations in Time: Interactive Visualization to Explore Structured Temporal Data</h2>
<div class="dt-tags"></div>
<p>Temporal data often has a hierarchical structure, defined by categorical variables describing different levels, such as political regions or sales products. The nesting of categorical variables produces a hierarchical structure. The \CRANpkg{tsibbletalk} package is developed to allow a user to interactively explore temporal data, relative to the nested or crossed structures. It can help to discover differences between category levels, and uncover interesting periodic or aperiodic slices. The package implements a shared `tsibble` object that allows for linked brushing between coordinated views, and a shiny module that aids in wrapping timelines for seasonal patterns. The tools are demonstrated using two data examples: domestic tourism in Australia and pedestrian traffic in Melbourne.</p>
</div>
</a>
<a href="articles/RJ-2021-108/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">May 24, 2021</div>
<div class="dt-authors">
<div class="dt-author">Melina Vidoni</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>Software Engineering and R Programming: A Call for Research</h2>
<div class="dt-tags"></div>
<p>Although R programming has been a part of research since its origins in the 1990s, few studies address scientific software development from a Software Engineering (SE) perspective. The past few years have seen unparalleled growth in the R community, and it is time to push the boundaries of SE research and R programming forwards. This paper discusses relevant studies that close this gap Additionally, it proposes a set of good practices derived from those findings aiming to act as a call-to-arms for both the R and RSE (Research SE) community to explore specific, interdisciplinary paths of research.</p>
</div>
</a>
<a href="articles/RJ-2021-106/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">May 24, 2021</div>
<div class="dt-authors">
<div class="dt-author">Anthony-Alexander Christidis</div>
<div class="dt-author">R. Douglas Martin</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>RPESE: Risk and Performance Estimators Standard Errors with Serially Dependent Data</h2>
<div class="dt-tags"></div>
<p>The R package RPESE (Risk and Performance Estimators Standard Errors) implements a new method for computing accurate standard errors of risk and performance estimators when returns are serially dependent. The new method makes use of the representation of a risk or performance estimator as a summation of a time series of influence-function (IF) transformed returns, and computes estimator standard errors using a sophisticated method of estimating the spectral density at frequency zero of the time series of IF-transformed returns. Two additional packages used by RPESE are introduced, namely RPEIF which computes and provides graphical displays of the IF of risk and performance estimators, and RPEGLMEN which implements a regularized Gamma generalized linear model polynomial fit to the periodogram of the time series of the IF-transformed returns. A Monte Carlo study shows that the new method provides more accurate estimates of standard errors for risk and performance estimators compared to well-known alternative methods in the presence of serial correlation.</p>
</div>
</a>
<a href="articles/RJ-2021-086/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">May 3, 2021</div>
<div class="dt-authors">
<div class="dt-author">Adrian E. Raftery</div>
<div class="dt-author">Hana evckov</div>
<div class="dt-author">Bernard W. Silverman</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>The vote Package: Single Transferable Vote and Other Electoral Systems in R</h2>
<div class="dt-tags"></div>
<p>We describe the vote package in R, which implements the plurality (or first-past-the-post), two-round runoff, score, approval, and Single Transferable Vote (STV) electoral systems, as well as methods for selecting the Condorcet winner and loser. We emphasize the STV system, which we have found to work well in practice for multi-winner elections with small electorates, such as committee and council elections, and the selection of multiple job candidates. For single-winner elections, STV is also called Instant Runoff Voting (IRV), Ranked Choice Voting (RCV), or the alternative vote (AV) system. The package also implements the STV system with equal preferences, for the first time in a software package, to our knowledge. It also implements a new variant of STV, in which a minimum number of candidates from a specified group are required to be elected. We illustrate the package with several real examples.</p>
</div>
</a>
<a href="articles/RJ-2021-105/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">April 12, 2021</div>
<div class="dt-authors">
<div class="dt-author">H. Sherry Zhang</div>
<div class="dt-author">Dianne Cook</div>
<div class="dt-author">Ursula Laa</div>
<div class="dt-author">Nicolas Langren</div>
<div class="dt-author">Patricia Menndez</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>Visual Diagnostics for Constrained Optimisation with Application to Guided Tours</h2>
<div class="dt-tags"></div>
<p>A guided tour helps to visualise high-dimensional data by showing low-dimensional projec tions along a projection pursuit optimisation path. Projection pursuit is a generalisation of principal component analysis in the sense that different indexes are used to define the interestingness of the projected data. While much work has been done in developing new indexes in the literature, less has been done on understanding the optimisation. Index functions can be noisy, might have multiple local maxima as well as an optimal maximum, and are constrained to generate orthonormal projection frames, which complicates the optimization. In addition, projection pursuit is primarily used for exploratory data analysis, and finding the local maxima is also useful. The guided tour is especially useful for exploration because it conducts geodesic interpolation connecting steps in the optimisation and shows how the projected data changes as a maxima is approached. This work provides new visual diagnostics for examining a choice of optimisation procedure based on the provision of a new data object which collects information throughout the optimisation. It has helped to diagnose and fix several problems with projection pursuit guided tour. This work might be useful more broadly for diagnosing optimisers and comparing their performance. The diagnostics are implemented in the R package ferrn.</p>
</div>
</a>
<a href="articles/RJ-2021-077/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">April 12, 2021</div>
<div class="dt-authors">
<div class="dt-author">Apostolos Chalkis</div>
<div class="dt-author">Vissarion Fisikopoulos</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>volesti: Volume Approximation and Sampling for Convex Polytopes in R</h2>
<div class="dt-tags"></div>
<p>Sampling from high-dimensional distributions and volume approximation of convex bodies are fundamental operations that appear in optimization, finance, engineering, artificial intelligence, and machine learning. In this paper, we present volesti, an R package that provides efficient, scalable algorithms for volume estimation, uniform, and Gaussian sampling from convex polytopes. volesti scales to hundreds of dimensions, handles efficiently three different types of polyhedra and pro vides non existing sampling routines to R. We demonstrate the power of volesti by solving several challenging problems using the R language.</p>
</div>
</a>
<a href="articles/RJ-2021-078/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">April 12, 2021</div>
<div class="dt-authors">
<div class="dt-author">Sladana Babic</div>
<div class="dt-author">Christophe Ley</div>
<div class="dt-author">Marko Palangetic</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>Elliptical Symmetry Tests in R</h2>
<div class="dt-tags"></div>
<p>The assumption of elliptical symmetry has an important role in many theoretical develop ments and applications. Hence, it is of primary importance to be able to test whether that assumption actually holds true or not. Various tests have been proposed in the literature for this problem. To the best of our knowledge, none of them has been implemented in R. This article describes the R package ellipticalsymmetry which implements several well-known tests for elliptical symmetry together with some recent tests. We demonstrate the testing procedures with a real data example.</p>
</div>
</a>
<a href="articles/RJ-2021-104/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">March 22, 2021</div>
<div class="dt-authors">
<div class="dt-author">Jean-Paul Fox</div>
<div class="dt-author">Konrad Klotzke</div>
<div class="dt-author">Duco Veen</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>Generalized Linear Randomized Response Modeling using GLMMRR</h2>
<div class="dt-tags"></div>
<p>Randomized response (RR) designs are used to collect response data about sensitive behaviors (e.g., criminal behavior, sexual desires). The modeling of RR data is more complex since it requires a description of the RR process. For the class of generalized linear mixed models (GLMMs), the RR process can be represented by an adjusted link function, which relates the expected RR to the linear predictor for most common RR designs. The package GLMMRR includes modified link functions for four different cumulative distributions (i.e., logistic, cumulative normal, Gumbel, Cauchy) for GLMs and GLMMs, where the package lme4 facilitates ML and REML estimation. The mixed modeling framework in GLMMRR can be used to jointly analyze data collected under different designs (e.g., dual questioning, multilevel, mixed mode, repeated measurements designs, multiple-group designs). Model-fit tests, tools for residual analyses, and plot functions to give support to a profound RR data analysis are added to the well-known features of the GLM and GLMM software (package lme4). Data of Hglinger and Jann (2018) and Hglinger, Jann, and Diekmann (2014) are used to illustrate the methodology and software.</p>
</div>
</a>
<a href="articles/RJ-2021-095/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">March 8, 2021</div>
<div class="dt-authors">
<div class="dt-author">Rafael Ayala</div>
<div class="dt-author">Daniel Ayala</div>
<div class="dt-author">Lara Sells Vidal</div>
<div class="dt-author">David Ruiz</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>openSkies - Integration of Aviation Data into the R Ecosystem</h2>
<div class="dt-tags"></div>
<p>Aviation data has become increasingly more accessible to the public thanks to the adoption of technologies such as Automatic Dependent Surveillance-Broadcast (ADS-B) and Mode S, which provide aircraft information over publicly accessible radio channels. Furthermore, the OpenSky Network provides multiple public resources to access such air traffic data from a large network of ADS-B receivers. Here, we present openSkies, the first R package for processing public air traffic data. The package provides an interface to the OpenSky Network resources, standardized data structures to represent the different entities involved in air traffic data, and functionalities to analyze and visualize such data. Furthermore, the portability of the implemented data structures makes openSkies easily reusable by other packages, therefore laying the foundation of aviation data engineering in R.</p>
</div>
</a>
<a href="articles/RJ-2021-063/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">March 1, 2021</div>
<div class="dt-authors">
<div class="dt-author">Rachael C. Aikens</div>
<div class="dt-author">Joseph Rigdon</div>
<div class="dt-author">Justin Lee</div>
<div class="dt-author">Michael Baiocchi</div>
<div class="dt-author">Andrew B. Goldstone</div>
<div class="dt-author">Peter Chiu</div>
<div class="dt-author">Y. Joseph Woo</div>
<div class="dt-author">Jonathan H. Chen</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>stratamatch: Prognostic Score Stratification Using a Pilot Design</h2>
<div class="dt-tags"></div>
<p>Optimal propensity score matching has emerged as one of the most ubiquitous approaches
 for causal inference studies on observational data. However, outstanding critiques of the statistical
 properties of propensity score matching have cast doubt on the statistical efficiency of this technique,
 and the poor scalability of optimal matching to large data sets makes this approach inconvenient
 if not infeasible for sample sizes that are increasingly commonplace in modern observational data.
 The stratamatch package provides implementation support and diagnostics for stratified matching
 designs, an approach that addresses both of these issues with optimal propensity score matching for
 large-sample observational studies. First, stratifying the data enables more computationally efficient
 matching of large data sets. Second, stratamatch implements a pilot design approach in order to
 stratify by a prognostic score, which may increase the precision of the effect estimate and increase
 power in sensitivity analyses of unmeasured confounding.</p>
</div>
</a>
<a href="articles/RJ-2021-103/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">March 1, 2021</div>
<div class="dt-authors">
<div class="dt-author">Jouni Helske</div>
<div class="dt-author">Matti Vihola</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>bssm: Bayesian Inference of Non-linear and Non-Gaussian State Space Models in R</h2>
<div class="dt-tags"></div>
<p>We present an R package bssm for Bayesian non-linear/non-Gaussian state space modeling. Unlike the existing packages, bssm allows for easy-to-use approximate inference based on Gaussian approximations such as the Laplace approximation and the extended Kalman filter. The package also accommodates discretely observed latent diffusion processes. The inference is based on fully automatic, adaptive Markov chain Monte Carlo (MCMC) on the hyperparameters, with optional importance sampling post-correction to eliminate any approximation bias. The package also implements a direct pseudo-marginal MCMC and a delayed acceptance pseudo-marginal MCMC using intermediate approximations. The package offers an easy-to-use interface to define models with linear-Gaussian state dynamics with non-Gaussian observation models and has an Rcpp interface for specifying custom non-linear and diffusion models.</p>
</div>
</a>
<a href="articles/RJ-2021-090/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Feb. 22, 2021</div>
<div class="dt-authors">
<div class="dt-author">Zhixin Lun</div>
<div class="dt-author">Ravindra Khattree</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>An R package for Non-Normal Multivariate Distributions: Simulation and Probability Calculations from Multivariate Lomax (Pareto Type II) and Other Related Distributions</h2>
<div class="dt-tags"></div>
<p>Convenient and easy-to-use programs are readily available in R to simulate data from and probability calculations for several common multivariate distributions such as normal and t. However, functions for doing so from other less common multivariate distributions, especially those which are asymmetric, are not as readily available, either in R or otherwise. We introduce the R package NonNorMvtDist to generate random numbers from multivariate Lomax distribution, which constitutes a very flexible family of skewed multivariate distributions. Further, by applying certain useful properties of multivariate Lomax distribution, multivariate cases of generalized Lomax, Mardias Pareto of Type I, Logistic, Burr, Cook-Johnsons uniform, F, and inverted beta can be also considered, and random numbers from these distributions can be generated. Methods for the probability and the equicoordinate quantile calculations for all these distributions are then provided. This work substantially enriches the existing R toolbox for nonnormal or nonsymmetric multivariate probability distributions.</p>
</div>
</a>
<a href="articles/RJ-2021-076/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Feb. 22, 2021</div>
<div class="dt-authors">
<div class="dt-author">Tony Pourmohamad</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>CompModels: A Suite of Computer Model Test Functions for Bayesian Optimization</h2>
<div class="dt-tags"></div>
<p>The CompModels package for R provides a suite of computer model test functions that can be used for computer model prediction/emulation, uncertainty quantification, and calibration. Moreover, the CompModels package is especially well suited for the sequential optimization of computer models. The package is a mix of real-world physics problems, known mathematical functions, and black-box functions that have been converted into computer models with the goal of Bayesian (i.e., sequential) optimization in mind. Likewise, the package contains computer models that represent either the constrained or unconstrained optimization case, each with varying levels of difficulty. In this paper, we illustrate the use of the package with both real-world examples and black-box functions by solving constrained optimization problems via Bayesian optimization. Ultimately, the package is shown to provide users with a source of computer model test functions that are reproducible, shareable, and that can be used for benchmarking of novel optimization methods.</p>
</div>
</a>
<a href="articles/RJ-2021-102/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Feb. 15, 2021</div>
<div class="dt-authors">
<div class="dt-author">Jeremy Gelb</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>spNetwork: A Package for Network Kernel Density Estimation</h2>
<div class="dt-tags"></div>
<p>This paper introduces the new package spNetwork that provides functions to perform Network Kernel Density Estimate analysis (NKDE). This method is an extension of the classical Kernel Density Estimate (KDE), a non parametric approach to estimate the intensity of a spatial process. More specifically, it adapts the KDE for cases when the study area is a network, constraining the location of events (such as accidents on roads, leaks in pipes, fish in rivers, etc.). We present and discuss in this paper the three main versions of NKDE: simple, discontinuous, and continuous that are implemented in spNetwork. We illustrate how to apply the three methods and map their results using a sample from a real dataset representing bike accidents in a central neighborhood of Montreal. We also describe the optimization techniques used to reduce calculation time and investigate their impacts when applying the three NKDE to a city-wide dataset.</p>
</div>
</a>
<a href="articles/RJ-2021-094/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Feb. 8, 2021</div>
<div class="dt-authors">
<div class="dt-author">Jinpu Li</div>
<div class="dt-author">Ryan P. Knigge</div>
<div class="dt-author">Kaiyi Chen</div>
<div class="dt-author">Emily V. Leary</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>PASSED: Calculate Power and Sample Size for Two Sample Tests</h2>
<div class="dt-tags"></div>
<p>Power and sample size estimation are critical aspects of study design to demonstrate minimized risk for subjects and justify the allocation of time, money, and other resources. Researchers often work with response variables that take the form of various distributions. Here, we present an R package, PASSED, that allows flexibility with seven common distributions and multiple options to accommodate sample size or power analysis. The relevant statistical theory, calculations, and examples for each distribution using PASSED are discussed in this paper.</p>
</div>
</a>
<a href="articles/RJ-2021-068/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Feb. 1, 2021</div>
<div class="dt-authors">
<div class="dt-author">Brandon Bolte</div>
<div class="dt-author">Nicols Schmidt</div>
<div class="dt-author">Sergio Bjar</div>
<div class="dt-author">Nguyen Huynh</div>
<div class="dt-author">Bumba Mukherjee</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>BayesSPsurv: An R Package to Estimate Bayesian (Spatial) Split-Population Survival Models</h2>
<div class="dt-tags"></div>
<p>Survival data often include a fraction of units that are susceptible to an event of interest
 as well as a fraction of immune units. In many applications, spatial clustering in unobserved risk
 factors across nearby units can also affect their survival rates and odds of becoming immune. To
 address these methodological challenges, this article introduces our BayesSPsurv R-package, which
 fits parametric Bayesian Spatial split-population survival (cure) models that can account for spatial
 autocorrelation in both subpopulations of the users time-to-event data. Spatial autocorrelation is
 modeled with spatially weighted frailties, which are estimated using a conditionally autoregressive
 prior. The user can also fit parametric cure models with or without nonspatial i.i.d. frailties, and
 each model can incorporate time-varying covariates. BayesSPsurv also includes various functions to
 conduct pre-estimation spatial autocorrelation tests, visualize results, and assess model performance,
 all of which are illustrated using data on post-civil war peace survival.</p>
</div>
</a>
<a href="articles/RJ-2021-053/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Jan. 15, 2021</div>
<div class="dt-authors">
<div class="dt-author">Daniel D. Sjoberg</div>
<div class="dt-author">Karissa Whiting</div>
<div class="dt-author">Michael Curry</div>
<div class="dt-author">Jessica A. Lavery</div>
<div class="dt-author">Joseph Larmarange</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>Reproducible Summary Tables with the gtsummary Package</h2>
<div class="dt-tags"></div>
<p>The gtsummary package provides an elegant and flexible way to create publication-ready
 summary tables in R. A critical part of the work of statisticians, data scientists, and analysts is
 summarizing data sets and regression models in R and publishing or sharing polished summary tables.
 The gtsummary package was created to streamline these everyday analysis tasks by allowing users
 to easily create reproducible summaries of data sets, regression models, survey data, and survival
 data with a simple interface and very little code. The package follows a tidy framework, making it
 easy to integrate with standard data workflows, and offers many table customization features through
 function arguments, helper functions, and custom themes.</p>
</div>
</a>
<a href="articles/RJ-2021-054/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Jan. 15, 2021</div>
<div class="dt-authors">
<div class="dt-author">Lucas Kook</div>
<div class="dt-author">Torsten Hothorn</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-054/feature.png"/>
</div>
<div class="description">
<h2>Regularized Transformation Models: The tramnet Package</h2>
<div class="dt-tags"></div>
<p>The tramnet package implements regularized linear transformation models by combining the
 flexible class of transformation models from tram with constrained convex optimization implemented
 in CVXR. Regularized transformation models unify many existing and novel regularized regression
 models under one theoretical and computational framework. Regularization strategies implemented
 for transformation models in tramnet include the Lasso, ridge regression, and the elastic net and
 follow the parameterization in glmnet. Several functionalities for optimizing the hyperparameters,
 including model-based optimization based on the mlrMBO package, are implemented. A multitude
 of S3 methods is deployed for visualization, handling, and simulation purposes. This work aims at
 illustrating all facets of tramnet in realistic settings and comparing regularized transformation models
 with existing implementations of similar models.</p>
</div>
</a>
<a href="articles/RJ-2021-085/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Jan. 15, 2021</div>
<div class="dt-authors">
<div class="dt-author">Sebastian Juhl</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>spfilteR: An R package for Semiparametric Spatial Filtering with Eigenvectors in (Generalized) Linear Models</h2>
<div class="dt-tags"></div>
<p>Eigenvector-based Spatial filtering constitutes a highly flexible semiparametric approach to account for spatial autocorrelation in a regression framework. It combines judiciously selected eigenvectors from a transformed connectivity matrix to construct a synthetic spatial filter and remove spatial patterns from model residuals. This article introduces the spfilteR package that provides several useful and flexible tools to estimate spatially filtered linear and generalized linear models in R. While the package features functions to identify relevant eigenvectors based on different selection criteria in an unsupervised fashion, it also helps users to perform supervised spatial filtering and to select eigenvectors based on alternative user-defined criteria. Besides a brief discussion of the eigenvector-based spatial filtering approach, this article presents the main functions of the package and illustrates their usage. Comparison to alternative implementations in other R packages highlights the added value of the spfilteR package.</p>
</div>
</a>
<a href="articles/RJ-2021-092/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Jan. 15, 2021</div>
<div class="dt-authors">
<div class="dt-author">Tianhai Zu</div>
<div class="dt-author">Yan Yu</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>SIQR: An R Package for Single-index Quantile Regression</h2>
<div class="dt-tags"></div>
<p>We develop an R package SIQR that implements the single-index quantile regression (SIQR) models via an efficient iterative local linear approach in Wu et al. (2010). Single-index quantile regression models are important tools in semiparametric regression to provide a comprehensive view of the conditional distributions of a response variable. It is especially useful when the data is heterogeneous or heavy-tailed. The package provides functions that allow users to fit SIQR models, predict, provide standard errors of the single-index coefficients via bootstrap, and visualize the estimated univariate function. We apply the R package SIQR to a well-known Boston Housing data.</p>
</div>
</a>
<a href="articles/RJ-2021-093/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Jan. 15, 2021</div>
<div class="dt-authors">
<div class="dt-author">Yuliang Xu</div>
<div class="dt-author">Shuo Shuo Liu</div>
<div class="dt-author">Grace Y. Yi</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>mgee2: An R package for marginal analysis of longitudinal ordinal data with misclassified responses and covariates</h2>
<div class="dt-tags"></div>
<p>Marginal methods have been widely used for analyzing longitudinal ordinal data due to their simplicity in model assumptions, robustness in inference results, and easiness in the implementation. However, they are often inapplicable in the presence of measurement errors in the variables. Under the setup of longitudinal studies with ordinal responses and covariates subject to misclassification, Chen et al. (2014) developed marginal methods for misclassification adjustments using the second-order estimating equations and proposed a two-stage estimation approach when the validation subsample is available. Parameter estimation is conducted through the Newton-Raphson algorithm, and the asymptotic distribution of the estimators is established. While the methods of Chen et al. (2014) can successfully correct the misclassification effects, its implementation is not accessible to general users due to the lack of a software package. In this paper, we develop an R package, mgee2, to implement the marginal methods proposed by Chen et al. (2014). To evaluate the performance and illustrate the features of the package, we conduct numerical studies.</p>
</div>
</a>
<a href="articles/RJ-2021-067/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Jan. 15, 2021</div>
<div class="dt-authors">
<div class="dt-author">Rolf Turner</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>A New Versatile Discrete Distribution</h2>
<div class="dt-tags"></div>
<p>This paper introduces a new flexible distribution for discrete data. Approximate moment estimators of the parameters of the distribution, to be used as starting values for numerical opti mization procedures, are discussed. Exact moment estimation, effected via a numerical procedure, and maximum likelihood estimation, are considered. The quality of the results produced by these estimators is assessed via simulation experiments. Several examples are given of fitting instances of the new distribution to real and simulated data. It is noted that the new distribution is a member of the exponential family. Expressions for the gradient and Hessian of the log-likelihood of the new distribution are derived. The former facilitates the numerical maximization of the likelihood with optim(); the latter provides means of calculating or estimating the covariance matrix of of the parame ter estimates. A discrepancy between estimates of the covariance matrix obtained by inverting the Hessian and those obtained by Monte Carlo methods is discussed.</p>
</div>
</a>
<a href="articles/RJ-2021-101/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Jan. 15, 2021</div>
<div class="dt-authors">
<div class="dt-author">Ali Sabri Taylan</div>
<div class="dt-author">Gkan Yapar</div>
<div class="dt-author">Hanife Taylan Selamlar</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>Automatic Time Series Forecasting with Ata Method in R: ATAforecasting Package</h2>
<div class="dt-tags"></div>
<p>Ata method is a new univariate time series forecasting method that provides innovative solutions to issues faced during the initialization and optimization stages of existing methods. The Ata methods forecasting performance is superior to existing methods in terms of easy implementation and accurate forecasting. It can be applied to non-seasonal or deseasonalized time series, where</p>
</div>
</a>
<a href="articles/RJ-2021-051/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Nov. 2, 2020</div>
<div class="dt-authors">
<div class="dt-author">Michael Kane</div>
<div class="dt-author">Xun Jiang</div>
<div class="dt-author">Simon Urbanek</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-051/waterfall.png"/>
</div>
<div class="description">
<h2>Automating Reproducible, Collaborative Clinical Trial Document Generation with the \pkg{listdown} Package</h2>
<div class="dt-tags"></div>
<p>The conveyance of clinical trial explorations and analysis results from a statistician to a clinical investigator is a critical component of the drug development and clinical research cycle. Automating the process of generating documents for data descriptions, summaries, exploration, and analysis allows the statistician to provide a more comprehensive view of the information captured by a clinical trial, and efficient generation of these documents allows the statistican to focus more on the conceptual development of a trial or trial analysis and less on the implementation of the summaries and results on which decisions are made. This paper explores the use of the \pkg{listdown} package for automating reproducible documents in clinical trials that facilitate the collaboration between statisticians and clinicians as well as defining an analysis pipeline for document generation.</p>
</div>
</a>
<a href="articles/RJ-2021-052/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Nov. 2, 2020</div>
<div class="dt-authors">
<div class="dt-author">Michael J. Kane</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>Towards a Grammar for Processing Clinical Trial Data</h2>
<div class="dt-tags"></div>
<p>The goal of this paper is to help define a path toward a grammar for processing clinical trials by a) defining a format in which we would like to represent data from standardized clinical trial data b) describing a standard set of operations to transform clinical trial data into this format, and c) to identify a set of verbs and other functionality to facilitate data processing and encourage reproducibility in the processing of these data. It provides a background on standard clinical trial data and goes through a simple preprocessing example illustrating the value of the proposed approach through the use of the \pkg{forceps} package, which is currently being used for data of this kind.</p>
</div>
</a>
<a href="articles/RJ-2021-100/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Nov. 2, 2020</div>
<div class="dt-authors">
<div class="dt-author">lvaro Briz-Redn</div>
<div class="dt-author">Francisco Martnez-Ruiz</div>
<div class="dt-author">Francisco Montes</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>DRHotNet: An R package for detecting differential risk hotspots on a linear network</h2>
<div class="dt-tags"></div>
<p>One of the most common applications of spatial data analysis is detecting zones, at a certain scale, where a point-referenced event under study is especially concentrated. The detection of such zones, which are usually referred to as hotspots, is essential in certain fields such as criminology, epidemiology, or traffic safety. Traditionally, hotspot detection procedures have been developed over areal units of analysis. Although working at this spatial scale can be suitable enough for many research or practical purposes, detecting hotspots at a more accurate level (for instance, at the road segment level) may be more convenient sometimes. Furthermore, it is typical that hotspot detection procedures are entirely focused on the determination of zones where an event is (overall) highly concentrated. It is less common, by far, that such procedures focus on detecting zones where a specific type of event is overrepresented in comparison with the other types observed, which have been denoted as differential risk hotspots. The R package DRHotNet provides several functionalities to facilitate the detection of differential risk hotspots within a linear network. In this paper, DRHotNet is depicted, and its usage in the R console is shown through a detailed analysis of a crime dataset.</p>
</div>
</a>
<a href="articles/RJ-2021-075/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Nov. 2, 2020</div>
<div class="dt-authors">
<div class="dt-author">Blint Tamsi</div>
<div class="dt-author">Torsten Hothorn</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>tramME: Mixed-Effects Transformation Models Using Template Model Builder</h2>
<div class="dt-tags"></div>
<p>Linear transformation models constitute a general family of parametric regression models for discrete and continuous responses. To accommodate correlated responses, the model is extended by incorporating mixed effects. This article presents the R package tramME, which builds on existing implementations of transformation models (mlt and tram packages) as well as Laplace approximation and automatic differentiation (using the TMB package), to calculate estimates and perform likelihood inference in mixed-effects transformation models. The resulting framework can be readily applied to a wide range of regression problems with grouped data structures.</p>
</div>
</a>
<a href="articles/RJ-2021-082/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 31, 2020</div>
<div class="dt-authors">
<div class="dt-author">David Moria</div>
<div class="dt-author">Gilma Hernndez-Herrera</div>
<div class="dt-author">Albert Navarro</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>miRecSurv Package: Prentice-Williams-Peterson Models with Multiple Imputation of Unknown Number of Previous Episodes</h2>
<div class="dt-tags"></div>
<p>Left censoring can occur with relative frequency when analyzing recurrent events in epi demiological studies, especially observational ones. Concretely, the inclusion of individuals that were already at risk before the effective initiation in a cohort study may cause the unawareness of prior episodes that have already been experienced, and this will easily lead to biased and inefficient estimates. The miRecSurv package is based on the use of models with specific baseline hazard, with multiple imputation of the number of prior episodes when unknown by means of the COMPoisson distribution, a very flexible count distribution that can handle over, sub, and equidispersion, with a stratified model depending on whether the individual had or had not previously been at risk, and the use of a frailty term. The usage of the package is illustrated by means of a real data example based on an occupational cohort study and a simulation study.</p>
</div>
</a>
<a href="articles/RJ-2021-046/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">John Hughes</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-046/feature.png"/>
</div>
<div class="description">
<h2>krippendorffsalpha: An R Package for Measuring Agreement Using Krippendorff's Alpha Coefficient</h2>
<div class="dt-tags"></div>
<p>R package krippendorffsalpha provides tools for measuring agreement using Krippendorffs
  coefficient, a well-known nonparametric measure of agreement (also called inter-rater reliability
 and various other names). This article first develops Krippendorffs  in a natural way and situates
  among statistical procedures. Then, the usage of package krippendorffsalpha is illustrated via
 analyses of two datasets, the latter of which was collected during an imaging study of hip cartilage.
 The package permits users to apply the  methodology using built-in distance functions for the
 nominal, ordinal, interval, or ratio levels of measurement. User-defined distance functions are also
 supported. The fitting function can accommodate any number of units, any number of coders, and
 missingness. Bootstrap inference is supported, and the bootstrap computation can be carried out in
 parallel.</p>
</div>
</a>
<a href="articles/RJ-2021-047/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">Majeed Simaan</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-047/feature.png"/>
</div>
<div class="description">
<h2>Working with CRSP/COMPUSTAT in R: Reproducible Empirical Asset Pricing</h2>
<div class="dt-tags"></div>
<p>It is common to come across SAS or Stata manuals while working on academic empirical
 finance research. Nonetheless, given the popularity of open-source programming languages such as R,
 there are fewer resources in R covering popular databases such as CRSP and COMPUSTAT. The aim
 of this article is to bridge the gap and illustrate how to leverage R in working with both datasets. As
 an application, we illustrate how to form size-value portfolios with respect to Fama and French (1993)
 and study the sensitivity of the results with respect to different inputs. Ultimately, the purpose of the
 article is to advocate reproducible finance research and contribute to the recent idea of Open Source
 Cross-Sectional Asset Pricing, proposed by Chen and Zimmermann (2020).</p>
</div>
</a>
<a href="articles/RJ-2021-049/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">Ana C. Cebrin</div>
<div class="dt-author">Jess Asn</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-049/feature.png"/>
</div>
<div class="description">
<h2>Analyzing Dependence between Point Processes in Time Using IndTestPP</h2>
<div class="dt-tags"></div>
<p>The need to analyze the dependence between two or more point processes in time appears in
 many modeling problems related to the occurrence of events, such as the occurrence of climate events
 at different spatial locations or synchrony detection in spike train analysis. The package IndTestPP
 provides a general framework for all the steps in this type of analysis, and one of its main features is the
 implementation of three families of tests to study independence given the intensities of the processes,
 which are not only useful to assess independence but also to identify factors causing dependence.
 The package also includes functions for generating different types of dependent point processes,
 and implements computational statistical inference tools using them. An application to characterize
 the dependence between the occurrence of extreme heat events in three Spanish locations using the
 package is shown.</p>
</div>
</a>
<a href="articles/RJ-2021-055/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">Raphael Sonabend</div>
<div class="dt-author">Franz J. Kirly</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>distr6: R6 Object-Oriented Probability Distributions Interface in R</h2>
<div class="dt-tags"></div>
<p>distr6 is an object-oriented (OO) probability distributions interface leveraging the extensibil
ity and scalability of R6 and the speed and efficiency of Rcpp. Over 50 probability distributions are
 currently implemented in the package with core methods, including density, distribution, and gener
ating functions, and more exotic ones, including hazards and distribution function anti-derivatives.
 In addition to simple distributions, distr6 supports compositions such as truncation, mixtures, and
 product distributions. This paper presents the core functionality of the package and demonstrates
 examples for key use-cases. In addition, this paper provides a critical review of the object-oriented
 programming paradigms in R and describes some novel implementations for design patterns and core
 object-oriented features introduced by the package for supporting distr6 components.</p>
</div>
</a>
<a href="articles/RJ-2021-060/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">Ostap Okhrin</div>
<div class="dt-author">Simon Trimborn</div>
<div class="dt-author">Martin Waltz</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>gofCopula: Goodness-of-Fit Tests for Copulae</h2>
<div class="dt-tags"></div>
<p>The last decades show an increased interest in modeling various types of data through
 copulae. Different copula models have been developed, which lead to the challenge of finding the
 best fitting model for a particular dataset. From the other side, a strand of literature developed a list
 of different Goodness-of-Fit (GoF) tests with different powers under different conditions. The usual
 practice is the selection of the best copula via the p-value of the GoF test. Although this method is not
 purely correct due to the fact that non-rejection does not imply acception, this strategy is favored by
 practitioners. Unfortunately, different GoF tests often provide contradicting outputs. The proposed
 R-package brings under one umbrella 13 most used copulae plus their rotated variants together
 with 16 GoF tests and a hybrid one. The package offers flexible margin modeling, automatized
 parallelization, parameter estimation, as well as a user-friendly interface, and pleasant visualizations
 of the results. To illustrate the functionality of the package, two exemplary applications are provided.</p>
</div>
</a>
<a href="articles/RJ-2021-066/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">Mara Xos Rodrguez-lvarez</div>
<div class="dt-author">Vanda Incio</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>ROCnReg: An R Package for Receiver Operating Characteristic Curve Inference With and Without Covariates</h2>
<div class="dt-tags"></div>
<p>This paper introduces the package ROCnReg that allows estimating the pooled ROC
 curve, the covariate-specific ROC curve, and the covariate-adjusted ROC curve by different methods,
 both from (semi) parametric and nonparametric perspectives and within Bayesian and frequentist
 paradigms. From the estimated ROC curve (pooled, covariate-specific, or covariate-adjusted), several
 summary measures of discriminatory accuracy, such as the (partial) area under the ROC curve and the
 Youden index, can be obtained. The package also provides functions to obtain ROC-based optimal
 threshold values using several criteria, namely, the Youden index criterion and the criterion that
 sets a target value for the false positive fraction. For the Bayesian methods, we provide tools for
 assessing model fit via posterior predictive checks, while the model choice can be carried out via
 several information criteria. Numerical and graphical outputs are provided for all methods. This is
 the only package implementing Bayesian procedures for ROC curves.</p>
</div>
</a>
<a href="articles/RJ-2021-072/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">Andrew G. Allmon</div>
<div class="dt-author">J.S. Marron</div>
<div class="dt-author">Michael G. Hudgens</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>diproperm: An R Package for the DiProPerm Test</h2>
<div class="dt-tags"></div>
<p>High-dimensional low sample size (HDLSS) data sets frequently emerge in many biomedical applications. The direction-projection-permutation (DiProPerm) test is a two-sample hypothesis test for comparing two high-dimensional distributions. The DiProPerm test is exact, i.e., the type I error is guaranteed to be controlled at the nominal level for any sample size, and thus is applicable in the HDLSS setting. This paper discusses the key components of the DiProPerm test, introduces the diproperm R package, and demonstrates the package on a real-world data set.</p>
</div>
</a>
<a href="articles/RJ-2021-048/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">Henrik Bengtsson</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>A Unifying Framework for Parallel and Distributed Processing in R using Futures</h2>
<div class="dt-tags"></div>
<p>A future is a programming construct designed for concurrent and asynchronous evaluation of code, making it particularly useful for parallel processing. The future package implements the Future API for programming with futures in R. This minimal API provides sufficient constructs for implementing parallel versions of well-established, high-level map-reduce APIs. The future ecosystem supports exception handling, output and condition relaying, parallel random number generation, and automatic identification of globals lowering the threshold to parallelize code. The Future API bridges parallel frontends with parallel backends, following the philosophy that end-users are the ones who choose the parallel backend while the developer focuses on what to parallelize. A variety of backends exist, and third-party contributions meeting the specifications, which ensure that the same code works on all backends, are automatically supported. The future framework solves several problems not addressed by other parallel frameworks in R.</p>
</div>
</a>
<a href="articles/RJ-2021-073/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">Farhad Pishgar</div>
<div class="dt-author">Noah Greifer</div>
<div class="dt-author">Clmence Leyrat</div>
<div class="dt-author">Elizabeth Stuart</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>MatchThem:: Matching and Weighting after Multiple Imputation</h2>
<div class="dt-tags"></div>
<p>Balancing the distributions of the confounders across the exposure levels in an observational study through matching or weighting is an accepted method to control for confounding due to these variables when estimating the association between an exposure and outcome and reducing the degree of dependence on certain modeling assumptions. Despite the increasing popularity in practice, these procedures cannot be immediately applied to datasets with missing values. Multiple imputation of the missing data is a popular approach to account for missing values while preserving the number of units in the dataset and accounting for the uncertainty in the missing values. However, to the best of our knowledge, there is no comprehensive matching and weighting software that can be easily implemented with multiply imputed datasets. In this paper, we review this problem and suggest a framework to map out the matching and weighting of multiply imputed datasets to 5 actions as well as the best practices to assess balance in these datasets after matching and weighting. We also illustrate these approaches using a companion package for R, MatchThem.</p>
</div>
</a>
<a href="articles/RJ-2021-098/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">Kexuan Yang</div>
<div class="dt-author">Sang Kyu Lee</div>
<div class="dt-author">Jun Zhao</div>
<div class="dt-author">Hyoung-Moon Kim</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>EMSS: New EM-type algorithms for the Heckman selection model in R</h2>
<div class="dt-tags"></div>
<p>When investigators observe non-random samples from populations, sample selectivity problems may occur. The Heckman selection model is widely used to deal with selectivity problems. Based on the EM algorithm, Zhao et al. (2020) developed three algorithms, namely, ECM, ECM(NR), and ECME(NR), which also have the EM algorithms main advantages: stability and ease of imple mentation. This paper provides the implementation of these three new EM-type algorithms in the package EMSS and illustrates the usage of the package on several simulated and real data examples. The comparison between the maximum likelihood estimation method (MLE) and three new EM-type algorithms in robustness issues is further discussed.</p>
</div>
</a>
<a href="articles/RJ-2021-099/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">Daro Ramos-Lpez</div>
<div class="dt-author">Ana D. Maldonado</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>Analysis of Corneal Data in R with the rPACI Package</h2>
<div class="dt-tags"></div>
<p>In ophthalmology, the early detection of keratoconus is still a crucial problem. Placido disk corneal topographers are essential in clinical practice, and many indices for diagnosing corneal irregularities exist. The main goal of this work is to present the R package rPACI, providing several functions to handle and analyze corneal data. This package implements primary indices of corneal irregularity (based on geometrical properties) and compound indices built from the primary ones, either using a generalized linear model or as a Bayesian classifier using a hybrid Bayesian network and performing approximate inference. rPACI aims to make the analysis of corneal data accessible for practitioners and researchers in the field. Moreover, a shiny app was developed to use rPACI in any web browser in a truly user-friendly graphical interface without installing R or writing any R code. It is openly deployed at https://admaldonado.shinyapps.io/rPACI/.</p>
</div>
</a>
<a href="articles/RJ-2021-074/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">A. Pedro Duarte Silva</div>
<div class="dt-author">Paula Brito</div>
<div class="dt-author">Peter Filzmoser</div>
<div class="dt-author">Jos G. Dias</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>MAINT.Data: Modelling and Analysing Interval Data in R</h2>
<div class="dt-tags"></div>
<p>We present the CRAN R package MAINT.Data for the modelling and analysis of multivariate interval data, i.e., where units are described by variables whose values are intervals of IR, representing intrinsic variability. Parametric inference methodologies based on probabilistic models for interval variables have been developed, where each interval is represented by its midpoint and log-range, for</p>
</div>
</a>
<a href="articles/RJ-2021-089/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">Viviane Philipps</div>
<div class="dt-author">Boris P. Hejblum</div>
<div class="dt-author">Mlanie Prague</div>
<div class="dt-author">Daniel Commenges</div>
<div class="dt-author">Ccile Proust-Lima</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>Robust and Efficient Optimization Using a Marquardt-Levenberg Algorithm with R Package marqLevAlg</h2>
<div class="dt-tags"></div>
<p>Implementations in R of classical general-purpose algorithms for local optimization generally have two major limitations which cause difficulties in applications to complex problems: too loose convergence criteria and too long calculation time. By relying on a Marquardt-Levenberg algorithm (MLA), a Newton-like method particularly robust for solving local optimization problems, we provide with marqLevAlg package an efficient and general-purpose local optimizer which (i) prevents con vergence to saddle points by using a stringent convergence criterion based on the relative distance to minimum/maximum in addition to the stability of the parameters and of the objective function; and (ii) reduces the computation time in complex settings by allowing parallel calculations at each iteration. We demonstrate through a variety of cases from the literature that our implementation reli ably and consistently reaches the optimum (even when other optimizers fail) and also largely reduces computational time in complex settings through the example of maximum likelihood estimation of different sophisticated statistical models.</p>
</div>
</a>
<a href="articles/RJ-2021-044/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 27, 2020</div>
<div class="dt-authors">
<div class="dt-author">Alexandre Brouste</div>
<div class="dt-author">Christophe Dutang</div>
<div class="dt-author">Darel Noutsa Mieniedou</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-044/feature.png"/>
</div>
<div class="description">
<h2>OneStep : Le Cam's One-step Estimation Procedure</h2>
<div class="dt-tags"></div>
<p>The OneStep package proposes principally an eponymic function that numerically computes
 Le Cams one-step estimator, which is asymptotically efficient and can be computed faster than the
 maximum likelihood estimator for large datasets. Monte Carlo simulations are carried out for several
 examples (discrete and continuous probability distributions) in order to exhibit the performance of Le
 Cams one-step estimation procedure in terms of efficiency and computational cost on observation
 samples of finite size.</p>
</div>
</a>
<a href="articles/RJ-2021-059/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 27, 2020</div>
<div class="dt-authors">
<div class="dt-author">Ezequiel Toum</div>
<div class="dt-author">Mariano H. Masiokas</div>
<div class="dt-author">Ricardo Villalba</div>
<div class="dt-author">Pierre Pitte</div>
<div class="dt-author">Lucas Ruiz</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>The HBV.IANIGLA Hydrological Model</h2>
<div class="dt-tags"></div>
<p>Over the past 40 years, the HBV (Hydrologiska Byrns Vattenbalansavdelning) hydrological
 model has been one of the most used worldwide due to its robustness, simplicity, and reliable results.
 Despite these advantages, the available versions impose some limitations for research studies in
 mountain watersheds dominated by ice-snow melt runoff (i.e., no glacier module, a limited number of
 elevation bands, among other constraints). Here we present HBV.IANIGLA, a tool for hydroclimatic
 studies in regions with steep topography and/or cryospheric processes which provides a modular
 and extended implementation of the HBV model as an R package. To our knowledge, this is the first
 modular version of the original HBV model. This feature can be very useful for teaching hydrological
 modeling, as it offers the possibility to build a customized, open-source model that can be adjusted to
 different requirements of students and users.</p>
</div>
</a>
<a href="articles/RJ-2021-088/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 19, 2020</div>
<div class="dt-authors">
<div class="dt-author">Shaobo Li</div>
<div class="dt-author">Xiaorui Zhu</div>
<div class="dt-author">Yuejie Chen</div>
<div class="dt-author">Dungang Liu</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>PAsso: an R Package for Assessing Partial Association between Ordinal Variables</h2>
<div class="dt-tags"></div>
<p>Partial association, the dependency between variables after adjusting for a set of covariates, is an important statistical notion for scientific research. However, if the variables of interest are ordered categorical data, the development of statistical methods and software for assessing their partial association is limited. Following the framework established by Liu et al. (2021), we develop an R package PAsso for assessing Partial Associations between ordinal variables. The package provides various functions that allow users to perform a wide spectrum of assessments, including quantification, visualization, and hypothesis testing. In this paper, we discuss the implementation of PAsso in detail and demonstrate its utility through an analysis of the 2016 American National Election Study.</p>
</div>
</a>
<a href="articles/RJ-2021-071/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 13, 2020</div>
<div class="dt-authors">
<div class="dt-author">Rachid Boumaza</div>
<div class="dt-author">Pierre Santagostini</div>
<div class="dt-author">Smail Yousfi</div>
<div class="dt-author">Sabine Demotes-Mainard</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>dad: an R Package for Visualisation, Classification and Discrimination of Multivariate Groups Modelled by their Densities</h2>
<div class="dt-tags"></div>
<p>Multidimensional scaling (MDS), hierarchical cluster analysis (HCA), and discriminant analysis (DA) are classical techniques which deal with data made of n individuals and p variables. When the individuals are divided into T groups, the R package dad associates with each group a multivariate probability density function and then carries out these techniques on the densities, which are estimated by the data under consideration. These techniques are based on distance measures between densities: chi-square, Hellinger, Jeffreys, Jensen-Shannon, and L p for discrete densities, Hellinger , Jeffreys, L2 , and 2-Wasserstein for Gaussian densities, and L2 for numeric non-Gaussian densities estimated by the Gaussian kernel method. Practical methods help the user to give meaning to the outputs in the context of MDS and HCA and to look for an optimal prediction in the context of DA based on the one-leave-out misclassification ratio. Some functions for data management or basic statistics calculations on groups are annexed.</p>
</div>
</a>
<a href="articles/RJ-2021-045/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">July 23, 2020</div>
<div class="dt-authors">
<div class="dt-author">Paul Walter</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-045/feature.png"/>
</div>
<div class="description">
<h2>The R Package smicd: Statistical Methods for Interval-Censored Data</h2>
<div class="dt-tags"></div>
<p>The package allows the use of two new statistical methods for the analysis of interval
censored data: 1) direct estimation/prediction of statistical indicators and 2) linear (mixed) regression
 analysis. Direct estimation of statistical indicators, for instance, poverty and inequality indicators,
 is facilitated by a non parametric kernel density algorithm. The algorithm is able to account for
 weights in the estimation of statistical indicators. The standard errors of the statistical indicators are
 estimated with a non parametric bootstrap. Furthermore, the package offers statistical methods for
 the estimation of linear and linear mixed regression models with an interval-censored dependent
 variable, particularly random slope and random intercept models. Parameter estimates are obtained
 through a stochastic expectation-maximization algorithm. Standard errors are estimated using a
 non parametric bootstrap in the linear regression model and by a parametric bootstrap in the linear
 mixed regression model. To handle departures from the model assumptions, fixed (logarithmic) and
 data-driven (Box-Cox) transformations are incorporated into the algorithm.</p>
</div>
</a>
<a href="articles/RJ-2021-083/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">July 14, 2020</div>
<div class="dt-authors">
<div class="dt-author">Kazushi Maruo</div>
<div class="dt-author">Ryota Ishii</div>
<div class="dt-author">Yusuke Yamaguchi</div>
<div class="dt-author">Masahiko Gosho</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>bcmixed: A Package for Median Inference on Longitudinal Data with the BoxCox Transformation</h2>
<div class="dt-tags"></div>
<p>This article illustrates the use of the bcmixed package and focuses on the two main functions: bcmarg and bcmmrm. The bcmarg function provides inference results for a marginal model of a mixed ef fect model using the BoxCox transformation. The bcmmrm function provides model median inferences based on the mixed effect models for repeated measures analysis using the BoxCox transformation for longitudinal randomized clinical trials. Using the bcmmrm function, analysis results with high power and high interpretability for treatment effects can be obtained for longitudinal randomized clinical trials with skewed outcomes. Further, the bcmixed package provides summarizing and visualization tools, which would be helpful to write clinical trial reports.</p>
</div>
</a>
<a href="articles/RJ-2021-043/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">July 6, 2020</div>
<div class="dt-authors">
<div class="dt-author">Sebastian Kreutzer</div>
<div class="dt-author">Johannes Friedrich</div>
<div class="dt-author">Vasilis Pagonis</div>
<div class="dt-author">Christian Laag</div>
<div class="dt-author">Ena Rajovic</div>
<div class="dt-author">Christoph Schmidt</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-043/feature.png"/>
</div>
<div class="description">
<h2>RLumCarlo: Simulating Cold Light using Monte Carlo Methods</h2>
<div class="dt-tags"></div>
<p>The luminescence phenomena of insulators and semiconductors (e.g., natural minerals such
 as quartz) have various application domains. For instance, Earth Sciences and archaeology exploit
 luminescence as a dating method. Herein, we present the R package RLumCarlo implementing sets of
 luminescence models to be simulated with Monte Carlo (MC) methods. MC methods make a powerful
 ally to all kinds of simulation attempts involving stochastic processes. Luminescence production
 is such a stochastic process in the form of charge (electron-hole pairs) interaction within insulators
 and semiconductors. To simulate luminescence-signal curves, we distribute single and independent
 MC processes to virtual MC clusters. RLumCarlo comes with a modularized design and consistent
 user interface: (1) C++ functions represent the modeling core and implement models for specific
 stimulations modes. (2) R functions give access to combinations of models and stimulation modes,
 start the simulation and render terminal and graphical feedback. The combination of MC clusters
 supports the simulation of complex luminescence phenomena.</p>
</div>
</a>
<a href="articles/RJ-2021-087/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">June 25, 2020</div>
<div class="dt-authors">
<div class="dt-author">Thiago R. Santos</div>
<div class="dt-author">Glaura C. Franco</div>
<div class="dt-author">Dani Gamerman</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>NGSSEML: Non-Gaussian State Space with Exact Marginal Likelihood</h2>
<div class="dt-tags"></div>
<p>The number of packages/software for Gaussian State Space models has increased over recent decades. However, there are very few codes available for non-Gaussian State Space (NGSS) models due to analytical intractability that prevents exact calculations. One of the few tractable exceptions is the family of NGSS with exact marginal likelihood, named NGSSEML. In this work, we present the wide range of data formats and distributions handled by NGSSEML and a package in the R language to perform classical and Bayesian inference for them. Special functions for filtering, forecasting, and smoothing procedures and the exact calculation of the marginal likelihood function are provided. The methods implemented in the package are illustrated for count and volatility time series and some reliability/survival models, showing that the codes are easy to handle. Therefore, the NGSSEML family emerges as a simple and interesting option/alternative for modeling non-Gaussian time-varying structures commonly encountered in time series and reliability/survival studies. Keywords: Bayesian, classical inference, reliability, smoothing, time series, software R</p>
</div>
</a>
<a href="articles/RJ-2021-097/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">June 5, 2020</div>
<div class="dt-authors">
<div class="dt-author">Jinhui Yang</div>
<div class="dt-author">Lifeng Lin</div>
<div class="dt-author">Haitao Chu</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>BayesSenMC: an R package for Bayesian Sensitivity Analysis of Misclassification</h2>
<div class="dt-tags"></div>
<p>In casecontrol studies, the odds ratio is commonly used to summarize the association be tween a binary exposure and a dichotomous outcome. However, exposure misclassification frequently appears in casecontrol studies due to inaccurate data reporting, which can produce bias in measures of association. In this article, we implement a Bayesian sensitivity analysis of misclassification to provide a full posterior inference on the corrected odds ratio under both non-differential and differen tial misclassification. We present an R (R Core Team, 2018) package BayesSenMC, which provides user-friendly functions for its implementation. The usage is illustrated by a real data analysis on the association between bipolar disorder and rheumatoid arthritis.</p>
</div>
</a>
<a href="articles/RJ-2021-035/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">June 3, 2020</div>
<div class="dt-authors">
<div class="dt-author">Markus Fritsch</div>
<div class="dt-author">Andrew Adrian Yu Pua</div>
<div class="dt-author">Joachim Schnurbus</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-035/feature.png"/>
</div>
<div class="description">
<h2>pdynmc: A Package for Estimating Linear Dynamic Panel Data Models Based on Nonlinear Moment Conditions</h2>
<div class="dt-tags"></div>
<p>This paper introduces pdynmc, an R package that provides users sufficient flexibility
 and precise control over the estimation and inference in linear dynamic panel data models. The
 package primarily allows for the inclusion of nonlinear moment conditions and the use of iterated
 GMM; additionally, visualizations for data structure and estimation results are provided. The current
 implementation reflects recent developments in literature, uses sensible argument defaults, and
 aligns commercial and noncommercial estimation commands. Since the understanding of the model
 assumptions is vital for setting up plausible estimation routines, we provide a broad introduction
 of linear dynamic panel data models directed towards practitioners before concisely describing the
 functionality available in pdynmc regarding instrument type, covariate type, estimation methodology,
 and general configuration. We then demonstrate the functionality by revisiting the popular firm-level
 dataset of Arellano and Bond (1991).</p>
</div>
</a>
<a href="articles/RJ-2021-036/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">June 3, 2020</div>
<div class="dt-authors">
<div class="dt-author">Julio E. Sandubete</div>
<div class="dt-author">Lorenzo Escot</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-036/feature.png"/>
</div>
<div class="description">
<h2>DChaos: An R Package for Chaotic Time Series Analysis</h2>
<div class="dt-tags"></div>
<p>Chaos theory has been hailed as a revolution of thoughts and attracting ever-increasing
 attention of many scientists from diverse disciplines. Chaotic systems are non-linear deterministic
 dynamic systems which can behave like an erratic and apparently random motion. A relevant field
 inside chaos theory is the detection of chaotic behavior from empirical time-series data. One of the
 main features of chaos is the well-known initial-value sensitivity property. Methods and techniques
 related to testing the hypothesis of chaos try to quantify the initial-value sensitive property estimating
 the so-called Lyapunov exponents. This paper describes the main estimation methods of the Lyapunov
 exponent from time series data. At the same time, we present the DChaos library. R users may
 compute the delayed-coordinate embedding vector from time series data, estimates the best-fitted
 neural net model from the delayed-coordinate embedding vectors, calculates analytically the partial
 derivatives from the chosen neural nets model. They can also obtain the neural net estimator of the
 Lyapunov exponent from the partial derivatives computed previously by two different procedures
 and four ways of subsampling by blocks. To sum up, the DChaos package allows the R users to test
 robustly the hypothesis of chaos in order to know if the data-generating process behind time series
 behaves chaotically or not. The packages functionality is illustrated by examples.</p>
</div>
</a>
<a href="articles/RJ-2021-038/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">June 3, 2020</div>
<div class="dt-authors">
<div class="dt-author">Alejandro Saavedra-Nieves</div>
<div class="dt-author">Paula Saavedra-Nieves</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>IndexNumber: An R Package for Measuring the Evolution of Magnitudes</h2>
<div class="dt-tags"></div>
<p>Index numbers are descriptive statistical measures useful in economic settings for comparing
 simple and complex magnitudes registered, usually in two time periods. Although this theory has
 a large history, it still plays an important role in modern todays societies where big amounts of
 economic data are available and need to be analyzed. After a detailed revision on classical index
 numbers in literature, this paper is focused on the description of the R package IndexNumber with
 strong capabilities for calculating them. Two of the four real data sets contained in this library
 are used for illustrating the determination of the index numbers in this work. Graphical tools are
 also implemented in order to show the time evolution of considered magnitudes simplifying the
 interpretation of the results.</p>
</div>
</a>
<a href="articles/RJ-2021-040/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">June 3, 2020</div>
<div class="dt-authors">
<div class="dt-author">Valeria Policastro</div>
<div class="dt-author">Dario Righelli</div>
<div class="dt-author">Annamaria Carissimo</div>
<div class="dt-author">Luisa Cutillo</div>
<div class="dt-author">Italia De Feis</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-040/feature.png"/>
</div>
<div class="description">
<h2>ROBustness In Network (robin): an R Package for Comparison and Validation of Communities </h2>
<div class="dt-tags"></div>
<p>In network analysis, many community detection algorithms have been developed. However,
 their implementation leaves unaddressed the question of the statistical validation of the results. Here,
 we present robin (ROBustness In Network), an R package to assess the robustness of the community
 structure of a network found by one or more methods to give indications about their reliability. The
 procedure initially detects if the community structure found by a set of algorithms is statistically
 significant and then compares two selected detection algorithms on the same graph to choose the
 one that better fits the network of interest. We demonstrate the use of our package on the American
 College Football benchmark dataset.</p>
</div>
</a>
<a href="articles/RJ-2021-041/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">June 3, 2020</div>
<div class="dt-authors">
<div class="dt-author">Ryan A. Peterson</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-041/feature.png"/>
</div>
<div class="description">
<h2>Finding Optimal Normalizing Transformations via \pkg{bestNormalize}</h2>
<div class="dt-tags"></div>
<p>The \pkg{bestNormalize} R package was designed to help users find a transformation that can effectively normalize a vector regardless of its actual distribution. Each of the many normalization techniques that have been developed has its own strengths and weaknesses, and deciding which to use until data are fully observed is difficult or impossible. This package facilitates choosing between a range of possible transformations and will automatically return  the best one, i.e., the one that makes data look the *most* normal. To evaluate and compare  the normalization efficacy across a suite of possible transformations, we developed a  statistic based on a goodness of fit test divided by its degrees of freedom.  Transformations can be seamlessly trained and applied to newly observed data  and can be implemented in conjunction with \pkg{caret} and \pkg{recipes} for  data preprocessing in machine learning workflows. Custom transformations and  normalization statistics are supported.</p>
</div>
</a>
<a href="articles/RJ-2021-042/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">June 3, 2020</div>
<div class="dt-authors">
<div class="dt-author">Javier Roca-Pardias</div>
<div class="dt-author">Mara Xos Rodrguez-lvarez</div>
<div class="dt-author">Stefan Sperlich</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-042/feature.png"/>
</div>
<div class="description">
<h2>Package wsbackfit for Smooth Backfitting Estimation of Generalized Structured Models</h2>
<div class="dt-tags"></div>
<p>A package is introduced that provides the weighted smooth backfitting estimator for
a large family of popular semiparametric regression models. This family is known as generalized
structured models, comprising, for example, generalized varying coefficient model, generalized
additive models, mixtures, potentially including parametric parts. The kernel-based weighted
smooth backfitting belongs to the statistically most efficient procedures for this model class. Its
asymptotic properties are well-understood thanks to the large body of literature about this estimator.
The introduced weights allow for the inclusion of sampling weights, trimming, and efficient estimation
under heteroscedasticity. Further options facilitate easy handling of aggregated data, prediction,
and the presentation of estimation results. Cross-validation methods are provided which can be used
for model and bandwidth selection.1</p>
</div>
</a>
<a href="articles/RJ-2021-057/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">June 3, 2020</div>
<div class="dt-authors">
<div class="dt-author">Genaro Sucarrat</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>garchx: Flexible and Robust GARCH-X Modeling</h2>
<div class="dt-tags"></div>
<p>The garchx package provides a user-friendly, fast, flexible, and robust framework for the
 estimation and inference of GARCH(p, q, r)-X models, where p is the ARCH order, q is the GARCH
 order, r is the asymmetry or leverage order, and X indicates that covariates can be included. Quasi
 Maximum Likelihood (QML) methods ensure estimates are consistent and standard errors valid,
 even when the standardized innovations are non-normal or dependent, or both. Zero-coefficient
 restrictions by omission enable parsimonious specifications, and functions to facilitate the non-standard
 inference associated with zero-restrictions in the null-hypothesis are provided. Finally, in the formal
 comparisons of precision and speed, the garchx package performs well relative to other prominent
 GARCH-packages on CRAN.</p>
</div>
</a>
<a href="articles/RJ-2021-091/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">June 3, 2020</div>
<div class="dt-authors">
<div class="dt-author">Javier Rodrguez-Cuadrado</div>
<div class="dt-author">Juan C. Laria</div>
<div class="dt-author">David Delgado-Gmez</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>cat.dt: An R package for fast construction of accurate Computerized Adaptive Tests using Decision Trees</h2>
<div class="dt-tags"></div>
<p>This article introduces the cat.dt package for the creation of Computerized Adaptive Tests (CATs). Unlike existing packages, the cat.dt package represents the CAT in a Decision Tree (DT) structure. This allows building the test before its administration, ensuring that the creation time of the test is independent of the number of participants. Moreover, to accelerate the construction of the tree, the package controls its growth by joining nodes with similar estimations or distributions of the ability level and uses techniques such as message passing and pre-calculations. The constructed tree, as well as the estimation procedure, can be visualized using the graphical tools included in the package. An experiment designed to evaluate its performance shows that the cat.dt package drastically reduces computational time in the creation of CATs without compromising accuracy.</p>
</div>
</a>
<a href="articles/RJ-2021-081/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">June 3, 2020</div>
<div class="dt-authors">
<div class="dt-author">Andrs RamrezHassan</div>
<div class="dt-author">Mateo Graciano-Londoo</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>A GUIded tour of Bayesian regression</h2>
<div class="dt-tags"></div>
<p>This paper presents a Graphical User Interface (GUI) to carry out a Bayesian regression analysis in a very friendly environment without any programming skills (drag and drop). This paper is designed for teaching and applied purposes at an introductory level. Our GUI is based on an interactive web application using shiny and libraries from R. We carry out some applications to highlight the potential of our GUI for applied researchers and practitioners. In addition, the Help option in the main tap panel has an extended version of this paper, where we present the basic theory underlying all regression models that we developed in our GUI and more applications associated with each model.</p>
</div>
</a>
<a href="articles/RJ-2021-039/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">June 3, 2020</div>
<div class="dt-authors">
<div class="dt-author">Sbastien Wouters</div>
<div class="dt-author">Anne-Christine Da Silva</div>
<div class="dt-author">Frdric Boulvain</div>
<div class="dt-author">Xavier Devleeschouwer</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>StratigrapheR: Concepts for Litholog Generation in R </h2>
<div class="dt-tags"></div>
<p>The StratigrapheR package proposes new concepts for the generation of lithological logs, or lithologs, in R. The generation of lithologs in a scripting environment opens new opportunities for the processing and analysis of stratified geological data. Among the new concepts presented: new plotting and data processing methodologies, new general R functions, and computer-oriented data conventions are provided. The package structure allows for these new concepts to be further improved, which can be done independently by any R user. The current limitations of the package are highlighted, along with the limitations in R for geological data processing, to help identify the best paths for improvements.</p>
</div>
</a>
<a href="articles/RJ-2021-034/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">May 7, 2020</div>
<div class="dt-authors">
<div class="dt-author">Miguel Flores</div>
<div class="dt-author">Rubn Fernndez-Casal</div>
<div class="dt-author">Salvador Naya</div>
<div class="dt-author">Javier Tarro-Saavedra</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-034/feature.png"/>
</div>
<div class="description">
<h2>Statistical Quality Control with the qcr Package</h2>
<div class="dt-tags"></div>
<p>The R package qcr for Statistical Quality Control (SQC) is introduced and described. It
     includes a comprehensive set of univariate and multivariate SQC tools that completes and increases
     the SQC techniques available in R. Apart from integrating different R packages devoted to SQC (qcc,
     MSQC), qcr provides nonparametric tools that are highly useful when Gaussian assumption is not
     met. This package computes standard univariate control charts for individual measurements, x, S, R,
     p, np, c, u, EWMA, and CUSUM. In addition, it includes functions to perform multivariate control
     charts such as Hotelling T2 , MEWMA and MCUSUM. As representative features, multivariate
     nonparametric alternatives based on data depth are implemented in this package: r, Q and S control
     charts. The qcr library also estimates the most complete set of capability indices from first to
     the fourth generation, covering the nonparametric alternatives, and performing the corresponding
     capability analysis graphical outputs, including the process capability plots. Moreover, Phase I and
     II control charts for functional data are included.

    Prcticas de CEC con R</p>
</div>
</a>
<a href="articles/RJ-2021-030/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">May 1, 2020</div>
<div class="dt-authors">
<div class="dt-author">Emmanuel Caron</div>
<div class="dt-author">Jrme Dedecker</div>
<div class="dt-author">Bertrand Michel</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-030/feature.png"/>
</div>
<div class="description">
<h2>Linear Regression with Stationary Errors: the R Package slm</h2>
<div class="dt-tags"></div>
<p>This paper introduces the R package slm, which stands for Stationary Linear Models.
The package contains a set of statistical procedures for linear regression in the general context where
the error process is strictly stationary with a short memory. We work in the setting of Hannan (1973),
who proved the asymptotic normality of the (normalized) least squares estimators (LSE) under
very mild conditions on the error process. We propose different ways to estimate the asymptotic
covariance matrix of the LSE and then to correct the type I error rates of the usual tests on the
parameters (as well as confidence intervals). The procedures are evaluated through different sets of
simulations.</p>
</div>
</a>
<a href="articles/RJ-2021-032/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">May 1, 2020</div>
<div class="dt-authors">
<div class="dt-author">Nora M. Villanueva</div>
<div class="dt-author">Marta Sestelo</div>
<div class="dt-author">Luis Meira-Machado</div>
<div class="dt-author">Javier Roca-Pardias</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-032/feature.png"/>
</div>
<div class="description">
<h2>clustcurv: An R Package for Determining Groups in Multiple Curves </h2>
<div class="dt-tags"></div>
<p>In many situations, it could be interesting to ascertain whether groups of curves can be
 performed, especially when confronted with a considerable number of curves. This paper introduces
 an R package, known as clustcurv, for determining clusters of curves with an automatic selection of
 their number. The package can be used for determining groups in multiple survival curves as well as
 for multiple regression curves. Moreover, it can be used with large numbers of curves. An illustration
 of the use of clustcurv is provided, using both real data examples and artificial data.
 Keywords: multiple curves, number of groups, nonparametric, survival analysis, regression models,
 cluster</p>
</div>
</a>
<a href="articles/RJ-2021-033/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">May 1, 2020</div>
<div class="dt-authors">
<div class="dt-author">Eashwar V. Somasundaram</div>
<div class="dt-author">Shael E. Brown</div>
<div class="dt-author">Adam Litzler</div>
<div class="dt-author">Jacob G. Scott</div>
<div class="dt-author">Raoul R. Wadhwa</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-033/feature.png"/>
</div>
<div class="description">
<h2>Benchmarking R packages for Calculation of Persistent Homology</h2>
<div class="dt-tags"></div>
<p>Several persistent homology software libraries have been implemented in R. Specifically,
 the Dionysus, GUDHI, and Ripser libraries have been wrapped by the TDA and TDAstats CRAN
 packages. These software represent powerful analysis tools that are computationally expensive and, to
 our knowledge, have not been formally benchmarked. Here, we analyze runtime and memory growth
 for the 2 R packages and the 3 underlying libraries. We find that datasets with less than 3 dimensions
 can be evaluated with persistent homology fastest by the GUDHI library in the TDA package. For
 higher-dimensional datasets, the Ripser library in the TDAstats package is the fastest. Ripser and
 TDAstats are also the most memory-efficient tools to calculate persistent homology.</p>
</div>
</a>
<a href="articles/RJ-2021-080/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">May 1, 2020</div>
<div class="dt-authors">
<div class="dt-author">Jingchen Hu</div>
<div class="dt-author">Olanrewaju Akande</div>
<div class="dt-author">Quanli Wang</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>Multiple Imputation and Synthetic Data Generation with NPBayesImputeCat</h2>
<div class="dt-tags"></div>
<p>In many contexts, missing data and disclosure control are ubiquitous and challenging issues. In particular, at statistical agencies, the respondent-level data they collect from surveys and censuses can suffer from high rates of missingness. Furthermore, agencies are obliged to protect respondents privacy when publishing the collected data for public use. The NPBayesImputeCat R package, introduced in this paper, provides routines to i) create multiple imputations for missing data and ii) create synthetic data for statistical disclosure control, for multivariate categorical data, with or without structural zeros. We describe the Dirichlet process mixture of products of the multinomial distributions model used in the package and illustrate various uses of the package using data samples from the American Community Survey (ACS). We also compare results of the missing data imputation to the mice R package and those of the synthetic data generation to the synthpop R package.</p>
</div>
</a>
<a href="articles/RJ-2021-096/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">May 1, 2020</div>
<div class="dt-authors">
<div class="dt-author">Novia Permatasari</div>
<div class="dt-author">Azka Ubaidillah</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>msae: An R Package of Multivariate Fay-Herriot Models for Small Area Estimation</h2>
<div class="dt-tags"></div>
<p>The paper introduces an R Package of multivariate Fay-Herriot models for small area estimation named msae. This package implements four types of Fay-Herriot models, including univariate Fay-Herriot model (model 0), multivariate Fay-Herriot model (model 1), autoregressive multivariate Fay-Herriot model (model 2), and heteroskedastic autoregressive multivariate Fay-Herriot model (model 3). It also contains some datasets generated based on multivariate Fay-Herriot models. We describe and implement functions through various practical examples. Multivariate Fay-Herriot models produce a more efficient parameter estimation than direct estimation and univariate model.</p>
</div>
</a>
<a href="articles/RJ-2021-029/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">April 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">Toby Dylan Hocking</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-029/feature.png"/>
</div>
<div class="description">
<h2>Wide-to-tall Data Reshaping Using Regular Expressions and the nc Package</h2>
<div class="dt-tags"></div>
<p>Regular expressions are powerful tools for extracting tables from non-tabular text data.
 Capturing regular expressions that describe the information to extract from column names can be
 especially useful when reshaping a data table from wide (few rows with many regularly named
 columns) to tall (fewer columns with more rows). We present the R package nc (short for named
 capture), which provides functions for wide-to-tall data reshaping using regular expressions. We
 describe the main new ideas of nc, and provide detailed comparisons with related R packages (stats,
 utils, data.table, tidyr, tidyfast, tidyfst, reshape2, cdata).</p>
</div>
</a>
<a href="articles/RJ-2021-062/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">April 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">Moreno I. Coco</div>
<div class="dt-author">Dan Mnster</div>
<div class="dt-author">Giuseppe Leonardi</div>
<div class="dt-author">Rick Dale</div>
<div class="dt-author">Sebastian Wallot</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>Unidimensional and Multidimensional Methods for Recurrence Quantification Analysis with crqa</h2>
<div class="dt-tags"></div>
<p>Recurrence quantification analysis is a widely used method for characterizing patterns in
 time series. This article presents a comprehensive survey for conducting a wide range of recurrence
based analyses to quantify the dynamical structure of single and multivariate time series and capture
 coupling properties underlying leader-follower relationships. The basics of recurrence quantification
 analysis (RQA) and all its variants are formally introduced step-by-step from the simplest auto
recurrence to the most advanced multivariate case. Importantly, we show how such RQA methods can
 be deployed under a single computational framework in R using a substantially renewed version of
 our crqa 2.0 package. This package includes implementations of several recent advances in recurrence
based analysis, among them applications to multivariate data and improved entropy calculations
 for categorical data. We show concrete applications of our package to example data, together with a
 detailed description of its functions and some guidelines on their usage.</p>
</div>
</a>
<a href="articles/RJ-2021-065/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">April 30, 2020</div>
<div class="dt-authors">
<div class="dt-author">Miguel Ferreiro-Daz</div>
<div class="dt-author">Toms R. Cotos-Yez</div>
<div class="dt-author">Jos R. Mndez</div>
<div class="dt-author">David Ruano-Ords</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>The bdpar Package: Big Data Pipelining Architecture for R</h2>
<div class="dt-tags"></div>
<p>In the last years, big data has become a useful paradigm for taking advantage of multiple
 sources to find relevant knowledge in real domains (such as the design of personalized marketing
 campaigns or helping to palliate the effects of several fatal diseases). Big data programming tools and
 methods have evolved over time from a MapReduce to a pipeline-based archetype. Concretely the use
 of pipelining schemes has become the most reliable way of processing and analyzing large amounts of
 data. To this end, this work introduces bdpar, a new highly customizable pipeline-based framework
 (using the OOP paradigm provided by R6 package) able to execute multiple preprocessing tasks over
 heterogeneous data sources. Moreover, to increase the flexibility and performance, bdpar provides
 helpful features such as (i) the definition of a novel object-based pipe operator (%&gt;|%), (ii) the ability to
 easily design and deploy new (and customized) input data parsers, tasks, and pipelines, (iii) only-once
 execution which avoids the execution of previously processed information (instances), guaranteeing
 that only new both input data and pipelines are executed, (iv) the capability to perform serial or
 parallel operations according to the user needs, (v) the inclusion of a debugging mechanism which
 allows users to check the status of each instance (and find possible errors) throughout the process.</p>
</div>
</a>
<a href="articles/RJ-2021-028/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">April 4, 2020</div>
<div class="dt-authors">
<div class="dt-author">Hong Wang</div>
<div class="dt-author">Ning Li</div>
<div class="dt-author">Shanpeng Li</div>
<div class="dt-author">Gang Li</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-028/feature.png"/>
</div>
<div class="description">
<h2>JMcmprsk: An R Package for Joint Modelling of Longitudinal and Survival Data with Competing Risks</h2>
<div class="dt-tags"></div>
<p>In this paper, we describe an R package named JMcmprsk, for joint modelling of longitudinal
 and survival data with competing risks. The package in its current version implements two joint
 models of longitudinal and survival data proposed to handle competing risks survival data together
 with continuous and ordinal longitudinal outcomes respectively (Elashoff et al., 2008; Li et al., 2010).
 The corresponding R implementations are further illustrated with real examples. The package also
 provides simulation functions to simulate datasets for joint modelling with continuous or ordinal
 outcomes under the competing risks scenario, which provide useful tools to validate and evaluate
 new joint modelling methods.</p>
</div>
</a>
<a href="articles/RJ-2021-031/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">April 4, 2020</div>
<div class="dt-authors">
<div class="dt-author">Falk Hee</div>
<div class="dt-author">Karina Cucchi</div>
<div class="dt-author">Nura Kawa</div>
<div class="dt-author">Yoram Rubin</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-031/feature.png"/>
</div>
<div class="description">
<h2>exPrior: An R Package for the Formulation of Ex-Situ Priors</h2>
<div class="dt-tags"></div>
<p>The exPrior package implements a procedure for formulating informative priors of geo
statistical properties for a target field site, called ex-situ priors and introduced in Cucchi et al. (2019).
 The procedure uses a Bayesian hierarchical model to assimilate multiple types of data coming from
 multiple sites considered as similar to the target site. This prior summarizes the information contained
 in the data in the form of a probability density function that can be used to better inform further
 geostatistical investigations at the site. The formulation of the prior uses ex-situ data, where the data
 set can either be gathered by the user or come in the form of a structured database. The package is
 designed to be flexible in that regard. For illustration purposes and for easiness of use, the package is
 ready to be used with the worldwide hydrogeological parameter database (WWHYPDA) Comunian
 and Renard (2009).</p>
</div>
</a>
<a href="articles/RJ-2021-079/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">April 4, 2020</div>
<div class="dt-authors">
<div class="dt-author">Hkon Otneim</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>lg: An R package for Local Gaussian Approximations</h2>
<div class="dt-tags"></div>
<p>The package lg for the R programming language provides implementations of recent methodological advances on applications of the local Gaussian correlation. This includes the estimation of the local Gaussian correlation itself, multivariate density estimation, conditional density estimation, various tests for independence and conditional independence, as well as a graphical module for creating dependence maps. This paper describes the lg package, its principles, and its practical use.</p>
</div>
</a>
<a href="articles/RJ-2021-070/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Feb. 24, 2020</div>
<div class="dt-authors">
<div class="dt-author">Gustavo Soutinho</div>
<div class="dt-author">Marta Sestelo</div>
<div class="dt-author">Lus Meira-Machado</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>survidm: An R package for Inference and Prediction in an Illness-Death Model</h2>
<div class="dt-tags"></div>
<p>Multi-state models are a useful way of describing a process in which an individual moves through a number of finite states in continuous time. The illness-death model plays a central role in the theory and practice of these models, describing the dynamics of healthy subjects who may move to an intermediate "diseased" state before entering into a terminal absorbing state. In these models, one important goal is the modeling of transition rates which is usually done by studying the relationship between covariates and disease evolution. However, biomedical researchers are also interested in reporting other interpretable results in a simple and summarized manner. These include estimates of predictive probabilities, such as the transition probabilities, occupation probabilities, cumulative incidence functions, and the sojourn time distributions. The development of survidm package has been motivated by recent contribution that provides answers to all these topics. An illustration of the software usage is included using real data.</p>
</div>
</a>
<a href="articles/RJ-2021-061/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Feb. 14, 2020</div>
<div class="dt-authors">
<div class="dt-author">Alessandro Beretta</div>
<div class="dt-author">Cdric Heuchenne</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>penPHcure: Variable Selection in Proportional Hazards Cure Model with Time-Varying Covariates</h2>
<div class="dt-tags"></div>
<p>We describe the penPHcure R package, which implements the semiparametric proportional
hazards (PH) cure model of Sy and Taylor (2000) extended to time-varying covariates and the variable
 selection technique based on its SCAD-penalized likelihood proposed by Beretta and Heuchenne
 (2019a). In survival analysis, cure models are a useful tool when a fraction of the population is likely to
 be immune from the event of interest. They can separate the effects of certain factors on the probability
 of being susceptible and on the time until the occurrence of the event. Moreover, the penPHcure
 package allows the user to simulate data from a PH cure model, where the event-times are generated
 on a continuous scale from a piecewise exponential distribution conditional on time-varying covariates,
 with a method similar to Hendry (2014). We present the results of a simulation study to assess the
 finite sample performance of the methodology and illustrate the functionalities of the penPHcure
 package using criminal recidivism data.</p>
</div>
</a>
<a href="articles/RJ-2021-069/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Nov. 15, 2019</div>
<div class="dt-authors">
<div class="dt-author">Ran Xu</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>Estimating Social Influence Effects in Networks Using A Latent Space Adjusted Approach in R </h2>
<div class="dt-tags"></div>
<p>Social influence effects have been extensively studied in various empirical network research. However, many challenges remain in estimating social influence effects in networks, as influence effects are often entangled with other factors, such as homophily in the selection process and the common social-environmental factors that individuals are embedded in. Methods currently available either do not solve these problems or require stringent assumptions. Recent works by Xu (2018) and others have shown that a latent space adjusted approach based on the latent space model has the potential to disentangle the influence effects from other processes, and the simulation evidence has shown that this approach outperforms other state-of-the-art approaches in terms of recovering the true social influence effect when there is an unobserved trait co-determining influence and selection. In this paper, I will further illustrate how the latent space adjusted approach can account for bias in the estimation of social influence effects and how this approach can be easily implemented in R.</p>
</div>
</a>
<a href="articles/RJ-2021-056/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 20, 2019</div>
<div class="dt-authors">
<div class="dt-author">Mark P.J. van der Loo</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>A Method for Deriving Information from Running R Code</h2>
<div class="dt-tags"></div>
<p>It is often useful to tap information from a running R script. Obvious use cases include
 monitoring the consumption of resources (time, memory) and logging. Perhaps less obvious cases
 include tracking changes in R objects or collecting the output of unit tests. In this paper, we demonstrate
 an approach that abstracts the collection and processing of such secondary information from the
 running R script. Our approach is based on a combination of three elements. The first element is
 to build a customized way to evaluate code. The second is labeled local masking and it involves
 temporarily masking a user-facing function so an alternative version of it is called. The third element
 we label local side effect. This refers to the fact that the masking function exports information to the
 secondary information flow without altering a global state. The result is a method for building systems
 in pure R that lets users create and control secondary flows of information with minimal impact on
 their workflow and no global side effects.</p>
</div>
</a>
<a href="articles/RJ-2021-064/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">Oct. 20, 2019</div>
<div class="dt-authors">
<div class="dt-author">Daniel Osorio</div>
<div class="dt-author">Kelly Botero</div>
<div class="dt-author">Andrs Pinzn Velasco</div>
<div class="dt-author">Nicols Mendoza-Meja</div>
<div class="dt-author">Felipe Rojas-            Rodrguez</div>
<div class="dt-author">George Barreto</div>
<div class="dt-author">Janneth Gonzlez</div>
</div>
</div>
<div class="thumbnail">
<img/>
</div>
<div class="description">
<h2>g2f as a Novel Tool to Find and Fill Gaps in Metabolic Networks</h2>
<div class="dt-tags"></div>
<p>During the building of a genome-scale metabolic model, there are several dead-end metabo lites and substrates which cannot be imported, produced, nor used by any reaction incorporated in the network. The presence of these dead-end metabolites can block out the net flux of the objective function when it is evaluated through Flux Balance Analysis (FBA), and when it is not blocked, bias in the biological conclusions increase. In this aspect, the refinement to restore the connectivity of the network can be carried out manually or using computational algorithms. The g2f package was designed as a tool to find the gaps from dead-end metabolites and fill them from the stoichiometric reactions of a reference, filtering candidate reactions using a weighting function. Additionally, this algorithm allows downloading all the sets of gene-associated stoichiometric reactions for a specific organism from the KEGG database. Our package is compatible with both 4.0.0 and 3.6.0 R versions.</p>
</div>
</a>
<a href="articles/RJ-2021-027/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">July 19, 2019</div>
<div class="dt-authors">
<div class="dt-author">Ana Lpez-Cheda</div>
<div class="dt-author">M. Amalia Jcome</div>
<div class="dt-author">Ignacio Lpez-de-Ullibarri</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-027/feature.png"/>
</div>
<div class="description">
<h2>npcure: An R Package for Nonparametric Inference in Mixture Cure Models</h2>
<div class="dt-tags"></div>
<p>Mixture cure models have been widely used to analyze survival data with a cure fraction.
 They assume that a subgroup of the individuals under study will never experience the event (cured
 subjects). So, the goal is twofold: to study both the cure probability and the failure time of the
 uncured individuals through a proper survival function (latency). The R package npcure implements a
 completely nonparametric approach for estimating these functions in mixture cure models, considering
 right-censored survival times. Nonparametric estimators for the cure probability and the latency as
 functions of a covariate are provided. Bootstrap bandwidth selectors for the estimators are included.
 The package also implements a nonparametric covariate significance test for the cure probability,
 which can be applied with a continuous, discrete, or qualitative covariate.</p>
</div>
</a>
<a href="articles/RJ-2021-026/" class="post-preview">
<script class="post-metadata" type="text/json">{"categories":[]}</script>
<div class="metadata">
<div class="publishedDate">July 18, 2019</div>
<div class="dt-authors">
<div class="dt-author">Bo-Young Kim</div>
<div class="dt-author">Yunju Im</div>
<div class="dt-author">Jae Keun Yoo</div>
</div>
</div>
<div class="thumbnail">
<img data-src="articles/RJ-2021-026/feature.png"/>
</div>
<div class="description">
<h2>SEEDCCA: An Integrated R-Package for Canonical Correlation Analysis and Partial Least Squares</h2>
<div class="dt-tags"></div>
<p>Canonical correlation analysis (CCA) has a long history as an explanatory statistical method
 in high-dimensional data analysis and has been successfully applied in many scientific fields such as
 chemometrics, pattern recognition, genomic sequence analysis, and so on. The so-called seedCCA is a
 newly developed R package that implements not only the standard and seeded CCA but also partial
 least squares. The package enables us to fit CCA to large-p and small-n data. The paper provides a
 complete guide. Also, the seeded CCA application results are compared with the regularized CCA in
 the existing R package. It is believed that the package, along with the paper, will contribute to high
dimensional data analysis in various science field practitioners and that the statistical methodologies
 in multivariate analysis become more fruitful.</p>
</div>
</a>
</div>
<div class="posts-more">
<a href="#">More articles &raquo;</a>
</div>
</div>
<!--/radix_placeholder_article_listing-->

<div class="d-title">
<h1>All articles</h1>
<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>All articles accepted into the R Journal.</p></p>
</div>


<div class="d-article">
<div class="sourceCode" id="cb1"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>


<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom"></div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<div class="distill-site-nav distill-site-footer">
<p> The R Foundation, <a href="mailto:r-journal@r-project.org">web page contact</a>.</p>
</div>
<!--/radix_placeholder_navigation_after_body-->


</body>

</html>
