<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
<title>The R Journal: TreeSearch: Morphological Phylogenetic Analysis in R</title>

<meta property="description" itemprop="description" content="TreeSearch is an R package for phylogenetic analysis, optimized for discrete character data. Tree search may be conducted using equal or implied step weights with an explicit (albeit inexact) allowance for inapplicable character entries, avoiding some of the pitfalls inherent in standard parsimony methods. Profile parsimony and user-specified optimality criteria are supported.&#10;A graphical interface, which requires no familiarity with R, is designed to help a user to improve the quality of datasets through critical review of underpinning character codings; and to obtain additional information from results by identifying and summarizing clusters of similar trees, mapping the distribution of trees, and removing &#39;rogue&#39; taxa that obscure underlying relationships.&#10;Taken together, the package aims to support methodological rigour at each step of data collection, analysis, and the exploration of phylogenetic results."/>

<link rel="canonical" href="https://doi.org/10.32614/RJ-2023-019/"/>
<link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>
<link rel="icon" type="image/vnd.microsoft.icon" href="../../resources/favicon.ico"/>

<!--  https://schema.org/Article -->
<meta property="article:published" itemprop="datePublished" content="2023-02-10"/>
<meta property="article:created" itemprop="dateCreated" content="2023-02-10"/>
<meta name="article:author" content="Martin R. Smith"/>

<!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
<meta property="og:title" content="The R Journal: TreeSearch: Morphological Phylogenetic Analysis in R"/>
<meta property="og:type" content="article"/>
<meta property="og:description" content="TreeSearch is an R package for phylogenetic analysis, optimized for discrete character data. Tree search may be conducted using equal or implied step weights with an explicit (albeit inexact) allowance for inapplicable character entries, avoiding some of the pitfalls inherent in standard parsimony methods. Profile parsimony and user-specified optimality criteria are supported.&#10;A graphical interface, which requires no familiarity with R, is designed to help a user to improve the quality of datasets through critical review of underpinning character codings; and to obtain additional information from results by identifying and summarizing clusters of similar trees, mapping the distribution of trees, and removing &#39;rogue&#39; taxa that obscure underlying relationships.&#10;Taken together, the package aims to support methodological rigour at each step of data collection, analysis, and the exploration of phylogenetic results."/>
<meta property="og:url" content="https://doi.org/10.32614/RJ-2023-019/"/>
<meta property="og:image" content="https://journal.r-project.org/articles/RJ-2023-019/Flow.svg"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="The R Journal"/>

<!--  https://dev.twitter.com/cards/types/summary -->
<meta property="twitter:card" content="summary_large_image"/>
<meta property="twitter:title" content="The R Journal: TreeSearch: Morphological Phylogenetic Analysis in R"/>
<meta property="twitter:description" content="TreeSearch is an R package for phylogenetic analysis, optimized for discrete character data. Tree search may be conducted using equal or implied step weights with an explicit (albeit inexact) allowance for inapplicable character entries, avoiding some of the pitfalls inherent in standard parsimony methods. Profile parsimony and user-specified optimality criteria are supported.&#10;A graphical interface, which requires no familiarity with R, is designed to help a user to improve the quality of datasets through critical review of underpinning character codings; and to obtain additional information from results by identifying and summarizing clusters of similar trees, mapping the distribution of trees, and removing &#39;rogue&#39; taxa that obscure underlying relationships.&#10;Taken together, the package aims to support methodological rigour at each step of data collection, analysis, and the exploration of phylogenetic results."/>
<meta property="twitter:url" content="https://doi.org/10.32614/RJ-2023-019/"/>
<meta property="twitter:image" content="https://journal.r-project.org/articles/RJ-2023-019/Flow.svg"/>

<!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
<meta name="citation_title" content="The R Journal: TreeSearch: Morphological Phylogenetic Analysis in R"/>
<meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2023-019"/>
<meta name="citation_pdf_url" content="RJ-2023-019.pdf"/>
<meta name="citation_volume" content="14"/>
<meta name="citation_issue" content="4"/>
<meta name="citation_doi" content="10.32614/RJ-2023-019"/>
<meta name="citation_journal_title" content="The R Journal"/>
<meta name="citation_issn" content="2073-4859"/>
<meta name="citation_firstpage" content="305"/>
<meta name="citation_lastpage" content="315"/>
<meta name="citation_fulltext_world_readable" content=""/>
<meta name="citation_online_date" content="2023/02/10"/>
<meta name="citation_publication_date" content="2023/02/10"/>
<meta name="citation_author" content="Martin R. Smith"/>
<meta name="citation_author_institution" content="University of Durham"/>
<!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Profile parsimony (PP): An analysis under implied weights (IW);citation_volume=20;citation_doi=10.1111/j.1096-0031.2003.00001.x;citation_author=J. Salvador Arias;citation_author=Daniel Rafael Miranda-Esquivel"/>
  <meta name="citation_reference" content="citation_title=K-means++: The advantages of careful seeding;citation_publisher=Society for Industrial and Applied Mathematics;citation_author=David Arthur;citation_author=Sergei Vassilvitskii"/>
  <meta name="citation_reference" content="citation_title=Phylogenetic signal and bias in paleontology;citation_volume=71;citation_doi=10.1093/sysbio/syab072;citation_author=Robert J. Asher;citation_author=Martin R. Smith"/>
  <meta name="citation_reference" content="citation_title=Landscapes on spaces of trees;citation_volume=131;citation_doi=10.1016/S0096-3003(01)00164-3;citation_author=Oliver Bastert;citation_author=Dan Rockmore;citation_author=Peter F. Stadler;citation_author=Gottfried Tinhofer"/>
  <meta name="citation_reference" content="citation_title=Hierarchical clustering with prototypes via minimax linkage;citation_volume=106;citation_doi=10.1198/jasa.2011.tm10183;citation_author=Jacob Bien;citation_author=Robert Tibshirani"/>
  <meta name="citation_reference" content="citation_title=Problematic character coding methods in morphology and their effects;citation_volume=104;citation_doi=10.1111/j.1095-8312.2011.01755.x;citation_author=Martin D. Brazeau"/>
  <meta name="citation_reference" content="citation_title=MorphyLib: A library for phylogenetic analysis of categorical trait data with inapplicability;citation_doi=10.5281/zenodo.815372;citation_author=Martin D. Brazeau;citation_author=Martin R. Smith;citation_author=Thomas Guillerme"/>
  <meta name="citation_reference" content="citation_title=An algorithm for morphological phylogenetic analysis with inapplicable data;citation_volume=68;citation_doi=10.1093/sysbio/syy083;citation_author=Martin D. Brazeau;citation_author=Thomas Guillerme;citation_author=Martin R. Smith"/>
  <meta name="citation_reference" content="citation_title=Bayesian and likelihood phylogenetic reconstructions of morphological traits are not discordant when taking uncertainty into consideration: A comment on Puttick &lt;i&gt;et al.&lt;/i&gt;;citation_volume=284;citation_doi=10.1098/rspb.2017.0986;citation_author=Joseph W. Brown;citation_author=Caroline Parins-Fukuchi;citation_author=Gregory W. Stull;citation_author=Oscar M. Vargas;citation_author=Stephen A. Smith"/>
  <meta name="citation_reference" content="citation_title=On the distribution of lengths of evolutionary trees;citation_volume=3;citation_doi=10.1137/0403005;citation_author=M. Carter;citation_author=M. Hendy;citation_author=D. Penny;citation_author=L. A. Székely;citation_author=N. C. Wormald"/>
  <meta name="citation_reference" content="citation_title=Parsimony and the problem of inapplicables in sequence data;citation_publisher=Oxford University Press;citation_doi=10.1093/acprof:oso/9780199297306.003.0006;citation_author=Jan E. De Laet"/>
  <meta name="citation_reference" content="citation_title=Distinguishing heat from light in debate over controversial fossils;citation_volume=31;citation_doi=10.1002/bies.200800128;citation_author=Philip C. J. Donoghue;citation_author=Mark A. Purnell"/>
  <meta name="citation_reference" content="citation_title=Comparison of undirected phylogenetic trees based on subtrees of four evolutionary units;citation_volume=34;citation_doi=10.2307/sysbio/34.2.193;citation_author=George F. Estabrook;citation_author=F. R. McMorris;citation_author=Christopher A. Meacham"/>
  <meta name="citation_reference" content="citation_title=Towards an inclusive philosophy for phylogenetic inference;citation_volume=50;citation_doi=10.1080/10635150118627;citation_author=Daniel P. Faith;citation_author=John W. H. Trueman"/>
  <meta name="citation_reference" content="citation_title=Toward defining the course of evolution: Minimum change for a specific tree topology;citation_volume=20;citation_doi=10.1093/sysbio/20.4.406;citation_author=Walter M. Fitch"/>
  <meta name="citation_reference" content="citation_title=Progressive Tree Neighborhood applied to the Maximum Parsimony problem;citation_volume=5;citation_doi=10.1109/TCBB.2007.1065;citation_author=A. Goeffon;citation_author=J.-M. Richer;citation_author=Progressive Tree Neighborhood applied to the Maximum Parsimony problem"/>
  <meta name="citation_reference" content="citation_title=Estimating character weights during tree search;citation_publisher=Blackwell Publishing Ltd;citation_volume=9;citation_doi=10.1111/j.1096-0031.1993.tb00209.x;citation_author=Pablo A. Goloboff"/>
  <meta name="citation_reference" content="citation_title=Weighting against homoplasy improves phylogenetic analysis of morphological data sets;citation_volume=24;citation_doi=10.1111/j.1096-0031.2008.00209.x;citation_author=Pablo A. Goloboff;citation_author=James M. Carpenter;citation_author=J. Salvador Arias;citation_author=Daniel Rafael Miranda Esquivel"/>
  <meta name="citation_reference" content="citation_title=Extended implied weighting;citation_volume=30;citation_doi=10.1111/cla.12047;citation_author=Pablo A. Goloboff"/>
  <meta name="citation_reference" content="citation_title=Weighted parsimony outperforms other methods of phylogenetic inference under models appropriate for morphology;citation_volume=34;citation_doi=10.1111/cla.12205;citation_author=Pablo A. Goloboff;citation_author=Ambrosio Torres;citation_author=J. Salvador Arias"/>
  <meta name="citation_reference" content="citation_title=Parsimony and model-based phylogenetic methods for morphological data: Comments on O’Reilly &lt;i&gt;et al&lt;/i&gt;.;citation_volume=61;citation_doi=10.1111/pala.12353;citation_author=Pablo A. Goloboff;citation_author=Ambrosio Torres Galvis;citation_author=J. Salvatdor Arias"/>
  <meta name="citation_reference" content="citation_title=A reconsideration of inapplicable characters, and an approximation with step-matrix recoding;citation_volume=37;citation_doi=10.1111/cla.12456;citation_author=Pablo A. Goloboff;citation_author=Jan De Laet;citation_author=Duniesky Ríos-Tamayo;citation_author=Claudia A. Szumik"/>
  <meta name="citation_reference" content="citation_title=Comparative cladistics: Identifying the sources for differing phylogenetic results between competing morphology-based datasets;citation_doi=10.1080/14772019.2021.1970038;citation_author=Pablo A. Goloboff;citation_author=Paul C. Sereno"/>
  <meta name="citation_reference" content="citation_title=Some distance properties of latent root and vector methods used in multivariate analysis;citation_volume=53;citation_doi=10.2307/2333639;citation_author=J. C. Gower"/>
  <meta name="citation_reference" content="citation_title=Minimum spanning trees and single linkage cluster analysis;citation_publisher=[Wiley, Royal Statistical Society];citation_volume=18;citation_doi=10.2307/2346439;citation_author=J. C. Gower;citation_author=G. J. S. Ross"/>
  <meta name="citation_reference" content="citation_title=Does time matter in phylogeny? A perspective from the fossil record;citation_doi=10.1101/2021.06.11.445746;citation_author=Pauline Guenser;citation_author=Rachel C. M. Warnock;citation_author=Walker Pett;citation_author=Philip C. J. Donoghue;citation_author=Emilia Jarochowska"/>
  <meta name="citation_reference" content="citation_title=Algorithm AS 136: A &lt;i&gt;K&lt;/i&gt;-means clustering algorithm;citation_publisher=[Wiley, Royal Statistical Society];citation_volume=28;citation_doi=10.2307/2346830;citation_author=J. A. Hartigan;citation_author=M. A. Wong"/>
  <meta name="citation_reference" content="citation_title=Phylogenetic systematics;citation_publisher=The University of Illinois Press;citation_author=Willi Hennig"/>
  <meta name="citation_reference" content="citation_title=Analysis and visualization of tree space;citation_volume=54;citation_doi=10.1080/10635150590946961;citation_author=David M. Hillis;citation_author=Tracy A. Heath;citation_author=Katherine St. John"/>
  <meta name="citation_reference" content="citation_title=Incorporating hierarchical characters into phylogenetic analysis;citation_volume=70;citation_doi=10.1093/sysbio/syab005;citation_author=Melanie J. Hopkins;citation_author=Katherine St. John"/>
  <meta name="citation_reference" content="citation_title=Bilaterian phylogeny and uncritical recycling of morphological data sets;citation_volume=50;citation_doi=10.1080/106351501753328857;citation_author=Ronald A. Jenner"/>
  <meta name="citation_reference" content="citation_title=Trustworthiness and metrics in visualizing similarity of gene expression;citation_volume=4;citation_doi=10.1186/1471-2105-4-48;citation_author=Samuel Kaski;citation_author=Janne Nikkilä;citation_author=Merja Oja;citation_author=Jarkko Venna;citation_author=Petri Törönen;citation_author=Eero Castrén"/>
  <meta name="citation_reference" content="citation_title=Partitioning around medoids (Program PAM);citation_publisher=John Wiley &amp; Sons, Ltd;citation_author=Leonard Kaufman;citation_author=Peter J. Rousseeuw"/>
  <meta name="citation_reference" content="citation_title=Fragmentary taxa, missing data, and ambiguity: Mistaken assumptions and conclusions;citation_volume=51;citation_doi=10.1080/10635150252899824;citation_author=Maureen Kearney"/>
  <meta name="citation_reference" content="citation_title=Illustrating phylogenetic placement of fossils using RoguePlots: An example from ichneumonid parasitoid wasps (Hymenoptera, Ichneumonidae) and an extensive morphological matrix;citation_publisher=Public Library of Science;citation_volume=14;citation_doi=10.1371/journal.pone.0212942;citation_author=Seraina Klopfstein;citation_author=Tamara Spasojevic"/>
  <meta name="citation_reference" content="citation_title=Death is on our side: Paleontological data drastically modify phylogenetic hypotheses;citation_volume=69;citation_doi=10.1093/sysbio/syaa023;citation_author=Nicolás Mongiardino Koch;citation_author=Luke A. Parry"/>
  <meta name="citation_reference" content="citation_title=Null models for the number of evolutionary steps in a character on a phylogenetic tree;citation_volume=45;citation_doi=10.1111/j.1558-5646.1991.tb04385.x;citation_author=Wayne P. Maddison;citation_author=M. Slatkin"/>
  <meta name="citation_reference" content="citation_title=Missing data versus missing characters in phylogenetic analysis;citation_volume=42;citation_doi=10.1093/sysbio/42.4.576;citation_author=Wayne P. Maddison"/>
  <meta name="citation_reference" content="citation_title=NEXUS: An extensible file format for systematic information;citation_volume=46;citation_doi=10.1093/sysbio/46.4.590;citation_author=David R. Maddison;citation_author=David L. Swofford;citation_author=Wayne P. Maddison"/>
  <meta name="citation_reference" content="citation_title=Cluster: Cluster Analysis Basics and Extensions;citation_volume=2.1.0;citation_author=Martin Maechler;citation_author=Peter Rousseeuw;citation_author=Anja Struyf;citation_author=Mia Hubert;citation_author=Kurt Hornik"/>
  <meta name="citation_reference" content="citation_title=New methods to calculate concordance factors for phylogenomic datasets;citation_volume=37;citation_doi=10.1093/molbev/msaa106;citation_author=Bui Quang Minh;citation_author=Matthew W. Hahn;citation_author=Robert Lanfear"/>
  <meta name="citation_reference" content="citation_title=Hennig’s auxiliary principle and reciprocal illumination revisited;citation_publisher=Cambridge University Press;citation_doi=10.1017/CBO9781316338797.013;citation_author=Randall D. Mooi;citation_author=Anthony C. Gill"/>
  <meta name="citation_reference" content="citation_title=A survey of recent advances in hierarchical clustering algorithms;citation_volume=26;citation_doi=10.1093/comjnl/26.4.354;citation_author=F. Murtagh"/>
  <meta name="citation_reference" content="citation_title=The Parsimony Ratchet, a new method for rapid parsimony analysis;citation_volume=15;citation_doi=10.1111/j.1096-0031.1999.tb00277.x;citation_author=Kevin C. Nixon"/>
  <meta name="citation_reference" content="citation_title=MorphoBank: Phylophenomics in the “cloud”;citation_volume=27;citation_doi=10.1111/j.1096-0031.2011.00355.x;citation_author=Maureen A. O’Leary;citation_author=Seth Kaufman"/>
  <meta name="citation_reference" content="citation_title=Bayesian methods outperform parsimony but at the expense of precision in the estimation of phylogeny from discrete morphological data;citation_volume=12;citation_doi=10.1098/rsbl.2016.0081;citation_author=Joseph E. O’Reilly;citation_author=Mark N. Puttick;citation_author=Luke Parry;citation_author=Alastair R. Tanner;citation_author=James E. Tarver;citation_author=James Fleming;citation_author=Davide Pisani;citation_author=Philip C. J. Donoghue"/>
  <meta name="citation_reference" content="citation_title=Ape 5.0: An environment for modern phylogenetics and evolutionary analyses in R;citation_volume=35;citation_doi=10.1093/bioinformatics/bty633;citation_author=Emmanuel Paradis;citation_author=Klaus Schliep"/>
  <meta name="citation_reference" content="citation_title=Unstable taxa in cladistic analysis: Identification and the assessment of relevant characters;citation_volume=25;citation_doi=10.1111/j.1096-0031.2009.00258.x;citation_author=Diego Pol;citation_author=Ignacio H. Escapa"/>
  <meta name="citation_reference" content="citation_title=Uncertain-tree: Discriminating among competing approaches to the phylogenetic analysis of phenotype data;citation_volume=284;citation_doi=10.1098/rspb.2016.2290;citation_author=Mark N. Puttick;citation_author=Joseph E. O’Reilly;citation_author=Alastair R. Tanner;citation_author=James F. Fleming;citation_author=James Clark;citation_author=Lucy Holloway;citation_author=Jesus Lozano-Fernandez;citation_author=Luke A. Parry;citation_author=James E. Tarver;citation_author=Davide Pisani;citation_author=Philip C. J. Donoghue"/>
  <meta name="citation_reference" content="citation_title=Comparison of phylogenetic trees;citation_volume=53;citation_doi=10.1016/0025-5564(81)90043-2;citation_author=David F. Robinson;citation_author=Leslie R. Foulds"/>
  <meta name="citation_reference" content="citation_title=tqDist: A library for computing the quartet and triplet distances between binary or general trees;citation_volume=30;citation_doi=10.1093/bioinformatics/btu157;citation_author=Andreas Sand;citation_author=Morten K. Holt;citation_author=Jens Johansen;citation_author=Gerth Stølting Brodal;citation_author=Thomas Mailund;citation_author=Christian N. S. Pedersen"/>
  <meta name="citation_reference" content="citation_title=Parsimony, not Bayesian analysis, recovers more stratigraphically congruent phylogenetic trees;citation_volume=14;citation_doi=10.1098/rsbl.2018.0263;citation_author=Robert S. Sansom;citation_author=Peter G. Choate;citation_author=Joseph N. Keating;citation_author=Emma Randle"/>
  <meta name="citation_reference" content="citation_title=Phangorn: Phylogenetic analysis in R;citation_volume=27;citation_doi=10.1093/bioinformatics/btq706;citation_author=Klaus Peter Schliep"/>
  <meta name="citation_reference" content="citation_title=Logical basis for morphological characters in phylogenetics;citation_volume=23;citation_doi=10.1111/j.1096-0031.2007.00161.x;citation_author=Paul C. Sereno"/>
  <meta name="citation_reference" content="citation_title=Comparative cladistics;citation_volume=25;citation_doi=10.1111/j.1096-0031.2009.00265.x;citation_author=Paul C. Sereno"/>
  <meta name="citation_reference" content="citation_title=Shiny: Web application framework for R;citation_author=Winston Chang;citation_author=Joe Cheng;citation_author=J. J. Allaire;citation_author=Carson Sievert;citation_author=Barret Schloerke;citation_author=Yi-Hui Xie;citation_author=Jeff Allen;citation_author=Jonathan McPherson;citation_author=Alan Dipert;citation_author=Barbara Borges"/>
  <meta name="citation_reference" content="citation_title=Shinyjs: Easily improve the user experience of your shiny apps in seconds;citation_author=Dean Attali"/>
  <meta name="citation_reference" content="citation_title=Giant taxon-character matrices: Quality of character constructions remains critical regardless of size;citation_volume=33;citation_doi=10.1111/cla.12163;citation_author=Tiago R. Simões;citation_author=Michael W. Caldwell;citation_author=Alessandro Palci;citation_author=Randall L. Nydam"/>
  <meta name="citation_reference" content="citation_title=Phyutility: A phyloinformatics tool for trees, alignments and molecular data;citation_volume=24;citation_doi=10.1093/bioinformatics/btm619;citation_author=Stephen A. Smith;citation_author=Casey W. Dunn"/>
  <meta name="citation_reference" content="citation_title=Bayesian and parsimony approaches reconstruct informative trees from simulated morphological datasets;citation_volume=15;citation_doi=10.1098/rsbl.2018.0632;citation_author=Martin R. Smith"/>
  <meta name="citation_reference" content="citation_title=Information theoretic Generalized Robinson–Foulds metrics for comparing phylogenetic trees;citation_volume=36;citation_doi=10.1093/bioinformatics/btaa614;citation_author=Martin R. Smith"/>
  <meta name="citation_reference" content="citation_title=Quartet: Comparison of phylogenetic trees using quartet and bipartition measures;citation_doi=10.5281/zenodo.2536318;citation_author=Martin R. Smith"/>
  <meta name="citation_reference" content="citation_title=Using information theory to detect rogue taxa and improve consensus trees;citation_volume=71;citation_doi=10.1093/sysbio/syab099;citation_author=Martin R. Smith"/>
  <meta name="citation_reference" content="citation_title=Robust analysis of phylogenetic tree space;citation_volume=71;citation_doi=10.1093/sysbio/syab100;citation_author=Martin R. Smith"/>
  <meta name="citation_reference" content="citation_title=Statistically based postprocessing of phylogenetic analysis by clustering;citation_volume=18;citation_doi=10.1093/bioinformatics/18.suppl_1.S285;citation_author=C. Stockham;citation_author=L.-S. Wang;citation_author=T. Warnow"/>
  <meta name="citation_reference" content="citation_title=Integration of anatomy ontologies and evo-devo using structured Markov models suggests a new framework for modeling discrete phenotypic traits;citation_volume=68;citation_doi=10.1093/sysbio/syz005;citation_author=Sergei Tarasov"/>
  <meta name="citation_reference" content="citation_title=New phylogenetic Markov models for inapplicable morphological characters;citation_doi=10.1101/2021.04.26.441495;citation_author=Sergei Tarasov"/>
  <meta name="citation_reference" content="citation_title=The information content of consensus trees;citation_publisher=Springer;citation_doi=10.1007/978-3-642-72253-0_12;citation_author=Joseph L. Thorley;citation_author=Mark Wilkinson;citation_author=Mike Charleston"/>
  <meta name="citation_reference" content="citation_title=TreeDist: Calculate and map distances between phylogenetic trees;citation_doi=10.5281/zenodo.3528123;citation_author=Martin R. Smith"/>
  <meta name="citation_reference" content="citation_title=TreeTools: Create, modify and analyse phylogenetic trees;citation_doi=10.5281/zenodo.3522725;citation_author=Martin R. Smith"/>
  <meta name="citation_reference" content="citation_title=Neighborhood preservation in nonlinear projection methods: An experimental study;citation_publisher=Springer;citation_doi=10.1007/3-540-44668-0_68;citation_author=Jarkko Venna;citation_author=Samuel Kaski"/>
  <meta name="citation_reference" content="citation_title=Machaeridians are Palaeozoic armoured annelids;citation_volume=451;citation_doi=10.1038/nature06474;citation_author=Jakob Vinther;citation_author=Peter Van Roy;citation_author=Derek E. G. Briggs"/>
  <meta name="citation_reference" content="citation_title=The prevalence of multifurcations in tree-space and their implications for tree-search;citation_volume=27;citation_doi=10.1093/molbev/msq163;citation_author=Simon Whelan;citation_author=Daniel Money"/>
  <meta name="citation_reference" content="citation_title=The role of morphological data in phylogeny reconstruction;citation_volume=53;citation_doi=10.1080/10635150490472959;citation_author=John J. Wiens"/>
  <meta name="citation_reference" content="citation_title=Common cladistic information and its consensus representation: Reduced Adams and reduced cladistic consensus trees and profiles;citation_volume=43;citation_doi=10.2307/2413673;citation_author=Mark Wilkinson"/>
  <meta name="citation_reference" content="citation_title=A comparison of two methods of character construction;citation_volume=11;citation_doi=10.1111/j.1096-0031.1995.tb00091.x;citation_author=Mark Wilkinson"/>
  <meta name="citation_reference" content="citation_title=Majority-rule reduced consensus trees and their use in bootstrapping;citation_volume=13;citation_doi=10.1093/oxfordjournals.molbev.a025604;citation_author=Mark Wilkinson"/>
  <meta name="citation_reference" content="citation_title=Missing entries and multiple trees: Instability, relationships, and support in parsimony analysis;citation_publisher=Taylor &amp; Francis;citation_volume=23;citation_doi=10.1671/0272-4634(2003)023[0311:MEAMTI]2.0.CO;2;citation_author=Mark Wilkinson"/>
  <meta name="citation_reference" content="citation_title=The effect of combining molecular and morphological data in published phylogenetic analyses;citation_volume=55;citation_doi=10.1080/10635150600899798;citation_author=A. H. Wortley;citation_author=Robert W. Scotland"/>
  <meta name="citation_reference" content="citation_title=Bayesian analyses in phylogenetic palaeontology: Interpreting the posterior sample;citation_volume=63;citation_doi=10.1111/pala.12500;citation_author=April M. Wright;citation_author=Graeme T. Lloyd"/>
  <!--radix_placeholder_rmarkdown_metadata-->

<script type="text/json" id="radix-rmarkdown-metadata">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","author","editor_options","description","draft","type","output","bibliography","date","date_received","volume","issue","slug","journal","pdf_url","citation_url","doi","creative_commons","packages","CTV","csl","canonical_url"]}},"value":[{"type":"character","attributes":{},"value":["TreeSearch: Morphological Phylogenetic Analysis in R"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","url","orcid_id","email"]}},"value":[{"type":"character","attributes":{},"value":["Martin R. Smith"]},{"type":"character","attributes":{},"value":["University of Durham"]},{"type":"character","attributes":{},"value":["Department of Earth Sciences","Durham, UK","DH1 3LE"]},{"type":"character","attributes":{},"value":["https://smithlabdurham.github.io/"]},{"type":"character","attributes":{},"value":["0000-0001-5660-1727"]},{"type":"character","attributes":{},"value":["martin.smith@durham.ac.uk"]}]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["markdown"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["wrap"]}},"value":[{"type":"integer","attributes":{},"value":[80]}]}]},{"type":"character","attributes":{},"value":["TreeSearch is an R package for phylogenetic analysis, optimized for discrete character data. Tree search may be conducted using equal or implied step weights with an explicit (albeit inexact) allowance for inapplicable character entries, avoiding some of the pitfalls inherent in standard parsimony methods. Profile parsimony and user-specified optimality criteria are supported.\nA graphical interface, which requires no familiarity with R, is designed to help a user to improve the quality of datasets through critical review of underpinning character codings; and to obtain additional information from results by identifying and summarizing clusters of similar trees, mapping the distribution of trees, and removing 'rogue' taxa that obscure underlying relationships.\nTaken together, the package aims to support methodological rigour at each step of data collection, analysis, and the exploration of phylogenetic results."]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["package"]},{"type":"character","attributes":{},"value":["rjtools::rjournal_web_article"]},{"type":"character","attributes":{},"value":["smith.bib"]},{"type":"character","attributes":{},"value":["2023-02-10"]},{"type":"character","attributes":{},"value":["2022-02-24"]},{"type":"integer","attributes":{},"value":[14]},{"type":"integer","attributes":{},"value":[4]},{"type":"character","attributes":{},"value":["RJ-2023-019"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"integer","attributes":{},"value":[305]},{"type":"integer","attributes":{},"value":[315]}]},{"type":"character","attributes":{},"value":["RJ-2023-019.pdf"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2023-019"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2023-019"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"character","attributes":{},"value":["TreeSearch","phangorn","ape","Quartet","Rogue","shiny","shinyjs","TreeDist","TreeTools"]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["Environmetrics","Phylogenetics","WebTechnologies"]},{"type":"character","attributes":{},"value":["/Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library/rjtools/rjournal.csl"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2023-019/"]}]}
</script>
<!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["2022-53-review-response.html","Flow.pdf","Flow.svg","RJ-2023-019_files/anchor-4.2.2/anchor.min.js","RJ-2023-019_files/bowser-1.9.3/bowser.min.js","RJ-2023-019_files/distill-2.2.21/template.v2.js","RJ-2023-019_files/figure-html5/map-results-1.png","RJ-2023-019_files/header-attrs-2.20/header-attrs.js","RJ-2023-019_files/jquery-3.6.0/jquery-3.6.0.js","RJ-2023-019_files/jquery-3.6.0/jquery-3.6.0.min.js","RJ-2023-019_files/jquery-3.6.0/jquery-3.6.0.min.map","RJ-2023-019_files/popper-2.6.0/popper.min.js","RJ-2023-019_files/tippy-6.2.7/tippy-bundle.umd.min.js","RJ-2023-019_files/tippy-6.2.7/tippy-light-border.css","RJ-2023-019_files/tippy-6.2.7/tippy.css","RJ-2023-019_files/tippy-6.2.7/tippy.umd.min.js","RJ-2023-019_files/webcomponents-2.0.0/webcomponents.js","RJ-2023-019.log","RJ-2023-019.pdf","RJ-2023-019.tex","RJ-2023-019.zip","RJtemplate.tex","RJwrapper.log","RJwrapper.tex","Rlogo-5.png","RoguePlot.png","smith_files/anchor-4.2.2/anchor.min.js","smith_files/bowser-1.9.3/bowser.min.js","smith_files/distill-2.2.21/template.v2.js","smith_files/figure-html5/map-results-1.png","smith_files/header-attrs-2.11/header-attrs.js","smith_files/header-attrs-2.14/header-attrs.js","smith_files/header-attrs-2.16/header-attrs.js","smith_files/header-attrs-2.19/header-attrs.js","smith_files/jquery-3.6.0/jquery-3.6.0.js","smith_files/jquery-3.6.0/jquery-3.6.0.min.js","smith_files/jquery-3.6.0/jquery-3.6.0.min.map","smith_files/popper-2.6.0/popper.min.js","smith_files/tippy-6.2.7/tippy-bundle.umd.min.js","smith_files/tippy-6.2.7/tippy-light-border.css","smith_files/tippy-6.2.7/tippy.css","smith_files/tippy-6.2.7/tippy.umd.min.js","smith_files/webcomponents-2.0.0/webcomponents.js","smith.bib","smith.tex","TreeSpace.pdf","TreeSpace.png","TreeSpace.svg","WillsCons.pdf","WillsCons.svg"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
<meta name="distill:offset" content="../.."/>

<script type="application/javascript">

  window.headroom_prevent_pin = false;

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // initialize headroom for banner
    var header = $('header').get(0);
    var headerHeight = header.offsetHeight;
    var headroom = new Headroom(header, {
      tolerance: 5,
      onPin : function() {
        if (window.headroom_prevent_pin) {
          window.headroom_prevent_pin = false;
          headroom.unpin();
        }
      }
    });
    headroom.init();
    if(window.location.hash)
      headroom.unpin();
    $(header).addClass('headroom--transition');

    // offset scroll location for banner on hash change
    // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
    window.addEventListener("hashchange", function(event) {
      window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
    });

    // responsive menu
    $('.distill-site-header').each(function(i, val) {
      var topnav = $(this);
      var toggle = topnav.find('.nav-toggle');
      toggle.on('click', function() {
        topnav.toggleClass('responsive');
      });
    });

    // nav dropdowns
    $('.nav-dropbtn').click(function(e) {
      $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
      $(this).parent().siblings('.nav-dropdown')
         .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $("body").click(function(e){
      $('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $(".nav-dropdown").click(function(e){
      e.stopPropagation();
    });
  });
</script>

<style type="text/css">

/* Theme (user-documented overrideables for nav appearance) */

.distill-site-nav {
  color: rgba(255, 255, 255, 0.8);
  background-color: #0F2E3D;
  font-size: 15px;
  font-weight: 300;
}

.distill-site-nav a {
  color: inherit;
  text-decoration: none;
}

.distill-site-nav a:hover {
  color: white;
}

@media print {
  .distill-site-nav {
    display: none;
  }
}

.distill-site-header {

}

.distill-site-footer {

}


/* Site Header */

.distill-site-header {
  width: 100%;
  box-sizing: border-box;
  z-index: 3;
}

.distill-site-header .nav-left {
  display: inline-block;
  margin-left: 8px;
}

@media screen and (max-width: 768px) {
  .distill-site-header .nav-left {
    margin-left: 0;
  }
}


.distill-site-header .nav-right {
  float: right;
  margin-right: 8px;
}

.distill-site-header a,
.distill-site-header .title {
  display: inline-block;
  text-align: center;
  padding: 14px 10px 14px 10px;
}

.distill-site-header .title {
  font-size: 18px;
  min-width: 150px;
}

.distill-site-header .logo {
  padding: 0;
}

.distill-site-header .logo img {
  display: none;
  max-height: 20px;
  width: auto;
  margin-bottom: -4px;
}

.distill-site-header .nav-image img {
  max-height: 18px;
  width: auto;
  display: inline-block;
  margin-bottom: -3px;
}



@media screen and (min-width: 1000px) {
  .distill-site-header .logo img {
    display: inline-block;
  }
  .distill-site-header .nav-left {
    margin-left: 20px;
  }
  .distill-site-header .nav-right {
    margin-right: 20px;
  }
  .distill-site-header .title {
    padding-left: 12px;
  }
}


.distill-site-header .nav-toggle {
  display: none;
}

.nav-dropdown {
  display: inline-block;
  position: relative;
}

.nav-dropdown .nav-dropbtn {
  border: none;
  outline: none;
  color: rgba(255, 255, 255, 0.8);
  padding: 16px 10px;
  background-color: transparent;
  font-family: inherit;
  font-size: inherit;
  font-weight: inherit;
  margin: 0;
  margin-top: 1px;
  z-index: 2;
}

.nav-dropdown-content {
  display: none;
  position: absolute;
  background-color: white;
  min-width: 200px;
  border: 1px solid rgba(0,0,0,0.15);
  border-radius: 4px;
  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
  z-index: 1;
  margin-top: 2px;
  white-space: nowrap;
  padding-top: 4px;
  padding-bottom: 4px;
}

.nav-dropdown-content hr {
  margin-top: 4px;
  margin-bottom: 4px;
  border: none;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.nav-dropdown-active {
  display: block;
}

.nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
  color: black;
  padding: 6px 24px;
  text-decoration: none;
  display: block;
  text-align: left;
}

.nav-dropdown-content .nav-dropdown-header {
  display: block;
  padding: 5px 24px;
  padding-bottom: 0;
  text-transform: uppercase;
  font-size: 14px;
  color: #999999;
  white-space: nowrap;
}

.nav-dropdown:hover .nav-dropbtn {
  color: white;
}

.nav-dropdown-content a:hover {
  background-color: #ddd;
  color: black;
}

.nav-right .nav-dropdown-content {
  margin-left: -45%;
  right: 0;
}

@media screen and (max-width: 768px) {
  .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
  .distill-site-header a.nav-toggle {
    float: right;
    display: block;
  }
  .distill-site-header .title {
    margin-left: 0;
  }
  .distill-site-header .nav-right {
    margin-right: 0;
  }
  .distill-site-header {
    overflow: hidden;
  }
  .nav-right .nav-dropdown-content {
    margin-left: 0;
  }
}


@media screen and (max-width: 768px) {
  .distill-site-header.responsive {position: relative; min-height: 500px; }
  .distill-site-header.responsive a.nav-toggle {
    position: absolute;
    right: 0;
    top: 0;
  }
  .distill-site-header.responsive a,
  .distill-site-header.responsive .nav-dropdown {
    display: block;
    text-align: left;
  }
  .distill-site-header.responsive .nav-left,
  .distill-site-header.responsive .nav-right {
    width: 100%;
  }
  .distill-site-header.responsive .nav-dropdown {float: none;}
  .distill-site-header.responsive .nav-dropdown-content {position: relative;}
  .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
    display: block;
    width: 100%;
    text-align: left;
  }
}

/* Site Footer */

.distill-site-footer {
  width: 100%;
  overflow: hidden;
  box-sizing: border-box;
  z-index: 3;
  margin-top: 30px;
  padding-top: 30px;
  padding-bottom: 30px;
  text-align: center;
}

/* Headroom */

d-title {
  padding-top: 6rem;
}

@media print {
  d-title {
    padding-top: 4rem;
  }
}

.headroom {
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
}

.headroom--transition {
  transition: all .4s ease-in-out;
}

.headroom--unpinned {
  top: -100px;
}

.headroom--pinned {
  top: 0;
}

/* adjust viewport for navbar height */
/* helps vertically center bootstrap (non-distill) content */
.min-vh-100 {
  min-height: calc(100vh - 100px) !important;
}

</style>

<script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="../../site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"/>
<link href="../../site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"/>
<script src="../../site_libs/headroom-0.9.4/headroom.min.js"></script>
<!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

<style type="text/css">

body {
  background-color: white;
}

.pandoc-table {
  width: 100%;
}

.pandoc-table>caption {
  margin-bottom: 10px;
}

.pandoc-table th:not([align]) {
  text-align: left;
}

.pagedtable-footer {
  font-size: 15px;
}

d-byline .byline {
  grid-template-columns: 2fr 2fr;
}

d-byline .byline h3 {
  margin-block-start: 1.5em;
}

d-byline .byline .authors-affiliations h3 {
  margin-block-start: 0.5em;
}

.authors-affiliations .orcid-id {
  width: 16px;
  height:16px;
  margin-left: 4px;
  margin-right: 4px;
  vertical-align: middle;
  padding-bottom: 2px;
}

d-title .dt-tags {
  margin-top: 1em;
  grid-column: text;
}

.dt-tags .dt-tag {
  text-decoration: none;
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0em 0.4em;
  margin-right: 0.5em;
  margin-bottom: 0.4em;
  font-size: 70%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

d-article table.gt_table td,
d-article table.gt_table th {
  border-bottom: none;
  font-size: 100%;
}

.html-widget {
  margin-bottom: 2.0em;
}

.l-screen-inset {
  padding-right: 16px;
}

.l-screen .caption {
  margin-left: 10px;
}

.shaded {
  background: rgb(247, 247, 247);
  padding-top: 20px;
  padding-bottom: 20px;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .html-widget {
  margin-bottom: 0;
  border: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .shaded-content {
  background: white;
}

.text-output {
  margin-top: 0;
  line-height: 1.5em;
}

.hidden {
  display: none !important;
}

hr.section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  margin: 0px;
}


d-byline {
  border-top: none;
}

d-article {
  padding-top: 2.5rem;
  padding-bottom: 30px;
  border-top: none;
}

d-appendix {
  padding-top: 30px;
}

d-article>p>img {
  width: 100%;
}

d-article h2 {
  margin: 1rem 0 1.5rem 0;
}

d-article h3 {
  margin-top: 1.5rem;
}

d-article iframe {
  border: 1px solid rgba(0, 0, 0, 0.1);
  margin-bottom: 2.0em;
  width: 100%;
}

/* Tweak code blocks */

d-article div.sourceCode code,
d-article pre code {
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
}

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: auto;
}

d-article div.sourceCode {
  background-color: white;
}

d-article div.sourceCode pre {
  padding-left: 10px;
  font-size: 12px;
  border-left: 2px solid rgba(0,0,0,0.1);
}

d-article pre {
  font-size: 12px;
  color: black;
  background: none;
  margin-top: 0;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

d-article pre a {
  border-bottom: none;
}

d-article pre a:hover {
  border-bottom: none;
  text-decoration: underline;
}

d-article details {
  grid-column: text;
  margin-bottom: 0.8em;
}

@media(min-width: 768px) {

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: visible !important;
}

d-article div.sourceCode pre {
  padding-left: 18px;
  font-size: 14px;
}

/* tweak for Pandoc numbered line within distill */
d-article pre.numberSource code > span {
    left: -2em;
}

d-article pre {
  font-size: 14px;
}

}

figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

/* CSS for d-contents */

.d-contents {
  grid-column: text;
  color: rgba(0,0,0,0.8);
  font-size: 0.9em;
  padding-bottom: 1em;
  margin-bottom: 1em;
  padding-bottom: 0.5em;
  margin-bottom: 1em;
  padding-left: 0.25em;
  justify-self: start;
}

@media(min-width: 1000px) {
  .d-contents.d-contents-float {
    height: 0;
    grid-column-start: 1;
    grid-column-end: 4;
    justify-self: center;
    padding-right: 3em;
    padding-left: 2em;
  }
}

.d-contents nav h3 {
  font-size: 18px;
  margin-top: 0;
  margin-bottom: 1em;
}

.d-contents li {
  list-style-type: none
}

.d-contents nav > ul {
  padding-left: 0;
}

.d-contents ul {
  padding-left: 1em
}

.d-contents nav ul li {
  margin-top: 0.6em;
  margin-bottom: 0.2em;
}

.d-contents nav a {
  font-size: 13px;
  border-bottom: none;
  text-decoration: none
  color: rgba(0, 0, 0, 0.8);
}

.d-contents nav a:hover {
  text-decoration: underline solid rgba(0, 0, 0, 0.6)
}

.d-contents nav > ul > li > a {
  font-weight: 600;
}

.d-contents nav > ul > li > ul {
  font-weight: inherit;
}

.d-contents nav > ul > li > ul > li {
  margin-top: 0.2em;
}


.d-contents nav ul {
  margin-top: 0;
  margin-bottom: 0.25em;
}

.d-article-with-toc h2:nth-child(2) {
  margin-top: 0;
}


/* Figure */

.figure {
  position: relative;
  margin-bottom: 2.5em;
  margin-top: 1.5em;
}

.figure .caption {
  color: rgba(0, 0, 0, 0.6);
  font-size: 12px;
  line-height: 1.5em;
}

.figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

.figure .caption a {
  color: rgba(0, 0, 0, 0.6);
}

.figure .caption b,
.figure .caption strong, {
  font-weight: 600;
  color: rgba(0, 0, 0, 1.0);
}

/* Citations */

d-article .citation {
  color: inherit;
  cursor: inherit;
}

div.hanging-indent{
  margin-left: 1em; text-indent: -1em;
}

/* Citation hover box */

.tippy-box[data-theme~=light-border] {
  background-color: rgba(250, 250, 250, 0.95);
}

.tippy-content > p {
  margin-bottom: 0;
  padding: 2px;
}


/* Tweak 1000px media break to show more text */

@media(min-width: 1000px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 16px;
  }

  .grid {
    grid-column-gap: 16px;
  }

  d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
  }
  figure .caption, .figure .caption, figure figcaption {
    font-size: 13px;
  }
}

@media(min-width: 1180px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 32px;
  }

  .grid {
    grid-column-gap: 32px;
  }
}


/* Get the citation styles for the appendix (not auto-injected on render since
   we do our own rendering of the citation appendix) */

d-appendix .citation-appendix,
.d-appendix .citation-appendix {
  font-size: 11px;
  line-height: 15px;
  border-left: 1px solid rgba(0, 0, 0, 0.1);
  padding-left: 18px;
  border: 1px solid rgba(0,0,0,0.1);
  background: rgba(0, 0, 0, 0.02);
  padding: 10px 18px;
  border-radius: 3px;
  color: rgba(150, 150, 150, 1);
  overflow: hidden;
  margin-top: -12px;
  white-space: pre-wrap;
  word-wrap: break-word;
}

/* Include appendix styles here so they can be overridden */

d-appendix {
  contain: layout style;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-top: 60px;
  margin-bottom: 0;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  color: rgba(0,0,0,0.5);
  padding-top: 60px;
  padding-bottom: 48px;
}

d-appendix h3 {
  grid-column: page-start / text-start;
  font-size: 15px;
  font-weight: 500;
  margin-top: 1em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.65);
}

d-appendix h3 + * {
  margin-top: 1em;
}

d-appendix ol {
  padding: 0 0 0 15px;
}

@media (min-width: 768px) {
  d-appendix ol {
    padding: 0 0 0 30px;
    margin-left: -30px;
  }
}

d-appendix li {
  margin-bottom: 1em;
}

d-appendix a {
  color: rgba(0, 0, 0, 0.6);
}

d-appendix > * {
  grid-column: text;
}

d-appendix > d-footnote-list,
d-appendix > d-citation-list,
d-appendix > distill-appendix {
  grid-column: screen;
}

/* Include footnote styles here so they can be overridden */

d-footnote-list {
  contain: layout style;
}

d-footnote-list > * {
  grid-column: text;
}

d-footnote-list a.footnote-backlink {
  color: rgba(0,0,0,0.3);
  padding-left: 0.5em;
}



/* Anchor.js */

.anchorjs-link {
  /*transition: all .25s linear; */
  text-decoration: none;
  border-bottom: none;
}
*:hover > .anchorjs-link {
  margin-left: -1.125em !important;
  text-decoration: none;
  border-bottom: none;
}

/* Social footer */

.social_footer {
  margin-top: 30px;
  margin-bottom: 0;
  color: rgba(0,0,0,0.67);
}

.disqus-comments {
  margin-right: 30px;
}

.disqus-comment-count {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  cursor: pointer;
}

#disqus_thread {
  margin-top: 30px;
}

.article-sharing a {
  border-bottom: none;
  margin-right: 8px;
}

.article-sharing a:hover {
  border-bottom: none;
}

.sidebar-section.subscribe {
  font-size: 12px;
  line-height: 1.6em;
}

.subscribe p {
  margin-bottom: 0.5em;
}


.article-footer .subscribe {
  font-size: 15px;
  margin-top: 45px;
}


.sidebar-section.custom {
  font-size: 12px;
  line-height: 1.6em;
}

.custom p {
  margin-bottom: 0.5em;
}

/* Styles for listing layout (hide title) */
.layout-listing d-title, .layout-listing .d-title {
  display: none;
}

/* Styles for posts lists (not auto-injected) */


.posts-with-sidebar {
  padding-left: 45px;
  padding-right: 45px;
}

.posts-list .description h2,
.posts-list .description p {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

.posts-list .description h2 {
  font-weight: 700;
  border-bottom: none;
  padding-bottom: 0;
}

.posts-list h2.post-tag {
  border-bottom: 1px solid rgba(0, 0, 0, 0.2);
  padding-bottom: 12px;
}
.posts-list {
  margin-top: 60px;
  margin-bottom: 24px;
}

.posts-list .post-preview {
  text-decoration: none;
  overflow: hidden;
  display: block;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  padding: 24px 0;
}

.post-preview-last {
  border-bottom: none !important;
}

.posts-list .posts-list-caption {
  grid-column: screen;
  font-weight: 400;
}

.posts-list .post-preview h2 {
  margin: 0 0 6px 0;
  line-height: 1.2em;
  font-style: normal;
  font-size: 24px;
}

.posts-list .post-preview p {
  margin: 0 0 12px 0;
  line-height: 1.4em;
  font-size: 16px;
}

.posts-list .post-preview .thumbnail {
  box-sizing: border-box;
  margin-bottom: 24px;
  position: relative;
  max-width: 500px;
}
.posts-list .post-preview img {
  width: 100%;
  display: block;
}

.posts-list .metadata {
  font-size: 12px;
  line-height: 1.4em;
  margin-bottom: 18px;
}

.posts-list .metadata > * {
  display: inline-block;
}

.posts-list .metadata .publishedDate {
  margin-right: 2em;
}

.posts-list .metadata .dt-authors {
  display: block;
  margin-top: 0.3em;
  margin-right: 2em;
}

.posts-list .dt-tags {
  display: block;
  line-height: 1em;
}

.posts-list .dt-tags .dt-tag {
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0.3em 0.4em;
  margin-right: 0.2em;
  margin-bottom: 0.4em;
  font-size: 60%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

.posts-list img {
  opacity: 1;
}

.posts-list img[data-src] {
  opacity: 0;
}

.posts-more {
  clear: both;
}


.posts-sidebar {
  font-size: 16px;
}

.posts-sidebar h3 {
  font-size: 16px;
  margin-top: 0;
  margin-bottom: 0.5em;
  font-weight: 400;
  text-transform: uppercase;
}

.sidebar-section {
  margin-bottom: 30px;
}

.categories ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.categories li {
  color: rgba(0, 0, 0, 0.8);
  margin-bottom: 0;
}

.categories li>a {
  border-bottom: none;
}

.categories li>a:hover {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
}

.categories .active {
  font-weight: 600;
}

.categories .category-count {
  color: rgba(0, 0, 0, 0.4);
}


@media(min-width: 768px) {
  .posts-list .post-preview h2 {
    font-size: 26px;
  }
  .posts-list .post-preview .thumbnail {
    float: right;
    width: 30%;
    margin-bottom: 0;
  }
  .posts-list .post-preview .description {
    float: left;
    width: 45%;
  }
  .posts-list .post-preview .metadata {
    float: left;
    width: 20%;
    margin-top: 8px;
  }
  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.5em;
    font-size: 16px;
  }
  .posts-with-sidebar .posts-list {
    float: left;
    width: 75%;
  }
  .posts-with-sidebar .posts-sidebar {
    float: right;
    width: 20%;
    margin-top: 60px;
    padding-top: 24px;
    padding-bottom: 24px;
  }
}


/* Improve display for browsers without grid (IE/Edge <= 15) */

.downlevel {
  line-height: 1.6em;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  margin: 0;
}

.downlevel .d-title {
  padding-top: 6rem;
  padding-bottom: 1.5rem;
}

.downlevel .d-title h1 {
  font-size: 50px;
  font-weight: 700;
  line-height: 1.1em;
  margin: 0 0 0.5rem;
}

.downlevel .d-title p {
  font-weight: 300;
  font-size: 1.2rem;
  line-height: 1.55em;
  margin-top: 0;
}

.downlevel .d-byline {
  padding-top: 0.8em;
  padding-bottom: 0.8em;
  font-size: 0.8rem;
  line-height: 1.8em;
}

.downlevel .section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
}

.downlevel .d-article {
  font-size: 1.06rem;
  line-height: 1.7em;
  padding-top: 1rem;
  padding-bottom: 2rem;
}


.downlevel .d-appendix {
  padding-left: 0;
  padding-right: 0;
  max-width: none;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.5);
  padding-top: 40px;
  padding-bottom: 48px;
}

.downlevel .footnotes ol {
  padding-left: 13px;
}

.downlevel .base-grid,
.downlevel .distill-header,
.downlevel .d-title,
.downlevel .d-abstract,
.downlevel .d-article,
.downlevel .d-appendix,
.downlevel .distill-appendix,
.downlevel .d-byline,
.downlevel .d-footnote-list,
.downlevel .d-citation-list,
.downlevel .distill-footer,
.downlevel .appendix-bottom,
.downlevel .posts-container {
  padding-left: 40px;
  padding-right: 40px;
}

@media(min-width: 768px) {
  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
  padding-left: 150px;
  padding-right: 150px;
  max-width: 900px;
}
}

.downlevel pre code {
  display: block;
  border-left: 2px solid rgba(0, 0, 0, .1);
  padding: 0 0 0 20px;
  font-size: 14px;
}

.downlevel code, .downlevel pre {
  color: black;
  background: none;
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

.downlevel .posts-list .post-preview {
  color: inherit;
}



</style>

<script type="application/javascript">

function is_downlevel_browser() {
  if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                 window.navigator.userAgent)) {
    return true;
  } else {
    return window.load_distill_framework === undefined;
  }
}

// show body when load is complete
function on_load_complete() {

  // add anchors
  if (window.anchors) {
    window.anchors.options.placement = 'left';
    window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
  }


  // set body to visible
  document.body.style.visibility = 'visible';

  // force redraw for leaflet widgets
  if (window.HTMLWidgets) {
    var maps = window.HTMLWidgets.findAll(".leaflet");
    $.each(maps, function(i, el) {
      var map = this.getMap();
      map.invalidateSize();
      map.eachLayer(function(layer) {
        if (layer instanceof L.TileLayer)
          layer.redraw();
      });
    });
  }

  // trigger 'shown' so htmlwidgets resize
  $('d-article').trigger('shown');
}

function init_distill() {

  init_common();

  // create front matter
  var front_matter = $('<d-front-matter></d-front-matter>');
  $('#distill-front-matter').wrap(front_matter);

  // create d-title
  $('.d-title').changeElementType('d-title');

  // separator
  var separator = '<hr class="section-separator" style="clear: both"/>';
  // prepend separator above appendix
  $('.d-byline').before(separator);
  $('.d-article').before(separator);

  // create d-byline
  var byline = $('<d-byline></d-byline>');
  $('.d-byline').replaceWith(byline);

  // create d-article
  var article = $('<d-article></d-article>');
  $('.d-article').wrap(article).children().unwrap();

  // move posts container into article
  $('.posts-container').appendTo($('d-article'));

  // create d-appendix
  $('.d-appendix').changeElementType('d-appendix');

  // flag indicating that we have appendix items
  var appendix = $('.appendix-bottom').children('h3').length > 0;

  // replace footnotes with <d-footnote>
  $('.footnote-ref').each(function(i, val) {
    appendix = true;
    var href = $(this).attr('href');
    var id = href.replace('#', '');
    var fn = $('#' + id);
    var fn_p = $('#' + id + '>p');
    fn_p.find('.footnote-back').remove();
    var text = fn_p.html();
    var dtfn = $('<d-footnote></d-footnote>');
    dtfn.html(text);
    $(this).replaceWith(dtfn);
  });
  // remove footnotes
  $('.footnotes').remove();

  // move refs into #references-listing
  $('#references-listing').replaceWith($('#refs'));

  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    var id = $(this).attr('id');
    $('.d-contents a[href="#' + id + '"]').parent().remove();
    appendix = true;
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
  });

  // show d-appendix if we have appendix content
  $("d-appendix").css('display', appendix ? 'grid' : 'none');

  // localize layout chunks to just output
  $('.layout-chunk').each(function(i, val) {

    // capture layout
    var layout = $(this).attr('data-layout');

    // apply layout to markdown level block elements
    var elements = $(this).children().not('details, div.sourceCode, pre, script');
    elements.each(function(i, el) {
      var layout_div = $('<div class="' + layout + '"></div>');
      if (layout_div.hasClass('shaded')) {
        var shaded_content = $('<div class="shaded-content"></div>');
        $(this).wrap(shaded_content);
        $(this).parent().wrap(layout_div);
      } else {
        $(this).wrap(layout_div);
      }
    });


    // unwrap the layout-chunk div
    $(this).children().unwrap();
  });

  // remove code block used to force  highlighting css
  $('.distill-force-highlighting-css').parent().remove();

  // remove empty line numbers inserted by pandoc when using a
  // custom syntax highlighting theme, except when numbering line
  // in code chunk
  $('pre:not(.numberLines) code.sourceCode a:empty').remove();

  // load distill framework
  load_distill_framework();

  // wait for window.distillRunlevel == 4 to do post processing
  function distill_post_process() {

    if (!window.distillRunlevel || window.distillRunlevel < 4)
      return;

    // hide author/affiliations entirely if we have no authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;
    if (!have_authors)
      $('d-byline').addClass('hidden');

    // article with toc class
    $('.d-contents').parent().addClass('d-article-with-toc');

    // strip links that point to #
    $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

    // add orcid ids
    $('.authors-affiliations').find('.author').each(function(i, el) {
      var orcid_id = front_matter.authors[i].orcidID;
      var author_name = front_matter.authors[i].author
      if (orcid_id) {
        var a = $('<a></a>');
        a.attr('href', 'https://orcid.org/' + orcid_id);
        var img = $('<img></img>');
        img.addClass('orcid-id');
        img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
        img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
        a.append(img);
        $(this).append(a);
      }
    });

    // hide elements of author/affiliations grid that have no value
    function hide_byline_column(caption) {
      $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
    }

    // affiliations
    var have_affiliations = false;
    for (var i = 0; i<front_matter.authors.length; ++i) {
      var author = front_matter.authors[i];
      if (author.affiliation !== "&nbsp;") {
        have_affiliations = true;
        break;
      }
    }
    if (!have_affiliations)
      $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

    // published date
    if (!front_matter.publishedDate)
      hide_byline_column("Published");

    // document object identifier
    var doi = $('d-byline').find('h3:contains("DOI")');
    var doi_p = doi.next().empty();
    if (!front_matter.doi) {
      // if we have a citation and valid citationText then link to that
      if ($('#citation').length > 0 && front_matter.citationText) {
        doi.html('Citation');
        $('<a href="#citation"></a>')
          .text(front_matter.citationText)
          .appendTo(doi_p);
      } else {
        hide_byline_column("DOI");
      }
    } else {
      $('<a></a>')
         .attr('href', "https://doi.org/" + front_matter.doi)
         .html(front_matter.doi)
         .appendTo(doi_p);
    }

     // change plural form of authors/affiliations
    if (front_matter.authors.length === 1) {
      var grid = $('.authors-affiliations');
      grid.children('h3:contains("Authors")').text('Author');
      grid.children('h3:contains("Affiliations")').text('Affiliation');
    }

    // remove d-appendix and d-footnote-list local styles
    $('d-appendix > style:first-child').remove();
    $('d-footnote-list > style:first-child').remove();

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // hoverable references
    $('span.citation[data-cites]').each(function() {
      const citeChild = $(this).children()[0]
      // Do not process if @xyz has been used without escaping and without bibliography activated
      // https://github.com/rstudio/distill/issues/466
      if (citeChild === undefined) return true

      if (citeChild.nodeName == "D-FOOTNOTE") {
        var fn = citeChild
        $(this).html(fn.shadowRoot.querySelector("sup"))
        $(this).id = fn.id
        fn.remove()
      }
      var refs = $(this).attr('data-cites').split(" ");
      var refHtml = refs.map(function(ref) {
        // Could use CSS.escape too here, we insure backward compatibility in navigator
        return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
      }).join("\n");
      window.tippy(this, {
        allowHTML: true,
        content: refHtml,
        maxWidth: 500,
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start'
      });
    });

    // fix footnotes in tables (#411)
    // replacing broken distill.pub feature
    $('table d-footnote').each(function() {
      // we replace internal showAtNode methode which is triggered when hovering a footnote
      this.hoverBox.showAtNode = function(node) {
        // ported from https://github.com/distillpub/template/pull/105/files
        calcOffset = function(elem) {
            let x = elem.offsetLeft;
            let y = elem.offsetTop;
            // Traverse upwards until an `absolute` element is found or `elem`
            // becomes null.
            while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                x += elem.offsetLeft;
                y += elem.offsetTop;
            }

            return { left: x, top: y };
        }
        // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
        const bbox = node.getBoundingClientRect();
        const offset = calcOffset(node);
        this.show([offset.left + bbox.width, offset.top + bbox.height]);
      }
    })

    // clear polling timer
    clearInterval(tid);

    // show body now that everything is ready
    on_load_complete();
  }

  var tid = setInterval(distill_post_process, 50);
  distill_post_process();

}

function init_downlevel() {

  init_common();

   // insert hr after d-title
  $('.d-title').after($('<hr class="section-separator"/>'));

  // check if we have authors
  var front_matter = JSON.parse($("#distill-front-matter").html());
  var have_authors = front_matter.authors && front_matter.authors.length > 0;

  // manage byline/border
  if (!have_authors)
    $('.d-byline').remove();
  $('.d-byline').after($('<hr class="section-separator"/>'));
  $('.d-byline a').remove();

  // remove toc
  $('.d-contents').remove();

  // move appendix elements
  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
  });


  // inject headers into references and footnotes
  var refs_header = $('<h3></h3>');
  refs_header.text('References');
  $('#refs').prepend(refs_header);

  var footnotes_header = $('<h3></h3');
  footnotes_header.text('Footnotes');
  $('.footnotes').children('hr').first().replaceWith(footnotes_header);

  // move appendix-bottom entries to the bottom
  $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
  $('.appendix-bottom').remove();

  // remove appendix if it's empty
  if ($('.d-appendix').children().length === 0)
    $('.d-appendix').remove();

  // prepend separator above appendix
  $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

  // trim code
  $('pre>code').each(function(i, val) {
    $(this).html($.trim($(this).html()));
  });

  // move posts-container right before article
  $('.posts-container').insertBefore($('.d-article'));

  $('body').addClass('downlevel');

  on_load_complete();
}


function init_common() {

  // jquery plugin to change element types
  (function($) {
    $.fn.changeElementType = function(newType) {
      var attrs = {};

      $.each(this[0].attributes, function(idx, attr) {
        attrs[attr.nodeName] = attr.nodeValue;
      });

      this.replaceWith(function() {
        return $("<" + newType + "/>", attrs).append($(this).contents());
      });
    };
  })(jQuery);

  // prevent underline for linked images
  $('a > img').parent().css({'border-bottom' : 'none'});

  // mark non-body figures created by knitr chunks as 100% width
  $('.layout-chunk').each(function(i, val) {
    var figures = $(this).find('img, .html-widget');
    // ignore leaflet img layers (#106)
    figures = figures.filter(':not(img[class*="leaflet"])')
    if ($(this).attr('data-layout') !== "l-body") {
      figures.css('width', '100%');
    } else {
      figures.css('max-width', '100%');
      figures.filter("[width]").each(function(i, val) {
        var fig = $(this);
        fig.css('width', fig.attr('width') + 'px');
      });

    }
  });

  // auto-append index.html to post-preview links in file: protocol
  // and in rstudio ide preview
  $('.post-preview').each(function(i, val) {
    if (window.location.protocol === "file:")
      $(this).attr('href', $(this).attr('href') + "index.html");
  });

  // get rid of index.html references in header
  if (window.location.protocol !== "file:") {
    $('.distill-site-header a[href]').each(function(i,val) {
      $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
    });
  }

  // add class to pandoc style tables
  $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
  $('.kable-table').children('table').addClass('pandoc-table');

  // add figcaption style to table captions
  $('caption').parent('table').addClass("figcaption");

  // initialize posts list
  if (window.init_posts_list)
    window.init_posts_list();

  // implmement disqus comment link
  $('.disqus-comment-count').click(function() {
    window.headroom_prevent_pin = true;
    $('#disqus_thread').toggleClass('hidden');
    if (!$('#disqus_thread').hasClass('hidden')) {
      var offset = $(this).offset();
      $(window).resize();
      $('html, body').animate({
        scrollTop: offset.top - 35
      });
    }
  });
}

document.addEventListener('DOMContentLoaded', function() {
  if (is_downlevel_browser())
    init_downlevel();
  else
    window.addEventListener('WebComponentsReady', init_distill);
});

</script>

<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.5);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

.distill-site-footer {
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */
</style>
<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.7);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #ffffff;
}

.distill-site-footer {
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #fafafa;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */

.nav-right > a {
  text-transform: uppercase;
}

d-title h1, d-title p, d-title figure,
d-abstract p, d-abstract b {
  grid-column: page;
}

.rj-blue {
  color: #2467bb;
}

ul li {
  line-height: 1.6;
  margin-top: 0em;
  margin-bottom: 0em;
}</style>
<style type="text/css">
/* base style */

/* FONT FAMILIES */

:root {
  --heading-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  --mono-default: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  --body-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

body,
.posts-list .post-preview p,
.posts-list .description p {
  font-family: var(--body-font), var(--body-default);
}

h1, h2, h3, h4, h5, h6,
.posts-list .post-preview h2,
.posts-list .description h2 {
  font-family: var(--heading-font), var(--heading-default);
}

d-article div.sourceCode code,
d-article pre code {
  font-family: var(--mono-font), var(--mono-default);
}


/*-- TITLE --*/
d-title h1,
.posts-list > h1 {
  color: var(--title-color, black);
}

d-title h1 {
  font-size: var(--title-size, 50px);
}

/*-- HEADERS --*/
d-article h1,
d-article h2,
d-article h3,
d-article h4,
d-article h5,
d-article h6 {
  color: var(--header-color, rgba(0, 0, 0, 0.8));
}

/*-- BODY --*/
d-article > p,  /* only text inside of <p> tags */
d-article > ul, /* lists */
d-article > ol {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
  font-size: var(--body-size, 1.06rem);
}


/*-- CODE --*/
d-article div.sourceCode code,
d-article pre code {
  font-size: var(--code-size, 14px);
}

/*-- ASIDE --*/
d-article aside {
  font-size: var(--aside-size, 12px);
  color: var(--aside-color, rgba(0, 0, 0, 0.6));
}

/*-- FIGURE CAPTIONS --*/
figure .caption,
figure figcaption,
.figure .caption {
  font-size: var(--fig-cap-size, 13px);
  color: var(--fig-cap-color, rgba(0, 0, 0, 0.6));
}

/*-- METADATA --*/
d-byline h3 {
  font-size: var(--heading-size, 0.6rem);
  color: var(--heading-color, rgba(0, 0, 0, 0.5));
}

d-byline {
  font-size: var(--body-size, 0.8rem);
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

d-byline a,
d-article d-byline a {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

/*-- TABLE OF CONTENTS --*/
.d-contents nav h3 {
  font-size: var(--heading-size, 18px);
}

.d-contents nav a {
  font-size: var(--contents-size, 13px);
}

/*-- APPENDIX --*/
d-appendix h3 {
  font-size: var(--heading-size, 15px);
  color: var(--heading-color, rgba(0, 0, 0, 0.65));
}

d-appendix {
  font-size: var(--text-size, 0.8em);
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

d-appendix d-footnote-list a.footnote-backlink {
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

/*-- WEBSITE HEADER + FOOTER --*/
.distill-site-header .title {
  font-size: var(--title-size, 18px);
  font-family: var(--navbar-font), var(--heading-default);
}

.distill-site-header a,
.nav-dropdown .nav-dropbtn {
  font-family: var(--navbar-font), var(--heading-default);
}

.nav-dropdown .nav-dropbtn {
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  font-size: var(--text-size, 15px);
}

.distill-site-header a:hover,
.nav-dropdown:hover .nav-dropbtn {
  color: var(--hover-color, white);
}

.distill-site-header {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer a:hover {
  color: var(--hover-color, white);
}</style>
<!--/radix_placeholder_distill-->
  <script src="../../site_libs/header-attrs-2.20/header-attrs.js"></script>
  <script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="../../site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="../../site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="../../site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="../../site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="../../site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="../../site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="../../site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="../../site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
<!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      buttons.append('<a href="https://twitter.com/intent/tweet?text='+title+'&url=' + cite_url + '" class="dt-tag"><i class="fab fa-twitter"></i> Tweet</a>')

      const abstract = $('<d-abstract>')
      abstract.append('<b>Abstract:</b><br>')
      abstract.append($("d-title p:not(:empty)").first()) // Move description to d-abstract
      $("d-title p:empty").remove() // Remove empty paragraphs after title
      abstract.append(buttons)
      abstract.insertAfter($('d-title')) // Add abstract section after title

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"TreeSearch: Morphological Phylogenetic Analysis in R","description":"TreeSearch is an R package for phylogenetic analysis, optimized for discrete character data. Tree search may be conducted using equal or implied step weights with an explicit (albeit inexact) allowance for inapplicable character entries, avoiding some of the pitfalls inherent in standard parsimony methods. Profile parsimony and user-specified optimality criteria are supported.\nA graphical interface, which requires no familiarity with R, is designed to help a user to improve the quality of datasets through critical review of underpinning character codings; and to obtain additional information from results by identifying and summarizing clusters of similar trees, mapping the distribution of trees, and removing 'rogue' taxa that obscure underlying relationships.\nTaken together, the package aims to support methodological rigour at each step of data collection, analysis, and the exploration of phylogenetic results.","doi":"10.32614/RJ-2023-019","authors":[{"author":"Martin R. Smith","authorURL":"https://smithlabdurham.github.io/","affiliation":"University of Durham","affiliationURL":"#","orcidID":"0000-0001-5660-1727"}],"publishedDate":"2023-02-10T00:00:00.000+00:00","citationText":"Smith, 2023"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a class="logo" href="../../index.html">
<img src="../../resources/rlogo.png" alt="Logo"/>
</a>
<a href="../../index.html" class="title">The R Journal</a>
</div>
<div class="nav-right">
<a href="../../index.html">Home</a>
<a href="../../issues.html">Issues</a>
<a href="../../news.html">News</a>
<a href="../../issues/2023-1">Current</a>
<a href="../../submissions.html">Submit</a>
<a href="../../articles.xml">
<i class="fa fa-rss" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>TreeSearch: Morphological Phylogenetic Analysis in R</h1>
<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>TreeSearch is an R package for phylogenetic analysis, optimized for discrete character data. Tree search may be conducted using equal or implied step weights with an explicit (albeit inexact) allowance for inapplicable character entries, avoiding some of the pitfalls inherent in standard parsimony methods. Profile parsimony and user-specified optimality criteria are supported.
A graphical interface, which requires no familiarity with R, is designed to help a user to improve the quality of datasets through critical review of underpinning character codings; and to obtain additional information from results by identifying and summarizing clusters of similar trees, mapping the distribution of trees, and removing ‘rogue’ taxa that obscure underlying relationships.
Taken together, the package aims to support methodological rigour at each step of data collection, analysis, and the exploration of phylogenetic results.</p></p>
</div>

<div class="d-byline">
  Martin R. Smith <a href="https://smithlabdurham.github.io/" class="uri">https://smithlabdurham.github.io/</a> (University of Durham)
  
<br/>2023-02-10
</div>

<div class="d-article">
<div class="layout-chunk" data-layout="l-body">

</div>
<h1 data-number="1" id="introduction"><span class="header-section-number">1</span> Introduction</h1>
<p>Even in the phylogenomic era, morphological data make an important contribution
to phylogenetic questions. Discrete phenotypic data improve the accuracy and
resolution of phylogenetic reconstruction even when outnumbered by molecular
characters, and are the only way to incorporate the unique perspective on
historical events that fossil taxa provide <span class="citation" data-cites="Wiens2004 Wortley2006 Koch2020 Asher2021">(<a href="#ref-Wiens2004" role="doc-biblioref">Wiens 2004</a>; <a href="#ref-Wortley2006" role="doc-biblioref">Wortley and Scotland 2006</a>; <a href="#ref-Koch2020" role="doc-biblioref">Koch and Parry 2020</a>; <a href="#ref-Asher2021" role="doc-biblioref">Asher and Smith 2022</a>)</span>.</p>
<p>One challenge with morphological analysis is the treatment of inapplicable
character states: for example, ‘tail colour’ cannot logically be ascribed either
of the states ‘red’ or ‘blue’ in a taxon that lacks a tail <span class="citation" data-cites="Maddison1993">(<a href="#ref-Maddison1993" role="doc-biblioref">Maddison 1993</a>)</span>. This
situation can profoundly mislead phylogenetic analysis, and is not handled
appropriately by any standard Markov model or parsimony method.</p>
<p>Solutions to this issue have recently been proposed <span class="citation" data-cites="DeLaet2005 Brazeau2019 Tarasov2019 Tarasov2022 Goloboff2021 Hopkins2021">(<a href="#ref-DeLaet2005" role="doc-biblioref">De Laet 2005</a>; <a href="#ref-Brazeau2019" role="doc-biblioref">Brazeau et al. 2019</a>; <a href="#ref-Tarasov2019" role="doc-biblioref">Tarasov 2019</a>, <a href="#ref-Tarasov2022" role="doc-biblioref">2022</a>; <a href="#ref-Goloboff2021" role="doc-biblioref">Goloboff et al. 2021</a>; <a href="#ref-Hopkins2021" role="doc-biblioref">Hopkins and St. John 2021</a>)</span>. Where a single
‘principal’ character (e.g. ‘tail’) exhibits <span class="math inline">\(n\)</span> ‘contingent’ characters (e.g.
‘tail colour’, ‘tail covering’), ‘exact’ solutions <span class="citation" data-cites="Tarasov2019 Tarasov2022 Goloboff2021">(<a href="#ref-Tarasov2019" role="doc-biblioref">Tarasov 2019</a>, <a href="#ref-Tarasov2022" role="doc-biblioref">2022</a>; <a href="#ref-Goloboff2021" role="doc-biblioref">Goloboff et al. 2021</a>)</span> require the construction of multi-state hierarchies containing
<span class="math inline">\(O(2^n)\)</span> entries, meaning that analysis is only computationally tractable for
simple hierarchies with few contingent characters. Moreover, these approaches
cannot accommodate characters that are contingent on more than one principal
character: for example, characters describing appendages on a differentiated
head may be contingent on the presence of the two characters ‘appendages’ and
‘differentiated head’.</p>
<p>Such situations can be approached using the flexible parsimony approximation
proposed by <span class="citation" data-cites="Brazeau2019">Brazeau et al. (<a href="#ref-Brazeau2019" role="doc-biblioref">2019</a>)</span>. <a href="#">TreeSearch</a> scores trees using the “Morphy” C
implementation of this algorithm <span class="citation" data-cites="Brazeau2017">(<a href="#ref-Brazeau2017" role="doc-biblioref">Brazeau et al. 2017</a>)</span>. Morphy implements tree search
under equal step weights. TreeSearch additionally implements implied step
weighting <span class="citation" data-cites="Goloboff1993">(<a href="#ref-Goloboff1993" role="doc-biblioref">Goloboff 1993</a>)</span>, a method which consistently finds more accurate and
precise trees than equal weights parsimony <span class="citation" data-cites="Goloboff2008 Goloboff2018 Smith2019">(<a href="#ref-Goloboff2008" role="doc-biblioref">Goloboff et al. 2008</a>, <a href="#ref-Goloboff2018" role="doc-biblioref">2018a</a>; <a href="#ref-Smith2019" role="doc-biblioref">Smith 2019a</a>)</span>.</p>
<p>There has been lively discussion as to whether, with the rise of probabilistic
approaches, parsimony remains a useful tool for morphological phylogenetics
<span class="citation" data-cites="OReilly2016 Puttick2017 Sansom2018 Goloboff2018b Brown2017">(e.g. <a href="#ref-OReilly2016" role="doc-biblioref">O’Reilly et al. 2016</a>; <a href="#ref-Brown2017" role="doc-biblioref">Brown et al. 2017</a>; <a href="#ref-Puttick2017" role="doc-biblioref">Puttick et al. 2017</a>; <a href="#ref-Goloboff2018b" role="doc-biblioref">Goloboff et al. 2018b</a>; <a href="#ref-Sansom2018" role="doc-biblioref">Sansom et al. 2018</a>)</span>.
Notwithstanding scenarios that go beyond the limits of parsimony, such as the
simultaneous incorporation of stratigraphic data and other prior knowledge <span class="citation" data-cites="Guenser2021">(e.g. <a href="#ref-Guenser2021" role="doc-biblioref">Guenser et al. 2021</a>)</span>, neither parsimony nor probabilistic methods consistently recover
‘better’ trees when gains in accuracy are balanced against losses in precision
<span class="citation" data-cites="Smith2019">(<a href="#ref-Smith2019" role="doc-biblioref">Smith 2019a</a>)</span>. Even if probabilistic methods may eventually be improved through
the creation of more sophisticated models that better reflect the nature of
morphological data <span class="citation" data-cites="Goloboff2018 Tarasov2019 Tarasov2022">(<a href="#ref-Goloboff2018" role="doc-biblioref">Goloboff et al. 2018a</a>; <a href="#ref-Tarasov2019" role="doc-biblioref">Tarasov 2019</a>, <a href="#ref-Tarasov2022" role="doc-biblioref">2022</a>)</span>, parsimony
analysis remains a useful tool – not only because treatments of inapplicable
character states are presently available, but also because it facilitates a
deeper understanding of the underpinning data by emphasizing the reciprocal
relationship between a tree and the synapomorphies that it implies.</p>
<p>Whatever method is used to find phylogenetic trees, a single consensus tree may
fail to convey all the signal in a set of phylogenetic results <span class="citation" data-cites="Wilkinson1994 Wilkinson1996 Wilkinson2003">(<a href="#ref-Wilkinson1994" role="doc-biblioref">Wilkinson 1994</a>, <a href="#ref-Wilkinson1996" role="doc-biblioref">1996</a>, <a href="#ref-Wilkinson2003" role="doc-biblioref">2003</a>)</span>. A set of optimal trees can be better
interpreted by examining consensus trees generated from clusters of similar
trees <span class="citation" data-cites="Stockham2002">(<a href="#ref-Stockham2002" role="doc-biblioref">Stockham et al. 2002</a>)</span>; by exploring tree space <span class="citation" data-cites="Wright2020 SmithSpace">(<a href="#ref-Wright2020" role="doc-biblioref">Wright and Lloyd 2020</a>; <a href="#ref-SmithSpace" role="doc-biblioref">Smith 2022a</a>)</span> and by
automatically identifying, annotating and removing ‘wildcard’ taxa <span class="citation" data-cites="SmithRogue">(<a href="#ref-SmithRogue" role="doc-biblioref">Smith 2022b</a>)</span>
whose ‘rogue’ behaviour may reflect underlying character conflict or ambiguity
<span class="citation" data-cites="Kearney2002">(<a href="#ref-Kearney2002" role="doc-biblioref">Kearney 2002</a>)</span>. These methods are not always easy to integrate into phylogenetic
workflows, so are not routinely included in empirical studies.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:flowchart"></span>
<img src="Flow.svg" alt="Flow charts summarizing key functions available in TreeSearch." width="60%" />
<p class="caption">
Figure 1: Flow charts summarizing key functions available in TreeSearch.
</p>
</div>
</div>
<p>TreeSearch provides functions that allow researchers to engage with the three
main aspects of morphological phylogenetic analysis: dataset construction and
validation; phylogenetic search (including with inapplicable data); and the
interrogation of optimal tree sets (Fig. <a href="#fig:flowchart">1</a>). These functions
can be accessed via the R command-line, as documented within the package and at
<a href="https://ms609.github.io/TreeSearch/">ms609.github.io/TreeSearch</a>, or through a
graphical user interface (GUI). The GUI includes options to export a log of
executed commands as a fully reproducible R script, and to save outputs in
graphical, Nexus or Newick formats.</p>
<h1 data-number="2" id="implementation"><span class="header-section-number">2</span> Implementation</h1>
<h2 class="unnumbered" id="tree-scoring">Tree scoring</h2>
<p><a href="#">TreeSearch</a> can score trees using equal weights, implied weighting
<span class="citation" data-cites="Goloboff1993">(<a href="#ref-Goloboff1993" role="doc-biblioref">Goloboff 1993</a>)</span>, or profile parsimony <span class="citation" data-cites="Faith2001">(<a href="#ref-Faith2001" role="doc-biblioref">Faith and Trueman 2001</a>)</span>. The function <code>TreeLength()</code>
calculates tree score using the “Morphy” phylogenetic library <span class="citation" data-cites="Brazeau2017">(<a href="#ref-Brazeau2017" role="doc-biblioref">Brazeau et al. 2017</a>)</span>,
which implements the <span class="citation" data-cites="Fitch1971">Fitch (<a href="#ref-Fitch1971" role="doc-biblioref">1971</a>)</span> and <span class="citation" data-cites="Brazeau2019">Brazeau et al. (<a href="#ref-Brazeau2019" role="doc-biblioref">2019</a>)</span> algorithms. Morphy returns the
equal weights parsimony score of a tree against a given dataset. Implied weights
and profile parsimony scores are computed by first making a separate call to
Morphy for each character in turn, passed as a single-character dataset; then
passing this value to the appropriate weighting formula and summing the total
score over all characters.</p>
<p>Implied weighting <span class="citation" data-cites="Goloboff1993">(<a href="#ref-Goloboff1993" role="doc-biblioref">Goloboff 1993</a>)</span> is an approximate method that treats each
additional step (i.e. transition between tokens) in a character as less
surprising – and thus requiring less penalty – than the previous step. Each
additional step demonstrates that a character is less reliable for phylogenetic
inference, and thus more likely to contain additional homoplasy. The score of a
tree under implied weighting is <span class="math inline">\(\sum{\frac{e_i}{e_i+k}}\)</span>, where <span class="math inline">\(e_i\)</span> denotes
the number of extra steps observed in character <span class="math inline">\(i\)</span>, and is derived by
subtracting the minimum score that the character can obtain on any tree from the
score observed on the tree in question <span class="citation" data-cites="Goloboff1993">(<a href="#ref-Goloboff1993" role="doc-biblioref">Goloboff 1993</a>)</span>. The minimum length of a
tree is one less than the number of unique tokens (excluding the inapplicable
token ‘<code>-</code>’) that must be present.</p>
<p>Profile parsimony <span class="citation" data-cites="Faith2001">(<a href="#ref-Faith2001" role="doc-biblioref">Faith and Trueman 2001</a>)</span> represents an alternative formulation of how
surprising each additional step in a character is <span class="citation" data-cites="Arias2004">(<a href="#ref-Arias2004" role="doc-biblioref">Arias and Miranda-Esquivel 2004</a>)</span>: the penalty
associated with each additional step in a character is a function of the
probability that a character will fit at least as well as is observed on a
uniformly selected tree. On this view, an additional step is less surprising if
observed in a character where there are more opportunities to observe homoplasy,
whether because a character contains fewer ambiguous codings (a motivation for
the ‘extended’ implied weighting of <span class="citation" data-cites="Goloboff2014">Goloboff (<a href="#ref-Goloboff2014" role="doc-biblioref">2014</a>)</span>) or because states are
distributed more evenly in a character, whose higher phylogenetic information
content <span class="citation" data-cites="Thorley1998">(<a href="#ref-Thorley1998" role="doc-biblioref">Thorley et al. 1998</a>)</span> corresponds to a lower proportion of trees in which no
additional steps are observed.</p>
<p>TreeSearch calculates the profile parsimony score by computing the logarithm of
the number of trees onto which a character can be mapped using <span class="math inline">\(m\)</span> steps, using
theorem 1 of <span class="citation" data-cites="Carter1990">Carter et al. (<a href="#ref-Carter1990" role="doc-biblioref">1990</a>)</span>. As computation for higher numbers of states
<span class="citation" data-cites="Maddison1991">(<a href="#ref-Maddison1991" role="doc-biblioref">Maddison and Slatkin 1991</a>)</span> is more computationally complex, the present implementation is
restricted to characters that contain two informative applicable states, and
uses the <span class="citation" data-cites="Fitch1971">Fitch (<a href="#ref-Fitch1971" role="doc-biblioref">1971</a>)</span> algorithm.</p>
<h2 class="unnumbered" id="tree-search">Tree search</h2>
<p>The TreeSearch GUI uses the routine <code>MaximizeParsimony()</code> to search for optimal
trees using tree bisection and reconnection (TBR) searches and the parsimony
ratchet <span class="citation" data-cites="Nixon1999">(<a href="#ref-Nixon1999" role="doc-biblioref">Nixon 1999</a>)</span>. This goes beyond the heuristic tree search implementation
in the R package <a href="https://cran.r-project.org/package=phangorn">phangorn</a> <span class="citation" data-cites="Schliep2011">(<a href="#ref-Schliep2011" role="doc-biblioref">Schliep 2011</a>)</span> by using compiled C++ code to
rearrange trees, dramatically accelerating computation, and thus increasing the
scale of dataset that can be analysed in reasonable time; and in supporting TBR
rearrangements, which explore larger neighbourhoods of tree space: TBR evaluates
more trees than nearest-neighbour interchanges or subtree pruning and
regrafting, leading to additional computational expense that is offset by a
decreased likelihood that search will become trapped in a local optimum
<span class="citation" data-cites="Goeffon2008 Whelan2010">(<a href="#ref-Goeffon2008" role="doc-biblioref">Goeffon et al. 2008</a>; <a href="#ref-Whelan2010" role="doc-biblioref">Whelan and Money 2010</a>)</span>.</p>
<p>By default, search begins from a greedy addition tree generated by function
<code>AdditionTree()</code>, which queues taxa in a random order, then attaches each taxon
in turn to the growing tree at the most parsimonious location. Search may also
be started from neighbour-joining trees, or the results of a previous search.</p>
<p>Search commences by conducting TBR rearrangements – a hill-climbing approach
that locates a locally optimal tree from which no tree accessible by a single
TBR rearrangement has a better score. A TBR iteration breaks a randomly selected
edge in the focal tree, and reconnects each possible pair of edges in the
resultant sub-trees to produce a list of candidate trees. Entries that are
inconsistent with user-specified topological constraints are removed; remaining
trees are inserted into a queue and scored in a random sequence. If the score of
a candidate tree is at least as good as the best yet encountered (within the
bounds of an optional tolerance parameter <span class="math inline">\(\epsilon\)</span>, which allows the retention
of almost-optimal trees in order to improve accuracy – see e.g. <span class="citation" data-cites="Smith2019">Smith (<a href="#ref-Smith2019" role="doc-biblioref">2019a</a>)</span>),
this tree is used as the starting point for a new TBR iteration. Otherwise, the
next tree in the list is considered. TBR search continues until the best score
is found a specified number of times; a specified number of TBR break points
have been evaluated without any improvement to tree score; or a set amount of
time has passed.</p>
<p>When TBR search is complete, iterations of the parsimony ratchet <span class="citation" data-cites="Nixon1999">(<a href="#ref-Nixon1999" role="doc-biblioref">Nixon 1999</a>)</span>
are conducted in order to search areas of tree space that are separated from the
best tree yet found by ‘valleys’ that cannot be traversed by TBR rearrangements
without passing through trees whose optimality score is below the threshold for
acceptance. Each ratchet iteration begins by resampling the original matrix. A
round of TBR search is conducted using this resampled matrix, and the tree thus
produced is used as a starting point for a new round of TBR search using the
original data. After a specified number of ratchet iterations, an optional final
round of TBR search allows a denser sampling of optimal trees from the final
region of tree space.</p>
<p>A simple example search can be conducted using a morphological dataset included
in the package, taken from <span class="citation" data-cites="Vinther2008">Vinther et al. (<a href="#ref-Vinther2008" role="doc-biblioref">2008</a>)</span>:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='st'><a href='https://ms609.github.io/TreeSearch/'>"TreeSearch"</a></span><span class='op'>)</span></span>
<span><span class='va'>vinther</span> <span class='op'>&lt;-</span> <span class='va'>inapplicable.phyData</span><span class='op'>[[</span><span class='st'>"Vinther2008"</span><span class='op'>]</span><span class='op'>]</span></span>
<span><span class='va'>trees</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://ms609.github.io/TreeSearch/reference/MaximizeParsimony.html'>MaximizeParsimony</a></span><span class='op'>(</span><span class='va'>vinther</span>, concavity <span class='op'>=</span> <span class='fl'>10</span>, tolerance <span class='op'>=</span> <span class='fl'>0.05</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>The <code>MaximizeParsimony()</code> command performs tree search under implied weights
with a concavity value of 10 (<code>concavity = Inf</code> would select equal weights),
retaining any tree whose score is within 0.05 of the best score.</p>
<p>The resulting trees can be summarised according to their scores (optionally,
against a different <code>dataset</code> or under a different weighting strategy, as
specified by <code>concavity</code>) and the iteration in which they were first hit:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://ms609.github.io/TreeSearch/reference/TreeLength.html'>TreeLength</a></span><span class='op'>(</span><span class='va'>trees</span>, dataset <span class='op'>=</span> <span class='va'>vinther</span>, concavity <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span> <span class='op'>|&gt;</span> </span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/Round.html'>signif</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>|&gt;</span>         <span class='co'># truncate non-significant digits</span></span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/table.html'>table</a></span><span class='op'>(</span><span class='op'>)</span>             <span class='co'># tabulate by score</span></span></code></pre>
</div>
<pre><code>
1.52814 1.54329  1.5641 
      3      45       4 </code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/attr.html'>attr</a></span><span class='op'>(</span><span class='va'>trees</span>, <span class='st'>"firstHit"</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>  seed  start ratch1 ratch2 ratch3 ratch4 ratch5 ratch6 ratch7  final 
     0     29      4      0     10      7      2      0      0      0 </code></pre>
</div>
<p>More flexible, if less computationally efficient, tree searches can be conducted
at the command line using the <code>TreeSearch()</code>, <code>Ratchet()</code> and <code>Bootstrap()</code>
commands, which support custom tree optimality criteria <span class="citation" data-cites="Hopkins2021">(e.g. <a href="#ref-Hopkins2021" role="doc-biblioref">Hopkins and St. John 2021</a>)</span>.</p>
<h2 class="unnumbered" id="visualization">Visualization</h2>
<p>The distribution of optimal trees, however obtained, can be visualized through
interactive mappings of tree space <span class="citation" data-cites="Hillis2005 SmithSpace">(<a href="#ref-Hillis2005" role="doc-biblioref">Hillis et al. 2005</a>; <a href="#ref-SmithSpace" role="doc-biblioref">Smith 2022a</a>)</span>. <!--
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>distances</span> <span class='op'>&lt;-</span> <span class='fu'>TreeDist</span><span class='fu'>::</span><span class='fu'><a href='https://ms609.github.io/TreeDist/reference/TreeDistance.html'>ClusteringInfoDistance</a></span><span class='op'>(</span><span class='va'>trees</span><span class='op'>)</span></span>
<span><span class='va'>map</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/cmdscale.html'>cmdscale</a></span><span class='op'>(</span><span class='va'>distances</span>, k <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span> <span class='co'># A serious analysis would consider &gt; 2 dims</span></span>
<span><span class='va'>scores</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>vapply</a></span><span class='op'>(</span><span class='va'>trees</span>, <span class='va'>TreeLength</span>, <span class='fl'>1</span>, <span class='va'>vinther</span>, concavity <span class='op'>=</span> <span class='fl'>10</span><span class='op'>)</span></span>
<span><span class='va'>scoreBins</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cut.html'>cut</a></span><span class='op'>(</span><span class='va'>scores</span>, <span class='va'>nBins</span> <span class='op'>&lt;-</span> <span class='fl'>4</span><span class='op'>)</span></span>
<span><span class='va'>cols</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/grDevices/palettes.html'>hcl.colors</a></span><span class='op'>(</span><span class='va'>nBins</span>, alpha <span class='op'>=</span> <span class='fl'>0.8</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/graphics/par.html'>par</a></span><span class='op'>(</span>mai <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>4</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>map</span>, col <span class='op'>=</span> <span class='va'>cols</span><span class='op'>[</span><span class='va'>scoreBins</span><span class='op'>]</span>, pch <span class='op'>=</span> <span class='fl'>16</span>, cex <span class='op'>=</span> <span class='fl'>2</span>,</span>
<span>     axes <span class='op'>=</span> <span class='cn'>FALSE</span>, xlab <span class='op'>=</span> <span class='st'>''</span>, ylab <span class='op'>=</span> <span class='st'>''</span>, asp <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span></span>
<span><span class='fu'>TreeTools</span><span class='fu'>::</span><span class='fu'><a href='https://ms609.github.io/TreeTools/reference/MSTEdges.html'>MSTEdges</a></span><span class='op'>(</span><span class='va'>distances</span>, plot <span class='op'>=</span> <span class='cn'>TRUE</span>, <span class='va'>map</span><span class='op'>[</span>, <span class='fl'>1</span><span class='op'>]</span>, <span class='va'>map</span><span class='op'>[</span>, <span class='fl'>2</span><span class='op'>]</span>,</span>
<span>                    col <span class='op'>=</span> <span class='st'>'#00000030'</span>, lty <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/graphics/legend.html'>legend</a></span><span class='op'>(</span><span class='st'>"topright"</span>, <span class='fu'><a href='https://rdrr.io/r/base/attr.html'>attr</a></span><span class='op'>(</span><span class='va'>scoreBins</span>, <span class='st'>"levels"</span><span class='op'>)</span>, col <span class='op'>=</span> <span class='va'>cols</span>, pch <span class='op'>=</span> <span class='fl'>16</span>,</span>
<span>       title <span class='op'>=</span> <span class='st'>"Tree score"</span>, bty <span class='op'>=</span> <span class='st'>"n"</span><span class='op'>)</span></span></code></pre></div>
<div class="figure">
<img src="RJ-2023-019_files/figure-html5/map-results-1.png" alt="Mapping of trees found in simple search, coloured according to parsimony score under implied weights, with concavity constant of 10" width="624" />
<p class="caption">(\#fig:map-results)Mapping of trees found in simple search, coloured according to parsimony score under implied weights, with concavity constant of 10</p>
</div>

</div>

--> The TreeSearch GUI supports the use of information theoretic distances
<span class="citation" data-cites="Smith2020">(<a href="#ref-Smith2020" role="doc-biblioref">Smith 2020a</a>)</span>; the quartet distance <span class="citation" data-cites="Estabrook1985">(<a href="#ref-Estabrook1985" role="doc-biblioref">Estabrook et al. 1985</a>)</span>; or the Robinson–Foulds
distance <span class="citation" data-cites="Robinson1981">(<a href="#ref-Robinson1981" role="doc-biblioref">Robinson and Foulds 1981</a>)</span> to construct tree spaces, which are mapped into 2–12
dimensions using principal coordinates analysis <span class="citation" data-cites="Gower1966">(<a href="#ref-Gower1966" role="doc-biblioref">Gower 1966</a>)</span>. The degree to
which a mapping faithfully depicts original tree-to-tree distances is measured
using the product of the trustworthiness and continuity metrics <span class="citation" data-cites="Venna2001 Kaski2003 SmithSpace">(<a href="#ref-Venna2001" role="doc-biblioref">Venna and Kaski 2001</a>; <a href="#ref-Kaski2003" role="doc-biblioref">Kaski et al. 2003</a>; <a href="#ref-SmithSpace" role="doc-biblioref">Smith 2022a</a>)</span>, a composite score denoting the degree to which points
that are nearby when mapped are truly close neighbours (trustworthiness), and
the degree to which nearby points remain nearby when mapped (continuity).
Plotting the minimum spanning tree – the shortest path that connects all trees
<span class="citation" data-cites="Gower1969">(<a href="#ref-Gower1969" role="doc-biblioref">Gower and Ross 1969</a>)</span> – can highlight stress in a mapping (grey lines in Fig.
<a href="#fig:treespace-html">2</a>): the spatial relationships of trees are distorted in
regions where the minimum spanning tree takes a circuitous route to connect
trees that are mapped close to one another <span class="citation" data-cites="SmithSpace">(see fig. 1a–b in <a href="#ref-SmithSpace" role="doc-biblioref">Smith 2022a</a>)</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:treespace-html"></span>
<img src="TreeSpace.svg" alt="TreeSpace GUI showing map of tree space with visualization of tree search progress" width="100%" />
<p class="caption">
Figure 2: Three-dimensional map visualizing progress in a tree search in the TreeSearch GUI. Optimal trees belong to three statistically distinct clusters with good support (silhouette coefficient <span class="math inline">\(&gt;\)</span> 0.5), characterized by different relationships between certain taxa (plotting symbols). Although multiple ratchet iterations have visited each cluster, limited overlap between ratchet iterations suggests that a continuation of tree search may sample novel optimal trees. High trustworthiness and continuity values and a simple minimum spanning tree (grey) indicate that the mapping does not exhibit severe distortion. This figure depicts the tree space GUI display after loading the <span class="citation" data-cites="Wills2012">Wills et al. (<a href="#ref-Wills2012" role="doc-biblioref">2012</a>)</span> dataset; clearing previous trees from memory (sample <em>n</em> trees = 0); and starting a new search (Search→Configure) with equal step weighting and 10<sup>1.5</sup> max hits. 93 trees were sampled, coloured by “When first found”, with plotting symbols depicting “Relationships” between the specified taxa.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<p>To relate the geometry of tree space to the underlying trees, each point in tree
space may be annotated according to the optimality score of its corresponding
tree under a selected step weighting scheme; by the relationships between chosen
taxa that are inferred by that tree; and by the search iteration in which the
tree was first found by tree search (Fig. <a href="#fig:treespace-html">2</a>).</p>
<p>Annotating trees by the iteration in which they were first found allows a user
to evaluate whether a continuation of tree search is likely to yield more
optimal trees. For example, if the retained trees were only recently found, the
search may not yet have located a global optimum. Alternatively, if certain
regions of tree space are visited only by a single ratchet iteration, it is
possible that further isolated ‘islands’ <span class="citation" data-cites="Bastert2002">(<a href="#ref-Bastert2002" role="doc-biblioref">Bastert et al. 2002</a>)</span> remain to be found;
continuing tree search until subsequent ratchet iterations no longer locate new
clusters of trees will reduce the chance that optimal regions of tree space
remain unvisited.</p>
<p>As the identification of clusters from mappings of tree space can be misleading
<span class="citation" data-cites="SmithSpace">(<a href="#ref-SmithSpace" role="doc-biblioref">Smith 2022a</a>)</span>, TreeSearch identifies clusters of trees from tree-to-tree
distances using K-means++ clustering, partitioning around medoids and
hierarchical clustering with minimax linkage <span class="citation" data-cites="Hartigan1979 Arthur2007 Murtagh1983 Bien2011 Maechler2019">(<a href="#ref-Hartigan1979" role="doc-biblioref">Hartigan and Wong 1979</a>; <a href="#ref-Murtagh1983" role="doc-biblioref">Murtagh 1983</a>; <a href="#ref-Arthur2007" role="doc-biblioref">Arthur and Vassilvitskii 2007</a>; <a href="#ref-Bien2011" role="doc-biblioref">Bien and Tibshirani 2011</a>; <a href="#ref-Maechler2019" role="doc-biblioref">Maechler et al. 2019</a>)</span>. Clusterings are evaluated using the
silhouette coefficient, a measure of the extent of overlap between clusters
<span class="citation" data-cites="Kaufman1990">(<a href="#ref-Kaufman1990" role="doc-biblioref">Kaufman and Rousseeuw 1990</a>)</span>. The clustering with the highest silhouette coefficient is
depicted if the silhouette coefficient exceeds a user-specified threshold; the
interpretation of the chosen threshold according to <span class="citation" data-cites="Kaufman1990">Kaufman and Rousseeuw (<a href="#ref-Kaufman1990" role="doc-biblioref">1990</a>)</span> is displayed to
the user. Plotting a separate consensus tree for each cluster often reveals
phylogenetic information that is concealed by polytomies in the single ‘plenary’
consensus of all optimal trees <span class="citation" data-cites="Stockham2002">(<a href="#ref-Stockham2002" role="doc-biblioref">Stockham et al. 2002</a>)</span>.</p>
<p>Plenary consensus trees can also lack resolution because of wildcard or ‘rogue’
taxa, in which conflict or ambiguity in their character codings leads to an
unsettled phylogenetic position <span class="citation" data-cites="Wilkinson1994 Wilkinson2003 Kearney2002">(<a href="#ref-Wilkinson1994" role="doc-biblioref">Wilkinson 1994</a>, <a href="#ref-Wilkinson2003" role="doc-biblioref">2003</a>; <a href="#ref-Kearney2002" role="doc-biblioref">Kearney 2002</a>)</span>.
TreeSearch detects rogue taxa using a heuristic approach <span class="citation" data-cites="SmithRogue">(<a href="#ref-SmithRogue" role="doc-biblioref">Smith 2022b</a>)</span> that
seeks to maximize the phylogenetic information content <span class="citation" data-cites="Thorley1998">(<em>sensu</em> <a href="#ref-Thorley1998" role="doc-biblioref">Thorley et al. 1998</a>)</span> of
a consensus tree created after removing rogue taxa from input trees. The
position of an excluded taxon is portrayed by shading each edge or node of the
consensus according to the number of times the specified taxon occurs at that
position on an underlying tree
[Fig. <a href="#fig:rogueplot-html">3</a>; after
<span class="citation" data-cites="Klopfstein2019">Klopfstein and Spasojevic (<a href="#ref-Klopfstein2019" role="doc-biblioref">2019</a>)</span>], equivalent to the ‘branch attachment frequency’ of Phyutility
<span class="citation" data-cites="Smith2008">(<a href="#ref-Smith2008" role="doc-biblioref">Smith and Dunn 2008</a>)</span>.</p>
<p>Identifying taxa with an unstable position, and splits with low support, can
help an investigator to critically re-examine character codings; to this end,
each edge of the resulting consensus can be annotated with the frequency of the
split amongst the tree set, or with a concordance factor <span class="citation" data-cites="Minh2020">(<a href="#ref-Minh2020" role="doc-biblioref">Minh et al. 2020</a>)</span> denoting
the strength of support from the underlying dataset.</p>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:rogueplot-html"></span>
<img src="WillsCons.svg" alt="TreeSearch GUI display showing plot of reduced consensus, coloured by leaf stability" width="100%" />
<p class="caption">
Figure 3: Reduced consensus of 48 cladograms generated by analysis of data from <span class="citation" data-cites="Wills2012">Wills et al. (<a href="#ref-Wills2012" role="doc-biblioref">2012</a>)</span> under different parsimony methods by <span class="citation" data-cites="Brazeau2019">Brazeau et al. (<a href="#ref-Brazeau2019" role="doc-biblioref">2019</a>)</span>, as displayed in the TreeSearch graphical user interface. Removal of taxa reveals strong support for relationships that would otherwise be masked by rogues such as <em>Palaeoscolex</em>, whose position in optimal trees is marked by the highlighted edges. The GUI state can be reproduced by selecting the options displayed in the figure.
</p>
</div>
</div>
<h2 class="unnumbered" id="dataset-review">Dataset review</h2>
<p>Ultimately, the quality of a dataset plays a central role in determining the
reliability of phylogenetic results, with changes to a relatively small number
of character codings potentially exhibiting an outsized impact on reconstructed
topologies <span class="citation" data-cites="Goloboff2021a">(<a href="#ref-Goloboff2021a" role="doc-biblioref">Goloboff and Sereno 2021</a>)</span>. Nevertheless, dataset quality does not always
receive commensurate attention <span class="citation" data-cites="Simoes2017">(<a href="#ref-Simoes2017" role="doc-biblioref">Simões et al. 2017</a>)</span>. One step towards improving the
rigour of morphological datasets is to annotate each cell in a dataset with an
explicit justification for each taxon’s coding <span class="citation" data-cites="Sereno2009">(<a href="#ref-Sereno2009" role="doc-biblioref">Sereno 2009</a>)</span>, which can be
accomplished in Nexus-formatted data files <span class="citation" data-cites="Maddison1997">(<a href="#ref-Maddison1997" role="doc-biblioref">Maddison et al. 1997</a>)</span> using software such
as <a href="https://morphobank.org">MorphoBank</a> <span class="citation" data-cites="OLeary2011">(<a href="#ref-OLeary2011" role="doc-biblioref">O’Leary and Kaufman 2011</a>)</span>.</p>
<p>TreeSearch presents such annotations alongside a reconstruction of each
character’s states on a specified tree, with inapplicable states mapped
according to the algorithm of <span class="citation" data-cites="Brazeau2019">Brazeau et al. (<a href="#ref-Brazeau2019" role="doc-biblioref">2019</a>)</span>. Neomorphic (presence/absence) and
transformational characters <span class="citation" data-cites="Sereno2007">(<a href="#ref-Sereno2007" role="doc-biblioref">Sereno 2007</a>)</span> are distinguished by reserving the
token <code>0</code> to solely denote the absence of a neomorphic character, with tokens
<code>1</code> … <code>n</code> used to denote the <span class="math inline">\(n\)</span> states of a transformational character
<span class="citation" data-cites="Brazeau2019">(<a href="#ref-Brazeau2019" role="doc-biblioref">Brazeau et al. 2019</a>)</span>. In order to identify character codings that contribute to taxon
instability, each leaf is coloured according to its mean contribution to tree
length for the visualized character <span class="citation" data-cites="Pol2009">(<a href="#ref-Pol2009" role="doc-biblioref">Pol and Escapa 2009</a>)</span>.</p>
<p>This visualization of reconstructed character transitions can help to identify
cases where the formulation of characters has unintended consequences
<span class="citation" data-cites="Wilkinson1995 Brazeau2011">(<a href="#ref-Wilkinson1995" role="doc-biblioref">Wilkinson 1995</a>; <a href="#ref-Brazeau2011" role="doc-biblioref">Brazeau 2011</a>)</span>; where inapplicable states have been
inconsistently applied <span class="citation" data-cites="Brazeau2019">(<a href="#ref-Brazeau2019" role="doc-biblioref">Brazeau et al. 2019</a>)</span>; where taphonomic absence is wrongly coded
as biological absence <span class="citation" data-cites="Donoghue2009">(<a href="#ref-Donoghue2009" role="doc-biblioref">Donoghue and Purnell 2009</a>)</span>; where previous datasets are uncritically
recycled <span class="citation" data-cites="Jenner2001">(<a href="#ref-Jenner2001" role="doc-biblioref">Jenner 2001</a>)</span>; or where taxa are coded with more confidence than a
critical evaluation of available evidence can truly support. Insofar as the
optimal tree and the underlying characters are reciprocally illuminating
<span class="citation" data-cites="Mooi2016">(<a href="#ref-Mooi2016" role="doc-biblioref">Mooi and Gill 2016</a>)</span>, successive cycles of phylogenetic analysis and character
re-formulation can improve the integrity of morphological datasets, and thus
increase their capacity to yield meaningful phylogenetic results <span class="citation" data-cites="Hennig1966">(<a href="#ref-Hennig1966" role="doc-biblioref">Hennig 1966</a>)</span>.</p>
<h1 data-number="3" id="availability"><span class="header-section-number">3</span> Availability</h1>
<p><a href="https://cran.r-project.org/package=TreeSearch">TreeSearch</a> can be installed through the Comprehensive R Archive
Network (CRAN) using
<code>install.packages("TreeSearch")</code>; the graphical user
interface is launched with the command
<code>TreeSearch::EasyTrees()</code>.
The package has been tested on Windows 10, Mac OS X 10 and Ubuntu 20, and
requires only packages available from the CRAN repository.
Source code is available at <a href="https://github.com/ms609/TreeSearch/" class="uri">https://github.com/ms609/TreeSearch/</a>,
and is permanently archived at Zenodo
(<a href="https://dx.doi.org/10.5281/zenodo.1042590" class="uri">https://dx.doi.org/10.5281/zenodo.1042590</a>). Online documentation is available
at <a href="https://ms609.github.io/TreeSearch/" class="uri">https://ms609.github.io/TreeSearch/</a>.</p>
<h1 data-number="4" id="acknowledgements"><span class="header-section-number">4</span> Acknowledgements</h1>
<p>I thank Alavya Dhungana and Joe Moysiuk for feedback on preliminary versions of
the software, and Martin Brazeau and anonymous referees for comments on the
manuscript. Functionality in TreeSearch employs the underlying R <!--[@R]-->
packages <a href="https://cran.r-project.org/package=ape">ape</a> <span class="citation" data-cites="Paradis2019">(<a href="#ref-Paradis2019" role="doc-biblioref">Paradis and Schliep 2019</a>)</span>, <a href="https://cran.r-project.org/package=phangorn">phangorn</a> <span class="citation" data-cites="Schliep2011">(<a href="#ref-Schliep2011" role="doc-biblioref">Schliep 2011</a>)</span>,
<a href="https://cran.r-project.org/package=Quartet">Quartet</a> <span class="citation" data-cites="Sand2014 SmithQuartet">(<a href="#ref-Sand2014" role="doc-biblioref">Sand et al. 2014</a>; <a href="#ref-SmithQuartet" role="doc-biblioref">Smith 2019b</a>)</span>, <a href="https://cran.r-project.org/package=Rogue">Rogue</a> <span class="citation" data-cites="SmithRogue">(<a href="#ref-SmithRogue" role="doc-biblioref">Smith 2022b</a>)</span>,
<a href="https://cran.r-project.org/package=shiny">shiny</a> <span class="citation" data-cites="shiny">(<a href="#ref-shiny" role="doc-biblioref">Chang et al. 2021</a>)</span>, <a href="https://cran.r-project.org/package=shinyjs">shinyjs</a> <span class="citation" data-cites="shinyjs">(<a href="#ref-shinyjs" role="doc-biblioref">Attali 2020</a>)</span>, <a href="https://cran.r-project.org/package=TreeDist">TreeDist</a>
<span class="citation" data-cites="TreeDist">(<a href="#ref-TreeDist" role="doc-biblioref">Smith 2020b</a>)</span>, and <a href="https://cran.r-project.org/package=TreeTools">TreeTools</a> <span class="citation" data-cites="TreeTools">(<a href="#ref-TreeTools" role="doc-biblioref">Smith 2019c</a>)</span>. Icons from R used under
GPL-3; Font Awesome, CC-BY-4.0.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h2 class="appendix unnumbered" id="supplementary-materials">Supplementary materials</h2>
<p>Supplementary materials are available in addition to this article. It can be downloaded at
<a href="RJ-2023-019.zip">RJ-2023-019.zip</a></p>
<h2 class="appendix unnumbered" id="cran-packages-used">CRAN packages used</h2>
<p><a href="https://cran.r-project.org/package=TreeSearch">TreeSearch</a>, <a href="https://cran.r-project.org/package=phangorn">phangorn</a>, <a href="https://cran.r-project.org/package=ape">ape</a>, <a href="https://cran.r-project.org/package=Quartet">Quartet</a>, <a href="https://cran.r-project.org/package=Rogue">Rogue</a>, <a href="https://cran.r-project.org/package=shiny">shiny</a>, <a href="https://cran.r-project.org/package=shinyjs">shinyjs</a>, <a href="https://cran.r-project.org/package=TreeDist">TreeDist</a>, <a href="https://cran.r-project.org/package=TreeTools">TreeTools</a></p>
<h2 class="appendix unnumbered" id="cran-task-views-implied-by-cited-packages">CRAN Task Views implied by cited packages</h2>
<p><a href="https://cran.r-project.org/view=Environmetrics">Environmetrics</a>, <a href="https://cran.r-project.org/view=Phylogenetics">Phylogenetics</a>, <a href="https://cran.r-project.org/view=WebTechnologies">WebTechnologies</a></p>
<div id="refs" class="references csl-bib-body hanging-indent" role="doc-bibliography">
<div id="ref-Arias2004" class="csl-entry" role="doc-biblioentry">
J. S. Arias and D. R. Miranda-Esquivel. Profile parsimony (<span>PP</span>): An analysis under implied weights (<span>IW</span>). <em>Cladistics</em>, 20(1): 56–63, 2004. DOI <a href="https://doi.org/10.1111/j.1096-0031.2003.00001.x">10.1111/j.1096-0031.2003.00001.x</a>.
</div>
<div id="ref-Arthur2007" class="csl-entry" role="doc-biblioentry">
D. Arthur and S. Vassilvitskii. K-means++: The advantages of careful seeding. In <em>Proceedings of the eighteenth annual <span>ACM-SIAM</span> symposium on <span>Discrete</span> algorithms</em>, pages. 1027–1035 2007. <span>USA</span>: <span>Society for Industrial and Applied Mathematics</span>. URL <a href="https://dl.acm.org/doi/10.5555/1283383.1283494">https://dl.acm.org/doi/10.5555/1283383.1283494</a>.
</div>
<div id="ref-Asher2021" class="csl-entry" role="doc-biblioentry">
R. J. Asher and M. R. Smith. Phylogenetic signal and bias in paleontology. <em>Systematic Biology</em>, 71(4): 986–1008, 2022. DOI <a href="https://doi.org/10.1093/sysbio/syab072">10.1093/sysbio/syab072</a>.
</div>
<div id="ref-shinyjs" class="csl-entry" role="doc-biblioentry">
D. Attali. <em>Shinyjs: Easily improve the user experience of your shiny apps in seconds.</em> 2020. URL <a href="https://CRAN.R-project.org/package=shinyjs">https://CRAN.R-project.org/package=shinyjs</a>. R package version 2.0.0.
</div>
<div id="ref-Bastert2002" class="csl-entry" role="doc-biblioentry">
O. Bastert, D. Rockmore, P. F. Stadler and G. Tinhofer. Landscapes on spaces of trees. <em>Applied Mathematics and Computation</em>, 131(2-3): 439–459, 2002. DOI <a href="https://doi.org/10.1016/S0096-3003(01)00164-3">10.1016/S0096-3003(01)00164-3</a>.
</div>
<div id="ref-Bien2011" class="csl-entry" role="doc-biblioentry">
J. Bien and R. Tibshirani. Hierarchical clustering with prototypes via minimax linkage. <em>Journal of the American Statistical Association</em>, 106(495): 1075–1084, 2011. DOI <a href="https://doi.org/10.1198/jasa.2011.tm10183">10.1198/jasa.2011.tm10183</a>.
</div>
<div id="ref-Brazeau2011" class="csl-entry" role="doc-biblioentry">
M. D. Brazeau. Problematic character coding methods in morphology and their effects. <em>Biological Journal of the Linnean Society</em>, 104(3): 489–498, 2011. DOI <a href="https://doi.org/10.1111/j.1095-8312.2011.01755.x">10.1111/j.1095-8312.2011.01755.x</a>.
</div>
<div id="ref-Brazeau2019" class="csl-entry" role="doc-biblioentry">
M. D. Brazeau, T. Guillerme and M. R. Smith. An algorithm for morphological phylogenetic analysis with inapplicable data. <em>Systematic Biology</em>, 68(4): 619–631, 2019. DOI <a href="https://doi.org/10.1093/sysbio/syy083">10.1093/sysbio/syy083</a>.
</div>
<div id="ref-Brazeau2017" class="csl-entry" role="doc-biblioentry">
M. D. Brazeau, M. R. Smith and T. Guillerme. <span>MorphyLib</span>: A library for phylogenetic analysis of categorical trait data with inapplicability. 2017. DOI <a href="https://doi.org/10.5281/zenodo.815372">10.5281/zenodo.815372</a>.
</div>
<div id="ref-Brown2017" class="csl-entry" role="doc-biblioentry">
J. W. Brown, C. Parins-Fukuchi, G. W. Stull, O. M. Vargas and S. A. Smith. Bayesian and likelihood phylogenetic reconstructions of morphological traits are not discordant when taking uncertainty into consideration: A comment on <span>Puttick</span> <span><em>et al.</em></span> <em>Proceedings of the Royal Society B: Biological Sciences</em>, 284(1864): 20170986, 2017. DOI <a href="https://doi.org/10.1098/rspb.2017.0986">10.1098/rspb.2017.0986</a>.
</div>
<div id="ref-Carter1990" class="csl-entry" role="doc-biblioentry">
M. Carter, M. Hendy, D. Penny, L. A. Székely and N. C. Wormald. On the distribution of lengths of evolutionary trees. <em>SIAM Journal on Discrete Mathematics</em>, 3(1): 38–47, 1990. DOI <a href="https://doi.org/10.1137/0403005">10.1137/0403005</a>.
</div>
<div id="ref-shiny" class="csl-entry" role="doc-biblioentry">
W. Chang, J. Cheng, J. J. Allaire, C. Sievert, B. Schloerke, Y.-H. Xie, J. Allen, J. McPherson, A. Dipert and B. Borges. <em>Shiny: Web application framework for <span>R</span>.</em> 2021. URL <a href="https://CRAN.R-project.org/package=shiny">https://CRAN.R-project.org/package=shiny</a>. R package version 1.7.1.
</div>
<div id="ref-DeLaet2005" class="csl-entry" role="doc-biblioentry">
J. E. De Laet. Parsimony and the problem of inapplicables in sequence data. <em>Parsimony, Phylogeny, and Genomics</em>, 81–116, 2005. DOI <a href="https://doi.org/10.1093/acprof:oso/9780199297306.003.0006">10.1093/acprof:oso/9780199297306.003.0006</a>.
</div>
<div id="ref-Donoghue2009" class="csl-entry" role="doc-biblioentry">
P. C. J. Donoghue and M. A. Purnell. Distinguishing heat from light in debate over controversial fossils. <em>BioEssays</em>, 31(2): 178–189, 2009. DOI <a href="https://doi.org/10.1002/bies.200800128">10.1002/bies.200800128</a>.
</div>
<div id="ref-Estabrook1985" class="csl-entry" role="doc-biblioentry">
G. F. Estabrook, F. R. McMorris and C. A. Meacham. Comparison of undirected phylogenetic trees based on subtrees of four evolutionary units. <em>Systematic Zoology</em>, 34(2): 193–200, 1985. DOI <a href="https://doi.org/10.2307/sysbio/34.2.193">10.2307/sysbio/34.2.193</a>.
</div>
<div id="ref-Faith2001" class="csl-entry" role="doc-biblioentry">
D. P. Faith and J. W. H. Trueman. Towards an inclusive philosophy for phylogenetic inference. <em>Systematic Biology</em>, 50(3): 331–350, 2001. DOI <a href="https://doi.org/10.1080/10635150118627">10.1080/10635150118627</a>.
</div>
<div id="ref-Fitch1971" class="csl-entry" role="doc-biblioentry">
W. M. Fitch. Toward defining the course of evolution: Minimum change for a specific tree topology. <em>Systematic Biology</em>, 20(4): 406–416, 1971. DOI <a href="https://doi.org/10.1093/sysbio/20.4.406">10.1093/sysbio/20.4.406</a>.
</div>
<div id="ref-Goeffon2008" class="csl-entry" role="doc-biblioentry">
A. Goeffon, J.-M. Richer and Jin-Kao Hao. Progressive <span>Tree Neighborhood</span> applied to the <span>Maximum Parsimony</span> problem. <em>IEEE/ACM Transactions on Computational Biology and Bioinformatics</em>, 5(1): 136–145, 2008. DOI <a href="https://doi.org/10.1109/TCBB.2007.1065">10.1109/TCBB.2007.1065</a>.
</div>
<div id="ref-Goloboff1993" class="csl-entry" role="doc-biblioentry">
P. A. Goloboff. Estimating character weights during tree search. <em>Cladistics</em>, 9(1): 83–91, 1993. DOI <a href="https://doi.org/10.1111/j.1096-0031.1993.tb00209.x">10.1111/j.1096-0031.1993.tb00209.x</a>.
</div>
<div id="ref-Goloboff2014" class="csl-entry" role="doc-biblioentry">
P. A. Goloboff. Extended implied weighting. <em>Cladistics</em>, 30(3): 260–272, 2014. DOI <a href="https://doi.org/10.1111/cla.12047">10.1111/cla.12047</a>.
</div>
<div id="ref-Goloboff2008" class="csl-entry" role="doc-biblioentry">
P. A. Goloboff, J. M. Carpenter, J. S. Arias and D. R. M. Esquivel. Weighting against homoplasy improves phylogenetic analysis of morphological data sets. <em>Cladistics</em>, 24(5): 758–773, 2008. DOI <a href="https://doi.org/10.1111/j.1096-0031.2008.00209.x">10.1111/j.1096-0031.2008.00209.x</a>.
</div>
<div id="ref-Goloboff2021" class="csl-entry" role="doc-biblioentry">
P. A. Goloboff, J. D. Laet, D. Ríos-Tamayo and C. A. Szumik. A reconsideration of inapplicable characters, and an approximation with step-matrix recoding. <em>Cladistics</em>, 37(5): 596–629, 2021. DOI <a href="https://doi.org/10.1111/cla.12456">10.1111/cla.12456</a>.
</div>
<div id="ref-Goloboff2021a" class="csl-entry" role="doc-biblioentry">
P. A. Goloboff and P. C. Sereno. Comparative cladistics: Identifying the sources for differing phylogenetic results between competing morphology-based datasets. <em>Journal of Systematic Palaeontology</em>, 1–26, 2021. DOI <a href="https://doi.org/10.1080/14772019.2021.1970038">10.1080/14772019.2021.1970038</a>.
</div>
<div id="ref-Goloboff2018" class="csl-entry" role="doc-biblioentry">
P. A. Goloboff, A. Torres and J. S. Arias. Weighted parsimony outperforms other methods of phylogenetic inference under models appropriate for morphology. <em>Cladistics</em>, 34(4): 407–437, 2018a. DOI <a href="https://doi.org/10.1111/cla.12205">10.1111/cla.12205</a>.
</div>
<div id="ref-Goloboff2018b" class="csl-entry" role="doc-biblioentry">
P. A. Goloboff, A. Torres Galvis and J. S. Arias. Parsimony and model-based phylogenetic methods for morphological data: Comments on <span>O</span>’<span>Reilly</span> <em>et al</em>. <em>Palaeontology</em>, 61(4): 625–630, 2018b. DOI <a href="https://doi.org/10.1111/pala.12353">10.1111/pala.12353</a>.
</div>
<div id="ref-Gower1966" class="csl-entry" role="doc-biblioentry">
J. C. Gower. Some distance properties of latent root and vector methods used in multivariate analysis. <em>Biometrika</em>, 53(3/4): 325–338, 1966. DOI <a href="https://doi.org/10.2307/2333639">10.2307/2333639</a>.
</div>
<div id="ref-Gower1969" class="csl-entry" role="doc-biblioentry">
J. C. Gower and G. J. S. Ross. Minimum spanning trees and single linkage cluster analysis. <em>Journal of the Royal Statistical Society. Series C (Applied Statistics)</em>, 18(1): 54–64, 1969. DOI <a href="https://doi.org/10.2307/2346439">10.2307/2346439</a>.
</div>
<div id="ref-Guenser2021" class="csl-entry" role="doc-biblioentry">
P. Guenser, R. C. M. Warnock, W. Pett, P. C. J. Donoghue and E. Jarochowska. Does time matter in phylogeny? A perspective from the fossil record. <em>bioR<span><span class="math inline">\(\chi\)</span></span>iv</em>, 2021. DOI <a href="https://doi.org/10.1101/2021.06.11.445746">10.1101/2021.06.11.445746</a>.
</div>
<div id="ref-Hartigan1979" class="csl-entry" role="doc-biblioentry">
J. A. Hartigan and M. A. Wong. Algorithm <span>AS</span> 136: A <span><em>K</em></span>-means clustering algorithm. <em>Journal of the Royal Statistical Society. Series C (Applied Statistics)</em>, 28(1): 100–108, 1979. DOI <a href="https://doi.org/10.2307/2346830">10.2307/2346830</a>.
</div>
<div id="ref-Hennig1966" class="csl-entry" role="doc-biblioentry">
W. Hennig. <em>Phylogenetic systematics.</em> <span>Urbana</span>: <span>The University of Illinois Press</span>, 1966.
</div>
<div id="ref-Hillis2005" class="csl-entry" role="doc-biblioentry">
D. M. Hillis, T. A. Heath and K. St. John. Analysis and visualization of tree space. <em>Systematic Biology</em>, 54(3): 471–482, 2005. DOI <a href="https://doi.org/10.1080/10635150590946961">10.1080/10635150590946961</a>.
</div>
<div id="ref-Hopkins2021" class="csl-entry" role="doc-biblioentry">
M. J. Hopkins and K. St. John. Incorporating hierarchical characters into phylogenetic analysis. <em>Systematic Biology</em>, 70(6): 1163–1180, 2021. DOI <a href="https://doi.org/10.1093/sysbio/syab005">10.1093/sysbio/syab005</a>.
</div>
<div id="ref-Jenner2001" class="csl-entry" role="doc-biblioentry">
R. A. Jenner. Bilaterian phylogeny and uncritical recycling of morphological data sets. <em>Systematic Biology</em>, 50(5): 730–742, 2001. DOI <a href="https://doi.org/10.1080/106351501753328857">10.1080/106351501753328857</a>.
</div>
<div id="ref-Kaski2003" class="csl-entry" role="doc-biblioentry">
S. Kaski, J. Nikkilä, M. Oja, J. Venna, P. Törönen and E. Castrén. Trustworthiness and metrics in visualizing similarity of gene expression. <em>BMC Bioinformatics</em>, 4: 48, 2003. DOI <a href="https://doi.org/10.1186/1471-2105-4-48">10.1186/1471-2105-4-48</a>.
</div>
<div id="ref-Kaufman1990" class="csl-entry" role="doc-biblioentry">
L. Kaufman and P. J. Rousseeuw. Partitioning around medoids (<span>Program PAM</span>). In <em>Finding groups in data: An introduction to cluster analysis</em>, pages. 68–125 1990. <span>John Wiley &amp; Sons, Ltd</span>.
</div>
<div id="ref-Kearney2002" class="csl-entry" role="doc-biblioentry">
M. Kearney. Fragmentary taxa, missing data, and ambiguity: Mistaken assumptions and conclusions. <em>Systematic Biology</em>, 51(2): 369–381, 2002. DOI <a href="https://doi.org/10.1080/10635150252899824">10.1080/10635150252899824</a>.
</div>
<div id="ref-Klopfstein2019" class="csl-entry" role="doc-biblioentry">
S. Klopfstein and T. Spasojevic. Illustrating phylogenetic placement of fossils using <span>RoguePlots</span>: An example from ichneumonid parasitoid wasps (<span>Hymenoptera</span>, <span>Ichneumonidae</span>) and an extensive morphological matrix. <em>PLoS ONE</em>, 14(4): e0212942, 2019. DOI <a href="https://doi.org/10.1371/journal.pone.0212942">10.1371/journal.pone.0212942</a>.
</div>
<div id="ref-Koch2020" class="csl-entry" role="doc-biblioentry">
N. M. Koch and L. A. Parry. Death is on our side: Paleontological data drastically modify phylogenetic hypotheses. <em>Systematic Biology</em>, 69(6): 1052–1067, 2020. DOI <a href="https://doi.org/10.1093/sysbio/syaa023">10.1093/sysbio/syaa023</a>.
</div>
<div id="ref-Maddison1997" class="csl-entry" role="doc-biblioentry">
D. R. Maddison, D. L. Swofford and W. P. Maddison. <span>NEXUS</span>: An extensible file format for systematic information. <em>Systematic Biology</em>, 46(4): 590–621, 1997. DOI <a href="https://doi.org/10.1093/sysbio/46.4.590">10.1093/sysbio/46.4.590</a>.
</div>
<div id="ref-Maddison1993" class="csl-entry" role="doc-biblioentry">
W. P. Maddison. Missing data versus missing characters in phylogenetic analysis. <em>Systematic Biology</em>, 42(4): 576–581, 1993. DOI <a href="https://doi.org/10.1093/sysbio/42.4.576">10.1093/sysbio/42.4.576</a>.
</div>
<div id="ref-Maddison1991" class="csl-entry" role="doc-biblioentry">
W. P. Maddison and M. Slatkin. Null models for the number of evolutionary steps in a character on a phylogenetic tree. <em>Evolution</em>, 45(5): 1184–1197, 1991. DOI <a href="https://doi.org/10.1111/j.1558-5646.1991.tb04385.x">10.1111/j.1558-5646.1991.tb04385.x</a>.
</div>
<div id="ref-Maechler2019" class="csl-entry" role="doc-biblioentry">
M. Maechler, P. Rousseeuw, A. Struyf, M. Hubert and K. Hornik. Cluster: Cluster <span>Analysis Basics</span> and <span>Extensions</span>. <em>Comprehensive R Archive Network</em>, 2.1.0: 2019.
</div>
<div id="ref-Minh2020" class="csl-entry" role="doc-biblioentry">
B. Q. Minh, M. W. Hahn and R. Lanfear. New methods to calculate concordance factors for phylogenomic datasets. <em>Molecular Biology and Evolution</em>, 37(9): 2727–2733, 2020. DOI <a href="https://doi.org/10.1093/molbev/msaa106">10.1093/molbev/msaa106</a>.
</div>
<div id="ref-Mooi2016" class="csl-entry" role="doc-biblioentry">
R. D. Mooi and A. C. Gill. Hennig’s auxiliary principle and reciprocal illumination revisited. In <em>The <span>Future</span> of <span>Phylogenetic Systematics</span></em>, Eds D. Williams, M. Schmitt and Q. Wheeler pages. 258–285 2016. <span>Cambridge</span>: <span>Cambridge University Press</span>. DOI <a href="https://doi.org/10.1017/CBO9781316338797.013">10.1017/CBO9781316338797.013</a>.
</div>
<div id="ref-Murtagh1983" class="csl-entry" role="doc-biblioentry">
F. Murtagh. A survey of recent advances in hierarchical clustering algorithms. <em>The Computer Journal</em>, 26(4): 354–359, 1983. DOI <a href="https://doi.org/10.1093/comjnl/26.4.354">10.1093/comjnl/26.4.354</a>.
</div>
<div id="ref-Nixon1999" class="csl-entry" role="doc-biblioentry">
K. C. Nixon. The <span>Parsimony Ratchet</span>, a new method for rapid parsimony analysis. <em>Cladistics</em>, 15(4): 407–414, 1999. DOI <a href="https://doi.org/10.1111/j.1096-0031.1999.tb00277.x">10.1111/j.1096-0031.1999.tb00277.x</a>.
</div>
<div id="ref-OLeary2011" class="csl-entry" role="doc-biblioentry">
M. A. O’Leary and S. Kaufman. <span>MorphoBank</span>: Phylophenomics in the <span>“cloud.”</span> <em>Cladistics</em>, 27(5): 529–537, 2011. DOI <a href="https://doi.org/10.1111/j.1096-0031.2011.00355.x">10.1111/j.1096-0031.2011.00355.x</a>.
</div>
<div id="ref-OReilly2016" class="csl-entry" role="doc-biblioentry">
J. E. O’Reilly, M. N. Puttick, L. Parry, A. R. Tanner, J. E. Tarver, J. Fleming, D. Pisani and P. C. J. Donoghue. Bayesian methods outperform parsimony but at the expense of precision in the estimation of phylogeny from discrete morphological data. <em>Biology Letters</em>, 12(4): 20160081, 2016. DOI <a href="https://doi.org/10.1098/rsbl.2016.0081">10.1098/rsbl.2016.0081</a>.
</div>
<div id="ref-Paradis2019" class="csl-entry" role="doc-biblioentry">
E. Paradis and K. Schliep. Ape 5.0: An environment for modern phylogenetics and evolutionary analyses in <span>R</span>. <em>Bioinformatics</em>, 35(3): 526–528, 2019. DOI <a href="https://doi.org/10.1093/bioinformatics/bty633">10.1093/bioinformatics/bty633</a>.
</div>
<div id="ref-Pol2009" class="csl-entry" role="doc-biblioentry">
D. Pol and I. H. Escapa. Unstable taxa in cladistic analysis: Identification and the assessment of relevant characters. <em>Cladistics</em>, 25(5): 515–527, 2009. DOI <a href="https://doi.org/10.1111/j.1096-0031.2009.00258.x">10.1111/j.1096-0031.2009.00258.x</a>.
</div>
<div id="ref-Puttick2017" class="csl-entry" role="doc-biblioentry">
M. N. Puttick, J. E. O’Reilly, A. R. Tanner, J. F. Fleming, J. Clark, L. Holloway, J. Lozano-Fernandez, L. A. Parry, J. E. Tarver, D. Pisani, et al. Uncertain-tree: Discriminating among competing approaches to the phylogenetic analysis of phenotype data. <em>Proceedings of the Royal Society B: Biological Sciences</em>, 284(1846): 20162290, 2017. DOI <a href="https://doi.org/10.1098/rspb.2016.2290">10.1098/rspb.2016.2290</a>.
</div>
<div id="ref-Robinson1981" class="csl-entry" role="doc-biblioentry">
D. F. Robinson and L. R. Foulds. Comparison of phylogenetic trees. <em>Mathematical Biosciences</em>, 53(1-2): 131–147, 1981. DOI <a href="https://doi.org/10.1016/0025-5564(81)90043-2">10.1016/0025-5564(81)90043-2</a>.
</div>
<div id="ref-Sand2014" class="csl-entry" role="doc-biblioentry">
A. Sand, M. K. Holt, J. Johansen, G. S. Brodal, T. Mailund and C. N. S. Pedersen. <span class="nocase">tqDist</span>: A library for computing the quartet and triplet distances between binary or general trees. <em>Bioinformatics</em>, 30(14): 2079–2080, 2014. DOI <a href="https://doi.org/10.1093/bioinformatics/btu157">10.1093/bioinformatics/btu157</a>.
</div>
<div id="ref-Sansom2018" class="csl-entry" role="doc-biblioentry">
R. S. Sansom, P. G. Choate, J. N. Keating and E. Randle. Parsimony, not <span>Bayesian</span> analysis, recovers more stratigraphically congruent phylogenetic trees. <em>Biology Letters</em>, 14(6): 20180263, 2018. DOI <a href="https://doi.org/10.1098/rsbl.2018.0263">10.1098/rsbl.2018.0263</a>.
</div>
<div id="ref-Schliep2011" class="csl-entry" role="doc-biblioentry">
K. P. Schliep. Phangorn: Phylogenetic analysis in <span>R</span>. <em>Bioinformatics</em>, 27(4): 592–593, 2011. DOI <a href="https://doi.org/10.1093/bioinformatics/btq706">10.1093/bioinformatics/btq706</a>.
</div>
<div id="ref-Sereno2009" class="csl-entry" role="doc-biblioentry">
P. C. Sereno. Comparative cladistics. <em>Cladistics</em>, 25(6): 624–659, 2009. DOI <a href="https://doi.org/10.1111/j.1096-0031.2009.00265.x">10.1111/j.1096-0031.2009.00265.x</a>.
</div>
<div id="ref-Sereno2007" class="csl-entry" role="doc-biblioentry">
P. C. Sereno. Logical basis for morphological characters in phylogenetics. <em>Cladistics</em>, 23(6): 565–587, 2007. DOI <a href="https://doi.org/10.1111/j.1096-0031.2007.00161.x">10.1111/j.1096-0031.2007.00161.x</a>.
</div>
<div id="ref-Simoes2017" class="csl-entry" role="doc-biblioentry">
T. R. Simões, M. W. Caldwell, A. Palci and R. L. Nydam. Giant taxon-character matrices: Quality of character constructions remains critical regardless of size. <em>Cladistics</em>, 33(2): 198–219, 2017. DOI <a href="https://doi.org/10.1111/cla.12163">10.1111/cla.12163</a>.
</div>
<div id="ref-Smith2019" class="csl-entry" role="doc-biblioentry">
M. R. Smith. Bayesian and parsimony approaches reconstruct informative trees from simulated morphological datasets. <em>Biology Letters</em>, 15(2): 20180632, 2019a. DOI <a href="https://doi.org/10.1098/rsbl.2018.0632">10.1098/rsbl.2018.0632</a>.
</div>
<div id="ref-Smith2020" class="csl-entry" role="doc-biblioentry">
M. R. Smith. Information theoretic <span>Generalized Robinson</span>–<span>Foulds</span> metrics for comparing phylogenetic trees. <em>Bioinformatics</em>, 36(20): 5007–5013, 2020a. DOI <a href="https://doi.org/10.1093/bioinformatics/btaa614">10.1093/bioinformatics/btaa614</a>.
</div>
<div id="ref-SmithQuartet" class="csl-entry" role="doc-biblioentry">
M. R. Smith. Quartet: Comparison of phylogenetic trees using quartet and bipartition measures. <em>Comprehensive R Archive Network</em>, doi:10.5281/zenodo.2536318, 2019b. DOI <a href="https://doi.org/10.5281/zenodo.2536318">10.5281/zenodo.2536318</a>.
</div>
<div id="ref-SmithSpace" class="csl-entry" role="doc-biblioentry">
M. R. Smith. Robust analysis of phylogenetic tree space. <em>Systematic Biology</em>, 71(5): 1255–1270, 2022a. DOI <a href="https://doi.org/10.1093/sysbio/syab100">10.1093/sysbio/syab100</a>.
</div>
<div id="ref-TreeDist" class="csl-entry" role="doc-biblioentry">
M. R. Smith. <span>TreeDist</span>: Calculate and map distances between phylogenetic trees. <em>Comprehensive R Archive Network</em>, doi:10.5281/zenodo.3528123, 2020b. DOI <a href="https://doi.org/10.5281/zenodo.3528123">10.5281/zenodo.3528123</a>.
</div>
<div id="ref-TreeTools" class="csl-entry" role="doc-biblioentry">
M. R. Smith. <span>TreeTools</span>: Create, modify and analyse phylogenetic trees. <em>Comprehensive R Archive Network</em>, doi:10.5281/zenodo.3522725, 2019c. DOI <a href="https://doi.org/10.5281/zenodo.3522725">10.5281/zenodo.3522725</a>.
</div>
<div id="ref-SmithRogue" class="csl-entry" role="doc-biblioentry">
M. R. Smith. Using information theory to detect rogue taxa and improve consensus trees. <em>Systematic Biology</em>, 71(5): 1088–1094, 2022b. DOI <a href="https://doi.org/10.1093/sysbio/syab099">10.1093/sysbio/syab099</a>.
</div>
<div id="ref-Smith2008" class="csl-entry" role="doc-biblioentry">
S. A. Smith and C. W. Dunn. Phyutility: A phyloinformatics tool for trees, alignments and molecular data. <em>Bioinformatics</em>, 24(5): 715–716, 2008. DOI <a href="https://doi.org/10.1093/bioinformatics/btm619">10.1093/bioinformatics/btm619</a>.
</div>
<div id="ref-Stockham2002" class="csl-entry" role="doc-biblioentry">
C. Stockham, L.-S. Wang and T. Warnow. Statistically based postprocessing of phylogenetic analysis by clustering. <em>Bioinformatics</em>, 18(Suppl 1): S285–S293, 2002. DOI <a href="https://doi.org/10.1093/bioinformatics/18.suppl_1.S285">10.1093/bioinformatics/18.suppl_1.S285</a>.
</div>
<div id="ref-Tarasov2019" class="csl-entry" role="doc-biblioentry">
S. Tarasov. Integration of anatomy ontologies and evo-devo using structured <span>Markov</span> models suggests a new framework for modeling discrete phenotypic traits. <em>Systematic Biology</em>, 68(5): 698–716, 2019. DOI <a href="https://doi.org/10.1093/sysbio/syz005">10.1093/sysbio/syz005</a>.
</div>
<div id="ref-Tarasov2022" class="csl-entry" role="doc-biblioentry">
S. Tarasov. New phylogenetic <span>Markov</span> models for inapplicable morphological characters. <em>bioR<span><span class="math inline">\(\chi\)</span></span>iv</em>, 2021.04.26.441495, 2022. DOI <a href="https://doi.org/10.1101/2021.04.26.441495">10.1101/2021.04.26.441495</a>.
</div>
<div id="ref-Thorley1998" class="csl-entry" role="doc-biblioentry">
J. L. Thorley, M. Wilkinson and M. Charleston. The information content of consensus trees. In <em>Advances in <span>Data Science</span> and <span>Classification</span></em>, Eds A. Rizzi, M. Vichi and H.-H. Bock pages. 91–98 1998. <span>Berlin</span>: <span>Springer</span>. DOI <a href="https://doi.org/10.1007/978-3-642-72253-0_12">10.1007/978-3-642-72253-0_12</a>.
</div>
<div id="ref-Venna2001" class="csl-entry" role="doc-biblioentry">
J. Venna and S. Kaski. Neighborhood preservation in nonlinear projection methods: An experimental study. In <em>Artificial <span>Neural Networks</span> <span>ICANN</span> 2001</em>, Eds G. Dorffner, H. Bischof and K. Hornik pages. 485–491 2001. <span>Berlin, Heidelberg</span>: <span>Springer</span>. DOI <a href="https://doi.org/10.1007/3-540-44668-0_68">10.1007/3-540-44668-0_68</a>.
</div>
<div id="ref-Vinther2008" class="csl-entry" role="doc-biblioentry">
J. Vinther, P. Van Roy and D. E. G. Briggs. <span class="nocase">Machaeridians are Palaeozoic armoured annelids</span>. <em>Nature</em>, 451(7175): 185–188, 2008. DOI <a href="https://doi.org/10.1038/nature06474">10.1038/nature06474</a>.
</div>
<div id="ref-Whelan2010" class="csl-entry" role="doc-biblioentry">
S. Whelan and D. Money. The prevalence of multifurcations in tree-space and their implications for tree-search. <em>Molecular Biology and Evolution</em>, 27(12): 2674–2677, 2010. DOI <a href="https://doi.org/10.1093/molbev/msq163">10.1093/molbev/msq163</a>.
</div>
<div id="ref-Wiens2004" class="csl-entry" role="doc-biblioentry">
J. J. Wiens. The role of morphological data in phylogeny reconstruction. <em>Systematic Biology</em>, 53(4): 653–661, 2004. DOI <a href="https://doi.org/10.1080/10635150490472959">10.1080/10635150490472959</a>.
</div>
<div id="ref-Wilkinson1995" class="csl-entry" role="doc-biblioentry">
M. Wilkinson. A comparison of two methods of character construction. <em>Cladistics</em>, 11(3): 297–308, 1995. DOI <a href="https://doi.org/10.1111/j.1096-0031.1995.tb00091.x">10.1111/j.1096-0031.1995.tb00091.x</a>.
</div>
<div id="ref-Wilkinson1994" class="csl-entry" role="doc-biblioentry">
M. Wilkinson. Common cladistic information and its consensus representation: Reduced <span>Adams</span> and reduced cladistic consensus trees and profiles. <em>Systematic Biology</em>, 43(3): 343–368, 1994. DOI <a href="https://doi.org/10.2307/2413673">10.2307/2413673</a>.
</div>
<div id="ref-Wilkinson1996" class="csl-entry" role="doc-biblioentry">
M. Wilkinson. Majority-rule reduced consensus trees and their use in bootstrapping. <em>Molecular Biology and Evolution</em>, 13(3): 437–444, 1996. DOI <a href="https://doi.org/10.1093/oxfordjournals.molbev.a025604">10.1093/oxfordjournals.molbev.a025604</a>.
</div>
<div id="ref-Wilkinson2003" class="csl-entry" role="doc-biblioentry">
M. Wilkinson. Missing entries and multiple trees: Instability, relationships, and support in parsimony analysis. <em>Journal of Vertebrate Paleontology</em>, 23(2): 311–323, 2003. DOI <a href="https://doi.org/10.1671/0272-4634(2003)023[0311:MEAMTI]2.0.CO;2">10.1671/0272-4634(2003)023[0311:MEAMTI]2.0.CO;2</a>.
</div>
<div id="ref-Wills2012" class="csl-entry" role="doc-biblioentry">
M. A. Wills, S. Gerber, M. Ruta and M. Hughes. The disparity of priapulid, archaeopriapulid and palaeoscolecid worms in the light of new data. <em>Journal of Evolutionary Biology</em>, 25(10): 2056–2076, 2012. DOI <a href="https://doi.org/10.1111/j.1420-9101.2012.02586.x">10.1111/j.1420-9101.2012.02586.x</a>.
</div>
<div id="ref-Wortley2006" class="csl-entry" role="doc-biblioentry">
A. H. Wortley and R. W. Scotland. The effect of combining molecular and morphological data in published phylogenetic analyses. <em>Systematic Biology</em>, 55(4): 677–685, 2006. DOI <a href="https://doi.org/10.1080/10635150600899798">10.1080/10635150600899798</a>.
</div>
<div id="ref-Wright2020" class="csl-entry" role="doc-biblioentry">
A. M. Wright and G. T. Lloyd. Bayesian analyses in phylogenetic palaeontology: Interpreting the posterior sample. <em>Palaeontology</em>, 63(6): 997–1006, 2020. DOI <a href="https://doi.org/10.1111/pala.12500">10.1111/pala.12500</a>.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
  <h3 id="references">References</h3>
  <div id="references-listing"></div>
  <h3 id="reuse">Reuse</h3>
  <p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
  <h3 id="citation">Citation</h3>
  <p>For attribution, please cite this work as</p>
  <pre class="citation-appendix short">Smith, "The R Journal: TreeSearch: Morphological Phylogenetic Analysis in R", The R Journal, 2023</pre>
  <p>BibTeX citation</p>
  <pre class="citation-appendix long">@article{RJ-2023-019,
  author = {Smith, Martin R.},
  title = {The R Journal: TreeSearch: Morphological Phylogenetic Analysis in R},
  journal = {The R Journal},
  year = {2023},
  note = {https://doi.org/10.32614/RJ-2023-019},
  doi = {10.32614/RJ-2023-019},
  volume = {14},
  issue = {4},
  issn = {2073-4859},
  pages = {305-315}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<!--radix_placeholder_navigation_after_body--><html><body>
<div class="distill-site-nav distill-site-footer">
<p>© The R Foundation, <a href="mailto:r-journal@r-project.org">web page
contact</a>.</p>
</div>
<!--/radix_placeholder_navigation_after_body-->
</body></html>


</body>

</html>
