---
title: News from the Bioconductor Project
abstract: The 'News from the Bioconductor Project' article from the 2014-2 issue.
author:
- name: The Bioconductor Team
  affiliation: |
    Program in Computational Biology, Fred Hutchinson Cancer Research
    Center
date: '2014-12-01'
date_received: ~
journal:
  firstpage: '227'
  lastpage: '228'
volume: 6
issue: 2
slug: RJ-2014-2-bioconductor
citation_url: https://rjournal.github.io/
packages:
  cran: []
  bioc:
  - GenomicAlignments
  - ballgown
  - derfinder
  - csaw
  - mQTL.NMR
  - DOQTL
  - facopy
  - focalCall
  - flowcatchR
  - flowCHIC
  - flowDensity
  - groHMM
  - FourCSeq
  - systemPipeR
  - IRanges
  - GenomicRanges
  - GenomicFeatures
  - AnnotationHub
  - BiocCheck
preview: preview.png
bibliography: bioc.bib
CTV: ~
output:
  rjtools::rjournal_web_article:
    self_contained: yes
    toc: no
    legacy_pdf: yes
    web_only: yes
    mathjax: https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js
    md_extension: -tex_math_single_backslash

---

::: article
The Bioconductor project provides tools for the analysis and
comprehension of high-throughput genomic data. The 934 software packages
available in Bioconductor can be viewed at
<http://bioconductor.org/packages/release/>. Navigate packages using
'biocViews' terms and title search. Each package has an html page with a
description, links to vignettes, reference manuals, and usage
statistics. Start using Bioconductor and R version 3.1 with

``` r
  source("http://bioconductor.org/biocLite.R")
  biocLite()
```

Install additional packages and dependencies, e.g.,
[*GenomicAlignments*](https://www.bioconductor.org/packages/release/bioc/html/GenomicAlignments.html),
with

``` r
  source("http://bioconductor.org/biocLite.R")
  biocLite("GenomicAlignments")
```

Upgrade installed packages with

``` r
  source("http://bioconductor.org/biocLite.R")
  biocLite()
```

# Bioconductor 3.0 Release Highlights

Bioconductor 3.0 was released on 14 October 2014. It is compatible with
R 3.1 and consists of 934 software packages, 219 experiment data
packages, and more than 870 current annotation packages. In total, the
release includes 114 new software packages and many updates and
improvements to existing packages. The [release
announcement](http://bioconductor.org/news/bioc_3_0_release/) includes
descriptions of new packages and updated NEWS files provided by package
maintainers.

The variety of research areas represented by Bioconductor packages are
organized (and searched) via the
[biocViews](http://www.bioconductor.org/packages/release/BiocViews.html#___Software)
interface. Here we highlight a few topics covered by the new packages.
Methods for differential expression analyses are offered in
[*ballgown*](https://www.bioconductor.org/packages/release/bioc/html/ballgown.html)
(assembled transcriptomes),
[*derfinder*](https://www.bioconductor.org/packages/release/bioc/html/derfinder.html)
(RNA-seq data at base-pair resolution) and
[*csaw*](https://www.bioconductor.org/packages/release/bioc/html/csaw.html)
(differentially bound regions in ChIP-seq data). Quantitative trait loci
(QTL) analysis for 1H NMR data are provided in
[*mQTL.NMR*](https://www.bioconductor.org/packages/release/bioc/html/mQTL.NMR.html);
[*DOQTL*](https://www.bioconductor.org/packages/release/bioc/html/DOQTL.html)
analyzes QTLs in multi-parent outbred populations. Copy number analysis
in tumoral phenotypes and genomic focal aberrations are available in
[*facopy*](https://www.bioconductor.org/packages/release/bioc/html/facopy.html)
and
[*focalCall*](https://www.bioconductor.org/packages/release/bioc/html/focalCall.html),
respectively. Additions to the flow cytometry family include
[*flowcatchR*](https://www.bioconductor.org/packages/release/bioc/html/flowcatchR.html)
with tools for analyzing in vivo microscopy imaging data of flowing
blood cells,
[*flowCHIC*](https://www.bioconductor.org/packages/release/bioc/html/flowCHIC.html)
for analyzing flow data of microbial communities based on histogram
images and
[*flowDensity*](https://www.bioconductor.org/packages/release/bioc/html/flowDensity.html)
which provides tools for automated sequential gating (analogous to
manual gating) based on data density. Several new packages take the
pipeline approach and facilitate steps from raw data to final analysis:
[*groHMM*](https://www.bioconductor.org/packages/release/bioc/html/groHMM.html)
(GRO-seq data),
[*FourCSeq*](https://www.bioconductor.org/packages/release/bioc/html/FourCSeq.html)
(multiplexed 4C sequencing data), and
[*systemPipeR*](https://www.bioconductor.org/packages/release/bioc/html/systemPipeR.html)
(NGS applications such as RNA-Seq, ChIP-Seq, VAR-Seq).

Bioconductor is built on the mature and flexible 'Ranges' infrastructure
defined in packages such as
[*IRanges*](https://www.bioconductor.org/packages/release/bioc/html/IRanges.html),
[*GenomicRanges*](https://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html),
[*GenomicAlignments*](https://www.bioconductor.org/packages/release/bioc/html/GenomicAlignments.html),
and
[*GenomicFeatures*](https://www.bioconductor.org/packages/release/bioc/html/GenomicFeatures.html).
Many packages rely on the Ranges framework for interoperable, re-usable
analysis; [@10.1371/journal.pcbi.1003118] provide an introduction and
[@10.1214/14-STS476] review strategies for processing, summarizing and
visualizing large genomic data.

Our collection of microarray, transcriptome and organism-specific
*annotation packages* use the 'select' interface (keys, columns,
keytypes) which enable programmatic access to the databases they
contain. The
[*AnnotationHub*](https://www.bioconductor.org/packages/release/bioc/html/AnnotationHub.html)
complements our traditional offerings with diverse whole genome
annotations from Ensembl, ENCODE, dbSNP, UCSC, and elsewhere.

# Other activities

Bioconductor offers an [Amazon Machine
Image](http://www.bioconductor.org/help/bioconductor-cloud-ami/)
optimized for running Bioconductor in the Amazon Elastic Compute Cloud
(EC2). The AMI comes pre-loaded with the latest release version of R,
and a subset of Bioconductor packages. The AMI can be customized by
installing R packages with 'biocLite()' or system-level packages with
the Ubuntu package manager 'apt-get'. Files can be transferred to the
EC2 instance via scp or the Rstudio interface.

A recent addition is our collection of Bioconductor [Docker
Images](http://www.bioconductor.org/help/docker/). These self-contained
environments run on Linux, Windows and Mac as well as virtual machines.
The containers provide convenient access to a 'fresh R session' or
specific version of Bioconductor without the overhead of installing
packages and dependencies. Analysis-specific images come pre-loaded with
packages of a common topic such as flow, proteomics, microarray and
sequencing.

New Bioconductor package contributors are encouraged to consult the
[package
guidelines](http://bioconductor.org/developers/package-guidelines/) and
[Package
Submission](http://bioconductor.org/developers/package-submission/)
sections of the Bioconductor web site, and use the new
[*BiocCheck*](https://www.bioconductor.org/packages/release/bioc/html/BiocCheck.html)
package, in addition to `R CMD check`, for guidance on conforming to
Bioconductor package standards.

The Bioconductor web site advertises [training and community
events](http://bioconductor.org/help/events/); [mailing
lists](http://bioconductor.org/help/mailing-list/) connect users with
each other, to domain experts, and to maintainers eager to ensure that
their packages satisfy the needs of leading edge approaches. Keep
abreast of packages added to the 'devel' branch and other activities by
following \@Bioconductor on Twitter.
:::
