<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
      margin-bottom: 0em;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
<title>rqPen: An R Package for Penalized Quantile Regression</title>

<meta property="description" itemprop="description" content="Quantile regression directly models a conditional quantile of interest. The R package rqPen provides penalized quantile regression estimation using the lasso, elastic net, adaptive lasso, SCAD, and MCP penalties. It also provides extensions to group penalties, both for groups of predictors and grouping variable selection across quantiles. This paper presents the different approaches to penalized quantile regression and how to use the provided methods in rqPen."/>

<link rel="canonical" href="https://doi.org/10.32614/RJ-2025-017/"/>
<link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>
<link rel="icon" type="image/vnd.microsoft.icon" href="../../resources/favicon.ico"/>

<!--  https://schema.org/Article -->
<meta property="article:published" itemprop="datePublished" content="2025-09-09"/>
<meta property="article:created" itemprop="dateCreated" content="2025-09-09"/>
<meta name="article:author" content="Ben Sherwood"/>
<meta name="article:author" content="Shaobo Li"/>
<meta name="article:author" content="Adam Maidman"/>

<!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
<meta property="og:title" content="rqPen: An R Package for Penalized Quantile Regression"/>
<meta property="og:type" content="article"/>
<meta property="og:description" content="Quantile regression directly models a conditional quantile of interest. The R package rqPen provides penalized quantile regression estimation using the lasso, elastic net, adaptive lasso, SCAD, and MCP penalties. It also provides extensions to group penalties, both for groups of predictors and grouping variable selection across quantiles. This paper presents the different approaches to penalized quantile regression and how to use the provided methods in rqPen."/>
<meta property="og:url" content="https://doi.org/10.32614/RJ-2025-017/"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="The R Journal"/>

<!--  https://dev.twitter.com/cards/types/summary -->
<meta property="twitter:card" content="summary"/>
<meta property="twitter:title" content="rqPen: An R Package for Penalized Quantile Regression"/>
<meta property="twitter:description" content="Quantile regression directly models a conditional quantile of interest. The R package rqPen provides penalized quantile regression estimation using the lasso, elastic net, adaptive lasso, SCAD, and MCP penalties. It also provides extensions to group penalties, both for groups of predictors and grouping variable selection across quantiles. This paper presents the different approaches to penalized quantile regression and how to use the provided methods in rqPen."/>
<meta property="twitter:url" content="https://doi.org/10.32614/RJ-2025-017/"/>

<!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
<meta name="citation_title" content="rqPen: An R Package for Penalized Quantile Regression"/>
<meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2025-017"/>
<meta name="citation_pdf_url" content="RJ-2025-017.pdf"/>
<meta name="citation_volume" content="17"/>
<meta name="citation_issue" content="2"/>
<meta name="citation_doi" content="10.32614/RJ-2025-017"/>
<meta name="citation_journal_title" content="The R Journal"/>
<meta name="citation_issn" content="2073-4859"/>
<meta name="citation_firstpage" content="146"/>
<meta name="citation_lastpage" content="175"/>
<meta name="citation_fulltext_world_readable" content=""/>
<meta name="citation_online_date" content="2025/09/09"/>
<meta name="citation_publication_date" content="2025/09/09"/>
<meta name="citation_author" content="Ben Sherwood"/>
<meta name="citation_author_institution" content="University of Kansas"/>
<meta name="citation_author" content="Shaobo Li"/>
<meta name="citation_author_institution" content="University of Kansas"/>
<meta name="citation_author" content="Adam Maidman"/>
<meta name="citation_author_institution" content="University of Minnesota"/>
<!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Quantile predictions for equity premium using penalized quantile regression with consistent variable selection across multiple quantiles;citation_author=Shaobo Li;citation_author=Ben Sherwood"/>
  <meta name="citation_reference" content="citation_title=Asymptotic properties of concave L1-norm group penalties;citation_volume=157;citation_doi=https://doi.org/10.1016/j.spl.2019.108631;citation_author=Ben Sherwood;citation_author=Aaron J. Molstad;citation_author=Sumanta Singha"/>
  <meta name="citation_reference" content="citation_title=Model selection via bayesian information criterion for quantile regression models;citation_publisher=[American Statistical Association, Taylor &amp; Francis, Ltd.];citation_volume=109;citation_doi=10.1080/01621459.2013.836975;citation_author=Eun Ryung Lee;citation_author=Hohsuk Noh;citation_author=Byeong U. Park"/>
  <meta name="citation_reference" content="citation_title=A sparse-group lasso;citation_publisher=Taylor &amp; Francis;citation_volume=22;citation_doi=10.1080/10618600.2012.681250;citation_author=Noah Simon;citation_author=Jerome Friedman;citation_author=Trevor Hastie;citation_author=Robert Tibshirani"/>
  <meta name="citation_reference" content="citation_title=Penalized methods for bi-level variable selection;citation_volume=2;citation_doi=10.4310/SII.2009.v2.n3.a10;citation_author=Patrick Breheny;citation_author=Jian Huang"/>
  <meta name="citation_reference" content="citation_title=Goodness of fit and related inference processes for quantile regression;citation_publisher=Taylor &amp; Francis;citation_volume=94;citation_doi=10.2307/2669943;citation_author=Roger Koenker;citation_author=Jos? A. F. Machado"/>
  <meta name="citation_reference" content="citation_title=Coordinate descent algorithms for nonconvex penalized regression, with applications to biological feature selection;citation_publisher=Institute of Mathematical Statistics;citation_volume=5;citation_doi=10.1214/10-AOAS388;citation_author=Patrick Breheny;citation_author=Jian Huang"/>
  <meta name="citation_reference" content="citation_title=Solution of an overdetermined system of equations in the $\ell_1$ norm.;citation_volume=17;citation_doi=10.1145/355616.36102;citation_author=I. Barrodale;citation_author=F. D. K. Roberts"/>
  <meta name="citation_reference" content="citation_title=L1-penalized quantile regression in high-dimensional sparse models;citation_volume=39;citation_doi=10.1214/10-AOS827;citation_author=Alexandre Belloni;citation_author=Victor Chernozhukov"/>
  <meta name="citation_reference" content="citation_title=Variable selection via nonconcave penalized likelihood and its oracle properties;citation_volume=96;citation_doi=10.1198/016214501753382273;citation_author=Jianqing Fan;citation_author=Runze Li"/>
  <meta name="citation_reference" content="citation_title=A selective review of group selection in high-dimensional models;citation_volume=27;citation_doi=10.1214/12-STS392;citation_author=Jian Huang;citation_author=Patrick Breheny;citation_author=Shuangge Ma"/>
  <meta name="citation_reference" content="citation_title=Group descent algorithms for nonconvex penalized linear and logistic regression models with grouped predictors;citation_volume=25;citation_doi=10.1007/s11222-013-9424-2;citation_author=Patrick Breheny;citation_author=Jian Huang"/>
  <meta name="citation_reference" content="citation_title=Robust estimation of a location parameter;citation_publisher=The Institute of Mathematical Statistics;citation_volume=35;citation_doi=10.1214/aoms/1177703732;citation_author=Peter J. Huber"/>
  <meta name="citation_reference" content="citation_title=Group lasso for high dimensional sparse quantile regression models;citation_author=Kengo Kato"/>
  <meta name="citation_reference" content="citation_title=Regression quantiles;citation_publisher=[Wiley, Econometric Society];citation_volume=46;citation_doi=10.2307/1913643;citation_issn=00129682, 14680262;citation_author=Roger Koenker;citation_author=Gilbert Bassett"/>
  <meta name="citation_reference" content="citation_title=Computing regression quantiles;citation_volume=36;citation_doi=10.2307/2347802;citation_author=Roger W. Koenker;citation_author=Vasco D’Orey"/>
  <meta name="citation_reference" content="citation_title=A remark on algorithm AS 229: Computing dual regression quantiles and regression rank score;citation_volume=43;citation_doi=10.2307/2986030;citation_author=Roger Koenker;citation_author=Vasco D’Orey"/>
  <meta name="citation_reference" content="citation_title=An iterative coordinate descent algorithm for high-dimensional nonconvex penalized quantile regression;citation_volume=24;citation_doi=10.1080/10618600.2014.913516;citation_author=Bo Peng;citation_author=Lan Wang"/>
  <meta name="citation_reference" content="citation_title=Quantile regression of analyzing heterogeneity in ultra-high dimension;citation_volume=107;citation_doi=10.1080/01621459.2012.656014;citation_author=Lan Wang;citation_author=Yichao Wu;citation_author=Runze Li"/>
  <meta name="citation_reference" content="citation_title=Group SCAD regression analysis for microarray time course gene expression data;citation_volume=23;citation_doi=10.1093/bioinformatics/btm125;citation_author=Lifeng Wang;citation_author=Guang Chen;citation_author=Hongzhe Li"/>
  <meta name="citation_reference" content="citation_title=Semismooth newton coordinate descent algorithm for elastic-net penalized huber loss regression and quantile regression;citation_volume=26;citation_doi=10.1080/10618600.2016.1256816;citation_author=Congrui Yi;citation_author=Jian Huang"/>
  <meta name="citation_reference" content="citation_title=Nearly unbiased variable selection under minimax concave penalty;citation_volume=38;citation_doi=10.1214/09-AOS729;citation_author=Cun-Hui Zhang"/>
  <meta name="citation_reference" content="citation_title=The adaptive lasso and its oracle properties;citation_volume=101;citation_doi=10.1198/016214506000000735;citation_author=Hui Zou"/>
  <meta name="citation_reference" content="citation_title=One-step sparse estimates in nonconcave penalized likelihood models;citation_volume=36;citation_doi=10.1214/009053607000000802;citation_author=Hui Zou;citation_author=Runze Li"/>
  <meta name="citation_reference" content="citation_title=Regularization and variable selection via the elastic net;citation_volume=67;citation_doi=10.1111/j.1467-9868.2005.00503.x;citation_author=Hui Zou;citation_author=Trevor Hastie"/>
  <meta name="citation_reference" content="citation_title=Regression shrinkage and selection via the lasso;citation_volume=58;citation_doi=10.1111/j.2517-6161.1996.tb02080.x;citation_author=Robert Tibshirani"/>
  <meta name="citation_reference" content="citation_title=Model selection and estimation in regression with grouped variables;citation_volume=68;citation_doi=10.1111/j.1467-9868.2005.00532.x;citation_author=Ming Yuan;citation_author=Yi Lin"/>
  <meta name="citation_reference" content="citation_title=The gaussian hare and the laplacian tortoise: Computability of squared-error versus absolute-error estimators;citation_publisher=The Institute of Mathematical Statistics;citation_volume=12;citation_doi=10.1214/ss/1030037960;citation_author=Stephen Portnoy;citation_author=Roger Koenker"/>
  <meta name="citation_reference" content="citation_title=A fast unified algorithm for solving group-lasso penalize learning problems;citation_volume=25;citation_doi=10.1007/s11222-014-9498-5;citation_author=Yi Yang;citation_author=Hui Zou"/>
  <meta name="citation_reference" content="citation_title=Convex optimization in R;citation_volume=60;citation_doi=10.18637/jss.v060.i05;citation_author=Roger Koenker;citation_author=Ivan Mizera"/>
  <meta name="citation_reference" content="citation_title=High-dimensional quantile regression: Convolution smoothing and concave regularization;citation_volume=84;citation_doi=https://doi.org/10.1111/rssb.12485;citation_author=Kean Ming Tan;citation_author=Lan Wang;citation_author=Wen-Xin Zhou"/>
  <meta name="citation_reference" content="citation_title=Smoothed quantile regression with large-scale inference;citation_volume=232;citation_doi=https://doi.org/10.1016/j.jeconom.2021.07.010;citation_author=Xuming He;citation_author=Xiaoou Pan;citation_author=Kean Ming Tan;citation_author=Wen-Xin Zhou"/>
  <meta name="citation_reference" content="citation_title=Quantile regression feature selection and estimation with grouped variables using huber approximation;citation_volume=32;citation_doi=10.1007/s11222-022-10135-w;citation_author=Ben Sherwood;citation_author=Shaobo Li"/>
  <meta name="citation_reference" content="citation_title=Regularization paths for generalized linear models via coordinate descent;citation_volume=33;citation_doi=10.18637/jss.v033.i01;citation_author=Jerome Friedman;citation_author=Trevor Hastie;citation_author=Robert Tibshirani"/>
  <meta name="citation_reference" content="citation_title=Strong rules for discarding predictors in lasso-type problems;citation_publisher=Wiley Online Library;citation_volume=74;citation_doi=10.1111/j.1467-9868.2011.01004.x;citation_author=Robert Tibshirani;citation_author=Strong rules for discarding predictors in lasso-type problems"/>
  <meta name="citation_reference" content="citation_title=Ames, Iowa: Alternative to the boston housing data as an end of semester regression project;citation_publisher=Taylor &amp; Francis;citation_volume=19;citation_doi=10.1080/10691898.2011.11889627;citation_author=Dean De Cock"/>
  <meta name="citation_reference" content="citation_title=Heteroscedasticity identification and variable selection via multiple quantile regression;citation_publisher=Taylor &amp; Francis;citation_volume=94;citation_doi=10.1080/00949655.2023.2243533;citation_author=Mingqiu Wang;citation_author=Xiaoning Kang;citation_author=Jiajuan Liang;citation_author=Kun Wang;citation_author=Yuanshan Wu"/>
  <meta name="citation_reference" content="citation_title=Valid post-selection inference in high-dimensional approximately sparse quantile regression models;citation_volume=114;citation_doi=10.1080/01621459.2018.1442339;citation_author=Victor Chernozhukov Alexandre Belloni;citation_author=Kengo Kato"/>
  <meta name="citation_reference" content="citation_title=Confidence intervals and hypothesis testing for high-dimensional quantile regression: Convolution smoothing and debiasing;citation_volume=24;citation_author=Yibo Yan;citation_author=Xiaozhou Wang;citation_author=Riquan Zhang"/>
  <meta name="citation_reference" content="citation_title=High dimensional censored quantile regression;citation_publisher=Institute of Mathematical Statistics;citation_volume=46;citation_doi=10.1214/17-AOS1551;citation_author=Qi Zheng;citation_author=Limin Peng;citation_author=Xuming He"/>
  <meta name="citation_reference" content="citation_title=Efficient estimation for censored quantile regression;citation_volume=118;citation_doi=10.1080/01621459.2022.2078331;citation_author=Tony Sit Sze Ming Lee;citation_author=Gongjun Xu"/>
  <meta name="citation_reference" content="citation_title=Quantiles for counts;citation_publisher=[American Statistical Association, Taylor &amp; Francis, Ltd.];citation_volume=100;citation_doi=10.1198/016214505000000330;citation_issn=01621459;citation_author=José A. F. Machado;citation_author=J. M. C. Santos Silva"/>
  <meta name="citation_reference" content="citation_title=Distributed inference for quantile regression processes;citation_publisher=Institute of Mathematical Statistics;citation_volume=47;citation_doi=10.1214/18-AOS1730;citation_author=Stanislav Volgushev;citation_author=Shih-Kang Chao;citation_author=Guang Cheng"/>
  <!--radix_placeholder_rmarkdown_metadata-->

<script type="text/json" id="radix-rmarkdown-metadata">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","date","description","draft","author","type","output","bibliography","preamble","date_received","volume","issue","slug","journal","pdf_url","citation_url","doi","creative_commons","packages","CTV","csl","canonical_url"]}},"value":[{"type":"character","attributes":{},"value":["rqPen: An R Package for Penalized Quantile Regression"]},{"type":"character","attributes":{},"value":["2025-09-09"]},{"type":"character","attributes":{},"value":["Quantile regression directly models a conditional quantile of interest. The R package rqPen provides penalized quantile regression estimation using the lasso, elastic net, adaptive lasso, SCAD, and MCP penalties. It also provides extensions to group penalties, both for groups of predictors and grouping variable selection across quantiles. This paper presents the different approaches to penalized quantile regression and how to use the provided methods in rqPen."]},{"type":"logical","attributes":{},"value":[false]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","email"]}},"value":[{"type":"character","attributes":{},"value":["Ben Sherwood"]},{"type":"character","attributes":{},"value":["University of Kansas"]},{"type":"character","attributes":{},"value":["School of Business","Lawrence, KS"]},{"type":"character","attributes":{},"value":["ben.sherwood@ku.edu"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","email"]}},"value":[{"type":"character","attributes":{},"value":["Shaobo Li"]},{"type":"character","attributes":{},"value":["University of Kansas"]},{"type":"character","attributes":{},"value":["School of Business","Lawrence, KS"]},{"type":"character","attributes":{},"value":["shaobo.li@ku.edu"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","email"]}},"value":[{"type":"character","attributes":{},"value":["Adam Maidman"]},{"type":"character","attributes":{},"value":["University of Minnesota"]},{"type":"character","attributes":{},"value":["School of Statistics","Minneapolis, MN"]},{"type":"character","attributes":{},"value":["abmaidman@gmail.com"]}]}]},{"type":"character","attributes":{},"value":["package"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[false]}]}]},{"type":"character","attributes":{},"value":["RJreferences.bib"]},{"type":"character","attributes":{},"value":["\\usepackage{thumbpdf,lmodern,amsfonts,amsmath,mathtools}"]},{"type":"character","attributes":{},"value":["2024-05-24"]},{"type":"integer","attributes":{},"value":[17]},{"type":"integer","attributes":{},"value":[2]},{"type":"character","attributes":{},"value":["RJ-2025-017"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"integer","attributes":{},"value":[146]},{"type":"integer","attributes":{},"value":[175]}]},{"type":"character","attributes":{},"value":["RJ-2025-017.pdf"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2025-017"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2025-017"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"character","attributes":{},"value":["rqPen","quantreg","Rmosek","hqreg","conquer","hrqglas","AmesHousing"]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["Econometrics","Environmetrics","Optimization","ReproducibleResearch","Robust"]},{"type":"character","attributes":{},"value":["/home/mitchell/R/x86_64-pc-linux-gnu-library/4.5/rjtools/rjournal.csl"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2025-017/"]}]}
</script>
<!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["figures/lassoObjCompare.pdf","figures/lassoObjCompareGroup.pdf","figures/lassoSpeedCompare.pdf","figures/lassoSpeedCompareGroup.pdf","lassoObjCompare.pdf","lassoObjCompareGroup.pdf","lassoSpeedCompare.pdf","lassoSpeedCompareGroup.pdf","motivation-letter/motivation-letter.pdf","motivation-letter/motivation-letter.tex","RJ-2025-017_files/figure-html5/lassoObjCompare-1.png","RJ-2025-017_files/figure-html5/lassoObjCompareGroup-1.png","RJ-2025-017_files/figure-html5/lassoSpeedCompare-1.png","RJ-2025-017_files/figure-html5/lassoSpeedCompareGroup-1.png","RJ-2025-017_files/figure-html5/unnamed-chunk-10-1.png","RJ-2025-017_files/figure-html5/unnamed-chunk-12-1.png","RJ-2025-017_files/figure-html5/unnamed-chunk-20-1.png","RJ-2025-017_files/figure-html5/unnamed-chunk-21-1.png","RJ-2025-017_files/figure-html5/unnamed-chunk-29-1.png","RJ-2025-017_files/figure-html5/unnamed-chunk-29-2.png","RJ-2025-017_files/figure-html5/unnamed-chunk-31-1.png","RJ-2025-017_files/figure-html5/unnamed-chunk-33-1.png","RJ-2025-017_files/figure-latex/lassoObjCompare-1.pdf","RJ-2025-017_files/figure-latex/lassoObjCompareGroup-1.pdf","RJ-2025-017_files/figure-latex/lassoSpeedCompare-1.pdf","RJ-2025-017_files/figure-latex/lassoSpeedCompareGroup-1.pdf","RJ-2025-017_files/figure-latex/unnamed-chunk-10-1.pdf","RJ-2025-017_files/figure-latex/unnamed-chunk-12-1.pdf","RJ-2025-017_files/figure-latex/unnamed-chunk-20-1.pdf","RJ-2025-017_files/figure-latex/unnamed-chunk-21-1.pdf","RJ-2025-017_files/figure-latex/unnamed-chunk-29-1.pdf","RJ-2025-017_files/figure-latex/unnamed-chunk-29-2.pdf","RJ-2025-017_files/figure-latex/unnamed-chunk-31-1.pdf","RJ-2025-017_files/figure-latex/unnamed-chunk-33-1.pdf","RJ-2025-017.log","RJ-2025-017.pdf","RJ-2025-017.tex","RJ-2025-017.zip","RJournal.sty","RJreferences.bib","RJwrapper.log","RJwrapper.tex","Rlogo-5.png","rqPenRJ.tex","simulationCode/lassoObjCompare2.rds","simulationCode/lassoObjCompareGroup.rds","simulationCode/lassoSpeedCompare2.rds","simulationCode/lassoSpeedCompareBen2.Rdata","simulationCode/lassoSpeedCompareGroup.rds","simulationCode/lassoSpeedGroupBen.Rdata","simulationCode/speedCompare.R","simulationCode/speedCompareGroup.R"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
<meta name="distill:offset" content="../.."/>

<script type="application/javascript">

  window.headroom_prevent_pin = false;

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // initialize headroom for banner
    var header = $('header').get(0);
    var headerHeight = header.offsetHeight;
    var headroom = new Headroom(header, {
      tolerance: 5,
      onPin : function() {
        if (window.headroom_prevent_pin) {
          window.headroom_prevent_pin = false;
          headroom.unpin();
        }
      }
    });
    headroom.init();
    if(window.location.hash)
      headroom.unpin();
    $(header).addClass('headroom--transition');

    // offset scroll location for banner on hash change
    // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
    window.addEventListener("hashchange", function(event) {
      window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
    });

    // responsive menu
    $('.distill-site-header').each(function(i, val) {
      var topnav = $(this);
      var toggle = topnav.find('.nav-toggle');
      toggle.on('click', function() {
        topnav.toggleClass('responsive');
      });
    });

    // nav dropdowns
    $('.nav-dropbtn').click(function(e) {
      $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
      $(this).parent().siblings('.nav-dropdown')
         .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $("body").click(function(e){
      $('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $(".nav-dropdown").click(function(e){
      e.stopPropagation();
    });
  });
</script>

<style type="text/css">

/* Theme (user-documented overrideables for nav appearance) */

.distill-site-nav {
  color: rgba(255, 255, 255, 0.8);
  background-color: #0F2E3D;
  font-size: 15px;
  font-weight: 300;
}

.distill-site-nav a {
  color: inherit;
  text-decoration: none;
}

.distill-site-nav a:hover {
  color: white;
}

@media print {
  .distill-site-nav {
    display: none;
  }
}

.distill-site-header {

}

.distill-site-footer {

}


/* Site Header */

.distill-site-header {
  width: 100%;
  box-sizing: border-box;
  z-index: 3;
}

.distill-site-header .nav-left {
  display: inline-block;
  margin-left: 8px;
}

@media screen and (max-width: 768px) {
  .distill-site-header .nav-left {
    margin-left: 0;
  }
}


.distill-site-header .nav-right {
  float: right;
  margin-right: 8px;
}

.distill-site-header a,
.distill-site-header .title {
  display: inline-block;
  text-align: center;
  padding: 14px 10px 14px 10px;
}

.distill-site-header .title {
  font-size: 18px;
  min-width: 150px;
}

.distill-site-header .logo {
  padding: 0;
}

.distill-site-header .logo img {
  display: none;
  max-height: 20px;
  width: auto;
  margin-bottom: -4px;
}

.distill-site-header .nav-image img {
  max-height: 18px;
  width: auto;
  display: inline-block;
  margin-bottom: -3px;
}



@media screen and (min-width: 1000px) {
  .distill-site-header .logo img {
    display: inline-block;
  }
  .distill-site-header .nav-left {
    margin-left: 20px;
  }
  .distill-site-header .nav-right {
    margin-right: 20px;
  }
  .distill-site-header .title {
    padding-left: 12px;
  }
}


.distill-site-header .nav-toggle {
  display: none;
}

.nav-dropdown {
  display: inline-block;
  position: relative;
}

.nav-dropdown .nav-dropbtn {
  border: none;
  outline: none;
  color: rgba(255, 255, 255, 0.8);
  padding: 16px 10px;
  background-color: transparent;
  font-family: inherit;
  font-size: inherit;
  font-weight: inherit;
  margin: 0;
  margin-top: 1px;
  z-index: 2;
}

.nav-dropdown-content {
  display: none;
  position: absolute;
  background-color: white;
  min-width: 200px;
  border: 1px solid rgba(0,0,0,0.15);
  border-radius: 4px;
  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
  z-index: 1;
  margin-top: 2px;
  white-space: nowrap;
  padding-top: 4px;
  padding-bottom: 4px;
}

.nav-dropdown-content hr {
  margin-top: 4px;
  margin-bottom: 4px;
  border: none;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.nav-dropdown-active {
  display: block;
}

.nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
  color: black;
  padding: 6px 24px;
  text-decoration: none;
  display: block;
  text-align: left;
}

.nav-dropdown-content .nav-dropdown-header {
  display: block;
  padding: 5px 24px;
  padding-bottom: 0;
  text-transform: uppercase;
  font-size: 14px;
  color: #999999;
  white-space: nowrap;
}

.nav-dropdown:hover .nav-dropbtn {
  color: white;
}

.nav-dropdown-content a:hover {
  background-color: #ddd;
  color: black;
}

.nav-right .nav-dropdown-content {
  margin-left: -45%;
  right: 0;
}

@media screen and (max-width: 768px) {
  .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
  .distill-site-header a.nav-toggle {
    float: right;
    display: block;
  }
  .distill-site-header .title {
    margin-left: 0;
  }
  .distill-site-header .nav-right {
    margin-right: 0;
  }
  .distill-site-header {
    overflow: hidden;
  }
  .nav-right .nav-dropdown-content {
    margin-left: 0;
  }
}


@media screen and (max-width: 768px) {
  .distill-site-header.responsive {position: relative; min-height: 500px; }
  .distill-site-header.responsive a.nav-toggle {
    position: absolute;
    right: 0;
    top: 0;
  }
  .distill-site-header.responsive a,
  .distill-site-header.responsive .nav-dropdown {
    display: block;
    text-align: left;
  }
  .distill-site-header.responsive .nav-left,
  .distill-site-header.responsive .nav-right {
    width: 100%;
  }
  .distill-site-header.responsive .nav-dropdown {float: none;}
  .distill-site-header.responsive .nav-dropdown-content {position: relative;}
  .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
    display: block;
    width: 100%;
    text-align: left;
  }
}

/* Site Footer */

.distill-site-footer {
  width: 100%;
  overflow: hidden;
  box-sizing: border-box;
  z-index: 3;
  margin-top: 30px;
  padding-top: 30px;
  padding-bottom: 30px;
  text-align: center;
}

/* Headroom */

d-title {
  padding-top: 6rem;
}

@media print {
  d-title {
    padding-top: 4rem;
  }
}

.headroom {
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
}

.headroom--transition {
  transition: all .4s ease-in-out;
}

.headroom--unpinned {
  top: -100px;
}

.headroom--pinned {
  top: 0;
}

/* adjust viewport for navbar height */
/* helps vertically center bootstrap (non-distill) content */
.min-vh-100 {
  min-height: calc(100vh - 100px) !important;
}

</style>

<script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="../../site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"/>
<link href="../../site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"/>
<script src="../../site_libs/headroom-0.9.4/headroom.min.js"></script>
<!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

<style type="text/css">

body {
  background-color: white;
}

.pandoc-table {
  width: 100%;
}

.pandoc-table>caption {
  margin-bottom: 10px;
}

.pandoc-table th:not([align]) {
  text-align: left;
}

.pagedtable-footer {
  font-size: 15px;
}

d-byline .byline {
  grid-template-columns: 2fr 2fr;
}

d-byline .byline h3 {
  margin-block-start: 1.5em;
}

d-byline .byline .authors-affiliations h3 {
  margin-block-start: 0.5em;
}

.authors-affiliations .orcid-id {
  width: 16px;
  height:16px;
  margin-left: 4px;
  margin-right: 4px;
  vertical-align: middle;
  padding-bottom: 2px;
}

d-title .dt-tags {
  margin-top: 1em;
  grid-column: text;
}

.dt-tags .dt-tag {
  text-decoration: none;
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0em 0.4em;
  margin-right: 0.5em;
  margin-bottom: 0.4em;
  font-size: 70%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

d-article table.gt_table td,
d-article table.gt_table th {
  border-bottom: none;
  font-size: 100%;
}

.html-widget {
  margin-bottom: 2.0em;
}

.l-screen-inset {
  padding-right: 16px;
}

.l-screen .caption {
  margin-left: 10px;
}

.shaded {
  background: rgb(247, 247, 247);
  padding-top: 20px;
  padding-bottom: 20px;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .html-widget {
  margin-bottom: 0;
  border: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .shaded-content {
  background: white;
}

.text-output {
  margin-top: 0;
  line-height: 1.5em;
}

.hidden {
  display: none !important;
}

hr.section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  margin: 0px;
}


d-byline {
  border-top: none;
}

d-article {
  padding-top: 2.5rem;
  padding-bottom: 30px;
  border-top: none;
}

d-appendix {
  padding-top: 30px;
}

d-article>p>img {
  width: 100%;
}

d-article h2 {
  margin: 1rem 0 1.5rem 0;
}

d-article h3 {
  margin-top: 1.5rem;
}

d-article iframe {
  border: 1px solid rgba(0, 0, 0, 0.1);
  margin-bottom: 2.0em;
  width: 100%;
}

/* Tweak code blocks */

d-article div.sourceCode code,
d-article pre code {
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
}

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: auto;
}

d-article div.sourceCode {
  background-color: white;
}

d-article div.sourceCode pre {
  padding-left: 10px;
  font-size: 12px;
  border-left: 2px solid rgba(0,0,0,0.1);
}

d-article pre {
  font-size: 12px;
  color: black;
  background: none;
  margin-top: 0;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

d-article pre a {
  border-bottom: none;
}

d-article pre a:hover {
  border-bottom: none;
  text-decoration: underline;
}

d-article details {
  grid-column: text;
  margin-bottom: 0.8em;
}

@media(min-width: 768px) {

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: visible !important;
}

d-article div.sourceCode pre {
  padding-left: 18px;
  font-size: 14px;
}

/* tweak for Pandoc numbered line within distill */
d-article pre.numberSource code > span {
    left: -2em;
}

d-article pre {
  font-size: 14px;
}

}

figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

/* CSS for d-contents */

.d-contents {
  grid-column: text;
  color: rgba(0,0,0,0.8);
  font-size: 0.9em;
  padding-bottom: 1em;
  margin-bottom: 1em;
  padding-bottom: 0.5em;
  margin-bottom: 1em;
  padding-left: 0.25em;
  justify-self: start;
}

@media(min-width: 1000px) {
  .d-contents.d-contents-float {
    height: 0;
    grid-column-start: 1;
    grid-column-end: 4;
    justify-self: center;
    padding-right: 3em;
    padding-left: 2em;
  }
}

.d-contents nav h3 {
  font-size: 18px;
  margin-top: 0;
  margin-bottom: 1em;
}

.d-contents li {
  list-style-type: none
}

.d-contents nav > ul {
  padding-left: 0;
}

.d-contents ul {
  padding-left: 1em
}

.d-contents nav ul li {
  margin-top: 0.6em;
  margin-bottom: 0.2em;
}

.d-contents nav a {
  font-size: 13px;
  border-bottom: none;
  text-decoration: none
  color: rgba(0, 0, 0, 0.8);
}

.d-contents nav a:hover {
  text-decoration: underline solid rgba(0, 0, 0, 0.6)
}

.d-contents nav > ul > li > a {
  font-weight: 600;
}

.d-contents nav > ul > li > ul {
  font-weight: inherit;
}

.d-contents nav > ul > li > ul > li {
  margin-top: 0.2em;
}


.d-contents nav ul {
  margin-top: 0;
  margin-bottom: 0.25em;
}

.d-article-with-toc h2:nth-child(2) {
  margin-top: 0;
}


/* Figure */

.figure {
  position: relative;
  margin-bottom: 2.5em;
  margin-top: 1.5em;
}

.figure .caption {
  color: rgba(0, 0, 0, 0.6);
  font-size: 12px;
  line-height: 1.5em;
}

.figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

.figure .caption a {
  color: rgba(0, 0, 0, 0.6);
}

.figure .caption b,
.figure .caption strong, {
  font-weight: 600;
  color: rgba(0, 0, 0, 1.0);
}

/* Citations */

d-article .citation {
  color: inherit;
  cursor: inherit;
}

div.hanging-indent{
  margin-left: 1em; text-indent: -1em;
}

/* Citation hover box */

.tippy-box[data-theme~=light-border] {
  background-color: rgba(250, 250, 250, 0.95);
}

.tippy-content > p {
  margin-bottom: 0;
  padding: 2px;
}


/* Tweak 1000px media break to show more text */

@media(min-width: 1000px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 16px;
  }

  .grid {
    grid-column-gap: 16px;
  }

  d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
  }
  figure .caption, .figure .caption, figure figcaption {
    font-size: 13px;
  }
}

@media(min-width: 1180px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 32px;
  }

  .grid {
    grid-column-gap: 32px;
  }
}


/* Get the citation styles for the appendix (not auto-injected on render since
   we do our own rendering of the citation appendix) */

d-appendix .citation-appendix,
.d-appendix .citation-appendix {
  font-size: 11px;
  line-height: 15px;
  border-left: 1px solid rgba(0, 0, 0, 0.1);
  padding-left: 18px;
  border: 1px solid rgba(0,0,0,0.1);
  background: rgba(0, 0, 0, 0.02);
  padding: 10px 18px;
  border-radius: 3px;
  color: rgba(150, 150, 150, 1);
  overflow: hidden;
  margin-top: -12px;
  white-space: pre-wrap;
  word-wrap: break-word;
}

/* Include appendix styles here so they can be overridden */

d-appendix {
  contain: layout style;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-top: 60px;
  margin-bottom: 0;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  color: rgba(0,0,0,0.5);
  padding-top: 60px;
  padding-bottom: 48px;
}

d-appendix h3 {
  grid-column: page-start / text-start;
  font-size: 15px;
  font-weight: 500;
  margin-top: 1em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.65);
}

d-appendix h3 + * {
  margin-top: 1em;
}

d-appendix ol {
  padding: 0 0 0 15px;
}

@media (min-width: 768px) {
  d-appendix ol {
    padding: 0 0 0 30px;
    margin-left: -30px;
  }
}

d-appendix li {
  margin-bottom: 1em;
}

d-appendix a {
  color: rgba(0, 0, 0, 0.6);
}

d-appendix > * {
  grid-column: text;
}

d-appendix > d-footnote-list,
d-appendix > d-citation-list,
d-appendix > distill-appendix {
  grid-column: screen;
}

/* Include footnote styles here so they can be overridden */

d-footnote-list {
  contain: layout style;
}

d-footnote-list > * {
  grid-column: text;
}

d-footnote-list a.footnote-backlink {
  color: rgba(0,0,0,0.3);
  padding-left: 0.5em;
}



/* Anchor.js */

.anchorjs-link {
  /*transition: all .25s linear; */
  text-decoration: none;
  border-bottom: none;
}
*:hover > .anchorjs-link {
  margin-left: -1.125em !important;
  text-decoration: none;
  border-bottom: none;
}

/* Social footer */

.social_footer {
  margin-top: 30px;
  margin-bottom: 0;
  color: rgba(0,0,0,0.67);
}

.disqus-comments {
  margin-right: 30px;
}

.disqus-comment-count {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  cursor: pointer;
}

#disqus_thread {
  margin-top: 30px;
}

.article-sharing a {
  border-bottom: none;
  margin-right: 8px;
}

.article-sharing a:hover {
  border-bottom: none;
}

.sidebar-section.subscribe {
  font-size: 12px;
  line-height: 1.6em;
}

.subscribe p {
  margin-bottom: 0.5em;
}


.article-footer .subscribe {
  font-size: 15px;
  margin-top: 45px;
}


.sidebar-section.custom {
  font-size: 12px;
  line-height: 1.6em;
}

.custom p {
  margin-bottom: 0.5em;
}

/* Styles for listing layout (hide title) */
.layout-listing d-title, .layout-listing .d-title {
  display: none;
}

/* Styles for posts lists (not auto-injected) */


.posts-with-sidebar {
  padding-left: 45px;
  padding-right: 45px;
}

.posts-list .description h2,
.posts-list .description p {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

.posts-list .description h2 {
  font-weight: 700;
  border-bottom: none;
  padding-bottom: 0;
}

.posts-list h2.post-tag {
  border-bottom: 1px solid rgba(0, 0, 0, 0.2);
  padding-bottom: 12px;
}
.posts-list {
  margin-top: 60px;
  margin-bottom: 24px;
}

.posts-list .post-preview {
  text-decoration: none;
  overflow: hidden;
  display: block;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  padding: 24px 0;
}

.post-preview-last {
  border-bottom: none !important;
}

.posts-list .posts-list-caption {
  grid-column: screen;
  font-weight: 400;
}

.posts-list .post-preview h2 {
  margin: 0 0 6px 0;
  line-height: 1.2em;
  font-style: normal;
  font-size: 24px;
}

.posts-list .post-preview p {
  margin: 0 0 12px 0;
  line-height: 1.4em;
  font-size: 16px;
}

.posts-list .post-preview .thumbnail {
  box-sizing: border-box;
  margin-bottom: 24px;
  position: relative;
  max-width: 500px;
}
.posts-list .post-preview img {
  width: 100%;
  display: block;
}

.posts-list .metadata {
  font-size: 12px;
  line-height: 1.4em;
  margin-bottom: 18px;
}

.posts-list .metadata > * {
  display: inline-block;
}

.posts-list .metadata .publishedDate {
  margin-right: 2em;
}

.posts-list .metadata .dt-authors {
  display: block;
  margin-top: 0.3em;
  margin-right: 2em;
}

.posts-list .dt-tags {
  display: block;
  line-height: 1em;
}

.posts-list .dt-tags .dt-tag {
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0.3em 0.4em;
  margin-right: 0.2em;
  margin-bottom: 0.4em;
  font-size: 60%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

.posts-list img {
  opacity: 1;
}

.posts-list img[data-src] {
  opacity: 0;
}

.posts-more {
  clear: both;
}


.posts-sidebar {
  font-size: 16px;
}

.posts-sidebar h3 {
  font-size: 16px;
  margin-top: 0;
  margin-bottom: 0.5em;
  font-weight: 400;
  text-transform: uppercase;
}

.sidebar-section {
  margin-bottom: 30px;
}

.categories ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.categories li {
  color: rgba(0, 0, 0, 0.8);
  margin-bottom: 0;
}

.categories li>a {
  border-bottom: none;
}

.categories li>a:hover {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
}

.categories .active {
  font-weight: 600;
}

.categories .category-count {
  color: rgba(0, 0, 0, 0.4);
}


@media(min-width: 768px) {
  .posts-list .post-preview h2 {
    font-size: 26px;
  }
  .posts-list .post-preview .thumbnail {
    float: right;
    width: 30%;
    margin-bottom: 0;
  }
  .posts-list .post-preview .description {
    float: left;
    width: 45%;
  }
  .posts-list .post-preview .metadata {
    float: left;
    width: 20%;
    margin-top: 8px;
  }
  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.5em;
    font-size: 16px;
  }
  .posts-with-sidebar .posts-list {
    float: left;
    width: 75%;
  }
  .posts-with-sidebar .posts-sidebar {
    float: right;
    width: 20%;
    margin-top: 60px;
    padding-top: 24px;
    padding-bottom: 24px;
  }
}


/* Improve display for browsers without grid (IE/Edge <= 15) */

.downlevel {
  line-height: 1.6em;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  margin: 0;
}

.downlevel .d-title {
  padding-top: 6rem;
  padding-bottom: 1.5rem;
}

.downlevel .d-title h1 {
  font-size: 50px;
  font-weight: 700;
  line-height: 1.1em;
  margin: 0 0 0.5rem;
}

.downlevel .d-title p {
  font-weight: 300;
  font-size: 1.2rem;
  line-height: 1.55em;
  margin-top: 0;
}

.downlevel .d-byline {
  padding-top: 0.8em;
  padding-bottom: 0.8em;
  font-size: 0.8rem;
  line-height: 1.8em;
}

.downlevel .section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
}

.downlevel .d-article {
  font-size: 1.06rem;
  line-height: 1.7em;
  padding-top: 1rem;
  padding-bottom: 2rem;
}


.downlevel .d-appendix {
  padding-left: 0;
  padding-right: 0;
  max-width: none;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.5);
  padding-top: 40px;
  padding-bottom: 48px;
}

.downlevel .footnotes ol {
  padding-left: 13px;
}

.downlevel .base-grid,
.downlevel .distill-header,
.downlevel .d-title,
.downlevel .d-abstract,
.downlevel .d-article,
.downlevel .d-appendix,
.downlevel .distill-appendix,
.downlevel .d-byline,
.downlevel .d-footnote-list,
.downlevel .d-citation-list,
.downlevel .distill-footer,
.downlevel .appendix-bottom,
.downlevel .posts-container {
  padding-left: 40px;
  padding-right: 40px;
}

@media(min-width: 768px) {
  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
  padding-left: 150px;
  padding-right: 150px;
  max-width: 900px;
}
}

.downlevel pre code {
  display: block;
  border-left: 2px solid rgba(0, 0, 0, .1);
  padding: 0 0 0 20px;
  font-size: 14px;
}

.downlevel code, .downlevel pre {
  color: black;
  background: none;
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

.downlevel .posts-list .post-preview {
  color: inherit;
}



</style>

<script type="application/javascript">

function is_downlevel_browser() {
  if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                 window.navigator.userAgent)) {
    return true;
  } else {
    return window.load_distill_framework === undefined;
  }
}

// show body when load is complete
function on_load_complete() {

  // add anchors
  if (window.anchors) {
    window.anchors.options.placement = 'left';
    window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
  }


  // set body to visible
  document.body.style.visibility = 'visible';

  // force redraw for leaflet widgets
  if (window.HTMLWidgets) {
    var maps = window.HTMLWidgets.findAll(".leaflet");
    $.each(maps, function(i, el) {
      var map = this.getMap();
      map.invalidateSize();
      map.eachLayer(function(layer) {
        if (layer instanceof L.TileLayer)
          layer.redraw();
      });
    });
  }

  // trigger 'shown' so htmlwidgets resize
  $('d-article').trigger('shown');
}

function init_distill() {

  init_common();

  // create front matter
  var front_matter = $('<d-front-matter></d-front-matter>');
  $('#distill-front-matter').wrap(front_matter);

  // create d-title
  $('.d-title').changeElementType('d-title');

  // separator
  var separator = '<hr class="section-separator" style="clear: both"/>';
  // prepend separator above appendix
  $('.d-byline').before(separator);
  $('.d-article').before(separator);

  // create d-byline
  var byline = $('<d-byline></d-byline>');
  $('.d-byline').replaceWith(byline);

  // create d-article
  var article = $('<d-article></d-article>');
  $('.d-article').wrap(article).children().unwrap();

  // move posts container into article
  $('.posts-container').appendTo($('d-article'));

  // create d-appendix
  $('.d-appendix').changeElementType('d-appendix');

  // flag indicating that we have appendix items
  var appendix = $('.appendix-bottom').children('h3').length > 0;

  // replace footnotes with <d-footnote>
  $('.footnote-ref').each(function(i, val) {
    appendix = true;
    var href = $(this).attr('href');
    var id = href.replace('#', '');
    var fn = $('#' + id);
    var fn_p = $('#' + id + '>p');
    fn_p.find('.footnote-back').remove();
    var text = fn_p.html();
    var dtfn = $('<d-footnote></d-footnote>');
    dtfn.html(text);
    $(this).replaceWith(dtfn);
  });
  // remove footnotes
  $('.footnotes').remove();

  // move refs into #references-listing
  $('#references-listing').replaceWith($('#refs'));

  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    var id = $(this).attr('id');
    $('.d-contents a[href="#' + id + '"]').parent().remove();
    appendix = true;
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
  });

  // show d-appendix if we have appendix content
  $("d-appendix").css('display', appendix ? 'grid' : 'none');

  // localize layout chunks to just output
  $('.layout-chunk').each(function(i, val) {

    // capture layout
    var layout = $(this).attr('data-layout');

    // apply layout to markdown level block elements
    var elements = $(this).children().not('details, div.sourceCode, pre, script');
    elements.each(function(i, el) {
      var layout_div = $('<div class="' + layout + '"></div>');
      if (layout_div.hasClass('shaded')) {
        var shaded_content = $('<div class="shaded-content"></div>');
        $(this).wrap(shaded_content);
        $(this).parent().wrap(layout_div);
      } else {
        $(this).wrap(layout_div);
      }
    });


    // unwrap the layout-chunk div
    $(this).children().unwrap();
  });

  // remove code block used to force  highlighting css
  $('.distill-force-highlighting-css').parent().remove();

  // remove empty line numbers inserted by pandoc when using a
  // custom syntax highlighting theme, except when numbering line
  // in code chunk
  $('pre:not(.numberLines) code.sourceCode a:empty').remove();

  // load distill framework
  load_distill_framework();

  // wait for window.distillRunlevel == 4 to do post processing
  function distill_post_process() {

    if (!window.distillRunlevel || window.distillRunlevel < 4)
      return;

    // hide author/affiliations entirely if we have no authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;
    if (!have_authors)
      $('d-byline').addClass('hidden');

    // article with toc class
    $('.d-contents').parent().addClass('d-article-with-toc');

    // strip links that point to #
    $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

    // add orcid ids
    $('.authors-affiliations').find('.author').each(function(i, el) {
      var orcid_id = front_matter.authors[i].orcidID;
      var author_name = front_matter.authors[i].author
      if (orcid_id) {
        var a = $('<a></a>');
        a.attr('href', 'https://orcid.org/' + orcid_id);
        var img = $('<img></img>');
        img.addClass('orcid-id');
        img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
        img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
        a.append(img);
        $(this).append(a);
      }
    });

    // hide elements of author/affiliations grid that have no value
    function hide_byline_column(caption) {
      $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
    }

    // affiliations
    var have_affiliations = false;
    for (var i = 0; i<front_matter.authors.length; ++i) {
      var author = front_matter.authors[i];
      if (author.affiliation !== "&nbsp;") {
        have_affiliations = true;
        break;
      }
    }
    if (!have_affiliations)
      $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

    // published date
    if (!front_matter.publishedDate)
      hide_byline_column("Published");

    // document object identifier
    var doi = $('d-byline').find('h3:contains("DOI")');
    var doi_p = doi.next().empty();
    if (!front_matter.doi) {
      // if we have a citation and valid citationText then link to that
      if ($('#citation').length > 0 && front_matter.citationText) {
        doi.html('Citation');
        $('<a href="#citation"></a>')
          .text(front_matter.citationText)
          .appendTo(doi_p);
      } else {
        hide_byline_column("DOI");
      }
    } else {
      $('<a></a>')
         .attr('href', "https://doi.org/" + front_matter.doi)
         .html(front_matter.doi)
         .appendTo(doi_p);
    }

     // change plural form of authors/affiliations
    if (front_matter.authors.length === 1) {
      var grid = $('.authors-affiliations');
      grid.children('h3:contains("Authors")').text('Author');
      grid.children('h3:contains("Affiliations")').text('Affiliation');
    }

    // remove d-appendix and d-footnote-list local styles
    $('d-appendix > style:first-child').remove();
    $('d-footnote-list > style:first-child').remove();

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // hoverable references
    $('span.citation[data-cites]').each(function() {
      const citeChild = $(this).children()[0]
      // Do not process if @xyz has been used without escaping and without bibliography activated
      // https://github.com/rstudio/distill/issues/466
      if (citeChild === undefined) return true

      if (citeChild.nodeName == "D-FOOTNOTE") {
        var fn = citeChild
        $(this).html(fn.shadowRoot.querySelector("sup"))
        $(this).id = fn.id
        fn.remove()
      }
      var refs = $(this).attr('data-cites').split(" ");
      var refHtml = refs.map(function(ref) {
        // Could use CSS.escape too here, we insure backward compatibility in navigator
        return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
      }).join("\n");
      window.tippy(this, {
        allowHTML: true,
        content: refHtml,
        maxWidth: 500,
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start'
      });
    });

    // fix footnotes in tables (#411)
    // replacing broken distill.pub feature
    $('table d-footnote').each(function() {
      // we replace internal showAtNode methode which is triggered when hovering a footnote
      this.hoverBox.showAtNode = function(node) {
        // ported from https://github.com/distillpub/template/pull/105/files
        calcOffset = function(elem) {
            let x = elem.offsetLeft;
            let y = elem.offsetTop;
            // Traverse upwards until an `absolute` element is found or `elem`
            // becomes null.
            while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                x += elem.offsetLeft;
                y += elem.offsetTop;
            }

            return { left: x, top: y };
        }
        // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
        const bbox = node.getBoundingClientRect();
        const offset = calcOffset(node);
        this.show([offset.left + bbox.width, offset.top + bbox.height]);
      }
    })

    // clear polling timer
    clearInterval(tid);

    // show body now that everything is ready
    on_load_complete();
  }

  var tid = setInterval(distill_post_process, 50);
  distill_post_process();

}

function init_downlevel() {

  init_common();

   // insert hr after d-title
  $('.d-title').after($('<hr class="section-separator"/>'));

  // check if we have authors
  var front_matter = JSON.parse($("#distill-front-matter").html());
  var have_authors = front_matter.authors && front_matter.authors.length > 0;

  // manage byline/border
  if (!have_authors)
    $('.d-byline').remove();
  $('.d-byline').after($('<hr class="section-separator"/>'));
  $('.d-byline a').remove();

  // remove toc
  $('.d-contents').remove();

  // move appendix elements
  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
  });


  // inject headers into references and footnotes
  var refs_header = $('<h3></h3>');
  refs_header.text('References');
  $('#refs').prepend(refs_header);

  var footnotes_header = $('<h3></h3');
  footnotes_header.text('Footnotes');
  $('.footnotes').children('hr').first().replaceWith(footnotes_header);

  // move appendix-bottom entries to the bottom
  $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
  $('.appendix-bottom').remove();

  // remove appendix if it's empty
  if ($('.d-appendix').children().length === 0)
    $('.d-appendix').remove();

  // prepend separator above appendix
  $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

  // trim code
  $('pre>code').each(function(i, val) {
    $(this).html($.trim($(this).html()));
  });

  // move posts-container right before article
  $('.posts-container').insertBefore($('.d-article'));

  $('body').addClass('downlevel');

  on_load_complete();
}


function init_common() {

  // jquery plugin to change element types
  (function($) {
    $.fn.changeElementType = function(newType) {
      var attrs = {};

      $.each(this[0].attributes, function(idx, attr) {
        attrs[attr.nodeName] = attr.nodeValue;
      });

      this.replaceWith(function() {
        return $("<" + newType + "/>", attrs).append($(this).contents());
      });
    };
  })(jQuery);

  // prevent underline for linked images
  $('a > img').parent().css({'border-bottom' : 'none'});

  // mark non-body figures created by knitr chunks as 100% width
  $('.layout-chunk').each(function(i, val) {
    var figures = $(this).find('img, .html-widget');
    // ignore leaflet img layers (#106)
    figures = figures.filter(':not(img[class*="leaflet"])')
    if ($(this).attr('data-layout') !== "l-body") {
      figures.css('width', '100%');
    } else {
      figures.css('max-width', '100%');
      figures.filter("[width]").each(function(i, val) {
        var fig = $(this);
        fig.css('width', fig.attr('width') + 'px');
      });

    }
  });

  // auto-append index.html to post-preview links in file: protocol
  // and in rstudio ide preview
  $('.post-preview').each(function(i, val) {
    if (window.location.protocol === "file:")
      $(this).attr('href', $(this).attr('href') + "index.html");
  });

  // get rid of index.html references in header
  if (window.location.protocol !== "file:") {
    $('.distill-site-header a[href]').each(function(i,val) {
      $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
    });
  }

  // add class to pandoc style tables
  $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
  $('.kable-table').children('table').addClass('pandoc-table');

  // add figcaption style to table captions
  $('caption').parent('table').addClass("figcaption");

  // initialize posts list
  if (window.init_posts_list)
    window.init_posts_list();

  // implmement disqus comment link
  $('.disqus-comment-count').click(function() {
    window.headroom_prevent_pin = true;
    $('#disqus_thread').toggleClass('hidden');
    if (!$('#disqus_thread').hasClass('hidden')) {
      var offset = $(this).offset();
      $(window).resize();
      $('html, body').animate({
        scrollTop: offset.top - 35
      });
    }
  });
}

document.addEventListener('DOMContentLoaded', function() {
  if (is_downlevel_browser())
    init_downlevel();
  else
    window.addEventListener('WebComponentsReady', init_distill);
});

</script>

<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.5);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

.distill-site-footer {
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */
</style>
<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.7);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #ffffff;
}

.distill-site-footer {
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #fafafa;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */

.nav-right > a {
  text-transform: uppercase;
}

d-title h1, d-title p, d-title figure,
d-abstract p, d-abstract b {
  grid-column: page;
}

.rj-blue {
  color: #2467bb;
}

ul li {
  line-height: 1.6;
  margin-top: 0em;
  margin-bottom: 0em;
}</style>
<style type="text/css">
/* base style */

/* FONT FAMILIES */

:root {
  --heading-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  --mono-default: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  --body-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

body,
.posts-list .post-preview p,
.posts-list .description p {
  font-family: var(--body-font), var(--body-default);
}

h1, h2, h3, h4, h5, h6,
.posts-list .post-preview h2,
.posts-list .description h2 {
  font-family: var(--heading-font), var(--heading-default);
}

d-article div.sourceCode code,
d-article pre code {
  font-family: var(--mono-font), var(--mono-default);
}


/*-- TITLE --*/
d-title h1,
.posts-list > h1 {
  color: var(--title-color, black);
}

d-title h1 {
  font-size: var(--title-size, 50px);
}

/*-- HEADERS --*/
d-article h1,
d-article h2,
d-article h3,
d-article h4,
d-article h5,
d-article h6 {
  color: var(--header-color, rgba(0, 0, 0, 0.8));
}

/*-- BODY --*/
d-article > p,  /* only text inside of <p> tags */
d-article > ul, /* lists */
d-article > ol {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
  font-size: var(--body-size, 1.06rem);
}


/*-- CODE --*/
d-article div.sourceCode code,
d-article pre code {
  font-size: var(--code-size, 14px);
}

/*-- ASIDE --*/
d-article aside {
  font-size: var(--aside-size, 12px);
  color: var(--aside-color, rgba(0, 0, 0, 0.6));
}

/*-- FIGURE CAPTIONS --*/
figure .caption,
figure figcaption,
.figure .caption {
  font-size: var(--fig-cap-size, 13px);
  color: var(--fig-cap-color, rgba(0, 0, 0, 0.6));
}

/*-- METADATA --*/
d-byline h3 {
  font-size: var(--heading-size, 0.6rem);
  color: var(--heading-color, rgba(0, 0, 0, 0.5));
}

d-byline {
  font-size: var(--body-size, 0.8rem);
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

d-byline a,
d-article d-byline a {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

/*-- TABLE OF CONTENTS --*/
.d-contents nav h3 {
  font-size: var(--heading-size, 18px);
}

.d-contents nav a {
  font-size: var(--contents-size, 13px);
}

/*-- APPENDIX --*/
d-appendix h3 {
  font-size: var(--heading-size, 15px);
  color: var(--heading-color, rgba(0, 0, 0, 0.65));
}

d-appendix {
  font-size: var(--text-size, 0.8em);
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

d-appendix d-footnote-list a.footnote-backlink {
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

/*-- WEBSITE HEADER + FOOTER --*/
.distill-site-header .title {
  font-size: var(--title-size, 18px);
  font-family: var(--navbar-font), var(--heading-default);
}

.distill-site-header a,
.nav-dropdown .nav-dropbtn {
  font-family: var(--navbar-font), var(--heading-default);
}

.nav-dropdown .nav-dropbtn {
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  font-size: var(--text-size, 15px);
}

.distill-site-header a:hover,
.nav-dropdown:hover .nav-dropbtn {
  color: var(--hover-color, white);
}

.distill-site-header {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer a:hover {
  color: var(--hover-color, white);
}</style>
<!--/radix_placeholder_distill-->
  <script src="../../site_libs/header-attrs-2.30/header-attrs.js"></script>
  <script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="../../site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="../../site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="../../site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="../../site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="../../site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="../../site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="../../site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="../../site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
<!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Mathjax config (add automatic linebreaks when supported)
      // MathJax = {
      //    tex: {
      //        inlineMath: [['$', '$'], ['\\(', '\\)']],
      //        displayMath: [['$$', '$$'], ['\\[', '\\]']],
      //        tags: 'ams',
      //        multline: true,
      //    },
      //    options: {
      //        linebreaks: { automatic: true },
      //    },
      // };

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      
      // Conditionally add supplementary files button
      if (document.getElementById('supplementary-materials')) {
        // create element safely
        const btn_suppl = document.createElement('a');
        btn_suppl.href = slug + '.zip';
        btn_suppl.className = 'dt-tag';
        btn_suppl.innerHTML = '<i class="fas fa-file-zipper"></i> Supplement';
        buttons.append(btn_suppl);
      }

      // adds Abstract: in front of the first <p> in the title section --
      // unless it happens to be the subtitle (FIXME: this is a bad hack - can't distill do this?)
      var tpar = $("d-title p:not(:empty)").filter(function() {
        return !$(this).hasClass("subtitle");
      }).first();
      if (tpar) {
        const abstract = $('<d-abstract>')
        abstract.append('<b>Abstract:</b><br>')
        abstract.append(tpar) // Move description to d-abstract
        $("d-title p:empty").remove() // Remove empty paragraphs after title
        abstract.append(buttons)
        abstract.insertAfter($('d-title')) // Add abstract section after title */
      }

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
    
    d-appendix .citation-appendix,
    .d-appendix .citation-appendix {
      color: rgb(60, 60, 60);
    }

    d-article h2 {
      border-bottom: 0px solid rgba(0, 0, 0, 0.1);
      padding-bottom: 0rem;
    }
    d-article h3 {
      font-size: 20px;
    }
    d-article h4 {
      font-size: 18px;
      text-transform: none;
    }

    @media (min-width: 1024px) {
      d-article h2 {
        font-size: 32px;
      }
      d-article h3 {
        font-size: 24px;
      }
      d-article h4 {
        font-size: 20px;
      }
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"rqPen: An R Package for Penalized Quantile Regression","description":"Quantile regression directly models a conditional quantile of interest. The R package rqPen provides penalized quantile regression estimation using the lasso, elastic net, adaptive lasso, SCAD, and MCP penalties. It also provides extensions to group penalties, both for groups of predictors and grouping variable selection across quantiles. This paper presents the different approaches to penalized quantile regression and how to use the provided methods in rqPen.","doi":"10.32614/RJ-2025-017","authors":[{"author":"Ben Sherwood","authorURL":"#","affiliation":"University of Kansas","affiliationURL":"#","orcidID":""},{"author":"Shaobo Li","authorURL":"#","affiliation":"University of Kansas","affiliationURL":"#","orcidID":""},{"author":"Adam Maidman","authorURL":"#","affiliation":"University of Minnesota","affiliationURL":"#","orcidID":""}],"publishedDate":"2025-09-09T00:00:00.000+00:00","citationText":"Sherwood, et al., 2025"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a class="logo" href="../../index.html">
<img src="../../resources/rlogo.png" alt="Logo"/>
</a>
<a href="../../index.html" class="title">The R Journal</a>
</div>
<div class="nav-right">
<a href="../../index.html">Home</a>
<a href="../../issues/2025-3">Current</a>
<a href="../../issues.html">Issues</a>
<a href="../../news.html">News</a>
<a href="../../submissions.html">Submit</a>
<a href="../../contribute.html">Contribute</a>
<a href="../../editors.html">Editorial board</a>
<a href="../../articles.xml">
<i class="fa fa-rss" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>rqPen: An R Package for Penalized Quantile Regression</h1>

<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>Quantile regression directly models a conditional quantile of interest. The R package rqPen provides penalized quantile regression estimation using the lasso, elastic net, adaptive lasso, SCAD, and MCP penalties. It also provides extensions to group penalties, both for groups of predictors and grouping variable selection across quantiles. This paper presents the different approaches to penalized quantile regression and how to use the provided methods in rqPen.</p></p>
</div>

<div class="d-byline">
  Ben Sherwood  (University of Kansas)
  
,   Shaobo Li  (University of Kansas)
  
,   Adam Maidman  (University of Minnesota)
  
<br/>2025-09-09
</div>

<div class="d-article">
<h2 data-number="1" id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>The package <a href="https://cran.r-project.org/package=rqPen">rqPen</a> allows users to model conditional quantiles using a penalized quantile regression approach. It supports a wide range of penalties and includes cross-validation and information criterion approaches for selecting tuning parameters. <span class="citation" data-cites="origQR">Koenker and Bassett (<a href="#ref-origQR" role="doc-biblioref">1978</a>)</span> proposed quantile regression as a robust alternative to mean regression that
directly models a conditional quantile of interest without the need for assumptions about the
distribution or moment conditions for the error term. Since <span class="citation" data-cites="lasso">Tibshirani (<a href="#ref-lasso" role="doc-biblioref">1996</a>)</span> introduced the lasso penalty, investigating the combination of different penalties and loss functions has been an active area of interest. Penalties provided in <a href="https://cran.r-project.org/package=rqPen">rqPen</a> are lasso, elastic net <span class="citation" data-cites="zou05">(<a href="#ref-zou05" role="doc-biblioref">Zou and Hastie 2005</a>)</span>, SCAD <span class="citation" data-cites="fanLi">(<a href="#ref-fanLi" role="doc-biblioref">Fan and Li 2001</a>)</span>, MCP <span class="citation" data-cites="mcp">(<a href="#ref-mcp" role="doc-biblioref">Zhang 2010</a>)</span>, adaptive lasso <span class="citation" data-cites="adaptiveLasso">(<a href="#ref-adaptiveLasso" role="doc-biblioref">Zou 2006</a>)</span>, group lasso <span class="citation" data-cites="yuan2007">(<a href="#ref-yuan2007" role="doc-biblioref">Yuan and Lin 2005</a>)</span>, group extensions of adaptive lasso, SCAD, and MCP <span class="citation" data-cites="grSCAD groupReview penBiLevel">(<a href="#ref-grSCAD" role="doc-biblioref">Wang et al. 2007</a>; <a href="#ref-penBiLevel" role="doc-biblioref">Breheny and Huang 2009</a>; <a href="#ref-groupReview" role="doc-biblioref">Huang et al. 2012</a>)</span>, and a group lasso penalty where variables are grouped across quantiles <span class="citation" data-cites="heteroIdQR">(<a href="#ref-heteroIdQR" role="doc-biblioref">Wang et al. 2024</a>)</span>. Extending the theoretical results of penalized estimators to the quantile regression setting has been an active area of research. Examples include deriving the rate of convergence for lasso <span class="citation" data-cites="qr_lasso">(<a href="#ref-qr_lasso" role="doc-biblioref">Belloni and Chernozhukov 2011</a>)</span> and group lasso <span class="citation" data-cites="qr_group_lasso">(<a href="#ref-qr_group_lasso" role="doc-biblioref">Kato 2012</a>)</span> and deriving oracle properties for non-convex penalties such as SCAD and MCP <span class="citation" data-cites="lan_scad">(<a href="#ref-lan_scad" role="doc-biblioref">Wang et al. 2012</a>)</span>. Discussed in these papers is how minimizing the penalized objective functions for quantile regression can be framed as linear programming problems or, in the case of group lasso, second-order cone programming problems. The linear programming formulation is particularly familiar to researchers in quantile regression because this is the most common approach for solving quantile regression problems, including in the <a href="https://cran.r-project.org/package=quantreg">quantreg</a> package <span class="citation" data-cites="crq1 crq2">(<a href="#ref-crq1" role="doc-biblioref">Koenker and D’Orey 1987</a>; <a href="#ref-crq2" role="doc-biblioref">Koenker and D’Orey 1994</a>)</span>, while a second-order cone programming problem can be solved using convex optimization software, including the R package <a href="https://cran.r-project.org/package=Rmosek">Rmosek</a> <span class="citation" data-cites="JSSv060i05">(<a href="#ref-JSSv060i05" role="doc-biblioref">Koenker and Mizera 2014</a>)</span>.</p>
<p>The ability to analyze large data sets is one of the major appeals of penalized regression methods. However, linear programming and second-order cone programming algorithms become computationally burdensome for large data sets. Further complicating matters is that the quantile loss function is non-differentiable, while popular algorithms for penalized objective functions rely on a differentiable loss function, for instance, <span class="citation" data-cites="friedman2009glmnet">Friedman et al. (<a href="#ref-friedman2009glmnet" role="doc-biblioref">2010</a>)</span> (elastic net), <span class="citation" data-cites="scadAlg">Breheny and Huang (<a href="#ref-scadAlg" role="doc-biblioref">2011</a>)</span> (non-convex penalties), <span class="citation" data-cites="brehenyglasso">Breheny and Huang (<a href="#ref-brehenyglasso" role="doc-biblioref">2015</a>)</span> (group non-convex penalties), and <span class="citation" data-cites="Yang2015">Yang and Zou (<a href="#ref-Yang2015" role="doc-biblioref">2015</a>)</span> (group lasso). <span class="citation" data-cites="huber_cd">Yi and Huang (<a href="#ref-huber_cd" role="doc-biblioref">2017</a>)</span> proposed using a Huber-type approximation of the quantile loss and coordinate descent algorithm for solving elastic net penalized quantile regression which is implemented in the R package <a href="https://cran.r-project.org/package=hqreg">hqreg</a>. Assuming a good initial estimator is provided, <span class="citation" data-cites="qr_cd">Peng and Wang (<a href="#ref-qr_cd" role="doc-biblioref">2015</a>)</span> proposed a coordinate descent algorithm (QICD) for non-convex penalties. The R package <a href="https://cran.r-project.org/package=conquer">conquer</a> approximates the quantile loss with a convolution of the quantile loss and Kernel function <span class="citation" data-cites="lowdConv highdConv">(<a href="#ref-highdConv" role="doc-biblioref">Tan et al. 2022</a>; <a href="#ref-lowdConv" role="doc-biblioref">He et al. 2023</a>)</span>.</p>
<p>The package <a href="https://cran.r-project.org/package=rqPen">rqPen</a> provides an implementation of the Huber-type approximation and linear programming algorithms. It allows users to fit quantile regression models with all of the penalty functions discussed in the first paragraph and provides tools for using cross-validation or information criterion to select tuning parameters. In addition, it provides plots for comparing cross-validation results and coefficient estimates as the sparsity parameter, <span class="math inline">\(\lambda\)</span>, changes. The package allows for estimating multiple quantiles with a call to a single function and provides plots of how coefficient values change with the different quantiles being modeled.</p>
<p>The packages <a href="https://cran.r-project.org/package=quantreg">quantreg</a>, <a href="https://cran.r-project.org/package=conquer">conquer</a>, <a href="https://cran.r-project.org/package=hrqglas">hrqglas</a>, and <a href="https://cran.r-project.org/package=hqreg">hqreg</a> are alternatives for penalized quantile regression in R. However, there are some substantial differences between <a href="https://cran.r-project.org/package=rqPen">rqPen</a> and these packages. The package <a href="https://cran.r-project.org/package=rqPen">rqPen</a> allows users to fit quantile regression models using the elastic net, SCAD, MCP, adaptive lasso, group lasso, group SCAD, group MCP and group adaptive lasso penalties, along with allowing users to choose between an <span class="math inline">\(L_1\)</span> or <span class="math inline">\(L_2\)</span> norm for all group penalties. In addition, users can fit models for multiple quantiles, use cross-validation or information criterion (IC) approaches to choose tuning parameters, and create plots of the change in coefficient estimates for different quantiles and tuning parameters. Also, commonly used functions such as <code>predict()</code>, <code>coef()</code>, and <code>plot()</code> can be used with the <code>rq.pen.seq</code> and <code>rq.pen.seq.cv</code> objects created by <a href="https://cran.r-project.org/package=rqPen">rqPen</a> functions. The package <a href="https://cran.r-project.org/package=quantreg">quantreg</a> fits unpenalized quantile regression models and provides the SCAD and lasso penalty. For both penalized and unpenalized quantile regression, <a href="https://cran.r-project.org/package=conquer">conquer</a> offers computationally efficient estimation for large <span class="math inline">\(n\)</span> or <span class="math inline">\(p\)</span> data sets. Users of <a href="https://cran.r-project.org/package=conquer">conquer</a> can choose between lasso, elastic net, group lasso, sparse-group <span class="citation" data-cites="sparseGroup">(<a href="#ref-sparseGroup" role="doc-biblioref">Simon et al. 2013</a>)</span>, group SCAD and group MCP. To the best of our knowledge, <a href="https://cran.r-project.org/package=conquer">conquer</a> is the only package that offers the sparse-group lasso penalty for quantile regression. While <a href="https://cran.r-project.org/package=hrqglas">hrqglas</a> provides penalized group lasso, it does not provide any of the other group penalties, nor does it allow for simultaneous estimation of multiple quantiles. Similarly, <a href="https://cran.r-project.org/package=hqreg">hqreg</a> provides the elastic net penalty using the Huber approximation, but not adaptive lasso, SCAD or MCP. However, <a href="https://cran.r-project.org/package=hrqglas">hrqglas</a> and <a href="https://cran.r-project.org/package=hqreg">hqreg</a> provide methods for robust mean regression using the Huber loss function, something <a href="https://cran.r-project.org/package=rqPen">rqPen</a> does not do. Both packages are required for <a href="https://cran.r-project.org/package=rqPen">rqPen</a> and provide the backbone for the algorithms that use a Huber approximation for the group, <a href="https://cran.r-project.org/package=hrqglas">hrqglas</a>, and non-group, <a href="https://cran.r-project.org/package=hqreg">hqreg</a>, penalties. The package <a href="https://cran.r-project.org/package=rqPen">rqPen</a> provides a variety of algorithms and penalty functions that users can choose from. In addition, it provides tools not available in other penalized quantile regression packages, such as plots of how coefficients change with the quantile being modeled or the sparsity parameter, allowing for multiple estimates of quantiles in a single line of code, and functions for information criterion based approaches to tuning parameter selection. Finally, it provides a penalty that guarantees consistent variable selection across quantiles, an option not available in the other packages mentioned.</p>
<h2 data-number="2" id="penalized-estimation-of-quantile-regression"><span class="header-section-number">2</span> Penalized estimation of quantile regression</h2>
<p>Consider observations <span class="math inline">\(\{y_i,{\bf x}_i\}_{i=1}^n\)</span>, where <span class="math inline">\({\bf x}_i = (1,x_{i1},\ldots,x_{ip})^\top \in \mathbb{R}^{p+1}\)</span>, and the model
<span class="math display" id="eq:qrModel">\[\begin{equation}
y_i = {\bf x}_i^\top {\boldsymbol\beta}^{\tau}_0 + \epsilon_i,
\tag{1}
\end{equation}\]</span>
where <span class="math inline">\(P(\epsilon_i &lt; 0 | {\bf x}_i)=\tau\)</span>. Define <span class="math inline">\(\rho_\tau(u) = u[\tau-I(u&lt;0)]\)</span>, <span class="citation" data-cites="origQR">Koenker and Bassett (<a href="#ref-origQR" role="doc-biblioref">1978</a>)</span> proposed estimating <a href="#eq:qrModel">(1)</a> by minimizing
<span class="math display" id="eq:qr">\[\begin{equation}
\sum_{i=1}^n \rho_\tau(y_i-{\bf x}_i^\top\boldsymbol \beta),
\tag{2}
\end{equation}\]</span>
which is available in the package <a href="https://cran.r-project.org/package=quantreg">quantreg</a>. Let <span class="math inline">\(m_i\)</span> be the weight for observation <span class="math inline">\(i\)</span>, <span class="math inline">\(\boldsymbol w\)</span> be a vector of weights for a predictor or group of predictors, and <span class="math inline">\(\lambda\)</span> and <span class="math inline">\(a\)</span> are penalty tuning parameters. The package <a href="https://cran.r-project.org/package=rqPen">rqPen</a> provides functions for estimating <span class="math inline">\(\boldsymbol \beta^\tau_0\)</span> by minimizing
<span class="math display" id="eq:genPen">\[\begin{equation}
\frac{1}{n} \sum_{i=1}^n m_i \rho_\tau(y_i - {\bf x}_i^\top\boldsymbol \beta) + P_{\boldsymbol w,\lambda,a}(\boldsymbol \beta),
\tag{3}
\end{equation}\]</span>
where <span class="math inline">\(\lambda&gt;0\)</span> and the plausible values of <span class="math inline">\(a\)</span> depend on the penalty being used. The penalty function <span class="math inline">\(P_{\lambda,a}(\boldsymbol \beta)\)</span> can take the form of a group or individual penalty.</p>
<p>Let <span class="math inline">\(w_j\)</span> be a weight for a variable <span class="math inline">\(j\)</span>. For an individual penalty, <a href="#eq:genPen">(3)</a> has the form of
<span class="math display" id="eq:indPen">\[\begin{equation}
\frac{1}{n} \sum_{i=1}^n m_i \rho_\tau(y_i - {\bf x}_i^\top\boldsymbol \beta) + \sum_{j=1}^p p_{w_j\lambda,a}(|\beta_j|).
\tag{4}
\end{equation}\]</span></p>
<p>The package <a href="https://cran.r-project.org/package=rqPen">rqPen</a> supports four different forms of individual penalties: elastic net, adaptive lasso, SCAD and MCP. Users can also specify ridge and lasso, which are special cases of the elastic net. The SCAD, <span class="math inline">\(p^s_{\lambda,a}()\)</span>, and MCP, <span class="math inline">\(p^m_{\lambda,a}()\)</span> penalty functions are
<span class="math display">\[\begin{eqnarray*}
p^s_{\lambda,a}(|x|) &amp;=&amp; \lambda|x|I(0 \leq |x| &lt; \lambda) + \frac{a\lambda|x|-(x^2+\lambda^2)/2}{a-1}I(\lambda \leq |x| \leq a\lambda) + \frac{(a+1)\lambda^2}{2}I(|x|&gt;a\lambda),  \\
p^m_{\lambda,a}(|x|) &amp;=&amp; \lambda\left(|x|-\frac{x^2}{2a\lambda}\right)I(0 \leq |x| &lt; a\lambda) + \frac{a\lambda^2}{2}I(|x|\geq a\lambda),
\end{eqnarray*}\]</span>
where <span class="math inline">\(a&gt;2\)</span> for the SCAD penalty function and <span class="math inline">\(a&gt;1\)</span> for MCP.</p>
<p>The following defines <span class="math inline">\(P_{\lambda,a}(\boldsymbol \beta)\)</span> for the four different penalty functions, plus two important special cases.</p>
<p>The weights, <span class="math inline">\(w_j\)</span>, allow for different weights for predictors and must be non-negative. If <span class="math inline">\(w_j=0\)</span> then that variable will be unpenalized. In <a href="https://cran.r-project.org/package=rqPen">rqPen</a> these weights are labeled as <code>penalty.factors</code> or <code>group.penalty.factors</code>. The lasso estimator provides the backbone for the algorithm of the three non-elastic net penalties. As <span class="math inline">\(\rho_\tau(x)+\rho_\tau(-x) = |x|\)</span>, the lasso estimator minimizes,
<span class="math display" id="eq:lasso">\[\begin{equation}
\frac{1}{n} \sum_{i=1}^n m_i \rho_\tau(y_i - {\bf x}_i^\top\boldsymbol \beta) +  \sum_{j=1}^p \rho_\tau(\lambda w_j\beta_j)+\rho_\tau(-\lambda w_j\beta_j).
\tag{5}
\end{equation}\]</span>
For <span class="math inline">\(i \in \{1,\ldots,n\}\)</span> define <span class="math inline">\(\tilde{y}_i = y_i\)</span>, <span class="math inline">\(\tilde{{\bf x}}_i = {\bf x}_i\)</span>, and <span class="math inline">\(\tilde{m}_i = m_i\)</span>. Let <span class="math inline">\({\bf e}_j \in \mathbb{R}^{p+1}\)</span> represent a unit vector with a value of one in the <span class="math inline">\(j\)</span>th position and zero in all other entries. For each <span class="math inline">\(j \in \{1,\ldots,p\}\)</span> define <span class="math inline">\(\tilde{{\bf x}}_{n+2j-1} = -n\lambda w_j {\bf e}_j\)</span> and <span class="math inline">\(\tilde{{\bf x}}_{n+2j} = n\lambda w_j {\bf e}_j\)</span>. In addition, for each <span class="math inline">\(i \in \{n+1,\ldots,n+2p\}\)</span> define <span class="math inline">\(\tilde{y}_i = 0\)</span> and <span class="math inline">\(\tilde{m}_i=1\)</span>. Then minimizing <a href="#eq:lasso">(5)</a> is equivalent to minimizing
<span class="math display" id="eq:lassoAlt">\[\begin{equation}
\frac{1}{n} \sum_{i=1}^{n+2p} \tilde{m}_i \rho_\tau(\tilde{y}_i-\tilde{{\bf x}}_i^\top\boldsymbol \beta).
\tag{6}
\end{equation}\]</span>
The objective function <a href="#eq:lassoAlt">(6)</a> has the same form as <a href="#eq:qr">(2)</a> except for the scaling factor, <span class="math inline">\(n^{-1}\)</span>, and the weights. This approach of creating the augmented <span class="math inline">\(2p\)</span> samples and then using standard quantile regression is implemented in <a href="https://cran.r-project.org/package=rqPen">rqPen</a> where the problem is solved using the <code>rq()</code> function from <a href="https://cran.r-project.org/package=quantreg">quantreg</a>. Note this approach is different from using <code>rq(method="lasso",...)</code> within <a href="https://cran.r-project.org/package=quantreg">quantreg</a>, which uses a linear programming approach but does not formulate the problem using augmented data.</p>
<p>For large values of <span class="math inline">\(n\)</span> and <span class="math inline">\(p\)</span> the linear programming algorithms become computationally burdensome. To decrease computational complexity, <span class="citation" data-cites="huber_cd">Yi and Huang (<a href="#ref-huber_cd" role="doc-biblioref">2017</a>)</span> proposed approximating the quantile loss function with a Huber-like function and a new coordinate descent algorithm that requires a differentiable loss function. The Huber loss function proposed by <span class="citation" data-cites="huber1964">Huber (<a href="#ref-huber1964" role="doc-biblioref">1964</a>)</span> is
<span class="math display">\[
h_\gamma(t) = \frac{t^2}{2\gamma}I(|t| \leq \gamma)+\left[|t|-\frac{\gamma}{2}\right]I(|t|&gt;\gamma).
\]</span>
Note <span class="math inline">\(\rho_\tau(u) = u[\tau-I(u&lt;0)]=\frac{1}{2}(|u|+(2\tau-1)u)\)</span> and for sufficiently small <span class="math inline">\(\gamma\)</span>, <span class="math inline">\(|u| \approx h_\gamma(u)\)</span>. The Huber-approximated quantile loss is
<span class="math display" id="eq:huberQuantile">\[\begin{equation}
h_{\gamma}^{\tau}(u) = h_\gamma(u) + (2\tau-1)u,
\tag{7}
\end{equation}\]</span>
and for small <span class="math inline">\(\gamma\)</span>, <span class="math inline">\(\rho_\tau(u) \approx \frac{1}{2}h_{\gamma}^{\tau}(u)\)</span>. The package <a href="https://cran.r-project.org/package=hqreg">hqreg</a> implements the approach of <span class="citation" data-cites="huber_cd">Yi and Huang (<a href="#ref-huber_cd" role="doc-biblioref">2017</a>)</span> and the function <code>hqreg()</code>, with <code>method="quantile"</code>, solves the problem of
<span class="math display" id="eq:huberLasso">\[\begin{equation}
\frac{1}{2n} \sum_{i=1}^n h_{\gamma}^{\tau}(y_i - {\bf x}_i^\top\boldsymbol \beta) +  \lambda \sum_{j=1}^p w_j|\beta_j|.
\tag{8}
\end{equation}\]</span></p>
<p>In <a href="https://cran.r-project.org/package=rqPen">rqPen</a>, solving <a href="#eq:huberLasso">(8)</a> is done by calling <code>hqreg::hqreg()</code> and thus <a href="https://cran.r-project.org/package=hqreg">hqreg</a> is a required package. The Huber approximation with lasso is the default for <code>rq.pen()</code>, one of the main functions in <a href="https://cran.r-project.org/package=rqPen">rqPen</a>. In addition, optimization of the adaptive lasso, SCAD and MCP problems can be solved using a version of <a href="#eq:huberLasso">(8)</a>, which is available in <a href="https://cran.r-project.org/package=rqPen">rqPen</a> but not <a href="https://cran.r-project.org/package=hqreg">hqreg</a>. The adaptive lasso can be solved using the same approach because it is a special case of a lasso problem with different weights for each coefficient. The initial estimators necessary for the weights are determined by a ridge estimator with the same value of <span class="math inline">\(\lambda\)</span>. The SCAD and MCP functions are approximated by a local linear approximation (LLA) as proposed by <span class="citation" data-cites="lla">Zou and Li (<a href="#ref-lla" role="doc-biblioref">2008</a>)</span>. Let <span class="math inline">\(p_{w_j\lambda,a}(|\beta_j|)\)</span> represent a generic penalty function and <span class="math inline">\(p&#39;_{w_j\lambda,a}(|\beta_j|)\)</span> be the derivative with respect to <span class="math inline">\(\beta_j\)</span>. Let <span class="math inline">\(\bar{\beta}_j\)</span> be the lasso estimator for the same value of <span class="math inline">\(\lambda\)</span> and weights. The LLA approach uses the following approximation,
<span class="math display" id="eq:lla">\[\begin{align}
\frac{1}{n} \sum_{i=1}^n m_i \rho_\tau(y_i-{\bf x}_i^\top\boldsymbol \beta)  + &amp; \sum_{j=1}^p p_{w_j\lambda,a}(|\beta_j|) \nonumber \\
&amp; \approx \frac{1}{n} \sum_{i=1}^n m_i \rho_\tau(y_i-{\bf x}_i^\top\boldsymbol \beta) + \sum_{j=1}^p p&#39;_{w_j\lambda,a}(|\bar{\beta}_j|)|\beta_j|.
\tag{9}
\end{align}\]</span>
Again, the problem becomes a special case of a lasso estimator with specific weights for each predictor. Thus all the non-group penalties, except for elastic net and ridge, can be solved using linear programming or Huber approximation algorithms.</p>
<p>The elastic net penalty of
<span class="math display" id="eq:enet">\[\begin{equation}
\frac{1}{n} \sum_{i=1}^n m_i \rho_\tau(y_i - {\bf x}_i^\top\boldsymbol \beta) + \lambda \sum_{j=1}^p w_j \left[ a|\beta_j|+(1-a)\beta_j^2\right],
\tag{10}
\end{equation}\]</span>
cannot be framed as a linear programming problem because of the ridge penalty. Thus for <span class="math inline">\(a \neq 1\)</span>, the <a href="https://cran.r-project.org/package=rqPen">rqPen</a> implementation of elastic net only uses the Huber approximation approach provided in <a href="https://cran.r-project.org/package=hqreg">hqreg</a>. While <a href="https://cran.r-project.org/package=hqreg">hqreg</a> provides a computational backbone, <a href="https://cran.r-project.org/package=rqPen">rqPen</a> provides SCAD, MCP, and adaptive lasso penalties that are not provided in <code>hqreg()</code>. In addition, <code>hqreg()</code> does not include any group penalties or allow for weights for observations, <span class="math inline">\(m_i\)</span>. If weights are specified then the group lasso with singleton groups, implemented in <a href="https://cran.r-project.org/package=hrqglas">hrqglas</a>, is used instead of using <code>hqreg()</code> from <a href="https://cran.r-project.org/package=hqreg">hqreg</a>.</p>
<h3 data-number="2.1" id="group-penalty-predictors"><span class="header-section-number">2.1</span> Group Penalty predictors</h3>
<p>When there exists a group structure to the predictors then a group penalty can often account for this structure. For instance, non-binary categorical variables or polynomial transformations of predictors. This section assumes the <span class="math inline">\(p\)</span> predictors are partitioned into <span class="math inline">\(G\)</span> groups and <span class="math inline">\(\boldsymbol \beta_g\)</span> represents the coefficients associated with the <span class="math inline">\(g\)</span>th group of predictors. Group penalized quantile regression estimators minimize
<span class="math display" id="eq:groupForm">\[\begin{equation}
\frac{1}{n} \sum_{i=1}^n m_i \rho_\tau(y_i - {\bf x}_i^\top\boldsymbol \beta) + \sum_{g=1}^G p_{w_g\lambda,a}(||\boldsymbol \beta_g||_q).
\tag{11}
\end{equation}\]</span></p>
<p>Users can choose four penalty functions for <span class="math inline">\(p_{w_g,\lambda,a}()\)</span>: (1) lasso; (2) Adaptive lasso; (3) SCAD; and (4) MCP. Specifically,</p>
<p>The values of <span class="math inline">\(q\)</span> in <code>rq.group.pen()</code> is limited to <span class="math inline">\(q \in \{1,2\}\)</span> and can be changed with <code>norm=q</code>. If <span class="math inline">\(q=2\)</span> then group variable selection will be all-or-nothing, that is all variables within a group will be selected or none of them will be. The choice of <span class="math inline">\(q=1\)</span> allows for bi-level variable selection <span class="citation" data-cites="penBiLevel">(<a href="#ref-penBiLevel" role="doc-biblioref">Breheny and Huang 2009</a>)</span>. Users cannot specify <code>norm=1</code> and <code>penalty="gLASSO"</code>, default penalty of group lasso, because that is equivalent to the individual lasso penalty implemented in <code>rq.pen()</code>.</p>
<p><span class="citation" data-cites="groupReview">Huang et al. (<a href="#ref-groupReview" role="doc-biblioref">2012</a>)</span> provide an excellent review of group penalties, including a discussion of the use of <span class="math inline">\(q=1\)</span>. For differentiable loss functions, <span class="citation" data-cites="penBiLevel">Breheny and Huang (<a href="#ref-penBiLevel" role="doc-biblioref">2009</a>)</span> provide a computational perspective on <span class="math inline">\(q=1\)</span>, while <span class="citation" data-cites="mestatp">Sherwood et al. (<a href="#ref-mestatp" role="doc-biblioref">2020</a>)</span> compare oracle model selection properties for <span class="math inline">\(q=1\)</span> and <span class="math inline">\(q=2\)</span>.</p>
<p>For <span class="math inline">\(q=1\)</span>, the group lasso estimator becomes a special case of the individual lasso estimator, and this is also true for the adaptive lasso penalty. For the SCAD and MCP penalties, LLA will be used to approximate the non-convex penalties. Let <span class="math inline">\(\bar{\boldsymbol \beta}_g\)</span> be an initial group lasso estimator and the penalized objective function is approximated by
<span class="math display" id="eq:groupFormlla">\[\begin{equation}
\frac{1}{n} \sum_{i=1}^n m_i \rho_\tau(y_i - {\bf x}_i^\top\boldsymbol \beta) + \sum_{g=1}^G p&#39;_{w_g\lambda,a}(||\bar{\boldsymbol \beta}_g||_q)||\boldsymbol \beta_g||_q.
\tag{12}
\end{equation}\]</span></p>
<p>Thus for the case of <span class="math inline">\(q=1\)</span> this becomes an individual lasso problem and the algorithms discussed in the previous section apply. In particular, the problem can be framed as a linear programming problem. While for <span class="math inline">\(q=2\)</span> <a href="#eq:groupFormlla">(12)</a> becomes a special case of the group lasso problem. These group lasso problems are not linear programming problems, but second-order cone programming problems. Second-order cone programming problems can be solved by existing convex optimization software such as <a href="https://cran.r-project.org/package=Rmosek">Rmosek</a> <span class="citation" data-cites="JSSv060i05">(<a href="#ref-JSSv060i05" role="doc-biblioref">Koenker and Mizera 2014</a>)</span>. However, there are some barriers to using <a href="https://cran.r-project.org/package=Rmosek">Rmosek</a>, for instance it requires a copy of Mosek installed on the user’s computer. In addition, similar to linear programming problems, second-order cone programming algorithms can be computationally burdensome for large values of <span class="math inline">\(n\)</span> or <span class="math inline">\(p\)</span>. For <span class="math inline">\(q=2\)</span>, the Huber approximation described in the previous subsection is used. However, <a href="https://cran.r-project.org/package=hqreg">hqreg</a> cannot be used to solve this problem because the approach of <span class="citation" data-cites="huber_cd">Yi and Huang (<a href="#ref-huber_cd" role="doc-biblioref">2017</a>)</span> is not for a group penalty. Instead, the algorithm of <span class="citation" data-cites="Yang2015">Yang and Zou (<a href="#ref-Yang2015" role="doc-biblioref">2015</a>)</span> is implemented. See <span class="citation" data-cites="SherwoodLi2022">Sherwood and Li (<a href="#ref-SherwoodLi2022" role="doc-biblioref">2022</a>)</span> for details regarding the application of this algorithm to the asymmetric Huber regression setting. This approach is available in the package <a href="https://cran.r-project.org/package=hrqglas">hrqglas</a>, which is maintained by two of the authors and provides the computational backbone for the <span class="math inline">\(L_2\)</span> group penalties.</p>
<h3 data-number="2.2" id="estimation-of-multiple-quantiles"><span class="header-section-number">2.2</span> Estimation of Multiple Quantiles</h3>
<p>For a set of quantiles of interest, <span class="math inline">\((\tau_1,\ldots,\tau_B)\)</span>, <span class="citation" data-cites="qr_lasso">Belloni and Chernozhukov (<a href="#ref-qr_lasso" role="doc-biblioref">2011</a>)</span> proposed minimizing,
<span class="math display" id="eq:qrBel">\[\begin{equation}
\frac{1}{n} \sum_{b=1}^B\sum_{i=1}^n m_i \rho_{\tau_b}\left(y_i - {\bf x}_i^\top\boldsymbol \beta^{\tau_b}\right) + \frac{\tilde{\lambda}}{n} \sum_{b=1}^B\sum_{j=1}^p \sqrt{\tau_b(1-\tau_b)} \hat{\sigma}_j |\beta^{\tau_b}_j|,
\tag{13}
\end{equation}\]</span>
where <span class="math inline">\(\hat{\sigma}_j = n^{-1}\sum_{i=1}^n x_{ij}^2\)</span>. We use <span class="math inline">\(\tilde{\lambda}\)</span> in <a href="#eq:qrBel">(13)</a> to emphasize that the scaling of the tuning parameter is different than that used in <a href="https://cran.r-project.org/package=rqPen">rqPen</a>, specifically <span class="math inline">\(\lambda=\frac{\tilde{\lambda}}{n}\)</span>. The penalized objective function of <a href="#eq:qrBel">(13)</a> provides a quantile specific weight of <span class="math inline">\(\sqrt{\tau_b(1-\tau_b)}\)</span> and a predictor specific weight of <span class="math inline">\(\hat{\sigma}_j\)</span>, while <span class="math inline">\(\lambda\)</span> is assumed to have the same value. In <a href="https://cran.r-project.org/package=rqPen">rqPen</a>, this approach is generalized to allow for different penalties and user choices of the quantile, predictor, and observation weights. For an individual penalty the estimator minimizes,
<span class="math display" id="eq:multtau">\[\begin{equation}
\frac{1}{n} \sum_{b=1}^B\sum_{i=1}^n m_i \rho_{\tau_b}\left(y_i - {\bf x}_i^\top\boldsymbol \beta^{\tau_b}\right) + \sum_{b=1}^B\sum_{j=1}^p p_{w_jd_b\lambda,a}(|\beta^{\tau_b}_j|).
\tag{14}
\end{equation}\]</span>
For a group penalty the penalized objective function is
<span class="math display" id="eq:multtaug">\[\begin{equation}
\frac{1}{n} \sum_{b=1}^B \sum_{i=1}^n m_i \rho_{\tau_b}\left(y_i - {\bf x}_i^\top\boldsymbol \beta^{\tau_b}\right) + \sum_{b=1}^B \sum_{g=1}^G p_{w_gd_b\lambda,a}(||\boldsymbol \beta_g||_q).
\tag{15}
\end{equation}\]</span>
Note, the <span class="math inline">\(w_jd_b\lambda\)</span> term allows users to specify specific weights for a predictor, <span class="math inline">\(w_j\)</span>, and specific weights for a quantile, <span class="math inline">\(d_b\)</span>. For instance, if a user chooses the lasso penalty with <span class="math inline">\(w_j=n^{-1}\sum_{i=1}^nx_{ij}^2\)</span>, <span class="math inline">\(d_b=n^{-1}\sqrt{\tau_b(1-\tau_b)}\)</span>, and <span class="math inline">\(m_i=1\)</span> then <a href="#eq:multtau">(14)</a> is equivalent to <a href="#eq:qrBel">(13)</a>. The optimization of <a href="#eq:multtau">(14)</a> and <a href="#eq:multtaug">(15)</a> consist of <span class="math inline">\(B\)</span> different optimization problems. In <a href="https://cran.r-project.org/package=rqPen">rqPen</a> they are solved separately using the algorithms that have been discussed previously, except the same sequence of <span class="math inline">\(\lambda\)</span> is used for all <span class="math inline">\(B\)</span> problems. In <span class="citation" data-cites="qr_lasso">Belloni and Chernozhukov (<a href="#ref-qr_lasso" role="doc-biblioref">2011</a>)</span> they suggest choosing the same value of <span class="math inline">\(\lambda\)</span> for all the quantiles, albeit with the quantile-specific weights. We later present how users can choose a single value of <span class="math inline">\(\lambda\)</span> and coefficient estimates come from either <a href="#eq:multtau">(14)</a> or <a href="#eq:multtaug">(15)</a>. Alternatively, users can choose <span class="math inline">\(B\)</span> different values of <span class="math inline">\(\lambda\)</span> and then the solutions come from <span class="math inline">\(B\)</span> different optimizations of <a href="#eq:indPen">(4)</a> or <a href="#eq:groupForm">(11)</a>. The latter approach is equivalent to a data-driven approach for estimating the optimal quantile-specific weights, <span class="math inline">\(d_b\)</span>.</p>
<h4 class="unnumbered" data-number="2.2.1" id="consistent-selection-across-quantiles">Consistent selection across quantiles</h4>
<p>The penalties discussed previously do not ensure consistent selection across the quantiles. That is for two quantiles <span class="math inline">\(\tau_1\)</span> and <span class="math inline">\(\tau_2\)</span>, it is possible for a variable’s coefficient to be non-zero at <span class="math inline">\(\tau_1\)</span>, but zero at <span class="math inline">\(\tau_2\)</span>. Our opinion is this is often hard to interpret. Consider the case where <span class="math inline">\(B\)</span> quantiles are considered. Define, <span class="math inline">\(\boldsymbol \beta^j = (\beta_{j}^{\tau_1},\ldots,\beta^{\tau_B}_j)^\top \in \mathbb{R}^{B}\)</span> as the vector of the <span class="math inline">\(j\)</span>th coefficient for all <span class="math inline">\(B\)</span> quantiles. Define <span class="math inline">\(||\boldsymbol \beta^j||_{2,\boldsymbol w,\boldsymbol d}=\sqrt{\sum_{b=1}^B d_b w_j (\beta^{\tau_b}_{j})^2}\)</span>, as weighted <span class="math inline">\(L_2\)</span>-norm that allows for predictor, <span class="math inline">\(m_j\)</span>, and quantile, <span class="math inline">\(w_k\)</span>, specific weights. If a user desires consistent variable selection across quantiles they can use the group quantile penalized objective function,
<span class="math display" id="eq:gql">\[\begin{equation}
\frac{1}{n} \sum_{b=1}^B \sum_{i=1}^n m_i \rho_{\tau_b}\left(y_i - {\bf x}_i^\top\boldsymbol \beta^{\tau_b}\right) + \lambda \sum_{j=1}^p ||\boldsymbol \beta^j||_{2,\boldsymbol w,\boldsymbol d}.
\tag{16}
\end{equation}\]</span></p>
<p>We refer to this penalty as the group quantile penalty <span class="citation" data-cites="gqlasso">(<a href="#ref-gqlasso" role="doc-biblioref">Li and Sherwood 2025</a>)</span>.</p>
<h3 data-number="2.3" id="tuning-parameter-selection"><span class="header-section-number">2.3</span> Tuning parameter selection</h3>
<p>There are two tuning parameters, <span class="math inline">\(\lambda\)</span> and <span class="math inline">\(a\)</span>, that need to be set. The value of <span class="math inline">\(a\)</span> defaults to commonly used values for each penalty: (1) elastic-net (<span class="math inline">\(a=1\)</span>); (2) adaptive lasso (<span class="math inline">\(a=1\)</span>); (3) SCAD (<span class="math inline">\(a=3.7\)</span>); and (4) MCP (<span class="math inline">\(a=3\)</span>). Users can also specify values of <span class="math inline">\(a\)</span>, except for the cases of Ridge (<span class="math inline">\(a=0\)</span>) and lasso (<span class="math inline">\(a=1\)</span>) where the value of <span class="math inline">\(a\)</span> is fixed. Typically the value of <span class="math inline">\(a\)</span> is not as much of a concern as the value of <span class="math inline">\(\lambda\)</span>, a potential notable exception to this would be the elastic net penalty. Users can specify a sequence for values of <span class="math inline">\(\lambda\)</span> otherwise, a sequence will be automatically generated. Define <span class="math inline">\(H_\gamma^\tau(\boldsymbol \beta) = \frac{1}{2n} \sum_{i=1}^n m_i h_\gamma^\tau(y_i - {\bf x}_i^\top\boldsymbol \beta)\)</span>. Define <span class="math inline">\(\tilde{a}=a\)</span> if the elastic net penalty is used and one otherwise. If <span class="math inline">\(a=0\)</span>, that is the ridge penalty is being used, then <span class="math inline">\(\tilde{a}=.001\)</span>. The default value for an individual penalty is <span class="math inline">\(\lambda_{\max} = \max_{j,b} 1.05\left|\frac{\partial}{\partial \beta_j} H_\gamma^{\tau_b}(\mathbf{0}_{p+1})\right|(w_jd_b\tilde{a})^{-1}\)</span> and for a group penalty is <span class="math inline">\(\lambda_{\max} = \max_{g,b} 1.05\left|\left| \frac{\partial}{\partial \boldsymbol \beta_g} H_\gamma^{\tau_b}(\mathbf{0}_{p+1})\right|\right|_q (w_gd_b)^{-1}\)</span>. The 1.05 multiplier is used to ensure that <span class="math inline">\(\lambda_{\max}\)</span> provides a completely sparse solution regardless of whether the Huber approximation is used or not. If any predictors or groups are unpenalized at a given quantile then they are excluded from these calculations.</p>
<p>Both cross-validation and information criterion are provided as tools for selecting the optimal pair of <span class="math inline">\((\lambda,a)\)</span>. Let <span class="math inline">\(\hat{\boldsymbol \beta}^\tau_{\lambda,a}\)</span> be the coefficient estimates for quantile <span class="math inline">\(\tau\)</span> using tuning parameters <span class="math inline">\(\lambda\)</span> and <span class="math inline">\(a\)</span>. Define <span class="math inline">\(k^\tau_{\lambda,a}\)</span> as the number of non-zero coefficients in <span class="math inline">\(\hat{\beta}_{\lambda,a}^\tau\)</span>. The quantile information criterion (QIC) is defined as
<span class="math display" id="eq:qic">\[\begin{equation}
QIC(\tau,\lambda,a) = \log\left[ \sum_{i=1}^n m_i \rho_\tau(y_i-{\bf x}_i^\top \hat{\boldsymbol \beta}^\tau_{\lambda,a}) \right] + \frac{mk^\tau_{\lambda,a}}{2n},
\tag{17}
\end{equation}\]</span>
where the value of <span class="math inline">\(m\)</span> depends on the criterion being used. Users can select between AIC (<span class="math inline">\(m=2\)</span>), BIC [<span class="math inline">\(m=\log(n)\)</span>], and a version of the large <span class="math inline">\(p\)</span> BIC proposed by <span class="citation" data-cites="qrbic">Lee et al. (<a href="#ref-qrbic" role="doc-biblioref">2014</a>, <span class="math inline">\(m=\log(n)\log(p)\)</span>)</span>.</p>
<p>Cross-validation is the other approach implemented for choosing the optimal pair of <span class="math inline">\((\lambda,a)\)</span>. Consider the case of <span class="math inline">\(K\)</span> folds, let <span class="math inline">\(n_k\)</span> be the number of observations in the <span class="math inline">\(k\)</span>th fold, <span class="math inline">\(y_i^k\)</span>, <span class="math inline">\({\bf x}_i^k\)</span>, and <span class="math inline">\(m_i^k\)</span> be the <span class="math inline">\(i\)</span>th response, predictor vector, and weights for the <span class="math inline">\(k\)</span>th fold and <span class="math inline">\(\hat{\boldsymbol \beta}^{\tau_b}_{-k,\lambda,a}\)</span> be the fit for the <span class="math inline">\(b\)</span>th quantile excluding the <span class="math inline">\(k\)</span>th fold for given values of <span class="math inline">\(\lambda\)</span> and <span class="math inline">\(a\)</span>. By default, even if Huber approximations are used, cross-validation is done using quantile loss. However, this can be changed by setting <code>cvFunc</code> in the later described functions <code>rq.group.pen.cv()</code>, <code>rq.gq.pen.cv()</code>, or <code>rq.pen.cv()</code>. For instance, <code>cvFunc=abs</code> will use absolute value loss regardless of the quantile being modeled. For simplicity of presentation, the following assumes quantile loss is used. The average quantile loss at a given fold is
<span class="math display" id="eq:cv1">\[\begin{equation}
C_k^\tau(\lambda,a) = \frac{1}{n_k} \sum_{i=1}^{n_k} m_i^k \rho_\tau(y_i^k-{{\bf x}_i^k}^\top\hat{\boldsymbol \beta}^{\tau}_{-k,\lambda,a}).
\tag{18}
\end{equation}\]</span>
The cross-validation functions return two summaries of the cross-validation for selecting <span class="math inline">\((\lambda,a)\)</span>. The return value <code>btr</code> provides a table of the values of <span class="math inline">\(a\)</span> and <span class="math inline">\(\lambda\)</span> that minimize
<span class="math display" id="eq:cv2">\[\begin{equation}
C^\tau(\lambda,a) = \frac{1}{K} \sum_{k=1}^K C_k^\tau(\lambda,a).
\tag{19}
\end{equation}\]</span>
In addition it provides the <span class="math inline">\(\lambda\)</span> value associated with the sparsest solution that is within one standard error of the value that minimizes <a href="#eq:cv2">(19)</a>. The standard error for a given pair of <span class="math inline">\((\lambda,a)\)</span> is calculated as
<span class="math display">\[\begin{equation}
\sqrt{\frac{1}{K-1} \sum_{k=1}^K \left[C_k^\tau(\lambda,a)-C^\tau(\lambda,a) \right]^2}.
\end{equation}\]</span>
The average summary can be replaced by changing the parameter <code>cvSummary</code>. For instance, <code>cvSummary=median</code> would use the median values of <span class="math inline">\(C_k^\tau(\lambda,a)\)</span>.</p>
<p>Users can also choose to optimize <span class="math inline">\(\lambda\)</span> and <span class="math inline">\(a\)</span>, so they are the same values across all quantiles. When evaluating the cross-validation error they can choose to not provide equal weight to each quantile by entering quantile specific weights of <span class="math inline">\(z_b\)</span>. The return value <code>gtr</code> provides results for the value of <span class="math inline">\(\lambda\)</span> and <span class="math inline">\(a\)</span> that minimize
<span class="math display" id="eq:cv3">\[\begin{equation}
C(\lambda,a) = \sum_{b=1}^B z_b C^{\tau_b}(\lambda,a).
\tag{20}
\end{equation}\]</span>
Users can choose between a single pair of <span class="math inline">\((\lambda,a)\)</span> that minimizes <a href="#eq:cv3">(20)</a> or <span class="math inline">\(B\)</span>, potentially different, pairs that minimize <a href="#eq:cv2">(19)</a> separately for each quantile except for when minimizing <a href="#eq:gql">(16)</a>. That is a joint optimization problem and thus only joint selection of tuning parameters using <a href="#eq:cv3">(20)</a> is allowed. In addition, results for the one standard error approach, where the standard error for a fixed <span class="math inline">\((\lambda,a)\)</span> pair is
<span class="math display">\[\begin{equation}
\sqrt{\frac{1}{K-1}\sum_{k=1}^K \sum_{b=1}^B z_b \left[C_k^\tau(\lambda,a) -C^\tau(\lambda,a) \right]^2},
\end{equation}\]</span>
are provided. The one standard error rule finds the sparsest solution, by increasing <span class="math inline">\(\lambda\)</span>, where the value of <a href="#eq:cv2">(19)</a> or <a href="#eq:cv3">(20)</a> are within one standard error of the optimal value, assuming the optimal <span class="math inline">\(a\)</span> is fixed.</p>
<p>By setting <code>groupError=FALSE</code> each observation is weighted equally and <a href="#eq:cv2">(19)</a> and <a href="#eq:cv3">(20)</a> are replaced by <span class="math inline">\(\sum_{k=1}^K n_k C_k^\tau(\lambda,a)\)</span>,
and <span class="math inline">\(\sum_{b=1}^B z_b \sum_{k=1}^K n_k C_k^\tau(\lambda,a)\)</span>, respectively.</p>
<h4 class="unnumbered" data-number="2.3.1" id="tuning-parameter-selection-for-multiple-quantiles-using-an-information-criterion">Tuning parameter selection for multiple quantiles using an information criterion</h4>
<p>For the case of multiple, <span class="math inline">\(B\)</span>, quantiles being modeled, <a href="https://cran.r-project.org/package=rqPen">rqPen</a> offers two different approaches for selecting the tuning parameters. One approach is to select <span class="math inline">\(B\)</span> pairs of <span class="math inline">\((\lambda,a)\)</span> that minimize the <span class="math inline">\(B\)</span> different versions of <a href="#eq:qic">(17)</a>. The other approach is given weights <span class="math inline">\((z_1,\ldots,z_b)\)</span>, the optimal pair of <span class="math inline">\((\lambda,a)\)</span> minimizes,
<span class="math display" id="eq:jointQic">\[\begin{equation}
\sum_{b=1}^B z_b QIC(\tau_b,\lambda,a).
\tag{21}
\end{equation}\]</span>
The weights offer flexibility to a user who wants to provide more weight to a specific quantile. The default value for all the weights is one thus giving equal weight to each quantile.</p>
<h3 data-number="2.4" id="additional-notes-on-huber-approximation"><span class="header-section-number">2.4</span> Additional notes on Huber approximation</h3>
<p>For Huber approximation, a value of <span class="math inline">\(\gamma\)</span> is needed. If <span class="math inline">\(\gamma\)</span> is too large then the estimator will have a large amount of bias. If <span class="math inline">\(\gamma\)</span> is too small the algorithms will become unstable. The values of <span class="math inline">\(\gamma\)</span> are updated for each value of <span class="math inline">\(\lambda\)</span>. Let <span class="math inline">\({\bf r}^k\)</span> be the vector of residuals corresponding to the estimator using the <span class="math inline">\(k\)</span>th value of <span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\lambda^k\)</span>, and <span class="math inline">\(Q_{.1}(|{\bf r}^k|)\)</span> be the 10th percentile of the absolute values of these residuals. For an individual penalty we use what is implemented in <a href="https://cran.r-project.org/package=hqreg">hqreg</a>, which is very similar to what is outlined by <span class="citation" data-cites="huber_cd">Yi and Huang (<a href="#ref-huber_cd" role="doc-biblioref">2017</a>)</span>. Specifically,
<span class="math display">\[\begin{eqnarray*}
\gamma^k &amp;=&amp; I(|1-\tau|&lt; .05)\max\left\{.0001,\min\left[\gamma^{k-1},Q_{.01}(|{\bf r}^k|)\right]\right\} \\
&amp;&amp; + I(|1-\tau|&gt;=.05)\max\left\{.001,\min\left[\gamma^{k-1},Q_{.1}(|{\bf r}^k|)\right]\right\}.
\end{eqnarray*}\]</span></p>
<p>For the group penalties, as presented in <span class="citation" data-cites="SherwoodLi2022">Sherwood and Li (<a href="#ref-SherwoodLi2022" role="doc-biblioref">2022</a>)</span>,
<span class="math display">\[\begin{equation*}
\gamma^k = \min\left\{4, \max\left[.001, Q_{.1}(|{\bf r}^k|) \right] \right\}.
\end{equation*}\]</span></p>
<p>One reason the linear and second-order cone programming problems are computationally slow for larger values of <span class="math inline">\(p\)</span> is they optimize across all potential predictors. <span class="citation" data-cites="tibshirani2012strong">Tibshirani et al. (<a href="#ref-tibshirani2012strong" role="doc-biblioref">2012</a>)</span> proposed a strong rule for discarding predictors that can greatly decrease the computational complexity of penalized optimization. The strong rule assumes a differentiable loss function and thus is only implemented for the Huber approximations. In addition, the Huber approximation approaches use warm starts across a path of potential <span class="math inline">\(\lambda\)</span> values. It starts with the largest potential value of <span class="math inline">\(\lambda\)</span> then uses that solution as the initial value for the next largest potential value of <span class="math inline">\(\lambda\)</span>. This iterative process is continued until all solutions have been found. The linear programming approaches rely on <code>quantreg::rq()</code> which does not have an option for initial values. Thus warm starts are not implemented for those approaches. For an individual penalty, the calculations of the strong rule are done in the package <a href="https://cran.r-project.org/package=hqreg">hqreg</a>, and details can be found in <span class="citation" data-cites="huber_cd">Yi and Huang (<a href="#ref-huber_cd" role="doc-biblioref">2017</a>)</span>. A similar approach has been used for group penalties <span class="citation" data-cites="SherwoodLi2022">(<a href="#ref-SherwoodLi2022" role="doc-biblioref">Sherwood and Li 2022</a>)</span>.</p>
<h2 data-number="3" id="description-of-functions"><span class="header-section-number">3</span> Description of functions</h2>
<p>Below are the main functions of <a href="https://cran.r-project.org/package=rqPen">rqPen</a> and an S3 method that is unique to <a href="https://cran.r-project.org/package=rqPen">rqPen</a>, <code>bytau.plot()</code>.</p>
<!-- \begin{itemize} -->
<!-- \item `rq.pen()` minimizes \@ref(eq:multtau) to provide estimates of a quantile regression model, for potentially multiple quantiles, using an individual penalty of either elastic-net, adaptive lasso, SCAD, or MCP. If not specified, a sequence of $\lambda$ values are automatically generated. Returns an `rq.pen.seq` S3 object that works with `bytau.plot()`, `coef()`, `plot()`, `predict()`, and `print()` methods. -->
<!-- \item `rq.gq.pen()` minimizes \@ref(eq:gql) to provide a quantile regression model for multiple quantiles using the group quantile lasso penalty. It also returns an `rq.pen.seq` S3 object and can be used with the same methods listed above. -->
<!-- \item `rq.group.pen()`  minimizes \@ref(eq:multtaug) and is a group penalty version of `rq.pen()`. Users have access to group versions of the lasso, adaptive lasso, SCAD, and MCP penalties. The lasso penalty is restricted to the L2-norm. For other penalties users can choose between the L1- or L2-norm. It also returns an `rq.pen.seq` S3 object  and can be used with the same methods listed above. -->
<!-- \item `rq.pen.cv()` automates K-folds cross-validation for selecting the tuning parameters $\lambda$ and $a$. If $\lambda$ is left undefined, a sequence of $\lambda$ values will be automatically generated. For $a$, the default values described earlier will be used, unless the user specifies a sequence. If multiple quantiles are modeled then it provides the optimal pair of $(\lambda,a)$ for each quantile and the optimal pair across all quantiles.  Returns an `rq.pen.seq.cv` S3 object that works with `bytau.plot()`, `coef()`, `plot()`, `predict()`, and `print()` methods. -->
<!-- \item `rq.gq.pen.cv()` is the group quantile lasso version of `rq.pen.cv()`. It does everything listed above, but for the group quantile penalty. It also returns an `rq.pen.seq.cv` object.  -->
<!-- \item `rq.group.pen.cv()` is the group version of `rq.pen.cv()`. It does everything listed above, but for group penalties. It also returns an `rq.pen.seq.cv` object. -->
<!-- \item `qic.select()` takes an `rq.pen.seq` or `rq.pen.seq.cv` object and provides the optimal values of $(\lambda,a)$ using an information criterion, as explained in the previous section. It returns an ``qic.select`' S3 object that works with `coef()`, `predict()`, and `print()` methods. -->
<!-- \item `bytau.plot()` plots coefficients estimates for each predictor as a function of $\tau$. -->
<!-- \end{itemize} -->
<ul>
<li><code>rq.pen()</code> minimizes <a href="#eq:multtau">(14)</a> to provide estimates of a quantile regression model, for potentially multiple quantiles, using an individual penalty of either elastic-net, adaptive lasso, SCAD, or MCP. If not specified, a sequence of <span class="math inline">\(\lambda\)</span> values is automatically generated. Returns an <code>rq.pen.seq</code> S3 object that works with <code>bytau.plot()</code>, <code>coef()</code>, <code>plot()</code>, <code>predict()</code>, and <code>print()</code> methods.</li>
<li><code>rq.gq.pen()</code> minimizes <a href="#eq:gql">(16)</a> to provide a quantile regression model for multiple quantiles using the group quantile lasso penalty. It also returns an <code>rq.pen.seq</code> S3 object and can be used with the same methods listed above.</li>
<li><code>rq.group.pen()</code> minimizes <a href="#eq:multtaug">(15)</a> and is a group penalty version of <code>rq.pen()</code>. Users have access to group versions of the lasso, adaptive lasso, SCAD, and MCP penalties. The lasso penalty is restricted to the L2-norm. For other penalties, users can choose between the L1- or L2-norm. It also returns an <code>rq.pen.seq</code> S3 object and can be used with the same methods listed above.</li>
<li><code>rq.pen.cv()</code> automates K-folds cross-validation for selecting the tuning parameters <span class="math inline">\(\lambda\)</span> and <span class="math inline">\(a\)</span>. If <span class="math inline">\(\lambda\)</span> is left undefined, a sequence of <span class="math inline">\(\lambda\)</span> values will be automatically generated. For <span class="math inline">\(a\)</span>, the default values described earlier will be used, unless the user specifies a sequence. If multiple quantiles are modeled then it provides the optimal pair of <span class="math inline">\((\lambda,a)\)</span> for each quantile and the optimal pair across all quantiles. Returns an <code>rq.pen.seq.cv</code> S3 object that works with <code>bytau.plot()</code>, <code>coef()</code>, <code>plot()</code>, <code>predict()</code>, and <code>print()</code> methods.</li>
<li><code>rq.gq.pen.cv()</code> is the group quantile lasso version of <code>rq.pen.cv()</code>. It does everything listed above, but for the group quantile penalty. It also returns an <code>rq.pen.seq.cv</code> object.</li>
<li><code>rq.group.pen.cv()</code> is the group version of <code>rq.pen.cv()</code>. It does everything listed above, but for group penalties. It also returns an <code>rq.pen.seq.cv</code> object.</li>
<li><code>qic.select()</code> takes an <code>rq.pen.seq</code> or <code>rq.pen.seq.cv</code> object and provides the optimal values of <span class="math inline">\((\lambda,a)\)</span> using an information criterion, as explained in the previous section. It returns a `<code>qic.select</code>’ S3 object that works with <code>coef()</code>, <code>predict()</code>, and <code>print()</code> methods.</li>
<li><code>bytau.plot()</code> plots coefficient estimates for each predictor as a function of <span class="math inline">\(\tau\)</span>.</li>
</ul>
<p>The <code>rq.pen()</code> function provides estimates of conditional quantiles from minimizing a penalized objective function for a sequence of <span class="math inline">\(\lambda\)</span> values.</p>
<!---
\code{rq.pen(x, y, tau = 0.5, lambda = NULL,
        penalty = c("LASSO", "Ridge", "ENet", "aLASSO", "SCAD", "MCP"),
        a = NULL, nlambda = 100, eps = ifelse(nrow(x) < ncol(x), 0.05, 0.01),
        penalty.factor = rep(1, ncol(x)),
        alg = ifelse(sum(dim(x)) < 200, "br", "huber"), scalex = TRUE,
        tau.penalty.factor = rep(1, length(tau)),
        coef.cutoff = 1e-08, max.iter = 10000, converge.eps = 1e-07,
        lambda.discard = TRUE, weights=NULL, ...)} \newline
--->
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>rq.pen</span><span class='op'>(</span><span class='va'>x</span>, <span class='va'>y</span>, tau <span class='op'>=</span> <span class='fl'>0.5</span>, lambda <span class='op'>=</span> <span class='cn'>NULL</span>,</span>
<span>        penalty <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"LASSO"</span>, <span class='st'>"Ridge"</span>, <span class='st'>"ENet"</span>, <span class='st'>"aLASSO"</span>, <span class='st'>"SCAD"</span>, <span class='st'>"MCP"</span><span class='op'>)</span>,</span>
<span>        a <span class='op'>=</span> <span class='cn'>NULL</span>, nlambda <span class='op'>=</span> <span class='fl'>100</span>, eps <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='op'>&lt;</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>, <span class='fl'>0.05</span>, <span class='fl'>0.01</span><span class='op'>)</span>,</span>
<span>        penalty.factor <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>ncol</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span>,</span>
<span>        alg <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/ifelse.html'>ifelse</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/dim.html'>dim</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>&lt;</span> <span class='fl'>200</span>, <span class='st'>"br"</span>, <span class='st'>"huber"</span><span class='op'>)</span>, scalex <span class='op'>=</span> <span class='cn'>TRUE</span>,</span>
<span>        tau.penalty.factor <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>tau</span><span class='op'>)</span><span class='op'>)</span>,</span>
<span>        coef.cutoff <span class='op'>=</span> <span class='fl'>1e-08</span>, max.iter <span class='op'>=</span> <span class='fl'>10000</span>, converge.eps <span class='op'>=</span> <span class='fl'>1e-07</span>,</span>
<span>        lambda.discard <span class='op'>=</span> <span class='cn'>TRUE</span>, weights<span class='op'>=</span><span class='cn'>NULL</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>The function <code>rq.pen()</code> requires a design matrix <code>x</code> and vector of response <code>y</code>. The predictors in the design matrix will be centered to have mean zero and standard deviation one when minimizing the penalized objective function unless <code>scalex</code> is set to FALSE. While the predictors are scaled, the coefficients are returned on the original scale of <code>x</code>. The quantiles modeled are set with <code>tau</code>. Users can choose the <code>penalty</code> function, with the default being lasso. A pre-specified lambda sequence can be set, <code>lambda</code>. If the user does not specify a sequence of <span class="math inline">\(\lambda\)</span> values then <span class="math inline">\(\lambda_{min}=\epsilon\lambda_{max}\)</span>, where <span class="math inline">\(\epsilon\)</span> can be set by <code>eps</code> and <code>nlambda</code> is the number of <span class="math inline">\(\lambda\)</span> values considered. Though very small values of <span class="math inline">\(\lambda\)</span> may be discarded if the coefficient estimates are not changing much with smaller values of <span class="math inline">\(\lambda\)</span>, unless <code>lambda.discard=FALSE</code>. For non-ridge or non-lasso penalties, a sequence of values can also be selected for <code>a</code>. If <code>a</code> is not set then default values depend on the penalty, elastic-net (<span class="math inline">\(a=0\)</span>), adaptive lasso (<span class="math inline">\(a=1\)</span>), SCAD (<span class="math inline">\(a=3.7\)</span>) and MCP (<span class="math inline">\(a=3\)</span>). Penalty factors can be set for predictors, <code>penalty.factor</code>, or quantiles, <code>tau.penalty.factor</code>. Observation weights are set using <code>weights</code>. The linear programming algorithms can provide very small, but non-zero estimates, and these coefficients are set to zero if they are below <code>coef.cutoff</code>. The choice of algorithm can be set using <code>alg</code>. Two linear programming algorithms are provided, the Barrodale and Roberts algorithm (br) <span class="citation" data-cites="br">(<a href="#ref-br" role="doc-biblioref">Barrodale and Roberts 1974</a>)</span>, as described in <span class="citation" data-cites="crq1">Koenker and D’Orey (<a href="#ref-crq1" role="doc-biblioref">1987</a>)</span> and <span class="citation" data-cites="crq2">Koenker and D’Orey (<a href="#ref-crq2" role="doc-biblioref">1994</a>)</span>, and the Frisch-Newton (fn) approach, described in <span class="citation" data-cites="portnoy1997">Portnoy and Koenker (<a href="#ref-portnoy1997" role="doc-biblioref">1997</a>)</span>. Both approaches rely on implementations in <a href="https://cran.r-project.org/package=quantreg">quantreg</a>. Setting <code>rq.pen(alg="huber")</code> uses the approach of <span class="citation" data-cites="huber_cd">Yi and Huang (<a href="#ref-huber_cd" role="doc-biblioref">2017</a>)</span> implemented in <a href="https://cran.r-project.org/package=hqreg">hqreg</a>. In addition, they can set the maximum number of iterations, <code>max.iter</code>, or convergence criteria, <code>converge.eps</code>, though these only apply to the ``huber’’ algorithms.</p>
<!---
\code{coef.rq.pen.seq(object, tau = NULL, a = NULL, lambda = NULL, modelsIndex = NULL,
    lambdaIndex = NULL, ...)} \newline
--->
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>coef.rq.pen.seq</span><span class='op'>(</span><span class='va'>object</span>, tau <span class='op'>=</span> <span class='cn'>NULL</span>, a <span class='op'>=</span> <span class='cn'>NULL</span>, lambda <span class='op'>=</span> <span class='cn'>NULL</span>, modelsIndex <span class='op'>=</span> <span class='cn'>NULL</span>,</span>
<span>    lambdaIndex <span class='op'>=</span> <span class='cn'>NULL</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>Using <code>coef()</code>, users can extract coefficients for specific quantiles, <code>tau</code>, and tuning parameters, <code>a</code> and <code>lambda</code>, from class <code>rq.pen.seq</code> object. Alternatively, they can directly specify the models and lambda values using <code>modelsIndex</code> and <code>lambdaIndex</code>, respectively. If none of these values are set then a matrix of coefficients for all quantiles and tuning parameters will be returned. </p>
<!---
\code{coef.rq.pen.seq.cv(object, septau = TRUE, cvmin = TRUE, useDefaults = TRUE,
    tau = NULL, ...)} \newline
--->
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>coef.rq.pen.seq.cv</span><span class='op'>(</span><span class='va'>object</span>, septau <span class='op'>=</span> <span class='cn'>TRUE</span>, cvmin <span class='op'>=</span> <span class='cn'>TRUE</span>, useDefaults <span class='op'>=</span> <span class='cn'>TRUE</span>,</span>
<span>    tau <span class='op'>=</span> <span class='cn'>NULL</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>With an <code>rq.pen.seq.cv</code> object, the <code>coef()</code> function returns coefficients based on the cross-validation results. If <code>cvmin=TRUE</code> then the tuning parameters associated with the minimum cross-validation error are returned, otherwise the one standard error rule is used. When <code>septau=TRUE</code> then the tuning parameters are optimized individually for each quantile. Users can identify the specific quantiles, <code>tau</code>, they wish to return. The default is to return results for all quantiles. Setting <code>useDefaults=FALSE</code> ignores the cross-validation results and users can use the arguments in <code>coef.rq.pen.seq()</code> to select coefficients. For instance, if <code>useDefaults=FALSE</code> and no other arguments are provided then all coefficients are returned.</p>
<p>Function <code>bytau.plot()</code> has a similar form for the <code>rq.pen.seq</code> and <code>rq.pen.seq.cv</code> objects. </p>
<!---
\code{bytau.plot.rq.pen.seq(x, a = NULL, lambda = NULL, lambdaIndex = NULL, ...)} \newline
--->
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>bytau.plot.rq.pen.seq</span><span class='op'>(</span><span class='va'>x</span>, a <span class='op'>=</span> <span class='cn'>NULL</span>, lambda <span class='op'>=</span> <span class='cn'>NULL</span>, lambdaIndex <span class='op'>=</span> <span class='cn'>NULL</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<!---
\code{bytau.plot.rq.pen.seq.cv(x, septau = TRUE, cvmin = TRUE, useDefaults = TRUE,
    ...)} \newline
--->
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>bytau.plot.rq.pen.seq.cv</span><span class='op'>(</span><span class='va'>x</span>, septau <span class='op'>=</span> <span class='cn'>TRUE</span>, cvmin <span class='op'>=</span> <span class='cn'>TRUE</span>, useDefaults <span class='op'>=</span> <span class='cn'>TRUE</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>These functions produce a plot of how coefficients change with quantiles. The arguments in <code>bytau.plot.rq.pen.seq()</code> and <code>bytau.plot.rq.pen.seq.cv()</code> are the same as those covered for <code>coef()</code>, but with one major difference. Only one vector of coefficients can be provided for each quantile. When using an <code>rq.pen.seq</code> object, users must specify a single pair of <span class="math inline">\((\lambda,a)\)</span>. When using an <code>rq.pen.seq.cv</code> object, users can rely on the default tools for selecting coefficients or specify a single pair of <span class="math inline">\(\lambda\)</span> and <span class="math inline">\(a\)</span>.</p>
<p>Both <code>rq.pen.seq</code> and <code>rq.pen.seq.cv</code> objects have <code>plot()</code> functions. Using <code>plot()</code> with an <code>rq.pen.seq.cv</code> object provides a plot of how the cross-validation error changes with <span class="math inline">\(\lambda\)</span> for each quantile.</p>
<!---
\code{plot.rq.pen.seq.cv(x, septau = TRUE, tau = NULL, logLambda = FALSE, main = NULL, ...)}
--->
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>plot.rq.pen.seq.cv</span><span class='op'>(</span><span class='va'>x</span>, septau <span class='op'>=</span> <span class='cn'>TRUE</span>, tau <span class='op'>=</span> <span class='cn'>NULL</span>, logLambda <span class='op'>=</span> <span class='cn'>FALSE</span>, main <span class='op'>=</span> <span class='cn'>NULL</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>If <code>septau=TRUE</code> then a separate plot is created for each quantile, otherwise one plot is created that combines the cross-validation error across all quantiles. Users can limit the plots to a subset of quantiles, <code>tau</code>. If <code>logLambda=TRUE</code> then the x-axis will be <span class="math inline">\(\log(\lambda)\)</span>. The <code>main</code> text can be set otherwise the main title depends on the name of the variable and the value of <code>septau</code>.</p>
<p>While <code>plot()</code> for an <code>rq.pen.seq</code> object creates a plot of the coefficients as they change with <span class="math inline">\(\lambda\)</span>.</p>
<!---
\code{plot.rq.pen.seq(x, vars = NULL, logLambda = TRUE, tau = NULL, a = NULL,
    lambda = NULL, modelsIndex = NULL, lambdaIndex = NULL, main = NULL,
    ...)}
--->
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>plot.rq.pen.seq</span><span class='op'>(</span><span class='va'>x</span>, vars <span class='op'>=</span> <span class='cn'>NULL</span>, logLambda <span class='op'>=</span> <span class='cn'>TRUE</span>, tau <span class='op'>=</span> <span class='cn'>NULL</span>, a <span class='op'>=</span> <span class='cn'>NULL</span>,</span>
<span>    lambda <span class='op'>=</span> <span class='cn'>NULL</span>, modelsIndex <span class='op'>=</span> <span class='cn'>NULL</span>, lambdaIndex <span class='op'>=</span> <span class='cn'>NULL</span>, main <span class='op'>=</span> <span class='cn'>NULL</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>The <code>plot()</code> function will create a plot for each combination of <span class="math inline">\(\tau\)</span> and <span class="math inline">\(a\)</span>. Users can specify specific values of <code>tau</code> and <code>a</code> if they only want to consider a subset. They can also limit the values of <code>lambda</code>. Choice of models and <span class="math inline">\(\lambda\)</span> values can also be done using <code>modelsIndex</code> and <code>lambdaIndex</code>. The <code>main</code> text can be set. If <code>logLambda=TRUE</code> then the <span class="math inline">\(\lambda\)</span> values are reported on the natural log scale.</p>
<h2 data-number="4" id="applications"><span class="header-section-number">4</span> Applications</h2>
<p>This section details how to use individual and group penalties. Different individual penalties and algorithms will be applied to the Barro data set from the <a href="https://cran.r-project.org/package=quantreg">quantreg</a> package. In addition, the group quantile penalty is used with the Barro data set. While the group penalties for predictors will be used with the Ames housing data <span class="citation" data-cites="de2011ames">(<a href="#ref-de2011ames" role="doc-biblioref">De Cock 2011</a>)</span> available in <a href="https://cran.r-project.org/package=AmesHousing">AmesHousing</a>.</p>
<h3 data-number="4.1" id="the-barro-data-set"><span class="header-section-number">4.1</span> The Barro data set</h3>
<p>The Barro data set, available in <a href="https://cran.r-project.org/package=quantreg">quantreg</a>, contains GDP growth rates and 13 other potential explanatory variables. See <span class="citation" data-cites="barro">Koenker and Machado (<a href="#ref-barro" role="doc-biblioref">1999</a>)</span> for more details.</p>
<h4 class="unnumbered" data-number="4.1.1" id="using-different-algorithms">Using different algorithms</h4>
<p>First, we look at fitting a model with the SCAD penalty for <span class="math inline">\(\tau=.5\)</span> using three different algorithms.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/bssherwood/rqpen/blob/master/ignore/rqPenArticle.pdf'>rqPen</a></span><span class='op'>)</span></span>
<span><span class='co'>#quantreg is required for rqPen, but call directly here</span></span>
<span><span class='co'>#because we need the Barro data set</span></span>
<span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://www.r-project.org'>quantreg</a></span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>barro</span><span class='op'>)</span></span>
<span><span class='va'>y</span> <span class='op'>&lt;-</span> <span class='va'>barro</span><span class='op'>$</span><span class='va'>y.net</span></span>
<span><span class='va'>x</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/SparseM/man/SparseM.ontology.html'>as.matrix</a></span><span class='op'>(</span><span class='va'>barro</span><span class='op'>[</span>,<span class='op'>-</span><span class='fl'>1</span><span class='op'>]</span><span class='op'>)</span></span>
<span><span class='va'>qbr</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/rq.pen.html'>rq.pen</a></span><span class='op'>(</span><span class='va'>x</span>,<span class='va'>y</span>,alg<span class='op'>=</span><span class='st'>"br"</span>, penalty<span class='op'>=</span><span class='st'>"SCAD"</span><span class='op'>)</span></span>
<span><span class='va'>qfn</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/rq.pen.html'>rq.pen</a></span><span class='op'>(</span><span class='va'>x</span>,<span class='va'>y</span>,alg<span class='op'>=</span><span class='st'>"fn"</span>, penalty<span class='op'>=</span><span class='st'>"SCAD"</span><span class='op'>)</span></span>
<span><span class='va'>qhuber</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/rq.pen.html'>rq.pen</a></span><span class='op'>(</span><span class='va'>x</span>,<span class='va'>y</span>,alg<span class='op'>=</span><span class='st'>"huber"</span>,penalty<span class='op'>=</span><span class='st'>"SCAD"</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>Where “br” and “fn” are the Barrodale and Roberts <span class="citation" data-cites="crq1 crq2">(<a href="#ref-crq1" role="doc-biblioref">Koenker and D’Orey 1987</a>; <a href="#ref-crq2" role="doc-biblioref">Koenker and D’Orey 1994</a>)</span> and Frisch-Newton <span class="citation" data-cites="portnoy1997">(<a href="#ref-portnoy1997" role="doc-biblioref">Portnoy and Koenker 1997</a>)</span> interior point method for linear programming problems implemented in <code>quantreg::rq()</code>. The following code takes the 25th value in the sequence of <span class="math inline">\(\lambda\)</span> values and compares the coefficients for the three different approaches creating three different <code>rq.pen.seq</code> objects.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>targetLambda</span> <span class='op'>&lt;-</span> <span class='va'>qbr</span><span class='op'>$</span><span class='va'>lambda</span><span class='op'>[</span><span class='fl'>25</span><span class='op'>]</span></span>
<span><span class='va'>brCoef</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/coef.html'>coefficients</a></span><span class='op'>(</span><span class='va'>qbr</span>,lambda<span class='op'>=</span><span class='va'>targetLambda</span><span class='op'>)</span></span>
<span><span class='va'>fnCoef</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/coef.html'>coefficients</a></span><span class='op'>(</span><span class='va'>qfn</span>, lambda<span class='op'>=</span><span class='va'>targetLambda</span><span class='op'>)</span></span>
<span><span class='va'>huberCoef</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/coef.html'>coefficients</a></span><span class='op'>(</span><span class='va'>qhuber</span>, lambda<span class='op'>=</span><span class='va'>targetLambda</span><span class='op'>)</span></span>
<span><span class='va'>coefDf</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='va'>brCoef</span>,<span class='va'>fnCoef</span>,<span class='va'>huberCoef</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>coefDf</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"br"</span>,<span class='st'>"fn"</span>,<span class='st'>"huber"</span><span class='op'>)</span></span>
<span><span class='va'>coefDf</span></span></code></pre>
</div>
<pre><code>                   br          fn       huber
intercept  0.02018841  0.02018852  0.02131606
lgdp2      0.00000000  0.00000000  0.00000000
mse2       0.00000000  0.00000000  0.00000000
fse2       0.00000000  0.00000000  0.00000000
fhe2       0.00000000  0.00000000  0.00000000
mhe2       0.00000000  0.00000000  0.00000000
lexp2      0.00000000  0.00000000  0.00000000
lintr2     0.00000000  0.00000000  0.00000000
gedy2     -0.01773640 -0.01773726 -0.06602351
Iy2        0.02366121  0.02366100  0.02917322
gcony2    -0.02007166 -0.02007206 -0.02269923
lblakp2   -0.01433958 -0.01433962 -0.01524323
pol2      -0.01112295 -0.01112292 -0.01067910
ttrad2     0.04986245  0.04986267  0.04863867</code></pre>
</div>
<p>The three different algorithms provide coefficient estimates that are different but similar. For larger data sets, with a large number of observations or variables, we recommend using the Huber approximation because this greatly reduces computational time while providing a reasonable solution <span class="citation" data-cites="huber_cd">(<a href="#ref-huber_cd" role="doc-biblioref">Yi and Huang 2017</a>)</span>. The following presents a plot of how the coefficient values from <code>qhuber</code> change with <span class="math inline">\(\lambda\)</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>qhuber</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="RJ-2025-017_files/figure-html5/unnamed-chunk-10-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<h4 class="unnumbered" data-number="4.1.2" id="multiple-quantiles">Multiple quantiles</h4>
<p>Next, we fit a model for multiple quantiles of <span class="math inline">\(\tau \in \{.1,.5,.9\}\)</span> and will also consider values of <span class="math inline">\(a \in \{3,4\}\)</span> for the additional tuning parameter for SCAD. A fit with multiple quantiles remains an <code>rq.pen.seq</code> object.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>qmult</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/rq.pen.html'>rq.pen</a></span><span class='op'>(</span><span class='va'>x</span>,<span class='va'>y</span>,tau<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>.1</span>,<span class='fl'>.5</span>,<span class='fl'>.9</span><span class='op'>)</span>,a<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>3</span>,<span class='fl'>4</span><span class='op'>)</span>,penalty<span class='op'>=</span><span class='st'>"SCAD"</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>The following code provides a plot of the coefficients for <span class="math inline">\(\tau=.1\)</span> and <span class="math inline">\(a=4\)</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>qmult</span>, tau<span class='op'>=</span><span class='fl'>.1</span>,a<span class='op'>=</span><span class='fl'>4</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="RJ-2025-017_files/figure-html5/unnamed-chunk-12-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<h4 class="unnumbered" data-number="4.1.3" id="using-ic-to-select-tuning-parameters">Using IC to select tuning parameters</h4>
<p>When jointly optimizing the tuning parameter selection the same values of <span class="math inline">\(\lambda\)</span> and <span class="math inline">\(a\)</span> will be selected for each quantile by minimizing <a href="#eq:jointQic">(21)</a>. Otherwise, the optimal tuning parameters will be selected by minimizing <a href="#eq:qic">(17)</a> for each quantile separately. Users can specify <code>method="AIC"</code>, <code>method="BIC"</code> or <code>method="PBIC"</code> for a large <span class="math inline">\(p\)</span> BIC proposed by <span class="citation" data-cites="qrbic">Lee et al. (<a href="#ref-qrbic" role="doc-biblioref">2014</a>)</span>. When doing joint selection, users can also specify <code>weights</code>, <span class="math inline">\(z_b\)</span> in <a href="#eq:jointQic">(21)</a>. The default is to provide an equal weight, of one, for all quantiles.</p>
<p>The below code selects the tuning parameters of <span class="math inline">\((\lambda,a)\)</span> for the <code>qmult</code> model using AIC. This code will create two <code>qic.select</code> objects, <code>qic_sep</code> and <code>qic_joint</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>qic_sep</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/qic.select.html'>qic.select</a></span><span class='op'>(</span><span class='va'>qmult</span>,method<span class='op'>=</span><span class='st'>"AIC"</span><span class='op'>)</span></span>
<span><span class='va'>qic_joint</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/qic.select.html'>qic.select</a></span><span class='op'>(</span><span class='va'>qmult</span>, method<span class='op'>=</span><span class='st'>"AIC"</span>, septau<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>Below provides the tuning parameters selected by the two different approaches. The <code>modelsInfo</code> attribute provides information on the model selected for each quantile.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>qic_sep</span><span class='op'>$</span><span class='va'>modelsInfo</span></span></code></pre>
</div>
<pre><code>     tau modelIndex     a      minQIC lambdaIndex      lambda
   &lt;num&gt;      &lt;int&gt; &lt;num&gt;       &lt;num&gt;       &lt;int&gt;       &lt;num&gt;
1:   0.1          1     3 -0.80983835         100 0.001744156
2:   0.5          3     3  0.05987224          71 0.006721156
3:   0.9          5     3 -0.86499611          82 0.004029227</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>qic_joint</span><span class='op'>$</span><span class='va'>modelsInfo</span></span></code></pre>
</div>
<pre><code>         modelIndex a tau      minQIC lambdaIndex      lambda
tau0.1a3          1 3 0.1 -0.80983835         100 0.001744156
tau0.5a3          3 3 0.5  0.06799888         100 0.001744156
tau0.9a3          5 3 0.9 -0.86105610         100 0.001744156</code></pre>
</div>
<p>The above tables provide the following information for the models selected for each quantile.</p>
<p>When defining <code>qic_joint</code> we set <code>septau=FALSE</code>. Thus only a single pair of <span class="math inline">\((\lambda,a)\)</span> is selected to the be the best for all nine quantiles. That is why the values of <code>lambda</code> and <code>a</code> are the same for all quantiles in <code>qic_joint$modelsInfo</code>, but can differ in <code>qic_sep$modelsInfo</code>. The coefficients can be extracted using <code>coef()</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/stats/coef.html'>coef</a></span><span class='op'>(</span><span class='va'>qic_sep</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>                tau=0.1      tau=0.5      tau=0.9
intercept  0.0238504916 -0.024276359 -0.045586339
lgdp2     -0.0267825117 -0.024713637 -0.034210392
mse2       0.0129219159  0.010444821  0.017737529
fse2       0.0008487373  0.000000000 -0.004439600
fhe2      -0.0186593726  0.000000000  0.000000000
mhe2       0.0081453667  0.000000000  0.000000000
lexp2      0.0487678850  0.057746988  0.085483398
lintr2    -0.0010966058 -0.001761961 -0.002291657
gedy2     -0.3527018635 -0.068097230 -0.035234865
Iy2        0.0853432506  0.081856907  0.057654273
gcony2    -0.2044733845 -0.098175742 -0.084273528
lblakp2   -0.0226578393 -0.026987474 -0.033076575
pol2      -0.0282535823 -0.025880939  0.000000000
ttrad2     0.1155784172  0.140594440  0.225825288</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/stats/coef.html'>coef</a></span><span class='op'>(</span><span class='va'>qic_joint</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>                tau=0.1       tau=0.5      tau=0.9
intercept  0.0238504916 -0.0326938848 -0.039997413
lgdp2     -0.0267825117 -0.0268792897 -0.034043618
mse2       0.0129219159  0.0108844720  0.018840828
fse2       0.0008487373  0.0000000000 -0.005800585
fhe2      -0.0186593726  0.0003025406  0.000000000
mhe2       0.0081453667  0.0052844922  0.002750732
lexp2      0.0487678850  0.0639833573  0.083730580
lintr2    -0.0010966058 -0.0020594254 -0.002348213
gedy2     -0.3527018635 -0.0471775148 -0.047890233
Iy2        0.0853432506  0.0797565855  0.059880455
gcony2    -0.2044733845 -0.0973417525 -0.088821669
lblakp2   -0.0226578393 -0.0264829810 -0.033128704
pol2      -0.0282535823 -0.0314157975  0.001284625
ttrad2     0.1155784172  0.1654870976  0.227607952</code></pre>
</div>
<h4 class="unnumbered" data-number="4.1.4" id="predictions-from-models">Predictions from models</h4>
<p>Users can make predictions from either the <code>qic.select</code> or <code>rq.pen.seq</code> objects. Prediction from a <code>qic.select</code> object will return predictions for all quantiles modeled.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># creating new data using the mean of all variables</span></span>
<span><span class='va'>newData</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='va'>barro</span>,<span class='fl'>2</span>,<span class='va'>mean</span><span class='op'>)</span><span class='op'>[</span><span class='op'>-</span><span class='fl'>1</span><span class='op'>]</span> <span class='co'>#removing response</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/stats/predict.html'>predict</a></span><span class='op'>(</span><span class='va'>qic_sep</span>,<span class='va'>newData</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>           tau=0.1    tau=0.5    tau=0.9
[1,] -0.0002466134 0.01894553 0.03922352</code></pre>
</div>
<p>The default for <code>predict()</code> for an <code>rq.pen.seq</code> object is to return predictions for all values of <span class="math inline">\(\lambda\)</span>, <span class="math inline">\(a\)</span>, and <span class="math inline">\(\tau\)</span>. For <code>qmult</code> this results in 900 predictions for one row of predictors because there are 100 <span class="math inline">\(\lambda\)</span> values, 3 <span class="math inline">\(a\)</span> values, and 3 quantiles being modeled.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>allPreds</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/predict.html'>predict</a></span><span class='op'>(</span><span class='va'>qmult</span>,<span class='va'>newData</span><span class='op'>)</span></span>
<span><span class='va'>allPreds</span><span class='op'>[</span>,<span class='fl'>1</span><span class='op'>:</span><span class='fl'>5</span><span class='op'>]</span> <span class='co'>#Present the first five predictions</span></span></code></pre>
</div>
<pre><code>tau0.1a3 L1 tau0.1a3 L2 tau0.1a3 L3 tau0.1a3 L4 tau0.1a3 L5 
-0.01338804 -0.01338804 -0.01338804 -0.01338804 -0.01338804 </code></pre>
</div>
<p>Following the <code>coef.rq.pen.seq</code> code presented in the previous section, users can specify predictions from specific values of <span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\tau\)</span>, or <span class="math inline">\(a\)</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>somePreds</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/predict.html'>predict</a></span><span class='op'>(</span><span class='va'>qmult</span>,<span class='va'>newData</span>, tau<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>.1</span>,<span class='fl'>.5</span>,<span class='fl'>.9</span><span class='op'>)</span>,a<span class='op'>=</span><span class='fl'>4</span>,lambda<span class='op'>=</span><span class='va'>qmult</span><span class='op'>$</span><span class='va'>lambda</span><span class='op'>[</span><span class='fl'>25</span><span class='op'>]</span><span class='op'>)</span></span>
<span><span class='va'>somePreds</span></span></code></pre>
</div>
<pre><code>         tau0.1a4   tau0.5a4   tau0.9a4
[1,] -0.008353503 0.01773852 0.05166048</code></pre>
</div>
<h4 class="unnumbered" data-number="4.1.5" id="cross-validation-for-tuning-parameter-selection">Cross-validation for tuning parameter selection</h4>
<p>The function <code>rq.pen.cv()</code> creates an <code>rq.pen.seq.cv</code> object that can be used to select tuning parameters. Below we consider a similar model but use elastic net instead of SCAD. We set <code>tauWeights</code> to have <span class="math inline">\(z_b=\tau_b(1-\tau_b)\)</span>, which provides more weights to errors at the median and less to the more extreme quantiles.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>)</span></span>
<span><span class='va'>tauVals</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>.1</span>,<span class='fl'>.5</span>,<span class='fl'>.9</span><span class='op'>)</span></span>
<span><span class='va'>qcv</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/rq.pen.cv.html'>rq.pen.cv</a></span><span class='op'>(</span><span class='va'>x</span>,<span class='va'>y</span>,tau<span class='op'>=</span><span class='va'>tauVals</span>,a<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>.1</span>,<span class='fl'>.5</span>,<span class='fl'>1</span><span class='op'>)</span>,penalty<span class='op'>=</span><span class='st'>"ENet"</span>,</span>
<span>                 tauWeights <span class='op'>=</span> <span class='va'>tauVals</span><span class='op'>*</span><span class='op'>(</span><span class='fl'>1</span><span class='op'>-</span><span class='va'>tauVals</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>The <code>plot()</code> function with an <code>rq.pen.seq.cv</code> object can create a plot of cross-validation error for a single quantile, <a href="#eq:cv1">(18)</a>, or across all quantiles, <a href="#eq:cv3">(20)</a>. The following provides a plot for cross-validation error for <span class="math inline">\(\tau=.5\)</span> as a function of <span class="math inline">\(\lambda\)</span> and <span class="math inline">\(a\)</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>qcv</span>,tau<span class='op'>=</span><span class='fl'>.5</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="RJ-2025-017_files/figure-html5/unnamed-chunk-20-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<p>The error bars are only provided for the sequence associated with the optimal value of <span class="math inline">\(a\)</span>, <span class="math inline">\(a=.5\)</span> in this example. The dashed lines are only for that optimal value, too. The first dashed line is at the value of <span class="math inline">\(\lambda\)</span> that minimizes the cross-validation error, while the second dashed line is at the value of <span class="math inline">\(\lambda\)</span> which is one standard error above the minimum value. Users can also get a similar plot that describes how the cross-validation error across all quantiles changes with <span class="math inline">\(\lambda\)</span> and <span class="math inline">\(a\)</span>. If a user is only interested in one quantile then the above and below plot would be identical.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>qcv</span>,septau<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="RJ-2025-017_files/figure-html5/unnamed-chunk-21-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<p>For an <code>rq.pen.seq.cv</code> object the <code>predict()</code> function relies on the <code>coef()</code> function. Below are four predictions using the four combinations of <code>septau</code> and <code>cvmin</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>p1</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/predict.html'>predict</a></span><span class='op'>(</span><span class='va'>qcv</span>,<span class='va'>newData</span><span class='op'>)</span></span>
<span><span class='va'>p2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/predict.html'>predict</a></span><span class='op'>(</span><span class='va'>qcv</span>,<span class='va'>newData</span>, septau <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></span>
<span><span class='va'>p3</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/predict.html'>predict</a></span><span class='op'>(</span><span class='va'>qcv</span>,<span class='va'>newData</span>, cvmin<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>)</span></span>
<span><span class='va'>p4</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/predict.html'>predict</a></span><span class='op'>(</span><span class='va'>qcv</span>,<span class='va'>newData</span>, septau<span class='op'>=</span><span class='cn'>FALSE</span>,cvmin<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>)</span></span>
<span><span class='va'>pred</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>p1</span><span class='op'>)</span>,<span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>p2</span><span class='op'>)</span>,<span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>p3</span><span class='op'>)</span>,<span class='fu'><a href='https://rdrr.io/r/base/t.html'>t</a></span><span class='op'>(</span><span class='va'>p4</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>pred</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"sepMin"</span>,<span class='st'>"jointMin"</span>,<span class='st'>"sep1se"</span>,<span class='st'>"joint1se"</span><span class='op'>)</span></span>
<span><span class='va'>pred</span></span></code></pre>
</div>
<pre><code>                sepMin     jointMin       sep1se     joint1se
tau0.1a0.1 -0.00170812 -0.002409442 -0.006801621 -0.006044798
tau0.5a0.1  0.01895257  0.018958411  0.019494565  0.017972817
tau0.9a1    0.03892322  0.038575608  0.051660477  0.051154967</code></pre>
</div>
<!-- For `cvMin=TRUE`, the default setting, the model predictions are the same for `septau=TRUE` and `septau=FALSE`, because the optimal $(\lambda,a)$ pair is the same for each quantile and thus will be the optimal pair across all quantiles. However, there are some differences in the one standard error approach.  -->
<p>The <code>print()</code> of an <code>rq.pen.seq.cv</code> object provides information about the cross-validation results.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>qcv</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/rq.pen.cv.html'>rq.pen.cv</a></span><span class='op'>(</span><span class='va'>x</span>,<span class='va'>y</span>,tau<span class='op'>=</span><span class='va'>tauVals</span>,a<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>.1</span>,<span class='fl'>.5</span>,<span class='fl'>1</span><span class='op'>)</span>,penalty<span class='op'>=</span><span class='st'>"ENet"</span>,</span>
<span>                 tauWeights <span class='op'>=</span> <span class='va'>tauVals</span><span class='op'>*</span><span class='op'>(</span><span class='fl'>1</span><span class='op'>-</span><span class='va'>tauVals</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='va'>qcv</span></span></code></pre>
</div>
<pre><code>
Cross validation tuning parameter optimized for each quantile
     tau       minCv      lambda lambdaIndex  lambda1se
   &lt;num&gt;       &lt;num&gt;       &lt;num&gt;       &lt;int&gt;      &lt;num&gt;
1:   0.1 0.003125878 0.002005356          98 0.09308038
2:   0.5 0.006990980 0.008095657          78 0.23056721
3:   0.9 0.002900349 0.013193903          71 0.08680707
   lambda1seIndex     a         cvse modelsIndex nonzero  nzse
            &lt;int&gt; &lt;num&gt;        &lt;num&gt;       &lt;int&gt;   &lt;int&gt; &lt;int&gt;
1:             43   0.5 0.0007264311           2      14     5
2:             30   0.5 0.0014415984           5      11     3
3:             44   0.1 0.0006872771           7      13    11

Cross validation tuning parameter optimized across all quantiles
     tau       minCv     lambda lambdaIndex lambda1se lambda1seIndex
   &lt;num&gt;       &lt;num&gt;      &lt;num&gt;       &lt;int&gt;     &lt;num&gt;          &lt;int&gt;
1:   0.1 0.002291483 0.01230468          72 0.3268231             25
2:   0.5 0.002291483 0.01230468          72 0.3268231             25
3:   0.9 0.002291483 0.01230468          72 0.3268231             25
       a         cvse modelsIndex nonzero  nzse
   &lt;num&gt;        &lt;num&gt;       &lt;int&gt;   &lt;int&gt; &lt;int&gt;
1:   0.1 0.0003916137           1      14     6
2:   0.1 0.0003916137           4      13    11
3:   0.1 0.0003916137           7      13     4</code></pre>
</div>
<p>The printed output returns results optimized at each quantile and grouping across all quantiles.</p>
<h4 class="unnumbered" data-number="4.1.6" id="crossing-quantiles">Crossing quantiles</h4>
<p>Crossing quantiles occurs when multiple quantiles are being modeled and predictions of a lower quantile are larger than predictions from an upper quantile. Take the following example using the Barro data set that looks at the fitted values for a model of the .1, .2 and .5 quantiles, for the 50th value in the <span class="math inline">\(\lambda\)</span> sequence.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>qCross</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/rq.pen.cv.html'>rq.pen.cv</a></span><span class='op'>(</span><span class='va'>x</span>,<span class='va'>y</span>,tau<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>.1</span>,<span class='fl'>.2</span>, <span class='fl'>.5</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='va'>fitCross</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/predict.html'>predict</a></span><span class='op'>(</span><span class='va'>qCross</span>,newx<span class='op'>=</span><span class='va'>x</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>Warning in checkCrossSep(preds, sort, object$fit$penalty): Crossing
quantiles at observations 17, 28, 40, 79, 94, 109, 139, 152 when
using seperate optimization for each lambda. Setting septau=FALSE may
reduce the number of crossings. In addition, using rq.gq.pen() may
reduce the number of crossings.</code></pre>
</div>
<p>Warnings are provided about potential crossings at observations 17, 28, 40, 79, 94, 109, 139, and 152. Further inspection verifies this is the case as the fitted values at these observations have the .1 quantile larger than the .2 quantile. Examining these predictions verifies that is the case.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>fitCross</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>17</span>, <span class='fl'>28</span>, <span class='fl'>40</span>, <span class='fl'>79</span>, <span class='fl'>94</span>, <span class='fl'>109</span>, <span class='fl'>139</span>, <span class='fl'>152</span><span class='op'>)</span>,<span class='op'>]</span></span></code></pre>
</div>
<pre><code>                     tau0.1a1     tau0.2a1     tau0.5a1
Zimbabwe75       2.477067e-02  0.024519238  0.031127465
United_States75  1.252791e-02  0.009830172  0.014629574
Indonesia75      2.545159e-02  0.024898722  0.037985364
Gambia85        -3.663382e-03 -0.004988545  0.009901519
Zaire85         -3.490489e-02 -0.034921445 -0.019417390
United_States85 -5.850236e-05 -0.005162521  0.005242369
Yemen85          1.213681e-02  0.006765797  0.016998950
Norway85         3.488170e-02  0.034015862  0.046534611</code></pre>
</div>
<p>There is an option to sort the predictions when there are crosses. The default is not to do this because crossing quantiles can often be evidence of a misspecified model. Even when using the sort option there is a warning provided that the original predictions had crossing values.
<a href="#">quantregGrowth</a></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>fitSort</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/predict.html'>predict</a></span><span class='op'>(</span><span class='va'>qCross</span>, newx<span class='op'>=</span><span class='va'>x</span>, sort<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>Warning in checkCrossSep(preds, sort, object$fit$penalty): Quantile
predictions sorted at observations 17, 28, 40, 79, 94, 109, 139, 152
when using seperate optimization for each lambda. Setting
septau=FALSE may reduce the number of crossings. In addition, using
rq.gq.pen() may reduce the number of crossings.</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>fitSort</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>17</span>, <span class='fl'>28</span>, <span class='fl'>40</span>, <span class='fl'>79</span>, <span class='fl'>94</span>, <span class='fl'>109</span>, <span class='fl'>139</span>, <span class='fl'>152</span><span class='op'>)</span>, <span class='op'>]</span></span></code></pre>
</div>
<pre><code>                    tau0.1a1      tau0.2a1     tau0.5a1
Zimbabwe75       0.024519238  2.477067e-02  0.031127465
United_States75  0.009830172  1.252791e-02  0.014629574
Indonesia75      0.024898722  2.545159e-02  0.037985364
Gambia85        -0.004988545 -3.663382e-03  0.009901519
Zaire85         -0.034921445 -3.490489e-02 -0.019417390
United_States85 -0.005162521 -5.850236e-05  0.005242369
Yemen85          0.006765797  1.213681e-02  0.016998950
Norway85         0.034015862  3.488170e-02  0.046534611</code></pre>
</div>
<h4 class="unnumbered" data-number="4.1.7" id="consistent-selection-across-quantiles-1">Consistent selection across quantiles</h4>
<p>Note the output for <code>coefficients(qic_sep)</code> and <code>coefficients(qic_joint)</code> both show the sparsity structure changes with the quantile being modeled. We find interpreting such results to be difficult. A penalty that enforces the same sparsity structure across quantiles would avoid this issue. The following code will minimize <a href="#eq:gql">(16)</a>. Coefficients from the 13th entry in the <span class="math inline">\(\lambda\)</span> sequence demonstrate that this approach provides consistency in variable selection across the quantiles. This property will hold for any value of <span class="math inline">\(\lambda\)</span>. The function <code>rq.gq.pen.cv()</code> can be used for cross-validation of this approach. The function <code>rq.gq.pen()</code> returns a <code>rq.pen.seq</code> object and thus functions such as <code>predict()</code>, <code>plot()</code>, and <code>qic.select()</code> will work as previously described.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>qCons</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/rq.gq.pen.html'>rq.gq.pen</a></span><span class='op'>(</span><span class='va'>x</span>,<span class='va'>y</span>,tau<span class='op'>=</span><span class='va'>tauVals</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/stats/coef.html'>coefficients</a></span><span class='op'>(</span><span class='va'>qCons</span>,lambdaIndex<span class='op'>=</span><span class='fl'>13</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>               tau0.1a1      tau0.5a1     tau0.9a1
intercept -0.1039717886  0.0565817889  0.212830365
lgdp2     -0.0002948174 -0.0003086188 -0.000304075
mse2       0.0000000000  0.0000000000  0.000000000
fse2       0.0000000000  0.0000000000  0.000000000
fhe2       0.0000000000  0.0000000000  0.000000000
mhe2       0.0000000000  0.0000000000  0.000000000
lexp2     -0.0094731519 -0.0093673950 -0.008568437
lintr2     0.0000000000  0.0000000000  0.000000000
gedy2      0.0000000000  0.0000000000  0.000000000
Iy2        0.0721650464  0.0699737010  0.070700812
gcony2    -0.0426233067 -0.0442306488 -0.038750494
lblakp2   -0.0290133965 -0.0290175588 -0.028373662
pol2      -0.0096515098 -0.0097285059 -0.010002021
ttrad2     0.0000000000  0.0000000000  0.000000000</code></pre>
</div>
<h3 data-number="4.2" id="group-penalty-example"><span class="header-section-number">4.2</span> Group Penalty Example</h3>
<p>The Ames Housing data contains many factor variables and is available in <a href="https://cran.r-project.org/package=AmesHousing">AmesHousing</a>. For simplicity of presentation, we consider only Overall_Cond, Garage_Type, Full_Bath, Fireplaces, and Lot_Config as potential predictors. Some of these are categorical predictors, and thus a group penalty can account for the group structure. For the group variables some factor levels are merged due to small number of observations in a group. For instance, we combine the groups FR2 and FR3, frontage on two or three sides, in Lot_Config because only 14 houses are in group FR3. The function <code>rq.group.pen()</code> creates an <code>rq.pen.seq</code> object, where the models are fit with a group penalty. In the below example, we fit a model for log sale price using these predictors but do not penalize the Lot_Config variable. For the other groups the square root of the group size is used as the group penalty factor. In addition, quantile-specific penalties of <span class="math inline">\(\tau(1-\tau)\)</span> are used. Two <code>rq.pen.seq</code> objects are fit, one with <span class="math inline">\(q=1\)</span>, <code>norm=1</code>, and the other with <span class="math inline">\(q=2\)</span>, <code>norm=2</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://github.com/topepo/AmesHousing'>AmesHousing</a></span><span class='op'>)</span></span>
<span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://forcats.tidyverse.org/'>forcats</a></span><span class='op'>)</span></span>
<span><span class='va'>ames</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/AmesHousing/man/make_ames.html'>make_ames</a></span><span class='op'>(</span><span class='op'>)</span></span>
<span><span class='va'>ames</span><span class='op'>$</span><span class='va'>Overall_Cond</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://forcats.tidyverse.org/reference/fct_collapse.html'>fct_collapse</a></span><span class='op'>(</span><span class='va'>ames</span><span class='op'>$</span><span class='va'>Overall_Cond</span>,</span>
<span>                                  below<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Very_Poor"</span>,<span class='st'>"Poor"</span>,<span class='st'>"Fair"</span>,<span class='st'>"Below_Average"</span><span class='op'>)</span>,</span>
<span>                                  above<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Above_Average"</span>,<span class='st'>"Good"</span>,<span class='st'>"Very_Good"</span>,</span>
<span>                                          <span class='st'>"Excellent"</span>, <span class='st'>"Very_Excellent"</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='va'>ames</span><span class='op'>$</span><span class='va'>Garage_Type</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://forcats.tidyverse.org/reference/fct_collapse.html'>fct_collapse</a></span><span class='op'>(</span><span class='va'>ames</span><span class='op'>$</span><span class='va'>Garage_Type</span>,</span>
<span>                      other<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"Basment"</span>,<span class='st'>"BuiltIn"</span>,<span class='st'>"CarPort"</span>,<span class='st'>"More_Than_Two_Types"</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='va'>ames</span><span class='op'>$</span><span class='va'>Lot_Config</span>  <span class='op'>&lt;-</span> <span class='fu'><a href='https://forcats.tidyverse.org/reference/fct_collapse.html'>fct_collapse</a></span><span class='op'>(</span><span class='va'>ames</span><span class='op'>$</span><span class='va'>Lot_Config</span>, frontage<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"FR2"</span>,<span class='st'>"FR3"</span><span class='op'>)</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>x_g</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/model.matrix.html'>model.matrix</a></span><span class='op'>(</span><span class='op'>~</span> <span class='va'>Overall_Cond</span><span class='op'>+</span><span class='va'>Garage_Type</span><span class='op'>+</span><span class='va'>Full_Bath</span></span>
<span>                          <span class='op'>+</span><span class='va'>Fireplaces</span><span class='op'>+</span><span class='va'>Lot_Config</span>,<span class='va'>ames</span><span class='op'>)</span><span class='op'>)</span><span class='op'>[</span>,<span class='op'>-</span><span class='fl'>1</span><span class='op'>]</span></span>
<span></span>
<span><span class='va'>y_g</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/Log.html'>log</a></span><span class='op'>(</span><span class='va'>ames</span><span class='op'>$</span><span class='va'>Sale_Price</span><span class='op'>)</span></span>
<span><span class='va'>g</span> <span class='op'>&lt;-</span>  <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>2</span><span class='op'>)</span>,<span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>2</span>,<span class='fl'>3</span><span class='op'>)</span>,<span class='fl'>3</span>,<span class='fl'>4</span>,<span class='fu'><a href='https://rdrr.io/r/base/rep.html'>rep</a></span><span class='op'>(</span><span class='fl'>5</span>,<span class='fl'>3</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='va'>gpf</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/MathFun.html'>sqrt</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/xtabs.html'>xtabs</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>g</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='va'>gpf</span><span class='op'>[</span><span class='fl'>5</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='fl'>0</span></span>
<span><span class='va'>qAmes1</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/rq.group.pen.html'>rq.group.pen</a></span><span class='op'>(</span><span class='va'>x_g</span>,<span class='va'>y_g</span>,groups<span class='op'>=</span><span class='va'>g</span>, group.pen.factor <span class='op'>=</span> <span class='va'>gpf</span>, tau<span class='op'>=</span><span class='va'>tauVals</span>,</span>
<span>                      penalty<span class='op'>=</span><span class='st'>"gSCAD"</span>,norm<span class='op'>=</span><span class='fl'>1</span>,tau.penalty.factor<span class='op'>=</span><span class='va'>tauVals</span><span class='op'>*</span><span class='op'>(</span><span class='fl'>1</span><span class='op'>-</span><span class='va'>tauVals</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='va'>qAmes2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/rq.group.pen.html'>rq.group.pen</a></span><span class='op'>(</span><span class='va'>x_g</span>,<span class='va'>y_g</span>,groups<span class='op'>=</span><span class='va'>g</span>, group.pen.factor <span class='op'>=</span> <span class='va'>gpf</span>, tau<span class='op'>=</span><span class='va'>tauVals</span>,</span>
<span>                      penalty<span class='op'>=</span><span class='st'>"gSCAD"</span>,tau.penalty.factor<span class='op'>=</span><span class='va'>tauVals</span><span class='op'>*</span><span class='op'>(</span><span class='fl'>1</span><span class='op'>-</span><span class='va'>tauVals</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>The below plots present how the median coefficient estimates change with <span class="math inline">\(\lambda\)</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>qAmes1</span>,tau<span class='op'>=</span><span class='fl'>.5</span><span class='op'>)</span></span></code></pre>
</div>
<img src="RJ-2025-017_files/figure-html5/unnamed-chunk-29-1.png" width="100%" style="display: block; margin: auto;" />
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>qAmes2</span>,tau<span class='op'>=</span><span class='fl'>.5</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="RJ-2025-017_files/figure-html5/unnamed-chunk-29-2.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<p>Two important things to note from the above two plots. First, for both models, the unpenalized coefficients for <code>Lot_Config</code> are never set to zero. Second, the group coefficients for <code>qAmes2</code> go to zero as a group. While the grouped coefficients for <code>qAmes1</code> tend to go to zero as a group but do not have to converge to zero at the same time. For instance, for <span class="math inline">\(\tau=.5\)</span> and using the 40th value of <span class="math inline">\(\lambda\)</span> the garage and overall condition coefficients from <code>qAmes1</code> provide a mix of zero and non-zero coefficients. This does not occur for the default setting of <span class="math inline">\(q=2\)</span>, <code>norm=2</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/stats/coef.html'>coefficients</a></span><span class='op'>(</span><span class='va'>qAmes1</span>,tau<span class='op'>=</span><span class='fl'>.5</span>, lambdaIndex<span class='op'>=</span><span class='fl'>40</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>                       tau0.5a3.7
intercept            11.425638982
Overall_CondAverage   0.005564408
Overall_Condabove     0.000000000
Garage_Typeother      0.000000000
Garage_TypeDetchd    -0.033878344
Garage_TypeNo_Garage -0.036836894
Full_Bath             0.346115172
Fireplaces            0.135504921
Lot_ConfigCulDSac     0.081517065
Lot_Configfrontage   -0.066503584
Lot_ConfigInside     -0.023914057</code></pre>
</div>
<p>The following plot presents how the coefficients for <code>Full_Bath</code> change with <span class="math inline">\(\tau\)</span> at the 40th <span class="math inline">\(\lambda\)</span> value, where <code>vars=7</code> because <code>Full_Bath</code> is the seventh coefficient.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/bytau.plot.html'>bytau.plot</a></span><span class='op'>(</span><span class='va'>qAmes1</span>,vars<span class='op'>=</span><span class='fl'>7</span>,lambdaIndex<span class='op'>=</span><span class='fl'>40</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="RJ-2025-017_files/figure-html5/unnamed-chunk-31-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<p>The <code>bytau.plot()</code> can also be used with an <code>rq.pen.seq.cv</code> object. The below code creates an <code>rq.pen.seq.cv</code> object using the group lasso penalty.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>qgl</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/rq.group.pen.cv.html'>rq.group.pen.cv</a></span><span class='op'>(</span><span class='va'>x_g</span>,<span class='va'>y_g</span>,<span class='va'>g</span>,tau<span class='op'>=</span><span class='va'>tauVals</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>The below code provides the plot for the coefficient of <code>GarageType_BuiltIn</code> using the one standard error rule and the <span class="math inline">\(\lambda\)</span> tuning parameter is selected to be optimal across all quantiles.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/pkg/rqPen/man/bytau.plot.html'>bytau.plot</a></span><span class='op'>(</span><span class='va'>qgl</span>,vars<span class='op'>=</span><span class='fl'>7</span>,septau<span class='op'>=</span><span class='cn'>FALSE</span>,cvmin<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>)</span></span></code></pre>
</div>
<p><img src="RJ-2025-017_files/figure-html5/unnamed-chunk-33-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<p>Other functions such as <code>predict()</code>, <code>coef()</code> and <code>qic.select()</code> work the same for a group penalty method as they do for an individual penalty, presented in the Barro data set example. The <code>rq.pen()</code>, <code>rq.group.pen()</code>, and <code>rq.gq.pen()</code> functions return <code>rq.pen.seq</code> objects. While, the <code>rq.pen.cv()</code>, <code>rq.group.pen.cv()</code> and <code>rq.gq.pen.cv()</code> functions return <code>rq.pen.seq.cv</code> objects.</p>
<h2 data-number="5" id="simulations"><span class="header-section-number">5</span> Simulations</h2>
<p>Alternative packages for convex penalized quantile regression that were discussed earlier are <a href="https://cran.r-project.org/package=quantreg">quantreg</a>, <a href="https://cran.r-project.org/package=conquer">conquer</a>, <a href="https://cran.r-project.org/package=hrqglas">hrqglas</a>, and <a href="https://cran.r-project.org/package=hqreg">hqreg</a>. The only one of these four that is not used by <a href="https://cran.r-project.org/package=rqPen">rqPen</a> is <a href="https://cran.r-project.org/package=conquer">conquer</a>. In this section two simulations settings are presented to compare the speed and quality of the solution found by <a href="https://cran.r-project.org/package=conquer">conquer</a> and <a href="https://cran.r-project.org/package=rqPen">rqPen</a> for the lasso and group lasso penalties. The maximum number of iterations for the algorithms are set to 5000 and the tolerance for convergence was set to .001. Otherwise the default options for both packages are used. Simulations were run on a Dell Latitude 7410 laptop with an Intel Core i7 vPRO processor.</p>
<h3 data-number="5.1" id="simulations-for-lasso-penalty"><span class="header-section-number">5.1</span> Simulations for lasso penalty</h3>
<p>Let <span class="math inline">\({\bf x}_i \sim N({\bf 0}, {\bf I}_p)\)</span> the responses are generated from a model of,<br />
<span class="math display" id="eq:dgp">\[\begin{equation}
y_i = {\bf x}_i^\top \boldsymbol \beta^* + \epsilon_i,
\tag{22}
\end{equation}\]</span>
where <span class="math inline">\(\boldsymbol \beta^*=(\tilde{\boldsymbol \beta},\mathbf{0}_{p-10})^\top\)</span>, <span class="math inline">\(\epsilon_i \sim N(0,1)\)</span>, and the errors are independent. The entries of <span class="math inline">\(\tilde{\boldsymbol \beta}\)</span> are generated from a standard normal distribution in each simulation. Data was generated fifty times for each combination of <span class="math inline">\(n \in \{200,2000,20000\}\)</span> and <span class="math inline">\(p \in \{30,100,300\}\)</span>. The methods were fit for 50 <span class="math inline">\(\lambda\)</span> values equally spaced from .05 to .25. For all methods we model <span class="math inline">\(\tau=.5\)</span>.</p>
<p>Three algorithms are considered, <code>rq.pen()</code> with <code>alg="br"</code> and <code>alg="huber"</code>, and <code>conquer.reg()</code>. In each simulation the time it took for the algorithms to run and the value of the objective functions at the 50 different <span class="math inline">\(\lambda\)</span> values were recorded. Using <code>rq.pen(alg="br")</code>, which relies on <a href="https://cran.r-project.org/package=quantreg">quantreg</a>, was dramatically slower than the other approaches. Results, that are consistent with the literature when comparing a penalized approximation approach to commonly used linear programming algorithms <span class="citation" data-cites="huber_cd highdConv">(<a href="#ref-huber_cd" role="doc-biblioref">Yi and Huang 2017</a>; <a href="#ref-highdConv" role="doc-biblioref">Tan et al. 2022</a>)</span>. Thus we focus on the speed differences between the Huber approximation in <a href="https://cran.r-project.org/package=rqPen">rqPen</a> and the convolution smoothing approach of <a href="https://cran.r-project.org/package=conquer">conquer</a>. Figure <a href="#fig:lassoSpeedCompare">1</a> presents a comparison of the base 10 log time of the two approaches for different values of <span class="math inline">\(n\)</span> and <span class="math inline">\(p\)</span>, which demonstrates that the Huber approximation approach in <a href="https://cran.r-project.org/package=rqPen">rqPen</a>, which relies on <a href="https://cran.r-project.org/package=hqreg">hqreg</a>, is consistently faster. To compare the accuracy of the algorithms, we evaluate a ratio of the objective function, using quantile loss, at the solution to the objective function from the BR algorithm. The BR approach does not replace the quantile loss with an approximation and thus this ratio is always greater than or equal to one for both alternative approaches. Figure <a href="#fig:lassoObjCompare">2</a> provides the ratios for different values of <span class="math inline">\(n\)</span> and <span class="math inline">\(p\)</span>. Note, the ratios also would depend on <span class="math inline">\(\lambda\)</span>, but that information is not included in the figure for ease of presentation. Both approaches are providing minimized objective values very similar to the exact approach, with the Huber approximation approach tending to be a little closer.</p>
<!-- \begin{figure} -->
<!--    \centering -->
<!--        \includegraphics[width=250px,height=250px]{figures/lassoSpeedCompare.pdf} -->
<!--        \caption{Log time in seconds of lasso quantile regression using conquer and the Huber approximation in rqPen} -->
<!--    \label{fig:lassoSpeedCompare} -->
<!-- \end{figure} -->
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:lassoSpeedCompare"></span>
<img src="RJ-2025-017_files/figure-html5/lassoSpeedCompare-1.png" alt="Log time in seconds of lasso quantile regression using conquer and the Huber approximation in rqPen." width="100%" />
<p class="caption">
Figure 1: Log time in seconds of lasso quantile regression using conquer and the Huber approximation in rqPen.
</p>
</div>
</div>
<!-- \begin{figure} -->
<!--    \centering -->
<!--        \includegraphics[width=250px,height=250px]{figures/lassoObjCompare.pdf} -->
<!--            \caption{Ratio of quantile regression lasso objective function at the rqPen-huber and conquer solutions, where rqPen-br is the baseline.} -->
<!--    \label{fig:lassoObjCompare} -->
<!-- \end{figure} -->
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:lassoObjCompare"></span>
<img src="RJ-2025-017_files/figure-html5/lassoObjCompare-1.png" alt="Ratio of quantile regression lasso objective function at the rqPen-huber and conquer solutions, where rqPen-br is the baseline." width="100%" />
<p class="caption">
Figure 2: Ratio of quantile regression lasso objective function at the rqPen-huber and conquer solutions, where rqPen-br is the baseline.
</p>
</div>
</div>
<h3 data-number="5.2" id="simulations-for-group-lasso-penalty"><span class="header-section-number">5.2</span> Simulations for group lasso penalty</h3>
<p>For group lasso let <span class="math inline">\(\mathbf{x}_{ij}=(x_{ij1},x_{ij2})^\top\)</span>, where <span class="math inline">\(P(x_{ij1}=1)=P(x_{ij2}=1)=1/3\)</span> and these are disjoint events. That is it represents a single observation of a multinomial random variable with 3 categories and equal probability of belonging to any of the three categories. Let <span class="math inline">\(\mathbf{x}_i = (\mathbf{x}_{i1},\ldots,\mathbf{x}_{ig})^\top\)</span>, where the predictors are independent of each other. The data generating mechanism for <span class="math inline">\(y\)</span> is the same as <a href="#eq:dgp">(22)</a>. Again, data was generated fifty times for each combination of <span class="math inline">\(n \in \{200,2000,20000\}\)</span> and <span class="math inline">\(g \in \{15,50,150\}\)</span>. The methods were fit for 50 <span class="math inline">\(\lambda\)</span> equally spaced values from .05 to .001. Again, for all methods we model <span class="math inline">\(\tau=.5\)</span>.</p>
<p>The group lasso methods implemented in rqPen, using <code>rq.group.pen()</code>, and conquer, using <code>conquer.reg(penalty="group",...)</code> are compared in these simulations. Both approaches replace the quantile loss function with a smooth approximation. When the quantile loss is used, the optimization problem is a second order cone programming problem that can be solved using convex optimization software. See <span class="citation" data-cites="SherwoodLi2022">Sherwood and Li (<a href="#ref-SherwoodLi2022" role="doc-biblioref">2022</a>)</span> for a comparison of the Huber approximation approach and using <a href="https://cran.r-project.org/package=Rmosek">Rmosek</a> for quantile regression with a group lasso penalty. To compare the two methods time and a ratio of the rqPen and conquer penalized objective functions evaluated at the solutions. Figure <a href="#fig:lassoSpeedCompareGroup">3</a> and <a href="#fig:lassoObjCompareGroup">4</a> present the time and objective function results, respectively. While, <a href="https://cran.r-project.org/package=rqPen">rqPen</a> tends to find a solution that provides a smaller value at the objective function, the results from <a href="https://cran.r-project.org/package=conquer">conquer</a> are noticeably faster. Note, the implementation of group lasso in <a href="https://cran.r-project.org/package=rqPen">rqPen</a> relies on <a href="https://cran.r-project.org/package=hrqglas">hrqglas</a>.</p>
<!-- \begin{figure} -->
<!--    \centering -->
<!--        \includegraphics[width=250px,height=250px]{figures/lassoSpeedCompareGroup.pdf} -->
<!--            \caption{Log time in seconds of group lasso quantile regression using conquer and the Huber approximation in rqPen.} -->
<!--              \label{fig:lassoSpeedCompareGroup} -->
<!-- \end{figure} -->
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:lassoSpeedCompareGroup"></span>
<img src="RJ-2025-017_files/figure-html5/lassoSpeedCompareGroup-1.png" alt="Log time in seconds of group lasso quantile regression using conquer and the Huber approximation in rqPen." width="100%" />
<p class="caption">
Figure 3: Log time in seconds of group lasso quantile regression using conquer and the Huber approximation in rqPen.
</p>
</div>
</div>
<!-- \begin{figure} -->
<!--    \centering -->
<!--        \includegraphics[width=250px,height=250px]{figures/lassoObjCompareGroup.pdf} -->
<!--    \caption{Ratio of quantile regression group lasso objective function at the rqPen and conquer group lasso solutions.} -->
<!--    \label{fig:lassoObjCompareGroup} -->
<!-- \end{figure} -->
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:lassoObjCompareGroup"></span>
<img src="RJ-2025-017_files/figure-html5/lassoObjCompareGroup-1.png" alt="Ratio of quantile regression group lasso objective function at the rqPen and conquer group lasso solutions." width="100%" />
<p class="caption">
Figure 4: Ratio of quantile regression group lasso objective function at the rqPen and conquer group lasso solutions.
</p>
</div>
</div>
<h2 data-number="6" id="summary"><span class="header-section-number">6</span> Summary</h2>
<p>The <a href="https://cran.r-project.org/package=rqPen">rqPen</a> package provides a penalized approach to estimating quantile regression models. The package supports the elastic-net, adaptive lasso, SCAD, and MCP penalty functions. For group predictors there are group penalty versions of the lasso, adaptive lasso, SCAD, and MCP penalties. In addition, users can use a group quantile penalty that ensures consistent variable selection across the quantiles. The package is available on CRAN.</p>
<p>Research in penalized quantile regression, and more generally quantile regression with large data sets, is an active area of research. The package currently does not include some recent innovations in the field. Censored quantile regression is a useful tool for survival analysis and an active area of research. <span class="citation" data-cites="hdCensQR">Zheng et al. (<a href="#ref-hdCensQR" role="doc-biblioref">2018</a>)</span> present a penalized quantile regression approach for estimating an interval of quantiles for censored data. While, <span class="citation" data-cites="effCenQR">Sze Ming Lee and Xu (<a href="#ref-effCenQR" role="doc-biblioref">2023</a>)</span> provides a unifying approach to estimation and inference for censored quantile regression that possibly could be extended to penalized settings. Another interesting work that provides a framework that could possibly be extended to penalized quantile regression is <span class="citation" data-cites="qrDist">Volgushev et al. (<a href="#ref-qrDist" role="doc-biblioref">2019</a>)</span>, which proposes a method for estimating quantile regression that splits up initial analysis across multiple computers and then combines the estimates. A challenge here would be dealing with the different sparsity patterns that could arise on the different machines. Inference for penalized methods has also been an active area of research. Recently, <span class="citation" data-cites="validQRInf">Alexandre Belloni and Kato (<a href="#ref-validQRInf" role="doc-biblioref">2019</a>)</span> propose an approach where they do not approximate the quantile loss function and <span class="citation" data-cites="ciHypHDQR">Yan et al. (<a href="#ref-ciHypHDQR" role="doc-biblioref">2023</a>)</span> provide an approach for when the loss function is approximated with a smoothing function. Quantiles are more easily understood when the response is continuous. For users interested in estimating conditional quantiles for count data, one easy solution is to jitter the response <span class="citation" data-cites="quantileCount">(<a href="#ref-quantileCount" role="doc-biblioref">Machado and Silva 2005</a>)</span>. All of these works suggest that there is a variety of existing work that could be developed into useful software and plenty of work to be done in the future.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h3 class="appendix" data-number="6.1" id="supplementary-materials"><span class="header-section-number">6.1</span> Supplementary materials</h3>
<p>Supplementary materials are available in addition to this article. It can be downloaded at
<a href="RJ-2025-017.zip">RJ-2025-017.zip</a></p>
<h3 class="appendix" data-number="6.2" id="cran-packages-used"><span class="header-section-number">6.2</span> CRAN packages used</h3>
<p><a href="https://cran.r-project.org/package=rqPen">rqPen</a>, <a href="https://cran.r-project.org/package=quantreg">quantreg</a>, <a href="https://cran.r-project.org/package=Rmosek">Rmosek</a>, <a href="https://cran.r-project.org/package=hqreg">hqreg</a>, <a href="https://cran.r-project.org/package=conquer">conquer</a>, <a href="https://cran.r-project.org/package=hrqglas">hrqglas</a>, <a href="https://cran.r-project.org/package=AmesHousing">AmesHousing</a></p>
<h3 class="appendix" data-number="6.3" id="cran-task-views-implied-by-cited-packages"><span class="header-section-number">6.3</span> CRAN Task Views implied by cited packages</h3>
<p><a href="https://cran.r-project.org/view=Econometrics">Econometrics</a>, <a href="https://cran.r-project.org/view=Environmetrics">Environmetrics</a>, <a href="https://cran.r-project.org/view=Optimization">Optimization</a>, <a href="https://cran.r-project.org/view=ReproducibleResearch">ReproducibleResearch</a>, <a href="https://cran.r-project.org/view=Robust">Robust</a></p>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-validQRInf" class="csl-entry" role="listitem">
V. C. Alexandre Belloni and K. Kato. Valid post-selection inference in high-dimensional approximately sparse quantile regression models. <em>Journal of the American Statistical Association</em>, 114(526): 749–758, 2019. URL <a href="https://doi.org/10.1080/01621459.2018.1442339">https://doi.org/10.1080/01621459.2018.1442339</a>.
</div>
<div id="ref-br" class="csl-entry" role="listitem">
I. Barrodale and F. D. K. Roberts. Solution of an overdetermined system of equations in the <span class="math inline">\(\ell_1\)</span> norm. <em>Communications of the ACM</em>, 17: 319–320, 1974. URL <a href="https://doi.org/10.1145/355616.36102">https://doi.org/10.1145/355616.36102</a>.
</div>
<div id="ref-qr_lasso" class="csl-entry" role="listitem">
A. Belloni and V. Chernozhukov. L1-penalized quantile regression in high-dimensional sparse models. <em>Ann. Statist.</em>, 39(1): 82–130, 2011. URL <a href="https://doi.org/10.1214/10-AOS827">https://doi.org/10.1214/10-AOS827</a>.
</div>
<div id="ref-scadAlg" class="csl-entry" role="listitem">
P. Breheny and J. Huang. Coordinate descent algorithms for nonconvex penalized regression, with applications to biological feature selection. <em>The Annals of Applied Statistics</em>, 5(1): 232–253, 2011. URL <a href="https://doi.org/10.1214/10-AOAS388">https://doi.org/10.1214/10-AOAS388</a>.
</div>
<div id="ref-brehenyglasso" class="csl-entry" role="listitem">
P. Breheny and J. Huang. Group descent algorithms for nonconvex penalized linear and logistic regression models with grouped predictors. <em>Stat. Comput.</em>, 25: 173–187, 2015. URL <a href="https://doi.org/10.1007/s11222-013-9424-2">https://doi.org/10.1007/s11222-013-9424-2</a>.
</div>
<div id="ref-penBiLevel" class="csl-entry" role="listitem">
P. Breheny and J. Huang. Penalized methods for bi-level variable selection. <em>Statistics and its Interface</em>, 2: 369–380, 2009. URL <a href="https://doi.org/10.4310/SII.2009.v2.n3.a10">https://doi.org/10.4310/SII.2009.v2.n3.a10</a>.
</div>
<div id="ref-de2011ames" class="csl-entry" role="listitem">
D. De Cock. Ames, <span>I</span>owa: Alternative to the boston housing data as an end of semester regression project. <em>J. Stat. Educ.</em>, 19(3): 2011. URL <a href="https://doi.org/10.1080/10691898.2011.11889627">https://doi.org/10.1080/10691898.2011.11889627</a>.
</div>
<div id="ref-fanLi" class="csl-entry" role="listitem">
J. Fan and R. Li. Variable selection via nonconcave penalized likelihood and its oracle properties. <em>J. Amer. Statist. Assoc.</em>, 96(456): 1348–1360, 2001. URL <a href="https://doi.org/10.1198/016214501753382273">https://doi.org/10.1198/016214501753382273</a>.
</div>
<div id="ref-friedman2009glmnet" class="csl-entry" role="listitem">
J. Friedman, T. Hastie and R. Tibshirani. Regularization paths for generalized linear models via coordinate descent. <em>J. Stat. Softw.</em>, 33(1): 1–22, 2010. URL <a href="https://www.jstatsoft.org/index.php/jss/article/view/v033i01">https://www.jstatsoft.org/index.php/jss/article/view/v033i01</a>.
</div>
<div id="ref-lowdConv" class="csl-entry" role="listitem">
X. He, X. Pan, K. M. Tan and W.-X. Zhou. Smoothed quantile regression with large-scale inference. <em>Journal of Econometrics</em>, 232(2): 367–388, 2023. URL <a href="https://www.sciencedirect.com/science/article/pii/S0304407621001950">https://www.sciencedirect.com/science/article/pii/S0304407621001950</a>.
</div>
<div id="ref-groupReview" class="csl-entry" role="listitem">
J. Huang, P. Breheny and S. Ma. A selective review of group selection in high-dimensional models. <em>Statistical Science</em>, 27(4): 481–499, 2012. URL <a href="https://doi.org/10.1214/12-STS392">https://doi.org/10.1214/12-STS392</a>.
</div>
<div id="ref-huber1964" class="csl-entry" role="listitem">
P. J. Huber. Robust estimation of a location parameter. <em>Ann. Math. Statist.</em>, 35(1): 73–101, 1964. URL <a href="https://doi.org/10.1214/aoms/1177703732">https://doi.org/10.1214/aoms/1177703732</a>.
</div>
<div id="ref-qr_group_lasso" class="csl-entry" role="listitem">
K. Kato. Group lasso for high dimensional sparse quantile regression models. 1–37, 2012. URL <a href="https://arxiv.org/pdf/1103.1458">https://arxiv.org/pdf/1103.1458</a>.
</div>
<div id="ref-crq1" class="csl-entry" role="listitem">
R. W. Koenker and V. D’Orey. Computing regression quantiles. <em>J. R. Stat. Soc. Ser. C. Appl. Stat.</em>, 36(3): 383–393, 1987. URL <a href="https://www.jstor.org/stable/2347802">https://www.jstor.org/stable/2347802</a>.
</div>
<div id="ref-origQR" class="csl-entry" role="listitem">
R. Koenker and G. Bassett. Regression quantiles. <em>Econometrica</em>, 46(1): 33–50, 1978. URL <a href="http://www.jstor.org/stable/1913643">http://www.jstor.org/stable/1913643</a> [online; last accessed May 20, 2025].
</div>
<div id="ref-crq2" class="csl-entry" role="listitem">
R. Koenker and V. D’Orey. A remark on algorithm AS 229: Computing dual regression quantiles and regression rank score. <em>J. R. Stat. Soc. Ser. C. Appl. Stat.</em>, 43(2): 410–414, 1994. URL <a href="http://www.jstor.org/stable/2986030">http://www.jstor.org/stable/2986030</a>.
</div>
<div id="ref-barro" class="csl-entry" role="listitem">
R. Koenker and J. A. F. Machado. Goodness of fit and related inference processes for quantile regression. <em>Journal of the American Statistical Association</em>, 94(448): 1296–1310, 1999. URL <a href="http://www.jstor.org/stable/2669943">http://www.jstor.org/stable/2669943</a>.
</div>
<div id="ref-JSSv060i05" class="csl-entry" role="listitem">
R. Koenker and I. Mizera. Convex optimization in <span>R</span>. <em>J. Stat. Softw.</em>, 60(5): 1–23, 2014. URL <a href="https://www.jstatsoft.org/index.php/jss/article/view/v060i05">https://www.jstatsoft.org/index.php/jss/article/view/v060i05</a>.
</div>
<div id="ref-qrbic" class="csl-entry" role="listitem">
E. R. Lee, H. Noh and B. U. Park. Model selection via bayesian information criterion for quantile regression models. <em>Journal of the American Statistical Association</em>, 109(505): 216–229, 2014. URL <a href="https://www.jstor.org/stable/24247149">https://www.jstor.org/stable/24247149</a>.
</div>
<div id="ref-gqlasso" class="csl-entry" role="listitem">
S. Li and B. Sherwood. Quantile predictions for equity premium using penalized quantile regression with consistent variable selection across multiple quantiles. 2025. URL <a href="https://arxiv.org/abs/2505.16019">https://arxiv.org/abs/2505.16019</a>.
</div>
<div id="ref-quantileCount" class="csl-entry" role="listitem">
J. A. F. Machado and J. M. C. S. Silva. Quantiles for counts. <em>Journal of the American Statistical Association</em>, 100(472): 1226–1237, 2005. URL <a href="https://doi.org/10.1198/016214505000000330">https://doi.org/10.1198/016214505000000330</a> [online; last accessed September 12, 2024].
</div>
<div id="ref-qr_cd" class="csl-entry" role="listitem">
B. Peng and L. Wang. An iterative coordinate descent algorithm for high-dimensional nonconvex penalized quantile regression. <em>J. Comput. Graph. Statist.</em>, 24(3): 676–694, 2015. URL <a href="https://doi.org/10.1080/10618600.2014.913516">https://doi.org/10.1080/10618600.2014.913516</a>.
</div>
<div id="ref-portnoy1997" class="csl-entry" role="listitem">
S. Portnoy and R. Koenker. The gaussian hare and the laplacian tortoise: Computability of squared-error versus absolute-error estimators. <em>Statist. Sci.</em>, 12(4): 279–300, 1997. URL <a href="https://doi.org/10.1214/ss/1030037960">https://doi.org/10.1214/ss/1030037960</a>.
</div>
<div id="ref-SherwoodLi2022" class="csl-entry" role="listitem">
B. Sherwood and S. Li. Quantile regression feature selection and estimation with grouped variables using huber approximation. <em>Statistics and Computing</em>, 32(5): 75, 2022. URL <a href="https://doi.org/10.1007/s11222-022-10135-w">https://doi.org/10.1007/s11222-022-10135-w</a>.
</div>
<div id="ref-mestatp" class="csl-entry" role="listitem">
B. Sherwood, A. J. Molstad and S. Singha. Asymptotic properties of concave L1-norm group penalties. <em>Statistics &amp; Probability Letters</em>, 157: 108631, 2020. URL <a href="https://www.sciencedirect.com/science/article/pii/S0167715219302779">https://www.sciencedirect.com/science/article/pii/S0167715219302779</a>.
</div>
<div id="ref-sparseGroup" class="csl-entry" role="listitem">
N. Simon, J. Friedman, T. Hastie and R. Tibshirani. A sparse-group lasso. <em>Journal of Computational and Graphical Statistics</em>, 22(2): 231–245, 2013. URL <a href="https://doi.org/10.1080/10618600.2012.681250">https://doi.org/10.1080/10618600.2012.681250</a>.
</div>
<div id="ref-effCenQR" class="csl-entry" role="listitem">
T. S. Sze Ming Lee and G. Xu. Efficient estimation for censored quantile regression. <em>Journal of the American Statistical Association</em>, 118(544): 2762–2775, 2023. URL <a href="https://doi.org/10.1080/01621459.2022.2078331">https://doi.org/10.1080/01621459.2022.2078331</a>.
</div>
<div id="ref-highdConv" class="csl-entry" role="listitem">
K. M. Tan, L. Wang and W.-X. Zhou. High-dimensional quantile regression: Convolution smoothing and concave regularization. <em>Journal of the Royal Statistical Society: Series B (Statistical Methodology)</em>, 84(1): 205–233, 2022. URL <a href="https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/rssb.12485">https://rss.onlinelibrary.wiley.com/doi/abs/10.1111/rssb.12485</a>.
</div>
<div id="ref-lasso" class="csl-entry" role="listitem">
R. Tibshirani. Regression shrinkage and selection via the lasso. <em>J. R. Stat. Soc. Ser. B. Stat. Methodol.</em>, 58(1): 267–288, 1996. URL <a href="http://www.jstor.org/stable/2346178">http://www.jstor.org/stable/2346178</a>.
</div>
<div id="ref-tibshirani2012strong" class="csl-entry" role="listitem">
R. Tibshirani et al. Strong rules for discarding predictors in lasso-type problems. <em>J. R. Stat. Soc. Ser. B. Stat. Methodol.</em>, 74(2): 245–266, 2012. URL <a href="https://doi.org/10.1111/j.1467-9868.2011.01004.x">https://doi.org/10.1111/j.1467-9868.2011.01004.x</a>.
</div>
<div id="ref-qrDist" class="csl-entry" role="listitem">
S. Volgushev, S.-K. Chao and G. Cheng. <span class="nocase">Distributed inference for quantile regression processes</span>. <em>The Annals of Statistics</em>, 47(3): 1634–1662, 2019. URL <a href="https://doi.org/10.1214/18-AOS1730">https://doi.org/10.1214/18-AOS1730</a>.
</div>
<div id="ref-grSCAD" class="csl-entry" role="listitem">
L. Wang, G. Chen and H. Li. Group SCAD regression analysis for microarray time course gene expression data. <em>Bioinformatics</em>, 23(12): 1486–1494, 2007. URL <a href="https://doi.org/10.1093/bioinformatics/btm125">https://doi.org/10.1093/bioinformatics/btm125</a>.
</div>
<div id="ref-lan_scad" class="csl-entry" role="listitem">
L. Wang, Y. Wu and R. Li. Quantile regression of analyzing heterogeneity in ultra-high dimension. <em>J. Amer. Statist. Assoc.</em>, 107(497): 214–222, 2012. URL <a href="https://doi.org/10.1080/01621459.2012.656014">https://doi.org/10.1080/01621459.2012.656014</a>.
</div>
<div id="ref-heteroIdQR" class="csl-entry" role="listitem">
M. Wang, X. Kang, J. Liang, K. Wang and Y. Wu. Heteroscedasticity identification and variable selection via multiple quantile regression. <em>Journal of Statistical Computation and Simulation</em>, 94(2): 297–314, 2024. URL <a href="https://doi.org/10.1080/00949655.2023.2243533">https://doi.org/10.1080/00949655.2023.2243533</a>.
</div>
<div id="ref-ciHypHDQR" class="csl-entry" role="listitem">
Y. Yan, X. Wang and R. Zhang. Confidence intervals and hypothesis testing for high-dimensional quantile regression: Convolution smoothing and debiasing. <em>Journal of Machine Learning Research</em>, 24(245): 1–49, 2023. URL <a href="http://jmlr.org/papers/v24/22-1217.html">http://jmlr.org/papers/v24/22-1217.html</a>.
</div>
<div id="ref-Yang2015" class="csl-entry" role="listitem">
Y. Yang and H. Zou. A fast unified algorithm for solving group-lasso penalize learning problems. <em>Stat. Comput.</em>, 25(6): 1129–1141, 2015. URL <a href="https://doi.org/10.1007/s11222-014-9498-5">https://doi.org/10.1007/s11222-014-9498-5</a>.
</div>
<div id="ref-huber_cd" class="csl-entry" role="listitem">
C. Yi and J. Huang. Semismooth newton coordinate descent algorithm for elastic-net penalized huber loss regression and quantile regression. <em>J. Comput. Graph. Statist.</em>, 26(3): 547–557, 2017. URL <a href="https://doi.org/10.1080/10618600.2016.1256816">https://doi.org/10.1080/10618600.2016.1256816</a>.
</div>
<div id="ref-yuan2007" class="csl-entry" role="listitem">
M. Yuan and Y. Lin. Model selection and estimation in regression with grouped variables. <em>J. R. Stat. Soc. Ser. B. Stat. Methodol.</em>, 68(1): 49–67, 2005. URL <a href="https://doi.org/10.1111/j.1467-9868.2005.00532.x">https://doi.org/10.1111/j.1467-9868.2005.00532.x</a>.
</div>
<div id="ref-mcp" class="csl-entry" role="listitem">
C.-H. Zhang. Nearly unbiased variable selection under minimax concave penalty. <em>Ann. Statist.</em>, 38(2): 894–942, 2010. URL <a href="https://doi.org/10.1214/09-AOS729">https://doi.org/10.1214/09-AOS729</a>.
</div>
<div id="ref-hdCensQR" class="csl-entry" role="listitem">
Q. Zheng, L. Peng and X. He. <span class="nocase">High dimensional censored quantile regression</span>. <em>The Annals of Statistics</em>, 46(1): 308–343, 2018. URL <a href="https://doi.org/10.1214/17-AOS1551">https://doi.org/10.1214/17-AOS1551</a>.
</div>
<div id="ref-adaptiveLasso" class="csl-entry" role="listitem">
H. Zou. The adaptive lasso and its oracle properties. <em>J. Amer. Statist. Assoc.</em>, 101(476): 1418–1429, 2006. URL <a href="https://doi.org/10.1198/016214506000000735">https://doi.org/10.1198/016214506000000735</a>.
</div>
<div id="ref-zou05" class="csl-entry" role="listitem">
H. Zou and T. Hastie. Regularization and variable selection via the elastic net. <em>J. R. Stat. Soc. Ser. B. Stat. Methodol.</em>, 67: 301–320, 2005. URL <a href="https://doi.org/10.1111/j.1467-9868.2005.00503.x">https://doi.org/10.1111/j.1467-9868.2005.00503.x</a>.
</div>
<div id="ref-lla" class="csl-entry" role="listitem">
H. Zou and R. Li. One-step sparse estimates in nonconcave penalized likelihood models. <em>Ann. Statist.</em>, 36(4): 1509–1533, 2008. URL <a href="https://doi.org/10.1214/009053607000000802">https://doi.org/10.1214/009053607000000802</a>.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
  <h3 id="references">References</h3>
  <div id="references-listing"></div>
  <h3 id="reuse">Reuse</h3>
  <p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
  <h3 id="citation">Citation</h3>
  <p>For attribution, please cite this work as</p>
  <pre class="citation-appendix short">Sherwood, et al., "rqPen: An R Package for Penalized Quantile Regression", The R Journal, 2025</pre>
  <p>BibTeX citation</p>
  <pre class="citation-appendix long">@article{RJ-2025-017,
  author = {Sherwood, Ben and Li, Shaobo and Maidman, Adam},
  title = {rqPen: An R Package for Penalized Quantile Regression},
  journal = {The R Journal},
  year = {2025},
  note = {https://doi.org/10.32614/RJ-2025-017},
  doi = {10.32614/RJ-2025-017},
  volume = {17},
  issue = {2},
  issn = {2073-4859},
  pages = {146-175}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<!--radix_placeholder_navigation_after_body--><html><body>
<div class="distill-site-nav distill-site-footer">
<p>© The R Foundation, <a href="mailto:r-journal@r-project.org">web page
contact</a>.</p>
</div>
<!--/radix_placeholder_navigation_after_body-->
</body></html>


</body>

</html>
