<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
      margin-bottom: 0em;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
<title>LUCIDus: An R Package For Implementing Latent Unknown Clustering By Integrating Multi-omics Data (LUCID) With Phenotypic Traits</title>

<meta property="description" itemprop="description" content="Many studies are leveraging current technologies to obtain multiple&#10;omics measurements on the same individuals. These measurements are&#10;usually cross-sectional, and methods developed and commonly used focus&#10;on omic integration at a single time point. More unique, and a growing&#10;area of interest, are studies that leverage biology or the temporal&#10;sequence of measurements to relate long-term exposures or germline&#10;genetics to intermediate measures capturing transitional processes&#10;that ultimately result in an outcome. In this context, we have&#10;previously introduced an integrative model named Latent Unknown&#10;Clustering by Integrating multi-omics Data (LUCID) aiming to&#10;distinguish unique effects of environmental exposures or germline&#10;genetics and informative omic effects while jointly estimating&#10;subgroups of individuals relevant to the outcome of interest. This&#10;multiple omics analysis consists of early integration (concatenation&#10;of omic layers to estimate common subgroups); intermediate integration&#10;(omic-layer-specific estimation of subgroups that are all related to&#10;the outcome); and late integration (omic-layer-specific estimation of&#10;subgroups that are then interrelated by a priori structures). In this&#10;article, we introduce LUCIDus version 3, an R package to implement the&#10;LUCID model. We review the statistical background of the model and&#10;introduce the workflow of LUCIDus, including model fitting, model&#10;selection, interpretation, inference, and prediction. Throughout, we&#10;use a realistic but simulated dataset based on an ongoing study, the&#10;Human Early Life Exposome Study (HELIX), to illustrate the workflow."/>

<link rel="canonical" href="https://doi.org/10.32614/RJ-2024-012/"/>
<link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>
<link rel="icon" type="image/vnd.microsoft.icon" href="../../resources/favicon.ico"/>

<!--  https://schema.org/Article -->
<meta property="article:published" itemprop="datePublished" content="2025-03-11"/>
<meta property="article:created" itemprop="dateCreated" content="2025-03-11"/>
<meta name="article:author" content="Yinqi Zhao"/>
<meta name="article:author" content="Qiran Jia"/>
<meta name="article:author" content="Jesse A. Goodrich"/>
<meta name="article:author" content="David V. Conti"/>

<!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
<meta property="og:title" content="LUCIDus: An R Package For Implementing Latent Unknown Clustering By Integrating Multi-omics Data (LUCID) With Phenotypic Traits"/>
<meta property="og:type" content="article"/>
<meta property="og:description" content="Many studies are leveraging current technologies to obtain multiple&#10;omics measurements on the same individuals. These measurements are&#10;usually cross-sectional, and methods developed and commonly used focus&#10;on omic integration at a single time point. More unique, and a growing&#10;area of interest, are studies that leverage biology or the temporal&#10;sequence of measurements to relate long-term exposures or germline&#10;genetics to intermediate measures capturing transitional processes&#10;that ultimately result in an outcome. In this context, we have&#10;previously introduced an integrative model named Latent Unknown&#10;Clustering by Integrating multi-omics Data (LUCID) aiming to&#10;distinguish unique effects of environmental exposures or germline&#10;genetics and informative omic effects while jointly estimating&#10;subgroups of individuals relevant to the outcome of interest. This&#10;multiple omics analysis consists of early integration (concatenation&#10;of omic layers to estimate common subgroups); intermediate integration&#10;(omic-layer-specific estimation of subgroups that are all related to&#10;the outcome); and late integration (omic-layer-specific estimation of&#10;subgroups that are then interrelated by a priori structures). In this&#10;article, we introduce LUCIDus version 3, an R package to implement the&#10;LUCID model. We review the statistical background of the model and&#10;introduce the workflow of LUCIDus, including model fitting, model&#10;selection, interpretation, inference, and prediction. Throughout, we&#10;use a realistic but simulated dataset based on an ongoing study, the&#10;Human Early Life Exposome Study (HELIX), to illustrate the workflow."/>
<meta property="og:url" content="https://doi.org/10.32614/RJ-2024-012/"/>
<meta property="og:image" content="https://journal.r-project.org/articles/RJ-2024-012/figures/fig1.png"/>
<meta property="og:image:width" content="3115"/>
<meta property="og:image:height" content="1763"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="The R Journal"/>

<!--  https://dev.twitter.com/cards/types/summary -->
<meta property="twitter:card" content="summary_large_image"/>
<meta property="twitter:title" content="LUCIDus: An R Package For Implementing Latent Unknown Clustering By Integrating Multi-omics Data (LUCID) With Phenotypic Traits"/>
<meta property="twitter:description" content="Many studies are leveraging current technologies to obtain multiple&#10;omics measurements on the same individuals. These measurements are&#10;usually cross-sectional, and methods developed and commonly used focus&#10;on omic integration at a single time point. More unique, and a growing&#10;area of interest, are studies that leverage biology or the temporal&#10;sequence of measurements to relate long-term exposures or germline&#10;genetics to intermediate measures capturing transitional processes&#10;that ultimately result in an outcome. In this context, we have&#10;previously introduced an integrative model named Latent Unknown&#10;Clustering by Integrating multi-omics Data (LUCID) aiming to&#10;distinguish unique effects of environmental exposures or germline&#10;genetics and informative omic effects while jointly estimating&#10;subgroups of individuals relevant to the outcome of interest. This&#10;multiple omics analysis consists of early integration (concatenation&#10;of omic layers to estimate common subgroups); intermediate integration&#10;(omic-layer-specific estimation of subgroups that are all related to&#10;the outcome); and late integration (omic-layer-specific estimation of&#10;subgroups that are then interrelated by a priori structures). In this&#10;article, we introduce LUCIDus version 3, an R package to implement the&#10;LUCID model. We review the statistical background of the model and&#10;introduce the workflow of LUCIDus, including model fitting, model&#10;selection, interpretation, inference, and prediction. Throughout, we&#10;use a realistic but simulated dataset based on an ongoing study, the&#10;Human Early Life Exposome Study (HELIX), to illustrate the workflow."/>
<meta property="twitter:url" content="https://doi.org/10.32614/RJ-2024-012/"/>
<meta property="twitter:image" content="https://journal.r-project.org/articles/RJ-2024-012/figures/fig1.png"/>
<meta property="twitter:image:width" content="3115"/>
<meta property="twitter:image:height" content="1763"/>

<!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
<meta name="citation_title" content="LUCIDus: An R Package For Implementing Latent Unknown Clustering By Integrating Multi-omics Data (LUCID) With Phenotypic Traits"/>
<meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2024-012"/>
<meta name="citation_pdf_url" content="RJ-2024-012.pdf"/>
<meta name="citation_volume" content="16"/>
<meta name="citation_issue" content="2"/>
<meta name="citation_doi" content="10.32614/RJ-2024-012"/>
<meta name="citation_journal_title" content="The R Journal"/>
<meta name="citation_issn" content="2073-4859"/>
<meta name="citation_firstpage" content="4"/>
<meta name="citation_lastpage" content="26"/>
<meta name="citation_fulltext_world_readable" content=""/>
<meta name="citation_online_date" content="2025/03/11"/>
<meta name="citation_publication_date" content="2025/03/11"/>
<meta name="citation_author" content="Yinqi Zhao"/>
<meta name="citation_author_institution" content="Division of Biostatistics, Department of Population and Public Health&#10;Sciences, University of Southern California"/>
<meta name="citation_author" content="Qiran Jia"/>
<meta name="citation_author_institution" content="Division of Biostatistics, Department of Population and Public Health&#10;Sciences, University of Southern California"/>
<meta name="citation_author" content="Jesse A. Goodrich"/>
<meta name="citation_author_institution" content="Division of Environmental Epidemiology, Department of Population and&#10;Public Health Sciences, University of Southern California"/>
<meta name="citation_author" content="David V. Conti"/>
<meta name="citation_author_institution" content="Division of Biostatistics, Department of Population and Public Health&#10;Sciences, University of Southern California"/>
<!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Coming of age: Ten years of next-generation sequencing technologies;citation_publisher=Nature Publishing Group;citation_volume=17;citation_author=Sara Goodwin;citation_author=John D McPherson;citation_author=W Richard McCombie"/>
  <meta name="citation_reference" content="citation_title=RNA sequencing: Advances, challenges and opportunities;citation_publisher=Nature Publishing Group;citation_volume=12;citation_author=Fatih Ozsolak;citation_author=Patrice M Milos"/>
  <meta name="citation_reference" content="citation_title=A review of applications of metabolomics in cancer;citation_publisher=Multidisciplinary Digital Publishing Institute;citation_volume=3;citation_author=Richard D Beger"/>
  <meta name="citation_reference" content="citation_title=Proteomics: Technologies and their applications;citation_publisher=Oxford University Press;citation_volume=55;citation_author=Bilal Aslam;citation_author=Madiha Basit;citation_author=Muhammad Atif Nisar;citation_author=Mohsin Khurshid;citation_author=Muhammad Hidayat Rasool"/>
  <meta name="citation_reference" content="citation_title=Complementing the genome with an “exposome”: The outstanding challenge of environmental exposure measurement in molecular epidemiology;citation_publisher=AACR;citation_volume=14;citation_author=Christopher Paul Wild"/>
  <meta name="citation_reference" content="citation_title=The human early-life exposome (HELIX): Project rationale and design;citation_publisher=NLM-Export;citation_volume=122;citation_author=Martine Vrijheid;citation_author=Rémy Slama;citation_author=Oliver Robinson;citation_author=Leda Chatzi;citation_author=Muireann Coen;citation_author=Peter Van den Hazel;citation_author=Cathrine Thomsen;citation_author=John Wright;citation_author=Toby J Athersuch;citation_author=Narcis Avellana;citation_author=The human early-life exposome (HELIX): Project rationale and design"/>
  <meta name="citation_reference" content="citation_title=Multi-omics data integration, interpretation, and its application;citation_publisher=SAGE Publications Sage UK: London, England;citation_volume=14;citation_author=Indhupriya Subramanian;citation_author=Srikant Verma;citation_author=Shiva Kumar;citation_author=Abhay Jere;citation_author=Krishanpal Anamika"/>
  <meta name="citation_reference" content="citation_title=Multi-omics integration—a comparison of unsupervised clustering methodologies;citation_publisher=Oxford University Press;citation_volume=20;citation_author=Giulia Tini;citation_author=Luca Marchetti;citation_author=Corrado Priami;citation_author=Marie-Pier Scott-Boyer"/>
  <meta name="citation_reference" content="citation_title=Omic association studies with r and bioconductor;citation_publisher=CRC Press;citation_author=Juan R González;citation_author=Alejandro Cáceres"/>
  <meta name="citation_reference" content="citation_title=Multi-omic and multi-view clustering algorithms: Review and cancer benchmark;citation_publisher=Oxford University Press;citation_volume=46;citation_author=Nimrod Rappoport;citation_author=Ron Shamir"/>
  <meta name="citation_reference" content="citation_title=Classification and diagnostic prediction of cancers using gene expression profiling and artificial neural networks;citation_publisher=Nature Publishing Group;citation_volume=7;citation_author=Javed Khan;citation_author=Jun S Wei;citation_author=Markus Ringner;citation_author=Lao H Saal;citation_author=Marc Ladanyi;citation_author=Frank Westermann;citation_author=Frank Berthold;citation_author=Manfred Schwab;citation_author=Cristina R Antonescu;citation_author=Carsten Peterson;citation_author=Classification and diagnostic prediction of cancers using gene expression profiling and artificial neural networks"/>
  <meta name="citation_reference" content="citation_title=Variable selection for generalized canonical correlation analysis;citation_publisher=Oxford University Press;citation_volume=15;citation_author=Arthur Tenenhaus;citation_author=Cathy Philippe;citation_author=Vincent Guillemot;citation_author=Kim-Anh Le Cao;citation_author=Jacques Grill;citation_author=Vincent Frouin"/>
  <meta name="citation_reference" content="citation_title=Similarity network fusion for aggregating data types on a genomic scale;citation_publisher=Nature Publishing Group;citation_volume=11;citation_author=Bo Wang;citation_author=Aziz M Mezlini;citation_author=Feyyaz Demir;citation_author=Marc Fiume;citation_author=Zhuowen Tu;citation_author=Michael Brudno;citation_author=Benjamin Haibe-Kains;citation_author=Anna Goldenberg"/>
  <meta name="citation_reference" content="citation_title=Pattern discovery and cancer gene identification in integrated cancer genomic data;citation_publisher=National Acad Sciences;citation_volume=110;citation_author=Qianxing Mo;citation_author=Sijian Wang;citation_author=Venkatraman E Seshan;citation_author=Adam B Olshen;citation_author=Nikolaus Schultz;citation_author=Chris Sander;citation_author=R Scott Powers;citation_author=Marc Ladanyi;citation_author=Ronglai Shen"/>
  <meta name="citation_reference" content="citation_title=Clustering and variable selection evaluation of 13 unsupervised methods for multi-omics data integration;citation_publisher=Oxford University Press;citation_volume=21;citation_author=Morgane Pierre-Jean;citation_author=Jean-François Deleuze;citation_author=Edith Le Floch;citation_author=Florence Mauger"/>
  <meta name="citation_reference" content="citation_title=Regression selection and shrinkage via the lasso;citation_volume=58;citation_author=Robert Tibshirani"/>
  <meta name="citation_reference" content="citation_title=Model selection and estimation in regression with grouped variables;citation_publisher=Wiley Online Library;citation_volume=68;citation_author=Ming Yuan;citation_author=Yi Lin"/>
  <meta name="citation_reference" content="citation_title=A latent unknown clustering integrating multi-omics data (LUCID) with phenotypic traits;citation_publisher=Oxford University Press;citation_volume=36;citation_author=Cheng Peng;citation_author=Jun Wang;citation_author=Isaac Asante;citation_author=Stan Louie;citation_author=Ran Jin;citation_author=Lida Chatzi;citation_author=Graham Casey;citation_author=Duncan C Thomas;citation_author=David V Conti"/>
  <meta name="citation_reference" content="citation_title=Dlstats: Download stats of r packages;citation_author=Guangchuang Yu"/>
  <meta name="citation_reference" content="citation_title=Perfluoroalkyl substances and severity of nonalcoholic fatty liver in children: An untargeted metabolomics approach;citation_publisher=Elsevier;citation_volume=134;citation_author=Ran Jin;citation_author=Rob McConnell;citation_author=Cioffi Catherine;citation_author=Shujing Xu;citation_author=Douglas I Walker;citation_author=Nikos Stratakis;citation_author=Dean P Jones;citation_author=Gary W Miller;citation_author=Cheng Peng;citation_author=David V. Conti;citation_author=Perfluoroalkyl substances and severity of nonalcoholic fatty liver in children: An untargeted metabolomics approach"/>
  <meta name="citation_reference" content="citation_title=Prenatal exposure to perfluoroalkyl substances associated with increased susceptibility to liver injury in children;citation_publisher=Wiley Online Library;citation_volume=72;citation_author=Nikos Stratakis;citation_author=David V. Conti;citation_author=Ran Jin;citation_author=Katerina Margetaki;citation_author=Damaskini Valvi;citation_author=Alexandros P Siskos;citation_author=Léa Maitre;citation_author=Erika Garcia;citation_author=Nerea Varo;citation_author=Yinqi Zhao;citation_author=Prenatal exposure to perfluoroalkyl substances associated with increased susceptibility to liver injury in children"/>
  <meta name="citation_reference" content="citation_title=Associations between persistent organic pollutants and endometriosis: A multiblock approach integrating metabolic and cytokine profiling;citation_publisher=Elsevier;citation_volume=158;citation_author=Komodo Matta;citation_author=Tiphaine Lefebvre;citation_author=Evelyne Vigneau;citation_author=Véronique Cariou;citation_author=Philippe Marchand;citation_author=Yann Guitton;citation_author=Anne-Lise Royer;citation_author=Stéphane Ploteau;citation_author=Bruno Le Bizec;citation_author=Jean-Philippe Antignac;citation_author=Associations between persistent organic pollutants and endometriosis: A multiblock approach integrating metabolic and cytokine profiling"/>
  <meta name="citation_reference" content="citation_title=State-of-the-art methods for exposure-health studies: Results from the exposome data challenge event;citation_author=Léa Maitre;citation_author=Jean-Baptiste Guimbaud;citation_author=Charline Warembourg;citation_author=Nuria Güil-Oumrait;citation_author=The Exposome Data Challenge Participant Consortium;citation_author=Paula Marcela Petrone;citation_author=Marc Chadeau-Hyam;citation_author=Martine Vrijheid;citation_author=Juan R Gonzalez;citation_author=Xavier Basagaña"/>
  <meta name="citation_reference" content="citation_title=Model-based gaussian and non-gaussian clustering;citation_publisher=JSTOR;citation_author=Jeffrey D Banfield;citation_author=Adrian E Raftery"/>
  <meta name="citation_reference" content="citation_title=Gaussian parsimonious clustering models;citation_publisher=Elsevier;citation_volume=28;citation_author=Gilles Celeux;citation_author=Gérard Govaert"/>
  <meta name="citation_reference" content="citation_title=Mclust 5: Clustering, classification and density estimation using gaussian finite mixture models;citation_publisher=NIH Public Access;citation_volume=8;citation_author=Luca Scrucca;citation_author=Michael Fop;citation_author=T Brendan Murphy;citation_author=Adrian E Raftery"/>
  <meta name="citation_reference" content="citation_title=The sankey diagram in energy and material flow management: Part II: Methodology and current applications;citation_publisher=Wiley Online Library;citation_volume=12;citation_author=Mario Schmidt"/>
  <meta name="citation_reference" content="citation_title=Tuning parameter selection in high dimensional penalized likelihood;citation_publisher=Wiley Online Library;citation_volume=75;citation_author=Yingying Fan;citation_author=Cheng Yong Tang"/>
  <meta name="citation_reference" content="citation_title=Variable selection methods for model-based clustering;citation_publisher=Amer. Statist. Assoc., the Bernoulli Soc., the Inst. Math. Statist.,; the …;citation_volume=12;citation_author=Michael Fop;citation_author=Thomas Brendan Murphy"/>
  <meta name="citation_reference" content="citation_title=Penalized model-based clustering with unconstrained covariance matrices;citation_publisher=NIH Public Access;citation_volume=3;citation_author=Hui Zhou;citation_author=Wei Pan;citation_author=Xiaotong Shen"/>
  <meta name="citation_reference" content="citation_title=Penalized model-based clustering with application to variable selection;citation_volume=8;citation_author=Wei Pan;citation_author=Xiaotong Shen"/>
  <meta name="citation_reference" content="citation_title=Bootstrap methods and their application;citation_publisher=Cambridge University Press;citation_author=Anthony Christopher Davison;citation_author=David Victor Hinkley"/>
  <meta name="citation_reference" content="citation_title=Boot: Bootstrap functions (originally by angelo canty for s);citation_author=Angelo Canty;citation_author=Brian Ripley"/>
  <meta name="citation_reference" content="citation_title=Gaussian mixture model clustering with incomplete data;citation_publisher=ACM New York, NY;citation_volume=17;citation_author=Yi Zhang;citation_author=Miaomiao Li;citation_author=Siwei Wang;citation_author=Sisi Dai;citation_author=Lei Luo;citation_author=En Zhu;citation_author=Huiying Xu;citation_author=Xinzhong Zhu;citation_author=Chaoyun Yao;citation_author=Haoran Zhou"/>
  <meta name="citation_reference" content="citation_title=Handling missing rows in multi-omics data integration: Multiple imputation in multiple factor analysis framework;citation_publisher=BioMed Central;citation_volume=17;citation_author=Valentin Voillet;citation_author=Philippe Besse;citation_author=Laurence Liaubet;citation_author=Magali San Cristobal;citation_author=Ignacio González"/>
  <meta name="citation_reference" content="citation_title=Performance of approaches relying on multidimensional intermediary data to decipher causal relationships between the exposome and health: A simulation study under various causal structures;citation_volume=153;citation_doi=https://doi.org/10.1016/j.envint.2021.106509;citation_issn=0160-4120;citation_author=Solène Cadiou;citation_author=Xavier Basagaña;citation_author=Juan R. Gonzalez;citation_author=Johanna Lepeule;citation_author=Martine Vrijheid;citation_author=Valérie Siroux;citation_author=Rémy Slama"/>
  <meta name="citation_reference" content="citation_title=Joint and individual variation explained (JIVE) for integrated analysis of multiple data types;citation_publisher=NIH Public Access;citation_volume=7;citation_author=Eric F Lock;citation_author=Katherine A Hoadley;citation_author=James Stephen Marron;citation_author=Andrew B Nobel"/>
  <meta name="citation_reference" content="citation_title=Estimating and testing high-dimensional mediation effects in epigenetic studies;citation_publisher=Oxford University Press;citation_volume=32;citation_author=Haixiang Zhang;citation_author=Yinan Zheng;citation_author=Zhou Zhang;citation_author=Tao Gao;citation_author=Brian Joyce;citation_author=Grace Yoon;citation_author=Wei Zhang;citation_author=Joel Schwartz;citation_author=Allan Just;citation_author=Elena Colicino;citation_author=Estimating and testing high-dimensional mediation effects in epigenetic studies"/>
  <meta name="citation_reference" content="citation_title=Bayesian shrinkage estimation of high dimensional causal mediation effects in omics studies;citation_publisher=Wiley Online Library;citation_volume=76;citation_author=Yanyi Song;citation_author=Xiang Zhou;citation_author=Min Zhang;citation_author=Wei Zhao;citation_author=Yongmei Liu;citation_author=Sharon LR Kardia;citation_author=Ana V Diez Roux;citation_author=Belinda L Needham;citation_author=Jennifer A Smith;citation_author=Bhramar Mukherjee"/>
  <meta name="citation_reference" content="citation_title=Causal mediation analysis with a latent mediator;citation_publisher=Wiley Online Library;citation_volume=58;citation_author=Jeffrey M Albert;citation_author=Cuiyu Geng;citation_author=Suchitra Nelson"/>
  <meta name="citation_reference" content="citation_title=High dimensional mediation analysis with latent variables;citation_publisher=Wiley Online Library;citation_volume=75;citation_author=Andriy Derkach;citation_author=Ruth M Pfeiffer;citation_author=Ting-Huei Chen;citation_author=Joshua N Sampson"/>
  <meta name="citation_reference" content="citation_title=Bayesian correlated clustering to integrate multiple datasets;citation_publisher=Oxford University Press;citation_volume=28;citation_author=Paul Kirk;citation_author=Jim E Griffin;citation_author=Richard S Savage;citation_author=Zoubin Ghahramani;citation_author=David L Wild"/>
  <meta name="citation_reference" content="citation_title=Bayesian consensus clustering;citation_publisher=Oxford University Press;citation_volume=29;citation_author=Eric F Lock;citation_author=David B Dunson"/>
  <meta name="citation_reference" content="citation_title=Time-course gene set analysis for longitudinal gene expression data;citation_publisher=Public Library of Science San Francisco, CA USA;citation_volume=11;citation_author=Boris P Hejblum;citation_author=Jason Skinner;citation_author=Rodolphe Thiébaut"/>
  <meta name="citation_reference" content="citation_title=The genomic and transcriptomic architecture of 2,000 breast tumours reveals novel subgroups;citation_publisher=Nature Publishing Group UK London;citation_volume=486;citation_author=Christina Curtis;citation_author=Sohrab P Shah;citation_author=Suet-Feung Chin;citation_author=Gulisa Turashvili;citation_author=Oscar M Rueda;citation_author=Mark J Dunning;citation_author=Doug Speed;citation_author=Andy G Lynch;citation_author=Shamith Samarajiwa;citation_author=Yinyin Yuan;citation_author=The genomic and transcriptomic architecture of 2,000 breast tumours reveals novel subgroups"/>
  <meta name="citation_reference" content="citation_title=Statistical mixture modeling for cell subtype identification in flow cytometry;citation_publisher=Wiley Online Library;citation_volume=73;citation_author=Cliburn Chan;citation_author=Feng Feng;citation_author=Janet Ottinger;citation_author=David Foster;citation_author=Mike West;citation_author=Thomas B Kepler"/>
  <meta name="citation_reference" content="citation_title=Sequential dirichlet process mixtures of multivariate skew t-distributions for model-based clustering of flow cytometry data;citation_author=Boris P Hejblum;citation_author=Chariff Alkhassim;citation_author=Raphael Gottardo;citation_author=François Caron;citation_author=Rodolphe Thiébaut"/>
  <meta name="citation_reference" content="citation_title=Dirichlet process mixture model for correcting technical variation in single-cell gene expression data;citation_publisher=PMLR;citation_author=Sandhya Prabhakaran;citation_author=Elham Azizi;citation_author=Ambrose Carr;citation_author=Dana Pe’er"/>
  <meta name="citation_reference" content="citation_title=A bayesian mixture modelling approach for spatial proteomics;citation_publisher=Public Library of Science San Francisco, CA USA;citation_volume=14;citation_author=Oliver M Crook;citation_author=Claire M Mulvey;citation_author=Paul DW Kirk;citation_author=Kathryn S Lilley;citation_author=Laurent Gatto"/>
  <meta name="citation_reference" content="citation_title=DIABLO: An integrative approach for identifying key molecular drivers from multi-omics assays;citation_publisher=Oxford University Press;citation_volume=35;citation_author=Amrit Singh;citation_author=Casey P Shannon;citation_author=Benoı̂t Gautier;citation_author=Florian Rohart;citation_author=Michaël Vacher;citation_author=Scott J Tebbutt;citation_author=Kim-Anh Lê Cao"/>
  <meta name="citation_reference" content="citation_title=Visdat: Visualising whole data frames including missing values;citation_author=Nicholas Tierney"/>
  <meta name="citation_reference" content="citation_title=LUCIDus: LUCID with multiple omics data;citation_publisher=Comprehensive R Archive Network (CRAN);citation_author=Qiran Jia;citation_author=Yinqi Zhao;citation_author=David Conti;citation_author=Jesse Goodrich"/>
  <!--radix_placeholder_rmarkdown_metadata-->

<script type="text/json" id="radix-rmarkdown-metadata">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","date_received","journal","volume","issue","slug","citation_url","packages","preview","bibliography","CTV","legacy_pdf","legacy_converted","output","draft","pdf_url","doi","creative_commons","csl","canonical_url"]}},"value":[{"type":"character","attributes":{},"value":["LUCIDus: An R Package For Implementing Latent Unknown Clustering By Integrating Multi-omics Data (LUCID) With Phenotypic Traits"]},{"type":"character","attributes":{},"value":["Many studies are leveraging current technologies to obtain multiple\nomics measurements on the same individuals. These measurements are\nusually cross-sectional, and methods developed and commonly used focus\non omic integration at a single time point. More unique, and a growing\narea of interest, are studies that leverage biology or the temporal\nsequence of measurements to relate long-term exposures or germline\ngenetics to intermediate measures capturing transitional processes\nthat ultimately result in an outcome. In this context, we have\npreviously introduced an integrative model named Latent Unknown\nClustering by Integrating multi-omics Data (LUCID) aiming to\ndistinguish unique effects of environmental exposures or germline\ngenetics and informative omic effects while jointly estimating\nsubgroups of individuals relevant to the outcome of interest. This\nmultiple omics analysis consists of early integration (concatenation\nof omic layers to estimate common subgroups); intermediate integration\n(omic-layer-specific estimation of subgroups that are all related to\nthe outcome); and late integration (omic-layer-specific estimation of\nsubgroups that are then interrelated by a priori structures). In this\narticle, we introduce LUCIDus version 3, an R package to implement the\nLUCID model. We review the statistical background of the model and\nintroduce the workflow of LUCIDus, including model fitting, model\nselection, interpretation, inference, and prediction. Throughout, we\nuse a realistic but simulated dataset based on an ongoing study, the\nHuman Early Life Exposome Study (HELIX), to illustrate the workflow."]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","orcid_id","address"]}},"value":[{"type":"character","attributes":{},"value":["Yinqi Zhao"]},{"type":"character","attributes":{},"value":["Division of Biostatistics, Department of Population and Public Health\nSciences, University of Southern California"]},{"type":"character","attributes":{},"value":["0000-0003-2413-732X"]},{"type":"character","attributes":{},"value":["2001 N Soto St., Los Angeles, CA, 90032","United States","[yinqiz@usc.edu](yinqiz@usc.edu){.uri}\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","orcid_id","address"]}},"value":[{"type":"character","attributes":{},"value":["Qiran Jia"]},{"type":"character","attributes":{},"value":["Division of Biostatistics, Department of Population and Public Health\nSciences, University of Southern California"]},{"type":"character","attributes":{},"value":["0000-0002-0790-5967"]},{"type":"character","attributes":{},"value":["2001 N Soto St., Los Angeles, CA, 90032","United States","[qiranjia@usc.edu](qiranjia@usc.edu){.uri}\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Jesse A. Goodrich"]},{"type":"character","attributes":{},"value":["Division of Environmental Epidemiology, Department of Population and\nPublic Health Sciences, University of Southern California"]},{"type":"character","attributes":{},"value":["2001 N Soto St., Los Angeles, CA, 90032","United States","ORCid: 0000-0001-6615-0472","[jagoodri@usc.edu](jagoodri@usc.edu){.uri}\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["David V. Conti"]},{"type":"character","attributes":{},"value":["Division of Biostatistics, Department of Population and Public Health\nSciences, University of Southern California"]},{"type":"character","attributes":{},"value":["2001 N Soto St., Los Angeles, CA, 90032","United States","ORCid: 0000-0002-2941-7833","[dconti@usc.med.edu](dconti@usc.med.edu){.uri}\n"]}]}]},{"type":"character","attributes":{},"value":["2025-03-11"]},{"type":"character","attributes":{},"value":["2022-09-13"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"integer","attributes":{},"value":[4]},{"type":"integer","attributes":{},"value":[26]}]},{"type":"integer","attributes":{},"value":[16]},{"type":"integer","attributes":{},"value":[2]},{"type":"character","attributes":{},"value":["RJ-2024-012"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2024-012"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"character","attributes":{},"value":["LUCIDus","dlstats","mclust","boot","visdat"]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["preview.png"]},{"type":"character","attributes":{},"value":["zhao.bib"]},{"type":"character","attributes":{},"value":["Cluster","Distributions","Econometrics","Environmetrics","MissingData","Omics","Optimization","Survival","TimeSeries"]},{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[true]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc","mathjax","md_extension"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js"]},{"type":"character","attributes":{},"value":["-tex_math_single_backslash"]}]}]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["RJ-2024-012.pdf"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2024-012"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"character","attributes":{},"value":["/home/mitchell/R/x86_64-pc-linux-gnu-library/4.4/rjtools/rjournal.csl"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2024-012/"]}]}
</script>
<!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["figures/fig1.pdf","figures/fig1.png","figures/fig2.pdf","figures/fig2.png","figures/fig3.pdf","figures/fig3.png","figures/fig4.pdf","figures/fig4.png","figures/fig5.pdf","figures/fig5.png","RJ-2024-012_files/anchor-4.2.2/anchor.min.js","RJ-2024-012_files/bowser-1.9.3/bowser.min.js","RJ-2024-012_files/distill-2.2.21/template.v2.js","RJ-2024-012_files/header-attrs-2.29/header-attrs.js","RJ-2024-012_files/jquery-3.6.0/jquery-3.6.0.js","RJ-2024-012_files/jquery-3.6.0/jquery-3.6.0.min.js","RJ-2024-012_files/jquery-3.6.0/jquery-3.6.0.min.map","RJ-2024-012_files/popper-2.6.0/popper.min.js","RJ-2024-012_files/tippy-6.2.7/tippy-bundle.umd.min.js","RJ-2024-012_files/tippy-6.2.7/tippy-light-border.css","RJ-2024-012_files/tippy-6.2.7/tippy.css","RJ-2024-012_files/tippy-6.2.7/tippy.umd.min.js","RJ-2024-012_files/webcomponents-2.0.0/webcomponents.js","RJ-2024-012.log","RJ-2024-012.tex","RJ-2024-012.zip","RJournal.sty","RJwrapper.log","RJwrapper.tex","table_data_1.csv","zhao.bib","zhao.tex"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
<meta name="distill:offset" content="../.."/>

<script type="application/javascript">

  window.headroom_prevent_pin = false;

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // initialize headroom for banner
    var header = $('header').get(0);
    var headerHeight = header.offsetHeight;
    var headroom = new Headroom(header, {
      tolerance: 5,
      onPin : function() {
        if (window.headroom_prevent_pin) {
          window.headroom_prevent_pin = false;
          headroom.unpin();
        }
      }
    });
    headroom.init();
    if(window.location.hash)
      headroom.unpin();
    $(header).addClass('headroom--transition');

    // offset scroll location for banner on hash change
    // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
    window.addEventListener("hashchange", function(event) {
      window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
    });

    // responsive menu
    $('.distill-site-header').each(function(i, val) {
      var topnav = $(this);
      var toggle = topnav.find('.nav-toggle');
      toggle.on('click', function() {
        topnav.toggleClass('responsive');
      });
    });

    // nav dropdowns
    $('.nav-dropbtn').click(function(e) {
      $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
      $(this).parent().siblings('.nav-dropdown')
         .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $("body").click(function(e){
      $('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $(".nav-dropdown").click(function(e){
      e.stopPropagation();
    });
  });
</script>

<style type="text/css">

/* Theme (user-documented overrideables for nav appearance) */

.distill-site-nav {
  color: rgba(255, 255, 255, 0.8);
  background-color: #0F2E3D;
  font-size: 15px;
  font-weight: 300;
}

.distill-site-nav a {
  color: inherit;
  text-decoration: none;
}

.distill-site-nav a:hover {
  color: white;
}

@media print {
  .distill-site-nav {
    display: none;
  }
}

.distill-site-header {

}

.distill-site-footer {

}


/* Site Header */

.distill-site-header {
  width: 100%;
  box-sizing: border-box;
  z-index: 3;
}

.distill-site-header .nav-left {
  display: inline-block;
  margin-left: 8px;
}

@media screen and (max-width: 768px) {
  .distill-site-header .nav-left {
    margin-left: 0;
  }
}


.distill-site-header .nav-right {
  float: right;
  margin-right: 8px;
}

.distill-site-header a,
.distill-site-header .title {
  display: inline-block;
  text-align: center;
  padding: 14px 10px 14px 10px;
}

.distill-site-header .title {
  font-size: 18px;
  min-width: 150px;
}

.distill-site-header .logo {
  padding: 0;
}

.distill-site-header .logo img {
  display: none;
  max-height: 20px;
  width: auto;
  margin-bottom: -4px;
}

.distill-site-header .nav-image img {
  max-height: 18px;
  width: auto;
  display: inline-block;
  margin-bottom: -3px;
}



@media screen and (min-width: 1000px) {
  .distill-site-header .logo img {
    display: inline-block;
  }
  .distill-site-header .nav-left {
    margin-left: 20px;
  }
  .distill-site-header .nav-right {
    margin-right: 20px;
  }
  .distill-site-header .title {
    padding-left: 12px;
  }
}


.distill-site-header .nav-toggle {
  display: none;
}

.nav-dropdown {
  display: inline-block;
  position: relative;
}

.nav-dropdown .nav-dropbtn {
  border: none;
  outline: none;
  color: rgba(255, 255, 255, 0.8);
  padding: 16px 10px;
  background-color: transparent;
  font-family: inherit;
  font-size: inherit;
  font-weight: inherit;
  margin: 0;
  margin-top: 1px;
  z-index: 2;
}

.nav-dropdown-content {
  display: none;
  position: absolute;
  background-color: white;
  min-width: 200px;
  border: 1px solid rgba(0,0,0,0.15);
  border-radius: 4px;
  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
  z-index: 1;
  margin-top: 2px;
  white-space: nowrap;
  padding-top: 4px;
  padding-bottom: 4px;
}

.nav-dropdown-content hr {
  margin-top: 4px;
  margin-bottom: 4px;
  border: none;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.nav-dropdown-active {
  display: block;
}

.nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
  color: black;
  padding: 6px 24px;
  text-decoration: none;
  display: block;
  text-align: left;
}

.nav-dropdown-content .nav-dropdown-header {
  display: block;
  padding: 5px 24px;
  padding-bottom: 0;
  text-transform: uppercase;
  font-size: 14px;
  color: #999999;
  white-space: nowrap;
}

.nav-dropdown:hover .nav-dropbtn {
  color: white;
}

.nav-dropdown-content a:hover {
  background-color: #ddd;
  color: black;
}

.nav-right .nav-dropdown-content {
  margin-left: -45%;
  right: 0;
}

@media screen and (max-width: 768px) {
  .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
  .distill-site-header a.nav-toggle {
    float: right;
    display: block;
  }
  .distill-site-header .title {
    margin-left: 0;
  }
  .distill-site-header .nav-right {
    margin-right: 0;
  }
  .distill-site-header {
    overflow: hidden;
  }
  .nav-right .nav-dropdown-content {
    margin-left: 0;
  }
}


@media screen and (max-width: 768px) {
  .distill-site-header.responsive {position: relative; min-height: 500px; }
  .distill-site-header.responsive a.nav-toggle {
    position: absolute;
    right: 0;
    top: 0;
  }
  .distill-site-header.responsive a,
  .distill-site-header.responsive .nav-dropdown {
    display: block;
    text-align: left;
  }
  .distill-site-header.responsive .nav-left,
  .distill-site-header.responsive .nav-right {
    width: 100%;
  }
  .distill-site-header.responsive .nav-dropdown {float: none;}
  .distill-site-header.responsive .nav-dropdown-content {position: relative;}
  .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
    display: block;
    width: 100%;
    text-align: left;
  }
}

/* Site Footer */

.distill-site-footer {
  width: 100%;
  overflow: hidden;
  box-sizing: border-box;
  z-index: 3;
  margin-top: 30px;
  padding-top: 30px;
  padding-bottom: 30px;
  text-align: center;
}

/* Headroom */

d-title {
  padding-top: 6rem;
}

@media print {
  d-title {
    padding-top: 4rem;
  }
}

.headroom {
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
}

.headroom--transition {
  transition: all .4s ease-in-out;
}

.headroom--unpinned {
  top: -100px;
}

.headroom--pinned {
  top: 0;
}

/* adjust viewport for navbar height */
/* helps vertically center bootstrap (non-distill) content */
.min-vh-100 {
  min-height: calc(100vh - 100px) !important;
}

</style>

<script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="../../site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"/>
<link href="../../site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"/>
<script src="../../site_libs/headroom-0.9.4/headroom.min.js"></script>
<!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

<style type="text/css">

body {
  background-color: white;
}

.pandoc-table {
  width: 100%;
}

.pandoc-table>caption {
  margin-bottom: 10px;
}

.pandoc-table th:not([align]) {
  text-align: left;
}

.pagedtable-footer {
  font-size: 15px;
}

d-byline .byline {
  grid-template-columns: 2fr 2fr;
}

d-byline .byline h3 {
  margin-block-start: 1.5em;
}

d-byline .byline .authors-affiliations h3 {
  margin-block-start: 0.5em;
}

.authors-affiliations .orcid-id {
  width: 16px;
  height:16px;
  margin-left: 4px;
  margin-right: 4px;
  vertical-align: middle;
  padding-bottom: 2px;
}

d-title .dt-tags {
  margin-top: 1em;
  grid-column: text;
}

.dt-tags .dt-tag {
  text-decoration: none;
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0em 0.4em;
  margin-right: 0.5em;
  margin-bottom: 0.4em;
  font-size: 70%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

d-article table.gt_table td,
d-article table.gt_table th {
  border-bottom: none;
  font-size: 100%;
}

.html-widget {
  margin-bottom: 2.0em;
}

.l-screen-inset {
  padding-right: 16px;
}

.l-screen .caption {
  margin-left: 10px;
}

.shaded {
  background: rgb(247, 247, 247);
  padding-top: 20px;
  padding-bottom: 20px;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .html-widget {
  margin-bottom: 0;
  border: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .shaded-content {
  background: white;
}

.text-output {
  margin-top: 0;
  line-height: 1.5em;
}

.hidden {
  display: none !important;
}

hr.section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  margin: 0px;
}


d-byline {
  border-top: none;
}

d-article {
  padding-top: 2.5rem;
  padding-bottom: 30px;
  border-top: none;
}

d-appendix {
  padding-top: 30px;
}

d-article>p>img {
  width: 100%;
}

d-article h2 {
  margin: 1rem 0 1.5rem 0;
}

d-article h3 {
  margin-top: 1.5rem;
}

d-article iframe {
  border: 1px solid rgba(0, 0, 0, 0.1);
  margin-bottom: 2.0em;
  width: 100%;
}

/* Tweak code blocks */

d-article div.sourceCode code,
d-article pre code {
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
}

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: auto;
}

d-article div.sourceCode {
  background-color: white;
}

d-article div.sourceCode pre {
  padding-left: 10px;
  font-size: 12px;
  border-left: 2px solid rgba(0,0,0,0.1);
}

d-article pre {
  font-size: 12px;
  color: black;
  background: none;
  margin-top: 0;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

d-article pre a {
  border-bottom: none;
}

d-article pre a:hover {
  border-bottom: none;
  text-decoration: underline;
}

d-article details {
  grid-column: text;
  margin-bottom: 0.8em;
}

@media(min-width: 768px) {

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: visible !important;
}

d-article div.sourceCode pre {
  padding-left: 18px;
  font-size: 14px;
}

/* tweak for Pandoc numbered line within distill */
d-article pre.numberSource code > span {
    left: -2em;
}

d-article pre {
  font-size: 14px;
}

}

figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

/* CSS for d-contents */

.d-contents {
  grid-column: text;
  color: rgba(0,0,0,0.8);
  font-size: 0.9em;
  padding-bottom: 1em;
  margin-bottom: 1em;
  padding-bottom: 0.5em;
  margin-bottom: 1em;
  padding-left: 0.25em;
  justify-self: start;
}

@media(min-width: 1000px) {
  .d-contents.d-contents-float {
    height: 0;
    grid-column-start: 1;
    grid-column-end: 4;
    justify-self: center;
    padding-right: 3em;
    padding-left: 2em;
  }
}

.d-contents nav h3 {
  font-size: 18px;
  margin-top: 0;
  margin-bottom: 1em;
}

.d-contents li {
  list-style-type: none
}

.d-contents nav > ul {
  padding-left: 0;
}

.d-contents ul {
  padding-left: 1em
}

.d-contents nav ul li {
  margin-top: 0.6em;
  margin-bottom: 0.2em;
}

.d-contents nav a {
  font-size: 13px;
  border-bottom: none;
  text-decoration: none
  color: rgba(0, 0, 0, 0.8);
}

.d-contents nav a:hover {
  text-decoration: underline solid rgba(0, 0, 0, 0.6)
}

.d-contents nav > ul > li > a {
  font-weight: 600;
}

.d-contents nav > ul > li > ul {
  font-weight: inherit;
}

.d-contents nav > ul > li > ul > li {
  margin-top: 0.2em;
}


.d-contents nav ul {
  margin-top: 0;
  margin-bottom: 0.25em;
}

.d-article-with-toc h2:nth-child(2) {
  margin-top: 0;
}


/* Figure */

.figure {
  position: relative;
  margin-bottom: 2.5em;
  margin-top: 1.5em;
}

.figure .caption {
  color: rgba(0, 0, 0, 0.6);
  font-size: 12px;
  line-height: 1.5em;
}

.figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

.figure .caption a {
  color: rgba(0, 0, 0, 0.6);
}

.figure .caption b,
.figure .caption strong, {
  font-weight: 600;
  color: rgba(0, 0, 0, 1.0);
}

/* Citations */

d-article .citation {
  color: inherit;
  cursor: inherit;
}

div.hanging-indent{
  margin-left: 1em; text-indent: -1em;
}

/* Citation hover box */

.tippy-box[data-theme~=light-border] {
  background-color: rgba(250, 250, 250, 0.95);
}

.tippy-content > p {
  margin-bottom: 0;
  padding: 2px;
}


/* Tweak 1000px media break to show more text */

@media(min-width: 1000px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 16px;
  }

  .grid {
    grid-column-gap: 16px;
  }

  d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
  }
  figure .caption, .figure .caption, figure figcaption {
    font-size: 13px;
  }
}

@media(min-width: 1180px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 32px;
  }

  .grid {
    grid-column-gap: 32px;
  }
}


/* Get the citation styles for the appendix (not auto-injected on render since
   we do our own rendering of the citation appendix) */

d-appendix .citation-appendix,
.d-appendix .citation-appendix {
  font-size: 11px;
  line-height: 15px;
  border-left: 1px solid rgba(0, 0, 0, 0.1);
  padding-left: 18px;
  border: 1px solid rgba(0,0,0,0.1);
  background: rgba(0, 0, 0, 0.02);
  padding: 10px 18px;
  border-radius: 3px;
  color: rgba(150, 150, 150, 1);
  overflow: hidden;
  margin-top: -12px;
  white-space: pre-wrap;
  word-wrap: break-word;
}

/* Include appendix styles here so they can be overridden */

d-appendix {
  contain: layout style;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-top: 60px;
  margin-bottom: 0;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  color: rgba(0,0,0,0.5);
  padding-top: 60px;
  padding-bottom: 48px;
}

d-appendix h3 {
  grid-column: page-start / text-start;
  font-size: 15px;
  font-weight: 500;
  margin-top: 1em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.65);
}

d-appendix h3 + * {
  margin-top: 1em;
}

d-appendix ol {
  padding: 0 0 0 15px;
}

@media (min-width: 768px) {
  d-appendix ol {
    padding: 0 0 0 30px;
    margin-left: -30px;
  }
}

d-appendix li {
  margin-bottom: 1em;
}

d-appendix a {
  color: rgba(0, 0, 0, 0.6);
}

d-appendix > * {
  grid-column: text;
}

d-appendix > d-footnote-list,
d-appendix > d-citation-list,
d-appendix > distill-appendix {
  grid-column: screen;
}

/* Include footnote styles here so they can be overridden */

d-footnote-list {
  contain: layout style;
}

d-footnote-list > * {
  grid-column: text;
}

d-footnote-list a.footnote-backlink {
  color: rgba(0,0,0,0.3);
  padding-left: 0.5em;
}



/* Anchor.js */

.anchorjs-link {
  /*transition: all .25s linear; */
  text-decoration: none;
  border-bottom: none;
}
*:hover > .anchorjs-link {
  margin-left: -1.125em !important;
  text-decoration: none;
  border-bottom: none;
}

/* Social footer */

.social_footer {
  margin-top: 30px;
  margin-bottom: 0;
  color: rgba(0,0,0,0.67);
}

.disqus-comments {
  margin-right: 30px;
}

.disqus-comment-count {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  cursor: pointer;
}

#disqus_thread {
  margin-top: 30px;
}

.article-sharing a {
  border-bottom: none;
  margin-right: 8px;
}

.article-sharing a:hover {
  border-bottom: none;
}

.sidebar-section.subscribe {
  font-size: 12px;
  line-height: 1.6em;
}

.subscribe p {
  margin-bottom: 0.5em;
}


.article-footer .subscribe {
  font-size: 15px;
  margin-top: 45px;
}


.sidebar-section.custom {
  font-size: 12px;
  line-height: 1.6em;
}

.custom p {
  margin-bottom: 0.5em;
}

/* Styles for listing layout (hide title) */
.layout-listing d-title, .layout-listing .d-title {
  display: none;
}

/* Styles for posts lists (not auto-injected) */


.posts-with-sidebar {
  padding-left: 45px;
  padding-right: 45px;
}

.posts-list .description h2,
.posts-list .description p {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

.posts-list .description h2 {
  font-weight: 700;
  border-bottom: none;
  padding-bottom: 0;
}

.posts-list h2.post-tag {
  border-bottom: 1px solid rgba(0, 0, 0, 0.2);
  padding-bottom: 12px;
}
.posts-list {
  margin-top: 60px;
  margin-bottom: 24px;
}

.posts-list .post-preview {
  text-decoration: none;
  overflow: hidden;
  display: block;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  padding: 24px 0;
}

.post-preview-last {
  border-bottom: none !important;
}

.posts-list .posts-list-caption {
  grid-column: screen;
  font-weight: 400;
}

.posts-list .post-preview h2 {
  margin: 0 0 6px 0;
  line-height: 1.2em;
  font-style: normal;
  font-size: 24px;
}

.posts-list .post-preview p {
  margin: 0 0 12px 0;
  line-height: 1.4em;
  font-size: 16px;
}

.posts-list .post-preview .thumbnail {
  box-sizing: border-box;
  margin-bottom: 24px;
  position: relative;
  max-width: 500px;
}
.posts-list .post-preview img {
  width: 100%;
  display: block;
}

.posts-list .metadata {
  font-size: 12px;
  line-height: 1.4em;
  margin-bottom: 18px;
}

.posts-list .metadata > * {
  display: inline-block;
}

.posts-list .metadata .publishedDate {
  margin-right: 2em;
}

.posts-list .metadata .dt-authors {
  display: block;
  margin-top: 0.3em;
  margin-right: 2em;
}

.posts-list .dt-tags {
  display: block;
  line-height: 1em;
}

.posts-list .dt-tags .dt-tag {
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0.3em 0.4em;
  margin-right: 0.2em;
  margin-bottom: 0.4em;
  font-size: 60%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

.posts-list img {
  opacity: 1;
}

.posts-list img[data-src] {
  opacity: 0;
}

.posts-more {
  clear: both;
}


.posts-sidebar {
  font-size: 16px;
}

.posts-sidebar h3 {
  font-size: 16px;
  margin-top: 0;
  margin-bottom: 0.5em;
  font-weight: 400;
  text-transform: uppercase;
}

.sidebar-section {
  margin-bottom: 30px;
}

.categories ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.categories li {
  color: rgba(0, 0, 0, 0.8);
  margin-bottom: 0;
}

.categories li>a {
  border-bottom: none;
}

.categories li>a:hover {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
}

.categories .active {
  font-weight: 600;
}

.categories .category-count {
  color: rgba(0, 0, 0, 0.4);
}


@media(min-width: 768px) {
  .posts-list .post-preview h2 {
    font-size: 26px;
  }
  .posts-list .post-preview .thumbnail {
    float: right;
    width: 30%;
    margin-bottom: 0;
  }
  .posts-list .post-preview .description {
    float: left;
    width: 45%;
  }
  .posts-list .post-preview .metadata {
    float: left;
    width: 20%;
    margin-top: 8px;
  }
  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.5em;
    font-size: 16px;
  }
  .posts-with-sidebar .posts-list {
    float: left;
    width: 75%;
  }
  .posts-with-sidebar .posts-sidebar {
    float: right;
    width: 20%;
    margin-top: 60px;
    padding-top: 24px;
    padding-bottom: 24px;
  }
}


/* Improve display for browsers without grid (IE/Edge <= 15) */

.downlevel {
  line-height: 1.6em;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  margin: 0;
}

.downlevel .d-title {
  padding-top: 6rem;
  padding-bottom: 1.5rem;
}

.downlevel .d-title h1 {
  font-size: 50px;
  font-weight: 700;
  line-height: 1.1em;
  margin: 0 0 0.5rem;
}

.downlevel .d-title p {
  font-weight: 300;
  font-size: 1.2rem;
  line-height: 1.55em;
  margin-top: 0;
}

.downlevel .d-byline {
  padding-top: 0.8em;
  padding-bottom: 0.8em;
  font-size: 0.8rem;
  line-height: 1.8em;
}

.downlevel .section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
}

.downlevel .d-article {
  font-size: 1.06rem;
  line-height: 1.7em;
  padding-top: 1rem;
  padding-bottom: 2rem;
}


.downlevel .d-appendix {
  padding-left: 0;
  padding-right: 0;
  max-width: none;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.5);
  padding-top: 40px;
  padding-bottom: 48px;
}

.downlevel .footnotes ol {
  padding-left: 13px;
}

.downlevel .base-grid,
.downlevel .distill-header,
.downlevel .d-title,
.downlevel .d-abstract,
.downlevel .d-article,
.downlevel .d-appendix,
.downlevel .distill-appendix,
.downlevel .d-byline,
.downlevel .d-footnote-list,
.downlevel .d-citation-list,
.downlevel .distill-footer,
.downlevel .appendix-bottom,
.downlevel .posts-container {
  padding-left: 40px;
  padding-right: 40px;
}

@media(min-width: 768px) {
  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
  padding-left: 150px;
  padding-right: 150px;
  max-width: 900px;
}
}

.downlevel pre code {
  display: block;
  border-left: 2px solid rgba(0, 0, 0, .1);
  padding: 0 0 0 20px;
  font-size: 14px;
}

.downlevel code, .downlevel pre {
  color: black;
  background: none;
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

.downlevel .posts-list .post-preview {
  color: inherit;
}



</style>

<script type="application/javascript">

function is_downlevel_browser() {
  if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                 window.navigator.userAgent)) {
    return true;
  } else {
    return window.load_distill_framework === undefined;
  }
}

// show body when load is complete
function on_load_complete() {

  // add anchors
  if (window.anchors) {
    window.anchors.options.placement = 'left';
    window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
  }


  // set body to visible
  document.body.style.visibility = 'visible';

  // force redraw for leaflet widgets
  if (window.HTMLWidgets) {
    var maps = window.HTMLWidgets.findAll(".leaflet");
    $.each(maps, function(i, el) {
      var map = this.getMap();
      map.invalidateSize();
      map.eachLayer(function(layer) {
        if (layer instanceof L.TileLayer)
          layer.redraw();
      });
    });
  }

  // trigger 'shown' so htmlwidgets resize
  $('d-article').trigger('shown');
}

function init_distill() {

  init_common();

  // create front matter
  var front_matter = $('<d-front-matter></d-front-matter>');
  $('#distill-front-matter').wrap(front_matter);

  // create d-title
  $('.d-title').changeElementType('d-title');

  // separator
  var separator = '<hr class="section-separator" style="clear: both"/>';
  // prepend separator above appendix
  $('.d-byline').before(separator);
  $('.d-article').before(separator);

  // create d-byline
  var byline = $('<d-byline></d-byline>');
  $('.d-byline').replaceWith(byline);

  // create d-article
  var article = $('<d-article></d-article>');
  $('.d-article').wrap(article).children().unwrap();

  // move posts container into article
  $('.posts-container').appendTo($('d-article'));

  // create d-appendix
  $('.d-appendix').changeElementType('d-appendix');

  // flag indicating that we have appendix items
  var appendix = $('.appendix-bottom').children('h3').length > 0;

  // replace footnotes with <d-footnote>
  $('.footnote-ref').each(function(i, val) {
    appendix = true;
    var href = $(this).attr('href');
    var id = href.replace('#', '');
    var fn = $('#' + id);
    var fn_p = $('#' + id + '>p');
    fn_p.find('.footnote-back').remove();
    var text = fn_p.html();
    var dtfn = $('<d-footnote></d-footnote>');
    dtfn.html(text);
    $(this).replaceWith(dtfn);
  });
  // remove footnotes
  $('.footnotes').remove();

  // move refs into #references-listing
  $('#references-listing').replaceWith($('#refs'));

  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    var id = $(this).attr('id');
    $('.d-contents a[href="#' + id + '"]').parent().remove();
    appendix = true;
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
  });

  // show d-appendix if we have appendix content
  $("d-appendix").css('display', appendix ? 'grid' : 'none');

  // localize layout chunks to just output
  $('.layout-chunk').each(function(i, val) {

    // capture layout
    var layout = $(this).attr('data-layout');

    // apply layout to markdown level block elements
    var elements = $(this).children().not('details, div.sourceCode, pre, script');
    elements.each(function(i, el) {
      var layout_div = $('<div class="' + layout + '"></div>');
      if (layout_div.hasClass('shaded')) {
        var shaded_content = $('<div class="shaded-content"></div>');
        $(this).wrap(shaded_content);
        $(this).parent().wrap(layout_div);
      } else {
        $(this).wrap(layout_div);
      }
    });


    // unwrap the layout-chunk div
    $(this).children().unwrap();
  });

  // remove code block used to force  highlighting css
  $('.distill-force-highlighting-css').parent().remove();

  // remove empty line numbers inserted by pandoc when using a
  // custom syntax highlighting theme, except when numbering line
  // in code chunk
  $('pre:not(.numberLines) code.sourceCode a:empty').remove();

  // load distill framework
  load_distill_framework();

  // wait for window.distillRunlevel == 4 to do post processing
  function distill_post_process() {

    if (!window.distillRunlevel || window.distillRunlevel < 4)
      return;

    // hide author/affiliations entirely if we have no authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;
    if (!have_authors)
      $('d-byline').addClass('hidden');

    // article with toc class
    $('.d-contents').parent().addClass('d-article-with-toc');

    // strip links that point to #
    $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

    // add orcid ids
    $('.authors-affiliations').find('.author').each(function(i, el) {
      var orcid_id = front_matter.authors[i].orcidID;
      var author_name = front_matter.authors[i].author
      if (orcid_id) {
        var a = $('<a></a>');
        a.attr('href', 'https://orcid.org/' + orcid_id);
        var img = $('<img></img>');
        img.addClass('orcid-id');
        img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
        img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
        a.append(img);
        $(this).append(a);
      }
    });

    // hide elements of author/affiliations grid that have no value
    function hide_byline_column(caption) {
      $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
    }

    // affiliations
    var have_affiliations = false;
    for (var i = 0; i<front_matter.authors.length; ++i) {
      var author = front_matter.authors[i];
      if (author.affiliation !== "&nbsp;") {
        have_affiliations = true;
        break;
      }
    }
    if (!have_affiliations)
      $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

    // published date
    if (!front_matter.publishedDate)
      hide_byline_column("Published");

    // document object identifier
    var doi = $('d-byline').find('h3:contains("DOI")');
    var doi_p = doi.next().empty();
    if (!front_matter.doi) {
      // if we have a citation and valid citationText then link to that
      if ($('#citation').length > 0 && front_matter.citationText) {
        doi.html('Citation');
        $('<a href="#citation"></a>')
          .text(front_matter.citationText)
          .appendTo(doi_p);
      } else {
        hide_byline_column("DOI");
      }
    } else {
      $('<a></a>')
         .attr('href', "https://doi.org/" + front_matter.doi)
         .html(front_matter.doi)
         .appendTo(doi_p);
    }

     // change plural form of authors/affiliations
    if (front_matter.authors.length === 1) {
      var grid = $('.authors-affiliations');
      grid.children('h3:contains("Authors")').text('Author');
      grid.children('h3:contains("Affiliations")').text('Affiliation');
    }

    // remove d-appendix and d-footnote-list local styles
    $('d-appendix > style:first-child').remove();
    $('d-footnote-list > style:first-child').remove();

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // hoverable references
    $('span.citation[data-cites]').each(function() {
      const citeChild = $(this).children()[0]
      // Do not process if @xyz has been used without escaping and without bibliography activated
      // https://github.com/rstudio/distill/issues/466
      if (citeChild === undefined) return true

      if (citeChild.nodeName == "D-FOOTNOTE") {
        var fn = citeChild
        $(this).html(fn.shadowRoot.querySelector("sup"))
        $(this).id = fn.id
        fn.remove()
      }
      var refs = $(this).attr('data-cites').split(" ");
      var refHtml = refs.map(function(ref) {
        // Could use CSS.escape too here, we insure backward compatibility in navigator
        return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
      }).join("\n");
      window.tippy(this, {
        allowHTML: true,
        content: refHtml,
        maxWidth: 500,
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start'
      });
    });

    // fix footnotes in tables (#411)
    // replacing broken distill.pub feature
    $('table d-footnote').each(function() {
      // we replace internal showAtNode methode which is triggered when hovering a footnote
      this.hoverBox.showAtNode = function(node) {
        // ported from https://github.com/distillpub/template/pull/105/files
        calcOffset = function(elem) {
            let x = elem.offsetLeft;
            let y = elem.offsetTop;
            // Traverse upwards until an `absolute` element is found or `elem`
            // becomes null.
            while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                x += elem.offsetLeft;
                y += elem.offsetTop;
            }

            return { left: x, top: y };
        }
        // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
        const bbox = node.getBoundingClientRect();
        const offset = calcOffset(node);
        this.show([offset.left + bbox.width, offset.top + bbox.height]);
      }
    })

    // clear polling timer
    clearInterval(tid);

    // show body now that everything is ready
    on_load_complete();
  }

  var tid = setInterval(distill_post_process, 50);
  distill_post_process();

}

function init_downlevel() {

  init_common();

   // insert hr after d-title
  $('.d-title').after($('<hr class="section-separator"/>'));

  // check if we have authors
  var front_matter = JSON.parse($("#distill-front-matter").html());
  var have_authors = front_matter.authors && front_matter.authors.length > 0;

  // manage byline/border
  if (!have_authors)
    $('.d-byline').remove();
  $('.d-byline').after($('<hr class="section-separator"/>'));
  $('.d-byline a').remove();

  // remove toc
  $('.d-contents').remove();

  // move appendix elements
  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
  });


  // inject headers into references and footnotes
  var refs_header = $('<h3></h3>');
  refs_header.text('References');
  $('#refs').prepend(refs_header);

  var footnotes_header = $('<h3></h3');
  footnotes_header.text('Footnotes');
  $('.footnotes').children('hr').first().replaceWith(footnotes_header);

  // move appendix-bottom entries to the bottom
  $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
  $('.appendix-bottom').remove();

  // remove appendix if it's empty
  if ($('.d-appendix').children().length === 0)
    $('.d-appendix').remove();

  // prepend separator above appendix
  $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

  // trim code
  $('pre>code').each(function(i, val) {
    $(this).html($.trim($(this).html()));
  });

  // move posts-container right before article
  $('.posts-container').insertBefore($('.d-article'));

  $('body').addClass('downlevel');

  on_load_complete();
}


function init_common() {

  // jquery plugin to change element types
  (function($) {
    $.fn.changeElementType = function(newType) {
      var attrs = {};

      $.each(this[0].attributes, function(idx, attr) {
        attrs[attr.nodeName] = attr.nodeValue;
      });

      this.replaceWith(function() {
        return $("<" + newType + "/>", attrs).append($(this).contents());
      });
    };
  })(jQuery);

  // prevent underline for linked images
  $('a > img').parent().css({'border-bottom' : 'none'});

  // mark non-body figures created by knitr chunks as 100% width
  $('.layout-chunk').each(function(i, val) {
    var figures = $(this).find('img, .html-widget');
    // ignore leaflet img layers (#106)
    figures = figures.filter(':not(img[class*="leaflet"])')
    if ($(this).attr('data-layout') !== "l-body") {
      figures.css('width', '100%');
    } else {
      figures.css('max-width', '100%');
      figures.filter("[width]").each(function(i, val) {
        var fig = $(this);
        fig.css('width', fig.attr('width') + 'px');
      });

    }
  });

  // auto-append index.html to post-preview links in file: protocol
  // and in rstudio ide preview
  $('.post-preview').each(function(i, val) {
    if (window.location.protocol === "file:")
      $(this).attr('href', $(this).attr('href') + "index.html");
  });

  // get rid of index.html references in header
  if (window.location.protocol !== "file:") {
    $('.distill-site-header a[href]').each(function(i,val) {
      $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
    });
  }

  // add class to pandoc style tables
  $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
  $('.kable-table').children('table').addClass('pandoc-table');

  // add figcaption style to table captions
  $('caption').parent('table').addClass("figcaption");

  // initialize posts list
  if (window.init_posts_list)
    window.init_posts_list();

  // implmement disqus comment link
  $('.disqus-comment-count').click(function() {
    window.headroom_prevent_pin = true;
    $('#disqus_thread').toggleClass('hidden');
    if (!$('#disqus_thread').hasClass('hidden')) {
      var offset = $(this).offset();
      $(window).resize();
      $('html, body').animate({
        scrollTop: offset.top - 35
      });
    }
  });
}

document.addEventListener('DOMContentLoaded', function() {
  if (is_downlevel_browser())
    init_downlevel();
  else
    window.addEventListener('WebComponentsReady', init_distill);
});

</script>

<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.5);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

.distill-site-footer {
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */
</style>
<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.7);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #ffffff;
}

.distill-site-footer {
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #fafafa;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */

.nav-right > a {
  text-transform: uppercase;
}

d-title h1, d-title p, d-title figure,
d-abstract p, d-abstract b {
  grid-column: page;
}

.rj-blue {
  color: #2467bb;
}

ul li {
  line-height: 1.6;
  margin-top: 0em;
  margin-bottom: 0em;
}</style>
<style type="text/css">
/* base style */

/* FONT FAMILIES */

:root {
  --heading-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  --mono-default: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  --body-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

body,
.posts-list .post-preview p,
.posts-list .description p {
  font-family: var(--body-font), var(--body-default);
}

h1, h2, h3, h4, h5, h6,
.posts-list .post-preview h2,
.posts-list .description h2 {
  font-family: var(--heading-font), var(--heading-default);
}

d-article div.sourceCode code,
d-article pre code {
  font-family: var(--mono-font), var(--mono-default);
}


/*-- TITLE --*/
d-title h1,
.posts-list > h1 {
  color: var(--title-color, black);
}

d-title h1 {
  font-size: var(--title-size, 50px);
}

/*-- HEADERS --*/
d-article h1,
d-article h2,
d-article h3,
d-article h4,
d-article h5,
d-article h6 {
  color: var(--header-color, rgba(0, 0, 0, 0.8));
}

/*-- BODY --*/
d-article > p,  /* only text inside of <p> tags */
d-article > ul, /* lists */
d-article > ol {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
  font-size: var(--body-size, 1.06rem);
}


/*-- CODE --*/
d-article div.sourceCode code,
d-article pre code {
  font-size: var(--code-size, 14px);
}

/*-- ASIDE --*/
d-article aside {
  font-size: var(--aside-size, 12px);
  color: var(--aside-color, rgba(0, 0, 0, 0.6));
}

/*-- FIGURE CAPTIONS --*/
figure .caption,
figure figcaption,
.figure .caption {
  font-size: var(--fig-cap-size, 13px);
  color: var(--fig-cap-color, rgba(0, 0, 0, 0.6));
}

/*-- METADATA --*/
d-byline h3 {
  font-size: var(--heading-size, 0.6rem);
  color: var(--heading-color, rgba(0, 0, 0, 0.5));
}

d-byline {
  font-size: var(--body-size, 0.8rem);
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

d-byline a,
d-article d-byline a {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

/*-- TABLE OF CONTENTS --*/
.d-contents nav h3 {
  font-size: var(--heading-size, 18px);
}

.d-contents nav a {
  font-size: var(--contents-size, 13px);
}

/*-- APPENDIX --*/
d-appendix h3 {
  font-size: var(--heading-size, 15px);
  color: var(--heading-color, rgba(0, 0, 0, 0.65));
}

d-appendix {
  font-size: var(--text-size, 0.8em);
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

d-appendix d-footnote-list a.footnote-backlink {
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

/*-- WEBSITE HEADER + FOOTER --*/
.distill-site-header .title {
  font-size: var(--title-size, 18px);
  font-family: var(--navbar-font), var(--heading-default);
}

.distill-site-header a,
.nav-dropdown .nav-dropbtn {
  font-family: var(--navbar-font), var(--heading-default);
}

.nav-dropdown .nav-dropbtn {
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  font-size: var(--text-size, 15px);
}

.distill-site-header a:hover,
.nav-dropdown:hover .nav-dropbtn {
  color: var(--hover-color, white);
}

.distill-site-header {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer a:hover {
  color: var(--hover-color, white);
}</style>
<!--/radix_placeholder_distill-->
  <script src="../../site_libs/header-attrs-2.29/header-attrs.js"></script>
  <script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="../../site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="../../site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="../../site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="../../site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="../../site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="../../site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="../../site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="../../site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
<!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Mathjax config (add automatic linebreaks when supported)
      // MathJax = {
      //    tex: {
      //        inlineMath: [['$', '$'], ['\\(', '\\)']],
      //        displayMath: [['$$', '$$'], ['\\[', '\\]']],
      //        tags: 'ams',
      //        multline: true,
      //    },
      //    options: {
      //        linebreaks: { automatic: true },
      //    },
      // };

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      buttons.append('<a href="' + slug + '.zip" class="dt-tag"><i class="fas fa-file-zipper"></i> Supplement</a>')

      // adds Abstract: in front of the first <p> in the title section --
      // unless it happens to be the subtitle (FIXME: this is a bad hack - can't distill do this?)
      var tpar = $("d-title p:not(:empty)").filter(function() {
        return !$(this).hasClass("subtitle");
      }).first();
      if (tpar) {
        const abstract = $('<d-abstract>')
        abstract.append('<b>Abstract:</b><br>')
        abstract.append(tpar) // Move description to d-abstract
        $("d-title p:empty").remove() // Remove empty paragraphs after title
        abstract.append(buttons)
        abstract.insertAfter($('d-title')) // Add abstract section after title */
      }

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
    
    d-appendix .citation-appendix,
    .d-appendix .citation-appendix {
      color: rgb(60, 60, 60);
    }

    d-article h2 {
      border-bottom: 0px solid rgba(0, 0, 0, 0.1);
      padding-bottom: 0rem;
    }
    d-article h3 {
      font-size: 20px;
    }
    d-article h4 {
      font-size: 18px;
      text-transform: none;
    }

    @media (min-width: 1024px) {
      d-article h2 {
        font-size: 32px;
      }
      d-article h3 {
        font-size: 24px;
      }
      d-article h4 {
        font-size: 20px;
      }
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"LUCIDus: An R Package For Implementing Latent Unknown Clustering By Integrating Multi-omics Data (LUCID) With Phenotypic Traits","description":"Many studies are leveraging current technologies to obtain multiple\nomics measurements on the same individuals. These measurements are\nusually cross-sectional, and methods developed and commonly used focus\non omic integration at a single time point. More unique, and a growing\narea of interest, are studies that leverage biology or the temporal\nsequence of measurements to relate long-term exposures or germline\ngenetics to intermediate measures capturing transitional processes\nthat ultimately result in an outcome. In this context, we have\npreviously introduced an integrative model named Latent Unknown\nClustering by Integrating multi-omics Data (LUCID) aiming to\ndistinguish unique effects of environmental exposures or germline\ngenetics and informative omic effects while jointly estimating\nsubgroups of individuals relevant to the outcome of interest. This\nmultiple omics analysis consists of early integration (concatenation\nof omic layers to estimate common subgroups); intermediate integration\n(omic-layer-specific estimation of subgroups that are all related to\nthe outcome); and late integration (omic-layer-specific estimation of\nsubgroups that are then interrelated by a priori structures). In this\narticle, we introduce LUCIDus version 3, an R package to implement the\nLUCID model. We review the statistical background of the model and\nintroduce the workflow of LUCIDus, including model fitting, model\nselection, interpretation, inference, and prediction. Throughout, we\nuse a realistic but simulated dataset based on an ongoing study, the\nHuman Early Life Exposome Study (HELIX), to illustrate the workflow.","doi":"10.32614/RJ-2024-012","authors":[{"author":"Yinqi Zhao","authorURL":"#","affiliation":"Division of Biostatistics, Department of Population and Public Health\nSciences, University of Southern California","affiliationURL":"#","orcidID":"0000-0003-2413-732X"},{"author":"Qiran Jia","authorURL":"#","affiliation":"Division of Biostatistics, Department of Population and Public Health\nSciences, University of Southern California","affiliationURL":"#","orcidID":"0000-0002-0790-5967"},{"author":"Jesse A. Goodrich","authorURL":"#","affiliation":"Division of Environmental Epidemiology, Department of Population and\nPublic Health Sciences, University of Southern California","affiliationURL":"#","orcidID":""},{"author":"David V. Conti","authorURL":"#","affiliation":"Division of Biostatistics, Department of Population and Public Health\nSciences, University of Southern California","affiliationURL":"#","orcidID":""}],"publishedDate":"2025-03-11T00:00:00.000+00:00","citationText":"Zhao, et al., 2025"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a class="logo" href="../../index.html">
<img src="../../resources/rlogo.png" alt="Logo"/>
</a>
<a href="../../index.html" class="title">The R Journal</a>
</div>
<div class="nav-right">
<a href="../../index.html">Home</a>
<a href="../../issues/2024-3">Current</a>
<a href="../../issues.html">Issues</a>
<a href="../../news.html">News</a>
<a href="../../submissions.html">Submit</a>
<a href="../../editors.html">Editorial board</a>
<a href="../../articles.xml">
<i class="fa fa-rss" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>LUCIDus: An R Package For Implementing Latent Unknown Clustering By Integrating Multi-omics Data (LUCID) With Phenotypic Traits</h1>

<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>Many studies are leveraging current technologies to obtain multiple
omics measurements on the same individuals. These measurements are
usually cross-sectional, and methods developed and commonly used focus
on omic integration at a single time point. More unique, and a growing
area of interest, are studies that leverage biology or the temporal
sequence of measurements to relate long-term exposures or germline
genetics to intermediate measures capturing transitional processes
that ultimately result in an outcome. In this context, we have
previously introduced an integrative model named Latent Unknown
Clustering by Integrating multi-omics Data (LUCID) aiming to
distinguish unique effects of environmental exposures or germline
genetics and informative omic effects while jointly estimating
subgroups of individuals relevant to the outcome of interest. This
multiple omics analysis consists of early integration (concatenation
of omic layers to estimate common subgroups); intermediate integration
(omic-layer-specific estimation of subgroups that are all related to
the outcome); and late integration (omic-layer-specific estimation of
subgroups that are then interrelated by a priori structures). In this
article, we introduce LUCIDus version 3, an R package to implement the
LUCID model. We review the statistical background of the model and
introduce the workflow of LUCIDus, including model fitting, model
selection, interpretation, inference, and prediction. Throughout, we
use a realistic but simulated dataset based on an ongoing study, the
Human Early Life Exposome Study (HELIX), to illustrate the workflow.</p></p>
</div>

<div class="d-byline">
  Yinqi Zhao  (Division of Biostatistics, Department of Population and Public Health
Sciences, University of Southern California)
  
,   Qiran Jia  (Division of Biostatistics, Department of Population and Public Health
Sciences, University of Southern California)
  
,   Jesse A. Goodrich  (Division of Environmental Epidemiology, Department of Population and
Public Health Sciences, University of Southern California)
  
,   David V. Conti  (Division of Biostatistics, Department of Population and Public Health
Sciences, University of Southern California)
  
<br/>2025-03-11
</div>

<div class="d-article">
<div class="article">
<h3 data-number="1" id="sec1"><span class="header-section-number">1</span> Introduction</h3>
<p>The rapid advancement in high-throughput technology has made it possible
to measure multiple omics (referred to as “multi-omics”) data on the
same person in many cohort studies. These datasets include DNA genome
sequences <span class="citation" data-cites="goodwin2016coming">(<a href="#ref-goodwin2016coming" role="doc-biblioref">Goodwin et al. 2016</a>)</span>, RNA expression data <span class="citation" data-cites="ozsolak2011rna">(<a href="#ref-ozsolak2011rna" role="doc-biblioref">Ozsolak and Milos 2011</a>)</span>,
metabolites in biofluids <span class="citation" data-cites="beger2013review">(<a href="#ref-beger2013review" role="doc-biblioref">Beger 2013</a>)</span>, proteins in cells or
tissues <span class="citation" data-cites="aslam2017proteomics">(<a href="#ref-aslam2017proteomics" role="doc-biblioref">Aslam et al. 2017</a>)</span>, as well as chemical exposures in the
environment <span class="citation" data-cites="wild2005complementing">(<a href="#ref-wild2005complementing" role="doc-biblioref">Wild 2005</a>)</span>. For example, the Human Early-Life
Exposome project (HELIX) measured molecular omics signatures including
RNA expression data, metabolites, plasma proteins, etc. from 1300
children at the age of 6-11 in six European countries
<span class="citation" data-cites="vrijheid2014human">(<a href="#ref-vrijheid2014human" role="doc-biblioref">Vrijheid et al. 2014</a>)</span>. Guided by biology or the temporal sequence of
measurements, these studies often share a common structure that relates
germline genetics or environmental exposures to intermediate factors
capturing transitional processes that ultimately result in an outcome.
While these suspected causal pathways can be measured with current omic
technology, the analysis often focuses on cross-sectional relationships
using cluster analysis or models focused on feature selection. This is
in part due to the analytic and computational challenges resulting from
the complex structure of the data collected from various sources,
limited sample sizes, and the high dimensionality of multi-omics
information <span class="citation" data-cites="tini2019multi">(<a href="#ref-tini2019multi" role="doc-biblioref">Tini et al. 2019</a>)</span>. Existing methods do not consider risk
factors that precede the omic measurements and link those factors with
disease or trait outcomes while using multi-omics data to characterize
the underlying mechanism. Thus, there is a great need to develop and
easily implement approaches leveraging these mediating relationships or
latent structures for the integration of multi-omics data
<span class="citation" data-cites="subramanian2020multi">(<a href="#ref-subramanian2020multi" role="doc-biblioref">Subramanian et al. 2020</a>)</span>.</p>
<p>Conventional analysis tools range from clustering approaches to variable
selection approaches <span class="citation" data-cites="gonzalez2019omic">(<a href="#ref-gonzalez2019omic" role="doc-biblioref">González and Cáceres 2019</a>)</span>. Clustering is a fundamental
method for analyzing both single omic and multi-omics data
<span class="citation" data-cites="rappoport2018multi">(<a href="#ref-rappoport2018multi" role="doc-biblioref">Rappoport and Shamir 2018</a>)</span>. Clustering aims to divide observations into
several groups (called clusters) so that observations within a group are
similar while samples in different groups are dissimilar. Clustering has
many important applications in the field of biological science and
medicine, such as defining gene sets <span class="citation" data-cites="hejblum2015time">(<a href="#ref-hejblum2015time" role="doc-biblioref">Hejblum et al. 2015</a>)</span>, identifying
subtypes of cancer, and performing diagnostic predictions
<span class="citation" data-cites="curtis2012genomic khan2001classification">(<a href="#ref-khan2001classification" role="doc-biblioref">Khan et al. 2001</a>; <a href="#ref-curtis2012genomic" role="doc-biblioref">Curtis et al. 2012</a>)</span>, defining cell types in
flow cytometry and scRNA-seq experiments
<span class="citation" data-cites="chan2008statistical hejblum2019sequential prabhakaran2016dirichlet">(<a href="#ref-chan2008statistical" role="doc-biblioref">Chan et al. 2008</a>; <a href="#ref-prabhakaran2016dirichlet" role="doc-biblioref">Prabhakaran et al. 2016</a>; <a href="#ref-hejblum2019sequential" role="doc-biblioref">Hejblum et al. 2019</a>)</span>,
and estimating protein localization <span class="citation" data-cites="crook2018bayesian">(<a href="#ref-crook2018bayesian" role="doc-biblioref">Crook et al. 2018</a>)</span>. Conventional
clustering methods applied to multi-omics data either concatenate
multiple datasets (often called early integration) or analyze each
dataset independently (i.e., late integration). However, both approaches
fall short of adequately capturing the variation across omic levels or
reflecting the heterogeneity of the integrated datasets. To overcome
these limitations, several specific integrated clustering methods for
analyzing multi-omics data are available in the community.
<span class="citation" data-cites="pierre2020clustering">Pierre-Jean et al. (<a href="#ref-pierre2020clustering" role="doc-biblioref">2020</a>)</span> conducted a thorough review comparing 13
unsupervised methods for multi-omics data integration via clustering.
Methods include SGCCA <span class="citation" data-cites="tenenhaus2014variable">(<a href="#ref-tenenhaus2014variable" role="doc-biblioref">Tenenhaus et al. 2014</a>)</span>, SNF
<span class="citation" data-cites="wang2014similarity">(<a href="#ref-wang2014similarity" role="doc-biblioref">Wang et al. 2014</a>)</span>, and iCluster plus <span class="citation" data-cites="mo2013pattern">(<a href="#ref-mo2013pattern" role="doc-biblioref">Mo et al. 2013</a>)</span>, etc. Building
upon the foundation of Principal Component Analysis (PCA), Joint and
Individual Variation Explained (JIVE) conducts integrated clustering of
multi-omics data via a variance decomposition framework to summarize
information across multiple omic layers <span class="citation" data-cites="lock2013joint">(<a href="#ref-lock2013joint" role="doc-biblioref">Lock et al. 2013</a>)</span>. Within the
Bayesian framework, <span class="citation" data-cites="kirk2012bayesian">Kirk et al. (<a href="#ref-kirk2012bayesian" role="doc-biblioref">2012</a>)</span> proposed an unsupervised approach,
Bayesian correlated clustering, that simultaneously models each omic
layer via a mixture model while considering the correlation between each
model. In a similar fashion, Bayesian consensus clustering can flexibly
describe the dependency and the heterogeneity of the multi-omics data
<span class="citation" data-cites="lock2013bayesian">(<a href="#ref-lock2013bayesian" role="doc-biblioref">Lock and Dunson 2013</a>)</span>. An alternative approach to clustering includes
dimension reduction or variable selection to reduce noise, improve model
interpretability, and avoid problems with overfitting. For example, the
least absolute shrinkage and selection operator (LASSO) induces sparsity
in model coefficients based on <span class="math inline">\(L_1\)</span> norm regularization
<span class="citation" data-cites="tibshirani1996regression">(<a href="#ref-tibshirani1996regression" role="doc-biblioref">Tibshirani 1996</a>)</span>. An extension of LASSO, Group LASSO, allows
for variable selection on both individual and pre-specified grouped
variables <span class="citation" data-cites="yuan2006model">(<a href="#ref-yuan2006model" role="doc-biblioref">Yuan and Lin 2006</a>)</span>. <span class="math inline">\(L_1\)</span> norm regularization is integrated into
many clustering approaches to reduce data dimensionality, such as SGCCA
and iCluster plus. Extensions of variable selection approaches exist for
mediation as well and include both high-dimensional and latent variable
approaches. HIMA first implements pre-screening of the high-dimensional
mediators and then utilizes an LASSO-type penalty to obtain a sparse
solution <span class="citation" data-cites="zhang2016estimating">(<a href="#ref-zhang2016estimating" role="doc-biblioref">Zhang et al. 2016</a>)</span>, while BAMA is a Bayesian shrinkage
approach that employs continuous shrinkage priors on the key
coefficients to select active mediators with large effects only
<span class="citation" data-cites="song2020bayesian">(<a href="#ref-song2020bayesian" role="doc-biblioref">Song et al. 2020</a>)</span>. Both <span class="citation" data-cites="albert2016causal">Albert et al. (<a href="#ref-albert2016causal" role="doc-biblioref">2016</a>)</span> and <span class="citation" data-cites="derkach2019high">Derkach et al. (<a href="#ref-derkach2019high" role="doc-biblioref">2019</a>)</span>
proposed statistical models within the causal mediation framework that
summarize the information from high-dimensional multi-omics data into a
latent variable to facilitate interpretation.</p>
<p>In the context of linking multi-omics to health outcomes and identifying
key omic features that drive the outcomes, <span class="citation" data-cites="singh2019diablo">Singh et al. (<a href="#ref-singh2019diablo" role="doc-biblioref">2019</a>)</span> proposed
Data Integration Analysis for Biomarker discovery using Latent
cOmponents (DIABLO), which extends the unsupervised sGCCA to a
supervised framework. DIABLO achieves summarizing common information
across different omic layers and conducting variable selection while
discriminating the outcome of interest. Nonetheless, DIABLO fails to
take into account the impact of risk factors related to the outcome and
their interplay with multi-omics data. Finally, <span class="citation" data-cites="peng2020latent">Peng et al. (<a href="#ref-peng2020latent" role="doc-biblioref">2020</a>)</span> proposed
a model called Latent Unknown Clustering by Integrating multi-omics Data
(LUCID) that incorporates both clustering and variable selection to
distinguish unique effects of germline genetics or environmental
exposures and informative omic effects while jointly estimating
subgroups of individuals relevant to the outcome of interest. LUCID is a
novel ‘quasi-mediation’ approach that uses latent variable analysis to
estimate subgroups of individuals characterized by key omic factors and
with differential associations to the outcome and exposures. They
demonstrated the performance of LUCID through extensive simulation
studies and real data applications to highlight the integration of
genomic, exposomic, and metabolomic data.
<a href="https://CRAN.R-project.org/package=LUCIDus"><strong>LUCIDus</strong></a>, the R package
to implement the LUCID model, provides an integrated clustering
framework and has numerous downloads (around 19,000 times since it was
first introduced according to
<a href="https://CRAN.R-project.org/package=dlstats"><strong>dlstats</strong></a>
<span class="citation" data-cites="yu2022dlstats">(<a href="#ref-yu2022dlstats" role="doc-biblioref">Yu 2022</a>)</span>). It has also been applied in several environmental
epidemiological studies
<span class="citation" data-cites="jin2020perfluoroalkyl stratakis2020prenatal matta2022associations">(<a href="#ref-jin2020perfluoroalkyl" role="doc-biblioref">Jin et al. 2020</a>; <a href="#ref-stratakis2020prenatal" role="doc-biblioref">Stratakis et al. 2020</a>; <a href="#ref-matta2022associations" role="doc-biblioref">Matta et al. 2022</a>)</span>.
In this paper, we introduce <strong>LUCIDus</strong> version 3, a major update and
enhancement from the original release <span class="citation" data-cites="LUCIDus">(<a href="#ref-LUCIDus" role="doc-biblioref">Jia et al. 2024</a>)</span>. To account for the
heterogeneity of the integrated datasets based on the study design,
<strong>LUCIDus</strong> version 3 incorporates three integration strategies into the
LUCID framework: (1) Early integration (Figure <a href="#fig:fig1">1</a> (a))
concatenates all multi-omics data matrices into a single matrix prior to
model estimation; (2) Intermediate integration (Figure <a href="#fig:fig1">1</a>
(b)) estimates latent clusters for each omic layer, and the resulting
clusters are then modeled in parallel in a joint model linking exposures
and the outcome; and (3) Late integration (Figure <a href="#fig:fig1">1</a> (c))
in which each omic layer is used to estimate omic-specific clusters that
are then linked via <span class="math inline">\(a priori\)</span> relationships with each other and the
exposure and outcome. <strong>LUCIDus</strong> version 3 also includes model
selection, model visualization, and inference based on bootstrap
resampling. It also incorporates an integrated imputation approach to
deal with missingness in multi-omics data. The paper is organized as
follows: we first briefly introduce the statistical model of LUCID; we
then go through the details of functions in the <strong>LUCIDus</strong> package; we
illustrate the workflow of fitting LUCID models by using a dataset
simulated based on real cases from the HELIX study
<span class="citation" data-cites="vrijheid2014human maitre2022state">(<a href="#ref-vrijheid2014human" role="doc-biblioref">Vrijheid et al. 2014</a>; <a href="#ref-maitre2022state" role="doc-biblioref">Maitre et al. 2022</a>)</span>.</p>
<h3 data-number="2" id="sec2"><span class="header-section-number">2</span> Model and software</h3>
<h4 class="unnumbered" data-number="2.1" id="sec2.1">Overview of the original LUCID model (early integration)</h4>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:fig1"></span>
<img src="figures/fig1.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 1: Three DAGs represent how three different types of the LUCID model integrate genetic/environmental exposures (<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>), other multi-omics data (<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>), and the phenotype trait (<span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span>). (a) LUCID early integration; (b) LUCID in parallel; (c) LUCID in serial. The squares represent observed data, the circles represent unobserved latent variables (clusters) and model parameters, and the diamond refers to <span class="math inline">\(L_1\)</span> penalty terms for regularization for (a). <span class="math inline">\(\boldsymbol{\mathbf{CoG}}\)</span> and <span class="math inline">\(\boldsymbol{\mathbf{CoY}}\)</span> represent covariates to be adjusted in the LUCID model. Missingness is allowed in multi-omics data. <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> is divided into subsets of observations with complete measurements and observations with missingness. For (b) and (c), <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> is partitioned into <span class="math inline">\(m\)</span> layers.
</p>
</div>
</div>
<p>In the LUCID model, genomic/exposomic exposures
<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>, other multi-omics data
<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, and phenotype trait <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span>
are integrated through a latent categorical variable
<span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span>. To model the complex correlation structure
between each layer of the multi-omics data, we propose three types of
LUCID models based on different integration strategies: (1) LUCID early
integration; (2) LUCID in parallel or intermediate integration; and (3)
LUCID in serial or late integration. Figure <a href="#fig:fig1">1</a> illustrates
the joint relationship between <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>,
<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span>, and
<span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> for each LUCID model. We will discuss LUCID in
parallel and LUCID in serial in detail under Sections
<a href="#sec_parallel">3.7</a> and <a href="#sec_serial">3.8</a> as two extensions, and here
we focus on LUCID early integration. Because <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span>
is an unobserved categorical variable, each category of
<span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> is interpreted as a latent cluster in the
data, jointly defined by <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>,
<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, and <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span>. Let
<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> be a <span class="math inline">\(N \times P\)</span> matrix with columns
representing genetic/environmental exposures and rows representing the
observations; <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> be a <span class="math inline">\(N \times M\)</span> matrix of
omics data (for example, gene expression data, DNA methylation profiles,
and metabolomic data, etc.) and <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> be a
<span class="math inline">\(N\)</span>-length vector of phenotype traits. We further assume
<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>, <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, and
<span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> are measured through a prospective sampling
procedure, so we do not model the distribution of
<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>. All three measured components
(<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>, <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, and
<span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span>) are linked by a latent variable
<span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> consisting of <span class="math inline">\(K\)</span> categories. The directed
acyclic graph (DAG) in Figure <a href="#fig:fig1">1</a> (a) implies that the
distributions of <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> given
<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>, <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> given
<span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span>, and <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> given
<span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> are conditionally independent of each other.
Let <span class="math inline">\(f(\cdot)\)</span> denote the probability mass functions (PMFs) for
categorical random variables or the probability density functions (PDFs)
for continuous random variables. The joint log-likelihood of the LUCID
model is constructed as:</p>
<p><span class="math display" id="eq:eq1">\[\begin{aligned}
         \log L(\boldsymbol{\mathbf{\Theta}}) &amp; = \sum_{i = 1}^N \log f(\boldsymbol{\mathbf{Z}}_i, Y_i|\boldsymbol{\mathbf{G}}_i;\boldsymbol{\mathbf{\Theta}}) \\
         &amp; = \sum_{i = 1}^N \log \sum_{j = 1}^K f(X_i = j| \boldsymbol{\mathbf{G}}_i; \boldsymbol{\mathbf{\Theta}}) f(\boldsymbol{\mathbf{Z}}_i| X_i = j; \boldsymbol{\mathbf{\Theta}}) f(Y_i|X_i = j; \boldsymbol{\mathbf{\Theta}})
    \end{aligned}
    \label{eq1}   \tag{1}\]</span></p>
<p>where <span class="math inline">\(\boldsymbol{\mathbf{\Theta}}\)</span> is a generic notation for all
parameters in the LUCID model.</p>
<p>Since <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> is a discrete variable with <span class="math inline">\(K\)</span>
categories, we assume <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> follows a multinomial
distribution conditioning on <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>, denoted by the
softmax function <span class="math inline">\(S(\cdot)\)</span>. We assume omics data
<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> follows a multivariate Gaussian distribution
conditioning on <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span>, denoted by
<span class="math inline">\(\phi(\boldsymbol{\mathbf{Z}}|\boldsymbol{\mathbf{X}} = j; \boldsymbol{\mathbf{\mu}}_j, \boldsymbol{\mathbf{\Sigma}}_j)\)</span>,
where <span class="math inline">\(\boldsymbol{\mathbf{\mu}}_j\)</span> and <span class="math inline">\(\boldsymbol{\mathbf{\Sigma}}_j\)</span>
are cluster-specific means and variance-covariance matrices for
<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>. For illustrative purposes, we assume
<span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> is a continuous outcome following a univariate
Gaussian distribution, denoted by
<span class="math inline">\(\phi(\boldsymbol{\mathbf{Y}}|\boldsymbol{\mathbf{X}} = j; \gamma_j, \sigma_j^2)\)</span>
(<span class="math inline">\(\gamma_j\)</span> and <span class="math inline">\(\sigma_j^2\)</span> are also interpreted as the
cluster-specific means and variances for the effect of
<span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> on <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span>). Similar
development for a binary outcome is detailed in <span class="citation" data-cites="peng2020latent">(<a href="#ref-peng2020latent" role="doc-biblioref">Peng et al. 2020</a>)</span>.
Because the latent cluster <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> is unobserved, the
maximum likelihood estimates (MLE) of the parameters associated with the
LUCID model based on Equation <a href="#eq:eq1">(1)</a> are not readily estimated.
Therefore, we use the Expectation-Maximization (EM) algorithm to obtain
the MLE of model parameters. We define <span class="math inline">\(I(X_i= j)\)</span> as an indicator
function representing that observation <span class="math inline">\(i\)</span> belongs to latent cluster
<span class="math inline">\(j\)</span>. Then, the log-likelihood function of model parameters in Equation
<a href="#eq:eq1">(1)</a> becomes:</p>
<p><span class="math display" id="eq:eq2">\[\begin{aligned}
         \log L(\boldsymbol{\mathbf{\Theta}}) &amp; = \sum_{i = 1}^N \sum_{j=1}^K I(X_i = j) \left( \log S(X_i = j| \boldsymbol{\mathbf{G}}_i; \boldsymbol{\mathbf{\Theta}}) + \log \phi(\boldsymbol{\mathbf{Z}}_i| X_i = j; \boldsymbol{\mathbf{\Theta}}) + \log \phi (Y_i|X_i = j; \boldsymbol{\mathbf{\Theta}}) \right)
    \end{aligned}
    \label{eq2}   \tag{2}\]</span></p>
<p>We denote the observed data as
<span class="math inline">\(\boldsymbol{\mathbf{D}} = \{\boldsymbol{\mathbf{G}}, \boldsymbol{\mathbf{Z}}, \boldsymbol{\mathbf{Y}}\}\)</span>,
and define the responsibility <span class="math inline">\(r\)</span> as the inclusion probability (IP) of
belonging to the latent cluster <span class="math inline">\(j\)</span> given observed data and current
estimation of model parameters at iteration <span class="math inline">\(t\)</span>, which is:</p>
<p><span class="math display" id="eq:eq3">\[\begin{aligned}
        r_{ij}^{(t)} &amp; = P(X_i = j|\boldsymbol{\mathbf{D}}; \boldsymbol{\mathbf{\Theta}}^{(t)}) \\
            &amp; = \frac{S\left(X_i = j| \boldsymbol{\mathbf{G}}_i; \boldsymbol{\mathbf{\beta}}^{(t)}\right) \phi\left(\boldsymbol{\mathbf{Z}}_i| X_i = j; \boldsymbol{\mathbf{\mu}}_j^{(t)}, \boldsymbol{\mathbf{\Sigma}}_j^{(t)}\right) \phi\left(Y_i|X_i = j; \gamma_j^{(t)}, \sigma^{2(t)}_j \right)}{\sum_{j = 1}^K S\left(X_i = j| \boldsymbol{\mathbf{G}}_i; \boldsymbol{\mathbf{\beta}}^{(t)}\right) \phi \left(\boldsymbol{\mathbf{Z}}_i| X_i = j; \boldsymbol{\mathbf{\mu}}_j^{(t)}, \boldsymbol{\mathbf{\Sigma}}_j^{(t)}\right) \phi \left(Y_i|X_i = j; \gamma_j^{(t)}, \sigma^{2(t)}_j\right)}
    \end{aligned}
    \label{eq3}   \tag{3}\]</span></p>
<p>At each iteration <span class="math inline">\(t\)</span>, in the E-step, we compute the expectation of the
complete data likelihood, which is:</p>
<p><span class="math display" id="eq:eq4">\[\begin{aligned}
        Q(\boldsymbol{\mathbf{\Theta}}|\boldsymbol{\mathbf{D}}, \boldsymbol{\mathbf{\Theta}}^{(t)}) = &amp; \sum_{i=1}^N\sum_{j=1}^K r_{ij} \log S\left(X_i = j|\boldsymbol{\mathbf{G}}_i;\boldsymbol{\mathbf{\beta}}\right) + \sum_{i=1}^N\sum_{j=1}^K r_{ij} \log \phi(\boldsymbol{\mathbf{Z}}_i|X_i = j;\boldsymbol{\mathbf{\mu}}_j, \boldsymbol{\mathbf{\Sigma}}_j) \\
        &amp; + \sum_{i=1}^N\sum_{j=1}^K r_{ij} \log \phi\left(Y|X_i = j; \gamma_j, \sigma_j^2\right)
    \end{aligned}
    \label{eq4}   \tag{4}\]</span></p>
<p>In the M-step, we update parameter estimates by maximizing Equation
<a href="#eq:eq4">(4)</a> in terms of <span class="math inline">\(\boldsymbol{\mathbf{\Theta}}\)</span>, which results
in the following:</p>
<p><span class="math display" id="eq:eq5">\[\begin{aligned}
    \boldsymbol{\mathbf{\beta}}^{(t+1)} &amp; = \arg \max_{\boldsymbol{\mathbf{\beta}}} \sum_{i=1}^N \sum_{j=1}^K r_{ij}^{(t)} \log S(X_i = j| \boldsymbol{\mathbf{G}}_i; \boldsymbol{\mathbf{\beta}}_j) \label{eq5} \\
    \boldsymbol{\mathbf{\mu}}_j^{(t+1)} &amp; = \frac{\sum_{i=1}^N r_{ij}^{(t)}\boldsymbol{\mathbf{Z}}_i}{\sum_{i=1}^N r_{ij}^{(t)}} \label{eq6} \\
    \boldsymbol{\mathbf{\Sigma}}_j^{(t+1)} &amp; = \frac{\sum_{i=1}^N r_{ij}^{(t)} \left(\boldsymbol{\mathbf{Z}}_i - \boldsymbol{\mathbf{\mu}}_j^{(t+1)} \right) \left(\boldsymbol{\mathbf{Z}}_i - \boldsymbol{\mathbf{\mu}}_j^{(t+1)} \right)^T}{\sum_{i=1}^N r_{ij}^{(t)}} \label{eq7} \\
    \gamma_j^{(t+1)} &amp; = \frac{\sum_{i=1}^N r_{ij}^{(t)}Y_i}{\sum_{i=1}^N r_{ij}^{(t)}} \label{eq8} \\
    \sigma_j^{2(t+1)} &amp; = \frac{\sum_{i=1}^N r_{ij}^{(t)} \left(Y_i - \gamma_j^{(t+1)} \right)^2}{\sum_{i=1}^N r_{ij}^{(t)}} \label{eq9}
\end{aligned}   \tag{5}\]</span></p>
<p>Although maximization of <span class="math inline">\(\boldsymbol{\mathbf{\beta}}_j^{(t+1)}\)</span> in
Equation <a href="#eq:eq5">(5)</a> does not have a closed-form solution, it is
equivalent to fitting a multinomial logistic regression by treating
<span class="math inline">\(r_{ij}\)</span> as the outcome and <span class="math inline">\(\boldsymbol{\mathbf{G}}_i\)</span> as the exposure.</p>
<p>For <strong>LUCIDus</strong>, we include a more flexible geometric feature of latent
clusters, such as volume, shape, and orientation determined by
<span class="math inline">\(\boldsymbol{\mathbf{\Sigma}}_j\)</span>. We use the parameterization of
covariance matrices by means of an eigenvalue decomposition in the form
below <span class="citation" data-cites="banfield1993model">(<a href="#ref-banfield1993model" role="doc-biblioref">Banfield and Raftery 1993</a>)</span>:</p>
<p><span class="math display" id="eq:eq10">\[\boldsymbol{\Sigma}_j = \lambda_j \boldsymbol{D}_j  \boldsymbol{A}_j \boldsymbol{D}_j^T \l\tag{6}   (#eq:eq10)\]</span></p>
<p>where <span class="math inline">\(\lambda_j\)</span> is a scalar, <span class="math inline">\(\boldsymbol{\mathbf{D}}_j\)</span> is the
orthogonal matrix of eigenvectors, and <span class="math inline">\(\boldsymbol{\mathbf{A}}_j\)</span> is a
diagonal matrix whose values are proportional to eigenvalues. A detailed
discussion of maximizing <span class="math inline">\(\boldsymbol{\mathbf{\Sigma}}_j\)</span> parameterized
by Equation <a href="#eq:eq10">(6)</a> is provided by <span class="citation" data-cites="celeux1995gaussian">Celeux and Govaert (<a href="#ref-celeux1995gaussian" role="doc-biblioref">1995</a>)</span>. Their
algorithm is implemented in the R package
<a href="https://CRAN.R-project.org/package=mclust"><strong>mclust</strong></a>
<span class="citation" data-cites="scrucca2016mclust">(<a href="#ref-scrucca2016mclust" role="doc-biblioref">Scrucca et al. 2016</a>)</span>, and we leverage <strong>mclust</strong> to update
<span class="math inline">\(\boldsymbol{\mathbf{\Sigma}}_j\)</span> in the M-step at each iteration of the
EM algorithm for LUCID.</p>
<div id="tbl1">
<div class="layout-chunk" data-layout="l-body">
<table>
<caption><span id="tab:table-1">Table 1: </span> Functions in the LUCIDus package. lucid() calls estimate_lucid() and tune_lucid() in the backend. The two workhorse functions are not normally called directly, but they can be useful when a user wants to examine the model fitting process in more detail.</caption>
<colgroup>
<col style="width: 4%" />
<col style="width: 95%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Main.function</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">lucid()</td>
<td style="text-align: left;">Main function to fit LUCID models, specified by giving integrated data, a distribution of the outcome and the type of the LUCID model (early, parallel, serial). It also conducts model selection and variable selection for LUCID early integration and model selection for candidate models with different numbers of latent clusters for LUCID in parallel and LUCID in serial.</td>
</tr>
<tr class="even">
<td style="text-align: left;">summary()</td>
<td style="text-align: left;">S3 method for LUCID. Create tables to summarize a LUCID model.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">plot()</td>
<td style="text-align: left;">S3 method for LUCID. Visualize LUCID models through a Sankey diagram.</td>
</tr>
<tr class="even">
<td style="text-align: left;">boot_lucid()</td>
<td style="text-align: left;">Derive confidence intervals based on bootstrap resampling.</td>
</tr>
<tr class="odd">
<td style="text-align: left;">predict_lucid()</td>
<td style="text-align: left;">Predict latent cluster assignment and outcome using integrated data.</td>
</tr>
<tr class="even">
<td style="text-align: left;">Workhorse function</td>
<td style="text-align: left;">Description</td>
</tr>
<tr class="odd">
<td style="text-align: left;">estimate_lucid()</td>
<td style="text-align: left;">Fit a LUCID model to estimate latent clusters by using integrated data.</td>
</tr>
<tr class="even">
<td style="text-align: left;">tune_lucid()</td>
<td style="text-align: left;">It fits a series of LUCID models over different combinations of tuning parameters and determines an optimal model with the minimum BIC.</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:fig2"></span>
<img src="figures/fig2.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 2: The workflow of the LUCIDus package. Dark blue nodes represent input data, light blue nodes represent output results. Green nodes and dashed arrows are optional steps for model estimation. Red texts correspond to 5 key functions in LUCIDus. Steps and functions marked with an asterisk currently work for LUCID early integration only.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="2.2" id="sec2.2">General workflow of <strong>LUCIDus</strong></h4>
<p>The <strong>LUCIDus</strong> package includes five main functions and two auxiliary
functions to implement the analysis framework based on LUCID. Brief
descriptions of each function are listed in Table <a href="#tbl1">1</a>. The
workflow of the <strong>LUCIDus</strong> package is shown in Figure <a href="#fig:fig2">2</a>.
Below we describe the typical workflow of analyzing integrated data
using the LUCID early integration model.</p>
<p>The function <code>lucid()</code> is the primary function in the package, which
fits a LUCID model based on an exposure matrix (argument <code>G</code>),
multi-omics data (argument <code>Z</code>), outcome data (argument <code>Y</code>), the number
of latent clusters (<code>K</code>; default is 2), the type of the LUCID model
(argument <code>lucid_model</code>; here it is <code>early</code> as we focus on the LUCID
early integration model in this section), and the family of the outcome
(argument <code>family</code>; default is <code>normal</code>). If a vector of <code>K</code> and/or
<span class="math inline">\(L_1\)</span> penalties are supplied, <code>lucid()</code> will automatically conduct model
selection on the number of clusters <span class="math inline">\(K\)</span>, select informative variables in
<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> or <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, or both, and
return a LUCID early integration model (an R object of class
<code>lucid_early</code>) with optimal <span class="math inline">\(K\)</span> and selected variables in
<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> and/or <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>. Several
additional functions can then be applied to a fitted LUCID object.
<code>summary()</code> summarizes the fitted LUCID model by producing summary
tables of parameter estimation and interpretation. Visualization is
performed via the <code>plot()</code> function to create a Sankey diagram showing
the interplay among the three components (<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>,
<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, and <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span>). In addition,
statistical inference can be accomplished for LUCID by constructing
confidence intervals (CIs) based on bootstrap resampling. This is
achieved by the function <code>boot_lucid()</code>. Finally, predictions on the
cluster assignment and the outcome can be obtained by calling the
function <code>predict_lucid()</code>. In practice, it might not be necessary to
implement the entire workflow above. For instance, if we have prior
knowledge of the number of latent clusters <span class="math inline">\(K\)</span>, model selection for the
number of clusters can be skipped. If a given dataset has limited
variables (for example, variables selected based on biological
annotations), then variable selection may not be necessary. Note that
for a LUCID in parallel model or a LUCID in serial model (the argument
<code>lucid_model</code> is <code>parallel</code> or <code>serial</code>, respectively), specification of
<code>K</code> is different, and the features of variable selection, constructing
confidence intervals (CIs) based on bootstrap resampling, and plotting
are not yet available (see Sections <a href="#sec_parallel">3.7</a> and
<a href="#sec_serial">3.8</a>).</p>
<p>In the following sections, we show example code and demonstrate the
usage of the LUCID early integration model using the simulated HELIX
data.</p>
<h3 data-number="3" id="sec3"><span class="header-section-number">3</span> Illustration</h3>
<h4 class="unnumbered" data-number="3.1" id="sec3.1">Fitting a LUCID model by <code>lucid()</code></h4>
<p>To illustrate integrative clustering with the LUCID early integration
model, we use simulated data based on real cases from the HELIX study
based on the correlation structure. A subset of the simulated data is
incorporated in the <strong>LUCIDus</strong> package to replicate the workflow
presented here. The dataset is a list of <code>dataframe</code> containing data
from 420 children, with 1 variable measuring maternal exposure to utero
mercury (referred to as the exposome), 10 methylomes measured in
children (referred to as the methylomics), 10 transcriptomes measured in
children (referred to as the transcriptomics), 10 miRNA measured in
children (referred to as the miRNA), childhood cytokeratin 18 level
(referred to as ck18, a continuous outcome as an indicator of
metabolic-dysfunction-associated fatty liver disease (MAFLD), childhood
cytokeratin 18 category (a binary outcome referred to as ck18_cat) and 2
covariates, including child’s sex and child’s age. To organize the data
into separate <code>dataframe</code>s for each type of data and to better
illustrate functionality within <strong>LUCIDus</strong>, we use the following code:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(LUCIDus)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># load data</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="fu">data</span>(<span class="st">&quot;simulated_HELIX_data&quot;</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> simulated_data <span class="ot">&lt;-</span> simulated_HELIX_data</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> exposome <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(simulated_data[[<span class="st">&quot;phenotype&quot;</span>]]<span class="sc">$</span>hs_hg_m_scaled)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">colnames</span>(exposome) <span class="ot">&lt;-</span> <span class="st">&quot;hs_hg_m_scaled&quot;</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> methylomics <span class="ot">&lt;-</span> simulated_data<span class="sc">$</span>methylome</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> ck18 <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(simulated_data[[<span class="st">&quot;phenotype&quot;</span>]]<span class="sc">$</span>ck18_scaled)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">colnames</span>(ck18) <span class="ot">&lt;-</span> <span class="st">&quot;ck18_scaled&quot;</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> ck18_cat <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(ck18 <span class="sc">&gt;</span> <span class="fu">mean</span>(ck18), <span class="dv">1</span>, <span class="dv">0</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> covars <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;hs_child_age_yrs_None&quot;</span>,<span class="st">&quot;e3_sex_None&quot;</span>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> covs <span class="ot">&lt;-</span> simulated_data[[<span class="st">&quot;phenotype&quot;</span>]][covars]</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> covs<span class="sc">$</span>e3_sex_None <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(covs<span class="sc">$</span>e3_sex_None <span class="sc">==</span> <span class="st">&quot;male&quot;</span>, <span class="dv">1</span>, <span class="dv">0</span>)</span></code></pre></div>
<p>Our goal is to conduct an integrated clustering of the exposome and
methylome and to relate the estimated latent clusters to the childhood
cytokeratin 18 level. The LUCID model is fitted using the function
<code>lucid()</code>. The input data are specified by arguments <code>G</code>, <code>Z</code>, and <code>Y</code>,
corresponding to the exposome, the methylome, and the childhood
cytokeratin 18 level, respectively. In practice, scaling of multi-omics
data is highly recommended to obtain more stable estimates; the
methylomics data used in this example are already scaled. If multi-omics
data are included in the LUCID analysis as the omic intermediate
<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, the user can concatenate them one by one into
a single data matrix and then input the concatenated matrix as
<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> for early integration. For illustrative
purposes, we assume that, in the example data, the optimal number of
latent clusters is 2 (determining the optimal number of clusters is a
major question prior to performing clustering analysis; more details are
discussed in the later Section <a href="#sec3.3">3.3</a>).</p>
<p>For illustration, we fit two LUCID models with continuous outcome ck18
and binary outcome ck18_cat, respectively. The parameter <code>family</code>
specifies the outcome type. The default setting is <code>normal</code>,
corresponding to a continuous outcome. For a binary outcome, <code>family</code>
should be specified as <code>binary</code>. <code>lucid()</code> returns an object of class
<code>lucid_early</code> providing the MLE of the LUCID model as well as IPs.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Fit a LUCID model with a continuous outcome</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> fit1 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics, <span class="at">Y =</span> ck18, <span class="at">init_omic.data.model =</span> <span class="cn">NULL</span>,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>               <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>, <span class="at">K =</span> <span class="dv">2</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># MLE of the LUCID model</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># fit1$res_Beta</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># fit1$res_Mu</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># fit1$res_Sigma</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># fit1$res_Gamma</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># IPs of the sample</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># fit1$inclusion.p</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># Fit LUCID model with a binary outcome</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> fit2 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics, <span class="at">Y =</span> ck18_cat, <span class="at">init_omic.data.model =</span> <span class="cn">NULL</span>,</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>               <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, <span class="at">family =</span> <span class="st">&quot;binary&quot;</span>, <span class="at">K =</span> <span class="dv">2</span>)</span></code></pre></div>
<h5 class="unnumbered" data-number="3.1.1" id="sec3.1.1">Initializing the parameters of LUCID</h5>
<p><span class="math inline">\(\boldsymbol{\mathbf{\beta}}\)</span> is initiated by randomly drawing from a
uniform distribution. By default, <code>init_par</code> is <code>mclust</code> and
<code>init_omic.data.model</code> is <code>EEV</code>, so <code>lucid()</code> calls <code>Mclust()</code> to fit
the mixture model of <code>EEV</code> on omics data <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> and
use the mixture model estimates to initiate <span class="math inline">\(\boldsymbol{\mathbf{\mu}}\)</span>
and <span class="math inline">\(\boldsymbol{\mathbf{\Sigma}}\)</span>, and then implement regression to
initiate <span class="math inline">\(\boldsymbol{\mathbf{\gamma}}\)</span>. Alternatively, the user can
also specify <code>init_omic.data.model</code> to choose a certain mixture model.
The available mixture models are listed in <code>mclust::mclustModelNames</code>
<span class="citation" data-cites="scrucca2016mclust">(<a href="#ref-scrucca2016mclust" role="doc-biblioref">Scrucca et al. 2016</a>)</span>. Note that <code>init_omic.data.model</code> can be <code>NULL</code> to
let <code>Mclust()</code> automatically choose the optimal mixture model. If not
using <code>Mclust()</code> for initiation, the user can also specify <code>init_par</code> to
be <code>random</code> to initiate <span class="math inline">\(\boldsymbol{\mathbf{\mu}}\)</span> and
<span class="math inline">\(\boldsymbol{\mathbf{\Sigma}}\)</span> by randomly drawing them from a uniform
distribution and then implement regression to initiate
<span class="math inline">\(\boldsymbol{\mathbf{\gamma}}\)</span>. We recommend using the default setting
to initiate the parameters by calling <code>Mclust()</code> for quick convergence
and stable performance.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Fit LUCID model with spherical shape, equal volume</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> fit3<span class="fl">.1</span> <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics, <span class="at">Y =</span> ck18, </span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                 <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>, <span class="at">K =</span> <span class="dv">2</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                 <span class="at">init_par =</span> <span class="st">&quot;mclust&quot;</span>,<span class="at">init_omic.data.model =</span> <span class="st">&quot;EII&quot;</span>)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Fit LUCID model with random guess</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> fit3<span class="fl">.2</span> <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics, <span class="at">Y =</span> ck18, </span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                 <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>, <span class="at">K =</span> <span class="dv">2</span>,</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                 <span class="at">init_par =</span> <span class="st">&quot;random&quot;</span>)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Fit LUCID model with ellipsoidal shape, varying volume, shape, and orientation </span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> fit4 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics, <span class="at">Y =</span> ck18, </span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>               <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>, <span class="at">K =</span> <span class="dv">2</span>,</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>               <span class="at">init_omic.data.model =</span> <span class="st">&quot;VVV&quot;</span>)</span></code></pre></div>
<h5 class="unnumbered" data-number="3.1.2" id="sec3.1.2">Supervised LUCID versus unsupervised LUCID</h5>
<p>In Equation <a href="#eq:eq1">(1)</a>, the latent clusters <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span>
are jointly defined by genomic/environmental exposure
<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>, other multi-omics data
<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, and outcome <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span>.
Because the outcome is explicitly incorporated in the likelihood
function, the estimation process based on Equation <a href="#eq:eq1">(1)</a> is
similar to a supervised learning process. <strong>LUCIDus</strong> also allows for an
unsupervised version of LUCID. In this unsupervised LUCID, the latent
clusters are estimated only by <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> and
<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>. Specifically, the joint likelihood of
unsupervised LUCID is written as</p>
<p><span class="math display" id="eq:eq11">\[\begin{aligned}
         \log L(\boldsymbol{\mathbf{\Theta}}) &amp; = \sum_{i = 1}^N \sum_{j=1}^K I(X_i = j) \left( \log S(X_i = j| \boldsymbol{\mathbf{G}}_i; \boldsymbol{\mathbf{\Theta}}) + \log \phi(\boldsymbol{\mathbf{Z}}_i| X_i = j; \boldsymbol{\mathbf{\Theta}})\right)
    \end{aligned}
    \label{eq11}   \tag{7}\]</span></p>
<p>and the corresponding responsibility based on Equation <a href="#eq:eq11">(7)</a> is
derived as</p>
<p><span class="math display" id="eq:eq12">\[\begin{aligned}
        r_{ij}^{(t)} &amp; = \frac{S\left(X_i = j| \boldsymbol{\mathbf{G}}_i; \boldsymbol{\mathbf{\beta}}^{(t)}\right) \phi\left(\boldsymbol{\mathbf{Z}}_i| X_i = j; \boldsymbol{\mathbf{\mu}}_j^{(t)}, \boldsymbol{\mathbf{\Sigma}}_j^{(t)}\right)}{\sum_{j = 1}^K S\left(X_i = j| \boldsymbol{\mathbf{G}}_i; \boldsymbol{\mathbf{\beta}}^{(t)}\right) \phi\left(\boldsymbol{\mathbf{Z}}_i| X_i = j; \boldsymbol{\mathbf{\mu}}_j^{(t)}, \boldsymbol{\mathbf{\Sigma}}_j^{(t)}\right)}
    \end{aligned}
    \label{eq12}   \tag{8}\]</span></p>
<p>Estimations for unsupervised LUCID are also obtained by the EM algorithm
discussed previously. The parameter <code>useY</code> in <code>lucid()</code> is a flag
indicating a supervised or unsupervised LUCID model. By default,
<code>useY = TRUE</code>, and <code>lucid()</code> fits supervised LUCID. To fit an
unsupervised LUCID model, a user should set <code>useY = FALSE</code>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Fit an unsupervised LUCID model</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> fit5 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics, <span class="at">Y =</span> ck18, <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>               <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>, <span class="at">K =</span> <span class="dv">2</span>, <span class="at">useY =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>The choice of a supervised LUCID or unsupervised LUCID model depends
both on the study design and the research question. A supervised LUCID
analysis may be appropriate, for example, if: (1) there is the belief
that the cluster structure is defined jointly by
<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>, <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, and
<span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> (for example, data collected from a
cross-sectional design); or (2) the goal is to build a predictive model
with data from one cohort to then apply directly to another cohort. If
the aim is to obtain an unbiased estimate for the association between
the latent cluster <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> and the outcome
<span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span>, an unsupervised LUCID model is more
appropriate.</p>
<h5 class="unnumbered" data-number="3.1.3" id="sec3.1.3">Adjusting for covariates</h5>
<p><strong>LUCIDus</strong> also allows for the adjustment of covariates. According to
the DAG in Figure <a href="#fig:fig1">1</a> (a), covariates may be included for
the association between the exposure and the latent cluster (referred to
as <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> to <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> covariate) or
for the association between the latent cluster and the outcome (referred
to as <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> to <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> covariate).
Covariates in the <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> to <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span>
relationship act more like predictors of the latent cluster while
covariates in the <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> to <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span>
relationship may be interpreted more in the context of an adjustment for
a potential confounding effect. A variable can serve as both a
<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> to <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> covariate and a
<span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> to <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> covariate. In the
<code>lucid()</code> function, <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> to
<span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> covariates are specified by parameter <code>CoG</code>
and <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> to <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> covariates
are specified by parameter <code>CoY</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Include covariates as G to X covariates</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> fit6 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics, <span class="at">Y =</span> ck18, <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, <span class="at">K =</span> <span class="dv">2</span>, </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>               <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>, <span class="at">CoG =</span> covs)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Include covariates as X to Y covariates</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> fit7 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics, <span class="at">Y =</span> ck18, <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, <span class="at">K =</span> <span class="dv">2</span>, </span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>               <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>, <span class="at">CoY =</span> covs)</span></code></pre></div>
<h4 class="unnumbered" data-number="3.2" id="section3.2">Interpreting LUCID</h4>
<p>Since LUCID is an integrated analysis framework consisting of
clustering, regression, and multiple data components, we provide two
utility functions, <code>summary()</code> and <code>plot()</code> to summarize the results of
LUCID and to facilitate interpretation.</p>
<h5 class="unnumbered" data-number="3.2.1" id="summarizing-lucid-in-tables-by-summary">Summarizing LUCID in tables by <code>summary()</code></h5>
<p>A direct call of <code>summary()</code> with the input of a model returned by
<code>lucid()</code> prints three tables in the console, corresponding to
associations among <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>, <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>,
and <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span>. We take the LUCID model fitted with the
continuous BMI as an example.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># summarize a simple lucid model with a continuous outcome</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="fu">summary</span>(fit1)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="sc">----------</span>Summary of the LUCID Early Integration model<span class="sc">----------</span> </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>K <span class="ot">=</span>  <span class="dv">2</span> , log likelihood <span class="ot">=</span> <span class="sc">-</span><span class="fl">6721.801</span> , BIC <span class="ot">=</span>  <span class="fl">14808.7</span> </span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>(<span class="dv">1</span>) <span class="fu">Y</span> (continuous outcome)<span class="sc">:</span> effect size of Y <span class="cf">for</span> each latent cluster</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>             Gamma</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>cluster1 <span class="fl">0.0000000</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>cluster2 <span class="fl">0.5040367</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>(<span class="dv">2</span>) Z<span class="sc">:</span> mean of omics data <span class="cf">for</span> each latent cluster </span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>              mu_cluster1 mu_cluster2</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>cg_GRHL3        <span class="fl">0.2452019</span>  <span class="sc">-</span><span class="fl">0.3627827</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>cg_BTF3L4       <span class="fl">0.1863510</span>  <span class="sc">-</span><span class="fl">0.2798288</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>cg_AL358472<span class="fl">.7</span>   <span class="fl">0.1954543</span>  <span class="sc">-</span><span class="fl">0.3290283</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>cg_HDGF         <span class="fl">0.1897587</span>  <span class="sc">-</span><span class="fl">0.2614602</span></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>cg_TDRD5        <span class="fl">0.1977435</span>  <span class="sc">-</span><span class="fl">0.2186209</span></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>cg_CSRNP3       <span class="fl">0.1742812</span>  <span class="sc">-</span><span class="fl">0.1794667</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>cg_HSPD1        <span class="fl">0.2829609</span>  <span class="sc">-</span><span class="fl">0.3282916</span></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>cg_EPM2AIP1    <span class="sc">-</span><span class="fl">0.1565522</span>   <span class="fl">0.3031461</span></span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>cg_AC025171<span class="fl">.1</span>   <span class="fl">0.2769363</span>  <span class="sc">-</span><span class="fl">0.2656729</span></span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>cg_VTRNA1_3    <span class="sc">-</span><span class="fl">0.2250234</span>   <span class="fl">0.1847493</span></span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>(<span class="dv">3</span>) E<span class="sc">:</span> odds ratio of being assigned to each latent cluster <span class="cf">for</span> each exposure </span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>                             beta       OR</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a>hs_hg_m_scaled.cluster2 <span class="fl">0.7909168</span> <span class="fl">2.205417</span></span></code></pre></div>
<p>The first table summarizes the association between the latent cluster
and the outcome (cytokeratin 18). The estimated effect size of
cytokeratin 18 for cluster 1 is 0 since it is the reference cluster
while that for cluster 2 is 0.504. In the case of a binary outcome, the
coefficient in the first table is interpreted as the log odds for the
variable for that specific cluster vs. the reference cluster. The second
table characterizes each cluster by its omics signature. The omics
signature is a cluster-specific mean for each variable in the omics
data. For instance, latent cluster 1 is characterized by low levels of
cg_EPM2AIP1 (<span class="math inline">\(\mu_1 (\text{cg\_EPM2AIP1})= -0.157\)</span>) and cg_VTRNA1_3
(<span class="math inline">\(\mu_1 (\text{cg\_VTRNA1\_3})= -0.225\)</span>), and high levels of all other
omic features; for example, cg_GRHL3
(<span class="math inline">\(\mu_1 (\text{cg{\_}GRHL3})= 0.2452019\)</span>). On the contrary, latent
cluster 2 features high levels of cg_EPM2AIP1
(<span class="math inline">\(\mu_2 (\text{cg{\_}EPM2AIP1})=  0.303\)</span>) and cg_VTRNA1_3
(<span class="math inline">\(\mu_2 (\text{cg{\_}VTRNA1{\_}3})= 0.185\)</span>), and low levels of all other
features including cg_GRHL3 (<span class="math inline">\(\mu_2 (\text{cg{\_}GRHL3})= -0.363\)</span>). The
third table relates the exposures to the latent clusters. For instance,
<code>hs``_``hg``_``m``_``scaled</code> represents the scaled level of maternal
mercury exposure. The coefficient OR for <code>hs``_``hg``_``m``_``scaled</code> is
0.791, meaning that for each doubling of the scaled level of maternal
mercury exposure, the odds ratio of being assigned to latent cluster 2
is 2.205. Since latent cluster 2 is associated with a higher child level
of cytokeratin 18, these results indicate that exposure to mercury is
associated with a higher child level of cytokeratin 18 and thus higher
risks of MAFLD.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:fig3"></span>
<img src="figures/fig3.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 3: An example of using the Sankey diagram to visualize a LUCID model. The dark grey nodes represent the exposure, the light grey node represents the outcome, the blue nodes are omics data, and the orange nodes are latent clusters. The width of links and nodes corresponds to effect size. Light-colored links represent negative associations while dark-colored links indicate positive associations.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="3.2.2" id="sec3.2.2">Visualizing LUCID by <code>plot()</code></h5>
<p>Visualization is another imperative way to interpret statistical models.
Here we use a Sankey diagram <span class="citation" data-cites="schmidt2008sankey">(<a href="#ref-schmidt2008sankey" role="doc-biblioref">Schmidt 2008</a>)</span> for the visualization
of the relationships among different components in LUCID. <code>plot()</code> takes
a fitted LUCID model as input and creates a Sankey diagram in <code>html</code>
format. It also accepts a user-defined color palette.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Visualize LUCID model via a Sankey diagram</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="fu">plot</span>(fit1)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Change the node color</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="fu">plot</span>(fit1, <span class="at">G_color =</span> <span class="st">&quot;yellow&quot;</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Change the link color</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="fu">plot</span>(fit1, <span class="at">pos_link_color =</span> <span class="st">&quot;red&quot;</span>, <span class="at">neg_link_color =</span> <span class="st">&quot;green&quot;</span>)</span></code></pre></div>
<p>Figure <a href="#fig:fig3">3</a> shows an example of a Sankey diagram for the
example data. Each node represents a component in the LUCID model.
Different components are labeled by user-defined colors. Each link
represents the statistical association between two nodes. The width of
the links reflects the effect size of the association, and the color of
the links indicates the direction of the association. By default,
dark-colored links represent positive associations while light-colored
links indicate negative associations. The output of the Sankey diagram
is an interactive html file in which the user can drag and rearrange the
layout of the elements to avoid overlapping. In practice, we recommend
limiting the number of variables in the Sankey diagram to approximately
10 to facilitate interpretation.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:fig4"></span>
<img src="figures/fig4.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 4: Choosing an optimal number of latent clusters K based on BIC.
</p>
</div>
</div>
<h4 class="unnumbered" data-number="3.3" id="sec3.3">Model selection and variable selection</h4>
<p>To determine the number of latent clusters, <span class="math inline">\(K\)</span>, LUCID implements model
selection by fitting a series of models, each with a different value for
<span class="math inline">\(K\)</span>, and then the Bayesian Information Criteria (BIC) is used to choose
the optimal model <span class="citation" data-cites="fan2013tuning">(<a href="#ref-fan2013tuning" role="doc-biblioref">Fan and Tang 2013</a>)</span>. The optimal model is the one with
the lowest BIC. The BIC for LUCID is defined as</p>
<p><span class="math display">\[\text{BIC} = -2 \log L(\hat{\boldsymbol{\mathbf{\Theta}}}) + D \log N\]</span></p>
<p>where <span class="math inline">\(\hat{\boldsymbol{\mathbf{\Theta}}}\)</span> is the maximum likelihood
estimation and <span class="math inline">\(D\)</span> is the number of parameters in LUCID. Specifically,
<span class="math inline">\(D = P(K - 1) + KM + KM(M + 1)/2 + n_Y\)</span>, where <span class="math inline">\(n_Y\)</span> is the number of
parameters dependent upon the type of outcome <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span>.
If <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> is a continuous outcome, then <span class="math inline">\(n_Y = 2K\)</span>.
If <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> is a binary outcome, then <span class="math inline">\(n_Y = K\)</span>. If a
vector of <span class="math inline">\(K\)</span> is used as input to the function <code>lucid()</code>, model
selection is automatically performed and returns the model with optimal
<span class="math inline">\(K\)</span>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> fit8 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics, <span class="at">Y =</span> ck18, <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>               <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>, <span class="at">K =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">6</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Check the optimal K</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> fit8<span class="sc">$</span>K</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="dv">2</span></span></code></pre></div>
<p>Separately, users can use the auxiliary function <code>tune_lucid()</code> to
explore model selection in more detail. This function returns the
optimal model as well as a table recording the tuning process. Below is
an example of tuning <span class="math inline">\(K\)</span> and visualizing the tuning process by using
<code>tune_lucid()</code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Look into the tuning process in more detail</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> tune_K <span class="ot">&lt;-</span> <span class="fu">tune_lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics, <span class="at">Y =</span> ck18, <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>,</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                      <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>, <span class="at">K =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">6</span>)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> fit9 <span class="ot">&lt;-</span> tune_K<span class="sc">$</span>best_model</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">ggplot</span>(<span class="at">data =</span> tune_K<span class="sc">$</span>tune_list, <span class="fu">aes</span>(<span class="at">x =</span> K, <span class="at">y =</span> BIC, <span class="at">label =</span> <span class="fu">round</span>(BIC, <span class="dv">0</span>))) <span class="sc">+</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>   <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>   <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>   <span class="fu">geom_text</span>(<span class="at">hjust =</span> <span class="sc">-</span><span class="fl">0.2</span>)</span></code></pre></div>
<p>Figure <a href="#fig:fig4">4</a> shows the tuning process with <span class="math inline">\(K = 2\)</span> resulting
in the lowest BIC.</p>
<p>Including additional or redundant variables in clustering models may
increase model complexity, impair prediction accuracy, and boost
computational time <span class="citation" data-cites="fop2018variable">(<a href="#ref-fop2018variable" role="doc-biblioref">Fop and Murphy 2018</a>)</span>. <strong>LUCIDus</strong> performs variable
selection for both the genetic/environmental exposures
<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> and for the other multi-omics data
<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> via <span class="math inline">\(L_1\)</span>-norm penalty terms. For variable
selection for <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>, we apply the LASSO regression
to obtain sparse solutions for Equation <a href="#eq:eq5">(5)</a>, which is</p>
<p><span class="math display" id="eq:eq14">\[\boldsymbol{\mathbf{\beta}}_{\text{LASSO}}^{(t+1)} = \arg \max_{\boldsymbol{\mathbf{\beta}}} \left\{ \sum_{i = 1}^N \sum_{j = 1}^K r_{ij}^{(t)}\log S(X_i = j | \boldsymbol{\mathbf{G}}_i; \boldsymbol{\mathbf{\beta}}_j) - \lambda_{\boldsymbol{\mathbf{\beta}}}\sum_{j = 1}^K\sum_{l = 1}^P |\beta_{jl}| \right\}
     \label{eq14}   \tag{9}\]</span></p>
<p>To obtain sparse estimation for <span class="math inline">\(\boldsymbol{\mathbf{\mu}}\)</span> and
<span class="math inline">\(\boldsymbol{\mathbf{\Sigma}}\)</span>, we implement the penalized model-based
method <span class="citation" data-cites="zhou2009penalized">(<a href="#ref-zhou2009penalized" role="doc-biblioref">Zhou et al. 2009</a>)</span>. <span class="math inline">\(\boldsymbol{\mathbf{\mu}}\)</span> is first
updated by maximizing the following equation,</p>
<p><span class="math display" id="eq:eq15">\[\boldsymbol{\mathbf{\mu}}_j^{(t+1)} = \arg \max_{\boldsymbol{\mathbf{\mu}}} \left \{ \sum_{i = 1}^N \sum_{j = 1}^K r_{ij}^{(t)}\log \phi(\boldsymbol{\mathbf{Z}}_i| \boldsymbol{\mathbf{\mu_j}}, \boldsymbol{\mathbf{\Sigma_j}}) - \lambda_{\boldsymbol{\mathbf{\mu}}}\sum_{j = 1}^K\sum_{l = 1}^M |\mu_{jl}| \right \}
    \label{eq15}   \tag{10}\]</span></p>
<p>Denote <span class="math inline">\(\boldsymbol{\mathbf{W}} = \boldsymbol{\mathbf{\Sigma}}^{-1}\)</span>.
Then <span class="math inline">\(\boldsymbol{\mathbf{\Sigma}}\)</span> is updated by maximizing its inverse
penalized by <span class="math inline">\(L_1\)</span>-norm</p>
<p><span class="math display" id="eq:eq16">\[\boldsymbol{\mathbf{W}}_j^{(t+1)} = \arg \max_{\boldsymbol{\mathbf{W}}} \left \{ \sum_{i = 1}^N \sum_{j = 1}^K r_{ij}^{(t)}\left( \det \boldsymbol{\mathbf{W}}_j - \text{trace}(\boldsymbol{\mathbf{S}}_j^{(t)} \boldsymbol{\mathbf{W}}_j) \right) - \lambda_{\boldsymbol{\mathbf{W}}}\sum_{ls}|w_{jls}| \right \}
    \label{eq16}   \tag{11}\]</span></p>
<p>where <span class="math inline">\(\boldsymbol{\mathbf{S}}_j\)</span> is the empirical covariance matrix at
each iteration, defined as</p>
<p><span class="math display" id="eq:eq17">\[\boldsymbol{\mathbf{S}}_j^{(t)} = \frac{\sum_{i=1}^N r_{ij}\left(\boldsymbol{\mathbf{Z}}_i - \boldsymbol{\mathbf{\mu}}_j^{(t)}\right)\left(\boldsymbol{\mathbf{Z}}_i - \boldsymbol{\mathbf{\mu}}_j^{(t)}\right)^T}{\sum_{i=1}^N r_{ij}^{(t)}}
    \label{eq17}   \tag{12}\]</span></p>
<p>To choose the optimal combination of the three <span class="math inline">\(L_1\)</span>-norm penalties, we
implement a grid search by adopting a modified BIC <span class="citation" data-cites="pan2007penalized">(<a href="#ref-pan2007penalized" role="doc-biblioref">Pan and Shen 2007</a>)</span>,
defined as</p>
<p><span class="math display" id="eq:eq18">\[\text{BIC}_p = -2 \log L(\hat{\boldsymbol{\mathbf{\Theta}}}) + (D - D_{\boldsymbol{\mathbf{G}}} - D_{\boldsymbol{\mathbf{Z}}})\log N
    \label{eq18}   \tag{13}\]</span></p>
<p>Where <span class="math inline">\(D_{\boldsymbol{\mathbf{G}}}\)</span> is the number of exposure variables
whose effect estimates are 0 across all latent clusters and
<span class="math inline">\(D_{\boldsymbol{\mathbf{Z}}}\)</span> is the number of omic variables whose
effect estimates are 0 across all latent clusters. We can tune
<span class="math inline">\(\lambda_{\boldsymbol{\mathbf{\beta}}}\)</span>,
<span class="math inline">\(\lambda_{\boldsymbol{\mathbf{\mu}}}\)</span>, and
<span class="math inline">\(\lambda_{\boldsymbol{\mathbf{W}}}\)</span> either separately or jointly. For
instance, if only exposome data <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> is
high-dimensional, we only need to tune
<span class="math inline">\(\boldsymbol{\mathbf{\lambda}}_{\boldsymbol{\mathbf{\beta}}}\)</span>. If
variable selection is desired for both <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> and
<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, then the three <span class="math inline">\(L_1\)</span>-norm penalties should be
tuned simultaneously. <span class="math inline">\(\lambda_{\boldsymbol{\mathbf{\beta}}}\)</span>,
<span class="math inline">\(\lambda_{\boldsymbol{\mathbf{\mu}}}\)</span>, and
<span class="math inline">\(\lambda_{\boldsymbol{\mathbf{W}}}\)</span> match the parameters <code>Rho_G</code>,
<code>Rho_Z_Mu</code>, and <code>Rho_Z_Cov</code> in the <code>lucid()</code> function. Each parameter
accepts a numeric vector or a scalar as input. For guidance, empirical
experiments utilizing normalized multi-omics data suggest integer values
in the range of <span class="math inline">\(0 - 100\)</span> for <code>Rho_Z_Mu</code> and values in the range of
<span class="math inline">\(0 - 1\)</span> for <code>Rho_G</code> and <code>Rho_Z_Cov</code>. The higher the values of
<code>Rho_Z_Mu</code>, <code>Rho_G</code>, and <code>Rho_Z_Cov</code>, the fewer omic features that will
be selected. Below are examples for conducting variable selection for
<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> and <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, separately and
jointly.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Variable selection for G</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># Add 10 more noise variables to the exposome</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> noise <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(<span class="dv">420</span> <span class="sc">*</span> <span class="dv">10</span>), <span class="at">nrow =</span> <span class="dv">420</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> exposome_noise <span class="ot">&lt;-</span> <span class="fu">cbind</span>(exposome, noise)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> fit10 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome_noise, <span class="at">Z =</span> methylomics, <span class="at">Y =</span> ck18,</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                 <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>, <span class="at">K =</span> <span class="dv">2</span>,</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                 <span class="at">Rho_G =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="fl">0.4</span>, <span class="at">by =</span> <span class="fl">0.01</span>), <span class="at">seed =</span> <span class="dv">1008</span>)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span><span class="sc">/</span><span class="dv">11</span> exposures are selected</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Summary of optimal lucid model</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># summary(fit10)</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># Variable selection for Z</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># add 10 more noise variables to the methylomics</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> methylomics_noise <span class="ot">&lt;-</span> <span class="fu">cbind</span>(methylomics, noise)</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> fit11 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics_noise, <span class="at">Y =</span> ck18,</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>, <span class="at">K =</span> <span class="dv">2</span>,</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">Rho_Z_Mu =</span> <span class="dv">5</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">Rho_Z_Cov =</span> <span class="fu">seq</span>(<span class="fl">0.1</span>, <span class="fl">0.4</span>, <span class="at">by =</span> <span class="fl">0.1</span>), <span class="at">seed =</span> <span class="dv">1008</span>)</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span><span class="sc">/</span><span class="dv">20</span> omics variables are selected</span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Summary of optimal lucid model</span></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># summary(fit11)</span></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># Variable selection for G and Z jointly</span></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> fit12 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome_noise, <span class="at">Z =</span> methylomics_noise, <span class="at">Y =</span> ck18,</span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>, <span class="at">K =</span> <span class="dv">2</span>,</span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">Rho_G =</span> <span class="fl">0.01</span>, <span class="at">Rho_Z_Mu =</span> <span class="dv">10</span>, <span class="at">Rho_Z_Cov =</span> <span class="fl">0.5</span>, <span class="at">seed =</span> <span class="dv">123</span>)</span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span><span class="sc">/</span><span class="dv">11</span> exposures are selected</span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a><span class="dv">11</span><span class="sc">/</span><span class="dv">20</span> omics variables are selected</span></code></pre></div>
<p>Oftentimes, the number of features <span class="math inline">\(M\)</span> for omics data
<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> is in the hundreds or even thousands. With a
high dimensionality of <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, LUCID can still have
stable performance and select the features with effect if the number of
observations <span class="math inline">\(N\)</span> is higher than <span class="math inline">\(M\)</span>. Below is an example of conducting
variable selection in <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> with <span class="math inline">\(M\)</span> = 210 features
(including 200 noise features) while <span class="math inline">\(N\)</span> = 420. We can see that with
higher <code>Rho_Z_Mu</code> and <code>Rho_Z_Cov</code>, LUCID successfully selects the
features with effect.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Variable selection for Z (high number of features M)</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="fu">set.seed</span>(<span class="dv">1008</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Add 200 more noise features to the methylomics</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> noise <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(<span class="dv">420</span> <span class="sc">*</span> <span class="dv">200</span>), <span class="at">nrow =</span> <span class="dv">420</span>)</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> methylomics_noise <span class="ot">&lt;-</span> <span class="fu">cbind</span>(methylomics, noise)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> fit13 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics_noise, <span class="at">Y =</span> ck18,</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>, <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, <span class="at">K =</span> <span class="dv">2</span>,</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">Rho_Z_Mu =</span> <span class="dv">20</span>, <span class="at">Rho_Z_Cov =</span> <span class="fl">0.6</span>)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="dv">17</span><span class="sc">/</span><span class="dv">210</span> omics variables are selected</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Summary of optimal lucid model</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># summary(fit13)</span></span></code></pre></div>
<p>Via simulations, the run time for LUCIDus increases linearly as the
number of features increases. While the integration of regularization
allows for a high number of features to be analyzed, when <span class="math inline">\(M\)</span> is higher
than <span class="math inline">\(N\)</span>, LUCID might result in unstable estimation; therefore,
pre-screening approaches for the omic features such as the
"meet-in-the-middle" become imperative to reduce dimensionality before
fitting the LUCID model <span class="citation" data-cites="CADIOU2021106509">(<a href="#ref-CADIOU2021106509" role="doc-biblioref">Cadiou et al. 2021</a>)</span>.</p>
<h4 class="unnumbered" data-number="3.4" id="sec3.4">Deriving confidence intervals</h4>
<p>We use the nonparametric bootstrap approach to derive confidence
intervals (CIs) for the MLE estimates from a LUCID model
<span class="citation" data-cites="davison1997bootstrap">(<a href="#ref-davison1997bootstrap" role="doc-biblioref">Davison and Hinkley 1997</a>)</span>. Nonparametric bootstrap draws observations from
a sample with replacement. We index the bootstrap samples by
<span class="math inline">\(b = 1, 2, \ldots, B\)</span>. For each bootstrap sample
<span class="math inline">\(\boldsymbol{\mathbf{D}}^b\)</span>, we fit the LUCID model and calculate MLE
<span class="math inline">\(\boldsymbol{\mathbf{\Theta}}^b\)</span>. Two types of bootstrap CIs are
constructed based on the distribution of
<span class="math inline">\(\boldsymbol{\mathbf{\Theta}}^b\)</span>: (1) normal CIs and (2) percentile CIs.</p>
<p>Given the confidence level <span class="math inline">\(1 - \alpha\)</span>, the normal CIs are constructed
by</p>
<p><span class="math display" id="eq:eq19">\[(1 - \alpha)\% \text{CI}_{\text{normal}} = \boldsymbol{\mathbf{\Theta}} - \text{bias} \pm Z_{1 - \alpha/2}\sigma(\boldsymbol{\mathbf{\Theta}})
    \label{eq19}   \tag{14}\]</span></p>
<p>where
<span class="math inline">\(\text{bias} = \bar{\boldsymbol{\mathbf{\Theta}}} - \boldsymbol{\mathbf{\Theta}}\)</span>,
<span class="math inline">\(\bar{\boldsymbol{\mathbf{\Theta}}}\)</span> is the mean of the bootstrap
estimators and <span class="math inline">\(\sigma(\boldsymbol{\mathbf{\Theta}})\)</span> is the bootstrap
standard error, which is</p>
<p><span class="math display" id="eq:eq20">\[\sigma(\boldsymbol{\mathbf{\Theta}}) = \sqrt{\frac{1}{B - 1}\sum_{b=1}^B\left(\boldsymbol{\mathbf{\Theta}}^b - \bar{\boldsymbol{\mathbf{\Theta}}}\right)}
    \label{eq20}   \tag{15}\]</span></p>
<p>All calculations in Equations <a href="#eq:eq19">(14)</a> and <a href="#eq:eq20">(15)</a> are
element-wise.</p>
<p>In addition, the <span class="math inline">\(1 - \alpha\)</span> percentile CIs are calculated by ordering
the bootstrap estimators from smallest to largest and selecting the
estimates at <span class="math inline">\(\alpha/2\)</span> percentile and <span class="math inline">\((1 - \alpha/2)\)</span> percentile as
the CIs.</p>
<p><code>boot_lucid()</code> constructs the two bootstrap CIs described above for
inference. Users can specify the confidence level (<code>conf</code>) and the
number of bootstrap replicates (<code>R</code>). While <code>boot_lucid()</code> is running, a
progress bar is displayed in the R console. We recommend <code>R</code> <span class="math inline">\(\geq 200\)</span>
to estimate the bootstrap standard error for normal CIs and <code>R</code>
<span class="math inline">\(\geq 800\)</span> to estimate quantiles to construct percentile CIs. Below are
examples of deriving bootstrap CIs for LUCID. By default, LUCID
calculates <span class="math inline">\(95\%\)</span> CIs.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Bootstrap to obtain 95% CI (by default) for LUCID</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> boot1 <span class="ot">&lt;-</span> <span class="fu">boot_lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics, <span class="at">Y =</span> ck18, <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                     <span class="at">model =</span> fit1, <span class="at">R =</span> <span class="dv">200</span>)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="co"># 90% CIs</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> boot2 <span class="ot">&lt;-</span> <span class="fu">boot_lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics, <span class="at">Y =</span> ck18, <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>,</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                     <span class="at">model =</span> fit1, <span class="at">R =</span> <span class="dv">200</span>, <span class="at">conf =</span> <span class="fl">0.9</span>)</span></code></pre></div>
<p>We can obtain a more comprehensive summary table with bootstrap CIs by
integrating <code>summary()</code> with output from <code>boot_lucid()</code>. The resulting
summary table has 5 columns: <code>t0</code> corresponds to parameters estimated
from the observed data; <code>norm_lower</code> and <code>norm_upper</code> represent the
lower and upper limit of normal CIs, respectively; <code>perc_lower</code> and
<code>perc_upper</code> represent the percentile CIs.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Summary table with 95% bootstrap CIs</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="fu">summary</span>(fit1, <span class="at">boot.se =</span> boot1)<span class="er">)</span></span></code></pre></div>
<p><code>boot_lucid()</code> is built upon the R library
<a href="https://CRAN.R-project.org/package=boot"><strong>boot</strong></a> <span class="citation" data-cites="canty2021boot">(<a href="#ref-canty2021boot" role="doc-biblioref">Canty and Ripley 2021</a>)</span>.
The original output from <strong>boot</strong> is also returned by <code>boot_lucid()</code> and
can be used to evaluate the normality of the distribution of the
bootstrap estimations.</p>
<h4 class="unnumbered" data-number="3.5" id="prediction">Prediction</h4>
<p>The prediction of LUCID includes two parts: prediction of the latent
clusters <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> and prediction of the outcome
<span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span>. The <code>predict_lucid()</code> can perform both tasks.
Prediction for the latent cluster is determined by IP using the optimal
rule, which is</p>
<p><span class="math display" id="eq:eq21">\[\hat{X}_i = \arg \max_j P(X_i = j| \boldsymbol{\mathbf{D}}, \boldsymbol{\mathbf{\Theta}})
    \label{eq21}   \tag{16}\]</span></p>
<p>Prediction for the outcome follows a conventional regression framework.
When making predictions, <code>predict_lucid()</code> for LUCID requires input of
new <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> and <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, and
optional input of <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span>. If
<span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> is provided, we use Equation <a href="#eq:eq3">(3)</a> to
calculate IPs; otherwise, Equation <a href="#eq:eq12">(8)</a> is applied. As a
result, we can either fit a supervised LUCID model to make predictions
in an unsupervised fashion or vice versa. This design allows for
flexibility. For instance, we can build a LUCID model in one cohort in a
supervised fashion and then make predictions on another new independent
cohort, regardless of whether the outcome is measured or not.
<code>predict_lucid()</code> for LUCID returns a <code>list</code> containing IPs for each
observation, predicted cluster assignment, and predicted outcome. Below
is an example code for prediction based on a supervised LUCID model,
<code>fit1</code>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Predict cluster with information of Y</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> pred1 <span class="ot">&lt;-</span> <span class="fu">predict_lucid</span>(<span class="at">model =</span> fit1, <span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics, <span class="at">Y =</span> ck18)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Predict cluster without information of Y</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> pred2 <span class="ot">&lt;-</span> <span class="fu">predict_lucid</span>(<span class="at">model =</span> fit1, <span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics)</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Predicted cluster label</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="fu">table</span>(pred1<span class="sc">$</span>pred.x)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="dv">1</span>   <span class="dv">2</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="dv">223</span> <span class="dv">197</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Predicted outcome</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> pred1<span class="sc">$</span>pred.y[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>]  <span class="fl">0.2580164</span> <span class="sc">-</span><span class="fl">0.3806054</span>  <span class="fl">0.2504817</span>  <span class="fl">0.4123169</span> <span class="sc">-</span><span class="fl">0.1898255</span></span></code></pre></div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:fig5"></span>
<img src="figures/fig5.png" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 5: Two simulated missing patterns in methylomics data
</p>
</div>
</div>
<h4 class="unnumbered" data-number="3.6" id="incorporating-missingness-in-multi-omics-data">Incorporating missingness in multi-omics data</h4>
<p>A major update in <strong>LUCIDus</strong> version 3 is the incorporation of missing
data. Missingness is a major challenge in the integrative analysis of
omics. In large cohort studies, it is common that some omics data are
not available for all participants for various reasons such as budgetary
constraints, low sample availability, or lack of consent for future use
of biospecimens <span class="citation" data-cites="voillet2016handling">(<a href="#ref-voillet2016handling" role="doc-biblioref">Voillet et al. 2016</a>)</span>. We refer to this type of
missingness as a list-wise missingness pattern. In addition, even when
omics data are measured for a given participant, it is common that some
omic features are randomly missing due to the measurement process. We
refer to this pattern as sporadic missingness. <strong>LUCIDus</strong> can deal with
the two missing patterns mentioned above or the combination of the two.
We assume the multi-omics data are missing completely at random (MCAR)
for both missing patterns (i.e., for the list-wise missing pattern, the
probability of multi-omics data being not available is the same for all
subjects; for sporadic missing patterns, the probability of a certain
omic feature being missing is the same for all omic features).</p>
<p>For list-wise missingness, we use a modified LUCID model based on a
likelihood partition. We divide the LUCID likelihood into two parts:
observations with complete multi-omics data and observations without any
measured multi-omics data. The likelihood of the former observations
remains the same as Equation <a href="#eq:eq2">(2)</a>, denoted by
<span class="math inline">\(l_o (\boldsymbol{\mathbf{\Theta}}|\boldsymbol{\mathbf{D}})\)</span>. The joint
likelihood of the latter observations becomes</p>
<p><span class="math display" id="eq:eq22">\[l_m(\boldsymbol{\mathbf{\Theta}}|\boldsymbol{\mathbf{D}}) = \sum_{l_o=1}^{N_o}\sum_{j=1}^K I(X_{i_o} = j) \left( \log S(X_{i_o} = j| \boldsymbol{\mathbf{G}}_{i_o};\boldsymbol{\mathbf{\beta}}_j) + \log \phi\left(Y_{i_o}|\gamma_j, \sigma_j^2\right)\right)
    \label{eq22}   \tag{17}\]</span></p>
<p>with corresponding responsibility defined as</p>
<p><span class="math display" id="eq:eq23">\[\begin{aligned}
        r_{ij}^{(t)} = \frac{S\left(X_i = j| \boldsymbol{\mathbf{G}}_i; \boldsymbol{\mathbf{\beta}}^{(t)}\right) \phi\left(Y_i|X_i = j; \gamma_j^{(t)}, \sigma^{2(t)}_j \right)}{\sum_{j = 1}^K S\left(X_i = j| \boldsymbol{\mathbf{G}}_i; \boldsymbol{\mathbf{\beta}}^{(t)}\right) \phi \left(Y_i|X_i = j; \gamma_j^{(t)}, \sigma^{2(t)}_j\right)}
    \end{aligned}
    \label{eq23}   \tag{18}\]</span></p>
<p>The joint likelihood of LUCID for all data becomes
<span class="math inline">\(l(\boldsymbol{\mathbf{\Theta}}) = l_o (\boldsymbol{\mathbf{\Theta}}|\boldsymbol{\mathbf{D}}) +l_m (\boldsymbol{\mathbf{\Theta}}|\boldsymbol{\mathbf{D}})\)</span>.
We then use the same EM algorithm described in Section <a href="#sec2.1">2.1</a> to
calculate the MLE of LUCID with list-wise missingness.</p>
<p>For sporadic missingness, we use an integrated imputation method
<span class="citation" data-cites="zhang2021gaussian">(<a href="#ref-zhang2021gaussian" role="doc-biblioref">Zhang et al. 2021</a>)</span>. Each missing value in the omics data
<span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> is treated as an unknown “parameter” to be
optimized. Specifically in the M-step, after maximizing the parameters
of the LUCID model with fixed <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>, we implement an
additional <em>imputation step</em> which maximizes <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span>
with fixed parameters within LUCID. Details of the statistical
derivation can be found in <span class="citation" data-cites="zhang2021gaussian">Zhang et al. (<a href="#ref-zhang2021gaussian" role="doc-biblioref">2021</a>)</span>.</p>
<p>To illustrate this new feature, we first simulate missing values in the
HELIX methylomics data by randomly setting values to <code>NA</code>. Both
list-wise and sporadic missing patterns are considered, as shown in
Figure <a href="#fig:fig5">5</a>. We use the <code>vis_miss()</code> function from the R
package <a href="https://CRAN.R-project.org/package=visdat"><strong>visdat</strong></a> to
visualize both missing patterns <span class="citation" data-cites="visdat">(<a href="#ref-visdat" role="doc-biblioref">Tierney 2023</a>)</span>.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">library</span>(visdat)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> methylomics_miss_sporadic <span class="ot">&lt;-</span> methylomics_miss_listwise <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(methylomics)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> index <span class="ot">&lt;-</span> <span class="fu">arrayInd</span>(<span class="fu">sample</span>(<span class="dv">1000</span>, <span class="fl">0.1</span> <span class="sc">*</span> <span class="dv">1000</span>), <span class="fu">dim</span>(methylomics))</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> methylomics_miss_sporadic[index] <span class="ot">&lt;-</span> <span class="cn">NA</span> <span class="co"># sporadic missing pattern</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> methylomics_miss_listwise[<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>, <span class="dv">30</span>), ] <span class="ot">&lt;-</span> <span class="cn">NA</span> <span class="co"># listwise missing pattern</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">vis_miss</span>(<span class="fu">as.data.frame</span>(methylomics_miss_sporadic))</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">vis_miss</span>(<span class="fu">as.data.frame</span>(methylomics_miss_listwise))</span></code></pre></div>
<p>Below are examples of fitting the LUCID model with missingness in
multi-omics data. <code>lucid()</code> will automatically examine missing patterns
present in the multi-omics data and select a corresponding method to
account for the missing data. For sporadic missing patterns, we use
<strong>mclust</strong> to initialize missing values in the multi-omics data.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> fit14 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics_miss_listwise, <span class="at">Y =</span> ck18,</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, <span class="at">K =</span> <span class="dv">2</span>, <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> fit15 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> methylomics_miss_sporadic, <span class="at">Y =</span> ck18,</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">lucid_model =</span> <span class="st">&quot;early&quot;</span>, <span class="at">K =</span> <span class="dv">2</span>, <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># summary(fit15)</span></span></code></pre></div>
<h4 class="unnumbered" data-number="3.7" id="sec_parallel">LUCID in Parallel</h4>
<p>In the previous sections, we focused on using the early integration
strategy by concatenating each omic layer one by one and inputting a
single data matrix into the LUCID model for estimation. However,
researchers may be interested in modeling the correlation structure of
each omic layer independently to investigate how multi-omics data may
act in parallel with an outcome. In this section, we extend the LUCID
model to incorporate multi-omics data by introducing multiple latent
variables into the framework. Suppose we have a sample of size <span class="math inline">\(n\)</span>,
indexed by <span class="math inline">\(i\)</span> = 1, 2, ..., <span class="math inline">\(n\)</span>. It has a collection of <span class="math inline">\(m\)</span> omic
layers, denoted by <span class="math inline">\(Z_1\)</span>, ..., <span class="math inline">\(Z_a\)</span>, ..., <span class="math inline">\(Z_m\)</span> with corresponding
dimensions <span class="math inline">\(p_1\)</span>, ..., <span class="math inline">\(p_a\)</span>, ..., <span class="math inline">\(p_m\)</span>. Each omic layer <span class="math inline">\(Z_a\)</span> is
summarized by a latent categorical variable <span class="math inline">\(X_a\)</span>, which contains <span class="math inline">\(k_a\)</span>
categories. Each category is interpreted as a latent cluster (or
subgroup) for that particular omic layer. All latent variables are
linked to the same exposure matrix <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> and the
outcome <span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> for LUCID in parallel, as shown in
Figure <a href="#fig:fig1">1</a> (b).</p>
<p>Suppose the latent variable <span class="math inline">\(X_{a i}\)</span> is observed for <span class="math inline">\(a=1, \ldots, m\)</span>.
According to the DAG in Figure <a href="#fig:fig1">1</a> (b), the distributions of
<span class="math inline">\(f\left(\boldsymbol{Z}_{a i} \mid \boldsymbol{G}\right)\)</span> are
conditionally independent of each other for <span class="math inline">\(a=1, \ldots, m\)</span>, and the
distributions of the latent variable
<span class="math inline">\(f\left(X_{a i} \mid \boldsymbol{G}_{i}\right)\)</span> are also conditionally
independent of each other.</p>
<p>Since <span class="math inline">\(\boldsymbol{X}_{a i}\)</span> is a discrete variable with <span class="math inline">\(k_{i}\)</span>
categories, we assume <span class="math inline">\(\boldsymbol{X}
_{a i}\)</span> follows a multinomial distribution conditioning on
<span class="math inline">\(\boldsymbol{G}\)</span>, denoted by the softmax function <span class="math inline">\(S(\cdot)\)</span>. We assume
multi-omics data <span class="math inline">\(\boldsymbol{Z}_{a i}\)</span> follows a multivariate Gaussian
distribution conditioning on <span class="math inline">\(\boldsymbol{X}_{a i}\)</span>, denoted by
<span class="math inline">\(\phi\left(\boldsymbol{Z}_{a i} \mid \boldsymbol{X}_{a i}=j_{a} ; \boldsymbol{\mu}_{j a}, \boldsymbol{\Sigma}_{j a}\right)\)</span>,
where <span class="math inline">\(\boldsymbol{\mu}_{j a}\)</span> and <span class="math inline">\(\boldsymbol{\Sigma}_{j a}\)</span> are
cluster-specific means and variance-covariance matrices for omic layer
<span class="math inline">\(a\)</span>. The outcome <span class="math inline">\(Y\)</span> can be either a continuous outcome or a binary
outcome. For illustration, here, we discuss the situation that <span class="math inline">\(Y\)</span> is a
continuous variable that follows a Gaussian distribution. The
relationship between latent variables <span class="math inline">\(X_{a}\)</span> and outcome <span class="math inline">\(Y\)</span> is
formulated as</p>
<p><span class="math display" id="eq:eq-24">\[\begin{aligned}
E Y_{i}=\gamma_{0}+\sum_{j_{1}=2}^{k_{1}} \gamma_{j_{1}} I\left(X_{1 i}=j_{1}\right)+\cdots+\sum_{j_{a}=2}^{k_{a}} \gamma_{j_{a}} I\left(X_{a i}=j_{a}\right)+\cdots+\sum_{j_{m}=2}^{k_{m}} \gamma_{j_{m}} I\left(X_{m i}=j_{m}\right)
\end{aligned}
    \label{eq_24}   \tag{19}\]</span></p>
<p>where the intercept <span class="math inline">\(\gamma_{0}\)</span> is the expected value of <span class="math inline">\(Y_{i}\)</span> given
<span class="math inline">\(X_{a i}=1\)</span> for <span class="math inline">\(a=1,2,\ldots, m\)</span> (the reference cluster for all
combinations of latent cluster assignments); <span class="math inline">\(\gamma_{j_{a}}\)</span> is the
change of <span class="math inline">\(Y\)</span> if the latent variable <span class="math inline">\(X_{a}\)</span> becomes <span class="math inline">\(j_{a}\)</span> instead of
1.</p>
<p>Let <span class="math inline">\(\boldsymbol{D}\)</span> be the generic notation for all observed data. The
log-likelihood of LUCID in Parallel is constructed below:</p>
<p><span class="math display" id="eq:eq-26">\[\begin{aligned}
\log L(\boldsymbol{\mathbf{\Theta}}\mid \boldsymbol{D}) &amp;  =\sum_{i=1}^{n} \log f\left(\boldsymbol{Z}_{1 i}, \ldots, \boldsymbol{Z}_{m i}, \boldsymbol{Y}_{i} \mid \boldsymbol{G}_{i} ; \boldsymbol{\Theta}\right) \\
&amp; =\sum_{i=1}^{n} \log \left[\prod_{j_{1}=1}^{k_{1}} \cdots \prod_{j_{m}=1}^{k_{m}} f\left(\boldsymbol{Z}_{1 i}, \ldots, \boldsymbol{Z}_{m i}, X_{1 i}, \ldots, X_{m i}, Y_{i} \mid \boldsymbol{G}_{i} ; \boldsymbol{\Theta}\right)^{I\left(X_{1 i}=j_{1}, \ldots, X_{m i}=j_{m}\right)}\right] \\
&amp; =\sum_{i=1}^{n} \sum_{j_{1}=1}^{k_{1}} \ldots \sum_{j_{m}=1}^{k_{m}} I\left(X_{1 i}=j_{1}, \ldots, X_{m i}=j_{m}\right) \log f\left(\boldsymbol{Z}_{1 i}, \ldots, \boldsymbol{Z}_{m i}, X_{1 i}, \ldots, X_{m i}, Y_{i} \mid \boldsymbol{G}_{i} ; \boldsymbol{\Theta}\right) \\
&amp; =\sum_{i=1}^{n} \sum_{j_{1}=1}^{k_{1}} \ldots \sum_{j_{m}=1}^{k_{m}} I\left(X_{1 i}=j_{1}, \ldots, X_{m i}=j_{m}\right) \log \phi\left(Y_{i} \mid X_{1 i}, \ldots, X_{m i}, \boldsymbol{\gamma}, \sigma^{2}\right) \\
&amp; +\sum_{i=1}^{n} \sum_{a=1}^{m} \sum_{j_{1}=1}^{k_{1}} \ldots \sum_{j_{m}=1}^{k_{m}} I\left(X_{1 i}=j_{1}, \ldots, X_{m i}=j_{m}\right) \log \phi\left(\boldsymbol{Z}_{a i} \mid X_{a i}=j_{a}, \boldsymbol{\mu}_{a,j_{a}}, \boldsymbol{\Sigma}_{a,j_{a}}\right) \\
&amp; +\sum_{i=1}^{n} \sum_{a=1}^{m} \sum_{j_{1}=1}^{k_{1}} \ldots \sum_{j_{m}=1}^{k_{m}} I\left(X_{1 i}=j_{1}, \ldots, X_{m i}=j_{m}\right) \log S\left(\boldsymbol{X}_{a i}=j_{a} \mid \boldsymbol{G}_{i}, \boldsymbol{\beta}_{a}\right)
\end{aligned}
    \label{eq_26}   \tag{20}\]</span></p>
<p>The log-likelihood of LUCID in parallel is similar to that with LUCID
early integration. It is natural to follow the same principles of the EM
algorithm for LUCID early integration and estimate the parameters of
LUCID in parallel with targeted modifications.</p>
<p>For illustration, we continue with the previous data example using
maternal exposure to mercury as the exposome <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>
and childhood cytokeratin 18 level as the outcome
<span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span>. For a LUCID in parallel model, multi-omics
data <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> is a list of three <span class="math inline">\(420 \times p_a\)</span>
matrices of omic layers where <span class="math inline">\(a = 1, 2, 3\)</span> (layer 1: methylome; layer
2: transcriptome; layer 3: miRNA). We fit LUCID in parallel models with
continuous ck18 using <code>lucid()</code>. We specify the argument <code>lucid_model</code>
to be <code>parallel</code>, and the argument <code>K</code> should be specified as a list of
three integers (or sequences of integers if tuning) indicating the
number of latent clusters for each omic layer. The variable selection
feature is not yet available for LUCID in parallel, and we can only tune
for <code>K</code>. We can use argument <code>CoG</code> and <code>CoY</code> to adjust for covariates
for the <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> to <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span>
association and the <span class="math inline">\(\boldsymbol{\mathbf{X}}\)</span> to
<span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> association, respectively. Similar to LUCID
early integration, the parameter <code>family</code> specifies the outcome type,
which is <code>normal</code> here, corresponding to a continuous outcome. For a
binary outcome, <code>family</code> should be specified as <code>binary</code>. <code>lucid()</code>
returns an object of class <code>lucid``_``parallel</code> providing the MLE of the
LUCID in parallel model as well as IPs. We can call <code>summary()</code> with the
input of a <code>lucid``_``parallel</code> object to print three tables of the
model results. Similar to the summary table of a LUCID early integration
model, the first table summarizes the association between the latent
clusters for each omic layer and the outcome (cytokeratin 18). The
second table characterizes each cluster by its omics signature within
each layer. The third table relates the exposures to the latent clusters
for each layer. The nuance is that the LUCID in parallel model has
multiple independent omic layers, and each layer corresponds to a
separate latent cluster variable, thus all the related results are
presented in the summary tables. See the two examples of fitting a LUCID
in parallel model below.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Create a list of multi-omics data</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> omics_lst <span class="ot">&lt;-</span> simulated_data[<span class="sc">-</span><span class="fu">which</span>(<span class="fu">names</span>(simulated_data) <span class="sc">==</span> <span class="st">&quot;phenotype&quot;</span>)]</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> Z <span class="ot">=</span> omics_lst[<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>)]</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># LUCID in parallel, adjusting for the covariates for the E-X and X-Y associations</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> fit16 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> Z, <span class="at">Y =</span> ck18, <span class="at">K =</span> <span class="fu">list</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>), <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>,</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">CoY =</span> covs, <span class="at">CoG =</span> covs,</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">lucid_model =</span> <span class="st">&quot;parallel&quot;</span>, <span class="at">useY =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># print the summary of the LUCID in parallel model</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="fu">summary</span>(fit16)</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="sc">----------</span>Summary of the LUCID <span class="cf">in</span> Parallel model<span class="sc">----------</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>K <span class="ot">=</span>  <span class="dv">2</span> <span class="dv">2</span> <span class="dv">2</span> , log likelihood <span class="ot">=</span> <span class="sc">-</span><span class="fl">16413.64</span> , BIC <span class="ot">=</span>  <span class="fl">36892.36</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>(<span class="dv">1</span>) <span class="fu">Y</span> (continuous outcome)<span class="sc">:</span> effects of each non<span class="sc">-</span>reference latent cluster</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> each layer of Y</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>(and effect of covariates <span class="cf">if</span> included)</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>                             Gamma</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>cluster2Layer1         <span class="fl">2.139337462</span></span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a>cluster2Layer2        <span class="sc">-</span><span class="fl">0.976575066</span></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a>cluster2Layer3         <span class="fl">1.617148373</span></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a>e3_sex_None            <span class="fl">0.023397524</span></span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a>hs_child_age_yrs_None <span class="sc">-</span><span class="fl">0.002297037</span></span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a>(<span class="dv">2</span>) Z<span class="sc">:</span> mean of omics data <span class="cf">for</span> each latent cluster of each layer</span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a>Layer  <span class="dv">1</span></span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a>              mu_cluster1 mu_cluster2</span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true" tabindex="-1"></a>cg_GRHL3       <span class="fl">0.09782065</span>  <span class="sc">-</span><span class="fl">0.2827826</span></span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true" tabindex="-1"></a>cg_BTF3L4      <span class="fl">0.12271079</span>  <span class="sc">-</span><span class="fl">0.3059186</span></span>
<span id="cb17-29"><a href="#cb17-29" aria-hidden="true" tabindex="-1"></a>cg_AL358472<span class="fl">.7</span>  <span class="fl">0.10019649</span>  <span class="sc">-</span><span class="fl">0.3164803</span></span>
<span id="cb17-30"><a href="#cb17-30" aria-hidden="true" tabindex="-1"></a>cg_HDGF        <span class="fl">0.15388979</span>  <span class="sc">-</span><span class="fl">0.3322801</span></span>
<span id="cb17-31"><a href="#cb17-31" aria-hidden="true" tabindex="-1"></a>cg_TDRD5       <span class="fl">0.10778098</span>  <span class="sc">-</span><span class="fl">0.1832600</span></span>
<span id="cb17-32"><a href="#cb17-32" aria-hidden="true" tabindex="-1"></a>cg_CSRNP3      <span class="fl">0.08692918</span>  <span class="sc">-</span><span class="fl">0.1300857</span></span>
<span id="cb17-33"><a href="#cb17-33" aria-hidden="true" tabindex="-1"></a>cg_HSPD1       <span class="fl">0.21255103</span>  <span class="sc">-</span><span class="fl">0.3855858</span></span>
<span id="cb17-34"><a href="#cb17-34" aria-hidden="true" tabindex="-1"></a>cg_EPM2AIP1   <span class="sc">-</span><span class="fl">0.11651397</span>   <span class="fl">0.3691065</span></span>
<span id="cb17-35"><a href="#cb17-35" aria-hidden="true" tabindex="-1"></a>cg_AC025171<span class="fl">.1</span>  <span class="fl">0.24747122</span>  <span class="sc">-</span><span class="fl">0.3750450</span></span>
<span id="cb17-36"><a href="#cb17-36" aria-hidden="true" tabindex="-1"></a>cg_VTRNA1_3   <span class="sc">-</span><span class="fl">0.13739036</span>   <span class="fl">0.1515515</span></span>
<span id="cb17-37"><a href="#cb17-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-38"><a href="#cb17-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-39"><a href="#cb17-39" aria-hidden="true" tabindex="-1"></a>Layer  <span class="dv">2</span></span>
<span id="cb17-40"><a href="#cb17-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-41"><a href="#cb17-41" aria-hidden="true" tabindex="-1"></a>                  mu_cluster1  mu_cluster2</span>
<span id="cb17-42"><a href="#cb17-42" aria-hidden="true" tabindex="-1"></a>tc_TC01006069_nc  <span class="fl">0.098203482</span> <span class="sc">-</span><span class="fl">0.169267971</span></span>
<span id="cb17-43"><a href="#cb17-43" aria-hidden="true" tabindex="-1"></a>tc_SLC9A4        <span class="sc">-</span><span class="fl">0.025858454</span>  <span class="fl">0.046609093</span></span>
<span id="cb17-44"><a href="#cb17-44" aria-hidden="true" tabindex="-1"></a>tc_RAB6C_AS1      <span class="fl">0.009567584</span> <span class="sc">-</span><span class="fl">0.087150028</span></span>
<span id="cb17-45"><a href="#cb17-45" aria-hidden="true" tabindex="-1"></a>tc_LOC100129029   <span class="fl">0.032557846</span>  <span class="fl">0.078278431</span></span>
<span id="cb17-46"><a href="#cb17-46" aria-hidden="true" tabindex="-1"></a>tc_BRE           <span class="sc">-</span><span class="fl">0.039411493</span>  <span class="fl">0.073329913</span></span>
<span id="cb17-47"><a href="#cb17-47" aria-hidden="true" tabindex="-1"></a>tc_TC03001220_nc <span class="sc">-</span><span class="fl">0.068457436</span>  <span class="fl">0.013811537</span></span>
<span id="cb17-48"><a href="#cb17-48" aria-hidden="true" tabindex="-1"></a>tc_TC04002114_nc <span class="sc">-</span><span class="fl">0.030761261</span>  <span class="fl">0.033415671</span></span>
<span id="cb17-49"><a href="#cb17-49" aria-hidden="true" tabindex="-1"></a>tc_TC04002369_nc  <span class="fl">0.297076410</span> <span class="sc">-</span><span class="fl">0.216123334</span></span>
<span id="cb17-50"><a href="#cb17-50" aria-hidden="true" tabindex="-1"></a>tc_BEND4         <span class="sc">-</span><span class="fl">0.128343956</span>  <span class="fl">0.074585000</span></span>
<span id="cb17-51"><a href="#cb17-51" aria-hidden="true" tabindex="-1"></a>tc_SLC9A3         <span class="fl">0.139869536</span>  <span class="fl">0.002458175</span></span>
<span id="cb17-52"><a href="#cb17-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-53"><a href="#cb17-53" aria-hidden="true" tabindex="-1"></a>Layer  <span class="dv">3</span></span>
<span id="cb17-54"><a href="#cb17-54" aria-hidden="true" tabindex="-1"></a>              mu_cluster1  mu_cluster2</span>
<span id="cb17-55"><a href="#cb17-55" aria-hidden="true" tabindex="-1"></a>miR.<span class="fl">101.3</span>p     <span class="fl">0.10748628</span>  <span class="fl">0.023648293</span></span>
<span id="cb17-56"><a href="#cb17-56" aria-hidden="true" tabindex="-1"></a>miR<span class="fl">.125</span>a<span class="fl">.5</span>p   <span class="sc">-</span><span class="fl">0.20581745</span>  <span class="fl">0.135670802</span></span>
<span id="cb17-57"><a href="#cb17-57" aria-hidden="true" tabindex="-1"></a>miR<span class="fl">.125</span>b.<span class="fl">1.3</span>p  <span class="fl">0.07644648</span>  <span class="fl">0.019200103</span></span>
<span id="cb17-58"><a href="#cb17-58" aria-hidden="true" tabindex="-1"></a>miR.<span class="fl">127.3</span>p    <span class="sc">-</span><span class="fl">0.05542160</span>  <span class="fl">0.184175902</span></span>
<span id="cb17-59"><a href="#cb17-59" aria-hidden="true" tabindex="-1"></a>miR.<span class="fl">140.5</span>p     <span class="fl">0.14979024</span>  <span class="fl">0.050720602</span></span>
<span id="cb17-60"><a href="#cb17-60" aria-hidden="true" tabindex="-1"></a>miR.<span class="fl">142.3</span>p     <span class="fl">0.11157913</span> <span class="sc">-</span><span class="fl">0.017566242</span></span>
<span id="cb17-61"><a href="#cb17-61" aria-hidden="true" tabindex="-1"></a>miR.<span class="fl">144.5</span>p     <span class="fl">0.07526727</span> <span class="sc">-</span><span class="fl">0.016027655</span></span>
<span id="cb17-62"><a href="#cb17-62" aria-hidden="true" tabindex="-1"></a>miR<span class="fl">.19</span>a<span class="fl">.3</span>p     <span class="fl">0.09193308</span> <span class="sc">-</span><span class="fl">0.002059547</span></span>
<span id="cb17-63"><a href="#cb17-63" aria-hidden="true" tabindex="-1"></a>miR<span class="fl">.19</span>b<span class="fl">.3</span>p     <span class="fl">0.08515311</span>  <span class="fl">0.032351416</span></span>
<span id="cb17-64"><a href="#cb17-64" aria-hidden="true" tabindex="-1"></a>miR.<span class="fl">21.5</span>p      <span class="fl">0.05800681</span>  <span class="fl">0.029022417</span></span>
<span id="cb17-65"><a href="#cb17-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-66"><a href="#cb17-66" aria-hidden="true" tabindex="-1"></a>(<span class="dv">3</span>) E<span class="sc">:</span> odds ratio of being assigned to each latent cluster <span class="cf">for</span> each exposure</span>
<span id="cb17-67"><a href="#cb17-67" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> each layer</span>
<span id="cb17-68"><a href="#cb17-68" aria-hidden="true" tabindex="-1"></a>Layer  <span class="dv">1</span></span>
<span id="cb17-69"><a href="#cb17-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-70"><a href="#cb17-70" aria-hidden="true" tabindex="-1"></a>                                     beta        OR</span>
<span id="cb17-71"><a href="#cb17-71" aria-hidden="true" tabindex="-1"></a>hs_hg_m_scaled.cluster2         <span class="fl">0.7352341</span> <span class="fl">2.0859703</span></span>
<span id="cb17-72"><a href="#cb17-72" aria-hidden="true" tabindex="-1"></a>e3_sex_None.cluster2           <span class="sc">-</span><span class="fl">0.4107395</span> <span class="fl">0.6631596</span></span>
<span id="cb17-73"><a href="#cb17-73" aria-hidden="true" tabindex="-1"></a>hs_child_age_yrs_None.cluster2 <span class="sc">-</span><span class="fl">0.2657909</span> <span class="fl">0.7665994</span></span>
<span id="cb17-74"><a href="#cb17-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-75"><a href="#cb17-75" aria-hidden="true" tabindex="-1"></a>Layer  <span class="dv">2</span></span>
<span id="cb17-76"><a href="#cb17-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-77"><a href="#cb17-77" aria-hidden="true" tabindex="-1"></a>                                     beta        OR</span>
<span id="cb17-78"><a href="#cb17-78" aria-hidden="true" tabindex="-1"></a>hs_hg_m_scaled.cluster2         <span class="fl">0.1313174</span> <span class="fl">1.1403296</span></span>
<span id="cb17-79"><a href="#cb17-79" aria-hidden="true" tabindex="-1"></a>e3_sex_None.cluster2            <span class="fl">0.1928708</span> <span class="fl">1.2127261</span></span>
<span id="cb17-80"><a href="#cb17-80" aria-hidden="true" tabindex="-1"></a>hs_child_age_yrs_None.cluster2 <span class="sc">-</span><span class="fl">0.1066509</span> <span class="fl">0.8988394</span></span>
<span id="cb17-81"><a href="#cb17-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-82"><a href="#cb17-82" aria-hidden="true" tabindex="-1"></a>Layer  <span class="dv">3</span></span>
<span id="cb17-83"><a href="#cb17-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-84"><a href="#cb17-84" aria-hidden="true" tabindex="-1"></a>                                     beta        OR</span>
<span id="cb17-85"><a href="#cb17-85" aria-hidden="true" tabindex="-1"></a>hs_hg_m_scaled.cluster2        <span class="sc">-</span><span class="fl">0.4896515</span> <span class="fl">0.6128399</span></span>
<span id="cb17-86"><a href="#cb17-86" aria-hidden="true" tabindex="-1"></a>e3_sex_None.cluster2            <span class="fl">0.4855559</span> <span class="fl">1.6250782</span></span>
<span id="cb17-87"><a href="#cb17-87" aria-hidden="true" tabindex="-1"></a>hs_child_age_yrs_None.cluster2  <span class="fl">0.1042404</span> <span class="fl">1.1098672</span></span>
<span id="cb17-88"><a href="#cb17-88" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># LUCID in parallel, tune for the number of clusters for methylomics</span></span>
<span id="cb17-89"><a href="#cb17-89" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> fit17 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> Z, <span class="at">Y =</span> ck18, <span class="at">K =</span> <span class="fu">list</span>(<span class="dv">2</span>, <span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">2</span>), <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>,</span>
<span id="cb17-90"><a href="#cb17-90" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">lucid_model =</span> <span class="st">&quot;parallel&quot;</span>, <span class="at">useY =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-91"><a href="#cb17-91" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># summary(fit17)</span></span></code></pre></div>
<h4 class="unnumbered" data-number="3.8" id="sec_serial">LUCID in Serial</h4>
<p>In a more late integration framework, LUCID can also be extended to
incorporate multiple latent variables in a serial fashion if researchers
believe that given an exposure, multi-omics data act serially through a
multistep process towards the outcome, as illustrated in Figure
<a href="#fig:fig1">1</a> (c). This framework can accommodate the following
situations: (1) longitudinal measurements on the same multi-omics data
type and (2) biological relationships of multi-omics data. The
estimation of latent clusters for each omic layer can be formulated as
an unsupervised LUCID early integration or LUCID in parallel sub-model.
For a LUCID in serial model, <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> includes <span class="math inline">\(m\)</span>
ordered omic layers in a sequential fashion. For illustration, assuming
all the sub-models are LUCID early integration models. For a certain
omic component <span class="math inline">\(a\)</span>, the cluster variable
<span class="math inline">\(\boldsymbol{\mathbf{X}}_{a - 1}\)</span> serves as the
“<span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span>” in the LUCID early integration model
framework, and <span class="math inline">\(\boldsymbol{\mathbf{X}}_a\)</span> is jointly estimated by
<span class="math inline">\(\boldsymbol{\mathbf{X}}_{a - 1}\)</span> and <span class="math inline">\(\boldsymbol{\mathbf{Z}}_a\)</span>.
However, the special cases are for the first and the last omic layer,
where the corresponding <span class="math inline">\(\boldsymbol{\mathbf{X}}_{1}\)</span> is estimated
jointly by <span class="math inline">\(\boldsymbol{\mathbf{G}}\)</span> and <span class="math inline">\(\boldsymbol{\mathbf{Z}}_1\)</span>,
and the corresponding <span class="math inline">\(\boldsymbol{\mathbf{X}}_{m}\)</span> is estimated jointly
by <span class="math inline">\(\boldsymbol{\mathbf{X}}_{m - 1}\)</span>, <span class="math inline">\(\boldsymbol{\mathbf{Z}}_m\)</span>, and
<span class="math inline">\(\boldsymbol{\mathbf{Y}}\)</span> if supervised, respectively. Alternatively,
for a certain omic component <span class="math inline">\(a\)</span>, we can add another transition
probability matrix component <span class="math inline">\(\boldsymbol{\mathbf{p}}\)</span> in the joint
likelihood relating <span class="math inline">\(\boldsymbol{\mathbf{X}}_{a - 1}\)</span> to
<span class="math inline">\(\boldsymbol{\mathbf{X}}_a\)</span> to describe the state change with
<span class="math inline">\(P(\boldsymbol{\mathbf{X}}_a|\boldsymbol{\mathbf{X}}_{a - 1}, \boldsymbol{\mathbf{C}}) = \boldsymbol{\mathbf{p}}_a\)</span>,
where <span class="math inline">\(\boldsymbol{\mathbf{p}}_a\)</span> is a general notation for the
inclusion probability of all the latent clusters for omic component <span class="math inline">\(a\)</span>
and <span class="math inline">\(\boldsymbol{\mathbf{C}}\)</span> dynamically represents other variables
that jointly estimate <span class="math inline">\(\boldsymbol{\mathbf{X}}_{a}\)</span>. Note that a list of
<span class="math inline">\(m\)</span> ordered omic components for <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> indicates <span class="math inline">\(m\)</span>
sub-models in a LUCID in serial model, and each sub-model can be either
a LUCID Early Integration model or a LUCID in Parallel model. Note that
the DAG in Figure <a href="#fig:fig1">1</a> (c) only shows the scenario where all
the sub-models are LUCID early integration models. Bold
<span class="math inline">\(\boldsymbol{\mathbf{X}}_{a}\)</span> and omic component
<span class="math inline">\(\boldsymbol{\mathbf{Z}}_a\)</span> indicate that they may include more than 1
latent variable or omic layer for each omic component if the
corresponding sub-model is LUCID in parallel.</p>
<p>Let <span class="math inline">\(m\)</span> be the number of sub-models in a LUCID in serial model. For each
sub-model <span class="math inline">\(a\)</span>, given <span class="math inline">\(\boldsymbol{\mathbf{\Theta_a}}\)</span> and
<span class="math inline">\(\boldsymbol{\mathbf{D_a}}\)</span>, its own <span class="math inline">\(\log L({\Theta_a} \mid {D_a})\)</span>
follows the log-likelihood of a LUCID Early Integration or a LUCID in
Parallel model defined in the previous sections. Except for the
sub-model <span class="math inline">\(m\)</span>, all sub-models are unsupervised. For sub-models 2 to <span class="math inline">\(m\)</span>,
the <span class="math inline">\(\beta\)</span> parameter defined previously becomes <span class="math inline">\(\delta\)</span>. The
log-likelihood of LUCID in Serial is constructed below,</p>
<p><span class="math display" id="eq:eq-27">\[\begin{aligned}
         \log L(\boldsymbol{\mathbf{\Theta}}\mid \boldsymbol{D}) &amp; = \sum_{a = 1}^m \log L(\boldsymbol{\mathbf{\Theta_a}}\mid \boldsymbol{\mathbf{D_a}}) \\ &amp;
         = \sum_{i = 1}^n \log f_1(\boldsymbol{\mathbf{Z}}_{1i}, \boldsymbol{\mathbf{G}}_i \mid \boldsymbol{\mathbf{\Theta_1}}) \\ &amp;
         + \sum_{a = 2}^{m-1} \sum_{i = 1}^n \log f_a(\boldsymbol{\mathbf{Z}}_{ai}, \boldsymbol{\mathbf{p}}_{ai} \mid \boldsymbol{\mathbf{\Theta_a}}) + \sum_{i = 1}^n \log f_m(\boldsymbol{\mathbf{Z}}_{mi}, \boldsymbol{\mathbf{p}}_{ai}, \boldsymbol{\mathbf{Y_i}} \mid \boldsymbol{\mathbf{\Theta_m}}) \\ &amp;
         = \sum_{i = 1}^n \log f_1(\boldsymbol{\mathbf{Z}}_{1i}, \boldsymbol{\mathbf{G}}_i \mid \boldsymbol{\mathbf{\beta}}, \boldsymbol{\mathbf{\mu_1}}, \boldsymbol{\mathbf{\Sigma_1}}) \\ &amp;
         + \sum_{a = 2}^{m-1} \sum_{i = 1}^n \log f_a(\boldsymbol{\mathbf{Z}}_{ai}, \boldsymbol{\mathbf{p}}_{ai} \mid \boldsymbol{\mathbf{\delta_a}}, \boldsymbol{\mathbf{\mu_a}}, \boldsymbol{\mathbf{\Sigma_a}}) \\ &amp;
         + \sum_{i = 1}^n \log f_m(\boldsymbol{\mathbf{Z}}_{mi}, \boldsymbol{\mathbf{p}}_{ai}, \boldsymbol{\mathbf{Y_i}} \mid \boldsymbol{\mathbf{\delta_m}}, \boldsymbol{\mathbf{\mu_m}}, \boldsymbol{\mathbf{\Sigma_m}}, \boldsymbol{\mathbf{\gamma}})
    \end{aligned}
    \label{eq_27}   \tag{21}\]</span></p>
<p>Here, we assume sub-models are independent of each other. The EM
algorithm for estimating the parameters of each sub-model is exactly the
same as estimating a single LUCID early integration model or a LUCID in
parallel model.</p>
<p>Using the previous data example for illustration, for a LUCID in serial
model, multi-omics data <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> can be an ordered list
of three <span class="math inline">\(420 \times p_a\)</span> matrices of omic layers where <span class="math inline">\(a = 1, 2, 3\)</span>
(layer 1: methylome; layer 2: transcriptome; layer 3: miRNA),
representing that all the sub-models are LUCID early integration models
and <code>K</code> is a list of three integers (or sequences of integers if tuning)
indicating the number of latent clusters for each successively linked
omic layer. We specify <code>lucid_model</code> to be <code>serial</code>. The use of
arguments <code>CoG</code>, <code>CoY</code>, <code>useY</code>, and <code>family</code> is exactly the same as for
a LUCID in parallel model. <code>lucid()</code> constructs the model and returns an
object of class <code>lucid``_``serial</code>, and calling <code>summary()</code> with the
input of a <code>lucid``_``serial</code> object prints the summarizing tables of
the model estimates. We have introduced summarizing tables for the LUCID
early integration model and the LUCID in parallel model. Since the LUCID
in serial model is a combination of LUCID early integration and LUCID in
parallel sub-models, the summarizing tables for each sub-model follow a
similar structure to what we have discussed for the two types of
sub-models, but with targeted modifications to conform to the setup of
the LUCID in serial model. See the summarizing tables for the first
LUCID in serial model example below. Alternatively, besides an integer,
the element of the ordered list of <span class="math inline">\(\boldsymbol{\mathbf{Z}}\)</span> can also be
a list, which indicates that the corresponding sub-model is a LUCID in
parallel model. Here, <code>K</code> should contain a list of lists of integers.
See the second LUCID in serial model example below.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># LUCID in serial, adjusting for the covariates of the E-X and X-Y associations</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># Tune the number of clusters for methylome and transcriptome</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="co"># All 3 sub-models are LUCID early integration models and each with two latent clusters for each layer</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> fit18 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> Z, <span class="at">Y =</span> ck18,</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">lucid_model =</span> <span class="st">&quot;serial&quot;</span>, <span class="at">CoY =</span> covs, <span class="at">CoG =</span> covs,</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">K =</span> <span class="fu">list</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">2</span>), <span class="at">useY =</span> <span class="cn">TRUE</span>, <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Print the summary of the LUCID in serial model</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="fu">summary</span>(fit18)</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="sc">----------</span>Summary of the First Component of the LUCID <span class="cf">in</span> Serial Model<span class="sc">----------</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="sc">----------</span>Summary of the LUCID Early Integration Sub Model<span class="sc">----------</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>K <span class="ot">=</span>  <span class="dv">2</span> , log likelihood <span class="ot">=</span> <span class="sc">-</span><span class="fl">5977.843</span> , BIC <span class="ot">=</span>  <span class="fl">13320.78</span></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>(<span class="dv">1</span>) Z<span class="sc">:</span> mean of omics data <span class="cf">for</span> each latent cluster</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>               mu_cluster1  mu_cluster2</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>cg_GRHL3       <span class="fl">0.295741860</span> <span class="sc">-</span><span class="fl">0.410975212</span></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>cg_BTF3L4      <span class="fl">0.021123889</span> <span class="sc">-</span><span class="fl">0.090792176</span></span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>cg_AL358472<span class="fl">.7</span>  <span class="fl">0.301523381</span> <span class="sc">-</span><span class="fl">0.439813049</span></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>cg_HDGF        <span class="fl">0.170527566</span> <span class="sc">-</span><span class="fl">0.234364618</span></span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>cg_TDRD5       <span class="fl">0.192293276</span> <span class="sc">-</span><span class="fl">0.207240424</span></span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>cg_CSRNP3      <span class="fl">0.235578302</span> <span class="sc">-</span><span class="fl">0.242839045</span></span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a>cg_HSPD1      <span class="sc">-</span><span class="fl">0.002680386</span> <span class="sc">-</span><span class="fl">0.003986538</span></span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a>cg_EPM2AIP1    <span class="fl">0.033545336</span>  <span class="fl">0.086638565</span></span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a>cg_AC025171<span class="fl">.1</span>  <span class="fl">0.303125436</span> <span class="sc">-</span><span class="fl">0.287725955</span></span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a>cg_VTRNA1_3   <span class="sc">-</span><span class="fl">0.090000154</span>  <span class="fl">0.029899266</span></span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a>(<span class="dv">2</span>) E<span class="sc">:</span> odds ratio of being assigned to each latent cluster <span class="cf">for</span> each exposure</span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a>                                     beta        OR</span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a>hs_hg_m_scaled.cluster2         <span class="fl">1.8300519</span> <span class="fl">6.2342102</span></span>
<span id="cb18-32"><a href="#cb18-32" aria-hidden="true" tabindex="-1"></a>e3_sex_None.cluster2           <span class="sc">-</span><span class="fl">0.1045747</span> <span class="fl">0.9007075</span></span>
<span id="cb18-33"><a href="#cb18-33" aria-hidden="true" tabindex="-1"></a>hs_child_age_yrs_None.cluster2 <span class="sc">-</span><span class="fl">0.4683347</span> <span class="fl">0.6260439</span></span>
<span id="cb18-34"><a href="#cb18-34" aria-hidden="true" tabindex="-1"></a><span class="sc">----------</span>Summary of the Middle Component of the LUCID <span class="cf">in</span> Serial Model<span class="sc">----------</span></span>
<span id="cb18-35"><a href="#cb18-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-36"><a href="#cb18-36" aria-hidden="true" tabindex="-1"></a><span class="sc">----------</span>Summary of the LUCID Early Integration Sub Model<span class="sc">----------</span></span>
<span id="cb18-37"><a href="#cb18-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-38"><a href="#cb18-38" aria-hidden="true" tabindex="-1"></a>K <span class="ot">=</span>  <span class="dv">2</span> , log likelihood <span class="ot">=</span> <span class="sc">-</span><span class="fl">6035.014</span> , BIC <span class="ot">=</span>  <span class="fl">13435.13</span></span>
<span id="cb18-39"><a href="#cb18-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-40"><a href="#cb18-40" aria-hidden="true" tabindex="-1"></a>(<span class="dv">1</span>) Z<span class="sc">:</span> mean of omics data <span class="cf">for</span> each latent cluster</span>
<span id="cb18-41"><a href="#cb18-41" aria-hidden="true" tabindex="-1"></a>                 mu_cluster1  mu_cluster2</span>
<span id="cb18-42"><a href="#cb18-42" aria-hidden="true" tabindex="-1"></a>tc_TC01006069_nc <span class="sc">-</span><span class="fl">0.32822703</span>  <span class="fl">0.321670957</span></span>
<span id="cb18-43"><a href="#cb18-43" aria-hidden="true" tabindex="-1"></a>tc_SLC9A4         <span class="fl">0.10853723</span> <span class="sc">-</span><span class="fl">0.106428335</span></span>
<span id="cb18-44"><a href="#cb18-44" aria-hidden="true" tabindex="-1"></a>tc_RAB6C_AS1     <span class="sc">-</span><span class="fl">0.24384843</span>  <span class="fl">0.195716238</span></span>
<span id="cb18-45"><a href="#cb18-45" aria-hidden="true" tabindex="-1"></a>tc_LOC100129029   <span class="fl">0.10572458</span> <span class="sc">-</span><span class="fl">0.005932059</span></span>
<span id="cb18-46"><a href="#cb18-46" aria-hidden="true" tabindex="-1"></a>tc_BRE           <span class="sc">-</span><span class="fl">0.10399915</span>  <span class="fl">0.125807476</span></span>
<span id="cb18-47"><a href="#cb18-47" aria-hidden="true" tabindex="-1"></a>tc_TC03001220_nc <span class="sc">-</span><span class="fl">0.30053137</span>  <span class="fl">0.248463710</span></span>
<span id="cb18-48"><a href="#cb18-48" aria-hidden="true" tabindex="-1"></a>tc_TC04002114_nc  <span class="fl">0.04931376</span> <span class="sc">-</span><span class="fl">0.060764599</span></span>
<span id="cb18-49"><a href="#cb18-49" aria-hidden="true" tabindex="-1"></a>tc_TC04002369_nc <span class="sc">-</span><span class="fl">0.38991227</span>  <span class="fl">0.586539770</span></span>
<span id="cb18-50"><a href="#cb18-50" aria-hidden="true" tabindex="-1"></a>tc_BEND4         <span class="sc">-</span><span class="fl">0.26952193</span>  <span class="fl">0.195503719</span></span>
<span id="cb18-51"><a href="#cb18-51" aria-hidden="true" tabindex="-1"></a>tc_SLC9A3         <span class="fl">0.26488011</span> <span class="sc">-</span><span class="fl">0.110649505</span></span>
<span id="cb18-52"><a href="#cb18-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-53"><a href="#cb18-53" aria-hidden="true" tabindex="-1"></a>(<span class="dv">2</span>) E<span class="sc">:</span> odds ratio of being assigned to each latent cluster <span class="cf">for</span> each cluster</span>
<span id="cb18-54"><a href="#cb18-54" aria-hidden="true" tabindex="-1"></a>    from the last sub model</span>
<span id="cb18-55"><a href="#cb18-55" aria-hidden="true" tabindex="-1"></a>               delta       OR</span>
<span id="cb18-56"><a href="#cb18-56" aria-hidden="true" tabindex="-1"></a>G1.cluster2 <span class="fl">1.554258</span> <span class="fl">4.731575</span></span>
<span id="cb18-57"><a href="#cb18-57" aria-hidden="true" tabindex="-1"></a><span class="sc">----------</span>Summary of the Last Component of the LUCID <span class="cf">in</span> Serial Model<span class="sc">----------</span></span>
<span id="cb18-58"><a href="#cb18-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-59"><a href="#cb18-59" aria-hidden="true" tabindex="-1"></a><span class="sc">----------</span>Summary of the LUCID Early Integration Sub Model<span class="sc">----------</span></span>
<span id="cb18-60"><a href="#cb18-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-61"><a href="#cb18-61" aria-hidden="true" tabindex="-1"></a>K <span class="ot">=</span>  <span class="dv">2</span> , log likelihood <span class="ot">=</span> <span class="sc">-</span><span class="fl">4533.659</span> , BIC <span class="ot">=</span>  <span class="fl">10444.5</span></span>
<span id="cb18-62"><a href="#cb18-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-63"><a href="#cb18-63" aria-hidden="true" tabindex="-1"></a>(<span class="dv">1</span>) <span class="fu">Y</span> (continuous outcome)<span class="sc">:</span> effect size of Y <span class="cf">for</span> each latent cluster</span>
<span id="cb18-64"><a href="#cb18-64" aria-hidden="true" tabindex="-1"></a>(and effect of covariates <span class="cf">if</span> included)</span>
<span id="cb18-65"><a href="#cb18-65" aria-hidden="true" tabindex="-1"></a>                            Gamma</span>
<span id="cb18-66"><a href="#cb18-66" aria-hidden="true" tabindex="-1"></a>cluster1               <span class="fl">0.00000000</span></span>
<span id="cb18-67"><a href="#cb18-67" aria-hidden="true" tabindex="-1"></a>cluster2               <span class="fl">1.64219533</span></span>
<span id="cb18-68"><a href="#cb18-68" aria-hidden="true" tabindex="-1"></a>e3_sex_None            <span class="fl">0.06626907</span></span>
<span id="cb18-69"><a href="#cb18-69" aria-hidden="true" tabindex="-1"></a>hs_child_age_yrs_None <span class="sc">-</span><span class="fl">0.16063912</span></span>
<span id="cb18-70"><a href="#cb18-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-71"><a href="#cb18-71" aria-hidden="true" tabindex="-1"></a>(<span class="dv">2</span>) Z<span class="sc">:</span> mean of omics data <span class="cf">for</span> each latent cluster</span>
<span id="cb18-72"><a href="#cb18-72" aria-hidden="true" tabindex="-1"></a>              mu_cluster1  mu_cluster2</span>
<span id="cb18-73"><a href="#cb18-73" aria-hidden="true" tabindex="-1"></a>miR.<span class="fl">101.3</span>p     <span class="fl">0.12404049</span>  <span class="fl">0.028070959</span></span>
<span id="cb18-74"><a href="#cb18-74" aria-hidden="true" tabindex="-1"></a>miR<span class="fl">.125</span>a<span class="fl">.5</span>p   <span class="sc">-</span><span class="fl">0.19583201</span>  <span class="fl">0.088128200</span></span>
<span id="cb18-75"><a href="#cb18-75" aria-hidden="true" tabindex="-1"></a>miR<span class="fl">.125</span>b.<span class="fl">1.3</span>p <span class="sc">-</span><span class="fl">0.02249475</span>  <span class="fl">0.064270997</span></span>
<span id="cb18-76"><a href="#cb18-76" aria-hidden="true" tabindex="-1"></a>miR.<span class="fl">127.3</span>p    <span class="sc">-</span><span class="fl">0.21746619</span>  <span class="fl">0.215300195</span></span>
<span id="cb18-77"><a href="#cb18-77" aria-hidden="true" tabindex="-1"></a>miR.<span class="fl">140.5</span>p     <span class="fl">0.27069168</span>  <span class="fl">0.017292491</span></span>
<span id="cb18-78"><a href="#cb18-78" aria-hidden="true" tabindex="-1"></a>miR.<span class="fl">142.3</span>p     <span class="fl">0.15068775</span> <span class="sc">-</span><span class="fl">0.015944126</span></span>
<span id="cb18-79"><a href="#cb18-79" aria-hidden="true" tabindex="-1"></a>miR.<span class="fl">144.5</span>p     <span class="fl">0.25253765</span> <span class="sc">-</span><span class="fl">0.071952430</span></span>
<span id="cb18-80"><a href="#cb18-80" aria-hidden="true" tabindex="-1"></a>miR<span class="fl">.19</span>a<span class="fl">.3</span>p     <span class="fl">0.15974513</span> <span class="sc">-</span><span class="fl">0.015887823</span></span>
<span id="cb18-81"><a href="#cb18-81" aria-hidden="true" tabindex="-1"></a>miR<span class="fl">.19</span>b<span class="fl">.3</span>p     <span class="fl">0.20886420</span> <span class="sc">-</span><span class="fl">0.008073847</span></span>
<span id="cb18-82"><a href="#cb18-82" aria-hidden="true" tabindex="-1"></a>miR.<span class="fl">21.5</span>p      <span class="fl">0.18868583</span> <span class="sc">-</span><span class="fl">0.017110888</span></span>
<span id="cb18-83"><a href="#cb18-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-84"><a href="#cb18-84" aria-hidden="true" tabindex="-1"></a>(<span class="dv">3</span>) E<span class="sc">:</span> odds ratio of being assigned to each latent cluster <span class="cf">for</span> each cluster</span>
<span id="cb18-85"><a href="#cb18-85" aria-hidden="true" tabindex="-1"></a>    from the last sub model</span>
<span id="cb18-86"><a href="#cb18-86" aria-hidden="true" tabindex="-1"></a>                delta       OR</span>
<span id="cb18-87"><a href="#cb18-87" aria-hidden="true" tabindex="-1"></a>G1.cluster2 <span class="fl">0.9911389</span> <span class="fl">2.694301</span></span>
<span id="cb18-88"><a href="#cb18-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-89"><a href="#cb18-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-90"><a href="#cb18-90" aria-hidden="true" tabindex="-1"></a><span class="sc">----------</span>Overall Summary of the LUCID <span class="cf">in</span> Serial model<span class="sc">----------</span></span>
<span id="cb18-91"><a href="#cb18-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-92"><a href="#cb18-92" aria-hidden="true" tabindex="-1"></a>log likelihood <span class="ot">=</span> <span class="sc">-</span><span class="fl">16546.52</span> , BIC <span class="ot">=</span>  <span class="fl">37200.41</span></span>
<span id="cb18-93"><a href="#cb18-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-94"><a href="#cb18-94" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># LUCID in serial with the first sub-model being LUCID in parallel</span></span>
<span id="cb18-95"><a href="#cb18-95" aria-hidden="true" tabindex="-1"></a>  and the second sub<span class="sc">-</span>model being LUCID early integration</span>
<span id="cb18-96"><a href="#cb18-96" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Rearrange the omics list to match the new structure of the LUCID in serial model</span></span>
<span id="cb18-97"><a href="#cb18-97" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> Z_new <span class="ot">=</span> <span class="fu">list</span>(<span class="fu">list</span>(omics_lst<span class="sc">$</span>methylome, omics_lst<span class="sc">$</span>transcriptome), omics_lst<span class="sc">$</span>miRNA)</span>
<span id="cb18-98"><a href="#cb18-98" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># Fit the LUCID in serial model with the new omics list and the new K list</span></span>
<span id="cb18-99"><a href="#cb18-99" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> fit19 <span class="ot">&lt;-</span> <span class="fu">lucid</span>(<span class="at">G =</span> exposome, <span class="at">Z =</span> Z_new, <span class="at">Y =</span> ck18,</span>
<span id="cb18-100"><a href="#cb18-100" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">lucid_model =</span> <span class="st">&quot;serial&quot;</span>, <span class="at">CoY =</span> covs, <span class="at">CoG =</span> covs,</span>
<span id="cb18-101"><a href="#cb18-101" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">K =</span> <span class="fu">list</span>(<span class="fu">list</span>(<span class="dv">2</span>, <span class="dv">2</span>), <span class="dv">2</span>), <span class="at">useY =</span> <span class="cn">TRUE</span>, <span class="at">family =</span> <span class="st">&quot;normal&quot;</span>)</span>
<span id="cb18-102"><a href="#cb18-102" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="co"># summary(fit19)</span></span></code></pre></div>
<h3 data-number="4" id="conclusion"><span class="header-section-number">4</span> Conclusion</h3>
<p>In this paper, we have introduced the <strong>LUCIDus</strong> package, version 3, to
perform estimation of the LUCID model in R <span class="citation" data-cites="LUCIDus">(<a href="#ref-LUCIDus" role="doc-biblioref">Jia et al. 2024</a>)</span>. <strong>LUCIDus</strong>
focuses on integrative clustering analysis using multi-omics data, both
with and without phenotypic traits. It consists of a set of toolkits for
modeling, interpretation, visualization, inference, and prediction.
Compared to the previous version, <strong>LUCIDus</strong> version 3 is faster (120
times faster than the previous version for a dataset with 20,000
observations), more stable, and includes several features for more
functionality, including options for early, intermediate, and late
integration of multi-omics data. The <strong>LUCIDus</strong> package is a useful
tool for performing integrated clustering analysis and can potentially
provide greater insights into environmental epidemiology studies with
measured multi-omics data.</p>
</div>
<div class="sourceCode" id="cb19"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h3 class="appendix" data-number="5" id="supplementary-materials"><span class="header-section-number">5</span> Supplementary materials</h3>
<p>Supplementary materials are available in addition to this article. It can be downloaded at
<a href="RJ-2024-012.zip">RJ-2024-012.zip</a></p>
<h3 class="appendix" data-number="6" id="cran-packages-used"><span class="header-section-number">6</span> CRAN packages used</h3>
<p><a href="https://cran.r-project.org/package=LUCIDus">LUCIDus</a>, <a href="https://cran.r-project.org/package=dlstats">dlstats</a>, <a href="https://cran.r-project.org/package=mclust">mclust</a>, <a href="https://cran.r-project.org/package=boot">boot</a>, <a href="https://cran.r-project.org/package=visdat">visdat</a></p>
<h3 class="appendix" data-number="7" id="cran-task-views-implied-by-cited-packages"><span class="header-section-number">7</span> CRAN Task Views implied by cited packages</h3>
<p><a href="https://cran.r-project.org/view=Cluster">Cluster</a>, <a href="https://cran.r-project.org/view=Distributions">Distributions</a>, <a href="https://cran.r-project.org/view=Econometrics">Econometrics</a>, <a href="https://cran.r-project.org/view=Environmetrics">Environmetrics</a>, <a href="https://cran.r-project.org/view=MissingData">MissingData</a>, <a href="https://cran.r-project.org/view=Omics">Omics</a>, <a href="https://cran.r-project.org/view=Optimization">Optimization</a>, <a href="https://cran.r-project.org/view=Survival">Survival</a>, <a href="https://cran.r-project.org/view=TimeSeries">TimeSeries</a></p>
<h3 class="appendix" data-number="8" id="note"><span class="header-section-number">8</span> Note</h3>
<p>This article is converted from a Legacy LaTeX article using the
<a href="https://cran.r-project.org/package=texor">texor</a> package.
The pdf version is the official version. To report a problem with the html,
refer to CONTRIBUTE on the R Journal homepage.</p>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-albert2016causal" class="csl-entry" role="listitem">
J. M. Albert, C. Geng and S. Nelson. Causal mediation analysis with a latent mediator. <em>Biometrical Journal</em>, 58(3): 535–548, 2016.
</div>
<div id="ref-aslam2017proteomics" class="csl-entry" role="listitem">
B. Aslam, M. Basit, M. A. Nisar, M. Khurshid and M. H. Rasool. Proteomics: Technologies and their applications. <em>Journal of Chromatographic Science</em>, 55(2): 182–196, 2017.
</div>
<div id="ref-banfield1993model" class="csl-entry" role="listitem">
J. D. Banfield and A. E. Raftery. Model-based gaussian and non-gaussian clustering. <em>Biometrics</em>, 803–821, 1993.
</div>
<div id="ref-beger2013review" class="csl-entry" role="listitem">
R. D. Beger. A review of applications of metabolomics in cancer. <em>Metabolites</em>, 3(3): 552–574, 2013.
</div>
<div id="ref-CADIOU2021106509" class="csl-entry" role="listitem">
S. Cadiou, X. Basagaña, J. R. Gonzalez, J. Lepeule, M. Vrijheid, V. Siroux and R. Slama. Performance of approaches relying on multidimensional intermediary data to decipher causal relationships between the exposome and health: A simulation study under various causal structures. <em>Environment International</em>, 153: 106509, 2021. URL <a href="https://www.sciencedirect.com/science/article/pii/S0160412021001343">https://www.sciencedirect.com/science/article/pii/S0160412021001343</a>.
</div>
<div id="ref-canty2021boot" class="csl-entry" role="listitem">
A. Canty and B. Ripley. <em>Boot: Bootstrap functions (originally by angelo canty for s).</em> 2021. URL <a href="https://CRAN.R-project.org/package=boot">https://CRAN.R-project.org/package=boot</a>. R package version 1.3-28.
</div>
<div id="ref-celeux1995gaussian" class="csl-entry" role="listitem">
G. Celeux and G. Govaert. Gaussian parsimonious clustering models. <em>Pattern Recognition</em>, 28(5): 781–793, 1995.
</div>
<div id="ref-chan2008statistical" class="csl-entry" role="listitem">
C. Chan, F. Feng, J. Ottinger, D. Foster, M. West and T. B. Kepler. Statistical mixture modeling for cell subtype identification in flow cytometry. <em>Cytometry Part A: The Journal of the International Society for Analytical Cytology</em>, 73(8): 693–701, 2008.
</div>
<div id="ref-crook2018bayesian" class="csl-entry" role="listitem">
O. M. Crook, C. M. Mulvey, P. D. Kirk, K. S. Lilley and L. Gatto. A bayesian mixture modelling approach for spatial proteomics. <em>PLoS Computational Biology</em>, 14(11): e1006516, 2018.
</div>
<div id="ref-curtis2012genomic" class="csl-entry" role="listitem">
C. Curtis, S. P. Shah, S.-F. Chin, G. Turashvili, O. M. Rueda, M. J. Dunning, D. Speed, A. G. Lynch, S. Samarajiwa, Y. Yuan, et al. The genomic and transcriptomic architecture of 2,000 breast tumours reveals novel subgroups. <em>Nature</em>, 486(7403): 346–352, 2012.
</div>
<div id="ref-davison1997bootstrap" class="csl-entry" role="listitem">
A. C. Davison and D. V. Hinkley. <em>Bootstrap methods and their application.</em> Cambridge University Press, 1997.
</div>
<div id="ref-derkach2019high" class="csl-entry" role="listitem">
A. Derkach, R. M. Pfeiffer, T.-H. Chen and J. N. Sampson. High dimensional mediation analysis with latent variables. <em>Biometrics</em>, 75(3): 745–756, 2019.
</div>
<div id="ref-fan2013tuning" class="csl-entry" role="listitem">
Y. Fan and C. Y. Tang. Tuning parameter selection in high dimensional penalized likelihood. <em>Journal of the Royal Statistical Society: Series B (Statistical Methodology)</em>, 75(3): 531–552, 2013.
</div>
<div id="ref-fop2018variable" class="csl-entry" role="listitem">
M. Fop and T. B. Murphy. Variable selection methods for model-based clustering. <em>Statistics Surveys</em>, 12: 18–65, 2018.
</div>
<div id="ref-gonzalez2019omic" class="csl-entry" role="listitem">
J. R. González and A. Cáceres. <em>Omic association studies with r and bioconductor.</em> CRC Press, 2019.
</div>
<div id="ref-goodwin2016coming" class="csl-entry" role="listitem">
S. Goodwin, J. D. McPherson and W. R. McCombie. Coming of age: Ten years of next-generation sequencing technologies. <em>Nature Reviews Genetics</em>, 17(6): 333–351, 2016.
</div>
<div id="ref-hejblum2019sequential" class="csl-entry" role="listitem">
B. P. Hejblum, C. Alkhassim, R. Gottardo, F. Caron and R. Thiébaut. Sequential dirichlet process mixtures of multivariate skew t-distributions for model-based clustering of flow cytometry data. 2019.
</div>
<div id="ref-hejblum2015time" class="csl-entry" role="listitem">
B. P. Hejblum, J. Skinner and R. Thiébaut. Time-course gene set analysis for longitudinal gene expression data. <em>PLoS Computational Biology</em>, 11(6): e1004310, 2015.
</div>
<div id="ref-LUCIDus" class="csl-entry" role="listitem">
Q. Jia, Y. Zhao, D. Conti and J. Goodrich. <em>LUCIDus: LUCID with multiple omics data.</em> Comprehensive R Archive Network (CRAN), 2024. URL <a href="https://CRAN.R-project.org/package=LUCIDus">https://CRAN.R-project.org/package=LUCIDus</a>. R package version 3.0.2.
</div>
<div id="ref-jin2020perfluoroalkyl" class="csl-entry" role="listitem">
R. Jin, R. McConnell, C. Catherine, S. Xu, D. I. Walker, N. Stratakis, D. P. Jones, G. W. Miller, C. Peng, D. V. Conti, et al. Perfluoroalkyl substances and severity of nonalcoholic fatty liver in children: An untargeted metabolomics approach. <em>Environment International</em>, 134: 105220, 2020.
</div>
<div id="ref-khan2001classification" class="csl-entry" role="listitem">
J. Khan, J. S. Wei, M. Ringner, L. H. Saal, M. Ladanyi, F. Westermann, F. Berthold, M. Schwab, C. R. Antonescu, C. Peterson, et al. Classification and diagnostic prediction of cancers using gene expression profiling and artificial neural networks. <em>Nature Medicine</em>, 7(6): 673–679, 2001.
</div>
<div id="ref-kirk2012bayesian" class="csl-entry" role="listitem">
P. Kirk, J. E. Griffin, R. S. Savage, Z. Ghahramani and D. L. Wild. Bayesian correlated clustering to integrate multiple datasets. <em>Bioinformatics</em>, 28(24): 3290–3297, 2012.
</div>
<div id="ref-lock2013bayesian" class="csl-entry" role="listitem">
E. F. Lock and D. B. Dunson. Bayesian consensus clustering. <em>Bioinformatics</em>, 29(20): 2610–2616, 2013.
</div>
<div id="ref-lock2013joint" class="csl-entry" role="listitem">
E. F. Lock, K. A. Hoadley, J. S. Marron and A. B. Nobel. Joint and individual variation explained (JIVE) for integrated analysis of multiple data types. <em>The Annals of Applied Statistics</em>, 7(1): 523, 2013.
</div>
<div id="ref-maitre2022state" class="csl-entry" role="listitem">
L. Maitre, J.-B. Guimbaud, C. Warembourg, N. Güil-Oumrait, T. E. D. C. P. Consortium, P. M. Petrone, M. Chadeau-Hyam, M. Vrijheid, J. R. Gonzalez and X. Basagaña. State-of-the-art methods for exposure-health studies: Results from the exposome data challenge event. <em>arXiv preprint arXiv:2202.01680</em>, 2022.
</div>
<div id="ref-matta2022associations" class="csl-entry" role="listitem">
K. Matta, T. Lefebvre, E. Vigneau, V. Cariou, P. Marchand, Y. Guitton, A.-L. Royer, S. Ploteau, B. Le Bizec, J.-P. Antignac, et al. Associations between persistent organic pollutants and endometriosis: A multiblock approach integrating metabolic and cytokine profiling. <em>Environment International</em>, 158: 106926, 2022.
</div>
<div id="ref-mo2013pattern" class="csl-entry" role="listitem">
Q. Mo, S. Wang, V. E. Seshan, A. B. Olshen, N. Schultz, C. Sander, R. S. Powers, M. Ladanyi and R. Shen. Pattern discovery and cancer gene identification in integrated cancer genomic data. <em>Proceedings of the National Academy of Sciences</em>, 110(11): 4245–4250, 2013.
</div>
<div id="ref-ozsolak2011rna" class="csl-entry" role="listitem">
F. Ozsolak and P. M. Milos. RNA sequencing: Advances, challenges and opportunities. <em>Nature Reviews Genetics</em>, 12(2): 87–98, 2011.
</div>
<div id="ref-pan2007penalized" class="csl-entry" role="listitem">
W. Pan and X. Shen. Penalized model-based clustering with application to variable selection. <em>Journal of Machine Learning Research</em>, 8(May): 1145–1164, 2007.
</div>
<div id="ref-peng2020latent" class="csl-entry" role="listitem">
C. Peng, J. Wang, I. Asante, S. Louie, R. Jin, L. Chatzi, G. Casey, D. C. Thomas and D. V. Conti. A latent unknown clustering integrating multi-omics data (LUCID) with phenotypic traits. <em>Bioinformatics</em>, 36(3): 842–850, 2020.
</div>
<div id="ref-pierre2020clustering" class="csl-entry" role="listitem">
M. Pierre-Jean, J.-F. Deleuze, E. Le Floch and F. Mauger. Clustering and variable selection evaluation of 13 unsupervised methods for multi-omics data integration. <em>Briefings in Bioinformatics</em>, 21(6): 2011–2030, 2020.
</div>
<div id="ref-prabhakaran2016dirichlet" class="csl-entry" role="listitem">
S. Prabhakaran, E. Azizi, A. Carr and D. Pe’er. Dirichlet process mixture model for correcting technical variation in single-cell gene expression data. In <em>International conference on machine learning</em>, pages. 1070–1079 2016. PMLR.
</div>
<div id="ref-rappoport2018multi" class="csl-entry" role="listitem">
N. Rappoport and R. Shamir. Multi-omic and multi-view clustering algorithms: Review and cancer benchmark. <em>Nucleic Acids Research</em>, 46(20): 10546–10562, 2018.
</div>
<div id="ref-schmidt2008sankey" class="csl-entry" role="listitem">
M. Schmidt. The sankey diagram in energy and material flow management: Part II: Methodology and current applications. <em>Journal of Industrial Ecology</em>, 12(2): 173–185, 2008.
</div>
<div id="ref-scrucca2016mclust" class="csl-entry" role="listitem">
L. Scrucca, M. Fop, T. B. Murphy and A. E. Raftery. Mclust 5: Clustering, classification and density estimation using gaussian finite mixture models. <em>The R Journal</em>, 8(1): 289, 2016.
</div>
<div id="ref-singh2019diablo" class="csl-entry" role="listitem">
A. Singh, C. P. Shannon, B. Gautier, F. Rohart, M. Vacher, S. J. Tebbutt and K.-A. Lê Cao. DIABLO: An integrative approach for identifying key molecular drivers from multi-omics assays. <em>Bioinformatics</em>, 35(17): 3055–3062, 2019.
</div>
<div id="ref-song2020bayesian" class="csl-entry" role="listitem">
Y. Song, X. Zhou, M. Zhang, W. Zhao, Y. Liu, S. L. Kardia, A. V. D. Roux, B. L. Needham, J. A. Smith and B. Mukherjee. Bayesian shrinkage estimation of high dimensional causal mediation effects in omics studies. <em>Biometrics</em>, 76(3): 700–710, 2020.
</div>
<div id="ref-stratakis2020prenatal" class="csl-entry" role="listitem">
N. Stratakis, D. V. Conti, R. Jin, K. Margetaki, D. Valvi, A. P. Siskos, L. Maitre, E. Garcia, N. Varo, Y. Zhao, et al. Prenatal exposure to perfluoroalkyl substances associated with increased susceptibility to liver injury in children. <em>Hepatology</em>, 72(5): 1758–1770, 2020.
</div>
<div id="ref-subramanian2020multi" class="csl-entry" role="listitem">
I. Subramanian, S. Verma, S. Kumar, A. Jere and K. Anamika. Multi-omics data integration, interpretation, and its application. <em>Bioinformatics and Biology Insights</em>, 14: 1177932219899051, 2020.
</div>
<div id="ref-tenenhaus2014variable" class="csl-entry" role="listitem">
A. Tenenhaus, C. Philippe, V. Guillemot, K.-A. Le Cao, J. Grill and V. Frouin. Variable selection for generalized canonical correlation analysis. <em>Biostatistics</em>, 15(3): 569–583, 2014.
</div>
<div id="ref-tibshirani1996regression" class="csl-entry" role="listitem">
R. Tibshirani. Regression selection and shrinkage via the lasso. <em>Journal of the Royal Statistical Society: Series B</em>, 58(1): 267–288, 1996.
</div>
<div id="ref-visdat" class="csl-entry" role="listitem">
N. Tierney. <em>Visdat: Visualising whole data frames including missing values.</em> 2023. URL <a href="https://cran.r-project.org/package=visdat">https://cran.r-project.org/package=visdat</a>. R package version 0.5.3.
</div>
<div id="ref-tini2019multi" class="csl-entry" role="listitem">
G. Tini, L. Marchetti, C. Priami and M.-P. Scott-Boyer. Multi-omics integration—a comparison of unsupervised clustering methodologies. <em>Briefings in Bioinformatics</em>, 20(4): 1269–1279, 2019.
</div>
<div id="ref-voillet2016handling" class="csl-entry" role="listitem">
V. Voillet, P. Besse, L. Liaubet, M. San Cristobal and I. González. Handling missing rows in multi-omics data integration: Multiple imputation in multiple factor analysis framework. <em>BMC Bioinformatics</em>, 17(1): 1–16, 2016.
</div>
<div id="ref-vrijheid2014human" class="csl-entry" role="listitem">
M. Vrijheid, R. Slama, O. Robinson, L. Chatzi, M. Coen, P. Van den Hazel, C. Thomsen, J. Wright, T. J. Athersuch, N. Avellana, et al. The human early-life exposome (HELIX): Project rationale and design. <em>Environmental Health Perspectives</em>, 122(6): 535–544, 2014.
</div>
<div id="ref-wang2014similarity" class="csl-entry" role="listitem">
B. Wang, A. M. Mezlini, F. Demir, M. Fiume, Z. Tu, M. Brudno, B. Haibe-Kains and A. Goldenberg. Similarity network fusion for aggregating data types on a genomic scale. <em>Nature Methods</em>, 11(3): 333–337, 2014.
</div>
<div id="ref-wild2005complementing" class="csl-entry" role="listitem">
C. P. Wild. Complementing the genome with an <span>“exposome”</span>: The outstanding challenge of environmental exposure measurement in molecular epidemiology. <em>Cancer Epidemiology, Biomarkers &amp; Prevention</em>, 14(8): 1847–1850, 2005.
</div>
<div id="ref-yu2022dlstats" class="csl-entry" role="listitem">
G. Yu. <em>Dlstats: Download stats of r packages.</em> 2022. URL <a href="https://CRAN.R-project.org/package=dlstats">https://CRAN.R-project.org/package=dlstats</a>. R package version 0.1.5.
</div>
<div id="ref-yuan2006model" class="csl-entry" role="listitem">
M. Yuan and Y. Lin. Model selection and estimation in regression with grouped variables. <em>Journal of the Royal Statistical Society: Series B (Statistical Methodology)</em>, 68(1): 49–67, 2006.
</div>
<div id="ref-zhang2016estimating" class="csl-entry" role="listitem">
H. Zhang, Y. Zheng, Z. Zhang, T. Gao, B. Joyce, G. Yoon, W. Zhang, J. Schwartz, A. Just, E. Colicino, et al. Estimating and testing high-dimensional mediation effects in epigenetic studies. <em>Bioinformatics</em>, 32(20): 3150–3154, 2016.
</div>
<div id="ref-zhang2021gaussian" class="csl-entry" role="listitem">
Y. Zhang, M. Li, S. Wang, S. Dai, L. Luo, E. Zhu, H. Xu, X. Zhu, C. Yao and H. Zhou. Gaussian mixture model clustering with incomplete data. <em>ACM Transactions on Multimedia Computing, Communications, and Applications (TOMM)</em>, 17(1s): 1–14, 2021.
</div>
<div id="ref-zhou2009penalized" class="csl-entry" role="listitem">
H. Zhou, W. Pan and X. Shen. Penalized model-based clustering with unconstrained covariance matrices. <em>Electronic Journal of Statistics</em>, 3: 1473, 2009.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
  <h3 id="references">References</h3>
  <div id="references-listing"></div>
  <h3 id="reuse">Reuse</h3>
  <p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
  <h3 id="citation">Citation</h3>
  <p>For attribution, please cite this work as</p>
  <pre class="citation-appendix short">Zhao, et al., "LUCIDus: An R Package For Implementing Latent Unknown Clustering By Integrating Multi-omics Data (LUCID) With Phenotypic Traits", The R Journal, 2025</pre>
  <p>BibTeX citation</p>
  <pre class="citation-appendix long">@article{RJ-2024-012,
  author = {Zhao, Yinqi and Jia, Qiran and Goodrich, Jesse A. and Conti, David V.},
  title = {LUCIDus: An R Package For Implementing Latent Unknown Clustering By Integrating Multi-omics Data (LUCID) With Phenotypic Traits},
  journal = {The R Journal},
  year = {2025},
  note = {https://doi.org/10.32614/RJ-2024-012},
  doi = {10.32614/RJ-2024-012},
  volume = {16},
  issue = {2},
  issn = {2073-4859},
  pages = {4-26}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<!--radix_placeholder_navigation_after_body--><html><body>
<div class="distill-site-nav distill-site-footer">
<p>© The R Foundation, <a href="mailto:r-journal@r-project.org">web page
contact</a>.</p>
</div>
<!--/radix_placeholder_navigation_after_body-->
</body></html>


</body>

</html>
