---
title: Changes on CRAN
abstract: The 'Changes on CRAN' article from the 2009-2 issue.
author:
- name: Kurt Hornik
  affiliation: WU Wirtschaftsuniversität Wien, Austria
- name: Achim Zeileis
  affiliation: WU Wirtschaftsuniversität Wien, Austria
date: '2009-12-01'
date_received: ~
journal:
  firstpage: 80
  lastpage: 94
volume: 1
issue: 2
slug: RJ-2009-2-cran
citation_url: https://rjournal.github.io/
packages:
  cran: []
  bioc: []
preview: preview.png
bibliography: ~
CTV: ~
output:
  rjtools::rjournal_web_article:
    self_contained: yes
    toc: no
    legacy_pdf: yes
    web_only: yes
    mathjax: https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js
    md_extension: -tex_math_single_backslash

---

::: article
# New CRAN task views

[*ClinicalTrials*](https://CRAN.R-project.org/view=ClinicalTrials)

:   Design, Monitoring, and Analysis of Clinical Trials.\
    Packages: *ClinicalRobustPriors*$^*$, *GroupSeq*$^*$, *HH*,
    *Hmisc*$^*$, *MChtest*$^*$, *PwrGSD*$^*$, *asypow*,
    *bifactorial*$^*$, *binomSamSize*, *blockrand*$^*$, *clinfun*$^*$,
    *coin*, *copas*, *epibasix*, *epicalc*, *experiment*$^*$,
    *gsDesign*$^*$, *ldbounds*$^*$, *meta*, *metafor*, *multcomp*,
    *rmeta*, *seqmon*$^*$, *speff2trial*$^*$, *ssanv*, *survival*$^*$.\
    Maintainer: Ed Zhang.

[*MedicalImaging*](https://CRAN.R-project.org/view=MedicalImaging)

:   Medical Image Analysis.\
    Packages: *AnalyzeFMRI*$^*$, *DICOM*$^*$, *PET*$^*$,
    *Rniftilib*$^*$, *VR*, *adimpro*$^*$, *bitops*, *dcemri*$^*$,
    *dti*$^*$, *fmri*$^*$, *minpack.lm*, *tractor.base*$^*$.\
    Maintainer: Brandon Whitcher.

(\* = core package)

# New packages in CRAN task views

[*Bayesian*](https://CRAN.R-project.org/view=Bayesian)

:   *BAS*, *glmmBUGS*.

[*ChemPhys*](https://CRAN.R-project.org/view=ChemPhys)

:   *AquaEnv*, *CHNOSZ*, *NMRS*, *gpls*, *nlreg*, *paltran*, *plspm*,
    *quantchem*, *spls*.

[*Cluster*](https://CRAN.R-project.org/view=Cluster)

:   *nnclust*.

[*Distributions*](https://CRAN.R-project.org/view=Distributions)

:   *Lmoments*, *VarianceGamma*, *denstrip*, *evdbayes*, *evir*,
    *fExtremes*, *fitdistrplus*, *gamlss.dist*$^*$, *gamlss.mx*, *ig*.

[*Econometrics*](https://CRAN.R-project.org/view=Econometrics)

:   *CADFtest*, *Mcomp*, *Zelig*, *aod*, *expsmooth*, *fma*, *forecast*.

[*Environmetrics*](https://CRAN.R-project.org/view=Environmetrics)

:   *aod*, *primer*, *vegetarian*.

[*ExperimentalDesign*](https://CRAN.R-project.org/view=ExperimentalDesign)

:   *DoE.base*$^*$, *DoE.wrapper*$^*$, *RcmdrPlugin.DoE*.

[*Finance*](https://CRAN.R-project.org/view=Finance)

:   *YieldCurve*, *atmi*, *forecast*, *ttrTests*.

[*HighPerformanceComputing*](https://CRAN.R-project.org/view=HighPerformanceComputing)

:   *cudaBayesreg*, *doMC*, *doSNOW*, *foreach*, *speedglm*.

[*MachineLearning*](https://CRAN.R-project.org/view=MachineLearning)

:   *penalizedSVM*.

[*NaturalLanguageProcessing*](https://CRAN.R-project.org/view=NaturalLanguageProcessing)

:   *openNLPmodels.en*, *openNLPmodels.es*.

[*Optimization*](https://CRAN.R-project.org/view=Optimization)

:   *Rcsdp*, *nleqslv*.

[*Psychometrics*](https://CRAN.R-project.org/view=Psychometrics)

:   *MLCM*, *SEMModComp*, *irtProb*, *latdiag*.

[*Spatial*](https://CRAN.R-project.org/view=Spatial)

:   *RgoogleMaps*, *nlme*.

[*Survival*](https://CRAN.R-project.org/view=Survival)

:   *AER*, *BMA*, *BayHaz*, *DAAG*, *Epi*, *ICE*, *LearnBayes*,
    *LogicReg*, *MAMSE*, *MCMCglmm*, *MCMCpack*, *OrdFacReg*, *SMIR*,
    *SMPracticals*, *SimHap*, *VGAM*, *XReg*, *clinfun*, *cmprskContin*,
    *coin*, *condGEE*, *coxme*, *epiR*, *etm*, *fitdistrplus*, *gbm*,
    *gof*, *gss*, *interval*, *ipred*, *km.ci*, *kmi*, *lmec*, *locfit*,
    *logspline*, *maxstat*, *mixAK*, *mstate*, *multcomp*, *multtest*,
    *mvpart*, *nltm*, *p3state.msm*, *pamr*, *party*, *penalized*,
    *phmm*, *polspline*, *quantreg*, *rankhazard*, *rhosp*,
    *risksetROC*, *rms*$^*$, *spatstat*, *superpc*, *survcomp*,
    *survey*, *uniCox*.

[*TimeSeries*](https://CRAN.R-project.org/view=TimeSeries)

:   *CADFtest*, *EvalEst*, *KFAS*, *Mcomp*, *dlnm*, *expsmooth*, *fma*,
    *forecast*$^*$, *fractalrock*, *mar1s*, *tiger*, *tis*.

(\* = core package)

# New contributed packages

*ADGofTest*

:   Anderson-Darling GoF test with p-value calculation based on
    Marsaglia's 2004 paper "Evaluating the Anderson-Darling
    Distribution". By Carlos J. Gil Bellosta.

*AGSDest*

:   Estimation in adaptive group sequential trials. By Niklas Hack and
    Werner Brannath.

*CDFt*

:   Statistical downscaling through CDF transform. Also performs
    computation of the Cramèr-von Mises and Kolmogorov-Smirnov
    statistics. By Mathieu Vrac and Paul-Antoine Michelangeli.

*COP*

:   Variables selection for index models via correlation pursuit. By
    Wenxuan Zhong.

*CORElearn*

:   A machine learning suite for classification, regression, feature
    evaluation and ordinal evaluation, based on C++ code. Contains
    several model learning techniques in classification and regression,
    for example decision and regression trees with optional constructive
    induction and models in the leafs, random forests, kNN, naive Bayes,
    and locally weighted regression. Especially strong in feature
    evaluation algorithms where it contains several variants of Relief
    algorithm and many impurity based attribute evaluation functions,
    e.g., Gini, information gain, MDL, DKM, etc. Additional strengths
    are its ordEval algorithm and its visualization used for ordinal
    features and classes. By Marko Robnik-Sikonja and Petr Savicky.

*CalciOMatic*

:   Simulate and analyze calcium imaging data obtained with ratiometric
    dyes. By Sebastien Joucla, Christophe Pouzat.

*CircNNTSR*

:   Statistical analysis of circular data using non-negative
    trigonometric sums (NNTS) models. Includes functions for calculation
    of densities and distributions, the estimation of parameters,
    plotting and more. By Juan José Fernández-Durán and Maria Mercedes
    Gregorio-Domínguez.

*CircSpatial*

:   A collection of functions for color continuous high resolution
    images of circular spatial data, circular kriging, and simulation of
    circular random fields. By Bill Morphet.

*DRI*

:   DR-Integrator: Integrative analysis of DNA copy number and gene
    expression data described in Salari et al (2009). By Keyan Salari,
    Robert Tibshirani, and Jonathan R. Pollack.

*DTDA*

:   Doubly Truncated Data Analysis. Implements different algorithms for
    analyzing randomly, one-sided and two-sided (i.e., doubly) truncated
    data. By Carla Moreira, Jacobo de Uña-Álvarez and Rosa Crujeiras.

*Daim*

:   Diagnostic accuracy of classification models. Several functions for
    evaluating the accuracy of classification models. Provides
    performance measures 'cv', '0.632' and '0.632+', estimation of the
    misclassification rate, sensitivity, specificity and AUC. If an
    application is computationally intensive, parallel execution can be
    used to reduce the time taken. By Sergej Potapov, Werner Adler and
    Berthold Lausen.

*Deducer*

:   An intuitive graphical data analysis system for use with *JGR*. By
    Ian Fellows.

*DiversitySampler*

:   Functions for re-sampling a community matrix to compute Shannon's
    Diversity index at different sampling levels. By Matthew K. Lau.

*DoE.base*

:   Creation of full factorial experimental designs and designs based on
    orthogonal arrays for (industrial) experiments. Additionally
    provides some utility functions used also by other DoE packages. By
    Ulrike

*DoE.wrapper*

:   Wrapper package for design of experiments functionality. Creates
    various kinds of designs for (industrial) experiments by using and
    possibly enhancing design generation routines from other packages.
    Currently, response surface designs from package *rsm* and latin
    hypercube samples from package *lhs* have been implemented. By
    Ulrike

*EQL*

:   Extended Quasi Likelihood function (EQL). Computation of the EQL for
    a given family of variance functions, Saddlepoint-approximations and
    related auxiliary functions (e.g., Hermite polynomials). By Thorn
    Thaler.

*ElectroGraph*

:   Enhanced routines for plotting and analyzing valued relational data,
    considering valued ties. In particular, relative distances are
    calculated using social conductance methods. By Andrew C. Thomas.

*EnQuireR*

:   A package dedicated to questionnaires. By Fournier Gwenaëlle,
    Cadoret Marine, Fournier Olivier, Le Poder François, Bouche Jérôme,
    and Lê Sébastien.

*EvalEst*

:   Dynamic Systems Estimation (DSE) extensions. By Paul Gilbert.

*FEST*

:   Identification of family relations using linked markers. By Øivind
    Skare.

*FME*

:   A Flexible Modeling Environment for inverse modeling, sensitivity,
    identifiability, Monte Carlo analysis. Intended to work with models
    written as a set of differential equations that are solved either by
    an integration routine from package *deSolve*, or a steady-state
    solver from package *rootSolve*, but can also be used with other
    types of functions. By Karline Soetaert and Thomas Petzoldt.

*FRB*

:   Robust inference based on applying Fast and Robust Bootstrap on
    robust estimators. Available methods are multivariate regression,
    PCA and Hotelling tests. By Ella Roelant, Stefan Van Aelst and Gert
    Willems.

*GGMselect*

:   Gaussian Graphs Models selection: graph estimation in Gaussian
    Graphical Models. The main functions return the adjacency matrix of
    an undirected graph estimated from a data matrix. By Annie Bouvier,
    Christophe Giraud, Sylvie Huet, and Verzelen N.

*GLMMarp*

:   Generalized Linear Multilevel Model with AR(p) errors. Functions to
    estimate the GLMM-AR(p) model for analyzing discrete time-series
    cross-sectional data via MCMC simulation. Also contains several
    useful utility functions, including an independent function for
    computing the Bayes factor with GLMM-AR(p) output, a function to
    recover the random coefficients at the individual level, and a
    function to do prediction by using the posterior distributions. By
    Xun Pang.

*GOFSN*

:   Goodness-Of-Fit tests for the family of Skew-Normal models.
    Implements a method for checking if a skew-normal model fits the
    observed dataset, when all parameters are unknown. While location
    and scale parameters are estimated by moment estimators, the shape
    parameter is integrated with respect to the prior predictive
    distribution, as proposed in Box (1980). A default and proper prior
    on skewness parameter is used to obtain the prior predictive
    distribution, as proposed in Cabras, & Castellanos (2008). By
    Veronica Paton Romero.

*GWAF*

:   Genome-Wide Association analyses with Family data: Functions to test
    genetic associations between SNPs and a continuous/dichotomous trait
    using family data, and to make genome-wide p-value plots and QQ
    plots. By Ming-Huei Chen and Qiong Yang.

*GeneReg*

:   Infer time delay gene regulatory networks using time course gene
    expression profiles. The main idea of the time delay linear model is
    to fit a linear regression model using a set of putative regulators
    to estimate the transcription pattern of a specific target gene. By
    Tao Huang.

*Geneclust*

:   Simulation and analysis of spatial structure of population genetics
    data. By Sophie Ancelet.

*HMR*

:   Statistical analysis of static chamber concentration data for trace
    gas flux estimation. By Asger R. Pedersen.

*HSAUR2*

:   A Handbook of Statistical Analyses Using R (2nd Edition). Functions,
    data sets, analyses and examples from the second edition of the book
    "A Handbook of Statistical Analyses Using R" (Brian S. Everitt and
    Torsten Hothorn, Chapman & Hall/CRC, 2008). By Brian S. Everitt and
    Torsten Hothorn.

*HTMLUtils*

:   Facilitates automated [HTML]{.smallcaps} report creation, in
    particular framed [HTML]{.smallcaps} pages and dynamically sortable
    tables. By Markus Loecher.

*HWEBayes*

:   Bayesian investigation of Hardy-Weinberg Equilibrium via estimation
    and testing. Three models are currently considered: HWE, a model
    parametrized in terms of the allele frequencies and a single
    inbreeding coefficient f, and the saturated model. Testing is based
    on Bayes factors. By Jon Wakefield.

*Haplin*

:   Performs a genetic association analysis of case-parent triad (trio)
    data with multiple markers. Can also incorporate complete or
    incomplete control triads, for instance independent control
    children. Estimation is based on haplotypes, for instance SNP
    haplotypes, even though phase is not known from the genetic data.
    Estimates relative risk and p-values associated with each haplotype.
    Uses MLE to make optimal use of data from triads with missing
    genotypic data, for instance if some SNPs has not been typed for
    some individuals. Also allows estimation of effects of maternal
    haplotypes, particularly appropriate in perinatal epidemiology.
    By H. K. Gjessing.

*HybridMC*

:   An implementation of the Hybrid Monte Carlo and Multipoint Hybrid
    Monte Carlo sampling techniques described in Liu (2001), "Monte
    Carlo Strategies in Computing". By Richard D. Morey.

*IsoGene*

:   Testing for monotonic relationship between gene expression and doses
    in a microarray experiment. Several testing procedures including the
    global likelihood ratio test (Bartholomew, 1961), Williams (1971,
    1972), Marcus (1976), M (Hu et al. 2005) and the modified M (Lin et
    al. 2007) are used to test for the monotonic trend in gene
    expression with respect to doses. BH (Benjamini and Hochberg 1995)
    and BY (Benjamini and Yekutilie 2004) FDR controlling procedures are
    applied to adjust the raw p-values obtained from the permutations.
    By Dan Lin et al.

*KFAS*

:   Functions for fast Kalman filtering, state and disturbance
    smoothing, forecasting and simulation of multivariate time-variant
    state space models. All functions can use exact diffuse
    initialization when distributions of some or all elements of initial
    state vector are unknown. Filtering, state smoothing and simulation
    functions use the sequential processing algorithm, which is faster
    than standard approach, and also allows singularity of the
    prediction error variance matrix. By Jouni Helske.

*LLdecomp*

:   Decomposes a set of variables into cliques and separators depending
    on their association which is measured using Random Forests. By
    Corinne Dahinden.

*MCMChybridGP*

:   Hybrid Markov chain Monte Carlo using Gaussian Processes. Uses
    hybrid MCMC to simulate from a multimodal target distribution. A
    Gaussian process approximation makes this possible when derivatives
    are unknown. Serves to minimize the number of function evaluations
    in Bayesian calibration of computer models using parallel tempering.
    Allows replacement of the true target distribution in high
    temperature chains, or complete replacement of the target. Mark J.
    Fielding.

*MLCM*

:   Maximum Likelihood Conjoint Measurement. Conjoint measurement is a
    psychophysical procedure in which stimulus pairs are presented that
    vary along 2 or more dimensions and the observer is required to
    compare the stimuli along one of them. This package contains
    functions to estimate the contribution of the $n$ scales to the
    judgment by a maximum likelihood method under several hypotheses of
    how the perceptual dimensions interact. By Kenneth Knoblauch and
    Laurence T. Maloney.

*MMIX*

:   Implement different types of model mixing and model selection
    methods for linear or logistic models. By Marie Morfin and David
    Makowski.

*NeatMap*

:   Non-clustered heatmap alternatives. Creates heatmap like plots in 2
    and 3 dimensions, without the need for cluster analysis. Like the
    heatmap, the plots created display both a dimensionally reduced
    representation of the data as well as the data itself. They are
    intended to be used in conjunction with dimensional reduction
    techniques such as PCA. By Satwik Rajaram and Yoshi Oono.

*ORIClust*

:   Order-restricted Information Criterion-based Clustering of genes.
    Clusters are given by genes matched to prespecified profiles across
    various ordered treatment groups. Particularly useful for analyzing
    data obtained from short time-course or dose-response microarray
    experiments. By Tianqing Liu, Nan Lin, Ningzhong Shi and Baoxue
    Zhang.

*OjaNP*

:   Functions for the Oja median, Oja signs and ranks and methods based
    upon them. By Daniel Fischer, Jyrki Klaus Nordhausen and Daniel
    Vogel.

*PCIT*

:   Partial Correlation Coefficient with Information Theory. Provides
    the PCIT algorithm developed by Reverter and Chan (2008) which
    identifies significant gene to gene associations to define edges in
    a weighted network. By Nathan S. Watson-Haigh.

*PCS*

:   Calculate the probability of correct selection (PCS). Given $k$
    populations (can be in thousands), what is the probability that a
    given subset of size $t$ contains the true top $t$ populations? This
    package finds this probability and offers three tuning parameters
    ($G$, $d$, $L$) to relax the definition. By Jason Wilson.

*PLIS*

:   Multiplicity control using Pooled LIS (PLIS) statistics. PLIS is a
    multiple testing procedure for testing several groups of hypotheses.
    Linear dependency is expected from the hypotheses within the same
    group and is modeled by hidden Markov Models. It is noted that, for
    PLIS, a smaller p value does not necessarily imply more significance
    because of dependency among the hypotheses. A typical application of
    PLIS is to analyze genome wide association studies datasets, where
    SNPs from the same chromosome are treated as a group and exhibit
    strong linear genomic dependency. By Zhi Wei & Wenguang Sun.

*PearsonDS*

:   Implementation of the Pearson distribution system, including full
    support for the (d,p,q,r)-family of functions for probability
    distributions and fitting via method of moments and maximum
    likelihood method. By Martin Becker.

*QCA3*

:   Functions for Qualitative Comparative Analysis (QCA). Can be used
    for all three types of QCA. Has methods for simplifying assumption
    and contradictory simplifying assumption, and can return constrained
    results by including or excluding specific conditions. By Ronggui
    Huang.

*R.filesets*

:   Easy handling of and access to files organized in structured
    directories. A file set refers to a set of files located in one or
    more directories on the file system. This package provides classes
    and methods to locate, setup, subset, navigate and iterative over
    such sets. The API is designed such that these classes can be
    subsetted to provide for instance a richer API for special file
    formats. By Henrik Bengtsson.

*R2PPT*

:   Simple R Interface to Microsoft PowerPoint using *rcom*. By Wayne
    Jones.

*R2wd*

:   Write MS-Word documents from R, using the statconnDCOM server to
    communicate with MS-Word via the COM interface. By Christian Ritter.

*RBerkeley*

:   Interface to Embedded Oracle Berkeley DB(tm). By Jeffrey A. Ryan.

*RC*

:   Reproducible Computing. Allows the user to create and use
    reproducible computations for the purpose of research and education.
    The meta data about the computations are stored in a remote
    repository which is hosted at
    [www.freestatistics.org](http://www.freestatistics.org){.uri}. By
    Patrick Wessa.

*REEMtree*

:   Estimates regression trees with random effects as a way to use data
    mining techniques to describe longitudinal or panel data. By Rebecca
    Sela and Jeffrey Simonoff.

*RImageJ*

:   Bindings between R and the ImageJ Java based image processing and
    analysis platform. By Romain François and Philippe Grosjean.

*RInside*

:   C++ classes to embed R in C++ applications. By Dirk Eddelbuettel.

*RLastFM*

:   R interface to last.fm [API]{.smallcaps}. By Greg Hirson.

*RPMM*

:   Recursively Partitioned Mixture Model for Beta and Gaussian
    Mixtures. This is a model-based clustering algorithm that returns a
    hierarchy of classes, similar to hierarchical clustering, but also
    similar to finite mixture models. By E. Andres Houseman.

*RSiena*

:   Simulation Investigation for Empirical Network Analysis. Fits models
    to longitudinal networks. By various authors.

*Rassoc*

:   Robust tests for case-control genetic association studies: allelic
    based test, Cochran-Armitage trend test, maximin efficiency robust
    test, MAX3 test and genetic model selection test. By Yong Zang,
    Wingkam Fung and Gang Zheng.

*RcmdrPlugin.DoE*

:   *Rcmdr* plugin for (industrial) Design of Experiments. Currently in
    beta status. By Ulrike

*RcmdrPlugin.qual*

:   An *Rcmdr* plug-in based on the Quality control class Stat 4300. By
    Erin Hodgess.

*ReacTran*

:   Routines for developing models that describe reaction and
    advective-diffusive transport in one, two or three dimensions.
    Includes transport routines in porous media, in estuaries, and in
    bodies with variable shape. By Karline Soetaert and Filip Meysman.

*ReadImages*

:   Functions for reading JPEG and PNG files, requiring libjpeg. By
    Markus Loecher.

*RelativeRisk*

:   Relative risk estimation for prospective and retrospective data. By
    Bob Wheeler.

*Rmpfr*

:   R MPFR - Multiple Precision Floating-Point Reliable. Aims to provide
    S4 classes and methods for arithmetic including transcendental
    ("special") functions for arbitrary precision floating point
    numbers. To this end, it interfaces to the LGPL'ed MPFR (Multiple
    Precision Floating-Point Reliable) Library which itself is based on
    the GMP (GNU Multiple Precision) Library. By Martin Mächler.

*Rniftilib*

:   R interface to nifticlib to read/write ANALYZE(TM)7.5/NIfTI-1 volume
    images. By Oliver Granert.

*RthroughExcelWorkbooksInstaller*

:   Workbooks in Excel that illustrate statistical concepts by accessing
    R functions from Excel. These workbooks use the automatic
    recalculation mode of Excel to update calculations and graphs in R.
    Downloads an executable which installs the workbooks on MS Windows
    systems where *RExcel* has already been installed. By Richard M.
    Heiberger and Erich Neuwirth.

*SAFD*

:   Statistical Analysis of Fuzzy Data. Aims to provide some basic
    functions for doing statistics with one dimensional fuzzy data (in
    the form of polygonal fuzzy numbers). Contains functions for the
    basic operations on the class of fuzzy numbers (sum, scalar product,
    mean, Hukuhara difference) as well as for calculating (Bertoluzza)
    distance, sample variance, sample covariance, sample correlation,
    and the Dempster-Shafer (levelwise) histogram. Includes
    functionality to simulate fuzzy random variables, bootstrap tests
    for the equality of means, and to do linear regression given
    trapezoidal fuzzy data. By Wolfgang Trutschnig and Asun Lubiano.

*SEL*

:   Semiparametric elicitation. Implements a novel method for
    transferring expert statements about an uncertain bounded quantity
    into a probability distribution (see Bornkamp and Ickstadt (2009)
    for a detailed description). For this purpose B-splines are used,
    and the density is obtained by penalized least squares, where the
    penalty encourages to distribute probability mass as uniformly as
    possible. Provides methods for fitting the expert's distribution as
    well as methods for evaluating the underlying density and cdf. In
    addition methods for plotting the expert's distribution, drawing
    random numbers and calculating quantiles of the expert's
    distribution are provided. By Bornkamp.

*SHARE*

:   SNP-Haplotype Adaptive REgression (SHARE): an adaptive algorithm to
    select the most informative set of SNPs for genetic association. By
    James Y. Dai.

*SIS*

:   (Iterative) Sure Independence Screening for Generalized Linear
    Models and Cox's Proportional Hazards models. By Jianqing Fan, Yang
    Feng, Richard Samworth and Yichao Wu.

*SWordInstaller*

:   SWord: Use R in Microsoft Word (Installer). Creating articles and
    reports in Word is easy. Adding R results (figures, tables,
    summaries) requires manually copying the data from R to Word.
    Changing the data requires redoing the analyses in R and
    re-inserting the results into Word manually. SWord integrates R
    scripts and results into Word documents. Documents can be edited and
    read even without R installed (or without any knowledge of R). The
    functionality of embedding the scripts is somewhat similar to what
    Sweave does for LaTeX documents. By Thomas Baier.

*SigWinR*

:   Implementation of the SigWin-detector for the detection of regions
    of increased or decreased gene expression (RIDGEs and anti-RIDGES)
    in transcriptome maps and the presentation in so called RIDGEOGRAMs
    as described by Marcia A. Inda et al. (2008). By Wim de Leeuw.

*TripleR*

:   Social Relation Model (SRM) analyses for single round-robin groups.
    These analyses are either based on one manifest variable, one latent
    construct measured by two manifest variables, two manifest variables
    and their bivariate relations, or two latent constructs each
    measured by two manifest variables (in the last case, four
    Round-Robin matrices have to be provided). By S.C. Schmukle, F.D.
    Schoenbrodt and M.D. Back.

*WMCapacity*

:   A GUI implementation of hierarchical Bayesian models of working
    memory, used for analyzing change detection data. By Richard D.
    Morey.

*YieldCurve*

:   Modeling and estimation of the yield curve, implementing the
    Nelson-Siegel, Diebold-Li and Svensson models. Also includes the
    data of the term structure of interest rate of Federal Reserve and
    European Central Bank. By Sergio Salvino Guirreri.

*aCGH.Spline*

:   Robust spline interpolation for dual color array comparative genomic
    hybridisation data. By Tomas William Fitzgerald.

*adaptTest*

:   Adaptive two-stage tests. Currently, four tests are included: Bauer
    and Koehne (1994), Lehmacher and Wassmer (1999), Vandemeulebroecke
    (2006), and the horizontal conditional error function. By Marc
    Vandemeulebroecke.

*amer*

:   Fit generalized additive mixed models based on the mixed model
    algorithm of *lme4*. By Fabian Scheipl.

*anchors*

:   Statistical analysis of surveys with anchoring vignettes. By
    Jonathan Wand, Gary King, and Olivia Lau.

*ant*

:   Version of the ant apache build tool, with a few R specific tasks to
    ease use of ant within an R package. By Romain François.

*aroma.affymetrix*

:   Analysis of large Affymetrix microarray data sets. Implements
    classes for files and sets of files for various Affymetrix file
    formats, e.g., AffymetrixCdfFile, AffymetrixCelFile, and
    AffymetrixCelSet. These are designed to be memory efficient but
    still being fast. The idea is to keep all data on file and only read
    data into memory when needed. Clever caching mechanisms are used to
    minimize the overhead of data I/O. All of the above is hidden in the
    package API and for the developer (and the end user), the data is
    queried as if it lives in memory. With this design it is only the
    diskspace that limits what number of arrays can be analyzed. By
    Henrik Bengtsson, Ken Simpson, Elizabeth Purdom, and Mark Robinson.

*aroma.core*

:   Private support package for *aroma.affymetrix* et al., with the
    package API still in alpha/beta. By Henrik Bengtsson.

*atm*

:   Creation of additive models with semiparametric predictors,
    emphasizing term objects, especially (1) implementation of a term
    class hierarchy, and (2) interpretation and evaluation of term
    estimates as functions of explanatories. By Charlotte Maia.

*atmi*

:   Analysis and usage of the trading rules, which are based on
    technical market indicators as well as on the time series analysis.
    By Waldemar Kemler and Peter Schaffner.

*bclust*

:   Bayesian clustering using spike-and-slab hierarchical model,
    suitable for clustering high-dimensional data. Builds a dendrogram
    with log posterior as a natural distance defined by the model. Can
    also compute Bayesian discrimination probabilities equivalent to the
    implemented Bayesian clustering. Spike-and-Slab models are adopted
    in a way to be able to produce an importance measure for clustering
    and discriminant variables. The method works properly for data with
    small sample size and high dimensions. The model parameter
    estimation maybe difficult, depending on data structure and the
    chosen distribution family. By Vahid Partovi Nia and Anthony C.
    Davison.

*bcv*

:   Cross-Validation for the SVD (Bi-Cross-Validation). This package
    implements methods for choosing the rank of an SVD approximation via
    cross validation. It provides both Gabriel-style "block" holdouts
    and Wold-style "speckled" holdouts. Also included is an
    implementation of the SVDImpute algorithm. See Owen & Perry's 2009
    AOAS article (<http://arxiv.org/abs/0908.2062>) and Perry's 2009 PhD
    thesis (<http://arxiv.org/abs/0909.3052>) for more information about
    Bi-crossvalidation. By Patrick O. Perry.

*bdoc*

:   Bayesian Discrete Ordered Classification of DNA Barcodes. By Michael
    Anderson.

*bdsmatrix*

:   Routines for Block Diagonal Symmetric matrices, a special case of
    sparse matrices, used by *coxme* and *kinship*. By Terry Therneau.

*bestglm*

:   Best subset GLM using AIC, BIC, EBIC, BICq or Cross-Validation. For
    the normal case, the "leaps" is used; otherwise, a slower exhaustive
    search. By A.I. McLeod and Changjiang Xu.

*binomSamSize*

:   Confidence intervals and sample size determination for a binomial
    proportion under simple random sampling and pooled sampling. Such
    computations are e.g. of interest when investigating the incidence
    or prevalence in populations. Contains functions to compute coverage
    probabilities and coverage coefficients of the provided confidence
    intervals procedures. Sample size calculations are based on expected
    length. By Michael with contributions by Wei Liu.

*bootRes*

:   Calculation of bootstrapped response and correlation functions for
    use in dendroclimatology. By Christian Zang.

*bvpSolve*

:   Solvers for boundary value problems (BVPs) of systems of ordinary
    differential equations (ODEs) via an interface to the FORTRAN
    function `twpbvp` and an R implementation of the shooting method. By
    Karline Soetaert.

*caGUI*

:   A Tcl/Tk GUI for computation and visualization of simple, multiple
    and joint correspondence analysis with the *ca* package. By Angelos
    Markos.

*caroline*

:   A collection of functions useful for various general situations. By
    David M. Schruth.

*cimis*

:   A set of functions for retrieving data from CIMIS, the California
    Irrigation Management Information System. By Greg Hirson.

*cmm*

:   Categorical Marginal Models: Quite extensive package for the
    estimation of marginal models for categorical data. By Wicher
    Bergsma and Andries van der Ark.

*cmprskContin*

:   Estimation and testing of continuous mark-specific relative risks in
    two groups as described in Gilbert, McKeague & Sun (2008).
    Implements the methods presented in the paper for testing
    mark-specific hazards ratios and for estimation of mark-specific
    incidence ratios that are cumulative in time or cumulative in both
    time and the continuous mark. By Peter Gilbert, Ian McKeague, and
    Yanqing Sun.

*cmrutils*

:   A collection of useful helper routines developed by students of the
    Center for the Mathematical Research, Stankin, Moscow. By Andrey
    Paramonov.

*colbycol*

:   Read big text files column by column: tries to solve the memory
    restrictions posed by such files by breaking them into columns which
    are subsequently read individually into R. By Carlos J. Gil
    Bellosta.

*condGEE*

:   Solves for the mean parameters, the variance parameter, and their
    asymptotic variance in a conditional GEE for recurrent event gap
    times, as described by Clement and Strawderman (2009). Makes a
    parametric assumption for the length of the censored gap time. By
    David Clement.

*countrycode*

:   Create and manipulate data frames that contain country names or
    country coding schemes. Standardizes country names, converts them
    into one of seven coding schemes, assigns region descriptors, and
    generates empty dyadic or country-year dataframes from the coding
    schemes. By Vincent Arel-Bundock.

*coxme*

:   Mixed Effects Cox Models: Cox proportional hazards models containing
    Gaussian random effects, also known as frailty models. By Terry
    Therneau.

*crantastic*

:   Various R tools for <http://crantastic.org/>. By Bjørn Arild
    Maeland.

*cshapes*

:   Package for CShapes, a GIS dataset of country boundaries
    (1946--2008). Includes functions for data extraction and the
    computation of weights matrices. By Nils B. Weidmann, Doreen Kuse,
    and Kristian Skrede Gleditsch.

*cudaBayesreg*

:   CUDA Parallel Implementation of a Bayesian Multilevel Model for fMRI
    Data Analysis. Compute Unified Device Architecture (CUDA) is a
    software platform for massively parallel high-performance computing
    on NVIDIA GPUs. This package provides a CUDA implementation of a
    Bayesian multilevel model for the analysis of brain fMRI data. By
    Adelino Ferreira da Silva.

*dcemri*

:   A collection of routines and documentation that allows one to
    perform a quantitative analysis of dynamic contrast-enhanced or
    diffusion-weighted MRI data. Medical imaging data should be
    organized using either the Analyze or NIfTI data formats. By Brandon
    Whitcher and Volker Schmid, with contributions from Andrew Thornton.

*descr*

:   Descriptive statistics: functions to describe weighted categorical
    variables and functions to facilitate the character encoding
    conversion of objects. By Jakson Aquino. Includes R source code
    and/or documentation written by Dirk Enzmann, Marc Schwartz, and
    Nitin Jain.

*desire*

:   Harrington and Derringer-Suich type desirability functions. By Heike
    Trautmann and Detlef Steuer and Olaf Mersmann.

*difR*

:   Some traditional methods to detect dichotomous differential item
    functioning (DIF) in psychometrics. Both uniform and non-uniform DIF
    effects can be detected, with methods relying upon item response
    models or not. Some methods deal with more than one focal group. By
    Sébastien Béland, David Magis and Gilles Raîche.

*digeR*

:   An easy to use Graphical User Interface for spots correlation
    analysis, score plot, classification, feature selection and power
    analysis for 2D DIGE experiment data. By Yue Fan, Thomas Brendan
    Murphy, and R. William G. Watson.

*dirmult*

:   Estimate parameters in Dirichlet-Multinomial and compute profile
    log-likelihoods. By Torben Tvedebrink.

*dlnm*

:   Distributed Lag Non-linear Models. Contains functions to specify
    basis and cross-basis matrices in order to run distributed lag
    models and their non-linear extension, then to predict and graph the
    results for a fitted model. By Antonio Gasparrini and Ben Armstrong.

*doMC*

:   Provides a parallel backend for the *foreach* `%dopar%` function
    using Simon Urbanek's *multicore* package. By REvolution Computing.

*doSNOW*

:   Provides a parallel backend for the *foreach* `%dopar%` function
    using Luke Tierney's *snow* package. By REvolution Computing.

*dummies*

:   Expands factors, characters and other eligible classes into
    dummy/indicator variables. By Christopher Brown.

*dynGraph*

:   Interactive visualization of dataframes and factorial planes. By
    Julien Durand, Sébastien Lê.

*el.convex*

:   Empirical likelihood ratio tests for means. By Dan Yang, Dylan
    Small.

*endogMNP*

:   Fits a Bayesian multinomial probit model with endogenous selection,
    which is sometimes called an endogenous switching model. This can be
    used to model discrete choice data when respondents select
    themselves into one of several groups. This package is based on the
    *MNP* package by Kosuke Imai and David A. van Dyk. By Lane F.
    Burgette.

*estout*

:   Estimates Output: stores the estimates of several models, and
    formats these to a table of the form estimate starred and standard
    error. below. Default output is LaTeX but output to CSV for later
    editing in a spreadsheet tool is possible as well. Works for linear
    models and panel models from package *plm*. By Felix Kaminsky.

*exact2x2*

:   Exact Conditional Tests and Confidence Intervals for $2 \times 2$
    tables. Calculates Fisher's exact and Blaker's exact tests. By M.P.
    Fay.

*farmR*

:   A mixed integer description of an arable farm. Finds optimal farming
    plans given economic and social preference information. By Ira R.
    Cooke.

*fds*

:   Functional data sets. By Han Lin Shang and Rob J Hyndman.

*flsa*

:   Path algorithm for the general Fused Lasso Signal Approximator. By
    Holger

*flubase*

:   Estimates the baseline of mortality free of influenza epidemics, and
    the respective excess deaths, for more than one time series (age
    groups, gender, regions, etc.). By Nunes B, Natario I and Carvalho
    L.

*foreach*

:   Support for the foreach looping construct. Foreach is an idiom that
    allows for iterating over elements in a collection, without the use
    of an explicit loop counter. Intended to be used for its return
    value, rather than for its side effects. Using *foreach* without
    side effects also facilitates executing the loop in parallel. By
    REvolution Computing.

*fractalrock*

:   Generate fractal time series with non-normal returns distribution.
    The basic principle driving fractal generation of time series is
    that data is generated iteratively based on increasing levels of
    resolution. The initial series is defined by a so-called initiator
    pattern and then generators are used to replace each segment of the
    initial pattern. Regular, repeatable patterns can be produced by
    using the same seed and generators. By using a set of generators,
    non-repeatable time series can be produced. This technique is the
    basis of the fractal time series process in this package. By Brian
    Lee Yung Rowe.

*frbf*

:   Implementation of the "Flexible kernels for RBF network" algorithm.
    By Fernando Martins.

*freqMAP*

:   Estimates a frequency Moving Average Plot (MAP) from multinomial
    data and a continuous covariate. The frequency MAP is a moving
    average estimate of category frequencies, where frequency means and
    posterior bounds are estimated. Comparisons of two frequency MAPs as
    well as odds ratios can be plotted. By Colin McCulloch.

*ftsa*

:   Functional time series analysis. By Rob J Hyndman and Han Lin Shang.

*gPdtest*

:   Computes the bootstrap goodness-of-fit test for the generalized
    Pareto distribution proposed by Villaseñor-Alva and González-Estrada
    (2009). The null hypothesis includes heavy and non-heavy tailded
    gPd's. Also provides functionality for fitting the gPd to data using
    the parameter estimation methods proposed in the same article. By
    Elizabeth González Estrada, José A. Villaseñor Alva.

*gRapHD*

:   Efficient selection of undirected graphical models for
    high-dimensional datasets. Provides tools for selecting trees,
    forests and decomposable models minimizing information criteria such
    as AIC or BIC, and for displaying the independence graphs of the
    models. Also some useful tools for analyzing graphical structures.
    It supports the use of discrete, continuous, or both types of
    variables. By Gabriel Coelho Goncalves de Abreu, Rodrigo Labouriau,
    and David Edwards.

*gamesNws*

:   Playing games using a NWS Server. Allows playing different card
    games (e.g. uno, poker, ...) using an NWS Server as the card table.
    By Markus Schmidberger and Fabian Grandke.

*gamlss.data*

:   Data for GAMLSS models. By Mikis Stasinopoulos and Bob Rigby.

*gamm4*

:   Fit generalized additive mixed models via a version of *mgcv*'s
    `gamm()` function, using *lme4* for estimation via Fabian Scheipl's
    trick. By Simon Wood.

*gausspred*

:   Predict the discrete response based on selected high dimensional
    features, such as gene expression data. The data are modeled with
    Bayesian Gaussian models. When a large number of features are
    available, one may like to select only a subset of features to use,
    typically those features strongly correlated with the response in
    training cases. Such a feature selection procedure is however
    invalid since the relationship between the response and the features
    will appear stronger. This package provides a way to avoid this bias
    and yields well-calibrated prediction for the test cases when one
    uses F-statistic to select features. By Longhai Li.

*geneARMA*

:   Simulate, model, and display data from a time-course microarray
    experiment with periodic gene expression. Fits models in a normal
    mixture model framework with mean approximated by a truncated
    Fourier series and covariance structure modeled by an ARMA$(p,q)$
    process. Estimation is performed with the EM algorithm. By Timothy
    McMurry and Arthur Berg.

*geneListPie*

:   Map a gene list to function categories defined in GOSlim or Kegg.
    The results can be plotted as a pie chart to provide a quick view of
    the distribution of the gene list among the function categories. By
    Xutao Deng.

*glmdm*

:   Simulation of GLMDM. By Jeff Gill, George Casella, Minjung Kyung and
    Jonathan Rapkin.

*glmperm*

:   A permutation test for inference in generalized linear models.
    Useful when parameter estimates in ordinary GLMs fail to converge or
    are unreliable due to small sample size. By Wiebke Werft and
    Douglas M. Potter.

*glmulti*

:   GLM model selection and multimodel inference made easy: automated
    model selection for GLMs. Provides a wrapper for `glm()` and similar
    functions, automatically generating all possible models (under
    constraints set by the user) with the specified response and
    explanatory variables, and finding the best models in terms of some
    Information Criterion (AIC, AICc or BIC). Can handle very large
    numbers of candidate models. Features a Genetic Algorithm to find
    the best models when an exhaustive screening of the candidates is
    not feasible. By Vincent Calcagno.

*gnumeric*

:   Read data from files readable by gnumeric. Can read a whole sheet or
    a range, from several file formats, including the native format of
    gnumeric. Reading is done by using ssconvert (a file converter
    utility included in the gnumeric distribution
    <http://projects.gnome.org/gnumeric/>) to convert the requested part
    to CSV. By Karoly Antal.

*gsDesign*

:   Derive group sequential designs and describe their properties. By
    Keaven Anderson.

*iGenomicViewer*

:   Tool for sending interactive bioinformatic heatmaps with tool-tip
    content. By Daniel P Gaile, Lori A. Shepherd, Lara Sucheston, Andrew
    Bruno, and Kenneth F. Manly.

*infotheo*

:   Implements various measures of information theory based on several
    entropy estimators. By Patrick E. Meyer.

*inlinedocs*

:   Convert inline comments to documentation. Generates Rd files from R
    source code with comments, providing for quick, sustainable package
    development. The syntax keeps code and documentation close together,
    and is inspired by the Don't Repeat Yourself principle. By Toby
    Dylan Hocking.

*intamap*

:   Classes and methods for automated spatial interpolation. By Edzer
    Pebesma, Jon Skøien and others.

*integrOmics*

:   Integrate Omics data project. Supplies two efficient methodologies,
    regularized CCA and sparse PLS, to unravel relationships between two
    heterogeneous data sets of size $n \times p$ and $n \times q$ where
    the $p$ and $q$ variables are measured on the same samples or
    individuals $n$. These data may come from high throughput
    technologies, such as omics data (e.g., transcriptomics,
    metabolomics or proteomics data) that require an integrative or
    joint analysis. However, *integrOmics* can also be applied to any
    other large data sets where $p+q \gg n$. rCCA is a regularized
    version of CCA to deal with the large number of variables. sPLS
    allows variable selection in a one step procedure and two frameworks
    are proposed: regression and canonical analysis. Numerous graphical
    outputs are provided to help interpreting the results. By Sébastien
    Dejean, Ignacio González and Kim-Anh Lê Cao.

*integrativeME*

:   integrative mixture of experts. Mixture of experts models (Jacobs et
    al., 1991) were introduced to account for nonlinearities and other
    complexities in the data, and are of interest due to their wide
    applicability and the advantages of fast learning via the
    expectation-maximization (EM) algorithm. The package features an ME
    extension to combine categorical clinical factors and continuous
    microarray data in a binary classification framework to analyze
    cancer studies. By Kim-Anh Lê Cao.

*interval*

:   Weighted Logrank Tests and NPMLE for interval censored data. By
    Michael P. Fay.

*isa2*

:   The Iterative Signature Algorithm (ISA), a biclustering algorithm
    that finds modules in an input matrix. A module or bicluster is a
    block of the reordered input matrix. By Gábor Csárdi.

*iterators*

:   Iterator construct for R. Iterators allow a programmer to traverse
    through all the elements of a vector, list, or other collection of
    data. By REvolution Computing.

*kmi*

:   Kaplan-Meier multiple imputation to recover the missing potential
    censoring information from competing risks events, so that standard
    right-censored methods could be applied to the imputed data sets to
    perform analyses of the cumulative incidence functions. By Arthur
    Allignol.

*latdiag*

:   Writes a file of commands for the dot program to draw a graph
    proposed by Rosenbaum and useful for checking some properties of
    various sorts of latent scale. By Michael Dewey.

*latticedl*

:   Direct labeling functions that use the *lattice* package. By Toby
    Dylan Hocking.

*lda*

:   Collapsed Gibbs sampling methods for topic models. Implements Latent
    Dirichlet allocation (LDA) and related models, including (but not
    limited to) sLDA, corrLDA, and the mixed-membership stochastic
    blockmodel. Inference for all of these models is implemented via a
    fast collapsed Gibbs sampler written in C. Utility functions for
    reading/writing data typically used in topic models, as well as
    tools for examining posterior distributions are also included. By
    Jonathan Chang.

*lemma*

:   Laplace approximated EM Microarray Analysis. LEMMA is used to detect
    "nonnull genes" --- genes for which the average response in
    treatment group 1 is significantly different from the average
    response in group 2, in normalized microarray data. LEMMA is an
    implementation of an approximate EM algorithm to estimate the
    parameters in the assumed linear model in Bar, Booth, Schifano,
    Wells (2009). By Haim Bar and Elizabeth Schifano.

*longitudinalData*

:   Tools for Longitudinal Data. By Christophe Genolini.

*lordif*

:   LOgistic Regression Differential Item Functioning (DIF) using Item
    Response Theory (IRT): analysis of DIF for dichotomous and
    polytomous items using an iterative hybrid of (ordinal) logistic
    regression and IRT. By Seung W. Choi, with contributions from
    Laura E. Gibbons and Paul K. Crane.

*mar1s*

:   Multiplicative AR(1) with Seasonal Processes, a stochastic process
    model built on top of AR(1). Provides the following procedures for
    MAR(1)S processes: fit, compose, decompose, advanced simulate and
    predict. By Andrey Paramonov.

*marelacTeaching*

:   Datasets and tutorials for use in the MArine, Riverine, Estuarine,
    LAcustrine and Coastal sciences. By Karline Soetaert, Thomas
    Petzoldt, and Filip Meysman.

*maticce*

:   MApping Transitions In Continuous Character Evolution. Tools for an
    information-theoretic approach to estimating the probability of
    continuous character transitions on phylogenetic trees. By Andrew
    Hipp, with contributions from Marcial Escudero.

*matrixStats*

:   Methods that apply to rows and columns of a matrix. Methods are
    optimized for speed and memory. Currently in a beta phase. By Henrik
    Bengtsson (partly by Robert Gentleman).

*mcclust*

:   Methods for processing a sample of (hard) clusterings, e.g., the
    MCMC output of a Bayesian clustering model. Includes methods that
    find a single best clustering to represent the sample, which are
    based on the posterior similarity matrix or a relabeling algorithm.
    By Arno Fritsch.

*mediation*

:   Parametric and nonparametric causal mediation analysis. Implements
    the methods and suggestions in Imai, Keele, and Yamamoto (2008) and
    Imai, Keele, Tingley (2009). In addition to the estimation of causal
    mediation effects, allows researchers to conduct sensitivity
    analysis for certain parametric models. By Luke Keele, Dustin
    Tingley, Teppei Yamamoto, Kosuke Imai.

*metafor*

:   A collection of functions for conducting meta-analyses in R. Fixed-
    and random-effects models (with and without moderators) can be
    fitted via the general linear (mixed-effects) model. For
    $2 \times 2$ table data, the Mantel-Haenszel and Peto's method are
    also implemented. By Wolfgang Viechtbauer.

*mixfdr*

:   Fits normal mixture models to data and uses them to compute effect
    size estimates and local and tail area false discovery rates. To
    make this precise, suppose you have many normally distributed $z$s,
    and each $z_i$ has mean $\delta_i$. This package will estimate
    $\delta_i$ based on the $z$s (effect sizes), $P(\delta_i=0|z_i)$
    (local false discovery rates) and $P(\delta_i=0\;|\;|z_i|>z)$ (tail
    area false discovery rates). By Omkar Muralidharan, with many
    suggestions from Bradley Efron.

*modTempEff*

:   Modeling temperature effects using time series data. Fits a
    constrained segmented distributed lag regression model to
    epidemiological time series of mortality, temperature, and other
    confounders. By Vito M. R. Muggeo.

*mrt*

:   Datasets and functions from Wright and London's Modern Regression
    Techniques. By Daniel B. Wright.

*mstate*

:   Functions for data preparation, descriptives, hazard estimation and
    prediction with Aalen-Johansen or simulation in competing risks and
    multi-state models. By Hein Putter, Liesbeth de Wreede, and Marta
    Fiocco.

*muRL*

:   Mailmerge using R, LaTeX, and the Web. Provides mailmerge methods
    for reading spreadsheets of addresses and other relevant information
    to create standardized but customizable letters. Provides a method
    for mapping US ZIP codes, including those of letter recipients.
    Provides a method for parsing and processing [HTML]{.smallcaps} code
    from online job postings of the American Political Science
    Association. By Ryan T. Moore and Andrew Reeves.

*multmod*

:   Testing of multiple outcomes using i.i.d. decompositions. By
    Christian B. Pipper, Christian Ritz.

*munsell*

:   Functions for exploring and using the Munsell colour system. By
    Charlotte Wickham.

*mvShapiroTest*

:   Generalized Shapiro-Wilk test for multivariate normality as proposed
    by Villaseñor-Alva and González-Estrada (2009). By Elizabeth
    González Estrada and José A. Villaseñor Alva.

*mvsf*

:   Multivariate generalization of the Shapiro-Francia test for
    normality. By David Delmail.

*nbpMatching*

:   Functions for non-bipartite optimal matching. By Bo Lu, Robert
    Greevy, Cole Beck.

*nnclust*

:   Nearest-neighbor tools for clustering. Finds nearest neighbors and
    the minimum spanning tree (MST) for large data sets, and does
    clustering using the MST. By Thomas Lumley.

*nodeHarvest*

:   Compute and visualize the Node Harvest estimator. Node harvest is a
    simple interpretable tree-like estimator for high-dimensional
    regression and classification. A few nodes are selected from an
    initially large ensemble of nodes, each associated with a positive
    weight. New observations can fall into one or several nodes and
    predictions are the weighted average response across all these
    groups. By Nicolai Meinshausen.

*nparLD*

:   Nonparametric Analysis of Longitudinal Data in Factorial
    Experiments. By Kimihiro Noguchi, Mahbub Latif, Karthinathan
    Thangavelu, Frank Konietschke, Yulia R. Gel, and Edgar Brunner.

*oblique.tree*

:   Grows oblique trees to classification data. By Alfred Truong.

*odfWeave.survey*

:   odfWeave support for the *survey* package. By Thomas Lumley.

*oosp*

:   Object Oriented Statistical Programming. Support for OOSP,
    especially by extending S3 capabilities, providing pointer and
    component objects, and providing basic support for symbolic-numeric
    statistical programming. By Charlotte Maia.

*optBiomarker*

:   Estimation of optimal number of biomarkers for two-group microarray
    based classifications at a given error tolerance level for various
    classification rules. By Mizanur Khondoker.

*p3state.msm*

:   Analyzing survival data from illness-death models. By Luís
    Meira-Machado and Javier Roca-Pardiñas.

*packS4*

:   Toy example of an S4 package, to illustrate the book "Petit Manuel
    de Programmation Orientee Objet sous R". By Christophe Genolini.

*packdep*

:   Elucidates the dependencies between user-contributed R packages and
    identifies key packages according to social network analysis
    metrics. By Radhakrishnan Nagarajan and Marco Scutari.

*pamm*

:   Simulation functions to assess or explore the power of a dataset to
    estimate significant random effects (intercept or slope) in a mixed
    model. Based on the *lme4* package. By Julien Martin.

*parser*

:   Detailed source code parser, based on the standard R parser but
    organizing the information differently. By Romain François.

*pedigreemm*

:   Fit pedigree-based mixed-effects models. By Douglas Bates and Ana
    Ines Vazquez.

*pegas*

:   Population and Evolutionary Genetics Analysis System. Provides
    functions for reading, writing, plotting, analysing, and
    manipulating allelic data, and for the analysis of population
    nucleotide sequences including coalescence analyses. By Emmanuel
    Paradis.

*pendensity*

:   Estimation of penalized (conditional) densities. By Christian
    Schellhase.

*perm*

:   Exact or Asymptotic permutation tests. By Michael Fay.

*pgs*

:   Precision of Geometric Sampling. Computes mean squared errors of
    stereological predictors. By Kien Kieu and Marianne Mora.

*phull*

:   Computes the p-hull of a finite planar set, which is a
    generalization of the convex hull, X-Y hull and bounding rectangle.
    A fast, $O(n \log n)$ Graham-scan based routine is used. By Marek
    Gagolewski.

*phybase*

:   Basic functions for phylogenetic analysis. Provides functions to
    read, write, manipulate, estimate, and summarize phylogenetic trees
    including species trees which contain not only the topology and
    branch lengths but also population sizes. The input/output functions
    can read tree files in which trees are presented in parenthetic
    format. The trees are read in as a string and then transformed to a
    matrix which describes the relationship of nodes and branch lengths.
    The nodes matrix provides an easy access for developers to further
    manipulate the tree, while the tree string provides interface with
    other phylogenetic R packages such as *ape*. The input/output
    functions can also be used to change the format of tree files
    between NEXUS and PHYLIP. Some basic functions have already been
    established in the package for manipulating trees such as deleting
    and swapping nodes, rooting and unrooting trees, changing the root
    of the tree. The package also includes functions for summarizing
    phylogenetic trees, calculating the coalescence time, population
    size, and tree distance, and to estimate the species tree from
    multiple gene trees. By Liang Liu.

*plus*

:   Penalized Linear Unbiased Selection. Efficient procedures for
    fitting an entire regression sequences with different model types.
    By Cun-Hui Zhang and Ofer Melnik.

*pooh*

:   Partial Orders and Relations: functions for computing closures of
    relations. By Charles J. Geyer.

*potts*

:   Markov Chain Monte Carlo for Potts Models. By Charles J. Geyer.

*psgp*

:   Projected Spatial Gaussian Process methods for package *intamap*. By
    Ben Ingram and Remi Barillec.

*rWMBAT*

:   The William and Mary Bayesian Analysis Tool. By Karl Kuschner, Qian
    Si and William Cooke, College of William and Mary.

*rainbow*

:   Rainbow plots, bagplots and boxplots for functional data. By Han Lin
    Shang and Rob J Hyndman.

*rankhazard*

:   Rank-hazard plots (Karvanen and Harrell, Statistics in
    Medicine, 2009) which visualize the relative importance of
    covariates in a proportional hazards model. The key idea is to rank
    the covariate values and plot the relative hazard as a function of
    ranks scaled to interval $[0,1]$. The relative hazard is plotted
    with respect to the reference hazard, which can be e.g. the hazard
    related to the median of the covariate. Transformation to scaled
    ranks allows plotting of covariates measured in different units in
    the same graph, which helps in the interpretation of the
    epidemiological relevance of the covariates. Rank-hazard plots show
    the difference of hazards between the extremes of the covariate
    values present in the data and can be used as a tool to check if the
    proportional hazards assumption leads to reasonable estimates for
    individuals with extreme covariate values. Alternative covariate
    definitions or different transformations applied to covariates can
    be also compared using rank-hazard plots. By Juha Karvanen.

*ringscale*

:   Implementation of the "Ringscale" method as proposed in the student
    research project "Detection of faint companions around young stars
    in speckle patterns of VLT/NACO cube mode images by means of
    post-processing" at the Friedrich-Schiller-University of Jena. By
    Daniel Haase.

*rioja*

:   Analysis of Quaternary science data, including constrained
    clustering, WA, WAPLS, IKFA, MLRC and MAT transfer functions, and
    stratigraphic diagrams. By Steve Juggins.

*ripa*

:   R Image Processing and Analysis. Makes it possible to process and
    analyze RGB, LAN (multispectral) and AVIRIS (hyperspectral) images.
    By Talita Perciano, with contributions from Alejandro C Frery.

*rms*

:   Regression Modeling Strategies: regression modeling, testing,
    estimation, validation, graphics, prediction, and typesetting by
    storing enhanced model design attributes in the fit. A collection of
    about 225 functions that assist with and streamline modeling. Also
    contains functions for binary and ordinal logistic regression models
    and the Buckley-James multiple regression model for right-censored
    responses, and implements penalized maximum likelihood estimation
    for logistic and ordinary linear models. *rms* works with almost any
    regression model, but it was especially written to work with binary
    or ordinal logistic regression, Cox regression, accelerated failure
    time models, ordinary linear models, the Buckley-James model,
    generalized least squares for serially or spatially correlated
    observations, generalized linear models, and quantile regression. By
    Frank E Harrell Jr.

*robustX*

:   eXperimental eXtraneous eXtraordinary functionality for Robust
    Statistics. By Werner Stahel, Martin Mächler and potentially others.

*rpartOrdinal*

:   Functions that can be called in conjunction with *rpart* for
    deriving a classification tree when the response to be predicted is
    ordinal. By Kellie J. Archer.

*rrv*

:   Random Return Variables. Currently provides limited support for
    formatting money. By Charlotte Maia.

*sBF*

:   Smooth Backfitting for additive models using Nadaraya-Watson
    estimator. By A. Arcagni, L. Bagnato.

*safeBinaryRegression*

:   Safe Binary Regression. Overloads the `glm()` function in the
    *stats* package so that a test for the existence of the maximum
    likelihood estimate is included in the fitting procedure for binary
    regression models. By Kjell Konis.

*sculpt3d*

:   A simple toolbar GUI for brushing RGL plots. Controls for simple
    brushing, highlighting, labeling, and mouseMode changes are provided
    by point-and-click rather than through the R terminal interface. By
    Justin Donaldson.

*sddpack*

:   Semidiscrete Decomposition (SDD), which approximates a matrix as a
    weighted sum of outer products formed by vectors with entries
    constrained to be in the set $\{-1, 0, 1\}$. By Tamara G. Kolda and
    Dianne P. O'Leary.

*sdef*

:   Synthesizing list of Differentially Expressed Features. Performs two
    tests to evaluate if experiments are associated and returns a list
    of interesting features common to all the experiments. By Alberto
    Cassese and Marta Blangiardo.

*season*

:   Routines for the seasonal analysis of health data, including
    regression models, time-stratified case-crossover, plotting
    functions and residual checks. By Adrian Barnett and Peter Baker.

*sendmailR*

:   A simple SMTP client which provides a portable solution for sending
    emails from within R. By Olaf Mersmann.

*simFrame*

:   A general framework for statistical simulation. By Andreas Alfons.

*simctest*

:   Sequential (or Safe) Implementation of Monte Carlo tests with
    uniformly bounded resampling risk. Features efficient computation of
    p-values for Monte Carlo tests, e.g., bootstrap tests. By Axel
    Gandy.

*skellam*

:   Functions for the Skellam distribution, including: pmf, cdf,
    quantiles and random variates. By Jerry W. Lewis.

*skmeans*

:   Algorithms to compute spherical $k$-means partitions. Features
    several methods, including a genetic and a simple fixed-point
    algorithm and an interface to the CLUTO vcluster program. By Kurt
    Hornik, Ingo Feinerer and Martin Kober.

*slam*

:   Sparse Lightweight Arrays and Matrices. Data structures and
    algorithms for sparse arrays and matrices, based on index arrays and
    simple triplet representations, respectively. By Kurt Hornik, David
    Meyer, and Christian Buchta.

*sos*

:   R related Search Engines. By Spencer Graves, Sundar Dorai-Raj, and
    Romain François.

*spatcounts*

:   Fit spatial CAR count regression models using MCMC. By Holger
    Schabenberger.

*speedglm*

:   Fitting LMs and GLMs to large data sets by updating algorithms. By
    Marco ENEA.

*speff2trial*

:   Semiparametric efficient estimation for a two-sample treatment
    effect: performs estimation and testing of the treatment effect in a
    2-group randomized clinical trial with a quantitative or dichotomous
    endpoint. The method is a special case of Robins, Rotnitzky, and
    Zhao (1994, JASA). It improves efficiency by leveraging baseline
    predictors of the endpoint. The method uses inverse probability
    weighting to provide unbiased estimation when the endpoint is
    missing at random. By Michal Juraska, with contributions from
    Peter B. Gilbert, Min Zhang, Marie Davidian, Anastasios A. Tsiatis
    and Xiaomin Lu.

*stringkernels*

:   Gapped and word-based string kernels for use with *kernlab*. By
    Martin Kober.

*sublogo*

:   Visualize correlation in biological sequence data using sublogo
    dendrogram plots. By Toby Dylan Hocking.

*sugaR*

:   Plots to help optimizing diabetes therapy. Provides a series of
    plots to integrate glucose levels, basal rate, activities, events
    and carbohydrate uptake on a single page in a humanely interpretable
    manner. It is meant for best-possibly representing the content of a
    well-curated diabetes diary of up to a week's time or of up to seven
    comparable days, from which conclusions for adjusting the individual
    treatment shall be drawn. By Steffen

*svDialogs*

:   SciViews [GUI]{.smallcaps} [API]{.smallcaps}: dialog boxes. Rapidly
    construct dialog boxes for your [GUI]{.smallcaps}, including an
    automatic function assistant. By Philippe Grosjean.

*svSweave*

:   SciViews [GUI]{.smallcaps} [API]{.smallcaps}: Sweave support
    functions. By Philippe Grosjean.

*svTools*

:   SciViews [GUI]{.smallcaps} [API]{.smallcaps}: tools, aimed at
    wrapping some of the functionalities of the packages *tools*,
    *utils* and *codetools* into a nicer format so that an
    [IDE]{.smallcaps} can use them. By Romain François.

*svUnit*

:   SciViews [GUI]{.smallcaps} [API]{.smallcaps}: unit testing. A
    complete unit test system and functions to implement its
    [GUI]{.smallcaps} part. By Philippe Grosjean.

*svWidgets*

:   SciViews [GUI]{.smallcaps} [API]{.smallcaps}: widgets & windows.
    High level management of widgets, windows and other graphical
    resources. By Philippe Grosjean.

*textcat*

:   N-Gram based text categorization. By Kurt Hornik, Johannes Rauch,
    Christian Buchta, and Ingo Feinerer.

*tikzDevice*

:   A device for R graphics output in PGF/TikZ format. Enables
    LaTeX-ready output from R graphics functions, with LaTeX mathematics
    that can be typeset directly into labels and annotations. Graphics
    produced this way can also be annotated with custom TikZ commands.
    By Charlie Sharpsteen and Cameron Bracken.

*tm.plugin.mail*

:   A plug-in for the *tm* text mining framework providing mail handling
    functionality. By Ingo Feinerer.

*tnet*

:   Analysis of Weighted, Two-mode, and Longitudinal networks. By Tore
    Opsahl.

*tolerance*

:   Functions for calculating tolerance intervals. Tolerance limits
    provide the limits between which we can expect to find a specified
    proportion of a population with a given level of confidence.
    Provides functions for estimating tolerance limits for various
    distributions, and plotting tolerance limits of continuous random
    variables. By Derek S. Young.

*tractor.base*

:   Basic functions for the TractoR (tractography with R) bundle.
    Consists of functions for working with magnetic resonance images.
    Can read and write image files stored in Analyze, NIfTI and DICOM
    formats (DICOM support is read only), generate images for use as
    regions of interest, and manipulate and visualize images. By Jon
    Clayden.

*treethresh*

:   Methods for Tree-based Local Adaptive Thresholding. By Ludger Evers
    and Tim Heaton.

*ttrTests*

:   Standard Backtests for Technical Trading Rules in Financial Data.
    Four core functions evaluate the efficacy of a technical trading
    rule: conditional return statistics, bootstrap resampling
    statistics, test for data snooping bias among parameter choices, and
    robustness of parameter choices. By David St John.

*ttutils*

:   Some utility functions. By Thorn Thaler.

*twitteR*

:   An R based Twitter client via an interface to the Twitter web API.
    By Jeff Gentry.

*vegetarian*

:   Jost Diversity Measures for Community Data. Computes diversity for
    community data sets using the methods outlined by Jost (2006, 2007),
    which offer the advantage of providing diversity numbers
    equivalents, independent alpha and beta diversities, and the ability
    to incorporate "order" as a continuous measure of the importance of
    rare species in the metrics. Computes alpha diversities, beta
    diversities, gamma diversities, and similarity indices. Confidence
    intervals for diversity measures are calculated using a bootstrap
    method described by Chao et al. (2008). By Noah Charney, Sydne
    Record.

*yhat*

:   Methods for variance partitioning for linear models and canonical
    correlation and methods for interpreting regression effects using
    beta weights, standardized beta weights, structure coefficients, and
    adjusted effect sizes. By Kim Nimon and J. Kyle Roberts.

# Other changes

-   Recommended bundle *VR* was unbundled into the recommended packages
    *MASS*, *class*, *nnet* and *spatial*, and moved to the Archive.

-   Bundle *BACCO* was unbundled into its packages *approximator*,
    *calibrator* and *emulator*, and moved to the Archive.

-   Bundle *forecasting* was unbundled into its packages *Mcomp*,
    *expsmooth*, *fma* and *forecast*, and moved to the Archive.

-   Bundle *hoa* was unbundled into its packages *cond*, *csampling*,
    *marg* and *nlreg*, and moved to the Archive.

-   Packages *ARES*, *MDD*, *GeneNT*, *HTMLapplets*, *RiboSort*,
    *SLmisc*, *UNF*, *WaveCGH*, *WeedMap*, *agreement*, *bivpois*,
    *boost*, *clac*, *crq*, *ggplot*, *mclust02*, *norm*, *partsm*,
    *pwr*, *sma*, *supclust*, *uroot* and *verify* were moved to the
    Archive.

-   Packages *Multiclasstesting*, *Rlab*, *intcox*, *km.ci*, *mixlow*,
    *sdtalt* and *survBayes* were resurrected from the Archive.

-   Package *spectrino* was removed from CRAN.

-   Package *openNLPmodels* was split into *openNLPmodels.en* and
    *openNLPmodels.es*.
:::
