---
title: News from the Bioconductor Project
abstract: The 'News from the Bioconductor Project' article from the 2013-2 issue.
author:
- name: Bioconductor Team
  affiliation: |
    Program in Computational Biology, Fred Hutchinson Cancer Research
    Center
date: '2013-12-01'
date_received: ~
journal:
  firstpage: '162'
  lastpage: '163'
volume: 5
issue: 2
slug: RJ-2013-2-bioconductor
citation_url: https://rjournal.github.io/
packages:
  cran: shiny
  bioc:
  - VariantTools
  - GenomicRanges
  - AnnotationHub
  - PSICQUIC
  - interactiveDisplay
  - BiocParallel
preview: preview.png
bibliography: bioc.bib
CTV: ~
output:
  rjtools::rjournal_web_article:
    self_contained: yes
    toc: no
    legacy_pdf: yes
    web_only: yes
    mathjax: https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js
    md_extension: -tex_math_single_backslash

---

::: article
Bioconductor 2.13 was released on 15 October 2013. It is compatible with
R 3.0.2, and consists of 749 software packages, 179 experiment data
packages, and more than 690 up-to-date annotation packages. The release
includes 84 new software packages, and enhancements to many others.
Descriptions of new packages and updated NEWS files provided by current
package maintainers are at
<http://bioconductor.org/news/bioc_2_13_release/>.

Start using Bioconductor and R version 3.0.2 with

``` r
> source("http://bioconductor.org/biocLite.R")
> biocLite()
```

Install additional packages, e.g.,
[*VariantTools*](https://www.bioconductor.org/packages/release/bioc/html/VariantTools.html),
with

``` r
> source("http://bioconductor.org/biocLite.R")
> biocLite("VariantTools")
```

Upgrade installed packages with

``` r
> source("http://bioconductor.org/biocLite.R")
> biocLite()
```

Explore available Bioconductor packages at
<http://bioconductor.org/packages/release/>. All packages are grouped by
'BiocViews' to identify coherent groups of packages. Each package has an
html page with the descriptions and links to vignettes, reference
manuals, and use statistics.

A Bioconductor Amazon Machine Instance is available and updated; see\
<http://bioconductor.org/help/bioconductor-cloud-ami>.

# Core Bioconductor packages

[*GenomicRanges*](https://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html)
and related packages continue to provide an extensive, mature and
extensible framework for interacting with high throughput sequence data.
Many contributed packages rely on this infrastructure for interoperable,
re-usable analysis; [@10.1371/journal.pcbi.1003118] provide an
introduction.

Our large collection of microarray, transcriptome and organism-specific
annotation packages have also been updated to include current
information. Most of these packages now provide access to the 'select'
interface (keys, columns, keytypes and select) which enable programmatic
access to the databases they contain. This uniform interface simplifies
the user experience; one illustration of this is in packages such as
*Homo.sapiens*, which coordinate access to model organism gene,
transcript, and pathway data bases. The
[*AnnotationHub*](https://www.bioconductor.org/packages/release/bioc/html/AnnotationHub.html),
now with more than 10,000 entries, complements our traditional offerings
with diverse whole genome annotations from Ensembl, ENCODE, dbSNP, UCSC,
and elsewhere.

This release includes the
[*PSICQUIC*](https://www.bioconductor.org/packages/release/bioc/html/PSICQUIC.html)
package, an R interface to a powerful and standardized query language
from the HUPO Proteomics Standard Initiative. More than two dozen
well-known interaction databases are include, of which BioGrid, BIND,
Reactome, STRING, IntAct, UniProt, DIP, and ChEMBL may be the best
known. The query language uses a controlled vocabulary honed over
several years by an active community, which returns documented and
annotated interactions. One can now curate a detailed and up-to-date
network illuminating functional relationships between genes and proteins
of interest.

# Other activities

In a change from the routine, our next Annual Meeting is in Boston, 30
July--1 August 2014, making it easier for east coast and European
attendees. The 2013 Annual Meeting was in Seattle, 17--19 July, with our
European developer community meeting in Cambridge, UK, 9--10 December.
Our very active training and community events are advertised at
<http://bioconductor.org/help/events/>. The Bioconductor mailing lists
(<http://bioconductor.org/help/mailing-list/>) connect users with each
other, to domain experts, and to maintainers eager to ensure that their
packages satisfy the needs of leading edge approaches. Keep abreast of
packages added to the 'devel' branch and other activities by following
\@Bioconductor on Twitter.

The past year marked our first participation in the Google Summer of
Code project. We had many excellent applicants, and sponsored two
projects to completion. GSoC participant Shawn Balcome, under Marc
Carlson's mentoring, produced the
[*interactiveDisplay*](https://www.bioconductor.org/packages/release/bioc/html/interactiveDisplay.html)
package for Bioconductor /
[*shiny*](https://CRAN.R-project.org/package=shiny) integration. Michel
Lang, mentored by Michael Lawrence, added very useful batch job and
error recovery functionality to the
[*BiocParallel*](https://www.bioconductor.org/packages/release/bioc/html/BiocParallel.html)
package. This was a very successful experience!
:::
