---
title: News from the Bioconductor Project
abstract: The 'News from the Bioconductor Project' article from the 2009-2 issue.
author:
- name: Bioconductor Team
  affiliation: Program in Computational Biology
  address: |
    Fred Hutchinson Cancer Research Center
date: '2009-12-01'
date_received: ~
journal:
  firstpage: 95
  lastpage: 95
volume: 1
issue: 2
slug: RJ-2009-2-bioconductor
citation_url: https://rjournal.github.io/
packages:
  cran: []
  bioc:
  - chipseq
  - ChIPseqR
  - chIPpeakAnno
  - ChIPsim
  - DEGseq
  - baySeq
  - GenomicFeatures
  - Rolexa
  - BioSeqClass
  - cycle
  - LiquidAssociation
  - SpeCond
  - RPA
  - CNTools
  - CNVtools
  - Starr
  - CGHnormaliter
  - mBPCR
  - AgiMicroRna
  - RmiR
  - MiChip
  - methylumi
  - flowFP
  - flowMerge
  - plateCore
  - ddCt
  - HTqPCR
  - clippda
  - RTCA
  - RTools4TB
  - GeneAnswers
  - BUS
  - ChromHeatMap
preview: preview.png
bibliography: ~
CTV: ~
output:
  rjtools::rjournal_web_article:
    self_contained: yes
    toc: no
    legacy_pdf: yes
    web_only: yes
    mathjax: https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js
    md_extension: -tex_math_single_backslash

---

::: article
We are pleased to announce Bioconductor 2.5, released on October 28,
2009. Bioconductor 2.5 is compatible with R 2.10.0, and consists of 352
packages. There are 34 new packages, and enhancements to many others.
Explore Bioconductor at <http://bioconductor.org>, and install packages
with

``` r
> source("http://bioconductor.org/biocLite.R")
> biocLite() # install standard packages...
> biocLite("IRanges") # ...or IRanges
```

# New and revised packages

This release includes new packages for diverse areas of high-throughput
analysis. Highlights include:

Next-generation sequence analysis

:   packages for ChIP-seq
    ([*chipseq*](https://www.bioconductor.org/packages/release/bioc/html/chipseq.html),
    [*ChIPseqR*](https://www.bioconductor.org/packages/release/bioc/html/ChIPseqR.html),
    [*chIPpeakAnno*](https://www.bioconductor.org/packages/release/bioc/html/chIPpeakAnno.html),
    [*ChIPsim*](https://www.bioconductor.org/packages/release/bioc/html/ChIPsim.html)),
    differential expression
    ([*DEGseq*](https://www.bioconductor.org/packages/release/bioc/html/DEGseq.html),
    [*baySeq*](https://www.bioconductor.org/packages/release/bioc/html/baySeq.html)),
    annotation
    ([*GenomicFeatures*](https://www.bioconductor.org/packages/release/bioc/html/GenomicFeatures.html)),
    and image processing
    ([*Rolexa*](https://www.bioconductor.org/packages/release/bioc/html/Rolexa.html)).

Advanced statistical methods

:   for microarray classification
    ([*BioSeqClass*](https://www.bioconductor.org/packages/release/bioc/html/BioSeqClass.html)),
    differential expression
    ([*cycle*](https://www.bioconductor.org/packages/release/bioc/html/cycle.html),
    [*LiquidAssociation*](https://www.bioconductor.org/packages/release/bioc/html/LiquidAssociation.html),
    [*SpeCond*](https://www.bioconductor.org/packages/release/bioc/html/SpeCond.html)),
    and probe reliability
    ([*RPA*](https://www.bioconductor.org/packages/release/bioc/html/RPA.html)).

Microarray domain-specific analysis

:   of copy number, array CGH, tiling
    ([*CNTools*](https://www.bioconductor.org/packages/release/bioc/html/CNTools.html),
    [*CNVtools*](https://www.bioconductor.org/packages/release/bioc/html/CNVtools.html),
    [*Starr*](https://www.bioconductor.org/packages/release/bioc/html/Starr.html),
    [*CGHnormaliter*](https://www.bioconductor.org/packages/release/bioc/html/CGHnormaliter.html),
    [*mBPCR*](https://www.bioconductor.org/packages/release/bioc/html/mBPCR.html)),
    micro-RNA
    ([*AgiMicroRna*](https://www.bioconductor.org/packages/release/bioc/html/AgiMicroRna.html),
    [*RmiR*](https://www.bioconductor.org/packages/release/bioc/html/RmiR.html),
    [*MiChip*](https://www.bioconductor.org/packages/release/bioc/html/MiChip.html)),
    and methylation
    ([*methylumi*](https://www.bioconductor.org/packages/release/bioc/html/methylumi.html))
    arrays.

Flow cytometry

:   fingerprinting
    ([*flowFP*](https://www.bioconductor.org/packages/release/bioc/html/flowFP.html)),
    cluster merging
    ([*flowMerge*](https://www.bioconductor.org/packages/release/bioc/html/flowMerge.html)),
    and plate-based assays
    ([*plateCore*](https://www.bioconductor.org/packages/release/bioc/html/plateCore.html)).

Diverse assays

:   related to high-throughput qPCR
    ([*ddCt*](https://www.bioconductor.org/packages/release/bioc/html/ddCt.html),
    [*HTqPCR*](https://www.bioconductor.org/packages/release/bioc/html/HTqPCR.html)),
    clinical proteomics
    ([*clippda*](https://www.bioconductor.org/packages/release/bioc/html/clippda.html)),
    and RTCA
    ([*RTCA*](https://www.bioconductor.org/packages/release/bioc/html/RTCA.html)).

Integrative tools

:   for data mining
    ([*RTools4TB*](https://www.bioconductor.org/packages/release/bioc/html/RTools4TB.html)),
    annotation
    ([*GeneAnswers*](https://www.bioconductor.org/packages/release/bioc/html/GeneAnswers.html)),
    network reconstruction
    ([*BUS*](https://www.bioconductor.org/packages/release/bioc/html/BUS.html)),
    and visualization
    ([*ChromHeatMap*](https://www.bioconductor.org/packages/release/bioc/html/ChromHeatMap.html)).

Our large collection of microarray- and organism-specific annotation
packages have been updated to include information current at the time of
the Bioconductor release. These annotation packages contain biological
information about microarray probes and the genes they are meant to
interrogate, or contain gene-based annotations of whole genomes. They
are particularly valuable in providing stable annotations for repeatable
research.

Further information on new and existing packages can be found on the
Bioconductor web site, which contains 'views' that identify coherent
groups of packages. The views link to on-line package descriptions,
vignettes, reference manuals, and use statistics.

# Other activities

The Bioconductor community met on July 27-28 at our annual conference in
Seattle for a combination of scientific talks and hands-on tutorials.
The active Bioconductor mailing lists
(<http://bioconductor.org/docs/mailList.html>) connect users with each
other, to domain experts, and to maintainers eager to ensure that their
packages satisfy the needs of leading edge approaches. Bioconductor
package maintainers and the Bioconductor team invest considerable effort
in producing high-quality software. The Bioconductor team continues to
ensure quality software through technical and scientific reviews of new
packages, and daily builds of released packages on Linux, Windows, and
Macintosh platforms.

# Looking forward

Contributions from the Bioconductor community play an important role in
shaping each release. In addition to development of high-quality
algorithms for microarray data, we anticipate continued efforts to
provide statistically informed analysis of next generation sequence
data. Areas of opportunity include the ChIP-seq, RNA-seq, rare variant,
and structural variant domains. Analysis of next generation sequence
data poses significant challenges in data representation, annotation,
and manipulation; the Bioconductor team is actively working on solutions
to address these software infrastructure challenges. We also anticipate
development of improved graph representations, important for
manipulating large networks of biological data. The next release cycle
promises to be one of active scientific growth and exploration!
:::
