---
title: News from the Bioconductor Project
abstract: The 'News from the Bioconductor Project' article from the 2010-2 issue.
author: 
- name:  Bioconductor Team
  affiliation: Program in Computational Biology, Fred Hutchinson Cancer Research Center
date: '2010-12-01'
date_received: ~
journal:
  firstpage: 101
  lastpage: 101
volume: 2
issue: 2
slug: RJ-2010-2-bioconductor
citation_url: https://rjournal.github.io/
packages:
  cran: []
  bioc:
  - iSeq
  - RMAPPER
  - MEDIPS
  - rnaSeqMap
  - R453Plus1Toolbox
  - OTUbase
  - ADaCGH2
  - les
  - LVSmiRNA
  - MBCB
  - fabia
  - farms
  - RDRToolbox
  - IsoGeneGUI
  - gage
  - HTSanalyzeR
  - PatientGeneSets
  - BioNet
  - netresponse
  - attract
  - CoGAPS
  - ontoCAT
  - DEgraph
  - NTW
  - RCytoscape
  - BHC
  - CGEN
  - SQUADD
  - imageHTS
  - CRImage
  - coRNAi
  - GeneGA
  - NuPoP
  - graph
  - GenomicRanges
  - GenomicFeatures
  - Biostrings
  - chipseq
preview: preview.png
bibliography: ~
CTV: ~
output:
  rjtools::rjournal_web_article:
    self_contained: yes
    toc: no
    legacy_pdf: yes
    web_only: yes
    mathjax: https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js
    md_extension: -tex_math_single_backslash

---

::: article
We are pleased to announce Bioconductor 2.7, released on October 18,
2010. Bioconductor 2.7 is compatible with R 2.12.0, and consists of 419
packages. There are 34 new packages, and enhancements to many others.
Explore Bioconductor at <http://bioconductor.org>, and install packages
with

``` r
> source("http://bioconductor.org/biocLite.R")
> biocLite() # install standard packages...
> biocLite("IRanges") # ...or IRanges
```

# New and revised packages

This release includes new packages for diverse areas of high-throughput
analysis. Highlights include:

Next-generation sequencing

:   packages for ChIP
    ([*iSeq*](https://www.bioconductor.org/packages/release/bioc/html/iSeq.html),
    [*RMAPPER*](https://www.bioconductor.org/packages/release/bioc/html/RMAPPER.html)),
    methylated DNA immunoprecipitation
    ([*MEDIPS*](https://www.bioconductor.org/packages/release/bioc/html/MEDIPS.html)),
    and RNA-seq
    ([*rnaSeqMap*](https://www.bioconductor.org/packages/release/bioc/html/rnaSeqMap.html))
    work flows, 454 sequencing
    ([*R453Plus1Toolbox*](https://www.bioconductor.org/packages/release/bioc/html/R453Plus1Toolbox.html))
    and management of microbial sequences
    ([*OTUbase*](https://www.bioconductor.org/packages/release/bioc/html/OTUbase.html)).

Microarray

:   analysis of domain-specific applications (array CGH,
    [*ADaCGH2*](https://www.bioconductor.org/packages/release/bioc/html/ADaCGH2.html);
    tiling arrays,
    [*les*](https://www.bioconductor.org/packages/release/bioc/html/les.html);
    miRNA,
    [*LVSmiRNA*](https://www.bioconductor.org/packages/release/bioc/html/LVSmiRNA.html);
    and bead arrays,
    [*MBCB*](https://www.bioconductor.org/packages/release/bioc/html/MBCB.html));
    specialized statistical methods
    ([*fabia*](https://www.bioconductor.org/packages/release/bioc/html/fabia.html),
    [*farms*](https://www.bioconductor.org/packages/release/bioc/html/farms.html),
    [*RDRToolbox*](https://www.bioconductor.org/packages/release/bioc/html/RDRToolbox.html)),
    and graphical tools
    ([*IsoGeneGUI*](https://www.bioconductor.org/packages/release/bioc/html/IsoGeneGUI.html)).

Gene set, network, and graph

:   oriented approaches and tools include
    [*gage*](https://www.bioconductor.org/packages/release/bioc/html/gage.html),
    [*HTSanalyzeR*](https://www.bioconductor.org/packages/release/bioc/html/HTSanalyzeR.html),
    [*PatientGeneSets*](https://www.bioconductor.org/packages/release/bioc/html/PatientGeneSets.html),
    [*BioNet*](https://www.bioconductor.org/packages/release/bioc/html/BioNet.html),
    [*netresponse*](https://www.bioconductor.org/packages/release/bioc/html/netresponse.html),
    [*attract*](https://www.bioconductor.org/packages/release/bioc/html/attract.html),
    [*CoGAPS*](https://www.bioconductor.org/packages/release/bioc/html/CoGAPS.html),
    [*ontoCAT*](https://www.bioconductor.org/packages/release/bioc/html/ontoCAT.html),
    [*DEgraph*](https://www.bioconductor.org/packages/release/bioc/html/DEgraph.html),
    [*NTW*](https://www.bioconductor.org/packages/release/bioc/html/NTW.html),
    and
    [*RCytoscape*](https://www.bioconductor.org/packages/release/bioc/html/RCytoscape.html).

Advanced statistical and modeling implementations

:   relevant to high-throughtput genetic analysis include
    [*BHC*](https://www.bioconductor.org/packages/release/bioc/html/BHC.html)
    (Bayesian Hierarchical Clustering),
    [*CGEN*](https://www.bioconductor.org/packages/release/bioc/html/CGEN.html)
    (case-control studies in genetic epidemiology), and
    [*SQUADD*](https://www.bioconductor.org/packages/release/bioc/html/SQUADD.html).

Image, cell-based, and other assay

:   packages, include
    [*imageHTS*](https://www.bioconductor.org/packages/release/bioc/html/imageHTS.html),
    [*CRImage*](https://www.bioconductor.org/packages/release/bioc/html/CRImage.html),
    [*coRNAi*](https://www.bioconductor.org/packages/release/bioc/html/coRNAi.html),
    [*GeneGA*](https://www.bioconductor.org/packages/release/bioc/html/GeneGA.html),
    [*NuPoP*](https://www.bioconductor.org/packages/release/bioc/html/NuPoP.html).

Our large collection of microarray- and organism-specific annotation
packages have been updated to include information current at the time of
the Bioconductor release. These annotation packages contain biological
information about microarray probes and the genes they are meant to
interrogate, or contain gene-based annotations of whole genomes. They
are particularly valuable in providing stable annotations for repeatable
research.

Several developments in packages maintained by the Bioconductor core
team are noteworthy. The *graphBAM* class in the
[*graph*](https://www.bioconductor.org/packages/release/bioc/html/graph.html)
package is available to manipulate very large graphs. The
[*GenomicRanges*](https://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html),
[*GenomicFeatures*](https://www.bioconductor.org/packages/release/bioc/html/GenomicFeatures.html),
and
[*Biostrings*](https://www.bioconductor.org/packages/release/bioc/html/Biostrings.html)
packages have enhanced classes such as *TranscriptDb* for representing
genome-scale 'track' annotations from common data resources,
*MultipleAlignment* for manipulating reference (and other
moderate-length) sequences in a microbiome project, and
*SummarizedExperiment* to collate range-based count data across samples
in sequence experiments. The
[*chipseq*](https://www.bioconductor.org/packages/release/bioc/html/chipseq.html)
package has enhanced functionality for peak calling, and has been
updated to use current data structures.

Further information on new and existing packages can be found on the
Bioconductor web site, which contains 'views' that identify coherent
groups of packages. The views link to on-line package descriptions,
vignettes, reference manuals, and use statistics.

# Other activities

The Bioconductor community met on July 28-30 at our annual conference in
Seattle for a combination of scientific talks and hands-on tutorials,
and on November 17-18 in Heidelberg, Germany for a meeting highlight
contributions from the European developer community. The active
Bioconductor mailing lists
(<http://bioconductor.org/docs/mailList.html>) connect users with each
other, to domain experts, and to maintainers eager to ensure that their
packages satisfy the needs of leading edge approaches. Bioconductor
package maintainers and the Bioconductor team invest considerable effort
in producing high-quality software. The Bioconductor team continues to
ensure quality software through technical and scientific reviews of new
packages, and daily builds of released packages on Linux, Windows, and
Macintosh platforms.

# Looking forward

Contributions from the Bioconductor community play an important role in
shaping each release. We anticipate continued efforts to provide
statistically informed analysis of next generation sequence data,
especially in the down-stream analysis of comprehensive, designed
sequencing experiments and integrative analyses. The next release cycle
promises to be one of active scientific growth and exploration.
:::
