<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
      margin-bottom: 0em;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
<title>BCClong: An R Package for Bayesian Consensus Clustering for Multiple Longitudinal Features</title>

<meta property="description" itemprop="description" content="It is very common nowadays for a study to collect multiple&#10;longitudinal features and appropriately integrating these features&#10;simultaneously for defining individual subgroups (i.e., clusters)&#10;becomes increasingly crucial to understanding population heterogeneity&#10;and predicting future outcomes. The aim of this paper is to describe a&#10;new package, BCClong, which implements a Bayesian consensus&#10;clustering (BCC) model for multiple longitudinal features. Compared to&#10;existing packages, several key features make the BCClong package&#10;appealing: (a) it allows simultaneous clustering of mixed-type (e.g.,&#10;continuous, discrete and categorical) longitudinal features, (b) it&#10;allows each longitudinal feature to be collected from different&#10;sources with measurements taken at distinct sets of time points (known&#10;as irregularly sampled longitudinal data), (c) it relaxes the&#10;assumption that all features have the same clustering structure by&#10;estimating the feature-specific (local) clusterings and consensus&#10;(global) clustering. Using two real data examples, we provide a&#10;tutorial with step-by-step instructions on using the package."/>

<link rel="canonical" href="https://doi.org/10.32614/RJ-2024-017/"/>
<link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>
<link rel="icon" type="image/vnd.microsoft.icon" href="../../resources/favicon.ico"/>

<!--  https://schema.org/Article -->
<meta property="article:published" itemprop="datePublished" content="2025-03-09"/>
<meta property="article:created" itemprop="dateCreated" content="2025-03-09"/>
<meta name="article:author" content="Zhiwen Tan"/>
<meta name="article:author" content="Chang Shen"/>
<meta name="article:author" content="Zihang Lu"/>

<!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
<meta property="og:title" content="BCClong: An R Package for Bayesian Consensus Clustering for Multiple Longitudinal Features"/>
<meta property="og:type" content="article"/>
<meta property="og:description" content="It is very common nowadays for a study to collect multiple&#10;longitudinal features and appropriately integrating these features&#10;simultaneously for defining individual subgroups (i.e., clusters)&#10;becomes increasingly crucial to understanding population heterogeneity&#10;and predicting future outcomes. The aim of this paper is to describe a&#10;new package, BCClong, which implements a Bayesian consensus&#10;clustering (BCC) model for multiple longitudinal features. Compared to&#10;existing packages, several key features make the BCClong package&#10;appealing: (a) it allows simultaneous clustering of mixed-type (e.g.,&#10;continuous, discrete and categorical) longitudinal features, (b) it&#10;allows each longitudinal feature to be collected from different&#10;sources with measurements taken at distinct sets of time points (known&#10;as irregularly sampled longitudinal data), (c) it relaxes the&#10;assumption that all features have the same clustering structure by&#10;estimating the feature-specific (local) clusterings and consensus&#10;(global) clustering. Using two real data examples, we provide a&#10;tutorial with step-by-step instructions on using the package."/>
<meta property="og:url" content="https://doi.org/10.32614/RJ-2024-017/"/>
<meta property="og:image" content="https://journal.r-project.org/articles/RJ-2024-017/Figures/traj.jpeg"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="The R Journal"/>

<!--  https://dev.twitter.com/cards/types/summary -->
<meta property="twitter:card" content="summary_large_image"/>
<meta property="twitter:title" content="BCClong: An R Package for Bayesian Consensus Clustering for Multiple Longitudinal Features"/>
<meta property="twitter:description" content="It is very common nowadays for a study to collect multiple&#10;longitudinal features and appropriately integrating these features&#10;simultaneously for defining individual subgroups (i.e., clusters)&#10;becomes increasingly crucial to understanding population heterogeneity&#10;and predicting future outcomes. The aim of this paper is to describe a&#10;new package, BCClong, which implements a Bayesian consensus&#10;clustering (BCC) model for multiple longitudinal features. Compared to&#10;existing packages, several key features make the BCClong package&#10;appealing: (a) it allows simultaneous clustering of mixed-type (e.g.,&#10;continuous, discrete and categorical) longitudinal features, (b) it&#10;allows each longitudinal feature to be collected from different&#10;sources with measurements taken at distinct sets of time points (known&#10;as irregularly sampled longitudinal data), (c) it relaxes the&#10;assumption that all features have the same clustering structure by&#10;estimating the feature-specific (local) clusterings and consensus&#10;(global) clustering. Using two real data examples, we provide a&#10;tutorial with step-by-step instructions on using the package."/>
<meta property="twitter:url" content="https://doi.org/10.32614/RJ-2024-017/"/>
<meta property="twitter:image" content="https://journal.r-project.org/articles/RJ-2024-017/Figures/traj.jpeg"/>

<!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
<meta name="citation_title" content="BCClong: An R Package for Bayesian Consensus Clustering for Multiple Longitudinal Features"/>
<meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2024-017"/>
<meta name="citation_pdf_url" content="RJ-2024-017.pdf"/>
<meta name="citation_volume" content="16"/>
<meta name="citation_issue" content="2"/>
<meta name="citation_doi" content="10.32614/RJ-2024-017"/>
<meta name="citation_journal_title" content="The R Journal"/>
<meta name="citation_issn" content="2073-4859"/>
<meta name="citation_firstpage" content="41"/>
<meta name="citation_lastpage" content="60"/>
<meta name="citation_fulltext_world_readable" content=""/>
<meta name="citation_online_date" content="2025/03/09"/>
<meta name="citation_publication_date" content="2025/03/09"/>
<meta name="citation_author" content="Zhiwen Tan"/>
<meta name="citation_author_institution" content="Department of Public Health Sciences"/>
<meta name="citation_author" content="Chang Shen"/>
<meta name="citation_author_institution" content="Department of Electrical and Computer Engineering"/>
<meta name="citation_author" content="Zihang Lu"/>
<meta name="citation_author_institution" content="Department of Public Health Sciences"/>
<!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Using a bayesian latent growth curve model to identify trajectories of positive affect and negative events following myocardial infarction;citation_publisher=Oxford University Press;citation_volume=6;citation_doi=10.1093/biostatistics/kxh022;citation_author=Michael R Elliott;citation_author=Joseph J Gallo;citation_author=Thomas R Ten Have;citation_author=Hillary R Bogner;citation_author=Ira R Katz"/>
  <meta name="citation_reference" content="citation_title=Identification of multivariate responders and non-responders by using Bayesian growth curve latent class models;citation_publisher=Wiley Online Library;citation_volume=58;citation_author=Benjamin E Leiby;citation_author=Mary D Sammel;citation_author=Thomas R Ten Have;citation_author=Kevin G Lynch"/>
  <meta name="citation_reference" content="citation_title=A Bayesian growth mixture model to examine maternal hypertension and birth outcomes;citation_publisher=Wiley Online Library;citation_volume=30;citation_author=Brian Neelon;citation_author=Geeta K Swamy;citation_author=Lane F Burgette;citation_author=Marie Lynn Miranda"/>
  <meta name="citation_reference" content="citation_title=Bayesian inference for finite mixtures of univariate and multivariate skew-normal and skew-t distributions;citation_publisher=Oxford University Press;citation_volume=11;citation_author=Sylvia Frühwirth-Schnatter;citation_author=Saumyadipta Pyne"/>
  <meta name="citation_reference" content="citation_title=Bayesian measures of model complexity and fit;citation_publisher=Wiley Online Library;citation_volume=64;citation_author=David J Spiegelhalter;citation_author=Nicola G Best;citation_author=Bradley P Carlin;citation_author=Angelika Van Der Linde"/>
  <meta name="citation_reference" content="citation_title=Deviance information criteria for missing data models;citation_publisher=International Society for Bayesian Analysis;citation_volume=1;citation_doi=10.1214/06-BA122;citation_author=Gilles Celeux;citation_author=Florence Forbes;citation_author=Christian P Robert;citation_author=D Michael Titterington;citation_author=Deviance information criteria for missing data models"/>
  <meta name="citation_reference" content="citation_title=Dealing with label switching in mixture models;citation_publisher=Wiley Online Library;citation_volume=62;citation_author=Matthew Stephens"/>
  <meta name="citation_reference" content="citation_title=Evaluating the accuracy of sampling-based approaches to the calculation of posterior moments;citation_publisher=Federal Reserve Bank of Minneapolis, Research Department Minneapolis, MN, USA;citation_volume=196;citation_author=John Geweke"/>
  <meta name="citation_reference" content="citation_title=Comparison of criteria for choosing the number of classes in Bayesian finite mixture models;citation_publisher=Public Library of Science;citation_volume=12;citation_author=Kazem Nasserinejad;citation_author=Joost Rosmalen;citation_author=Wim Kort;citation_author=Emmanuel Lesaffre"/>
  <meta name="citation_reference" content="citation_title=Posterior predictive assessment of model fitness via realized discrepancies;citation_publisher=JSTOR;citation_author=Andrew Gelman;citation_author=Xiao-Li Meng;citation_author=Hal Stern"/>
  <meta name="citation_reference" content="citation_title=Estimating normal means with a Dirichlet process prior;citation_publisher=Taylor &amp; Francis;citation_volume=89;citation_author=Michael D Escobar"/>
  <meta name="citation_reference" content="citation_title=Asymptotic equivalence of bayes cross validation and widely applicable information criterion in singular learning theory;citation_volume=11;citation_author=Sumio Watanabe"/>
  <meta name="citation_reference" content="citation_title=Bayesian comparison of latent variable models: Conditional versus marginal likelihoods;citation_publisher=Springer;citation_volume=84;citation_author=Edgar C Merkle;citation_author=Daniel Furr;citation_author=Sophia Rabe-Hesketh"/>
  <meta name="citation_reference" content="citation_title=Bayesian approaches to variable selection: A comparative study from practical perspectives;citation_publisher=SAGE Publications Sage UK: London, England;citation_volume=18;citation_author=Zihang Lu;citation_author=Wendy Lou"/>
  <meta name="citation_reference" content="citation_title=Clustering for multivariate continuous and discrete longitudinal data;citation_publisher=Institute of Mathematical Statistics;citation_volume=7;citation_doi=10.1214/12-AOAS580;citation_author=Arnošt Komárek;citation_author=Lenka Komárková"/>
  <meta name="citation_reference" content="citation_title=Bayesian consensus clustering;citation_publisher=Oxford University Press;citation_volume=29;citation_doi=10.1093/bioinformatics/btt425;citation_author=Eric F Lock;citation_author=David B Dunson"/>
  <meta name="citation_reference" content="citation_title=Shape invariant mixture model for clustering non-linear longitudinal growth trajectories;citation_publisher=SAGE Publications Sage UK: London, England;citation_volume=28;citation_author=Zihang Lu;citation_author=Wendy Lou"/>
  <meta name="citation_reference" content="citation_title=Capabilities of R package mixAK for clustering based on multivariate continuous and discrete longitudinal data;citation_publisher=Foundation for Open Access Statistics;citation_volume=59;citation_author=Arnošt Komárek;citation_author=Lenka Komárková"/>
  <meta name="citation_reference" content="citation_title=Group-based multi-trajectory modeling;citation_publisher=SAGE Publications Sage UK: London, England;citation_volume=27;citation_author=Daniel S Nagin;citation_author=Bobby L Jones;citation_author=Valeria Lima Passos;citation_author=Richard E Tremblay"/>
  <meta name="citation_reference" content="citation_title=Variational inference: A review for statisticians;citation_publisher=Taylor &amp; Francis;citation_volume=112;citation_author=David M Blei;citation_author=Alp Kucukelbir;citation_author=Jon D McAuliffe"/>
  <meta name="citation_reference" content="citation_title=Bayesian consensus clustering for multivariate longitudinal data;citation_publisher=Wiley Online Library;citation_volume=41;citation_doi=10.1002/sim.9225;citation_author=Zihang Lu;citation_author=Wendy Lou"/>
  <meta name="citation_reference" content="citation_title=Bayesian approaches to variable selection in mixture models with application to disease clustering;citation_publisher=Taylor &amp; Francis;citation_author=Zihang Lu;citation_author=Wendy Lou"/>
  <meta name="citation_reference" content="citation_title=KmL: K-means for longitudinal data;citation_volume=25;citation_doi=10.1007/s00180-009-0178-4;citation_author=Christophe Genolini;citation_author=Béatrice Falissard"/>
  <meta name="citation_reference" content="citation_title=KmL3D: A non-parametric algorithm for clustering joint trajectories;citation_publisher=Elsevier;citation_volume=109;citation_author=Christophe Genolini;citation_author=Jean-Baptiste Pingault;citation_author=Tarak Driss;citation_author=Sylvana Côté;citation_author=Richard E Tremblay;citation_author=Franck Vitaro;citation_author=Catherine Arnaud;citation_author=Bruno Falissard"/>
  <meta name="citation_reference" content="citation_title=A randomised controlled trial examining the longer-term outcomes of standard versus new antiepileptic drugs. The SANAD trial;citation_publisher=National Coordinating Centre for Health Technology Assessment;citation_volume=11;citation_author=AG Marson;citation_author=R Appleton;citation_author=GA Baker;citation_author=DW Chadwick;citation_author=J Doughty;citation_author=B Eaton;citation_author=C Gamble;citation_author=A Jacoby;citation_author=P Shackley;citation_author=DF Smith;citation_author=A randomised controlled trial examining the longer-term outcomes of standard versus new antiepileptic drugs. The SANAD trial"/>
  <meta name="citation_reference" content="citation_title=FlexMix: A general framework for finite mixture models and latent class regression in R;citation_volume=11;citation_doi=10.18637/jss.v011.i08;citation_author=Friedrich Leisch"/>
  <meta name="citation_reference" content="citation_title=Gbmt: Group-based multivariate trajectory modeling;citation_author=Alessandro Magrini"/>
  <meta name="citation_reference" content="citation_title=joineRML: A joint model and software package for time-to-event and multivariate longitudinal outcomes;citation_publisher=Springer;citation_volume=18;citation_author=Graeme L Hickey;citation_author=Pete Philipson;citation_author=Andrea Jorgensen;citation_author=Ruwanthi Kolamunnage-Dona"/>
  <meta name="citation_reference" content="citation_title=A joint modeling approach for clustering mixed-type multivariate longitudinal data: Application to the CHILD cohort study;citation_author=Zhiwen Tan;citation_author=Chang Shen;citation_author=Padmaja Subbarao;citation_author=Wendy Lou;citation_author=Zihang Lu"/>
  <meta name="citation_reference" content="citation_title=Cowplot: Streamlined plot theme and plot annotations for ’ggplot2’;citation_author=Claus O. Wilke"/>
  <meta name="citation_reference" content="citation_title=Understanding predictive information criteria for Bayesian models;citation_publisher=Springer;citation_volume=24;citation_author=Andrew Gelman;citation_author=Jessica Hwang;citation_author=Aki Vehtari"/>
  <meta name="citation_reference" content="citation_title=LaplacesDemon: Complete environment for bayesian inference;citation_author=Byron Hall"/>
  <meta name="citation_reference" content="citation_title=Cluster: Cluster analysis basics and extensions;citation_author=Martin Maechler;citation_author=Peter Rousseeuw;citation_author=Anja Struyf;citation_author=Mia Hubert;citation_author=Kurt Hornik"/>
  <meta name="citation_reference" content="citation_title=ggplot2: Elegant graphics for data analysis;citation_publisher=Springer-Verlag New York;citation_author=Hadley Wickham"/>
  <meta name="citation_reference" content="citation_title=label.switching: An R package for dealing with the label switching problem in MCMC outputs;citation_volume=69;citation_doi=10.18637/jss.v069.c01;citation_author=Panagiotis Papastamoulis"/>
  <meta name="citation_reference" content="citation_title=Fitting linear mixed-effects models using lme4;citation_volume=67;citation_doi=10.18637/jss.v067.i01;citation_author=Douglas Bates;citation_author=Martin Mächler;citation_author=Ben Bolker;citation_author=Steve Walker"/>
  <meta name="citation_reference" content="citation_title=MCMCpack: Markov chain Monte Carlo in R;citation_volume=42;citation_doi=10.18637/jss.v042.i09;citation_author=Andrew D. Martin;citation_author=Kevin M. Quinn;citation_author=Jong Hee Park"/>
  <meta name="citation_reference" content="citation_title=CODA: Convergence diagnosis and output analysis for MCMC;citation_volume=6;citation_author=Martyn Plummer;citation_author=Nicky Best;citation_author=Kate Cowles;citation_author=Karen Vines"/>
  <meta name="citation_reference" content="citation_title=mvtnorm: Multivariate normal and t distributions;citation_author=Alan Genz;citation_author=Frank Bretz;citation_author=Tetsuhisa Miwa;citation_author=Xuefei Mi;citation_author=Friedrich Leisch;citation_author=Fabian Scheipl;citation_author=Torsten Hothorn"/>
  <meta name="citation_reference" content="citation_title=Rcpp: Seamless R and C++ integration;citation_volume=40;citation_doi=10.18637/jss.v040.i08;citation_author=Dirk Eddelbuettel;citation_author=Romain François"/>
  <meta name="citation_reference" content="citation_title=Rmpfr: R MPFR - multiple precision floating-point reliable;citation_author=Martin Maechler"/>
  <meta name="citation_reference" content="citation_title=MASS: Support functions and datasets for venables and ripley’s MASS;citation_author=Brian Ripley;citation_author=Bill Venables;citation_author=Douglas M Bates;citation_author=Kurt Hornik;citation_author=Albrecht Gebhardt;citation_author=David Firth;citation_author= Ripley"/>
  <meta name="citation_reference" content="citation_title=Model-based clustering, classification, and density estimation using mclust in R;citation_publisher=Chapman; Hall/CRC;citation_doi=10.1201/9781003277965;citation_author=Luca Scrucca;citation_author=Chris Fraley;citation_author=T. Brendan Murphy;citation_author=Adrian E. Raftery"/>
  <meta name="citation_reference" content="citation_title=Truncdist: Truncated random variables;citation_author=Frederick Novomestky;citation_author=Saralees Nadarajah;citation_author=Maintainer Frederick Novomestky"/>
  <meta name="citation_reference" content="citation_title=Modern applied statistics with s;citation_publisher=Springer;citation_author=W. N. Venables;citation_author=B. D. Ripley"/>
  <meta name="citation_reference" content="citation_title=Assessment of agricultural sustainability in european union countries: A group-based multivariate trajectory approach;citation_publisher=Springer;citation_volume=106;citation_author=Alessandro Magrini"/>
  <meta name="citation_reference" content="citation_title=Distributed stochastic gradient MCMC;citation_publisher=PMLR;citation_author=Sungjin Ahn;citation_author=Babak Shahbaba;citation_author=Max Welling"/>
  <meta name="citation_reference" content="citation_title=How Gaussian mixture models might miss detecting factors that impact growth patterns;citation_publisher=JSTOR;citation_volume=12;citation_author=Brianna C Heggeseth;citation_author=Nicholas P Jewell"/>
  <meta name="citation_reference" content="citation_title=A Bayesian latent class model for integrating multi-source longitudinal data: Application to the CHILD cohort study;citation_publisher=Oxford University Press US;citation_volume=73;citation_author=Zihang Lu;citation_author=Padmaja Subbarao;citation_author=Wendy Lou"/>
  <meta name="citation_reference" content="citation_title=Joint clustering multiple longitudinal features: A comparison of methods and software packages with practical guidance;citation_publisher=Wiley Online Library;citation_volume=42;citation_author=Zihang Lu;citation_author=Mojtaba Ahmadiankalati;citation_author=Zhiwen Tan"/>
  <meta name="citation_reference" content="citation_title=A sparse factor model for clustering high-dimensional longitudinal data;citation_volume=43;citation_author=Zihang Lu;citation_author=Noirrit Kiran Chandra"/>
  <!--radix_placeholder_rmarkdown_metadata-->

<script type="text/json" id="radix-rmarkdown-metadata">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","date_received","journal","volume","issue","slug","citation_url","packages","preview","bibliography","CTV","legacy_pdf","legacy_converted","output","draft","pdf_url","doi","creative_commons","csl","canonical_url"]}},"value":[{"type":"character","attributes":{},"value":["BCClong: An R Package for Bayesian Consensus Clustering for Multiple Longitudinal Features"]},{"type":"character","attributes":{},"value":["It is very common nowadays for a study to collect multiple\nlongitudinal features and appropriately integrating these features\nsimultaneously for defining individual subgroups (i.e., clusters)\nbecomes increasingly crucial to understanding population heterogeneity\nand predicting future outcomes. The aim of this paper is to describe a\nnew package, BCClong, which implements a Bayesian consensus\nclustering (BCC) model for multiple longitudinal features. Compared to\nexisting packages, several key features make the BCClong package\nappealing: (a) it allows simultaneous clustering of mixed-type (e.g.,\ncontinuous, discrete and categorical) longitudinal features, (b) it\nallows each longitudinal feature to be collected from different\nsources with measurements taken at distinct sets of time points (known\nas irregularly sampled longitudinal data), (c) it relaxes the\nassumption that all features have the same clustering structure by\nestimating the feature-specific (local) clusterings and consensus\n(global) clustering. Using two real data examples, we provide a\ntutorial with step-by-step instructions on using the package."]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Zhiwen Tan"]},{"type":"character","attributes":{},"value":["Department of Public Health Sciences"]},{"type":"character","attributes":{},"value":["Queen's University\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Chang Shen"]},{"type":"character","attributes":{},"value":["Department of Electrical and Computer Engineering"]},{"type":"character","attributes":{},"value":["Queen's University\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Zihang Lu"]},{"type":"character","attributes":{},"value":["Department of Public Health Sciences"]},{"type":"character","attributes":{},"value":["*and* Department of Mathematics and Statistics","Queen's University","[zihang.lu@queensu.ca](zihang.lu@queensu.ca){.uri}\n"]}]}]},{"type":"character","attributes":{},"value":["2025-03-09"]},{"type":"character","attributes":{},"value":["2023-04-20"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"integer","attributes":{},"value":[41]},{"type":"integer","attributes":{},"value":[60]}]},{"type":"integer","attributes":{},"value":[16]},{"type":"integer","attributes":{},"value":[2]},{"type":"character","attributes":{},"value":["RJ-2024-017"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2024-017"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"list","attributes":{},"value":[]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["preview.png"]},{"type":"character","attributes":{},"value":["BCClong.bib"]},{"type":"list","attributes":{},"value":[]},{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[true]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc","mathjax","md_extension"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js"]},{"type":"character","attributes":{},"value":["-tex_math_single_backslash"]}]}]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["RJ-2024-017.pdf"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2024-017"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"character","attributes":{},"value":["/home/mitchell/R/x86_64-pc-linux-gnu-library/4.4/rjtools/rjournal.csl"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2024-017/"]}]}
</script>
<!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["BCClong.bib","BCClong.tex","Figures/num_cluster.jpeg","Figures/posterior_check.jpeg","Figures/posterior_prob.jpeg","Figures/trace_alpha.jpeg","Figures/trace_ga11.jpeg","Figures/trace_ga12.jpeg","Figures/trace_ga13.jpeg","Figures/trace_ga21.jpeg","Figures/trace_ga22.jpeg","Figures/trace_ga23.jpeg","Figures/trace_ppi.jpeg","Figures/traj_ga11.jpeg","Figures/traj.jpeg","Figures/trajplot_PBC910.jpeg","Figures/trajplot.jpeg","RJ-2024-017_files/anchor-4.2.2/anchor.min.js","RJ-2024-017_files/bowser-1.9.3/bowser.min.js","RJ-2024-017_files/distill-2.2.21/template.v2.js","RJ-2024-017_files/header-attrs-2.29/header-attrs.js","RJ-2024-017_files/jquery-3.6.0/jquery-3.6.0.js","RJ-2024-017_files/jquery-3.6.0/jquery-3.6.0.min.js","RJ-2024-017_files/jquery-3.6.0/jquery-3.6.0.min.map","RJ-2024-017_files/popper-2.6.0/popper.min.js","RJ-2024-017_files/tippy-6.2.7/tippy-bundle.umd.min.js","RJ-2024-017_files/tippy-6.2.7/tippy-light-border.css","RJ-2024-017_files/tippy-6.2.7/tippy.css","RJ-2024-017_files/tippy-6.2.7/tippy.umd.min.js","RJ-2024-017_files/webcomponents-2.0.0/webcomponents.js","RJ-2024-017.zip","RJournal.sty","RJreferences.bib","RJwrapper.log","RJwrapper.tex","texput.log"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
<meta name="distill:offset" content="../.."/>

<script type="application/javascript">

  window.headroom_prevent_pin = false;

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // initialize headroom for banner
    var header = $('header').get(0);
    var headerHeight = header.offsetHeight;
    var headroom = new Headroom(header, {
      tolerance: 5,
      onPin : function() {
        if (window.headroom_prevent_pin) {
          window.headroom_prevent_pin = false;
          headroom.unpin();
        }
      }
    });
    headroom.init();
    if(window.location.hash)
      headroom.unpin();
    $(header).addClass('headroom--transition');

    // offset scroll location for banner on hash change
    // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
    window.addEventListener("hashchange", function(event) {
      window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
    });

    // responsive menu
    $('.distill-site-header').each(function(i, val) {
      var topnav = $(this);
      var toggle = topnav.find('.nav-toggle');
      toggle.on('click', function() {
        topnav.toggleClass('responsive');
      });
    });

    // nav dropdowns
    $('.nav-dropbtn').click(function(e) {
      $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
      $(this).parent().siblings('.nav-dropdown')
         .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $("body").click(function(e){
      $('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $(".nav-dropdown").click(function(e){
      e.stopPropagation();
    });
  });
</script>

<style type="text/css">

/* Theme (user-documented overrideables for nav appearance) */

.distill-site-nav {
  color: rgba(255, 255, 255, 0.8);
  background-color: #0F2E3D;
  font-size: 15px;
  font-weight: 300;
}

.distill-site-nav a {
  color: inherit;
  text-decoration: none;
}

.distill-site-nav a:hover {
  color: white;
}

@media print {
  .distill-site-nav {
    display: none;
  }
}

.distill-site-header {

}

.distill-site-footer {

}


/* Site Header */

.distill-site-header {
  width: 100%;
  box-sizing: border-box;
  z-index: 3;
}

.distill-site-header .nav-left {
  display: inline-block;
  margin-left: 8px;
}

@media screen and (max-width: 768px) {
  .distill-site-header .nav-left {
    margin-left: 0;
  }
}


.distill-site-header .nav-right {
  float: right;
  margin-right: 8px;
}

.distill-site-header a,
.distill-site-header .title {
  display: inline-block;
  text-align: center;
  padding: 14px 10px 14px 10px;
}

.distill-site-header .title {
  font-size: 18px;
  min-width: 150px;
}

.distill-site-header .logo {
  padding: 0;
}

.distill-site-header .logo img {
  display: none;
  max-height: 20px;
  width: auto;
  margin-bottom: -4px;
}

.distill-site-header .nav-image img {
  max-height: 18px;
  width: auto;
  display: inline-block;
  margin-bottom: -3px;
}



@media screen and (min-width: 1000px) {
  .distill-site-header .logo img {
    display: inline-block;
  }
  .distill-site-header .nav-left {
    margin-left: 20px;
  }
  .distill-site-header .nav-right {
    margin-right: 20px;
  }
  .distill-site-header .title {
    padding-left: 12px;
  }
}


.distill-site-header .nav-toggle {
  display: none;
}

.nav-dropdown {
  display: inline-block;
  position: relative;
}

.nav-dropdown .nav-dropbtn {
  border: none;
  outline: none;
  color: rgba(255, 255, 255, 0.8);
  padding: 16px 10px;
  background-color: transparent;
  font-family: inherit;
  font-size: inherit;
  font-weight: inherit;
  margin: 0;
  margin-top: 1px;
  z-index: 2;
}

.nav-dropdown-content {
  display: none;
  position: absolute;
  background-color: white;
  min-width: 200px;
  border: 1px solid rgba(0,0,0,0.15);
  border-radius: 4px;
  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
  z-index: 1;
  margin-top: 2px;
  white-space: nowrap;
  padding-top: 4px;
  padding-bottom: 4px;
}

.nav-dropdown-content hr {
  margin-top: 4px;
  margin-bottom: 4px;
  border: none;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.nav-dropdown-active {
  display: block;
}

.nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
  color: black;
  padding: 6px 24px;
  text-decoration: none;
  display: block;
  text-align: left;
}

.nav-dropdown-content .nav-dropdown-header {
  display: block;
  padding: 5px 24px;
  padding-bottom: 0;
  text-transform: uppercase;
  font-size: 14px;
  color: #999999;
  white-space: nowrap;
}

.nav-dropdown:hover .nav-dropbtn {
  color: white;
}

.nav-dropdown-content a:hover {
  background-color: #ddd;
  color: black;
}

.nav-right .nav-dropdown-content {
  margin-left: -45%;
  right: 0;
}

@media screen and (max-width: 768px) {
  .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
  .distill-site-header a.nav-toggle {
    float: right;
    display: block;
  }
  .distill-site-header .title {
    margin-left: 0;
  }
  .distill-site-header .nav-right {
    margin-right: 0;
  }
  .distill-site-header {
    overflow: hidden;
  }
  .nav-right .nav-dropdown-content {
    margin-left: 0;
  }
}


@media screen and (max-width: 768px) {
  .distill-site-header.responsive {position: relative; min-height: 500px; }
  .distill-site-header.responsive a.nav-toggle {
    position: absolute;
    right: 0;
    top: 0;
  }
  .distill-site-header.responsive a,
  .distill-site-header.responsive .nav-dropdown {
    display: block;
    text-align: left;
  }
  .distill-site-header.responsive .nav-left,
  .distill-site-header.responsive .nav-right {
    width: 100%;
  }
  .distill-site-header.responsive .nav-dropdown {float: none;}
  .distill-site-header.responsive .nav-dropdown-content {position: relative;}
  .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
    display: block;
    width: 100%;
    text-align: left;
  }
}

/* Site Footer */

.distill-site-footer {
  width: 100%;
  overflow: hidden;
  box-sizing: border-box;
  z-index: 3;
  margin-top: 30px;
  padding-top: 30px;
  padding-bottom: 30px;
  text-align: center;
}

/* Headroom */

d-title {
  padding-top: 6rem;
}

@media print {
  d-title {
    padding-top: 4rem;
  }
}

.headroom {
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
}

.headroom--transition {
  transition: all .4s ease-in-out;
}

.headroom--unpinned {
  top: -100px;
}

.headroom--pinned {
  top: 0;
}

/* adjust viewport for navbar height */
/* helps vertically center bootstrap (non-distill) content */
.min-vh-100 {
  min-height: calc(100vh - 100px) !important;
}

</style>

<script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="../../site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"/>
<link href="../../site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"/>
<script src="../../site_libs/headroom-0.9.4/headroom.min.js"></script>
<!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

<style type="text/css">

body {
  background-color: white;
}

.pandoc-table {
  width: 100%;
}

.pandoc-table>caption {
  margin-bottom: 10px;
}

.pandoc-table th:not([align]) {
  text-align: left;
}

.pagedtable-footer {
  font-size: 15px;
}

d-byline .byline {
  grid-template-columns: 2fr 2fr;
}

d-byline .byline h3 {
  margin-block-start: 1.5em;
}

d-byline .byline .authors-affiliations h3 {
  margin-block-start: 0.5em;
}

.authors-affiliations .orcid-id {
  width: 16px;
  height:16px;
  margin-left: 4px;
  margin-right: 4px;
  vertical-align: middle;
  padding-bottom: 2px;
}

d-title .dt-tags {
  margin-top: 1em;
  grid-column: text;
}

.dt-tags .dt-tag {
  text-decoration: none;
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0em 0.4em;
  margin-right: 0.5em;
  margin-bottom: 0.4em;
  font-size: 70%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

d-article table.gt_table td,
d-article table.gt_table th {
  border-bottom: none;
  font-size: 100%;
}

.html-widget {
  margin-bottom: 2.0em;
}

.l-screen-inset {
  padding-right: 16px;
}

.l-screen .caption {
  margin-left: 10px;
}

.shaded {
  background: rgb(247, 247, 247);
  padding-top: 20px;
  padding-bottom: 20px;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .html-widget {
  margin-bottom: 0;
  border: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .shaded-content {
  background: white;
}

.text-output {
  margin-top: 0;
  line-height: 1.5em;
}

.hidden {
  display: none !important;
}

hr.section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  margin: 0px;
}


d-byline {
  border-top: none;
}

d-article {
  padding-top: 2.5rem;
  padding-bottom: 30px;
  border-top: none;
}

d-appendix {
  padding-top: 30px;
}

d-article>p>img {
  width: 100%;
}

d-article h2 {
  margin: 1rem 0 1.5rem 0;
}

d-article h3 {
  margin-top: 1.5rem;
}

d-article iframe {
  border: 1px solid rgba(0, 0, 0, 0.1);
  margin-bottom: 2.0em;
  width: 100%;
}

/* Tweak code blocks */

d-article div.sourceCode code,
d-article pre code {
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
}

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: auto;
}

d-article div.sourceCode {
  background-color: white;
}

d-article div.sourceCode pre {
  padding-left: 10px;
  font-size: 12px;
  border-left: 2px solid rgba(0,0,0,0.1);
}

d-article pre {
  font-size: 12px;
  color: black;
  background: none;
  margin-top: 0;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

d-article pre a {
  border-bottom: none;
}

d-article pre a:hover {
  border-bottom: none;
  text-decoration: underline;
}

d-article details {
  grid-column: text;
  margin-bottom: 0.8em;
}

@media(min-width: 768px) {

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: visible !important;
}

d-article div.sourceCode pre {
  padding-left: 18px;
  font-size: 14px;
}

/* tweak for Pandoc numbered line within distill */
d-article pre.numberSource code > span {
    left: -2em;
}

d-article pre {
  font-size: 14px;
}

}

figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

/* CSS for d-contents */

.d-contents {
  grid-column: text;
  color: rgba(0,0,0,0.8);
  font-size: 0.9em;
  padding-bottom: 1em;
  margin-bottom: 1em;
  padding-bottom: 0.5em;
  margin-bottom: 1em;
  padding-left: 0.25em;
  justify-self: start;
}

@media(min-width: 1000px) {
  .d-contents.d-contents-float {
    height: 0;
    grid-column-start: 1;
    grid-column-end: 4;
    justify-self: center;
    padding-right: 3em;
    padding-left: 2em;
  }
}

.d-contents nav h3 {
  font-size: 18px;
  margin-top: 0;
  margin-bottom: 1em;
}

.d-contents li {
  list-style-type: none
}

.d-contents nav > ul {
  padding-left: 0;
}

.d-contents ul {
  padding-left: 1em
}

.d-contents nav ul li {
  margin-top: 0.6em;
  margin-bottom: 0.2em;
}

.d-contents nav a {
  font-size: 13px;
  border-bottom: none;
  text-decoration: none
  color: rgba(0, 0, 0, 0.8);
}

.d-contents nav a:hover {
  text-decoration: underline solid rgba(0, 0, 0, 0.6)
}

.d-contents nav > ul > li > a {
  font-weight: 600;
}

.d-contents nav > ul > li > ul {
  font-weight: inherit;
}

.d-contents nav > ul > li > ul > li {
  margin-top: 0.2em;
}


.d-contents nav ul {
  margin-top: 0;
  margin-bottom: 0.25em;
}

.d-article-with-toc h2:nth-child(2) {
  margin-top: 0;
}


/* Figure */

.figure {
  position: relative;
  margin-bottom: 2.5em;
  margin-top: 1.5em;
}

.figure .caption {
  color: rgba(0, 0, 0, 0.6);
  font-size: 12px;
  line-height: 1.5em;
}

.figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

.figure .caption a {
  color: rgba(0, 0, 0, 0.6);
}

.figure .caption b,
.figure .caption strong, {
  font-weight: 600;
  color: rgba(0, 0, 0, 1.0);
}

/* Citations */

d-article .citation {
  color: inherit;
  cursor: inherit;
}

div.hanging-indent{
  margin-left: 1em; text-indent: -1em;
}

/* Citation hover box */

.tippy-box[data-theme~=light-border] {
  background-color: rgba(250, 250, 250, 0.95);
}

.tippy-content > p {
  margin-bottom: 0;
  padding: 2px;
}


/* Tweak 1000px media break to show more text */

@media(min-width: 1000px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 16px;
  }

  .grid {
    grid-column-gap: 16px;
  }

  d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
  }
  figure .caption, .figure .caption, figure figcaption {
    font-size: 13px;
  }
}

@media(min-width: 1180px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 32px;
  }

  .grid {
    grid-column-gap: 32px;
  }
}


/* Get the citation styles for the appendix (not auto-injected on render since
   we do our own rendering of the citation appendix) */

d-appendix .citation-appendix,
.d-appendix .citation-appendix {
  font-size: 11px;
  line-height: 15px;
  border-left: 1px solid rgba(0, 0, 0, 0.1);
  padding-left: 18px;
  border: 1px solid rgba(0,0,0,0.1);
  background: rgba(0, 0, 0, 0.02);
  padding: 10px 18px;
  border-radius: 3px;
  color: rgba(150, 150, 150, 1);
  overflow: hidden;
  margin-top: -12px;
  white-space: pre-wrap;
  word-wrap: break-word;
}

/* Include appendix styles here so they can be overridden */

d-appendix {
  contain: layout style;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-top: 60px;
  margin-bottom: 0;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  color: rgba(0,0,0,0.5);
  padding-top: 60px;
  padding-bottom: 48px;
}

d-appendix h3 {
  grid-column: page-start / text-start;
  font-size: 15px;
  font-weight: 500;
  margin-top: 1em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.65);
}

d-appendix h3 + * {
  margin-top: 1em;
}

d-appendix ol {
  padding: 0 0 0 15px;
}

@media (min-width: 768px) {
  d-appendix ol {
    padding: 0 0 0 30px;
    margin-left: -30px;
  }
}

d-appendix li {
  margin-bottom: 1em;
}

d-appendix a {
  color: rgba(0, 0, 0, 0.6);
}

d-appendix > * {
  grid-column: text;
}

d-appendix > d-footnote-list,
d-appendix > d-citation-list,
d-appendix > distill-appendix {
  grid-column: screen;
}

/* Include footnote styles here so they can be overridden */

d-footnote-list {
  contain: layout style;
}

d-footnote-list > * {
  grid-column: text;
}

d-footnote-list a.footnote-backlink {
  color: rgba(0,0,0,0.3);
  padding-left: 0.5em;
}



/* Anchor.js */

.anchorjs-link {
  /*transition: all .25s linear; */
  text-decoration: none;
  border-bottom: none;
}
*:hover > .anchorjs-link {
  margin-left: -1.125em !important;
  text-decoration: none;
  border-bottom: none;
}

/* Social footer */

.social_footer {
  margin-top: 30px;
  margin-bottom: 0;
  color: rgba(0,0,0,0.67);
}

.disqus-comments {
  margin-right: 30px;
}

.disqus-comment-count {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  cursor: pointer;
}

#disqus_thread {
  margin-top: 30px;
}

.article-sharing a {
  border-bottom: none;
  margin-right: 8px;
}

.article-sharing a:hover {
  border-bottom: none;
}

.sidebar-section.subscribe {
  font-size: 12px;
  line-height: 1.6em;
}

.subscribe p {
  margin-bottom: 0.5em;
}


.article-footer .subscribe {
  font-size: 15px;
  margin-top: 45px;
}


.sidebar-section.custom {
  font-size: 12px;
  line-height: 1.6em;
}

.custom p {
  margin-bottom: 0.5em;
}

/* Styles for listing layout (hide title) */
.layout-listing d-title, .layout-listing .d-title {
  display: none;
}

/* Styles for posts lists (not auto-injected) */


.posts-with-sidebar {
  padding-left: 45px;
  padding-right: 45px;
}

.posts-list .description h2,
.posts-list .description p {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

.posts-list .description h2 {
  font-weight: 700;
  border-bottom: none;
  padding-bottom: 0;
}

.posts-list h2.post-tag {
  border-bottom: 1px solid rgba(0, 0, 0, 0.2);
  padding-bottom: 12px;
}
.posts-list {
  margin-top: 60px;
  margin-bottom: 24px;
}

.posts-list .post-preview {
  text-decoration: none;
  overflow: hidden;
  display: block;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  padding: 24px 0;
}

.post-preview-last {
  border-bottom: none !important;
}

.posts-list .posts-list-caption {
  grid-column: screen;
  font-weight: 400;
}

.posts-list .post-preview h2 {
  margin: 0 0 6px 0;
  line-height: 1.2em;
  font-style: normal;
  font-size: 24px;
}

.posts-list .post-preview p {
  margin: 0 0 12px 0;
  line-height: 1.4em;
  font-size: 16px;
}

.posts-list .post-preview .thumbnail {
  box-sizing: border-box;
  margin-bottom: 24px;
  position: relative;
  max-width: 500px;
}
.posts-list .post-preview img {
  width: 100%;
  display: block;
}

.posts-list .metadata {
  font-size: 12px;
  line-height: 1.4em;
  margin-bottom: 18px;
}

.posts-list .metadata > * {
  display: inline-block;
}

.posts-list .metadata .publishedDate {
  margin-right: 2em;
}

.posts-list .metadata .dt-authors {
  display: block;
  margin-top: 0.3em;
  margin-right: 2em;
}

.posts-list .dt-tags {
  display: block;
  line-height: 1em;
}

.posts-list .dt-tags .dt-tag {
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0.3em 0.4em;
  margin-right: 0.2em;
  margin-bottom: 0.4em;
  font-size: 60%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

.posts-list img {
  opacity: 1;
}

.posts-list img[data-src] {
  opacity: 0;
}

.posts-more {
  clear: both;
}


.posts-sidebar {
  font-size: 16px;
}

.posts-sidebar h3 {
  font-size: 16px;
  margin-top: 0;
  margin-bottom: 0.5em;
  font-weight: 400;
  text-transform: uppercase;
}

.sidebar-section {
  margin-bottom: 30px;
}

.categories ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.categories li {
  color: rgba(0, 0, 0, 0.8);
  margin-bottom: 0;
}

.categories li>a {
  border-bottom: none;
}

.categories li>a:hover {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
}

.categories .active {
  font-weight: 600;
}

.categories .category-count {
  color: rgba(0, 0, 0, 0.4);
}


@media(min-width: 768px) {
  .posts-list .post-preview h2 {
    font-size: 26px;
  }
  .posts-list .post-preview .thumbnail {
    float: right;
    width: 30%;
    margin-bottom: 0;
  }
  .posts-list .post-preview .description {
    float: left;
    width: 45%;
  }
  .posts-list .post-preview .metadata {
    float: left;
    width: 20%;
    margin-top: 8px;
  }
  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.5em;
    font-size: 16px;
  }
  .posts-with-sidebar .posts-list {
    float: left;
    width: 75%;
  }
  .posts-with-sidebar .posts-sidebar {
    float: right;
    width: 20%;
    margin-top: 60px;
    padding-top: 24px;
    padding-bottom: 24px;
  }
}


/* Improve display for browsers without grid (IE/Edge <= 15) */

.downlevel {
  line-height: 1.6em;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  margin: 0;
}

.downlevel .d-title {
  padding-top: 6rem;
  padding-bottom: 1.5rem;
}

.downlevel .d-title h1 {
  font-size: 50px;
  font-weight: 700;
  line-height: 1.1em;
  margin: 0 0 0.5rem;
}

.downlevel .d-title p {
  font-weight: 300;
  font-size: 1.2rem;
  line-height: 1.55em;
  margin-top: 0;
}

.downlevel .d-byline {
  padding-top: 0.8em;
  padding-bottom: 0.8em;
  font-size: 0.8rem;
  line-height: 1.8em;
}

.downlevel .section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
}

.downlevel .d-article {
  font-size: 1.06rem;
  line-height: 1.7em;
  padding-top: 1rem;
  padding-bottom: 2rem;
}


.downlevel .d-appendix {
  padding-left: 0;
  padding-right: 0;
  max-width: none;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.5);
  padding-top: 40px;
  padding-bottom: 48px;
}

.downlevel .footnotes ol {
  padding-left: 13px;
}

.downlevel .base-grid,
.downlevel .distill-header,
.downlevel .d-title,
.downlevel .d-abstract,
.downlevel .d-article,
.downlevel .d-appendix,
.downlevel .distill-appendix,
.downlevel .d-byline,
.downlevel .d-footnote-list,
.downlevel .d-citation-list,
.downlevel .distill-footer,
.downlevel .appendix-bottom,
.downlevel .posts-container {
  padding-left: 40px;
  padding-right: 40px;
}

@media(min-width: 768px) {
  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
  padding-left: 150px;
  padding-right: 150px;
  max-width: 900px;
}
}

.downlevel pre code {
  display: block;
  border-left: 2px solid rgba(0, 0, 0, .1);
  padding: 0 0 0 20px;
  font-size: 14px;
}

.downlevel code, .downlevel pre {
  color: black;
  background: none;
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

.downlevel .posts-list .post-preview {
  color: inherit;
}



</style>

<script type="application/javascript">

function is_downlevel_browser() {
  if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                 window.navigator.userAgent)) {
    return true;
  } else {
    return window.load_distill_framework === undefined;
  }
}

// show body when load is complete
function on_load_complete() {

  // add anchors
  if (window.anchors) {
    window.anchors.options.placement = 'left';
    window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
  }


  // set body to visible
  document.body.style.visibility = 'visible';

  // force redraw for leaflet widgets
  if (window.HTMLWidgets) {
    var maps = window.HTMLWidgets.findAll(".leaflet");
    $.each(maps, function(i, el) {
      var map = this.getMap();
      map.invalidateSize();
      map.eachLayer(function(layer) {
        if (layer instanceof L.TileLayer)
          layer.redraw();
      });
    });
  }

  // trigger 'shown' so htmlwidgets resize
  $('d-article').trigger('shown');
}

function init_distill() {

  init_common();

  // create front matter
  var front_matter = $('<d-front-matter></d-front-matter>');
  $('#distill-front-matter').wrap(front_matter);

  // create d-title
  $('.d-title').changeElementType('d-title');

  // separator
  var separator = '<hr class="section-separator" style="clear: both"/>';
  // prepend separator above appendix
  $('.d-byline').before(separator);
  $('.d-article').before(separator);

  // create d-byline
  var byline = $('<d-byline></d-byline>');
  $('.d-byline').replaceWith(byline);

  // create d-article
  var article = $('<d-article></d-article>');
  $('.d-article').wrap(article).children().unwrap();

  // move posts container into article
  $('.posts-container').appendTo($('d-article'));

  // create d-appendix
  $('.d-appendix').changeElementType('d-appendix');

  // flag indicating that we have appendix items
  var appendix = $('.appendix-bottom').children('h3').length > 0;

  // replace footnotes with <d-footnote>
  $('.footnote-ref').each(function(i, val) {
    appendix = true;
    var href = $(this).attr('href');
    var id = href.replace('#', '');
    var fn = $('#' + id);
    var fn_p = $('#' + id + '>p');
    fn_p.find('.footnote-back').remove();
    var text = fn_p.html();
    var dtfn = $('<d-footnote></d-footnote>');
    dtfn.html(text);
    $(this).replaceWith(dtfn);
  });
  // remove footnotes
  $('.footnotes').remove();

  // move refs into #references-listing
  $('#references-listing').replaceWith($('#refs'));

  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    var id = $(this).attr('id');
    $('.d-contents a[href="#' + id + '"]').parent().remove();
    appendix = true;
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
  });

  // show d-appendix if we have appendix content
  $("d-appendix").css('display', appendix ? 'grid' : 'none');

  // localize layout chunks to just output
  $('.layout-chunk').each(function(i, val) {

    // capture layout
    var layout = $(this).attr('data-layout');

    // apply layout to markdown level block elements
    var elements = $(this).children().not('details, div.sourceCode, pre, script');
    elements.each(function(i, el) {
      var layout_div = $('<div class="' + layout + '"></div>');
      if (layout_div.hasClass('shaded')) {
        var shaded_content = $('<div class="shaded-content"></div>');
        $(this).wrap(shaded_content);
        $(this).parent().wrap(layout_div);
      } else {
        $(this).wrap(layout_div);
      }
    });


    // unwrap the layout-chunk div
    $(this).children().unwrap();
  });

  // remove code block used to force  highlighting css
  $('.distill-force-highlighting-css').parent().remove();

  // remove empty line numbers inserted by pandoc when using a
  // custom syntax highlighting theme, except when numbering line
  // in code chunk
  $('pre:not(.numberLines) code.sourceCode a:empty').remove();

  // load distill framework
  load_distill_framework();

  // wait for window.distillRunlevel == 4 to do post processing
  function distill_post_process() {

    if (!window.distillRunlevel || window.distillRunlevel < 4)
      return;

    // hide author/affiliations entirely if we have no authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;
    if (!have_authors)
      $('d-byline').addClass('hidden');

    // article with toc class
    $('.d-contents').parent().addClass('d-article-with-toc');

    // strip links that point to #
    $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

    // add orcid ids
    $('.authors-affiliations').find('.author').each(function(i, el) {
      var orcid_id = front_matter.authors[i].orcidID;
      var author_name = front_matter.authors[i].author
      if (orcid_id) {
        var a = $('<a></a>');
        a.attr('href', 'https://orcid.org/' + orcid_id);
        var img = $('<img></img>');
        img.addClass('orcid-id');
        img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
        img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
        a.append(img);
        $(this).append(a);
      }
    });

    // hide elements of author/affiliations grid that have no value
    function hide_byline_column(caption) {
      $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
    }

    // affiliations
    var have_affiliations = false;
    for (var i = 0; i<front_matter.authors.length; ++i) {
      var author = front_matter.authors[i];
      if (author.affiliation !== "&nbsp;") {
        have_affiliations = true;
        break;
      }
    }
    if (!have_affiliations)
      $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

    // published date
    if (!front_matter.publishedDate)
      hide_byline_column("Published");

    // document object identifier
    var doi = $('d-byline').find('h3:contains("DOI")');
    var doi_p = doi.next().empty();
    if (!front_matter.doi) {
      // if we have a citation and valid citationText then link to that
      if ($('#citation').length > 0 && front_matter.citationText) {
        doi.html('Citation');
        $('<a href="#citation"></a>')
          .text(front_matter.citationText)
          .appendTo(doi_p);
      } else {
        hide_byline_column("DOI");
      }
    } else {
      $('<a></a>')
         .attr('href', "https://doi.org/" + front_matter.doi)
         .html(front_matter.doi)
         .appendTo(doi_p);
    }

     // change plural form of authors/affiliations
    if (front_matter.authors.length === 1) {
      var grid = $('.authors-affiliations');
      grid.children('h3:contains("Authors")').text('Author');
      grid.children('h3:contains("Affiliations")').text('Affiliation');
    }

    // remove d-appendix and d-footnote-list local styles
    $('d-appendix > style:first-child').remove();
    $('d-footnote-list > style:first-child').remove();

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // hoverable references
    $('span.citation[data-cites]').each(function() {
      const citeChild = $(this).children()[0]
      // Do not process if @xyz has been used without escaping and without bibliography activated
      // https://github.com/rstudio/distill/issues/466
      if (citeChild === undefined) return true

      if (citeChild.nodeName == "D-FOOTNOTE") {
        var fn = citeChild
        $(this).html(fn.shadowRoot.querySelector("sup"))
        $(this).id = fn.id
        fn.remove()
      }
      var refs = $(this).attr('data-cites').split(" ");
      var refHtml = refs.map(function(ref) {
        // Could use CSS.escape too here, we insure backward compatibility in navigator
        return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
      }).join("\n");
      window.tippy(this, {
        allowHTML: true,
        content: refHtml,
        maxWidth: 500,
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start'
      });
    });

    // fix footnotes in tables (#411)
    // replacing broken distill.pub feature
    $('table d-footnote').each(function() {
      // we replace internal showAtNode methode which is triggered when hovering a footnote
      this.hoverBox.showAtNode = function(node) {
        // ported from https://github.com/distillpub/template/pull/105/files
        calcOffset = function(elem) {
            let x = elem.offsetLeft;
            let y = elem.offsetTop;
            // Traverse upwards until an `absolute` element is found or `elem`
            // becomes null.
            while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                x += elem.offsetLeft;
                y += elem.offsetTop;
            }

            return { left: x, top: y };
        }
        // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
        const bbox = node.getBoundingClientRect();
        const offset = calcOffset(node);
        this.show([offset.left + bbox.width, offset.top + bbox.height]);
      }
    })

    // clear polling timer
    clearInterval(tid);

    // show body now that everything is ready
    on_load_complete();
  }

  var tid = setInterval(distill_post_process, 50);
  distill_post_process();

}

function init_downlevel() {

  init_common();

   // insert hr after d-title
  $('.d-title').after($('<hr class="section-separator"/>'));

  // check if we have authors
  var front_matter = JSON.parse($("#distill-front-matter").html());
  var have_authors = front_matter.authors && front_matter.authors.length > 0;

  // manage byline/border
  if (!have_authors)
    $('.d-byline').remove();
  $('.d-byline').after($('<hr class="section-separator"/>'));
  $('.d-byline a').remove();

  // remove toc
  $('.d-contents').remove();

  // move appendix elements
  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
  });


  // inject headers into references and footnotes
  var refs_header = $('<h3></h3>');
  refs_header.text('References');
  $('#refs').prepend(refs_header);

  var footnotes_header = $('<h3></h3');
  footnotes_header.text('Footnotes');
  $('.footnotes').children('hr').first().replaceWith(footnotes_header);

  // move appendix-bottom entries to the bottom
  $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
  $('.appendix-bottom').remove();

  // remove appendix if it's empty
  if ($('.d-appendix').children().length === 0)
    $('.d-appendix').remove();

  // prepend separator above appendix
  $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

  // trim code
  $('pre>code').each(function(i, val) {
    $(this).html($.trim($(this).html()));
  });

  // move posts-container right before article
  $('.posts-container').insertBefore($('.d-article'));

  $('body').addClass('downlevel');

  on_load_complete();
}


function init_common() {

  // jquery plugin to change element types
  (function($) {
    $.fn.changeElementType = function(newType) {
      var attrs = {};

      $.each(this[0].attributes, function(idx, attr) {
        attrs[attr.nodeName] = attr.nodeValue;
      });

      this.replaceWith(function() {
        return $("<" + newType + "/>", attrs).append($(this).contents());
      });
    };
  })(jQuery);

  // prevent underline for linked images
  $('a > img').parent().css({'border-bottom' : 'none'});

  // mark non-body figures created by knitr chunks as 100% width
  $('.layout-chunk').each(function(i, val) {
    var figures = $(this).find('img, .html-widget');
    // ignore leaflet img layers (#106)
    figures = figures.filter(':not(img[class*="leaflet"])')
    if ($(this).attr('data-layout') !== "l-body") {
      figures.css('width', '100%');
    } else {
      figures.css('max-width', '100%');
      figures.filter("[width]").each(function(i, val) {
        var fig = $(this);
        fig.css('width', fig.attr('width') + 'px');
      });

    }
  });

  // auto-append index.html to post-preview links in file: protocol
  // and in rstudio ide preview
  $('.post-preview').each(function(i, val) {
    if (window.location.protocol === "file:")
      $(this).attr('href', $(this).attr('href') + "index.html");
  });

  // get rid of index.html references in header
  if (window.location.protocol !== "file:") {
    $('.distill-site-header a[href]').each(function(i,val) {
      $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
    });
  }

  // add class to pandoc style tables
  $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
  $('.kable-table').children('table').addClass('pandoc-table');

  // add figcaption style to table captions
  $('caption').parent('table').addClass("figcaption");

  // initialize posts list
  if (window.init_posts_list)
    window.init_posts_list();

  // implmement disqus comment link
  $('.disqus-comment-count').click(function() {
    window.headroom_prevent_pin = true;
    $('#disqus_thread').toggleClass('hidden');
    if (!$('#disqus_thread').hasClass('hidden')) {
      var offset = $(this).offset();
      $(window).resize();
      $('html, body').animate({
        scrollTop: offset.top - 35
      });
    }
  });
}

document.addEventListener('DOMContentLoaded', function() {
  if (is_downlevel_browser())
    init_downlevel();
  else
    window.addEventListener('WebComponentsReady', init_distill);
});

</script>

<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.5);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

.distill-site-footer {
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */
</style>
<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.7);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #ffffff;
}

.distill-site-footer {
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #fafafa;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */

.nav-right > a {
  text-transform: uppercase;
}

d-title h1, d-title p, d-title figure,
d-abstract p, d-abstract b {
  grid-column: page;
}

.rj-blue {
  color: #2467bb;
}

ul li {
  line-height: 1.6;
  margin-top: 0em;
  margin-bottom: 0em;
}</style>
<style type="text/css">
/* base style */

/* FONT FAMILIES */

:root {
  --heading-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  --mono-default: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  --body-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

body,
.posts-list .post-preview p,
.posts-list .description p {
  font-family: var(--body-font), var(--body-default);
}

h1, h2, h3, h4, h5, h6,
.posts-list .post-preview h2,
.posts-list .description h2 {
  font-family: var(--heading-font), var(--heading-default);
}

d-article div.sourceCode code,
d-article pre code {
  font-family: var(--mono-font), var(--mono-default);
}


/*-- TITLE --*/
d-title h1,
.posts-list > h1 {
  color: var(--title-color, black);
}

d-title h1 {
  font-size: var(--title-size, 50px);
}

/*-- HEADERS --*/
d-article h1,
d-article h2,
d-article h3,
d-article h4,
d-article h5,
d-article h6 {
  color: var(--header-color, rgba(0, 0, 0, 0.8));
}

/*-- BODY --*/
d-article > p,  /* only text inside of <p> tags */
d-article > ul, /* lists */
d-article > ol {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
  font-size: var(--body-size, 1.06rem);
}


/*-- CODE --*/
d-article div.sourceCode code,
d-article pre code {
  font-size: var(--code-size, 14px);
}

/*-- ASIDE --*/
d-article aside {
  font-size: var(--aside-size, 12px);
  color: var(--aside-color, rgba(0, 0, 0, 0.6));
}

/*-- FIGURE CAPTIONS --*/
figure .caption,
figure figcaption,
.figure .caption {
  font-size: var(--fig-cap-size, 13px);
  color: var(--fig-cap-color, rgba(0, 0, 0, 0.6));
}

/*-- METADATA --*/
d-byline h3 {
  font-size: var(--heading-size, 0.6rem);
  color: var(--heading-color, rgba(0, 0, 0, 0.5));
}

d-byline {
  font-size: var(--body-size, 0.8rem);
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

d-byline a,
d-article d-byline a {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

/*-- TABLE OF CONTENTS --*/
.d-contents nav h3 {
  font-size: var(--heading-size, 18px);
}

.d-contents nav a {
  font-size: var(--contents-size, 13px);
}

/*-- APPENDIX --*/
d-appendix h3 {
  font-size: var(--heading-size, 15px);
  color: var(--heading-color, rgba(0, 0, 0, 0.65));
}

d-appendix {
  font-size: var(--text-size, 0.8em);
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

d-appendix d-footnote-list a.footnote-backlink {
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

/*-- WEBSITE HEADER + FOOTER --*/
.distill-site-header .title {
  font-size: var(--title-size, 18px);
  font-family: var(--navbar-font), var(--heading-default);
}

.distill-site-header a,
.nav-dropdown .nav-dropbtn {
  font-family: var(--navbar-font), var(--heading-default);
}

.nav-dropdown .nav-dropbtn {
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  font-size: var(--text-size, 15px);
}

.distill-site-header a:hover,
.nav-dropdown:hover .nav-dropbtn {
  color: var(--hover-color, white);
}

.distill-site-header {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer a:hover {
  color: var(--hover-color, white);
}</style>
<!--/radix_placeholder_distill-->
  <script src="../../site_libs/header-attrs-2.29/header-attrs.js"></script>
  <script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="../../site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="../../site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="../../site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="../../site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="../../site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="../../site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="../../site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="../../site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
<!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Mathjax config (add automatic linebreaks when supported)
      // MathJax = {
      //    tex: {
      //        inlineMath: [['$', '$'], ['\\(', '\\)']],
      //        displayMath: [['$$', '$$'], ['\\[', '\\]']],
      //        tags: 'ams',
      //        multline: true,
      //    },
      //    options: {
      //        linebreaks: { automatic: true },
      //    },
      // };

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      buttons.append('<a href="' + slug + '.zip" class="dt-tag"><i class="fas fa-file-zipper"></i> Supplement</a>')

      // adds Abstract: in front of the first <p> in the title section --
      // unless it happens to be the subtitle (FIXME: this is a bad hack - can't distill do this?)
      var tpar = $("d-title p:not(:empty)").filter(function() {
        return !$(this).hasClass("subtitle");
      }).first();
      if (tpar) {
        const abstract = $('<d-abstract>')
        abstract.append('<b>Abstract:</b><br>')
        abstract.append(tpar) // Move description to d-abstract
        $("d-title p:empty").remove() // Remove empty paragraphs after title
        abstract.append(buttons)
        abstract.insertAfter($('d-title')) // Add abstract section after title */
      }

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
    
    d-appendix .citation-appendix,
    .d-appendix .citation-appendix {
      color: rgb(60, 60, 60);
    }

    d-article h2 {
      border-bottom: 0px solid rgba(0, 0, 0, 0.1);
      padding-bottom: 0rem;
    }
    d-article h3 {
      font-size: 20px;
    }
    d-article h4 {
      font-size: 18px;
      text-transform: none;
    }

    @media (min-width: 1024px) {
      d-article h2 {
        font-size: 32px;
      }
      d-article h3 {
        font-size: 24px;
      }
      d-article h4 {
        font-size: 20px;
      }
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"BCClong: An R Package for Bayesian Consensus Clustering for Multiple Longitudinal Features","description":"It is very common nowadays for a study to collect multiple\nlongitudinal features and appropriately integrating these features\nsimultaneously for defining individual subgroups (i.e., clusters)\nbecomes increasingly crucial to understanding population heterogeneity\nand predicting future outcomes. The aim of this paper is to describe a\nnew package, BCClong, which implements a Bayesian consensus\nclustering (BCC) model for multiple longitudinal features. Compared to\nexisting packages, several key features make the BCClong package\nappealing: (a) it allows simultaneous clustering of mixed-type (e.g.,\ncontinuous, discrete and categorical) longitudinal features, (b) it\nallows each longitudinal feature to be collected from different\nsources with measurements taken at distinct sets of time points (known\nas irregularly sampled longitudinal data), (c) it relaxes the\nassumption that all features have the same clustering structure by\nestimating the feature-specific (local) clusterings and consensus\n(global) clustering. Using two real data examples, we provide a\ntutorial with step-by-step instructions on using the package.","doi":"10.32614/RJ-2024-017","authors":[{"author":"Zhiwen Tan","authorURL":"#","affiliation":"Department of Public Health Sciences","affiliationURL":"#","orcidID":""},{"author":"Chang Shen","authorURL":"#","affiliation":"Department of Electrical and Computer Engineering","affiliationURL":"#","orcidID":""},{"author":"Zihang Lu","authorURL":"#","affiliation":"Department of Public Health Sciences","affiliationURL":"#","orcidID":""}],"publishedDate":"2025-03-09T00:00:00.000+00:00","citationText":"Tan, et al., 2025"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a class="logo" href="../../index.html">
<img src="../../resources/rlogo.png" alt="Logo"/>
</a>
<a href="../../index.html" class="title">The R Journal</a>
</div>
<div class="nav-right">
<a href="../../index.html">Home</a>
<a href="../../issues/2024-4">Current</a>
<a href="../../issues.html">Issues</a>
<a href="../../news.html">News</a>
<a href="../../submissions.html">Submit</a>
<a href="../../editors.html">Editorial board</a>
<a href="../../articles.xml">
<i class="fa fa-rss" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>BCClong: An R Package for Bayesian Consensus Clustering for Multiple Longitudinal Features</h1>

<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>It is very common nowadays for a study to collect multiple
longitudinal features and appropriately integrating these features
simultaneously for defining individual subgroups (i.e., clusters)
becomes increasingly crucial to understanding population heterogeneity
and predicting future outcomes. The aim of this paper is to describe a
new package, BCClong, which implements a Bayesian consensus
clustering (BCC) model for multiple longitudinal features. Compared to
existing packages, several key features make the BCClong package
appealing: (a) it allows simultaneous clustering of mixed-type (e.g.,
continuous, discrete and categorical) longitudinal features, (b) it
allows each longitudinal feature to be collected from different
sources with measurements taken at distinct sets of time points (known
as irregularly sampled longitudinal data), (c) it relaxes the
assumption that all features have the same clustering structure by
estimating the feature-specific (local) clusterings and consensus
(global) clustering. Using two real data examples, we provide a
tutorial with step-by-step instructions on using the package.</p></p>
</div>

<div class="d-byline">
  Zhiwen Tan  (Department of Public Health Sciences)
  
,   Chang Shen  (Department of Electrical and Computer Engineering)
  
,   Zihang Lu  (Department of Public Health Sciences)
  
<br/>2025-03-09
</div>

<div class="d-article">
<div class="article">
<h3 data-number="1" id="sec:intro"><span class="header-section-number">1</span> Introduction</h3>
<p>Cluster analysis has been widely used to identify subgroups of the study
population, and it is also known as an unsupervised learning method in
the machine learning literature. When longitudinal data is available,
identifying distinct developmental patterns is a common goal in many
research studies. Traditional statistical models often focus on
clustering a single longitudinal feature, which may ignore the
complexity of the clustering structures and the dynamic relationship
when multiple longitudinal features are available. It is very common
nowadays for a study to collect multiple features and appropriately
integrating multiple longitudinal features simultaneously for defining
individual subgroups (i.e., clusters) becomes increasingly crucial to
understanding population heterogeneity and predicting future outcomes.
In particular, compared to clustering a single longitudinal feature,
simultaneously clustering multiple longitudinal features (also known as
joint clustering) facilitates the discovery of co-occurring trajectory
patterns for multiple features of interest.</p>
<p>Several R packages have been developed to cluster longitudinal data with
multiple features. For model-based clustering, <strong>gbmt</strong> package
<span class="citation" data-cites="Magrini2022">(<a href="#ref-Magrini2022" role="doc-biblioref">Magrini 2022b</a>)</span> and <strong>flexmix</strong> package <span class="citation" data-cites="Leisch2004">(<a href="#ref-Leisch2004" role="doc-biblioref">Leisch 2004</a>)</span> are developed to
implement the group-based multi-trajectory analysis (GBMT)
<span class="citation" data-cites="Nagin2018 Magrini2022a">(<a href="#ref-Nagin2018" role="doc-biblioref">Nagin et al. 2018</a>; <a href="#ref-Magrini2022a" role="doc-biblioref">Magrini 2022a</a>)</span>. Both packages implement EM algorithm
allowing unbalanced data and missing values, and can only be applied to
continuous features. Notably, a SAS procedure PROC TRAJ is also
developed for this model and has been widely used <span class="citation" data-cites="Nagin2018">(<a href="#ref-Nagin2018" role="doc-biblioref">Nagin et al. 2018</a>)</span>. The GBMT
assumes the features and repeated measures are independent conditional
on the cluster membership. Alternatively, <strong>lcmm</strong> and <strong>mixAK</strong>
packages are developed under the framework of mixed-effect models. The
dependence between features and repeated measurements is captured by
joint modeling of the random effects. Specifically, the <strong>lcmm</strong> package
is designed for performing analyses such as latent class mixed models
for a single longitudinal feature, joint latent class mixed models for
multiple longitudinal features as well as joint latent class mixed
models for longitudinal features and time-to-event outcomes. While this
package allows modeling multiple continuous or categorical longitudinal
data of the same type, it does not support clustering mixed data types
(e.g., continuous and categorical) simultaneously. On the other hand,
the <strong>mixAK</strong> package can handle same-type (e.g., multiple continuous)
or mixed-type (e.g., continuous, discrete and categorical) features. For
nonparametric clustering, the <strong>kml3d</strong> package <span class="citation" data-cites="Genolini2013">(<a href="#ref-Genolini2013" role="doc-biblioref">Genolini et al. 2013</a>)</span> is
developed to implement the K-means clustering for multiple longitudinal
features, which is an extension of <strong>kml</strong> package <span class="citation" data-cites="Genolini2010">(<a href="#ref-Genolini2010" role="doc-biblioref">Genolini and Falissard 2009</a>)</span> for a
single longitudinal feature. However, this package can only be applied
to continuous features and does not allow data to be missing. See
<span class="citation" data-cites="Lu2023">Lu et al. (<a href="#ref-Lu2023" role="doc-biblioref">2023</a>)</span> for a review and comparison of these packages.</p>
<p>The aim of this paper is to describe a new package, <strong>BCClong</strong> package,
which implements the BCC model for multiple longitudinal features
<span class="citation" data-cites="Lu2022 Tan2022a">(<a href="#ref-Lu2022" role="doc-biblioref">Lu and Lou 2022b</a>; <a href="#ref-Tan2022a" role="doc-biblioref">Tan et al. 2022</a>)</span>. Compared to existing packages, several key
features make the <strong>BCClong</strong> package appealing: (a) it allows
simultaneous clustering of mixed-type (e.g., continuous, discrete and
categorical) longitudinal features, (b) it allows each longitudinal
feature to be collected from different sources with measurements taken
at distinct sets of time points (known as irregularly sampled
longitudinal data), (c) it relaxes the assumption that all features have
the same clustering structure by estimating the feature-specific (local)
clusterings and overall (global) clustering. This model is very
flexible, which facilitates the interpretation of clustering results and
enhances its practical utilities.</p>
<h3 data-number="2" id="sec:models"><span class="header-section-number">2</span> Models and software</h3>
<p>The methodology of the BCC model for clustering multiple longitudinal
features implemented in the <strong>BCClong</strong> package is described in detail
in <span class="citation" data-cites="Lu2022">Lu and Lou (<a href="#ref-Lu2022" role="doc-biblioref">2022b</a>)</span> (for multiple continuous features) and in <span class="citation" data-cites="Tan2022a">Tan et al. (<a href="#ref-Tan2022a" role="doc-biblioref">2022</a>)</span> (for
multiple mixed-type features). In this section, we provide a brief
overview of this model and the <strong>BCClong</strong> package. Without loss of
generality, we use the term "features" to represent variables of
interest used for the cluster analysis, however, these variables can
also be patient’s traits, biomarkers, risk factors, exposures and
outcomes of interest in different contexts and studies.</p>
<h4 class="unnumbered" data-number="2.1" id="bayesian-consensus-clustering-for-multiple-longitudinal-features">Bayesian Consensus Clustering for Multiple Longitudinal Features</h4>
<p>The BCC model is originally proposed as a flexible approach to model the
dependence and heterogeneity of the data collected from multiple sources
<span class="citation" data-cites="Lock2013">(<a href="#ref-Lock2013" role="doc-biblioref">Lock and Dunson 2013</a>)</span>. Instead of arriving at a single overall clustering
structure, the BCC model allows each feature to follow feature-specific
(local) clustering and these local clusterings are aggregated to find a
consensus (global) clustering.</p>
<p>Let <span class="math inline">\(L_{i,r} = 1,...,K\)</span> denote the feature-specific (local) cluster
label for individual <span class="math inline">\(i\)</span> and feature <span class="math inline">\(r\)</span> for <span class="math inline">\(r=1,...,R\)</span>, and
<span class="math inline">\(C_i = 1,...,K\)</span> denote the consensus (global) cluster label for
individual <span class="math inline">\(i\)</span>, where <span class="math inline">\(K\)</span> denotes the number of clusters and <span class="math inline">\(R\)</span> denotes
the number of longitudinal features. Let
<span class="math inline">\(\boldsymbol{y}_{i,r} = (y_{i1,r}, ..., y_{in_{i,r},r})^\top\)</span> denote the
measurement for individual <span class="math inline">\(i\)</span> and feature <span class="math inline">\(r\)</span>, where <span class="math inline">\(y_{ij,r}\)</span> denotes
the observation <span class="math inline">\(j\)</span> of individual <span class="math inline">\(i\)</span> for feature <span class="math inline">\(r\)</span>, and <span class="math inline">\(n_{i,r}\)</span> is
the number of measurements, for <span class="math inline">\(i=1,...,N, j=1,...,n_{i,r}, r=1,...,R\)</span>.</p>
<p>The BCC model assumes that <span class="math inline">\(\boldsymbol{L}_{r} = (L_{1,r},...,L_{N,r})\)</span>
are dependent on <span class="math inline">\(\boldsymbol{C} = (C_1,...,C_N)\)</span> through
<span class="math inline">\(P(L_{i,r}=k|C_i) = \vartheta(k,C_i, \alpha_r)\)</span>, where <span class="math inline">\(\alpha_r\)</span>
adjusts the dependence function <span class="math inline">\(\vartheta(\cdot)\)</span>. The conditional
model can be specified as
<span class="math display">\[P(L_{i,r}=k|\boldsymbol{y}_{i,r},C_i,\boldsymbol{\gamma}_{k,r}, \boldsymbol{\beta}_{ik}) \propto\vartheta(k,C_i,\alpha_r)f_{k,r}(\boldsymbol{y}_{i,r}|\boldsymbol{\gamma}_{k,r}, \boldsymbol{\beta}_{ik,r})\]</span>
where <span class="math inline">\(\boldsymbol{\gamma}_{k,r}\)</span> and <span class="math inline">\(\boldsymbol{\beta}_{ik,r}\)</span> denote
the fixed effects and random effects for individual <span class="math inline">\(i\)</span> belong to
cluster <span class="math inline">\(k\)</span>, where <span class="math inline">\(k=1,...,K\)</span>. To specify <span class="math inline">\(\vartheta(k,C_i,\alpha_r)\)</span>,
the package uses the following function <span class="citation" data-cites="Lock2013">(<a href="#ref-Lock2013" role="doc-biblioref">Lock and Dunson 2013</a>)</span>.
<span class="math display">\[\vartheta(k,C_i,\alpha_r)  =
    \begin{cases}
      \alpha_r &amp; \text{if } C_i = L_{i,r}\\
      (1- \alpha_r)/(K-1) &amp; \text{otherwise}
    \end{cases}\]</span>
Therefore, <span class="math inline">\(\alpha_r\)</span> (for <span class="math inline">\(r = 1,...,R\)</span>) can also be viewed as an
adherence parameter representing the degree of agreement between
<span class="math inline">\(L_{i,r}\)</span> and <span class="math inline">\(C_i\)</span>. Moreover,
<span class="math inline">\(f_{k,r}(\boldsymbol{y}_{i,r}|\boldsymbol{\gamma}_{k,r}, \boldsymbol{\beta}_{ik,r})\)</span>
is a distribution from the exponential family with the dispersion
parameter <span class="math inline">\(\phi_{k,r}\)</span> and the fully specified mean function is given by
<span class="math display">\[h^{-1}_{k,r}(E(\boldsymbol{y}_{i,r}|\boldsymbol{\beta}_{ik,r},\boldsymbol{\gamma}_{k,r})) = \boldsymbol{\eta}_{ik,r} = \boldsymbol{x}_{i,r}^\top\boldsymbol{\gamma}_{k,r} +  \boldsymbol{Z}_{i,r}^\top\boldsymbol{\beta}_{ik,r}\]</span>
where <span class="math inline">\(h_{k,r}^{-1}\)</span> is a canonical link function for the mean of the
feature <span class="math inline">\(r\)</span> in cluster <span class="math inline">\(k\)</span> . Also,
<span class="math inline">\(\boldsymbol{\eta}_{ik,r} = (\eta_{i1k,r},...,\eta_{in_{i,r}k,r})^\top\)</span>
is the linear predictor for feature <span class="math inline">\(r\)</span>, <span class="math inline">\(\boldsymbol{x}_{i,r}\)</span> is a
<span class="math inline">\(p_r\times n_{i,r}\)</span> vector of predictors, <span class="math inline">\(\boldsymbol{\gamma}_{k,r}\)</span> is
the corresponding vector of coefficients, <span class="math inline">\(\boldsymbol{Z}_{i,r}\)</span> is a
<span class="math inline">\(q_r\times n_{i,r}\)</span> vector of predictors, and
<span class="math inline">\(\boldsymbol{\beta}_{ik,r}|\cdot \sim \text{MVN}(\boldsymbol{0},\boldsymbol{\Sigma}_{k,r})\)</span>.
In addition, the distribution of <span class="math inline">\(\boldsymbol{y}_{i,r}\)</span> can be written
as
<span class="math display">\[P(\boldsymbol{y}_{i,r}|\boldsymbol{\beta}_{ik,r},\boldsymbol{\gamma}_{k,r}, \phi_{k,r}) = \prod_{j=1}^{n_{i,r}} \text{exp}\bigg\{ \frac{y_{ij,r}\eta_{ijk,r} - g_{k,r}(\eta_{ijk,r})}{\phi_{k,r}} + w_{k,r}(y_{ij,r},\phi_{k,r}) \bigg\}\]</span>
where <span class="math inline">\(g(\eta)\)</span> and <span class="math inline">\(w(y,\phi)\)</span> are appropriate distribution-specific
functions. For example, for feature with Gaussian distribution,
<span class="math inline">\(g(\eta) = \eta^2/2\)</span> and
<span class="math inline">\(w(y,\phi) = \frac{\log(2\pi\phi)}{2} - \frac{y^2}{2\phi}\)</span>. Furthermore,
for three features with Gaussian, Poisson and Binomial distributions,
respectively. The model can be specified as:
<span class="math display">\[\begin{aligned}
  \text{(1): }  E(\boldsymbol{y}_{i,1}|\boldsymbol{\gamma}_{k,1}, \boldsymbol{\beta}_{ik,1}) &amp; = \boldsymbol{\eta}_{ik,1}  =   \boldsymbol{x}_{i,1}^\top\boldsymbol{\gamma}_{k,1} +  \boldsymbol{Z}_{i,1}^\top\boldsymbol{\beta}_{ik,1}  \\
  \text{(2): }    \text{log}(E(\boldsymbol{y}_{i,2}|\boldsymbol{\gamma}_{k,2}, \boldsymbol{\beta}_{ik,2})) &amp; = \boldsymbol{\eta}_{ik,2}  =  \boldsymbol{x}_{i,2}^\top\boldsymbol{\gamma}_{k,2} +  \boldsymbol{Z}_{i,2}^\top\boldsymbol{\beta}_{ik,2} \\
  \text{(3): }   \text{logit}(E(\boldsymbol{y}_{i,3}|\boldsymbol{\gamma}_{k,3},\boldsymbol{\beta}_{ik,3})) &amp; =  \boldsymbol{\eta}_{ik,3} = \boldsymbol{x}_{i,3}^\top\boldsymbol{\gamma}_{k,3} +  \boldsymbol{Z}_{i,3}^\top\boldsymbol{\beta}_{ik,3}
\end{aligned}\]</span>
where <span class="math inline">\(i=1,...,N\)</span>, <span class="math inline">\(j=1,...,n_{i,r}\)</span> and <span class="math inline">\(r=1,2,3\)</span>. The model also
involves dispersion parameters <span class="math inline">\(\phi_{k,r} = \sigma^2_{k,r}\)</span>, for
Gaussian distribution. The corresponding dispersion parameters for
logistic regression and Poisson regression are both 1, for <span class="math inline">\(r=1,...,R\)</span>
and <span class="math inline">\(k=1,...,K\)</span>. Given this proposed model, the complete-data likelihood
can be written as
<span class="math display">\[\mathbb{L}^{Bayes}(\boldsymbol{\Theta}| \boldsymbol{y}_i,\boldsymbol{\beta}_i; \boldsymbol{x}_{i})   =  \sum_{k_1,...,k_R}   \bigg(\sum_k \prod_{r=1}^R \vartheta(L_{i,r}=k_r,C_i=k,\alpha_r) \pi_k \bigg)  \prod_{r=1}^R f(\boldsymbol{y}_{i,r}|\boldsymbol{\gamma}_{ik_r,r},\boldsymbol{\beta}_{ik_r,r}; \boldsymbol{x}_{i,r})\]</span>
where
<span class="math inline">\(\boldsymbol{y}_i = (\boldsymbol{y}_{i,1},...,\boldsymbol{y}_{i,R})\)</span>,
<span class="math inline">\(\boldsymbol{x}_i = (\boldsymbol{x}_{i,1},...,\boldsymbol{x}_{i,R})\)</span>,
<span class="math inline">\(\boldsymbol{\beta}_i = (\boldsymbol{\beta}_{i,1},..., \boldsymbol{\beta}_{i,R})\)</span>,
<span class="math inline">\(\pi_k=P(C_i = k)\)</span> and
<span class="math inline">\(\boldsymbol{\Theta}= (\boldsymbol{\alpha}, \boldsymbol{\pi},\boldsymbol{\gamma},\boldsymbol{\Sigma},\boldsymbol{\phi})\)</span>.</p>
<h4 class="unnumbered" data-number="2.2" id="bayesian-inference">Bayesian Inference</h4>
<p>For computational convenience, the <strong>BCClong</strong> package uses independent
conjugate priors for the model parameters. Specifically, for the
adherence parameters the prior distribution is
<span class="math inline">\(\alpha_r \sim \text{TBeta}(\delta_{1,r}, \delta_{2,r}, 1/K)\)</span>, for
<span class="math inline">\(r = 1,...,R\)</span>, where <span class="math inline">\(\text{TBeta}\)</span> denotes a truncated Beta
distribution ranged from <span class="math inline">\([1/K, 1]\)</span>. For the cluster probabilities, the
prior distribution is
<span class="math inline">\(\boldsymbol{\pi}\sim \text{Dirichlet}(\boldsymbol{\varphi}_0)\)</span>. For the
fixed effect coefficients, the prior distribution is
<span class="math inline">\(\boldsymbol{\gamma}_{k,r} \sim \text{MVN}(\boldsymbol{0},\boldsymbol{V}_{0k,r})\)</span>,
where <span class="math inline">\(\boldsymbol{V}_{0k,r}\)</span> is a <span class="math inline">\(m_{k,r} \times m_{k,r}\)</span>
variance-covariance matrix, and <span class="math inline">\(m_{k,r}\)</span> is the dimension of
<span class="math inline">\(\boldsymbol{\gamma}_{k,r}\)</span>. The dispersion parameter <span class="math inline">\(\phi_{k,r}\)</span> is
not constant only when the feature is a Gaussian distribution. In such a
case, <span class="math inline">\(\phi_{k,r} = \sigma^2_{k,r}\)</span>. The prior distribution is
<span class="math inline">\(\sigma^2_{k,r} \sim \text{IG}(a_{0k,r},b_{0k,r})\)</span>, where <span class="math inline">\(a_{0k,r}\)</span> and
<span class="math inline">\(b_{0k,r}\)</span> are the parameters of an inverse gamma distribution. By
definition, <span class="math inline">\(\phi_{k,r} = 1\)</span> for features follow Poisson or Binomial
distribution, for <span class="math inline">\(k=1,...,K\)</span> and <span class="math inline">\(r=1,...,R\)</span>. For the
variance-covariance matrix of the random effect, the prior distribution
is
<span class="math inline">\(\Sigma_{k,r}^{-1} \sim \text{Wishart}(\lambda_{0k,r}, (\lambda_{0k,r}\boldsymbol{\Lambda}_{0k,r})^{-1} )\)</span>,
where the prior for the Wishart distribution is parametrized such that
the mean is <span class="math inline">\(\boldsymbol{\Lambda}_{0k,r}^{-1}\)</span>. In special case when
<span class="math inline">\(\Sigma_{k,r}\)</span> is a diagonal matrix, that is,
<span class="math inline">\(\Sigma_{k,r} = \text{diag}(\xi^2_{k1,r},...,\xi^2_{kq_r,r})\)</span>, the prior
is an inverse gamma distribution, i.e.,
<span class="math inline">\(\xi^2_{km,r} \sim \text{IG}(c_{0k,r},d_{0k,r})\)</span>, for <span class="math inline">\(m = 1,..., q_r\)</span>.</p>
<p>For posterior computation, a Gibbs sampling approach to update the model
parameters is described as follows. After initializing the parameters,
the algorithm repeats the following steps until convergence. At the <span class="math inline">\(s\)</span>
step of the iteration,</p>
<ul>
<li><p>Update local cluster membership <span class="math inline">\(L_{i,r}^{(s)}\)</span> given
<span class="math inline">\(\{\boldsymbol{y}_{i,r}, \boldsymbol{\Theta}_r^{(s-1)}, \alpha_r^{(s-1)}, C_i^{(s-1)}, \boldsymbol{\beta}_r^{(s-1)} \}\)</span>,
for <span class="math inline">\(i=1,...,N\)</span> and <span class="math inline">\(r=1,...,R\)</span>.</p></li>
<li><p>Update <span class="math inline">\(\alpha_r^{(s)}\)</span> given
<span class="math inline">\(\{ \boldsymbol{C}^{(s-1)}, \boldsymbol{L}^{(s)} \}\)</span>, for
<span class="math inline">\(r=1,...,R\)</span>.</p></li>
<li><p>Update <span class="math inline">\(C_i^{(s)}\)</span> given
<span class="math inline">\(\{ \boldsymbol{L}^{(s)},  \boldsymbol{\alpha}^{(s)}, \boldsymbol{\pi}^{(s-1)}\}\)</span>,
for <span class="math inline">\(i=1,...,N\)</span>.</p></li>
<li><p>Update <span class="math inline">\(\boldsymbol{\pi}^{(s)}\)</span> given <span class="math inline">\(\boldsymbol{C}^{(s)}\)</span>.</p></li>
<li><p>Update cluster-specific parameters
<span class="math inline">\(\boldsymbol{\Theta}^{(s)} = (\boldsymbol{\gamma}^{(s)},\boldsymbol{\Sigma}^{(s)}, \boldsymbol{\sigma}^{2(s)})\)</span>
and <span class="math inline">\(\boldsymbol{\beta}^{(s)}\)</span>. In particular,
<span class="math inline">\(\boldsymbol{\gamma}^{(s)}\)</span> and <span class="math inline">\(\boldsymbol{\beta}^{(s)}\)</span> are
updated via Metropolis-Hastings algorithm.</p></li>
</ul>
<p>Each MCMC iteration will produce a realization of cluster membership for
<span class="math inline">\(\boldsymbol{C}\)</span>, <span class="math inline">\(\boldsymbol{L}_1\)</span>,..., <span class="math inline">\(\boldsymbol{L}_R\)</span>. The
package uses the mode over all the MCMC samples (after burn-in and
thinning) as the point estimates for both feature-specific and global
clusterings.</p>
<p>Label switching is a common phenomenon in mixture models. This problem
arises due to both the likelihood and posterior distributions being
invariant to permutations of the parameters. The <strong>BCClong</strong> package
applies a post-processing algorithm, via the <strong>label.switching</strong> package
<span class="citation" data-cites="Papastamoulis2016">(<a href="#ref-Papastamoulis2016" role="doc-biblioref">Papastamoulis 2016</a>)</span>, to reorder the labels based on Kullback-Leibler
divergence <span class="citation" data-cites="Stephens2000">(<a href="#ref-Stephens2000" role="doc-biblioref">Stephens 2000</a>)</span>.</p>
<p>The computation of the BCC model can be carried out using the main
function in the <strong>BCClong</strong> package:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">BCC.multi</span>(mydat, dist, id, time, formula, num.cluster,  </span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>         hyper.par, initials, initial.cluster.membership, </span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>         input.initial.local.cluster.membership, </span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>         input.initial.global.cluster.membership,</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>         burn.in, thin, per, max.iter, ...)</span></code></pre></div>
<p>The meaning of the key arguments of the <strong>BCC.multi()</strong> function is
described as follows.</p>
<ul>
<li><p><code>mydat</code>: list of <span class="math inline">\(R\)</span> longitudinal features (i.e., with a length of
<span class="math inline">\(R\)</span>), where <span class="math inline">\(R\)</span> is the number of features. The data should be
prepared in a long-format (each row is one time point per
individual).</p></li>
<li><p><code>dist</code>: a character vector (with a length of <span class="math inline">\(R\)</span>) that determines
the distribution for each feature. Possible values are "gaussian"
for a continuous feature, "poisson" for a discrete feature (e.g.,
count data) using a log link and "binomial" for a dichotomous
feature (0/1) using a logit link. Single value (i.e., a length of 1)
is recycled if necessary.</p></li>
<li><p><code>id</code>: a list (with a length of <span class="math inline">\(R\)</span>) of vectors of the study id of
individuals for each feature. Single value (i.e., a length of 1) is
recycled if necessary.</p></li>
<li><p><code>time</code>: a list (with a length of <span class="math inline">\(R\)</span>) of vectors of time (or age) at
which the feature measurements are recorded.</p></li>
<li><p><code>formula</code>: a list (with a length of <span class="math inline">\(R\)</span>) of formula for each
feature. Each formula is a two-sided linear formula object
describing both the fixed-effects and random effects part of the
model, with the response (i.e., longitudinal feature) on the left of
a <code> </code> operator and the terms, separated by <code>+</code> operations, or the
right. Random-effects terms are distinguished by vertical bars <code>(|)</code>
separating expressions for design matrices from grouping factors.
See <code>formula</code> argument from the <strong>lme4</strong> package.</p></li>
<li><p><code>num.cluster</code>: number of clusters <span class="math inline">\(K\)</span>. A minimum value of <span class="math inline">\(K\)</span> is 2.</p></li>
<li><p><code>hyper.par</code>: hyper-parameters of the prior distributions for the
model parameters. The default hyper-parameters values will result in
weakly informative prior distributions.</p></li>
<li><p><code>initials</code>: a list of initial values for model parameters.</p></li>
<li><p><code>initial.cluster.membership</code>: a character, which can be "random",
"mixAK" or "input". For "random" (default), the local cluster
membership is randomly generated from 1 to <span class="math inline">\(K\)</span> for each feature. For
"mixAK", the local cluster membership is obtained by fitting the
Bayesian mixture model for longitudinal data to each feature through
the mixAK package for the given <span class="math inline">\(K\)</span>. For both methods, the global
cluster membership is set to be equal to the local cluster
membership of the first feature. For "input", the user can input
initial values for local and global cluster memberships, using the
argument <code>input.initial.local.cluster.membership</code> and
<code>input.initial.global.cluster.membership</code>.</p></li>
<li><p><code>input.initial.local.cluster.membership</code>: if
<code>initial.cluster.membership = "input"</code>, then this augment must not
be empty. It is a list with a length of <span class="math inline">\(R\)</span> and each element of the
list corresponds to the local clustering of each feature. See the
Illustration section for an example.</p></li>
<li><p><code>input.initial.global.cluster.membership</code>: if
<code>initial.cluster.membership = "input"</code>, by default the
<code>input.initial.global.cluster.membership</code> is the local cluster
membership of the first feature, and therefore this argument needs
not to be supplied, but the user can also use this argument to
supply a different initial cluster membership for the global
clustering. See the Illustration section for an example.</p></li>
<li><p><code>burn.in</code>: the number of samples disgarded. This value must be
smaller than <code>max.iter</code>.</p></li>
<li><p><code>thin</code>: the number of thinning. For example, if thin = 10, then the
MCMC chain will keep one sample every 10 iterations.</p></li>
<li><p><code>per</code>: specify how often the MCMC chain will print the iteration
number.</p></li>
<li><p><code>max.iter</code>: the number of MCMC iterations.</p></li>
</ul>
<p>In the <strong>BCClong</strong> package, the default hyper-parameter values for the
prior distributions are set to reflect no prior information regarding
the values of these parameters, that is, uninformative priors are used.
The argument <code>hyper.par</code> allows setting these hyper-parameter values.
The default values are:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>hyper.par <span class="ot">=</span> <span class="fu">list</span>(<span class="at">a.star =</span> <span class="dv">1</span>, <span class="at">b.star =</span> <span class="dv">1</span>, <span class="at">delta =</span> <span class="dv">1</span>, </span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">aa0 =</span> <span class="fl">0.001</span>, <span class="at">bb0 =</span> <span class="fl">0.001</span>, <span class="at">cc0 =</span> <span class="fl">0.001</span>, <span class="at">dd0 =</span> <span class="fl">0.001</span>,  </span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">vv0 =</span> <span class="dv">1000</span>)</span></code></pre></div>
<p>This indicates that for <span class="math inline">\(\delta_{1,r} = \delta_{2,r} = 1\)</span> for
<span class="math inline">\(r = 1,...,R\)</span> (corresponding to <code>a.star = 1, b.star = 1</code>). The elements
of <span class="math inline">\(\boldsymbol{\varphi}\)</span> are set to 1 (corresponding to <code>delta = 1</code>),
and <span class="math inline">\(a_{0k,r} = b_{0k,r} = 0.001\)</span> for <span class="math inline">\(k = 1,...,K\)</span> and <span class="math inline">\(r = 1,...,R\)</span>
(corresponding to <code>aa0 = 0.001, bb0 = 0.001</code>). Also,
<span class="math inline">\(\boldsymbol{V}_{0k,r}\)</span> is set to be a diagonal matrix with elements of
1000 (corresponding to <code>vv0 = 1000</code>). Finally, <span class="math inline">\(\Sigma_{k,r}\)</span> is set to
be a diagonal matrix with elements following an inverse gamma
distribution (corresponding to <code>cc0 = 0.001, dd0 = 0.001</code>). Customized
or informative priors can be used by modifying the values in
<code>hyper.par</code>.</p>
<p>By default, the <code>initials</code> is set to <code>NULL</code> and the <strong>BCClong</strong> will
automately generate initial values for both the local and global cluster
memberships and the model parameters.</p>
<p>Convergence of MCMC is diagnosed using visual inspection of trace plots
as well as using the Geweke statistics <span class="citation" data-cites="Geweke1991">(<a href="#ref-Geweke1991" role="doc-biblioref">Geweke 1991</a>)</span>. The <strong>BCClong</strong>
package provides a function called <strong>traceplot()</strong> to produce trace
plots for model parameters of interest. The <strong>traceplot()</strong> function has
the following arguments:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">traceplot</span>(fit, cluster.indx, feature.indx, parameter)</span></code></pre></div>
<p>The meaning of these arguments is described as follows.</p>
<ul>
<li><p><code>fit</code>: an objective output from <code>BCC.multi</code>() function.</p></li>
<li><p><code>cluster.indx</code>: a numeric value. For cluster-specific parameters,
specifying <code>cluster.indx</code> will generate the trace plot for the
corresponding cluster.</p></li>
<li><p><code>feature.indx</code>: a numeric value. For cluster-specific parameters,
specifying <code>feature.indx</code> will generate the trace plot for the
corresponding cluster.</p></li>
<li><p><code>parameter</code>: a character value. Specify which parameter for which
the trace plot will be generated. The value can be "PPI" for
<span class="math inline">\(\boldsymbol{\pi}\)</span>, <span class="math inline">\(\boldsymbol{\alpha}\)</span> for <span class="math inline">\(\boldsymbol{\alpha}\)</span>,
"GA" for <span class="math inline">\(\boldsymbol{\gamma}\)</span>, "SIGMA.SQ.U" for
<span class="math inline">\(\boldsymbol{\Sigma}\)</span> and "SIGMA.SQ.E" for <span class="math inline">\(\boldsymbol{\sigma}\)</span>.</p></li>
</ul>
<p>In addition, the output of summary statistics from the <strong>BCC.multi()</strong>
function also includes the Geweke statistics calculated using the
<strong>coda</strong> package <span class="citation" data-cites="Martyn2006">(<a href="#ref-Martyn2006" role="doc-biblioref">Plummer et al. 2006</a>)</span>. Briefly, the Geweke statistics determine
whether a Markov chain converges or not based on a test for equality of
the means of the first and last part of a Markov chain (by default the
first 10% and the last 50%). If the samples are drawn from the
stationary distribution of the chain, the two means are equal and
Geweke’s statistic has an asymptotically standard normal distribution.
Practically, if the estimated Geweke statistics fall between -1.96 and
1.96, we consider the model parameter is converged.</p>
<h4 class="unnumbered" data-number="2.3" id="missing-data-measurement-time-and-frequency">Missing Data, Measurement Time and Frequency</h4>
<p>The proposed BCC model for longitudinal data and the <strong>BCClong</strong> package
allow the modeling of missing data under the assumption that the data
are missing at random <span class="citation" data-cites="Lu2022 Tan2022a">(<a href="#ref-Lu2022" role="doc-biblioref">Lu and Lou 2022b</a>; <a href="#ref-Tan2022a" role="doc-biblioref">Tan et al. 2022</a>)</span>. No extra steps are required
from the users to handle the missing data. The package will keep
individuals with at least one observation for any input features and use
them for the analysis, whereas the package will remove individuals with
no data on all features. The package also allows the number of
observations to be different between individuals and features, and these
observations can also be measured at time points that are different
between individuals and features.</p>
<h4 class="unnumbered" data-number="2.4" id="determining-the-number-of-clusters">Determining the Number of Clusters</h4>
<p>Determining the number of clusters in model-based clustering is
challenging and there are no widely accepted approaches. Review of
different approaches within the Bayesian framework can be found in
<span class="citation" data-cites="Celeux2006 Nasserinejad2017 Merkle2019">Celeux et al. (<a href="#ref-Celeux2006" role="doc-biblioref">2006</a>; <a href="#ref-Nasserinejad2017" role="doc-biblioref">Nasserinejad et al. 2017</a>; <a href="#ref-Merkle2019" role="doc-biblioref">Merkle et al. 2019</a>)</span>. The <strong>BCClong</strong> package
offers three criteria to determine the number of clusters, namely the
mean adjusted adherence <span class="citation" data-cites="Lock2013">(<a href="#ref-Lock2013" role="doc-biblioref">Lock and Dunson 2013</a>)</span>, the Deviance Information Criterion
(DIC) <span class="citation" data-cites="Spiegelhalter2002 Celeux2006">(<a href="#ref-Spiegelhalter2002" role="doc-biblioref">Spiegelhalter et al. 2002</a>; <a href="#ref-Celeux2006" role="doc-biblioref">Celeux et al. 2006</a>)</span> and the widely Applicable
Information Criterion (WAIC) <span class="citation" data-cites="Watanabe2010 Gelman2014a">(<a href="#ref-Watanabe2010" role="doc-biblioref">Watanabe 2010</a>; <a href="#ref-Gelman2014a" role="doc-biblioref">Gelman et al. 2014</a>)</span>. We briefly
describe these three approaches in this subsection.</p>
<p>For consensus clustering, <span class="citation" data-cites="Lock2013">Lock and Dunson (<a href="#ref-Lock2013" role="doc-biblioref">2013</a>)</span> proposed an empirical separability
criterion, which selects the value of <span class="math inline">\(K\)</span> that gives maximum adherence
to an overall clustering. For each <span class="math inline">\(K \ge 2\)</span>, the estimated adherence
parameters <span class="math inline">\(\alpha_r \in [\frac{1}{K},1]\)</span> are mapped to the unit
interval by the linear transformation
<span class="math inline">\(\alpha^{*}_r = \frac{K\alpha_r - 1}{K-1}\)</span>, thus
<span class="math inline">\(\alpha^{*}_r \in [0,1]\)</span>. One then selects the value of <span class="math inline">\(K\)</span> that results
in the highest mean adjusted adherence, which is defined as
<span class="math inline">\(\bar{\alpha}^{*} = \frac{1}{R}\sum_{m=1}^R \alpha^{*}_r\)</span>. This approach
selects a model with which the feature-specific clusterings contribute
the most information (on average) to the global clustering. A model with
a higher <span class="math inline">\(\bar{\alpha}^{*}\)</span> is preferred.</p>
<p>DIC and its variants have also been widely used for clustering
multivariate longitudinal data, see references
<span class="citation" data-cites="Lu2019 Neelon2011 Leiby2009 Elliott2005 Fruehwirth-Schnatter2010">(<a href="#ref-Elliott2005" role="doc-biblioref">Elliott et al. 2005</a>; <a href="#ref-Leiby2009" role="doc-biblioref">Leiby et al. 2009</a>; <a href="#ref-Fruehwirth-Schnatter2010" role="doc-biblioref">Frühwirth-Schnatter and Pyne 2010</a>; <a href="#ref-Neelon2011" role="doc-biblioref">Neelon et al. 2011</a>; <a href="#ref-Lu2019" role="doc-biblioref">Lu and Lou 2019</a>)</span>
for example. In the <strong>BCClong</strong> package, the DIC is computed based on
the complete likelihood (also known as <span class="math inline">\(\text{DIC}_4\)</span> in <span class="citation" data-cites="Celeux2006">Celeux et al. (<a href="#ref-Celeux2006" role="doc-biblioref">2006</a>)</span>),
which is defined as
<span class="math inline">\(\text{DIC} = -4E_{\boldsymbol{\Theta},\boldsymbol{C},\boldsymbol{L}} [\log f(\boldsymbol{y},\boldsymbol{C},\boldsymbol{L}|\boldsymbol{\Theta})|\boldsymbol{y}]+ 2E_{\boldsymbol{C},\boldsymbol{L}}[\log f(\boldsymbol{y},\boldsymbol{C},\boldsymbol{L}|E[\boldsymbol{\Theta}|\boldsymbol{y},\boldsymbol{C},\boldsymbol{L}])|\boldsymbol{y})]\)</span>,
where <span class="math inline">\(\boldsymbol{y} = (\boldsymbol{y}_1, ..., \boldsymbol{y}_N)\)</span> and
<span class="math inline">\(\boldsymbol{\Theta}\)</span> is defined in equation (5). The effective number
of parameters and the posterior mean of deviance are defined as
<span class="math inline">\(p_D = - E_{\boldsymbol{\Theta},\boldsymbol{C},\boldsymbol{L}}[2\log f(\boldsymbol{y},\boldsymbol{C},\boldsymbol{L}|\boldsymbol{\Theta})|\boldsymbol{y}] +  E_{\boldsymbol{C},\boldsymbol{L}}[2\log f(\boldsymbol{y},\boldsymbol{C},\boldsymbol{L}|E[\boldsymbol{\Theta}|\boldsymbol{y},\boldsymbol{C},\boldsymbol{L}])|\boldsymbol{y}]\)</span>
and
<span class="math inline">\(\overline{D(\boldsymbol{\Theta})} = -E_{\boldsymbol{\Theta},\boldsymbol{C},\boldsymbol{L}}[2\log f(\boldsymbol{y},\boldsymbol{C},\boldsymbol{L}|\boldsymbol{\Theta})|\boldsymbol{y}]\)</span>,
respectively. A model with a lower DIC is preferred.</p>
<p>WAIC is an extension of the Akaike Information Criterion (AIC) that is
more fully Bayesian than the DIC. Similar to DIC, WAIC estimates the
effective number of parameters to adjust for overfitting. The
<strong>BCClong</strong> package uses the <strong>LaplacesDemon</strong> package <span class="citation" data-cites="Hall2022">(<a href="#ref-Hall2022" role="doc-biblioref">Hall 2021</a>)</span> to
calculate the WAIC, which is defined as
<span class="math inline">\(\text{WAIC} = -2(lppd - p_{\text{WAIC}})\)</span>, where lppd is the log
pointwise predictive density which is equivalent to
<span class="math inline">\(lppd=\text{log}\prod_{i=1}^Np_{\text{post}}(\boldsymbol{y},\boldsymbol{C},\boldsymbol{L})\)</span>.
Also,
<span class="math inline">\(p_{\text{WAIC}} = \sum_{i=1}^N \text{var}_{\text{post}} (\text{log}p(\boldsymbol{y},\boldsymbol{C},\boldsymbol{L}|\boldsymbol{\Theta}))\)</span>
is an approximation to the number of unconstrained and uninformed
parameters, where a parameter counts as 1 when estimated without
constraint or any prior information, 0 if fully constrained or all
information comes from the prior distribution, or an intermediate number
if both the data and prior are informative. The <span class="math inline">\(p_{\text{WAIC}}\)</span>
defined here is corresponding to <span class="math inline">\(p_{\text{WAIC2}}\)</span> in <span class="citation" data-cites="Gelman2014a">Gelman et al. (<a href="#ref-Gelman2014a" role="doc-biblioref">2014</a>)</span>,
which is recommended because its results are closer in practice to the
results of leave-one-out cross-validation. A model with a lower WAIC is
preferred.</p>
<p>In the <strong>BCClong</strong> package, the <strong>model.selection.criteria()</strong> function
is used to compute the DIC and WAIC:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">model.selection.criteria</span>(fit, fast_version)</span></code></pre></div>
<ul>
<li><p><code>fit</code>: an objective output from <code>BCC.multi</code>() function.</p></li>
<li><p><code>fast_version</code>: if <code>fast_verion = TRUE</code> (default), then compute the
DIC and WAIC using the first 100 MCMC samples (after burn-in and
thinning) . If <code>fast_version = FALSE</code>, then compute the DIC and WAIC
using all MCMC samples (after burn-in and thinning) .</p></li>
</ul>
<h4 class="unnumbered" data-number="2.5" id="goodness-of-fit">Goodness of Fit</h4>
<p>To assess the model goodness of fit, one can compute the posterior
predictive check <span class="citation" data-cites="Gelman1996">(<a href="#ref-Gelman1996" role="doc-biblioref">Gelman et al. 1996</a>)</span>. This approach is performed by comparing
the observed data with data replicated from the posterior predictive
distribution. If the model fits the data well, the replicated data,
denoted as <span class="math inline">\(\boldsymbol{y}^{rep}\)</span> should have similar distribution as
the observed data <span class="math inline">\(\boldsymbol{y}\)</span>. Let
<span class="math inline">\(T = T(\boldsymbol{y},\boldsymbol{\Phi})\)</span> denote a discrepancy measure,
where
<span class="math inline">\(\boldsymbol{\Phi} = (\boldsymbol{\Theta}, \boldsymbol{\beta}, \boldsymbol{L}, \boldsymbol{C})\)</span>
denotes all model parameters (including random effects and cluster
labels) and <span class="math inline">\(T\)</span> is sample quantiles or residual-based measures.
Following <span class="citation" data-cites="Gelman1996">Gelman et al. (<a href="#ref-Gelman1996" role="doc-biblioref">1996</a>)</span>, a natural discrepancy measure is a <span class="math inline">\(\chi^2\)</span>
measure. For BCC model it is defined as
<span class="math display">\[T^{obs}(\boldsymbol{y},\boldsymbol{\Phi})= \sum^K_{k=1} \sum_{i=1}^N \sum_{j=1}^{n_{i,r}}   \sum_{r=1}^R \frac{ z_{ik,r} ||\boldsymbol{y}_{i,r} - h^{-1}_{k,r}(E(\boldsymbol{y}_{i,r}|\boldsymbol{\beta}_{ik,r},\boldsymbol{\gamma}_{k,r}))||^2}{\sigma_{k,r}^2}\]</span>
where <span class="math inline">\(z_{ik,r} = 1\)</span> if <span class="math inline">\(L_{i,r}=k\)</span>, and 0 otherwise. This measure is
computed at each MCMC step by treating cluster indicator <span class="math inline">\(z_{ik,r}\)</span> and
random effects <span class="math inline">\(\boldsymbol{\beta}_{ik,r}\)</span> as observed data.</p>
<p>The Bayesian predictive p-value is the probability that the discrepancy
measure based on a predictive sample,
<span class="math inline">\(T^{rep}(\boldsymbol{y}^{rep},\boldsymbol{\Phi})\)</span> is more extreme than
the observed measure <span class="math inline">\(T^{obs}(\boldsymbol{y},\boldsymbol{\Phi})\)</span>. This
quantity is estimated by computing the proportion of draws in which
<span class="math inline">\(T^{rep} &gt; T^{obs}\)</span>. A p-value close to 0.5 indicates the model provides
a good fit to the data, whereas a p-value close to 0 or 1 indicates a
poor fit.</p>
<p>In the <strong>BCClong</strong> package, <strong>BayesT()</strong> is used for computing the
posterior predictive check.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">BayesT</span>(fit)</span></code></pre></div>
<ul>
<li><code>fit</code>: an objective output from <code>BCC.multi</code>() function.</li>
</ul>
<h4 class="unnumbered" data-number="2.6" id="visualization-of-trajectory-patterns-by-clusters">Visualization of Trajectory Patterns by Clusters</h4>
<p>To plot the longitudinal trajectory of features by local and global
clusterings derived from the BCC model, <strong>trajplot()</strong> function can be
used. This function uses the <strong>ggplot2</strong> package internally.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">trajplot</span>(fit,feature.ind,which.cluster)</span></code></pre></div>
<ul>
<li><p><code>fit</code>: an objective output from <strong>BCC.multi()</strong> function.</p></li>
<li><p><code>feature.indx</code>: a numeric value indicating which feature to plot.
The number indicates the order of the feature specified in <code>mydat</code>
argument of the <strong>BCC.multi()</strong> function.</p></li>
<li><p><code>which.cluster</code>: a character value: "global" or "local",
indicating whether to plot the trajectory by global cluster or local
cluster indices.</p></li>
</ul>
<h3 data-number="3" id="sec:illustrations"><span class="header-section-number">3</span> Illustrations</h3>
<p>In this section, we demonstrate the utility of the <strong>BCClong</strong> package
(version 1.0.3) using two real data examples. The first example was
quality of life data following epilepsy drug treatment, which included
three continuous features. This dataset can be found in <strong>joineRML</strong>
package <span class="citation" data-cites="Hickey2018">(<a href="#ref-Hickey2018" role="doc-biblioref">Hickey et al. 2018</a>)</span>. The second example was the mayo clinic primary
biliary cholangitis (PBC) data which included three features of
mixed-type (i.e., continuous, discrete and categorical). This dataset
can be found in <strong>mixAK</strong> package <span class="citation" data-cites="Komarek2014">(<a href="#ref-Komarek2014" role="doc-biblioref">Komárek and Komárková 2014</a>)</span>.</p>
<h4 class="unnumbered" data-number="3.1" id="example-1-quality-of-life-data-following-epilepsy-drug-treatment">Example 1: Quality of Life Data Following Epilepsy Drug Treatment</h4>
<p>The data called <strong>epileptic.qol</strong> came from the Standard and New
Antiepileptic Drugs (SAND) study <span class="citation" data-cites="Marson2007">(<a href="#ref-Marson2007" role="doc-biblioref">Marson et al. 2007</a>)</span>. This is a randomized
control trial of standard and new antiepileptic drugs, comparing effects
on long-term clinical outcomes. Quality of life (QoL) data were
collected by mail at baseline, 3 months, and at 1 and 2 years using
validated measures. The first measurement for each individual was the
baseline measurement, however, there was variability in the time taken
to return the questionnaires. Similarly, the second, third, and fourth
follow-up times, which were scheduled for 3 months, 1-year, and 2 years,
respectively, also had variability in completion times.</p>
<p>The data can be loaded by</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">library</span>(<span class="st">&quot;joineRML&quot;</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">head</span>(epileptic.qol[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">5</span>,<span class="dv">6</span>,<span class="dv">7</span>,<span class="dv">8</span>)])</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>  id time anxiety depress aep</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>  <span class="dv">1</span>  <span class="dv">147</span>      <span class="dv">11</span>      <span class="dv">14</span>  <span class="dv">43</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span>  <span class="dv">1</span>  <span class="dv">259</span>      <span class="dv">12</span>      <span class="dv">12</span>  <span class="dv">51</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>  <span class="dv">1</span>  <span class="dv">519</span>      <span class="dv">20</span>      <span class="dv">21</span>  <span class="dv">63</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="dv">4</span>  <span class="dv">1</span>  <span class="dv">906</span>      <span class="dv">17</span>      <span class="dv">20</span>  <span class="dv">53</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span>  <span class="dv">2</span>  <span class="dv">134</span>      <span class="dv">19</span>      <span class="dv">13</span>  <span class="dv">45</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="dv">6</span>  <span class="dv">2</span>  <span class="dv">258</span>      <span class="dv">21</span>      <span class="dv">16</span>  <span class="dv">50</span></span></code></pre></div>
<p>This is a longitudinal dataset with one row per visit. At each visit,
measurements of three continuous features <span class="math inline">\((R = 3)\)</span> are recorded, namely
the anxiety score (anxiety), depress score (depress) and Liverpool
Adverse Events Profile (aep) defined according to the NEWQOL (Newly
Diagnosed Epilepsy Quality of Life) assessment. One could use a
spaghetti plot to visualize the trajectories of these features (Figure
<a href="#fig:figtraj-demo">1</a>). The goal of the analysis was to identify
distinct longitudinal QoL patterns based on these three features,
simultaneously. Based on the observed trajectories, we used a linear
form with a random intercept for each feature with a Gaussian
distribution. The model is written as:
<span class="math display">\[\begin{aligned}
  \text{anxiety: }  E(\boldsymbol{y}_{i,1}|\boldsymbol{\gamma}_{k,1}, \boldsymbol{\beta}_{ik,1}) &amp; = \boldsymbol{\eta}_{ik,1}  =   \boldsymbol{x}_{i,1}^\top\boldsymbol{\gamma}_{k,1} +  \boldsymbol{Z}_{i,1}^\top\boldsymbol{\beta}_{ik,1}  \\
  \text{depress: }    E(\boldsymbol{y}_{i,2}|\boldsymbol{\gamma}_{k,2}, \boldsymbol{\beta}_{ik,2}) &amp; = \boldsymbol{\eta}_{ik,2}  =  \boldsymbol{x}_{i,2}^\top\boldsymbol{\gamma}_{k,2} +  \boldsymbol{Z}_{i,2}^\top\boldsymbol{\beta}_{ik,2} \\
  \text{aep: }   E(\boldsymbol{y}_{i,3}|\boldsymbol{\gamma}_{k,3},\boldsymbol{\beta}_{ik,3}) &amp; =  \boldsymbol{\eta}_{ik,3} = \boldsymbol{x}_{i,3}^\top\boldsymbol{\gamma}_{k,3} +  \boldsymbol{Z}_{i,3}^\top\boldsymbol{\beta}_{ik,3}
\end{aligned}\]</span>
where <span class="math inline">\(\boldsymbol{x}_{i,r} = (1,  \boldsymbol{t}_{i,r})^\top\)</span> for
<span class="math inline">\(r=1,2,3\)</span> and
<span class="math inline">\(\boldsymbol{t}_{i,r} = (t_{i1,r},...,t_{in_{i,r},r})^\top\)</span> is a vector
of time values at which the feature <span class="math inline">\(r\)</span> is recorded for individual <span class="math inline">\(i\)</span>.
The corresponding coefficients are
<span class="math inline">\(\boldsymbol{\gamma}_{k,r} =  (\gamma_{k,r1}, \gamma_{k,r2})^\top\)</span>.
Also, <span class="math inline">\(\boldsymbol{Z}_{i,r} =\boldsymbol{1}_{n_{i,r}}\)</span> for <span class="math inline">\(r=1,2,3\)</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figtraj-demo"></span>
<img src="Figures/traj.jpeg" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 1:  Longitudinal trajectories of three features for the epileptic.qol dataset. A locally weighted scatterplot smoothing curve is overlaid on each panel to provide an estimate of the overall trend. (A) longitudinal trajectories of Anxiety score (anxiety), (B) longitudinal trajectories of depression score (depress), (C) longitudinal trajectories of Liverpool Adverse Events Profile (aep).
</p>
</div>
</div>
<p>For model-based clustering based on a finite mixture model, it is
necessary to normalize the data within each feature in order to model
the growth pattern not the level. This is because the level could
dominate the sample variability so that the clusters will be mainly
determined by the level and will not necessarily be homogeneous in trend
<span class="citation" data-cites="Heggeseth2018">(<a href="#ref-Heggeseth2018" role="doc-biblioref">Heggeseth and Jewell 2018</a>)</span>. Therefore, in the current analysis, the three features
were normalized using the <strong>scale</strong> function before entering the model,
using the following commands,</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> epileptic.qol<span class="sc">$</span>anxiety_scale <span class="ot">&lt;-</span> <span class="fu">scale</span>(epileptic.qol<span class="sc">$</span>anxiety)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> epileptic.qol<span class="sc">$</span>depress_scale <span class="ot">&lt;-</span> <span class="fu">scale</span>(epileptic.qol<span class="sc">$</span>depress)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> epileptic.qol<span class="sc">$</span>aep_scale <span class="ot">&lt;-</span> <span class="fu">scale</span>(epileptic.qol<span class="sc">$</span>aep)</span></code></pre></div>
<h5 class="unnumbered" data-number="3.1.1" id="determing-the-number-of-clusters">Determing the Number of Clusters</h5>
<p>The following codes fit the BCC model with the number of clusters
ranging from 2 to 5, i.e, <span class="math inline">\(K = 2, ..., 5\)</span>. For each value of <span class="math inline">\(K\)</span>, we
saved the mean adjusted adherence (alpha.adjust). For each model, we
generated 2000 samples and discarded the first 1000 samples. Note that
the following part of the codes may be computationally intensive, if
users just want to get familiar with the codes and outputs, a smaller
number of iterations and burn-in can be used to reduce the computational
time, for example, by setting
<code>burn.in = 10, per = 1, and max.iter = 20</code>.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> dat <span class="ot">&lt;-</span> epileptic.qol</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> alpha.adjust <span class="ot">&lt;-</span>  <span class="cn">NULL</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="cf">for</span> (k <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>){ </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        fit.BCC <span class="ot">&lt;-</span> <span class="fu">BCC.multi</span> (</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">mydat =</span> <span class="fu">list</span>(dat<span class="sc">$</span>anxiety_scale,dat<span class="sc">$</span>depress_scale,dat<span class="sc">$</span>aep_scale),</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">dist =</span> <span class="fu">c</span>(<span class="st">&quot;gaussian&quot;</span>),</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">id =</span> <span class="fu">list</span>(dat<span class="sc">$</span>id),</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">time =</span> <span class="fu">list</span>(dat<span class="sc">$</span>time),</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">formula =</span> <span class="fu">list</span>(y <span class="sc">~</span> time <span class="sc">+</span>  (<span class="dv">1</span><span class="sc">|</span>id)),</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">num.cluster =</span> k,</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">burn.in =</span> <span class="dv">1000</span>,   </span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">thin =</span> <span class="dv">1</span>,     </span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">per =</span> <span class="dv">100</span>,       </span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">max.iter =</span> <span class="dv">2000</span>)  </span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        alpha.adjust <span class="ot">&lt;-</span> <span class="fu">c</span>(alpha.adjust, fit.BCC<span class="sc">$</span>alpha.adjust) </span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        }</span></code></pre></div>
<p>An alternative specification for the <code>formula</code> argument is</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>    formula <span class="ot">=</span> <span class="fu">list</span>(y <span class="sc">~</span> time <span class="sc">+</span>  (<span class="dv">1</span><span class="sc">|</span>id), y <span class="sc">~</span> time <span class="sc">+</span>  (<span class="dv">1</span><span class="sc">|</span>id), y <span class="sc">~</span> time <span class="sc">+</span>  (<span class="dv">1</span><span class="sc">|</span>id))</span></code></pre></div>
<p>This specification can be used if the users would like to use different
functional forms for different features. Each element within the list
corresponds to a model for a feature listed in <code>mydat</code> argument.</p>
<p>To compare the model under different numbers of clusters, one can plot
the mean adjusted adherence index using the following commands:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> num.cluster <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">5</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot</span>(num.cluster, alpha.adjust, <span class="at">type =</span> <span class="st">&quot;o&quot;</span>,</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">cex.lab =</span> <span class="fl">1.5</span>, <span class="at">cex.axis =</span> <span class="fl">1.5</span>, <span class="at">cex.main =</span> <span class="fl">1.5</span>, <span class="at">lwd =</span> <span class="dv">2</span>,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">xlab =</span> <span class="st">&quot;Number of Clusters&quot;</span>,</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">ylab =</span> <span class="st">&quot;mean adjusted adherence&quot;</span>, <span class="at">main=</span><span class="st">&quot;mean adjusted adherence&quot;</span>)</span></code></pre></div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:fignum-cluster"></span>
<img src="Figures/num_cluster.jpeg" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 2:  Mean adjusted adherence for K = 2 to 5 for the epileptic.qol dataset.
</p>
</div>
</div>
<p>From Figure <a href="#fig:fignum-cluster">2</a>, it was clear that the model with
the number of clusters <span class="math inline">\(K=2\)</span> had the largest mean adjusted adherence
(0.88), suggesting that a two-cluster model provided the best fit for
the data. To determine the number of clusters based on DIC and WAIC, one
can also use the <strong>model.selection.criteria()</strong> function to compute
these indices.</p>
<p>Next, we used the following commands to fit the final model with <span class="math inline">\(K=2\)</span>.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> fit.BCC2 <span class="ot">&lt;-</span> <span class="fu">BCC.multi</span> (</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">mydat =</span> <span class="fu">list</span>(dat<span class="sc">$</span>anxiety_scale,dat<span class="sc">$</span>depress_scale,dat<span class="sc">$</span>aep_scale),</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">dist =</span> <span class="fu">c</span>(<span class="st">&quot;gaussian&quot;</span>),</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">id =</span> <span class="fu">list</span>(dat<span class="sc">$</span>id),</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">time =</span> <span class="fu">list</span>(dat<span class="sc">$</span>time),</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">formula =</span> <span class="fu">list</span>(y <span class="sc">~</span> time <span class="sc">+</span>  (<span class="dv">1</span><span class="sc">|</span>id)),</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">num.cluster =</span> <span class="dv">2</span>,</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">burn.in =</span> <span class="dv">1000</span>,   </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">thin =</span> <span class="dv">1</span>,     </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">per =</span> <span class="dv">100</span>,       </span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">max.iter =</span> <span class="dv">2000</span>)  </span></code></pre></div>
<p>One can apply <code>print()</code> and <code>summary()</code> functions to the object returned
from <code>BCC.multi()</code> to obtain summary information of the result, for
example,</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">print</span>(fit.BCC2)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">summary</span>(fit.BCC2)</span></code></pre></div>
<p>Both functions will return result summary such as number of individuals,
number of features included in the analysis, tabulation for global and
local clusterings.</p>
<h5 class="unnumbered" data-number="3.1.2" id="posterior-summary-statistics-and-trace-plots">Posterior Summary Statistics and Trace Plots</h5>
<p>One can print the summary statistics for all the model parameters using
the following commands (results not shown):</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(fit.BCC2<span class="sc">$</span>summary.stat) </span></code></pre></div>
<p>As an illustration, we print the estimated cluster probabilities
<span class="math inline">\(\boldsymbol{\pi} = (\pi_1, \pi_2)\)</span> as follow:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">print</span>(fit.BCC2<span class="sc">$</span>summary.stat<span class="sc">$</span>PPI)</span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>                   [,<span class="dv">1</span>]       [,<span class="dv">2</span>]</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>mean         <span class="fl">0.43608373</span> <span class="fl">0.56391627</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>sd           <span class="fl">0.02367214</span> <span class="fl">0.02367214</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fl">2.5</span>%         <span class="fl">0.39326189</span> <span class="fl">0.51584563</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="fl">97.5</span>%        <span class="fl">0.48415437</span> <span class="fl">0.60673811</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>geweke.stat <span class="sc">-</span><span class="fl">0.78671117</span> <span class="fl">0.78671117</span></span></code></pre></div>
<p>The summary statistics provided the posterior mean, standard deviation
(sd), 2.5%tile, 97.5%tile and the geweke statistics. For example, the
posterior mean for <span class="math inline">\(\pi_1\)</span> and <span class="math inline">\(\pi_2\)</span> were 0.44 and 0.56, respectively.
This suggested that the estimated cluster proportion for Clusters 1 and
2 are 44% and 56%, respectively. The geweke statistics for <span class="math inline">\(\pi_1\)</span> and
<span class="math inline">\(\pi_2\)</span> fell between -2 and 2, suggesting that the two parameters
converge to a stationary distribution.</p>
<p>One can also use the following commands to print the adherence
parameters for the three features, i.e.,
<span class="math inline">\(\boldsymbol{\alpha}=(\alpha_1,\alpha_2,\alpha_3)\)</span>:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">print</span>(fit.BCC2<span class="sc">$</span>summary.stat<span class="sc">$</span>ALPHA)</span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>                  [,<span class="dv">1</span>]       [,<span class="dv">2</span>]        [,<span class="dv">3</span>]</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>mean        <span class="fl">0.97572877</span> <span class="fl">0.91727462</span>  <span class="fl">0.92427049</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>sd          <span class="fl">0.01230074</span> <span class="fl">0.01764104</span>  <span class="fl">0.01771331</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="fl">2.5</span>%        <span class="fl">0.94682192</span> <span class="fl">0.88115409</span>  <span class="fl">0.88730582</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="fl">97.5</span>%       <span class="fl">0.99620462</span> <span class="fl">0.95091755</span>  <span class="fl">0.95581869</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>geweke.stat <span class="fl">0.10674171</span> <span class="fl">0.05585855</span> <span class="sc">-</span><span class="fl">1.13303537</span></span></code></pre></div>
<p>The output indicated that the posterior mean for the adherence
parameters were 0.98, 0.92 and 0.92, respectively. This suggested that
all three features highly adhered to the global clustering. The geweke
statistics suggested that all the parameters converge to a stationary
distribution. One can print similar information for the fixed effect
coefficients <span class="math inline">\(\boldsymbol{\gamma}_k\)</span> using commands
<code>print(fit.BCC2$summary.stat$GA)</code>, for the variance-covariance matrix of
the random effects <span class="math inline">\(\boldsymbol{\Sigma}_k\)</span> using commands
<code>print(fit.BCC2$summary.stat$SIGMA.SQ.U)</code>, for the residual variance of
continuous features <span class="math inline">\(\boldsymbol{\sigma}_k\)</span> using commands
<code>print(fit.BCC2$summary.stat$SIGMA.SQ.E)</code>, for <span class="math inline">\(k=1,...,K\)</span>.</p>
<p>To visualize the MCMC chain for model parameters, one can use the
<strong>traceplot()</strong> function to generate the trace plots. For example, to
generate a trace plot for cluster probabilities
<span class="math inline">\(\boldsymbol{\pi} = (\pi_1, \pi_2)\)</span> and the adherence parameters
<span class="math inline">\(\boldsymbol{\alpha}=(\alpha_1,\alpha_2,\alpha_3)\)</span>, the following
commands were used:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">traceplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">parameter =</span> <span class="st">&quot;PPI&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;pi&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;MCMC samples&quot;</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">traceplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">parameter =</span> <span class="st">&quot;ALPHA&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;alpha&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;MCMC samples&quot;</span>)</span></code></pre></div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figtrace-ppi"></span>
<img src="Figures/trace_ppi.jpeg" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 3:  Trace plots for cluster probabilities <span class="math inline">\(\pi_1\)</span> and <span class="math inline">\(\pi_2\)</span>.
</p>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figtrace-alpha"></span>
<img src="Figures/trace_alpha.jpeg" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 4:  Trace plots for adherence parameters <span class="math inline">\(\alpha_1\)</span>, <span class="math inline">\(\alpha_2\)</span>, <span class="math inline">\(\alpha_3\)</span>.
</p>
</div>
</div>
<p>The outputs are displayed in Figure  <a href="#fig:figtrace-ppi">3</a> and Figure
 <a href="#fig:figtrace-alpha">4</a>, respectively. These trace plots suggested
that <span class="math inline">\(\pi_1\)</span>, <span class="math inline">\(\pi_2\)</span>, <span class="math inline">\(\alpha_1\)</span>, <span class="math inline">\(\alpha_2\)</span> and <span class="math inline">\(\alpha_3\)</span> are mixing
well and converge.</p>
<p>In addition, one can generate trace plots for cluster-specific
parameters, for example, the fixed effect regression coefficients,
<span class="math inline">\(\boldsymbol{\gamma}_{k, r}\)</span> =
<span class="math inline">\((\boldsymbol{\gamma}_{k, r1}, \boldsymbol{\gamma}_{k, r2})\)</span>, with the
first element the intercept and the second element the slope. The
following commands generate trace plots (results not shown) for Cluster
1 of the first feature (anxiety), second (depress) and third (aep)
features, that is,
<span class="math inline">\(\boldsymbol{\gamma}_{1,1} = (\gamma_{1,11}, \gamma_{1,12})\)</span>,
<span class="math inline">\(\boldsymbol{\gamma}_{1,2} = (\gamma_{1,21}, \gamma_{1,22})\)</span>, and
<span class="math inline">\(\boldsymbol{\gamma}_{1,3} = (\gamma_{1,31}, \gamma_{1,32})\)</span>.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">traceplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">cluster.indx =</span> <span class="dv">1</span>, <span class="at">feature.indx =</span> <span class="dv">1</span>,</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">parameter =</span> <span class="st">&quot;GA&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;GA&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;MCMC samples&quot;</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">traceplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">cluster.indx =</span> <span class="dv">1</span>, <span class="at">feature.indx =</span> <span class="dv">2</span>,</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">parameter =</span> <span class="st">&quot;GA&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;GA&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;MCMC samples&quot;</span>)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">traceplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">cluster.indx =</span> <span class="dv">1</span>, <span class="at">feature.indx =</span> <span class="dv">3</span>,</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">parameter =</span> <span class="st">&quot;GA&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;GA&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;MCMC samples&quot;</span>)</span></code></pre></div>
<p>Also, the following commands generated trace plots (results not shown)
for Cluster 2 of the first feature (anxiety), second (depress) and third
(aep) features, that is,
<span class="math inline">\(\boldsymbol{\gamma}_{2,1} = (\gamma_{2,11}, \gamma_{2,12})\)</span>,
<span class="math inline">\(\boldsymbol{\gamma}_{2,2} = (\gamma_{2,21}, \gamma_{2,22})\)</span>, and
<span class="math inline">\(\boldsymbol{\gamma}_{2,3} = (\gamma_{2,31}, \gamma_{2,32})\)</span>.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">traceplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">cluster.indx =</span> <span class="dv">2</span>, <span class="at">feature.indx =</span> <span class="dv">1</span>,</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">parameter =</span> <span class="st">&quot;GA&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;GA&quot;</span>, <span class="at">xlab=</span><span class="st">&quot;MCMC samples&quot;</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">traceplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">cluster.indx =</span> <span class="dv">2</span>, <span class="at">feature.indx =</span> <span class="dv">2</span>,</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">parameter =</span> <span class="st">&quot;GA&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;GA&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;MCMC samples&quot;</span>)</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">traceplot</span>(<span class="at">fit =</span> fit.BCC2,<span class="at">cluster.indx =</span> <span class="dv">2</span>, <span class="at">feature.indx =</span> <span class="dv">3</span>,</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">parameter =</span> <span class="st">&quot;GA&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;GA&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;MCMC samples&quot;</span>)</span></code></pre></div>
<p>Similarly, <code>parameter = "SIGMA.SQ.U"</code> to generate the elements of the
variance-covariance matrix (<span class="math inline">\(\boldsymbol{\Sigma}_k\)</span>) for the random
effects, and <code>parameter = "SIGMA.SQ.E"</code> the residual variance of
continuous features <span class="math inline">\(\boldsymbol{\sigma}_k\)</span>, for <span class="math inline">\(k=1,...,K\)</span>.</p>
<h5 class="unnumbered" data-number="3.1.3" id="longitudinal-profiles-by-local-and-global-clusters">Longitudinal Profiles by Local and Global Clusters</h5>
<p>The package uses the mode over all the MCMC samples (after burn-in and
thinning) as the point estimates for both feature-specific and global
clusterings. The following commands computed the number of individuals
belonging to each cluster for both feature-specific and global
clusterings.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">table</span>(fit.BCC2<span class="sc">$</span>cluster.local[[<span class="dv">1</span>]])</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="dv">1</span>   <span class="dv">2</span> </span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="dv">231</span> <span class="dv">309</span> </span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">table</span>(fit.BCC2<span class="sc">$</span>cluster.local[[<span class="dv">2</span>]])</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="dv">1</span>   <span class="dv">2</span> </span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="dv">223</span> <span class="dv">317</span>  </span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">table</span>(fit.BCC2<span class="sc">$</span>cluster.local[[<span class="dv">3</span>]])</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>  <span class="dv">1</span>   <span class="dv">2</span> </span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a><span class="dv">267</span> <span class="dv">273</span> </span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">table</span>(fit.BCC2<span class="sc">$</span>cluster.global)</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>  <span class="dv">1</span>   <span class="dv">2</span> </span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="dv">236</span> <span class="dv">304</span> </span></code></pre></div>
<p>That is, for clustering based on the first feature (anxiety)
<span class="math inline">\(\boldsymbol{L}_1\)</span>, 231 individuals were assigned to Cluster 1 and 309
were assigned to Cluster 2. For clustering based on the second feature
(depress) <span class="math inline">\(\boldsymbol{L}_2\)</span>, 223 individuals were assigned to Cluster 1
and 317 were assigned to Cluster 2. For clustering based on the third
feature (aep) <span class="math inline">\(\boldsymbol{L}_3\)</span>, 267 individuals were assigned to
Cluster 1 and 273 were assigned to Cluster 2. For global clustering,
<span class="math inline">\(\boldsymbol{C}\)</span>, 236 individuals were assigned to Cluster 1 and 304
were assigned to Cluster 2.</p>
<p>To plot the longitudinal trajectory of features by local and global
clusterings, <strong>trajplot()</strong> function can be used. Example codes for
plotting the trajectory of the features in the <strong>epileptic.qol</strong> data
were provided as follows:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> gp1 <span class="ot">&lt;-</span> <span class="fu">trajplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">feature.ind =</span> <span class="dv">1</span>,</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">which.cluster =</span> <span class="st">&quot;local.cluster&quot;</span>,</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">title=</span> <span class="fu">bquote</span>(<span class="fu">paste</span>(<span class="st">&quot;Local Clustering (&quot;</span>,</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="fu">hat</span>(alpha)[<span class="dv">1</span>] <span class="sc">==</span>.(<span class="fu">round</span>(fit.BCC2<span class="sc">$</span>alpha[<span class="dv">1</span>],<span class="dv">2</span>)),<span class="st">&quot;)&quot;</span>)),</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">xlab =</span> <span class="st">&quot;time (months)&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;anxiety&quot;</span>, <span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;#00BA38&quot;</span>, <span class="st">&quot;#619CFF&quot;</span>))</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> gp2 <span class="ot">&lt;-</span> <span class="fu">trajplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">feature.ind =</span> <span class="dv">2</span>,</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">which.cluster =</span> <span class="st">&quot;local.cluster&quot;</span>,</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">title =</span> <span class="fu">bquote</span>(<span class="fu">paste</span>(<span class="st">&quot;Local Clustering (&quot;</span>,</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="fu">hat</span>(alpha)[<span class="dv">2</span>] <span class="sc">==</span>.(<span class="fu">round</span>(fit.BCC2<span class="sc">$</span>alpha[<span class="dv">2</span>],<span class="dv">2</span>)),<span class="st">&quot;)&quot;</span>)),</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">xlab =</span> <span class="st">&quot;time (months)&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;depress&quot;</span>, <span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;#00BA38&quot;</span>, <span class="st">&quot;#619CFF&quot;</span>))</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> gp3 <span class="ot">&lt;-</span> <span class="fu">trajplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">feature.ind =</span> <span class="dv">3</span>,</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">which.cluster =</span> <span class="st">&quot;local.cluster&quot;</span>,</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">title =</span> <span class="fu">bquote</span>(<span class="fu">paste</span>(<span class="st">&quot;Local Clustering (&quot;</span>,</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="fu">hat</span>(alpha)[<span class="dv">3</span>] <span class="sc">==</span>.(<span class="fu">round</span>(fit.BCC2<span class="sc">$</span>alpha[<span class="dv">3</span>],<span class="dv">2</span>)),<span class="st">&quot;)&quot;</span>)),</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">xlab =</span> <span class="st">&quot;time (months)&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;aep&quot;</span>, <span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;#00BA38&quot;</span>, <span class="st">&quot;#619CFF&quot;</span>))</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> gp4 <span class="ot">&lt;-</span> <span class="fu">trajplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">feature.ind =</span> <span class="dv">1</span>,</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">which.cluster =</span> <span class="st">&quot;global.cluster&quot;</span>,</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">title =</span> <span class="st">&quot;Global Clustering&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;time (months)&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;anxiety&quot;</span>, </span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;#00BA38&quot;</span>, <span class="st">&quot;#619CFF&quot;</span>))</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> gp5 <span class="ot">&lt;-</span> <span class="fu">trajplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">feature.ind =</span> <span class="dv">2</span>,</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">which.cluster =</span> <span class="st">&quot;global.cluster&quot;</span>,</span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">title =</span> <span class="st">&quot;Global Clustering&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;time (months)&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;depress&quot;</span>, </span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;#00BA38&quot;</span>, <span class="st">&quot;#619CFF&quot;</span>))</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> gp6 <span class="ot">&lt;-</span> <span class="fu">trajplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">feature.ind =</span> <span class="dv">3</span>,</span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">which.cluster =</span> <span class="st">&quot;global.cluster&quot;</span>,</span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">title =</span> <span class="st">&quot;Global Clustering&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;time (months)&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;aep&quot;</span>, </span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;#00BA38&quot;</span>, <span class="st">&quot;#619CFF&quot;</span>))</span></code></pre></div>
<p>The <strong>cowplot</strong> package <span class="citation" data-cites="Wilke2019">(<a href="#ref-Wilke2019" role="doc-biblioref">Wilke 2024</a>)</span> is used here to combine all six
plots together into one figure.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">library</span>(<span class="st">&quot;cowplot&quot;</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">dev.new</span>(<span class="at">width =</span> <span class="dv">180</span>, <span class="at">height =</span> <span class="dv">120</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot_grid</span>(gp1, gp2, gp3, gp4, gp5, gp6, </span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">&quot;(A)&quot;</span>, <span class="st">&quot;(B)&quot;</span>, <span class="st">&quot;(C)&quot;</span>, <span class="st">&quot;(D)&quot;</span>, <span class="st">&quot;(E)&quot;</span>, <span class="st">&quot;(F)&quot;</span>), </span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">ncol =</span> <span class="dv">3</span>,   <span class="at">align =</span> <span class="st">&quot;v&quot;</span> )  </span></code></pre></div>
<p>The results are displayed in Figure <a href="#fig:figtraj-epileptic-qol">5</a>.
The top panel (A, B and C) shows the three features plotted by local
clustering indices whereas the bottom panel (D, E, F) shows the three
features plotted by global clustering indices. Individuals in Cluster 1
had lower anxiety scores, depress scores and aep scores, which
represented a better health condition compared to those in Cluster 2.
The local clustering based on anxiety score highly adhered to the global
clustering (<span class="math inline">\(\hat{\alpha}_1 = 0.98\)</span>), followed by aep
(<span class="math inline">\(\hat{\alpha}_2 = 0.92\)</span>) and depress scores (<span class="math inline">\(\hat{\alpha}_3 = 0.92\)</span>),
respectively. This suggested all three features highly adhered to the
global clustering and the anxiety score contributed the most information
to determine the global clustering.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figtraj-epileptic-qol"></span>
<img src="Figures/trajplot.jpeg" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 5:  Longitudinal trajectories for features by local and global clusterings epileptic.qol data. Locally weighted scatterplot smoothing curves are overlaid on each panel to provide an estimate of the overall trend. (A) Longitudinal trajectories of anxiety plotted by local clustering <span class="math inline">\(\boldsymbol{L}_1\)</span>. (B) Longitudinal trajectories of depress plotted by local clustering <span class="math inline">\(\boldsymbol{L}_2\)</span>. (C) Longitudinal trajectories of aep plotted by local clustering <span class="math inline">\(\boldsymbol{L}_3\)</span>. (D) Longitudinal trajectories of anxiety plotted by global clustering <span class="math inline">\(\boldsymbol{C}\)</span>. (E) Longitudinal trajectories of depress plotted by global clustering <span class="math inline">\(\boldsymbol{C}\)</span>. (F) Longitudinal trajectories of aep plotted by global clustering <span class="math inline">\(\boldsymbol{C}\)</span>.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="3.1.4" id="goodness-of-fit-1">Goodness of Fit</h5>
<p>To assess the model goodness of fit, one can use <strong>BayesT()</strong> for
computing the posterior predictive check. The commands of the
two-cluster model for <strong>epileptic.qol</strong> data are as follows:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> res <span class="ot">&lt;-</span> <span class="fu">BayesT</span>(<span class="at">fit =</span> fit.BCC2)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot</span>(<span class="fu">log</span>(res<span class="sc">$</span>T.obs), <span class="fu">log</span>(res<span class="sc">$</span>T.rep), <span class="at">xlim =</span> <span class="fu">c</span>(<span class="fl">8.45</span>,<span class="fl">8.7</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="fl">8.45</span>,<span class="fl">8.7</span>), <span class="at">cex =</span> <span class="fl">1.5</span>,</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">xlab =</span> <span class="st">&quot;Observed T statistics (in log scale)&quot;</span>, </span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">ylab =</span> <span class="st">&quot;Predicted T statistics (in log scale)&quot;</span>)</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">abline</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">lwd =</span> <span class="dv">2</span>,<span class="at">col =</span> <span class="dv">2</span>)</span></code></pre></div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figposterior-check"></span>
<img src="Figures/posterior_check.jpeg" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 6:  Posterior predictive check: observed T (in log scale) and predicted T (in log scale) statistics.
</p>
</div>
</div>
<p>The result is displayed in Figure <a href="#fig:figposterior-check">6</a>. This
figure suggested that there was no systematic pattern (e.g., consistent
overestimation or underestimation) between the observed <span class="math inline">\(T\)</span> (in log
scale) and predicted <span class="math inline">\(T\)</span> (in log scale) statistics, indicating that the
model provided a good fit to the data. To further evaluate the goodness
of fit using an objective measure, the Bayesian p-value can be
calculated, using the following commands,</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> p.value <span class="ot">&lt;-</span> <span class="fu">sum</span>(res<span class="sc">$</span>T.rep <span class="sc">&gt;</span> res<span class="sc">$</span>T.obs)<span class="sc">/</span><span class="fu">length</span>(res<span class="sc">$</span>T.rep)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> p.value </span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">0.559</span></span></code></pre></div>
<p>The Bayesian p-value of 0.559 corresponds to the proportion of samples
above the diagonal, further suggesting that there is no clear evidence
of model misspecification or that the model provides a poor fit to the
data. In addition, the output also provided the posterior cluster
probability, which is a measure of cluster uncertainty. The following
commands generated a boxplot for the posterior cluster probabilities by
clusters.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> fit.BCC2<span class="sc">$</span>cluster.global <span class="ot">&lt;-</span> <span class="fu">factor</span>(fit.BCC2<span class="sc">$</span>cluster.global,</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">&quot;Cluster 1&quot;</span>, <span class="st">&quot;Cluster 2&quot;</span>))</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">boxplot</span>(fit.BCC2<span class="sc">$</span>postprob <span class="sc">~</span> fit.BCC2<span class="sc">$</span>cluster.global, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="dv">1</span>),</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Posterior Cluster Probability&quot;</span>)</span></code></pre></div>
<p>The results were displayed in Figure <a href="#fig:figposterior-prob">7</a>. The
posterior cluster probabilities for a majority of individuals in both
Clusters 1 and 2 were close to 1, except for a few individuals. This
suggested that cluster membership for individuals is robust.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figposterior-prob"></span>
<img src="Figures/posterior_prob.jpeg" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 7:  Boxplot for the posterior cluster probabilities by clusters.
</p>
</div>
</div>
<h5 class="unnumbered" data-number="3.1.5" id="using-a-different-set-of-initial-values-for-local-and-global-cluster-memberhips">Using a Different Set of Initial Values for Local and Global Cluster Memberhips</h5>
<p>It is a good practice to fit the model with different starting values to
ensure the model converges and the clustering results are stable. By
default, the initial values of the local cluster membership are randomly
generated from 1 to <span class="math inline">\(K\)</span> for each feature and the global cluster
membership is set to be equal to the local cluster membership of the
first feature. One can also supply the local cluster membership by
fitting the Bayesian mixture model for longitudinal data to each feature
through the mixAK package for the given <span class="math inline">\(K\)</span>. To do so, the argument
<code>initial.cluster.membership = "mixAK"</code> can be added to the <code>BCC.multi</code>
function. One can also provide user input initial values for the cluster
membership, by using <code>initial.cluster.membership = "input"</code>, then
supply the initial values using the
<code>input.initial.local.cluster.membership</code> and
<code>input.initial.global.cluster.membership</code> augments. For example, if the
user would like to refit the model using clustering results generated
from the previous model (<code>fit.BCC2</code>), the following codes can be used:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> fit.BCC2a <span class="ot">&lt;-</span>  <span class="fu">BCC.multi</span> (</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">mydat =</span> <span class="fu">list</span>(dat<span class="sc">$</span>anxiety_scale, dat<span class="sc">$</span>depress_scale, dat<span class="sc">$</span>aep_scale),</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">dist =</span> <span class="fu">c</span>(<span class="st">&quot;gaussian&quot;</span>),</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">id =</span> <span class="fu">list</span>(dat<span class="sc">$</span>id),</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">time =</span> <span class="fu">list</span>(dat<span class="sc">$</span>time),</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">formula =</span> <span class="fu">list</span>(y <span class="sc">~</span> time <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>id)),</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">initial.cluster.membership =</span> <span class="st">&quot;input&quot;</span>,</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">input.initial.local.cluster.membership =</span> <span class="fu">list</span>(fit.BCC2<span class="sc">$</span>cluster.local[[<span class="dv">1</span>]],</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        fit.BCC2<span class="sc">$</span>cluster.local[[<span class="dv">2</span>]], fit.BCC2<span class="sc">$</span>cluster.local[[<span class="dv">3</span>]]),</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">input.initial.global.cluster.membership =</span> fit.BCC2<span class="sc">$</span>cluster.global,</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">num.cluster =</span> <span class="dv">2</span>,</span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">burn.in =</span> <span class="dv">1000</span>, </span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">thin =</span> <span class="dv">1</span>, </span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">per =</span> <span class="dv">100</span>,  </span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">max.iter =</span> <span class="dv">2000</span>)  </span></code></pre></div>
<h5 class="unnumbered" data-number="3.1.6" id="comparing-clustering-results-to-other-model-specifications">Comparing Clustering Results to Other Model Specifications</h5>
<p>To evaluate the robustness of the resulting clusters, one can compare
the clustering results to more complicated models. We fit a BCC model
with a linear form using a random intercept and slope, and a BCC model
with a quadratic form using a random intercept and slope. The commands
are as follows:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> fit.BCC2b <span class="ot">&lt;-</span>  <span class="fu">BCC.multi</span> (</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">mydat =</span> <span class="fu">list</span>(dat<span class="sc">$</span>anxiety_scale, dat<span class="sc">$</span>depress_scale, dat<span class="sc">$</span>aep_scale),</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">dist =</span> <span class="fu">c</span>(<span class="st">&quot;gaussian&quot;</span>),</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">id =</span> <span class="fu">list</span>(dat<span class="sc">$</span>id),</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">time =</span> <span class="fu">list</span>(dat<span class="sc">$</span>time),</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">formula =</span> <span class="fu">list</span>(y <span class="sc">~</span> time <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">+</span> time<span class="sc">|</span>id)),</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">num.cluster =</span> <span class="dv">2</span>,</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">print.info =</span> <span class="st">&quot;FALSE&quot;</span>,</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">burn.in =</span> <span class="dv">1000</span>,             </span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">thin =</span> <span class="dv">1</span>,               </span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">per =</span> <span class="dv">100</span>,                   </span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">max.iter =</span> <span class="dv">2000</span>) </span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> fit.BCC2c <span class="ot">&lt;-</span>  <span class="fu">BCC.multi</span> (</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">mydat =</span> <span class="fu">list</span>(dat<span class="sc">$</span>anxiety_scale, dat<span class="sc">$</span>depress_scale, dat<span class="sc">$</span>aep_scale),</span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">dist =</span> <span class="fu">c</span>(<span class="st">&quot;gaussian&quot;</span>),</span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">id =</span> <span class="fu">list</span>(dat<span class="sc">$</span>id),</span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">time =</span> <span class="fu">list</span>(dat<span class="sc">$</span>time),</span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">formula =</span> <span class="fu">list</span>(y <span class="sc">~</span> time <span class="sc">+</span> time2 <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">+</span> time<span class="sc">|</span>id)),</span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">num.cluster =</span> <span class="dv">2</span>,</span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">print.info =</span> <span class="st">&quot;FALSE&quot;</span>,</span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">burn.in =</span> <span class="dv">1000</span>,  </span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">thin =</span> <span class="dv">1</span>,  </span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">per =</span> <span class="dv">100</span>,  </span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">max.iter =</span> <span class="dv">2000</span>) </span>
<span id="cb31-26"><a href="#cb31-26" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> fit.BCC2b<span class="sc">$</span>cluster.global <span class="ot">&lt;-</span> <span class="fu">factor</span>(fit.BCC2b<span class="sc">$</span>cluster.global,</span>
<span id="cb31-27"><a href="#cb31-27" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">&quot;Cluster 1&quot;</span>, <span class="st">&quot;Cluster 2&quot;</span>))</span>
<span id="cb31-28"><a href="#cb31-28" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">table</span>(fit.BCC2<span class="sc">$</span>cluster.global, fit.BCC2b<span class="sc">$</span>cluster.global)</span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>           Cluster <span class="dv">1</span> Cluster <span class="dv">2</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  Cluster <span class="dv">1</span>       <span class="dv">223</span>        <span class="dv">13</span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  Cluster <span class="dv">2</span>         <span class="dv">1</span>       <span class="dv">303</span></span></code></pre></div>
<p>The agreement between the first model (linear model with a random
intercept) and the second model (linear model with a random intercept
and random slope) was <span class="math inline">\((223 + 303)/540 = 97\%\)</span>.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> fit.BCC2c<span class="sc">$</span>cluster.global <span class="ot">&lt;-</span> <span class="fu">factor</span>(fit.BCC2c<span class="sc">$</span>cluster.global,</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">&quot;Cluster 1&quot;</span>, <span class="st">&quot;Cluster 2&quot;</span>))</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">table</span>(fit.BCC2<span class="sc">$</span>cluster.global, fit.BCC2c<span class="sc">$</span>cluster.global)</span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>            Cluster <span class="dv">1</span> Cluster <span class="dv">2</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  Cluster <span class="dv">1</span>       <span class="dv">227</span>       <span class="dv">9</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  Cluster <span class="dv">2</span>         <span class="dv">2</span>       <span class="dv">302</span></span></code></pre></div>
<p>The agreement between the first model (linear model with a random
intercept) and the second model (quadratic model with a random intercept
and random slope) was <span class="math inline">\((226 + 301)/540 = 98\%\)</span>. This suggested that
fitting a more complicated model does not substantially change the
individual cluster membership, and further attests that the estimated
cluster membership was robust to different model specifications.</p>
<h4 class="unnumbered" data-number="3.2" id="example-2-mayo-clinic-primary-biliary-cholangitis-data">Example 2: Mayo Clinic Primary Biliary Cholangitis Data</h4>
<p>The second dataset was from a randomized placebo-controlled trial of the
drug D-penicillamine for Primary Biliary Cholangitis conducted between
1974 and 1984 at the Mayo Clinic Primary Biliary Cholangitis (PBC). The
dataset used here contains multiple laboratory results collected
longitudinally on 312 randomized PBC patients. Example features
(variables) include serum bilirubin (mg/dl), serum albumin (mg/dl),
platelet count, serum glutamic-oxaloacetic transaminase and the presence
of blood vessel malformations in the skin. The dataset was analyzed
previously by <span class="citation" data-cites="Komarek2013">Komárek and Komárková (<a href="#ref-Komarek2013" role="doc-biblioref">2013</a>)</span> using the <strong>mixAK</strong> package. The analysis
goal was to identify subgroups of patients with similar profiles based
on multiple features which may serve as critical information regarding
patients’ prognostics.</p>
<p>To illustrate the clustering process of mixed-type longitudinal
features, we followed <span class="citation" data-cites="Komarek2013">Komárek and Komárková (<a href="#ref-Komarek2013" role="doc-biblioref">2013</a>)</span> analysis and chose three features of
interest, namely the lbili (log of serum bilirubin), platelet count and
spiders (presence of blood vessel malformations in the ski). Also,
following <span class="citation" data-cites="Komarek2013">Komárek and Komárková (<a href="#ref-Komarek2013" role="doc-biblioref">2013</a>)</span> we included <span class="math inline">\(N = 260\)</span> individuals known to be
alive at 910 days of follow-up, and only the longitudinal measurements
up to this point were considered (known as the <strong>PBC910</strong> data). The
median (min, max) number of observations for these markers was 4 (1, 5).
Given a small number of observations available for each individual, we
considered a model with only a random intercept, i.e., using the
following specification:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>            formula <span class="ot">=</span> <span class="fu">list</span>(y <span class="sc">~</span> time <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>id))</span></code></pre></div>
<p>For a complete analysis, one can follow the steps outlined in Example 1
to determine the number of clusters and to evaluate the model
performance using several different approaches (e.g., posterior
predictive check). Here, as an illustration, and for the purpose of
comparison, we only fitted a model with two clusters (i.e., <span class="math inline">\(K=2\)</span>) as in
<span class="citation" data-cites="Komarek2013">Komárek and Komárková (<a href="#ref-Komarek2013" role="doc-biblioref">2013</a>)</span>.</p>
<p>The data is available through the <strong>mixAK</strong> package. The following
commands prepared the dataset for analysis.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">library</span>(<span class="st">&quot;mixAK&quot;</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">data</span>(PBC910)</span></code></pre></div>
<p>Using the argument <code>dist = c("gaussian","poisson","binomial")</code> to
specify the distributions for the three features, the following commands
performed a BCC model with <span class="math inline">\(K=2\)</span>.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>R <span class="sc">&gt;</span> fit.BCC2 <span class="ot">&lt;-</span> <span class="fu">BCC.multi</span>(</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">mydat =</span> <span class="fu">list</span>(PBC910<span class="sc">$</span>lbili, PBC910<span class="sc">$</span>platelet, PBC910<span class="sc">$</span>spiders),</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">dist =</span> <span class="fu">c</span>(<span class="st">&quot;gaussian&quot;</span>, <span class="st">&quot;poisson&quot;</span>, <span class="st">&quot;binomial&quot;</span>),</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">id =</span> <span class="fu">list</span>(PBC910<span class="sc">$</span>id),</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">time =</span> <span class="fu">list</span>(PBC910<span class="sc">$</span>month),</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">formula =</span> <span class="fu">list</span>(y <span class="sc">~</span> time <span class="sc">+</span> (<span class="dv">1</span><span class="sc">|</span>id)),</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">num.cluster =</span> <span class="dv">2</span>,</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">burn.in =</span> <span class="dv">10000</span>,   </span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">thin =</span> <span class="dv">10</span>,     </span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">per =</span> <span class="dv">1000</span>,       </span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">max.iter =</span> <span class="dv">20000</span>)</span></code></pre></div>
<p>The following commands generated the longitudinal profile plots for the
three features (lbili, platelet, spiders).</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> gp1 <span class="ot">&lt;-</span> <span class="fu">trajplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">feature.ind =</span> <span class="dv">1</span>, </span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">which.cluster =</span> <span class="st">&quot;local.cluster&quot;</span>,</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">title =</span> <span class="fu">bquote</span>(<span class="fu">paste</span>(<span class="st">&quot;Local Clustering (&quot;</span>,</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="fu">hat</span>(alpha)[<span class="dv">1</span>] <span class="sc">==</span>.(<span class="fu">round</span>(fit.BCC2<span class="sc">$</span>alpha[<span class="dv">1</span>],<span class="dv">2</span>)),<span class="st">&quot;)&quot;</span>)),</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">xlab =</span> <span class="st">&quot;months&quot;</span>,<span class="at">ylab =</span> <span class="st">&quot;lbili&quot;</span>, <span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;#00BA38&quot;</span>, <span class="st">&quot;#619CFF&quot;</span>))</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> gp2 <span class="ot">&lt;-</span> <span class="fu">trajplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">feature.ind =</span> <span class="dv">2</span>, </span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">which.cluster =</span> <span class="st">&quot;local.cluster&quot;</span>,</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">title =</span> <span class="fu">bquote</span>(<span class="fu">paste</span>(<span class="st">&quot;Local Clustering (&quot;</span>,</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="fu">hat</span>(alpha)[<span class="dv">2</span>] <span class="sc">==</span>.(<span class="fu">round</span>(fit.BCC2<span class="sc">$</span>alpha[<span class="dv">2</span>],<span class="dv">2</span>)),<span class="st">&quot;)&quot;</span>)),</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">xlab =</span> <span class="st">&quot;months&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;platelet&quot;</span>, <span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;#00BA38&quot;</span>, <span class="st">&quot;#619CFF&quot;</span>))</span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> gp3 <span class="ot">&lt;-</span> <span class="fu">trajplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">feature.ind =</span> <span class="dv">3</span>, <span class="at">which.cluster =</span> <span class="st">&quot;local.cluster&quot;</span>,</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">title =</span> <span class="fu">bquote</span>(<span class="fu">paste</span>(<span class="st">&quot;Local Clustering (&quot;</span>,</span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="fu">hat</span>(alpha)[<span class="dv">3</span>] <span class="sc">==</span>.(<span class="fu">round</span>(fit.BCC2<span class="sc">$</span>alpha[<span class="dv">3</span>],<span class="dv">2</span>)),<span class="st">&quot;)&quot;</span>)),</span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">xlab =</span> <span class="st">&quot;months&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;spiders&quot;</span>, <span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;#00BA38&quot;</span>, <span class="st">&quot;#619CFF&quot;</span>))</span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> gp4 <span class="ot">&lt;-</span> <span class="fu">trajplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">feature.ind =</span> <span class="dv">1</span>, <span class="at">which.cluster =</span> <span class="st">&quot;global.cluster&quot;</span>,</span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">title =</span> <span class="st">&quot;Global Clustering&quot;</span>,</span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">xlab =</span><span class="st">&quot;months&quot;</span>,<span class="at">ylab =</span> <span class="st">&quot;lbili&quot;</span>, <span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;#00BA38&quot;</span>, <span class="st">&quot;#619CFF&quot;</span>))</span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> gp5 <span class="ot">&lt;-</span> <span class="fu">trajplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">feature.ind =</span> <span class="dv">2</span>, <span class="at">which.cluster =</span> <span class="st">&quot;global.cluster&quot;</span>,</span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">title =</span> <span class="st">&quot;Global Clustering&quot;</span>,</span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">xlab =</span> <span class="st">&quot;months&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;platelet&quot;</span>, <span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;#00BA38&quot;</span>, <span class="st">&quot;#619CFF&quot;</span>))</span>
<span id="cb38-21"><a href="#cb38-21" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> gp6 <span class="ot">&lt;-</span> <span class="fu">trajplot</span>(<span class="at">fit =</span> fit.BCC2, <span class="at">feature.ind =</span> <span class="dv">3</span>, <span class="at">which.cluster =</span> <span class="st">&quot;global.cluster&quot;</span>,</span>
<span id="cb38-22"><a href="#cb38-22" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">title =</span> <span class="st">&quot;Global Clustering&quot;</span>,</span>
<span id="cb38-23"><a href="#cb38-23" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">xlab =</span> <span class="st">&quot;months&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;spiders&quot;</span>,<span class="at">color =</span> <span class="fu">c</span>(<span class="st">&quot;#00BA38&quot;</span>, <span class="st">&quot;#619CFF&quot;</span>))</span>
<span id="cb38-24"><a href="#cb38-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb38-25"><a href="#cb38-25" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">library</span>(<span class="st">&quot;cowplot&quot;</span>)</span>
<span id="cb38-26"><a href="#cb38-26" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot_grid</span>(gp1, gp2, gp3, gp4, gp5, gp6, </span>
<span id="cb38-27"><a href="#cb38-27" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">&quot;(A)&quot;</span>, <span class="st">&quot;(B)&quot;</span>, <span class="st">&quot;(C)&quot;</span>, <span class="st">&quot;(D)&quot;</span>, <span class="st">&quot;(E)&quot;</span>, <span class="st">&quot;(F)&quot;</span>), </span>
<span id="cb38-28"><a href="#cb38-28" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">ncol =</span> <span class="dv">3</span>,   <span class="at">align =</span> <span class="st">&quot;v&quot;</span> )  </span></code></pre></div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:figtraj"></span>
<img src="Figures/trajplot_PBC910.jpeg" alt="graphic without alt text" width="100%" />
<p class="caption">
Figure 8:  Longitudinal trajectories for features by local and global clusterings for PBC910 data. Locally weighted scatterplot smoothing curves are overlaid on each panel to provide an estimate of the overall trend. (A) Longitudinal trajectories of lbili plotted by local clustering <span class="math inline">\(\boldsymbol{L}_1\)</span>. (B) Longitudinal trajectories of platelet plotted by local clustering <span class="math inline">\(\boldsymbol{L}_2\)</span>. (C) Longitudinal trajectories of spiders plotted by local clustering <span class="math inline">\(\boldsymbol{L}_3\)</span>. (D) Longitudinal trajectories of lbili plotted by global clustering <span class="math inline">\(\boldsymbol{C}\)</span>. (E) Longitudinal trajectories of platelet plotted by global clustering <span class="math inline">\(\boldsymbol{C}\)</span>. (F) Longitudinal trajectories of spiders plotted by global clustering <span class="math inline">\(\boldsymbol{C}\)</span>.
</p>
</div>
</div>
<p>To compare the results with an existing model, we fit a Bayesian mixture
model using <strong>mixAK</strong> package using the following commands (see
<span class="citation" data-cites="Komarek2014">(<a href="#ref-Komarek2014" role="doc-biblioref">Komárek and Komárková 2014</a>)</span> for details of using this package):</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> mod <span class="ot">&lt;-</span> <span class="fu">GLMM_MCMC</span>(<span class="at">y =</span> PBC910[, <span class="fu">c</span>(<span class="st">&quot;lbili&quot;</span>, <span class="st">&quot;platelet&quot;</span>, <span class="st">&quot;spiders&quot;</span>)],</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">dist =</span> <span class="fu">c</span>(<span class="st">&quot;gaussian&quot;</span>, <span class="st">&quot;poisson(log)&quot;</span>, <span class="st">&quot;binomial(logit)&quot;</span>),</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">id =</span> PBC910[, <span class="st">&quot;id&quot;</span>], </span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">x =</span> <span class="fu">list</span>(<span class="at">lbili =</span> <span class="st">&quot;empty&quot;</span>, <span class="at">platelet =</span> <span class="st">&quot;empty&quot;</span>, <span class="at">spiders =</span> PBC910[, <span class="st">&quot;month&quot;</span>]), </span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">z =</span> <span class="fu">list</span>(<span class="at">lbili =</span> PBC910[, <span class="st">&quot;month&quot;</span>], <span class="at">platelet =</span> PBC910[, <span class="st">&quot;month&quot;</span>], <span class="at">spiders =</span> <span class="st">&quot;empty&quot;</span>),</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">random.intercept =</span> <span class="fu">rep</span>(<span class="cn">TRUE</span>, <span class="dv">3</span>), <span class="at">prior.b =</span> <span class="fu">list</span>(<span class="at">Kmax =</span> <span class="dv">2</span>),</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">nMCMC =</span> <span class="fu">c</span>(<span class="at">burn =</span> <span class="dv">100</span>, <span class="at">keep =</span> <span class="dv">1000</span>, <span class="at">thin =</span> <span class="dv">10</span>, <span class="at">info =</span> <span class="dv">100</span>),</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>        <span class="at">parallel =</span> <span class="cn">FALSE</span>)</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> mod <span class="ot">&lt;-</span> <span class="fu">NMixRelabel</span>(mod, <span class="at">type =</span> <span class="st">&quot;stephens&quot;</span>, <span class="at">keep.comp.prob =</span> <span class="cn">TRUE</span>)</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> cluster.mixAK <span class="ot">&lt;-</span> <span class="fu">apply</span>(mod[[<span class="dv">1</span>]]<span class="sc">$</span>poster.comp.prob, <span class="dv">1</span>, which.max)</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">table</span>(<span class="at">mixAK =</span> cluster.mixAK, <span class="at">BCClong =</span> fit.BCC2<span class="sc">$</span>cluster.global)</span></code></pre></div>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>     BCClong</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>mixAK   <span class="dv">1</span>   <span class="dv">2</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>    <span class="dv">1</span> <span class="dv">158</span>   <span class="dv">3</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>    <span class="dv">2</span>  <span class="dv">65</span>  <span class="dv">34</span></span></code></pre></div>
<p>The agreement between the two models for the <strong>PBC910</strong> data was (158 +
34)/260 = 74%.</p>
<h3 data-number="4" id="sec:summary"><span class="header-section-number">4</span> Summary and discussion</h3>
<p>In this paper, we described a BCC model for clustering longitudinal data
with multiple features. Using two real-life and open-access data, we
provided step-by-step guidance on using the <strong>BCClong</strong> package and
interpreting the results. The results also demonstrated that the
<strong>BCClong</strong> package is a useful tool for clustering longitudinal data
and it enhanced our understanding of the heterogeneity underlying each
feature and across features. Of note, the reason we require the user to
create three separate lists of data vectors (instead of a data frame)
when using BCC.multi(), in specifying the arguments mydat, id and time
is to allow more flexibility in modeling longitudinal features with
distinct structures. This is particularly useful when these features are
collected from multiple data sources (with different study designs), and
therefore the time scale and data collection frequency could differ
between features. Additional details regarding the functions and their
arguments can be found in the package vignettes.</p>
<p>The BCC model implemented using the <strong>BCClong</strong> package yields both
feature-specific (local) clusterings and consensus (global) clustering.
In practice, global clustering is often of greater interest as it
encompassed information from all features. It can be used to associate
with exposures and predict long-term outcomes. Conceptually,
feature-specific clustering can be viewed as clustering based on a
single feature alone, whereas global clustering can be viewed as a
weighted “average” clustering across all features, with weights being
the adherence parameters. In order to capture the complexity of the
underlying population heterogeneity, it is of great importance to
consider multiple features of interest simultaneously, as we often do
for cross-sectional data.</p>
<p>Several directions can be considered in future studies. For example,
incorporating variable selection (e.g., spike-and-slab and shrinkage)
priors <span class="citation" data-cites="Lu2021b Lu2021">(<a href="#ref-Lu2021" role="doc-biblioref">Lu and Lou 2021</a>, <a href="#ref-Lu2021b" role="doc-biblioref">2022a</a>)</span> for fixed and random effects will enhance the
flexibility of the package and allow it to be applied to
high-dimensional settings. In addition, determining the number of
clusters can be achieved by using a Dirichlet process mixture model
<span class="citation" data-cites="Escobar1994 Lu2023a Lu2024">(<a href="#ref-Escobar1994" role="doc-biblioref">Escobar 1994</a>; <a href="#ref-Lu2024" role="doc-biblioref">Lu et al. 2024</a>; <a href="#ref-Lu2023a" role="doc-biblioref">Lu and Chandra 2024</a>)</span>. The package does not support
predicting the cluster membership of a new individual or future
trajectories for a given individual. These functions will be considered
in future versions of the package. Finally, the current version of the
<strong>BCClong</strong> package relies on a standard MCMC algorithm to estimate the
model which is generally slow and requires particular care for the user
to check convergence. While the computational time is reasonable for our
current applications, to scale up the model to a large dataset, other
computational algorithms such as distributed stochastic gradient MCMC
<span class="citation" data-cites="Ahn2014">(<a href="#ref-Ahn2014" role="doc-biblioref">Ahn et al. 2014</a>)</span> and variational inference <span class="citation" data-cites="Blei2017">(<a href="#ref-Blei2017" role="doc-biblioref">Blei et al. 2017</a>)</span> can be considered in
future studies.</p>
<h3 class="unnumbered" id="computational-details">Computational details</h3>
<p>Most of the functions in the <strong>BCClong</strong> package are written in C++. The
output shown in this article was obtained using R version 4.2.1 (June,
2022), <strong>BCClong</strong> 1.0.3, and the following contributed packages which
are the dependencies or imports of the <strong>BCClong</strong> package: cluster
(2.1.4) <span class="citation" data-cites="Maechler2022">(<a href="#ref-Maechler2022" role="doc-biblioref">Maechler et al. 2022</a>)</span>, coda (0.19-4) <span class="citation" data-cites="Martyn2006">(<a href="#ref-Martyn2006" role="doc-biblioref">Plummer et al. 2006</a>)</span>, ggplot2 (3.4.0)
<span class="citation" data-cites="Wickham2016">(<a href="#ref-Wickham2016" role="doc-biblioref">Wickham 2016</a>)</span>, label.switching (1.8) <span class="citation" data-cites="Papastamoulis2016">(<a href="#ref-Papastamoulis2016" role="doc-biblioref">Papastamoulis 2016</a>)</span>,
LaplacesDemon (16.1.6) <span class="citation" data-cites="Hall2022">(<a href="#ref-Hall2022" role="doc-biblioref">Hall 2021</a>)</span>, lme4 (1.1-31) <span class="citation" data-cites="Bates2015">(<a href="#ref-Bates2015" role="doc-biblioref">Bates et al. 2015</a>)</span>, MASS
(7.3-58.1) <span class="citation" data-cites="Ripley2013">(<a href="#ref-Ripley2013" role="doc-biblioref">Ripley et al. 2013</a>)</span>, mclust (6.0.0) <span class="citation" data-cites="Fraley2006">(<a href="#ref-Fraley2006" role="doc-biblioref">Scrucca et al. 2023</a>)</span>, MCMCpack (1.6-3)
<span class="citation" data-cites="Martin2011">(<a href="#ref-Martin2011" role="doc-biblioref">Martin et al. 2011</a>)</span>, mixAK (5.5) <span class="citation" data-cites="Komarek2014">(<a href="#ref-Komarek2014" role="doc-biblioref">Komárek and Komárková 2014</a>)</span>, mvtnorm (1.1-3) <span class="citation" data-cites="Genz2021">(<a href="#ref-Genz2021" role="doc-biblioref">Genz et al. 2021</a>)</span>,
nnet (7.3-18) <span class="citation" data-cites="Ripley2016">(<a href="#ref-Ripley2016" role="doc-biblioref">Venables and Ripley 2002</a>)</span>, Rcpp (1.0.9) <span class="citation" data-cites="Eddelbuettel2011">(<a href="#ref-Eddelbuettel2011" role="doc-biblioref">Eddelbuettel and François 2011</a>)</span>, Rmpfr
(0.8-9) <span class="citation" data-cites="Maechler2022a">(<a href="#ref-Maechler2022a" role="doc-biblioref">Maechler 2022</a>)</span>, truncdist (1.0-2) <span class="citation" data-cites="Novomestky2016">(<a href="#ref-Novomestky2016" role="doc-biblioref">Novomestky et al. 2016</a>)</span>. R itself
and all packages used are available from the Comprehensive R Archive
Network (CRAN) at <a href="https://CRAN.R-project.org/" class="uri">https://CRAN.R-project.org/</a>.</p>
<h3 class="unnumbered" id="acknowledgments">Acknowledgments</h3>
<p>ZT is supported by a Dean’s Doctoral Award from Queen’s University. ZL
is supported by a Discovery Grant funded by the Natural Sciences and
Engineering Research Council of Canada.</p>
</div>
<div class="sourceCode" id="cb41"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h3 class="appendix" data-number="5" id="supplementary-materials"><span class="header-section-number">5</span> Supplementary materials</h3>
<p>Supplementary materials are available in addition to this article. It can be downloaded at
<a href="RJ-2024-017.zip">RJ-2024-017.zip</a></p>
<h3 class="appendix" data-number="6" id="note"><span class="header-section-number">6</span> Note</h3>
<p>This article is converted from a Legacy LaTeX article using the
<a href="https://cran.r-project.org/package=texor">texor</a> package.
The pdf version is the official version. To report a problem with the html,
refer to CONTRIBUTE on the R Journal homepage.</p>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Ahn2014" class="csl-entry" role="listitem">
S. Ahn, B. Shahbaba and M. Welling. Distributed stochastic gradient <span>MCMC</span>. In <em>International conference on machine learning</em>, pages. 1044–1052 2014. PMLR. URL <a href="https://proceedings.mlr.press/v32/ahn14.html">https://proceedings.mlr.press/v32/ahn14.html</a>.
</div>
<div id="ref-Bates2015" class="csl-entry" role="listitem">
D. Bates, M. Mächler, B. Bolker and S. Walker. Fitting linear mixed-effects models using <span class="nocase">lme4</span>. <em>Journal of Statistical Software</em>, 67(1): 1–48, 2015. DOI <a href="https://doi.org/10.18637/jss.v067.i01">10.18637/jss.v067.i01</a>. R package version 1.1-31.
</div>
<div id="ref-Blei2017" class="csl-entry" role="listitem">
D. M. Blei, A. Kucukelbir and J. D. McAuliffe. Variational inference: A review for statisticians. <em>Journal of the American Statistical Association</em>, 112(518): 859–877, 2017. URL <a href="https://doi.org/10.1080/01621459.2017.1285773">https://doi.org/10.1080/01621459.2017.1285773</a>.
</div>
<div id="ref-Celeux2006" class="csl-entry" role="listitem">
G. Celeux, F. Forbes, C. P. Robert, D. M. Titterington, et al. Deviance information criteria for missing data models. <em>Bayesian Analysis</em>, 1(4): 651–673, 2006. DOI <a href="https://doi.org/10.1214/06-BA122">10.1214/06-BA122</a>.
</div>
<div id="ref-Eddelbuettel2011" class="csl-entry" role="listitem">
D. Eddelbuettel and R. François. <span>Rcpp</span>: Seamless <span>R</span> and <span>C++</span> integration. <em>Journal of Statistical Software</em>, 40(8): 1–18, 2011. DOI <a href="https://doi.org/10.18637/jss.v040.i08">10.18637/jss.v040.i08</a>. R package version 1.0.9.
</div>
<div id="ref-Elliott2005" class="csl-entry" role="listitem">
M. R. Elliott, J. J. Gallo, T. R. Ten Have, H. R. Bogner and I. R. Katz. Using a bayesian latent growth curve model to identify trajectories of positive affect and negative events following myocardial infarction. <em>Biostatistics</em>, 6(1): 119–143, 2005. DOI <a href="https://doi.org/10.1093/biostatistics/kxh022">10.1093/biostatistics/kxh022</a>.
</div>
<div id="ref-Escobar1994" class="csl-entry" role="listitem">
M. D. Escobar. Estimating normal means with a <span>D</span>irichlet process prior. <em>Journal of the American Statistical Association</em>, 89(425): 268–277, 1994. URL <a href="https://doi.org/10.2307/2291223">https://doi.org/10.2307/2291223</a>.
</div>
<div id="ref-Fruehwirth-Schnatter2010" class="csl-entry" role="listitem">
S. Frühwirth-Schnatter and S. Pyne. Bayesian inference for finite mixtures of univariate and multivariate skew-normal and skew-t distributions. <em>Biostatistics</em>, 11(2): 317–336, 2010. URL <a href="https://doi.org/10.1093/biostatistics/kxp062">https://doi.org/10.1093/biostatistics/kxp062</a>.
</div>
<div id="ref-Gelman2014a" class="csl-entry" role="listitem">
A. Gelman, J. Hwang and A. Vehtari. Understanding predictive information criteria for <span>B</span>ayesian models. <em>Statistics and Computing</em>, 24(6): 997–1016, 2014. URL <a href="https://doi.org/10.1007/s11222-013-9416-2">https://doi.org/10.1007/s11222-013-9416-2</a>.
</div>
<div id="ref-Gelman1996" class="csl-entry" role="listitem">
A. Gelman, X.-L. Meng and H. Stern. Posterior predictive assessment of model fitness via realized discrepancies. <em>Statistica Sinica</em>, 733–760, 1996. URL <a href="https://www.jstor.org/stable/24306036">https://www.jstor.org/stable/24306036</a>.
</div>
<div id="ref-Genolini2010" class="csl-entry" role="listitem">
C. Genolini and B. Falissard. KmL: K-means for longitudinal data. <em>Computational Statistics</em>, 25(2): 317–328, 2009. DOI <a href="https://doi.org/10.1007/s00180-009-0178-4">10.1007/s00180-009-0178-4</a>.
</div>
<div id="ref-Genolini2013" class="csl-entry" role="listitem">
C. Genolini, J.-B. Pingault, T. Driss, S. Côté, R. E. Tremblay, F. Vitaro, C. Arnaud and B. Falissard. KmL3D: A non-parametric algorithm for clustering joint trajectories. <em>Computer Methods and Programs in Biomedicine</em>, 109(1): 104–111, 2013. URL <a href="https://doi.org/10.1016/j.cmpb.2012.08.016">https://doi.org/10.1016/j.cmpb.2012.08.016</a>.
</div>
<div id="ref-Genz2021" class="csl-entry" role="listitem">
A. Genz, F. Bretz, T. Miwa, X. Mi, F. Leisch, F. Scheipl and T. Hothorn. <em><span class="nocase">mvtnorm</span>: Multivariate normal and t distributions.</em> 2021. URL <a href="https://CRAN.R-project.org/package=mvtnorm">https://CRAN.R-project.org/package=mvtnorm</a>. R package version 1.1-3.
</div>
<div id="ref-Geweke1991" class="csl-entry" role="listitem">
J. Geweke. <em>Evaluating the accuracy of sampling-based approaches to the calculation of posterior moments.</em> Federal Reserve Bank of Minneapolis, Research Department Minneapolis, MN, USA, 1991.
</div>
<div id="ref-Hall2022" class="csl-entry" role="listitem">
B. Hall. <em>LaplacesDemon: Complete environment for bayesian inference.</em> 2021. URL <a href="https://CRAN.R-project.org/package=LaplacesDemon">https://CRAN.R-project.org/package=LaplacesDemon</a>. R package version 16.1.6.
</div>
<div id="ref-Heggeseth2018" class="csl-entry" role="listitem">
B. C. Heggeseth and N. P. Jewell. How <span>G</span>aussian mixture models might miss detecting factors that impact growth patterns. <em>The Annals of Applied Statistics</em>, 12(1): 222–245, 2018. URL <a href="https://doi.org/10.1214/17-aoas1066">https://doi.org/10.1214/17-aoas1066</a>.
</div>
<div id="ref-Hickey2018" class="csl-entry" role="listitem">
G. L. Hickey, P. Philipson, A. Jorgensen and R. Kolamunnage-Dona. <span class="nocase">joineRML</span>: A joint model and software package for time-to-event and multivariate longitudinal outcomes. <em>BMC Medical Research Methodology</em>, 18(1): 1–14, 2018. URL <a href="https://doi.org/10.1186/s12874-018-0502-1">https://doi.org/10.1186/s12874-018-0502-1</a>.
</div>
<div id="ref-Komarek2014" class="csl-entry" role="listitem">
A. Komárek and L. Komárková. Capabilities of <span>R</span> package mixAK for clustering based on multivariate continuous and discrete longitudinal data. <em>Journal of Statistical Software</em>, 59(12): 1–38, 2014. URL <a href="https://doi.org/10.18637/jss.v059.i12">https://doi.org/10.18637/jss.v059.i12</a>. R package version 5.5.
</div>
<div id="ref-Komarek2013" class="csl-entry" role="listitem">
A. Komárek and L. Komárková. Clustering for multivariate continuous and discrete longitudinal data. <em>The Annals of Applied Statistics</em>, 7(1): 177–200, 2013. DOI <a href="https://doi.org/10.1214/12-AOAS580">10.1214/12-AOAS580</a>.
</div>
<div id="ref-Leiby2009" class="csl-entry" role="listitem">
B. E. Leiby, M. D. Sammel, T. R. Ten Have and K. G. Lynch. Identification of multivariate responders and non-responders by using <span>B</span>ayesian growth curve latent class models. <em>Journal of the Royal Statistical Society: Series C (Applied Statistics)</em>, 58(4): 505–524, 2009. URL <a href="https://doi.org/10.1111/j.1467-9876.2009.00663.x">https://doi.org/10.1111/j.1467-9876.2009.00663.x</a>.
</div>
<div id="ref-Leisch2004" class="csl-entry" role="listitem">
F. Leisch. <span>FlexMix</span>: A general framework for finite mixture models and latent class regression in <span>R</span>. <em>Journal of Statistical Software</em>, 11(8): 1–18, 2004. DOI <a href="https://doi.org/10.18637/jss.v011.i08">10.18637/jss.v011.i08</a>.
</div>
<div id="ref-Lock2013" class="csl-entry" role="listitem">
E. F. Lock and D. B. Dunson. Bayesian consensus clustering. <em>Bioinformatics</em>, 29(20): 2610–2616, 2013. DOI <a href="https://doi.org/10.1093/bioinformatics/btt425">10.1093/bioinformatics/btt425</a>.
</div>
<div id="ref-Lu2023" class="csl-entry" role="listitem">
Z. Lu, M. Ahmadiankalati and Z. Tan. Joint clustering multiple longitudinal features: A comparison of methods and software packages with practical guidance. <em>Statistics in Medicine</em>, 42(29): 5513–5540, 2023. URL <a href="https://doi.org/10.1002/sim.9917">https://doi.org/10.1002/sim.9917</a>.
</div>
<div id="ref-Lu2023a" class="csl-entry" role="listitem">
Z. Lu and N. K. Chandra. A sparse factor model for clustering high-dimensional longitudinal data. <em>Statistics in Medicine</em>, 43(19): 3633–3648, 2024. URL <a href="https://doi: 10.1002/sim.10151">doi: 10.1002/sim.10151</a>.
</div>
<div id="ref-Lu2021" class="csl-entry" role="listitem">
Z. Lu and W. Lou. Bayesian approaches to variable selection in mixture models with application to disease clustering. <em>Journal of Applied Statistics</em>, 1–21, 2021. URL <a href="https://doi.org/10.1080/02664763.2021.1994529">https://doi.org/10.1080/02664763.2021.1994529</a>.
</div>
<div id="ref-Lu2021b" class="csl-entry" role="listitem">
Z. Lu and W. Lou. Bayesian approaches to variable selection: A comparative study from practical perspectives. <em>International Journal of Biostatistics</em>, 18(1): 89–108, 2022a. URL <a href="https://doi.org/10.1515/ijb-2020-0130">https://doi.org/10.1515/ijb-2020-0130</a>.
</div>
<div id="ref-Lu2022" class="csl-entry" role="listitem">
Z. Lu and W. Lou. Bayesian consensus clustering for multivariate longitudinal data. <em>Statistics in Medicine</em>, 41(1): 108–127, 2022b. DOI <a href="https://doi.org/10.1002/sim.9225">10.1002/sim.9225</a>.
</div>
<div id="ref-Lu2019" class="csl-entry" role="listitem">
Z. Lu and W. Lou. Shape invariant mixture model for clustering non-linear longitudinal growth trajectories. <em>Statistical Methods in Medical Research</em>, 28(12): 3769–3784, 2019. URL <a href="https://doi.org/10.1177/0962280218815301">https://doi.org/10.1177/0962280218815301</a>.
</div>
<div id="ref-Lu2024" class="csl-entry" role="listitem">
Z. Lu, P. Subbarao and W. Lou. A <span>B</span>ayesian latent class model for integrating multi-source longitudinal data: Application to the <span>CHILD</span> cohort study. <em>Journal of the Royal Statistical Society Series C: Applied Statistics</em>, 73(2): 398–419, 2024. URL <a href="https://doi.org/10.1093/jrsssc/qlad100">https://doi.org/10.1093/jrsssc/qlad100</a>.
</div>
<div id="ref-Maechler2022a" class="csl-entry" role="listitem">
M. Maechler. <em>Rmpfr: R MPFR - multiple precision floating-point reliable.</em> 2022. URL <a href="https://CRAN.R-project.org/package=Rmpfr">https://CRAN.R-project.org/package=Rmpfr</a>. R package version 0.8-9.
</div>
<div id="ref-Maechler2022" class="csl-entry" role="listitem">
M. Maechler, P. Rousseeuw, A. Struyf, M. Hubert and K. Hornik. <em>Cluster: Cluster analysis basics and extensions.</em> 2022. URL <a href="https://CRAN.R-project.org/package=cluster">https://CRAN.R-project.org/package=cluster</a>. R package version 2.1.4.
</div>
<div id="ref-Magrini2022a" class="csl-entry" role="listitem">
A. Magrini. Assessment of agricultural sustainability in european union countries: A group-based multivariate trajectory approach. <em>Asta Advances in Statistical Analysis</em>, 106(4): 673–703, 2022a. URL <a href="https://doi.org/10.1007/s10182-022-00437-9">https://doi.org/10.1007/s10182-022-00437-9</a>.
</div>
<div id="ref-Magrini2022" class="csl-entry" role="listitem">
A. Magrini. <em>Gbmt: Group-based multivariate trajectory modeling.</em> 2022b. URL <a href="https://CRAN.R-project.org/package=gbmt">https://CRAN.R-project.org/package=gbmt</a>. R package version 0.1.3.
</div>
<div id="ref-Marson2007" class="csl-entry" role="listitem">
A. Marson, R. Appleton, G. Baker, D. Chadwick, J. Doughty, B. Eaton, C. Gamble, A. Jacoby, P. Shackley, D. Smith, et al. A randomised controlled trial examining the longer-term outcomes of standard versus new antiepileptic drugs. The SANAD trial. <em>Health Technology Assessment-southampton-</em>, 11(37): 2007. URL <a href="https://doi.org/10.3310/hta11370">https://doi.org/10.3310/hta11370</a>.
</div>
<div id="ref-Martin2011" class="csl-entry" role="listitem">
A. D. Martin, K. M. Quinn and J. H. Park. <span>MCMCpack</span>: <span>M</span>arkov chain <span>M</span>onte <span>C</span>arlo in <span>R</span>. <em>Journal of Statistical Software</em>, 42(9): 22, 2011. DOI <a href="https://doi.org/10.18637/jss.v042.i09">10.18637/jss.v042.i09</a>. R package version 1.6-3.
</div>
<div id="ref-Merkle2019" class="csl-entry" role="listitem">
E. C. Merkle, D. Furr and S. Rabe-Hesketh. Bayesian comparison of latent variable models: Conditional versus marginal likelihoods. <em>Psychometrika</em>, 84(3): 802–829, 2019. URL <a href="https://doi.org/10.1007/s11336-019-09679-0">https://doi.org/10.1007/s11336-019-09679-0</a>.
</div>
<div id="ref-Nagin2018" class="csl-entry" role="listitem">
D. S. Nagin, B. L. Jones, V. L. Passos and R. E. Tremblay. Group-based multi-trajectory modeling. <em>Statistical Methods in Medical Research</em>, 27(7): 2015–2023, 2018. URL <a href="https://doi.org/10.1177/0962280216673085">https://doi.org/10.1177/0962280216673085</a>.
</div>
<div id="ref-Nasserinejad2017" class="csl-entry" role="listitem">
K. Nasserinejad, J. van Rosmalen, W. de Kort and E. Lesaffre. Comparison of criteria for choosing the number of classes in <span>B</span>ayesian finite mixture models. <em>Plos One</em>, 12(1): e0168838, 2017. URL <a href="https://doi.org/10.1371/journal.pone.0168838">https://doi.org/10.1371/journal.pone.0168838</a>.
</div>
<div id="ref-Neelon2011" class="csl-entry" role="listitem">
B. Neelon, G. K. Swamy, L. F. Burgette and M. L. Miranda. A <span>B</span>ayesian growth mixture model to examine maternal hypertension and birth outcomes. <em>Statistics in Medicine</em>, 30(22): 2721–2735, 2011. URL <a href="https://doi.org/10.1002/sim.4291">https://doi.org/10.1002/sim.4291</a>.
</div>
<div id="ref-Novomestky2016" class="csl-entry" role="listitem">
F. Novomestky, S. Nadarajah and M. F. Novomestky. <em>Truncdist: Truncated random variables.</em> 2016. URL <a href="https://CRAN.R-project.org/package=truncdist">https://CRAN.R-project.org/package=truncdist</a>. R package version 1.0-2.
</div>
<div id="ref-Papastamoulis2016" class="csl-entry" role="listitem">
P. Papastamoulis. <span class="nocase">label.switching</span>: An <span>R</span> package for dealing with the label switching problem in <span>MCMC</span> outputs. <em>Journal of Statistical Software, Code Snippets</em>, 69(1): 1–24, 2016. DOI <a href="https://doi.org/10.18637/jss.v069.c01">10.18637/jss.v069.c01</a>. R package version 1.8.
</div>
<div id="ref-Martyn2006" class="csl-entry" role="listitem">
M. Plummer, N. Best, K. Cowles and K. Vines. CODA: Convergence diagnosis and output analysis for MCMC. <em>R News</em>, 6(1): 7–11, 2006. URL <a href="https://journal.r-project.org/archive/">https://journal.r-project.org/archive/</a>. R package version 0.19-4.
</div>
<div id="ref-Ripley2013" class="csl-entry" role="listitem">
B. Ripley, B. Venables, D. M. Bates, K. Hornik, A. Gebhardt, D. Firth and Ripley. <em><span>MASS</span>: Support functions and datasets for venables and ripley’s MASS.</em> 2013. URL <a href="https://CRAN.R-project.org/package=crosstalk">https://CRAN.R-project.org/package=crosstalk</a>. R package version 7.3-58.1.
</div>
<div id="ref-Fraley2006" class="csl-entry" role="listitem">
L. Scrucca, C. Fraley, T. B. Murphy and A. E. Raftery. <em>Model-based clustering, classification, and density estimation using <span class="nocase">mclust</span> in <span>R</span>.</em> Chapman; Hall/CRC, 2023. URL <a href="https://mclust-org.github.io/book/">https://mclust-org.github.io/book/</a>. R package version 6.0.0.
</div>
<div id="ref-Spiegelhalter2002" class="csl-entry" role="listitem">
D. J. Spiegelhalter, N. G. Best, B. P. Carlin and A. Van Der Linde. Bayesian measures of model complexity and fit. <em>Journal of the Royal Statistical Society: Series B (Statistical Methodology)</em>, 64(4): 583–639, 2002. URL <a href="https://doi.org/10.1111/1467-9868.00353">https://doi.org/10.1111/1467-9868.00353</a>.
</div>
<div id="ref-Stephens2000" class="csl-entry" role="listitem">
M. Stephens. Dealing with label switching in mixture models. <em>Journal of the Royal Statistical Society: Series B (Statistical Methodology)</em>, 62(4): 795–809, 2000. URL <a href="https://doi.org/10.1111/1467-9868.00265">https://doi.org/10.1111/1467-9868.00265</a>.
</div>
<div id="ref-Tan2022a" class="csl-entry" role="listitem">
Z. Tan, C. Shen, P. Subbarao, W. Lou and Z. Lu. A joint modeling approach for clustering mixed-type multivariate longitudinal data: Application to the <span>CHILD</span> cohort study. <em>Arxiv Preprint Arxiv:2210.08385</em>, 2022. URL <a href="https://doi.org/10.48550/arXiv.2210.08385">https://doi.org/10.48550/arXiv.2210.08385</a>.
</div>
<div id="ref-Ripley2016" class="csl-entry" role="listitem">
W. N. Venables and B. D. Ripley. <em>Modern applied statistics with s.</em> Fourth New York: Springer, 2002. URL <a href="https://www.stats.ox.ac.uk/pub/MASS4/">https://www.stats.ox.ac.uk/pub/MASS4/</a>. ISBN 0-387-95457-0, R package version 7.3-18.
</div>
<div id="ref-Watanabe2010" class="csl-entry" role="listitem">
S. Watanabe. Asymptotic equivalence of bayes cross validation and widely applicable information criterion in singular learning theory. <em>Journal of Machine Learning Research</em>, 11(Dec): 3571–3594, 2010. URL <a href="http://jmlr.org/papers/v11/watanabe10a.html">http://jmlr.org/papers/v11/watanabe10a.html</a>.
</div>
<div id="ref-Wickham2016" class="csl-entry" role="listitem">
H. Wickham. <em>ggplot2: Elegant graphics for data analysis.</em> Springer-Verlag New York, 2016. URL <a href="https://ggplot2.tidyverse.org">https://ggplot2.tidyverse.org</a>.
</div>
<div id="ref-Wilke2019" class="csl-entry" role="listitem">
C. O. Wilke. <em>Cowplot: Streamlined plot theme and plot annotations for ’ggplot2’.</em> 2024. URL <a href="https://CRAN.R-project.org/package=cowplot">https://CRAN.R-project.org/package=cowplot</a>. R package version 1.1.3.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
  <h3 id="references">References</h3>
  <div id="references-listing"></div>
  <h3 id="reuse">Reuse</h3>
  <p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
  <h3 id="citation">Citation</h3>
  <p>For attribution, please cite this work as</p>
  <pre class="citation-appendix short">Tan, et al., "BCClong: An R Package for Bayesian Consensus Clustering for Multiple Longitudinal Features", The R Journal, 2025</pre>
  <p>BibTeX citation</p>
  <pre class="citation-appendix long">@article{RJ-2024-017,
  author = {Tan, Zhiwen and Shen, Chang and Lu, Zihang},
  title = {BCClong: An R Package for Bayesian Consensus Clustering for Multiple Longitudinal Features},
  journal = {The R Journal},
  year = {2025},
  note = {https://doi.org/10.32614/RJ-2024-017},
  doi = {10.32614/RJ-2024-017},
  volume = {16},
  issue = {2},
  issn = {2073-4859},
  pages = {41-60}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<!--radix_placeholder_navigation_after_body--><html><body>
<div class="distill-site-nav distill-site-footer">
<p>© The R Foundation, <a href="mailto:r-journal@r-project.org">web page
contact</a>.</p>
</div>
<!--/radix_placeholder_navigation_after_body-->
</body></html>


</body>

</html>
