---
title: News from the Bioconductor Project
abstract: The 'News from the Bioconductor Project' article from the 2015-1 issue.
author:
- name: The Bioconductor Team
  affiliation: |
    Program in Computational Biology, Fred Hutchinson Cancer Research
    Center
date: '2015-06-01'
date_received: ~
journal:
  firstpage: '239'
  lastpage: '240'
volume: 7
issue: 1
slug: RJ-2015-1-bioconductor
citation_url: https://rjournal.github.io/
packages:
  cran: []
  bioc:
  - AnnotationHub
  - sincell
  - RCyjs
  - mogsa
  - nethet
  - pandaR
  - pwOmics
  - seq2pathway
  - BEclear
  - conumee
  - DMRcaller
  - ENmix
  - RnBeads
  - skewr
  - FlowRepositoryR
  - FlowSOM
  - flowVS
  - immunoClus
  - LEA
  - muscle
  - msa
  - gtrellis
  - ggtree
  - ComplexHeatmap
  - seqPattern
  - soGGI
  - Rhtslib
  - GoogleGenomics
  - biomaRt
  - VariantAnnotation
  - VariantFiltering
  - ensemblVEP
  - BiocCheck
preview: preview.png
bibliography: ~
CTV: ~
output:
  rjtools::rjournal_web_article:
    self_contained: yes
    toc: no
    legacy_pdf: yes
    web_only: yes
    mathjax: https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js
    md_extension: -tex_math_single_backslash

---

::: article
The [Bioconductor](http://bioconductor.org) project provides tools for
the analysis and comprehension of high-throughput genomic data. The 1024
software packages available in Bioconductor can be viewed at
<http://bioconductor.org/packages/>. Navigate packages using 'biocViews'
terms and title search. Each package has an html page with a
description, links to vignettes, reference manuals, and usage
statistics. Start using Bioconductor version 3.1 by installing R 3.2.1
and evaluating the commands

``` r
  source("http://bioconductor.org/biocLite.R")
  biocLite()
```

Install additional packages and dependencies, e.g.,
[*AnnotationHub*](https://www.bioconductor.org/packages/release/bioc/html/AnnotationHub.html),
with

``` r
  source("http://bioconductor.org/biocLite.R")
  biocLite("AnnotationHub")
```

# Bioconductor 3.1 release highlights

Bioconductor 3.1 was released on 17 April 2015. It is compatible with R
3.2 and consists of 1024 software packages, 214 experiment data
packages, and more 917 up-to-date annotation packages. There are 95 new
software packages and many updates and improvements to existing
packages. The [release
announcement](http://bioconductor.org/news/bioc_3_1_release/) includes
descriptions of new packages and updated NEWS files provided by package
maintainers. There are a great diversity of packages represented.
Highlights include:
[*sincell*](https://www.bioconductor.org/packages/release/bioc/html/sincell.html)
for assessment of cell-state hierarchies from single-cell data; gene set
and network analysis packages (e.g.,
[*RCyjs*](https://www.bioconductor.org/packages/release/bioc/html/RCyjs.html)
[*mogsa*](https://www.bioconductor.org/packages/release/bioc/html/mogsa.html),
[*nethet*](https://www.bioconductor.org/packages/release/bioc/html/nethet.html),
[*pandaR*](https://www.bioconductor.org/packages/release/bioc/html/pandaR.html),
[*pwOmics*](https://www.bioconductor.org/packages/release/bioc/html/pwOmics.html),
[*seq2pathway*](https://www.bioconductor.org/packages/release/bioc/html/seq2pathway.html));
packages for methylation (e.g.,
[*BEclear*](https://www.bioconductor.org/packages/release/bioc/html/BEclear.html),
[*conumee*](https://www.bioconductor.org/packages/release/bioc/html/conumee.html),
[*DMRcaller*](https://www.bioconductor.org/packages/release/bioc/html/DMRcaller.html),
[*ENmix*](https://www.bioconductor.org/packages/release/bioc/html/ENmix.html),
[*RnBeads*](https://www.bioconductor.org/packages/release/bioc/html/RnBeads.html),
[*skewr*](https://www.bioconductor.org/packages/release/bioc/html/skewr.html)),
flow cytometry (e.g.,
[*FlowRepositoryR*](https://www.bioconductor.org/packages/release/bioc/html/FlowRepositoryR.html),
[*FlowSOM*](https://www.bioconductor.org/packages/release/bioc/html/FlowSOM.html),
[*flowVS*](https://www.bioconductor.org/packages/release/bioc/html/flowVS.html),
[*immunoClus*](https://www.bioconductor.org/packages/release/bioc/html/immunoClus.html)),
and other domain-specific analysis (e.g.,
[*LEA*](https://www.bioconductor.org/packages/release/bioc/html/LEA.html),
for landscape genomics); access to multiple-sequence alignment
algorithms
([*muscle*](https://www.bioconductor.org/packages/release/bioc/html/muscle.html),
[*msa*](https://www.bioconductor.org/packages/release/bioc/html/msa.html));
visualization packages such as
[*gtrellis*](https://www.bioconductor.org/packages/release/bioc/html/gtrellis.html)
(genome level Trellis graph visualizes),
[*ggtree*](https://www.bioconductor.org/packages/release/bioc/html/ggtree.html)
(phylogenetic tree and associated annotation data),
[*ComplexHeatmap*](https://www.bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html),
[*seqPattern*](https://www.bioconductor.org/packages/release/bioc/html/seqPattern.html)
(oligonucleotide patterns and sequence motifs centred at a common
reference point), and
[*soGGI*](https://www.bioconductor.org/packages/release/bioc/html/soGGI.html)
(genomic interval aggregate and summary plots of signal or motif
occurrence); and infrastructure packages such as
[*Rhtslib*](https://www.bioconductor.org/packages/release/bioc/html/Rhtslib.html)
(wrapping a recent version of the htslib C library for processing CRAM,
BAM, and other high-throughput sequence files) and
[*GoogleGenomics*](https://www.bioconductor.org/packages/release/bioc/html/GoogleGenomics.html)
(to interact with the Google Genomics interface).

Our collection of microarray, transcriptome and organism-specific
*annotation packages* use the 'select' interface (keys, columns,
keytypes) to access static information on gene annotations (*org.\**
packages) and gene models (*TxDb.\** packages); these augment packages
such as
[*biomaRt*](https://www.bioconductor.org/packages/release/bioc/html/biomaRt.html)
for interactive querying of web-based resources and
[*VariantAnnotation*](https://www.bioconductor.org/packages/release/bioc/html/VariantAnnotation.html),
[*VariantFiltering*](https://www.bioconductor.org/packages/release/bioc/html/VariantFiltering.html),
and
[*ensemblVEP*](https://www.bioconductor.org/packages/release/bioc/html/ensemblVEP.html)
for annotation of DNA sequences. Use of these resources are documented
in updated annotation
[workflows](http://bioconductor.org/help/workflows/). The
*AnnotationHub* complements our traditional offerings with diverse whole
genome annotations from Ensembl, ENCODE, dbSNP, UCSC, and elsewhere; a
recent addition includes Roadmap Epigenomics resources, with use
described in the [AnnotationHub
How-To](http://bioconductor.org/packages/devel/bioc/vignettes/AnnotationHub/inst/doc/AnnotationHub-HOWTO.html)
vignette.

# Other activities

Several enhancements to the Bioconductor web and support sites aim to
help users to identify appropriate packages while encouraging developers
to provide high-quality software. In addition to
[biocViews](http://www.bioconductor.org/packages/BiocViews.html#___Software)
terms to classify packages, each 'landing' page (e.g.,
<http://bioconductor.org/packages/DESeq2>) contains graphical shields
that highlight cross-platform availability, download percentile, support
site queries and responses, years in Bioconductor, and maintenance
activity during the last 6 months. Landing pages also contain shields
more useful to package maintainers, including current build status and
test coverage. Users posting questions on the [support
site](https://support.bioconductor.org) are encouraged to add tags
identifying the package the question is about; maintainers are
subscribed to the tag and receive email notification of the question.
This enables maintainers to stay informed of problems with their
package, without requiring daily monitoring of the support site.

Continued availability of Bioconductor
[Docker](http://www.bioconductor.org/help/docker/) and
[Amazon](http://www.bioconductor.org/help/bioconductor-cloud-ami/)
images provides a very effective on-ramp for power users to rapidly
obtain access to standardized and scalable computing environments.
Docker images are available for release and development versions of
Bioconductor, with analysis-specific images pre-loaded with packages
relevant to common analysis scenarios, e.g., of sequencing, microarray,
flow cell, or proteomic data. Both Amazon and Docker images include
RStudio Server for easy web-browser based access.

New Bioconductor package contributors are encouraged to consult the
package
[guidelines](http://bioconductor.org/developers/package-guidelines/) and
[submission](http://bioconductor.org/developers/package-submission/)
sections of the Bioconductor web site, and use the
[*BiocCheck*](https://www.bioconductor.org/packages/release/bioc/html/BiocCheck.html)
package, in addition to `R CMD check`, for guidance on conforming to
Bioconductor package standards. New package submissions are
automatically built across Linux, Mac, and Windows platforms, providing
an opportunity to address cross-platform issues; many new package
contributors take advantage of this facility to refine their package
before it is subject to technical preview. Keep abreast of packages
added to the 'devel' branch and other activities by following
\@Bioconductor on Twitter.

The Bioconductor web site advertises [training and community
events](http://bioconductor.org/help/events/), including the [BioC
2015](http://bioconductor.org/BioC2015), the Bioconductor annual
conference, to be held in Seattle, 20--22 July.
:::
