<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
      margin-bottom: 0em;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
<title>simlandr: Simulation-Based Landscape Construction for Dynamical Systems</title>

<meta property="description" itemprop="description" content="We present the simlandr package for R, which provides a set of tools for constructing potential landscapes for dynamical systems using Monte Carlo simulation. Potential landscapes can be used to quantify the stability of system states. While the canonical form of a potential function is defined for gradient systems, generalized potential functions can also be defined for non-gradient dynamical systems. Our method is based on the potential landscape definition from the steady-state distribution, and can be used for a large variety of models. To facilitate simulation and computation, we introduce several novel features, including data structures optimized for batch simulations under varying conditions, an out-of-memory computation tool with integrated hash-based file-saving systems, and an algorithm for efficiently searching the minimum energy path. Using a multistable cell differentiation model as examples, we illustrate how simlandr can be used for model simulation, landscape construction, and barrier height calculation. The simlandr package is available at https://CRAN.R-project.org/package=simlandr, under GPL-3 license."/>

<link rel="canonical" href="https://doi.org/10.32614/RJ-2025-039/"/>
<link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>
<link rel="icon" type="image/vnd.microsoft.icon" href="../../resources/favicon.ico"/>

<!--  https://schema.org/Article -->
<meta property="article:published" itemprop="datePublished" content="2026-02-04"/>
<meta property="article:created" itemprop="dateCreated" content="2026-02-04"/>
<meta name="article:author" content="Jingmeng Cui"/>
<meta name="article:author" content="Merlijn Olthof"/>
<meta name="article:author" content="Anna Lichtwarck-Aschoff"/>
<meta name="article:author" content="Tiejun Li"/>
<meta name="article:author" content="Fred Hasselman"/>

<!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
<meta property="og:title" content="simlandr: Simulation-Based Landscape Construction for Dynamical Systems"/>
<meta property="og:type" content="article"/>
<meta property="og:description" content="We present the simlandr package for R, which provides a set of tools for constructing potential landscapes for dynamical systems using Monte Carlo simulation. Potential landscapes can be used to quantify the stability of system states. While the canonical form of a potential function is defined for gradient systems, generalized potential functions can also be defined for non-gradient dynamical systems. Our method is based on the potential landscape definition from the steady-state distribution, and can be used for a large variety of models. To facilitate simulation and computation, we introduce several novel features, including data structures optimized for batch simulations under varying conditions, an out-of-memory computation tool with integrated hash-based file-saving systems, and an algorithm for efficiently searching the minimum energy path. Using a multistable cell differentiation model as examples, we illustrate how simlandr can be used for model simulation, landscape construction, and barrier height calculation. The simlandr package is available at https://CRAN.R-project.org/package=simlandr, under GPL-3 license."/>
<meta property="og:url" content="https://doi.org/10.32614/RJ-2025-039/"/>
<meta property="og:image" content="https://journal.r-project.org/articles/RJ-2025-039/figures/metaphor.png"/>
<meta property="og:image:width" content="2680"/>
<meta property="og:image:height" content="603"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="The R Journal"/>

<!--  https://dev.twitter.com/cards/types/summary -->
<meta property="twitter:card" content="summary_large_image"/>
<meta property="twitter:title" content="simlandr: Simulation-Based Landscape Construction for Dynamical Systems"/>
<meta property="twitter:description" content="We present the simlandr package for R, which provides a set of tools for constructing potential landscapes for dynamical systems using Monte Carlo simulation. Potential landscapes can be used to quantify the stability of system states. While the canonical form of a potential function is defined for gradient systems, generalized potential functions can also be defined for non-gradient dynamical systems. Our method is based on the potential landscape definition from the steady-state distribution, and can be used for a large variety of models. To facilitate simulation and computation, we introduce several novel features, including data structures optimized for batch simulations under varying conditions, an out-of-memory computation tool with integrated hash-based file-saving systems, and an algorithm for efficiently searching the minimum energy path. Using a multistable cell differentiation model as examples, we illustrate how simlandr can be used for model simulation, landscape construction, and barrier height calculation. The simlandr package is available at https://CRAN.R-project.org/package=simlandr, under GPL-3 license."/>
<meta property="twitter:url" content="https://doi.org/10.32614/RJ-2025-039/"/>
<meta property="twitter:image" content="https://journal.r-project.org/articles/RJ-2025-039/figures/metaphor.png"/>
<meta property="twitter:image:width" content="2680"/>
<meta property="twitter:image:height" content="603"/>

<!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
<meta name="citation_title" content="simlandr: Simulation-Based Landscape Construction for Dynamical Systems"/>
<meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2025-039"/>
<meta name="citation_pdf_url" content="RJ-2025-039.pdf"/>
<meta name="citation_volume" content="17"/>
<meta name="citation_issue" content="4"/>
<meta name="citation_doi" content="10.32614/RJ-2025-039"/>
<meta name="citation_journal_title" content="The R Journal"/>
<meta name="citation_issn" content="2073-4859"/>
<meta name="citation_firstpage" content="173"/>
<meta name="citation_lastpage" content="191"/>
<meta name="citation_fulltext_world_readable" content=""/>
<meta name="citation_online_date" content="2026/02/04"/>
<meta name="citation_publication_date" content="2026/02/04"/>
<meta name="citation_author" content="Jingmeng Cui"/>
<meta name="citation_author_institution" content="University of Groningen Radboud University"/>
<meta name="citation_author" content="Merlijn Olthof"/>
<meta name="citation_author_institution" content="University of Groningen Radboud University"/>
<meta name="citation_author" content="Anna Lichtwarck-Aschoff"/>
<meta name="citation_author_institution" content="University of Groningen Radboud University"/>
<meta name="citation_author" content="Tiejun Li"/>
<meta name="citation_author_institution" content="Peking University"/>
<meta name="citation_author" content="Fred Hasselman"/>
<meta name="citation_author_institution" content="Radboud University"/>
<!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Potential in stochastic differential equations: Novel construction;citation_publisher=IOP Publishing;citation_volume=37;citation_doi=10.1088/0305-4470/37/3/L01;citation_issn=0305-4470;citation_author=P. Ao"/>
  <meta name="citation_reference" content="citation_title=From metaphor to computation: Constructing the potential landscape for multivariate psychological formal models;citation_volume=58;citation_doi=10.1080/00273171.2022.2119927;citation_author=Jingmeng Cui;citation_author=Anna Lichtwarck-Aschoff;citation_author=Merlijn Olthof;citation_author=Tiejun Li;citation_author=Fred Hasselman"/>
  <meta name="citation_reference" content="citation_title=Ordered line integral methods for computing the quasi-potential;citation_volume=75;citation_doi=10.1007/s10915-017-0590-9;citation_issn=1573-7691;citation_author=Daisy Dahiya;citation_author=Maria Cameron"/>
  <meta name="citation_reference" content="citation_title=A note on two problems in connexion with graphs;citation_publisher=Springer;citation_volume=1;citation_doi=10.1007/BF01386390;citation_author=Edsger W Dijkstra"/>
  <meta name="citation_reference" content="citation_title=Multivariate kernel smoothing and its applications;citation_publisher=Chapman and Hall/CRC;citation_doi=10.1201/9780429485572;citation_author=José E. Chacón;citation_author=Tarn Duong"/>
  <meta name="citation_reference" content="citation_title=digest: Create compact hash digests of R objects;citation_doi=10.32614/CRAN.package.digest;citation_author=Dirk Eddelbuettel;citation_author=Antoine Lucas;citation_author=Jarek Tuszynski;citation_author=Henrik Bengtsson;citation_author=Simon Urbanek;citation_author=Mario Frasca;citation_author=Bryan Lewis;citation_author=Murray Stokely;citation_author=Hannes Muehleisen;citation_author=Duncan Murdoch;citation_author=Jim Hester;citation_author=Wush Wu;citation_author=Qiang Kou;citation_author=Thierry Onkelinx;citation_author=Michel Lang;citation_author=Viliam Simko;citation_author=Kurt Hornik;citation_author=Radford Neal;citation_author=Kendon Bell;citation_author=Matthew Queljoe;citation_author=Ion Suruceanu;citation_author=Bill Denney;citation_author=Dirk Schumacher;citation_author=Winston Chang"/>
  <meta name="citation_reference" content="citation_title=Transition-path theory and path-finding algorithms for the study of rare events;citation_volume=61;citation_doi=10.1146/annurev.physchem.040808.090412;citation_issn=0066-426X;citation_author=Weinan E;citation_author=Eric Vanden-Eijnden"/>
  <meta name="citation_reference" content="citation_title=Random perturbations of dynamical systems;citation_publisher=Springer;citation_doi=10.1007/978-3-642-25847-3;citation_author=M. I. Freidlin;citation_author=A. D. Wentzell"/>
  <meta name="citation_reference" content="citation_title=Pathways of maximum likelihood for rare events in nonequilibrium systems: Application to nucleation in the presence of shear;citation_publisher=American Physical Society;citation_volume=100;citation_doi=10.1103/PhysRevLett.100.140601;citation_author=Matthias Heymann;citation_author=Eric Vanden-Eijnden"/>
  <meta name="citation_reference" content="citation_title=Scalable strategies for computing with massive data;citation_volume=55;citation_doi=10.18637/jss.v055.i14;citation_author=Michael J. Kane;citation_author=John Emerson;citation_author=Stephen Weston"/>
  <meta name="citation_reference" content="citation_title=Bayesian estimates of equation system parameters: An application of integration by Monte Carlo;citation_publisher=Wiley, Econometric Society;citation_volume=46;citation_doi=10.2307/1913641;citation_issn=0012-9682;citation_author=T. Kloek;citation_author=H. K. Dijk"/>
  <meta name="citation_reference" content="citation_title=Linking the ball-and-cup analogy and ordination trajectories to describe ecosystem stability, resistance, and resilience;citation_volume=10;citation_doi=10.1002/ecs2.2629;citation_issn=2150-8925;citation_author=Karl A. Lamothe;citation_author=Keith M. Somers;citation_author=Donald A. Jackson"/>
  <meta name="citation_reference" content="citation_title=QPot: An R package for stochastic differential equation quasi-potential analysis;citation_volume=8;citation_doi=10.32614/RJ-2016-031;citation_author=Christopher M. Moore;citation_author=Christopher R. Stieha;citation_author=Ben C. Nolting;citation_author=Maria K. Cameron;citation_author=Karen C. Abbott"/>
  <meta name="citation_reference" content="citation_title=Complexity theory of psychopathology;citation_volume=132;citation_doi=10.1037/abn0000740;citation_author=Merlijn Olthof;citation_author=Fred Hasselman;citation_author=Freek Oude Maatman;citation_author=Anna Maria Theodora Bosman;citation_author=Anna Lichtwarck-Aschoff"/>
  <meta name="citation_reference" content="citation_title=CODA: Convergence diagnosis and output analysis for MCMC;citation_volume=6;citation_author=Martyn Plummer;citation_author=Nicky Best;citation_author=Kate Cowles;citation_author=Karen Vines"/>
  <meta name="citation_reference" content="citation_title=PabRod/rolldown: Post-publication update;citation_doi=10.5281/zenodo.3763038;citation_author=Pablo Rodríguez-Sánchez"/>
  <meta name="citation_reference" content="citation_title=Interactive web-based data visualization with R, plotly, and shiny;citation_publisher=Chapman and Hall/CRC;citation_author=Carson Sievert"/>
  <meta name="citation_reference" content="citation_title=Principles of development and differentiation;citation_publisher=Macmillan;citation_author=Conrad Hal Waddington"/>
  <meta name="citation_reference" content="citation_title=Potential landscape and flux framework of nonequilibrium networks: Robustness, dissipation, and coherence of biochemical oscillations;citation_publisher=National Academy of Sciences;citation_volume=105;citation_doi=10.1073/pnas.0800579105;citation_issn=0027-8424;citation_author=Jin Wang;citation_author=Li Xu;citation_author=Erkang Wang"/>
  <meta name="citation_reference" content="citation_title=Quantifying the Waddington landscape and biological paths for development and differentiation;citation_publisher=National Academy of Sciences;citation_volume=108;citation_doi=10.1073/pnas.1017017108;citation_issn=0027-8424;citation_author=Jin Wang;citation_author=Kun Zhang;citation_author=Li Xu;citation_author=Erkang Wang"/>
  <meta name="citation_reference" content="citation_title=ggplot2: Elegant graphics for data analysis;citation_publisher=Springer-Verlag New York;citation_author=Hadley Wickham"/>
  <meta name="citation_reference" content="citation_title=Improving performance;citation_publisher=Chapman and Hall/CRC;citation_author=Hadley Wickham"/>
  <meta name="citation_reference" content="citation_title=Construction of the landscape for multi-stable systems: Potential landscape, quasi-potential, a-type integral and beyond;citation_publisher=American Institute of Physics;citation_volume=144;citation_doi=10.1063/1.4943096;citation_issn=0021-9606;citation_author=Peijie Zhou;citation_author=Tiejun Li"/>
  <meta name="citation_reference" content="citation_title=Comments on &quot;Climbing Escher’s Stairs: A way to approximate stability landscapes in multidimensional systems&quot;;citation_publisher=arXiv;citation_doi=10.48550/arXiv.2312.09690;citation_author=Jingmeng Cui;citation_author=Anna Lichtwarck-Aschoff;citation_author=Fred Hasselman"/>
  <meta name="citation_reference" content="citation_title=Modeling stem cell fates using non-Markov processes;citation_volume=28;citation_doi=10.1016/j.stem.2021.01.009;citation_author=Patrick S. Stumpf;citation_author=Fumio Arai;citation_author=Ben D. MacArthur"/>
  <meta name="citation_reference" content="citation_title=Performing parallel Monte Carlo and moment equations methods for Itô and Stratonovich stochastic differential systems: R package Sim.DiffProc;citation_volume=96;citation_doi=10.18637/jss.v096.i02;citation_issn=1548-7660;citation_author=Arsalane Chouaib Guidoum;citation_author=Kamal Boukhetala"/>
  <meta name="citation_reference" content="citation_title=Advancing the network theory of mental disorders: A computational model of panic disorder;citation_volume=131;citation_doi=10.1037/rev0000515;citation_author=Donald J. Robinaugh;citation_author=Jonas M. B. Haslbeck;citation_author=Lourens J. Waldorp;citation_author=Jolanda J. Kossakowski;citation_author=Eiko I. Fried;citation_author=Alexander J. Millner;citation_author=Richard J. McNally;citation_author=Oisín Ryan;citation_author=Jill Ron;citation_author=Han L. J. Maas;citation_author=Egbert H. Nes;citation_author=Marten Scheffer;citation_author=Kenneth S. Kendler;citation_author=Denny Borsboom"/>
  <meta name="citation_reference" content="citation_title=A unifying framework for parallel and distributed processing in R using futures;citation_volume=13;citation_doi=10.32614/RJ-2021-048;citation_author=Henrik Bengtsson"/>
  <meta name="citation_reference" content="citation_title=Rare event simulation using Monte Carlo methods;citation_publisher=John Wiley &amp; Sons, Incorporated;citation_doi=10.1002/9780470745403;citation_author=Gerardo Rubino;citation_author=Bruno Tuffin"/>
  <meta name="citation_reference" content="citation_title=Finding transition state and minimum energy path of bistable elastic continua through energy landscape explorations;citation_volume=183;citation_doi=10.1016/j.jmps.2023.105503;citation_issn=0022-5096;citation_author=Guangchao Wan;citation_author=Samuel J. Avis;citation_author=Zizheng Wang;citation_author=Xueju Wang;citation_author=Halim Kusumaatmaja;citation_author=Teng Zhang"/>
  <!--radix_placeholder_rmarkdown_metadata-->

<script type="text/json" id="radix-rmarkdown-metadata">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","date","description","draft","author","type","output","bibliography","date_received","volume","issue","slug","journal","pdf_url","citation_url","doi","creative_commons","packages","CTV","csl","canonical_url"]}},"value":[{"type":"character","attributes":{},"value":["simlandr: Simulation-Based Landscape Construction for Dynamical Systems"]},{"type":"character","attributes":{},"value":["2026-02-04"]},{"type":"character","attributes":{},"value":["We present the simlandr package for R, which provides a set of tools for constructing potential landscapes for dynamical systems using Monte Carlo simulation. Potential landscapes can be used to quantify the stability of system states. While the canonical form of a potential function is defined for gradient systems, generalized potential functions can also be defined for non-gradient dynamical systems. Our method is based on the potential landscape definition from the steady-state distribution, and can be used for a large variety of models. To facilitate simulation and computation, we introduce several novel features, including data structures optimized for batch simulations under varying conditions, an out-of-memory computation tool with integrated hash-based file-saving systems, and an algorithm for efficiently searching the minimum energy path. Using a multistable cell differentiation model as examples, we illustrate how simlandr can be used for model simulation, landscape construction, and barrier height calculation. The simlandr package is available at https://CRAN.R-project.org/package=simlandr, under GPL-3 license."]},{"type":"logical","attributes":{},"value":[false]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","email","orcid_id"]}},"value":[{"type":"character","attributes":{},"value":["Jingmeng Cui"]},{"type":"character","attributes":{},"value":["University of Groningen","Radboud University"]},{"type":"character","attributes":{},"value":["Faculty of Behavioural and Social Sciences, Groningen, the Netherlands","Behavioural Science Institute, Nijmegen, the Netherlands"]},{"type":"character","attributes":{},"value":["jingmeng.cui@rug.nl"]},{"type":"character","attributes":{},"value":["0000-0003-3421-8457"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","email","orcid_id"]}},"value":[{"type":"character","attributes":{},"value":["Merlijn Olthof"]},{"type":"character","attributes":{},"value":["University of Groningen","Radboud University"]},{"type":"character","attributes":{},"value":["Faculty of Behavioural and Social Sciences, Groningen, the Netherlands","Behavioural Science Institute, Nijmegen, the Netherlands"]},{"type":"character","attributes":{},"value":["merlijn.olthof@ru.nl"]},{"type":"character","attributes":{},"value":["0000-0002-5975-6588"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","email","orcid_id"]}},"value":[{"type":"character","attributes":{},"value":["Anna Lichtwarck-Aschoff"]},{"type":"character","attributes":{},"value":["University of Groningen","Radboud University"]},{"type":"character","attributes":{},"value":["Faculty of Behavioural and Social Sciences, Groningen, the Netherlands","Behavioural Science Institute, Nijmegen, the Netherlands"]},{"type":"character","attributes":{},"value":["a.lichtwarck-aschoff@rug.nl"]},{"type":"character","attributes":{},"value":["0000-0002-4365-1538"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","email","orcid_id"]}},"value":[{"type":"character","attributes":{},"value":["Tiejun Li"]},{"type":"character","attributes":{},"value":["Peking University"]},{"type":"character","attributes":{},"value":["\\textit{(corresponding author)} LMAM and School of Mathematical Sciences, Beijing, China"]},{"type":"character","attributes":{},"value":["tieli@pku.edu.cn"]},{"type":"character","attributes":{},"value":["0000-0002-2086-1620"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address","email","orcid_id"]}},"value":[{"type":"character","attributes":{},"value":["Fred Hasselman"]},{"type":"character","attributes":{},"value":["Radboud University"]},{"type":"character","attributes":{},"value":["\\textit{(corresponding author)} Behavioural Science Institute, Nijmegen, the Netherlands"]},{"type":"character","attributes":{},"value":["fred.hasselman@ru.nl"]},{"type":"character","attributes":{},"value":["0000-0003-1384-8361"]}]}]},{"type":"character","attributes":{},"value":["package"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["rjtools::rjournal_pdf_article","distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["toc"]}},"value":[{"type":"logical","attributes":{},"value":[false]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[false]}]}]},{"type":"character","attributes":{},"value":["RJreferences.bib"]},{"type":"character","attributes":{},"value":["2024-10-13"]},{"type":"integer","attributes":{},"value":[17]},{"type":"integer","attributes":{},"value":[4]},{"type":"character","attributes":{},"value":["RJ-2025-039"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"integer","attributes":{},"value":[173]},{"type":"integer","attributes":{},"value":[191]}]},{"type":"character","attributes":{},"value":["RJ-2025-039.pdf"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2025-039"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2025-039"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"character","attributes":{},"value":["simlandr","waydown","Sim.DiffProc","future","coda","bigmemory","digest","ks","ggplot2","plotly"]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["Bayesian","ChemPhys","DifferentialEquations","DynamicVisualizations","Finance","GraphicalModels","HighPerformanceComputing","NetworkAnalysis","Phylogenetics","Psychometrics","Spatial","TeachingStatistics","TimeSeries","WebTechnologies"]},{"type":"character","attributes":{},"value":["/home/mitchell/R/x86_64-pc-linux-gnu-library/4.5/rjtools/rjournal.csl"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2025-039/"]}]}
</script>
<!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["bp/07d724553cd87ff154d1f0d190d891f6.bin","bp/07d724553cd87ff154d1f0d190d891f6.desc","bp/18e668a304b7f129aad27ed028589f7d.bin","bp/18e668a304b7f129aad27ed028589f7d.desc","bp/1c1d941510edff68cc909722c417fe23.bin","bp/1c1d941510edff68cc909722c417fe23.desc","bp/47d347c592dcdbc74dbe951f5ffd0478.bin","bp/47d347c592dcdbc74dbe951f5ffd0478.desc","bp/8af6a534432e217c199a499dd6760d7d.bin","bp/8af6a534432e217c199a499dd6760d7d.desc","bp/92869d244fbdbd41a58ed1632cf4857f.bin","bp/92869d244fbdbd41a58ed1632cf4857f.desc","bp/93d8721a031e7db2a9d5aec57b7933c5.bin","bp/93d8721a031e7db2a9d5aec57b7933c5.desc","bp/a62440cca2f59c0c06d291160c0ea40b.bin","bp/a62440cca2f59c0c06d291160c0ea40b.desc","bp/c5de4a8b1661617893e0d36b246c598d.bin","bp/c5de4a8b1661617893e0d36b246c598d.desc","data/batch_output_gene.RDS","data/batch_output_panic.RDS","data/single_output_gene.RDS","data/single_output_panic.RDS","figures/3dstatic_gene.png","figures/diagram-e1.png","figures/diagram-e2.png","figures/diagram.png","figures/metaphor.png","figures/wang2011.png","RJ-2025-039_files/anchor-4.2.2/anchor.min.js","RJ-2025-039_files/bowser-1.9.3/bowser.min.js","RJ-2025-039_files/distill-2.2.21/template.v2.js","RJ-2025-039_files/figure-html5/3dstaticpanic-1.png","RJ-2025-039_files/figure-html5/bbatch3dgene-1.png","RJ-2025-039_files/figure-html5/bsingle3dgene-1.png","RJ-2025-039_files/figure-html5/converge-gene-1.png","RJ-2025-039_files/figure-html5/converge-panic-1.png","RJ-2025-039_files/figure-html5/lbatch3dpanic-1.png","RJ-2025-039_files/header-attrs-2.30/header-attrs.js","RJ-2025-039_files/jquery-3.6.0/jquery-3.6.0.js","RJ-2025-039_files/jquery-3.6.0/jquery-3.6.0.min.js","RJ-2025-039_files/jquery-3.6.0/jquery-3.6.0.min.map","RJ-2025-039_files/popper-2.6.0/popper.min.js","RJ-2025-039_files/tippy-6.2.7/tippy-bundle.umd.min.js","RJ-2025-039_files/tippy-6.2.7/tippy-light-border.css","RJ-2025-039_files/tippy-6.2.7/tippy.css","RJ-2025-039_files/tippy-6.2.7/tippy.umd.min.js","RJ-2025-039_files/webcomponents-2.0.0/webcomponents.js","RJ-2025-039.log","RJ-2025-039.tex","RJ-2025-039.zip","RJournal.sty","RJreferences.bib","RJwrapper.log","RJwrapper.tex","simlandr-manuscript.Rmd.bak","simlandr-manuscript.tex","supplementary-materials.zip"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
<meta name="distill:offset" content="../.."/>

<script type="application/javascript">

  window.headroom_prevent_pin = false;

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // initialize headroom for banner
    var header = $('header').get(0);
    var headerHeight = header.offsetHeight;
    var headroom = new Headroom(header, {
      tolerance: 5,
      onPin : function() {
        if (window.headroom_prevent_pin) {
          window.headroom_prevent_pin = false;
          headroom.unpin();
        }
      }
    });
    headroom.init();
    if(window.location.hash)
      headroom.unpin();
    $(header).addClass('headroom--transition');

    // offset scroll location for banner on hash change
    // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
    window.addEventListener("hashchange", function(event) {
      window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
    });

    // responsive menu
    $('.distill-site-header').each(function(i, val) {
      var topnav = $(this);
      var toggle = topnav.find('.nav-toggle');
      toggle.on('click', function() {
        topnav.toggleClass('responsive');
      });
    });

    // nav dropdowns
    $('.nav-dropbtn').click(function(e) {
      $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
      $(this).parent().siblings('.nav-dropdown')
         .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $("body").click(function(e){
      $('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $(".nav-dropdown").click(function(e){
      e.stopPropagation();
    });
  });
</script>

<style type="text/css">

/* Theme (user-documented overrideables for nav appearance) */

.distill-site-nav {
  color: rgba(255, 255, 255, 0.8);
  background-color: #0F2E3D;
  font-size: 15px;
  font-weight: 300;
}

.distill-site-nav a {
  color: inherit;
  text-decoration: none;
}

.distill-site-nav a:hover {
  color: white;
}

@media print {
  .distill-site-nav {
    display: none;
  }
}

.distill-site-header {

}

.distill-site-footer {

}


/* Site Header */

.distill-site-header {
  width: 100%;
  box-sizing: border-box;
  z-index: 3;
}

.distill-site-header .nav-left {
  display: inline-block;
  margin-left: 8px;
}

@media screen and (max-width: 768px) {
  .distill-site-header .nav-left {
    margin-left: 0;
  }
}


.distill-site-header .nav-right {
  float: right;
  margin-right: 8px;
}

.distill-site-header a,
.distill-site-header .title {
  display: inline-block;
  text-align: center;
  padding: 14px 10px 14px 10px;
}

.distill-site-header .title {
  font-size: 18px;
  min-width: 150px;
}

.distill-site-header .logo {
  padding: 0;
}

.distill-site-header .logo img {
  display: none;
  max-height: 20px;
  width: auto;
  margin-bottom: -4px;
}

.distill-site-header .nav-image img {
  max-height: 18px;
  width: auto;
  display: inline-block;
  margin-bottom: -3px;
}



@media screen and (min-width: 1000px) {
  .distill-site-header .logo img {
    display: inline-block;
  }
  .distill-site-header .nav-left {
    margin-left: 20px;
  }
  .distill-site-header .nav-right {
    margin-right: 20px;
  }
  .distill-site-header .title {
    padding-left: 12px;
  }
}


.distill-site-header .nav-toggle {
  display: none;
}

.nav-dropdown {
  display: inline-block;
  position: relative;
}

.nav-dropdown .nav-dropbtn {
  border: none;
  outline: none;
  color: rgba(255, 255, 255, 0.8);
  padding: 16px 10px;
  background-color: transparent;
  font-family: inherit;
  font-size: inherit;
  font-weight: inherit;
  margin: 0;
  margin-top: 1px;
  z-index: 2;
}

.nav-dropdown-content {
  display: none;
  position: absolute;
  background-color: white;
  min-width: 200px;
  border: 1px solid rgba(0,0,0,0.15);
  border-radius: 4px;
  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
  z-index: 1;
  margin-top: 2px;
  white-space: nowrap;
  padding-top: 4px;
  padding-bottom: 4px;
}

.nav-dropdown-content hr {
  margin-top: 4px;
  margin-bottom: 4px;
  border: none;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.nav-dropdown-active {
  display: block;
}

.nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
  color: black;
  padding: 6px 24px;
  text-decoration: none;
  display: block;
  text-align: left;
}

.nav-dropdown-content .nav-dropdown-header {
  display: block;
  padding: 5px 24px;
  padding-bottom: 0;
  text-transform: uppercase;
  font-size: 14px;
  color: #999999;
  white-space: nowrap;
}

.nav-dropdown:hover .nav-dropbtn {
  color: white;
}

.nav-dropdown-content a:hover {
  background-color: #ddd;
  color: black;
}

.nav-right .nav-dropdown-content {
  margin-left: -45%;
  right: 0;
}

@media screen and (max-width: 768px) {
  .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
  .distill-site-header a.nav-toggle {
    float: right;
    display: block;
  }
  .distill-site-header .title {
    margin-left: 0;
  }
  .distill-site-header .nav-right {
    margin-right: 0;
  }
  .distill-site-header {
    overflow: hidden;
  }
  .nav-right .nav-dropdown-content {
    margin-left: 0;
  }
}


@media screen and (max-width: 768px) {
  .distill-site-header.responsive {position: relative; min-height: 500px; }
  .distill-site-header.responsive a.nav-toggle {
    position: absolute;
    right: 0;
    top: 0;
  }
  .distill-site-header.responsive a,
  .distill-site-header.responsive .nav-dropdown {
    display: block;
    text-align: left;
  }
  .distill-site-header.responsive .nav-left,
  .distill-site-header.responsive .nav-right {
    width: 100%;
  }
  .distill-site-header.responsive .nav-dropdown {float: none;}
  .distill-site-header.responsive .nav-dropdown-content {position: relative;}
  .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
    display: block;
    width: 100%;
    text-align: left;
  }
}

/* Site Footer */

.distill-site-footer {
  width: 100%;
  overflow: hidden;
  box-sizing: border-box;
  z-index: 3;
  margin-top: 30px;
  padding-top: 30px;
  padding-bottom: 30px;
  text-align: center;
}

/* Headroom */

d-title {
  padding-top: 6rem;
}

@media print {
  d-title {
    padding-top: 4rem;
  }
}

.headroom {
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
}

.headroom--transition {
  transition: all .4s ease-in-out;
}

.headroom--unpinned {
  top: -100px;
}

.headroom--pinned {
  top: 0;
}

/* adjust viewport for navbar height */
/* helps vertically center bootstrap (non-distill) content */
.min-vh-100 {
  min-height: calc(100vh - 100px) !important;
}

</style>

<script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="../../site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"/>
<link href="../../site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"/>
<script src="../../site_libs/headroom-0.9.4/headroom.min.js"></script>
<!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

<style type="text/css">

body {
  background-color: white;
}

.pandoc-table {
  width: 100%;
}

.pandoc-table>caption {
  margin-bottom: 10px;
}

.pandoc-table th:not([align]) {
  text-align: left;
}

.pagedtable-footer {
  font-size: 15px;
}

d-byline .byline {
  grid-template-columns: 2fr 2fr;
}

d-byline .byline h3 {
  margin-block-start: 1.5em;
}

d-byline .byline .authors-affiliations h3 {
  margin-block-start: 0.5em;
}

.authors-affiliations .orcid-id {
  width: 16px;
  height:16px;
  margin-left: 4px;
  margin-right: 4px;
  vertical-align: middle;
  padding-bottom: 2px;
}

d-title .dt-tags {
  margin-top: 1em;
  grid-column: text;
}

.dt-tags .dt-tag {
  text-decoration: none;
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0em 0.4em;
  margin-right: 0.5em;
  margin-bottom: 0.4em;
  font-size: 70%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

d-article table.gt_table td,
d-article table.gt_table th {
  border-bottom: none;
  font-size: 100%;
}

.html-widget {
  margin-bottom: 2.0em;
}

.l-screen-inset {
  padding-right: 16px;
}

.l-screen .caption {
  margin-left: 10px;
}

.shaded {
  background: rgb(247, 247, 247);
  padding-top: 20px;
  padding-bottom: 20px;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .html-widget {
  margin-bottom: 0;
  border: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .shaded-content {
  background: white;
}

.text-output {
  margin-top: 0;
  line-height: 1.5em;
}

.hidden {
  display: none !important;
}

hr.section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  margin: 0px;
}


d-byline {
  border-top: none;
}

d-article {
  padding-top: 2.5rem;
  padding-bottom: 30px;
  border-top: none;
}

d-appendix {
  padding-top: 30px;
}

d-article>p>img {
  width: 100%;
}

d-article h2 {
  margin: 1rem 0 1.5rem 0;
}

d-article h3 {
  margin-top: 1.5rem;
}

d-article iframe {
  border: 1px solid rgba(0, 0, 0, 0.1);
  margin-bottom: 2.0em;
  width: 100%;
}

/* Tweak code blocks */

d-article div.sourceCode code,
d-article pre code {
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
}

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: auto;
}

d-article div.sourceCode {
  background-color: white;
}

d-article div.sourceCode pre {
  padding-left: 10px;
  font-size: 12px;
  border-left: 2px solid rgba(0,0,0,0.1);
}

d-article pre {
  font-size: 12px;
  color: black;
  background: none;
  margin-top: 0;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

d-article pre a {
  border-bottom: none;
}

d-article pre a:hover {
  border-bottom: none;
  text-decoration: underline;
}

d-article details {
  grid-column: text;
  margin-bottom: 0.8em;
}

@media(min-width: 768px) {

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: visible !important;
}

d-article div.sourceCode pre {
  padding-left: 18px;
  font-size: 14px;
}

/* tweak for Pandoc numbered line within distill */
d-article pre.numberSource code > span {
    left: -2em;
}

d-article pre {
  font-size: 14px;
}

}

figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

/* CSS for d-contents */

.d-contents {
  grid-column: text;
  color: rgba(0,0,0,0.8);
  font-size: 0.9em;
  padding-bottom: 1em;
  margin-bottom: 1em;
  padding-bottom: 0.5em;
  margin-bottom: 1em;
  padding-left: 0.25em;
  justify-self: start;
}

@media(min-width: 1000px) {
  .d-contents.d-contents-float {
    height: 0;
    grid-column-start: 1;
    grid-column-end: 4;
    justify-self: center;
    padding-right: 3em;
    padding-left: 2em;
  }
}

.d-contents nav h3 {
  font-size: 18px;
  margin-top: 0;
  margin-bottom: 1em;
}

.d-contents li {
  list-style-type: none
}

.d-contents nav > ul {
  padding-left: 0;
}

.d-contents ul {
  padding-left: 1em
}

.d-contents nav ul li {
  margin-top: 0.6em;
  margin-bottom: 0.2em;
}

.d-contents nav a {
  font-size: 13px;
  border-bottom: none;
  text-decoration: none
  color: rgba(0, 0, 0, 0.8);
}

.d-contents nav a:hover {
  text-decoration: underline solid rgba(0, 0, 0, 0.6)
}

.d-contents nav > ul > li > a {
  font-weight: 600;
}

.d-contents nav > ul > li > ul {
  font-weight: inherit;
}

.d-contents nav > ul > li > ul > li {
  margin-top: 0.2em;
}


.d-contents nav ul {
  margin-top: 0;
  margin-bottom: 0.25em;
}

.d-article-with-toc h2:nth-child(2) {
  margin-top: 0;
}


/* Figure */

.figure {
  position: relative;
  margin-bottom: 2.5em;
  margin-top: 1.5em;
}

.figure .caption {
  color: rgba(0, 0, 0, 0.6);
  font-size: 12px;
  line-height: 1.5em;
}

.figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

.figure .caption a {
  color: rgba(0, 0, 0, 0.6);
}

.figure .caption b,
.figure .caption strong, {
  font-weight: 600;
  color: rgba(0, 0, 0, 1.0);
}

/* Citations */

d-article .citation {
  color: inherit;
  cursor: inherit;
}

div.hanging-indent{
  margin-left: 1em; text-indent: -1em;
}

/* Citation hover box */

.tippy-box[data-theme~=light-border] {
  background-color: rgba(250, 250, 250, 0.95);
}

.tippy-content > p {
  margin-bottom: 0;
  padding: 2px;
}


/* Tweak 1000px media break to show more text */

@media(min-width: 1000px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 16px;
  }

  .grid {
    grid-column-gap: 16px;
  }

  d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
  }
  figure .caption, .figure .caption, figure figcaption {
    font-size: 13px;
  }
}

@media(min-width: 1180px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 32px;
  }

  .grid {
    grid-column-gap: 32px;
  }
}


/* Get the citation styles for the appendix (not auto-injected on render since
   we do our own rendering of the citation appendix) */

d-appendix .citation-appendix,
.d-appendix .citation-appendix {
  font-size: 11px;
  line-height: 15px;
  border-left: 1px solid rgba(0, 0, 0, 0.1);
  padding-left: 18px;
  border: 1px solid rgba(0,0,0,0.1);
  background: rgba(0, 0, 0, 0.02);
  padding: 10px 18px;
  border-radius: 3px;
  color: rgba(150, 150, 150, 1);
  overflow: hidden;
  margin-top: -12px;
  white-space: pre-wrap;
  word-wrap: break-word;
}

/* Include appendix styles here so they can be overridden */

d-appendix {
  contain: layout style;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-top: 60px;
  margin-bottom: 0;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  color: rgba(0,0,0,0.5);
  padding-top: 60px;
  padding-bottom: 48px;
}

d-appendix h3 {
  grid-column: page-start / text-start;
  font-size: 15px;
  font-weight: 500;
  margin-top: 1em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.65);
}

d-appendix h3 + * {
  margin-top: 1em;
}

d-appendix ol {
  padding: 0 0 0 15px;
}

@media (min-width: 768px) {
  d-appendix ol {
    padding: 0 0 0 30px;
    margin-left: -30px;
  }
}

d-appendix li {
  margin-bottom: 1em;
}

d-appendix a {
  color: rgba(0, 0, 0, 0.6);
}

d-appendix > * {
  grid-column: text;
}

d-appendix > d-footnote-list,
d-appendix > d-citation-list,
d-appendix > distill-appendix {
  grid-column: screen;
}

/* Include footnote styles here so they can be overridden */

d-footnote-list {
  contain: layout style;
}

d-footnote-list > * {
  grid-column: text;
}

d-footnote-list a.footnote-backlink {
  color: rgba(0,0,0,0.3);
  padding-left: 0.5em;
}



/* Anchor.js */

.anchorjs-link {
  /*transition: all .25s linear; */
  text-decoration: none;
  border-bottom: none;
}
*:hover > .anchorjs-link {
  margin-left: -1.125em !important;
  text-decoration: none;
  border-bottom: none;
}

/* Social footer */

.social_footer {
  margin-top: 30px;
  margin-bottom: 0;
  color: rgba(0,0,0,0.67);
}

.disqus-comments {
  margin-right: 30px;
}

.disqus-comment-count {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  cursor: pointer;
}

#disqus_thread {
  margin-top: 30px;
}

.article-sharing a {
  border-bottom: none;
  margin-right: 8px;
}

.article-sharing a:hover {
  border-bottom: none;
}

.sidebar-section.subscribe {
  font-size: 12px;
  line-height: 1.6em;
}

.subscribe p {
  margin-bottom: 0.5em;
}


.article-footer .subscribe {
  font-size: 15px;
  margin-top: 45px;
}


.sidebar-section.custom {
  font-size: 12px;
  line-height: 1.6em;
}

.custom p {
  margin-bottom: 0.5em;
}

/* Styles for listing layout (hide title) */
.layout-listing d-title, .layout-listing .d-title {
  display: none;
}

/* Styles for posts lists (not auto-injected) */


.posts-with-sidebar {
  padding-left: 45px;
  padding-right: 45px;
}

.posts-list .description h2,
.posts-list .description p {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

.posts-list .description h2 {
  font-weight: 700;
  border-bottom: none;
  padding-bottom: 0;
}

.posts-list h2.post-tag {
  border-bottom: 1px solid rgba(0, 0, 0, 0.2);
  padding-bottom: 12px;
}
.posts-list {
  margin-top: 60px;
  margin-bottom: 24px;
}

.posts-list .post-preview {
  text-decoration: none;
  overflow: hidden;
  display: block;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  padding: 24px 0;
}

.post-preview-last {
  border-bottom: none !important;
}

.posts-list .posts-list-caption {
  grid-column: screen;
  font-weight: 400;
}

.posts-list .post-preview h2 {
  margin: 0 0 6px 0;
  line-height: 1.2em;
  font-style: normal;
  font-size: 24px;
}

.posts-list .post-preview p {
  margin: 0 0 12px 0;
  line-height: 1.4em;
  font-size: 16px;
}

.posts-list .post-preview .thumbnail {
  box-sizing: border-box;
  margin-bottom: 24px;
  position: relative;
  max-width: 500px;
}
.posts-list .post-preview img {
  width: 100%;
  display: block;
}

.posts-list .metadata {
  font-size: 12px;
  line-height: 1.4em;
  margin-bottom: 18px;
}

.posts-list .metadata > * {
  display: inline-block;
}

.posts-list .metadata .publishedDate {
  margin-right: 2em;
}

.posts-list .metadata .dt-authors {
  display: block;
  margin-top: 0.3em;
  margin-right: 2em;
}

.posts-list .dt-tags {
  display: block;
  line-height: 1em;
}

.posts-list .dt-tags .dt-tag {
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0.3em 0.4em;
  margin-right: 0.2em;
  margin-bottom: 0.4em;
  font-size: 60%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

.posts-list img {
  opacity: 1;
}

.posts-list img[data-src] {
  opacity: 0;
}

.posts-more {
  clear: both;
}


.posts-sidebar {
  font-size: 16px;
}

.posts-sidebar h3 {
  font-size: 16px;
  margin-top: 0;
  margin-bottom: 0.5em;
  font-weight: 400;
  text-transform: uppercase;
}

.sidebar-section {
  margin-bottom: 30px;
}

.categories ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.categories li {
  color: rgba(0, 0, 0, 0.8);
  margin-bottom: 0;
}

.categories li>a {
  border-bottom: none;
}

.categories li>a:hover {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
}

.categories .active {
  font-weight: 600;
}

.categories .category-count {
  color: rgba(0, 0, 0, 0.4);
}


@media(min-width: 768px) {
  .posts-list .post-preview h2 {
    font-size: 26px;
  }
  .posts-list .post-preview .thumbnail {
    float: right;
    width: 30%;
    margin-bottom: 0;
  }
  .posts-list .post-preview .description {
    float: left;
    width: 45%;
  }
  .posts-list .post-preview .metadata {
    float: left;
    width: 20%;
    margin-top: 8px;
  }
  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.5em;
    font-size: 16px;
  }
  .posts-with-sidebar .posts-list {
    float: left;
    width: 75%;
  }
  .posts-with-sidebar .posts-sidebar {
    float: right;
    width: 20%;
    margin-top: 60px;
    padding-top: 24px;
    padding-bottom: 24px;
  }
}


/* Improve display for browsers without grid (IE/Edge <= 15) */

.downlevel {
  line-height: 1.6em;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  margin: 0;
}

.downlevel .d-title {
  padding-top: 6rem;
  padding-bottom: 1.5rem;
}

.downlevel .d-title h1 {
  font-size: 50px;
  font-weight: 700;
  line-height: 1.1em;
  margin: 0 0 0.5rem;
}

.downlevel .d-title p {
  font-weight: 300;
  font-size: 1.2rem;
  line-height: 1.55em;
  margin-top: 0;
}

.downlevel .d-byline {
  padding-top: 0.8em;
  padding-bottom: 0.8em;
  font-size: 0.8rem;
  line-height: 1.8em;
}

.downlevel .section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
}

.downlevel .d-article {
  font-size: 1.06rem;
  line-height: 1.7em;
  padding-top: 1rem;
  padding-bottom: 2rem;
}


.downlevel .d-appendix {
  padding-left: 0;
  padding-right: 0;
  max-width: none;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.5);
  padding-top: 40px;
  padding-bottom: 48px;
}

.downlevel .footnotes ol {
  padding-left: 13px;
}

.downlevel .base-grid,
.downlevel .distill-header,
.downlevel .d-title,
.downlevel .d-abstract,
.downlevel .d-article,
.downlevel .d-appendix,
.downlevel .distill-appendix,
.downlevel .d-byline,
.downlevel .d-footnote-list,
.downlevel .d-citation-list,
.downlevel .distill-footer,
.downlevel .appendix-bottom,
.downlevel .posts-container {
  padding-left: 40px;
  padding-right: 40px;
}

@media(min-width: 768px) {
  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
  padding-left: 150px;
  padding-right: 150px;
  max-width: 900px;
}
}

.downlevel pre code {
  display: block;
  border-left: 2px solid rgba(0, 0, 0, .1);
  padding: 0 0 0 20px;
  font-size: 14px;
}

.downlevel code, .downlevel pre {
  color: black;
  background: none;
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

.downlevel .posts-list .post-preview {
  color: inherit;
}



</style>

<script type="application/javascript">

function is_downlevel_browser() {
  if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                 window.navigator.userAgent)) {
    return true;
  } else {
    return window.load_distill_framework === undefined;
  }
}

// show body when load is complete
function on_load_complete() {

  // add anchors
  if (window.anchors) {
    window.anchors.options.placement = 'left';
    window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
  }


  // set body to visible
  document.body.style.visibility = 'visible';

  // force redraw for leaflet widgets
  if (window.HTMLWidgets) {
    var maps = window.HTMLWidgets.findAll(".leaflet");
    $.each(maps, function(i, el) {
      var map = this.getMap();
      map.invalidateSize();
      map.eachLayer(function(layer) {
        if (layer instanceof L.TileLayer)
          layer.redraw();
      });
    });
  }

  // trigger 'shown' so htmlwidgets resize
  $('d-article').trigger('shown');
}

function init_distill() {

  init_common();

  // create front matter
  var front_matter = $('<d-front-matter></d-front-matter>');
  $('#distill-front-matter').wrap(front_matter);

  // create d-title
  $('.d-title').changeElementType('d-title');

  // separator
  var separator = '<hr class="section-separator" style="clear: both"/>';
  // prepend separator above appendix
  $('.d-byline').before(separator);
  $('.d-article').before(separator);

  // create d-byline
  var byline = $('<d-byline></d-byline>');
  $('.d-byline').replaceWith(byline);

  // create d-article
  var article = $('<d-article></d-article>');
  $('.d-article').wrap(article).children().unwrap();

  // move posts container into article
  $('.posts-container').appendTo($('d-article'));

  // create d-appendix
  $('.d-appendix').changeElementType('d-appendix');

  // flag indicating that we have appendix items
  var appendix = $('.appendix-bottom').children('h3').length > 0;

  // replace footnotes with <d-footnote>
  $('.footnote-ref').each(function(i, val) {
    appendix = true;
    var href = $(this).attr('href');
    var id = href.replace('#', '');
    var fn = $('#' + id);
    var fn_p = $('#' + id + '>p');
    fn_p.find('.footnote-back').remove();
    var text = fn_p.html();
    var dtfn = $('<d-footnote></d-footnote>');
    dtfn.html(text);
    $(this).replaceWith(dtfn);
  });
  // remove footnotes
  $('.footnotes').remove();

  // move refs into #references-listing
  $('#references-listing').replaceWith($('#refs'));

  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    var id = $(this).attr('id');
    $('.d-contents a[href="#' + id + '"]').parent().remove();
    appendix = true;
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
  });

  // show d-appendix if we have appendix content
  $("d-appendix").css('display', appendix ? 'grid' : 'none');

  // localize layout chunks to just output
  $('.layout-chunk').each(function(i, val) {

    // capture layout
    var layout = $(this).attr('data-layout');

    // apply layout to markdown level block elements
    var elements = $(this).children().not('details, div.sourceCode, pre, script');
    elements.each(function(i, el) {
      var layout_div = $('<div class="' + layout + '"></div>');
      if (layout_div.hasClass('shaded')) {
        var shaded_content = $('<div class="shaded-content"></div>');
        $(this).wrap(shaded_content);
        $(this).parent().wrap(layout_div);
      } else {
        $(this).wrap(layout_div);
      }
    });


    // unwrap the layout-chunk div
    $(this).children().unwrap();
  });

  // remove code block used to force  highlighting css
  $('.distill-force-highlighting-css').parent().remove();

  // remove empty line numbers inserted by pandoc when using a
  // custom syntax highlighting theme, except when numbering line
  // in code chunk
  $('pre:not(.numberLines) code.sourceCode a:empty').remove();

  // load distill framework
  load_distill_framework();

  // wait for window.distillRunlevel == 4 to do post processing
  function distill_post_process() {

    if (!window.distillRunlevel || window.distillRunlevel < 4)
      return;

    // hide author/affiliations entirely if we have no authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;
    if (!have_authors)
      $('d-byline').addClass('hidden');

    // article with toc class
    $('.d-contents').parent().addClass('d-article-with-toc');

    // strip links that point to #
    $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

    // add orcid ids
    $('.authors-affiliations').find('.author').each(function(i, el) {
      var orcid_id = front_matter.authors[i].orcidID;
      var author_name = front_matter.authors[i].author
      if (orcid_id) {
        var a = $('<a></a>');
        a.attr('href', 'https://orcid.org/' + orcid_id);
        var img = $('<img></img>');
        img.addClass('orcid-id');
        img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
        img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
        a.append(img);
        $(this).append(a);
      }
    });

    // hide elements of author/affiliations grid that have no value
    function hide_byline_column(caption) {
      $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
    }

    // affiliations
    var have_affiliations = false;
    for (var i = 0; i<front_matter.authors.length; ++i) {
      var author = front_matter.authors[i];
      if (author.affiliation !== "&nbsp;") {
        have_affiliations = true;
        break;
      }
    }
    if (!have_affiliations)
      $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

    // published date
    if (!front_matter.publishedDate)
      hide_byline_column("Published");

    // document object identifier
    var doi = $('d-byline').find('h3:contains("DOI")');
    var doi_p = doi.next().empty();
    if (!front_matter.doi) {
      // if we have a citation and valid citationText then link to that
      if ($('#citation').length > 0 && front_matter.citationText) {
        doi.html('Citation');
        $('<a href="#citation"></a>')
          .text(front_matter.citationText)
          .appendTo(doi_p);
      } else {
        hide_byline_column("DOI");
      }
    } else {
      $('<a></a>')
         .attr('href', "https://doi.org/" + front_matter.doi)
         .html(front_matter.doi)
         .appendTo(doi_p);
    }

     // change plural form of authors/affiliations
    if (front_matter.authors.length === 1) {
      var grid = $('.authors-affiliations');
      grid.children('h3:contains("Authors")').text('Author');
      grid.children('h3:contains("Affiliations")').text('Affiliation');
    }

    // remove d-appendix and d-footnote-list local styles
    $('d-appendix > style:first-child').remove();
    $('d-footnote-list > style:first-child').remove();

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // hoverable references
    $('span.citation[data-cites]').each(function() {
      const citeChild = $(this).children()[0]
      // Do not process if @xyz has been used without escaping and without bibliography activated
      // https://github.com/rstudio/distill/issues/466
      if (citeChild === undefined) return true

      if (citeChild.nodeName == "D-FOOTNOTE") {
        var fn = citeChild
        $(this).html(fn.shadowRoot.querySelector("sup"))
        $(this).id = fn.id
        fn.remove()
      }
      var refs = $(this).attr('data-cites').split(" ");
      var refHtml = refs.map(function(ref) {
        // Could use CSS.escape too here, we insure backward compatibility in navigator
        return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
      }).join("\n");
      window.tippy(this, {
        allowHTML: true,
        content: refHtml,
        maxWidth: 500,
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start'
      });
    });

    // fix footnotes in tables (#411)
    // replacing broken distill.pub feature
    $('table d-footnote').each(function() {
      // we replace internal showAtNode methode which is triggered when hovering a footnote
      this.hoverBox.showAtNode = function(node) {
        // ported from https://github.com/distillpub/template/pull/105/files
        calcOffset = function(elem) {
            let x = elem.offsetLeft;
            let y = elem.offsetTop;
            // Traverse upwards until an `absolute` element is found or `elem`
            // becomes null.
            while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                x += elem.offsetLeft;
                y += elem.offsetTop;
            }

            return { left: x, top: y };
        }
        // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
        const bbox = node.getBoundingClientRect();
        const offset = calcOffset(node);
        this.show([offset.left + bbox.width, offset.top + bbox.height]);
      }
    })

    // clear polling timer
    clearInterval(tid);

    // show body now that everything is ready
    on_load_complete();
  }

  var tid = setInterval(distill_post_process, 50);
  distill_post_process();

}

function init_downlevel() {

  init_common();

   // insert hr after d-title
  $('.d-title').after($('<hr class="section-separator"/>'));

  // check if we have authors
  var front_matter = JSON.parse($("#distill-front-matter").html());
  var have_authors = front_matter.authors && front_matter.authors.length > 0;

  // manage byline/border
  if (!have_authors)
    $('.d-byline').remove();
  $('.d-byline').after($('<hr class="section-separator"/>'));
  $('.d-byline a').remove();

  // remove toc
  $('.d-contents').remove();

  // move appendix elements
  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
  });


  // inject headers into references and footnotes
  var refs_header = $('<h3></h3>');
  refs_header.text('References');
  $('#refs').prepend(refs_header);

  var footnotes_header = $('<h3></h3');
  footnotes_header.text('Footnotes');
  $('.footnotes').children('hr').first().replaceWith(footnotes_header);

  // move appendix-bottom entries to the bottom
  $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
  $('.appendix-bottom').remove();

  // remove appendix if it's empty
  if ($('.d-appendix').children().length === 0)
    $('.d-appendix').remove();

  // prepend separator above appendix
  $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

  // trim code
  $('pre>code').each(function(i, val) {
    $(this).html($.trim($(this).html()));
  });

  // move posts-container right before article
  $('.posts-container').insertBefore($('.d-article'));

  $('body').addClass('downlevel');

  on_load_complete();
}


function init_common() {

  // jquery plugin to change element types
  (function($) {
    $.fn.changeElementType = function(newType) {
      var attrs = {};

      $.each(this[0].attributes, function(idx, attr) {
        attrs[attr.nodeName] = attr.nodeValue;
      });

      this.replaceWith(function() {
        return $("<" + newType + "/>", attrs).append($(this).contents());
      });
    };
  })(jQuery);

  // prevent underline for linked images
  $('a > img').parent().css({'border-bottom' : 'none'});

  // mark non-body figures created by knitr chunks as 100% width
  $('.layout-chunk').each(function(i, val) {
    var figures = $(this).find('img, .html-widget');
    // ignore leaflet img layers (#106)
    figures = figures.filter(':not(img[class*="leaflet"])')
    if ($(this).attr('data-layout') !== "l-body") {
      figures.css('width', '100%');
    } else {
      figures.css('max-width', '100%');
      figures.filter("[width]").each(function(i, val) {
        var fig = $(this);
        fig.css('width', fig.attr('width') + 'px');
      });

    }
  });

  // auto-append index.html to post-preview links in file: protocol
  // and in rstudio ide preview
  $('.post-preview').each(function(i, val) {
    if (window.location.protocol === "file:")
      $(this).attr('href', $(this).attr('href') + "index.html");
  });

  // get rid of index.html references in header
  if (window.location.protocol !== "file:") {
    $('.distill-site-header a[href]').each(function(i,val) {
      $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
    });
  }

  // add class to pandoc style tables
  $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
  $('.kable-table').children('table').addClass('pandoc-table');

  // add figcaption style to table captions
  $('caption').parent('table').addClass("figcaption");

  // initialize posts list
  if (window.init_posts_list)
    window.init_posts_list();

  // implmement disqus comment link
  $('.disqus-comment-count').click(function() {
    window.headroom_prevent_pin = true;
    $('#disqus_thread').toggleClass('hidden');
    if (!$('#disqus_thread').hasClass('hidden')) {
      var offset = $(this).offset();
      $(window).resize();
      $('html, body').animate({
        scrollTop: offset.top - 35
      });
    }
  });
}

document.addEventListener('DOMContentLoaded', function() {
  if (is_downlevel_browser())
    init_downlevel();
  else
    window.addEventListener('WebComponentsReady', init_distill);
});

</script>

<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.5);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

.distill-site-footer {
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */
</style>
<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.7);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #ffffff;
}

.distill-site-footer {
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #fafafa;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */

.nav-right > a {
  text-transform: uppercase;
}

d-title h1, d-title p, d-title figure,
d-abstract p, d-abstract b {
  grid-column: page;
}

.rj-blue {
  color: #2467bb;
}

ul li {
  line-height: 1.6;
  margin-top: 0em;
  margin-bottom: 0em;
}</style>
<style type="text/css">
/* base style */

/* FONT FAMILIES */

:root {
  --heading-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  --mono-default: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  --body-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

body,
.posts-list .post-preview p,
.posts-list .description p {
  font-family: var(--body-font), var(--body-default);
}

h1, h2, h3, h4, h5, h6,
.posts-list .post-preview h2,
.posts-list .description h2 {
  font-family: var(--heading-font), var(--heading-default);
}

d-article div.sourceCode code,
d-article pre code {
  font-family: var(--mono-font), var(--mono-default);
}


/*-- TITLE --*/
d-title h1,
.posts-list > h1 {
  color: var(--title-color, black);
}

d-title h1 {
  font-size: var(--title-size, 50px);
}

/*-- HEADERS --*/
d-article h1,
d-article h2,
d-article h3,
d-article h4,
d-article h5,
d-article h6 {
  color: var(--header-color, rgba(0, 0, 0, 0.8));
}

/*-- BODY --*/
d-article > p,  /* only text inside of <p> tags */
d-article > ul, /* lists */
d-article > ol {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
  font-size: var(--body-size, 1.06rem);
}


/*-- CODE --*/
d-article div.sourceCode code,
d-article pre code {
  font-size: var(--code-size, 14px);
}

/*-- ASIDE --*/
d-article aside {
  font-size: var(--aside-size, 12px);
  color: var(--aside-color, rgba(0, 0, 0, 0.6));
}

/*-- FIGURE CAPTIONS --*/
figure .caption,
figure figcaption,
.figure .caption {
  font-size: var(--fig-cap-size, 13px);
  color: var(--fig-cap-color, rgba(0, 0, 0, 0.6));
}

/*-- METADATA --*/
d-byline h3 {
  font-size: var(--heading-size, 0.6rem);
  color: var(--heading-color, rgba(0, 0, 0, 0.5));
}

d-byline {
  font-size: var(--body-size, 0.8rem);
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

d-byline a,
d-article d-byline a {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

/*-- TABLE OF CONTENTS --*/
.d-contents nav h3 {
  font-size: var(--heading-size, 18px);
}

.d-contents nav a {
  font-size: var(--contents-size, 13px);
}

/*-- APPENDIX --*/
d-appendix h3 {
  font-size: var(--heading-size, 15px);
  color: var(--heading-color, rgba(0, 0, 0, 0.65));
}

d-appendix {
  font-size: var(--text-size, 0.8em);
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

d-appendix d-footnote-list a.footnote-backlink {
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

/*-- WEBSITE HEADER + FOOTER --*/
.distill-site-header .title {
  font-size: var(--title-size, 18px);
  font-family: var(--navbar-font), var(--heading-default);
}

.distill-site-header a,
.nav-dropdown .nav-dropbtn {
  font-family: var(--navbar-font), var(--heading-default);
}

.nav-dropdown .nav-dropbtn {
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  font-size: var(--text-size, 15px);
}

.distill-site-header a:hover,
.nav-dropdown:hover .nav-dropbtn {
  color: var(--hover-color, white);
}

.distill-site-header {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer a:hover {
  color: var(--hover-color, white);
}</style>
<!--/radix_placeholder_distill-->
  <script src="../../site_libs/header-attrs-2.30/header-attrs.js"></script>
  <script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="../../site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="../../site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="../../site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="../../site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="../../site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="../../site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="../../site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="../../site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
<!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Mathjax config (add automatic linebreaks when supported)
      // MathJax = {
      //    tex: {
      //        inlineMath: [['$', '$'], ['\\(', '\\)']],
      //        displayMath: [['$$', '$$'], ['\\[', '\\]']],
      //        tags: 'ams',
      //        multline: true,
      //    },
      //    options: {
      //        linebreaks: { automatic: true },
      //    },
      // };

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      
      // Conditionally add supplementary files button
      if (document.getElementById('supplementary-materials')) {
        // create element safely
        const btn_suppl = document.createElement('a');
        btn_suppl.href = slug + '.zip';
        btn_suppl.className = 'dt-tag';
        btn_suppl.innerHTML = '<i class="fas fa-file-zipper"></i> Supplement';
        buttons.append(btn_suppl);
      }

      // adds Abstract: in front of the first <p> in the title section --
      // unless it happens to be the subtitle (FIXME: this is a bad hack - can't distill do this?)
      var tpar = $("d-title p:not(:empty)").filter(function() {
        return !$(this).hasClass("subtitle");
      }).first();
      if (tpar) {
        const abstract = $('<d-abstract>')
        abstract.append('<b>Abstract:</b><br>')
        abstract.append(tpar) // Move description to d-abstract
        $("d-title p:empty").remove() // Remove empty paragraphs after title
        abstract.append(buttons)
        abstract.insertAfter($('d-title')) // Add abstract section after title */
      }

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
    
    d-appendix .citation-appendix,
    .d-appendix .citation-appendix {
      color: rgb(60, 60, 60);
    }

    d-article h2 {
      border-bottom: 0px solid rgba(0, 0, 0, 0.1);
      padding-bottom: 0rem;
    }
    d-article h3 {
      font-size: 20px;
    }
    d-article h4 {
      font-size: 18px;
      text-transform: none;
    }

    @media (min-width: 1024px) {
      d-article h2 {
        font-size: 32px;
      }
      d-article h3 {
        font-size: 24px;
      }
      d-article h4 {
        font-size: 20px;
      }
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"simlandr: Simulation-Based Landscape Construction for Dynamical Systems","description":"We present the simlandr package for R, which provides a set of tools for constructing potential landscapes for dynamical systems using Monte Carlo simulation. Potential landscapes can be used to quantify the stability of system states. While the canonical form of a potential function is defined for gradient systems, generalized potential functions can also be defined for non-gradient dynamical systems. Our method is based on the potential landscape definition from the steady-state distribution, and can be used for a large variety of models. To facilitate simulation and computation, we introduce several novel features, including data structures optimized for batch simulations under varying conditions, an out-of-memory computation tool with integrated hash-based file-saving systems, and an algorithm for efficiently searching the minimum energy path. Using a multistable cell differentiation model as examples, we illustrate how simlandr can be used for model simulation, landscape construction, and barrier height calculation. The simlandr package is available at https://CRAN.R-project.org/package=simlandr, under GPL-3 license.","doi":"10.32614/RJ-2025-039","authors":[{"author":"Jingmeng Cui","authorURL":"#","affiliation":["University of Groningen","Radboud University"],"affiliationURL":"#","orcidID":"0000-0003-3421-8457"},{"author":"Merlijn Olthof","authorURL":"#","affiliation":["University of Groningen","Radboud University"],"affiliationURL":"#","orcidID":"0000-0002-5975-6588"},{"author":"Anna Lichtwarck-Aschoff","authorURL":"#","affiliation":["University of Groningen","Radboud University"],"affiliationURL":"#","orcidID":"0000-0002-4365-1538"},{"author":"Tiejun Li","authorURL":"#","affiliation":"Peking University","affiliationURL":"#","orcidID":"0000-0002-2086-1620"},{"author":"Fred Hasselman","authorURL":"#","affiliation":"Radboud University","affiliationURL":"#","orcidID":"0000-0003-1384-8361"}],"publishedDate":"2026-02-04T00:00:00.000+00:00","citationText":"Cui, et al., 2026"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a class="logo" href="../../index.html">
<img src="../../resources/rlogo.png" alt="Logo"/>
</a>
<a href="../../index.html" class="title">The R Journal</a>
</div>
<div class="nav-right">
<a href="../../index.html">Home</a>
<a href="../../issues/2025-4">Current</a>
<a href="../../issues.html">Issues</a>
<a href="../../news.html">News</a>
<a href="../../submissions.html">Submit</a>
<a href="../../contribute.html">Contribute</a>
<a href="../../editors.html">Editorial board</a>
<a href="../../articles.xml">
<i class="fa fa-rss" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>simlandr: Simulation-Based Landscape Construction for Dynamical Systems</h1>

<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>We present the simlandr package for R, which provides a set of tools for constructing potential landscapes for dynamical systems using Monte Carlo simulation. Potential landscapes can be used to quantify the stability of system states. While the canonical form of a potential function is defined for gradient systems, generalized potential functions can also be defined for non-gradient dynamical systems. Our method is based on the potential landscape definition from the steady-state distribution, and can be used for a large variety of models. To facilitate simulation and computation, we introduce several novel features, including data structures optimized for batch simulations under varying conditions, an out-of-memory computation tool with integrated hash-based file-saving systems, and an algorithm for efficiently searching the minimum energy path. Using a multistable cell differentiation model as examples, we illustrate how simlandr can be used for model simulation, landscape construction, and barrier height calculation. The simlandr package is available at <a href="https://CRAN.R-project.org/package=simlandr" class="uri">https://CRAN.R-project.org/package=simlandr</a>, under GPL-3 license.</p></p>
</div>

<div class="d-byline">
  Jingmeng Cui  (University of GroningenRadboud University)
  
,   Merlijn Olthof  (University of GroningenRadboud University)
  
,   Anna Lichtwarck-Aschoff  (University of GroningenRadboud University)
  
,   Tiejun Li  (Peking University)
  
,   Fred Hasselman  (Radboud University)
  
<br/>2026-02-04
</div>

<div class="d-article">
<div class="layout-chunk" data-layout="l-body">

</div>
<h2 data-number="1" id="sec:intro"><span class="header-section-number">1</span> Introduction</h2>
<p>To better understand a dynamical system, it is often important to know the stability of different states. The metaphor of a potential landscape consisting of hills and valleys has been used to illustrate differences in stability in many fields, including genetics <span class="citation" data-cites="wangQuantifyingWaddingtonLandscape2011 WaddingtonPrinciplesDevelopmentDifferentiation1966">(<a href="#ref-WaddingtonPrinciplesDevelopmentDifferentiation1966" role="doc-biblioref">Waddington 1966</a>; <a href="#ref-wangQuantifyingWaddingtonLandscape2011" role="doc-biblioref">Wang et al. 2011</a>)</span>, ecology <span class="citation" data-cites="lamotheLinkingBallandcupAnalogy2019">(<a href="#ref-lamotheLinkingBallandcupAnalogy2019" role="doc-biblioref">Lamothe et al. 2019</a>)</span>, and psychology <span class="citation" data-cites="olthofComplexityTheoryPsychopathology2020">(<a href="#ref-olthofComplexityTheoryPsychopathology2020" role="doc-biblioref">Olthof et al. 2023</a>)</span>. In such a landscape, the stable states of the system correspond to the lowest points (minima) in the valleys of the landscape. Just like a ball that is thrown in such a landscape will eventually gravitate towards such a minimum, the dynamical system is conceptually more likely to visit its stable states in which the system is also more resilient to noise. For example, in the landscape metaphor of psychopathology (Figure <a href="#fig:metaphor">1</a>), the valleys represent different mental health states, their relative depth represents the relative stability of the states, and the barriers between valleys represent the difficulty of transitioning between these states <span class="citation" data-cites="olthofComplexityTheoryPsychopathology2020">Hayes and Andrews (<a href="#ref-hayesComplexSystemsApproach2020" role="doc-biblioref">2020</a>)</span>. When the healthy state is more stable, the person is more likely to stay mentally healthy, whereas when the maladaptive state is more stable, the person is more likely to suffer from mental disorders.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:metaphor"></span>
<img src="figures/metaphor.png" alt="Three landscape plots, each with a ball resting in one of two basins. The left basin is labeled maladaptive, the right basin healthy. In the first plot, the maladaptive basin is deeper; in the second, both basins are equally deep; in the third, the healthy basin is deeper." width="100%" />
<p class="caption">
Figure 1: Illustration of the ball-and-landscape metaphor commonly used in the field of psychopathology.
</p>
</div>
</div>
<p>Yet, formally quantifying the stability of states is a nontrivial question. Here we present an R package, <a href="https://cran.r-project.org/package=simlandr">simlandr</a>, that can quantify the stability of various kinds of systems without many mathematical restrictions.</p>
<p>Dynamical systems are usually modeled by stochastic differential equations, which may depend on the past history (i.e., may be non-Markovian, <span class="citation" data-cites="StumpfEtAlModelingStemCell2021">Stumpf et al. (<a href="#ref-StumpfEtAlModelingStemCell2021" role="doc-biblioref">2021</a>)</span>). They take the general form of
<span class="math display" id="eq:sde">\[\begin{equation}
\mathrm{d} \boldsymbol{X}_t = \boldsymbol{b}(\boldsymbol{X}_t, \boldsymbol{H}_t){\mathrm{d}t} + \boldsymbol{\sigma}( \boldsymbol{X}_t, \boldsymbol{H}_t)\mathrm{d}\boldsymbol{W},
\tag{1}
\end{equation}\]</span>
where <span class="math inline">\(\boldsymbol{X}_t\)</span> is the random variable representing the current state of the system and <span class="math inline">\(\boldsymbol{H_t}\)</span> represents the past history of the system <span class="math inline">\(\boldsymbol{H}_t=\{\boldsymbol{X_s} | s \in [0, t)\}\)</span> The first term on the right-hand side of Eq <a href="#eq:sde">(1)</a> represents the deterministic part of the dynamics, which is a function of the system’s current state <span class="math inline">\(\boldsymbol{b}(\boldsymbol{X}_t, \boldsymbol{H}_t)\)</span>. The second term represents the stochastic part, which is standard white noise <span class="math inline">\(\mathrm{d}\boldsymbol{W}\)</span> multiplied by the noise strength <span class="math inline">\(\boldsymbol{\sigma}( \boldsymbol{X}_t, \boldsymbol{H}_t)\)</span>.</p>
<p>If the dynamical equation (Eq <a href="#eq:sde">(1)</a>) can be written in the following form
<span class="math display" id="eq:canon">\[\begin{equation}
\mathrm{d} \boldsymbol{X} = - \nabla U {\mathrm{d} t} +\sqrt{2}\mathrm{d}\boldsymbol{W},
\tag{2}
\end{equation}\]</span>
then <span class="math inline">\(U\)</span> is the potential function of the system.. However, this is not possible for general dynamical systems. The trajectory of such a system may contain loops which are not possible to be represented by a gradient system (this issue was compared to Escher’s stairs by <span class="citation" data-cites="rodriguez-sanchezClimbingEscherStairs2020">Rodríguez-Sánchez et al. (<a href="#ref-rodriguez-sanchezClimbingEscherStairs2020" role="doc-biblioref">2020</a>)</span>). In this case, further generalization is needed. The theoretical background of <a href="https://cran.r-project.org/package=simlandr">simlandr</a> is the generalized potential landscape by <span class="citation" data-cites="wangPotentialLandscapeFlux2008">Wang et al. (<a href="#ref-wangPotentialLandscapeFlux2008" role="doc-biblioref">2008</a>)</span>, which is based on the Boltzmann distribution and the steady-state distribution of the system. The Boltzmann distribution is a distribution law in physics, which states the distribution of classical particles depends on the energy level they occupy. When the energy is higher, the particle is exponentially less likely to be in such states
<span class="math display" id="eq:Boltzmann">\[\begin{equation}
P(\boldsymbol{x}) \propto \exp (-U).
\tag{3}
\end{equation}\]</span></p>
<p>This is then linked to dynamical systems by the steady-state distribution. The steady-state distribution of stochastic differential equations is the distribution that does not change over time, denoted by <span class="math inline">\(P_{\mathrm{SS}}\)</span> which satisfies
<span class="math display" id="eq:ss">\[\begin{equation}
\frac{\partial P_{\mathrm{SS}} (\boldsymbol{x},t)}{\partial t} = 0.
\tag{4}
\end{equation}\]</span>
The steady-state distribution is important because it extracts time-invariant information from a set of stochastic differential equations. Substituting the steady-state distribution to Eq <a href="#eq:Boltzmann">(3)</a> gives Wang’s generalized potential landscape function <span class="citation" data-cites="wangPotentialLandscapeFlux2008">(<a href="#ref-wangPotentialLandscapeFlux2008" role="doc-biblioref">Wang et al. 2008</a>)</span>
<span class="math display" id="eq:Wang">\[\begin{equation}
U(\boldsymbol{x}) = - \ln P_{\mathrm{SS}}(\boldsymbol{x}).
\tag{5}
\end{equation}\]</span>
If the system has ergodicity (i.e., after sufficient time it can travel to all possible states in the state space), the long-term sample distribution can be used to estimate the steady-state distribution, and the generalized potential function can be calculated.</p>
<p>Our approach is not the only possible way for constructing potential landscapes. Many other theoretical approaches are available, including the SDE decomposition method by <span class="citation" data-cites="AoPotentialStochasticDifferential2004">Ao (<a href="#ref-AoPotentialStochasticDifferential2004" role="doc-biblioref">2004</a>)</span> and the quasi-potential by <span class="citation" data-cites="FreidlinWentzellRandomPerturbationsDynamical2012">Freidlin and Wentzell (<a href="#ref-FreidlinWentzellRandomPerturbationsDynamical2012" role="doc-biblioref">2012</a>)</span>. Various strategies to numerically compute these landscapes have been proposed (see <span class="citation" data-cites="zhouConstructionLandscapeMultistable2016">Zhou and Li (<a href="#ref-zhouConstructionLandscapeMultistable2016" role="doc-biblioref">2016</a>)</span> for a review). However, available realizations are still scarce. To our knowledge, besides <a href="https://cran.r-project.org/package=simlandr">simlandr</a>, there are two existing packages specifically for computing potential landscapes: the <a href="https://cran.r-project.org/package=waydown">waydown</a> package <span class="citation" data-cites="rodriguez-sanchezPabRodRolldownPostpublication2020">(<a href="#ref-rodriguez-sanchezPabRodRolldownPostpublication2020" role="doc-biblioref">Rodríguez-Sánchez 2020</a>)</span> and the <a href="#">QPot</a> package <span class="citation" data-cites="MooreEtAlQPotPackageStochastic2016 dahiyaOrderedLineIntegral2018">(<a href="#ref-MooreEtAlQPotPackageStochastic2016" role="doc-biblioref">Moore et al. 2016</a>; <a href="#ref-dahiyaOrderedLineIntegral2018" role="doc-biblioref">Dahiya and Cameron 2018</a>)</span>. The <a href="https://cran.r-project.org/package=waydown">waydown</a> package uses the skew-symmetric decomposition of the Jacobian, which theoretically produces landscapes that are similar to <span class="citation" data-cites="wangPotentialLandscapeFlux2008">Wang et al. (<a href="#ref-wangPotentialLandscapeFlux2008" role="doc-biblioref">2008</a>)</span> (but see <span class="citation" data-cites="CuiEtAlCommentsClimbingEscher2023">Cui et al. (<a href="#ref-CuiEtAlCommentsClimbingEscher2023" role="doc-biblioref">2023a</a>)</span> for a potential technical issue with this package.). The <a href="#">QPot</a> package uses a path integral method that produces quasi-landscapes following the definition by <span class="citation" data-cites="FreidlinWentzellRandomPerturbationsDynamical2012">Freidlin and Wentzell (<a href="#ref-FreidlinWentzellRandomPerturbationsDynamical2012" role="doc-biblioref">2012</a>)</span>. Because of the analytical methods used by <a href="https://cran.r-project.org/package=waydown">waydown</a> and <a href="#">QPot</a>, they both require the dynamic function to be Markovian and differentiable in the whole state space. Moreover, they can only be used for systems of up to two dimensions. <a href="https://cran.r-project.org/package=simlandr">simlandr</a>, in contrast, is based on Monte Carlo simulation and the steady-state distribution. It does not have specific requirements for the model. Even for models that are not globally differentiable, have history-dependence, and are defined in a high-dimensional space, <a href="https://cran.r-project.org/package=simlandr">simlandr</a> is still applicable (e.g., <span class="citation" data-cites="CuiEtAlMetaphorComputationConstructing2021">Cui et al. (<a href="#ref-CuiEtAlMetaphorComputationConstructing2021" role="doc-biblioref">2023b</a>)</span>). Therefore, <a href="https://cran.r-project.org/package=simlandr">simlandr</a> can be applied to a much wider range of dynamical systems, illustrate a big picture of dominated attractors, and investigate how the stability of different attractors may be influenced by model parameters. As a trade-off, <a href="https://cran.r-project.org/package=simlandr">simlandr</a> is not designed for rare events sampling in which the noise strength <span class="math inline">\(\boldsymbol{\sigma}(\boldsymbol{X})\)</span> is extremely small, nor for the precise calculation of the tail probability and transition paths. Instead, it is better to view <a href="https://cran.r-project.org/package=simlandr">simlandr</a> as a semi-quantitative tool that provides a broad overview of key attractors in dynamic systems, allowing for comparisons of their relative stability and the investigation of system parameter influences. We will show some typical use cases of <a href="https://cran.r-project.org/package=simlandr">simlandr</a> in later sections. Some key terms used in this article are summarized in Table <a href="#tab:terms">1</a>.</p>
<table>
<caption><span id="tab:terms">Table 1: </span> Summary of key terms used in this article.</caption>
<colgroup>
<col style="width: 31%" />
<col style="width: 68%" />
</colgroup>
<thead>
<tr class="header">
<th>Term</th>
<th>Explanation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Potential landscape metaphor</td>
<td>A conceptual metaphor representing the stability of a complex dynamic system as an uneven landscape, with a ball on it representing the system’s state. This can be quantitatively realized in various ways <span class="citation" data-cites="zhouConstructionLandscapeMultistable2016">(<a href="#ref-zhouConstructionLandscapeMultistable2016" role="doc-biblioref">Zhou and Li 2016</a>)</span>.</td>
</tr>
<tr class="even">
<td>Gradient system</td>
<td>A system whose deterministic motion can be described solely by the gradient of a potential function.</td>
</tr>
<tr class="odd">
<td>Non-Markovian system</td>
<td>A system whose future evolution depends not only on its current state but also on its past history.</td>
</tr>
<tr class="even">
<td>Steady-state distribution</td>
<td>The probability distribution of a dynamic system that remains unchanged over time.</td>
</tr>
<tr class="odd">
<td>Ergodic system</td>
<td>A dynamic system that, given enough time, will eventually pass through all possible states.</td>
</tr>
<tr class="even">
<td>Minimum energy path (MEP)</td>
<td>A transition path linking two local minima and passing through a saddle point <span class="citation" data-cites="WanEtAlFindingTransitionState2024">(<a href="#ref-WanEtAlFindingTransitionState2024" role="doc-biblioref">Wan et al. 2024</a>)</span>. It is always parallel to the gradient of the energy landscape, representing an efficient transition route.</td>
</tr>
</tbody>
</table>
<h2 data-number="2" id="design-and-implementation"><span class="header-section-number">2</span> Design and implementation</h2>
<p>The general workflow of <a href="https://cran.r-project.org/package=simlandr">simlandr</a> involves three steps: model simulation, landscape construction, and barrier height computation. See Figure <a href="#fig:diagram">2</a> for a summary.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:diagram"></span>
<img src="figures/diagram.png" alt="A flow chart showing the analysis steps in simlandr, with functions listed under each step." width="100%" />
<p class="caption">
Figure 2: The structure and workflow of simlandr.
</p>
</div>
</div>
<h3 data-number="2.1" id="step-1-model-simulation"><span class="header-section-number">2.1</span> Step 1: model simulation</h3>
<p>For the first step, a simulation function should be created by the user. This function should be able to simulate how the dynamical system of interest evolves over time and record its state in every time step. This can often be done with the Euler-Maruyama method. If the SDEs are up to three dimensions and Markovian, a helper function from <a href="https://cran.r-project.org/package=simlandr">simlandr</a>, <code>sim_SDE()</code>, can also be used. This function is based on the simulation utilities from the <a href="https://cran.r-project.org/package=Sim.DiffProc">Sim.DiffProc</a> <span class="citation" data-cites="GuidoumBoukhetalaPerformingParallelMonte2020">(<a href="#ref-GuidoumBoukhetalaPerformingParallelMonte2020" role="doc-biblioref">Guidoum and Boukhetala 2020</a>)</span> and the output can be directly used for later steps. Moreover, the <code>multi_init_simulation()</code> function can be used to simulate trajectories from various starting points, thus reducing the possibility for the system to be trapped in a local minimum. The <code>multi_init_simulation()</code> function also supports parallel simulation based on the <a href="https://cran.r-project.org/package=future">future</a> framework to improve time efficiency <span class="citation" data-cites="RJ-2021-048">Bengtsson (<a href="#ref-RJ-2021-048" role="doc-biblioref">2021</a>)</span>.</p>
<p>For Monte Carlo methods, it is important that the simulation <em>converges</em>, which means the distribution of the system is roughly stable. The precision of the steady-state distribution estimation, according to Eq <a href="#eq:Wang">(5)</a>, determines the precision of the distribution estimation. <a href="https://cran.r-project.org/package=simlandr">simlandr</a> provides a visual tool to compare the sample distributions in different stages (<code>check_conv()</code>), whereas the <a href="https://cran.r-project.org/package=coda">coda</a> package <span class="citation" data-cites="PlummerEtAlCODAConvergenceDiagnosis2006">(<a href="#ref-PlummerEtAlCODAConvergenceDiagnosis2006" role="doc-biblioref">Plummer et al. 2006</a>)</span> can be used for more advanced diagnostics. The output of the <code>sim_SDE()</code> and <code>multi_init_simulation()</code> functions also uses the classes from the <a href="https://cran.r-project.org/package=coda">coda</a> package to enable easy convergence diagnosis. To achieve ergodicity in reasonable time, sometimes stronger noises need to be added to the system.</p>
<p>A simulation function is sufficient if the user is only interested in a single model setting. If the model is parameterized and the user wants to investigate the influence of parameters on the stability of the system, then multiple simulations need to be run with different parameter settings. <a href="https://cran.r-project.org/package=simlandr">simlandr</a> provides functions to perform batch simulations and store the outputs for landscape construction as one object. This can later be used to compare the stability under different parameter settings or produce animations to show how a model parameter influences the stability of the model.</p>
<p>In many cases, the output of the simulation is so large that it cannot be properly stored in the memory. <a href="https://cran.r-project.org/package=simlandr">simlandr</a> provides a hash_big_matrix class, which is a modification of the big.matrix class from the <a href="https://cran.r-project.org/package=bigmemory">bigmemory</a> package <span class="citation" data-cites="KaneEtAlScalableStrategiesComputing2013">(<a href="#ref-KaneEtAlScalableStrategiesComputing2013" role="doc-biblioref">Kane et al. 2013</a>)</span>, that can perform out-of-memory computation and organize the data files in the disk. In an out-of-memory computation, the majority of the data is not loaded into the memory, but only the small subset of data that is used for the current computation step. Therefore, the memory occupation is dramatically reduced. The big.matrix class in the <a href="https://cran.r-project.org/package=bigmemory">bigmemory</a> package provides a powerful tool for out-of-memory computation. It, however, requires an explicit file name for each matrix, which can be cumbersome if there are many matrices to be handled, and this is likely to be the case in a batch simulation. The <code>hash_big_matrix</code> class automatically generates the file names using the md5 values of the matrices with the <a href="https://cran.r-project.org/package=digest">digest</a> package<span class="citation" data-cites="EddelbuettelEtAlPkgdigestCreateCompact2021">(<a href="#ref-EddelbuettelEtAlPkgdigestCreateCompact2021" role="doc-biblioref">Eddelbuettel et al. 2021</a>)</span> and stores it within the object. Therefore, the file links can also be restored automatically.</p>
<h3 data-number="2.2" id="step-2-landscape-construction"><span class="header-section-number">2.2</span> Step 2: landscape construction</h3>
<p><a href="https://cran.r-project.org/package=simlandr">simlandr</a> provides a set of tools to construct 2D, 3D, and 4D landscapes from single or multiple simulation results. The steady-state distribution for selected variables of the system is first estimated using the kernel density estimates (KDE). The density function in R is used for 2D landscapes, whereas the <a href="https://cran.r-project.org/package=ks">ks</a> package <span class="citation" data-cites="ChaconDuongMultivariateKernelSmoothing2018">(<a href="#ref-ChaconDuongMultivariateKernelSmoothing2018" role="doc-biblioref">Chacón and Duong 2018</a>)</span> is used by default for 3D and 4D landscapes because of its higher efficiency. Then the potential function <span class="math inline">\(U\)</span> is calculated from Eq <a href="#eq:Wang">(5)</a>. The landscape plots without a z-axis are created with <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> <span class="citation" data-cites="WickhamGgplot2ElegantGraphics2016">(<a href="#ref-WickhamGgplot2ElegantGraphics2016" role="doc-biblioref">Wickham 2016</a>)</span>, and those with a z-axis are created with <a href="https://cran.r-project.org/package=plotly">plotly</a> <span class="citation" data-cites="SievertInteractiveWebbasedData2020">(<a href="#ref-SievertInteractiveWebbasedData2020" role="doc-biblioref">Sievert 2020</a>)</span>. These plots can be further refined using the standard <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> or <a href="https://cran.r-project.org/package=plotly">plotly</a> methods. See Table <a href="#tab:overview">2</a> for an overview for the family of landscape functions.</p>
<table>
<caption><span id="tab:overview">Table 2: </span> Overview of various landscape functions provided by <code>simlandr</code>. Dimensions in bold represent the potential U calculated by the function. Dimensions in italic represent model parameters. Dimensions in parentheses are optional.</caption>
<colgroup>
<col style="width: 26%" />
<col style="width: 26%" />
<col style="width: 47%" />
</colgroup>
<thead>
<tr class="header">
<th>Type of Input</th>
<th>Function</th>
<th>Dimensions</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Single simulation data</td>
<td><code>make_2d_static()</code></td>
<td>x, <strong>y</strong></td>
</tr>
<tr class="even">
<td></td>
<td><code>make_3d_static()</code></td>
<td><span class="math inline">\(1\)</span> x, y, <strong>z+color</strong>; (2) x, y, <strong>color</strong></td>
</tr>
<tr class="odd">
<td></td>
<td><code>make_4d_static()</code></td>
<td>x, y, z, <strong>color</strong></td>
</tr>
<tr class="even">
<td>Multiple simulation data</td>
<td><code>make_2d_matrix()</code></td>
<td>x, <strong>y</strong>, <em>cols</em>, <em>(rows)</em></td>
</tr>
<tr class="odd">
<td></td>
<td><code>make_3d_matrix()</code></td>
<td>x, y, <strong>z+color</strong>, <em>cols</em>, <em>(rows)</em></td>
</tr>
<tr class="even">
<td></td>
<td><code>make_3d_animation()</code></td>
<td><span class="math inline">\(1\)</span> x, y, <strong>z+color</strong>, <em>fr</em>; (2) x, y, <strong>color</strong>, <em>fr</em>; (3) x, y, <strong>z+color</strong>, <em>cols</em></td>
</tr>
</tbody>
</table>
<h3 data-number="2.3" id="sec:singlel"><span class="header-section-number">2.3</span> Step 3: barrier height calculation</h3>
<p>An important property of the states in a landscape is their stability, which can be indicated by the barrier height that the system has to overcome when it transitions between one stable state to another adjacent state (see <span class="citation" data-cites="CuiEtAlMetaphorComputationConstructing2021">Cui et al. (<a href="#ref-CuiEtAlMetaphorComputationConstructing2021" role="doc-biblioref">2023b</a>)</span> for further discussions about different stability indicators). The barrier height is also related to the escape time that the system transitions from one valley to another, which can be tested empirically <span class="citation" data-cites="wangPotentialLandscapeFlux2008">(<a href="#ref-wangPotentialLandscapeFlux2008" role="doc-biblioref">Wang et al. 2008</a>)</span>. <a href="https://cran.r-project.org/package=simlandr">simlandr</a> provides tools to calculate the barrier heights from landscapes. These functions look for the local minima in given regions and try to find the saddle point between two local minima. The potential differences between the saddle point and local minima are calculated as barrier heights.</p>
<p>In 2D cases, there is only one possible path connecting two local minima. The point on the path with the highest <span class="math inline">\(U\)</span> is identified as the saddle point. For 3D landscapes, there are multiple paths between two local minima. If we treat the system <em>as if</em> it is a gradient system with Brownian noise, then the most probable path (termed as the <em>minimum energy path</em>, MEP) that the system transitions is that it first goes along the steepest <em>ascent</em> path from the starting point, and then goes along the steepest <em>descent</em> path to the end point <span class="citation" data-cites="EVanden-EijndenTransitionpathTheoryPathfinding2010">(<a href="#ref-EVanden-EijndenTransitionpathTheoryPathfinding2010" role="doc-biblioref">E and Vanden-Eijnden 2010</a>)</span>. We find this path by minimizing the following action using the <span class="citation" data-cites="dijkstra1959note">Dijkstra (<a href="#ref-dijkstra1959note" role="doc-biblioref">1959</a>)</span> algorithm <span class="citation" data-cites="HeymannVanden-EijndenPathwaysMaximumLikelihood2008">(<a href="#ref-HeymannVanden-EijndenPathwaysMaximumLikelihood2008" role="doc-biblioref">Heymann and Vanden-Eijnden 2008</a>)</span>
<span class="math display" id="eq:optim">\[\begin{equation}
\varphi_{\mathrm{MEP}} = \arg\min_\varphi \int_A^B |\nabla U||\mathrm{d}\varphi|\left( \approx \min_\varphi \Sigma_i |\nabla U_i||\Delta \varphi_i|\right),
\tag{6}
\end{equation}\]</span>
where <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span> are the starting and end points and <span class="math inline">\(\varphi\)</span> is the path starting at <span class="math inline">\(A\)</span> and ending in <span class="math inline">\(B\)</span>. After that, the point with the maximum potential value on the MEP is identified as the saddle point. Note that while the barrier height still indicates the stability of local minima, the MEP may not be the true most probable path for a nongradient system to transition between stable states.</p>
<h2 data-number="3" id="examples"><span class="header-section-number">3</span> Examples</h2>
<p>We use two dynamical systems to illustrate the usage of the <a href="https://cran.r-project.org/package=simlandr">simlandr</a> package. The first one is a two-dimensional stochastic non-gradient gene expression model, which was used by <span class="citation" data-cites="wangQuantifyingWaddingtonLandscape2011">Wang et al. (<a href="#ref-wangQuantifyingWaddingtonLandscape2011" role="doc-biblioref">2011</a>)</span> to represent cell development and differentiation. The second example is a dynamic model of panic disorder <span class="citation" data-cites="robinaugh2024">(<a href="#ref-robinaugh2024" role="doc-biblioref">Robinaugh et al. 2024</a>)</span> which contains many more variables and parameters, non-Markovian property, and non-differentiable formulas. We mainly use the first example to show the agreement of the results from <a href="https://cran.r-project.org/package=simlandr">simlandr</a> with previous analytic results, and the second example as a typical use case of a complex dynamic model which is not treatable with other methods (also see <span class="citation" data-cites="CuiEtAlMetaphorComputationConstructing2021">Cui et al. (<a href="#ref-CuiEtAlMetaphorComputationConstructing2021" role="doc-biblioref">2023b</a>)</span> for more substantive discussions). Note that, both systems include more than two variables, making it impossible to perform the landscape analysis with other available R packages.</p>
<h3 data-number="3.1" id="example-1-the-gene-expression-model"><span class="header-section-number">3.1</span> Example 1: the gene expression model</h3>
<p>This model is built on the mutual regulations of the expressions of two genes, in which <span class="math inline">\(X_1\)</span> and <span class="math inline">\(X_2\)</span> represent the expression levels of two genes which activate themselves and inhibit each other. A graphical illustration is shown in Figure <a href="#fig:e1">3</a> (adapted from <span class="citation" data-cites="wangQuantifyingWaddingtonLandscape2011">Wang et al. (<a href="#ref-wangQuantifyingWaddingtonLandscape2011" role="doc-biblioref">2011</a>)</span>). Their dynamic functions can be written as
<span class="math display" id="eq:sim2a" id="eq:sim2x2" id="eq:sim2x1">\[\begin{align}
\frac {\mathrm{d}X_ {1}}{\mathrm{d}t}  &amp;=  \frac {ax_ {1}^ {n}}{S^ {n}+x_ {1}^ {n}} + \frac {bS^ {n}}{S^ {n}+x_ {2}^ {n}} - kx_ {1}+ \sigma_1 \frac{\mathrm{d}W_1}{\mathrm{d}t},
\tag{7}
\\
\frac {\mathrm{d}X_ {2}}{\mathrm{d}t}  &amp;=  \frac {ax_ {2}^ {n}}{S^ {n}+x_ {2}^ {n}} + \frac {bS^ {n}}{S^ {n}+x_ {1}^ {n}} - kx_ {2}+ \sigma_2 \frac{\mathrm{d}W_2}{\mathrm{d}t},
\tag{8}
\\
\frac {\mathrm{d}a}{\mathrm{d}t} &amp;= -\lambda a+ \sigma_3 \frac{\mathrm{d}W_3}{\mathrm{d}t},
\tag{9}
\end{align}\]</span>
where <span class="math inline">\(a\)</span> represents the strength of self-activation, <span class="math inline">\(b\)</span> represents the strength of mutual-inhibition, and <span class="math inline">\(k\)</span> represents the speed of degradation. The development of an organism is modeled as <span class="math inline">\(a\)</span> decreasing at a certain speed <span class="math inline">\(\lambda\)</span>. In the beginning, there is only one possible state for the cell. After a certain milestone, the cell differentiates into one of the two possible states.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:e1"></span>
<img src="figures/diagram-e1.png" alt="Two large circles labeled X1 and X2. Dashed arrows labeled b connect the circles in both directions. Each circle has a solid self-loop labeled a and a dashed self-loop labeled k." width="50%" />
<p class="caption">
Figure 3: A graphical illustration of the relationship between the activation levels of the two genes. Solid arrows represent positive relationships (i.e., activation) and dashed arrows represent negative relationships (i.e., inhibition).
</p>
</div>
</div>
<p>This model can be simulated using the <code>sim_SDE()</code> function in <a href="https://cran.r-project.org/package=simlandr">simlandr</a>, with the default parameter setting <span class="math inline">\(b = 1, k = 1, S = 0.5, n = 4, \lambda = 0.01\)</span>, and <span class="math inline">\(\sigma_1 = \sigma_2 = \sigma_3 = 0.2\)</span>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># Load the package.</span></span>
<span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://sciurus365.github.io/simlandr/'>simlandr</a></span><span class='op'>)</span></span>
<span></span>
<span><span class='co'># Specify the simulation function.</span></span>
<span><span class='va'>b</span> <span class='op'>&lt;-</span> <span class='fl'>1</span></span>
<span><span class='va'>k</span> <span class='op'>&lt;-</span> <span class='fl'>1</span></span>
<span><span class='va'>S</span> <span class='op'>&lt;-</span> <span class='fl'>0.5</span></span>
<span><span class='va'>n</span> <span class='op'>&lt;-</span> <span class='fl'>4</span></span>
<span><span class='va'>lambda</span> <span class='op'>&lt;-</span> <span class='fl'>0.01</span></span>
<span></span>
<span><span class='va'>drift_gene</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'>rlang</span><span class='fu'>::</span><span class='fu'><a href='https://rlang.r-lib.org/reference/expr.html'>expr</a></span><span class='op'>(</span><span class='va'>z</span> <span class='op'>*</span> <span class='va'>x</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>/</span><span class='op'>(</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>S</span><span class='op'>)</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span> <span class='op'>+</span> <span class='va'>x</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> <span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>b</span><span class='op'>)</span> <span class='op'>*</span></span>
<span>    <span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>S</span><span class='op'>)</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>/</span><span class='op'>(</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>S</span><span class='op'>)</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span> <span class='op'>+</span> <span class='va'>y</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>-</span> <span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>k</span><span class='op'>)</span> <span class='op'>*</span> <span class='va'>x</span><span class='op'>)</span>, <span class='fu'>rlang</span><span class='fu'>::</span><span class='fu'><a href='https://rlang.r-lib.org/reference/expr.html'>expr</a></span><span class='op'>(</span><span class='va'>z</span> <span class='op'>*</span> <span class='va'>y</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>/</span><span class='op'>(</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>S</span><span class='op'>)</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span> <span class='op'>+</span></span>
<span>    <span class='va'>y</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> <span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>b</span><span class='op'>)</span> <span class='op'>*</span> <span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>S</span><span class='op'>)</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>/</span><span class='op'>(</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>S</span><span class='op'>)</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span> <span class='op'>+</span> <span class='va'>x</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>-</span> <span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>k</span><span class='op'>)</span> <span class='op'>*</span> <span class='va'>y</span><span class='op'>)</span>,</span>
<span>    <span class='fu'>rlang</span><span class='fu'>::</span><span class='fu'><a href='https://rlang.r-lib.org/reference/expr.html'>expr</a></span><span class='op'>(</span><span class='op'>-</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>lambda</span><span class='op'>)</span> <span class='op'>*</span> <span class='va'>z</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>|&gt;</span></span>
<span>    <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>as.expression</a></span><span class='op'>(</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>diffusion_gene</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>expression</a></span><span class='op'>(</span><span class='fl'>0.2</span>, <span class='fl'>0.2</span>, <span class='fl'>0.2</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># Perform a simulation and save the output.</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>1614</span><span class='op'>)</span></span>
<span><span class='va'>single_output_gene</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/sim_SDE.html'>sim_SDE</a></span><span class='op'>(</span>drift <span class='op'>=</span> <span class='va'>drift_gene</span>, diffusion <span class='op'>=</span> <span class='va'>diffusion_gene</span>,</span>
<span>    N <span class='op'>=</span> <span class='fl'>1e+06</span>, M <span class='op'>=</span> <span class='fl'>10</span>, Dt <span class='op'>=</span> <span class='fl'>0.1</span>, x0 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>0</span>, <span class='fl'>1</span><span class='op'>)</span>, keep_full <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<p>After the simulation, we perform some basic data wrangling to produce a dataset that can be used for further analysis. We create a new variable <code>delta_x</code> as the difference between X1 (X) and X2 (Y), and we rename the variable Z as <code>a</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>single_output_gene2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/do.call.html'>do.call</a></span><span class='op'>(</span><span class='va'>rbind</span>, <span class='va'>single_output_gene</span><span class='op'>)</span></span>
<span><span class='va'>single_output_gene2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='va'>single_output_gene2</span><span class='op'>[</span>, <span class='st'>"X"</span><span class='op'>]</span> <span class='op'>-</span> <span class='va'>single_output_gene2</span><span class='op'>[</span>,</span>
<span>    <span class='st'>"Y"</span><span class='op'>]</span>, <span class='va'>single_output_gene2</span><span class='op'>[</span>, <span class='st'>"Z"</span><span class='op'>]</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>single_output_gene2</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"delta_x"</span>, <span class='st'>"a"</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>We then perform the convergence check on the simulation result. First, we convert the simulation output to the format for the <code>coda</code> package, and thin the output to speed up the convergence check.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>single_output_gene_mcmc_thin</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/coda/man/mcmc.list.html'>as.mcmc.list</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span><span class='va'>single_output_gene</span>,</span>
<span>    <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='va'>x</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/seq.html'>seq</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>, by <span class='op'>=</span> <span class='fl'>100</span><span class='op'>)</span>, <span class='op'>]</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>We then show the convergence diagnosis plot to check the convergence of the simulation in Figure <a href="#fig:converge-gene">4</a>. The distribution of the two key variables in different simulation stages are converging, indicating that the simulation is long enough to provide reliable estimation of the steady-state distribution. Other convergence checks can also be readily performed using the <a href="https://cran.r-project.org/package=coda">coda</a> package.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>single_output_gene_mcmc_thin</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure"><span style="display:block;" id="fig:converge-gene"></span>
<img src="RJ-2025-039_files/figure-html5/converge-gene-1.png" alt="Trace and density plots for variables x, y, and z." width="100%" />
<p class="caption">
Figure 4: The convergence check result for the simulation of the gene expression model. The variables in different simulation stages did not show distributional differences, indicating that the simulation is long enough to provide a reliable estimation of the steady-state distribution.
</p>
</div>
</div>
<p>We generated the 3D landscape for this model with <code>make_3d_single()</code>. Here, we use <code>x</code>, <code>y</code> to specify the variables of interest, and use <code>lims</code> to specify the limits of the x and y axes for the landscape. The <code>lims</code> argument can be left blank, then the limits will be automatically calculated.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>l_single_gene_3d</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/make_3d_static.html'>make_3d_single</a></span><span class='op'>(</span><span class='va'>single_output_gene2</span>, x <span class='op'>=</span> <span class='st'>"delta_x"</span>,</span>
<span>    y <span class='op'>=</span> <span class='st'>"a"</span>, lims <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>1.5</span>, <span class='fl'>1.5</span>, <span class='fl'>0</span>, <span class='fl'>1.5</span><span class='op'>)</span>, Umax <span class='op'>=</span> <span class='fl'>8</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>The resulting landscape is shown in the left panel of Figure <a href="#fig:3dstaticgene">5</a>. In this plot, the x-axis represents <span class="math inline">\(\Delta x (= x_1-x_2)\)</span>, and the y-axis represents <span class="math inline">\(a\)</span>. To compare with, the potential landscape obtained analytically by <span class="citation" data-cites="wangQuantifyingWaddingtonLandscape2011">Wang et al. (<a href="#ref-wangQuantifyingWaddingtonLandscape2011" role="doc-biblioref">2011</a>)</span> is shown in the right panel Figure <a href="#fig:3dstaticgene">5</a>. The result of <a href="https://cran.r-project.org/package=simlandr">simlandr</a> appears to be very close to the result based on the analytical derivation. Note that because different normalization methods were used, the <span class="math inline">\(U\)</span> values of the two landscapes are not directly comparable. Here, we are mainly interested in their relative shape.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>l_single_non_grad_3d</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:3dstaticgene"></span>
<img src="figures/3dstatic_gene.png" alt="Two similar landscape plots, each with three basins." width="50%" /><img src="figures/wang2011.png" alt="Two similar landscape plots, each with three basins." width="50%" />
<p class="caption">
Figure 5: The 3D landscape (potential value as z-axis) for the gene expression model. The left panel is the plot produced by simlandr; the right panel is the potential landscape obtained analytically by Wang et al. (2008), reproduced with the permission of the authors and in accordance with the journal policy.
</p>
</div>
</div>
<p>We then calculate the barrier for the landscape using <code>calculate_barrier()</code>. The barrier is calculated by specifying the start and end locations, and the radii of the start and end locations. The height of the barrier from two sides can be calculated with <code>get_barrier_height()</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>b_single_gene_3d</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/calculate_barrier.html'>calculate_barrier</a></span><span class='op'>(</span><span class='va'>l_single_gene_3d</span>, start_location_value <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,</span>
<span>    <span class='fl'>1.2</span><span class='op'>)</span>, end_location_value <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>0.2</span><span class='op'>)</span>, start_r <span class='op'>=</span> <span class='fl'>0.3</span>, end_r <span class='op'>=</span> <span class='fl'>0.3</span><span class='op'>)</span></span>
<span></span>
<span><span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/get_barrier_height.html'>get_barrier_height</a></span><span class='op'>(</span><span class='va'>b_single_gene_3d</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>delta_U_start   delta_U_end 
     2.506326      2.810604 </code></pre>
</div>
<p>The local minima, the saddle point, and the MEP can be added to the landscape with <code>autolayer()</code>, shown in Figure <a href="#fig:bsingle3dgene">6</a>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>l_single_gene_3d</span>, <span class='fl'>2</span><span class='op'>)</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/autolayer.html'>autolayer</a></span><span class='op'>(</span><span class='va'>b_single_gene_3d</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure"><span style="display:block;" id="fig:bsingle3dgene"></span>
<img src="RJ-2025-039_files/figure-html5/bsingle3dgene-1.png" alt="A landscape plot with a white line connecting two white dots, passing through a red dot in the middle." width="50%" />
<p class="caption">
Figure 6: The landscape for the gene expression model. The local minima are marked as white dots, the saddle points are marked as red dots, and the MEPs are marked as white lines.
</p>
</div>
</div>
<p>Next, we use multiple simulations to investigate the influence of two parameters, <span class="math inline">\(k\)</span> and <span class="math inline">\(b\)</span>, on the stability of the system. As explained above, the parameter <span class="math inline">\(b\)</span> represents the strength of mutual-inhibition between the two genes. Therefore, as <span class="math inline">\(b\)</span> increases, we expect the differentiation is more extreme, that is, the cell is more likely to develop into one of the two cell types with very different gene expression levels. The valleys in the landscape representing the two types will become further apart and the barrier becomes clearer. The parameter <span class="math inline">\(k\)</span> represents the speed of degradation of the gene products. As <span class="math inline">\(k\)</span> increases, the gene products degrade faster, and this effect is more pronounced when the gene products are at high levels. Therefore, the dominant gene will express at a less extreme level, and we expect that the two valleys become closer, and the barrier will become less clear as <span class="math inline">\(k\)</span> increases.</p>
<p>We use the batch simulation functions of <a href="https://cran.r-project.org/package=simlandr">simlandr</a>. First, we create the argument set for the batch simulation. This specifies the parameters to be varied. We examined three <span class="math inline">\(b\)</span> values, 0.5, 1, 1.5, and three <span class="math inline">\(k\)</span> values, 0.5, 1, 1.5, which form 9 possible combinations.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>batch_arg_set_gene</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/arg_set-class.html'>new_arg_set</a></span><span class='op'>(</span><span class='op'>)</span></span>
<span><span class='va'>batch_arg_set_gene</span> <span class='op'>&lt;-</span> <span class='va'>batch_arg_set_gene</span> <span class='op'>|&gt;</span></span>
<span>    <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/arg_set-class.html'>add_arg_ele</a></span><span class='op'>(</span>arg_name <span class='op'>=</span> <span class='st'>"parameter"</span>, ele_name <span class='op'>=</span> <span class='st'>"b"</span>, start <span class='op'>=</span> <span class='fl'>0.5</span>, end <span class='op'>=</span> <span class='fl'>1.5</span>,</span>
<span>        by <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>|&gt;</span></span>
<span>    <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/arg_set-class.html'>add_arg_ele</a></span><span class='op'>(</span>arg_name <span class='op'>=</span> <span class='st'>"parameter"</span>, ele_name <span class='op'>=</span> <span class='st'>"k"</span>, start <span class='op'>=</span> <span class='fl'>0.5</span>, end <span class='op'>=</span> <span class='fl'>1.5</span>,</span>
<span>        by <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span></span>
<span><span class='va'>batch_grid_gene</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/arg_set-class.html'>make_arg_grid</a></span><span class='op'>(</span><span class='va'>batch_arg_set_gene</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>We then perform the batch simulation with the <code>batch_simulation()</code> function. Here, we specify the simulation function to be used, which is similar to the single simulation we showed above, together with the data wrangling procedure. The simulation function is defined with the <code>sim_fun</code> argument. We also use <code>bigmemory = TRUE</code> to store the simulation results in the <code>hash_big_matrix</code> format, which is more memory-efficient.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>batch_output_gene</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/batch_simulation.html'>batch_simulation</a></span><span class='op'>(</span><span class='va'>batch_grid_gene</span>, sim_fun <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>parameter</span><span class='op'>)</span> <span class='op'>{</span></span>
<span>    <span class='va'>b</span> <span class='op'>&lt;-</span> <span class='va'>parameter</span><span class='op'>[[</span><span class='st'>"b"</span><span class='op'>]</span><span class='op'>]</span></span>
<span>    <span class='va'>k</span> <span class='op'>&lt;-</span> <span class='va'>parameter</span><span class='op'>[[</span><span class='st'>"k"</span><span class='op'>]</span><span class='op'>]</span></span>
<span>    <span class='va'>drift_gene</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fu'>rlang</span><span class='fu'>::</span><span class='fu'><a href='https://rlang.r-lib.org/reference/expr.html'>expr</a></span><span class='op'>(</span><span class='va'>z</span> <span class='op'>*</span> <span class='va'>x</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>/</span><span class='op'>(</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>S</span><span class='op'>)</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span> <span class='op'>+</span> <span class='va'>x</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> <span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>b</span><span class='op'>)</span> <span class='op'>*</span></span>
<span>        <span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>S</span><span class='op'>)</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>/</span><span class='op'>(</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>S</span><span class='op'>)</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span> <span class='op'>+</span> <span class='va'>y</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>-</span> <span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>k</span><span class='op'>)</span> <span class='op'>*</span> <span class='va'>x</span><span class='op'>)</span>, <span class='fu'>rlang</span><span class='fu'>::</span><span class='fu'><a href='https://rlang.r-lib.org/reference/expr.html'>expr</a></span><span class='op'>(</span><span class='va'>z</span> <span class='op'>*</span></span>
<span>        <span class='va'>y</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>/</span><span class='op'>(</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>S</span><span class='op'>)</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span> <span class='op'>+</span> <span class='va'>y</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> <span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>b</span><span class='op'>)</span> <span class='op'>*</span> <span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>S</span><span class='op'>)</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>/</span><span class='op'>(</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>S</span><span class='op'>)</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span> <span class='op'>+</span></span>
<span>        <span class='va'>x</span><span class='op'>^</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>n</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>-</span> <span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>k</span><span class='op'>)</span> <span class='op'>*</span> <span class='va'>y</span><span class='op'>)</span>, <span class='fu'>rlang</span><span class='fu'>::</span><span class='fu'><a href='https://rlang.r-lib.org/reference/expr.html'>expr</a></span><span class='op'>(</span><span class='op'>-</span><span class='op'>(</span><span class='op'>!</span><span class='op'>!</span><span class='va'>lambda</span><span class='op'>)</span> <span class='op'>*</span> <span class='va'>z</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>|&gt;</span></span>
<span>        <span class='fu'><a href='https://rdrr.io/r/base/expression.html'>as.expression</a></span><span class='op'>(</span><span class='op'>)</span></span>
<span>    <span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>1614</span><span class='op'>)</span></span>
<span>    <span class='va'>single_output_gene</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/sim_SDE.html'>sim_SDE</a></span><span class='op'>(</span>drift <span class='op'>=</span> <span class='va'>drift_gene</span>, diffusion <span class='op'>=</span> <span class='va'>diffusion_gene</span>,</span>
<span>        N <span class='op'>=</span> <span class='fl'>1e+06</span>, M <span class='op'>=</span> <span class='fl'>10</span>, Dt <span class='op'>=</span> <span class='fl'>0.1</span>, x0 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>0</span>, <span class='fl'>1</span><span class='op'>)</span>, keep_full <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span></span>
<span>    <span class='va'>single_output_gene2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/do.call.html'>do.call</a></span><span class='op'>(</span><span class='va'>rbind</span>, <span class='va'>single_output_gene</span><span class='op'>)</span></span>
<span>    <span class='va'>single_output_gene2</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='va'>single_output_gene2</span><span class='op'>[</span>, <span class='st'>"X"</span><span class='op'>]</span> <span class='op'>-</span> <span class='va'>single_output_gene2</span><span class='op'>[</span>,</span>
<span>        <span class='st'>"Y"</span><span class='op'>]</span>, <span class='va'>single_output_gene2</span><span class='op'>[</span>, <span class='st'>"Z"</span><span class='op'>]</span><span class='op'>)</span></span>
<span>    <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>colnames</a></span><span class='op'>(</span><span class='va'>single_output_gene2</span><span class='op'>)</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"delta_x"</span>, <span class='st'>"a"</span><span class='op'>)</span></span>
<span>    <span class='va'>single_output_gene2</span></span>
<span><span class='op'>}</span>, bigmemory <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>If the output is saved in an RDS file, upon next use, it can be read as follows.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>saveRDS</a></span><span class='op'>(</span><span class='va'>batch_output_gene</span>, <span class='st'>"batch_output_gene.RDS"</span><span class='op'>)</span></span>
<span><span class='va'>batch_output_gene</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/readRDS.html'>readRDS</a></span><span class='op'>(</span><span class='st'>"batch_output_gene.RDS"</span><span class='op'>)</span> <span class='op'>|&gt;</span></span>
<span>    <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/attach_all_matrices.html'>attach_all_matrices</a></span><span class='op'>(</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<p>We then make the 3D matrix for the batch output, using <code>make_3d_matrix()</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>l_batch_gene_3d</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/make_3d_matrix.html'>make_3d_matrix</a></span><span class='op'>(</span><span class='va'>batch_output_gene</span>, x <span class='op'>=</span> <span class='st'>"delta_x"</span>, y <span class='op'>=</span> <span class='st'>"a"</span>,</span>
<span>    cols <span class='op'>=</span> <span class='st'>"b"</span>, rows <span class='op'>=</span> <span class='st'>"k"</span>, lims <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>5</span>, <span class='fl'>5</span>, <span class='op'>-</span><span class='fl'>0.5</span>, <span class='fl'>2</span><span class='op'>)</span>, h <span class='op'>=</span> <span class='fl'>0.005</span>, Umax <span class='op'>=</span> <span class='fl'>8</span>,</span>
<span>    kde_fun <span class='op'>=</span> <span class='st'>"ks"</span>, individual_landscape <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>For the barrier calculation step, the start and end points of the barrier may be different for each landscape plot. The following code shows how to create a barrier grid for each landscape plot. First, we create a barrier grid template using the function <code>make_barrier_grid_3d()</code>. Next, we modify the barrier grid template to create a barrier grid for the landscape plot.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/make_barrier_grid_3d.html'>make_barrier_grid_3d</a></span><span class='op'>(</span><span class='va'>batch_grid_gene</span>, start_location_value <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1.5</span><span class='op'>)</span>,</span>
<span>    end_location_value <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='op'>-</span><span class='fl'>0.5</span><span class='op'>)</span>, start_r <span class='op'>=</span> <span class='fl'>1</span>, end_r <span class='op'>=</span> <span class='fl'>1</span>, print_template <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>structure(list(start_location_value = list(c(0, 1.5), c(0, 1.5
), c(0, 1.5), c(0, 1.5), c(0, 1.5), c(0, 1.5), c(0, 1.5), c(0, 
1.5), c(0, 1.5)), start_r = list(c(1, 1), c(1, 1), c(1, 1), c(1, 
1), c(1, 1), c(1, 1), c(1, 1), c(1, 1), c(1, 1)), end_location_value = list(
    c(1, -0.5), c(1, -0.5), c(1, -0.5), c(1, -0.5), c(1, -0.5
    ), c(1, -0.5), c(1, -0.5), c(1, -0.5), c(1, -0.5)), end_r = list(
    c(1, 1), c(1, 1), c(1, 1), c(1, 1), c(1, 1), c(1, 1), c(1, 
    1), c(1, 1), c(1, 1))), row.names = c(NA, -9L), class = c(&quot;arg_grid&quot;, 
&quot;data.frame&quot;))</code></pre>
<pre><code>  ele_list   b   k start_location_value start_r end_location_value
1 0.5, 0.5 0.5 0.5             0.0, 1.5    1, 1          1.0, -0.5
2 1.0, 0.5 1.0 0.5             0.0, 1.5    1, 1          1.0, -0.5
3 1.5, 0.5 1.5 0.5             0.0, 1.5    1, 1          1.0, -0.5
4 0.5, 1.0 0.5 1.0             0.0, 1.5    1, 1          1.0, -0.5
5     1, 1 1.0 1.0             0.0, 1.5    1, 1          1.0, -0.5
6 1.5, 1.0 1.5 1.0             0.0, 1.5    1, 1          1.0, -0.5
7 0.5, 1.5 0.5 1.5             0.0, 1.5    1, 1          1.0, -0.5
8 1.0, 1.5 1.0 1.5             0.0, 1.5    1, 1          1.0, -0.5
9 1.5, 1.5 1.5 1.5             0.0, 1.5    1, 1          1.0, -0.5
  end_r
1  1, 1
2  1, 1
3  1, 1
4  1, 1
5  1, 1
6  1, 1
7  1, 1
8  1, 1
9  1, 1</code></pre>
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>bg_gene</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/make_barrier_grid_3d.html'>make_barrier_grid_3d</a></span><span class='op'>(</span><span class='va'>batch_grid_gene</span>, df <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/structure.html'>structure</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>start_location_value <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,</span>
<span>    <span class='fl'>1.5</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1.5</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1.5</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1.5</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1.5</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1.5</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1.5</span><span class='op'>)</span>,</span>
<span>    <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1.5</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1.5</span><span class='op'>)</span><span class='op'>)</span>, start_r <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.2</span>, <span class='fl'>1</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.2</span>, <span class='fl'>1</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.2</span>,</span>
<span>    <span class='fl'>1</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.2</span>, <span class='fl'>0.5</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.2</span>, <span class='fl'>0.5</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.2</span>, <span class='fl'>0.5</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.2</span>, <span class='fl'>0.3</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.2</span>, <span class='fl'>0.3</span><span class='op'>)</span>,</span>
<span>    <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.2</span>, <span class='fl'>0.3</span><span class='op'>)</span><span class='op'>)</span>, end_location_value <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>2</span>, <span class='fl'>0</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>2</span>, <span class='fl'>0</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>2</span>, <span class='fl'>0</span><span class='op'>)</span>,</span>
<span>    <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>0</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>0</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>0</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>0</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>0</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>0</span><span class='op'>)</span><span class='op'>)</span>, end_r <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,</span>
<span>    <span class='fl'>1</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span>,</span>
<span>    <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span>, row.names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='cn'>NA</span>, <span class='op'>-</span><span class='fl'>9L</span><span class='op'>)</span>, class <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"arg_grid"</span>, <span class='st'>"data.frame"</span><span class='op'>)</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>With the barrier grid template, we can calculate the barrier for each landscape plot.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>b_batch_gene_3d</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/calculate_barrier.html'>calculate_barrier</a></span><span class='op'>(</span><span class='va'>l_batch_gene_3d</span>, bg <span class='op'>=</span> <span class='va'>bg_gene</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>If the barrier grid was not needed, the following code can be used to calculate the barrier.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>b_batch_gene_3d</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/calculate_barrier.html'>calculate_barrier</a></span><span class='op'>(</span><span class='va'>l_batch_gene_3d</span>, start_location_value <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,</span>
<span>    <span class='fl'>1.5</span><span class='op'>)</span>, end_location_value <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>0</span><span class='op'>)</span>, start_r <span class='op'>=</span> <span class='fl'>1</span>, end_r <span class='op'>=</span> <span class='fl'>1</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>The resulting landscapes and the MEPs between states are shown in Figure <a href="#fig:bbatch3dgene">7</a>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>l_batch_gene_3d</span><span class='op'>)</span> <span class='op'>+</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/autolayer.html'>autolayer</a></span><span class='op'>(</span><span class='va'>b_batch_gene_3d</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure"><span style="display:block;" id="fig:bbatch3dgene"></span>
<img src="RJ-2025-039_files/figure-html5/bbatch3dgene-1.png" alt="Nine landscape plots arranged in a 3×3 grid. The x-axis is labeled delta x, the y-axis is labeled a. Rows correspond to k values (0.5, 1, 1.5), and columns to b values (0.5, 1, 1.5)." width="100%" />
<p class="caption">
Figure 7: The landscape for the gene expression model of different <span class="math inline">\(b\)</span> and <span class="math inline">\(k\)</span> values. The local minima are marked as white dots, the saddle points are marked as red dots, and the MEPs are marked as white lines.
</p>
</div>
</div>
<p>From the landscapes, it is clear that increasing <span class="math inline">\(b\)</span>, which represents higher strength of gene mutual prohibition, makes the two differentiated states further apart from each other. Increasing <span class="math inline">\(k\)</span>, which represents faster degradation, makes the undifferentiated state disappear earlier, thus makes the differentiation earlier. When <span class="math inline">\(b\)</span> is low enough and <span class="math inline">\(k\)</span> is high enough, there is no differentiation anymore because the two differentiated states merge together and form a more stable state at <span class="math inline">\(\Delta x = 0\)</span>. In this case, there is no actual differentiation in the system, but only a one-to-one conversion of cell types. Only when <span class="math inline">\(b\)</span> is high enough and <span class="math inline">\(k\)</span> is low enough is it possible for the cell to differentiate into two types.</p>
<h3 data-number="3.2" id="example-2-panic-disorder-model"><span class="header-section-number">3.2</span> Example 2: panic disorder model</h3>
<p>The second example we use is the panic disorder model proposed by <span class="citation" data-cites="robinaugh2024">Robinaugh et al. (<a href="#ref-robinaugh2024" role="doc-biblioref">2024</a>)</span>. The model is implemented in the <a href="#">PanicModel</a> package (<a href="https://github.com/jmbh/PanicModel/" class="uri">https://github.com/jmbh/PanicModel/</a>). It contains 12 variables and 33 parameters and also involves history dependency and non-differentiable formula (such as if-else conditions) to model the complex interplay of individual and environmental elements in different time scales. The most important variables of the model are the physical arousal (<span class="math inline">\(A\)</span>) of a person (e.g., heart beat, muscle tension, sweating), the person’s perceived threat (<span class="math inline">\(PT\)</span>, how dangerous the person cognitively evaluates the environment), and the person’s tendency to escape from the situation (<span class="math inline">\(E\)</span>). The core theoretical idea of the model is that physical arousal and perceived threat of a person may strengthen each other in certain circumstances, leading to sudden increases in both variables, manifesting as panic attacks. The tendency that a person tends to use physical arousal as cognitive evidence of threat is represented by another variable, arousal schema <span class="math inline">\(S\)</span>. A comprehensive introduction of the model is beyond the scope of the current article, and we would like to refer interested readers to <span class="citation" data-cites="robinaugh2024">Robinaugh et al. (<a href="#ref-robinaugh2024" role="doc-biblioref">2024</a>)</span>. Here, to simplify the context, we assume that <span class="math inline">\(AS\)</span> does not change over time, and no psychotherapy is being administered. We focus on the influence of <span class="math inline">\(AS\)</span> on the system’s stability represented by <span class="math inline">\(A\)</span> and <span class="math inline">\(PT\)</span>. A graphical illustration of several core variables of this model is shown in Figure <a href="#fig:e2">8</a> (adapted from <span class="citation" data-cites="CuiEtAlMetaphorComputationConstructing2021">Cui et al. (<a href="#ref-CuiEtAlMetaphorComputationConstructing2021" role="doc-biblioref">2023b</a>)</span>).</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:e2"></span>
<img src="figures/diagram-e2.png" alt="Four large circles labeled H, A, PT, and E from left to right. Solid arrows go from A to H, A to PT, PT to A, and PT to E. Dashed arrows go from H to A and from E to PT. The solid arrow from A to PT is labeled AS." width="100%" />
<p class="caption">
Figure 8: A graphical illustration of the relationships between several important psychological variables in the panic disorder model. Solid arrows represent positive relationships and dashed arrows represent negative relationships.
</p>
</div>
</div>
<p>To construct the potential landscapes for this model, we first create a function that performs a simulation using the <code>simPanic()</code> function from <a href="#">PanicModel</a>. This is required as we need to modify some default options for illustration.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>PanicModel</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>sim_fun_panic</span> <span class='op'>&lt;-</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>x0</span>, <span class='va'>par</span><span class='op'>)</span> <span class='op'>{</span></span>
<span></span>
<span>    <span class='co'># Change several default parameters</span></span>
<span>    <span class='va'>pars</span> <span class='op'>&lt;-</span> <span class='va'>pars_default</span></span>
<span>    <span class='co'># Increase the noise strength to improve sampling efficiency</span></span>
<span>    <span class='va'>pars</span><span class='op'>$</span><span class='va'>N</span><span class='op'>$</span><span class='va'>lambda_N</span> <span class='op'>&lt;-</span> <span class='fl'>200</span></span>
<span>    <span class='co'># Make S constant through the simulation</span></span>
<span>    <span class='va'>pars</span><span class='op'>$</span><span class='va'>TS</span><span class='op'>$</span><span class='va'>r_S_a</span> <span class='op'>&lt;-</span> <span class='fl'>0</span></span>
<span>    <span class='va'>pars</span><span class='op'>$</span><span class='va'>TS</span><span class='op'>$</span><span class='va'>r_S_e</span> <span class='op'>&lt;-</span> <span class='fl'>0</span></span>
<span></span>
<span>    <span class='co'># Specify the initial values of A and PT according to the format</span></span>
<span>    <span class='co'># requirement by `multi_init_simulation()`, while the other</span></span>
<span>    <span class='co'># variables use the default initial values.</span></span>
<span>    <span class='va'>initial</span> <span class='op'>&lt;-</span> <span class='va'>initial_default</span></span>
<span>    <span class='va'>initial</span><span class='op'>$</span><span class='va'>A</span> <span class='op'>&lt;-</span> <span class='va'>x0</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span></span>
<span>    <span class='va'>initial</span><span class='op'>$</span><span class='va'>PT</span> <span class='op'>&lt;-</span> <span class='va'>x0</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span></span>
<span></span>
<span>    <span class='co'># Specify the value of S according to the format requirement by</span></span>
<span>    <span class='co'># `batch_simulation()`.</span></span>
<span>    <span class='va'>initial</span><span class='op'>$</span><span class='va'>S</span> <span class='op'>&lt;-</span> <span class='va'>par</span><span class='op'>$</span><span class='va'>S</span></span>
<span></span>
<span>    <span class='co'># Extract the simulation output from the result by simPanic().</span></span>
<span>    <span class='co'># Only keep the core variables.</span></span>
<span>    <span class='kw'><a href='https://rdrr.io/r/base/function.html'>return</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/pkg/PanicModel/man/simPanic.html'>simPanic</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fl'>5000</span>, initial <span class='op'>=</span> <span class='va'>initial</span>, parameters <span class='op'>=</span> <span class='va'>pars</span><span class='op'>)</span><span class='op'>$</span><span class='va'>outmat</span><span class='op'>[</span>,</span>
<span>        <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"A"</span>, <span class='st'>"PT"</span>, <span class='st'>"E"</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='op'>}</span></span></code></pre>
</div>
</div>
<p>We then perform a single simulation from multiple starting points. To speed up the simulation, we use parallel computing.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>future</span><span class='fu'>::</span><span class='fu'><a href='https://future.futureverse.org/reference/plan.html'>plan</a></span><span class='op'>(</span><span class='st'>"multisession"</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>1614</span>, kind <span class='op'>=</span> <span class='st'>"L'Ecuyer-CMRG"</span><span class='op'>)</span></span>
<span><span class='va'>single_output_panic</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/multi_init_simulation.html'>multi_init_simulation</a></span><span class='op'>(</span>sim_fun <span class='op'>=</span> <span class='va'>sim_fun_panic</span>, range_x0 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>,</span>
<span>    <span class='fl'>1</span>, <span class='fl'>0</span>, <span class='fl'>1</span><span class='op'>)</span>, R <span class='op'>=</span> <span class='fl'>4</span>, par <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>S <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<p>The convergence check results of the simulation, shown in Figure <a href="#fig:converge-panic">9</a>, indicate that the time series of the first 100 data points are strongly influenced by the choice of initial value. Therefore, we remove the first 100 data points in the following analysis.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>single_output_panic</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure"><span style="display:block;" id="fig:converge-panic"></span>
<img src="RJ-2025-039_files/figure-html5/converge-panic-1.png" alt="Trace and density plots for variables x, y, and z." width="100%" />
<p class="caption">
Figure 9: The convergence check result for the simulation of the panic disorder model.
</p>
</div>
</div>
<p>We then create the 3D landscape for the panic disorder model, shown in Figure <a href="#fig:3dstaticpanic">10</a>. The landscape shows that the system has two stable states, which are represented by the valleys in the landscape. The system can be trapped in these valleys, leading to different levels of physical arousal and perceived threat. The valley with higher potential value represents a state with higher physical arousal and perceived threat, which corresponds to a panic attack. In contrast, the valley with lower potential value represents a state with lower physical arousal and perceived threat, which corresponds to a healthy state.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="figure"><span style="display:block;" id="fig:3dstaticpanic"></span>
<img src="RJ-2025-039_files/figure-html5/3dstaticpanic-1.png" alt="A landscape plot with two basins." width="50%" />
<p class="caption">
Figure 10: The 3D landscape (potential value as color) for the panic disorder model
</p>
</div>
</div>
<p>We now investigate the effect of the parameter <span class="math inline">\(S\)</span> on the potential landscape. This parameter represents the tendency that a person considers physical arousal as a sign of danger. Therefore, we expect that higher <span class="math inline">\(S\)</span> will stabilize the panic state and destabilize the healthy state.</p>
<p>We perform a batch simulation with varying <span class="math inline">\(S\)</span> values to construct the potential landscapes for different <span class="math inline">\(S\)</span> values. This, again, starts with the creation of a grid of parameter values.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>batch_arg_grid_panic</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/arg_set-class.html'>new_arg_set</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'>|&gt;</span></span>
<span>    <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/arg_set-class.html'>add_arg_ele</a></span><span class='op'>(</span>arg_name <span class='op'>=</span> <span class='st'>"par"</span>, ele_name <span class='op'>=</span> <span class='st'>"S"</span>, start <span class='op'>=</span> <span class='fl'>0</span>, end <span class='op'>=</span> <span class='fl'>1</span>, by <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span> <span class='op'>|&gt;</span></span>
<span>    <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/arg_set-class.html'>make_arg_grid</a></span><span class='op'>(</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>We then perform the batch simulation using parallel computing.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>future</span><span class='fu'>::</span><span class='fu'><a href='https://future.futureverse.org/reference/plan.html'>plan</a></span><span class='op'>(</span><span class='st'>"multisession"</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>1614</span>, kind <span class='op'>=</span> <span class='st'>"L'Ecuyer-CMRG"</span><span class='op'>)</span></span>
<span><span class='va'>batch_output_panic</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/batch_simulation.html'>batch_simulation</a></span><span class='op'>(</span><span class='va'>batch_arg_grid_panic</span>, sim_fun <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>par</span><span class='op'>)</span> <span class='op'>{</span></span>
<span>    <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/multi_init_simulation.html'>multi_init_simulation</a></span><span class='op'>(</span><span class='va'>sim_fun_panic</span>, range_x0 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1</span>, <span class='fl'>0</span>, <span class='fl'>1</span><span class='op'>)</span>, R <span class='op'>=</span> <span class='fl'>4</span>,</span>
<span>        par <span class='op'>=</span> <span class='va'>par</span><span class='op'>)</span> <span class='op'>|&gt;</span></span>
<span>        <span class='fu'><a href='https://rdrr.io/r/stats/window.html'>window</a></span><span class='op'>(</span>start <span class='op'>=</span> <span class='fl'>100</span><span class='op'>)</span></span>
<span><span class='op'>}</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">

</div>
<p>The 3D landscapes for different <span class="math inline">\(S\)</span> values are shown in Figure <a href="#fig:lbatch3dpanic">11</a>. The landscapes show that the system only has one stable state when <span class="math inline">\(S\)</span> is low, but has two stable states when <span class="math inline">\(S\)</span> is high. The stability of the panic state also increases when <span class="math inline">\(S\)</span> is higher. This indicates that a higher <span class="math inline">\(S\)</span> value corresponds to a higher risk of panic attacks.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>l_batch_panic_3d</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://sciurus365.github.io/simlandr/reference/make_3d_matrix.html'>make_3d_matrix</a></span><span class='op'>(</span><span class='va'>batch_output_panic</span>, x <span class='op'>=</span> <span class='st'>"A"</span>, y <span class='op'>=</span> <span class='st'>"PT"</span>,</span>
<span>    cols <span class='op'>=</span> <span class='st'>"S"</span>, h <span class='op'>=</span> <span class='fl'>0.005</span>, lims <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='op'>-</span><span class='fl'>1</span>, <span class='fl'>1.5</span>, <span class='op'>-</span><span class='fl'>0.5</span>, <span class='fl'>1.5</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span><span class='va'>l_batch_panic_3d</span><span class='op'>)</span></span></code></pre>
</div>
<div class="figure"><span style="display:block;" id="fig:lbatch3dpanic"></span>
<img src="RJ-2025-039_files/figure-html5/lbatch3dpanic-1.png" alt="Three landscape plots arranged in a row. The x-axis is labeled A, the y-axis is labeled PT. Columns correspond to S values of 0, 0.5, and 1." width="100%" />
<p class="caption">
Figure 11: The landscape for the panic disorder model of different <span class="math inline">\(S\)</span> values. Two landscapes are shown for different variable combinations, <span class="math inline">\(A\)</span> and <span class="math inline">\(PT\)</span>, or <span class="math inline">\(A\)</span> and <span class="math inline">\(E\)</span>.
</p>
</div>
</div>
<h2 data-number="4" id="discussion"><span class="header-section-number">4</span> Discussion</h2>
<p>Potential landscapes can show the stability of states for a dynamical system in an intuitive and quantitative way. They are especially informative for multistable systems. In this article, we illustrated how to construct potential landscapes using <a href="https://cran.r-project.org/package=simlandr">simlandr</a>. The potential landscapes generated by <a href="https://cran.r-project.org/package=simlandr">simlandr</a> are based on the steady-state distribution of the system, which is in turn estimated using Monte Carlo simulation. Compared to analytic methods, Monte Carlo estimation is more flexible and thus more applicable for complex models. The flexibility comes together with a higher demand for time and storage, which is necessary to make the estimation precise enough. The hash_big_matrix class partly solved this problem by dumping the memory storage to hard disk space. Also, it is important that the simulation function itself is efficient enough. The functions <code>sim_SDE()</code> and <code>multi_init_simulation()</code> make use of the efficient simulations provided by <a href="https://cran.r-project.org/package=Sim.DiffProc">Sim.DiffProc</a> <span class="citation" data-cites="GuidoumBoukhetalaPerformingParallelMonte2020">(<a href="#ref-GuidoumBoukhetalaPerformingParallelMonte2020" role="doc-biblioref">Guidoum and Boukhetala 2020</a>)</span> and the parallel computing with the <a href="https://cran.r-project.org/package=future">future</a> framework <span class="citation" data-cites="RJ-2021-048">(<a href="#ref-RJ-2021-048" role="doc-biblioref">Bengtsson 2021</a>)</span>. For customized simulation functions, there are also multiple approaches that can be used to improve the performance, for which we refer interested readers to <span class="citation" data-cites="WickhamImprovingPerformance2019a">Wickham (<a href="#ref-WickhamImprovingPerformance2019a" role="doc-biblioref">2019</a>)</span>. In Supplementary Materials A, we provide a benchmark of the typical time and memory usage of the procedures in <a href="https://cran.r-project.org/package=simlandr">simlandr</a>. From there, we can see that time and memory usage are acceptable in most cases on a personal computer. When the transition between attractors is rare, the <code>multi_init_simulation()</code> function may help to speed up the convergence, and more advanced sampling methods like importance sampling or rare event sampling may be needed in more complex situations. The detailed ways to implement such methods are highly dependent on the specific model and are beyond the scope of this package. We direct interested readers to <span class="citation" data-cites="RubinoTuffinRareEventSimulation2009">Rubino and Tuffin (<a href="#ref-RubinoTuffinRareEventSimulation2009" role="doc-biblioref">2009</a>)</span> and <span class="citation" data-cites="KloekvanDijkBayesianEstimatesEquation1978">Kloek and van Dijk (<a href="#ref-KloekvanDijkBayesianEstimatesEquation1978" role="doc-biblioref">1978</a>)</span> for a comprehensive review of rare event simulation methods. Nevertheless, the landscape construction functions in <a href="https://cran.r-project.org/package=simlandr">simlandr</a> allow users to provide weights for the simulation results, which can be used to adjust the sampling distribution.</p>
<p>In addition, the length of the simulation and the choice of noise strength may also have an important influence on the results. If the length is too short, the density estimation will be inaccurate, resulting in rugged landscapes. If the length is too long, the simulation part would require more computational resources, which is not always realistic. If the noise is too weak, the system may not be able to converge in a reasonable time, resulting in problems in convergence checks, overly noisy landscapes, or failure to show valleys that are theoretically present. If the noise is too strong, the simulation may be unstable and the boundaries between valleys may be blurry. In Supplementary Materials B, we showed the influence of simulation length and noise strength on the landscape output. With some theoretical expectation of the system’s behavior, it is not difficult to spot that the simulation is too short, or the noise level might be unsuitable. In that case, some adjustments are required before the landscape can be well constructed.</p>
<p>All landscape construction and barrier calculation functions in <a href="https://cran.r-project.org/package=simlandr">simlandr</a> contain both visual aids and numerical data that can be used for further processing. The html plots based on <a href="https://cran.r-project.org/package=plotly">plotly</a> are more suitable for interactive illustrations, while it is also possible to export them to static plots using <code>plotly::orca()</code>. The <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> plots are readily usable for flat printing.</p>
<p>We also want to note some limitations of the potential landscape generated by <a href="https://cran.r-project.org/package=simlandr">simlandr</a>. First, the generalized potential landscape is not a complete description of all dynamics in a system. It emphasizes the stability of different states by filtering out other dynamical information. Some behaviors are not possible in gradient systems (e.g., oscillations and loops), thus cannot be shown in a potential landscape <span class="citation" data-cites="zhouConstructionLandscapeMultistable2016">(<a href="#ref-zhouConstructionLandscapeMultistable2016" role="doc-biblioref">Zhou and Li 2016</a>)</span>. Second, since the steady-state distribution is estimated using a kernel smoothing method, which depends on stochastic simulations, the resulting potential function may not be highly accurate. Its accuracy is further affected by the choice of kernel bandwidth and noise strength. This issue is particularly pronounced at valley edges, where fewer samples are available for estimation. Similar limitations apply to MEP calculations, as they are derived from the generalized landscape rather than the original dynamics. Therefore, we do not recommend directly interpreting the potential function or barrier height results for applications requiring high precision. Instead, the potential landscape is best used as a semi-quantitative tool to gain insights into the system’s overall behavior, guide further analysis, and compare system behavior under different parameter settings, provided the same simulation and kernel estimation conditions are used. The examples in this article illustrated some typical use cases we recommend.</p>
<h2 data-number="5" id="availability-and-future-directions"><span class="header-section-number">5</span> Availability and future directions</h2>
<p>This package is publicly available from the Comprehensive R Archive Network (CRAN) at , under GPL-3 license. The results in the current article were generated with <a href="https://cran.r-project.org/package=simlandr">simlandr</a> 0.4.0 version. R script to replicate all the results in this article can be found in the supporting information.</p>
<p>The barrier height data calculated by <a href="https://cran.r-project.org/package=simlandr">simlandr</a> can also be further analyzed and visualized. For example, sometimes it is helpful to look into how the barrier height changes with varying parameters (e.g., <span class="citation" data-cites="CuiEtAlMetaphorComputationConstructing2021">Cui et al. (<a href="#ref-CuiEtAlMetaphorComputationConstructing2021" role="doc-biblioref">2023b</a>)</span>). We encourage users to explore other ways of analyzing and visualizing the various results provided by <a href="https://cran.r-project.org/package=simlandr">simlandr</a>.</p>
<p>The method we chose for <a href="https://cran.r-project.org/package=simlandr">simlandr</a> is not the only possible one. The generalized landscape by <span class="citation" data-cites="wangPotentialLandscapeFlux2008">Wang et al. (<a href="#ref-wangPotentialLandscapeFlux2008" role="doc-biblioref">2008</a>)</span>, which we implemented, is more flexible and emphasizes the possibility that the system is in a specific state, while other methods may have other strengths (e.g., the method by <span class="citation" data-cites="rodriguez-sanchezPabRodRolldownPostpublication2020">Rodríguez-Sánchez (<a href="#ref-rodriguez-sanchezPabRodRolldownPostpublication2020" role="doc-biblioref">2020</a>)</span> emphasizes the gradient part of the vector field, and the method by <span class="citation" data-cites="MooreEtAlQPotPackageStochastic2016">Moore et al. (<a href="#ref-MooreEtAlQPotPackageStochastic2016" role="doc-biblioref">2016</a>)</span> emphasizes the possibility of transition processes under small noise). We look forward to future theoretical and methodological developments in this direction.</p>
<h2 data-number="6" id="acknowledgments"><span class="header-section-number">6</span> Acknowledgments</h2>
<p>TL was supported by the NSFC under Grant No. 11825102 and the Beijing Academy of Artificial Intelligence (BAAI). ALA and MO were supported by an NWO VIDI grant, Grant No. VI.Vidi.191.178.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h3 class="appendix" data-number="6.1" id="supplementary-materials"><span class="header-section-number">6.1</span> Supplementary materials</h3>
<p>Supplementary materials are available in addition to this article. It can be downloaded at
<a href="RJ-2025-039.zip">RJ-2025-039.zip</a></p>
<h3 class="appendix" data-number="6.2" id="cran-packages-used"><span class="header-section-number">6.2</span> CRAN packages used</h3>
<p><a href="https://cran.r-project.org/package=simlandr">simlandr</a>, <a href="https://cran.r-project.org/package=waydown">waydown</a>, <a href="https://cran.r-project.org/package=Sim.DiffProc">Sim.DiffProc</a>, <a href="https://cran.r-project.org/package=future">future</a>, <a href="https://cran.r-project.org/package=coda">coda</a>, <a href="https://cran.r-project.org/package=bigmemory">bigmemory</a>, <a href="https://cran.r-project.org/package=digest">digest</a>, <a href="https://cran.r-project.org/package=ks">ks</a>, <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a>, <a href="https://cran.r-project.org/package=plotly">plotly</a></p>
<h3 class="appendix" data-number="6.3" id="cran-task-views-implied-by-cited-packages"><span class="header-section-number">6.3</span> CRAN Task Views implied by cited packages</h3>
<p><a href="https://cran.r-project.org/view=Bayesian">Bayesian</a>, <a href="https://cran.r-project.org/view=ChemPhys">ChemPhys</a>, <a href="https://cran.r-project.org/view=DifferentialEquations">DifferentialEquations</a>, <a href="https://cran.r-project.org/view=DynamicVisualizations">DynamicVisualizations</a>, <a href="https://cran.r-project.org/view=Finance">Finance</a>, <a href="https://cran.r-project.org/view=GraphicalModels">GraphicalModels</a>, <a href="https://cran.r-project.org/view=HighPerformanceComputing">HighPerformanceComputing</a>, <a href="https://cran.r-project.org/view=NetworkAnalysis">NetworkAnalysis</a>, <a href="https://cran.r-project.org/view=Phylogenetics">Phylogenetics</a>, <a href="https://cran.r-project.org/view=Psychometrics">Psychometrics</a>, <a href="https://cran.r-project.org/view=Spatial">Spatial</a>, <a href="https://cran.r-project.org/view=TeachingStatistics">TeachingStatistics</a>, <a href="https://cran.r-project.org/view=TimeSeries">TimeSeries</a>, <a href="https://cran.r-project.org/view=WebTechnologies">WebTechnologies</a></p>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-AoPotentialStochasticDifferential2004" class="csl-entry" role="listitem">
P. Ao. Potential in stochastic differential equations: Novel construction. <em>Journal of Physics A: Mathematical and General</em>, 37(3): L25–L30, 2004. DOI <a href="https://doi.org/10.1088/0305-4470/37/3/L01">10.1088/0305-4470/37/3/L01</a>.
</div>
<div id="ref-RJ-2021-048" class="csl-entry" role="listitem">
H. Bengtsson. A unifying framework for parallel and distributed processing in <span>R</span> using futures. <em>The R Journal</em>, 13(2): 208–227, 2021. DOI <a href="https://doi.org/10.32614/RJ-2021-048">10.32614/RJ-2021-048</a>.
</div>
<div id="ref-ChaconDuongMultivariateKernelSmoothing2018" class="csl-entry" role="listitem">
J. E. Chacón and T. Duong. <em>Multivariate kernel smoothing and its applications.</em> <span>New York</span>: <span>Chapman and Hall/CRC</span>, 2018. DOI <a href="https://doi.org/10.1201/9780429485572">10.1201/9780429485572</a>.
</div>
<div id="ref-CuiEtAlCommentsClimbingEscher2023" class="csl-entry" role="listitem">
J. Cui, A. Lichtwarck-Aschoff and F. Hasselman. Comments on "<span class="nocase"><span>Climbing <span>Escher</span></span>’s Stairs</span>: <span>A</span> way to approximate stability landscapes in multidimensional systems". 2023a. DOI <a href="https://doi.org/10.48550/arXiv.2312.09690">10.48550/arXiv.2312.09690</a>.
</div>
<div id="ref-CuiEtAlMetaphorComputationConstructing2021" class="csl-entry" role="listitem">
J. Cui, A. Lichtwarck-Aschoff, M. Olthof, T. Li and F. Hasselman. From metaphor to computation: Constructing the potential landscape for multivariate psychological formal models. <em>Multivariate Behavioral Research</em>, 58(4): 743–761, 2023b. DOI <a href="https://doi.org/10.1080/00273171.2022.2119927">10.1080/00273171.2022.2119927</a>.
</div>
<div id="ref-dahiyaOrderedLineIntegral2018" class="csl-entry" role="listitem">
D. Dahiya and M. Cameron. Ordered line integral methods for computing the quasi-potential. <em>Journal of Scientific Computing</em>, 75(3): 1351–1384, 2018. DOI <a href="https://doi.org/10.1007/s10915-017-0590-9">10.1007/s10915-017-0590-9</a>.
</div>
<div id="ref-dijkstra1959note" class="csl-entry" role="listitem">
E. W. Dijkstra. A note on two problems in connexion with graphs. <em>Numerische Mathematik</em>, 1(1): 269–271, 1959. DOI <a href="https://doi.org/10.1007/BF01386390">10.1007/BF01386390</a>.
</div>
<div id="ref-EVanden-EijndenTransitionpathTheoryPathfinding2010" class="csl-entry" role="listitem">
W. E and E. Vanden-Eijnden. Transition-path theory and path-finding algorithms for the study of rare events. <em>Annual Review of Physical Chemistry</em>, 61(1): 391–420, 2010. DOI <a href="https://doi.org/10.1146/annurev.physchem.040808.090412">10.1146/annurev.physchem.040808.090412</a>.
</div>
<div id="ref-EddelbuettelEtAlPkgdigestCreateCompact2021" class="csl-entry" role="listitem">
D. Eddelbuettel, A. Lucas, J. Tuszynski, H. Bengtsson, S. Urbanek, M. Frasca, B. Lewis, M. Stokely, H. Muehleisen, D. Murdoch, et al. <em><span class="nocase">digest</span>: Create compact hash digests of <span>R</span> objects.</em> 2021. DOI <a href="https://doi.org/10.32614/CRAN.package.digest">10.32614/CRAN.package.digest</a>.
</div>
<div id="ref-FreidlinWentzellRandomPerturbationsDynamical2012" class="csl-entry" role="listitem">
M. I. Freidlin and A. D. Wentzell. <em>Random perturbations of dynamical systems.</em> 3rd ed Berlin: Springer, 2012. DOI <a href="https://doi.org/10.1007/978-3-642-25847-3">10.1007/978-3-642-25847-3</a>.
</div>
<div id="ref-GuidoumBoukhetalaPerformingParallelMonte2020" class="csl-entry" role="listitem">
A. C. Guidoum and K. Boukhetala. Performing parallel <span>Monte Carlo</span> and moment equations methods for <span>Itô</span> and <span>Stratonovich</span> stochastic differential systems: <span>R</span> package <span>Sim</span>.<span>DiffProc</span>. <em>Journal of Statistical Software</em>, 96: 1–82, 2020. DOI <a href="https://doi.org/10.18637/jss.v096.i02">10.18637/jss.v096.i02</a>.
</div>
<div id="ref-hayesComplexSystemsApproach2020" class="csl-entry" role="listitem">
A. M. Hayes and L. A. Andrews. A complex systems approach to the study of change in psychotherapy. <em>BMC Medicine</em>, 18(1): 197, 2020. DOI <a href="https://doi.org/10.1186/s12916-020-01662-2">10.1186/s12916-020-01662-2</a>.
</div>
<div id="ref-HeymannVanden-EijndenPathwaysMaximumLikelihood2008" class="csl-entry" role="listitem">
M. Heymann and E. Vanden-Eijnden. Pathways of maximum likelihood for rare events in nonequilibrium systems: Application to nucleation in the presence of shear. <em>Physical Review Letters</em>, 100(14): 140601, 2008. DOI <a href="https://doi.org/10.1103/PhysRevLett.100.140601">10.1103/PhysRevLett.100.140601</a>.
</div>
<div id="ref-KaneEtAlScalableStrategiesComputing2013" class="csl-entry" role="listitem">
M. J. Kane, J. Emerson and S. Weston. Scalable strategies for computing with massive data. <em>Journal of Statistical Software</em>, 55(14): 1–19, 2013. DOI <a href="https://doi.org/10.18637/jss.v055.i14">10.18637/jss.v055.i14</a>.
</div>
<div id="ref-KloekvanDijkBayesianEstimatesEquation1978" class="csl-entry" role="listitem">
T. Kloek and H. K. van Dijk. <span>Bayesian</span> estimates of equation system parameters: An application of integration by <span>Monte Carlo</span>. <em>Econometrica</em>, 46(1): 1–19, 1978. DOI <a href="https://doi.org/10.2307/1913641">10.2307/1913641</a>.
</div>
<div id="ref-lamotheLinkingBallandcupAnalogy2019" class="csl-entry" role="listitem">
K. A. Lamothe, K. M. Somers and D. A. Jackson. Linking the ball-and-cup analogy and ordination trajectories to describe ecosystem stability, resistance, and resilience. <em>Ecosphere</em>, 10(3): e02629, 2019. DOI <a href="https://doi.org/10.1002/ecs2.2629">10.1002/ecs2.2629</a>.
</div>
<div id="ref-MooreEtAlQPotPackageStochastic2016" class="csl-entry" role="listitem">
C. M. Moore, C. R. Stieha, B. C. Nolting, M. K. Cameron and K. C. Abbott. <span>QPot</span>: An <span>R</span> package for stochastic differential equation quasi-potential analysis. <em><span>The R Journal</span></em>, 8(2): 19–38, 2016. DOI <a href="https://doi.org/10.32614/RJ-2016-031">10.32614/RJ-2016-031</a>.
</div>
<div id="ref-olthofComplexityTheoryPsychopathology2020" class="csl-entry" role="listitem">
M. Olthof, F. Hasselman, F. Oude Maatman, A. M. T. Bosman and A. Lichtwarck-Aschoff. Complexity theory of psychopathology. <em>Journal of Psychopathology and Clinical Science</em>, 132(3): 314–323, 2023. DOI <a href="https://doi.org/10.1037/abn0000740">10.1037/abn0000740</a>.
</div>
<div id="ref-PlummerEtAlCODAConvergenceDiagnosis2006" class="csl-entry" role="listitem">
M. Plummer, N. Best, K. Cowles and K. Vines. CODA: Convergence diagnosis and output analysis for MCMC. <em>R News</em>, 6(1): 7–11, 2006. URL <a href="https://journal.r-project.org/articles/RN-2006-002/RN-2006-002.pdf">https://journal.r-project.org/articles/RN-2006-002/RN-2006-002.pdf</a>.
</div>
<div id="ref-robinaugh2024" class="csl-entry" role="listitem">
D. J. Robinaugh, J. M. B. Haslbeck, L. J. Waldorp, J. J. Kossakowski, E. I. Fried, A. J. Millner, R. J. McNally, O. Ryan, J. de Ron, H. L. J. van der Maas, et al. Advancing the network theory of mental disorders: A computational model of panic disorder. <em>Psychological Review</em>, 131(6): 1482–1508, 2024. DOI <a href="https://doi.org/10.1037/rev0000515">10.1037/rev0000515</a>.
</div>
<div id="ref-rodriguez-sanchezPabRodRolldownPostpublication2020" class="csl-entry" role="listitem">
P. Rodríguez-Sánchez. PabRod/rolldown: Post-publication update. 2020. DOI <a href="https://doi.org/10.5281/zenodo.3763038">10.5281/zenodo.3763038</a>.
</div>
<div id="ref-rodriguez-sanchezClimbingEscherStairs2020" class="csl-entry" role="listitem">
P. Rodríguez-Sánchez, E. H. van Nes and M. Scheffer. Climbing <span>Escher</span>’s stairs: A way to approximate stability landscapes in multidimensional systems. <em>PLOS Computational Biology</em>, 16(4): e1007788, 2020. DOI <a href="https://doi.org/10.1371/journal.pcbi.1007788">10.1371/journal.pcbi.1007788</a>.
</div>
<div id="ref-RubinoTuffinRareEventSimulation2009" class="csl-entry" role="listitem">
G. Rubino and B. Tuffin. <em>Rare event simulation using <span>Monte Carlo</span> methods.</em> <span>Newark, UNITED KINGDOM</span>: <span>John Wiley &amp; Sons, Incorporated</span>, 2009. DOI <a href="https://doi.org/10.1002/9780470745403">10.1002/9780470745403</a>.
</div>
<div id="ref-SievertInteractiveWebbasedData2020" class="csl-entry" role="listitem">
C. Sievert. <em>Interactive web-based data visualization with <span>R</span>, plotly, and shiny.</em> <span>Chapman and Hall/CRC</span>, 2020. URL <a href="https://plotly-r.com">https://plotly-r.com</a>.
</div>
<div id="ref-StumpfEtAlModelingStemCell2021" class="csl-entry" role="listitem">
P. S. Stumpf, F. Arai and B. D. MacArthur. Modeling stem cell fates using non-<span>Markov</span> processes. <em>Cell Stem Cell</em>, 28(2): 187–190, 2021. DOI <a href="https://doi.org/10.1016/j.stem.2021.01.009">10.1016/j.stem.2021.01.009</a>.
</div>
<div id="ref-WaddingtonPrinciplesDevelopmentDifferentiation1966" class="csl-entry" role="listitem">
C. H. Waddington. <em>Principles of development and differentiation.</em> New York: Macmillan, 1966.
</div>
<div id="ref-WanEtAlFindingTransitionState2024" class="csl-entry" role="listitem">
G. Wan, S. J. Avis, Z. Wang, X. Wang, H. Kusumaatmaja and T. Zhang. Finding transition state and minimum energy path of bistable elastic continua through energy landscape explorations. <em>Journal of the Mechanics and Physics of Solids</em>, 183: 105503, 2024. DOI <a href="https://doi.org/10.1016/j.jmps.2023.105503">10.1016/j.jmps.2023.105503</a>.
</div>
<div id="ref-wangPotentialLandscapeFlux2008" class="csl-entry" role="listitem">
J. Wang, L. Xu and E. Wang. Potential landscape and flux framework of nonequilibrium networks: Robustness, dissipation, and coherence of biochemical oscillations. <em>Proceedings of the National Academy of Sciences</em>, 105(34): 12271–12276, 2008. DOI <a href="https://doi.org/10.1073/pnas.0800579105">10.1073/pnas.0800579105</a>.
</div>
<div id="ref-wangQuantifyingWaddingtonLandscape2011" class="csl-entry" role="listitem">
J. Wang, K. Zhang, L. Xu and E. Wang. Quantifying the <span>Waddington</span> landscape and biological paths for development and differentiation. <em>Proceedings of the National Academy of Sciences</em>, 108(20): 8257–8262, 2011. DOI <a href="https://doi.org/10.1073/pnas.1017017108">10.1073/pnas.1017017108</a>.
</div>
<div id="ref-WickhamGgplot2ElegantGraphics2016" class="csl-entry" role="listitem">
H. Wickham. <em><span class="nocase">ggplot2</span>: Elegant graphics for data analysis.</em> Springer-Verlag New York, 2016. URL <a href="https://ggplot2.tidyverse.org">https://ggplot2.tidyverse.org</a>.
</div>
<div id="ref-WickhamImprovingPerformance2019a" class="csl-entry" role="listitem">
H. Wickham. Improving performance. In <em>Advanced <span>R</span></em>, 2nd ed pages. 531–546 2019. Boca Raton: <span>Chapman and Hall/CRC</span>. URL <a href="https://adv-r.hadley.nz/">https://adv-r.hadley.nz/</a>.
</div>
<div id="ref-zhouConstructionLandscapeMultistable2016" class="csl-entry" role="listitem">
P. Zhou and T. Li. Construction of the landscape for multi-stable systems: Potential landscape, quasi-potential, a-type integral and beyond. <em>The Journal of Chemical Physics</em>, 144(9): 094109, 2016. DOI <a href="https://doi.org/10.1063/1.4943096">10.1063/1.4943096</a>.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
  <h3 id="references">References</h3>
  <div id="references-listing"></div>
  <h3 id="reuse">Reuse</h3>
  <p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
  <h3 id="citation">Citation</h3>
  <p>For attribution, please cite this work as</p>
  <pre class="citation-appendix short">Cui, et al., "simlandr: Simulation-Based Landscape Construction for Dynamical Systems", The R Journal, 2026</pre>
  <p>BibTeX citation</p>
  <pre class="citation-appendix long">@article{RJ-2025-039,
  author = {Cui, Jingmeng and Olthof, Merlijn and Lichtwarck-Aschoff, Anna and Li, Tiejun and Hasselman, Fred},
  title = {simlandr: Simulation-Based Landscape Construction for Dynamical Systems},
  journal = {The R Journal},
  year = {2026},
  note = {https://doi.org/10.32614/RJ-2025-039},
  doi = {10.32614/RJ-2025-039},
  volume = {17},
  issue = {4},
  issn = {2073-4859},
  pages = {173-191}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<!--radix_placeholder_navigation_after_body--><html><body>
<div class="distill-site-nav distill-site-footer">
<p>© The R Foundation, <a href="mailto:r-journal@r-project.org">web page
contact</a>.</p>
</div>
<!--/radix_placeholder_navigation_after_body-->
</body></html>


</body>

</html>
