<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
      margin-bottom: 0em;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
  <title>movieROC: Visualizing the Decision Rules Underlying Binary Classification</title>

  <meta property="description" itemprop="description" content="The receiver operating characteristic (ROC) curve is a graphical tool&#10;commonly used to depict the binary classification accuracy of a&#10;continuous marker in terms of its sensitivity and specificity. The&#10;standard ROC curve assumes a monotone relationship between the marker&#10;and the response, inducing classification subsets of the form&#10;$(c,\infty)$ with $c \in \mathbb{R}$. However, in non-standard cases,&#10;the involved classification regions are not so clear, highlighting the&#10;importance of tracking the decision rules. This paper introduces the R&#10;package movieROC,&#10;which provides visualization tools for understanding the ability of&#10;markers to identify a characteristic of interest, complementing the&#10;ROC curve representation. This tool accommodates multivariate&#10;scenarios and generalizations involving different decision rules. The&#10;main contribution of this package is the visualization of the&#10;underlying classification regions, with the associated gain in&#10;interpretability. Adding the time (videos) as a third dimension, this&#10;package facilitates the visualization of binary classification in&#10;multivariate problems. It constitutes a good tool to generate&#10;graphical material for presentations."/>

  <link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>

  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2026-02-03"/>
  <meta property="article:created" itemprop="dateCreated" content="2026-02-03"/>
  <meta name="article:author" content="Sonia Pérez-Fernández"/>
  <meta name="article:author" content="Pablo Martínez-Camblor"/>
  <meta name="article:author" content="Norberto Corral-Blanco"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="movieROC: Visualizing the Decision Rules Underlying Binary Classification"/>
  <meta property="og:type" content="article"/>
  <meta property="og:description" content="The receiver operating characteristic (ROC) curve is a graphical tool&#10;commonly used to depict the binary classification accuracy of a&#10;continuous marker in terms of its sensitivity and specificity. The&#10;standard ROC curve assumes a monotone relationship between the marker&#10;and the response, inducing classification subsets of the form&#10;$(c,\infty)$ with $c \in \mathbb{R}$. However, in non-standard cases,&#10;the involved classification regions are not so clear, highlighting the&#10;importance of tracking the decision rules. This paper introduces the R&#10;package movieROC,&#10;which provides visualization tools for understanding the ability of&#10;markers to identify a characteristic of interest, complementing the&#10;ROC curve representation. This tool accommodates multivariate&#10;scenarios and generalizations involving different decision rules. The&#10;main contribution of this package is the visualization of the&#10;underlying classification regions, with the associated gain in&#10;interpretability. Adding the time (videos) as a third dimension, this&#10;package facilitates the visualization of binary classification in&#10;multivariate problems. It constitutes a good tool to generate&#10;graphical material for presentations."/>
  <meta property="og:locale" content="en_US"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="movieROC: Visualizing the Decision Rules Underlying Binary Classification"/>
  <meta property="twitter:description" content="The receiver operating characteristic (ROC) curve is a graphical tool&#10;commonly used to depict the binary classification accuracy of a&#10;continuous marker in terms of its sensitivity and specificity. The&#10;standard ROC curve assumes a monotone relationship between the marker&#10;and the response, inducing classification subsets of the form&#10;$(c,\infty)$ with $c \in \mathbb{R}$. However, in non-standard cases,&#10;the involved classification regions are not so clear, highlighting the&#10;importance of tracking the decision rules. This paper introduces the R&#10;package movieROC,&#10;which provides visualization tools for understanding the ability of&#10;markers to identify a characteristic of interest, complementing the&#10;ROC curve representation. This tool accommodates multivariate&#10;scenarios and generalizations involving different decision rules. The&#10;main contribution of this package is the visualization of the&#10;underlying classification regions, with the associated gain in&#10;interpretability. Adding the time (videos) as a third dimension, this&#10;package facilitates the visualization of binary classification in&#10;multivariate problems. It constitutes a good tool to generate&#10;graphical material for presentations."/>

  <!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
  <meta name="citation_title" content="movieROC: Visualizing the Decision Rules Underlying Binary Classification"/>
  <meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2025-035"/>
  <meta name="citation_pdf_url" content="RJ-2025-035.pdf"/>
  <meta name="citation_volume" content="17"/>
  <meta name="citation_issue" content="4"/>
  <meta name="citation_doi" content="10.32614/RJ-2025-035"/>
  <meta name="citation_journal_title" content="The R Journal"/>
  <meta name="citation_issn" content="2073-4859"/>
  <meta name="citation_firstpage" content="59"/>
  <meta name="citation_lastpage" content="79"/>
  <meta name="citation_fulltext_world_readable" content=""/>
  <meta name="citation_online_date" content="2026/02/03"/>
  <meta name="citation_publication_date" content="2026/02/03"/>
  <meta name="citation_author" content="Sonia Pérez-Fernández"/>
  <meta name="citation_author_institution" content="Department of Statistics and Operations Research and Mathematics Didactics, University of Oviedo (Asturias, Spain)"/>
  <meta name="citation_author" content="Pablo Martínez-Camblor"/>
  <meta name="citation_author_institution" content="Department of Anesthesiology, Geisel School of Medicine at Dartmouth (NH, USA); Faculty of Health Sciences, Universidad Autónoma de Chile (Chile)"/>
  <meta name="citation_author" content="Norberto Corral-Blanco"/>
  <meta name="citation_author_institution" content="Department of Statistics and Operations Research and Mathematics Didactics, University of Oviedo (Asturias, Spain)"/>
  <!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Animation: A gallery of animations in statistics and utilities to create animations;citation_author=Yihui Xie;citation_author=Christian Mueller;citation_author=Lijia Yu;citation_author=Weicheng Zhu"/>
  <meta name="citation_reference" content="citation_title=Ks: Kernel smoothing;citation_author=Tarn Duong"/>
  <meta name="citation_reference" content="citation_title=e1071: Misc functions of the department of statistics, probability theory group (formerly: E1071), TU wien;citation_author=David Meyer;citation_author=Evgenia Dimitriadou;citation_author=Kurt Hornik;citation_author=Andreas Weingessel;citation_author=Friedrich Leisch"/>
  <meta name="citation_reference" content="citation_title=The robustness of the &quot;binormal&quot; assumptions used in fitting ROC curves;citation_volume=8;citation_author=James A. Hanley"/>
  <meta name="citation_reference" content="citation_title=smoothROCtime: an R package for time-dependent ROC curve estimation;citation_volume=35;citation_author=Susana Dı́az-Coto"/>
  <meta name="citation_reference" content="citation_title=Analysis of a complex of statistical variables into principal components;citation_publisher=Warwick &amp; York;citation_volume=24;citation_author=Harold Hotelling"/>
  <meta name="citation_reference" content="citation_title=Proper receiver operating characteristic analysis: The bigamma model;citation_volume=4;citation_author=D. D. Dorfman;citation_author=K. S. Berbaum;citation_author=C. E. Metz;citation_author=R. V. Lenth;citation_author=J. A. Hanley;citation_author=H. A. Dagga"/>
  <meta name="citation_reference" content="citation_title=ROC analysis for classification and prediction in practice;citation_publisher=CRC Press;citation_author=C. T. Nakas;citation_author=L. E. Bantis;citation_author=C. Gatsonis"/>
  <meta name="citation_reference" content="citation_title=Ks: Kernel density estimation and kernel discriminant analysis for multivariate data in R;citation_volume=21;citation_author=Tarn Duong"/>
  <meta name="citation_reference" content="citation_title=Optimal classification scores based on multivariate marker transformations;citation_volume=105;citation_author=Pablo Martı́nez-Camblor;citation_author=Sonia Pérez-Fernández;citation_author=Susana Dı́az-Coto"/>
  <meta name="citation_reference" content="citation_title=The area under the generalized receiver-operating characteristic curve;citation_publisher=De Gruyter;citation_volume=18;citation_author=Pablo Martı́nez-Camblor;citation_author=Sonia Pérez-Fernández;citation_author=Susana Dı́az-Coto"/>
  <meta name="citation_reference" content="citation_title=The length of the receiver operating characteristic curve and the two cutoff Youden index within a robust framework for discovery, evaluation, and cutoff estimation in biomarker studies involving improper receiver operating characteristic curves;citation_volume=40;citation_author=Leonidas E. Bantis;citation_author=John V. Tsimikas;citation_author=Gregory R. Chambers;citation_author=Michela Capello;citation_author=Samir Hanash;citation_author=Ziding Feng"/>
  <meta name="citation_reference" content="citation_title=A min–max combination of biomarkers to improve diagnostic accuracy;citation_volume=30;citation_author=Chunling Liu;citation_author=Aiyi Liu;citation_author=Susan Halabi"/>
  <meta name="citation_reference" content="citation_title=Smooth non-parametric receiver operating characteristic (ROC) curves for continuous diagnostic tests;citation_volume=16;citation_author=Kelly H. Zou;citation_author=W. J. Hall;citation_author=David E. Shapiro"/>
  <meta name="citation_reference" content="citation_title=Nonparametric and semiparametric estimation of the receiver operating characteristic curve;citation_publisher=The Institute of Mathematical Statistics;citation_volume=24;citation_author=Fushing Hsieh;citation_author=Bruce W. Turnbull"/>
  <meta name="citation_reference" content="citation_title=On the paper “Notes on the overlap measure as an alternative to the Youden index”;citation_volume=37;citation_author=P. Martı́nez-Camblor"/>
  <meta name="citation_reference" content="citation_title=Biomarkers: Potential uses and limitations;citation_volume=1;citation_author=R. Mayeux"/>
  <meta name="citation_reference" content="citation_title=ROCnReg: An R package for receiver operating characteristic curve inference with and without covariates;citation_volume=13;citation_author=Marı́a Xosé Rodrı́guez-Álvarez;citation_author=Vanda Inácio"/>
  <meta name="citation_reference" content="citation_title=nsROC: An R package for non-standard ROC curve analysis;citation_volume=10;citation_author=Sonia Pérez-Fernández;citation_author=Pablo Martı́nez-Camblor;citation_author=Peter Filzmoser;citation_author=Norberto Corral"/>
  <meta name="citation_reference" content="citation_title=OptimalCutpoints: An R package for selecting optimal cutpoints in diagnostic tests;citation_volume=61;citation_author=Mónica López-Ratón;citation_author=Marı́a Xosé Rodrı́guez-Álvarez;citation_author=Carmen Cadarso Suárez;citation_author=Francisco Gude Sampedro"/>
  <meta name="citation_reference" content="citation_title=PROC: An open-source package for R and S+ to analyze and compare ROC curves;citation_volume=12;citation_author=Xavier Robin;citation_author=Natacha Turck;citation_author=Alexandre Hainard;citation_author=Natalia Tiberti;citation_author=Frédérique Lisacek;citation_author=Jean-Charles Sanchez;citation_author=Markus Müller"/>
  <meta name="citation_reference" content="citation_title=Genome-wide DNA methylation profiles in hepatocellular carcinoma;citation_publisher=Wiley Online Library;citation_volume=55;citation_author=Jing Shen;citation_author=Shuang Wang;citation_author=Yu-Jing Zhang;citation_author=Maya Kappil;citation_author=Hui-Chen Wu;citation_author=Muhammad G Kibriya;citation_author=Qiao Wang;citation_author=Farzana Jasmine;citation_author=Habib Ahsan;citation_author=Po-Huang Lee;citation_author=Genome-wide DNA methylation profiles in hepatocellular carcinoma"/>
  <meta name="citation_reference" content="citation_title=The generalized receiver operating characteristic curve;citation_publisher=Aboa Centre for Economics;citation_number=114;citation_author=Heikki Kauppi"/>
  <meta name="citation_reference" content="citation_title=Statistical evaluation of medical tests;citation_volume=8;citation_author=Vanda Inácio;citation_author=Marı́a Xosé Rodrı́guez-Álvarez;citation_author=Pilar Gayoso-Diz"/>
  <meta name="citation_reference" content="citation_title=Combining biomarkers by maximizing the true positive rate for a fixed false positive rate;citation_volume=63;citation_author=Allison Meisner;citation_author=Marco Carone;citation_author=Margaret Sullivan Pepe;citation_author=Kathleen F. Kerr"/>
  <meta name="citation_reference" content="citation_title=The statistical evaluation of medical tests for classification and prediction;citation_publisher=Oxford University Press;citation_author=Margaret Sullivan Pepe"/>
  <meta name="citation_reference" content="citation_title=Statistical methods in diagnostic medicine;citation_publisher=Wiley;citation_volume=414;citation_author=Xiao-Hua Zhou;citation_author=Donna K. McClish;citation_author=Nancy A. Obuchowski"/>
  <meta name="citation_reference" content="citation_title=Linear combinations of multiple diagnostic markers;citation_volume=88;citation_author=John Q. Su;citation_author=Jun S. Liu"/>
  <meta name="citation_reference" content="citation_title=The meaning and use of the area under a receiver operating characteristic (ROC) curve;citation_volume=143;citation_author=James A. Hanley;citation_author=Barbara J McNeil"/>
  <meta name="citation_reference" content="citation_title=Index for rating diagnostic tests;citation_volume=3;citation_author=W. J. Youden"/>
  <meta name="citation_reference" content="citation_title=Visualizing the decision rules behind the ROC curves: Understanding the classification process;citation_volume=105;citation_author=Sonia Pérez-Fernández;citation_author=Pablo Martı́nez-Camblor;citation_author=Peter Filzmoser;citation_author=Norberto Corral"/>
  <meta name="citation_reference" content="citation_title=Combining diagnostic test results to increase accuracy;citation_volume=1;citation_author=Margaret Sullivan Pepe;citation_author=Mary Lou Thompson"/>
  <meta name="citation_reference" content="citation_title=The Youden index in the generalized receiver operating characteristic curve context;citation_volume=15;citation_author=Pablo Martı́nez-Camblor;citation_author=Juan Carlos Pardo-Fernández"/>
  <meta name="citation_reference" content="citation_title=Improving the biomarker diagnostic capacity via functional transformations;citation_volume=46;citation_author=Pablo Martı́nez-Camblor;citation_author=Sonia Pérez-Fernández;citation_author=Susana Dı́az-Coto"/>
  <meta name="citation_reference" content="citation_title=Parametric estimates for the receiver operating characteristic curve generalization for non-monotone relationships;citation_volume=28;citation_author=Pablo Martı́nez-Camblor;citation_author=Juan Carlos Pardo-Fernández"/>
  <meta name="citation_reference" content="citation_title=Receiver operating characteristic curve generalization for non-monotone relationships;citation_volume=26;citation_author=Pablo Martı́nez-Camblor;citation_author=Norberto Corral;citation_author=Corsino Rey;citation_author=Julio Pascual;citation_author=Eva Cernuda-Morollón"/>
  <meta name="citation_reference" content="citation_title=Linear combination methods to improve diagnostic/prognostic accuracy on future observations;citation_volume=25;citation_author=Le Kang;citation_author=Aiyi Liu;citation_author=Lili Tian"/>
  <meta name="citation_reference" content="citation_title=Combining several screening tests: Optimality of the risk score;citation_volume=58;citation_author=Martin W. McIntosh;citation_author=Margaret Sullivan Pepe"/>
  <meta name="citation_reference" content="citation_title=Rms: Regression modeling strategies;citation_author=Frank E Harrell Jr"/>
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","date_received","journal","volume","issue","slug","packages","preview","bibliography","CTV","legacy_pdf","legacy_converted","output","draft","pdf_url","citation_url","doi","creative_commons","csl"]}},"value":[{"type":"character","attributes":{},"value":["movieROC: Visualizing the Decision Rules Underlying Binary Classification"]},{"type":"character","attributes":{},"value":["The receiver operating characteristic (ROC) curve is a graphical tool\ncommonly used to depict the binary classification accuracy of a\ncontinuous marker in terms of its sensitivity and specificity. The\nstandard ROC curve assumes a monotone relationship between the marker\nand the response, inducing classification subsets of the form\n$(c,\\infty)$ with $c \\in \\mathbb{R}$. However, in non-standard cases,\nthe involved classification regions are not so clear, highlighting the\nimportance of tracking the decision rules. This paper introduces the R\npackage movieROC,\nwhich provides visualization tools for understanding the ability of\nmarkers to identify a characteristic of interest, complementing the\nROC curve representation. This tool accommodates multivariate\nscenarios and generalizations involving different decision rules. The\nmain contribution of this package is the visualization of the\nunderlying classification regions, with the associated gain in\ninterpretability. Adding the time (videos) as a third dimension, this\npackage facilitates the visualization of binary classification in\nmultivariate problems. It constitutes a good tool to generate\ngraphical material for presentations.\n"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","orcid_id","email"]}},"value":[{"type":"character","attributes":{},"value":["Sonia Pérez-Fernández"]},{"type":"character","attributes":{},"value":["Department of Statistics and Operations Research and Mathematics Didactics, University of Oviedo (Asturias, Spain)"]},{"type":"character","attributes":{},"value":["0000-0002-2767-6399"]},{"type":"character","attributes":{},"value":["perezsonia@uniovi.es"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","orcid_id","email"]}},"value":[{"type":"character","attributes":{},"value":["Pablo Martínez-Camblor"]},{"type":"character","attributes":{},"value":["Department of Anesthesiology, Geisel School of Medicine at Dartmouth (NH, USA); Faculty of Health Sciences, Universidad Autónoma de Chile (Chile)"]},{"type":"character","attributes":{},"value":["0000-0001-7845-3905"]},{"type":"character","attributes":{},"value":["Pablo.Martinez-Camblor@hitchcock.org"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","orcid_id","email"]}},"value":[{"type":"character","attributes":{},"value":["Norberto Corral-Blanco"]},{"type":"character","attributes":{},"value":["Department of Statistics and Operations Research and Mathematics Didactics, University of Oviedo (Asturias, Spain)"]},{"type":"character","attributes":{},"value":["0000-0002-6962-6154"]},{"type":"character","attributes":{},"value":["norbert@uniovi.es"]}]}]},{"type":"character","attributes":{},"value":["2026-02-03"]},{"type":"character","attributes":{},"value":["2024-02-09"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"integer","attributes":{},"value":[59]},{"type":"integer","attributes":{},"value":[79]}]},{"type":"integer","attributes":{},"value":[17]},{"type":"integer","attributes":{},"value":[4]},{"type":"character","attributes":{},"value":["RJ-2025-035"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"character","attributes":{},"value":["movieROC","pROC","ROCnReg","OptimalCutpoints","nsROC","rms","ks","e1071","animation"]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["preview.png"]},{"type":"character","attributes":{},"value":["PerFer-MarCam_movieROC.bib"]},{"type":"character","attributes":{},"value":["Cluster","Distributions","DynamicVisualizations","Econometrics","Environmetrics","MachineLearning","Psychometrics","ReproducibleResearch","Survival","TeachingStatistics"]},{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[true]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc","mathjax","md_extension"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js"]},{"type":"character","attributes":{},"value":["-tex_math_single_backslash"]}]}]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["RJ-2025-035.pdf"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2025-035"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2025-035"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"character","attributes":{},"value":["/home/mitchell/R/x86_64-pc-linux-gnu-library/4.5/rjtools/rjournal.csl"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["figures/buildROC_multi3_full.pdf","figures/buildROC_multi3_full.png","figures/buildROC_multi3.pdf","figures/Histogram_3genes.pdf","figures/Histogram_3genes.png","figures/KernelDens.gif","figures/LRM.gif","figures/Meisner.gif","figures/movieROC_extraFunctions.pdf","figures/movieROC_extraFunctions.png","figures/movieROC_g1g2_lrm.pdf","figures/movieROC_g1g2_Meis.pdf","figures/movieROC_g1g2_optimalT.pdf","figures/movieROC_g1g2_PT.pdf","figures/movieROC_mainFunctions.pdf","figures/movieROC_mainFunctions.png","figures/PepeTh.gif","figures/plotfunregions_hroc_lrm_gene18384097.pdf","figures/plotfunregions_hroc_lrm_gene18384097.png","figures/plotregions_gene18384097_1.png","figures/plotregions_gene18384097_2.png","figures/plotregions_gene18384097.pdf","figures/plotregions_gene20202438_1.png","figures/plotregions_gene20202438_2.png","figures/plotregions_gene20202438_3.png","figures/plotregions_gene20202438.pdf","figures/ROCcurves_3genes.pdf","figures/ROCcurves_3genes.png","figures/StandardROC_gene20202438.gif","PerFer-MarCam_movieROC.bib","PerFer-MarCam_movieROC.tex","RJ-2025-035_files/anchor-4.2.2/anchor.min.js","RJ-2025-035_files/bowser-1.9.3/bowser.min.js","RJ-2025-035_files/distill-2.2.21/template.v2.js","RJ-2025-035_files/header-attrs-2.30/header-attrs.js","RJ-2025-035_files/jquery-3.6.0/jquery-3.6.0.js","RJ-2025-035_files/jquery-3.6.0/jquery-3.6.0.min.js","RJ-2025-035_files/jquery-3.6.0/jquery-3.6.0.min.map","RJ-2025-035_files/popper-2.6.0/popper.min.js","RJ-2025-035_files/tippy-6.2.7/tippy-bundle.umd.min.js","RJ-2025-035_files/tippy-6.2.7/tippy-light-border.css","RJ-2025-035_files/tippy-6.2.7/tippy.css","RJ-2025-035_files/tippy-6.2.7/tippy.umd.min.js","RJ-2025-035_files/webcomponents-2.0.0/webcomponents.js","RJournal.sty","RJwrapper.log","RJwrapper.tex"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
    font-size: 100%;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  hr.section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    margin: 0px;
  }


  d-byline {
    border-top: none;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
    border-top: none;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  /* tweak for Pandoc numbered line within distill */
  d-article pre.numberSource code > span {
      left: -2em;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // separator
    var separator = '<hr class="section-separator" style="clear: both"/>';
    // prepend separator above appendix
    $('.d-byline').before(separator);
    $('.d-article').before(separator);

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme, except when numbering line
    // in code chunk
    $('pre:not(.numberLines) code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        var author_name = front_matter.authors[i].author
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        const citeChild = $(this).children()[0]
        // Do not process if @xyz has been used without escaping and without bibliography activated
        // https://github.com/rstudio/distill/issues/466
        if (citeChild === undefined) return true

        if (citeChild.nodeName == "D-FOOTNOTE") {
          var fn = citeChild
          $(this).html(fn.shadowRoot.querySelector("sup"))
          $(this).id = fn.id
          fn.remove()
        }
        var refs = $(this).attr('data-cites').split(" ");
        var refHtml = refs.map(function(ref) {
          // Could use CSS.escape too here, we insure backward compatibility in navigator
          return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
        }).join("\n");
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // fix footnotes in tables (#411)
      // replacing broken distill.pub feature
      $('table d-footnote').each(function() {
        // we replace internal showAtNode methode which is triggered when hovering a footnote
        this.hoverBox.showAtNode = function(node) {
          // ported from https://github.com/distillpub/template/pull/105/files
          calcOffset = function(elem) {
              let x = elem.offsetLeft;
              let y = elem.offsetTop;
              // Traverse upwards until an `absolute` element is found or `elem`
              // becomes null.
              while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                  x += elem.offsetLeft;
                  y += elem.offsetTop;
              }

              return { left: x, top: y };
          }
          // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
          const bbox = node.getBoundingClientRect();
          const offset = calcOffset(node);
          this.show([offset.left + bbox.width, offset.top + bbox.height]);
        }
      })

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      // ignore leaflet img layers (#106)
      figures = figures.filter(':not(img[class*="leaflet"])')
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="RJ-2025-035_files/header-attrs-2.30/header-attrs.js"></script>
  <script src="RJ-2025-035_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="RJ-2025-035_files/popper-2.6.0/popper.min.js"></script>
  <link href="RJ-2025-035_files/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="RJ-2025-035_files/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="RJ-2025-035_files/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="RJ-2025-035_files/anchor-4.2.2/anchor.min.js"></script>
  <script src="RJ-2025-035_files/bowser-1.9.3/bowser.min.js"></script>
  <script src="RJ-2025-035_files/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="RJ-2025-035_files/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Mathjax config (add automatic linebreaks when supported)
      // MathJax = {
      //    tex: {
      //        inlineMath: [['$', '$'], ['\\(', '\\)']],
      //        displayMath: [['$$', '$$'], ['\\[', '\\]']],
      //        tags: 'ams',
      //        multline: true,
      //    },
      //    options: {
      //        linebreaks: { automatic: true },
      //    },
      // };

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      
      // Conditionally add supplementary files button
      if (document.getElementById('supplementary-materials')) {
        // create element safely
        const btn_suppl = document.createElement('a');
        btn_suppl.href = slug + '.zip';
        btn_suppl.className = 'dt-tag';
        btn_suppl.innerHTML = '<i class="fas fa-file-zipper"></i> Supplement';
        buttons.append(btn_suppl);
      }

      // adds Abstract: in front of the first <p> in the title section --
      // unless it happens to be the subtitle (FIXME: this is a bad hack - can't distill do this?)
      var tpar = $("d-title p:not(:empty)").filter(function() {
        return !$(this).hasClass("subtitle");
      }).first();
      if (tpar) {
        const abstract = $('<d-abstract>')
        abstract.append('<b>Abstract:</b><br>')
        abstract.append(tpar) // Move description to d-abstract
        $("d-title p:empty").remove() // Remove empty paragraphs after title
        abstract.append(buttons)
        abstract.insertAfter($('d-title')) // Add abstract section after title */
      }

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
    
    d-appendix .citation-appendix,
    .d-appendix .citation-appendix {
      color: rgb(60, 60, 60);
    }

    d-article h2 {
      border-bottom: 0px solid rgba(0, 0, 0, 0.1);
      padding-bottom: 0rem;
    }
    d-article h3 {
      font-size: 20px;
    }
    d-article h4 {
      font-size: 18px;
      text-transform: none;
    }

    @media (min-width: 1024px) {
      d-article h2 {
        font-size: 32px;
      }
      d-article h3 {
        font-size: 24px;
      }
      d-article h4 {
        font-size: 20px;
      }
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"movieROC: Visualizing the Decision Rules Underlying Binary Classification","description":"The receiver operating characteristic (ROC) curve is a graphical tool\ncommonly used to depict the binary classification accuracy of a\ncontinuous marker in terms of its sensitivity and specificity. The\nstandard ROC curve assumes a monotone relationship between the marker\nand the response, inducing classification subsets of the form\n$(c,\\infty)$ with $c \\in \\mathbb{R}$. However, in non-standard cases,\nthe involved classification regions are not so clear, highlighting the\nimportance of tracking the decision rules. This paper introduces the R\npackage movieROC,\nwhich provides visualization tools for understanding the ability of\nmarkers to identify a characteristic of interest, complementing the\nROC curve representation. This tool accommodates multivariate\nscenarios and generalizations involving different decision rules. The\nmain contribution of this package is the visualization of the\nunderlying classification regions, with the associated gain in\ninterpretability. Adding the time (videos) as a third dimension, this\npackage facilitates the visualization of binary classification in\nmultivariate problems. It constitutes a good tool to generate\ngraphical material for presentations.","doi":"10.32614/RJ-2025-035","authors":[{"author":"Sonia Pérez-Fernández","authorURL":"#","affiliation":"Department of Statistics and Operations Research and Mathematics Didactics, University of Oviedo (Asturias, Spain)","affiliationURL":"#","orcidID":"0000-0002-2767-6399"},{"author":"Pablo Martínez-Camblor","authorURL":"#","affiliation":"Department of Anesthesiology, Geisel School of Medicine at Dartmouth (NH, USA); Faculty of Health Sciences, Universidad Autónoma de Chile (Chile)","affiliationURL":"#","orcidID":"0000-0001-7845-3905"},{"author":"Norberto Corral-Blanco","authorURL":"#","affiliation":"Department of Statistics and Operations Research and Mathematics Didactics, University of Oviedo (Asturias, Spain)","affiliationURL":"#","orcidID":"0000-0002-6962-6154"}],"publishedDate":"2026-02-03T00:00:00.000+11:00","citationText":"Pérez-Fernández, et al., 2026"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>movieROC: Visualizing the Decision Rules Underlying Binary Classification</h1>

<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>The receiver operating characteristic (ROC) curve is a graphical tool
commonly used to depict the binary classification accuracy of a
continuous marker in terms of its sensitivity and specificity. The
standard ROC curve assumes a monotone relationship between the marker
and the response, inducing classification subsets of the form
<span class="math inline">\((c,\infty)\)</span> with <span class="math inline">\(c \in \mathbb{R}\)</span>. However, in non-standard cases,
the involved classification regions are not so clear, highlighting the
importance of tracking the decision rules. This paper introduces the R
package movieROC,
which provides visualization tools for understanding the ability of
markers to identify a characteristic of interest, complementing the
ROC curve representation. This tool accommodates multivariate
scenarios and generalizations involving different decision rules. The
main contribution of this package is the visualization of the
underlying classification regions, with the associated gain in
interpretability. Adding the time (videos) as a third dimension, this
package facilitates the visualization of binary classification in
multivariate problems. It constitutes a good tool to generate
graphical material for presentations.</p></p>
</div>

<div class="d-byline">
  Sonia Pérez-Fernández  (Department of Statistics and Operations Research and Mathematics Didactics, University of Oviedo (Asturias, Spain))
  
,   Pablo Martínez-Camblor  (Department of Anesthesiology, Geisel School of Medicine at Dartmouth (NH, USA); Faculty of Health Sciences, Universidad Autónoma de Chile (Chile))
  
,   Norberto Corral-Blanco  (Department of Statistics and Operations Research and Mathematics Didactics, University of Oviedo (Asturias, Spain))
  
<br/>2026-02-03
</div>

<div class="d-article">
<div class="article">
<h2 data-number="1" id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>The use of data to detect a characteristic of interest is a cornerstone
of many disciplines such as medicine (to diagnose a pathology or to
predict a patient outcome), finance (to detect fraud) or machine
learning (to evaluate a classification algorithm), among others.
Continuous markers are surrogate measures for the characteristic under
study, or predictors of a potential subsequent event. They are measured
in subjects, some of whom have the characteristic (<em>positive</em>), and some
without it (<em>negative</em>). In addition to reliability and feasibility, a
good marker must have two relevant properties: interpretability and
accuracy <span class="citation" data-cites="Mayeux2004">(<a href="#ref-Mayeux2004" role="doc-biblioref">Mayeux 2004</a>)</span>. High binary classification <em>accuracy</em> can be
achieved if there exists a strong relationship between the marker and
the <em>response</em>. The latter is assessed by a <em>gold standard</em> for the presence
or absence of the characteristic of interest. <em>Interpretability</em> refers to
the <em>decision rules</em> or <em>subsets</em> considered in the classification process.
This piece of research seeks to elucidate both desirable properties for
a marker by the implementation of a graphical tool in R language. We
propose a novel approach involving the generation of videos as a
solution to effectively capture the classification procedure for
univariate and multivariate markers. Graphical analysis plays a pivotal
role in data exploration, interpretation, and communication. Its
burgeoning potential is underscored by the fast pace of technological
advances, which empower the creation of insightful graphical
representations.</p>
<p>A usual practice when the binary classification accuracy of a marker is
of interest involves the representation of the <em>Receiver Operating
Characteristic (ROC) curve</em>, summarized by the <em>Area Under the Curve</em> (<em>AUC</em>)
<span class="citation" data-cites="Hanley1982">(<a href="#ref-Hanley1982" role="doc-biblioref">Hanley and McNeil 1982</a>)</span>. The resulting plot reflects the trade-off between the
sensitivity and the complement of the specificity. <em>Sensitivity</em> and
<em>specificity</em> are probabilities of correctly classifying subjects, either
positive or negative, respectively. Mathematically, let <span class="math inline">\(\xi\)</span> and <span class="math inline">\(\chi\)</span>
be the random variables modeling the marker values in the positive and
the negative population, respectively, with <span class="math inline">\(F_\xi(\cdot)\)</span> and
<span class="math inline">\(F_\chi(\cdot)\)</span> their associated cumulative distribution functions.
Assuming that the expected value of the marker is larger in the positive
than in the negative population, the standard ROC curve is based on
<em>classification subsets</em> of the form <span class="math inline">\(s = (c, \infty)\)</span>, where <span class="math inline">\(c\)</span> is the
so-called <em>cut-off value or threshold</em> in the support of the marker <span class="math inline">\(X\)</span>,
<span class="math inline">\(\mathcal{S}(X)\)</span>. One subject is classified as a positive if its marker
value is within this region, and as a negative otherwise. This type of
subsets has two important advantages: first, their interpretability is
clear; second, for each specificity <span class="math inline">\(1-t \in [0,1]\)</span>, the corresponding
<span class="math inline">\(s_t = (c_t, \infty)\)</span> is univocally defined by <span class="math inline">\(c_t = F_\chi^{-1}(1-t)\)</span>
for absolutely continuous markers.</p>
<p>When differences in marker distribution between the negative and the
positive population are only in location but not in shape, then
<span class="math inline">\(F_\chi(\cdot) &lt; F_\xi(\cdot)\)</span>, and the classification is direct by
using these decision rules. However, when this is not the case, the
standard ROC curve may cross the main diagonal, resulting in an
<em>improper</em> curve <span class="citation" data-cites="Dorfman1997">(<a href="#ref-Dorfman1997" role="doc-biblioref">Dorfman et al. 1997</a>)</span>. This may be due to three different
scenarios:</p>
<ol type="i">
<li><p>the behavior of the marker in the two studied populations is
different but it is not possible to determine the decision rules.
Notice that the binary classification problem goes further than
distinction between the two populations: the classification subsets
should be highly likely in one population and highly unlikely in the
other one <span class="citation" data-cites="Camblor2018a">(<a href="#ref-Camblor2018a" role="doc-biblioref">Martı́nez-Camblor 2018</a>)</span>;</p></li>
<li><p>there exists a relationship between the marker and the response with
a potential classification use, but this is not monotone;</p></li>
<li><p>there is no relationship between the marker and the response at all
(main diagonal ROC curve).</p></li>
</ol>
<p>In the second case, we have to define classification subsets different
from standard <span class="math inline">\(s_t=(c_t,\infty)\)</span>. Therefore, the use of the marker
becomes more complex. With the aim of accommodating scenarios where both
higher and lower values of the marker are associated with a higher risk
of having the characteristic, <span class="citation" data-cites="Camblor2017a">Martı́nez-Camblor et al. (<a href="#ref-Camblor2017a" role="doc-biblioref">2017</a>)</span> proposed the so-called
<em>generalized ROC (gROC) curve</em>. This curve tracks the highest sensitivity
for every specificity in the unit interval resulting from subsets of the
form <span class="math inline">\(s_t=(-\infty, x_t^L] \cup (x_t^U, \infty)\)</span> with
<span class="math inline">\(x_t^L \leq x_t^U \in \mathcal{S}(X)\)</span>.</p>
<p>Although final decisions are based on the underlying classification
subsets, they are typically not depicted. This omission is not a
shortcoming in standard cases, as for each specificity <span class="math inline">\(1-t \in [0,1]\)</span>,
there is only one rule of the form <span class="math inline">\(s_t = (c_t, \infty)\)</span> with such
specificity. Particularly, <span class="math inline">\(s_t\)</span> is univocally defined by
<span class="math inline">\(c_t = 1 - F_\chi^{-1}(1-t)\)</span>; and the same applies if we fix a
sensitivity. Nevertheless, if the gROC curve is taken, there are
infinite subsets of the form <span class="math inline">\(s_t=(-\infty, x_t^L] \cup (x_t^U, \infty)\)</span>
resulting in <span class="math inline">\(\mathbb{P}(\chi \in s_t) = t\)</span>. This loss of univocity underlines
the importance of reporting (numerically and/or graphically) the
decision rules actually proposed for classifying. This gap is covered in
the presented package.</p>
<p>An alternative approach to assess the classification performance of a
marker involves considering a transformation of it. This transformation
<span class="math inline">\(h(\cdot)\)</span> aims to capture differences in distribution between the two
populations in the ROC sense. Once <span class="math inline">\(h(\cdot)\)</span> is identified, the
standard ROC curve for <span class="math inline">\(h(X)\)</span> is represented, resulting in the <em>efficient
ROC (eROC) curve</em> <span class="citation" data-cites="Kauppi2016">(<a href="#ref-Kauppi2016" role="doc-biblioref">Kauppi 2016</a>)</span>. Arguing as before, for a fixed
specificity, the classification subsets <span class="math inline">\(s_t=(c_t, \infty)\)</span> in the
transformed space are univocally defined, where a subject is classified
as positive if <span class="math inline">\(h(x) \in s_t\)</span> and negative otherwise (with <span class="math inline">\(x\)</span>
representing its marker value). However, they may have any shape in the
original space, depending on the monotonicity of the functional
transformation <span class="math inline">\(h(\cdot)\)</span> <span class="citation" data-cites="Camblor2019a">(<a href="#ref-Camblor2019a" role="doc-biblioref">Martı́nez-Camblor et al. 2019</a>)</span>. Emphasizing the importance of
tracking the decision rules underlying the eROC curve, this monitoring
process enables an assessment of whether the improved accuracy of the
marker justifies the potential loss in interpretability.</p>
<p>The ROC curve is defined for classification accuracy evaluation of
univariate markers. To deal with multivariate markers, the usual
practice is to consider a transformation <span class="math inline">\(\boldsymbol{h}(\cdot)\)</span> to
reduce it to a univariate one, and then to construct the standard ROC
curve. Same considerations as before apply when a functional
transformation is taken. In the proposed R library, we consider methods
from the literature to define and estimate <span class="math inline">\(\boldsymbol{h}(\cdot)\)</span> in
the multivariate case <span class="citation" data-cites="Kang2016 Meisner2021">(<a href="#ref-Kang2016" role="doc-biblioref">Kang et al. 2016</a>; <a href="#ref-Meisner2021" role="doc-biblioref">Meisner et al. 2021</a>)</span>.</p>
<p>Focusing on the classification subsets underlying the decision rules,
the <a href="https://CRAN.R-project.org/package=movieROC"><strong>movieROC</strong></a> package
incorporates methods to visualize the construction process of ROC curves
by presenting the classification accuracy of these subsets. For
univariate markers, the library includes both the classical (standard
ROC curve) and the generalized (gROC curve) approach. Besides, it enables
the display of decision rules for various transformations of the marker,
seeking to maximize performance and allowing for flexibility in the final
shape of the subsets (eROC curve). For multidimensional markers, the
proposed tool visualizes the evolution of decision subsets when
different objective functions are employed for optimization, even
imposing restrictions on the underlying regions. In this case,
displaying the decision rules associated with every specificity in a
single static image is no longer feasible. Therefore, <em>dynamic
representations</em> (videos) are implemented, drawing on time as an extra
dimension to capture the variation in specificity.</p>
<p>Much software available in R could be discussed here covering diverse
topics related to ROC curves: the
<a href="https://CRAN.R-project.org/package=pROC"><strong>pROC</strong></a> package is a main
reference including tools to visualize, estimate and compare ROC curves
<span class="citation" data-cites="pROC2011">(<a href="#ref-pROC2011" role="doc-biblioref">Robin et al. 2011</a>)</span>; <a href="https://CRAN.R-project.org/package=ROCnReg"><strong>ROCnReg</strong></a>
explicitly considers covariate information to estimate the
covariate-specific and the covariate-adjusted ROC curves <span class="citation" data-cites="ROCnReg2021">(<a href="#ref-ROCnReg2021" role="doc-biblioref">Rodrı́guez-Álvarez and Inácio 2021</a>)</span>;
<a href="https://CRAN.R-project.org/package=smoothROCtime"><strong>smoothROCtime</strong></a>
implements smooth estimation of time-dependent ROC curves based on the
bivariate kernel density estimator for <span class="math inline">\((X, \textit{time-to-event})\)</span>
<span class="citation" data-cites="DiazCoto2020">(<a href="#ref-DiazCoto2020" role="doc-biblioref">Dı́az-Coto 2020</a>)</span>;
<a href="https://CRAN.R-project.org/package=OptimalCutpoints"><strong>OptimalCutpoints</strong></a>
includes point and interval estimation methods for optimal thresholds
<span class="citation" data-cites="OptimalCutpoints2014">(<a href="#ref-OptimalCutpoints2014" role="doc-biblioref">López-Ratón et al. 2014</a>)</span>; and
<a href="https://CRAN.R-project.org/package=nsROC"><strong>nsROC</strong></a> performs
non-standard analysis such as gROC estimation <span class="citation" data-cites="nsROC2018">(<a href="#ref-nsROC2018" role="doc-biblioref">Pérez-Fernández et al. 2018</a>)</span>; among
others.</p>
<p>This paper introduces and elucidates the diverse functionalities of the
newly developed
<a href="https://CRAN.R-project.org/package=movieROC"><strong>movieROC</strong></a> package,
aimed at facilitating the visualization and comprehension of the
decision rules underlying the binary classification process,
encompassing various generalizations. Despite the availability of
numerous R packages implementing related analyses, we have identified
the main gaps covered in this library: tracking the decision rules
underlying the ROC curve, including multivariate markers and
non-standard scenarios (i.e. non-monotonic). The rest of the paper is
structured as follows. In
Section <a href="#section:functionality" data-reference-type="ref" data-reference="section:functionality">2</a>, we introduce the main R functions
and objects implemented, and briefly explain the dataset employed
throughout this manuscript to demonstrate the utility of the R library.
Section <a href="#section:regularroc" data-reference-type="ref" data-reference="section:regularroc">3</a> is devoted to reconsidering the
definition of the standard ROC curve from the perspective of
classification subsets, including an extension to multivariate
scenarios. Sections <a href="#section:groc" data-reference-type="ref" data-reference="section:groc">4</a> and
<a href="#section:efficientroc" data-reference-type="ref" data-reference="section:efficientroc">5</a> revisit the gROC curve and the eROC
curve, respectively, covering various methods to capture the potential
classification accuracy of the marker under study. Each of these
sections begins with a state-of-the-art overview, followed by the main
syntax of the corresponding R functions. In addition, examples of
implementation using the dataset presented in Section
<a href="#subsection:dataset" data-reference-type="ref" data-reference="subsection:dataset">2.3</a> are provided. Finally, the paper
concludes with a concise summary and computational details regarding the
implemented tool.</p>
<h2 data-number="2" id="section:functionality"><span class="header-section-number">2</span> Main functions of the movieROC package and illustrative dataset</h2>
<p>Sections <a href="#subsec:functionality" data-reference-type="ref" data-reference="subsec:functionality">2.1</a> and
<a href="#subsec:functions" data-reference-type="ref" data-reference="subsec:functions">2.2</a> provide a detailed description of the main
objectives of the implemented R functions. To reflect the practical
usage of the developed R package, we employ a real dataset throughout
this manuscript, which is introduced in
Section <a href="#subsection:dataset" data-reference-type="ref" data-reference="subsection:dataset">2.3</a>.</p>
<h3 data-number="2.1" id="subsec:functionality"><span class="header-section-number">2.1</span> Functionality of the movieROC package</h3>
<p>A graphical tool was developed to showcase static and dynamic graphics
displaying the classification subsets derived from maximizing diagnostic
accuracy under certain assumptions, ensuring the preservation of the
interpretability. The R package facilitates the construction of the ROC
curve across various specificities, providing visualizations of the
resulting classification regions. The proposed tool comprises multiple R
functions that generate objects with distinct class attributes (see
function names where red arrows depart from and red nodes in
Figure <a href="#figure:functionality" data-reference-type="ref" data-reference="figure:functionality">1</a>, respectively). Once the object of
interest is created, different methods may be used, in order to plot the
underlying regions (<code>plot_regions()</code>, <code>plot_funregions()</code>), to track the
resulting ROC curve (<code>plot_buildROC()</code>, <code>plot()</code>), to <code>predict</code> decision
rules for a particular specificity, and to <code>print</code> relevant information,
among others. The main function of the package, <code>movieROC()</code>, produces
videos to exhibit the classification procedure.</p>
<figure id="figure:functionality">
<img src="figures/movieROC_mainFunctions.png" style="width:100.0%"
alt="graphic without alt text" />
<img src="figures/movieROC_extraFunctions.png" style="width:100.0%"
alt="graphic without alt text" />
<figcaption>
Figure 1: R functions of the <a
href="https://CRAN.R-project.org/package=movieROC"><strong>movieROC</strong></a>
package. The blue nodes include the names of the R functions and the red
nodes indicate the different R objects that can be created and worked
with. The red arrows depart from those R functions engaged in creating R
objects and the black arrows indicate which R functions can be applied
to which R objects. The grey dashed arrows show internal
dependencies.
</figcaption>
</figure>
<p>It includes algorithms to visualize the regions that underlie the binary
classification problem, considering different approaches:</p>
<ul>
<li><p>make the classification subsets flexible in order to cover
non-standard scenarios, by considering two cut-off values (<code>gROC()</code>
function); explained in
Section <a href="#section:groc" data-reference-type="ref" data-reference="section:groc">4</a>;</p></li>
<li><p>transform the marker by a proper function <span class="math inline">\(h(\cdot)\)</span> (<code>hROC()</code>
function); introduced in
Section <a href="#section:efficientroc" data-reference-type="ref" data-reference="section:efficientroc">5</a>;</p></li>
<li><p>when dealing with multivariate markers, consider a functional
transformation with some fixed or dynamic parameters resulting from
different methods available in the literature (<code>multiROC()</code>
function); covered in
Section <a href="#section:multiroc" data-reference-type="ref" data-reference="section:multiroc">3.1</a>.</p></li>
</ul>
<h3 data-number="2.2" id="subsec:functions"><span class="header-section-number">2.2</span> Class methods for movieROC objects</h3>
<p>By using the <code>gROC()</code>, the <code>multiROC()</code> or the <code>hROC()</code> function, the
user obtains an R object of class ‘<code>groc</code>’, ‘<code>multiroc</code>’ or ‘<code>hroc</code>’,
respectively. These will be called
<a href="https://CRAN.R-project.org/package=movieROC"><strong>movieROC</strong></a> objects.
Once the object of interest is created, the implemented package includes
many functions (methods) to pass to it. Some of them are generic methods
(<code>print()</code>, <code>plot()</code> and <code>predict()</code>), commonly used in R language over
different objects according to their class attributes. The rest of the
functions are specific for this library and therefore only applicable to
<a href="https://CRAN.R-project.org/package=movieROC"><strong>movieROC</strong></a> objects.
The following outline summarizes all these functions and
provides their target and main syntax (with default input parameters).</p>
<h4 class="unnumbered" data-number="2.2.1" id="generic-functions">Generic functions</h4>
<ul>
<li><strong><code>print()</code></strong>: Print some relevant information.</li>
<li><strong><code>plot()</code></strong>: Plot the ROC curve estimate.</li>
<li><strong><code>predict()</code></strong>: Print the classification subsets corresponding to a particular false-positive rate (<code>FPR</code>) introduced by the user. For a ‘<code>groc</code>’ object, the user may specify a cut-off value <code>C</code> (for the standard ROC curve) or two cut-off values <code>XL</code> and <code>XU</code> (for the gROC curve).</li>
</ul>
<h4 class="unnumbered" data-number="2.2.2" id="specific-functions">Specific functions</h4>
<ul>
<li><strong><code>plot_regions()</code></strong></li>
</ul>
<p>Applicable to a ‘<code>groc</code>’ or a ‘<code>hroc</code>’ object. Plot two graphics in the same figure: left, classification subsets for each false-positive rate (grey color by default); right, <span class="math inline">\(90^\circ\)</span> rotated ROC curve.</p>
<em>Main syntax:</em>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span>  <span class='fu'>plot_regions</span><span class='op'>(</span><span class='va'>obj</span>, plot.roc <span class='op'>=</span> <span class='cn'>TRUE</span>, plot.auc <span class='op'>=</span> <span class='cn'>FALSE</span>, FPR <span class='op'>=</span> <span class='fl'>0.15</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>If the input parameter <code>FPR</code> is specified, the corresponding classification region reporting such false-positive rate and the point in the ROC curve are highlighted in blue color.</p>
<ul>
<li><strong><code>plot_funregions()</code></strong></li>
</ul>
<p>Applicable to a ‘<code>groc</code>’ or a ‘<code>hroc</code>’ object.
Plot the transforming function and the classification subsets reporting the false-positive rate(s) indicated in the input parameter(s) <code>FPR</code> and <code>FPR2</code>.</p>
<em>Main syntax:</em>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>plot_funregions</span><span class='op'>(</span><span class='va'>obj</span>, FPR <span class='op'>=</span> <span class='fl'>0.15</span>, FPR2 <span class='op'>=</span> <span class='cn'>NULL</span>, plot.subsets <span class='op'>=</span> <span class='cn'>TRUE</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<ul>
<li><strong><code>plot_buildROC()</code></strong></li>
</ul>
<p>Applicable to a ‘<code>groc</code>’ or a ‘<code>multiroc</code>’ object.</p>
<ul>
<li>For a ‘<code>groc</code>’ object: Plot four (if input <code>reduce</code> is FALSE) or two (if <code>reduce</code> is TRUE, only those on the top) graphics in the same figure: top-left, density function estimates for the marker in both populations with the areas corresponding to FPR and TPR colored (blue and red, respectively) for the optional input parameter <code>FPR</code>, <code>C</code> or <code>XL, XU</code>; top-right, the empirical ROC curve estimate; bottom-left, boxplots in both groups; bottom-right, classification subsets for every FPR (grey color).</li>
</ul>
<em>Main syntax:</em>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>plot_buildROC</span><span class='op'>(</span><span class='va'>obj</span>, FPR <span class='op'>=</span> <span class='fl'>0.15</span>, <span class='va'>C</span>, <span class='va'>XL</span>, <span class='va'>XU</span>, h <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>1</span><span class='op'>)</span>, </span>
<span>              histogram <span class='op'>=</span> <span class='cn'>FALSE</span>, breaks <span class='op'>=</span> <span class='fl'>15</span>, reduce <span class='op'>=</span> <span class='cn'>TRUE</span>, </span>
<span>              build.process <span class='op'>=</span> <span class='cn'>FALSE</span>, completeROC <span class='op'>=</span> <span class='cn'>FALSE</span>,  <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>If <code>build.process</code> is FALSE, the whole ROC curve is displayed; otherwise, if <code>completeROC</code> is TRUE, the portion of the ROC curve until the fixed FPR is highlighted in black and the rest is shown in gray, while if <code>completeROC</code> is FALSE, only the first portion of the curve is displayed.</p>
<ul>
<li>For a ‘<code>multiroc</code>’ object: Plot two graphics in the same figure: right, the ROC curve highlighting the point and the threshold for the resulting univariate marker; left, scatterplot with the marker values in both positive (red color) and negative (blue color) subjects. About the left graphic: for <span class="math inline">\(p=2\)</span>, over the original/feature bivariate space; for <span class="math inline">\(p&gt;2\)</span>, projected over two selected components of the marker (if <code>display.method = "OV"</code> with components selection in <code>displayOV</code>, <code>c(1,2)</code> by default) or the first two principal components from PCA (if <code>display.method = "PCA"</code>, default). The classification subset reporting the <code>FPR</code> selected by the user (<code>FPR</code> <span class="math inline">\(\neq\)</span> <code>NULL</code>) is displayed in gold color.</li>
</ul>
<p><em>Main syntax:</em></p>
for <span class="math inline">\(p=2\)</span>:
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>plot_buildROC</span><span class='op'>(</span><span class='va'>obj</span>, FPR <span class='op'>=</span> <span class='fl'>0.15</span>, </span>
<span>              build.process <span class='op'>=</span> <span class='cn'>FALSE</span>, completeROC <span class='op'>=</span> <span class='cn'>TRUE</span>,  <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
for <span class="math inline">\(p&gt;2\)</span>:
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>plot_buildROC</span><span class='op'>(</span><span class='va'>obj</span>, FPR <span class='op'>=</span> <span class='fl'>0.15</span>, </span>
<span>              display.method <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"PCA"</span>,<span class='st'>"OV"</span><span class='op'>)</span>, displayOV <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>2</span><span class='op'>)</span>, </span>
<span>              build.process <span class='op'>=</span> <span class='cn'>FALSE</span>, completeROC <span class='op'>=</span> <span class='cn'>TRUE</span>,  <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>If <code>build.process</code> is FALSE, the whole ROC curve is displayed; otherwise, if <code>completeROC</code> is TRUE, the portion of the ROC curve until the fixed FPR is highlighted in black and the rest is shown in gray, while if <code>completeROC</code> is FALSE, only the first portion of the curve is shown.</p>
<ul>
<li><strong><code>movieROC()</code></strong></li>
</ul>
<p>Applicable to a ‘<code>groc</code>’ or a ‘<code>multiroc</code>’ object. Save a video as a GIF illustrating the construction of the ROC curve.</p>
<ul>
<li>For a ‘<code>groc</code>’ object:</li>
</ul>
<em>Main syntax:</em>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>movieROC</span><span class='op'>(</span><span class='va'>obj</span>, fpr <span class='op'>=</span> <span class='cn'>NULL</span>, </span>
<span>         h <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>1</span><span class='op'>)</span>, histogram <span class='op'>=</span> <span class='cn'>FALSE</span>, breaks <span class='op'>=</span> <span class='fl'>15</span>, </span>
<span>         reduce <span class='op'>=</span> <span class='cn'>TRUE</span>, completeROC <span class='op'>=</span> <span class='cn'>FALSE</span>, videobar <span class='op'>=</span> <span class='cn'>TRUE</span>, </span>
<span>         file <span class='op'>=</span> <span class='st'>"animation1.gif"</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>For each element in vector <code>fpr</code> (optional input parameter), the function executed is <code>plot_buildROC(obj, FPR = fpr[i], build.process = TRUE, ...)</code>. The vector of false-positive rates illustrated in the video is <code>NULL</code> by default: if length of output parameter <code>t</code> for <code>gROC()</code> function is lower than 150, such vector is taken as <code>fpr</code>; otherwise, an equally-spaced vector of length 100 covering the range of the marker values is considered.</p>
<ul>
<li>For a ‘<code>multiroc</code>’ object:</li>
</ul>
<p><em>Main syntax:</em></p>
for <span class="math inline">\(p=2\)</span>:
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>movieROC</span><span class='op'>(</span><span class='va'>obj</span>, fpr <span class='op'>=</span> <span class='cn'>NULL</span>, </span>
<span>         file <span class='op'>=</span> <span class='st'>"animation1.gif"</span>, save <span class='op'>=</span> <span class='cn'>TRUE</span>, </span>
<span>         border <span class='op'>=</span> <span class='cn'>TRUE</span>, completeROC <span class='op'>=</span> <span class='cn'>FALSE</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
for <span class="math inline">\(p&gt;2\)</span>:
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>movieROC</span><span class='op'>(</span><span class='va'>obj</span>, fpr <span class='op'>=</span> <span class='cn'>NULL</span>, </span>
<span>         display.method <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"PCA"</span>,<span class='st'>"OV"</span><span class='op'>)</span>, displayOV <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>2</span><span class='op'>)</span>, </span>
<span>         file <span class='op'>=</span> <span class='st'>"animation1.gif"</span>, save <span class='op'>=</span> <span class='cn'>TRUE</span>, </span>
<span>         border <span class='op'>=</span> <span class='cn'>TRUE</span>, completeROC <span class='op'>=</span> <span class='cn'>FALSE</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>The video is <code>save</code>d by default as a GIF with the name indicated in argument <code>file</code> (extension <code>.gif</code> should be added). A <code>border</code> for the classification subsets is drawn by default.</p>
<p>For each element in vector <code>fpr</code> (optional input parameter), the function executed is</p>
for <span class="math inline">\(p=2\)</span>:
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>plot_buildROC</span><span class='op'>(</span><span class='va'>obj</span>, FPR <span class='op'>=</span> <span class='va'>fpr</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span>, build.process <span class='op'>=</span> <span class='cn'>TRUE</span>, <span class='va'>completeROC</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
for <span class="math inline">\(p&gt;2\)</span>:
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>plot_buildROC</span><span class='op'>(</span><span class='va'>obj</span>, FPR <span class='op'>=</span> <span class='va'>fpr</span><span class='op'>[</span><span class='va'>i</span><span class='op'>]</span>, build.process <span class='op'>=</span> <span class='cn'>TRUE</span>, <span class='va'>completeROC</span>, </span>
<span>              <span class='va'>display.method</span>, <span class='va'>displayOV</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>Same considerations about the input <code>fpr</code> as those for <code>movieROC()</code> over a ‘<code>groc</code>’ object.</p>
<h3 data-number="2.3" id="subsection:dataset"><span class="header-section-number">2.3</span> Illustrative dataset</h3>
<p>In order to illustrate the functionality of our R package, we consider
the <code>HCC</code> data. This dataset is derived from gene expression arrays of
tumor and adjacent non-tumor tissues of 62 Taiwanese cases of
hepatocellular carcinoma (HCC). The goal of the original study
<span class="citation" data-cites="Shen2012">(<a href="#ref-Shen2012" role="doc-biblioref">Shen et al. 2012</a>)</span> was to identify, with a genome-wide approach, additional
genes hypermethylated in HCC that could be used for more accurate
analysis of plasma DNA for early diagnosis, by using Illumina
methylation arrays (Illumina, Inc., San Diego, CA) that screen 27,578
autosomal CpG sites. The complete dataset was deposited in NCBI’s Gene
Expression Omnibus (GEO) and it is available through series accession
number GSE37988
(<a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE37988" class="uri">www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE37988</a>).
It is included in the presented package (<code>HCC</code> dataset), selecting 948
genes with complete information.</p>
<p>The following code loads the R package and the <code>HCC</code> dataset (see the
<a href="https://cran.r-project.org/web/packages/movieROC/vignettes/movieROC_vignette.pdf">vignette</a>
for main structure).</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">library</span>(movieROC)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">data</span>(HCC)</span></code></pre></div>
<p>We selected the genes 20202438, 18384097, and 03515901. On the one hand,
we chose the gene 03515901 as an example of a monotone relationship
between the marker and the response, reporting a good ROC curve. On the
other hand, relative gene expression intensities of the genes 20202438
and 18384097 tend to be more extreme in tissues with tumor than in those
without it. These are non-standard cases, so if we limit ourselves to
detect “appropriate” genes on the basis of the standard ROC curve, they
would not be chosen. However, extending the decision rules by means of
the gROC curve, those genes may be considered as potential biomarkers
(locations) to differ between the two groups. The R code estimating and
displaying the density probability function for gene expression
intensities of the selected genes in each group
(Figure <a href="#fig:densities" data-reference-type="ref" data-reference="fig:densities">2</a>) is included in the
<a href="https://cran.r-project.org/web/packages/movieROC/vignettes/movieROC_vignette.pdf">vignette</a>.</p>
<figure id="fig:densities">
<img src="figures/Histogram_3genes.png" style="width:100.0%"
alt="graphic without alt text" />
<figcaption>
Figure 2: Density histograms and kernel density estimations
(lighter) for gene expression intensities of the genes 20202438,
18384097 and 03515901 in negative (non-tumor) and positive (tumor)
tissues.
</figcaption>
</figure>
<h2 data-number="3" id="section:regularroc"><span class="header-section-number">3</span> Regular ROC curve</h2>
<p>Assuming that there exists a monotone relationship between the marker
and the response, the <em>regular, right-sided or standard ROC curve</em>
associated with the marker <span class="math inline">\(X\)</span> considers classification subsets of the
form <span class="math inline">\(s_t=(c_t,\infty)\)</span>. For each specificity
<span class="math inline">\(1-t=\mathbb{P}(\chi \notin s_t) \in [0,1]\)</span>, also called <em>true-negative rate</em>,
there exists only one subset <span class="math inline">\(s_t\)</span> reporting such specificity and thus a
particular sensitivity, also called <em>true-positive rate</em>,
<span class="math inline">\(\mathbb{P}(\xi \in s_t)\)</span>. This results in a simple correspondence between each
point of the ROC curve
<span class="math inline">\(\mathcal{R}_r(t) = 1-F_\xi \big(F_\chi^{-1}(1-t)\big)\)</span> and its
associated classification region <span class="math inline">\(s_t \in \mathcal{I}_r(t)\)</span>, where
<span class="math display">\[\mathcal{I}_r(t) = \Big\{ s_t = (c_t, \infty)  :  c_t \in \mathcal{S}(X)  ,  \mathbb{P}(\chi \in s_t) = t \Big\}\]</span>
is the <em>right-sided family of eligible classification subsets</em>. The
definition of this family captures the shape of the decision rules and
the target specificity.</p>
<p>If higher values of the marker are associated with a higher probability
of not having the characteristic (see gene 03515901 in
Figure <a href="#fig:densities" data-reference-type="ref" data-reference="fig:densities">2</a>), the ROC curve would be defined by the
<em>left-sided family of eligible classification subsets</em> <span class="citation" data-cites="Camblor2017a">(<a href="#ref-Camblor2017a" role="doc-biblioref">Martı́nez-Camblor et al. 2017</a>)</span>,
<span class="math inline">\(\mathcal{I}_l(t)\)</span>, similarly to <span class="math inline">\(\mathcal{I}_r(t)\)</span> but with the form
<span class="math inline">\(s_t = (\infty, c_t]\)</span>. It results in
<span class="math inline">\(\mathcal{R}_l(t) = F_\xi \big(F_\chi^{-1}(t) \big)\)</span>, <span class="math inline">\(t\in[0,1]\)</span>, and
the decision rules are also univocally defined in this case.</p>
<p>The ROC curve and related problems were widely studied in the
literature; interested readers are referred to the monographs of
<span class="citation" data-cites="Zhou2002">Zhou et al. (<a href="#ref-Zhou2002" role="doc-biblioref">2002</a>)</span>, <span class="citation" data-cites="Pepe2003a">Pepe (<a href="#ref-Pepe2003a" role="doc-biblioref">2003</a>)</span>, and <span class="citation" data-cites="Nakas2023">Nakas et al. (<a href="#ref-Nakas2023" role="doc-biblioref">2023</a>)</span>, as well as the review by
<span class="citation" data-cites="Inacio2021">Inácio et al. (<a href="#ref-Inacio2021" role="doc-biblioref">2021</a>)</span>. By definition, the ROC curve is confined within the unit
square, with optimal performance achieved when it approaches the
left-upper corner (AUC closer to 1). Conversely, proximity to the main
diagonal (AUC closer to 0.5) means diminished discriminatory ability,
resembling a random classifier.</p>
<p>In practice, let <span class="math inline">\((\xi_1, \xi_2, \dots, \xi_n)\)</span> and
<span class="math inline">\((\chi_1, \chi_2, \dots, \chi_m)\)</span> be two independent and identically
distributed (i.i.d.) samples from the positive and the negative
population, respectively. Different estimation procedures are
implemented in the
<a href="https://CRAN.R-project.org/package=movieROC"><strong>movieROC</strong></a> package,
such as the empirical estimator <span class="citation" data-cites="Hsieh1996">(<a href="#ref-Hsieh1996" role="doc-biblioref">Hsieh and Turnbull 1996</a>)</span> (by default in the <code>gROC()</code>
function), accompanied by its summary indices: the AUC and the Youden
index <span class="citation" data-cites="Youden1950">(<a href="#ref-Youden1950" role="doc-biblioref">Youden 1950</a>)</span>. Alternatively, semiparametric approaches based on
kernel density estimation for the involved distributions may be
considered <span class="citation" data-cites="Zou1997">(<a href="#ref-Zou1997" role="doc-biblioref">Zou et al. 1997</a>)</span>. The <code>plot_densityROC()</code> function provides plots
for both right- and left-sided ROC curves estimated by this method. On
the other hand, assuming that the marker follows a gaussian distribution
in both populations, that is,
<span class="math inline">\(\xi \sim \mathcal{N}(\mu_\xi, \sigma_\xi)\)</span> and
<span class="math inline">\(\chi \sim \mathcal{N}(\mu_\chi, \sigma_\chi)\)</span>, parametric approaches
propose plug-in estimators by estimating the unknown parameters while
using the known distributions <span class="citation" data-cites="Hanley1988">(<a href="#ref-Hanley1988" role="doc-biblioref">Hanley 1988</a>)</span>. This parametric estimation
is included in the <code>gROC_param()</code> function, which works similarly to
<code>gROC()</code>.</p>
<p><em>Main syntax:</em></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>gROC</span><span class='op'>(</span><span class='va'>X</span>, <span class='va'>D</span>, side <span class='op'>=</span> <span class='st'>"right"</span>, <span class='va'>...</span><span class='op'>)</span></span>
<span><span class='fu'>gROC_param</span><span class='op'>(</span><span class='va'>X</span>, <span class='va'>D</span>, side <span class='op'>=</span> <span class='st'>"right"</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>Table 1 in the
<a href="https://cran.r-project.org/web/packages/movieROC/vignettes/movieROC_vignette.pdf">vignette</a>
provides the main input and output parameters of these R functions,
which estimate the regular ROC curve (right-sided or left-sided with
<code>side = "right"</code> or <code>"left"</code>, respectively) and associated decision
rules. Its output is an R object of class ‘<code>groc</code>’, to which the
functions listed in Section <a href="#subsec:functions" data-reference-type="ref" data-reference="subsec:functions">2.2</a> can be applied. Most of them are
visualization tools, but the user may also <code>print()</code> summary information
and <code>predict()</code> classification regions for a particular specificity.</p>
<p>Figure <a href="#fig:roccurves" data-reference-type="ref" data-reference="fig:roccurves">3</a> graphically represents the empirical
estimation of the standard (gray line) and generalized (black line) ROC
curves for each gene in Figure <a href="#fig:densities" data-reference-type="ref" data-reference="fig:densities">2</a>. To construct the standard ROC curve for the
first two genes (20202438 and 18384097), the right-sided ROC curve is
considered; and the left-sided curve for the third one (03515901). As
expected following the discussion about
Figure <a href="#fig:densities" data-reference-type="ref" data-reference="fig:densities">2</a>, the standard and gROC curves are similar for
the third gene because there exists a monotone relationship between the
marker and the response. However, these curves differ for the first two
genes due to the lack of monotonicity in those scenarios. The empirical
gROC curve estimator is explained in detail in
Section <a href="#section:groc" data-reference-type="ref" data-reference="section:groc">4</a>.</p>
<p>Next chunk of code generates the figure, providing an example of the use
of <code>gROC()</code> function, <code>plot()</code> and how to get access to the AUC.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="cf">for</span>(gene <span class="cf">in</span> <span class="fu">c</span>(<span class="st">&quot;20202438&quot;</span>, <span class="st">&quot;18384097&quot;</span>, <span class="st">&quot;03515901&quot;</span>)){</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>   roc <span class="ot">&lt;-</span> <span class="fu">gROC</span>(<span class="at">X =</span> HCC[,<span class="fu">paste0</span>(<span class="st">&quot;cg&quot;</span>,gene)], <span class="at">D =</span> HCC<span class="sc">$</span>tumor, </span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>               <span class="at">side =</span> <span class="fu">ifelse</span>(gene <span class="sc">==</span> <span class="st">&quot;03515901&quot;</span>, <span class="st">&quot;left&quot;</span>, <span class="st">&quot;right&quot;</span>))</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>   <span class="fu">plot</span>(roc, <span class="at">col =</span> <span class="st">&quot;gray50&quot;</span>, <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">&quot;Gene&quot;</span>, gene), <span class="at">lwd =</span> <span class="dv">3</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>   groc <span class="ot">&lt;-</span> <span class="fu">gROC</span>(<span class="at">X =</span> HCC[,<span class="fu">paste0</span>(<span class="st">&quot;cg&quot;</span>,gene)], <span class="at">D =</span> HCC<span class="sc">$</span>tumor, <span class="at">side =</span> <span class="st">&quot;both&quot;</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>   <span class="fu">plot</span>(groc, <span class="at">new =</span> <span class="cn">FALSE</span>, <span class="at">lwd =</span> <span class="dv">3</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>   <span class="fu">legend</span>(<span class="st">&quot;bottomright&quot;</span>, <span class="fu">paste</span>(<span class="fu">c</span>(<span class="st">&quot;AUC =&quot;</span>, <span class="st">&quot;gAUC =&quot;</span>), <span class="fu">format</span>(<span class="fu">c</span>(roc<span class="sc">$</span>auc, groc<span class="sc">$</span>auc), </span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>          <span class="at">digits =</span> <span class="dv">3</span>)), <span class="at">col =</span> <span class="fu">c</span>(<span class="st">&quot;gray50&quot;</span>, <span class="st">&quot;black&quot;</span>), <span class="at">lwd =</span> <span class="dv">3</span>, <span class="at">bty =</span> <span class="st">&quot;n&quot;</span>, <span class="at">inset =</span> .<span class="dv">01</span>)}</span></code></pre></div>
<figure id="fig:roccurves">
<img src="figures/ROCcurves_3genes.png" style="width:100.0%"
alt="graphic without alt text" />
<figcaption>
Figure 3: Standard ROC curve (in gray) and gROC curve (in
black) empirical estimation for the capacity of genes 20202438, 18384097
and 03515901 to differ between tumor and non-tumor group.
</figcaption>
</figure>
<p>The following code snippet estimates the standard ROC curve for gene
20202438, prints its basic information, and predicts the classification
region and sensitivity resulting in a specificity of 0.9. It provides an
illustrative example of utilizing the <code>print()</code> and <code>predict()</code>
functions.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> roc_selg1 <span class="ot">&lt;-</span> <span class="fu">gROC</span>(<span class="at">X =</span> HCC<span class="sc">$</span>cg20202438, <span class="at">D =</span> HCC<span class="sc">$</span>tumor, <span class="at">side =</span> <span class="st">&quot;right&quot;</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> roc_selg1</span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>Data was encoded with <span class="fu">nontumor</span> (controls) and <span class="fu">tumor</span> (cases).</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>It is assumed that larger values of the marker indicate larger confidence that a </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a> given subject is a case.</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>There are <span class="dv">62</span> controls and <span class="dv">62</span> cases.</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>The specificity and sensitivity reported by the Youden index are <span class="fl">0.855</span> and <span class="fl">0.403</span>, </span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a> respectively, corresponding to the following classification subset<span class="sc">:</span> (<span class="fl">0.799</span>, <span class="cn">Inf</span>).</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>The area under the right<span class="sc">-</span>sided ROC <span class="fu">curve</span> (AUC) is <span class="dv">0</span>.<span class="fl">547.</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">predict</span>(roc_selg1, <span class="at">FPR =</span> .<span class="dv">1</span>)</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>ClassSubsets           <span class="sc">$</span>Specificity        <span class="sc">$</span>Sensitivity</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>] <span class="fl">0.8063487</span>   <span class="cn">Inf</span>     [<span class="dv">1</span>] <span class="fl">0.9032258</span>       [<span class="dv">1</span>] <span class="fl">0.3064516</span></span></code></pre></div>
<p>The following line of code displays the whole construction of the
empirical standard ROC curve for gene 20202438. The video is saved by
default as a GIF with the name provided.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">movieROC</span>(roc_selg1, <span class="at">reduce =</span> <span class="cn">FALSE</span>, <span class="at">file =</span> <span class="st">&quot;StandardROC_gene20202438.gif&quot;</span>)</span></code></pre></div>
<p><img src="figures/StandardROC_gene20202438.gif" style="width:60.0%" /></p>
<h3 data-number="3.1" id="section:multiroc"><span class="header-section-number">3.1</span> Multivariate ROC curve</h3>
<p>In practice, many cases may benefit from combining information from
different markers to enhance classification accuracy. Rather than
assessing univariate markers separately, taking the multivariate marker
resulting from merging them can yield relevant gain. However, note that
the ROC curve and related indices are defined only for univariate
markers, as they require the existence of a total order. To address this
limitation, a common approach involves transforming the <span class="math inline">\(p\)</span>-dimensional
multivariate marker <span class="math inline">\(\boldsymbol{X}\)</span> into a univariate one through a
functional transformation
<span class="math inline">\(\boldsymbol{h}: \mathbb{R}^p \longrightarrow \mathbb{R}\)</span>. This
transformation <span class="math inline">\(\boldsymbol{h}(\cdot)\)</span> seeks to optimize an objective
function related to the classification accuracy, usually the AUC
<span class="citation" data-cites="Su1993 McIntosh2002 Camblor2019a">(<a href="#ref-Su1993" role="doc-biblioref">Su and Liu 1993</a>; <a href="#ref-McIntosh2002" role="doc-biblioref">McIntosh and Pepe 2002</a>; <a href="#ref-Camblor2019a" role="doc-biblioref">Martı́nez-Camblor et al. 2019</a>)</span>.</p>
<p>We enumerate the methods included in the proposed R tool by the
<code>multiROC()</code> function (with the input parameter <code>method</code>), listed
according to the objective function to optimize. Recall that the output
of <code>multiROC()</code> is an object of class ‘<code>multiroc</code>’, containing
information about the estimation of the ROC curve and subsets for
multivariate scenarios. Table 2 in the
<a href="https://cran.r-project.org/web/packages/movieROC/vignettes/movieROC_vignette.pdf">vignette</a>
includes the usage of this function.</p>
<p><em>Main syntax:</em></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>multiROC</span><span class='op'>(</span><span class='va'>X</span>, <span class='va'>D</span>, method <span class='op'>=</span> <span class='st'>"lrm"</span>, </span>
<span>         formula <span class='op'>=</span> <span class='st'>'D ~ X.1 + I(X.1^2) + X.2 + I(X.2^2) + I(X.1*X.2)'</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<ol type="1">
<li>AUC: Different procedures to estimate the <span class="math inline">\(\boldsymbol{h}(\cdot)\)</span>
maximizing the AUC in the multidimensional case have been studied in
the literature. Among all families of functions, linear combinations
(<span class="math inline">\(\mathcal{L}_{\boldsymbol{\beta}}(\boldsymbol{X}) = \beta_1 X_1 + \dots + \beta_p X_p\)</span>)
are widely used due to their simplicity; an extensive review of the
existing methods was conducted by <span class="citation" data-cites="Kang2016">Kang et al. (<a href="#ref-Kang2016" role="doc-biblioref">2016</a>)</span>.</li>
</ol>
<ul>
<li>Computation: In the <code>multiROC()</code> function, fixing input parameters
<code>method = "fixedLinear"</code> and <code>methodLinear</code> to one from <code>"SuLiu"</code>
<span class="citation" data-cites="Su1993">(<a href="#ref-Su1993" role="doc-biblioref">Su and Liu 1993</a>)</span>, <code>"PepeThompson"</code> <span class="citation" data-cites="Pepe2000">(<a href="#ref-Pepe2000" role="doc-biblioref">Pepe and Thompson 2000</a>)</span>, or <code>"minmax"</code> <span class="citation" data-cites="Liu2011">(<a href="#ref-Liu2011" role="doc-biblioref">Liu et al. 2011</a>)</span>.
The R function also admits quadratic combinations when <span class="math inline">\(p=2\)</span>, i.e.
<span class="math inline">\(\mathcal{Q}_{\boldsymbol{\beta}}(\boldsymbol{X}) = \beta_1 X_1 + \beta_2 X_2 + \beta_3 X_1 X_2 + \beta_4 X_1^2 + \beta_5 X_2^2\)</span>,
by fixing <code>method = "fixedQuadratic"</code> for a particular
<code>coefQuadratic</code> <span class="math inline">\(= \boldsymbol{\beta} = (\beta_1, \dots, \beta_5)\)</span>.</li>
</ul>
<ol start="2" type="1">
<li><p>The risk score function
logit<span class="math inline">\(\left\{ \mathbb{P}(D = 1 \, | \, \boldsymbol{X}) \right\}\)</span>: Our
package allows the user to fit a logistic regression model
(<code>method = "lrm"</code>) considering any family of functions (linear,
quadratic, whether considering interactions or not...) by means of
the input parameter <code>formula.lrm</code>. A stepwise regression model is
fitted if <code>stepModel = TRUE</code>. Details are explained in
Section <a href="#section:efficientroc" data-reference-type="ref" data-reference="section:efficientroc">5</a>.</p></li>
<li><p>The sensitivity for a particular specificity:</p></li>
</ol>
<ol type="a">
<li>Considering the theoretical discussion about the search of the
optimal transformation <span class="math inline">\(\boldsymbol{h}(\cdot)\)</span> pointed out in
Section <a href="#section:efficientroc" data-reference-type="ref" data-reference="section:efficientroc">5</a>, <span class="citation" data-cites="Camblor2021b">Martı́nez-Camblor et al. (<a href="#ref-Camblor2021b" role="doc-biblioref">2021a</a>)</span> proposed to
estimate it by multivariate kernel density estimation for
positive and negative groups separately.</li>
</ol>
<ul>
<li>Computation: The <code>multiROC()</code> function integrates the estimation
procedures for the bandwidth matrix developed by <span class="citation" data-cites="Duong2007">Duong (<a href="#ref-Duong2007" role="doc-biblioref">2007</a>)</span>, by
fixing <code>method = "kernelOptimal"</code> and choosing a proper method
to estimate the bandwidth (<code>"kernelOptimal.H"</code>).</li>
</ul>
<ol start="2" type="a">
<li>Mainly linear combinations have been explored to date in the
scientific literature <span class="citation" data-cites="Meisner2021 PerezFernandez2020">(<a href="#ref-Meisner2021" role="doc-biblioref">Meisner et al. 2021</a>; <a href="#ref-PerezFernandez2020" role="doc-biblioref">Pérez-Fernández et al. 2021</a>)</span>. For a
fixed specificity <span class="math inline">\(t\in[0,1]\)</span>, we seek the linear combination
<span class="math inline">\(\mathcal{L}_{\boldsymbol{\beta}(t)}(\boldsymbol{X}) = \beta_1(t) X_1 + \dots + \beta_p(t) X_p\)</span>
maximizing the true-positive rate by considering standard
subsets for the transformed marker. The coefficients
<span class="math inline">\(\boldsymbol{\beta}(t)\)</span> are called ‘dynamic parameters’ because
they may be different for each <span class="math inline">\(t \in [0,1]\)</span>.</li>
</ol>
<ul>
<li>Computation: Since our objective is to display the ROC curve,
<span class="math inline">\(\mathcal{L}_{\boldsymbol{\beta}(t)}(\boldsymbol{X})\)</span> is
estimated for every <span class="math inline">\(t\)</span> in a grid of the unit interval,
resulting in one <span class="math inline">\(\boldsymbol{\hat{\beta}}(t)\)</span> for each <span class="math inline">\(t\)</span>.
This approach is time-consuming, especially when it is based on
the plug-in empirical estimators involved
(<code>method = "dynamicEmpirical"</code>, only implemented for <span class="math inline">\(p=2\)</span>), and
may result in overfitting. Instead, <span class="citation" data-cites="Meisner2021">Meisner et al. (<a href="#ref-Meisner2021" role="doc-biblioref">2021</a>)</span> method is
recommended (<code>method = "dynamicMeisner"</code>).</li>
</ul>
<p>Once the classification subsets for a multivariate marker are
constructed by the <code>multiROC()</code> function, several R methods may be used
for the output object (see Section <a href="#subsec:functions" data-reference-type="ref" data-reference="subsec:functions">2.2</a>). They include <code>print</code> relevant
information or <code>plot</code> the resulting ROC curve. The main contribution of
the package is to plot the construction of the ROC curve together with
the classification subsets in a static figure for a particular FPR
(<code>plot_buildROC()</code> function), or in a video for tracking the whole
process (<code>movieROC()</code> function).</p>
<p>Figure <a href="#fig:biroc" data-reference-type="ref" data-reference="fig:biroc">4</a>
illustrates the videos resulting from <code>movieROC()</code> function.
Particularly, classification accuracy of the
bivariate marker <code>(cg20202438, cg18384097)</code> was studied by using four
different approaches indicated on the captions, considering linear
combinations (top) and nonlinear transformations (bottom). This figure
was implemented by the code below, integrating <code>multiROC()</code> and
<code>movieROC()</code> functions. Four videos are saved as GIF files with names
<code>"PepeTh.gif"</code> (a), <code>"Meisner.gif"</code> (b), <code>"LRM.gif"</code> (c), and
<code>"KernelDens.gif"</code> (d).</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> X <span class="ot">&lt;-</span> HCC[ ,<span class="fu">c</span>(<span class="st">&quot;cg20202438&quot;</span>, <span class="st">&quot;cg18384097&quot;</span>)]; D <span class="ot">&lt;-</span> HCC<span class="sc">$</span>tumor</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> biroc_12_PT <span class="ot">&lt;-</span> <span class="fu">multiROC</span>(X, D, <span class="at">method =</span> <span class="st">&quot;fixedLinear&quot;</span>, <span class="at">methodLinear =</span> <span class="st">&quot;PepeThompson&quot;</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> biroc_12_Meis <span class="ot">&lt;-</span> <span class="fu">multiROC</span>(X, D, <span class="at">method =</span> <span class="st">&quot;dynamicMeisner&quot;</span>, <span class="at">verbose =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> biroc_12_lrm <span class="ot">&lt;-</span> <span class="fu">multiROC</span>(X, D)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> biroc_12_kernel <span class="ot">&lt;-</span> <span class="fu">multiROC</span>(X, D, <span class="at">method =</span> <span class="st">&quot;kernelOptimal&quot;</span>)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> list_biroc <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">PepeTh =</span> biroc_12_PT, <span class="at">Meisner =</span> biroc_12_Meis, </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>           <span class="at">LRM =</span> biroc_12_lrm, <span class="at">KernelDens =</span> biroc_12_kernel)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">lapply</span>(<span class="fu">names</span>(list_biroc), <span class="cf">function</span>(x) <span class="fu">movieROC</span>(list_biroc[[x]],  </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>           <span class="at">display.method =</span> <span class="st">&quot;OV&quot;</span>, <span class="at">xlab =</span> <span class="st">&quot;Gene 20202438&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Gene 18384097&quot;</span>, </span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>           <span class="at">cex =</span> <span class="fl">1.2</span>, <span class="at">alpha.points =</span> <span class="dv">1</span>, <span class="at">lwd.curve =</span> <span class="dv">4</span>, <span class="at">file =</span> <span class="fu">paste0</span>(x, <span class="st">&quot;.gif&quot;</span>)))</span></code></pre></div>
<div id="fig:biroc" class="figure">
<table>
<colgroup>
<col style="width: 51%" />
<col style="width: 48%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;"><img src="figures/PepeTh.gif" style="width:100.0%" alt="a) Pepe and Thompson" /></th>
<th style="text-align: center;"><img src="figures/Meisner.gif" style="width:100.0%" alt="b) Meisner et al." /></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;"><strong>(a)</strong> Linear combinations with fixed parameters by <span class="citation" data-cites="Pepe2000">Pepe and Thompson (<a href="#ref-Pepe2000" role="doc-biblioref">2000</a>)</span>.</td>
<td style="text-align: center;"><strong>(b)</strong> Linear combinations with dynamic parameters by <span class="citation" data-cites="Meisner2021">Meisner et al. (<a href="#ref-Meisner2021" role="doc-biblioref">2021</a>)</span>.</td>
</tr>
</tbody>
</table>
<table>
<colgroup>
<col style="width: 49%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: center;"><img src="figures/LRM.gif" style="width:100.0%" alt="c) Logistic Regression" /></th>
<th style="text-align: center;"><img src="figures/KernelDens.gif" style="width:100.0%" alt="d) Kernel Density" /></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;"><strong>(c)</strong> Logistic regression model with quadratic formula by default (see <code>formula.lrm</code> in Table 2 of the <a href="https://cran.r-project.org/web/packages/movieROC/vignettes/movieROC_vignette.pdf">vignette</a>).</td>
<td style="text-align: center;"><strong>(d)</strong> Optimal transformation by multivariate kernel density estimation with <code>"Hbcv"</code> method by default <span class="citation" data-cites="Camblor2021b">(<a href="#ref-Camblor2021b" role="doc-biblioref">Martı́nez-Camblor et al. 2021a</a>)</span>.</td>
</tr>
</tbody>
</table>
<p>Figure 4: Videos (from <code>movieROC()</code> function) of the classification procedure and ROC curve for the bivariate marker (<code>cg20202438</code>, <code>cg18384097</code>). Four different methods for classification are displayed.</p>
</div>
<p>When the marker has a dimension higher than two it is difficult to
visualize the data and the classification regions. Therefore, the
<code>movieROC()</code> function offers two options for showing the results, both
on a bidimensional space. On the one hand, to choose two of the
components of the multivariate marker and project the classification
subsets on the plane defined by them
(Figure <a href="#fig:multiroccurve" data-reference-type="ref" data-reference="fig:multiroccurve">5</a>, middle). On the other, to project the
classification regions on the plane defined by the two first principal
components (Figure <a href="#fig:multiroccurve" data-reference-type="ref" data-reference="fig:multiroccurve">5</a>, left). The R function <code>prcomp()</code> from <code>stats</code>
is used to perform Principal Components Analysis (PCA) <span class="citation" data-cites="Hotelling1933">(<a href="#ref-Hotelling1933" role="doc-biblioref">Hotelling 1933</a>)</span>.</p>
<p>Figure <a href="#fig:multiroccurve" data-reference-type="ref" data-reference="fig:multiroccurve">5</a> shows the difficulty in displaying the
decision rules when <span class="math inline">\(p&gt;2\)</span> (the 3 genes used along this manuscript), even
with the two options implemented in our package. It was generated using
<code>multiROC()</code> and <code>plot_buildROC()</code>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> multiroc_PT <span class="ot">&lt;-</span> <span class="fu">multiROC</span>(<span class="at">X =</span> HCC[ ,<span class="fu">c</span>(<span class="st">&quot;cg20202438&quot;</span>, <span class="st">&quot;cg18384097&quot;</span>, <span class="st">&quot;cg03515901&quot;</span>)], </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                 <span class="at">D =</span> HCC<span class="sc">$</span>tumor, <span class="at">method =</span> <span class="st">&quot;fixedLinear&quot;</span>, <span class="at">methodLinear =</span> <span class="st">&quot;PepeThompson&quot;</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> multiroc_PT</span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>Data was encoded with <span class="fu">nontumor</span> (controls) and <span class="fu">tumor</span> (cases).</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>There are <span class="dv">62</span> controls and <span class="dv">62</span> cases.</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>A total of <span class="dv">3</span> variables have been considered.</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>A linear combination with fixed parameters estimated by PepeThompson approach has </span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a> been considered.</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>The specificity and sensitivity reported by the Youden index are <span class="fl">0.855</span> and <span class="fl">0.742</span>, </span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a> respectively, corresponding to the cut<span class="sc">-</span>off point <span class="sc">-</span><span class="fl">0.0755</span> <span class="cf">for</span> the transformation </span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a> <span class="fu">h</span>(X) <span class="ot">=</span>  <span class="fl">0.81</span><span class="sc">*</span>cg20202438 <span class="sc">-</span> <span class="fl">0.1</span><span class="sc">*</span>cg18384097 <span class="sc">-</span> <span class="dv">1</span><span class="sc">*</span>cg03515901.</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>The area under the ROC <span class="fu">curve</span> (AUC) is <span class="dv">0</span>.<span class="fl">811.</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot_buildROC</span>(multiroc_PT, <span class="at">cex =</span> <span class="fl">1.2</span>, <span class="at">lwd.curve =</span> <span class="dv">4</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot_buildROC</span>(multiroc_PT, <span class="at">display.method =</span> <span class="st">&quot;OV&quot;</span>, <span class="at">displayOV =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>), <span class="at">cex =</span> <span class="fl">1.2</span>,</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>                <span class="at">xlab =</span> <span class="st">&quot;Gene 20202438&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Gene 03515901&quot;</span>, <span class="at">lwd.curve =</span> <span class="dv">4</span>)</span></code></pre></div>
<figure id="fig:multiroccurve">
<img src="figures/buildROC_multi3_full.png" style="width:100.0%" />
<figcaption>
Figure 5: Multivariate ROC curve estimation for the
simultaneous diagnostic accuracy of genes 20202438, 18384097 and
03515901. <span class="citation" data-cites="Pepe2000">Pepe and Thompson
(2000)</span> approach was used to estimate the linear coefficients and
classification rules (yellow and gray border for positive and negative
class, respectively) for a FPR 0.15 are displayed. Left, projected over
the 2 principal components from PCA; middle, over the 1st and the 3rd
selected genes.
</figcaption>
</figure>
<h2 data-number="4" id="section:groc"><span class="header-section-number">4</span> Generalized ROC curve</h2>
<p>There are scenarios whose standard ROC curves are not concave (first two
genes in Figure <a href="#fig:roccurves" data-reference-type="ref" data-reference="fig:roccurves">3</a>, gray solid line), reflecting that the
standard initial assumption of existence of a monotone relationship
between the marker and the response is misleading. In
Figure <a href="#fig:densities" data-reference-type="ref" data-reference="fig:densities">2</a>, we may see that difference in gene 20202438
distribution between those tissues which have the characteristic and those
which do not is mainly in dispersion. To accommodate this common type of
scenarios, <span class="citation" data-cites="Camblor2017a">Martı́nez-Camblor et al. (<a href="#ref-Camblor2017a" role="doc-biblioref">2017</a>)</span> extended the ROC curve definition to the case
where both extremes for marker values are associated with a higher risk
of having the characteristic of interest, by considering the <em>both-sided
family of eligible classification subsets</em>:
<span class="math display">\[\mathcal{I}_g(t) = \Big\{ s_t = (-\infty,x_t^L] \cup (x_t^U, \infty)  :  x_t^L \leq x_t^U \in \mathcal{S}(X)  ,  \mathbb{P}(\chi \in s_t) = t \Big\}.\]</span></p>
<p>It becomes crucial to consider the supremum in the definition of the
generalized ROC curve because the decision rule for each <span class="math inline">\(t \in [0,1]\)</span>
is not univocally defined: there exist infinite pairs <span class="math inline">\(x_t^L \leq x_t^U\)</span>
reporting a specificity <span class="math inline">\(1-t\)</span> (i.e. <span class="math inline">\(\mathcal{I}_g(t)\)</span> is uncountably
infinite). Computationally, this optimization process incurs a
time-consuming estimation, depending on the number of different marker
values in the sample.</p>
<p>After the introduction of this extension, several studies followed up
regarding estimation of the gROC curve <span class="citation" data-cites="Camblor2017a Camblor2019b">(<a href="#ref-Camblor2017a" role="doc-biblioref">Martı́nez-Camblor et al. 2017</a>; <a href="#ref-Camblor2019b" role="doc-biblioref">Martı́nez-Camblor and Pardo-Fernández 2019a</a>)</span>
and related measures such as its area (gAUC) <span class="citation" data-cites="Camblor2021a">(<a href="#ref-Camblor2021a" role="doc-biblioref">Martı́nez-Camblor et al. 2021b</a>)</span> and the
Youden index <span class="citation" data-cites="Camblor2019c Bantis2021a">(<a href="#ref-Camblor2019c" role="doc-biblioref">Martı́nez-Camblor and Pardo-Fernández 2019b</a>; <a href="#ref-Bantis2021a" role="doc-biblioref">Bantis et al. 2021</a>)</span>. By considering this
generalization, another property of the classification subsets may be
lost: the regions may not be self-contained over the increase in
false-positive rate. It may happen that a subject is classified as a
positive for a particular FPR <span class="math inline">\(t_1\)</span>, but as a negative for a higher FPR
<span class="math inline">\(t_2\)</span>. Therefore, it is natural to establish a restriction <em>(C)</em> on the
classification subsets, ensuring that any subject classified as a
positive for a fixed specificity (or sensitivity) will also be
classified as a positive for any classification subset with lower
specificity (higher sensitivity). <span class="citation" data-cites="PerezFernandez2020">Pérez-Fernández et al. (<a href="#ref-PerezFernandez2020" role="doc-biblioref">2021</a>)</span> proposed an
algorithm to estimate the gROC curve under restriction <em>(C)</em>, included
in the <code>gROC()</code> function of the presented R package. See final Section
for computational details about this algorithm implementation.</p>
<p><em>Main syntax:</em></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>gROC</span><span class='op'>(</span><span class='va'>X</span>, <span class='va'>D</span>, side <span class='op'>=</span> <span class='st'>"both"</span>, <span class='va'>...</span><span class='op'>)</span></span>
<span><span class='fu'>gROC_param</span><span class='op'>(</span><span class='va'>X</span>, <span class='va'>D</span>, side <span class='op'>=</span> <span class='st'>"both"</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>Table 1 in the
<a href="https://cran.r-project.org/web/packages/movieROC/vignettes/movieROC_vignette.pdf">vignette</a>
collects the input and output parameters of the <code>gROC()</code> function, which
estimates the gROC curve, both in the mentioned direction
(<code>side = "both"</code>) and in the opposite, i.e. when classification subsets
of the form <span class="math inline">\(s_t=(x_t^L, x_t^U]\)</span> are considered (<code>side = "both2"</code>). In
addition, all the particular methods for a ‘<code>groc</code>’ object collected in
Section <a href="#subsec:functions" data-reference-type="ref" data-reference="subsec:functions">2.2</a> may be used in this general
scenario.</p>
<p>Following, the gene 20202438 expression intensity diagnostic accuracy is
evaluated by the gROC curve without restrictions (<code>groc_selg1</code> object)
and under the restriction <em>(C)</em> (<code>groc_selg1_C</code> object). The
classification subsets and sensitivity for a specificity of <span class="math inline">\(0.9\)</span> are
displayed with the <code>predict()</code> function.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> groc_selg1 <span class="ot">&lt;-</span> <span class="fu">gROC</span>(<span class="at">X =</span> HCC<span class="sc">$</span>cg20202438, <span class="at">D =</span> HCC<span class="sc">$</span>tumor, <span class="at">side =</span> <span class="st">&quot;both&quot;</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">predict</span>(groc_selg1, <span class="at">FPR =</span> .<span class="dv">1</span>)</span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="sc">$</span>ClassSubsets                     <span class="sc">$</span>Specificity      <span class="sc">$</span>Sensitivity    </span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>          [,<span class="dv">1</span>]      [,<span class="dv">2</span>]        [<span class="dv">1</span>] <span class="fl">0.9032258</span>       [<span class="dv">1</span>] <span class="fl">0.4032258</span>   </span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>[<span class="dv">1</span>,]      <span class="sc">-</span><span class="cn">Inf</span> <span class="fl">0.7180623</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>[<span class="dv">2</span>,] <span class="fl">0.8296072</span>       <span class="cn">Inf</span>        </span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> groc_selg1_C <span class="ot">&lt;-</span> <span class="fu">gROC</span>(<span class="at">X =</span> HCC<span class="sc">$</span>cg20202438, <span class="at">D =</span> HCC<span class="sc">$</span>tumor, <span class="at">side =</span> <span class="st">&quot;both&quot;</span>, </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>              <span class="at">restric =</span> <span class="cn">TRUE</span>, <span class="at">optim =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>All the classification regions underlying the standard and the
generalized ROC curves without and with restrictions are represented in
Figure <a href="#fig:regionsgroc" data-reference-type="ref" data-reference="fig:regionsgroc">6</a>. The following code was used to generate
the figure, illustrating the usage and output of the <code>plot_regions()</code>
function. Besides displaying all the classification regions underlying
every specificity (in gray), the one chosen by the user (FPR = 0.15 by
default) is highlighted in blue. Note that the ROC curves are rotated
<span class="math inline">\(90^\circ\)</span> to the right, in order to use the vertical axis for FPR in
both plots.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot_regions</span>(roc_selg1, <span class="at">cex.legend =</span> <span class="fl">1.5</span>, <span class="at">plot.auc =</span> <span class="cn">TRUE</span>, </span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>       <span class="at">main =</span> <span class="st">&quot;Standard right-sided assumption [Classification subsets]&quot;</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot_regions</span>(groc_selg1, <span class="at">plot.auc =</span> <span class="cn">TRUE</span>, <span class="at">legend =</span> F, </span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>       <span class="at">main.plotroc =</span> <span class="st">&quot;gROC curve&quot;</span>,</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>       <span class="at">main =</span> <span class="st">&quot;General approach [Classification subsets]&quot;</span>)</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">plot_regions</span>(groc_selg1_C, <span class="at">plot.auc =</span> <span class="cn">TRUE</span>, <span class="at">legend =</span> F, </span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>       <span class="at">main.plotroc =</span> <span class="st">&quot;gROC curve&quot;</span>,</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>       <span class="at">main =</span> <span class="st">&quot;General approach with restriction (C) [Classific. subsets]&quot;</span>, </span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>       <span class="at">xlab =</span> <span class="st">&quot;Gene 20202438 expression intensity&quot;</span>)</span></code></pre></div>
<figure id="fig:regionsgroc">
<img src="figures/plotregions_gene20202438_1.png" style="width:100.0%"
alt="graphic without alt text" />
<img src="figures/plotregions_gene20202438_2.png" style="width:100.0%"
alt="graphic without alt text" />
<img src="figures/plotregions_gene20202438_3.png" style="width:100.0%"
alt="graphic without alt text" />
<figcaption>
Figure 6: Classification regions and the ROC curve (90º
rotated) for evaluation of gene 20202438 expression intensity assuming
i) standard scenario (top), ii) generalized scenario without
restrictions (middle), iii) generalized scenario under restriction
<em>(C)</em> over the subsets (bottom).
</figcaption>
</figure>
<p>It is clear the gain achieved for considering the generalized scenario
for this marker, which fits better its distribution in each group.
Standard estimated AUC is 0.547, while the gAUC increases to 0.765. The
gAUC is not especially affected by imposing the restriction <em>(C)</em>,
resulting in 0.762.</p>
<h2 data-number="5" id="section:efficientroc"><span class="header-section-number">5</span> Efficient ROC curve: pursuing an optimal transformation</h2>
<p>By keeping classification subsets of the form <span class="math inline">\(s_t = (c_t, \infty)\)</span>, an
alternative approach can be explored: transforming the univariate marker
through a suitable function <span class="math inline">\(h: \mathbb{R} \longrightarrow \mathbb{R}\)</span>
to enhance its accuracy. Henceforth, the transformation <span class="math inline">\(h^*(\cdot)\)</span>
reporting the dominant ROC curve compared to the one from any other
function (i.e. <span class="math inline">\(\mathcal{R}_{h^*}(\cdot) \geq \mathcal{R}_h(\cdot)\)</span>)
will be referred to as <em>optimal transformation</em> (in the ROC sense), and
the resulting ROC curve is called eROC <span class="citation" data-cites="Kauppi2016">(<a href="#ref-Kauppi2016" role="doc-biblioref">Kauppi 2016</a>)</span>. Following the
well-known Neyman–Pearson lemma, <span class="citation" data-cites="McIntosh2002">McIntosh and Pepe (<a href="#ref-McIntosh2002" role="doc-biblioref">2002</a>)</span> proved that <span class="math inline">\(h^*(\cdot)\)</span>
is the likelihood ratio.</p>
<p>We enumerate the methods included in the proposed R tool by the <code>hROC()</code>
function (with the input parameter <code>type</code>), listed according to the
procedure considered to estimate <span class="math inline">\(h^*(\cdot)\)</span>. The output of this
function is an object of class ‘<code>hroc</code>’. See Table 3 in the
<a href="https://cran.r-project.org/web/packages/movieROC/vignettes/movieROC_vignette.pdf">vignette</a>
for function usage and output details.</p>
<p><em>Main syntax:</em></p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'>hROC</span><span class='op'>(</span><span class='va'>X</span>, <span class='va'>D</span>, type <span class='op'>=</span> <span class='st'>"lrm"</span>, formula.lrm <span class='op'>=</span> <span class='st'>'D ~ pol(X,3)'</span>, <span class='va'>...</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<ol type="1">
<li><span class="citation" data-cites="Camblor2019a">Martı́nez-Camblor et al. (<a href="#ref-Camblor2019a" role="doc-biblioref">2019</a>)</span> exploited the result proved by <span class="citation" data-cites="McIntosh2002">McIntosh and Pepe (<a href="#ref-McIntosh2002" role="doc-biblioref">2002</a>)</span>,
suggesting to estimate the logit of the risk function by logistic
regression, since it is a monotone increasing transformation of the
likelihood ratio.</li>
</ol>
<ul>
<li>Computation: By the proposed R tool, the user can define any
transformation <span class="math inline">\(h(\cdot)\)</span> for the right-hand side of the logistic
regression model to be fitted,
logit<span class="math inline">\(\big\{ \mathbb{P}(D = 1 \, | \, x) \big\} = h(x)\)</span>. Particularly, by
fixing the input parameters: <code>type = "lrm"</code> and defining the
function <span class="math inline">\(h(\cdot)\)</span> as <code>formula.lrm</code>.</li>
</ul>
<ol start="2" type="1">
<li>Arguing as in <span class="citation" data-cites="Camblor2021b">Martı́nez-Camblor et al. (<a href="#ref-Camblor2021b" role="doc-biblioref">2021a</a>)</span> for univariate markers instead of
multivariate, the optimal transformation in the ROC sense is
equivalent to
<span class="math inline">\(h^*(\cdot)=f_\xi(\cdot)/\big(f_\xi(\cdot) + f_\chi(\cdot)\big)\)</span>,
where <span class="math inline">\(f(\cdot)\)</span> denotes the density function. In order to estimate
<span class="math inline">\(h^*(\cdot)\)</span>, different estimation procedures for the density
functions separately may be used, such as the kernel density
estimator.</li>
</ol>
<ul>
<li>Computation: By the <code>hROC()</code> function, the user may fix
<code>type = "kernel"</code> and choose a proper bandwidth for the kernel
estimation by <code>kernel.h</code> in order to compute this method.</li>
</ul>
<ol start="3" type="1">
<li><span class="citation" data-cites="Camblor2019a">Martı́nez-Camblor et al. (<a href="#ref-Camblor2019a" role="doc-biblioref">2019</a>)</span> also included the estimation of the <em>overfitting
function</em>, <span class="math inline">\(h_{of}(\cdot)\)</span>, defined as the optimal one when no
restrictions on the shape of <span class="math inline">\(h^*(\cdot)\)</span> are imposed. It takes the
value 1 for the positive marker values and 0 for the negative ones,
reporting an estimated AUC of 1, but totally depending on the
available sample (the resulting rules cannot be extended).</li>
</ol>
<ul>
<li>Computation: <span class="math inline">\(h_{of}(\cdot)\)</span> may be estimated by fixing the input
parameter <code>type = "overfitting"</code>.</li>
</ul>
<p>The following code and figures study the capacity of improving the
classification performance of the gene 18384097 expression intensity via
the above functional transformations and its impact on the final
decision rules. The first one considers an ordinary cubic polynomial
formula (<code>hroc_cubic_selg2</code>), and a linear tail-restricted cubic spline
(<code>hroc_rcs_selg2</code>) for the right-hand side of the logistic regression model.
The second one uses two different bandwidths (<span class="math inline">\(h=1\)</span> and <span class="math inline">\(h=3\)</span>) for
density function estimation. For a comparative purpose, the last one
estimates the gROC curve under restriction <em>(C)</em>.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> X <span class="ot">&lt;-</span> HCC<span class="sc">$</span>cg18384097; D <span class="ot">&lt;-</span> HCC<span class="sc">$</span>tumor</span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> hroc_cubic_selg2 <span class="ot">&lt;-</span> <span class="fu">hROC</span>(X, D); hroc_cubic_selg2</span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>Data was encoded with <span class="fu">nontumor</span> (controls) and <span class="fu">tumor</span> (cases).</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>There are <span class="dv">62</span> controls and <span class="dv">62</span> cases.</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>A logistic regression model of the form D <span class="sc">~</span> <span class="fu">pol</span>(X,<span class="dv">3</span>) has been performed.</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>The estimated parameters of the model are the following<span class="sc">:</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a> Intercept          X        X<span class="sc">^</span><span class="dv">2</span>        X<span class="sc">^</span><span class="dv">3</span> </span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>   <span class="st">&quot;1.551&quot;</span>   <span class="st">&quot;32.054&quot;</span> <span class="st">&quot;-120.713&quot;</span>  <span class="st">&quot;100.449&quot;</span> </span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>The specificity and sensitivity reported by the Youden index are <span class="fl">0.935</span> and <span class="fl">0.532</span>, </span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a> respectively, corresponding to the following classification subset<span class="sc">:</span> </span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a> (<span class="sc">-</span><span class="cn">Inf</span>, <span class="fl">0.442</span>) <span class="fu">U</span> (<span class="fl">0.78</span>, <span class="cn">Inf</span>).</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>The area under the ROC <span class="fu">curve</span> (AUC) is <span class="dv">0</span>.<span class="fl">759.</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> hroc_rcs_selg2 <span class="ot">&lt;-</span> <span class="fu">hROC</span>(X, D, <span class="at">formula.lrm =</span> <span class="st">&quot;D ~ rcs(X,8)&quot;</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> hroc_lkr1_selg2 <span class="ot">&lt;-</span> <span class="fu">hROC</span>(X, D, <span class="at">type =</span> <span class="st">&quot;kernel&quot;</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> hroc_lkr3_selg2 <span class="ot">&lt;-</span> <span class="fu">hROC</span>(X, D, <span class="at">type =</span> <span class="st">&quot;kernel&quot;</span>, <span class="at">kernel.h =</span> <span class="dv">3</span>)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> hroc_overfit_selg2 <span class="ot">&lt;-</span> <span class="fu">hROC</span>(X, D, <span class="at">type =</span> <span class="st">&quot;overfitting&quot;</span>)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> groc_selg2_C <span class="ot">&lt;-</span> <span class="fu">gROC</span>(X, D, <span class="at">side =</span> <span class="st">&quot;both&quot;</span>, <span class="at">restric =</span> <span class="cn">TRUE</span>, <span class="at">optim =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>The following code snippet compares the AUC achieved from each approach
considered above:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> list_hroc <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">Cubic =</span> hroc_cubic_selg2, <span class="at">Splines =</span> hroc_rcs_selg2, </span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>               <span class="at">Overfit =</span> hroc_overfit_selg2, <span class="at">LikRatioEst_h3 =</span> hroc_lkr3_selg2,</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="sc">+</span>               <span class="at">LikRatioEst_h1 =</span> hroc_lkr1_selg2, <span class="at">gAUC_restC =</span> groc_selg2_C)</span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> AUCs <span class="ot">&lt;-</span> <span class="fu">sapply</span>(list_hroc, <span class="cf">function</span>(x) x<span class="sc">$</span>auc)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">round</span>(AUCs, <span class="dv">3</span>)</span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>Cubic        Splines   Overfit   LikRatioEst_h3    LikRatioEst_h1       gAUC_restC </span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fl">0.759</span>          <span class="fl">0.807</span>     <span class="fl">1.000</span>            <span class="fl">0.781</span>             <span class="fl">0.799</span>            <span class="fl">0.836</span></span></code></pre></div>
<p>The shape of the classification regions over the original space
<span class="math inline">\(\mathcal{S}(X)\)</span> depends on the monotonicity of <span class="math inline">\(h^*(\cdot)\)</span>, which may
be graphically studied by the <code>plot_funregions()</code> function (see
Figure <a href="#fig:transformations" data-reference-type="ref" data-reference="fig:transformations">7</a>). These regions can be visualized by
the R function <code>plot_regions()</code> (see
Figure <a href="#fig:regionshroc" data-reference-type="ref" data-reference="fig:regionshroc">8</a>). Both are explained in
Section <a href="#subsec:functions" data-reference-type="ref" data-reference="subsec:functions">2.2</a> and illustrated below. The next chunk of
code produced Figure <a href="#fig:transformations" data-reference-type="ref" data-reference="fig:transformations">7</a>, representing the different functional
transformations estimated previously:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>R<span class="sc">&gt;</span> <span class="fu">lapply</span>(list_hroc, <span class="cf">function</span>(x) <span class="fu">plot_funregions</span>(x, <span class="at">FPR =</span> .<span class="dv">15</span>, <span class="at">FPR2 =</span> .<span class="dv">5</span>))</span></code></pre></div>
<figure id="fig:transformations">
<img src="figures/plotfunregions_hroc_lrm_gene18384097.png"
style="width:100.0%" alt="graphic without alt text" />
<figcaption>
Figure 7: Different functional transformations and
resulting classification subsets for gene 18384097. Rules for FPR 0.15
(blue) and 0.50 (red) are remarked. Top, from left to right: cubic
polynomial function, restricted cubic splines (with 8 knots), and
overfitted transformation. Bottom: likelihood ratio estimation with
bandwidths 3 (left) and 1 (middle), and transformation resulting in the
gROC curve under restriction <em>(C)</em>.
</figcaption>
</figure>
<p>Finally, using the <code>plot_regions()</code> function,
Figure <a href="#fig:regionshroc" data-reference-type="ref" data-reference="fig:regionshroc">8</a> shows the resulting classification subsets
over the original space for the best two of the six methods above. The first
method (fitting a logistic regression model with restricted cubic
splines with 8 knots) reports an AUC of 0.804 (compared to 0.684 by the
standard ROC curve), but the shape of some classification rules is
complex, such as <span class="math inline">\(s_t=(-\infty,a_t] \cup (b_t,c_t] \cup (d_t,\infty)\)</span>.
This area increases to 0.836 by considering subsets of the form
<span class="math inline">\(s_t=(-\infty,x_t^L] \cup (x_t^U,\infty)\)</span>, even imposing the restriction
<span class="math inline">\(\textit{(C)}\)</span> to get a functional transformation <span class="math inline">\(h(\cdot)\)</span>.</p>
<figure id="fig:regionshroc">
<img src="figures/plotregions_gene18384097_1.png" style="width:100.0%"
alt="graphic without alt text" />
<img src="figures/plotregions_gene18384097_2.png" style="width:100.0%"
alt="graphic without alt text" />
<figcaption>
Figure 8: Classification regions and the resulting ROC
curve (90º rotated) for the gene 18384097. Top, ROC curve for restricted
cubic splines transformation with 8 knots; bottom, gROC curve under
restriction <em>(C)</em> for the original marker.
</figcaption>
</figure>
<h2 data-number="6" id="section:conclusion"><span class="header-section-number">6</span> Summary and conclusion</h2>
<p>Conducting binary classification using continuous markers requires
establishment of decision rules. In the standard case, each specificity
<span class="math inline">\(t \in [0,1]\)</span> entails a classification subset of the form
<span class="math inline">\(s_t = (c_t,\infty)\)</span> univocally defined. However, in more complex
situations – such as where there is a non-monotone relationship between the
marker and the response or in multivariate scenarios – these become not
clear. Visualization of the decision rules becomes crucial in these
cases. To address this, the
<a href="https://CRAN.R-project.org/package=movieROC"><strong>movieROC</strong></a> package
incorporates novel visualization tools complementing the ROC curve
representation.</p>
<p>This R package offers a user-friendly and easily comprehensible software
solution tailored for practical researchers. It implements statistical
techniques to estimate and compare, and finally to graphically represent
different classification procedures. While several R packages address
ROC curve estimation, the proposed one emphasizes the classification
process, tracking the decision rules underlying the studied binary
classification problem. This tool incorporates different considerations
and transformations which may be useful to capture the potential of the
marker to classify in non-standard scenarios. Nevertheless, this library
is also useful in standard cases, as well as when the marker itself
comes from a classification or regression method (such as support vector
machines), because it provides nice visuals and additional information
not usually reported with the ROC curve.</p>
<p>The main function of the package, <code>movieROC()</code>, allows users to monitor how
the resulting classification subsets change along different
specificities, thereby building the corresponding ROC curve. Notably, it
introduces time as a third dimension to keep those specificities,
generating informative videos. For interested readers or potential users
of <a href="https://CRAN.R-project.org/package=movieROC"><strong>movieROC</strong></a>, the
<a href="https://cran.r-project.org/web/packages/movieROC/movieROC.pdf">manual</a>
available in CRAN provides complete information about the implemented
functions and their parameters. In addition, a
<a href="https://cran.r-project.org/web/packages/movieROC/vignettes/movieROC_vignette.pdf">vignette</a>
is accessible, including mathematical formalism and details about the
algorithms implemented.</p>
<h2 class="unnumbered" id="sec:complim">Computational considerations</h2>
<h4 class="unnumbered" data-number="6.0.1" id="dependencies">Dependencies</h4>
<p>Some functions of our package depend on other libraries available on
CRAN:</p>
<ul>
<li><p><code>gROC(X, D, side = "both", restric = TRUE, optim = TRUE, ...)</code> uses
the <code>allShortestPaths()</code> function in the
<a href="https://CRAN.R-project.org/package=e1071"><strong>e1071</strong></a> package
<span class="citation" data-cites="e1071CRAN2023">(<a href="#ref-e1071CRAN2023" role="doc-biblioref">Meyer et al. 2023</a>)</span>.</p></li>
<li><p><code>hROC(X, D, type = "lrm", ...)</code> and
<code>multiROC(X, D, method = "lrm", ...)</code> use the <code>lrm()</code> function in
the <a href="https://CRAN.R-project.org/package=rms"><strong>rms</strong></a> package
<span class="citation" data-cites="rmsCRAN2023">(<a href="#ref-rmsCRAN2023" role="doc-biblioref">Harrell Jr 2023</a>)</span>.</p></li>
<li><p><code>multiROC(X, D, method = "kernelOptimal", ...)</code> uses the <code>kde()</code>
function in the <a href="https://CRAN.R-project.org/package=ks"><strong>ks</strong></a>
package <span class="citation" data-cites="ksCRAN2023">(<a href="#ref-ksCRAN2023" role="doc-biblioref">Duong 2023</a>)</span>.</p></li>
<li><p><code>multiROC(X, D, method = "dynamicMeisner", ...)</code> uses the <code>maxTPR()</code>
function in the <strong>maxTPR</strong> package <span class="citation" data-cites="Meisner2021">(<a href="#ref-Meisner2021" role="doc-biblioref">Meisner et al. 2021</a>)</span>. This package was
removed from the CRAN repository, so we integrated the code of the
<code>maxTPR()</code> function into our package. This function uses
<code>Rsolnp::solnp()</code> and <code>robustbase::BYlogreg()</code>.</p></li>
<li><p><code>multiROC(X, D, method = "fixedLinear", methodLinear, ...)</code> uses the
R functions included in <span class="citation" data-cites="Kang2016">Kang et al. (<a href="#ref-Kang2016" role="doc-biblioref">2016</a>)</span> (Appendix). We integrated this
code into our package.</p></li>
<li><p><code>movieROC(obj, save = TRUE, ...)</code> uses the <code>saveGIF()</code> function in
the <a href="https://CRAN.R-project.org/package=animation"><strong>animation</strong></a>
package <span class="citation" data-cites="animationCRAN2021">(<a href="#ref-animationCRAN2021" role="doc-biblioref">Xie et al. 2021</a>)</span>.</p></li>
</ul>
<h4 class="unnumbered" data-number="6.0.2" id="limitations">Limitations</h4>
<p>Users should be aware of certain limitations while working with this
package:</p>
<ul>
<li><p>Some methods are potentially time-consuming, especially with medium
to large sample sizes:</p>
<ul>
<li><p>The estimation of the gROC curve under restriction <em>(C)</em> can be
computationally intensive, especially when considering different FPR
to locally optimize the search using
<code>gROC(X, D, side = "both", restric = TRUE, optim = TRUE, t0max = TRUE)</code>.
Note that this method involves a quite exhaustive search of the
self-contained classification subsets leading to the optimal gROC
curve estimate. However, even selecting different false-positive
rates <span class="math inline">\(t_0\)</span> to start from, it may not result in the optimal
achievable estimate under restriction <em>(C)</em>. Input parameters
<code>restric</code>, <code>optim</code>, <code>t0</code> and <code>t0max</code> for <code>gROC()</code> function included
in Table 1 of the
<a href="https://cran.r-project.org/web/packages/movieROC/vignettes/movieROC_vignette.pdf">vignette</a>
serve to control this search.</p></li>
<li><p>Similarly, it also occurs for multivariate markers when considering
linear frontiers with dynamic parameters (by using<br>
<code>multiROC(X, D, method = "dynamicMeisner" | "dynamicEmpirical")</code>).</p></li>
</ul></li>
<li><p>Most implemented R functions consider empirical estimation for the
resulting ROC curve, even if the procedure to estimate the decision
rules is semi-parametric. An exception is the <code>gROC_param()</code>
function, which accommodates the binormal scenario.</p></li>
<li><p>When visualizing classification regions for multivariate markers
with high dimension (<code>plot_buildROC()</code> and <code>movieROC()</code> functions
for a ‘<code>multiroc</code>’ object), our package provides some alternatives,
but additional improvements could provide further aid in
interpretation.</p></li>
</ul>
<h2 class="unnumbered" id="acknowledgements">Acknowledgements</h2>
<p>The authors acknowledge support by the Grants PID2019-104486GB-I00 and
PID2020-118101GB-I00 from Ministerio de Ciencia e Innovación (Spanish
Government), and by a financial Grant for Excellence Mobility for
lecturers and researchers subsidized by the University of Oviedo in
collaboration with Banco Santander.</p>
</div>
<div class="sourceCode" id="cb24"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h3 class="appendix" data-number="6.1" id="cran-packages-used"><span class="header-section-number">6.1</span> CRAN packages used</h3>
<p><a href="https://cran.r-project.org/package=movieROC">movieROC</a>, <a href="https://cran.r-project.org/package=pROC">pROC</a>, <a href="https://cran.r-project.org/package=ROCnReg">ROCnReg</a>, <a href="https://cran.r-project.org/package=OptimalCutpoints">OptimalCutpoints</a>, <a href="https://cran.r-project.org/package=nsROC">nsROC</a>, <a href="https://cran.r-project.org/package=rms">rms</a>, <a href="https://cran.r-project.org/package=ks">ks</a>, <a href="https://cran.r-project.org/package=e1071">e1071</a>, <a href="https://cran.r-project.org/package=animation">animation</a></p>
<h3 class="appendix" data-number="6.2" id="cran-task-views-implied-by-cited-packages"><span class="header-section-number">6.2</span> CRAN Task Views implied by cited packages</h3>
<p><a href="https://cran.r-project.org/view=Cluster">Cluster</a>, <a href="https://cran.r-project.org/view=Distributions">Distributions</a>, <a href="https://cran.r-project.org/view=DynamicVisualizations">DynamicVisualizations</a>, <a href="https://cran.r-project.org/view=Econometrics">Econometrics</a>, <a href="https://cran.r-project.org/view=Environmetrics">Environmetrics</a>, <a href="https://cran.r-project.org/view=MachineLearning">MachineLearning</a>, <a href="https://cran.r-project.org/view=Psychometrics">Psychometrics</a>, <a href="https://cran.r-project.org/view=ReproducibleResearch">ReproducibleResearch</a>, <a href="https://cran.r-project.org/view=Survival">Survival</a>, <a href="https://cran.r-project.org/view=TeachingStatistics">TeachingStatistics</a></p>
<h3 class="appendix" data-number="6.3" id="note"><span class="header-section-number">6.3</span> Note</h3>
<p>This article is converted from a Legacy LaTeX article using the
<a href="https://cran.r-project.org/package=texor">texor</a> package.
The pdf version is the official version. To report a problem with the html,
refer to CONTRIBUTE on the R Journal homepage.</p>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Bantis2021a" class="csl-entry" role="listitem">
L. E. Bantis, J. V. Tsimikas, G. R. Chambers, M. Capello, S. Hanash and Z. Feng. The length of the receiver operating characteristic curve and the two cutoff <span>Youden</span> index within a robust framework for discovery, evaluation, and cutoff estimation in biomarker studies involving improper receiver operating characteristic curves. <em>Statistics in Medicine</em>, 40(7): 1767–1789, 2021. URL <a href="https://doi.org/10.1002/sim.8869">https://doi.org/10.1002/sim.8869</a>.
</div>
<div id="ref-DiazCoto2020" class="csl-entry" role="listitem">
S. Dı́az-Coto. <span class="nocase">smoothROCtime: an <span>R</span> package for time-dependent <span>ROC</span> curve estimation</span>. <em>Computational Statistics</em>, 35: 1231–1251, 2020. URL <a href="https://doi.org/10.1007/s00180-020-00955-7">https://doi.org/10.1007/s00180-020-00955-7</a>.
</div>
<div id="ref-Dorfman1997" class="csl-entry" role="listitem">
D. D. Dorfman, K. S. Berbaum, C. E. Metz, R. V. Lenth, J. A. Hanley and H. A. Dagga. Proper receiver operating characteristic analysis: The bigamma model. <em>Academic Radiology</em>, 4(2): 138–149, 1997. URL <a href="https://doi.org/10.1016/s1076-6332(97)80013-x">https://doi.org/10.1016/s1076-6332(97)80013-x</a>.
</div>
<div id="ref-Duong2007" class="csl-entry" role="listitem">
T. Duong. Ks: Kernel density estimation and kernel discriminant analysis for multivariate data in <span>R</span>. <em>Journal of Statistical Software</em>, 21(7): 1–16, 2007. URL <a href="https://doi.org/10.18637/jss.v021.i07">https://doi.org/10.18637/jss.v021.i07</a>.
</div>
<div id="ref-ksCRAN2023" class="csl-entry" role="listitem">
T. Duong. <em>Ks: Kernel smoothing.</em> 2023. URL <a href="https://CRAN.R-project.org/package=ks">https://CRAN.R-project.org/package=ks</a>. R package version 1.14.1.
</div>
<div id="ref-Hanley1988" class="csl-entry" role="listitem">
J. A. Hanley. The robustness of the "binormal" assumptions used in fitting <span>ROC</span> curves. <em>Medical Decision Making</em>, 8(3): 197–203, 1988. URL <a href="https://doi.org/10.1177/0272989X8800800308">https://doi.org/10.1177/0272989X8800800308</a>. PMID: 3398748.
</div>
<div id="ref-Hanley1982" class="csl-entry" role="listitem">
J. A. Hanley and B. J. McNeil. The meaning and use of the area under a receiver operating characteristic (<span>ROC</span>) curve. <em>Radiology</em>, 143(1): 29–36, 1982. URL <a href="https://doi.org/110.1148/radiology.143.1.7063747">https://doi.org/110.1148/radiology.143.1.7063747</a>.
</div>
<div id="ref-rmsCRAN2023" class="csl-entry" role="listitem">
F. E. Harrell Jr. <em>Rms: Regression modeling strategies.</em> 2023. URL <a href="https://CRAN.R-project.org/package=rms">https://CRAN.R-project.org/package=rms</a>. R package version 6.7-1.
</div>
<div id="ref-Hotelling1933" class="csl-entry" role="listitem">
H. Hotelling. Analysis of a complex of statistical variables into principal components. <em>Journal of Educational Psychology</em>, 24(6): 417–441, 1933. URL <a href="https://doi.org/10.1037/h0071325">https://doi.org/10.1037/h0071325</a>.
</div>
<div id="ref-Hsieh1996" class="csl-entry" role="listitem">
F. Hsieh and B. W. Turnbull. Nonparametric and semiparametric estimation of the receiver operating characteristic curve. <em>The Annals of Statistics</em>, 24(1): 25–40, 1996. URL <a href="https://doi.org/10.1214/aos/1033066197">https://doi.org/10.1214/aos/1033066197</a>.
</div>
<div id="ref-Inacio2021" class="csl-entry" role="listitem">
V. Inácio, M. X. Rodrı́guez-Álvarez and P. Gayoso-Diz. Statistical evaluation of medical tests. <em>Annual Review of Statistics and Its Application</em>, 8(1): 41–67, 2021. URL <a href="https://doi.org/10.1146/annurev-statistics-040720-022432">https://doi.org/10.1146/annurev-statistics-040720-022432</a>.
</div>
<div id="ref-Kang2016" class="csl-entry" role="listitem">
L. Kang, A. Liu and L. Tian. Linear combination methods to improve diagnostic/prognostic accuracy on future observations. <em>Statistical Methods in Medical Research</em>, 25(4): 1359–1380, 2016. URL <a href="https://doi.org/10.1177/0962280213481053">https://doi.org/10.1177/0962280213481053</a>.
</div>
<div id="ref-Kauppi2016" class="csl-entry" role="listitem">
H. Kauppi. The generalized receiver operating characteristic curve. 114. Aboa Centre for Economics. 2016. URL <a href="https://www.econstor.eu/bitstream/10419/233329/1/aboa-ce-dp114.pdf">https://www.econstor.eu/bitstream/10419/233329/1/aboa-ce-dp114.pdf</a>.
</div>
<div id="ref-Liu2011" class="csl-entry" role="listitem">
C. Liu, A. Liu and S. Halabi. A min–max combination of biomarkers to improve diagnostic accuracy. <em>Statistics in Medicine</em>, 30(16): 2005–2014, 2011. URL <a href="https://doi.org/10.1002/sim.4238">https://doi.org/10.1002/sim.4238</a>.
</div>
<div id="ref-OptimalCutpoints2014" class="csl-entry" role="listitem">
M. López-Ratón, M. X. Rodrı́guez-Álvarez, C. C. Suárez and F. G. Sampedro. <span>OptimalCutpoints</span>: An <span>R</span> package for selecting optimal cutpoints in diagnostic tests. <em>Journal of Statistical Software</em>, 61(8): 1–36, 2014. URL <a href="https://doi.org/10.18637/jss.v061.i08">https://doi.org/10.18637/jss.v061.i08</a>.
</div>
<div id="ref-Camblor2018a" class="csl-entry" role="listitem">
P. Martı́nez-Camblor. <span class="nocase">On the paper <span>“Notes on the overlap measure as an alternative to the <span>Youden</span> index”</span></span>. <em>Statistics in Medicine</em>, 37(7): 1222–1224, 2018. URL <a href="https://doi.org/10.1002/sim.7517">https://doi.org/10.1002/sim.7517</a>.
</div>
<div id="ref-Camblor2017a" class="csl-entry" role="listitem">
P. Martı́nez-Camblor, N. Corral, C. Rey, J. Pascual and E. Cernuda-Morollón. Receiver operating characteristic curve generalization for non-monotone relationships. <em>Statistical Methods in Medical Research</em>, 26(1): 113–123, 2017. URL <a href="https://doi.org/10.1177/0962280214541095">https://doi.org/10.1177/0962280214541095</a>.
</div>
<div id="ref-Camblor2019b" class="csl-entry" role="listitem">
P. Martı́nez-Camblor and J. C. Pardo-Fernández. Parametric estimates for the receiver operating characteristic curve generalization for non-monotone relationships. <em>Statistical Methods in Medical Research</em>, 28(7): 2032–2048, 2019a. URL <a href="https://doi.org/10.1177/0962280217747009">https://doi.org/10.1177/0962280217747009</a>.
</div>
<div id="ref-Camblor2019c" class="csl-entry" role="listitem">
P. Martı́nez-Camblor and J. C. Pardo-Fernández. The <span>Youden</span> index in the generalized receiver operating characteristic curve context. <em>The International Journal of Biostatistics</em>, 15(1): 2019b. URL <a href="https://doi.org/10.1515/ijb-2018-0060">https://doi.org/10.1515/ijb-2018-0060</a>.
</div>
<div id="ref-Camblor2019a" class="csl-entry" role="listitem">
P. Martı́nez-Camblor, S. Pérez-Fernández and S. Dı́az-Coto. Improving the biomarker diagnostic capacity via functional transformations. <em>Journal of Applied Statistics</em>, 46(9): 1550–1566, 2019. URL <a href="https://doi.org/10.1080/02664763.2018.1554628">https://doi.org/10.1080/02664763.2018.1554628</a>.
</div>
<div id="ref-Camblor2021b" class="csl-entry" role="listitem">
P. Martı́nez-Camblor, S. Pérez-Fernández and S. Dı́az-Coto. Optimal classification scores based on multivariate marker transformations. <em>AStA Advances in Statistical Analysis</em>, 105(4): 581–599, 2021a. URL <a href="https://doi.org/10.1007/s10182-020-00388-z">https://doi.org/10.1007/s10182-020-00388-z</a>.
</div>
<div id="ref-Camblor2021a" class="csl-entry" role="listitem">
P. Martı́nez-Camblor, S. Pérez-Fernández and S. Dı́az-Coto. The area under the generalized receiver-operating characteristic curve. <em>The International Journal of Biostatistics</em>, 18(1): 293–306, 2021b. URL <a href="https://doi.org/10.1515/ijb-2020-0091">https://doi.org/10.1515/ijb-2020-0091</a>.
</div>
<div id="ref-Mayeux2004" class="csl-entry" role="listitem">
R. Mayeux. Biomarkers: Potential uses and limitations. <em>NeuroRx</em>, 1(2): 182–188, 2004. URL <a href="https://doi.org/10.1602/neurorx.1.2.182">https://doi.org/10.1602/neurorx.1.2.182</a>.
</div>
<div id="ref-McIntosh2002" class="csl-entry" role="listitem">
M. W. McIntosh and M. S. Pepe. Combining several screening tests: Optimality of the risk score. <em>Biometrics</em>, 58(3): 657–664, 2002. URL <a href="https://doi.org/10.1111/j.0006-341X.2002.00657.x">https://doi.org/10.1111/j.0006-341X.2002.00657.x</a>.
</div>
<div id="ref-Meisner2021" class="csl-entry" role="listitem">
A. Meisner, M. Carone, M. S. Pepe and K. F. Kerr. Combining biomarkers by maximizing the true positive rate for a fixed false positive rate. <em>Biometrical Journal</em>, 63(6): 1223–1240, 2021. URL <a href="https://doi.org/10.1002/bimj.202000210">https://doi.org/10.1002/bimj.202000210</a>.
</div>
<div id="ref-e1071CRAN2023" class="csl-entry" role="listitem">
D. Meyer, E. Dimitriadou, K. Hornik, A. Weingessel and F. Leisch. <em>e1071: Misc functions of the department of statistics, probability theory group (formerly: E1071), TU wien.</em> 2023. URL <a href="https://CRAN.R-project.org/package=e1071">https://CRAN.R-project.org/package=e1071</a>. R package version 1.7-13.
</div>
<div id="ref-Nakas2023" class="csl-entry" role="listitem">
C. T. Nakas, L. E. Bantis and C. Gatsonis. <em>ROC analysis for classification and prediction in practice.</em> CRC Press, 2023. URL <a href="https://www.routledge.com/ROC-Analysis-for-Classification-and-Prediction-in-Practice/Nakas-Bantis-Gatsonis/p/book/9781482233704">https://www.routledge.com/ROC-Analysis-for-Classification-and-Prediction-in-Practice/Nakas-Bantis-Gatsonis/p/book/9781482233704</a>.
</div>
<div id="ref-Pepe2003a" class="csl-entry" role="listitem">
M. S. Pepe. <em>The statistical evaluation of medical tests for classification and prediction.</em> Oxford University Press, 2003. URL <a href="https://global.oup.com/academic/product/the-statistical-evaluation-of-medical-tests-for-classification-and-prediction-9780198565826?cc=es&amp;lang=en&amp;">https://global.oup.com/academic/product/the-statistical-evaluation-of-medical-tests-for-classification-and-prediction-9780198565826?cc=es&amp;lang=en&amp;</a>.
</div>
<div id="ref-Pepe2000" class="csl-entry" role="listitem">
M. S. Pepe and M. L. Thompson. Combining diagnostic test results to increase accuracy. <em>Biostatistics</em>, 1(2): 123–140, 2000. URL <a href="https://doi.org/10.1093/biostatistics/1.2.123">https://doi.org/10.1093/biostatistics/1.2.123</a>.
</div>
<div id="ref-nsROC2018" class="csl-entry" role="listitem">
S. Pérez-Fernández, P. Martı́nez-Camblor, P. Filzmoser and N. Corral. <span class="nocase">nsROC</span>: An <span>R</span> package for non-standard <span>ROC</span> curve analysis. <em><span>The R Journal</span></em>, 10(2): 55–74, 2018. URL <a href="https://doi.org/10.32614/RJ-2018-043">https://doi.org/10.32614/RJ-2018-043</a>.
</div>
<div id="ref-PerezFernandez2020" class="csl-entry" role="listitem">
S. Pérez-Fernández, P. Martı́nez-Camblor, P. Filzmoser and N. Corral. Visualizing the decision rules behind the <span>ROC</span> curves: Understanding the classification process. <em><span>AStA Advances in Statistical Analysis</span></em>, 105(1): 135–161, 2021. URL <a href="https://doi.org/10.1007/s10182-020-00385-2">https://doi.org/10.1007/s10182-020-00385-2</a>.
</div>
<div id="ref-pROC2011" class="csl-entry" role="listitem">
X. Robin, N. Turck, A. Hainard, N. Tiberti, F. Lisacek, J.-C. Sanchez and M. Müller. P<span>ROC</span>: An open-source package for <span>R</span> and <span>S+</span> to analyze and compare <span>ROC</span> curves. <em>BMC Bioinformatics</em>, 12(77): 1–8, 2011. URL <a href="https://doi.org/10.1186/1471-2105-12-77">https://doi.org/10.1186/1471-2105-12-77</a>.
</div>
<div id="ref-ROCnReg2021" class="csl-entry" role="listitem">
M. X. Rodrı́guez-Álvarez and V. Inácio. <span>ROCnReg</span>: An <span>R</span> package for receiver operating characteristic curve inference with and without covariates. <em>The R Journal</em>, 13(1): 525–555, 2021. URL <a href="https://doi.org/10.32614/RJ-2021-066">https://doi.org/10.32614/RJ-2021-066</a>.
</div>
<div id="ref-Shen2012" class="csl-entry" role="listitem">
J. Shen, S. Wang, Y.-J. Zhang, M. Kappil, H.-C. Wu, M. G. Kibriya, Q. Wang, F. Jasmine, H. Ahsan, P.-H. Lee, et al. Genome-wide <span>DNA</span> methylation profiles in hepatocellular carcinoma. <em>Hepatology</em>, 55(6): 1799–1808, 2012. URL <a href="https://doi.org/10.1002/hep.25569">https://doi.org/10.1002/hep.25569</a>.
</div>
<div id="ref-Su1993" class="csl-entry" role="listitem">
J. Q. Su and J. S. Liu. Linear combinations of multiple diagnostic markers. <em>Journal of the American Statistical Association</em>, 88(424): 1350–1355, 1993. URL <a href="https://doi.org/10.1080/01621459.1993.10476417">https://doi.org/10.1080/01621459.1993.10476417</a>.
</div>
<div id="ref-animationCRAN2021" class="csl-entry" role="listitem">
Y. Xie, C. Mueller, L. Yu and W. Zhu. <em>Animation: A gallery of animations in statistics and utilities to create animations.</em> 2021. URL <a href="https://yihui.org/animation/">https://yihui.org/animation/</a>. R package version 2.7.
</div>
<div id="ref-Youden1950" class="csl-entry" role="listitem">
W. J. Youden. Index for rating diagnostic tests. <em>Cancer</em>, 3(1): 32–35, 1950. URL <a href="https://doi.org/10.1002/1097-0142(1950)3:1&lt;32::AID-CNCR2820030106&gt;3.0.CO;2-3">https://doi.org/10.1002/1097-0142(1950)3:1&lt;32::AID-CNCR2820030106&gt;3.0.CO;2-3</a>.
</div>
<div id="ref-Zhou2002" class="csl-entry" role="listitem">
X.-H. Zhou, D. K. McClish and N. A. Obuchowski. <em>Statistical methods in diagnostic medicine.</em> Wiley, 2002. URL <a href="https://doi.org/10.1002/9780470317082">https://doi.org/10.1002/9780470317082</a>.
</div>
<div id="ref-Zou1997" class="csl-entry" role="listitem">
K. H. Zou, W. J. Hall and D. E. Shapiro. Smooth non-parametric receiver operating characteristic (<span>ROC</span>) curves for continuous diagnostic tests. <em>Statistics in Medicine</em>, 16(19): 2143–2156, 1997. URL <a href="https://doi.org/10.1002/(SICI)1097-0258(19971015)16:19&lt;2143::AID-SIM655&gt;3.0.CO;2-3">https://doi.org/10.1002/(SICI)1097-0258(19971015)16:19&lt;2143::AID-SIM655&gt;3.0.CO;2-3</a>.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="references">References</h3>
<div id="references-listing"></div>
<h3 id="reuse">Reuse</h3>
<p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
<h3 id="citation">Citation</h3>
<p>For attribution, please cite this work as</p>
<pre class="citation-appendix short">Pérez-Fernández, et al., "movieROC: Visualizing the Decision Rules Underlying Binary Classification", The R Journal, 2026</pre>
<p>BibTeX citation</p>
<pre class="citation-appendix long">@article{RJ-2025-035,
  author = {Pérez-Fernández, Sonia and Martínez-Camblor, Pablo and Corral-Blanco, Norberto},
  title = {movieROC: Visualizing the Decision Rules Underlying Binary Classification},
  journal = {The R Journal},
  year = {2026},
  note = {https://doi.org/10.32614/RJ-2025-035},
  doi = {10.32614/RJ-2025-035},
  volume = {17},
  issue = {4},
  issn = {2073-4859},
  pages = {59-79}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
