---
title: News from the Bioconductor Project
abstract: The 'News from the Bioconductor Project' article from the 2009-1 issue.
author: |
  Bioconductor Team
  Program in Computational Biology
  Fred Hutchinson Cancer Research Center
date: '2009-06-01'
date_received: ~
journal:
  firstpage: 91
  lastpage: 92
volume: 1
issue: 1
slug: RJ-2009-1-bioconductor
citation_url: https://rjournal.github.io/
packages:
  cran: []
  bioc:
  - SSPA
  - dyebias
  - spkTools
  - Rmagpie
  - MCRestimate
  - flowflowJo
  - flowStats
  - GeneSelectMMD
  - tspair
  - metahdep
  - betr
  - qpgraph
  - KEGGgraph
  - geen2pathway
  - GOSemSim
  - SPIA
  - AffyTiling
  - rMAT
  - crlmm
  - GeneRegionScan
  - biocDatasets
  - PAnnBuilder
  - DAVIDQuery
  - Biostrings
  - ShortRead
preview: preview.png
bibliography: ~
CTV: ~
output:
  rjtools::rjournal_web_article:
    self_contained: yes
    toc: no
    legacy_pdf: yes
    web_only: yes
    mathjax: https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js
    md_extension: -tex_math_single_backslash

---

::: article
We are pleased to announce Bioconductor 2.4, released on April 21, 2009.
Bioconductor 2.4 is compatible with R 2.9.0, and consists of 320
packages. There are 28 new packages, and enhancements to many others.
Explore Bioconductor at <http://bioconductor.org>, and install packages
with

``` r
> source("http://bioconductor.org/biocLite.R")
> biocLite() # install standard packages...
> biocLite("IRanges") # ...or IRanges
```

# New packages

This release includes powerful new packages for diverse areas of
high-throughput analysis, including:

Refined differential expression

:   power analysis, pre-processing and error estimation
    ([*SSPA*](https://www.bioconductor.org/packages/release/bioc/html/SSPA.html),
    [*dyebias*](https://www.bioconductor.org/packages/release/bioc/html/dyebias.html),
    [*spkTools*](https://www.bioconductor.org/packages/release/bioc/html/spkTools.html),
    [*Rmagpie*](https://www.bioconductor.org/packages/release/bioc/html/Rmagpie.html),
    [*MCRestimate*](https://www.bioconductor.org/packages/release/bioc/html/MCRestimate.html)).

Flow cytometry

:   tools for data import
    ([*flowflowJo*](https://www.bioconductor.org/packages/release/bioc/html/flowflowJo.html))
    and auto-gating
    ([*flowStats*](https://www.bioconductor.org/packages/release/bioc/html/flowStats.html)).

Advanced clustering and gene selection

:   approaches
    ([*Rmagpie*](https://www.bioconductor.org/packages/release/bioc/html/Rmagpie.html),
    [*MCRestimate*](https://www.bioconductor.org/packages/release/bioc/html/MCRestimate.html),
    [*GeneSelectMMD*](https://www.bioconductor.org/packages/release/bioc/html/GeneSelectMMD.html),
    [*tspair*](https://www.bioconductor.org/packages/release/bioc/html/tspair.html),
    [*metahdep*](https://www.bioconductor.org/packages/release/bioc/html/metahdep.html),
    [*betr*](https://www.bioconductor.org/packages/release/bioc/html/betr.html)).

Probabilistic graphical models

:   for reverse engineering regulatory networks
    ([*qpgraph*](https://www.bioconductor.org/packages/release/bioc/html/qpgraph.html)).

Pathway analysis

:   using novel approaches
    ([*KEGGgraph*](https://www.bioconductor.org/packages/release/bioc/html/KEGGgraph.html),
    [*geen2pathway*](https://www.bioconductor.org/packages/release/bioc/html/geen2pathway.html),
    [*GOSemSim*](https://www.bioconductor.org/packages/release/bioc/html/GOSemSim.html),
    [*SPIA*](https://www.bioconductor.org/packages/release/bioc/html/SPIA.html)).

Technology-specific

:   packages
    ([*AffyTiling*](https://www.bioconductor.org/packages/release/bioc/html/AffyTiling.html),
    [*rMAT*](https://www.bioconductor.org/packages/release/bioc/html/rMAT.html),
    [*crlmm*](https://www.bioconductor.org/packages/release/bioc/html/crlmm.html),
    [*GeneRegionScan*](https://www.bioconductor.org/packages/release/bioc/html/GeneRegionScan.html)).

Interfaces

:   to data base and other external resources
    ([*biocDatasets*](https://www.bioconductor.org/packages/release/bioc/html/biocDatasets.html),
    [*PAnnBuilder*](https://www.bioconductor.org/packages/release/bioc/html/PAnnBuilder.html),
    [*DAVIDQuery*](https://www.bioconductor.org/packages/release/bioc/html/DAVIDQuery.html)).

# Annotation

Bioconductor 'annotation' packages contain biological information about
microarray probes and the genes they are meant to interrogate, or
contain ENTREZ gene based annotations of whole genomes. This release
updates existing database content, and lays the groundwork for 4 new
species: *Pan troglodytes*, *Macaca mulatta*, *Anopheles gambiae* and
*Xenopus laevis*. These species will be available in the development
branch starting in May. In addition, the 'yeast' package now contains
NCBI identifiers. A similarly enhanced *Arabidopsis* package will be in
the development branch in May.

# High-throughput sequencing

The stable of tools for high-throughput sequence analysis has developed
considerably during this release cycle, particularly data structures and
methods for conveniently navigating this complex data. Examples include
the `IRanges` and related classes and methods for manipulating ranged
(interval-based) data, the `Rle` class and its rich functionality for
run-length encoded data (e.g., genome-scale 'pileup' or coverage data),
the `XDataFrame` class allowing data frame-like functionality but with
more flexible column types (requiring only that the column object have
methods for length and subsetting), and the `GenomeData` and
`GenomeDataList` objects and methods for manipulating collections of
structured (e.g., by chromosome or locus) data. The
[*Biostrings*](https://www.bioconductor.org/packages/release/bioc/html/Biostrings.html)
package continues to provide very flexible pattern matching facilities,
while
[*ShortRead*](https://www.bioconductor.org/packages/release/bioc/html/ShortRead.html)
introduces new I/O functionality and the generation of HTML-based
quality assessment reports from diverse data sources.

# Other activities

Bioconductor package maintainers and the Bioconductor team invest
considerable effort in producing high-quality software. A focus during
development of Bioconductor 2.4 has been on more consistent and
widespread use of name spaces and package imports. These changes reduce
'collisions' between user and package variable names, and make package
code more robust. The Bioconductor team continues to ensure quality
software through technical and scientific reviews of new packages, and
daily builds of released packages on Linux, Windows, and Macintosh
platforms. The Bioconductor web site is also evolving. Bioconductor
'views' describing software functionality have been re-organized, and
package vignettes, reference manuals, and use statistics are readily
accessible from package home pages.

# Looking forward

The Bioconductor community will meet on July 27-28 at our annual
conference in Seattle for a combination of scientific talks and hands-on
tutorials. The active Bioconductor mailing lists
(<http://bioconductor.org/docs/mailList.html>) connect users with each
other, to domain experts, and to maintainers eager to ensure that their
packages satisfy the needs of leading edge approaches.

This will be a dynamic release cycle. New contributed packages are
already under review, and our build machines have started tracking the
latest development versions of R. In addition to development of
high-quality algorithms to address microarray data analysis, we
anticipate continued efforts to leverage diverse external data sources
and to meet the challenges of presenting high volume data in rich
graphical contexts.
:::
