---
title: Allele Imputation and Haplotype Determination from Databases Composed of Nuclear
  Families
abstract: 'The alleHap package is designed for imputing genetic missing data and reconstruct
  non recombinant haplotypes from pedigree databases in a deterministic way. When
  genotypes of related individuals are available in a number of linked genetic markers,
  the program starts by identifying haplotypes compatible with the observed genotypes
  in those markers without missing values. If haplotypes are identified in parents
  or offspring, missing alleles can be imputed in subjects containing missing values.
  Several scenarios are analyzed: family completely genotyped, children partially
  genotyped and parents completely genotyped, children fully genotyped and parents
  containing entirely or partially missing genotypes, and founders and their offspring
  both only partially genotyped. The alleHap package also has a function to simulate
  pedigrees including all these scenarios. This article describes in detail how our
  package works for the desired applications, including illustrated explanations and
  easily reproducible examples.'
author:
- Nathan Medina-Rodríguez
- Ángelo Santana
date: '2017-11-16'
date_received: '2016-08-25'
journal:
  firstpage: 35
  lastpage: 55
volume: 9
issue: 2
slug: RJ-2017-057
packages:
  cran:
  - haplo.ccs
  - haplo.stats
  - hsphase
  - linkim
  - rrBLUP
  - synbreed
  - alleHap
  bioc: ~
preview: preview.png
CTV: Genetics
output:
  rjtools::rjournal_web_article:
    self_contained: no
    toc: no
    legacy_pdf: yes

---
