<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
<title>Likelihood Ratio Test-Based Drug Safety Assessment using R Package pvLRT</title>

<meta property="description" itemprop="description" content="Medical product safety continues to be a key concern of the twenty-first century. Several spontaneous adverse events reporting databases established across the world continuously collect and archive adverse events data on various medical products. Determining signals of disproportional reporting (SDR) of product/adverse event pairs from these large-scale databases require the use of principled statistical techniques. Likelihood ratio test (LRT)-based approaches are particularly noteworthy in this context as they permit objective SDR detection without requiring ad hoc thresholds. However, their implementation is non-trivial due to analytical complexities, which necessitate the use of computation-heavy methods. Here we introduce R package pvLRT which implements a suite of LRT approaches, along with various post-processing and graphical summary functions, to facilitate simplified use of the methodologies. Detailed examples are provided to illustrate the package through analyses of three real product safety datasets obtained from publicly available FDA FAERS and VAERS databases."/>

<link rel="canonical" href="https://doi.org/10.32614/RJ-2023-027/"/>
<link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>
<link rel="icon" type="image/vnd.microsoft.icon" href="../../resources/favicon.ico"/>

<!--  https://schema.org/Article -->
<meta property="article:published" itemprop="datePublished" content="2023-08-26"/>
<meta property="article:created" itemprop="dateCreated" content="2023-08-26"/>
<meta name="article:author" content="Saptarshi Chakraborty"/>
<meta name="article:author" content="Marianthi Markatou"/>
<meta name="article:author" content="Robert Ball"/>

<!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
<meta property="og:title" content="Likelihood Ratio Test-Based Drug Safety Assessment using R Package pvLRT"/>
<meta property="og:type" content="article"/>
<meta property="og:description" content="Medical product safety continues to be a key concern of the twenty-first century. Several spontaneous adverse events reporting databases established across the world continuously collect and archive adverse events data on various medical products. Determining signals of disproportional reporting (SDR) of product/adverse event pairs from these large-scale databases require the use of principled statistical techniques. Likelihood ratio test (LRT)-based approaches are particularly noteworthy in this context as they permit objective SDR detection without requiring ad hoc thresholds. However, their implementation is non-trivial due to analytical complexities, which necessitate the use of computation-heavy methods. Here we introduce R package pvLRT which implements a suite of LRT approaches, along with various post-processing and graphical summary functions, to facilitate simplified use of the methodologies. Detailed examples are provided to illustrate the package through analyses of three real product safety datasets obtained from publicly available FDA FAERS and VAERS databases."/>
<meta property="og:url" content="https://doi.org/10.32614/RJ-2023-027/"/>
<meta property="og:image" content="https://journal.r-project.org/articles/RJ-2023-027/RJ-2023-027_files/figure-html5/show-heatmap-poisson-statin46-1.png"/>
<meta property="og:image:width" content="2496"/>
<meta property="og:image:height" content="960"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="The R Journal"/>

<!--  https://dev.twitter.com/cards/types/summary -->
<meta property="twitter:card" content="summary_large_image"/>
<meta property="twitter:title" content="Likelihood Ratio Test-Based Drug Safety Assessment using R Package pvLRT"/>
<meta property="twitter:description" content="Medical product safety continues to be a key concern of the twenty-first century. Several spontaneous adverse events reporting databases established across the world continuously collect and archive adverse events data on various medical products. Determining signals of disproportional reporting (SDR) of product/adverse event pairs from these large-scale databases require the use of principled statistical techniques. Likelihood ratio test (LRT)-based approaches are particularly noteworthy in this context as they permit objective SDR detection without requiring ad hoc thresholds. However, their implementation is non-trivial due to analytical complexities, which necessitate the use of computation-heavy methods. Here we introduce R package pvLRT which implements a suite of LRT approaches, along with various post-processing and graphical summary functions, to facilitate simplified use of the methodologies. Detailed examples are provided to illustrate the package through analyses of three real product safety datasets obtained from publicly available FDA FAERS and VAERS databases."/>
<meta property="twitter:url" content="https://doi.org/10.32614/RJ-2023-027/"/>
<meta property="twitter:image" content="https://journal.r-project.org/articles/RJ-2023-027/RJ-2023-027_files/figure-html5/show-heatmap-poisson-statin46-1.png"/>
<meta property="twitter:image:width" content="2496"/>
<meta property="twitter:image:height" content="960"/>

<!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
<meta name="citation_title" content="Likelihood Ratio Test-Based Drug Safety Assessment using R Package pvLRT"/>
<meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2023-027"/>
<meta name="citation_pdf_url" content="RJ-2023-027.pdf"/>
<meta name="citation_volume" content="15"/>
<meta name="citation_issue" content="1"/>
<meta name="citation_doi" content="10.32614/RJ-2023-027"/>
<meta name="citation_journal_title" content="The R Journal"/>
<meta name="citation_issn" content="2073-4859"/>
<meta name="citation_firstpage" content="101"/>
<meta name="citation_lastpage" content="121"/>
<meta name="citation_fulltext_world_readable" content=""/>
<meta name="citation_online_date" content="2023/08/26"/>
<meta name="citation_publication_date" content="2023/08/26"/>
<meta name="citation_author" content="Saptarshi Chakraborty"/>
<meta name="citation_author_institution" content="Department of Biostatistics, SUNY University at Buffalo"/>
<meta name="citation_author" content="Marianthi Markatou"/>
<meta name="citation_author_institution" content="Department of Biostatistics, SUNY University at Buffalo"/>
<meta name="citation_author" content="Robert Ball"/>
<meta name="citation_author_institution" content="Center for Drug Evaluation and Research, U.S. Food and Drug Administration"/>
<!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=A likelihood ratio test based method for signal detection with application to FDA’s drug safety data;citation_volume=106;citation_doi=10.1198/jasa.2011.ap10243;citation_author=Lan Huang;citation_author=Jyoti Zalkikar;citation_author=Ram C. Tiwari"/>
  <meta name="citation_reference" content="citation_title=Zero-inflated poisson model based likelihood ratio test for drug safety signal detection;citation_volume=26;citation_doi=10.1177/0962280214549590;citation_author=Lan Huang;citation_author=Dan Zheng;citation_author=Jyoti Zalkikar;citation_author=Ram Tiwari"/>
  <meta name="citation_reference" content="citation_title=An evaluation of statistical approaches to postmarketing surveillance;citation_volume=39;citation_doi=10.1002/sim.8447;citation_author=Yuxin Ding;citation_author=Marianthi Markatou;citation_author=Robert Ball"/>
  <meta name="citation_reference" content="citation_title=A pattern discovery framework for adverse event evaluation and inference in spontaneous reporting systems;citation_doi=10.1002/sam.11233;citation_author=M. Markatou;citation_author=R. Ball"/>
  <meta name="citation_reference" content="citation_title=On the use of the likelihood ratio test methodology in pharmacovigilance;citation_publisher=Wiley Online Library;citation_volume=41;citation_author=Saptarshi Chakraborty;citation_author=Anran Liu;citation_author=Robert Ball;citation_author=Marianthi Markatou"/>
  <meta name="citation_reference" content="citation_title=ggplot2: Elegant graphics for data analysis;citation_publisher=Springer-Verlag New York;citation_author=Hadley Wickham"/>
  <meta name="citation_reference" content="citation_title=Interactive web-based data visualization with r, plotly, and shiny;citation_publisher=Chapman; Hall/CRC;citation_author=Carson Sievert"/>
  <meta name="citation_reference" content="citation_title=Magrittr: A forward-pipe operator for r;citation_author=Stefan Milton Bache;citation_author=Hadley Wickham"/>
  <meta name="citation_reference" content="citation_title=Patchwork: The composer of plots;citation_author=Thomas Lin Pedersen"/>
  <meta name="citation_reference" content="citation_title=Data mining in the US Vaccine Adverse Event Reporting System (VAERS): early detection of intussusception and other events after rotavirus vaccination;citation_volume=19;citation_doi=10.1016/S0264-410X(01)00237-7;citation_author=Manette T. Niu;citation_author=Diane E. Erwin;citation_author=M. Miles Braun"/>
  <meta name="citation_reference" content="citation_title=An analysis of rotavirus vaccine reports to the vaccine adverse event reporting system: More than intussusception alone?;citation_volume=113;citation_doi=10.1542/peds.113.4.e353;citation_author=Penina Haber;citation_author=Robert T. Chen;citation_author=Lynn R. Zanardi;citation_author=Gina T. Mootrey;citation_author=Roseanne English;citation_author=Miles M. Braun;citation_author=An analysis of rotavirus vaccine reports to the vaccine adverse event reporting system: More than intussusception alone?"/>
  <meta name="citation_reference" content="citation_title=Extended likelihood ratio test-based methods for signal detection in a drug class with application to FDA’s adverse event reporting system database;citation_volume=27;citation_doi=10.1177/0962280216646678;citation_author=Yueqin Zhao;citation_author=Min Yi;citation_author=Ram C Tiwari"/>
  <meta name="citation_reference" content="citation_title=Package “parallel”;citation_publisher=https://stat.ethz.ch/R-manual/R-devel/library/parallel/doc/parallel.pdf;citation_author=Package “parallel”"/>
  <!--radix_placeholder_rmarkdown_metadata-->

<script type="text/json" id="radix-rmarkdown-metadata">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","draft","author","preamble","type","output","bibliography","params","always_allow_html","date","date_received","volume","issue","slug","journal","pdf_url","citation_url","doi","creative_commons","packages","CTV","csl","canonical_url"]}},"value":[{"type":"character","attributes":{},"value":["Likelihood Ratio Test-Based Drug Safety Assessment using R Package pvLRT"]},{"type":"character","attributes":{},"value":["Medical product safety continues to be a key concern of the twenty-first century. Several spontaneous adverse events reporting databases established across the world continuously collect and archive adverse events data on various medical products. Determining signals of disproportional reporting (SDR) of product/adverse event pairs from these large-scale databases require the use of principled statistical techniques. Likelihood ratio test (LRT)-based approaches are particularly noteworthy in this context as they permit objective SDR detection without requiring ad hoc thresholds. However, their implementation is non-trivial due to analytical complexities, which necessitate the use of computation-heavy methods. Here we introduce R package pvLRT which implements a suite of LRT approaches, along with various post-processing and graphical summary functions, to facilitate simplified use of the methodologies. Detailed examples are provided to illustrate the package through analyses of three real product safety datasets obtained from publicly available FDA FAERS and VAERS databases."]},{"type":"logical","attributes":{},"value":[false]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","orcid_id","email","address"]}},"value":[{"type":"character","attributes":{},"value":["Saptarshi Chakraborty"]},{"type":"character","attributes":{},"value":["Department of Biostatistics, SUNY University at Buffalo"]},{"type":"character","attributes":{},"value":["0000-0002-3121-9174"]},{"type":"character","attributes":{},"value":["chakrab2@buffalo.edu"]},{"type":"character","attributes":{},"value":["School of Public Health and Health Professions","704 Kimball Tower, Buffalo, NY 14214, USA"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","email","orcid_id","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Marianthi Markatou"]},{"type":"character","attributes":{},"value":["markatou@buffalo.edu"]},{"type":"character","attributes":{},"value":["0000-0002-1453-8229"]},{"type":"character","attributes":{},"value":["Department of Biostatistics, SUNY University at Buffalo"]},{"type":"character","attributes":{},"value":["School of Public Health and Health Professions","726 Kimball Tower, Buffalo, NY 14214, USA"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","orcid_id","email","address"]}},"value":[{"type":"character","attributes":{},"value":["Robert Ball"]},{"type":"character","attributes":{},"value":["Center for Drug Evaluation and Research, U.S. Food and Drug Administration"]},{"type":"character","attributes":{},"value":["0000-0002-1609-7420"]},{"type":"character","attributes":{},"value":["Robert.Ball@fda.hhs.gov"]},{"type":"character","attributes":{},"value":["Office of Surveillance and Epidemiology","Silver Spring, Maryland, USA"]}]}]},{"type":"character","attributes":{},"value":["% Any extra LaTeX you need in the preamble\n\\renewcommand{\\texttt}{\\code}\n\\usepackage{multirow, makecell}\n\\usepackage{multicol}\n\\usepackage{tabularx, booktabs}    \n"]},{"type":"character","attributes":{},"value":["package"]},{"type":"character","attributes":{},"value":["rjtools::rjournal_web_article"]},{"type":"character","attributes":{},"value":["pvlrt-package-paper-Rjournal.bib"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["arxiv"]}},"value":[{"type":"logical","attributes":{},"value":[false]}]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2023-08-26"]},{"type":"character","attributes":{},"value":["2022-03-02"]},{"type":"integer","attributes":{},"value":[15]},{"type":"integer","attributes":{},"value":[1]},{"type":"character","attributes":{},"value":["RJ-2023-027"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"integer","attributes":{},"value":[101]},{"type":"integer","attributes":{},"value":[121]}]},{"type":"character","attributes":{},"value":["RJ-2023-027.pdf"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2023-027"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2023-027"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"character","attributes":{},"value":["<<x[2]>>","PhViD","openEBGM","sglr","Sequential","SPRT","AEenrich","mds","pvLRT","ggplot2","plotly","data.table","tidyverse","magrittr","patchwork","parallel"]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["Bayesian","Finance","HighPerformanceComputing","Phylogenetics","Spatial","TeachingStatistics","TimeSeries","WebTechnologies"]},{"type":"character","attributes":{},"value":["/Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/library/rjtools/rjournal.csl"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2023-027/"]}]}
</script>
<!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["2022-54-copyedit.pdf","build/2022-54-copyedit.pdf","build/letter-to-editor-revision.pdf","build/pvLRT_0.5.tar.gz","build/pvlrt-package-paper-Rjournal_cache/html5/__packages","build/pvlrt-package-paper-Rjournal_cache/html5/run-poisson-gbca_ac53cbad78a415937fa5ba50fc6e0094.RData","build/pvlrt-package-paper-Rjournal_cache/html5/run-poisson-gbca_ac53cbad78a415937fa5ba50fc6e0094.rdb","build/pvlrt-package-paper-Rjournal_cache/html5/run-poisson-gbca_ac53cbad78a415937fa5ba50fc6e0094.rdx","build/pvlrt-package-paper-Rjournal_cache/html5/run-poisson-rv-list_18c018010d47c0c14a787e722e4e1325.RData","build/pvlrt-package-paper-Rjournal_cache/html5/run-poisson-rv-list_18c018010d47c0c14a787e722e4e1325.rdb","build/pvlrt-package-paper-Rjournal_cache/html5/run-poisson-rv-list_18c018010d47c0c14a787e722e4e1325.rdx","build/pvlrt-package-paper-Rjournal_cache/html5/run-poisson-statin_638d8560a95f831982bf1cd8aededf7d.RData","build/pvlrt-package-paper-Rjournal_cache/html5/run-poisson-statin_638d8560a95f831982bf1cd8aededf7d.rdb","build/pvlrt-package-paper-Rjournal_cache/html5/run-poisson-statin_638d8560a95f831982bf1cd8aededf7d.rdx","build/pvlrt-package-paper-Rjournal_cache/html5/run-poisson-statin46_52a99367268efa3823cfbbac36898e87.RData","build/pvlrt-package-paper-Rjournal_cache/html5/run-poisson-statin46_52a99367268efa3823cfbbac36898e87.rdb","build/pvlrt-package-paper-Rjournal_cache/html5/run-poisson-statin46_52a99367268efa3823cfbbac36898e87.rdx","build/pvlrt-package-paper-Rjournal_cache/html5/run-zip-gbca_a1cdb2f43fc71dc945acfc47e380742a.RData","build/pvlrt-package-paper-Rjournal_cache/html5/run-zip-gbca_a1cdb2f43fc71dc945acfc47e380742a.rdb","build/pvlrt-package-paper-Rjournal_cache/html5/run-zip-gbca_a1cdb2f43fc71dc945acfc47e380742a.rdx","build/pvlrt-package-paper-Rjournal_cache/html5/run-zip-rv-list_6bb836cc0e8e4d54e524e61a45ba9174.RData","build/pvlrt-package-paper-Rjournal_cache/html5/run-zip-rv-list_6bb836cc0e8e4d54e524e61a45ba9174.rdb","build/pvlrt-package-paper-Rjournal_cache/html5/run-zip-rv-list_6bb836cc0e8e4d54e524e61a45ba9174.rdx","build/pvlrt-package-paper-Rjournal_cache/html5/run-zip-statin_269110bce50747b57b4f87f847e9d1a8.RData","build/pvlrt-package-paper-Rjournal_cache/html5/run-zip-statin_269110bce50747b57b4f87f847e9d1a8.rdb","build/pvlrt-package-paper-Rjournal_cache/html5/run-zip-statin_269110bce50747b57b4f87f847e9d1a8.rdx","build/pvlrt-package-paper-Rjournal_cache/html5/run-zip-statin46_0a8507ae92006cb9a0f974bb42d051b2.RData","build/pvlrt-package-paper-Rjournal_cache/html5/run-zip-statin46_0a8507ae92006cb9a0f974bb42d051b2.rdb","build/pvlrt-package-paper-Rjournal_cache/html5/run-zip-statin46_0a8507ae92006cb9a0f974bb42d051b2.rdx","build/pvlrt-package-paper-Rjournal_cache/latex/__packages","build/pvlrt-package-paper-Rjournal_cache/latex/run-poisson-gbca_47b1d7be63154e343c039d9259bedae5.RData","build/pvlrt-package-paper-Rjournal_cache/latex/run-poisson-gbca_47b1d7be63154e343c039d9259bedae5.rdb","build/pvlrt-package-paper-Rjournal_cache/latex/run-poisson-gbca_47b1d7be63154e343c039d9259bedae5.rdx","build/pvlrt-package-paper-Rjournal_cache/latex/run-poisson-statin_339b42bf5add5cf6560c8451c42cc51d.RData","build/pvlrt-package-paper-Rjournal_cache/latex/run-poisson-statin_339b42bf5add5cf6560c8451c42cc51d.rdb","build/pvlrt-package-paper-Rjournal_cache/latex/run-poisson-statin_339b42bf5add5cf6560c8451c42cc51d.rdx","build/pvlrt-package-paper-Rjournal_cache/latex/run-poisson-statin46_9f021bba50994e5c8d33e7cbf6d72869.RData","build/pvlrt-package-paper-Rjournal_cache/latex/run-poisson-statin46_9f021bba50994e5c8d33e7cbf6d72869.rdb","build/pvlrt-package-paper-Rjournal_cache/latex/run-poisson-statin46_9f021bba50994e5c8d33e7cbf6d72869.rdx","build/pvlrt-package-paper-Rjournal_cache/latex/run-zip-gbca_07ecbddb4d7ca96fc093de8d5885574c.RData","build/pvlrt-package-paper-Rjournal_cache/latex/run-zip-gbca_07ecbddb4d7ca96fc093de8d5885574c.rdb","build/pvlrt-package-paper-Rjournal_cache/latex/run-zip-gbca_07ecbddb4d7ca96fc093de8d5885574c.rdx","build/pvlrt-package-paper-Rjournal_cache/latex/run-zip-statin_ccdfab4c5cd5ecbf2168e72a3b5559a7.RData","build/pvlrt-package-paper-Rjournal_cache/latex/run-zip-statin_ccdfab4c5cd5ecbf2168e72a3b5559a7.rdb","build/pvlrt-package-paper-Rjournal_cache/latex/run-zip-statin_ccdfab4c5cd5ecbf2168e72a3b5559a7.rdx","build/pvlrt-package-paper-Rjournal_cache/latex/run-zip-statin46_79e50452feef3e30ba42abea6c0fb4cc.RData","build/pvlrt-package-paper-Rjournal_cache/latex/run-zip-statin46_79e50452feef3e30ba42abea6c0fb4cc.rdb","build/pvlrt-package-paper-Rjournal_cache/latex/run-zip-statin46_79e50452feef3e30ba42abea6c0fb4cc.rdx","build/pvlrt-package-paper-Rjournal_files/anchor-4.2.2/anchor.min.js","build/pvlrt-package-paper-Rjournal_files/bowser-1.9.3/bowser.min.js","build/pvlrt-package-paper-Rjournal_files/crosstalk-1.2.0/css/crosstalk.min.css","build/pvlrt-package-paper-Rjournal_files/crosstalk-1.2.0/js/crosstalk.js","build/pvlrt-package-paper-Rjournal_files/crosstalk-1.2.0/js/crosstalk.js.map","build/pvlrt-package-paper-Rjournal_files/crosstalk-1.2.0/js/crosstalk.min.js","build/pvlrt-package-paper-Rjournal_files/crosstalk-1.2.0/js/crosstalk.min.js.map","build/pvlrt-package-paper-Rjournal_files/crosstalk-1.2.0/scss/crosstalk.scss","build/pvlrt-package-paper-Rjournal_files/distill-2.2.21/template.v2.js","build/pvlrt-package-paper-Rjournal_files/figure-html5/barplot-poisson-statin46-1.png","build/pvlrt-package-paper-Rjournal_files/figure-html5/create-heatmap-poisson-statin46-1.png","build/pvlrt-package-paper-Rjournal_files/figure-html5/heatmap-rv-all-1.png","build/pvlrt-package-paper-Rjournal_files/figure-html5/heatmap-zip-gbca-latex-1.png","build/pvlrt-package-paper-Rjournal_files/figure-html5/heatmap-zip-statin-latex-1.png","build/pvlrt-package-paper-Rjournal_files/figure-html5/show-heatmap-poisson-statin46-1.png","build/pvlrt-package-paper-Rjournal_files/figure-latex/barplot-poisson-statin46-1.pdf","build/pvlrt-package-paper-Rjournal_files/figure-latex/create-heatmap-poisson-statin46-1.pdf","build/pvlrt-package-paper-Rjournal_files/figure-latex/heatmap-zip-gbca-html-1.pdf","build/pvlrt-package-paper-Rjournal_files/figure-latex/heatmap-zip-gbca-latex-1.pdf","build/pvlrt-package-paper-Rjournal_files/figure-latex/heatmap-zip-statin-latex-1.pdf","build/pvlrt-package-paper-Rjournal_files/figure-latex/show-heatmap-poisson-statin46-1.pdf","build/pvlrt-package-paper-Rjournal_files/figure-latex/statin-46-heatmap-interactive-ploty-1.pdf","build/pvlrt-package-paper-Rjournal_files/header-attrs-2.23/header-attrs.js","build/pvlrt-package-paper-Rjournal_files/header-attrs-2.24/header-attrs.js","build/pvlrt-package-paper-Rjournal_files/htmlwidgets-1.6.2/htmlwidgets.js","build/pvlrt-package-paper-Rjournal_files/jquery-3.6.0/jquery-3.6.0.js","build/pvlrt-package-paper-Rjournal_files/jquery-3.6.0/jquery-3.6.0.min.js","build/pvlrt-package-paper-Rjournal_files/jquery-3.6.0/jquery-3.6.0.min.map","build/pvlrt-package-paper-Rjournal_files/plotly-binding-4.10.2/plotly.js","build/pvlrt-package-paper-Rjournal_files/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css","build/pvlrt-package-paper-Rjournal_files/plotly-main-2.11.1/plotly-latest.min.js","build/pvlrt-package-paper-Rjournal_files/popper-2.6.0/popper.min.js","build/pvlrt-package-paper-Rjournal_files/tippy-6.2.7/tippy-bundle.umd.min.js","build/pvlrt-package-paper-Rjournal_files/tippy-6.2.7/tippy-light-border.css","build/pvlrt-package-paper-Rjournal_files/tippy-6.2.7/tippy.css","build/pvlrt-package-paper-Rjournal_files/tippy-6.2.7/tippy.umd.min.js","build/pvlrt-package-paper-Rjournal_files/typedarray-0.1/typedarray.min.js","build/pvlrt-package-paper-Rjournal_files/webcomponents-2.0.0/webcomponents.js","build/pvlrt-package-paper-Rjournal.bib","build/pvlrt-package-paper-Rjournal.pdf","build/pvlrt-package-paper-Rjournal.R","build/pvlrt-package-paper-Rjournal.Rmd","build/pvlrt-package-paper-Rjournal.Rmd~","build/pvlrt-package-paper-Rjournal.tex","build/Response-to-Reviewers.pdf","build/rjarticle.zip","build/RJreferences.bib","build/RJwrapper.tex","letter-to-editor-revision.pdf","pvLRT_0.5.tar.gz","pvlrt-package-paper-Rjournal_files/anchor-4.2.2/anchor.min.js","pvlrt-package-paper-Rjournal_files/bowser-1.9.3/bowser.min.js","pvlrt-package-paper-Rjournal_files/crosstalk-1.2.0/css/crosstalk.min.css","pvlrt-package-paper-Rjournal_files/crosstalk-1.2.0/js/crosstalk.js","pvlrt-package-paper-Rjournal_files/crosstalk-1.2.0/js/crosstalk.js.map","pvlrt-package-paper-Rjournal_files/crosstalk-1.2.0/js/crosstalk.min.js","pvlrt-package-paper-Rjournal_files/crosstalk-1.2.0/js/crosstalk.min.js.map","pvlrt-package-paper-Rjournal_files/crosstalk-1.2.0/scss/crosstalk.scss","pvlrt-package-paper-Rjournal_files/distill-2.2.21/template.v2.js","pvlrt-package-paper-Rjournal_files/figure-html5/barplot-poisson-statin46-1.png","pvlrt-package-paper-Rjournal_files/figure-html5/create-heatmap-poisson-statin46-1.png","pvlrt-package-paper-Rjournal_files/figure-html5/heatmap-rv-all-1.png","pvlrt-package-paper-Rjournal_files/figure-html5/heatmap-zip-gbca-latex-1.png","pvlrt-package-paper-Rjournal_files/figure-html5/heatmap-zip-statin-latex-1.png","pvlrt-package-paper-Rjournal_files/figure-html5/show-heatmap-poisson-statin46-1.png","pvlrt-package-paper-Rjournal_files/figure-latex/barplot-poisson-statin46-1.pdf","pvlrt-package-paper-Rjournal_files/figure-latex/create-heatmap-poisson-statin46-1.pdf","pvlrt-package-paper-Rjournal_files/figure-latex/heatmap-rv-all-1.pdf","pvlrt-package-paper-Rjournal_files/figure-latex/heatmap-zip-gbca-latex-1.pdf","pvlrt-package-paper-Rjournal_files/figure-latex/heatmap-zip-statin-latex-1.pdf","pvlrt-package-paper-Rjournal_files/figure-latex/show-heatmap-poisson-statin46-1.pdf","pvlrt-package-paper-Rjournal_files/header-attrs-2.23/header-attrs.js","pvlrt-package-paper-Rjournal_files/htmlwidgets-1.6.0/htmlwidgets.js","pvlrt-package-paper-Rjournal_files/htmlwidgets-1.6.2/htmlwidgets.js","pvlrt-package-paper-Rjournal_files/jquery-3.6.0/jquery-3.6.0.js","pvlrt-package-paper-Rjournal_files/jquery-3.6.0/jquery-3.6.0.min.js","pvlrt-package-paper-Rjournal_files/jquery-3.6.0/jquery-3.6.0.min.map","pvlrt-package-paper-Rjournal_files/plotly-binding-4.10.0/plotly.js","pvlrt-package-paper-Rjournal_files/plotly-binding-4.10.2/plotly.js","pvlrt-package-paper-Rjournal_files/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css","pvlrt-package-paper-Rjournal_files/plotly-main-2.11.1/plotly-latest.min.js","pvlrt-package-paper-Rjournal_files/popper-2.6.0/popper.min.js","pvlrt-package-paper-Rjournal_files/tippy-6.2.7/tippy-bundle.umd.min.js","pvlrt-package-paper-Rjournal_files/tippy-6.2.7/tippy-light-border.css","pvlrt-package-paper-Rjournal_files/tippy-6.2.7/tippy.css","pvlrt-package-paper-Rjournal_files/tippy-6.2.7/tippy.umd.min.js","pvlrt-package-paper-Rjournal_files/typedarray-0.1/typedarray.min.js","pvlrt-package-paper-Rjournal_files/webcomponents-2.0.0/webcomponents.js","pvlrt-package-paper-Rjournal.bib","pvlrt-package-paper-Rjournal.tex","Response-to-Reviewers.pdf","RJ-2023-027_files/anchor-4.2.2/anchor.min.js","RJ-2023-027_files/bowser-1.9.3/bowser.min.js","RJ-2023-027_files/crosstalk-1.2.0/css/crosstalk.min.css","RJ-2023-027_files/crosstalk-1.2.0/js/crosstalk.js","RJ-2023-027_files/crosstalk-1.2.0/js/crosstalk.js.map","RJ-2023-027_files/crosstalk-1.2.0/js/crosstalk.min.js","RJ-2023-027_files/crosstalk-1.2.0/js/crosstalk.min.js.map","RJ-2023-027_files/crosstalk-1.2.0/scss/crosstalk.scss","RJ-2023-027_files/distill-2.2.21/template.v2.js","RJ-2023-027_files/figure-html5/barplot-poisson-statin46-1.png","RJ-2023-027_files/figure-html5/create-heatmap-poisson-statin46-1.png","RJ-2023-027_files/figure-html5/heatmap-rv-all-1.png","RJ-2023-027_files/figure-html5/heatmap-zip-gbca-latex-1.png","RJ-2023-027_files/figure-html5/heatmap-zip-statin-latex-1.png","RJ-2023-027_files/figure-html5/show-heatmap-poisson-statin46-1.png","RJ-2023-027_files/figure-latex/barplot-poisson-statin46-1.pdf","RJ-2023-027_files/figure-latex/create-heatmap-poisson-statin46-1.pdf","RJ-2023-027_files/figure-latex/heatmap-rv-all-1.pdf","RJ-2023-027_files/figure-latex/heatmap-zip-gbca-html-1.pdf","RJ-2023-027_files/figure-latex/heatmap-zip-gbca-latex-1.pdf","RJ-2023-027_files/figure-latex/heatmap-zip-statin-latex-1.pdf","RJ-2023-027_files/figure-latex/show-heatmap-poisson-statin46-1.pdf","RJ-2023-027_files/figure-latex/statin-46-heatmap-interactive-ploty-1.pdf","RJ-2023-027_files/header-attrs-2.23/header-attrs.js","RJ-2023-027_files/header-attrs-2.24/header-attrs.js","RJ-2023-027_files/htmlwidgets-1.6.2/htmlwidgets.js","RJ-2023-027_files/jquery-3.6.0/jquery-3.6.0.js","RJ-2023-027_files/jquery-3.6.0/jquery-3.6.0.min.js","RJ-2023-027_files/jquery-3.6.0/jquery-3.6.0.min.map","RJ-2023-027_files/plotly-binding-4.10.2/plotly.js","RJ-2023-027_files/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css","RJ-2023-027_files/plotly-main-2.11.1/plotly-latest.min.js","RJ-2023-027_files/popper-2.6.0/popper.min.js","RJ-2023-027_files/tippy-6.2.7/tippy-bundle.umd.min.js","RJ-2023-027_files/tippy-6.2.7/tippy-light-border.css","RJ-2023-027_files/tippy-6.2.7/tippy.css","RJ-2023-027_files/tippy-6.2.7/tippy.umd.min.js","RJ-2023-027_files/typedarray-0.1/typedarray.min.js","RJ-2023-027_files/webcomponents-2.0.0/webcomponents.js","RJ-2023-027.log","RJ-2023-027.pdf","RJ-2023-027.tex","rjarticle.zip","RJreferences.bib","RJwrapper.log","RJwrapper.tex"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
<meta name="distill:offset" content="../.."/>

<script type="application/javascript">

  window.headroom_prevent_pin = false;

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // initialize headroom for banner
    var header = $('header').get(0);
    var headerHeight = header.offsetHeight;
    var headroom = new Headroom(header, {
      tolerance: 5,
      onPin : function() {
        if (window.headroom_prevent_pin) {
          window.headroom_prevent_pin = false;
          headroom.unpin();
        }
      }
    });
    headroom.init();
    if(window.location.hash)
      headroom.unpin();
    $(header).addClass('headroom--transition');

    // offset scroll location for banner on hash change
    // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
    window.addEventListener("hashchange", function(event) {
      window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
    });

    // responsive menu
    $('.distill-site-header').each(function(i, val) {
      var topnav = $(this);
      var toggle = topnav.find('.nav-toggle');
      toggle.on('click', function() {
        topnav.toggleClass('responsive');
      });
    });

    // nav dropdowns
    $('.nav-dropbtn').click(function(e) {
      $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
      $(this).parent().siblings('.nav-dropdown')
         .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $("body").click(function(e){
      $('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $(".nav-dropdown").click(function(e){
      e.stopPropagation();
    });
  });
</script>

<style type="text/css">

/* Theme (user-documented overrideables for nav appearance) */

.distill-site-nav {
  color: rgba(255, 255, 255, 0.8);
  background-color: #0F2E3D;
  font-size: 15px;
  font-weight: 300;
}

.distill-site-nav a {
  color: inherit;
  text-decoration: none;
}

.distill-site-nav a:hover {
  color: white;
}

@media print {
  .distill-site-nav {
    display: none;
  }
}

.distill-site-header {

}

.distill-site-footer {

}


/* Site Header */

.distill-site-header {
  width: 100%;
  box-sizing: border-box;
  z-index: 3;
}

.distill-site-header .nav-left {
  display: inline-block;
  margin-left: 8px;
}

@media screen and (max-width: 768px) {
  .distill-site-header .nav-left {
    margin-left: 0;
  }
}


.distill-site-header .nav-right {
  float: right;
  margin-right: 8px;
}

.distill-site-header a,
.distill-site-header .title {
  display: inline-block;
  text-align: center;
  padding: 14px 10px 14px 10px;
}

.distill-site-header .title {
  font-size: 18px;
  min-width: 150px;
}

.distill-site-header .logo {
  padding: 0;
}

.distill-site-header .logo img {
  display: none;
  max-height: 20px;
  width: auto;
  margin-bottom: -4px;
}

.distill-site-header .nav-image img {
  max-height: 18px;
  width: auto;
  display: inline-block;
  margin-bottom: -3px;
}



@media screen and (min-width: 1000px) {
  .distill-site-header .logo img {
    display: inline-block;
  }
  .distill-site-header .nav-left {
    margin-left: 20px;
  }
  .distill-site-header .nav-right {
    margin-right: 20px;
  }
  .distill-site-header .title {
    padding-left: 12px;
  }
}


.distill-site-header .nav-toggle {
  display: none;
}

.nav-dropdown {
  display: inline-block;
  position: relative;
}

.nav-dropdown .nav-dropbtn {
  border: none;
  outline: none;
  color: rgba(255, 255, 255, 0.8);
  padding: 16px 10px;
  background-color: transparent;
  font-family: inherit;
  font-size: inherit;
  font-weight: inherit;
  margin: 0;
  margin-top: 1px;
  z-index: 2;
}

.nav-dropdown-content {
  display: none;
  position: absolute;
  background-color: white;
  min-width: 200px;
  border: 1px solid rgba(0,0,0,0.15);
  border-radius: 4px;
  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
  z-index: 1;
  margin-top: 2px;
  white-space: nowrap;
  padding-top: 4px;
  padding-bottom: 4px;
}

.nav-dropdown-content hr {
  margin-top: 4px;
  margin-bottom: 4px;
  border: none;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.nav-dropdown-active {
  display: block;
}

.nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
  color: black;
  padding: 6px 24px;
  text-decoration: none;
  display: block;
  text-align: left;
}

.nav-dropdown-content .nav-dropdown-header {
  display: block;
  padding: 5px 24px;
  padding-bottom: 0;
  text-transform: uppercase;
  font-size: 14px;
  color: #999999;
  white-space: nowrap;
}

.nav-dropdown:hover .nav-dropbtn {
  color: white;
}

.nav-dropdown-content a:hover {
  background-color: #ddd;
  color: black;
}

.nav-right .nav-dropdown-content {
  margin-left: -45%;
  right: 0;
}

@media screen and (max-width: 768px) {
  .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
  .distill-site-header a.nav-toggle {
    float: right;
    display: block;
  }
  .distill-site-header .title {
    margin-left: 0;
  }
  .distill-site-header .nav-right {
    margin-right: 0;
  }
  .distill-site-header {
    overflow: hidden;
  }
  .nav-right .nav-dropdown-content {
    margin-left: 0;
  }
}


@media screen and (max-width: 768px) {
  .distill-site-header.responsive {position: relative; min-height: 500px; }
  .distill-site-header.responsive a.nav-toggle {
    position: absolute;
    right: 0;
    top: 0;
  }
  .distill-site-header.responsive a,
  .distill-site-header.responsive .nav-dropdown {
    display: block;
    text-align: left;
  }
  .distill-site-header.responsive .nav-left,
  .distill-site-header.responsive .nav-right {
    width: 100%;
  }
  .distill-site-header.responsive .nav-dropdown {float: none;}
  .distill-site-header.responsive .nav-dropdown-content {position: relative;}
  .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
    display: block;
    width: 100%;
    text-align: left;
  }
}

/* Site Footer */

.distill-site-footer {
  width: 100%;
  overflow: hidden;
  box-sizing: border-box;
  z-index: 3;
  margin-top: 30px;
  padding-top: 30px;
  padding-bottom: 30px;
  text-align: center;
}

/* Headroom */

d-title {
  padding-top: 6rem;
}

@media print {
  d-title {
    padding-top: 4rem;
  }
}

.headroom {
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
}

.headroom--transition {
  transition: all .4s ease-in-out;
}

.headroom--unpinned {
  top: -100px;
}

.headroom--pinned {
  top: 0;
}

/* adjust viewport for navbar height */
/* helps vertically center bootstrap (non-distill) content */
.min-vh-100 {
  min-height: calc(100vh - 100px) !important;
}

</style>

<script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="../../site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"/>
<link href="../../site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"/>
<script src="../../site_libs/headroom-0.9.4/headroom.min.js"></script>
<!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

<style type="text/css">

body {
  background-color: white;
}

.pandoc-table {
  width: 100%;
}

.pandoc-table>caption {
  margin-bottom: 10px;
}

.pandoc-table th:not([align]) {
  text-align: left;
}

.pagedtable-footer {
  font-size: 15px;
}

d-byline .byline {
  grid-template-columns: 2fr 2fr;
}

d-byline .byline h3 {
  margin-block-start: 1.5em;
}

d-byline .byline .authors-affiliations h3 {
  margin-block-start: 0.5em;
}

.authors-affiliations .orcid-id {
  width: 16px;
  height:16px;
  margin-left: 4px;
  margin-right: 4px;
  vertical-align: middle;
  padding-bottom: 2px;
}

d-title .dt-tags {
  margin-top: 1em;
  grid-column: text;
}

.dt-tags .dt-tag {
  text-decoration: none;
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0em 0.4em;
  margin-right: 0.5em;
  margin-bottom: 0.4em;
  font-size: 70%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

d-article table.gt_table td,
d-article table.gt_table th {
  border-bottom: none;
  font-size: 100%;
}

.html-widget {
  margin-bottom: 2.0em;
}

.l-screen-inset {
  padding-right: 16px;
}

.l-screen .caption {
  margin-left: 10px;
}

.shaded {
  background: rgb(247, 247, 247);
  padding-top: 20px;
  padding-bottom: 20px;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .html-widget {
  margin-bottom: 0;
  border: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .shaded-content {
  background: white;
}

.text-output {
  margin-top: 0;
  line-height: 1.5em;
}

.hidden {
  display: none !important;
}

hr.section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  margin: 0px;
}


d-byline {
  border-top: none;
}

d-article {
  padding-top: 2.5rem;
  padding-bottom: 30px;
  border-top: none;
}

d-appendix {
  padding-top: 30px;
}

d-article>p>img {
  width: 100%;
}

d-article h2 {
  margin: 1rem 0 1.5rem 0;
}

d-article h3 {
  margin-top: 1.5rem;
}

d-article iframe {
  border: 1px solid rgba(0, 0, 0, 0.1);
  margin-bottom: 2.0em;
  width: 100%;
}

/* Tweak code blocks */

d-article div.sourceCode code,
d-article pre code {
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
}

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: auto;
}

d-article div.sourceCode {
  background-color: white;
}

d-article div.sourceCode pre {
  padding-left: 10px;
  font-size: 12px;
  border-left: 2px solid rgba(0,0,0,0.1);
}

d-article pre {
  font-size: 12px;
  color: black;
  background: none;
  margin-top: 0;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

d-article pre a {
  border-bottom: none;
}

d-article pre a:hover {
  border-bottom: none;
  text-decoration: underline;
}

d-article details {
  grid-column: text;
  margin-bottom: 0.8em;
}

@media(min-width: 768px) {

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: visible !important;
}

d-article div.sourceCode pre {
  padding-left: 18px;
  font-size: 14px;
}

/* tweak for Pandoc numbered line within distill */
d-article pre.numberSource code > span {
    left: -2em;
}

d-article pre {
  font-size: 14px;
}

}

figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

/* CSS for d-contents */

.d-contents {
  grid-column: text;
  color: rgba(0,0,0,0.8);
  font-size: 0.9em;
  padding-bottom: 1em;
  margin-bottom: 1em;
  padding-bottom: 0.5em;
  margin-bottom: 1em;
  padding-left: 0.25em;
  justify-self: start;
}

@media(min-width: 1000px) {
  .d-contents.d-contents-float {
    height: 0;
    grid-column-start: 1;
    grid-column-end: 4;
    justify-self: center;
    padding-right: 3em;
    padding-left: 2em;
  }
}

.d-contents nav h3 {
  font-size: 18px;
  margin-top: 0;
  margin-bottom: 1em;
}

.d-contents li {
  list-style-type: none
}

.d-contents nav > ul {
  padding-left: 0;
}

.d-contents ul {
  padding-left: 1em
}

.d-contents nav ul li {
  margin-top: 0.6em;
  margin-bottom: 0.2em;
}

.d-contents nav a {
  font-size: 13px;
  border-bottom: none;
  text-decoration: none
  color: rgba(0, 0, 0, 0.8);
}

.d-contents nav a:hover {
  text-decoration: underline solid rgba(0, 0, 0, 0.6)
}

.d-contents nav > ul > li > a {
  font-weight: 600;
}

.d-contents nav > ul > li > ul {
  font-weight: inherit;
}

.d-contents nav > ul > li > ul > li {
  margin-top: 0.2em;
}


.d-contents nav ul {
  margin-top: 0;
  margin-bottom: 0.25em;
}

.d-article-with-toc h2:nth-child(2) {
  margin-top: 0;
}


/* Figure */

.figure {
  position: relative;
  margin-bottom: 2.5em;
  margin-top: 1.5em;
}

.figure .caption {
  color: rgba(0, 0, 0, 0.6);
  font-size: 12px;
  line-height: 1.5em;
}

.figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

.figure .caption a {
  color: rgba(0, 0, 0, 0.6);
}

.figure .caption b,
.figure .caption strong, {
  font-weight: 600;
  color: rgba(0, 0, 0, 1.0);
}

/* Citations */

d-article .citation {
  color: inherit;
  cursor: inherit;
}

div.hanging-indent{
  margin-left: 1em; text-indent: -1em;
}

/* Citation hover box */

.tippy-box[data-theme~=light-border] {
  background-color: rgba(250, 250, 250, 0.95);
}

.tippy-content > p {
  margin-bottom: 0;
  padding: 2px;
}


/* Tweak 1000px media break to show more text */

@media(min-width: 1000px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 16px;
  }

  .grid {
    grid-column-gap: 16px;
  }

  d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
  }
  figure .caption, .figure .caption, figure figcaption {
    font-size: 13px;
  }
}

@media(min-width: 1180px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 32px;
  }

  .grid {
    grid-column-gap: 32px;
  }
}


/* Get the citation styles for the appendix (not auto-injected on render since
   we do our own rendering of the citation appendix) */

d-appendix .citation-appendix,
.d-appendix .citation-appendix {
  font-size: 11px;
  line-height: 15px;
  border-left: 1px solid rgba(0, 0, 0, 0.1);
  padding-left: 18px;
  border: 1px solid rgba(0,0,0,0.1);
  background: rgba(0, 0, 0, 0.02);
  padding: 10px 18px;
  border-radius: 3px;
  color: rgba(150, 150, 150, 1);
  overflow: hidden;
  margin-top: -12px;
  white-space: pre-wrap;
  word-wrap: break-word;
}

/* Include appendix styles here so they can be overridden */

d-appendix {
  contain: layout style;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-top: 60px;
  margin-bottom: 0;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  color: rgba(0,0,0,0.5);
  padding-top: 60px;
  padding-bottom: 48px;
}

d-appendix h3 {
  grid-column: page-start / text-start;
  font-size: 15px;
  font-weight: 500;
  margin-top: 1em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.65);
}

d-appendix h3 + * {
  margin-top: 1em;
}

d-appendix ol {
  padding: 0 0 0 15px;
}

@media (min-width: 768px) {
  d-appendix ol {
    padding: 0 0 0 30px;
    margin-left: -30px;
  }
}

d-appendix li {
  margin-bottom: 1em;
}

d-appendix a {
  color: rgba(0, 0, 0, 0.6);
}

d-appendix > * {
  grid-column: text;
}

d-appendix > d-footnote-list,
d-appendix > d-citation-list,
d-appendix > distill-appendix {
  grid-column: screen;
}

/* Include footnote styles here so they can be overridden */

d-footnote-list {
  contain: layout style;
}

d-footnote-list > * {
  grid-column: text;
}

d-footnote-list a.footnote-backlink {
  color: rgba(0,0,0,0.3);
  padding-left: 0.5em;
}



/* Anchor.js */

.anchorjs-link {
  /*transition: all .25s linear; */
  text-decoration: none;
  border-bottom: none;
}
*:hover > .anchorjs-link {
  margin-left: -1.125em !important;
  text-decoration: none;
  border-bottom: none;
}

/* Social footer */

.social_footer {
  margin-top: 30px;
  margin-bottom: 0;
  color: rgba(0,0,0,0.67);
}

.disqus-comments {
  margin-right: 30px;
}

.disqus-comment-count {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  cursor: pointer;
}

#disqus_thread {
  margin-top: 30px;
}

.article-sharing a {
  border-bottom: none;
  margin-right: 8px;
}

.article-sharing a:hover {
  border-bottom: none;
}

.sidebar-section.subscribe {
  font-size: 12px;
  line-height: 1.6em;
}

.subscribe p {
  margin-bottom: 0.5em;
}


.article-footer .subscribe {
  font-size: 15px;
  margin-top: 45px;
}


.sidebar-section.custom {
  font-size: 12px;
  line-height: 1.6em;
}

.custom p {
  margin-bottom: 0.5em;
}

/* Styles for listing layout (hide title) */
.layout-listing d-title, .layout-listing .d-title {
  display: none;
}

/* Styles for posts lists (not auto-injected) */


.posts-with-sidebar {
  padding-left: 45px;
  padding-right: 45px;
}

.posts-list .description h2,
.posts-list .description p {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

.posts-list .description h2 {
  font-weight: 700;
  border-bottom: none;
  padding-bottom: 0;
}

.posts-list h2.post-tag {
  border-bottom: 1px solid rgba(0, 0, 0, 0.2);
  padding-bottom: 12px;
}
.posts-list {
  margin-top: 60px;
  margin-bottom: 24px;
}

.posts-list .post-preview {
  text-decoration: none;
  overflow: hidden;
  display: block;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  padding: 24px 0;
}

.post-preview-last {
  border-bottom: none !important;
}

.posts-list .posts-list-caption {
  grid-column: screen;
  font-weight: 400;
}

.posts-list .post-preview h2 {
  margin: 0 0 6px 0;
  line-height: 1.2em;
  font-style: normal;
  font-size: 24px;
}

.posts-list .post-preview p {
  margin: 0 0 12px 0;
  line-height: 1.4em;
  font-size: 16px;
}

.posts-list .post-preview .thumbnail {
  box-sizing: border-box;
  margin-bottom: 24px;
  position: relative;
  max-width: 500px;
}
.posts-list .post-preview img {
  width: 100%;
  display: block;
}

.posts-list .metadata {
  font-size: 12px;
  line-height: 1.4em;
  margin-bottom: 18px;
}

.posts-list .metadata > * {
  display: inline-block;
}

.posts-list .metadata .publishedDate {
  margin-right: 2em;
}

.posts-list .metadata .dt-authors {
  display: block;
  margin-top: 0.3em;
  margin-right: 2em;
}

.posts-list .dt-tags {
  display: block;
  line-height: 1em;
}

.posts-list .dt-tags .dt-tag {
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0.3em 0.4em;
  margin-right: 0.2em;
  margin-bottom: 0.4em;
  font-size: 60%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

.posts-list img {
  opacity: 1;
}

.posts-list img[data-src] {
  opacity: 0;
}

.posts-more {
  clear: both;
}


.posts-sidebar {
  font-size: 16px;
}

.posts-sidebar h3 {
  font-size: 16px;
  margin-top: 0;
  margin-bottom: 0.5em;
  font-weight: 400;
  text-transform: uppercase;
}

.sidebar-section {
  margin-bottom: 30px;
}

.categories ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.categories li {
  color: rgba(0, 0, 0, 0.8);
  margin-bottom: 0;
}

.categories li>a {
  border-bottom: none;
}

.categories li>a:hover {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
}

.categories .active {
  font-weight: 600;
}

.categories .category-count {
  color: rgba(0, 0, 0, 0.4);
}


@media(min-width: 768px) {
  .posts-list .post-preview h2 {
    font-size: 26px;
  }
  .posts-list .post-preview .thumbnail {
    float: right;
    width: 30%;
    margin-bottom: 0;
  }
  .posts-list .post-preview .description {
    float: left;
    width: 45%;
  }
  .posts-list .post-preview .metadata {
    float: left;
    width: 20%;
    margin-top: 8px;
  }
  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.5em;
    font-size: 16px;
  }
  .posts-with-sidebar .posts-list {
    float: left;
    width: 75%;
  }
  .posts-with-sidebar .posts-sidebar {
    float: right;
    width: 20%;
    margin-top: 60px;
    padding-top: 24px;
    padding-bottom: 24px;
  }
}


/* Improve display for browsers without grid (IE/Edge <= 15) */

.downlevel {
  line-height: 1.6em;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  margin: 0;
}

.downlevel .d-title {
  padding-top: 6rem;
  padding-bottom: 1.5rem;
}

.downlevel .d-title h1 {
  font-size: 50px;
  font-weight: 700;
  line-height: 1.1em;
  margin: 0 0 0.5rem;
}

.downlevel .d-title p {
  font-weight: 300;
  font-size: 1.2rem;
  line-height: 1.55em;
  margin-top: 0;
}

.downlevel .d-byline {
  padding-top: 0.8em;
  padding-bottom: 0.8em;
  font-size: 0.8rem;
  line-height: 1.8em;
}

.downlevel .section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
}

.downlevel .d-article {
  font-size: 1.06rem;
  line-height: 1.7em;
  padding-top: 1rem;
  padding-bottom: 2rem;
}


.downlevel .d-appendix {
  padding-left: 0;
  padding-right: 0;
  max-width: none;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.5);
  padding-top: 40px;
  padding-bottom: 48px;
}

.downlevel .footnotes ol {
  padding-left: 13px;
}

.downlevel .base-grid,
.downlevel .distill-header,
.downlevel .d-title,
.downlevel .d-abstract,
.downlevel .d-article,
.downlevel .d-appendix,
.downlevel .distill-appendix,
.downlevel .d-byline,
.downlevel .d-footnote-list,
.downlevel .d-citation-list,
.downlevel .distill-footer,
.downlevel .appendix-bottom,
.downlevel .posts-container {
  padding-left: 40px;
  padding-right: 40px;
}

@media(min-width: 768px) {
  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
  padding-left: 150px;
  padding-right: 150px;
  max-width: 900px;
}
}

.downlevel pre code {
  display: block;
  border-left: 2px solid rgba(0, 0, 0, .1);
  padding: 0 0 0 20px;
  font-size: 14px;
}

.downlevel code, .downlevel pre {
  color: black;
  background: none;
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

.downlevel .posts-list .post-preview {
  color: inherit;
}



</style>

<script type="application/javascript">

function is_downlevel_browser() {
  if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                 window.navigator.userAgent)) {
    return true;
  } else {
    return window.load_distill_framework === undefined;
  }
}

// show body when load is complete
function on_load_complete() {

  // add anchors
  if (window.anchors) {
    window.anchors.options.placement = 'left';
    window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
  }


  // set body to visible
  document.body.style.visibility = 'visible';

  // force redraw for leaflet widgets
  if (window.HTMLWidgets) {
    var maps = window.HTMLWidgets.findAll(".leaflet");
    $.each(maps, function(i, el) {
      var map = this.getMap();
      map.invalidateSize();
      map.eachLayer(function(layer) {
        if (layer instanceof L.TileLayer)
          layer.redraw();
      });
    });
  }

  // trigger 'shown' so htmlwidgets resize
  $('d-article').trigger('shown');
}

function init_distill() {

  init_common();

  // create front matter
  var front_matter = $('<d-front-matter></d-front-matter>');
  $('#distill-front-matter').wrap(front_matter);

  // create d-title
  $('.d-title').changeElementType('d-title');

  // separator
  var separator = '<hr class="section-separator" style="clear: both"/>';
  // prepend separator above appendix
  $('.d-byline').before(separator);
  $('.d-article').before(separator);

  // create d-byline
  var byline = $('<d-byline></d-byline>');
  $('.d-byline').replaceWith(byline);

  // create d-article
  var article = $('<d-article></d-article>');
  $('.d-article').wrap(article).children().unwrap();

  // move posts container into article
  $('.posts-container').appendTo($('d-article'));

  // create d-appendix
  $('.d-appendix').changeElementType('d-appendix');

  // flag indicating that we have appendix items
  var appendix = $('.appendix-bottom').children('h3').length > 0;

  // replace footnotes with <d-footnote>
  $('.footnote-ref').each(function(i, val) {
    appendix = true;
    var href = $(this).attr('href');
    var id = href.replace('#', '');
    var fn = $('#' + id);
    var fn_p = $('#' + id + '>p');
    fn_p.find('.footnote-back').remove();
    var text = fn_p.html();
    var dtfn = $('<d-footnote></d-footnote>');
    dtfn.html(text);
    $(this).replaceWith(dtfn);
  });
  // remove footnotes
  $('.footnotes').remove();

  // move refs into #references-listing
  $('#references-listing').replaceWith($('#refs'));

  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    var id = $(this).attr('id');
    $('.d-contents a[href="#' + id + '"]').parent().remove();
    appendix = true;
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
  });

  // show d-appendix if we have appendix content
  $("d-appendix").css('display', appendix ? 'grid' : 'none');

  // localize layout chunks to just output
  $('.layout-chunk').each(function(i, val) {

    // capture layout
    var layout = $(this).attr('data-layout');

    // apply layout to markdown level block elements
    var elements = $(this).children().not('details, div.sourceCode, pre, script');
    elements.each(function(i, el) {
      var layout_div = $('<div class="' + layout + '"></div>');
      if (layout_div.hasClass('shaded')) {
        var shaded_content = $('<div class="shaded-content"></div>');
        $(this).wrap(shaded_content);
        $(this).parent().wrap(layout_div);
      } else {
        $(this).wrap(layout_div);
      }
    });


    // unwrap the layout-chunk div
    $(this).children().unwrap();
  });

  // remove code block used to force  highlighting css
  $('.distill-force-highlighting-css').parent().remove();

  // remove empty line numbers inserted by pandoc when using a
  // custom syntax highlighting theme, except when numbering line
  // in code chunk
  $('pre:not(.numberLines) code.sourceCode a:empty').remove();

  // load distill framework
  load_distill_framework();

  // wait for window.distillRunlevel == 4 to do post processing
  function distill_post_process() {

    if (!window.distillRunlevel || window.distillRunlevel < 4)
      return;

    // hide author/affiliations entirely if we have no authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;
    if (!have_authors)
      $('d-byline').addClass('hidden');

    // article with toc class
    $('.d-contents').parent().addClass('d-article-with-toc');

    // strip links that point to #
    $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

    // add orcid ids
    $('.authors-affiliations').find('.author').each(function(i, el) {
      var orcid_id = front_matter.authors[i].orcidID;
      var author_name = front_matter.authors[i].author
      if (orcid_id) {
        var a = $('<a></a>');
        a.attr('href', 'https://orcid.org/' + orcid_id);
        var img = $('<img></img>');
        img.addClass('orcid-id');
        img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
        img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
        a.append(img);
        $(this).append(a);
      }
    });

    // hide elements of author/affiliations grid that have no value
    function hide_byline_column(caption) {
      $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
    }

    // affiliations
    var have_affiliations = false;
    for (var i = 0; i<front_matter.authors.length; ++i) {
      var author = front_matter.authors[i];
      if (author.affiliation !== "&nbsp;") {
        have_affiliations = true;
        break;
      }
    }
    if (!have_affiliations)
      $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

    // published date
    if (!front_matter.publishedDate)
      hide_byline_column("Published");

    // document object identifier
    var doi = $('d-byline').find('h3:contains("DOI")');
    var doi_p = doi.next().empty();
    if (!front_matter.doi) {
      // if we have a citation and valid citationText then link to that
      if ($('#citation').length > 0 && front_matter.citationText) {
        doi.html('Citation');
        $('<a href="#citation"></a>')
          .text(front_matter.citationText)
          .appendTo(doi_p);
      } else {
        hide_byline_column("DOI");
      }
    } else {
      $('<a></a>')
         .attr('href', "https://doi.org/" + front_matter.doi)
         .html(front_matter.doi)
         .appendTo(doi_p);
    }

     // change plural form of authors/affiliations
    if (front_matter.authors.length === 1) {
      var grid = $('.authors-affiliations');
      grid.children('h3:contains("Authors")').text('Author');
      grid.children('h3:contains("Affiliations")').text('Affiliation');
    }

    // remove d-appendix and d-footnote-list local styles
    $('d-appendix > style:first-child').remove();
    $('d-footnote-list > style:first-child').remove();

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // hoverable references
    $('span.citation[data-cites]').each(function() {
      const citeChild = $(this).children()[0]
      // Do not process if @xyz has been used without escaping and without bibliography activated
      // https://github.com/rstudio/distill/issues/466
      if (citeChild === undefined) return true

      if (citeChild.nodeName == "D-FOOTNOTE") {
        var fn = citeChild
        $(this).html(fn.shadowRoot.querySelector("sup"))
        $(this).id = fn.id
        fn.remove()
      }
      var refs = $(this).attr('data-cites').split(" ");
      var refHtml = refs.map(function(ref) {
        // Could use CSS.escape too here, we insure backward compatibility in navigator
        return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
      }).join("\n");
      window.tippy(this, {
        allowHTML: true,
        content: refHtml,
        maxWidth: 500,
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start'
      });
    });

    // fix footnotes in tables (#411)
    // replacing broken distill.pub feature
    $('table d-footnote').each(function() {
      // we replace internal showAtNode methode which is triggered when hovering a footnote
      this.hoverBox.showAtNode = function(node) {
        // ported from https://github.com/distillpub/template/pull/105/files
        calcOffset = function(elem) {
            let x = elem.offsetLeft;
            let y = elem.offsetTop;
            // Traverse upwards until an `absolute` element is found or `elem`
            // becomes null.
            while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                x += elem.offsetLeft;
                y += elem.offsetTop;
            }

            return { left: x, top: y };
        }
        // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
        const bbox = node.getBoundingClientRect();
        const offset = calcOffset(node);
        this.show([offset.left + bbox.width, offset.top + bbox.height]);
      }
    })

    // clear polling timer
    clearInterval(tid);

    // show body now that everything is ready
    on_load_complete();
  }

  var tid = setInterval(distill_post_process, 50);
  distill_post_process();

}

function init_downlevel() {

  init_common();

   // insert hr after d-title
  $('.d-title').after($('<hr class="section-separator"/>'));

  // check if we have authors
  var front_matter = JSON.parse($("#distill-front-matter").html());
  var have_authors = front_matter.authors && front_matter.authors.length > 0;

  // manage byline/border
  if (!have_authors)
    $('.d-byline').remove();
  $('.d-byline').after($('<hr class="section-separator"/>'));
  $('.d-byline a').remove();

  // remove toc
  $('.d-contents').remove();

  // move appendix elements
  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
  });


  // inject headers into references and footnotes
  var refs_header = $('<h3></h3>');
  refs_header.text('References');
  $('#refs').prepend(refs_header);

  var footnotes_header = $('<h3></h3');
  footnotes_header.text('Footnotes');
  $('.footnotes').children('hr').first().replaceWith(footnotes_header);

  // move appendix-bottom entries to the bottom
  $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
  $('.appendix-bottom').remove();

  // remove appendix if it's empty
  if ($('.d-appendix').children().length === 0)
    $('.d-appendix').remove();

  // prepend separator above appendix
  $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

  // trim code
  $('pre>code').each(function(i, val) {
    $(this).html($.trim($(this).html()));
  });

  // move posts-container right before article
  $('.posts-container').insertBefore($('.d-article'));

  $('body').addClass('downlevel');

  on_load_complete();
}


function init_common() {

  // jquery plugin to change element types
  (function($) {
    $.fn.changeElementType = function(newType) {
      var attrs = {};

      $.each(this[0].attributes, function(idx, attr) {
        attrs[attr.nodeName] = attr.nodeValue;
      });

      this.replaceWith(function() {
        return $("<" + newType + "/>", attrs).append($(this).contents());
      });
    };
  })(jQuery);

  // prevent underline for linked images
  $('a > img').parent().css({'border-bottom' : 'none'});

  // mark non-body figures created by knitr chunks as 100% width
  $('.layout-chunk').each(function(i, val) {
    var figures = $(this).find('img, .html-widget');
    // ignore leaflet img layers (#106)
    figures = figures.filter(':not(img[class*="leaflet"])')
    if ($(this).attr('data-layout') !== "l-body") {
      figures.css('width', '100%');
    } else {
      figures.css('max-width', '100%');
      figures.filter("[width]").each(function(i, val) {
        var fig = $(this);
        fig.css('width', fig.attr('width') + 'px');
      });

    }
  });

  // auto-append index.html to post-preview links in file: protocol
  // and in rstudio ide preview
  $('.post-preview').each(function(i, val) {
    if (window.location.protocol === "file:")
      $(this).attr('href', $(this).attr('href') + "index.html");
  });

  // get rid of index.html references in header
  if (window.location.protocol !== "file:") {
    $('.distill-site-header a[href]').each(function(i,val) {
      $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
    });
  }

  // add class to pandoc style tables
  $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
  $('.kable-table').children('table').addClass('pandoc-table');

  // add figcaption style to table captions
  $('caption').parent('table').addClass("figcaption");

  // initialize posts list
  if (window.init_posts_list)
    window.init_posts_list();

  // implmement disqus comment link
  $('.disqus-comment-count').click(function() {
    window.headroom_prevent_pin = true;
    $('#disqus_thread').toggleClass('hidden');
    if (!$('#disqus_thread').hasClass('hidden')) {
      var offset = $(this).offset();
      $(window).resize();
      $('html, body').animate({
        scrollTop: offset.top - 35
      });
    }
  });
}

document.addEventListener('DOMContentLoaded', function() {
  if (is_downlevel_browser())
    init_downlevel();
  else
    window.addEventListener('WebComponentsReady', init_distill);
});

</script>

<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.5);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

.distill-site-footer {
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */
</style>
<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.7);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #ffffff;
}

.distill-site-footer {
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #fafafa;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */

.nav-right > a {
  text-transform: uppercase;
}

d-title h1, d-title p, d-title figure,
d-abstract p, d-abstract b {
  grid-column: page;
}

.rj-blue {
  color: #2467bb;
}

ul li {
  line-height: 1.6;
  margin-top: 0em;
  margin-bottom: 0em;
}</style>
<style type="text/css">
/* base style */

/* FONT FAMILIES */

:root {
  --heading-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  --mono-default: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  --body-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

body,
.posts-list .post-preview p,
.posts-list .description p {
  font-family: var(--body-font), var(--body-default);
}

h1, h2, h3, h4, h5, h6,
.posts-list .post-preview h2,
.posts-list .description h2 {
  font-family: var(--heading-font), var(--heading-default);
}

d-article div.sourceCode code,
d-article pre code {
  font-family: var(--mono-font), var(--mono-default);
}


/*-- TITLE --*/
d-title h1,
.posts-list > h1 {
  color: var(--title-color, black);
}

d-title h1 {
  font-size: var(--title-size, 50px);
}

/*-- HEADERS --*/
d-article h1,
d-article h2,
d-article h3,
d-article h4,
d-article h5,
d-article h6 {
  color: var(--header-color, rgba(0, 0, 0, 0.8));
}

/*-- BODY --*/
d-article > p,  /* only text inside of <p> tags */
d-article > ul, /* lists */
d-article > ol {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
  font-size: var(--body-size, 1.06rem);
}


/*-- CODE --*/
d-article div.sourceCode code,
d-article pre code {
  font-size: var(--code-size, 14px);
}

/*-- ASIDE --*/
d-article aside {
  font-size: var(--aside-size, 12px);
  color: var(--aside-color, rgba(0, 0, 0, 0.6));
}

/*-- FIGURE CAPTIONS --*/
figure .caption,
figure figcaption,
.figure .caption {
  font-size: var(--fig-cap-size, 13px);
  color: var(--fig-cap-color, rgba(0, 0, 0, 0.6));
}

/*-- METADATA --*/
d-byline h3 {
  font-size: var(--heading-size, 0.6rem);
  color: var(--heading-color, rgba(0, 0, 0, 0.5));
}

d-byline {
  font-size: var(--body-size, 0.8rem);
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

d-byline a,
d-article d-byline a {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

/*-- TABLE OF CONTENTS --*/
.d-contents nav h3 {
  font-size: var(--heading-size, 18px);
}

.d-contents nav a {
  font-size: var(--contents-size, 13px);
}

/*-- APPENDIX --*/
d-appendix h3 {
  font-size: var(--heading-size, 15px);
  color: var(--heading-color, rgba(0, 0, 0, 0.65));
}

d-appendix {
  font-size: var(--text-size, 0.8em);
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

d-appendix d-footnote-list a.footnote-backlink {
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

/*-- WEBSITE HEADER + FOOTER --*/
.distill-site-header .title {
  font-size: var(--title-size, 18px);
  font-family: var(--navbar-font), var(--heading-default);
}

.distill-site-header a,
.nav-dropdown .nav-dropbtn {
  font-family: var(--navbar-font), var(--heading-default);
}

.nav-dropdown .nav-dropbtn {
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  font-size: var(--text-size, 15px);
}

.distill-site-header a:hover,
.nav-dropdown:hover .nav-dropbtn {
  color: var(--hover-color, white);
}

.distill-site-header {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer a:hover {
  color: var(--hover-color, white);
}</style>
<!--/radix_placeholder_distill-->
  <script src="../../site_libs/header-attrs-2.24/header-attrs.js"></script>
  <script src="../../site_libs/htmlwidgets-1.6.2/htmlwidgets.js"></script>
  <script src="../../site_libs/plotly-binding-4.10.2/plotly.js"></script>
  <script src="../../site_libs/typedarray-0.1/typedarray.min.js"></script>
  <script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <link href="../../site_libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
  <script src="../../site_libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>
  <link href="../../site_libs/plotly-htmlwidgets-css-2.11.1/plotly-htmlwidgets.css" rel="stylesheet" />
  <script src="../../site_libs/plotly-main-2.11.1/plotly-latest.min.js"></script>
  <script src="../../site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="../../site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="../../site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="../../site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="../../site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="../../site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="../../site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="../../site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
<!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      buttons.append('<a href="https://twitter.com/intent/tweet?text='+title+'&url=' + cite_url + '" class="dt-tag"><i class="fab fa-twitter"></i> Tweet</a>')

      const abstract = $('<d-abstract>')
      abstract.append('<b>Abstract:</b><br>')
      abstract.append($("d-title p:not(:empty)").first()) // Move description to d-abstract
      $("d-title p:empty").remove() // Remove empty paragraphs after title
      abstract.append(buttons)
      abstract.insertAfter($('d-title')) // Add abstract section after title

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Likelihood Ratio Test-Based Drug Safety Assessment using R Package pvLRT","description":"Medical product safety continues to be a key concern of the twenty-first century. Several spontaneous adverse events reporting databases established across the world continuously collect and archive adverse events data on various medical products. Determining signals of disproportional reporting (SDR) of product/adverse event pairs from these large-scale databases require the use of principled statistical techniques. Likelihood ratio test (LRT)-based approaches are particularly noteworthy in this context as they permit objective SDR detection without requiring ad hoc thresholds. However, their implementation is non-trivial due to analytical complexities, which necessitate the use of computation-heavy methods. Here we introduce R package pvLRT which implements a suite of LRT approaches, along with various post-processing and graphical summary functions, to facilitate simplified use of the methodologies. Detailed examples are provided to illustrate the package through analyses of three real product safety datasets obtained from publicly available FDA FAERS and VAERS databases.","doi":"10.32614/RJ-2023-027","authors":[{"author":"Saptarshi Chakraborty","authorURL":"#","affiliation":"Department of Biostatistics, SUNY University at Buffalo","affiliationURL":"#","orcidID":"0000-0002-3121-9174"},{"author":"Marianthi Markatou","authorURL":"#","affiliation":"Department of Biostatistics, SUNY University at Buffalo","affiliationURL":"#","orcidID":"0000-0002-1453-8229"},{"author":"Robert Ball","authorURL":"#","affiliation":"Center for Drug Evaluation and Research, U.S. Food and Drug Administration","affiliationURL":"#","orcidID":"0000-0002-1609-7420"}],"publishedDate":"2023-08-26T00:00:00.000+00:00","citationText":"Chakraborty, et al., 2023"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a class="logo" href="../../index.html">
<img src="../../resources/rlogo.png" alt="Logo"/>
</a>
<a href="../../index.html" class="title">The R Journal</a>
</div>
<div class="nav-right">
<a href="../../index.html">Home</a>
<a href="../../issues/2024-2">Current</a>
<a href="../../issues.html">Issues</a>
<a href="../../news.html">News</a>
<a href="../../submissions.html">Submit</a>
<a href="../../editors.html">Editorial board</a>
<a href="../../articles.xml">
<i class="fa fa-rss" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Likelihood Ratio Test-Based Drug Safety Assessment using R Package pvLRT</h1>
<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>Medical product safety continues to be a key concern of the twenty-first century. Several spontaneous adverse events reporting databases established across the world continuously collect and archive adverse events data on various medical products. Determining signals of disproportional reporting (SDR) of product/adverse event pairs from these large-scale databases require the use of principled statistical techniques. Likelihood ratio test (LRT)-based approaches are particularly noteworthy in this context as they permit objective SDR detection without requiring ad hoc thresholds. However, their implementation is non-trivial due to analytical complexities, which necessitate the use of computation-heavy methods. Here we introduce R package pvLRT which implements a suite of LRT approaches, along with various post-processing and graphical summary functions, to facilitate simplified use of the methodologies. Detailed examples are provided to illustrate the package through analyses of three real product safety datasets obtained from publicly available FDA FAERS and VAERS databases.</p></p>
</div>

<div class="d-byline">
  Saptarshi Chakraborty  (Department of Biostatistics, SUNY University at Buffalo)
  
,   Marianthi Markatou  (Department of Biostatistics, SUNY University at Buffalo)
  
,   Robert Ball  (Center for Drug Evaluation and Research, U.S. Food and Drug Administration)
  
<br/>2023-08-26
</div>

<div class="d-article">
<h1 data-number="1" id="introduction"><span class="header-section-number">1</span> Introduction</h1>
<p>Adverse events of medical products are a major concern worldwide, with serious global health implications particularly in the post-pandemic world. While clinical trials conducted pre-licensure remain the primary source of safety information about medical products, and other data sources such as claims and electronic health records are increasingly being used post-licensure, they exhibit several limitations. First, detecting extremely rare adverse events in clinical trials is difficult due to their strict inclusion/exclusion criteria and their relatively small counts compared to the population which will eventually receive the product. Second, randomized clinical trials for detecting multiple adverse events may be logistically and ethically infeasible. Despite having known issues such as self-selection, confounding, and missing data, etc., large-scale observational studies have been used in surveillance and post-licensure epidemiological studies to supplement traditional clinical trial-based approaches <span class="citation" data-cites="markatou2014">(<a href="#ref-markatou2014" role="doc-biblioref">Markatou and Ball 2014</a>)</span>. Thanks to the proliferation of electronic health record systems worldwide, a vast trove of such data are now available and can collectively provide critical first alerts for emerging medical product safety concerns. Curated large-scale databases such as US Food and Drug Administration (FDA)’s Adverse Event Reporting System (FAERS) and Vaccine Adverse Event Reporting System (VAERS), the European Medicines Agency’s Eudravigilance, and the World Health Organization’s VigiBase provide collective lists of drugs/vaccines and adverse events from large number of reports, and thus are invaluable resources for product safety assessment.</p>
<p>Mining these databases to assess medical product safety is however challenging due to limitations of the data (e.g., biases in reporting, presence of duplicate reports, and missing/incomplete information) and the lack of a universally accepted statistical methodology. Several approaches to detecting signals of disproportionate reporting (SDR) as an indicator of a possible safety concern from adverse event data exist; these constitute a part of the pharmacovigilance process for assessing whether a medical product is the cause of an adverse event. Of the existing approaches, Proportional Relative Reporting Proportion (PRR), Reporting Odds Ratio (ROR), Multi-item Gamma Poisson Shrinker (MGPS), Bayesian Confidence Propagation Neural Network (BCPNN) and their false discovery rate (FDR)-adjusted variants are notable. Many of these approaches require one or both types of thresholding: (a) on the number of reports needed for the method to operate, and (b) on the underlying statistic to identify important/severe adverse events. However, determining appropriate thresholds remains a key challenge for these approaches, and in practice ad hoc thresholds are often used. This is problematic because both too low and too high thresholds lead to drastically different detection rates, and no separate measure is usually available to assess correctness of the detected signals. A principled approach permitting data-driven threshold determination with optimal guarantees on the detected signals is therefore essential to ensure statistical validity of results.</p>
<p>The LRT-based approaches to signal detection in the medical product safety datasets <span class="citation" data-cites="ding2020 huangLRT2011 huang2017 zhao2018 chakraborty2022">(<a href="#ref-huangLRT2011" role="doc-biblioref">Huang et al. 2011</a>, <a href="#ref-huang2017" role="doc-biblioref">2017</a>; <a href="#ref-zhao2018" role="doc-biblioref">Zhao et al. 2018</a>; <a href="#ref-ding2020" role="doc-biblioref">Ding et al. 2020</a>; <a href="#ref-chakraborty2022" role="doc-biblioref">Chakraborty et al. 2022</a>)</span> provide a highly rigorous suite of statistical methods to address these problems. To provide a high-level summary, these methods assume a Poisson/zero-inflated Poisson model to parametrize associations between drugs and their adverse events. Then conditional on the total reported number of cases for each adverse event and each medical product, these methods quantify signal strength in the observed number of reports per adverse event/medical product pair through a (likelihood ratio) test statistic. Finally, significance of the pair is determined by comparing the observed test statistic to its null (independence) sampling distribution. There are several key advantages of these formal hypothesis test-based approaches which make them ideal for practical use. First, they aid coherent frequetist sampling distribution-based quantification of signal strengths, ensuring strong statistical and probabilistic rigor of the results. Second, they do not require the use of ad hoc thresholds for the observed report counts; instead, they quantify significance through probabilistic measures of uncertainty, e.g., <span class="math inline">\(p\)</span>-values. This ensures that the identified signals are valid up to a pre-specified level of tolerance (usually 0.05). Third, because they are based on the highly formal likelihood ratio test theory for null hypothesis significance testing, they can rigorously control the type I error and the false discovery rate, while ensuring high power and sensitivity for signal detection. Moreover, by appropriately defining a maximum likelihood ratio test statistic, the method achieves automated FDR adjustment without requiring any separate <span class="math inline">\(p\)</span>-value adjustment step.</p>
<div class="layout-chunk" data-layout="l-body">
<table>
<caption>
<span id="tab:table-summary-existing-packages">Table 1: </span>Existing R packages on CRAN with functionalities for pharmacovigilance.
</caption>
<thead>
<tr>
<th style="text-align:left;">
Package
</th>
<th style="text-align:left;">
Method
</th>
<th style="text-align:left;">
Notes
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
<strong>PhViD</strong>: Pharmacovigilance Signal Detection
</td>
<td style="text-align:left;">
PRR, ROR, BCPNN, GPS
</td>
<td style="text-align:left;">
Aimed towards drug safety.
</td>
</tr>
<tr>
<td style="text-align:left;">
<strong>openEBGM</strong>: EBGM Disproportionality Scores for Adverse Event Data Mining
</td>
<td style="text-align:left;">
MGPS
</td>
<td style="text-align:left;">
Aimed towards drug safety.
</td>
</tr>
<tr>
<td style="text-align:left;">
<strong>sglr</strong>: An R package for power and boundary calculations in pre-licensure vaccine trials using a sequential generalized likelihood ratio test
</td>
<td style="text-align:left;">
Sequential Generalized Likelihood Ratio decision boundaries
</td>
<td style="text-align:left;">
Aimed towards sequential testing-based vaccine safety assessments as used by the FDA.
</td>
</tr>
<tr>
<td style="text-align:left;">
<strong>Sequential</strong>: Exact sequential analysis for Binomial and Poisson data
</td>
<td style="text-align:left;">
Max SPRT statistic
</td>
<td style="text-align:left;">
Aimed towards sequential testing-based vaccine safety assessments as used by the FDA.
</td>
</tr>
<tr>
<td style="text-align:left;">
<strong>AEenrich</strong>: Vaccine adverse event enrichment tests
</td>
<td style="text-align:left;">
<ol type="a">
<li>Modified Fisher’s exact test (for pre-selected significant AEs); b) Modified Kolmogorov Smirnov statistic
</td>
<td style="text-align:left;">
Aimed towards sequential testing-based vaccine safety assessments as used by the FDA.
</td>
</tr>
<tr>
<td style="text-align:left;">
<strong>mds</strong>: Medical Devices Surveillance
</td>
<td style="text-align:left;">
Data preprocessing
</td>
<td style="text-align:left;">
Provides functions for handling messy/unstructured medical devices data.
</td>
</tr>
<tr>
<td style="text-align:left;">
<strong>pvLRT</strong>: A suite of likelihood ratio test based methods to use in pharmacovigilance
</td>
<td style="text-align:left;">
(Pseudo) LRT approaches based on log-linear models
</td>
<td style="text-align:left;">
Our package.
</td>
</tr>
</tbody>
</table></li>
</ol>
</div>
<p>There currently exist a few R packages in CRAN with functionalities relevant for medical product safety. This includes the packages <a href="https://cran.r-project.org/package=PhViD">PhViD</a> <span class="citation" data-cites="pkg_phvid">(<a href="#ref-pkg_phvid" role="doc-biblioref">Ahmed and Poncet 2016</a>)</span>, <a href="https://cran.r-project.org/package=openEBGM">openEBGM</a> <span class="citation" data-cites="canida2017">(<a href="#ref-canida2017" role="doc-biblioref">Canida and Ihrie 2017</a>)</span>, <a href="https://cran.r-project.org/package=AEenrich">AEenrich</a> <span class="citation" data-cites="pkg_aeenrich">(<a href="#ref-pkg_aeenrich" role="doc-biblioref">Li et al. 2021</a>)</span>, <a href="https://cran.r-project.org/package=Sequential">Sequential</a> <span class="citation" data-cites="pkg_sequential">(<a href="#ref-pkg_sequential" role="doc-biblioref">Silva and Kulldorff 2021</a>)</span>, <a href="https://cran.r-project.org/package=SPRT">SPRT</a> <a href="https://cran.r-project.org/package=sglr">sglr</a> <span class="citation" data-cites="pkg_sglr">(<a href="#ref-pkg_sglr" role="doc-biblioref">Narasimhan and Shih 2012</a>)</span>, and <a href="https://cran.r-project.org/package=mds">mds</a> <span class="citation" data-cites="pkg_mds">(<a href="#ref-pkg_mds" role="doc-biblioref">Chung 2020</a>)</span>. Among these <a href="https://cran.r-project.org/package=PhViD">PhViD</a> and <a href="https://cran.r-project.org/package=openEBGM">openEBGM</a> provide functionalities for spontaneous adverse event data-driven pharmacovigilance: <a href="https://cran.r-project.org/package=PhViD">PhViD</a> implements methods such as PRR, ROR, and BCPNN, and <a href="https://cran.r-project.org/package=openEBGM">openEBGM</a> implements the method of MGPS. By contrast, the other packages provide sequential testing-based approaches to vaccine safety. Table 1 lists these packages and their functionalities together with some high-level notes on their targeted uses. We note however that none of these packages implement LRT-based approaches to drug safety based on spontaneous adverse event data as considered in <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a>. Indeed the development of <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> was motivated by the lack of easily accessible and comprehensive open source computational solutions to the LRT-based pharmacovigilance approaches. An important common ingredient of these LRT-based methods is a computation-intensive Monte Carlo simulation step required to facilitate (<em>exact/non-asymptotic</em>) inference from the analytically intractable <em>null</em> sampling distributions of the relevant test statistics. However, this requirement precludes immediate applications of the methods by pharmacovigilance practitioners in the absence of easily accessible software implementations. This is particularly relevant for zero-inflation based models where, in addition to the computation-heavy Monte Carlo step additional numerical optimization steps are necessary for the estimation of the zero-inflation parameter. We have developed <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> <span class="citation" data-cites="pkg_pvLRT">(<a href="#ref-pkg_pvLRT" role="doc-biblioref">Chakraborty and Markatou 2022</a>)</span> to cater to these needs while ensuring proper statistical rigor in the implementations.</p>
<p>The following are the major contributions of the <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> package. First, the package serves as a comprehensive software implementation of several LRT-based methods proposed over the past decade, including some recently developed methods. Both Poisson and Zero-inflated Poisson (ZIP) based models are implemented, and tests of both single and multiple simultaneous drugs (or medical products) are provided. Formal model comparison methods, such as Akaike Information Criterion (AIC) and Bayesian Information Criterion (BIC), can be used on the model fits to aid data-driven selection of Poisson vs. ZIP models on individual datasets. Second, three visual summary plots for the test results , namely bubble plot, barplot, and heatmap, are provided for immediate visualization. These plots provide information on sample size and test statistic and p-values for all AE/drug pairs tested, and hence permit quick exploration of signals in a (possibly large) dataset. These visualizations use <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> <span class="citation" data-cites="pkg_ggplot2">(<a href="#ref-pkg_ggplot2" role="doc-biblioref">Wickham 2016</a>)</span> as the plotting engine, and hence its attributes (color, text size, etc.) can be easily modified post hoc by changing some of the <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> graphical sttributes. In addition, leveraging the modern R ecosystem, the resultant plots can be easily made interactive via additional functions from other packages, such as <code>ggplotly()</code> from the package <a href="https://cran.r-project.org/package=plotly">plotly</a> <span class="citation" data-cites="pkg_plotly">(<a href="#ref-pkg_plotly" role="doc-biblioref">Sievert 2020</a>)</span>. Third, aside from implementations of likelihood ratio tests, simple functions for random contingency table data generation from the Poisson and zero-inflated Poisson models are provided. This is particularly helpful in simulation experiments where different statistical methods for pharmacovigilance are compared. Finally, we have included processed contingency tables of AE/drug reports for two specific drug groups, viz. statins and Gadolinium-based Contrast Agents (GBCA; <span class="citation" data-cites="zhao2018">Zhao et al. (<a href="#ref-zhao2018" role="doc-biblioref">2018</a>)</span>), raw AE/drug incidence data, and a vaccine/AE dataset on rotavirus vaccines as R datasets. The drug data were obtained from the publicly available FDA FAERS database for the quarters 2014 Q1 – 2020 Q3 (for the processed contingency tables) and 2022 Q3 (raw incidence data), and the vaccine data were obtained from the FDA VAERS database for the year 1999; the package provides a convenient approach to accessing these processed data, particularly beneficial in methodological studies.</p>
<p>We note here that <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> is primarily an analysis package providing functions implementing statistical methodologies and subsequent post-processing and visualizations of results. The input data for the main analysis functions in <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> are always assumed to be pre-processed contingency tables (matrix-like objects) enumerating adverse event report counts with AEs along the rows and Drugs (or other medical products) along the columns. The package provides a convenience function to convert raw AE/drug incidence data (e.g., those downloaded from the FDA FAERS database) into analysis-ready processed contingency tables; however, no other functions for input raw data exploration and pre-processing, e.g. for filtering specific AEs or Drugs of interest or grouping specific AEs or drugs together are provided. We have deliberately made this design choice to (a) make the scope of the package well-defined, and (b) encourage the user to explore the raw adverse event data well before analyzing them. The modern R ecosystem contains several excellent general-purpose packages/package-collections, including <a href="https://cran.r-project.org/package=data.table">data.table</a> <span class="citation" data-cites="pkg_datatable">(<a href="#ref-pkg_datatable" role="doc-biblioref">Dowle and Srinivasan 2021</a>)</span> and <a href="https://cran.r-project.org/package=tidyverse">tidyverse</a> <span class="citation" data-cites="pkg_tidyverse">(<a href="#ref-pkg_tidyverse" role="doc-biblioref">Wickham et al. 2019</a>)</span> that provide a suite of principled and easy to use data pre-processing, munging, and visualization functions. We highly recommend leveraging these packages/functions to understand and preprocess the data well, before likelihood ratio tests are performed using <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a>.</p>
<p>The purpose of the current article is to provide a high-level overview of the <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> package with detailed notes on its use, and guided examples with real world adverse event data exemplifying the use of LRT-based methodologies for pharmacovigilance. We note however that this article is not meant to serve as a comprehensive manual for <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a>; all functions/objects provided in <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> come with detailed documentation, and this documentation serves as the definitive resource for <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a>. It is also important to note that the LRT-based disproportionality analysis exemplified herein is only one part of the collective medical product/adverse event relationship assessment process. Current pharmacovigilance practice for medical product safety assessment includes review of summary statistics, disproportionality scores, cases summaries, and other sources of information about the medical products and the adverse events. The remainder of the article is organized as follows. We begin with a brief review of the LRT-based approaches to pharmacovigilance implemented in <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a>. We then exemplify <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> by analyzing three sets of real pharmacovigilance datasets, two concerning drug safety and one concerning vaccine safety. We conclude the article with a brief discussion and some potential future directions.</p>
<h1 data-number="2" id="a-brief-review-of-the-likelihood-ratio-test-based-approaches-to-pharmacovigilance"><span class="header-section-number">2</span> A brief review of the Likelihood Ratio test-based approaches to pharmacovigilance</h1>
<p>This section reviews the underlying theories behind the likelihood ratio test (LRT)-based methods for drug safety assessment. We begin by fixing our notation. Consider a drug safety database cataloging <span class="math inline">\(I\)</span> AEs detected among <span class="math inline">\(J\)</span> medical products. Hereinafter we describe the methodologies in terms of ‘drugs’ as the medical products of concern; however, the methodologies trivially extend to other medical products including vaccines and medical devices. Let <span class="math inline">\(n_{ij}\)</span> denote the number of reported cases for the <span class="math inline">\(i\)</span>-th AE in the <span class="math inline">\(j\)</span>-th drug. For the <span class="math inline">\(I \times J\)</span> contingency table <span class="math inline">\((n_{ij})\)</span>, let <span class="math inline">\(n_{i\bullet} = \sum_{j=1}^{J} n_{ij}\)</span> and <span class="math inline">\(n_{\bullet j} = \sum_{i=1}^{I} n_{ij}\)</span> denote the <span class="math inline">\(i\)</span>-th row total and <span class="math inline">\(j\)</span>-th column total respectively, <span class="math inline">\(i = 1, \dots, I\)</span>; <span class="math inline">\(j = 1, \dots, J\)</span>; and let <span class="math inline">\(n_{\bullet \bullet}\)</span> denote the grand total. Interest lies in determining whether the observed report count <span class="math inline">\(n_{ij}\)</span> for <span class="math inline">\((i, j)\)</span>-th AE/drug pair is substantially larger than what is expected had there been no association between the <span class="math inline">\(i\)</span>-th AE and the <span class="math inline">\(j\)</span>-th drug. Within the null hypothesis significance testing framework this is achieved by testing a hypothesis of absence of a ``signal” against that of its presence at the corresponding AE/drug pair. We formally define these hypotheses and the underlying parametric models below.</p>
<h4 class="unnumbered" id="model-and-parametrization-for-lrt-based-pharmacovigilance.">Model and Parametrization for LRT-based pharmacovigilance.</h4>
<p>A Poisson log-linear model for LRT-based SDR detection has been proposed in the literature to parametrize AE/drug associations based on the observed report counts <span class="math inline">\(\{n_{ij}\}\)</span>. At the outset we note that these AE/drug associations can also be studied through a series of logistic regression models each with binary occurrences of each individual AE as the response and the presence/absence of various drugs (or other medical products) as predictors. When a Poisson model is assumed for the report counts <span class="math inline">\(\{n_{ij}\}\)</span>, the resulting log-linear model and the collective logistic regression models lead to identical maximum likelihood-based inferences <span class="citation" data-cites="agresti2013">(<a href="#ref-agresti2013" role="doc-biblioref">Agresti 2013</a>)</span>. However, the log linear model is more flexible than the logistic models in that, by relaxing the underlying Poisson assumption for the report counts, it can handle a richer set of data. Two different parametrizations, namely the reporting proportion <span class="citation" data-cites="huangLRT2011">(<a href="#ref-huangLRT2011" role="doc-biblioref">Huang et al. 2011</a>)</span> and the relative reporting proportion <span class="citation" data-cites="dumouchel1999">(<a href="#ref-dumouchel1999" role="doc-biblioref">Dumouchel 1999</a>)</span>, have been considered within the log-linear modeling framework for pharmacovigilance. These parametrizations differ by how they define and handle “signals”. In the reporting proportion parametrization a signal is determined at a specific AE for a given drug if its reporting proportion is substantially larger than the overall reporting proportions of all other AEs combined. In contrast, in the relative reporting proportion parametrization one focuses on the relative reporting proportion of a specific AE for a given drug and tests whether it is substantially larger than 1. The latter parametrization does not require consideration of the category of “all other AEs combined” explicitly in each comparison, and hence has a simpler interpretation. In applications the two parametrizations incur similar computational expenses, and produce virtually identical results when the true data generating process is Poisson. However, the results may vary substantially when there is a model misspecification, including the presence of zero-inflation, i.e., excess zero-report counts that are inherently different from the <em>sampling</em> zero counts governed by the Poisson law and are produced by an exogenous process. A note on different types of zero-inflation possibly occurring in adverse event datasets is provided in the next paragraph. Our recent methodological study <span class="citation" data-cites="chakraborty2022">(<a href="#ref-chakraborty2022" role="doc-biblioref">Chakraborty et al. 2022</a>)</span> has shown that the relative reporting proportion-based Poisson LRT is more robust against zero-inflated data than its reporting proportion based counterpart. Extensions of the Poisson LRTs have been proposed to explicitly handle excess zeros under a zero-inflated Poisson (ZIP) model. The relative reporting proportion parametrization, within a pseudo likelihood ratio testing framework, aids a straightforward extension to the ZIP model that adds only a small overhead (one single additional optimization of a smooth univariate function) to the overall computational burden, and produces a test statistic whose functional form is identical to the ordinary (non-zero inflated) Poisson LRT. This computational simplicity makes the approach highly scalable for large data sizes, similar to the Poisson LRT. Collectively, these simplifications aid the construction of a unified LRT framework based on the relative reporting proportion-based parametrization that can simultaneously handle a Poisson model and a ZIP model, with near identical computational burden. By contrast, the reporting proportion-based LRT under a ZIP model suffers from a substantially increased computation burden <span class="citation" data-cites="huang2017">(<a href="#ref-huang2017" role="doc-biblioref">Huang et al. 2017</a>)</span> which hinders its scalability. For these reasons, both in this article and in the <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> package we primarily focus on the relative reporting proportion parametrization; we do note however that the <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> package does provide an implementation of the reporting proportion parametrization under an ordinary Poisson model setup.</p>
<h4 class="unnumbered" id="a-brief-note-on-excess-zeros-or-zero-inflation-in-medical-product-safety-data.">A brief note on excess zeros or <em>zero-inflation</em> in medical product safety data.</h4>
<p>We now briefly note how zero-inflation occurs in pharmacovigilance. In adverse event reporting datasets, an important source of excess zero report counts is <em>structural zeros</em> associated with AE/drug pairs that are physically impossible to occur. If structural zeros are present, then the corresponding cells in a contingency table enumerating the number of reported cases of various AE/drug pairs will always be zero, regardless of the total number of reported counts. Moreover, inference on these structural zero positions can be made from the observed data: intuitively, a cell with a positive report count in any observed table cannot be a structural zero, and only an observed zero cell can be a structural zero. This is in contrast with another source of zero-inflation, namely data corruption, that may also occur in adverse events data. Here, due to noise in data recording process some report counts in the AE/drug contingency tables are randomly recorded as zero. Under this setting, there is no population level true zero positions, and the zero inflated positions vary from one sample to another. In this article we primarily focus on the structural zero-type zero-inflation, as it is the primary type of zero-inflation that the FAERS datasets stored in <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> feature. Indeed, the FAERS data undergo several rounds of rigorous reviews and checks to safeguard against data corruption. However, we note that <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> does provide functions for handling corruption-type zero-inflation; interested readers are referred to the the documentation of the <code>pvlrt()</code> function.</p>
<h4 class="unnumbered" id="parametric-hypothesis-testing-for-signals-in-pharmacovigilance-based-on-a-zero-inflated-poisson-zip-model.">Parametric hypothesis testing for signals in pharmacovigilance based on a Zero-inflated Poisson (ZIP) model.</h4>
<p>To facilitate a unified treatment we consider a zero-inflated Poisson model to describe the test procedure. A discrete random variable <span class="math inline">\(X\)</span> is said to follow a ZIP<span class="math inline">\((\theta, \omega)\)</span> distribution where <span class="math inline">\(\omega \in [0, 1)\)</span> and <span class="math inline">\(\theta &gt; 0\)</span> if the probability mass function of <span class="math inline">\(X\)</span> is given by <span class="math display">\[
\operatorname{P}(X = x) = \begin{cases}
\omega +  (1 - \omega) \exp(-\theta), &amp; x = 0 \\
(1 - \omega) \exp(-\theta) \frac{\theta^x}{x!}, &amp; x = 1, 2, 3, \dots
\end{cases},
\]</span> i.e., if the distribution of <span class="math inline">\(X\)</span> is a <span class="math inline">\((1-\omega)\)</span> and <span class="math inline">\(\omega\)</span> mixture of a Poisson(<span class="math inline">\(\theta\)</span>) distribution and a point mass at zero. The model reduces to an ordinary Poisson model when the zero-inflation probability <span class="math inline">\(\omega = 0\)</span>. For adverse event data report counts we consider the following zero-inflated Poisson model for the observed count <span class="math inline">\(n_{i j}\)</span> for the <span class="math inline">\(i\)</span>-th AE of a specific drug <span class="math inline">\(j\)</span>: <span class="math inline">\(n_{i j} \sim \text{ZIP}(\lambda_{i j} \times E_{i j}, \omega_{j})\)</span>, where <span class="math inline">\(E_{i j} = n_{i \bullet} n_{\bullet j}/n_{\bullet \bullet }\)</span> is the expected number of reports for the <span class="math inline">\((i, j)\)</span>-th pair when there is no association between the pair, and <span class="math inline">\(\lambda_{i j}\)</span> is the relative reporting proportion. Under this set up, null hypothesis significance tests are performed to determine signals at specific pairs. Suppose we are interested in determining significant adverse events among <span class="math inline">\(K\)</span> out of the <span class="math inline">\(J\)</span> drugs, labeled <span class="math inline">\(1, \dots, K\)</span>. The global null hypothesis is <span class="math inline">\(H_0^{1:K}: \{\lambda_{i j} = 1\)</span> for all <span class="math inline">\(i = 1, \dots, I\)</span> and <span class="math inline">\(j = 1, \dots, J\)</span>} and is tested against the <em>global alternative</em> <span class="math inline">\(H_a^{1:K} = \{ \lambda_{i j} &gt; 1\)</span> for at least one <span class="math inline">\((i, j): i = 1, \dots, I; j = 1, \dots, K\}\)</span> which represents the hypothesis of ``at least one signal’’. If the global null hypothesis is rejected in favor of the global alternative, our focus then shifts on identifying the individual AE/drug pairs with strong associations. This is achieved by post hoc test of the global null hypothesis against the individual alternative hypotheses <span class="math inline">\(H_{a, i j}: \lambda_{i j} &gt; 1\)</span>.</p>
<h4 class="unnumbered" id="pseudo-likelihood-ratio-test.">Pseudo Likelihood Ratio Test.</h4>
<p>A pseudo likelihood ratio test of the above hypotheses involves computation of the individual pseudo likelihood ratios <span class="math display">\[\begin{equation*}
{\text{LR}}_{i j} =
\begin{cases}
1 &amp; n_{i j} = 0 \\
\exp(-(\hat \lambda_{i j}-1) E_{i j}) \hat \lambda_{i j}^{n_{i j}} &amp; n_{i j} &gt; 0
\end{cases},
\end{equation*}\]</span> where <span class="math inline">\(\hat \lambda_{i j} = \max\{n_{i j}/E_{i j}, 1\}\)</span> is the maximum likelihood estimator of <span class="math inline">\(\lambda_{i j}\)</span> under <span class="math inline">\(H_0^{1:K} \cup H_a^{1:K}\)</span>, and subsequently, the maximum likelihood ratio statistic <span class="math inline">\(\text{MLR}^{1:K} = \max_{i = 1, \dots, I; j = 1, \dots, K} \text{LR}_{i j}\)</span>. The global null is rejected in favor of the global alternative hypothesis if <span class="math inline">\(\text{MLR}^{1:K}\)</span> is large, in which case subsequent post hoc tests of individual alternatives <span class="math inline">\(H_{a, i j}\)</span> can be performed based on the observed values of <span class="math inline">\({\text{LR}}_{i j}\)</span>. Computation of the global and post hoc <span class="math inline">\(p\)</span>-values are described in the following paragraph. Note that the computed or “observed” values of <span class="math inline">\({\text{LR}}_{i j}\)</span> aid rigorous statistical quantification of the signal strength in a pair <span class="math inline">\((i, j)\)</span>. Thus, we can use these quantities to rank the individual AE/drug pairs in terms of their signal strengths detected in the dataset under consideration. This is particularly useful for determining, the “most prominent” adverse events for a specific drug or a groups of drugs as detected in a dataset. Note also that the functional form of <span class="math inline">\(\text{LR}_{i j}\)</span> and hence that of <span class="math inline">\(\text{MLR}^{1:K}\)</span> does not contain the zero-inflation parameters <span class="math inline">\({\omega_j}\)</span>. However, the null sampling distributions necessary for performing the hypothesis tests do involve them. Computational approximations of these null sampling distributions are described next.</p>
<h4 class="unnumbered" id="bootstrapped-null-sampling-distribution-of-the-pseudo-likelihood-ratio-test-statistic-and-computing-p-values.">Bootstrapped null sampling distribution of the pseudo likelihood ratio test statistic, and computing <span class="math inline">\(p\)</span>-values.</h4>
<p>The null sampling distributions of the individual likelihood ratio statistics <span class="math inline">\(\{\text{LR}_{i j}\}\)</span> and the maximum likelihood ratio statistic <span class="math inline">\(\{\text{MLR}^{1:K}\}\)</span> are analytically intractable, and they must be approximated using computational techniques to facilitate inference. In <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> we use a parametric bootstrap resampling scheme for this purpose <span class="citation" data-cites="davison1997">(<a href="#ref-davison1997" role="doc-biblioref">Davison and Hinkley 1997</a>)</span>. To this end, zero-inflated Poisson models are fitted to the observed data under the global null hypothesis <span class="math inline">\(H_{0, ij}^{1:K}\)</span>, and null resampled counts <span class="math inline">\(\{\tilde n_{i j}\}\)</span> are subsequently generated from the fitted model. Fitting zero-inflated Poisson models requires estimation of the zero-inflation parameters <span class="math inline">\(\{\omega_j\}\)</span> from the observed data. In <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> the maximum likelihood estimates of <span class="math inline">\(\{\omega_j\}\)</span> are used for this purpose, which are obtained by maximizing the profile likelihood for <span class="math inline">\(\omega_j\)</span>:<br />
<span class="math display">\[
l_j(\omega) = (I - I_{0j}) \log (1 - \omega) + \sum_{\{i: n_{i j} = 0\}} \log \left(\omega + (1-\omega) \exp(-E_{i j}) \right),
\]</span> for <span class="math inline">\(j=1, \dots, J\)</span>. Here <span class="math inline">\(I_{0 j}\)</span> denotes the number of zero <span class="math inline">\(n_{ij}\)</span> values for each drug <span class="math inline">\(j\)</span>. The conditional (posterior) probability of encountering a zero-inflation at the <span class="math inline">\((i, j)\)</span> pair, given the corresponding observed count <span class="math inline">\(n_{ij}\)</span> is subsequently estimated and computed as: <span class="math display">\[\begin{aligned}
\hat \eta_{i j}^\text{obs}=
\begin{cases}
0, &amp; n_{i j} &gt; 0 \\
\frac{\hat \omega_{j}^\text{obs}}{\hat \omega_{j}^\text{obs} + (1-\hat \omega_{j}^\text{obs})\exp( - E_{i j})} &amp; n_{i j} = 0
\end{cases},
\end{aligned}
\]</span> for all <span class="math inline">\(i = 1, \dots, I\)</span> and <span class="math inline">\(j = 1, \dots, J\)</span>. Then, using these estimated conditional zero-inflation probabilities, null parametric bootstrap resamples are generated from <span class="math inline">\(\tilde n_{i j} \sim \text{ZIP}(E_{i j}, \hat \eta_{i j})\)</span>. Under an ordinary Poisson model, the parameters <span class="math inline">\(\omega_{j}\)</span> and <span class="math inline">\(\eta_{i j}\)</span> do not arise (they are equal to zero from a ZIP model perspective), and the null bootstrap resamples are simply generated from <span class="math inline">\(\tilde n_{i j} \sim \text{Poisson}(E_{i j})\)</span>; hence, the <em>bootstrap</em> null resamples become <em>exact</em> null resamples. In either case, these null resamples <span class="math inline">\(\tilde n_{ij}\)</span> produce null resampled values of the test statistic <span class="math inline">\(\text{MLR}^{1:K}\)</span>, and they collectively approximate the corresponding sampling distributions. The <span class="math inline">\(p\)</span>-value for the global null against the global alternative hypothesis is <span class="math inline">\(P^{1:K} = \operatorname{P}(\text{MLR}^{1:K}\geq \text{MLR}^{1:K, \text{obs}} \mid H_{0}^{1:K})\)</span>, and can be approximated by the proportion of the null resampled <span class="math inline">\(\text{MLR}^{1:K}\)</span> values that exceed the observed value of <span class="math inline">\(\text{MLR}^{1:K, \text{obs}}\)</span>: <span class="math display">\[
P^{1:K, \text{computed}} = \frac{1 + \sum_{h=1}^M \mathbb{1}\left(\widetilde{\text{MLR}}^{1:K, (h)} \geq \text{MLR}^{1:K, \text{obs}} \right)}{M+1},
\]</span> where <span class="math inline">\(\{\widetilde{\text{MLR}}^{1:K, (h)}: h = 1, \dots, M\}\)</span> are bootstrapped resampled null values of <span class="math inline">\(\text{MLR}^{1:K}\)</span>. The global null can be rejected at level <span class="math inline">\(\alpha\)</span> in favor of the global alternative hypothesis if <span class="math inline">\(P^{1:K, \text{computed}} &lt; \alpha\)</span>. The post hoc <span class="math inline">\(p\)</span>-values <span class="math inline">\(\{P_{ij} = \operatorname{P}(\text{MLR}^{1:K}\geq \text{LR}_{ij}^{\text{obs}} \mid H_{0}^{1:K}) \}\)</span> for the individual alternative hypotheses <span class="math inline">\(\{H_{a, i j}\}\)</span> are also approximated using the same null resampled values of <span class="math inline">\(\text{MLR}^{1:K}\)</span>: <span class="math display">\[
P^{\text{computed}}_{ij} = \frac{1 + \sum_{h=1}^M \mathbb{1}\left(\widetilde{\text{MLR}}^{1:K, (h)} \geq \text{LR}_{ij}^{\text{obs}} \right)}{M+1}.
\]</span> Note that these post hoc <span class="math inline">\(p\)</span>-values are based on the null sampling distribution of the <em>same</em> random test statistic <span class="math inline">\(\text{MLR}^{1:K}\)</span>; only the thresholds where the tail probabilities are evaluated are different (the individual observed <span class="math inline">\(\text{LR}_{ij}^{\text{obs}}\)</span> values). Consequently, no inflation in the overall type I error occurs due to multiplicity of hypothesis tests with independent test statistics. This also ensures that the false discovery rate of the entire procedure is controlled at the same level as the overall type I error.</p>
<h4 class="unnumbered" id="model-selection-determining-whether-to-use-a-poisson-model-or-a-zip-model-on-a-given-dataset.">Model selection: determining whether to use a Poisson model or a ZIP model on a given dataset.</h4>
<p>As noted before, the ZIP model above assumes that the report counts for a specific drug <span class="math inline">\(j\)</span> are zero-inflated with probability <span class="math inline">\(\omega_{j}\)</span>. If <span class="math inline">\(\omega_{j}\)</span> is small, then the ZIP model effectively reduces to an ordinary Poisson model; however in general, the ZIP model produces probabilities that are different from the Poisson probabilities. A natural question of particular practical importance is therefore which of these two models should be used on a given dataset. Clearly, if there is knowledge about the existence (or non-existence) of structural zero pairs in a given dataset (the exact positions of the structural zeros may possibly be unknown), then one should use a ZIP (ordinary Poisson) model. However, in practice no such knowledge may be available, and the optimal model must be determined entirely on the basis of the observed data. In other words, one must address what is known as a model selection problem. Because we are in a parametric likelihood framework, a convenient and canonical approach would involve the use of some penalized likelihood model selection criterion, such as the AIC or the BIC <span class="citation" data-cites="sakamoto1986">(<a href="#ref-sakamoto1986" role="doc-biblioref">Sakamoto et al. 1986</a>)</span> defined as:<br />
<span class="math display">\[
\operatorname{AIC}(k) = -2 \log(\hat{L}) + k p,
\]</span> where <span class="math inline">\(\hat{L}\)</span> denotes the maximized value of the likelihood function for a given model, <span class="math inline">\(k\)</span> is a penalty parameter, and <span class="math inline">\(p\)</span> is the total number of parameters in the model. The choice <span class="math inline">\(k = 2\)</span> produces the classical AIC, and <span class="math inline">\(k = \log N\)</span> where <span class="math inline">\(N\)</span> denotes the total number of data points, produces the BIC. The model with the minimum AIC (or BIC) is regarded as optimal. Note that in the current context for both the Poisson and the ZIP models, <span class="math inline">\(\log(\hat{L})\)</span> is obtained from the unrestricted model with each <span class="math inline">\(\{\lambda_{ij}\}\)</span> assuming values in <span class="math inline">\([1, \infty)\)</span>; the total number of parameters are <span class="math inline">\(p = I \times J\)</span> in the Poisson model and <span class="math inline">\((I + 1)J\)</span> in the ZIP model, and the total sample size <span class="math inline">\(N\)</span> is the grand total <span class="math inline">\(n_{\bullet \bullet}\)</span> of the contingency table.</p>
<h4 class="unnumbered" id="testing-for-zero-inflation-among-individual-drugs.">Testing for zero-inflation among individual drugs.</h4>
<p>The above model selection procedure addresses the question of whether to use the ZIP model or an ordinary Poisson model for a given dataset. When a ZIP model is determined to be optimal, interest may then lie in determining the statistical strength/significance of individual zero-inflation parameters <span class="math inline">\(\{\omega_{j}\}\)</span> specific to the individual drugs. In a null hypothesis significance testing framework, this can be translated as a test of <span class="math inline">\(H_{0, j^*}^\omega: \omega_{j^*} = 0\)</span> against <span class="math inline">\(H_{a, j^*}^\omega : \omega_{j^*} &gt; 0\)</span>, for each drug <span class="math inline">\(j^*\)</span>. The likelihood ratio test statistic for this problem is <span class="math inline">\(\text{LR}_{j^*}^\omega = \exp \left[l_j(\hat \omega_{j^*}) - l_j(0) \right]\)</span>, where <span class="math inline">\(l_{j^*}(\omega)\)</span> is the profile log likelihood function for <span class="math inline">\(\omega_{j^*}\)</span> as described before, and <span class="math inline">\(\hat \omega_{j^*}\)</span> is the corresponding maximum likelihood estimator obtained by maximizing <span class="math inline">\(l_{j^*}(\omega)\)</span>. Similar to the LRT for <span class="math inline">\(\lambda_{ij}\)</span> described before, the null sampling distribution of <span class="math inline">\(\text{LR}_{j^*}^\omega\)</span> is analytically intractable, but can be approximated using a parametric bootstrap. To this end, null resamples <span class="math inline">\(\{\tilde n_{ij^*}\}\)</span> are generated from the <span class="math inline">\(\text{ZIP}(\hat \lambda_{i j^*} E_{i j^*}, \omega_{j^*} = 0) \equiv \text{Poisson}(\hat \lambda_{i j^*} E_{i j^*})\)</span> distribution, where <span class="math inline">\(\hat \lambda{i j^*} = \max\{n_{ij^*}/E_{ij^*}, 1\}\)</span> is the maximum likelihood estimator of <span class="math inline">\(\lambda_{ij^*}\)</span>. From these null resamples <span class="math inline">\(\{\tilde{n}_{i j}\}\)</span>, null values of the likelihood ratio test statistics <span class="math inline">\(\{\widetilde{\text{LR}}_{j^*}^\omega\}\)</span> are computed. The <span class="math inline">\(p\)</span>-value of the test of <span class="math inline">\(H_{0, j^*}^\omega\)</span> against <span class="math inline">\(H_{a, j^*}^\omega\)</span> is subsequently computed using Monte Carlo estimated probability (proportion) of <em>null</em> <span class="math inline">\(\{\widetilde{\text{LR}}_{j^*}^\omega\}\)</span> values that are bigger than or equal to the “observed” value <span class="math inline">\(\text{LR}_{j^*}^{\omega,\ \text{obs}}\)</span>. If zero-inflation in several drugs <span class="math inline">\({j^*}\)</span> are tested simultaneously, then the overall <span class="math inline">\(p\)</span>-value needs to be adjusted for multiplicity; in <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> the Benjamini-Hochberg adjusted <span class="math inline">\(p\)</span>-values (<span class="math inline">\(q\)</span>-values; <span class="citation" data-cites="benjamini1995">Benjamini and Hochberg (<a href="#ref-benjamini1995" role="doc-biblioref">1995</a>)</span>) are used by default for this purpose.</p>
<h1 data-number="3" id="adverse-event-data-analysis-with"><span class="header-section-number">3</span> Adverse Event Data Analysis with <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a></h1>
<p>In this section we provide data analysis examples based on four real datasets contained in the <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> package. These datasets are stored as pre-processed contingency tables (in matrix-like forms) cataloging the aggregated report count for each AE/drug pair. To produce such contingency tables from raw adverse event data the <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> function <code>convert_raw_to_contin_table()</code> may be used. We note, however, that care must be taken before summarizing raw data into such contingency tables. First, a decision needs to be made regarding which reports to include in the analysis, e.g., only consider primary and/or secondary suspect drugs or also include concomitant and interacting drugs as cataloged in the FAERS database. The processed data stored in <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> only considers reports for the primary and secondary suspect drugs. Second, one needs to identify which drugs to include in the “Other Drug” category which plays the key role of “baseline” drugs. Appropriate specification of this baseline category is required because of its contribution to the total report counts for each AE, even though the AE/baseline drug category associations are not of primary importance. Typically, one identifies a handful of drugs of particular interest (e.g., the statins as discussed in the examples below) and combines all remaining drugs as baseline <span class="citation" data-cites="ding2020">(<a href="#ref-ding2020" role="doc-biblioref">Ding et al. 2020</a>)</span>. Finally, one needs to determine which AEs to include in the testing. The proposed LRT-based methods permit strong control over the global type I error and false discovery rates while ensuring high power <span class="citation" data-cites="chakraborty2022">(<a href="#ref-chakraborty2022" role="doc-biblioref">Chakraborty et al. 2022</a>)</span>, regardless of the total number of AEs. However, computation of the bootstrap p-values become increasingly more expensive with a growing number of AEs, and thus in practice one may encounter certain upper bounds to the total number of AEs that can be tested at a given time. We refer to <span class="citation" data-cites="ding2020">Ding et al. (<a href="#ref-ding2020" role="doc-biblioref">2020</a>)</span> for a detailed discussion on these considerations. The analyses provided below (divided into subsections) showcase examples with both small (<span class="math inline">\(I = 46\)</span>) and large (<span class="math inline">\(I \approx 6000\)</span>) number of AEs. We start by loading the <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> package into in <code>R</code> using <code>library()</code>.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>pvLRT</span><span class='op'>)</span></span></code></pre>
</div>
</div>
<h2 class="unnumbered" id="analysis-of-the-statin-dataset-with-47-adverse-events">Analysis of the statin dataset with 47 adverse events</h2>
<p>We begin with an analysis of the statin dataset labeled <code>statin46</code>, previously identified as being associated with the statin drugs. There are 47 rows in the data, corresponding to these 46 adverse events, and a collapsed category labeled “other” that consists of all other adverse events tabulated in the FAERS database during this time span. An older version of the data with the same 46 adverse events have been previously considered in <span class="citation" data-cites="ding2020">(<a href="#ref-ding2020" role="doc-biblioref">Ding et al. 2020</a>)</span>; our analysis that follows corroborates findings from that study. To begin the analysis, we first load the data into an R session and view the first few rows and columns with the following codes:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"statin46"</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>statin46</span><span class='op'>)</span><span class='op'>[</span>, <span class='fl'>1</span><span class='op'>:</span><span class='fl'>3</span><span class='op'>]</span></span></code></pre>
</div>
<pre><code>                                          Atorvastatin Fluvastatin
Acute Kidney Injury                               1353          42
Anuria                                              71           0
Blood Calcium Decreased                             14           2
Blood Creatine Phosphokinase Abnormal               34           0
Blood Creatine Phosphokinase Increased            1175         125
Blood Creatine Phosphokinase Mm Increased            2           0
                                          Lovastatin
Acute Kidney Injury                                7
Anuria                                             0
Blood Calcium Decreased                            0
Blood Creatine Phosphokinase Abnormal              0
Blood Creatine Phosphokinase Increased            32
Blood Creatine Phosphokinase Mm Increased          0</code></pre>
</div>
<p>Our interest lies in finding the most important adverse events of these 6 statin drugs. We note at the outset that the adverse events potentially caused by these drugs cannot be considered in isolation due to similarity among functions of statins, and thus we must consider all 6 drugs simultaneously while determining their important adverse events. This is achieved by performing a simultaneous test for all drugs of interest collectively within an <em>extended LRT</em> framework, which ensures that the overall type-I error of the process is preserved and the false discovery rate is controlled. See the previous section for a note on the statistical properties of the <em>extended LRT</em> test.</p>
<h4 class="unnumbered" id="analysis-based-on-an-ordinary-poisson-model.">Analysis based on an ordinary Poisson model.</h4>
<p>We first perform an LRT analysis based on the ordinary Poisson model on this dataset. The R codes for performing this task are as follows.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>100</span><span class='op'>)</span> </span>
<span><span class='va'>test_statin46_poisson</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/pvLRT/man/pvlrt.html'>pvlrt</a></span><span class='op'>(</span>   </span>
<span>  <span class='va'>statin46</span>,   </span>
<span>  zi_prob <span class='op'>=</span> <span class='fl'>0</span>,   </span>
<span>  test_drug_idx <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>6</span>   </span>
<span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>The first argument specifies the contingency table on which the LRT is to be performed, which is <code>statin46</code> in the current analysis. The argument <code>zi_prob = 0</code> pre-specifies the zero-inflation probability in the model for each drug to be 0, thus reducing the model to an ordinary Poisson model. Note that the ordinary Poisson model-based test can also be invoked through the wrapper function <code>lrt_poisson()</code>. Under the hood, <code>lrt_poisson()</code> sets <code>zi_prob = 0</code> and then calls <code>pvlrt()</code> itself. Next, the argument <code>test_drug_idx = 1:6</code> indicates that hypothesis tests are to be performed only on the first 6 columns, i.e., excluding the last column for “other” drugs. By default, all drugs supplied in <code>test_drug_idx</code> are tested simultaneously as a single group/class in an extended LRT framework; if other class/group structures exist among the drugs or if the drugs are to be tested separately (i.e., each tested drug forms its own class) those can be passed through <code>drug_class_idx</code>.</p>
<p>Running the code above creates a <code>pvlrt</code> S3 object. A <code>pvlrt</code> object is simply a reclassified <code>matrix</code> of computed (log) LRTs; it has the same dimensions as the input <code>contin_table</code>, with each cell containing the computed test statistic value for the corresponding AE/Drug pair. However, in addition to the usual <code>dimnames</code> attributes, a <code>pvlrt</code> object has several other attributes including the p-values associated with these log LRTs and estimates for zero-inflation parameters, and some meta-data including the type of test performed, the input data, etc. The printing method for <code>pvlrt</code> objects provides a high-level textual summary of the test. To glance at the overall results from the above LR tests, we simply print <code>test_statin46_poisson</code> on the R console:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>test_statin46_poisson</span></span></code></pre>
</div>
<pre><code>Relative reporting rate (lambda)-based ordinary-LRT on 47 AE &amp;
7 drugs. Hypothesis tests performed on 6 drugs using
parametric bootstrap.

Running time of the original pvlrt call: 1.475351 secs

No zi considered in the model.
Total 110 AE-drug pairs are significant at level = 0.05.

Extract all LR statistics and p-values using `summary()`.</code></pre>
</div>
<p>The above output shows that there are 110 significant AE/drug pairs at level <span class="math inline">\(\alpha = 0.05\)</span>. The likelihood ratio test statistics and <span class="math inline">\(p\)</span>-values for all AE/drug pairs can be extracted using <code>summary()</code>. This will produce a data table, with each row providing the sample size, likelihood ratio, and <span class="math inline">\(p\)</span>-value for each AE/drug pair present in the data<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># summary(test_statin46_poisson) </span></span></code></pre>
</div>
</div>
<p>The output is omitted for brevity. Note that for pairs that are not tested (here, all entries in the 7-th column of the <code>statin46</code> table) the test results will be missing (denoted by <code>NA</code> in <code>R</code>). <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> currently has implementations for three visual summary methods for the test results, namely bubble-plot, barplot, and heatmap. These can either be accessed explicitly through the functions <code>bubbleplot_pvrlt</code>, <code>barplot</code> (which is an <code>S</code>3 method for <code>pvlrt</code> objects), and <code>heatmap_pvlrt</code>, or directly through the <code>plot</code> method for <code>pvlrt</code> objects with the specifications of <code>type = "bubbleplot"</code> or <code>type = "barplot"</code>, or <code>type = "barplot"</code> respectively. As noted before in the Introduction, these plots are constructed using the <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> package <span class="citation" data-cites="pkg_ggplot2">(<a href="#ref-pkg_ggplot2" role="doc-biblioref">Wickham 2016</a>)</span>. As a result, several aspects of the produced plots can be updated post hoc, as facilitated by the principles of <em>grammar of graphics</em> that <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> implements. Examples of resizing axes label text sizes for <code>pvlrt</code> plots are provided below.</p>
<p>In the following we first create a bubble-plot top 15 AE/drug pairs across all 6 statin drugs, with pairs being ranked by the magnitudes of their likelihood ratio test statistic values. The figure is stored in <code>pl_bubble_statin46_poisson</code>, which is visualized upon printing.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>pl_bubble_statin46_poisson</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span></span>
<span>  <span class='va'>test_statin46_poisson</span>,</span>
<span>  type <span class='op'>=</span> <span class='st'>"bubbleplot"</span>,</span>
<span>  x_axis_measure <span class='op'>=</span> <span class='st'>"lrstat"</span>,</span>
<span>  size_measure <span class='op'>=</span> <span class='st'>"n"</span>,</span>
<span>  fill_measure <span class='op'>=</span> <span class='st'>"p_value"</span>,</span>
<span>  AE <span class='op'>=</span> <span class='fl'>15</span></span>
<span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>The arguments used in <code>plot()</code> above are described as follows. The first two arguments specify the <code>pvlrt</code> object under consideration and the type of plot to be made. All three visualizations implemented (<code>bubbleplot</code>, <code>barplot</code> and <code>heatmap</code>) display the test results by plotting AEs across the rows and Drugs across the columns, in their respective plots. The arguments <code>x_axis_measure = "lrstat"</code>, <code>size_measure = "n"</code>, and <code>fill_measure = "p_value"</code> indicate that the log LRT values will be displayed along the x-axis, the circle sizes in the bubbleplot will represent the corresponding sample size of an AE/drug pair, and that the circles will be color coded according to the p-values of the corresponding test statistics, respectively. Note that the three dimensions (<code>lrstat</code>, <code>n</code> and <code>p_value</code>) displayed through these three arguments can be interchanged. The argument <code>AE = 15</code> indicates that the top 15 adverse events with the largest log likelihood ratio test statistic values are to be plotted. Here these “largest” values are obtained by comparing across all drugs tested during the original <code>pvlrt</code> run. Instead of looking at “all” drugs while determining the top AEs, one can also focus on a specific subset of drugs by supplying the appropriate subset through the argument <code>Drug</code>. Moreover, instead of looking at the top adverse events, one can also explicitly specify which adverse events to display by directly passing their names through the same <code>AE</code> argument. When passing names through <code>AE</code> and <code>Drug</code>, partial string patterns can be supplied and then matched against their full names via regular expressions; this is achieved by setting <code>grep = TRUE</code>. For example, specifying <code>AE = "muscle"</code> and setting <code>grep = TRUE</code> will show all adverse events that have the character <code>"muscle"</code> in its name. We refer to the package documentation for further details on these two arguments. To visualize the plot produced, we can simply print <code>pl_bubble_statin46_poisson</code> on the R console or save it to a graphic device via <code>ggsave</code>. As noted before, post hoc editing of various graphical parameters in <code>pl_heat_statin46_poisson</code> can be made using <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> functions. Note that to use ggplot2 functions, we need to load the ggplot2 package. Below we show how the axes label sizes and legend text sizes can be modified. (Changing the default labels etc. may cause <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> to throw some warnings). We refer to the <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> package documentations and vignettes for further details on modifications of <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> objects and saving <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> objects to graphic devices.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://ggplot2.tidyverse.org'>ggplot2</a></span><span class='op'>)</span></span>
<span><span class='va'>pl_bubble_statin46_poisson</span> <span class='op'>+</span></span>
<span>  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span></span>
<span>    axis.text.y <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>13</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span>,</span>
<span>    axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>9</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span>,</span>
<span>    legend.title <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>12</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span>,</span>
<span>    strip.text <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>14</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span>,</span>
<span>    legend.position <span class='op'>=</span> <span class='st'>"top"</span></span>
<span>  <span class='op'>)</span></span></code></pre>
</div>
<p><img src="RJ-2023-027_files/figure-html5/show-heatmap-poisson-statin46-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<p>The above bubbleplot visualizes three simultaneous dimensions, viz., <code>lrstat</code>, <code>n</code> and <code>p_value</code> along the <span class="math inline">\(x\)</span>-axis, circle size, and circle color respectively. An alternative mode of visualization is made by paneled barplots which can visualize two simultaneous dimensions, e.g., the <span class="math inline">\(x\)</span>-axis showing the log likelihood ratio statistic, and the colors representing p-values as in the bubbleplot. The sample sizes can be displayed on the bar as text, by specifying <code>show_n = TRUE</code>. The following R codes create the barplot.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>pl_bar_statin46_poisson</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span></span>
<span>  <span class='va'>test_statin46_poisson</span>,</span>
<span>  type <span class='op'>=</span> <span class='st'>"barplot"</span>,</span>
<span>  fill_measure <span class='op'>=</span> <span class='st'>"p_value"</span>,</span>
<span>  x_axis_measure <span class='op'>=</span> <span class='st'>"lrstat"</span>,</span>
<span>  AE <span class='op'>=</span> <span class='fl'>15</span>,</span>
<span>  show_n <span class='op'>=</span> <span class='cn'>TRUE</span>,</span>
<span>  border_color <span class='op'>=</span> <span class='st'>"black"</span></span>
<span><span class='op'>)</span> </span>
<span><span class='va'>pl_bar_statin46_poisson</span> <span class='op'>+</span></span>
<span>  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span></span>
<span>    axis.text.y <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>13</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span>,</span>
<span>    axis.text.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>13</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span>,</span>
<span>    axis.title.x <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>14</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span>,</span>
<span>    legend.title <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>12</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span>,</span>
<span>    strip.text <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>14</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span>,</span>
<span>    legend.position <span class='op'>=</span> <span class='st'>"top"</span></span>
<span>  <span class='op'>)</span></span></code></pre>
</div>
<p><img src="RJ-2023-027_files/figure-html5/barplot-poisson-statin46-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<p>Finally, we can visualize only one dimension, say <code>p_value</code> as cell-colors in a heatmap, and inscribe as texts the other two dimensions, i.e., <code>lrstat</code> and <code>n</code> (by specifying <code>show_n = TRUE</code> and <code>show_lrstat = TRUE</code>). The following R codes perform this task.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>pl_heat_statin46_poisson</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/graphics/plot.default.html'>plot</a></span><span class='op'>(</span></span>
<span>  <span class='va'>test_statin46_poisson</span>,</span>
<span>  type <span class='op'>=</span> <span class='st'>"heatmap"</span>,</span>
<span>  fill_measure <span class='op'>=</span> <span class='st'>"p_value"</span>,</span>
<span>  AE <span class='op'>=</span> <span class='fl'>15</span>,</span>
<span>  show_n <span class='op'>=</span> <span class='cn'>TRUE</span>,</span>
<span>  show_lrstat <span class='op'>=</span> <span class='cn'>TRUE</span>,</span>
<span>  border_color <span class='op'>=</span> <span class='st'>"black"</span></span>
<span><span class='op'>)</span></span>
<span><span class='va'>pl_heat_statin46_poisson</span> <span class='op'>+</span></span>
<span>  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span></span>
<span>    axis.text <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>13</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span>,</span>
<span>    legend.title <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>12</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span>,</span>
<span>    strip.text <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>14</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span>,</span>
<span>    legend.position <span class='op'>=</span> <span class='st'>"top"</span></span>
<span>  <span class='op'>)</span></span></code></pre>
</div>
<p><img src="RJ-2023-027_files/figure-html5/create-heatmap-poisson-statin46-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<p>Note that all three plots above can be made interactive using external R packages; consider, e.g., <code>plotly::ggplotly(pl_bubble_statin46_poisson)</code>. Such interactive plots can be particularly useful in exploratory analyses performed on large adverse event datasets. We refer to the <a href="https://cran.r-project.org/package=plotly">plotly</a> <span class="citation" data-cites="pkg_plotly">(<a href="#ref-pkg_plotly" role="doc-biblioref">Sievert 2020</a>)</span> package documentation for further details on <code>plotly::ggplotly</code> and other interactive plotting functions. An interactive version of the above figure is displayed below.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'>## install plotly if not available</span></span>
<span><span class='co'># if (!requireNamespace("plotly")) install.packages("plotly")</span></span>
<span><span class='co'># interactive heatmap. turn all inscribed texts off</span></span>
<span><span class='fu'>plotly</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/plotly/man/ggplotly.html'>ggplotly</a></span><span class='op'>(</span><span class='va'>pl_heat_statin46_poisson</span><span class='op'>)</span></span></code></pre>
</div>
<div class="plotly html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-5cb92b4c836fdc2995ee" style="width:624px;height:384px;"></div>
<script type="application/json" data-for="htmlwidget-5cb92b4c836fdc2995ee">{"x":{"data":[{"x":[1,2,3,4,5,6],"y":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15],"z":[[0,0,0,1,0.98449999999999993,1],[0,0,0,0,0,0],[0,0,0,0.087099999999999997,0.00050000000000000001,1],[1,1,0,1,1,1],[0,0,0.00050000000000000001,1,0.18840000000000001,1],[0,0,0,0,0.0001,0],[0,0,0,0,0.0024000000000000002,0.076700000000000004],[0,0,0,0,0,1],[0,0,0,0,1,0],[0,0,0.0001,0.90629999999999999,1,1],[0,0,0,0,0,0],[0,0,0,0,0,0],[0,0,0,0,0,0],[0,0,0,0,0,0],[0,0,0,0,0,0]],"text":[["p_value: 0.00009999<br />info: n=48; logLR=120.35<br />Drug: Atorvastatin<br />AE: Muscle Enzyme Increased","p_value: 0.00009999<br />info: n=9; logLR=15.4<br />Drug: Simvastatin<br />AE: Muscle Enzyme Increased","p_value: 0.00009999<br />info: n=13; logLR=22.34<br />Drug: Rosuvastatin<br />AE: Muscle Enzyme Increased","p_value: 1.00000000<br />info: n=0; logLR=0<br />Drug: Pravastatin<br />AE: Muscle Enzyme Increased","p_value: 0.98450155<br />info: n=1; logLR=2.19<br />Drug: Fluvastatin<br />AE: Muscle Enzyme Increased","p_value: 1.00000000<br />info: n=0; logLR=0<br />Drug: Lovastatin<br />AE: Muscle Enzyme Increased"],["p_value: 0.00009999<br />info: n=137; logLR=32.48<br />Drug: Atorvastatin<br />AE: Musculoskeletal Discomfort","p_value: 0.00009999<br />info: n=93; logLR=48.4<br />Drug: Simvastatin<br />AE: Musculoskeletal Discomfort","p_value: 0.00009999<br />info: n=187; logLR=144.92<br />Drug: Rosuvastatin<br />AE: Musculoskeletal Discomfort","p_value: 0.00009999<br />info: n=25; logLR=13.39<br />Drug: Pravastatin<br />AE: Musculoskeletal Discomfort","p_value: 0.00009999<br />info: n=18; logLR=24.97<br />Drug: Fluvastatin<br />AE: Musculoskeletal Discomfort","p_value: 0.00009999<br />info: n=15; logLR=26.2<br />Drug: Lovastatin<br />AE: Musculoskeletal Discomfort"],["p_value: 0.00009999<br />info: n=71; logLR=146.05<br />Drug: Atorvastatin<br />AE: Myoglobin Blood Increased","p_value: 0.00009999<br />info: n=39; logLR=89.11<br />Drug: Simvastatin<br />AE: Myoglobin Blood Increased","p_value: 0.00009999<br />info: n=16; logLR=18.1<br />Drug: Rosuvastatin<br />AE: Myoglobin Blood Increased","p_value: 0.08719128<br />info: n=4; logLR=5.61<br />Drug: Pravastatin<br />AE: Myoglobin Blood Increased","p_value: 0.00059994<br />info: n=4; logLR=10.78<br />Drug: Fluvastatin<br />AE: Myoglobin Blood Increased","p_value: 1.00000000<br />info: n=0; logLR=0<br />Drug: Lovastatin<br />AE: Myoglobin Blood Increased"],["p_value: 1.00000000<br />info: n=6; logLR=0<br />Drug: Atorvastatin<br />AE: Creatinine Renal Clearance Decreased","p_value: 1.00000000<br />info: n=6; logLR=0<br />Drug: Simvastatin<br />AE: Creatinine Renal Clearance Decreased","p_value: 0.00009999<br />info: n=124; logLR=151.96<br />Drug: Rosuvastatin<br />AE: Creatinine Renal Clearance Decreased","p_value: 1.00000000<br />info: n=2; logLR=0<br />Drug: Pravastatin<br />AE: Creatinine Renal Clearance Decreased","p_value: 1.00000000<br />info: n=0; logLR=0<br />Drug: Fluvastatin<br />AE: Creatinine Renal Clearance Decreased","p_value: 1.00000000<br />info: n=0; logLR=0<br />Drug: Lovastatin<br />AE: Creatinine Renal Clearance Decreased"],["p_value: 0.00009999<br />info: n=68; logLR=163.06<br />Drug: Atorvastatin<br />AE: Muscle Necrosis","p_value: 0.00009999<br />info: n=20; logLR=40.56<br />Drug: Simvastatin<br />AE: Muscle Necrosis","p_value: 0.00059994<br />info: n=10; logLR=10.7<br />Drug: Rosuvastatin<br />AE: Muscle Necrosis","p_value: 1.00000000<br />info: n=1; logLR=0.58<br />Drug: Pravastatin<br />AE: Muscle Necrosis","p_value: 0.18848115<br />info: n=2; logLR=4.82<br />Drug: Fluvastatin<br />AE: Muscle Necrosis","p_value: 1.00000000<br />info: n=0; logLR=0<br />Drug: Lovastatin<br />AE: Muscle Necrosis"],["p_value: 0.00009999<br />info: n=219; logLR=284.97<br />Drug: Atorvastatin<br />AE: Myositis","p_value: 0.00009999<br />info: n=141; logLR=234.28<br />Drug: Simvastatin<br />AE: Myositis","p_value: 0.00009999<br />info: n=62; logLR=40.31<br />Drug: Rosuvastatin<br />AE: Myositis","p_value: 0.00009999<br />info: n=28; logLR=39.32<br />Drug: Pravastatin<br />AE: Myositis","p_value: 0.00019998<br />info: n=8; logLR=12.05<br />Drug: Fluvastatin<br />AE: Myositis","p_value: 0.00009999<br />info: n=10; logLR=22.56<br />Drug: Lovastatin<br />AE: Myositis"],["p_value: 0.00009999<br />info: n=340; logLR=302.34<br />Drug: Atorvastatin<br />AE: Chromaturia","p_value: 0.00009999<br />info: n=114; logLR=78.26<br />Drug: Simvastatin<br />AE: Chromaturia","p_value: 0.00009999<br />info: n=174; logLR=127.94<br />Drug: Rosuvastatin<br />AE: Chromaturia","p_value: 0.00009999<br />info: n=33; logLR=25.1<br />Drug: Pravastatin<br />AE: Chromaturia","p_value: 0.00249975<br />info: n=10; logLR=8.98<br />Drug: Fluvastatin<br />AE: Chromaturia","p_value: 0.07679232<br />info: n=6; logLR=5.71<br />Drug: Lovastatin<br />AE: Chromaturia"],["p_value: 0.00009999<br />info: n=181; logLR=333.29<br />Drug: Atorvastatin<br />AE: Muscle Rupture","p_value: 0.00009999<br />info: n=120; logLR=269.54<br />Drug: Simvastatin<br />AE: Muscle Rupture","p_value: 0.00009999<br />info: n=36; logLR=30.08<br />Drug: Rosuvastatin<br />AE: Muscle Rupture","p_value: 0.00009999<br />info: n=61; logLR=175.98<br />Drug: Pravastatin<br />AE: Muscle Rupture","p_value: 0.00009999<br />info: n=25; logLR=83.75<br />Drug: Fluvastatin<br />AE: Muscle Rupture","p_value: 1.00000000<br />info: n=0; logLR=0<br />Drug: Lovastatin<br />AE: Muscle Rupture"],["p_value: 0.00009999<br />info: n=291; logLR=454.05<br />Drug: Atorvastatin<br />AE: Muscle Disorder","p_value: 0.00009999<br />info: n=87; logLR=106.82<br />Drug: Simvastatin<br />AE: Muscle Disorder","p_value: 0.00009999<br />info: n=191; logLR=307.92<br />Drug: Rosuvastatin<br />AE: Muscle Disorder","p_value: 0.00009999<br />info: n=21; logLR=24.2<br />Drug: Pravastatin<br />AE: Muscle Disorder","p_value: 1.00000000<br />info: n=2; logLR=0.78<br />Drug: Fluvastatin<br />AE: Muscle Disorder","p_value: 0.00009999<br />info: n=7; logLR=13.43<br />Drug: Lovastatin<br />AE: Muscle Disorder"],["p_value: 0.00009999<br />info: n=279; logLR=1112.61<br />Drug: Atorvastatin<br />AE: Necrotising Myositis","p_value: 0.00009999<br />info: n=52; logLR=164.02<br />Drug: Simvastatin<br />AE: Necrotising Myositis","p_value: 0.00019998<br />info: n=10; logLR=12.49<br />Drug: Rosuvastatin<br />AE: Necrotising Myositis","p_value: 0.90630937<br />info: n=2; logLR=2.65<br />Drug: Pravastatin<br />AE: Necrotising Myositis","p_value: 1.00000000<br />info: n=0; logLR=0<br />Drug: Fluvastatin<br />AE: Necrotising Myositis","p_value: 1.00000000<br />info: n=0; logLR=0<br />Drug: Lovastatin<br />AE: Necrotising Myositis"],["p_value: 0.00009999<br />info: n=1857; logLR=1574.58<br />Drug: Atorvastatin<br />AE: Muscular Weakness","p_value: 0.00009999<br />info: n=859; logLR=774.85<br />Drug: Simvastatin<br />AE: Muscular Weakness","p_value: 0.00009999<br />info: n=808; logLR=464.12<br />Drug: Rosuvastatin<br />AE: Muscular Weakness","p_value: 0.00009999<br />info: n=152; logLR=89.99<br />Drug: Pravastatin<br />AE: Muscular Weakness","p_value: 0.00009999<br />info: n=45; logLR=31.66<br />Drug: Fluvastatin<br />AE: Muscular Weakness","p_value: 0.00009999<br />info: n=31; logLR=26.79<br />Drug: Lovastatin<br />AE: Muscular Weakness"],["p_value: 0.00009999<br />info: n=1175; logLR=2031.52<br />Drug: Atorvastatin<br />AE: Blood Creatine Phosphokinase Increased","p_value: 0.00009999<br />info: n=768; logLR=1626.15<br />Drug: Simvastatin<br />AE: Blood Creatine Phosphokinase Increased","p_value: 0.00009999<br />info: n=562; logLR=837.22<br />Drug: Rosuvastatin<br />AE: Blood Creatine Phosphokinase Increased","p_value: 0.00009999<br />info: n=200; logLR=421.48<br />Drug: Pravastatin<br />AE: Blood Creatine Phosphokinase Increased","p_value: 0.00009999<br />info: n=125; logLR=371.84<br />Drug: Fluvastatin<br />AE: Blood Creatine Phosphokinase Increased","p_value: 0.00009999<br />info: n=32; logLR=70.73<br />Drug: Lovastatin<br />AE: Blood Creatine Phosphokinase Increased"],["p_value: 0.00009999<br />info: n=849; logLR=2108.56<br />Drug: Atorvastatin<br />AE: Myopathy","p_value: 0.00009999<br />info: n=544; logLR=1558.83<br />Drug: Simvastatin<br />AE: Myopathy","p_value: 0.00009999<br />info: n=377; logLR=815.01<br />Drug: Rosuvastatin<br />AE: Myopathy","p_value: 0.00009999<br />info: n=145; logLR=417.31<br />Drug: Pravastatin<br />AE: Myopathy","p_value: 0.00009999<br />info: n=64; logLR=218.64<br />Drug: Fluvastatin<br />AE: Myopathy","p_value: 0.00009999<br />info: n=45; logLR=163.66<br />Drug: Lovastatin<br />AE: Myopathy"],["p_value: 0.00009999<br />info: n=2041; logLR=3994.43<br />Drug: Atorvastatin<br />AE: Rhabdomyolysis","p_value: 0.00009999<br />info: n=1376; logLR=3275.56<br />Drug: Simvastatin<br />AE: Rhabdomyolysis","p_value: 0.00009999<br />info: n=936; logLR=1567.2<br />Drug: Rosuvastatin<br />AE: Rhabdomyolysis","p_value: 0.00009999<br />info: n=163; logLR=264.93<br />Drug: Pravastatin<br />AE: Rhabdomyolysis","p_value: 0.00009999<br />info: n=52; logLR=95.28<br />Drug: Fluvastatin<br />AE: Rhabdomyolysis","p_value: 0.00009999<br />info: n=44; logLR=97.66<br />Drug: Lovastatin<br />AE: Rhabdomyolysis"],["p_value: 0.00009999<br />info: n=5362; logLR=8026.54<br />Drug: Atorvastatin<br />AE: Myalgia","p_value: 0.00009999<br />info: n=3216; logLR=5782.26<br />Drug: Simvastatin<br />AE: Myalgia","p_value: 0.00009999<br />info: n=2757; logLR=3660.8<br />Drug: Rosuvastatin<br />AE: Myalgia","p_value: 0.00009999<br />info: n=939; logLR=1779.83<br />Drug: Pravastatin<br />AE: Myalgia","p_value: 0.00009999<br />info: n=341; logLR=760.5<br />Drug: Fluvastatin<br />AE: Myalgia","p_value: 0.00009999<br />info: n=151; logLR=302.31<br />Drug: Lovastatin<br />AE: Myalgia"]],"colorscale":[[0,"#8B0000"],[0.0001,"#8B0000"],[0.00050000000000000001,"#8B0000"],[0.0024000000000000002,"#8B0101"],[0.076700000000000004,"#972314"],[0.087099999999999997,"#992616"],[0.18840000000000001,"#A8412D"],[0.90629999999999999,"#F8E8E4"],[0.98449999999999993,"#FEFBFB"],[1,"#FFFFFF"]],"type":"heatmap","showscale":false,"autocolorscale":false,"showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1],"y":[1],"name":"99_0ac207b18207b2ce934100174f14c321","type":"scatter","mode":"markers","opacity":0,"hoverinfo":"skip","showlegend":false,"marker":{"color":[0,1],"colorscale":[[0,"#8B0000"],[0.0033444816053511705,"#8C0201"],[0.006688963210702341,"#8C0402"],[0.010033444816053512,"#8D0603"],[0.013377926421404682,"#8D0804"],[0.016722408026755852,"#8E0A05"],[0.020066889632107024,"#8E0C06"],[0.023411371237458196,"#8F0E07"],[0.026755852842809364,"#8F1007"],[0.030100334448160532,"#901108"],[0.033444816053511704,"#901309"],[0.03678929765886288,"#91140A"],[0.040133779264214048,"#91160B"],[0.043478260869565216,"#92170C"],[0.046822742474916391,"#92180D"],[0.05016722408026756,"#931A0E"],[0.053511705685618728,"#941B0E"],[0.056856187290969896,"#941C0F"],[0.060200668896321065,"#951D10"],[0.06354515050167224,"#951E11"],[0.066889632107023408,"#961F11"],[0.070234113712374577,"#962012"],[0.073578595317725759,"#972213"],[0.076923076923076927,"#972314"],[0.080267558528428096,"#982415"],[0.083612040133779264,"#982515"],[0.086956521739130432,"#992616"],[0.090301003344481601,"#992717"],[0.093645484949832783,"#9A2818"],[0.096989966555183937,"#9A2918"],[0.10033444816053512,"#9B2A19"],[0.10367892976588629,"#9B2B1A"],[0.10702341137123746,"#9C2B1B"],[0.11036789297658864,"#9C2C1B"],[0.11371237458193979,"#9D2D1C"],[0.11705685618729098,"#9D2E1D"],[0.12040133779264213,"#9E2F1E"],[0.12374581939799331,"#9E301F"],[0.12709030100334448,"#9F311F"],[0.13043478260869565,"#9F3220"],[0.13377926421404682,"#A03321"],[0.13712374581939801,"#A03422"],[0.14046822742474915,"#A13522"],[0.14381270903010032,"#A13523"],[0.14715719063545152,"#A23624"],[0.15050167224080269,"#A23725"],[0.15384615384615385,"#A33825"],[0.157190635451505,"#A33926"],[0.16053511705685619,"#A43A27"],[0.16387959866220736,"#A43B28"],[0.16722408026755853,"#A53B28"],[0.1705685618729097,"#A53C29"],[0.17391304347826086,"#A63D2A"],[0.17725752508361203,"#A63E2B"],[0.1806020066889632,"#A73F2B"],[0.1839464882943144,"#A7402C"],[0.18729096989966557,"#A8402D"],[0.19063545150501671,"#A8412E"],[0.19397993311036787,"#A9422E"],[0.19732441471571907,"#A9432F"],[0.20066889632107024,"#AA4430"],[0.20401337792642141,"#AA4431"],[0.20735785953177258,"#AB4531"],[0.21070234113712374,"#AB4632"],[0.21404682274247491,"#AC4733"],[0.21739130434782608,"#AC4834"],[0.22073578595317728,"#AD4935"],[0.22408026755852842,"#AD4935"],[0.22742474916387959,"#AE4A36"],[0.23076923076923078,"#AE4B37"],[0.23411371237458195,"#AF4C38"],[0.23745819397993312,"#AF4D38"],[0.24080267558528426,"#B04D39"],[0.24414715719063546,"#B04E3A"],[0.24749163879598662,"#B14F3B"],[0.25083612040133779,"#B1503B"],[0.25418060200668896,"#B1503C"],[0.25752508361204013,"#B2513D"],[0.2608695652173913,"#B2523E"],[0.26421404682274247,"#B3533F"],[0.26755852842809363,"#B3543F"],[0.2709030100334448,"#B45440"],[0.27424749163879603,"#B45541"],[0.27759197324414714,"#B55642"],[0.28093645484949831,"#B55742"],[0.28428093645484953,"#B65843"],[0.28762541806020064,"#B65844"],[0.29096989966555187,"#B75945"],[0.29431438127090304,"#B75A46"],[0.29765886287625415,"#B85B46"],[0.30100334448160537,"#B85B47"],[0.30434782608695649,"#B85C48"],[0.30769230769230771,"#B95D49"],[0.31103678929765888,"#B95E4A"],[0.31438127090300999,"#BA5F4A"],[0.31772575250836121,"#BA5F4B"],[0.32107023411371238,"#BB604C"],[0.32441471571906355,"#BB614D"],[0.32775919732441472,"#BC624E"],[0.33110367892976589,"#BC624E"],[0.33444816053511706,"#BC634F"],[0.33779264214046822,"#BD6450"],[0.34113712374581939,"#BD6551"],[0.34448160535117056,"#BE6552"],[0.34782608695652173,"#BE6652"],[0.3511705685618729,"#BF6753"],[0.35451505016722407,"#BF6854"],[0.35785953177257529,"#C06955"],[0.3612040133779264,"#C06956"],[0.36454849498327757,"#C06A56"],[0.3678929765886288,"#C16B57"],[0.37123745819397991,"#C16C58"],[0.37458193979933113,"#C26C59"],[0.3779264214046823,"#C26D5A"],[0.38127090301003341,"#C36E5A"],[0.38461538461538464,"#C36F5B"],[0.38795986622073575,"#C36F5C"],[0.39130434782608697,"#C4705D"],[0.39464882943143814,"#C4715E"],[0.39799331103678925,"#C5725E"],[0.40133779264214048,"#C5735F"],[0.40468227424749165,"#C67360"],[0.40802675585284282,"#C67461"],[0.41137123745819398,"#C67562"],[0.41471571906354515,"#C77663"],[0.41806020066889632,"#C77663"],[0.42140468227424749,"#C87764"],[0.42474916387959871,"#C87865"],[0.42809364548494983,"#C87966"],[0.43143812709030099,"#C97967"],[0.43478260869565216,"#C97A67"],[0.43812709030100333,"#CA7B68"],[0.44147157190635455,"#CA7C69"],[0.44481605351170567,"#CB7C6A"],[0.44816053511705684,"#CB7D6B"],[0.45150501672240806,"#CB7E6C"],[0.45484949832775917,"#CC7F6C"],[0.4581939799331104,"#CC806D"],[0.46153846153846156,"#CD806E"],[0.46488294314381268,"#CD816F"],[0.4682274247491639,"#CD8270"],[0.47157190635451507,"#CE8371"],[0.47491638795986624,"#CE8371"],[0.47826086956521741,"#CF8472"],[0.48160535117056852,"#CF8573"],[0.48494983277591974,"#CF8674"],[0.48829431438127091,"#D08675"],[0.49163879598662208,"#D08776"],[0.49498327759197325,"#D18877"],[0.49832775919732442,"#D18977"],[0.50167224080267558,"#D18A78"],[0.50501672240802675,"#D28A79"],[0.50836120401337792,"#D28B7A"],[0.51170568561872909,"#D38C7B"],[0.51505016722408026,"#D38D7C"],[0.51839464882943143,"#D38D7C"],[0.52173913043478259,"#D48E7D"],[0.52508361204013376,"#D48F7E"],[0.52842809364548493,"#D4907F"],[0.5317725752508361,"#D59080"],[0.53511705685618727,"#D59181"],[0.53846153846153855,"#D69282"],[0.5418060200668896,"#D69382"],[0.54515050167224077,"#D69483"],[0.54849498327759205,"#D79484"],[0.55183946488294311,"#D79585"],[0.55518394648829428,"#D79686"],[0.55852842809364556,"#D89787"],[0.56187290969899661,"#D89788"],[0.56521739130434778,"#D99888"],[0.56856187290969906,"#D99989"],[0.57190635451505023,"#D99A8A"],[0.57525083612040129,"#DA9A8B"],[0.57859531772575257,"#DA9B8C"],[0.58193979933110374,"#DA9C8D"],[0.58528428093645479,"#DB9D8E"],[0.58862876254180607,"#DB9E8E"],[0.59197324414715724,"#DB9E8F"],[0.5953177257525083,"#DC9F90"],[0.59866220735785947,"#DCA091"],[0.60200668896321075,"#DDA192"],[0.60535117056856191,"#DDA193"],[0.60869565217391297,"#DDA294"],[0.61204013377926425,"#DEA395"],[0.61538461538461542,"#DEA495"],[0.61872909698996648,"#DEA596"],[0.62207357859531776,"#DFA597"],[0.62541806020066892,"#DFA698"],[0.62876254180601998,"#DFA799"],[0.63210702341137126,"#E0A89A"],[0.63545150501672243,"#E0A89B"],[0.6387959866220736,"#E0A99C"],[0.64214046822742477,"#E1AA9C"],[0.64548494983277593,"#E1AB9D"],[0.6488294314381271,"#E1AB9E"],[0.65217391304347827,"#E2AC9F"],[0.65551839464882944,"#E2ADA0"],[0.65886287625418061,"#E2AEA1"],[0.66220735785953178,"#E3AFA2"],[0.66555183946488294,"#E3AFA3"],[0.66889632107023411,"#E4B0A4"],[0.67224080267558539,"#E4B1A4"],[0.67558528428093645,"#E4B2A5"],[0.67892976588628762,"#E5B3A6"],[0.68227424749163879,"#E5B3A7"],[0.68561872909698995,"#E5B4A8"],[0.68896321070234112,"#E5B5A9"],[0.69230769230769229,"#E6B6AA"],[0.69565217391304346,"#E6B6AB"],[0.69899665551839463,"#E6B7AC"],[0.7023411371237458,"#E7B8AC"],[0.70568561872909707,"#E7B9AD"],[0.70903010033444813,"#E7BAAE"],[0.7123745819397993,"#E8BAAF"],[0.71571906354515058,"#E8BBB0"],[0.71906354515050164,"#E8BCB1"],[0.72240802675585281,"#E9BDB2"],[0.72575250836120409,"#E9BDB3"],[0.72909698996655514,"#E9BEB4"],[0.73244147157190631,"#EABFB5"],[0.73578595317725759,"#EAC0B5"],[0.73913043478260876,"#EAC1B6"],[0.74247491638795982,"#EBC1B7"],[0.7458193979933111,"#EBC2B8"],[0.74916387959866226,"#EBC3B9"],[0.75250836120401332,"#ECC4BA"],[0.7558528428093646,"#ECC5BB"],[0.75919732441471577,"#ECC5BC"],[0.76254180602006683,"#ECC6BD"],[0.76588628762541811,"#EDC7BE"],[0.76923076923076927,"#EDC8BE"],[0.77257525083612044,"#EDC9BF"],[0.7759197324414715,"#EEC9C0"],[0.77926421404682278,"#EECAC1"],[0.78260869565217395,"#EECBC2"],[0.785953177257525,"#EFCCC3"],[0.78929765886287628,"#EFCCC4"],[0.79264214046822745,"#EFCDC5"],[0.79598662207357851,"#EFCEC6"],[0.79933110367892979,"#F0CFC7"],[0.80267558528428096,"#F0D0C8"],[0.80602006688963213,"#F0D0C9"],[0.80936454849498329,"#F1D1C9"],[0.81270903010033446,"#F1D2CA"],[0.81605351170568563,"#F1D3CB"],[0.8193979933110368,"#F1D4CC"],[0.82274247491638797,"#F2D4CD"],[0.82608695652173914,"#F2D5CE"],[0.8294314381270903,"#F2D6CF"],[0.83277591973244147,"#F3D7D0"],[0.83612040133779264,"#F3D8D1"],[0.83946488294314392,"#F3D8D2"],[0.84280936454849498,"#F3D9D3"],[0.84615384615384615,"#F4DAD4"],[0.84949832775919742,"#F4DBD5"],[0.85284280936454848,"#F4DCD5"],[0.85618729096989965,"#F4DCD6"],[0.85953177257525093,"#F5DDD7"],[0.86287625418060199,"#F5DED8"],[0.86622073578595316,"#F5DFD9"],[0.86956521739130432,"#F6E0DA"],[0.87290969899665549,"#F6E0DB"],[0.87625418060200666,"#F6E1DC"],[0.87959866220735783,"#F6E2DD"],[0.88294314381270911,"#F7E3DE"],[0.88628762541806017,"#F7E4DF"],[0.88963210702341133,"#F7E4E0"],[0.89297658862876261,"#F7E5E1"],[0.89632107023411367,"#F8E6E2"],[0.89966555183946484,"#F8E7E3"],[0.90301003344481612,"#F8E8E3"],[0.90635451505016718,"#F8E8E4"],[0.90969899665551834,"#F9E9E5"],[0.91304347826086962,"#F9EAE6"],[0.91638795986622079,"#F9EBE7"],[0.91973244147157185,"#F9ECE8"],[0.92307692307692313,"#FAECE9"],[0.9264214046822743,"#FAEDEA"],[0.92976588628762535,"#FAEEEB"],[0.93311036789297663,"#FAEFEC"],[0.9364548494983278,"#FBF0ED"],[0.93979933110367886,"#FBF0EE"],[0.94314381270903014,"#FBF1EF"],[0.94648829431438131,"#FBF2F0"],[0.94983277591973247,"#FCF3F1"],[0.95317725752508364,"#FCF4F2"],[0.95652173913043481,"#FCF4F3"],[0.95986622073578598,"#FCF5F4"],[0.96321070234113704,"#FDF6F4"],[0.96655518394648832,"#FDF7F5"],[0.96989966555183948,"#FDF8F6"],[0.97324414715719054,"#FDF9F7"],[0.97658862876254182,"#FDF9F8"],[0.97993311036789299,"#FEFAF9"],[0.98327759197324416,"#FEFBFA"],[0.98662207357859533,"#FEFCFB"],[0.98996655518394649,"#FEFDFC"],[0.99331103678929766,"#FFFDFD"],[0.99665551839464883,"#FFFEFE"],[1,"#FFFFFF"]],"colorbar":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.8897637795275593,"thickness":23.039999999999996,"title":"p_value","titlefont":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724},"tickmode":"array","ticktext":["0.25","0.50","0.75","1.00"],"tickvals":[0.24992500000000001,0.49995000000000001,0.74997499999999995,1],"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":11.68949771689498},"ticklen":2,"len":0.5}},"xaxis":"x","yaxis":"y","frame":null}],"layout":{"margin":{"t":27.872146118721464,"r":7.3059360730593621,"b":85.662100456621033,"l":233.05936073059365},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.40000000000000002,6.5999999999999996],"tickmode":"array","ticktext":["Atorvastatin","Simvastatin","Rosuvastatin","Pravastatin","Fluvastatin","Lovastatin"],"tickvals":[1,2,3,3.9999999999999996,5,6],"categoryorder":"array","categoryarray":["Atorvastatin","Simvastatin","Rosuvastatin","Pravastatin","Fluvastatin","Lovastatin"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.6529680365296811,"tickwidth":0.66417600664176002,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.68949771689498},"tickangle":-90,"showline":false,"linecolor":null,"linewidth":0,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.40000000000000002,15.6],"tickmode":"array","ticktext":["Muscle Enzyme Increased","Musculoskeletal Discomfort","Myoglobin Blood Increased","Creatinine Renal Clearance Decreased","Muscle Necrosis","Myositis","Chromaturia","Muscle Rupture","Muscle Disorder","Necrotising Myositis","Muscular Weakness","Blood Creatine Phosphokinase Increased","Myopathy","Rhabdomyolysis","Myalgia"],"tickvals":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15],"categoryorder":"array","categoryarray":["Muscle Enzyme Increased","Musculoskeletal Discomfort","Myoglobin Blood Increased","Creatinine Renal Clearance Decreased","Muscle Necrosis","Myositis","Chromaturia","Muscle Rupture","Muscle Disorder","Necrotising Myositis","Muscular Weakness","Blood Creatine Phosphokinase Increased","Myopathy","Rhabdomyolysis","Myalgia"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.6529680365296811,"tickwidth":0.66417600664176002,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.68949771689498},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.8897637795275593,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.68949771689498},"title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724}}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"source":"A","attrs":{"dd4b19513a81":{"fill":{},"label":{},"x":{},"y":{},"type":"heatmap"},"dd4b23d7f823":{"fill":{},"label":{},"x":{},"y":{}}},"cur_data":"dd4b19513a81","visdat":{"dd4b19513a81":["function (y) ","x"],"dd4b23d7f823":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.20000000000000001,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
<p>The following observations are made from the above plots. First, Myalgia appears to be the AE with the largest computed log likelihood ratio (LR) statistic across all 6 statin drugs combined. It has the largest LR in Atorvastatin, followed by Simvastatin, Rosuvastatin, Pravastatin, Fluvastatin, and Lovastatin. Next is Rhabdomyolysis, which displays a near similar pattern as Myalgia in terms of computed signal/association strength among the six drugs, except here Lovastatin has a slightly larger computed LR value than Fluvastatin. Second, not all of the 15 adverse events are statistically significant across all six statin drugs. For example, Necrotising Myositis is a statistically significant adverse event of Atorvastatin, Simvastatin, and Rosuvastatin, but not for Pravastatin, Fluvastatin, and Lovastatin. Third, the sample size <span class="math inline">\(n_{i j}\)</span> of a specific AE/drug pair <span class="math inline">\((i, j)\)</span> alone does not provide much information about its strength of association; for example, the pairs (Muscle Necrosis, Rosuvastatin), (Myositis, Lovastatin), (Chromaturia, Fluvastatin), and (Necrotising Myositis, Rosuvastatin) all have sample sizes <span class="math inline">\(n_{ij} = 10\)</span>, but their computed likelihood ratio statistics, and hence the strengths of their associations vary. Moreover, a pair with a smaller sample size may harbor a stronger signal than one with a larger sample size; for example for Fluvastatin, Chromaturia has a sample size of <span class="math inline">\(n_{ij} = 10\)</span> and Myoglobin Blood Increased has a sample size of <span class="math inline">\(n_{ij} = 4\)</span>; yet the former AE has a smaller LRT statistic value than the latter.</p>
<h4 class="unnumbered" id="analysis-based-on-a-zero-inflated-poisson-model.">Analysis based on a zero-inflated Poisson model.</h4>
<p>We now consider zero-inflated Poisson model-based LRTs to identify signals in the <code>statin46</code> data. To this end, the same <code>pvlrt</code> function can be used, with two changes in the parameters: (a) the argument <code>zi_prob</code> is now set to <code>NULL</code>, which ensures that the zero-inflation parameters are estimated from the data, and (b) the logical argument <code>test_zi</code> is now set to <code>TRUE</code>, which specifies that a separate (pseudo) likelihood ratio test for the significance of each drug-specific zero inflation parameter will be performed. The following are the R codes used:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>100</span><span class='op'>)</span> </span>
<span><span class='va'>test_statin46_zip</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/pvLRT/man/pvlrt.html'>pvlrt</a></span><span class='op'>(</span>   </span>
<span>  <span class='va'>statin46</span>,   </span>
<span>  test_drug_idx <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>6</span>,   </span>
<span>  zi_prob <span class='op'>=</span> <span class='cn'>NULL</span>,</span>
<span>  test_zi <span class='op'>=</span> <span class='cn'>TRUE</span></span>
<span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>A textual summary of the analysis can be obtained by printing the output of <code>test_statin46_zip</code> on the R console:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>test_statin46_zip</span></span></code></pre>
</div>
<pre><code>Relative reporting rate (lambda)-based pseudo-LRT on 47 AE &amp; 7
drugs. Hypothesis tests performed on 6 drugs using parametric
bootstrap.

Running time of the original pvlrt call: 6.557479 secs

Drug-specific estimated zi probabilities: 0 (Atorvastatin),
0.07 (Fluvastatin), 0.16 (Lovastatin), 0.06 (Pravastatin), 0
(Rosuvastatin), 0 (Simvastatin), 0 (Other)

LR test of significance for zi probabilities: LRstat= 0.00,
q&gt;0.90 (Atorvastatin), LRstat= 5.28, q&lt;0.001 (Fluvastatin),
LRstat= 3.81, q&lt;0.001 (Lovastatin), LRstat=32.49, q&lt;0.001
(Pravastatin), LRstat= 0.00, q&gt;0.90 (Rosuvastatin), LRstat=
0.00, q&gt;0.90 (Simvastatin), LRstat= 0.00, q=&lt;NA&gt; (Other)
Total 110 AE-drug pairs are significant at level = 0.05.

Extract all LR statistics and p-values using `summary()`.</code></pre>
</div>
<p>It follows from the output that the estimated drug-specific zero-inflation probabilities are significantly different from zero (q-value or FDR adjusted <span class="math inline">\(p\)</span>-value &lt; 0.05) only for Fluvastatin and Lovastatin, where the point estimates of the zero-inflation parameters are <span class="math inline">\(\omega = 0.07\)</span> and <span class="math inline">\(\omega = 0.16\)</span> respectively. The total number of significant AE/drug is 110, which is the same as the number of significant pairs detected from the Poisson model. To see which 110 pairs detected to be significant in the two models, we can use the function <code>extract_significant_pairs</code> to extract a <code>data.table</code> containing the test results of significant AE/drug pairs and at a given level <span class="math inline">\(\alpha = 0.05\)</span> from the two models, and subsequently compare these results.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># extract_significant_pairs(test_statin46_poisson)</span></span>
<span><span class='co'># extract_significant_pairs(test_statin46_zip)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/all.equal.html'>all.equal</a></span><span class='op'>(</span></span>
<span>  <span class='fu'><a href='https://rdrr.io/pkg/pvLRT/man/extract_lrstat_matrix.html'>extract_significant_pairs</a></span><span class='op'>(</span><span class='va'>test_statin46_poisson</span><span class='op'>)</span>,</span>
<span>  <span class='fu'><a href='https://rdrr.io/pkg/pvLRT/man/extract_lrstat_matrix.html'>extract_significant_pairs</a></span><span class='op'>(</span><span class='va'>test_statin46_zip</span><span class='op'>)</span></span>
<span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>[1] &quot;Column &#39;p_value&#39;: Mean relative difference: 0.06278027&quot;</code></pre>
</div>
<p>As it appears the above two <code>pvlrt</code> objects differ only in the computed <span class="math inline">\(p\)</span>-values, but not in any other fields/statistics. That the computed LR statistics are identical is not surprising, since the functional form of the test statistic is exactly the same for the two models. The <span class="math inline">\(p\)</span>-values, on the other hand, depend on the null sampling distributions of LR statistics, which are different for the two models. Moreover, here the <span class="math inline">\(p\)</span>-values are numerically computed using Monte Carlo methods based on bootstrap resampling, which means that even for the same null sampling distributions, the computed <span class="math inline">\(p\)</span>-values may show random variations between two instances of the computation. However, that there is no difference between the identified significant (computed <span class="math inline">\(p\)</span>-value &lt; 0.05) AE/drug pairs for the two models essentially indicates concordance of the signals identified.</p>
<p>While the two models identify the same AE/drug pairs to be significant at the <span class="math inline">\(0.05\)</span> level, they may differ in terms of predictive accuracy/goodness of fit, which can be compared through model selection criteria such as BIC. The <code>logLik</code> methods for <code>pvlrt</code> objects implemented in the package (see the documentation for <code>logLik.pvlrt</code>) allow automated computations of the BIC and AIC of the fitted models, using the <code>stats</code> <code>S3</code> functions <code>BIC</code> and <code>AIC</code>. Below we use these two functions to compare the predictive accuracies of the fitted Poisson and ZIP models.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/AIC.html'>BIC</a></span><span class='op'>(</span><span class='va'>test_statin46_poisson</span>, <span class='va'>test_statin46_zip</span><span class='op'>)</span>, </span>
<span>      <span class='fu'><a href='https://rdrr.io/r/stats/AIC.html'>AIC</a></span><span class='op'>(</span><span class='va'>test_statin46_poisson</span>, <span class='va'>test_statin46_zip</span><span class='op'>)</span><span class='op'>)</span><span class='op'>[</span>, <span class='op'>-</span><span class='fl'>3</span><span class='op'>]</span> <span class='co'># remove duplicated df col</span></span></code></pre>
</div>
<pre><code>                       df      BIC      AIC
test_statin46_poisson 329 15962.47 10707.02
test_statin46_zip     336 16005.13 10637.85</code></pre>
</div>
<p>It follows that the Poisson model has smaller BIC but larger AIC than the ZIP model for the current dataset. This essentially implies that the Poisson model and the ZIP model both perform equally well in modeling the <code>statin46</code> dataset.</p>
<h2 class="unnumbered" id="analysis-of-the-full-statin-data-with-6039-adverse-events">Analysis of the full <code>statin</code> data with 6039 adverse events</h2>
<p>We next analyze the full <code>statin</code> dataset with all 6039 adverse events. This dataset differs from the smaller <code>statin46</code> dataset in that it explicitly catalogs report-counts of many adverse events that are grouped together as “Other AE” in <code>statin46</code>. The first few rows of the bigger <code>statin</code> dataset can be displayed using <code>head(statin)</code>; we omit the output for brevity.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"statin"</span><span class='op'>)</span></span>
<span><span class='co'># head(statin)</span></span></code></pre>
</div>
</div>
<p>In the following we first run a Poisson LRT, followed by a ZIP LRT, and then compare the two models. The model that fits the data better (assessed through BIC/AIC) will subsequently be used for inference. The following commands run a Poisson LRT on the <code>statin</code> dataset, with the default 10,000 resamples for computation of <span class="math inline">\(p\)</span>-values. The remaining arguments are analogous to the case of analysis of <code>statin46</code>. Once created, the <code>pvlrt</code> objects will output brief textual summary of the results; the outputs are omitted for brevity.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>100</span><span class='op'>)</span></span>
<span><span class='va'>test_statin_poisson</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/pvLRT/man/pvlrt.html'>pvlrt</a></span><span class='op'>(</span></span>
<span>  <span class='va'>statin</span>,</span>
<span>  zi_prob <span class='op'>=</span> <span class='fl'>0</span>,</span>
<span>  test_drug_idx <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>6</span></span>
<span><span class='op'>)</span></span>
<span><span class='co'># test_statin_poisson # print the results</span></span></code></pre>
</div>
</div>
<p>Next we run the zero-inflated Poisson model. Codes are as follows.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>100</span><span class='op'>)</span></span>
<span><span class='va'>test_statin_zip</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/pvLRT/man/pvlrt.html'>pvlrt</a></span><span class='op'>(</span></span>
<span>  <span class='va'>statin</span>,</span>
<span>  test_drug_idx <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>6</span>,</span>
<span>  zi_prob <span class='op'>=</span> <span class='cn'>NULL</span>,</span>
<span>  test_zi <span class='op'>=</span> <span class='cn'>TRUE</span></span>
<span><span class='op'>)</span></span>
<span><span class='co'># test_statin_zip # print the results</span></span></code></pre>
</div>
</div>
<!-- ```{r print-zip-statin} -->
<!-- # print the results -->
<!-- # test_statin_zip -->
<!-- ``` -->
<p>As it appears, in this bigger dataset all drugs (except the “other” group) have &gt; 10% zero inflation probabilities with statistically significant FDR adjusted <span class="math inline">\(p\)</span>-values (<span class="math inline">\(q\)</span>-values). This indicates that the zero-inflated Poisson model based test would likely be more appropriate here. We formally check this via model comparison using AIC and BIC as follows:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/AIC.html'>BIC</a></span><span class='op'>(</span><span class='va'>test_statin_poisson</span>, <span class='va'>test_statin_zip</span><span class='op'>)</span>,</span>
<span>      <span class='fu'><a href='https://rdrr.io/r/stats/AIC.html'>AIC</a></span><span class='op'>(</span><span class='va'>test_statin_poisson</span>, <span class='va'>test_statin_zip</span><span class='op'>)</span><span class='op'>)</span><span class='op'>[</span>, <span class='op'>-</span><span class='fl'>3</span><span class='op'>]</span> <span class='co'>#remove duplicated df col</span></span></code></pre>
</div>
<pre><code>                       df     BIC      AIC
test_statin_poisson 42273 1063654 388384.2
test_statin_zip     42280 1043314 367931.7</code></pre>
</div>
<p>It follows that here the zero-inflated Poisson model has both smaller AIC and BIC than the Poisson model, thus confirming our suspicion that the former model is a better fit for the data. We shall therefore conduct the subsequent analysis based on the zero-inflated Poisson model-based tests.</p>
<p>To get a visual understanding of the results we plot a heatmap of the <span class="math inline">\(p\)</span>-values with LR test statistics and samples sizes inscribed as texts:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>pl_heat_statin_zip</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/pvLRT/man/heatmap_pvlrt.html'>heatmap_pvlrt</a></span><span class='op'>(</span></span>
<span>  <span class='va'>test_statin_zip</span>,</span>
<span>  AE <span class='op'>=</span> <span class='fl'>15</span>,</span>
<span>  show_n <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span>  show_lrstat <span class='op'>=</span> <span class='cn'>FALSE</span></span>
<span><span class='op'>)</span> </span>
<span><span class='fu'>plotly</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/plotly/man/ggplotly.html'>ggplotly</a></span><span class='op'>(</span><span class='va'>pl_heat_statin_zip</span><span class='op'>)</span></span></code></pre>
</div>
<div class="plotly html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-54f18e87994fa40b7f7a" style="width:100%;height:576px;"></div>
<script type="application/json" data-for="htmlwidget-54f18e87994fa40b7f7a">{"x":{"data":[{"x":[1,2,3,4,5,6],"y":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15],"z":[[0,0.00050000000000000001,0,0.013300000000000001,0.012200000000000001,0.76169999999999993],[0,0,0,1,1,1],[0,0,0.19800000000000001,1,1,1],[0,0,0,0,0,0],[0,0,0,0,1,0.48159999999999997],[0,0,0,0,0,0.0001],[0,0,0,0,1,1],[1,1,0,1,1,1],[0,0,0,0,0,0],[0,0,0,0,0,0],[0,0,0,0,0.69850000000000001,0],[0,0,0,0,0.024500000000000001,0],[0,0,0,0,0,0],[0,0,0,0,0,0],[0,0,0,0,0.038600000000000002,0]],"text":[["p_value: 0.00009999<br />info: n=215; logLR=162.94; p<0.001<br />Drug: Atorvastatin<br />AE: Blood Triglycerides Increased","p_value: 0.00059994<br />info: n=58; logLR=24.01; p<0.001<br />Drug: Simvastatin<br />AE: Blood Triglycerides Increased","p_value: 0.00009999<br />info: n=427; logLR=757.8; p<0.001<br />Drug: Rosuvastatin<br />AE: Blood Triglycerides Increased","p_value: 0.01339866<br />info: n=24; logLR=17.81; p=0.013<br />Drug: Pravastatin<br />AE: Blood Triglycerides Increased","p_value: 0.01229877<br />info: n=13; logLR=17.92; p=0.012<br />Drug: Fluvastatin<br />AE: Blood Triglycerides Increased","p_value: 0.76172383<br />info: n=7; logLR=9.38; p=0.762<br />Drug: Lovastatin<br />AE: Blood Triglycerides Increased"],["p_value: 0.00009999<br />info: n=275; logLR=1101.65; p<0.001<br />Drug: Atorvastatin<br />AE: Immune-Mediated Myositis","p_value: 0.00009999<br />info: n=70; logLR=243.58; p<0.001<br />Drug: Simvastatin<br />AE: Immune-Mediated Myositis","p_value: 0.00009999<br />info: n=26; logLR=56.22; p<0.001<br />Drug: Rosuvastatin<br />AE: Immune-Mediated Myositis","p_value: 1.00000000<br />info: n=0; logLR=0; p>0.90<br />Drug: Pravastatin<br />AE: Immune-Mediated Myositis","p_value: 1.00000000<br />info: n=0; logLR=0; p>0.90<br />Drug: Fluvastatin<br />AE: Immune-Mediated Myositis","p_value: 1.00000000<br />info: n=0; logLR=0; p>0.90<br />Drug: Lovastatin<br />AE: Immune-Mediated Myositis"],["p_value: 0.00009999<br />info: n=279; logLR=1112.61; p<0.001<br />Drug: Atorvastatin<br />AE: Necrotising Myositis","p_value: 0.00009999<br />info: n=52; logLR=164.02; p<0.001<br />Drug: Simvastatin<br />AE: Necrotising Myositis","p_value: 0.19808019<br />info: n=10; logLR=12.49; p=0.198<br />Drug: Rosuvastatin<br />AE: Necrotising Myositis","p_value: 1.00000000<br />info: n=2; logLR=2.65; p>0.90<br />Drug: Pravastatin<br />AE: Necrotising Myositis","p_value: 1.00000000<br />info: n=0; logLR=0; p>0.90<br />Drug: Fluvastatin<br />AE: Necrotising Myositis","p_value: 1.00000000<br />info: n=0; logLR=0; p>0.90<br />Drug: Lovastatin<br />AE: Necrotising Myositis"],["p_value: 0.00009999<br />info: n=2072; logLR=1200.78; p<0.001<br />Drug: Atorvastatin<br />AE: Muscle Spasms","p_value: 0.00009999<br />info: n=1103; logLR=811.72; p<0.001<br />Drug: Simvastatin<br />AE: Muscle Spasms","p_value: 0.00009999<br />info: n=1458; logLR=982.1; p<0.001<br />Drug: Rosuvastatin<br />AE: Muscle Spasms","p_value: 0.00009999<br />info: n=301; logLR=230.08; p<0.001<br />Drug: Pravastatin<br />AE: Muscle Spasms","p_value: 0.00009999<br />info: n=97; logLR=92.45; p<0.001<br />Drug: Fluvastatin<br />AE: Muscle Spasms","p_value: 0.00009999<br />info: n=69; logLR=79.54; p<0.001<br />Drug: Lovastatin<br />AE: Muscle Spasms"],["p_value: 0.00009999<br />info: n=1352; logLR=730.6; p<0.001<br />Drug: Atorvastatin<br />AE: Drug Interaction","p_value: 0.00009999<br />info: n=1303; logLR=1542.51; p<0.001<br />Drug: Simvastatin<br />AE: Drug Interaction","p_value: 0.00009999<br />info: n=571; logLR=165.31; p<0.001<br />Drug: Rosuvastatin<br />AE: Drug Interaction","p_value: 0.00009999<br />info: n=179; logLR=116.52; p<0.001<br />Drug: Pravastatin<br />AE: Drug Interaction","p_value: 1.00000000<br />info: n=15; logLR=0.58; p>0.90<br />Drug: Fluvastatin<br />AE: Drug Interaction","p_value: 0.48165183<br />info: n=21; logLR=10.57; p=0.482<br />Drug: Lovastatin<br />AE: Drug Interaction"],["p_value: 0.00009999<br />info: n=1857; logLR=1574.58; p<0.001<br />Drug: Atorvastatin<br />AE: Muscular Weakness","p_value: 0.00009999<br />info: n=859; logLR=774.85; p<0.001<br />Drug: Simvastatin<br />AE: Muscular Weakness","p_value: 0.00009999<br />info: n=808; logLR=464.12; p<0.001<br />Drug: Rosuvastatin<br />AE: Muscular Weakness","p_value: 0.00009999<br />info: n=152; logLR=89.99; p<0.001<br />Drug: Pravastatin<br />AE: Muscular Weakness","p_value: 0.00009999<br />info: n=45; logLR=31.66; p<0.001<br />Drug: Fluvastatin<br />AE: Muscular Weakness","p_value: 0.00019998<br />info: n=31; logLR=26.79; p<0.001<br />Drug: Lovastatin<br />AE: Muscular Weakness"],["p_value: 0.00009999<br />info: n=459; logLR=1869.37; p<0.001<br />Drug: Atorvastatin<br />AE: Immune-Mediated Necrotising Myopathy","p_value: 0.00009999<br />info: n=113; logLR=396.98; p<0.001<br />Drug: Simvastatin<br />AE: Immune-Mediated Necrotising Myopathy","p_value: 0.00009999<br />info: n=44; logLR=98.55; p<0.001<br />Drug: Rosuvastatin<br />AE: Immune-Mediated Necrotising Myopathy","p_value: 0.00009999<br />info: n=20; logLR=62.44; p<0.001<br />Drug: Pravastatin<br />AE: Immune-Mediated Necrotising Myopathy","p_value: 1.00000000<br />info: n=0; logLR=0; p>0.90<br />Drug: Fluvastatin<br />AE: Immune-Mediated Necrotising Myopathy","p_value: 1.00000000<br />info: n=2; logLR=5.53; p>0.90<br />Drug: Lovastatin<br />AE: Immune-Mediated Necrotising Myopathy"],["p_value: 1.00000000<br />info: n=344; logLR=0; p>0.90<br />Drug: Atorvastatin<br />AE: Intentional Product Misuse","p_value: 1.00000000<br />info: n=76; logLR=0; p>0.90<br />Drug: Simvastatin<br />AE: Intentional Product Misuse","p_value: 0.00009999<br />info: n=1754; logLR=1883.48; p<0.001<br />Drug: Rosuvastatin<br />AE: Intentional Product Misuse","p_value: 1.00000000<br />info: n=12; logLR=0; p>0.90<br />Drug: Pravastatin<br />AE: Intentional Product Misuse","p_value: 1.00000000<br />info: n=0; logLR=0; p>0.90<br />Drug: Fluvastatin<br />AE: Intentional Product Misuse","p_value: 1.00000000<br />info: n=1; logLR=0; p>0.90<br />Drug: Lovastatin<br />AE: Intentional Product Misuse"],["p_value: 0.00009999<br />info: n=1175; logLR=2031.52; p<0.001<br />Drug: Atorvastatin<br />AE: Blood Creatine Phosphokinase Increased","p_value: 0.00009999<br />info: n=768; logLR=1626.15; p<0.001<br />Drug: Simvastatin<br />AE: Blood Creatine Phosphokinase Increased","p_value: 0.00009999<br />info: n=562; logLR=837.22; p<0.001<br />Drug: Rosuvastatin<br />AE: Blood Creatine Phosphokinase Increased","p_value: 0.00009999<br />info: n=200; logLR=421.48; p<0.001<br />Drug: Pravastatin<br />AE: Blood Creatine Phosphokinase Increased","p_value: 0.00009999<br />info: n=125; logLR=371.84; p<0.001<br />Drug: Fluvastatin<br />AE: Blood Creatine Phosphokinase Increased","p_value: 0.00009999<br />info: n=32; logLR=70.73; p<0.001<br />Drug: Lovastatin<br />AE: Blood Creatine Phosphokinase Increased"],["p_value: 0.00009999<br />info: n=849; logLR=2108.56; p<0.001<br />Drug: Atorvastatin<br />AE: Myopathy","p_value: 0.00009999<br />info: n=544; logLR=1558.83; p<0.001<br />Drug: Simvastatin<br />AE: Myopathy","p_value: 0.00009999<br />info: n=377; logLR=815.01; p<0.001<br />Drug: Rosuvastatin<br />AE: Myopathy","p_value: 0.00009999<br />info: n=145; logLR=417.31; p<0.001<br />Drug: Pravastatin<br />AE: Myopathy","p_value: 0.00009999<br />info: n=64; logLR=218.64; p<0.001<br />Drug: Fluvastatin<br />AE: Myopathy","p_value: 0.00009999<br />info: n=45; logLR=163.66; p<0.001<br />Drug: Lovastatin<br />AE: Myopathy"],["p_value: 0.00009999<br />info: n=3778; logLR=2231.05; p<0.001<br />Drug: Atorvastatin<br />AE: Drug Hypersensitivity","p_value: 0.00009999<br />info: n=1306; logLR=565.09; p<0.001<br />Drug: Simvastatin<br />AE: Drug Hypersensitivity","p_value: 0.00009999<br />info: n=1009; logLR=100.4; p<0.001<br />Drug: Rosuvastatin<br />AE: Drug Hypersensitivity","p_value: 0.00009999<br />info: n=500; logLR=351.73; p<0.001<br />Drug: Pravastatin<br />AE: Drug Hypersensitivity","p_value: 0.69853015<br />info: n=56; logLR=9.65; p=0.699<br />Drug: Fluvastatin<br />AE: Drug Hypersensitivity","p_value: 0.00009999<br />info: n=210; logLR=341.41; p<0.001<br />Drug: Lovastatin<br />AE: Drug Hypersensitivity"],["p_value: 0.00009999<br />info: n=654; logLR=464.95; p<0.001<br />Drug: Atorvastatin<br />AE: Blood Cholesterol Increased","p_value: 0.00009999<br />info: n=227; logLR=123.97; p<0.001<br />Drug: Simvastatin<br />AE: Blood Cholesterol Increased","p_value: 0.00009999<br />info: n=1328; logLR=2309.81; p<0.001<br />Drug: Rosuvastatin<br />AE: Blood Cholesterol Increased","p_value: 0.00009999<br />info: n=69; logLR=45.01; p<0.001<br />Drug: Pravastatin<br />AE: Blood Cholesterol Increased","p_value: 0.02459754<br />info: n=21; logLR=16.56; p=0.025<br />Drug: Fluvastatin<br />AE: Blood Cholesterol Increased","p_value: 0.00009999<br />info: n=23; logLR=31.16; p<0.001<br />Drug: Lovastatin<br />AE: Blood Cholesterol Increased"],["p_value: 0.00009999<br />info: n=2041; logLR=3994.43; p<0.001<br />Drug: Atorvastatin<br />AE: Rhabdomyolysis","p_value: 0.00009999<br />info: n=1376; logLR=3275.56; p<0.001<br />Drug: Simvastatin<br />AE: Rhabdomyolysis","p_value: 0.00009999<br />info: n=936; logLR=1567.2; p<0.001<br />Drug: Rosuvastatin<br />AE: Rhabdomyolysis","p_value: 0.00009999<br />info: n=163; logLR=264.93; p<0.001<br />Drug: Pravastatin<br />AE: Rhabdomyolysis","p_value: 0.00009999<br />info: n=52; logLR=95.28; p<0.001<br />Drug: Fluvastatin<br />AE: Rhabdomyolysis","p_value: 0.00009999<br />info: n=44; logLR=97.66; p<0.001<br />Drug: Lovastatin<br />AE: Rhabdomyolysis"],["p_value: 0.00009999<br />info: n=5362; logLR=8026.54; p<0.001<br />Drug: Atorvastatin<br />AE: Myalgia","p_value: 0.00009999<br />info: n=3216; logLR=5782.26; p<0.001<br />Drug: Simvastatin<br />AE: Myalgia","p_value: 0.00009999<br />info: n=2757; logLR=3660.8; p<0.001<br />Drug: Rosuvastatin<br />AE: Myalgia","p_value: 0.00009999<br />info: n=939; logLR=1779.83; p<0.001<br />Drug: Pravastatin<br />AE: Myalgia","p_value: 0.00009999<br />info: n=341; logLR=760.5; p<0.001<br />Drug: Fluvastatin<br />AE: Myalgia","p_value: 0.00009999<br />info: n=151; logLR=302.31; p<0.001<br />Drug: Lovastatin<br />AE: Myalgia"],["p_value: 0.00009999<br />info: n=2.0e+04; logLR=8.1e+04; p<0.001<br />Drug: Atorvastatin<br />AE: Type 2 Diabetes Mellitus","p_value: 0.00009999<br />info: n=221; logLR=123.63; p<0.001<br />Drug: Simvastatin<br />AE: Type 2 Diabetes Mellitus","p_value: 0.00009999<br />info: n=611; logLR=661.63; p<0.001<br />Drug: Rosuvastatin<br />AE: Type 2 Diabetes Mellitus","p_value: 0.00009999<br />info: n=146; logLR=191.68; p<0.001<br />Drug: Pravastatin<br />AE: Type 2 Diabetes Mellitus","p_value: 0.03869613<br />info: n=20; logLR=15.71; p=0.039<br />Drug: Fluvastatin<br />AE: Type 2 Diabetes Mellitus","p_value: 0.00009999<br />info: n=47; logLR=96.75; p<0.001<br />Drug: Lovastatin<br />AE: Type 2 Diabetes Mellitus"]],"colorscale":[[0,"#8B0000"],[0.0001,"#8B0000"],[0.00050000000000000001,"#8B0000"],[0.012200000000000001,"#8D0703"],[0.013300000000000001,"#8D0804"],[0.024500000000000001,"#8F0F07"],[0.038600000000000002,"#91150B"],[0.19800000000000001,"#A9432F"],[0.48159999999999997,"#CF8573"],[0.69850000000000001,"#E6B7AB"],[0.76169999999999993,"#ECC6BC"],[1,"#FFFFFF"]],"type":"heatmap","showscale":false,"autocolorscale":false,"showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1],"y":[1],"name":"99_0ac207b18207b2ce934100174f14c321","type":"scatter","mode":"markers","opacity":0,"hoverinfo":"skip","showlegend":false,"marker":{"color":[0,1],"colorscale":[[0,"#8B0000"],[0.0033444816053511705,"#8C0201"],[0.006688963210702341,"#8C0402"],[0.010033444816053512,"#8D0603"],[0.013377926421404682,"#8D0804"],[0.016722408026755852,"#8E0A05"],[0.020066889632107024,"#8E0C06"],[0.023411371237458196,"#8F0E07"],[0.026755852842809364,"#8F1007"],[0.030100334448160532,"#901108"],[0.033444816053511704,"#901309"],[0.03678929765886288,"#91140A"],[0.040133779264214048,"#91160B"],[0.043478260869565216,"#92170C"],[0.046822742474916391,"#92180D"],[0.05016722408026756,"#931A0E"],[0.053511705685618728,"#941B0E"],[0.056856187290969896,"#941C0F"],[0.060200668896321065,"#951D10"],[0.06354515050167224,"#951E11"],[0.066889632107023408,"#961F11"],[0.070234113712374577,"#962012"],[0.073578595317725759,"#972213"],[0.076923076923076927,"#972314"],[0.080267558528428096,"#982415"],[0.083612040133779264,"#982515"],[0.086956521739130432,"#992616"],[0.090301003344481601,"#992717"],[0.093645484949832783,"#9A2818"],[0.096989966555183937,"#9A2918"],[0.10033444816053512,"#9B2A19"],[0.10367892976588629,"#9B2B1A"],[0.10702341137123746,"#9C2B1B"],[0.11036789297658864,"#9C2C1B"],[0.11371237458193979,"#9D2D1C"],[0.11705685618729098,"#9D2E1D"],[0.12040133779264213,"#9E2F1E"],[0.12374581939799331,"#9E301F"],[0.12709030100334448,"#9F311F"],[0.13043478260869565,"#9F3220"],[0.13377926421404682,"#A03321"],[0.13712374581939801,"#A03422"],[0.14046822742474915,"#A13522"],[0.14381270903010032,"#A13523"],[0.14715719063545152,"#A23624"],[0.15050167224080269,"#A23725"],[0.15384615384615385,"#A33825"],[0.157190635451505,"#A33926"],[0.16053511705685619,"#A43A27"],[0.16387959866220736,"#A43B28"],[0.16722408026755853,"#A53B28"],[0.1705685618729097,"#A53C29"],[0.17391304347826086,"#A63D2A"],[0.17725752508361203,"#A63E2B"],[0.1806020066889632,"#A73F2B"],[0.1839464882943144,"#A7402C"],[0.18729096989966557,"#A8402D"],[0.19063545150501671,"#A8412E"],[0.19397993311036787,"#A9422E"],[0.19732441471571907,"#A9432F"],[0.20066889632107024,"#AA4430"],[0.20401337792642141,"#AA4431"],[0.20735785953177258,"#AB4531"],[0.21070234113712374,"#AB4632"],[0.21404682274247491,"#AC4733"],[0.21739130434782608,"#AC4834"],[0.22073578595317728,"#AD4935"],[0.22408026755852842,"#AD4935"],[0.22742474916387959,"#AE4A36"],[0.23076923076923078,"#AE4B37"],[0.23411371237458195,"#AF4C38"],[0.23745819397993312,"#AF4D38"],[0.24080267558528426,"#B04D39"],[0.24414715719063546,"#B04E3A"],[0.24749163879598662,"#B14F3B"],[0.25083612040133779,"#B1503B"],[0.25418060200668896,"#B1503C"],[0.25752508361204013,"#B2513D"],[0.2608695652173913,"#B2523E"],[0.26421404682274247,"#B3533F"],[0.26755852842809363,"#B3543F"],[0.2709030100334448,"#B45440"],[0.27424749163879603,"#B45541"],[0.27759197324414714,"#B55642"],[0.28093645484949831,"#B55742"],[0.28428093645484953,"#B65843"],[0.28762541806020064,"#B65844"],[0.29096989966555187,"#B75945"],[0.29431438127090304,"#B75A46"],[0.29765886287625415,"#B85B46"],[0.30100334448160537,"#B85B47"],[0.30434782608695649,"#B85C48"],[0.30769230769230771,"#B95D49"],[0.31103678929765888,"#B95E4A"],[0.31438127090300999,"#BA5F4A"],[0.31772575250836121,"#BA5F4B"],[0.32107023411371238,"#BB604C"],[0.32441471571906355,"#BB614D"],[0.32775919732441472,"#BC624E"],[0.33110367892976589,"#BC624E"],[0.33444816053511706,"#BC634F"],[0.33779264214046822,"#BD6450"],[0.34113712374581939,"#BD6551"],[0.34448160535117056,"#BE6552"],[0.34782608695652173,"#BE6652"],[0.3511705685618729,"#BF6753"],[0.35451505016722407,"#BF6854"],[0.35785953177257529,"#C06955"],[0.3612040133779264,"#C06956"],[0.36454849498327757,"#C06A56"],[0.3678929765886288,"#C16B57"],[0.37123745819397991,"#C16C58"],[0.37458193979933113,"#C26C59"],[0.3779264214046823,"#C26D5A"],[0.38127090301003341,"#C36E5A"],[0.38461538461538464,"#C36F5B"],[0.38795986622073575,"#C36F5C"],[0.39130434782608697,"#C4705D"],[0.39464882943143814,"#C4715E"],[0.39799331103678925,"#C5725E"],[0.40133779264214048,"#C5735F"],[0.40468227424749165,"#C67360"],[0.40802675585284282,"#C67461"],[0.41137123745819398,"#C67562"],[0.41471571906354515,"#C77663"],[0.41806020066889632,"#C77663"],[0.42140468227424749,"#C87764"],[0.42474916387959871,"#C87865"],[0.42809364548494983,"#C87966"],[0.43143812709030099,"#C97967"],[0.43478260869565216,"#C97A67"],[0.43812709030100333,"#CA7B68"],[0.44147157190635455,"#CA7C69"],[0.44481605351170567,"#CB7C6A"],[0.44816053511705684,"#CB7D6B"],[0.45150501672240806,"#CB7E6C"],[0.45484949832775917,"#CC7F6C"],[0.4581939799331104,"#CC806D"],[0.46153846153846156,"#CD806E"],[0.46488294314381268,"#CD816F"],[0.4682274247491639,"#CD8270"],[0.47157190635451507,"#CE8371"],[0.47491638795986624,"#CE8371"],[0.47826086956521741,"#CF8472"],[0.48160535117056852,"#CF8573"],[0.48494983277591974,"#CF8674"],[0.48829431438127091,"#D08675"],[0.49163879598662208,"#D08776"],[0.49498327759197325,"#D18877"],[0.49832775919732442,"#D18977"],[0.50167224080267558,"#D18A78"],[0.50501672240802675,"#D28A79"],[0.50836120401337792,"#D28B7A"],[0.51170568561872909,"#D38C7B"],[0.51505016722408026,"#D38D7C"],[0.51839464882943143,"#D38D7C"],[0.52173913043478259,"#D48E7D"],[0.52508361204013376,"#D48F7E"],[0.52842809364548493,"#D4907F"],[0.5317725752508361,"#D59080"],[0.53511705685618727,"#D59181"],[0.53846153846153855,"#D69282"],[0.5418060200668896,"#D69382"],[0.54515050167224077,"#D69483"],[0.54849498327759205,"#D79484"],[0.55183946488294311,"#D79585"],[0.55518394648829428,"#D79686"],[0.55852842809364556,"#D89787"],[0.56187290969899661,"#D89788"],[0.56521739130434778,"#D99888"],[0.56856187290969906,"#D99989"],[0.57190635451505023,"#D99A8A"],[0.57525083612040129,"#DA9A8B"],[0.57859531772575257,"#DA9B8C"],[0.58193979933110374,"#DA9C8D"],[0.58528428093645479,"#DB9D8E"],[0.58862876254180607,"#DB9E8E"],[0.59197324414715724,"#DB9E8F"],[0.5953177257525083,"#DC9F90"],[0.59866220735785947,"#DCA091"],[0.60200668896321075,"#DDA192"],[0.60535117056856191,"#DDA193"],[0.60869565217391297,"#DDA294"],[0.61204013377926425,"#DEA395"],[0.61538461538461542,"#DEA495"],[0.61872909698996648,"#DEA596"],[0.62207357859531776,"#DFA597"],[0.62541806020066892,"#DFA698"],[0.62876254180601998,"#DFA799"],[0.63210702341137126,"#E0A89A"],[0.63545150501672243,"#E0A89B"],[0.6387959866220736,"#E0A99C"],[0.64214046822742477,"#E1AA9C"],[0.64548494983277593,"#E1AB9D"],[0.6488294314381271,"#E1AB9E"],[0.65217391304347827,"#E2AC9F"],[0.65551839464882944,"#E2ADA0"],[0.65886287625418061,"#E2AEA1"],[0.66220735785953178,"#E3AFA2"],[0.66555183946488294,"#E3AFA3"],[0.66889632107023411,"#E4B0A4"],[0.67224080267558539,"#E4B1A4"],[0.67558528428093645,"#E4B2A5"],[0.67892976588628762,"#E5B3A6"],[0.68227424749163879,"#E5B3A7"],[0.68561872909698995,"#E5B4A8"],[0.68896321070234112,"#E5B5A9"],[0.69230769230769229,"#E6B6AA"],[0.69565217391304346,"#E6B6AB"],[0.69899665551839463,"#E6B7AC"],[0.7023411371237458,"#E7B8AC"],[0.70568561872909707,"#E7B9AD"],[0.70903010033444813,"#E7BAAE"],[0.7123745819397993,"#E8BAAF"],[0.71571906354515058,"#E8BBB0"],[0.71906354515050164,"#E8BCB1"],[0.72240802675585281,"#E9BDB2"],[0.72575250836120409,"#E9BDB3"],[0.72909698996655514,"#E9BEB4"],[0.73244147157190631,"#EABFB5"],[0.73578595317725759,"#EAC0B5"],[0.73913043478260876,"#EAC1B6"],[0.74247491638795982,"#EBC1B7"],[0.7458193979933111,"#EBC2B8"],[0.74916387959866226,"#EBC3B9"],[0.75250836120401332,"#ECC4BA"],[0.7558528428093646,"#ECC5BB"],[0.75919732441471577,"#ECC5BC"],[0.76254180602006683,"#ECC6BD"],[0.76588628762541811,"#EDC7BE"],[0.76923076923076927,"#EDC8BE"],[0.77257525083612044,"#EDC9BF"],[0.7759197324414715,"#EEC9C0"],[0.77926421404682278,"#EECAC1"],[0.78260869565217395,"#EECBC2"],[0.785953177257525,"#EFCCC3"],[0.78929765886287628,"#EFCCC4"],[0.79264214046822745,"#EFCDC5"],[0.79598662207357851,"#EFCEC6"],[0.79933110367892979,"#F0CFC7"],[0.80267558528428096,"#F0D0C8"],[0.80602006688963213,"#F0D0C9"],[0.80936454849498329,"#F1D1C9"],[0.81270903010033446,"#F1D2CA"],[0.81605351170568563,"#F1D3CB"],[0.8193979933110368,"#F1D4CC"],[0.82274247491638797,"#F2D4CD"],[0.82608695652173914,"#F2D5CE"],[0.8294314381270903,"#F2D6CF"],[0.83277591973244147,"#F3D7D0"],[0.83612040133779264,"#F3D8D1"],[0.83946488294314392,"#F3D8D2"],[0.84280936454849498,"#F3D9D3"],[0.84615384615384615,"#F4DAD4"],[0.84949832775919742,"#F4DBD5"],[0.85284280936454848,"#F4DCD5"],[0.85618729096989965,"#F4DCD6"],[0.85953177257525093,"#F5DDD7"],[0.86287625418060199,"#F5DED8"],[0.86622073578595316,"#F5DFD9"],[0.86956521739130432,"#F6E0DA"],[0.87290969899665549,"#F6E0DB"],[0.87625418060200666,"#F6E1DC"],[0.87959866220735783,"#F6E2DD"],[0.88294314381270911,"#F7E3DE"],[0.88628762541806017,"#F7E4DF"],[0.88963210702341133,"#F7E4E0"],[0.89297658862876261,"#F7E5E1"],[0.89632107023411367,"#F8E6E2"],[0.89966555183946484,"#F8E7E3"],[0.90301003344481612,"#F8E8E3"],[0.90635451505016718,"#F8E8E4"],[0.90969899665551834,"#F9E9E5"],[0.91304347826086962,"#F9EAE6"],[0.91638795986622079,"#F9EBE7"],[0.91973244147157185,"#F9ECE8"],[0.92307692307692313,"#FAECE9"],[0.9264214046822743,"#FAEDEA"],[0.92976588628762535,"#FAEEEB"],[0.93311036789297663,"#FAEFEC"],[0.9364548494983278,"#FBF0ED"],[0.93979933110367886,"#FBF0EE"],[0.94314381270903014,"#FBF1EF"],[0.94648829431438131,"#FBF2F0"],[0.94983277591973247,"#FCF3F1"],[0.95317725752508364,"#FCF4F2"],[0.95652173913043481,"#FCF4F3"],[0.95986622073578598,"#FCF5F4"],[0.96321070234113704,"#FDF6F4"],[0.96655518394648832,"#FDF7F5"],[0.96989966555183948,"#FDF8F6"],[0.97324414715719054,"#FDF9F7"],[0.97658862876254182,"#FDF9F8"],[0.97993311036789299,"#FEFAF9"],[0.98327759197324416,"#FEFBFA"],[0.98662207357859533,"#FEFCFB"],[0.98996655518394649,"#FEFDFC"],[0.99331103678929766,"#FFFDFD"],[0.99665551839464883,"#FFFEFE"],[1,"#FFFFFF"]],"colorbar":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.8897637795275593,"thickness":23.039999999999996,"title":"p_value","titlefont":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724},"tickmode":"array","ticktext":["0.25","0.50","0.75","1.00"],"tickvals":[0.24992500000000001,0.49995000000000001,0.74997499999999995,1],"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":11.68949771689498},"ticklen":2,"len":0.5}},"xaxis":"x","yaxis":"y","frame":null}],"layout":{"margin":{"t":30.611872146118724,"r":7.3059360730593621,"b":88.401826484018301,"l":233.05936073059365},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.40000000000000002,6.5999999999999996],"tickmode":"array","ticktext":["Atorvastatin","Simvastatin","Rosuvastatin","Pravastatin","Fluvastatin","Lovastatin"],"tickvals":[1,2,3,3.9999999999999996,5,6],"categoryorder":"array","categoryarray":["Atorvastatin","Simvastatin","Rosuvastatin","Pravastatin","Fluvastatin","Lovastatin"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.6529680365296811,"tickwidth":0.66417600664176002,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.68949771689498},"tickangle":-90,"showline":false,"linecolor":null,"linewidth":0,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.40000000000000002,15.6],"tickmode":"array","ticktext":["Blood Triglycerides Increased","Immune-Mediated Myositis","Necrotising Myositis","Muscle Spasms","Drug Interaction","Muscular Weakness","Immune-Mediated Necrotising Myopathy","Intentional Product Misuse","Blood Creatine Phosphokinase Increased","Myopathy","Drug Hypersensitivity","Blood Cholesterol Increased","Rhabdomyolysis","Myalgia","Type 2 Diabetes Mellitus"],"tickvals":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15],"categoryorder":"array","categoryarray":["Blood Triglycerides Increased","Immune-Mediated Myositis","Necrotising Myositis","Muscle Spasms","Drug Interaction","Muscular Weakness","Immune-Mediated Necrotising Myopathy","Intentional Product Misuse","Blood Creatine Phosphokinase Increased","Myopathy","Drug Hypersensitivity","Blood Cholesterol Increased","Rhabdomyolysis","Myalgia","Type 2 Diabetes Mellitus"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.6529680365296811,"tickwidth":0.66417600664176002,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.68949771689498},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.8897637795275593,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.68949771689498},"title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724}}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"source":"A","attrs":{"dd4ba0e12bd":{"fill":{},"label":{},"x":{},"y":{},"type":"heatmap"}},"cur_data":"dd4ba0e12bd","visdat":{"dd4ba0e12bd":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.20000000000000001,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
<p>The following observations are made. First, (Type 2 Diabetes Mellitus, Atorvastatin) appears to be the <em>most prominent</em> association in terms of LR statistic (log LR = 8.0632317^{4}) across all AE/drug pairs. The next four prominent signals are observed in (Myalgia, Atorvastatin), (Myalgia, Simvastatin), (Rhabdomyolysis, Atorvastatin), and (Myalgia, Rosuvastatin) respectively.</p>
<h2 class="unnumbered" id="analysis-of-the-full-gbca-data-with-1707-adverse-events">Analysis of the full <code>gbca</code> data with 1707 adverse events</h2>
<p>We next analyze the full <code>gbca</code> dataset with all 1707 adverse events and 9 drugs. This example arises in the context of an FDA evaluation of Gadolinium-based contrast agents (GBCA) from which FDA concluded there were no adverse events from GBCA retention (<a href="https://www.fda.gov/drugs/postmarket-drug-safety-information-patients-and-providers/information-gadolinium-based-contrast-agents" class="uri">https://www.fda.gov/drugs/postmarket-drug-safety-information-patients-and-providers/information-gadolinium-based-contrast-agents</a>). Several disproportionality methods were compared in the FAERS database <span class="citation" data-cites="zhao2018">(<a href="#ref-zhao2018" role="doc-biblioref">Zhao et al. 2018</a>)</span> using the GBCAs as an example. We use this example to further illustrate the LRT methods, and assess their computational efficiency in datasets with many zero-report counts. The observed proportions of zero counts are noticeably large in most drugs:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"gbca"</span><span class='op'>)</span></span>
<span><span class='co'># head(gbca) # show the first few rows</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>obs_prop_0</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='va'>gbca</span>, <span class='fl'>2</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/mean.html'>mean</a></span><span class='op'>(</span><span class='va'>x</span> <span class='op'>==</span> <span class='fl'>0</span><span class='op'>)</span><span class='op'>)</span></span>
<span><span class='fu'><a href='https://rdrr.io/r/base/Round.html'>round</a></span><span class='op'>(</span><span class='va'>obs_prop_0</span>, <span class='fl'>2</span><span class='op'>)</span></span></code></pre>
</div>
<pre><code>     Gadobenate      Gadobutrol     Gadodiamide    Gadofosveset 
           0.60            0.42            0.71            0.99 
  Gadopentetate      Gadoterate     Gadoteridol Gadoversetamide 
           0.61            0.51            0.70            0.87 
     Gadoxetate           Other 
           0.82            0.00 </code></pre>
</div>
<p>These large proportions of zero counts provide a heuristic justification for using a ZIP-model based LRT in this example: because the total zero counts comprise both <em>true/structural</em> zero counts (zero-inflation) and <em>sampling</em> zero counts, some substantial parts of these observed zeros may be attributable to zero inflation. To formalize this intuition, in the following we first run a Poisson LRT, followed by a ZIP LRT, and then compare goodness of fits for the two models in the dataset. The model that fits the data better (assessed through BIC/AIC) will be subsequently used for inference. We begin with the Poisson LRT, with the default 10,000 resamples for computation of <span class="math inline">\(p\)</span>-values. The codes are as follows. The resulting output, when printed on console will display textual summary of test results which we omit for brevity.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>100</span><span class='op'>)</span></span>
<span><span class='va'>test_gbca_poisson</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/pvLRT/man/pvlrt.html'>pvlrt</a></span><span class='op'>(</span></span>
<span>  <span class='va'>gbca</span>,</span>
<span>  zi_prob <span class='op'>=</span> <span class='fl'>0</span>,</span>
<span>  test_drug_idx <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>9</span></span>
<span><span class='op'>)</span></span>
<span><span class='co'># test_gbca_poisson ## print results</span></span></code></pre>
</div>
</div>
<!-- ```{r print-poisson-gbca} -->
<!-- ## print results -->
<!-- # test_gbca_poisson -->
<!-- ``` -->
<p>Next we run the zero-inflated Poisson model. Codes are as follows.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>100</span><span class='op'>)</span></span>
<span><span class='va'>test_gbca_zip</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/pvLRT/man/pvlrt.html'>pvlrt</a></span><span class='op'>(</span></span>
<span>  <span class='va'>gbca</span>,</span>
<span>  test_drug_idx <span class='op'>=</span> <span class='fl'>1</span><span class='op'>:</span><span class='fl'>9</span>,</span>
<span>  zi_prob <span class='op'>=</span> <span class='cn'>NULL</span>,</span>
<span>  test_zi <span class='op'>=</span> <span class='cn'>TRUE</span></span>
<span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># print results</span></span>
<span><span class='va'>test_gbca_zip</span></span></code></pre>
</div>
<pre><code>Relative reporting rate (lambda)-based pseudo-LRT on 1707 AE &amp;
10 drugs. Hypothesis tests performed on 9 drugs using
parametric bootstrap.

Running time of the original pvlrt call: 1.048432 mins

Drug-specific estimated zi probabilities: 0.39 (Gadobenate),
0.2 (Gadobutrol), 0.52 (Gadodiamide), 0.44 (Gadofosveset),
0.41 (Gadopentetate), 0.26 (Gadoterate), 0.32 (Gadoteridol),
0.76 (Gadoversetamide), 0.66 (Gadoxetate), 0 (Other)

LR test of significance for zi probabilities: LRstat=2318.57,
q&lt;0.001 (Gadobenate), LRstat=1377.85, q&lt;0.001 (Gadobutrol),
LRstat=2671.46, q&lt;0.001 (Gadodiamide), LRstat= 4.23, q=0.13
(Gadofosveset), LRstat=2112.43, q&lt;0.001 (Gadopentetate),
LRstat=1160.36, q&lt;0.001 (Gadoterate), LRstat= 433.91, q&lt;0.001
(Gadoteridol), LRstat=3671.24, q&lt;0.001 (Gadoversetamide),
LRstat=2284.60, q&lt;0.001 (Gadoxetate), LRstat= 0.00, q=&lt;NA&gt;
(Other)
Total 203 AE-drug pairs are significant at level = 0.05.

Extract all LR statistics and p-values using `summary()`.</code></pre>
</div>
<p>It follows that several drug groups have substantially large (&gt; 40%) zero inflation probabilities with statistically significant FDR adjusted <span class="math inline">\(p\)</span>-values (<span class="math inline">\(q\)</span>-values). This agrees with the earlier exploratory observation of large proportions of observed zeros. Similar to the full <code>statin</code> data analysis, it therefore seems that the zero-inflated Poisson model would be more appropriate here. We use AIC and BIC to formally check this:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/cbind.html'>cbind</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/stats/AIC.html'>BIC</a></span><span class='op'>(</span><span class='va'>test_gbca_poisson</span>, <span class='va'>test_gbca_zip</span><span class='op'>)</span>,</span>
<span>      <span class='fu'><a href='https://rdrr.io/r/stats/AIC.html'>AIC</a></span><span class='op'>(</span><span class='va'>test_gbca_poisson</span>, <span class='va'>test_gbca_zip</span><span class='op'>)</span><span class='op'>)</span><span class='op'>[</span>, <span class='op'>-</span><span class='fl'>3</span><span class='op'>]</span> <span class='co'># remove </span></span></code></pre>
</div>
<pre><code>                     df      BIC      AIC
test_gbca_poisson 17070 416208.7 147441.5
test_gbca_zip     17080 384316.9 115392.2</code></pre>
</div>
<p>Indeed, both AIC and BIC are smaller for the zero-inflated Poisson model than the Poisson model. We therefore use the former model for subsequent analysis.</p>
<p>To get a visual understanding of the results we plot a heatmap of the <span class="math inline">\(p\)</span>-values with LR test statistics and samples sizes inscribed as texts:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>pl_heat_gbca_zip</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/pvLRT/man/heatmap_pvlrt.html'>heatmap_pvlrt</a></span><span class='op'>(</span></span>
<span>  <span class='va'>test_gbca_zip</span>,</span>
<span>  AE <span class='op'>=</span> <span class='fl'>15</span>,</span>
<span>  show_n <span class='op'>=</span> <span class='cn'>FALSE</span>,</span>
<span>  show_lrstat <span class='op'>=</span> <span class='cn'>TRUE</span></span>
<span><span class='op'>)</span> </span>
<span><span class='fu'>plotly</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/plotly/man/ggplotly.html'>ggplotly</a></span><span class='op'>(</span><span class='va'>pl_heat_gbca_zip</span><span class='op'>)</span></span></code></pre>
</div>
<div class="plotly html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-b38fe8872c08a37b39df" style="width:624px;height:384px;"></div>
<script type="application/json" data-for="htmlwidget-b38fe8872c08a37b39df">{"x":{"data":[{"x":[1,2,3,4,5,6,7,8,9],"y":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15],"z":[[0,0,0,0,0,1,1,1,1],[0,0,0,0,0,1,1,1,1],[0,0,0,0,0,1,1,1,1],[0,0,0,0,0,0,0.0011000000000000001,0.99390000000000001,1],[0,0,0,0,0,1,1,1,1],[0,0.95010000000000006,1,0,1,0,0,0,1],[0,0,0,0,0,1,1,1,1],[0,0,0,0,0,1,1,1,1],[0,0,0,0,0,1,1,1,1],[0,0,0,0,0,0,0.0024000000000000002,1,1],[0,0,0,0,0,1,1,1,1],[1,1,1,0,1,0,0,0.99980000000000002,1],[0,0,0,0,1,0,0,0,1],[0,0,0,0,0,0,0,0.16720000000000002,1],[0,0,0,0,0,0,0.98849999999999993,0,1]],"text":[["p_value: 0.00009999<br />info: ; logLR=453.48<br />Drug: Gadopentetate<br />AE: Emotional Distress","p_value: 0.00009999<br />info: ; logLR=467.63<br />Drug: Gadodiamide<br />AE: Emotional Distress","p_value: 0.00009999<br />info: ; logLR=512.38<br />Drug: Gadoversetamide<br />AE: Emotional Distress","p_value: 0.00009999<br />info: ; logLR=434.33<br />Drug: Gadobenate<br />AE: Emotional Distress","p_value: 0.00009999<br />info: ; logLR=446.08<br />Drug: Gadoxetate<br />AE: Emotional Distress","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadobutrol<br />AE: Emotional Distress","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadoterate<br />AE: Emotional Distress","p_value: 1.00000000<br />info: ; logLR=1.02<br />Drug: Gadoteridol<br />AE: Emotional Distress","p_value: 1.00000000<br />info: ; logLR=8.48<br />Drug: Gadofosveset<br />AE: Emotional Distress"],["p_value: 0.00009999<br />info: ; logLR=467.88<br />Drug: Gadopentetate<br />AE: Confusional State","p_value: 0.00009999<br />info: ; logLR=470.12<br />Drug: Gadodiamide<br />AE: Confusional State","p_value: 0.00009999<br />info: ; logLR=519.95<br />Drug: Gadoversetamide<br />AE: Confusional State","p_value: 0.00009999<br />info: ; logLR=442.1<br />Drug: Gadobenate<br />AE: Confusional State","p_value: 0.00009999<br />info: ; logLR=544.49<br />Drug: Gadoxetate<br />AE: Confusional State","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadobutrol<br />AE: Confusional State","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadoterate<br />AE: Confusional State","p_value: 1.00000000<br />info: ; logLR=12.98<br />Drug: Gadoteridol<br />AE: Confusional State","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadofosveset<br />AE: Confusional State"],["p_value: 0.00009999<br />info: ; logLR=546.46<br />Drug: Gadopentetate<br />AE: Hypoaesthesia","p_value: 0.00009999<br />info: ; logLR=483.29<br />Drug: Gadodiamide<br />AE: Hypoaesthesia","p_value: 0.00009999<br />info: ; logLR=567.05<br />Drug: Gadoversetamide<br />AE: Hypoaesthesia","p_value: 0.00009999<br />info: ; logLR=487.93<br />Drug: Gadobenate<br />AE: Hypoaesthesia","p_value: 0.00009999<br />info: ; logLR=569.61<br />Drug: Gadoxetate<br />AE: Hypoaesthesia","p_value: 1.00000000<br />info: ; logLR=23.42<br />Drug: Gadobutrol<br />AE: Hypoaesthesia","p_value: 1.00000000<br />info: ; logLR=24.14<br />Drug: Gadoterate<br />AE: Hypoaesthesia","p_value: 1.00000000<br />info: ; logLR=19.86<br />Drug: Gadoteridol<br />AE: Hypoaesthesia","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadofosveset<br />AE: Hypoaesthesia"],["p_value: 0.00009999<br />info: ; logLR=589.66<br />Drug: Gadopentetate<br />AE: Paraesthesia","p_value: 0.00009999<br />info: ; logLR=536.87<br />Drug: Gadodiamide<br />AE: Paraesthesia","p_value: 0.00009999<br />info: ; logLR=566.24<br />Drug: Gadoversetamide<br />AE: Paraesthesia","p_value: 0.00009999<br />info: ; logLR=531.79<br />Drug: Gadobenate<br />AE: Paraesthesia","p_value: 0.00009999<br />info: ; logLR=566.1<br />Drug: Gadoxetate<br />AE: Paraesthesia","p_value: 0.00009999<br />info: ; logLR=86.65<br />Drug: Gadobutrol<br />AE: Paraesthesia","p_value: 0.00119988<br />info: ; logLR=74.19<br />Drug: Gadoterate<br />AE: Paraesthesia","p_value: 0.99390061<br />info: ; logLR=31.57<br />Drug: Gadoteridol<br />AE: Paraesthesia","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadofosveset<br />AE: Paraesthesia"],["p_value: 0.00009999<br />info: ; logLR=612.51<br />Drug: Gadopentetate<br />AE: Mental Impairment","p_value: 0.00009999<br />info: ; logLR=651.61<br />Drug: Gadodiamide<br />AE: Mental Impairment","p_value: 0.00009999<br />info: ; logLR=702.65<br />Drug: Gadoversetamide<br />AE: Mental Impairment","p_value: 0.00009999<br />info: ; logLR=603.85<br />Drug: Gadobenate<br />AE: Mental Impairment","p_value: 0.00009999<br />info: ; logLR=700.82<br />Drug: Gadoxetate<br />AE: Mental Impairment","p_value: 1.00000000<br />info: ; logLR=4.01<br />Drug: Gadobutrol<br />AE: Mental Impairment","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadoterate<br />AE: Mental Impairment","p_value: 1.00000000<br />info: ; logLR=0.09<br />Drug: Gadoteridol<br />AE: Mental Impairment","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadofosveset<br />AE: Mental Impairment"],["p_value: 0.00009999<br />info: ; logLR=131.16<br />Drug: Gadopentetate<br />AE: Contrast Media Reaction","p_value: 0.95010499<br />info: ; logLR=35.24<br />Drug: Gadodiamide<br />AE: Contrast Media Reaction","p_value: 1.00000000<br />info: ; logLR=11.97<br />Drug: Gadoversetamide<br />AE: Contrast Media Reaction","p_value: 0.00009999<br />info: ; logLR=349.75<br />Drug: Gadobenate<br />AE: Contrast Media Reaction","p_value: 1.00000000<br />info: ; logLR=23.33<br />Drug: Gadoxetate<br />AE: Contrast Media Reaction","p_value: 0.00009999<br />info: ; logLR=704.64<br />Drug: Gadobutrol<br />AE: Contrast Media Reaction","p_value: 0.00009999<br />info: ; logLR=220.95<br />Drug: Gadoterate<br />AE: Contrast Media Reaction","p_value: 0.00009999<br />info: ; logLR=122.48<br />Drug: Gadoteridol<br />AE: Contrast Media Reaction","p_value: 1.00000000<br />info: ; logLR=13.52<br />Drug: Gadofosveset<br />AE: Contrast Media Reaction"],["p_value: 0.00009999<br />info: ; logLR=680.87<br />Drug: Gadopentetate<br />AE: Anhedonia","p_value: 0.00009999<br />info: ; logLR=647.67<br />Drug: Gadodiamide<br />AE: Anhedonia","p_value: 0.00009999<br />info: ; logLR=713.32<br />Drug: Gadoversetamide<br />AE: Anhedonia","p_value: 0.00009999<br />info: ; logLR=626.6<br />Drug: Gadobenate<br />AE: Anhedonia","p_value: 0.00009999<br />info: ; logLR=719.54<br />Drug: Gadoxetate<br />AE: Anhedonia","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadobutrol<br />AE: Anhedonia","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadoterate<br />AE: Anhedonia","p_value: 1.00000000<br />info: ; logLR=0.14<br />Drug: Gadoteridol<br />AE: Anhedonia","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadofosveset<br />AE: Anhedonia"],["p_value: 0.00009999<br />info: ; logLR=653.4<br />Drug: Gadopentetate<br />AE: Inflammation","p_value: 0.00009999<br />info: ; logLR=675.63<br />Drug: Gadodiamide<br />AE: Inflammation","p_value: 0.00009999<br />info: ; logLR=768.7<br />Drug: Gadoversetamide<br />AE: Inflammation","p_value: 0.00009999<br />info: ; logLR=649.19<br />Drug: Gadobenate<br />AE: Inflammation","p_value: 0.00009999<br />info: ; logLR=781.34<br />Drug: Gadoxetate<br />AE: Inflammation","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadobutrol<br />AE: Inflammation","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadoterate<br />AE: Inflammation","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadoteridol<br />AE: Inflammation","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadofosveset<br />AE: Inflammation"],["p_value: 0.00009999<br />info: ; logLR=806.14<br />Drug: Gadopentetate<br />AE: Injury","p_value: 0.00009999<br />info: ; logLR=743.96<br />Drug: Gadodiamide<br />AE: Injury","p_value: 0.00009999<br />info: ; logLR=793.67<br />Drug: Gadoversetamide<br />AE: Injury","p_value: 0.00009999<br />info: ; logLR=749.18<br />Drug: Gadobenate<br />AE: Injury","p_value: 0.00009999<br />info: ; logLR=799.49<br />Drug: Gadoxetate<br />AE: Injury","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadobutrol<br />AE: Injury","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadoterate<br />AE: Injury","p_value: 1.00000000<br />info: ; logLR=4.92<br />Drug: Gadoteridol<br />AE: Injury","p_value: 1.00000000<br />info: ; logLR=9.67<br />Drug: Gadofosveset<br />AE: Injury"],["p_value: 0.00009999<br />info: ; logLR=799.32<br />Drug: Gadopentetate<br />AE: Burning Sensation","p_value: 0.00009999<br />info: ; logLR=772.19<br />Drug: Gadodiamide<br />AE: Burning Sensation","p_value: 0.00009999<br />info: ; logLR=862.49<br />Drug: Gadoversetamide<br />AE: Burning Sensation","p_value: 0.00009999<br />info: ; logLR=838.75<br />Drug: Gadobenate<br />AE: Burning Sensation","p_value: 0.00009999<br />info: ; logLR=810.67<br />Drug: Gadoxetate<br />AE: Burning Sensation","p_value: 0.00009999<br />info: ; logLR=85.87<br />Drug: Gadobutrol<br />AE: Burning Sensation","p_value: 0.00249975<br />info: ; logLR=71.4<br />Drug: Gadoterate<br />AE: Burning Sensation","p_value: 1.00000000<br />info: ; logLR=20.36<br />Drug: Gadoteridol<br />AE: Burning Sensation","p_value: 1.00000000<br />info: ; logLR=18.82<br />Drug: Gadofosveset<br />AE: Burning Sensation"],["p_value: 0.00009999<br />info: ; logLR=1019.73<br />Drug: Gadopentetate<br />AE: Mass","p_value: 0.00009999<br />info: ; logLR=1016.37<br />Drug: Gadodiamide<br />AE: Mass","p_value: 0.00009999<br />info: ; logLR=1123.06<br />Drug: Gadoversetamide<br />AE: Mass","p_value: 0.00009999<br />info: ; logLR=1007.39<br />Drug: Gadobenate<br />AE: Mass","p_value: 0.00009999<br />info: ; logLR=1128.29<br />Drug: Gadoxetate<br />AE: Mass","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadobutrol<br />AE: Mass","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadoterate<br />AE: Mass","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadoteridol<br />AE: Mass","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadofosveset<br />AE: Mass"],["p_value: 1.00000000<br />info: ; logLR=19.39<br />Drug: Gadopentetate<br />AE: Urticaria","p_value: 1.00000000<br />info: ; logLR=3.88<br />Drug: Gadodiamide<br />AE: Urticaria","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadoversetamide<br />AE: Urticaria","p_value: 0.00009999<br />info: ; logLR=160.04<br />Drug: Gadobenate<br />AE: Urticaria","p_value: 1.00000000<br />info: ; logLR=0.09<br />Drug: Gadoxetate<br />AE: Urticaria","p_value: 0.00009999<br />info: ; logLR=1393.05<br />Drug: Gadobutrol<br />AE: Urticaria","p_value: 0.00009999<br />info: ; logLR=677.36<br />Drug: Gadoterate<br />AE: Urticaria","p_value: 0.99980002<br />info: ; logLR=27.46<br />Drug: Gadoteridol<br />AE: Urticaria","p_value: 1.00000000<br />info: ; logLR=1.4<br />Drug: Gadofosveset<br />AE: Urticaria"],["p_value: 0.00009999<br />info: ; logLR=417.2<br />Drug: Gadopentetate<br />AE: Nephrogenic Systemic Fibrosis","p_value: 0.00009999<br />info: ; logLR=1507.28<br />Drug: Gadodiamide<br />AE: Nephrogenic Systemic Fibrosis","p_value: 0.00009999<br />info: ; logLR=135.34<br />Drug: Gadoversetamide<br />AE: Nephrogenic Systemic Fibrosis","p_value: 0.00009999<br />info: ; logLR=165.75<br />Drug: Gadobenate<br />AE: Nephrogenic Systemic Fibrosis","p_value: 1.00000000<br />info: ; logLR=1.64<br />Drug: Gadoxetate<br />AE: Nephrogenic Systemic Fibrosis","p_value: 0.00009999<br />info: ; logLR=101.59<br />Drug: Gadobutrol<br />AE: Nephrogenic Systemic Fibrosis","p_value: 0.00009999<br />info: ; logLR=203.69<br />Drug: Gadoterate<br />AE: Nephrogenic Systemic Fibrosis","p_value: 0.00009999<br />info: ; logLR=237.56<br />Drug: Gadoteridol<br />AE: Nephrogenic Systemic Fibrosis","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadofosveset<br />AE: Nephrogenic Systemic Fibrosis"],["p_value: 0.00009999<br />info: ; logLR=2053.62<br />Drug: Gadopentetate<br />AE: Gadolinium Deposition Disease","p_value: 0.00009999<br />info: ; logLR=2013.07<br />Drug: Gadodiamide<br />AE: Gadolinium Deposition Disease","p_value: 0.00009999<br />info: ; logLR=2007.27<br />Drug: Gadoversetamide<br />AE: Gadolinium Deposition Disease","p_value: 0.00009999<br />info: ; logLR=2000.49<br />Drug: Gadobenate<br />AE: Gadolinium Deposition Disease","p_value: 0.00009999<br />info: ; logLR=1786.61<br />Drug: Gadoxetate<br />AE: Gadolinium Deposition Disease","p_value: 0.00009999<br />info: ; logLR=250.48<br />Drug: Gadobutrol<br />AE: Gadolinium Deposition Disease","p_value: 0.00009999<br />info: ; logLR=102.79<br />Drug: Gadoterate<br />AE: Gadolinium Deposition Disease","p_value: 0.16728327<br />info: ; logLR=51.68<br />Drug: Gadoteridol<br />AE: Gadolinium Deposition Disease","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadofosveset<br />AE: Gadolinium Deposition Disease"],["p_value: 0.00009999<br />info: ; logLR=2084.09<br />Drug: Gadopentetate<br />AE: Contrast Media Deposition","p_value: 0.00009999<br />info: ; logLR=1525.55<br />Drug: Gadodiamide<br />AE: Contrast Media Deposition","p_value: 0.00009999<br />info: ; logLR=1332.18<br />Drug: Gadoversetamide<br />AE: Contrast Media Deposition","p_value: 0.00009999<br />info: ; logLR=1534.29<br />Drug: Gadobenate<br />AE: Contrast Media Deposition","p_value: 0.00009999<br />info: ; logLR=1319.67<br />Drug: Gadoxetate<br />AE: Contrast Media Deposition","p_value: 0.00009999<br />info: ; logLR=252.72<br />Drug: Gadobutrol<br />AE: Contrast Media Deposition","p_value: 0.98850115<br />info: ; logLR=32.47<br />Drug: Gadoterate<br />AE: Contrast Media Deposition","p_value: 0.00009999<br />info: ; logLR=166.61<br />Drug: Gadoteridol<br />AE: Contrast Media Deposition","p_value: 1.00000000<br />info: ; logLR=0<br />Drug: Gadofosveset<br />AE: Contrast Media Deposition"]],"colorscale":[[0,"#8B0000"],[0.0011000000000000001,"#8B0100"],[0.0024000000000000002,"#8B0101"],[0.16720000000000002,"#A53B28"],[0.95010000000000006,"#FCF3F1"],[0.98849999999999993,"#FEFCFC"],[0.99390000000000001,"#FFFEFD"],[0.99980000000000002,"#FFFFFF"],[1,"#FFFFFF"]],"type":"heatmap","showscale":false,"autocolorscale":false,"showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"showlegend":false,"xaxis":"x","yaxis":"y","hoverinfo":"text","frame":null},{"x":[1],"y":[1],"name":"99_0ac207b18207b2ce934100174f14c321","type":"scatter","mode":"markers","opacity":0,"hoverinfo":"skip","showlegend":false,"marker":{"color":[0,1],"colorscale":[[0,"#8B0000"],[0.0033444816053511705,"#8C0201"],[0.006688963210702341,"#8C0402"],[0.010033444816053512,"#8D0603"],[0.013377926421404682,"#8D0804"],[0.016722408026755852,"#8E0A05"],[0.020066889632107024,"#8E0C06"],[0.023411371237458196,"#8F0E07"],[0.026755852842809364,"#8F1007"],[0.030100334448160532,"#901108"],[0.033444816053511704,"#901309"],[0.03678929765886288,"#91140A"],[0.040133779264214048,"#91160B"],[0.043478260869565216,"#92170C"],[0.046822742474916391,"#92180D"],[0.05016722408026756,"#931A0E"],[0.053511705685618728,"#941B0E"],[0.056856187290969896,"#941C0F"],[0.060200668896321065,"#951D10"],[0.06354515050167224,"#951E11"],[0.066889632107023408,"#961F11"],[0.070234113712374577,"#962012"],[0.073578595317725759,"#972213"],[0.076923076923076927,"#972314"],[0.080267558528428096,"#982415"],[0.083612040133779264,"#982515"],[0.086956521739130432,"#992616"],[0.090301003344481601,"#992717"],[0.093645484949832783,"#9A2818"],[0.096989966555183937,"#9A2918"],[0.10033444816053512,"#9B2A19"],[0.10367892976588629,"#9B2B1A"],[0.10702341137123746,"#9C2B1B"],[0.11036789297658864,"#9C2C1B"],[0.11371237458193979,"#9D2D1C"],[0.11705685618729098,"#9D2E1D"],[0.12040133779264213,"#9E2F1E"],[0.12374581939799331,"#9E301F"],[0.12709030100334448,"#9F311F"],[0.13043478260869565,"#9F3220"],[0.13377926421404682,"#A03321"],[0.13712374581939801,"#A03422"],[0.14046822742474915,"#A13522"],[0.14381270903010032,"#A13523"],[0.14715719063545152,"#A23624"],[0.15050167224080269,"#A23725"],[0.15384615384615385,"#A33825"],[0.157190635451505,"#A33926"],[0.16053511705685619,"#A43A27"],[0.16387959866220736,"#A43B28"],[0.16722408026755853,"#A53B28"],[0.1705685618729097,"#A53C29"],[0.17391304347826086,"#A63D2A"],[0.17725752508361203,"#A63E2B"],[0.1806020066889632,"#A73F2B"],[0.1839464882943144,"#A7402C"],[0.18729096989966557,"#A8402D"],[0.19063545150501671,"#A8412E"],[0.19397993311036787,"#A9422E"],[0.19732441471571907,"#A9432F"],[0.20066889632107024,"#AA4430"],[0.20401337792642141,"#AA4431"],[0.20735785953177258,"#AB4531"],[0.21070234113712374,"#AB4632"],[0.21404682274247491,"#AC4733"],[0.21739130434782608,"#AC4834"],[0.22073578595317728,"#AD4935"],[0.22408026755852842,"#AD4935"],[0.22742474916387959,"#AE4A36"],[0.23076923076923078,"#AE4B37"],[0.23411371237458195,"#AF4C38"],[0.23745819397993312,"#AF4D38"],[0.24080267558528426,"#B04D39"],[0.24414715719063546,"#B04E3A"],[0.24749163879598662,"#B14F3B"],[0.25083612040133779,"#B1503B"],[0.25418060200668896,"#B1503C"],[0.25752508361204013,"#B2513D"],[0.2608695652173913,"#B2523E"],[0.26421404682274247,"#B3533F"],[0.26755852842809363,"#B3543F"],[0.2709030100334448,"#B45440"],[0.27424749163879603,"#B45541"],[0.27759197324414714,"#B55642"],[0.28093645484949831,"#B55742"],[0.28428093645484953,"#B65843"],[0.28762541806020064,"#B65844"],[0.29096989966555187,"#B75945"],[0.29431438127090304,"#B75A46"],[0.29765886287625415,"#B85B46"],[0.30100334448160537,"#B85B47"],[0.30434782608695649,"#B85C48"],[0.30769230769230771,"#B95D49"],[0.31103678929765888,"#B95E4A"],[0.31438127090300999,"#BA5F4A"],[0.31772575250836121,"#BA5F4B"],[0.32107023411371238,"#BB604C"],[0.32441471571906355,"#BB614D"],[0.32775919732441472,"#BC624E"],[0.33110367892976589,"#BC624E"],[0.33444816053511706,"#BC634F"],[0.33779264214046822,"#BD6450"],[0.34113712374581939,"#BD6551"],[0.34448160535117056,"#BE6552"],[0.34782608695652173,"#BE6652"],[0.3511705685618729,"#BF6753"],[0.35451505016722407,"#BF6854"],[0.35785953177257529,"#C06955"],[0.3612040133779264,"#C06956"],[0.36454849498327757,"#C06A56"],[0.3678929765886288,"#C16B57"],[0.37123745819397991,"#C16C58"],[0.37458193979933113,"#C26C59"],[0.3779264214046823,"#C26D5A"],[0.38127090301003341,"#C36E5A"],[0.38461538461538464,"#C36F5B"],[0.38795986622073575,"#C36F5C"],[0.39130434782608697,"#C4705D"],[0.39464882943143814,"#C4715E"],[0.39799331103678925,"#C5725E"],[0.40133779264214048,"#C5735F"],[0.40468227424749165,"#C67360"],[0.40802675585284282,"#C67461"],[0.41137123745819398,"#C67562"],[0.41471571906354515,"#C77663"],[0.41806020066889632,"#C77663"],[0.42140468227424749,"#C87764"],[0.42474916387959871,"#C87865"],[0.42809364548494983,"#C87966"],[0.43143812709030099,"#C97967"],[0.43478260869565216,"#C97A67"],[0.43812709030100333,"#CA7B68"],[0.44147157190635455,"#CA7C69"],[0.44481605351170567,"#CB7C6A"],[0.44816053511705684,"#CB7D6B"],[0.45150501672240806,"#CB7E6C"],[0.45484949832775917,"#CC7F6C"],[0.4581939799331104,"#CC806D"],[0.46153846153846156,"#CD806E"],[0.46488294314381268,"#CD816F"],[0.4682274247491639,"#CD8270"],[0.47157190635451507,"#CE8371"],[0.47491638795986624,"#CE8371"],[0.47826086956521741,"#CF8472"],[0.48160535117056852,"#CF8573"],[0.48494983277591974,"#CF8674"],[0.48829431438127091,"#D08675"],[0.49163879598662208,"#D08776"],[0.49498327759197325,"#D18877"],[0.49832775919732442,"#D18977"],[0.50167224080267558,"#D18A78"],[0.50501672240802675,"#D28A79"],[0.50836120401337792,"#D28B7A"],[0.51170568561872909,"#D38C7B"],[0.51505016722408026,"#D38D7C"],[0.51839464882943143,"#D38D7C"],[0.52173913043478259,"#D48E7D"],[0.52508361204013376,"#D48F7E"],[0.52842809364548493,"#D4907F"],[0.5317725752508361,"#D59080"],[0.53511705685618727,"#D59181"],[0.53846153846153855,"#D69282"],[0.5418060200668896,"#D69382"],[0.54515050167224077,"#D69483"],[0.54849498327759205,"#D79484"],[0.55183946488294311,"#D79585"],[0.55518394648829428,"#D79686"],[0.55852842809364556,"#D89787"],[0.56187290969899661,"#D89788"],[0.56521739130434778,"#D99888"],[0.56856187290969906,"#D99989"],[0.57190635451505023,"#D99A8A"],[0.57525083612040129,"#DA9A8B"],[0.57859531772575257,"#DA9B8C"],[0.58193979933110374,"#DA9C8D"],[0.58528428093645479,"#DB9D8E"],[0.58862876254180607,"#DB9E8E"],[0.59197324414715724,"#DB9E8F"],[0.5953177257525083,"#DC9F90"],[0.59866220735785947,"#DCA091"],[0.60200668896321075,"#DDA192"],[0.60535117056856191,"#DDA193"],[0.60869565217391297,"#DDA294"],[0.61204013377926425,"#DEA395"],[0.61538461538461542,"#DEA495"],[0.61872909698996648,"#DEA596"],[0.62207357859531776,"#DFA597"],[0.62541806020066892,"#DFA698"],[0.62876254180601998,"#DFA799"],[0.63210702341137126,"#E0A89A"],[0.63545150501672243,"#E0A89B"],[0.6387959866220736,"#E0A99C"],[0.64214046822742477,"#E1AA9C"],[0.64548494983277593,"#E1AB9D"],[0.6488294314381271,"#E1AB9E"],[0.65217391304347827,"#E2AC9F"],[0.65551839464882944,"#E2ADA0"],[0.65886287625418061,"#E2AEA1"],[0.66220735785953178,"#E3AFA2"],[0.66555183946488294,"#E3AFA3"],[0.66889632107023411,"#E4B0A4"],[0.67224080267558539,"#E4B1A4"],[0.67558528428093645,"#E4B2A5"],[0.67892976588628762,"#E5B3A6"],[0.68227424749163879,"#E5B3A7"],[0.68561872909698995,"#E5B4A8"],[0.68896321070234112,"#E5B5A9"],[0.69230769230769229,"#E6B6AA"],[0.69565217391304346,"#E6B6AB"],[0.69899665551839463,"#E6B7AC"],[0.7023411371237458,"#E7B8AC"],[0.70568561872909707,"#E7B9AD"],[0.70903010033444813,"#E7BAAE"],[0.7123745819397993,"#E8BAAF"],[0.71571906354515058,"#E8BBB0"],[0.71906354515050164,"#E8BCB1"],[0.72240802675585281,"#E9BDB2"],[0.72575250836120409,"#E9BDB3"],[0.72909698996655514,"#E9BEB4"],[0.73244147157190631,"#EABFB5"],[0.73578595317725759,"#EAC0B5"],[0.73913043478260876,"#EAC1B6"],[0.74247491638795982,"#EBC1B7"],[0.7458193979933111,"#EBC2B8"],[0.74916387959866226,"#EBC3B9"],[0.75250836120401332,"#ECC4BA"],[0.7558528428093646,"#ECC5BB"],[0.75919732441471577,"#ECC5BC"],[0.76254180602006683,"#ECC6BD"],[0.76588628762541811,"#EDC7BE"],[0.76923076923076927,"#EDC8BE"],[0.77257525083612044,"#EDC9BF"],[0.7759197324414715,"#EEC9C0"],[0.77926421404682278,"#EECAC1"],[0.78260869565217395,"#EECBC2"],[0.785953177257525,"#EFCCC3"],[0.78929765886287628,"#EFCCC4"],[0.79264214046822745,"#EFCDC5"],[0.79598662207357851,"#EFCEC6"],[0.79933110367892979,"#F0CFC7"],[0.80267558528428096,"#F0D0C8"],[0.80602006688963213,"#F0D0C9"],[0.80936454849498329,"#F1D1C9"],[0.81270903010033446,"#F1D2CA"],[0.81605351170568563,"#F1D3CB"],[0.8193979933110368,"#F1D4CC"],[0.82274247491638797,"#F2D4CD"],[0.82608695652173914,"#F2D5CE"],[0.8294314381270903,"#F2D6CF"],[0.83277591973244147,"#F3D7D0"],[0.83612040133779264,"#F3D8D1"],[0.83946488294314392,"#F3D8D2"],[0.84280936454849498,"#F3D9D3"],[0.84615384615384615,"#F4DAD4"],[0.84949832775919742,"#F4DBD5"],[0.85284280936454848,"#F4DCD5"],[0.85618729096989965,"#F4DCD6"],[0.85953177257525093,"#F5DDD7"],[0.86287625418060199,"#F5DED8"],[0.86622073578595316,"#F5DFD9"],[0.86956521739130432,"#F6E0DA"],[0.87290969899665549,"#F6E0DB"],[0.87625418060200666,"#F6E1DC"],[0.87959866220735783,"#F6E2DD"],[0.88294314381270911,"#F7E3DE"],[0.88628762541806017,"#F7E4DF"],[0.88963210702341133,"#F7E4E0"],[0.89297658862876261,"#F7E5E1"],[0.89632107023411367,"#F8E6E2"],[0.89966555183946484,"#F8E7E3"],[0.90301003344481612,"#F8E8E3"],[0.90635451505016718,"#F8E8E4"],[0.90969899665551834,"#F9E9E5"],[0.91304347826086962,"#F9EAE6"],[0.91638795986622079,"#F9EBE7"],[0.91973244147157185,"#F9ECE8"],[0.92307692307692313,"#FAECE9"],[0.9264214046822743,"#FAEDEA"],[0.92976588628762535,"#FAEEEB"],[0.93311036789297663,"#FAEFEC"],[0.9364548494983278,"#FBF0ED"],[0.93979933110367886,"#FBF0EE"],[0.94314381270903014,"#FBF1EF"],[0.94648829431438131,"#FBF2F0"],[0.94983277591973247,"#FCF3F1"],[0.95317725752508364,"#FCF4F2"],[0.95652173913043481,"#FCF4F3"],[0.95986622073578598,"#FCF5F4"],[0.96321070234113704,"#FDF6F4"],[0.96655518394648832,"#FDF7F5"],[0.96989966555183948,"#FDF8F6"],[0.97324414715719054,"#FDF9F7"],[0.97658862876254182,"#FDF9F8"],[0.97993311036789299,"#FEFAF9"],[0.98327759197324416,"#FEFBFA"],[0.98662207357859533,"#FEFCFB"],[0.98996655518394649,"#FEFDFC"],[0.99331103678929766,"#FFFDFD"],[0.99665551839464883,"#FFFEFE"],[1,"#FFFFFF"]],"colorbar":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.8897637795275593,"thickness":23.039999999999996,"title":"p_value","titlefont":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724},"tickmode":"array","ticktext":["0.25","0.50","0.75","1.00"],"tickvals":[0.24992500000000001,0.49995000000000001,0.74997499999999995,1],"tickfont":{"color":"rgba(0,0,0,1)","family":"","size":11.68949771689498},"ticklen":2,"len":0.5}},"xaxis":"x","yaxis":"y","frame":null}],"layout":{"margin":{"t":27.872146118721464,"r":7.3059360730593621,"b":103.1963470319635,"l":180.45662100456624},"plot_bgcolor":"rgba(255,255,255,1)","paper_bgcolor":"rgba(255,255,255,1)","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724},"xaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.40000000000000002,9.5999999999999996],"tickmode":"array","ticktext":["Gadopentetate","Gadodiamide","Gadoversetamide","Gadobenate","Gadoxetate","Gadobutrol","Gadoterate","Gadoteridol","Gadofosveset"],"tickvals":[1,2,3,4,5,6,6.9999999999999991,8,9],"categoryorder":"array","categoryarray":["Gadopentetate","Gadodiamide","Gadoversetamide","Gadobenate","Gadoxetate","Gadobutrol","Gadoterate","Gadoteridol","Gadofosveset"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.6529680365296811,"tickwidth":0.66417600664176002,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.68949771689498},"tickangle":-90,"showline":false,"linecolor":null,"linewidth":0,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"y","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724}},"hoverformat":".2f"},"yaxis":{"domain":[0,1],"automargin":true,"type":"linear","autorange":false,"range":[0.40000000000000002,15.6],"tickmode":"array","ticktext":["Emotional Distress","Confusional State","Hypoaesthesia","Paraesthesia","Mental Impairment","Contrast Media Reaction","Anhedonia","Inflammation","Injury","Burning Sensation","Mass","Urticaria","Nephrogenic Systemic Fibrosis","Gadolinium Deposition Disease","Contrast Media Deposition"],"tickvals":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15],"categoryorder":"array","categoryarray":["Emotional Distress","Confusional State","Hypoaesthesia","Paraesthesia","Mental Impairment","Contrast Media Reaction","Anhedonia","Inflammation","Injury","Burning Sensation","Mass","Urticaria","Nephrogenic Systemic Fibrosis","Gadolinium Deposition Disease","Contrast Media Deposition"],"nticks":null,"ticks":"outside","tickcolor":"rgba(51,51,51,1)","ticklen":3.6529680365296811,"tickwidth":0.66417600664176002,"showticklabels":true,"tickfont":{"color":"rgba(77,77,77,1)","family":"","size":11.68949771689498},"tickangle":-0,"showline":false,"linecolor":null,"linewidth":0,"showgrid":false,"gridcolor":null,"gridwidth":0,"zeroline":false,"anchor":"x","title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724}},"hoverformat":".2f"},"shapes":[{"type":"rect","fillcolor":null,"line":{"color":null,"width":0,"linetype":[]},"yref":"paper","xref":"paper","x0":0,"x1":1,"y0":0,"y1":1}],"showlegend":false,"legend":{"bgcolor":"rgba(255,255,255,1)","bordercolor":"transparent","borderwidth":1.8897637795275593,"font":{"color":"rgba(0,0,0,1)","family":"","size":11.68949771689498},"title":{"text":"","font":{"color":"rgba(0,0,0,1)","family":"","size":14.611872146118724}}},"hovermode":"closest","barmode":"relative"},"config":{"doubleClick":"reset","modeBarButtonsToAdd":["hoverclosest","hovercompare"],"showSendToCloud":false},"source":"A","attrs":{"dd4b24008ea3":{"fill":{},"label":{},"x":{},"y":{},"type":"heatmap"},"dd4b209481cc":{"fill":{},"label":{},"x":{},"y":{}}},"cur_data":"dd4b24008ea3","visdat":{"dd4b24008ea3":["function (y) ","x"],"dd4b209481cc":["function (y) ","x"]},"highlight":{"on":"plotly_click","persistent":false,"dynamic":false,"selectize":false,"opacityDim":0.20000000000000001,"selected":{"opacity":1},"debounce":0},"shinyEvents":["plotly_hover","plotly_click","plotly_selected","plotly_relayout","plotly_brushed","plotly_brushing","plotly_clickannotation","plotly_doubleclick","plotly_deselect","plotly_afterplot","plotly_sunburstclick"],"base_url":"https://plot.ly"},"evals":[],"jsHooks":[]}</script>
</div>
<p>The following observations are made. First, (Contrast Media Deposition, Gadopentetate) appears to be the most prominent AE/drug pair in terms of LR statistic (log LR = 2084.09) across all AE/drug pairs. The next four prominent signals are observed in (Gadolinium Deposition Disease, Gadopentetate), (Gadolinium Deposition Disease, Gadodiamide), (Gadolinium Deposition Disease, Gadoversetamide), and (Gadolinium Deposition Disease, Gadobenate) respectively.</p>
<h2 class="unnumbered" id="analysis-of-the-rotavirus-vaccine-datasets-rv-rvold-and-rvyoung">Analysis of the rotavirus vaccine datasets <code>rv</code>, <code>rvold</code>, and <code>rvyoung</code></h2>
<p>Finally, to exemplify the use of <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> in analyzing vaccine safety we consider rotavirus vaccine datasets. These vaccines are an important example for vaccine safety assessment. Intussusception, a telescoping of the intestines that can cause serious illness, had led to the withdrawal of the first rotavirus vaccines from the market. Several AEs have since been identified in analyses of the FDA/CDC VAERS data <span class="citation" data-cites="niu2001 haber2004">(<a href="#ref-niu2001" role="doc-biblioref">Niu et al. 2001</a>; <a href="#ref-haber2004" role="doc-biblioref">Haber et al. 2004</a>)</span> as being associated with intusscusception. Below in our analysis we consider several of these vaccine/AE combinations; these combinations are well-known and well-studied and hence they aid informative assessment of the performance of the disproportionality methods. In <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a>, the rotavirus vaccine datasets are obtained from the FDA VAERS database for the year of 1999, and they summarize adverse event report counts for two vaccine groups, namely the rotavirus vaccine (“Rv”) and “Other” vaccines (37 different vaccines combined). <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> contains three such datasets namely <code>rvold</code>, <code>rvyoung</code>, and <code>rv</code> tabulating the number of occurrences of 727, 346, and 794 AEs among individuals of “old” (age <span class="math inline">\(\geq 1\)</span> year), “young” (age <span class="math inline">\(&lt; 1\)</span> year), and “combined” (collective old and young) age groups respectively, with most counts in the combined dataset arising from the young age group. Our interest here lies in identifying the most prominent AEs of the rotavirus vaccine and then comparing them across the three datasets/age groups. The rotavirus data has been previously analyzed in the literature using other approaches such as the EBGM <span class="citation" data-cites="niu2001">(<a href="#ref-niu2001" role="doc-biblioref">Niu et al. 2001</a>)</span> and proportional morbidity analysis for vaccine safety <span class="citation" data-cites="haber2004">(<a href="#ref-haber2004" role="doc-biblioref">Haber et al. 2004</a>)</span>. Here we will utilize the LRT method as implemented in <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a>: first, in each dataset we will perform an LRT with an <em>optimally selected</em> model (Poisson or ZIP); subsequently, based on the LRT results we will determine a few prominent AEs with the largest LR statistics in each dataset and compare these AEs across datasets.</p>
<p>We first fit Poisson and ZIP models to the three datasets:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>100</span><span class='op'>)</span></span>
<span><span class='va'>test_rv_poisson_list</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span></span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>old <span class='op'>=</span> <span class='va'>rvold</span>, young <span class='op'>=</span> <span class='va'>rvyoung</span>, combined <span class='op'>=</span> <span class='va'>rv</span><span class='op'>)</span>,</span>
<span>  <span class='kw'>function</span><span class='op'>(</span><span class='va'>this_data</span><span class='op'>)</span> <span class='op'>{</span></span>
<span>    <span class='fu'><a href='https://rdrr.io/pkg/pvLRT/man/pvlrt.html'>pvlrt</a></span><span class='op'>(</span></span>
<span>      <span class='va'>this_data</span>,</span>
<span>      zi_prob <span class='op'>=</span> <span class='fl'>0</span>,</span>
<span>      test_drug_idx <span class='op'>=</span> <span class='fl'>1</span></span>
<span>    <span class='op'>)</span></span>
<span>  <span class='op'>}</span></span>
<span><span class='op'>)</span></span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='fu'><a href='https://rdrr.io/r/base/Random.html'>set.seed</a></span><span class='op'>(</span><span class='fl'>100</span><span class='op'>)</span></span>
<span><span class='va'>test_rv_zip_list</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span></span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>old <span class='op'>=</span> <span class='va'>rvold</span>, young <span class='op'>=</span> <span class='va'>rvyoung</span>, combined <span class='op'>=</span> <span class='va'>rv</span><span class='op'>)</span>,</span>
<span>  <span class='kw'>function</span><span class='op'>(</span><span class='va'>this_data</span><span class='op'>)</span> <span class='op'>{</span></span>
<span>    <span class='fu'><a href='https://rdrr.io/pkg/pvLRT/man/pvlrt.html'>pvlrt</a></span><span class='op'>(</span></span>
<span>      <span class='va'>this_data</span>,</span>
<span>      test_drug_idx <span class='op'>=</span> <span class='fl'>1</span>,</span>
<span>      zi_prob <span class='op'>=</span> <span class='cn'>NULL</span></span>
<span>    <span class='op'>)</span></span>
<span>  <span class='op'>}</span></span>
<span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>The objects <code>test_rv_poisson_list</code> and <code>test_rv_zip_list</code> list the Poisson and ZIP LRT results for the three datasets respectively. The estimated zero-inflation probabilities for the “Rv” vaccine can be extracted from the latter list as follows. In the code chunks below we use the <a href="https://cran.r-project.org/package=magrittr">magrittr</a> <span class="citation" data-cites="pkg_magrittr">(<a href="#ref-pkg_magrittr" role="doc-biblioref">Bache and Wickham 2020</a>)</span> forward pipe <code>%&gt;%</code> infix operator for better readability of chained computations. The operator is automatically loaded with <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a>; see the <a href="https://cran.r-project.org/package=magrittr">magrittr</a> package manual for further details on the pipe operator.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>rv_zi_prob</span> <span class='op'>&lt;-</span> <span class='va'>test_rv_zip_list</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> </span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='kw'>function</span><span class='op'>(</span><span class='va'>test</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/pkg/pvLRT/man/extract_lrstat_matrix.html'>extract_zi_probability</a></span><span class='op'>(</span><span class='va'>test</span><span class='op'>)</span><span class='op'>[</span><span class='st'>"Rv"</span><span class='op'>]</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> </span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/Round.html'>round</a></span><span class='op'>(</span><span class='fl'>3</span><span class='op'>)</span></span>
<span><span class='va'>rv_zi_prob</span></span></code></pre>
</div>
<pre><code>     old.Rv    young.Rv combined.Rv 
      0.248       0.000       0.215 </code></pre>
</div>
<p>It follows from the above estimated zero inflation probabilities that a ZIP model could be appropriate for the combined and the old age groups whereas an ordinary Poisson model is likely adequate for the young age groups. We formalize the model selection via AIC and BIC, and find the optimal models for the three datasets. Codes are as follows.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>compare_models_rv</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/mapply.html'>mapply</a></span><span class='op'>(</span></span>
<span>  <span class='kw'>function</span><span class='op'>(</span><span class='va'>test_poisson</span>, <span class='va'>test_zip</span>, <span class='va'>data_name</span><span class='op'>)</span> <span class='op'>{</span></span>
<span>    <span class='va'>AIC_both</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/AIC.html'>AIC</a></span><span class='op'>(</span><span class='va'>test_poisson</span>, <span class='va'>test_zip</span><span class='op'>)</span><span class='op'>$</span><span class='va'>AIC</span></span>
<span>    <span class='va'>BIC_both</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/AIC.html'>BIC</a></span><span class='op'>(</span><span class='va'>test_poisson</span>, <span class='va'>test_zip</span><span class='op'>)</span><span class='op'>$</span><span class='va'>BIC</span></span>
<span>    <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span></span>
<span>      name <span class='op'>=</span> <span class='va'>data_name</span>, </span>
<span>      AIC_poisson <span class='op'>=</span> <span class='va'>AIC_both</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span>,</span>
<span>      BIC_poisson <span class='op'>=</span> <span class='va'>BIC_both</span><span class='op'>[</span><span class='fl'>1</span><span class='op'>]</span>,</span>
<span>      AIC_zip <span class='op'>=</span> <span class='va'>AIC_both</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span>,</span>
<span>      BIC_zip <span class='op'>=</span> <span class='va'>BIC_both</span><span class='op'>[</span><span class='fl'>2</span><span class='op'>]</span>,</span>
<span>      optimal_AIC <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"poisson"</span>, <span class='st'>"zip"</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/which.min.html'>which.min</a></span><span class='op'>(</span><span class='va'>AIC_both</span><span class='op'>)</span><span class='op'>]</span>,</span>
<span>      optimal_BIC <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"poisson"</span>, <span class='st'>"zip"</span><span class='op'>)</span><span class='op'>[</span><span class='fu'><a href='https://rdrr.io/r/base/which.min.html'>which.min</a></span><span class='op'>(</span><span class='va'>AIC_both</span><span class='op'>)</span><span class='op'>]</span></span>
<span>    <span class='op'>)</span></span>
<span>  <span class='op'>}</span>,</span>
<span>  <span class='va'>test_rv_poisson_list</span>,</span>
<span>  <span class='va'>test_rv_zip_list</span>,</span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>test_rv_poisson_list</span><span class='op'>)</span>,</span>
<span>  SIMPLIFY <span class='op'>=</span> <span class='cn'>FALSE</span></span>
<span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> </span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/unname.html'>unname</a></span><span class='op'>(</span><span class='op'>)</span> <span class='op'><a href='https://magrittr.tidyverse.org/reference/pipe.html'>%&gt;%</a></span> </span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/do.call.html'>do.call</a></span><span class='op'>(</span><span class='va'>rbind</span>, <span class='va'>.</span><span class='op'>)</span></span>
<span></span>
<span><span class='va'>compare_models_rv</span></span></code></pre>
</div>
<pre><code>      name AIC_poisson BIC_poisson  AIC_zip  BIC_zip optimal_AIC
1      old    5749.153   18631.535 5751.783 18651.88     poisson
2    young    3829.600    9000.445 3833.600  9019.39     poisson
3 combined    8308.330   22800.414 8051.049 22561.38         zip
  optimal_BIC
1     poisson
2     poisson
3         zip</code></pre>
</div>
<p>It follows from the above AIC and BIC values that while the ZIP model is a clear winner for the combined age group dataset, the Poisson model displays slightly smaller AIC and BIC values in the other two datasets. We therefore first create a list of <em>optimal</em> tests from these two models fitted to the three datasets, and use these optimal models for subsequent determination of prominent AEs as follows:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>test_rv_optimal_list</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span></span>
<span>  old <span class='op'>=</span> <span class='va'>test_rv_poisson_list</span><span class='op'>$</span><span class='va'>old</span>,</span>
<span>  young <span class='op'>=</span> <span class='va'>test_rv_poisson_list</span><span class='op'>$</span><span class='va'>young</span>,</span>
<span>  combined <span class='op'>=</span> <span class='va'>test_rv_zip_list</span><span class='op'>$</span><span class='va'>combined</span></span>
<span><span class='op'>)</span></span></code></pre>
</div>
</div>
<p>We may extract the statistically significant AE/vaccine pairs (<span class="math inline">\(p\)</span>-value &lt; 0.05) from these tests as follows.</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>signif_pairs_rv</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span><span class='op'>(</span><span class='va'>test_rv_zip_list</span>, <span class='va'>extract_significant_pairs</span><span class='op'>)</span> </span></code></pre>
</div>
</div>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='co'># print for summarized results</span></span>
<span><span class='va'>signif_pairs_rv</span></span></code></pre>
</div>
<pre><code>$old
                             AE Drug n    lrstat    p_value
1: Gastrointestinal Haemorrhage   Rv 2 10.768628 0.00049995
2:                    Diarrhoea   Rv 4 10.386030 0.00049995
3:       Intestinal Obstruction   Rv 1  7.087679 0.03909609
4:       Secondary Transmission   Rv 1  7.087679 0.03909609

$young
                             AE Drug   n    lrstat    p_value
1:                    Diarrhoea   Rv 114 35.680505 0.00009999
2: Gastrointestinal Haemorrhage   Rv  88 33.708041 0.00009999
3:                     Vomiting   Rv 129 22.512090 0.00009999
4:              Intussusception   Rv  32 11.980492 0.00009999
5:               Abdominal Pain   Rv  40 11.911067 0.00009999
6:                  Dehydration   Rv  37 10.643293 0.00009999
7:              Gastroenteritis   Rv  21  6.150401 0.00809919
8:       Intestinal Obstruction   Rv  19  5.457423 0.02109789

$combined
                              AE Drug   n     lrstat    p_value
 1:                    Diarrhoea   Rv 135 154.320951 0.00009999
 2: Gastrointestinal Haemorrhage   Rv  91 153.884769 0.00009999
 3:                     Vomiting   Rv 134  99.195439 0.00009999
 4:              Intussusception   Rv  34  58.610616 0.00009999
 5:                    Agitation   Rv 105  45.670818 0.00009999
 6:                  Dehydration   Rv  39  42.857250 0.00009999
 7:       Intestinal Obstruction   Rv  21  34.022145 0.00009999
 8:               Abdominal Pain   Rv  43  33.794857 0.00009999
 9:              Gastroenteritis   Rv  23  33.200344 0.00009999
10:              Abnormal Faeces   Rv  16  23.570006 0.00009999
11:                 Constipation   Rv  18  23.020318 0.00009999
12:                     Anorexia   Rv  45  22.292411 0.00009999
13:             Weight Decreased   Rv  13  13.939082 0.00019998
14:       Diarrhoea Haemorrhagic   Rv   7  13.573366 0.00039996
15:                   Flatulence   Rv  10  13.093022 0.00079992
16:                   Somnolence   Rv  37  13.077281 0.00079992
17:                      Melaena   Rv   7  12.255722 0.00099990
18:         Abdominal Distension   Rv   8  12.226474 0.00109989
19:    Gastrointestinal Disorder   Rv  16  10.774175 0.00369963
20:                 Haematemesis   Rv   4   9.818926 0.00809919
21:                       Sepsis   Rv  10   8.919282 0.01769823
                              AE Drug   n     lrstat    p_value</code></pre>
</div>
<p>For visualization, we consider heatmaps for top 15 AEs (in terms of LRT statistics) in the three datasets, and place them side by side. Because <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> uses <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> as its plotting engine, these individual heatmaps can be easily combined via external <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a> post-processing packages such as <a href="https://cran.r-project.org/package=patchwork">patchwork</a> <span class="citation" data-cites="pkg_patchwork">(<a href="#ref-pkg_patchwork" role="doc-biblioref">Pedersen 2020</a>)</span>. Codes are as follows:</p>
<div class="layout-chunk" data-layout="l-body">
<div class="sourceCode">
<pre class="sourceCode r"><code class="sourceCode r"><span><span class='va'>pl_heatmap_rv_list</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/mapply.html'>mapply</a></span><span class='op'>(</span></span>
<span>  <span class='kw'>function</span><span class='op'>(</span><span class='va'>this_test</span>, <span class='va'>this_name</span><span class='op'>)</span> <span class='op'>{</span></span>
<span>    <span class='fu'><a href='https://rdrr.io/pkg/pvLRT/man/heatmap_pvlrt.html'>heatmap_pvlrt</a></span><span class='op'>(</span></span>
<span>      <span class='va'>this_test</span>,</span>
<span>      AE <span class='op'>=</span> <span class='fl'>15</span>,</span>
<span>      show_n <span class='op'>=</span> <span class='cn'>TRUE</span>,</span>
<span>      show_lrstat <span class='op'>=</span> <span class='cn'>TRUE</span></span>
<span>    <span class='op'>)</span> <span class='op'>+</span> </span>
<span>      <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/labs.html'>ggtitle</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste</a></span><span class='op'>(</span><span class='st'>"age group:\n"</span>, <span class='va'>this_name</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>+</span> </span>
<span>      <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span>plot.title <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>hjust <span class='op'>=</span> <span class='fl'>0.5</span><span class='op'>)</span><span class='op'>)</span></span>
<span>  <span class='op'>}</span>,</span>
<span>  <span class='va'>test_rv_optimal_list</span>,</span>
<span>  <span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span><span class='op'>(</span><span class='va'>test_rv_optimal_list</span><span class='op'>)</span>,</span>
<span>  SIMPLIFY <span class='op'>=</span> <span class='cn'>FALSE</span></span>
<span><span class='op'>)</span></span>
<span><span class='co'>## install patchwork if not installed</span></span>
<span><span class='co'># if (!requireNamespace("patchwork")) install.packages("pacthwork")</span></span>
<span><span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='https://patchwork.data-imaginist.com'>patchwork</a></span><span class='op'>)</span></span>
<span><span class='va'>pl_heatmap_rv_comb</span> <span class='op'>&lt;-</span> <span class='va'>pl_heatmap_rv_list</span><span class='op'>$</span><span class='va'>old</span> <span class='op'>+</span> </span>
<span>  <span class='va'>pl_heatmap_rv_list</span><span class='op'>$</span><span class='va'>young</span> <span class='op'>+</span> </span>
<span>  <span class='va'>pl_heatmap_rv_list</span><span class='op'>$</span><span class='va'>combined</span> <span class='op'>+</span> </span>
<span>  <span class='co'># combine the legends</span></span>
<span>  <span class='fu'><a href='https://patchwork.data-imaginist.com/reference/plot_layout.html'>plot_layout</a></span><span class='op'>(</span>guides <span class='op'>=</span> <span class='st'>"collect"</span><span class='op'>)</span> <span class='op'>&amp;</span> </span>
<span>  <span class='co'># unified coloring scheme, </span></span>
<span>  <span class='co'># see the 'fill_gradient_range' argument for heatmap_pvlrt</span></span>
<span>  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/scale_gradient.html'>scale_fill_gradientn</a></span><span class='op'>(</span></span>
<span>    colors <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"darkred"</span>, <span class='st'>"white"</span><span class='op'>)</span>, </span>
<span>    limits <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0</span>, <span class='fl'>1</span><span class='op'>)</span></span>
<span>  <span class='op'>)</span> <span class='op'>&amp;</span></span>
<span>  <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/theme.html'>theme</a></span><span class='op'>(</span></span>
<span>    axis.text <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>16</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span>,</span>
<span>    legend.title <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>16</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span>,</span>
<span>    legend.text <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>14</span><span class='op'>)</span>,</span>
<span>    title <span class='op'>=</span> <span class='fu'><a href='https://ggplot2.tidyverse.org/reference/element.html'>element_text</a></span><span class='op'>(</span>size <span class='op'>=</span> <span class='fl'>16</span>, face <span class='op'>=</span> <span class='st'>"bold"</span><span class='op'>)</span></span>
<span>  <span class='op'>)</span></span>
<span></span>
<span><span class='va'>pl_heatmap_rv_comb</span></span></code></pre>
</div>
<p><img src="RJ-2023-027_files/figure-html5/heatmap-rv-all-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<p>The following observations are made from the above heatmaps and associated test results (stored in <code>signif_pairs_rv</code>). First, in terms of LR statistics and <span class="math inline">\(p\)</span>-values, the “old” age group has only 4 statistically significant (<span class="math inline">\(p\)</span>-value <span class="math inline">\(&lt; 0.05\)</span>) AEs, viz., Gastrointestinal Haemorrhage, Diarrhoea, Intestinal Obstruction, and Secondary Transmission. Note, however, the extremely small sample sizes for these AEs within the dataset; care must therefore be taken while interpreting these results. By contrast, the young and the combined age group data sets contain sufficient sample sizes. There are 8 and 21 significant AEs in the young and the combined age group datasets respectively. These prominent AEs for the combined and the young age groups agree, which is expected as the most reported cases in the combined dataset correspond to the young age group. Several of these prominent AEs detected in our analysis associate with Intussusception, which has been determined to be of particular concern in the literature <span class="citation" data-cites="niu2001 haber2004">(<a href="#ref-niu2001" role="doc-biblioref">Niu et al. 2001</a>; <a href="#ref-haber2004" role="doc-biblioref">Haber et al. 2004</a>)</span>. However, unlike the approaches used in those studies, our LRT-based approach for signal detection does not require ad hoc thresholding, and provides rigorous statistical guarantees on the results obtained.</p>
<h1 data-number="4" id="discussion-and-future-directions"><span class="header-section-number">4</span> Discussion and Future Directions</h1>
<p>Likelihood ratio test (LRT)-based approaches to pharmacovigilance constitute a class of rigorous statistical methods that permit objective signal determination without the need for specifying ad hoc thresholds. The lack of comprehensive software implementing the LRT methods (in R or otherwise) however has hindered their applicability in practice – especially in cases where the adverse event count data are zero-inflated.</p>
<p>The package <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a> is our contribution to this area, providing a platform that implements a suite of formal statistical tools and post-processing functions for analyzing medical product safety datasets, and is readily accessible to practitioners. There are various directions in which the software can be extended in future releases. Some possible features/updates under consideration include the following.</p>
<ol type="1">
<li><p>Parallelize bootstrap resampling. This may permit substantial computational gains in sizable datasets. Care must however be taken while generating random draws in the parallel workers to ensure reproducibility and validity (see the discussion of page 5, Section 6 of the R package <a href="https://cran.r-project.org/package=parallel">parallel</a> <span class="citation" data-cites="pkg_parallel">(<a href="#ref-pkg_parallel" role="doc-biblioref">R Core Team 2022</a>)</span>).</p></li>
<li><p>Include additional processed datasets, and expand the current documentations with examples on these additional datasets.</p></li>
<li><p>Create visualizations for the estimated zero inflation probabilities along with the LRT results.</p></li>
</ol>
<h1 data-number="5" id="acknowledgment"><span class="header-section-number">5</span> Acknowledgment</h1>
<p>Dr. Markatou acknowledges grant support from KALEIDA Health Foundation (Troup Fund) and BAA HDDISWP#97, FDA. These resources supported the work of the University at Buffalo affiliated authors.</p>
<div class="layout-chunk" data-layout="l-body">

</div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h2 class="appendix unnumbered" id="cran-packages-used">CRAN packages used</h2>
<p><a href="https://cran.r-project.org/package=%3C%3Cx%5B2%5D%3E%3E">&lt;&lt;x[2]&gt;&gt;</a>, <a href="https://cran.r-project.org/package=PhViD">PhViD</a>, <a href="https://cran.r-project.org/package=openEBGM">openEBGM</a>, <a href="https://cran.r-project.org/package=sglr">sglr</a>, <a href="https://cran.r-project.org/package=Sequential">Sequential</a>, <a href="https://cran.r-project.org/package=SPRT">SPRT</a>, <a href="https://cran.r-project.org/package=AEenrich">AEenrich</a>, <a href="https://cran.r-project.org/package=mds">mds</a>, <a href="https://cran.r-project.org/package=pvLRT">pvLRT</a>, <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a>, <a href="https://cran.r-project.org/package=plotly">plotly</a>, <a href="https://cran.r-project.org/package=data.table">data.table</a>, <a href="https://cran.r-project.org/package=tidyverse">tidyverse</a>, <a href="https://cran.r-project.org/package=magrittr">magrittr</a>, <a href="https://cran.r-project.org/package=patchwork">patchwork</a>, <a href="https://cran.r-project.org/package=parallel">parallel</a></p>
<h2 class="appendix unnumbered" id="cran-task-views-implied-by-cited-packages">CRAN Task Views implied by cited packages</h2>
<p><a href="https://cran.r-project.org/view=Bayesian">Bayesian</a>, <a href="https://cran.r-project.org/view=Finance">Finance</a>, <a href="https://cran.r-project.org/view=HighPerformanceComputing">HighPerformanceComputing</a>, <a href="https://cran.r-project.org/view=Phylogenetics">Phylogenetics</a>, <a href="https://cran.r-project.org/view=Spatial">Spatial</a>, <a href="https://cran.r-project.org/view=TeachingStatistics">TeachingStatistics</a>, <a href="https://cran.r-project.org/view=TimeSeries">TimeSeries</a>, <a href="https://cran.r-project.org/view=WebTechnologies">WebTechnologies</a></p>
<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-agresti2013" class="csl-entry" role="listitem">
A. Agresti. <em>Categorical Data Analysis.</em> John Wiley &amp; Sons, 2013. Google-Books-ID: 6PHHE1Cr44AC.
</div>
<div id="ref-pkg_phvid" class="csl-entry" role="listitem">
I. Ahmed and A. Poncet. <em><span class="nocase">PhViD: an R package for PharmacoVigilance signal Detection</span>.</em> 2016. R package version 1.0.8.
</div>
<div id="ref-pkg_magrittr" class="csl-entry" role="listitem">
S. M. Bache and H. Wickham. <em>Magrittr: A forward-pipe operator for r.</em> 2020. URL <a href="https://CRAN.R-project.org/package=magrittr">https://CRAN.R-project.org/package=magrittr</a>. R package version 2.0.1.
</div>
<div id="ref-benjamini1995" class="csl-entry" role="listitem">
Y. Benjamini and Y. Hochberg. Controlling the false discovery rate: A practical and powerful approach to multiple testing. <em>Journal of the Royal Statistical Society: Series B (Methodological)</em>, 1995. DOI <a href="https://doi.org/10.1111/j.2517-6161.1995.tb02031.x">10.1111/j.2517-6161.1995.tb02031.x</a>.
</div>
<div id="ref-canida2017" class="csl-entry" role="listitem">
T. Canida and J. Ihrie. <span class="nocase">openEBGM: An R Implementation of the Gamma-Poisson Shrinker Data Mining Model</span>. <em>The R Journal</em>, 9(2): 499–519, 2017. URL <a href="https://journal.r-project.org/archive/2017/RJ-2017-063/index.html">https://journal.r-project.org/archive/2017/RJ-2017-063/index.html</a>.
</div>
<div id="ref-chakraborty2022" class="csl-entry" role="listitem">
S. Chakraborty, A. Liu, R. Ball and M. Markatou. On the use of the likelihood ratio test methodology in pharmacovigilance. <em>Statistics in Medicine</em>, 41(27): 5395–5420, 2022.
</div>
<div id="ref-pkg_pvLRT" class="csl-entry" role="listitem">
S. Chakraborty and M. Markatou. <em>pvLRT: Likelihood ratio test-based approaches to pharmacovigilance.</em> 2022. R package version 0.4.
</div>
<div id="ref-pkg_mds" class="csl-entry" role="listitem">
G. Chung. <em><span class="nocase">mds: Medical Devices Surveillance</span>.</em> 2020. URL <a href="https://CRAN.R-project.org/package=mds">https://CRAN.R-project.org/package=mds</a>. R package version 0.3.2.
</div>
<div id="ref-davison1997" class="csl-entry" role="listitem">
A. C. Davison and D. V. Hinkley. <em>Bootstrap methods and their application.</em> Cambridge: Cambridge University Press, 1997. DOI <a href="https://doi.org/10.1017/CBO9780511802843">10.1017/CBO9780511802843</a>.
</div>
<div id="ref-ding2020" class="csl-entry" role="listitem">
Y. Ding, M. Markatou and R. Ball. An evaluation of statistical approaches to postmarketing surveillance. <em>Statistics in Medicine</em>, 39(7): 845–874, 2020. URL <a href="https://onlinelibrary.wiley.com/doi/abs/10.1002/sim.8447">https://onlinelibrary.wiley.com/doi/abs/10.1002/sim.8447</a>.
</div>
<div id="ref-pkg_datatable" class="csl-entry" role="listitem">
M. Dowle and A. Srinivasan. <em>Data.table: Extension of ‘data.frame‘.</em> 2021. URL <a href="https://CRAN.R-project.org/package=data.table">https://CRAN.R-project.org/package=data.table</a>. R package version 1.14.2.
</div>
<div id="ref-dumouchel1999" class="csl-entry" role="listitem">
W. Dumouchel. Bayesian data mining in large frequency tables, with an application to the <span>FDA</span> spontaneous reporting system. <em>The American Statistician</em>, 53(3): 177–190, 1999. URL <a href="https://www.tandfonline.com/doi/abs/10.1080/00031305.1999.10474456">https://www.tandfonline.com/doi/abs/10.1080/00031305.1999.10474456</a>.
</div>
<div id="ref-haber2004" class="csl-entry" role="listitem">
P. Haber, R. T. Chen, L. R. Zanardi, G. T. Mootrey, R. English, M. M. Braun and the VAERS Working Group. An analysis of rotavirus vaccine reports to the vaccine adverse event reporting system: More than intussusception alone? <em>Pediatrics</em>, 113(4): e353–e359, 2004. URL <a href="https://doi.org/10.1542/peds.113.4.e353">https://doi.org/10.1542/peds.113.4.e353</a>.
</div>
<div id="ref-huangLRT2011" class="csl-entry" role="listitem">
L. Huang, J. Zalkikar and R. C. Tiwari. A likelihood ratio test based method for signal detection with application to FDA<span>’</span>s drug safety data. <em>Journal of the American Statistical Association</em>, 106(496): 1230–1241, 2011. URL <a href="https://doi.org/10.1198/jasa.2011.ap10243">https://doi.org/10.1198/jasa.2011.ap10243</a>.
</div>
<div id="ref-huang2017" class="csl-entry" role="listitem">
L. Huang, D. Zheng, J. Zalkikar and R. Tiwari. Zero-inflated poisson model based likelihood ratio test for drug safety signal detection. <em>Statistical Methods in Medical Research</em>, 26(1): 471–488, 2017. URL <a href="https://doi.org/10.1177/0962280214549590">https://doi.org/10.1177/0962280214549590</a>.
</div>
<div id="ref-pkg_aeenrich" class="csl-entry" role="listitem">
S. Li, H. Chen, L. Zhao and M. Kleinsasser. <em><span>AEenrich: Adverse Event Enrichment Tests</span>.</em> 2021. URL <a href="https://CRAN.R-project.org/package=AEenrich">https://CRAN.R-project.org/package=AEenrich</a>. R package version 1.1.0.
</div>
<div id="ref-markatou2014" class="csl-entry" role="listitem">
M. Markatou and R. Ball. A pattern discovery framework for adverse event evaluation and inference in spontaneous reporting systems. <em>Stat. Anal. Data Min.</em>, 2014. DOI <a href="https://doi.org/10.1002/sam.11233">10.1002/sam.11233</a>.
</div>
<div id="ref-pkg_sglr" class="csl-entry" role="listitem">
B. Narasimhan and M.-C. Shih. <em><span class="nocase">sglr</span>: An r package for computing the boundaries for sequential generalized likelihood ratio test for pre-licensure vaccine studies.</em> 2012. URL <a href="http://CRAN.R-project.org/package=sglr">http://CRAN.R-project.org/package=sglr</a>. R package version 0.7.
</div>
<div id="ref-niu2001" class="csl-entry" role="listitem">
M. T. Niu, D. E. Erwin and M. M. Braun. <span>Data mining in the US Vaccine Adverse Event Reporting System (VAERS): early detection of intussusception and other events after rotavirus vaccination</span>. <em>Vaccine</em>, 19(32): 4627–4634, 2001. URL <a href="https://www.sciencedirect.com/science/article/pii/S0264410X01002377">https://www.sciencedirect.com/science/article/pii/S0264410X01002377</a>.
</div>
<div id="ref-pkg_patchwork" class="csl-entry" role="listitem">
T. L. Pedersen. <em>Patchwork: The composer of plots.</em> 2020. URL <a href="https://CRAN.R-project.org/package=patchwork">https://CRAN.R-project.org/package=patchwork</a>. R package version 1.1.1.
</div>
<div id="ref-pkg_parallel" class="csl-entry" role="listitem">
R Core Team. Package <span>“parallel.”</span> 2022. (Accessed on 08/04/2023).
</div>
<div id="ref-sakamoto1986" class="csl-entry" role="listitem">
Y. Sakamoto, M. Ishiguro and G. Kitagawa. Akaike information criterion statistics. <em>Dordrecht, The Netherlands: D. Reidel</em>, 81(10.5555): 26853, 1986. Publisher: Taylor &amp; Francis.
</div>
<div id="ref-pkg_plotly" class="csl-entry" role="listitem">
C. Sievert. <em>Interactive web-based data visualization with r, plotly, and shiny.</em> Chapman; Hall/CRC, 2020. URL <a href="https://plotly-r.com">https://plotly-r.com</a>.
</div>
<div id="ref-pkg_sequential" class="csl-entry" role="listitem">
I. R. Silva and M. Kulldorff. <em><span class="nocase">Sequential: Exact Sequential Analysis for Poisson and Binomial Data</span>.</em> 2021. URL <a href="https://CRAN.R-project.org/package=Sequential">https://CRAN.R-project.org/package=Sequential</a>. R package version 4.1.
</div>
<div id="ref-pkg_ggplot2" class="csl-entry" role="listitem">
H. Wickham. <em>ggplot2: Elegant graphics for data analysis.</em> Springer-Verlag New York, 2016. URL <a href="https://ggplot2.tidyverse.org">https://ggplot2.tidyverse.org</a>.
</div>
<div id="ref-pkg_tidyverse" class="csl-entry" role="listitem">
H. Wickham, M. Averick, J. Bryan, W. Chang, L. D. McGowan, R. François, G. Grolemund, A. Hayes, L. Henry, J. Hester, et al. Welcome to the <span class="nocase">tidyverse</span>. <em>Journal of Open Source Software</em>, 4(43): 1686, 2019. DOI <a href="https://doi.org/10.21105/joss.01686">10.21105/joss.01686</a>.
</div>
<div id="ref-zhao2018" class="csl-entry" role="listitem">
Y. Zhao, M. Yi and R. C. Tiwari. Extended likelihood ratio test-based methods for signal detection in a drug class with application to FDA<span>’</span>s adverse event reporting system database. <em>Statistical Methods in Medical Research</em>, 27(3): 876–890, 2018. URL <a href="https://doi.org/10.1177/0962280216646678">https://doi.org/10.1177/0962280216646678</a>. Publisher: SAGE Publications Ltd STM.
</div>
</div>
<aside id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr />
<ol>
<li id="fn1"><p>The default printing specifications for a <code>data.table</code> object applies to the resulting summary, which abbreviates the output to show results for the top 5 and bottom 5 AE/drug pairs (arranged with respect to the computed LRT values) on the console for large summary tables. To print more pairs, use e.g., <code>summary(test_statin46_poisson) %&gt;% print(n)</code> with a larger <code>n</code>. Storing the summary output to an object, e.g., <code>summary_test &lt;- summary(test_statin46_poisson)</code> allows access to results from all pairs.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</aside>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
  <h3 id="references">References</h3>
  <div id="references-listing"></div>
  <h3 id="reuse">Reuse</h3>
  <p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
  <h3 id="citation">Citation</h3>
  <p>For attribution, please cite this work as</p>
  <pre class="citation-appendix short">Chakraborty, et al., "Likelihood Ratio Test-Based Drug Safety Assessment using R Package pvLRT", The R Journal, 2023</pre>
  <p>BibTeX citation</p>
  <pre class="citation-appendix long">@article{RJ-2023-027,
  author = {Chakraborty, Saptarshi and Markatou, Marianthi and Ball, Robert},
  title = {Likelihood Ratio Test-Based Drug Safety Assessment using R Package pvLRT},
  journal = {The R Journal},
  year = {2023},
  note = {https://doi.org/10.32614/RJ-2023-027},
  doi = {10.32614/RJ-2023-027},
  volume = {15},
  issue = {1},
  issn = {2073-4859},
  pages = {101-121}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<!--radix_placeholder_navigation_after_body--><html><body>
<div class="distill-site-nav distill-site-footer">
<p>© The R Foundation, <a href="mailto:r-journal@r-project.org">web page
contact</a>.</p>
</div>
<!--/radix_placeholder_navigation_after_body-->
</body></html>


</body>

</html>
