<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
      margin-bottom: 0em;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
  <title>dtComb: A Comprehensive R Library and Web Tool for Combining Diagnostic Tests</title>

  <meta property="description" itemprop="description" content="The combination of diagnostic tests has become a crucial area of&#10;research, aiming to improve the accuracy and robustness of medical&#10;diagnostics. While existing tools focus primarily on linear&#10;combination methods, there is a lack of comprehensive tools that&#10;integrate diverse methodologies. In this study, we present dtComb, a&#10;comprehensive R package and web tool designed to address the&#10;limitations of existing diagnostic test combination platforms. One of&#10;the unique contributions of dtComb is offering a range of 142 methods&#10;to combine two diagnostic tests, including linear, non-linear, machine&#10;learning algorithms, and mathematical operators. Another significant&#10;contribution of dtComb is its inclusion of advanced tools for ROC&#10;analysis, diagnostic performance metrics, and visual outputs such as&#10;sensitivity-specificity curves. Furthermore, dtComb offers&#10;classification functions for new observations, making it an&#10;easy-to-use tool for clinicians and researchers. The web-based version&#10;is also available at &lt;https://biotools.erciyes.edu.tr/dtComb/&gt; for&#10;non-R users, providing an intuitive interface for test combination and&#10;model training."/>

  <link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>

  <!--  https://schema.org/Article -->
  <meta property="article:published" itemprop="datePublished" content="2026-02-03"/>
  <meta property="article:created" itemprop="dateCreated" content="2026-02-03"/>
  <meta name="article:author" content="S. Ilayda Yerlitaş Taştan"/>
  <meta name="article:author" content="Serra Bersan Gengeç"/>
  <meta name="article:author" content="Necla Koçhan"/>
  <meta name="article:author" content="Ertürk Zararsız"/>
  <meta name="article:author" content="Selçuk Korkmaz"/>
  <meta name="article:author" content="Zararsız"/>

  <!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
  <meta property="og:title" content="dtComb: A Comprehensive R Library and Web Tool for Combining Diagnostic Tests"/>
  <meta property="og:type" content="article"/>
  <meta property="og:description" content="The combination of diagnostic tests has become a crucial area of&#10;research, aiming to improve the accuracy and robustness of medical&#10;diagnostics. While existing tools focus primarily on linear&#10;combination methods, there is a lack of comprehensive tools that&#10;integrate diverse methodologies. In this study, we present dtComb, a&#10;comprehensive R package and web tool designed to address the&#10;limitations of existing diagnostic test combination platforms. One of&#10;the unique contributions of dtComb is offering a range of 142 methods&#10;to combine two diagnostic tests, including linear, non-linear, machine&#10;learning algorithms, and mathematical operators. Another significant&#10;contribution of dtComb is its inclusion of advanced tools for ROC&#10;analysis, diagnostic performance metrics, and visual outputs such as&#10;sensitivity-specificity curves. Furthermore, dtComb offers&#10;classification functions for new observations, making it an&#10;easy-to-use tool for clinicians and researchers. The web-based version&#10;is also available at &lt;https://biotools.erciyes.edu.tr/dtComb/&gt; for&#10;non-R users, providing an intuitive interface for test combination and&#10;model training."/>
  <meta property="og:locale" content="en_US"/>

  <!--  https://dev.twitter.com/cards/types/summary -->
  <meta property="twitter:card" content="summary"/>
  <meta property="twitter:title" content="dtComb: A Comprehensive R Library and Web Tool for Combining Diagnostic Tests"/>
  <meta property="twitter:description" content="The combination of diagnostic tests has become a crucial area of&#10;research, aiming to improve the accuracy and robustness of medical&#10;diagnostics. While existing tools focus primarily on linear&#10;combination methods, there is a lack of comprehensive tools that&#10;integrate diverse methodologies. In this study, we present dtComb, a&#10;comprehensive R package and web tool designed to address the&#10;limitations of existing diagnostic test combination platforms. One of&#10;the unique contributions of dtComb is offering a range of 142 methods&#10;to combine two diagnostic tests, including linear, non-linear, machine&#10;learning algorithms, and mathematical operators. Another significant&#10;contribution of dtComb is its inclusion of advanced tools for ROC&#10;analysis, diagnostic performance metrics, and visual outputs such as&#10;sensitivity-specificity curves. Furthermore, dtComb offers&#10;classification functions for new observations, making it an&#10;easy-to-use tool for clinicians and researchers. The web-based version&#10;is also available at &lt;https://biotools.erciyes.edu.tr/dtComb/&gt; for&#10;non-R users, providing an intuitive interface for test combination and&#10;model training."/>

  <!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
  <meta name="citation_title" content="dtComb: A Comprehensive R Library and Web Tool for Combining Diagnostic Tests"/>
  <meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2025-036"/>
  <meta name="citation_pdf_url" content="RJ-2025-036.pdf"/>
  <meta name="citation_volume" content="17"/>
  <meta name="citation_issue" content="4"/>
  <meta name="citation_doi" content="10.32614/RJ-2025-036"/>
  <meta name="citation_journal_title" content="The R Journal"/>
  <meta name="citation_issn" content="2073-4859"/>
  <meta name="citation_firstpage" content="80"/>
  <meta name="citation_lastpage" content="102"/>
  <meta name="citation_fulltext_world_readable" content=""/>
  <meta name="citation_online_date" content="2026/02/03"/>
  <meta name="citation_publication_date" content="2026/02/03"/>
  <meta name="citation_author" content="S. Ilayda Yerlitaş Taştan"/>
  <meta name="citation_author_institution" content="Department of Biostatistics"/>
  <meta name="citation_author" content="Serra Bersan Gengeç"/>
  <meta name="citation_author_institution" content="Department of Biostatistics"/>
  <meta name="citation_author" content="Necla Koçhan"/>
  <meta name="citation_author_institution" content="Department of Mathematics"/>
  <meta name="citation_author" content="Ertürk Zararsız"/>
  <meta name="citation_author_institution" content="Department of Biostatistics"/>
  <meta name="citation_author" content="Selçuk Korkmaz"/>
  <meta name="citation_author_institution" content="Department of Biostatistics"/>
  <meta name="citation_author" content="Zararsız"/>
  <meta name="citation_author_institution" content="Department of Biostatistics"/>
  <!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Construction and evaluation of optimal diagnostic tests with application to hepatocellular carcinoma diagnosis;citation_author=Ainesh Sewak;citation_author=Sandra Siegfried;citation_author=Torsten Hothorn"/>
  <meta name="citation_reference" content="citation_title=Likelihood ratio combination of multiple biomarkers via smoothing spline estimated densities;citation_publisher=Wiley Online Library;citation_volume=43;citation_author=Zhiyuan Du;citation_author=Pang Du;citation_author=Aiyi Liu"/>
  <meta name="citation_reference" content="citation_title=AI algorithm combined with RNA editing-based blood biomarkers to discriminate bipolar from major depressive disorders in an external validation multicentric cohort;citation_publisher=Elsevier;citation_volume=356;citation_author=Nicolas Salvetat;citation_author=Francisco Jesus Checa-Robles;citation_author=Aurélie Delacrétaz;citation_author=Christopher Cayzac;citation_author=Benjamin Dubuc;citation_author=Diana Vetter;citation_author=Jacques Dainat;citation_author=Jean-Philippe Lang;citation_author=Franziska Gamma;citation_author=Dinah Weissmann"/>
  <meta name="citation_reference" content="citation_title=Explainable machine-learning models for COVID-19 prognosis prediction using clinical, laboratory and radiomic features;citation_publisher=IEEE;citation_volume=11;citation_author=Francesco Prinzi;citation_author=Carmelo Militello;citation_author=Nicola Scichilone;citation_author=Salvatore Gaglio;citation_author=Salvatore Vitabile"/>
  <meta name="citation_reference" content="citation_title=A min–max combination of biomarkers to improve diagnostic accuracy;citation_publisher=Wiley Online Library;citation_volume=30;citation_author=Chunling Liu;citation_author=Aiyi Liu;citation_author=Susan Halabi"/>
  <meta name="citation_reference" content="citation_title=Binary classification using multivariate receiver operating characteristic curve for continuous data;citation_publisher=Taylor &amp; Francis;citation_volume=26;citation_author=G Sameera;citation_author=R Vishnu Vardhan;citation_author=KVS Sarma"/>
  <meta name="citation_reference" content="citation_title=Linear combinations of multiple diagnostic markers;citation_publisher=Taylor &amp; Francis;citation_volume=88;citation_author=John Q Su;citation_author=Jun S Liu"/>
  <meta name="citation_reference" content="citation_title=Combining diagnostic test results to increase accuracy;citation_publisher=Oxford University Press;citation_volume=1;citation_author=Margaret Sullivan Pepe;citation_author=Mary Lou Thompson"/>
  <meta name="citation_reference" content="citation_title=Classification and selection of biomarkers in genomic data using LASSO;citation_publisher=Wiley Online Library;citation_volume=2005;citation_author=Debashis Ghosh;citation_author=Arul M Chinnaiyan"/>
  <meta name="citation_reference" content="citation_title=Neurofilament light chain is elevated in patients with newly diagnosed idiopathic intracranial hypertension: A prospective study;citation_publisher=SAGE Publications Sage UK: London, England;citation_volume=44;citation_author=Katrine Svart;citation_author=Johanne Juhl Korsbæk;citation_author=Rigmor Højland Jensen;citation_author=Tina Parkner;citation_author=Cindy Søndersø Knudsen;citation_author=Steen Gregers Hasselbalch;citation_author=Snorre Malm Hagen;citation_author=Elisabeth Arnberg Wibroe;citation_author=Laleh Dehghani Molander;citation_author=Dagmar Beier"/>
  <meta name="citation_reference" content="citation_title=AST/ALT ratio is an independent risk factor for diabetic retinopathy: A cross-sectional study;citation_publisher=LWW;citation_volume=103;citation_author=Jian Luo;citation_author=Fang Yu;citation_author=Haifeng Zhou;citation_author=Xueyan Wu;citation_author=Quan Zhou;citation_author=Qin Liu;citation_author=Shenglian Gan"/>
  <meta name="citation_reference" content="citation_title=Significance of neutrophil to lymphocyte ratio (NLR) and platelet lymphocyte ratio (PLR) in diabetic foot ulcer and potential new therapeutic targets;citation_publisher=SAGE Publications Sage CA: Los Angeles, CA;citation_volume=23;citation_author=Dragos Serban;citation_author=Nikolaos Papanas;citation_author=Ana Maria Dascalu;citation_author=Peter Kempler;citation_author=Itamar Raz;citation_author=Ali A Rizvi;citation_author=Manfredi Rizzo;citation_author=Corneliu Tudor;citation_author=Mihail Silviu Tudosie;citation_author=Denisa Tanasescu;citation_author=Significance of neutrophil to lymphocyte ratio (NLR) and platelet lymphocyte ratio (PLR) in diabetic foot ulcer and potential new therapeutic targets"/>
  <meta name="citation_reference" content="citation_title=Advancements in medical diagnosis and treatment through machine learning: A review;citation_publisher=Wiley Online Library;citation_volume=41;citation_author=Mohammad Ahsan;citation_author=Anam Khan;citation_author=Kaif Rehman Khan;citation_author=Bam Bahadur Sinha;citation_author=Anamika Sharma"/>
  <meta name="citation_reference" content="citation_title=By artificial intelligence algorithms and machine learning models to diagnosis cancer;citation_publisher=Elsevier;citation_volume=80;citation_author=Seema Agarwal;citation_author=Ajay Singh Yadav;citation_author=Vennapoosa Dinesh;citation_author=Kolluru Sai Sri Vatsav;citation_author=Kolluru Sai Surya Prakash;citation_author=Sushma Jaiswal"/>
  <meta name="citation_reference" content="citation_title=A game changer for bipolar disorder diagnosis using RNA editing-based biomarkers;citation_publisher=Nature Publishing Group UK London;citation_volume=12;citation_author=Nicolas Salvetat;citation_author=Francisco Jesus Checa-Robles;citation_author=Vipul Patel;citation_author=Christopher Cayzac;citation_author=Benjamin Dubuc;citation_author=Fabrice Chimienti;citation_author=Jean-Daniel Abraham;citation_author=Pierrick Dupré;citation_author=Diana Vetter;citation_author=Sandie Méreuze;citation_author=A game changer for bipolar disorder diagnosis using RNA editing-based biomarkers"/>
  <meta name="citation_reference" content="citation_title=Comparison of diagnostic accuracy of models combining the renal biomarkers in predicting renal scarring in pediatric population with vesicoureteral reflux (VUR);citation_publisher=Springer;citation_volume=192;citation_author=Sachit Ganapathy;citation_author=Harichandrakumar KT;citation_author=Bibekanand Jindal;citation_author=Prathibha S Naik;citation_author=Sreekumaran Nair N"/>
  <meta name="citation_reference" content="citation_title=Diagnosing diabetes mellitus using machine learning techniques;citation_volume=8;citation_author=Mazen Alzyoud;citation_author=Raed Alazaidah;citation_author=Mohammad Aljaidi;citation_author=Ghassan Samara;citation_author=M Qasem;citation_author=Muhammad Khalid;citation_author=Najah Al-Shanableh"/>
  <meta name="citation_reference" content="citation_title=Statistical learning approaches in diagnosing patients with nontraumatic acute abdomen;citation_volume=24;citation_author=GÖKMEN Zararsiz;citation_author=Hizir Yakup Akyildiz;citation_author=DİNÇER GÖKSÜLÜK;citation_author=Selcuk Korkmaz;citation_author=AHMET ÖZTÜRK"/>
  <meta name="citation_reference" content="citation_title=Discovering biomarkers associated and predicting cardiovascular disease with high accuracy using a novel nexus of machine learning techniques for precision medicine;citation_publisher=Nature Publishing Group UK London;citation_volume=14;citation_author=William DeGroat;citation_author=Habiba Abdelhalim;citation_author=Kush Patel;citation_author=Dinesh Mendhe;citation_author=Saman Zeeshan;citation_author=Zeeshan Ahmed"/>
  <meta name="citation_reference" content="citation_title=mROC: A computer program for combining tumour markers in predicting disease states;citation_publisher=Elsevier;citation_volume=66;citation_author=Andrew Kramar;citation_author=David Faraggi;citation_author=Antoine Fortuné;citation_author=Benjamin Reiser"/>
  <meta name="citation_reference" content="citation_title=Visualizing the decision rules behind the ROC curves: Understanding the classification process;citation_publisher=Springer;citation_volume=105;citation_author=Sonia Pérez-Fernández;citation_author=Pablo Martı́nez-Camblor;citation_author=Peter Filzmoser;citation_author=Norberto Corral"/>
  <meta name="citation_reference" content="citation_title=Two simple algorithms on linear combination of multiple biomarkers to maximize partial area under the ROC curve;citation_publisher=Elsevier;citation_volume=88;citation_author=Wenbao Yu;citation_author=Taesung Park"/>
  <meta name="citation_reference" content="citation_title=Building predictive models in R using the caret package;citation_volume=28;citation_author=Max Kuhn"/>
  <meta name="citation_reference" content="citation_title=Gam: Generalized additive models.;citation_author=Trevor Hastie"/>
  <meta name="citation_reference" content="citation_title=Asymptotic efficiency of logistic regression relative to linear discriminant analysis;citation_publisher=Oxford University Press;citation_volume=78;citation_author=S Ruiz-Velasco"/>
  <meta name="citation_reference" content="citation_title=The efficiency of logistic regression compared to normal discriminant analysis;citation_publisher=Taylor &amp; Francis;citation_volume=70;citation_author=Bradley Efron"/>
  <meta name="citation_reference" content="citation_title=Analysis of binary data;citation_publisher=Chapman; Hall/CRC;citation_author=D. R. Cox;citation_author=E. J. Snell"/>
  <meta name="citation_reference" content="citation_title=A bedside scoring system (“Candida score”) for early antifungal treatment in nonneutropenic critically ill patients with Candida colonization;citation_publisher=LWW;citation_volume=34;citation_author=Cristóbal León;citation_author=Sergio Ruiz-Santana;citation_author=Pedro Saavedra;citation_author=Benito Almirante;citation_author=Juan Nolla-Salas;citation_author=Francisco Álvarez-Lerma;citation_author=José Garnacho-Montero;citation_author=Marı́a Ángeles León;citation_author=EPCAN Study Group;citation_author=A bedside scoring system (“Candida score”) for early antifungal treatment in nonneutropenic critically ill patients with Candida colonization"/>
  <meta name="citation_reference" content="citation_title=Combining predictors for classification using the area under the receiver operating characteristic curve;citation_publisher=Oxford University Press;citation_volume=62;citation_author=Margaret Sullivan Pepe;citation_author=Tianxi Cai;citation_author=Gary Longton"/>
  <meta name="citation_reference" content="citation_title=The statistical evaluation of medical tests for classification and prediction;citation_publisher=Oxford university press;citation_author=Margaret Sullivan Pepe"/>
  <meta name="citation_reference" content="citation_title=Linear combination methods to improve diagnostic/prognostic accuracy on future observations;citation_publisher=SAGE Publications Sage UK: London, England;citation_volume=25;citation_author=Le Kang;citation_author=Aiyi Liu;citation_author=Lili Tian"/>
  <meta name="citation_reference" content="citation_title=Classification into two multivariate normal distributions with different covariance matrices;citation_publisher=JSTOR;citation_author=Theodore W Anderson;citation_author=Raghu Raj Bahadur"/>
  <meta name="citation_reference" content="citation_title=Tools to identify linear combination of prognostic factors which maximizes area under receiver operator curve;citation_publisher=Springer;citation_volume=4;citation_author=Nicolae Todor;citation_author=Irina Todor;citation_author=Gavril Săplăcan"/>
  <meta name="citation_reference" content="citation_title=An introduction to statistical learning;citation_publisher=Springer;citation_volume=112;citation_author=Gareth James;citation_author=Daniela Witten;citation_author=Trevor Hastie;citation_author=Robert Tibshirani;citation_author=An introduction to statistical learning"/>
  <meta name="citation_reference" content="citation_title=Regularization paths for generalized linear models via coordinate descent;citation_publisher=NIH Public Access;citation_volume=33;citation_author=Jerome Friedman;citation_author=Trevor Hastie;citation_author=Rob Tibshirani"/>
  <meta name="citation_reference" content="citation_title=Using linear and natural cubic splines, SITAR, and latent trajectory models to characterise nonlinear longitudinal growth trajectories in cohort studies;citation_publisher=Springer;citation_volume=22;citation_author=Ahmed Elhakeem;citation_author=Rachael A Hughes;citation_author=Kate Tilling;citation_author=Diana L Cousminer;citation_author=Stefan A Jackowski;citation_author=Tim J Cole;citation_author=Alex SF Kwong;citation_author=Zheyuan Li;citation_author=Struan FA Grant;citation_author=Adam DG Baxter-Jones;citation_author=Using linear and natural cubic splines, SITAR, and latent trajectory models to characterise nonlinear longitudinal growth trajectories in cohort studies"/>
  <meta name="citation_reference" content="citation_title=Distance measures for classification of numerical features;citation_publisher=Tampere University of Technology;citation_author=Georgy Minaev;citation_author=Robert Piché;citation_author=Ari Visa"/>
  <meta name="citation_reference" content="citation_title=A comparative study on distance measuring approaches for clustering;citation_publisher=Citeseer;citation_volume=2;citation_author=Shraddha Pandit;citation_author=Suchita Gupta;citation_author=A comparative study on distance measuring approaches for clustering"/>
  <meta name="citation_reference" content="citation_title=Comprehensive survey on distance/similarity measures between probability density functions;citation_volume=1;citation_author=Sung-Hyuk Cha"/>
  <meta name="citation_reference" content="citation_title=Optimal linear combinations of multiple diagnostic biomarkers based on Youden index;citation_publisher=Wiley Online Library;citation_volume=33;citation_author=Jingjing Yin;citation_author=Lili Tian"/>
  <meta name="citation_reference" content="citation_title=epiR: Tools for the analysis of epidemiological data.;citation_author=Mark Stevenson;citation_author=Telmo Nunes;citation_author=Cord Heuer;citation_author=Jonathon Marshall;citation_author=Javier Sanchez;citation_author=Ron Thornton;citation_author=Jeno Reiczigel;citation_author=Jim Robison-Cox;citation_author=Paola Sebastiani;citation_author=Peter Solymos;citation_author=epiR: Tools for the analysis of epidemiological data."/>
  <meta name="citation_reference" content="citation_title=Devtools: Tools to make developing R packages easier;citation_author=H. Wickham;citation_author=J. Hester;citation_author=W. Chang;citation_author=J. Bryan"/>
  <meta name="citation_reference" content="citation_title=roxygen2: In-source documentation for R;citation_author=H. Wickham;citation_author=P. Danenberg;citation_author=M. Eugster"/>
  <meta name="citation_reference" content="citation_title=Testthat: Get started with testing;citation_author=H. Wickham"/>
  <meta name="citation_reference" content="citation_title=Programming validation: Perspectives and strategies;citation_author=Parag Shiralkar"/>
  <meta name="citation_reference" content="citation_title=The value of D-dimer test in the diagnosis of patients with nontraumatic acute abdomen;citation_volume=16;citation_author=Hizir Yakup Akyildiz;citation_author=Erdogan Sozuer;citation_author=Alper Akcan;citation_author=Can Kuçuk;citation_author=Tarik Artis;citation_author=İsmail Biri;citation_author=Namık Yılmaz"/>
  <meta name="citation_reference" content="citation_title=Shiny: Web application framework for R;citation_author=Winston Chang;citation_author=Joe Cheng;citation_author=JJ Allaire;citation_author=Carson Sievert;citation_author=Barret Schloerke;citation_author=Yihui Xie;citation_author=Jeff Allen;citation_author=Jonathan McPherson;citation_author=Alan Dipert;citation_author=Barbara Borges"/>
  <meta name="citation_reference" content="citation_title=ROCR: Visualizing classifier performance in R;citation_publisher=Oxford University Press;citation_volume=21;citation_author=Tobias Sing;citation_author=Oliver Sander;citation_author=Niko Beerenwinkel;citation_author=Thomas Lengauer"/>
  <meta name="citation_reference" content="citation_title=pROC: An open-source package for R and S+ to analyze and compare ROC curves;citation_publisher=Springer;citation_volume=12;citation_author=Xavier Robin;citation_author=Natacha Turck;citation_author=Alexandre Hainard;citation_author=Natalia Tiberti;citation_author=Frédérique Lisacek;citation_author=Jean-Charles Sanchez;citation_author=Markus Müller"/>
  <meta name="citation_reference" content="citation_title=PRROC: Computing and visualizing precision-recall and receiver operating characteristic curves in R;citation_publisher=Oxford University Press;citation_volume=31;citation_author=Jan Grau;citation_author=Ivo Grosse;citation_author=Jens Keilwagen"/>
  <meta name="citation_reference" content="citation_title=plotROC: A tool for plotting ROC curves;citation_publisher=NIH Public Access;citation_volume=79;citation_author=Michael C Sachs"/>
  <meta name="citation_reference" content="citation_title=Linear combination of leukocyte count and D-Dimer levels in the diagnosis of patients with non-traumatic acute abdomen;citation_volume=5;citation_doi=10.37990/medr.1166531;citation_author=G. Ertürk Zararsız"/>
  <meta name="citation_reference" content="citation_title=Comparing the min–max–median/IQR approach with the min–max approach, logistic regression and XGBoost, maximising the Youden index;citation_volume=15;citation_doi=10.3390/sym15030756;citation_author=R. Aznar-Gimeno;citation_author=L. M. Esteban;citation_author=G. Sanz;citation_author=R. del-Hoyo-Alonso"/>
  <meta name="citation_reference" content="citation_title=Amyloid-$\beta$ PET—correlation with cerebrospinal fluid biomarkers and prediction of Alzheimer’s disease diagnosis in a memory clinic;citation_publisher=Public Library of Science San Francisco, CA USA;citation_volume=14;citation_author=Ebba Gløersen Müller;citation_author=Trine Holt Edwin;citation_author=Caroline Stokke;citation_author=Sigrid Stensby Navelsaker;citation_author=Almira Babovic;citation_author=Nenad Bogdanovic;citation_author=Anne Brita Knapskog;citation_author=Mona Elisabeth Revheim"/>
  <meta name="citation_reference" content="citation_title=Combining multiple multimodal speech features into an interpretable index score for capturing disease progression in Amyotrophic Lateral Sclerosis;citation_volume=2353;citation_author=Michael Neumann;citation_author=Hardik Kothare;citation_author=Vikram Ramanarayanan"/>
  <meta name="citation_reference" content="citation_title=A stepwise algorithm for linearly combining biomarkers under Youden index maximization;citation_publisher=MDPI;citation_volume=10;citation_author=Rocı́o Aznar-Gimeno;citation_author=Luis M Esteban;citation_author=Rafael del-Hoyo-Alonso;citation_author=Ángel Borque-Fernando;citation_author=Gerardo Sanz"/>
  <meta name="citation_reference" content="citation_title=When does combining markers improve classification performance and what are implications for practice?;citation_publisher=Wiley Online Library;citation_volume=32;citation_author=Aasthaa Bansal;citation_author=Margaret Sullivan Pepe"/>
  <meta name="citation_reference" content="citation_title=The neutrophil-to-lymphocyte ratio: A narrative review;citation_publisher=ecancer Global Foundation;citation_volume=10;citation_author=Sara Socorro Faria;citation_author=Paulo César Fernandes Jr;citation_author=Marcelo José Barbosa Silva;citation_author=Vladmir C Lima;citation_author=Wagner Fontes;citation_author=Ruffo Freitas-Junior;citation_author=Agda Karina Eterovic;citation_author=Patrice Forget"/>
  <meta name="citation_reference" content="citation_title=The AST/ALT ratio as an indicator of cirrhosis in patients with PBC;citation_publisher=Wiley Online Library;citation_volume=26;citation_author=Helena Nyblom;citation_author=Einar Björnsson;citation_author=Magnus Simrén;citation_author=Frank Aldenborg;citation_author=Sven Almer;citation_author=Rolf Olsson"/>
  <meta name="citation_reference" content="citation_title=Scikit-learn: Machine learning in Python;citation_publisher=JMLR. org;citation_volume=12;citation_author=Fabian Pedregosa;citation_author=Gaël Varoquaux;citation_author=Alexandre Gramfort;citation_author=Vincent Michel;citation_author=Bertrand Thirion;citation_author=Olivier Grisel;citation_author=Mathieu Blondel;citation_author=Peter Prettenhofer;citation_author=Ron Weiss;citation_author=Vincent Dubourg;citation_author=Scikit-learn: Machine learning in Python"/>
  <meta name="citation_reference" content="citation_title=TensorFlow: Large-scale machine learning on heterogeneous systems;citation_author=Martín Abadi;citation_author=Ashish Agarwal;citation_author=Paul Barham;citation_author=Eugene Brevdo;citation_author=Zhifeng Chen;citation_author=Craig Citro;citation_author=Greg S. Corrado;citation_author=Andy Davis;citation_author=Jeffrey Dean;citation_author=Matthieu Devin;citation_author=Sanjay Ghemawat;citation_author=Ian Goodfellow;citation_author=Andrew Harp;citation_author=Geoffrey Irving;citation_author=Michael Isard;citation_author=Yangqing Jia;citation_author=Rafal Jozefowicz;citation_author=Lukasz Kaiser;citation_author=Manjunath Kudlur;citation_author=Josh Levenberg;citation_author=Dandelion Mané;citation_author=Rajat Monga;citation_author=Sherry Moore;citation_author=Derek Murray;citation_author=Chris Olah;citation_author=Mike Schuster;citation_author=Jonathon Shlens;citation_author=Benoit Steiner;citation_author=Ilya Sutskever;citation_author=Kunal Talwar;citation_author=Paul Tucker;citation_author=Vincent Vanhoucke;citation_author=Vijay Vasudevan;citation_author=Fernanda Viégas;citation_author=Oriol Vinyals;citation_author=Pete Warden;citation_author=Martin Wattenberg;citation_author=Martin Wicke;citation_author=Yuan Yu;citation_author=Xiaoqiang Zheng"/>
  <!--radix_placeholder_rmarkdown_metadata-->

  <script type="text/json" id="radix-rmarkdown-metadata">
  {"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","date_received","journal","volume","issue","slug","packages","preview","bibliography","CTV","legacy_pdf","legacy_converted","output","draft","pdf_url","citation_url","doi","creative_commons","csl"]}},"value":[{"type":"character","attributes":{},"value":["dtComb: A Comprehensive R Library and Web Tool for Combining Diagnostic Tests"]},{"type":"character","attributes":{},"value":["The combination of diagnostic tests has become a crucial area of\nresearch, aiming to improve the accuracy and robustness of medical\ndiagnostics. While existing tools focus primarily on linear\ncombination methods, there is a lack of comprehensive tools that\nintegrate diverse methodologies. In this study, we present dtComb, a\ncomprehensive R package and web tool designed to address the\nlimitations of existing diagnostic test combination platforms. One of\nthe unique contributions of dtComb is offering a range of 142 methods\nto combine two diagnostic tests, including linear, non-linear, machine\nlearning algorithms, and mathematical operators. Another significant\ncontribution of dtComb is its inclusion of advanced tools for ROC\nanalysis, diagnostic performance metrics, and visual outputs such as\nsensitivity-specificity curves. Furthermore, dtComb offers\nclassification functions for new observations, making it an\neasy-to-use tool for clinicians and researchers. The web-based version\nis also available at <https://biotools.erciyes.edu.tr/dtComb/> for\nnon-R users, providing an intuitive interface for test combination and\nmodel training.\n"]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["S. Ilayda Yerlitaş Taştan"]},{"type":"character","attributes":{},"value":["Department of Biostatistics"]},{"type":"character","attributes":{},"value":["Erciyes University","Türkiye","(ORCiD: 0000-0003-2830-3006)","[ilaydayerlitas340@gmail.com](ilaydayerlitas340@gmail.com){.uri}\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Serra Bersan Gengeç"]},{"type":"character","attributes":{},"value":["Department of Biostatistics"]},{"type":"character","attributes":{},"value":["Erciyes University","Türkiye","[serrabersan@gmail.com](serrabersan@gmail.com){.uri}\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Necla Koçhan"]},{"type":"character","attributes":{},"value":["Department of Mathematics"]},{"type":"character","attributes":{},"value":["Izmir University of Economics","Türkiye","(ORCiD: 0000-0003-2355-4826)","[necla.kayaalp@gmail.com](necla.kayaalp@gmail.com){.uri}\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Ertürk Zararsız"]},{"type":"character","attributes":{},"value":["Department of Biostatistics"]},{"type":"character","attributes":{},"value":["Erciyes University","Türkiye","(ORCiD if desired)","[gozdeerturk9@gmail.com](gozdeerturk9@gmail.com){.uri}\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Selçuk Korkmaz"]},{"type":"character","attributes":{},"value":["Department of Biostatistics"]},{"type":"character","attributes":{},"value":["Trakya University","Türkiye","(ORCiD if desired)","[selcukorkmaz@gmail.com](selcukorkmaz@gmail.com){.uri}\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Zararsız"]},{"type":"character","attributes":{},"value":["Department of Biostatistics"]},{"type":"character","attributes":{},"value":["Erciyes University","Türkiye","(ORCiD: 0000-0001-5801-1835)","[gokmen.zararsiz@gmail.com](gokmen.zararsiz@gmail.com){.uri}\n"]}]}]},{"type":"character","attributes":{},"value":["2026-02-03"]},{"type":"character","attributes":{},"value":["2024-03-08"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"integer","attributes":{},"value":[80]},{"type":"integer","attributes":{},"value":[102]}]},{"type":"integer","attributes":{},"value":[17]},{"type":"integer","attributes":{},"value":[4]},{"type":"character","attributes":{},"value":["RJ-2025-036"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"list","attributes":{},"value":[]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["preview.png"]},{"type":"character","attributes":{},"value":["dtCombreferences.bib"]},{"type":"list","attributes":{},"value":[]},{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[true]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc","mathjax","md_extension"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js"]},{"type":"character","attributes":{},"value":["-tex_math_single_backslash"]}]}]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["RJ-2025-036.pdf"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2025-036"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2025-036"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"character","attributes":{},"value":["/home/mitchell/R/x86_64-pc-linux-gnu-library/4.5/rjtools/rjournal.csl"]}]}
  </script>
  <!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["diff.pdf","diff.tex","dtComb.tex","dtComb2.tex","dtComb3.tex","dtCombreferences.bib","Figure/Figure_1.pdf","Figure/Figure_1.png","Figure/Figure_2.pdf","Figure/Figure_2.png","Figure/Figure_3.pdf","Figure/Figure_3.png","Figure/Figure_4.pdf","Figure/Figure_4.png","Figure/Figure_5.pdf","Figure/Figure_5.png","manuscript.pdf","RJ-2025-036_files/anchor-4.2.2/anchor.min.js","RJ-2025-036_files/bowser-1.9.3/bowser.min.js","RJ-2025-036_files/distill-2.2.21/template.v2.js","RJ-2025-036_files/header-attrs-2.30/header-attrs.js","RJ-2025-036_files/jquery-3.6.0/jquery-3.6.0.js","RJ-2025-036_files/jquery-3.6.0/jquery-3.6.0.min.js","RJ-2025-036_files/jquery-3.6.0/jquery-3.6.0.min.map","RJ-2025-036_files/popper-2.6.0/popper.min.js","RJ-2025-036_files/tippy-6.2.7/tippy-bundle.umd.min.js","RJ-2025-036_files/tippy-6.2.7/tippy-light-border.css","RJ-2025-036_files/tippy-6.2.7/tippy.css","RJ-2025-036_files/tippy-6.2.7/tippy.umd.min.js","RJ-2025-036_files/webcomponents-2.0.0/webcomponents.js","RJournal.sty","RJwrapper.log","RJwrapper.tex","Rlogo-5.png"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
  <!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

  <style type="text/css">

  body {
    background-color: white;
  }

  .pandoc-table {
    width: 100%;
  }

  .pandoc-table>caption {
    margin-bottom: 10px;
  }

  .pandoc-table th:not([align]) {
    text-align: left;
  }

  .pagedtable-footer {
    font-size: 15px;
  }

  d-byline .byline {
    grid-template-columns: 2fr 2fr;
  }

  d-byline .byline h3 {
    margin-block-start: 1.5em;
  }

  d-byline .byline .authors-affiliations h3 {
    margin-block-start: 0.5em;
  }

  .authors-affiliations .orcid-id {
    width: 16px;
    height:16px;
    margin-left: 4px;
    margin-right: 4px;
    vertical-align: middle;
    padding-bottom: 2px;
  }

  d-title .dt-tags {
    margin-top: 1em;
    grid-column: text;
  }

  .dt-tags .dt-tag {
    text-decoration: none;
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0em 0.4em;
    margin-right: 0.5em;
    margin-bottom: 0.4em;
    font-size: 70%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  d-article table.gt_table td,
  d-article table.gt_table th {
    border-bottom: none;
    font-size: 100%;
  }

  .html-widget {
    margin-bottom: 2.0em;
  }

  .l-screen-inset {
    padding-right: 16px;
  }

  .l-screen .caption {
    margin-left: 10px;
  }

  .shaded {
    background: rgb(247, 247, 247);
    padding-top: 20px;
    padding-bottom: 20px;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .html-widget {
    margin-bottom: 0;
    border: 1px solid rgba(0, 0, 0, 0.1);
  }

  .shaded .shaded-content {
    background: white;
  }

  .text-output {
    margin-top: 0;
    line-height: 1.5em;
  }

  .hidden {
    display: none !important;
  }

  hr.section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    margin: 0px;
  }


  d-byline {
    border-top: none;
  }

  d-article {
    padding-top: 2.5rem;
    padding-bottom: 30px;
    border-top: none;
  }

  d-appendix {
    padding-top: 30px;
  }

  d-article>p>img {
    width: 100%;
  }

  d-article h2 {
    margin: 1rem 0 1.5rem 0;
  }

  d-article h3 {
    margin-top: 1.5rem;
  }

  d-article iframe {
    border: 1px solid rgba(0, 0, 0, 0.1);
    margin-bottom: 2.0em;
    width: 100%;
  }

  /* Tweak code blocks */

  d-article div.sourceCode code,
  d-article pre code {
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  }

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: auto;
  }

  d-article div.sourceCode {
    background-color: white;
  }

  d-article div.sourceCode pre {
    padding-left: 10px;
    font-size: 12px;
    border-left: 2px solid rgba(0,0,0,0.1);
  }

  d-article pre {
    font-size: 12px;
    color: black;
    background: none;
    margin-top: 0;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  d-article pre a {
    border-bottom: none;
  }

  d-article pre a:hover {
    border-bottom: none;
    text-decoration: underline;
  }

  d-article details {
    grid-column: text;
    margin-bottom: 0.8em;
  }

  @media(min-width: 768px) {

  d-article pre,
  d-article div.sourceCode,
  d-article div.sourceCode pre {
    overflow: visible !important;
  }

  d-article div.sourceCode pre {
    padding-left: 18px;
    font-size: 14px;
  }

  /* tweak for Pandoc numbered line within distill */
  d-article pre.numberSource code > span {
      left: -2em;
  }

  d-article pre {
    font-size: 14px;
  }

  }

  figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  /* CSS for d-contents */

  .d-contents {
    grid-column: text;
    color: rgba(0,0,0,0.8);
    font-size: 0.9em;
    padding-bottom: 1em;
    margin-bottom: 1em;
    padding-bottom: 0.5em;
    margin-bottom: 1em;
    padding-left: 0.25em;
    justify-self: start;
  }

  @media(min-width: 1000px) {
    .d-contents.d-contents-float {
      height: 0;
      grid-column-start: 1;
      grid-column-end: 4;
      justify-self: center;
      padding-right: 3em;
      padding-left: 2em;
    }
  }

  .d-contents nav h3 {
    font-size: 18px;
    margin-top: 0;
    margin-bottom: 1em;
  }

  .d-contents li {
    list-style-type: none
  }

  .d-contents nav > ul {
    padding-left: 0;
  }

  .d-contents ul {
    padding-left: 1em
  }

  .d-contents nav ul li {
    margin-top: 0.6em;
    margin-bottom: 0.2em;
  }

  .d-contents nav a {
    font-size: 13px;
    border-bottom: none;
    text-decoration: none
    color: rgba(0, 0, 0, 0.8);
  }

  .d-contents nav a:hover {
    text-decoration: underline solid rgba(0, 0, 0, 0.6)
  }

  .d-contents nav > ul > li > a {
    font-weight: 600;
  }

  .d-contents nav > ul > li > ul {
    font-weight: inherit;
  }

  .d-contents nav > ul > li > ul > li {
    margin-top: 0.2em;
  }


  .d-contents nav ul {
    margin-top: 0;
    margin-bottom: 0.25em;
  }

  .d-article-with-toc h2:nth-child(2) {
    margin-top: 0;
  }


  /* Figure */

  .figure {
    position: relative;
    margin-bottom: 2.5em;
    margin-top: 1.5em;
  }

  .figure .caption {
    color: rgba(0, 0, 0, 0.6);
    font-size: 12px;
    line-height: 1.5em;
  }

  .figure img.external {
    background: white;
    border: 1px solid rgba(0, 0, 0, 0.1);
    box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
    padding: 18px;
    box-sizing: border-box;
  }

  .figure .caption a {
    color: rgba(0, 0, 0, 0.6);
  }

  .figure .caption b,
  .figure .caption strong, {
    font-weight: 600;
    color: rgba(0, 0, 0, 1.0);
  }

  /* Citations */

  d-article .citation {
    color: inherit;
    cursor: inherit;
  }

  div.hanging-indent{
    margin-left: 1em; text-indent: -1em;
  }

  /* Citation hover box */

  .tippy-box[data-theme~=light-border] {
    background-color: rgba(250, 250, 250, 0.95);
  }

  .tippy-content > p {
    margin-bottom: 0;
    padding: 2px;
  }


  /* Tweak 1000px media break to show more text */

  @media(min-width: 1000px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 16px;
    }

    .grid {
      grid-column-gap: 16px;
    }

    d-article {
      font-size: 1.06rem;
      line-height: 1.7em;
    }
    figure .caption, .figure .caption, figure figcaption {
      font-size: 13px;
    }
  }

  @media(min-width: 1180px) {
    .base-grid,
    distill-header,
    d-title,
    d-abstract,
    d-article,
    d-appendix,
    distill-appendix,
    d-byline,
    d-footnote-list,
    d-citation-list,
    distill-footer {
      grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
      grid-column-gap: 32px;
    }

    .grid {
      grid-column-gap: 32px;
    }
  }


  /* Get the citation styles for the appendix (not auto-injected on render since
     we do our own rendering of the citation appendix) */

  d-appendix .citation-appendix,
  .d-appendix .citation-appendix {
    font-size: 11px;
    line-height: 15px;
    border-left: 1px solid rgba(0, 0, 0, 0.1);
    padding-left: 18px;
    border: 1px solid rgba(0,0,0,0.1);
    background: rgba(0, 0, 0, 0.02);
    padding: 10px 18px;
    border-radius: 3px;
    color: rgba(150, 150, 150, 1);
    overflow: hidden;
    margin-top: -12px;
    white-space: pre-wrap;
    word-wrap: break-word;
  }

  /* Include appendix styles here so they can be overridden */

  d-appendix {
    contain: layout style;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-top: 60px;
    margin-bottom: 0;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
    color: rgba(0,0,0,0.5);
    padding-top: 60px;
    padding-bottom: 48px;
  }

  d-appendix h3 {
    grid-column: page-start / text-start;
    font-size: 15px;
    font-weight: 500;
    margin-top: 1em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.65);
  }

  d-appendix h3 + * {
    margin-top: 1em;
  }

  d-appendix ol {
    padding: 0 0 0 15px;
  }

  @media (min-width: 768px) {
    d-appendix ol {
      padding: 0 0 0 30px;
      margin-left: -30px;
    }
  }

  d-appendix li {
    margin-bottom: 1em;
  }

  d-appendix a {
    color: rgba(0, 0, 0, 0.6);
  }

  d-appendix > * {
    grid-column: text;
  }

  d-appendix > d-footnote-list,
  d-appendix > d-citation-list,
  d-appendix > distill-appendix {
    grid-column: screen;
  }

  /* Include footnote styles here so they can be overridden */

  d-footnote-list {
    contain: layout style;
  }

  d-footnote-list > * {
    grid-column: text;
  }

  d-footnote-list a.footnote-backlink {
    color: rgba(0,0,0,0.3);
    padding-left: 0.5em;
  }



  /* Anchor.js */

  .anchorjs-link {
    /*transition: all .25s linear; */
    text-decoration: none;
    border-bottom: none;
  }
  *:hover > .anchorjs-link {
    margin-left: -1.125em !important;
    text-decoration: none;
    border-bottom: none;
  }

  /* Social footer */

  .social_footer {
    margin-top: 30px;
    margin-bottom: 0;
    color: rgba(0,0,0,0.67);
  }

  .disqus-comments {
    margin-right: 30px;
  }

  .disqus-comment-count {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
    cursor: pointer;
  }

  #disqus_thread {
    margin-top: 30px;
  }

  .article-sharing a {
    border-bottom: none;
    margin-right: 8px;
  }

  .article-sharing a:hover {
    border-bottom: none;
  }

  .sidebar-section.subscribe {
    font-size: 12px;
    line-height: 1.6em;
  }

  .subscribe p {
    margin-bottom: 0.5em;
  }


  .article-footer .subscribe {
    font-size: 15px;
    margin-top: 45px;
  }


  .sidebar-section.custom {
    font-size: 12px;
    line-height: 1.6em;
  }

  .custom p {
    margin-bottom: 0.5em;
  }

  /* Styles for listing layout (hide title) */
  .layout-listing d-title, .layout-listing .d-title {
    display: none;
  }

  /* Styles for posts lists (not auto-injected) */


  .posts-with-sidebar {
    padding-left: 45px;
    padding-right: 45px;
  }

  .posts-list .description h2,
  .posts-list .description p {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  }

  .posts-list .description h2 {
    font-weight: 700;
    border-bottom: none;
    padding-bottom: 0;
  }

  .posts-list h2.post-tag {
    border-bottom: 1px solid rgba(0, 0, 0, 0.2);
    padding-bottom: 12px;
  }
  .posts-list {
    margin-top: 60px;
    margin-bottom: 24px;
  }

  .posts-list .post-preview {
    text-decoration: none;
    overflow: hidden;
    display: block;
    border-bottom: 1px solid rgba(0, 0, 0, 0.1);
    padding: 24px 0;
  }

  .post-preview-last {
    border-bottom: none !important;
  }

  .posts-list .posts-list-caption {
    grid-column: screen;
    font-weight: 400;
  }

  .posts-list .post-preview h2 {
    margin: 0 0 6px 0;
    line-height: 1.2em;
    font-style: normal;
    font-size: 24px;
  }

  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.4em;
    font-size: 16px;
  }

  .posts-list .post-preview .thumbnail {
    box-sizing: border-box;
    margin-bottom: 24px;
    position: relative;
    max-width: 500px;
  }
  .posts-list .post-preview img {
    width: 100%;
    display: block;
  }

  .posts-list .metadata {
    font-size: 12px;
    line-height: 1.4em;
    margin-bottom: 18px;
  }

  .posts-list .metadata > * {
    display: inline-block;
  }

  .posts-list .metadata .publishedDate {
    margin-right: 2em;
  }

  .posts-list .metadata .dt-authors {
    display: block;
    margin-top: 0.3em;
    margin-right: 2em;
  }

  .posts-list .dt-tags {
    display: block;
    line-height: 1em;
  }

  .posts-list .dt-tags .dt-tag {
    display: inline-block;
    color: rgba(0,0,0,0.6);
    padding: 0.3em 0.4em;
    margin-right: 0.2em;
    margin-bottom: 0.4em;
    font-size: 60%;
    border: 1px solid rgba(0,0,0,0.2);
    border-radius: 3px;
    text-transform: uppercase;
    font-weight: 500;
  }

  .posts-list img {
    opacity: 1;
  }

  .posts-list img[data-src] {
    opacity: 0;
  }

  .posts-more {
    clear: both;
  }


  .posts-sidebar {
    font-size: 16px;
  }

  .posts-sidebar h3 {
    font-size: 16px;
    margin-top: 0;
    margin-bottom: 0.5em;
    font-weight: 400;
    text-transform: uppercase;
  }

  .sidebar-section {
    margin-bottom: 30px;
  }

  .categories ul {
    list-style-type: none;
    margin: 0;
    padding: 0;
  }

  .categories li {
    color: rgba(0, 0, 0, 0.8);
    margin-bottom: 0;
  }

  .categories li>a {
    border-bottom: none;
  }

  .categories li>a:hover {
    border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  }

  .categories .active {
    font-weight: 600;
  }

  .categories .category-count {
    color: rgba(0, 0, 0, 0.4);
  }


  @media(min-width: 768px) {
    .posts-list .post-preview h2 {
      font-size: 26px;
    }
    .posts-list .post-preview .thumbnail {
      float: right;
      width: 30%;
      margin-bottom: 0;
    }
    .posts-list .post-preview .description {
      float: left;
      width: 45%;
    }
    .posts-list .post-preview .metadata {
      float: left;
      width: 20%;
      margin-top: 8px;
    }
    .posts-list .post-preview p {
      margin: 0 0 12px 0;
      line-height: 1.5em;
      font-size: 16px;
    }
    .posts-with-sidebar .posts-list {
      float: left;
      width: 75%;
    }
    .posts-with-sidebar .posts-sidebar {
      float: right;
      width: 20%;
      margin-top: 60px;
      padding-top: 24px;
      padding-bottom: 24px;
    }
  }


  /* Improve display for browsers without grid (IE/Edge <= 15) */

  .downlevel {
    line-height: 1.6em;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
    margin: 0;
  }

  .downlevel .d-title {
    padding-top: 6rem;
    padding-bottom: 1.5rem;
  }

  .downlevel .d-title h1 {
    font-size: 50px;
    font-weight: 700;
    line-height: 1.1em;
    margin: 0 0 0.5rem;
  }

  .downlevel .d-title p {
    font-weight: 300;
    font-size: 1.2rem;
    line-height: 1.55em;
    margin-top: 0;
  }

  .downlevel .d-byline {
    padding-top: 0.8em;
    padding-bottom: 0.8em;
    font-size: 0.8rem;
    line-height: 1.8em;
  }

  .downlevel .section-separator {
    border: none;
    border-top: 1px solid rgba(0, 0, 0, 0.1);
  }

  .downlevel .d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
    padding-top: 1rem;
    padding-bottom: 2rem;
  }


  .downlevel .d-appendix {
    padding-left: 0;
    padding-right: 0;
    max-width: none;
    font-size: 0.8em;
    line-height: 1.7em;
    margin-bottom: 0;
    color: rgba(0,0,0,0.5);
    padding-top: 40px;
    padding-bottom: 48px;
  }

  .downlevel .footnotes ol {
    padding-left: 13px;
  }

  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
    padding-left: 40px;
    padding-right: 40px;
  }

  @media(min-width: 768px) {
    .downlevel .base-grid,
    .downlevel .distill-header,
    .downlevel .d-title,
    .downlevel .d-abstract,
    .downlevel .d-article,
    .downlevel .d-appendix,
    .downlevel .distill-appendix,
    .downlevel .d-byline,
    .downlevel .d-footnote-list,
    .downlevel .d-citation-list,
    .downlevel .distill-footer,
    .downlevel .appendix-bottom,
    .downlevel .posts-container {
    padding-left: 150px;
    padding-right: 150px;
    max-width: 900px;
  }
  }

  .downlevel pre code {
    display: block;
    border-left: 2px solid rgba(0, 0, 0, .1);
    padding: 0 0 0 20px;
    font-size: 14px;
  }

  .downlevel code, .downlevel pre {
    color: black;
    background: none;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
    text-align: left;
    white-space: pre;
    word-spacing: normal;
    word-break: normal;
    word-wrap: normal;
    line-height: 1.5;

    -moz-tab-size: 4;
    -o-tab-size: 4;
    tab-size: 4;

    -webkit-hyphens: none;
    -moz-hyphens: none;
    -ms-hyphens: none;
    hyphens: none;
  }

  .downlevel .posts-list .post-preview {
    color: inherit;
  }



  </style>

  <script type="application/javascript">

  function is_downlevel_browser() {
    if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                   window.navigator.userAgent)) {
      return true;
    } else {
      return window.load_distill_framework === undefined;
    }
  }

  // show body when load is complete
  function on_load_complete() {

    // add anchors
    if (window.anchors) {
      window.anchors.options.placement = 'left';
      window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
    }


    // set body to visible
    document.body.style.visibility = 'visible';

    // force redraw for leaflet widgets
    if (window.HTMLWidgets) {
      var maps = window.HTMLWidgets.findAll(".leaflet");
      $.each(maps, function(i, el) {
        var map = this.getMap();
        map.invalidateSize();
        map.eachLayer(function(layer) {
          if (layer instanceof L.TileLayer)
            layer.redraw();
        });
      });
    }

    // trigger 'shown' so htmlwidgets resize
    $('d-article').trigger('shown');
  }

  function init_distill() {

    init_common();

    // create front matter
    var front_matter = $('<d-front-matter></d-front-matter>');
    $('#distill-front-matter').wrap(front_matter);

    // create d-title
    $('.d-title').changeElementType('d-title');

    // separator
    var separator = '<hr class="section-separator" style="clear: both"/>';
    // prepend separator above appendix
    $('.d-byline').before(separator);
    $('.d-article').before(separator);

    // create d-byline
    var byline = $('<d-byline></d-byline>');
    $('.d-byline').replaceWith(byline);

    // create d-article
    var article = $('<d-article></d-article>');
    $('.d-article').wrap(article).children().unwrap();

    // move posts container into article
    $('.posts-container').appendTo($('d-article'));

    // create d-appendix
    $('.d-appendix').changeElementType('d-appendix');

    // flag indicating that we have appendix items
    var appendix = $('.appendix-bottom').children('h3').length > 0;

    // replace footnotes with <d-footnote>
    $('.footnote-ref').each(function(i, val) {
      appendix = true;
      var href = $(this).attr('href');
      var id = href.replace('#', '');
      var fn = $('#' + id);
      var fn_p = $('#' + id + '>p');
      fn_p.find('.footnote-back').remove();
      var text = fn_p.html();
      var dtfn = $('<d-footnote></d-footnote>');
      dtfn.html(text);
      $(this).replaceWith(dtfn);
    });
    // remove footnotes
    $('.footnotes').remove();

    // move refs into #references-listing
    $('#references-listing').replaceWith($('#refs'));

    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      var id = $(this).attr('id');
      $('.d-contents a[href="#' + id + '"]').parent().remove();
      appendix = true;
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
    });

    // show d-appendix if we have appendix content
    $("d-appendix").css('display', appendix ? 'grid' : 'none');

    // localize layout chunks to just output
    $('.layout-chunk').each(function(i, val) {

      // capture layout
      var layout = $(this).attr('data-layout');

      // apply layout to markdown level block elements
      var elements = $(this).children().not('details, div.sourceCode, pre, script');
      elements.each(function(i, el) {
        var layout_div = $('<div class="' + layout + '"></div>');
        if (layout_div.hasClass('shaded')) {
          var shaded_content = $('<div class="shaded-content"></div>');
          $(this).wrap(shaded_content);
          $(this).parent().wrap(layout_div);
        } else {
          $(this).wrap(layout_div);
        }
      });


      // unwrap the layout-chunk div
      $(this).children().unwrap();
    });

    // remove code block used to force  highlighting css
    $('.distill-force-highlighting-css').parent().remove();

    // remove empty line numbers inserted by pandoc when using a
    // custom syntax highlighting theme, except when numbering line
    // in code chunk
    $('pre:not(.numberLines) code.sourceCode a:empty').remove();

    // load distill framework
    load_distill_framework();

    // wait for window.distillRunlevel == 4 to do post processing
    function distill_post_process() {

      if (!window.distillRunlevel || window.distillRunlevel < 4)
        return;

      // hide author/affiliations entirely if we have no authors
      var front_matter = JSON.parse($("#distill-front-matter").html());
      var have_authors = front_matter.authors && front_matter.authors.length > 0;
      if (!have_authors)
        $('d-byline').addClass('hidden');

      // article with toc class
      $('.d-contents').parent().addClass('d-article-with-toc');

      // strip links that point to #
      $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

      // add orcid ids
      $('.authors-affiliations').find('.author').each(function(i, el) {
        var orcid_id = front_matter.authors[i].orcidID;
        var author_name = front_matter.authors[i].author
        if (orcid_id) {
          var a = $('<a></a>');
          a.attr('href', 'https://orcid.org/' + orcid_id);
          var img = $('<img></img>');
          img.addClass('orcid-id');
          img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
          img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
          a.append(img);
          $(this).append(a);
        }
      });

      // hide elements of author/affiliations grid that have no value
      function hide_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
      }

      // affiliations
      var have_affiliations = false;
      for (var i = 0; i<front_matter.authors.length; ++i) {
        var author = front_matter.authors[i];
        if (author.affiliation !== "&nbsp;") {
          have_affiliations = true;
          break;
        }
      }
      if (!have_affiliations)
        $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

      // published date
      if (!front_matter.publishedDate)
        hide_byline_column("Published");

      // document object identifier
      var doi = $('d-byline').find('h3:contains("DOI")');
      var doi_p = doi.next().empty();
      if (!front_matter.doi) {
        // if we have a citation and valid citationText then link to that
        if ($('#citation').length > 0 && front_matter.citationText) {
          doi.html('Citation');
          $('<a href="#citation"></a>')
            .text(front_matter.citationText)
            .appendTo(doi_p);
        } else {
          hide_byline_column("DOI");
        }
      } else {
        $('<a></a>')
           .attr('href', "https://doi.org/" + front_matter.doi)
           .html(front_matter.doi)
           .appendTo(doi_p);
      }

       // change plural form of authors/affiliations
      if (front_matter.authors.length === 1) {
        var grid = $('.authors-affiliations');
        grid.children('h3:contains("Authors")').text('Author');
        grid.children('h3:contains("Affiliations")').text('Affiliation');
      }

      // remove d-appendix and d-footnote-list local styles
      $('d-appendix > style:first-child').remove();
      $('d-footnote-list > style:first-child').remove();

      // move appendix-bottom entries to the bottom
      $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
      $('.appendix-bottom').remove();

      // hoverable references
      $('span.citation[data-cites]').each(function() {
        const citeChild = $(this).children()[0]
        // Do not process if @xyz has been used without escaping and without bibliography activated
        // https://github.com/rstudio/distill/issues/466
        if (citeChild === undefined) return true

        if (citeChild.nodeName == "D-FOOTNOTE") {
          var fn = citeChild
          $(this).html(fn.shadowRoot.querySelector("sup"))
          $(this).id = fn.id
          fn.remove()
        }
        var refs = $(this).attr('data-cites').split(" ");
        var refHtml = refs.map(function(ref) {
          // Could use CSS.escape too here, we insure backward compatibility in navigator
          return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
        }).join("\n");
        window.tippy(this, {
          allowHTML: true,
          content: refHtml,
          maxWidth: 500,
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        });
      });

      // fix footnotes in tables (#411)
      // replacing broken distill.pub feature
      $('table d-footnote').each(function() {
        // we replace internal showAtNode methode which is triggered when hovering a footnote
        this.hoverBox.showAtNode = function(node) {
          // ported from https://github.com/distillpub/template/pull/105/files
          calcOffset = function(elem) {
              let x = elem.offsetLeft;
              let y = elem.offsetTop;
              // Traverse upwards until an `absolute` element is found or `elem`
              // becomes null.
              while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                  x += elem.offsetLeft;
                  y += elem.offsetTop;
              }

              return { left: x, top: y };
          }
          // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
          const bbox = node.getBoundingClientRect();
          const offset = calcOffset(node);
          this.show([offset.left + bbox.width, offset.top + bbox.height]);
        }
      })

      // clear polling timer
      clearInterval(tid);

      // show body now that everything is ready
      on_load_complete();
    }

    var tid = setInterval(distill_post_process, 50);
    distill_post_process();

  }

  function init_downlevel() {

    init_common();

     // insert hr after d-title
    $('.d-title').after($('<hr class="section-separator"/>'));

    // check if we have authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;

    // manage byline/border
    if (!have_authors)
      $('.d-byline').remove();
    $('.d-byline').after($('<hr class="section-separator"/>'));
    $('.d-byline a').remove();

    // remove toc
    $('.d-contents').remove();

    // move appendix elements
    $('h1.appendix, h2.appendix').each(function(i, val) {
      $(this).changeElementType('h3');
    });
    $('h3.appendix').each(function(i, val) {
      $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
    });


    // inject headers into references and footnotes
    var refs_header = $('<h3></h3>');
    refs_header.text('References');
    $('#refs').prepend(refs_header);

    var footnotes_header = $('<h3></h3');
    footnotes_header.text('Footnotes');
    $('.footnotes').children('hr').first().replaceWith(footnotes_header);

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // remove appendix if it's empty
    if ($('.d-appendix').children().length === 0)
      $('.d-appendix').remove();

    // prepend separator above appendix
    $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

    // trim code
    $('pre>code').each(function(i, val) {
      $(this).html($.trim($(this).html()));
    });

    // move posts-container right before article
    $('.posts-container').insertBefore($('.d-article'));

    $('body').addClass('downlevel');

    on_load_complete();
  }


  function init_common() {

    // jquery plugin to change element types
    (function($) {
      $.fn.changeElementType = function(newType) {
        var attrs = {};

        $.each(this[0].attributes, function(idx, attr) {
          attrs[attr.nodeName] = attr.nodeValue;
        });

        this.replaceWith(function() {
          return $("<" + newType + "/>", attrs).append($(this).contents());
        });
      };
    })(jQuery);

    // prevent underline for linked images
    $('a > img').parent().css({'border-bottom' : 'none'});

    // mark non-body figures created by knitr chunks as 100% width
    $('.layout-chunk').each(function(i, val) {
      var figures = $(this).find('img, .html-widget');
      // ignore leaflet img layers (#106)
      figures = figures.filter(':not(img[class*="leaflet"])')
      if ($(this).attr('data-layout') !== "l-body") {
        figures.css('width', '100%');
      } else {
        figures.css('max-width', '100%');
        figures.filter("[width]").each(function(i, val) {
          var fig = $(this);
          fig.css('width', fig.attr('width') + 'px');
        });

      }
    });

    // auto-append index.html to post-preview links in file: protocol
    // and in rstudio ide preview
    $('.post-preview').each(function(i, val) {
      if (window.location.protocol === "file:")
        $(this).attr('href', $(this).attr('href') + "index.html");
    });

    // get rid of index.html references in header
    if (window.location.protocol !== "file:") {
      $('.distill-site-header a[href]').each(function(i,val) {
        $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
      });
    }

    // add class to pandoc style tables
    $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
    $('.kable-table').children('table').addClass('pandoc-table');

    // add figcaption style to table captions
    $('caption').parent('table').addClass("figcaption");

    // initialize posts list
    if (window.init_posts_list)
      window.init_posts_list();

    // implmement disqus comment link
    $('.disqus-comment-count').click(function() {
      window.headroom_prevent_pin = true;
      $('#disqus_thread').toggleClass('hidden');
      if (!$('#disqus_thread').hasClass('hidden')) {
        var offset = $(this).offset();
        $(window).resize();
        $('html, body').animate({
          scrollTop: offset.top - 35
        });
      }
    });
  }

  document.addEventListener('DOMContentLoaded', function() {
    if (is_downlevel_browser())
      init_downlevel();
    else
      window.addEventListener('WebComponentsReady', init_distill);
  });

  </script>

  <!--/radix_placeholder_distill-->
  <script src="RJ-2025-036_files/header-attrs-2.30/header-attrs.js"></script>
  <script src="RJ-2025-036_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="RJ-2025-036_files/popper-2.6.0/popper.min.js"></script>
  <link href="RJ-2025-036_files/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="RJ-2025-036_files/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="RJ-2025-036_files/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="RJ-2025-036_files/anchor-4.2.2/anchor.min.js"></script>
  <script src="RJ-2025-036_files/bowser-1.9.3/bowser.min.js"></script>
  <script src="RJ-2025-036_files/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="RJ-2025-036_files/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
  <!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Mathjax config (add automatic linebreaks when supported)
      // MathJax = {
      //    tex: {
      //        inlineMath: [['$', '$'], ['\\(', '\\)']],
      //        displayMath: [['$$', '$$'], ['\\[', '\\]']],
      //        tags: 'ams',
      //        multline: true,
      //    },
      //    options: {
      //        linebreaks: { automatic: true },
      //    },
      // };

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      
      // Conditionally add supplementary files button
      if (document.getElementById('supplementary-materials')) {
        // create element safely
        const btn_suppl = document.createElement('a');
        btn_suppl.href = slug + '.zip';
        btn_suppl.className = 'dt-tag';
        btn_suppl.innerHTML = '<i class="fas fa-file-zipper"></i> Supplement';
        buttons.append(btn_suppl);
      }

      // adds Abstract: in front of the first <p> in the title section --
      // unless it happens to be the subtitle (FIXME: this is a bad hack - can't distill do this?)
      var tpar = $("d-title p:not(:empty)").filter(function() {
        return !$(this).hasClass("subtitle");
      }).first();
      if (tpar) {
        const abstract = $('<d-abstract>')
        abstract.append('<b>Abstract:</b><br>')
        abstract.append(tpar) // Move description to d-abstract
        $("d-title p:empty").remove() // Remove empty paragraphs after title
        abstract.append(buttons)
        abstract.insertAfter($('d-title')) // Add abstract section after title */
      }

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
    
    d-appendix .citation-appendix,
    .d-appendix .citation-appendix {
      color: rgb(60, 60, 60);
    }

    d-article h2 {
      border-bottom: 0px solid rgba(0, 0, 0, 0.1);
      padding-bottom: 0rem;
    }
    d-article h3 {
      font-size: 20px;
    }
    d-article h4 {
      font-size: 18px;
      text-transform: none;
    }

    @media (min-width: 1024px) {
      d-article h2 {
        font-size: 32px;
      }
      d-article h3 {
        font-size: 24px;
      }
      d-article h4 {
        font-size: 20px;
      }
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"dtComb: A Comprehensive R Library and Web Tool for Combining Diagnostic Tests","description":"The combination of diagnostic tests has become a crucial area of\nresearch, aiming to improve the accuracy and robustness of medical\ndiagnostics. While existing tools focus primarily on linear\ncombination methods, there is a lack of comprehensive tools that\nintegrate diverse methodologies. In this study, we present dtComb, a\ncomprehensive R package and web tool designed to address the\nlimitations of existing diagnostic test combination platforms. One of\nthe unique contributions of dtComb is offering a range of 142 methods\nto combine two diagnostic tests, including linear, non-linear, machine\nlearning algorithms, and mathematical operators. Another significant\ncontribution of dtComb is its inclusion of advanced tools for ROC\nanalysis, diagnostic performance metrics, and visual outputs such as\nsensitivity-specificity curves. Furthermore, dtComb offers\nclassification functions for new observations, making it an\neasy-to-use tool for clinicians and researchers. The web-based version\nis also available at <https://biotools.erciyes.edu.tr/dtComb/> for\nnon-R users, providing an intuitive interface for test combination and\nmodel training.","doi":"10.32614/RJ-2025-036","authors":[{"author":"S. Ilayda Yerlitaş Taştan","authorURL":"#","affiliation":"Department of Biostatistics","affiliationURL":"#","orcidID":""},{"author":"Serra Bersan Gengeç","authorURL":"#","affiliation":"Department of Biostatistics","affiliationURL":"#","orcidID":""},{"author":"Necla Koçhan","authorURL":"#","affiliation":"Department of Mathematics","affiliationURL":"#","orcidID":""},{"author":"Ertürk Zararsız","authorURL":"#","affiliation":"Department of Biostatistics","affiliationURL":"#","orcidID":""},{"author":"Selçuk Korkmaz","authorURL":"#","affiliation":"Department of Biostatistics","affiliationURL":"#","orcidID":""},{"author":"Zararsız","authorURL":"#","affiliation":"Department of Biostatistics","affiliationURL":"#","orcidID":""}],"publishedDate":"2026-02-03T00:00:00.000+11:00","citationText":"Taştan, et al., 2026"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>dtComb: A Comprehensive R Library and Web Tool for Combining Diagnostic Tests</h1>

<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>The combination of diagnostic tests has become a crucial area of
research, aiming to improve the accuracy and robustness of medical
diagnostics. While existing tools focus primarily on linear
combination methods, there is a lack of comprehensive tools that
integrate diverse methodologies. In this study, we present dtComb, a
comprehensive R package and web tool designed to address the
limitations of existing diagnostic test combination platforms. One of
the unique contributions of dtComb is offering a range of 142 methods
to combine two diagnostic tests, including linear, non-linear, machine
learning algorithms, and mathematical operators. Another significant
contribution of dtComb is its inclusion of advanced tools for ROC
analysis, diagnostic performance metrics, and visual outputs such as
sensitivity-specificity curves. Furthermore, dtComb offers
classification functions for new observations, making it an
easy-to-use tool for clinicians and researchers. The web-based version
is also available at <a href="https://biotools.erciyes.edu.tr/dtComb/" class="uri">https://biotools.erciyes.edu.tr/dtComb/</a> for
non-R users, providing an intuitive interface for test combination and
model training.</p></p>
</div>

<div class="d-byline">
  S. Ilayda Yerlitaş Taştan  (Department of Biostatistics)
  
,   Serra Bersan Gengeç  (Department of Biostatistics)
  
,   Necla Koçhan  (Department of Mathematics)
  
,   Ertürk Zararsız  (Department of Biostatistics)
  
,   Selçuk Korkmaz  (Department of Biostatistics)
  
,   Zararsız  (Department of Biostatistics)
  
<br/>2026-02-03
</div>

<div class="d-article">
<div class="article">
<h3 data-number="1" id="introduction"><span class="header-section-number">1</span> Introduction</h3>
<p>A typical scenario often encountered in combining diagnostic tests is
when the gold standard method combines two-category and two continuous
diagnostic tests. In such cases, clinicians usually seek to compare
these two diagnostic tests and improve the performance of these
diagnostic test results by dividing the results into proportional
results <span class="citation" data-cites="muller2019amyloid faria2016neutrophil nyblom2006ast">(<a href="#ref-nyblom2006ast" role="doc-biblioref">Nyblom et al. 2006</a>; <a href="#ref-faria2016neutrophil" role="doc-biblioref">Faria et al. 2016</a>; <a href="#ref-muller2019amyloid" role="doc-biblioref">Müller et al. 2019</a>)</span>.
However, this technique is straightforward and may not fully capture all
potential interactions and relationships between the diagnostic tests.
Linear combination methods have been developed to overcome such problems
<span class="citation" data-cites="erturkzararsiz2023linear">(<a href="#ref-erturkzararsiz2023linear" role="doc-biblioref">Ertürk Zararsız 2023</a>)</span>.<br />
Linear methods combine two diagnostic tests into a single score/index by
assigning weights to each test, optimizing their performance in
diagnosing the condition of interest <span class="citation" data-cites="neumann2023combining">(<a href="#ref-neumann2023combining" role="doc-biblioref">Neumann et al. 2023</a>)</span>. Such
methods improve accuracy by leveraging the strengths of both tests
<span class="citation" data-cites="aznar2022stepwise bansal2013does">(<a href="#ref-bansal2013does" role="doc-biblioref">Bansal and Sullivan Pepe 2013</a>; <a href="#ref-aznar2022stepwise" role="doc-biblioref">Aznar-Gimeno et al. 2022</a>)</span>. For instance, Su and Liu
<span class="citation" data-cites="su1993linear">(<a href="#ref-su1993linear" role="doc-biblioref">Su and Liu 1993</a>)</span> found that Fisher’s linear discriminant function
generates a linear combination of markers with either proportional or
disproportional covariance matrices, aiming to maximize sensitivity
consistently across the entire selectivity spectrum under a multivariate
normal distribution model. In contrast, another approach introduced by
Pepe and Thomson <span class="citation" data-cites="pepe2000combining">(<a href="#ref-pepe2000combining" role="doc-biblioref">Pepe and Thompson 2000</a>)</span> relies on ranking scores,
eliminating the need for linear distributional assumptions when
combining diagnostic tests. Despite the theoretical advances, when
existing tools were examined, it was seen that they contained a limited
number of methods. For instance, Kramar et al. developed a computer
program called <strong>mROC</strong> that includes only the Su and Liu method
<span class="citation" data-cites="kramar2001mroc">(<a href="#ref-kramar2001mroc" role="doc-biblioref">Kramar et al. 2001</a>)</span>. Pérez-Fernández et al. presented a
<a href="https://cran.r-project.org/web/packages/movieROC/index.html"><code>movieROC</code></a>
R package that includes methods such as Su and Liu, min-max, and
logistic regression methods <span class="citation" data-cites="perez2021visualizing">(<a href="#ref-perez2021visualizing" role="doc-biblioref">Pérez-Fernández et al. 2021</a>)</span>. An R package called
<a href="https://github.com/wbaopaul/MaxmzpAUC-R"><code>maxmzpAUC</code></a> that includes
similar methods was developed by Yu and Park <span class="citation" data-cites="yu2015two">(<a href="#ref-yu2015two" role="doc-biblioref">Yu and Park 2015</a>)</span>.</p>
<p>On the other hand, non-linear approaches incorporating the non-linearity
between the diagnostic tests have been developed and employed to
integrate the diagnostic tests
<span class="citation" data-cites="du2024likelihood ghosh2005classification">(<a href="#ref-ghosh2005classification" role="doc-biblioref">Ghosh and Chinnaiyan 2005</a>; <a href="#ref-du2024likelihood" role="doc-biblioref">Du et al. 2024</a>)</span>. These approaches
incorporate the non-linear structure of tests into the model, which
might improve the accuracy and reliability of the diagnosis. In contrast
to some existing packages, which permit the use of non-linear approaches
such as splines<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>, lasso<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a> and ridge regression, there is currently
no package that employs these methods directly for combination and
offers diagnostic performance. Machine-learning (ML) algorithms have
recently been adopted to combine diagnostic tests
<span class="citation" data-cites="ahsan2024advancements sewak2024construction agarwal2023artificial prinzi2023explainable">(<a href="#ref-agarwal2023artificial" role="doc-biblioref">Agarwal et al. 2023</a>; <a href="#ref-prinzi2023explainable" role="doc-biblioref">Prinzi et al. 2023</a>; <a href="#ref-ahsan2024advancements" role="doc-biblioref">Ahsan et al. 2024</a>; <a href="#ref-sewak2024construction" role="doc-biblioref">Sewak et al. 2024</a>)</span>.
Many publications/studies focus on implementing ML algorithms in
diagnostic tests
<span class="citation" data-cites="salvetat2022game salvetat2024ai ganapathy2023comparison alzyoud2024diagnosing zararsiz2016statistical">(<a href="#ref-zararsiz2016statistical" role="doc-biblioref">Zararsiz et al. 2016</a>; <a href="#ref-salvetat2022game" role="doc-biblioref">Salvetat et al. 2022</a>, <a href="#ref-salvetat2024ai" role="doc-biblioref">2024</a>; <a href="#ref-ganapathy2023comparison" role="doc-biblioref">Ganapathy et al. 2023</a>; <a href="#ref-alzyoud2024diagnosing" role="doc-biblioref">Alzyoud et al. 2024</a>)</span>.
For instance, DeGroat et al. performed four different classification
algorithms (Random Forest, Support Vector Machine, Extreme Gradient
Boosting Decision Trees, and k-Nearest Neighbors) to combine markers for
the diagnosis of cardiovascular disease <span class="citation" data-cites="degroat2024discovering">(<a href="#ref-degroat2024discovering" role="doc-biblioref">DeGroat et al. 2024</a>)</span>. The
results showed that patients with cardiovascular disease can be
diagnosed with up to 96% accuracy using these ML techniques. There are
numerous applications where ML methods can be implemented
(<a href="https://scikit-learn.org/stable/"><code>scikit-learn</code></a>
<span class="citation" data-cites="pedregosa2011scikit">(<a href="#ref-pedregosa2011scikit" role="doc-biblioref">Pedregosa et al. 2011</a>)</span>,
<a href="https://www.tensorflow.org/learn?hl=tr"><code>TensorFlow</code></a>
<span class="citation" data-cites="tensorflow2015-whitepaper">(<a href="#ref-tensorflow2015-whitepaper" role="doc-biblioref">Abadi et al. 2015</a>)</span>,
<a href="https://cran.r-project.org/web/packages/caret/index.html"><code>caret</code></a>
<span class="citation" data-cites="kuhn2008building">(<a href="#ref-kuhn2008building" role="doc-biblioref">Kuhn 2008</a>)</span>). The
<a href="https://cran.r-project.org/web/packages/caret/index.html"><code>caret</code></a>
library is one of the most comprehensive tools developed in the R
language<span class="citation" data-cites="kuhn2008building">(<a href="#ref-kuhn2008building" role="doc-biblioref">Kuhn 2008</a>)</span>. However, these are general tools developed
only for ML algorithms and do not directly combine two diagnostic tests
or provide diagnostic performance measures.</p>
<p>Apart from the aforementioned methods, several basic mathematical
operations such as addition, multiplication, subtraction, and division
can also be used to combine markers
<span class="citation" data-cites="svart2024neurofilament luo2024ast serban2024significance">(<a href="#ref-luo2024ast" role="doc-biblioref">Luo et al. 2024</a>; <a href="#ref-serban2024significance" role="doc-biblioref">Serban et al. 2024</a>; <a href="#ref-svart2024neurofilament" role="doc-biblioref">Svart et al. 2024</a>)</span>. For
instance, addition can enhance diagnostic sensitivity by combining the
effects of markers, whereas subtraction can more distinctly
differentiate disease states by illustrating the variance across
markers. On the other hand, there are several commercial (e.g. IBM SPSS,
MedCalc, Stata, etc.) and open source (R) software packages
(<a href="https://cran.r-project.org/web/packages/ROCR/index.html"><code>ROCR</code></a>
<span class="citation" data-cites="sing2005rocr">(<a href="#ref-sing2005rocr" role="doc-biblioref">Sing et al. 2005</a>)</span>,
(<a href="https://cran.r-project.org/web/packages/pROC/index.html"><code>pROC</code></a>
<span class="citation" data-cites="robin2011proc">(<a href="#ref-robin2011proc" role="doc-biblioref">Robin et al. 2011</a>)</span>,
<a href="https://cran.r-project.org/web/packages/PRROC/index.html"><code>PRROC</code></a>
<span class="citation" data-cites="grau2015prroc">(<a href="#ref-grau2015prroc" role="doc-biblioref">Grau et al. 2015</a>)</span>,
<a href="https://cran.r-project.org/web/packages/plotROC/index.html"><code>plotROC</code></a>
<span class="citation" data-cites="sachs2017plotroc">(<a href="#ref-sachs2017plotroc" role="doc-biblioref">Sachs 2017</a>)</span>) that researchers can use for Receiver operating
characteristic (ROC) curve analysis. However, these tools are designed
to perform a single marker ROC analysis. As a result, there is currently
no software tool that covers almost all combination methods.</p>
<p>In this study, we developed
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>,
an R package encompassing nearly all existing combination approaches in
the literature.
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
has two key advantages, making it easy to apply and superior to the
other packages: (1) it provides users with a comprehensive 142 methods,
including linear and non-linear approaches, ML approaches and
mathematical operators; (2) it produces turnkey solutions to users from
the stage of uploading data to the stage of performing analyses,
performance evaluation and reporting. Furthermore, it is the only
package that illustrates linear approaches such as Minimax and Todor &amp;
Saplacan <span class="citation" data-cites="sameera2016binary todor2014tools">(<a href="#ref-todor2014tools" role="doc-biblioref">Todor et al. 2014</a>; <a href="#ref-sameera2016binary" role="doc-biblioref">Sameera et al. 2016</a>)</span>. In addition, it allows
for the classification of new, previously unseen observations using
trained models. To our knowledge, no other tools were designed and
developed to combine two diagnostic tests on a single platform with 142
different methods. In other words,
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
has made more effective and robust combination methods ready for
application instead of traditional approaches such as simple ratio-based
methods. First, we review the theoretical basis of the related
combination methods; then, we present an example implementation to
demonstrate the applicability of the package. Finally, we present a
user-friendly, up-to-date, and comprehensive web tool developed to
facilitate
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
for physicians and healthcare professionals who do not use the R
programming language. The
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package is freely available on the CRAN network, the web application is
freely available at <a href="https://biotools.erciyes.edu.tr/dtComb/" class="uri">https://biotools.erciyes.edu.tr/dtComb/</a>, and all
source code is available on GitHub<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a>.</p>
<h3 data-number="2" id="material-and-methods"><span class="header-section-number">2</span> Material and methods</h3>
<p>This section will provide an overview of the combination methods
implemented in the literature. Before applying these methods, we will
also discuss the standardization techniques available for the markers,
the resampling methods during model training, and, ultimately, the
metrics used to evaluate the model’s performance.</p>
<h4 class="unnumbered" data-number="2.1" id="combination-approaches">Combination approaches</h4>
<h5 class="unnumbered" data-number="2.1.1" id="linear-combination-methods">Linear combination methods</h5>
<p>The
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package comprises eight distinct linear combination methods, which will
be elaborated in this section. Before investigating these methods, we
briefly introduce some notations which will be used throughout this
section.<br />
Notations:<br />
Let <span class="math inline">\(D_{i}, i = 1, 2, …, n_1\)</span> be the marker values of the <span class="math inline">\(i\)</span>th
individual in the diseased group, where <span class="math inline">\(D_i=(D_{i1},D_{i2})\)</span>, and
<span class="math inline">\(H_j, j=1,2,…,n_2\)</span> be the marker values of the <span class="math inline">\(j\)</span>th individual in the
healthy group, where <span class="math inline">\(H_j=(H_{j1},H_{j2})\)</span>. Let
<span class="math inline">\(x_{i1}=c(D_{{i1}},H_{j1})\)</span> be the values of the first marker, and
<span class="math inline">\(x_{i2}=c(D_{i2},H_{j2})\)</span> be values of the second marker for the <span class="math inline">\(i\)</span>th
individual <span class="math inline">\(i=1,2,...,n\)</span>. Let <span class="math inline">\(D_{i,min}=\min(D_{i1},D_{i2})\)</span>,
<span class="math inline">\(D_{i,max}=\max(D_{i1},D_{i2})\)</span>, <span class="math inline">\(H_{j,min}=\min(H_{j1},H_{j2})\)</span>,
<span class="math inline">\(H_{j,max}=\max(H_{j1},H_{j2})\)</span> and <span class="math inline">\(c_i\)</span> be the resulting combination
score of the <span class="math inline">\(i\)</span>th individual.</p>
<ul>
<li><p><em>Logistic regression:</em> Logistic regression is a statistical method
used for binary classification. The logistic regression model
estimates the probability of the binary outcome occurring based on
the values of the independent variables. It is one of the most
commonly applied methods in diagnostic tests, and it generates a
linear combination of markers that can distinguish between control
and diseased individuals. Logistic regression is generally less
effective than normal-based discriminant analysis, like Su and Liu’s
multivariate normality-based method, when the normal assumption is
met <span class="citation" data-cites="ruiz1991asymptotic efron1975efficiency">(<a href="#ref-efron1975efficiency" role="doc-biblioref">Efron 1975</a>; <a href="#ref-ruiz1991asymptotic" role="doc-biblioref">Ruiz-Velasco 1991</a>)</span>. On the other hand,
others have argued that logistic regression is more robust because
it does not require any assumptions about the joint distribution of
multiple markers <span class="citation" data-cites="cox1989analysis">(<a href="#ref-cox1989analysis" role="doc-biblioref">Cox and Snell 1989</a>)</span>. Therefore, it is essential to
investigate the performance of linear combination methods derived
from the logistic regression approach with non-normally distributed
data.<br />
The objective of the logistic regression model is to maximize the
logistic likelihood function. In other words, the logistic
likelihood function is maximized to estimate the logistic regression
model coefficients.<br />
</p>
<p><span class="math display" id="eq:1">\[\label{eq:1}
c=\frac{exp(\beta_0 + \beta_1 x_{i1} + \beta_2 x_{i2}}{1+exp(\beta_0 + \beta_1 x_{i1} + \beta_2 x_{i2}}   \tag{1}\]</span>
The logistic regression coefficients can provide the maximum
likelihood estimation of the model, producing an easily
interpretable value for distinguishing between the two groups.</p></li>
<li><p><em>Scoring based on logistic regression:</em> The method primarily uses a
binary logistic regression model, with slight modifications to
enhance the combination score. The regression coefficients, as
predicted in Eq <a href="#eq:1">(1)</a>, are rounded to a user-specified number
of decimal places and subsequently used to calculate the combination
score <span class="citation" data-cites="leon2006bedside">(<a href="#ref-leon2006bedside" role="doc-biblioref">León et al. 2006</a>)</span>.
<span class="math display">\[c= \beta_1 x_{i1}+\beta_2 x_{i2}\]</span></p></li>
<li><p><em>Pepe &amp; Thompson’s method:</em> Pepe &amp; Thompson have aimed to maximize
the AUC or partial AUC to combine diagnostic tests, regardless of
the distribution of markers <span class="citation" data-cites="pepe2000combining">(<a href="#ref-pepe2000combining" role="doc-biblioref">Pepe and Thompson 2000</a>)</span>. They developed an
empirical solution of optimal linear combinations that maximize the
Mann-Whitney U statistic, an empirical estimate of the ROC curve.
Notably, this approach is distribution-free. Mathematically, they
maximized the following objective function:
<span class="math display">\[\text{maximize} \; U(a)= \frac{1}{n_1 n_2} \sum_{i=1}^{n_1} \sum_{j=1}^{n_2} I\left[D_{i1}+\alpha D_{i2}\geq H_{j1}+\alpha H_{j2}\right]\]</span></p>
<p><span class="math display" id="eq:4">\[c= x_{i1}+\alpha x_{i2}
\label{eq:4}   \tag{2}\]</span>
where <span class="math inline">\(a \in [-1,1]\)</span> is interpreted as the relative weight of
<span class="math inline">\(x_{i2}\)</span> to <span class="math inline">\(x_{i1}\)</span> in the combination, the weight of the second
marker. This formula aims to find <span class="math inline">\(\alpha\)</span> to maximize <span class="math inline">\(U(a)\)</span>.
Readers are referred to see (Pepe and Thomson) <span class="citation" data-cites="pepe2000combining">(<a href="#ref-pepe2000combining" role="doc-biblioref">Pepe and Thompson 2000</a>)</span>.</p></li>
<li><p><em>Pepe, Cai &amp; Langton’s method:</em> Pepe et al. observed that when the
disease status and the levels of markers conform to a generalized
linear model, the regression coefficients represent the optimal
linear combinations that maximize the area under the ROC curves
<span class="citation" data-cites="pepe2006combining">(<a href="#ref-pepe2006combining" role="doc-biblioref">Pepe et al. 2006</a>)</span>. The following objective function is maximized
to achieve a higher AUC value:
<span class="math display">\[\text{maximize} \; U(a)= \frac{1}{n_1 n_2} \sum_{i=1}^{n_1} \sum_{j=1}^{n_2} I\left[D_{i1}+\alpha D_{i2}&gt; H_{j1}+\alpha H_{j2}\right] + \frac{1}{2}I\left[D_{i1}+\alpha = H_{j1} + \alpha H_{j2}\right]\]</span>
Before calculating the combination score using Eq <a href="#eq:4">(2)</a>, the
marker values are normalized or scaled to be constrained within the
scale of 0 to 1. In addition, it is noted that the estimate obtained
by maximizing the empirical AUC can be considered as a particular
case of the maximum rank correlation estimator from which the
general asymptotic distribution theory was developed. Readers are
referred to Pepe (2003, Chapters 4–6) for a review of the ROC curve
approach and more details <span class="citation" data-cites="pepe2003statistical">(<a href="#ref-pepe2003statistical" role="doc-biblioref">Pepe 2003</a>)</span>.</p></li>
<li><p><em>Min-Max method:</em> The Pepe &amp; Thomson method is straightforward if
there are two markers. It is computationally challenging if we have
more than two markers to be combined. To overcome the computational
complexity issue of this method, Liu et al. <span class="citation" data-cites="liu2011min">(<a href="#ref-liu2011min" role="doc-biblioref">Liu et al. 2011</a>)</span> proposed a
non-parametric approach that linearly combines the minimum and
maximum values of the observed markers of each subject. This
approach, which does not rely on the normality assumption of data
distributions (i.e., distribution-free), is known as the Min-Max
method and may provide higher sensitivity than any single marker.
The objective function of the Min-Max method is as follows:
<span class="math display">\[\text{maximize} \; U(a)= \frac{1}{n_1 n_2} \sum_{i=1}^{n_1} \sum_{j=1}^{n_2} I[D_{i,max}+\alpha D_{i,min}&gt; H_{j,max}+\alpha H_{j,min}]\]</span></p>
<p><span class="math display">\[c= x_{i,max}+\alpha x_{i,min}\]</span><br />
where <span class="math inline">\(x_{i,max}=\max⁡(x_{i1},x_{i2})\)</span> and
<span class="math inline">\(x_{i,min}=\min⁡(x_{i1},x_{i2})\)</span>.<br />
The Min-Max method aims to combine repeated measurements of a single
marker over time or multiple markers that are measured with the same
unit. While the Min-Max method is relatively simple to implement, it
has some limitations. For example, markers may have different units
of measurement, so standardization can be needed to ensure
uniformity during the combination process. Furthermore, it is
unclear whether all available information is fully utilized when
combining markers, as this method incorporates only the markers’
minimum and maximum values into the model <span class="citation" data-cites="kang2016linear">(<a href="#ref-kang2016linear" role="doc-biblioref">Kang et al. 2016</a>)</span>.</p></li>
<li><p><em>Su &amp; Liu’s method:</em> Su and Liu examined the combination score
separately under the assumption of two multivariate normal
distributions when the covariance matrices were proportional or
disproportionate <span class="citation" data-cites="su1993linear">(<a href="#ref-su1993linear" role="doc-biblioref">Su and Liu 1993</a>)</span>. Multivariate normal distributions
with different covariances were first utilized in classification
problems <span class="citation" data-cites="anderson1962classification">(<a href="#ref-anderson1962classification" role="doc-biblioref">Anderson and Bahadur 1962</a>)</span>. Then, Su and Liu also
developed a linear combination method by extending the idea of using
multivariate distributions to the AUC, showing that the best
coefficients that maximize AUC are Fisher’s discriminant
coefficients. Assuming that <span class="math inline">\(D~N(\mu_D, \sum_D)\)</span> and
<span class="math inline">\(H~N(\mu_H, \sum_H)\)</span> represent the multivariate normal distributions
for the diseased and non-diseased groups, respectively. The Fisher’s
coefficients are as follows:
<span class="math display" id="eq:alpha-beta">\[(\alpha, \beta) = (\sum_{D} + \sum_{H})^{-1} \mu \labe\tag{3}   (#eq:alpha-beta)\]</span>
where <span class="math inline">\(\mu=\mu_D-\mu_H\)</span>. The combination score in this case is:
<span class="math display" id="eq:9">\[c= \alpha x_{i1}+ \beta x_{i2}
\label{eq:9}   \tag{4}\]</span></p></li>
<li><p><em>The Minimax method:</em> The Minimax method is an extension of Su &amp;
Liu’s method <span class="citation" data-cites="sameera2016binary">(<a href="#ref-sameera2016binary" role="doc-biblioref">Sameera et al. 2016</a>)</span>. Suppose that D follows a
multivariate normal distribution <span class="math inline">\(D\sim N(\mu_D, \sum_D)\)</span>,
representing the diseased group, and H follows a multivariate normal
distribution <span class="math inline">\(H\sim N(\mu_H, \sum_H)\)</span>, representing the non-diseased
group. Then Fisher’s coefficients are as follows:
<span class="math display" id="eq:alpha-beta-expression">\[(\alpha, \beta) = \left[t\sum_{D} + (1-t)\sum_{H}\right]^{-1} (\mu_D - \mu_H) \labe\tag{5}   (#eq:alpha-beta-expression)\]</span></p>
<p>Given these coefficients, the combination score is calculated using
Eq <a href="#eq:9">(4)</a>. In this formula, <em>t</em> is a constant with values
ranging from 0 to 1. This value can be hyper-tuned by maximizing the
AUC.</p></li>
<li><p><em>Todor &amp; Saplacan’s method:</em> Todor and Saplacan’s method uses the
sine and cosine trigonometric functions to calculate the combination
score <span class="citation" data-cites="todor2014tools">(<a href="#ref-todor2014tools" role="doc-biblioref">Todor et al. 2014</a>)</span>. The combination score is calculated using
<span class="math inline">\(\theta \in[-\frac{\pi}{2},\frac{\pi}{2}]\)</span>, which maximizes the AUC
within this interval. The formula for the combination score is given
as follows:
<span class="math display">\[c= \sin{(\theta)}x_{i1}+\cos{(\theta)}x_{i2}\]</span></p></li>
</ul>
<h5 class="unnumbered" data-number="2.1.2" id="non-linear-combination-methods">Non-linear combination methods</h5>
<p>In addition to linear combination methods, the
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package includes seven non-linear approaches, which will be discussed in
this subsection. In this subsection, we will use the following
notations: <span class="math inline">\(x_{ij}\)</span>: the value of the <em>j</em>th marker for the <em>i</em>th
individual, <span class="math inline">\(i=1,2,...,n\)</span> and <span class="math inline">\(j=1,2\)</span> <em>d</em>: degree of polynomial
regressions and splines, <span class="math inline">\(d = 1,2,…,p\)</span>.</p>
<ul>
<li><p><em>Logistic Regression with Polynomial Feature Space:</em> This approach
extends the logistic regression model by adding extra predictors
created by raising the original predictor variables to a certain
power. This transformation enables the model to capture and model
non-linear relationships in the data by including polynomial terms
in the feature space <span class="citation" data-cites="james2013introduction">(<a href="#ref-james2013introduction" role="doc-biblioref">James et al. 2013</a>)</span>. The combination score
is calculated as follows:
<span class="math display">\[c=\frac{exp\left(\beta_0 + \beta_1 x_{ij} + \beta_2 x_{ij}^2+...+\beta_p x_{ij}^p\right)}{1+exp\left(\beta_0 + \beta_1 x_{ij} + \beta_2 x_{ij}^2+...+\beta_p x_{ij}^p\right)}\]</span>
where <span class="math inline">\(c_i\)</span> is the combination score for the <em>i</em>th individual and
represents the posterior probabilities.</p></li>
<li><p><em>Ridge Regression with Polynomial Feature Space:</em> This method
combines Ridge regression with expanded feature space created by
adding polynomial terms to the original predictor variables. It is a
widely used shrinkage method when we have multicollinearity between
the variables, which may be an issue for least squares regression.
This method aims to estimate the coefficients of these correlated
variables by minimizing the residual sum of squares (RSS) while
adding a term (referred to as a regularization term) to prevent
overfitting. The objective function is based on the L2 norm of the
coefficient vector, which prevents overfitting in the model (Eq
<a href="#eq:beta-hat-r">(6)</a>). The Ridge estimate is defined as follows:
<span class="math display" id="eq:beta-hat-r">\[\hat{\beta}^R = \text{argmin}_{\beta} \text{RSS} + \lambda \sum_{j=1}^{2} \sum_{d=1}^{p} \beta_j^{d^2} \labe\tag{6}   (#eq:beta-hat-r)\]</span></p>
<p>where
<span class="math display">\[RSS=\sum_{i=1}^{n}\left(y_i-\beta_0-\sum_{j=1}^{2}\sum_{d=1}^{p} \beta_j^d x_{ij}^d\right)\]</span>
and <span class="math inline">\(\hat{\beta}^R\)</span> denotes the estimates of the coefficients of the
Ridge regression, and the second term is called a penalty term where
<span class="math inline">\(\lambda \geq 0\)</span> is a shrinkage parameter. The shrinkage parameter,
<span class="math inline">\(\lambda\)</span>, controls the amount of shrinkage applied to regression
coefficients. A cross-validation is implemented to find the
shrinkage parameter. We used the
<a href="https://cran.r-project.org/web/packages/glmnet/index.html"><code>glmnet</code></a>
package <span class="citation" data-cites="friedman2010regularization">(<a href="#ref-friedman2010regularization" role="doc-biblioref">Friedman et al. 2010</a>)</span> to implement the Ridge
regression in combining the diagnostic tests.</p></li>
<li><p><em>Lasso Regression with Polynomial Feature Space:</em> Similar to Ridge
regression, Lasso regression is also a shrinkage method that adds a
penalty term to the objective function of the least square
regression. The objective function, in this case, is based on the L1
norm of the coefficient vector, which leads to the sparsity in the
model. Some of the regression coefficients are precisely zero when
the tuning parameter <span class="math inline">\(\lambda\)</span> is sufficiently large. This property
of the Lasso method allows the model to automatically identify and
remove less relevant variables and reduce the algorithm’s
complexity. The Lasso estimates are defined as follows:</p>
<p><span class="math display" id="eq:beta-hat-l">\[\hat{\beta}^L = \text{argmin}_{\beta} \text{RSS} + \lambda \sum_{j=1}^{2} \sum_{d=1}^{d} | \beta_j^d | \labe\tag{7}   (#eq:beta-hat-l)\]</span></p>
<p>To implement the Lasso regression in combining the diagnostic tests,
we used the
<a href="https://cran.r-project.org/web/packages/glmnet/index.html"><code>glmnet</code></a>
package <span class="citation" data-cites="friedman2010regularization">(<a href="#ref-friedman2010regularization" role="doc-biblioref">Friedman et al. 2010</a>)</span>.</p></li>
<li><p><em>Elastic-Net Regression with Polynomial Feature Space:</em> Elastic-Net
Regression is a method that combines Lasso (L1 regularization) and
Ridge (L2 regularization) penalties to address some of the
limitations of each technique. The combination of the two penalties
is controlled by two hyperparameters, <span class="math inline">\(\alpha\in\)</span><span class="math display">\[0,1\]</span> and
<span class="math inline">\(\lambda\)</span>, which enable you to adjust the trade-off between the L1
and L2 regularization terms <span class="citation" data-cites="james2013introduction">(<a href="#ref-james2013introduction" role="doc-biblioref">James et al. 2013</a>)</span>. For the
implementation of the method, the
<a href="https://cran.r-project.org/web/packages/glmnet/index.html"><code>glmnet</code></a>
package is used <span class="citation" data-cites="friedman2010regularization">(<a href="#ref-friedman2010regularization" role="doc-biblioref">Friedman et al. 2010</a>)</span>.</p></li>
<li><p><em>Splines:</em> Another non-linear combination technique frequently
applied in diagnostic tests is the splines. Splines are a versatile
mathematical and computational technique that has a wide range of
applications. These splines are piecewise functions that make
interpolating or approximating data points possible. There are
several types of splines, such as cubic splines. Smooth curves are
created by approximating a set of control points using cubic
polynomial functions. When implementing splines, two critical
parameters come into play: degrees of freedom and the choice of
polynomial degrees (i.e., degrees of the fitted polynomials). These
user-adjustable parameters, which influence the flexibility and
smoothness of the resulting curve, are critical for controlling the
behavior of splines. We used the
<a href="https://rdocumentation.org/packages/splines/versions/3.6.2"><code>splines</code></a>
package in the R programming language to implement splines.</p></li>
<li><p><em>Generalized Additive Models with Smoothing Splines and Generalized
Additive Models with Natural Cubic Splines:</em> Regression models are
of great interest in many fields to understand the importance of
different inputs. Even though regression is widely used, the
traditional linear models often fail in real life as effects may not
be linear. Another method called generalized additive models was
introduced to identify and characterize non-linear regression
<span class="citation" data-cites="james2013introduction">(<a href="#ref-james2013introduction" role="doc-biblioref">James et al. 2013</a>)</span>. Smoothing Splines and Natural Cubic
Splines are two standard methods used within GAMs to model
non-linear relationships. To implement these two methods, we used
the <a href="https://cran.r-project.org/web/packages/gam/index.html"><code>gam</code></a>
package in R <span class="citation" data-cites="Trevor2015gam">(<a href="#ref-Trevor2015gam" role="doc-biblioref">Hastie 2015</a>)</span>. The method of GAMs with Smoothing
Splines is a more data-driven and adaptive approach where smoothing
splines can automatically capture non-linear relationships without
specifying the number of knots (specific points where two or more
polynomial segments are joined together to create a
piecewise-defined curve or surface) or the shape of the spline in
advance. On the other hand, natural cubic splines are preferred when
we have prior knowledge or assumptions about the shape of the
non-linear relationship. Natural cubic splines are more
interpretable and can be controlled by the number of knots
<span class="citation" data-cites="elhakeem2022using">(<a href="#ref-elhakeem2022using" role="doc-biblioref">Elhakeem et al. 2022</a>)</span>.</p></li>
</ul>
<h5 class="unnumbered" data-number="2.1.3" id="mathematical-operators">Mathematical Operators</h5>
<p>This section will mention four arithmetic operators, eight distance
measurements, and the exponential approach. Also, unlike other
approaches, in this section, users can apply logarithmic, exponential,
and trigonometric (sinus and cosine) transformations on the markers. Let
<span class="math inline">\(x_{ij}\)</span> represent the value of the <em>j</em>th variable for the <em>i</em>th
observation, with <span class="math inline">\(i=1,2,...,n\)</span> and <span class="math inline">\(j=1,2\)</span>. Let the resulting
combination score for the <em>i</em>th individual be <span class="math inline">\(c_i\)</span>.</p>
<ul>
<li><p><em>Arithmetic Operators:</em> Arithmetic operators such as addition,
multiplication, division, and subtraction can also be used in
diagnostic tests to optimize the AUC, a measure of diagnostic test
performance. These mathematical operations can potentially increase
the AUC and improve the efficacy of diagnostic tests by combining
markers in specific ways. For example, if high values in one test
indicate risk, while low values in the other indicate risk,
subtraction or division can effectively combine these markers.</p></li>
<li><p><em>Distance Measurements:</em> While combining markers with mathematical
operators, a distance measure is used to evaluate the relationships
or similarities between marker values. It’s worth noting that, as
far as we know, no studies have integrated various distinct distance
measures with arithmetic operators in this context. Euclidean
distance is the most commonly used distance measure, which may not
accurately reflect the relationship between markers. Therefore, we
incorporated a variety of distances into the package we developed.
These distances are given as follows
<span class="citation" data-cites="minaev2018distance pandit2011comparative cha2007comprehensive">(<a href="#ref-cha2007comprehensive" role="doc-biblioref">Cha 2007</a>; <a href="#ref-pandit2011comparative" role="doc-biblioref">Pandit et al. 2011</a>; <a href="#ref-minaev2018distance" role="doc-biblioref">Minaev et al. 2018</a>)</span>:<br />
<em>Euclidean:</em>
<span class="math display" id="eq:euclidean-distance">\[c = \sqrt{(x_{i1} - 0)^2 + (x_{i2} - 0)^2} \labe\tag{8}   (#eq:euclidean-distance)\]</span><br />
<em>Manhattan:</em>
<span class="math display" id="eq:manhattan-distance">\[c = |x_{i1} - 0| + |x_{i2} - 0| \labe\tag{9}   (#eq:manhattan-distance)\]</span><br />
<em>Chebyshev:</em>
<span class="math display" id="eq:max-absolute">\[c = \max\{|x_{i1} - 0|, |x_{i2} - 0|\} \labe\tag{10}   (#eq:max-absolute)\]</span><br />
<em>Kulczynskid:</em>
<span class="math display" id="eq:custom-expression">\[c = \frac{|x_{i1} - 0| + |x_{i2} - 0|}{\min\{x_{i1}, x_{i2}\}} \labe\tag{11}   (#eq:custom-expression)\]</span><br />
<em>Lorentzian:</em>
<span class="math display" id="eq:ln-expression">\[c = \ln(1 + |x_{i1} - 0|) + \ln(1 + |x_{i2} - 0|) \labe\tag{12}   (#eq:ln-expression)\]</span><br />
<em>Taneja:</em>
<span class="math display" id="eq:log-expression">\[c = z_1 \left( \log \left( \frac{z_1}{\sqrt{x_{i1} \epsilon}} \right) \right) + z_2 \left( \log \left( \frac{z_2}{\sqrt{x_{i2} \epsilon}} \right) \right) \labe\tag{13}   (#eq:log-expression)\]</span><br />
where
<span class="math inline">\(z_1 = \frac{x_{i1} - 0}{2}, \quad z_2 = \frac{x_{i2} - 0}{2}\)</span><br />
<em>Kumar-Johnson:</em>
<span class="math display" id="eq:c-expression">\[c = \frac{{(x_{i1}^2 - 0)^2}}{{2(x_{i1} \epsilon)^{\frac{3}{2}}}} + \frac{{(x_{i2}^2 - 0)^2}}{{2(x_{i2} \epsilon)^{\frac{3}{2}}}}, \quad  \epsilon=0.0000) \labe\tag{14}   (#eq:c-expression)\]</span><br />
<em>Avg:</em>
<span class="math display" id="eq:c-expression">\[c = \frac{{|x_{i1} - 0| + |x_{i2} - 0| + \max\{(x_{i1} - 0),(x_{i2} - 0)\}}}{2} \labe\tag{14}   (#eq:c-expression)\]</span><br />
</p></li>
<li><p><em>Exponential approach:</em> The exponential approach is another
technique to explore different relationships between the diagnostic
measurements. The methods in which one of the two diagnostic tests
is considered as the base and the other as an exponent can be
represented as <span class="math inline">\(x_{i1}^{(x_{i2})}\)</span> and <span class="math inline">\(x_{i2}^{(x_{i1})}\)</span>. The
specific goals or hypothesis of the analysis, as well as the
characteristics of the diagnostic tests, will determine which method
to use.</p></li>
</ul>
<h5 class="unnumbered" data-number="2.1.4" id="machine-learning-algorithms">Machine-Learning algorithms</h5>
<p>Machine-learning algorithms have been increasingly implemented in
various fields, including the medical field, to combine diagnostic
tests. Integrating diagnostic tests through ML can lead to more
accurate, timely, and personalized diagnoses, which are particularly
valuable in complex medical cases where multiple factors must be
considered. In this study, we aimed to incorporate almost all ML
algorithms in the package we developed. We took advantage of the
<a href="https://cran.r-project.org/web/packages/caret/index.html"><code>caret</code></a>
package in R <span class="citation" data-cites="kuhn2008building">(<a href="#ref-kuhn2008building" role="doc-biblioref">Kuhn 2008</a>)</span> to achieve this goal. This package
includes 190 classification algorithms that could be used to train
models and make predictions. Our study focused on models that use
numerical inputs and produce binary responses depending on the
variables/features and the desired outcome. This selection process
resulted in 113 models we further implemented in our study. We then
classified these 113 models into five classes using the same idea given
in <span class="citation" data-cites="zararsiz2016statistical">(<a href="#ref-zararsiz2016statistical" role="doc-biblioref">Zararsiz et al. 2016</a>)</span>: (i) discriminant classifiers, (ii)
decision tree models, (iii) kernel-based classifiers, (iv) ensemble
classifiers, and (v) others. Like in the
<a href="https://cran.r-project.org/web/packages/caret/index.html"><code>caret</code></a>
package, <code>mlComb()</code> sets up a grid of tuning parameters for a number of
classification routines, fits each model, and calculates a performance
measure based on resampling. After the model fitting, it uses the
<code>predict()</code> function to calculate the probability of the "event"
occurring for each observation. Finally, it performs ROC analysis based
on the probabilities obtained from the prediction step.</p>
<h4 class="unnumbered" data-number="2.2" id="standardization">Standardization</h4>
<p>Standardization is converting/transforming data into a standard scale to
facilitate meaningful comparisons and statistical inference. Many
statistical techniques frequently employ standardization to improve the
interpretability and comparability of data. We implemented five
different standardization methods that can be applied for each marker,
the formulas of which are listed below:</p>
<ul>
<li><p>Z-score: <span class="math inline">\(\frac{{x - \text{mean}(x)}}{{\text{sd}(x)}}\)</span></p></li>
<li><p>T-score:
<span class="math inline">\(\left( \frac{{x - \text{mean}(x)}}{{\text{sd}(x)}} \times 10 \right) + 50\)</span></p></li>
<li><p>min_max_scale: <span class="math inline">\(\frac{{x - \min(x)}}{{\max(x) - \min(x)}}\)</span></p></li>
<li><p>scale_mean_to_one: <span class="math inline">\(\frac{x}{{\text{mean}(x)}}\)</span></p></li>
<li><p>scale_sd_to_one: <span class="math inline">\(\frac{x}{{\text{sd}(x)}}\)</span></p></li>
</ul>
<h4 class="unnumbered" data-number="2.3" id="model-building">Model building</h4>
<p>After specifying a combination method from the
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package, users can build and optimize model parameters using functions
like <code>mlComb()</code>, <code>linComb()</code>, <code>nonlinComb()</code>, and <code>mathComb()</code>,
depending on the specific model selected. Parameter optimization is done
using n-fold cross-validation, repeated n-fold cross-validation, and
bootstrapping methods for linear and non-linear approaches (i.e.,
<code>linComb()</code>, <code>nonlinComb()</code>). Additionally, for machine-learning
approaches (i.e., <code>mlComb()</code>), all of the resampling methods from the
<a href="https://cran.r-project.org/web/packages/caret/index.html"><code>caret</code></a>
package are used to optimize the model parameters. The total number of
parameters being optimized varies across models, and these parameters
are fine-tuned to maximize the AUC. The returned object stores input
data, preprocessed and transformed data, trained model, and resampling
results.</p>
<h4 class="unnumbered" data-number="2.4" id="evaluation-of-model-performances">Evaluation of model performances</h4>
<p>A confusion matrix, as shown in Table <a href="#tab:T1" data-reference-type="ref" data-reference="tab:confusion_matrix">1</a>, is a table used to evaluate the
performance of a classification model and shows the number of correct
and incorrect predictions. It compares predicted and actual</p>
<div id="tab:confusion_matrix">
<table style="width:83%;">
<caption><span id="tab:T1">Table 1: </span> Confusion Matrix</caption>
<colgroup>
<col style="width: 26%" />
<col style="width: 30%" />
<col style="width: 15%" />
<col style="width: 11%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Predicted labels</th>
<th style="text-align: left;">Actual class labels</th>
<th></th>
<th style="text-align: left;">Total</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"></td>
<td style="text-align: left;">Positive</td>
<td>Negative</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">Positive</td>
<td style="text-align: left;">TP</td>
<td>FP</td>
<td style="text-align: left;">TP+FP</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Negative</td>
<td style="text-align: left;">FN</td>
<td>TN</td>
<td style="text-align: left;">FN+TN</td>
</tr>
<tr class="even">
<td style="text-align: left;">Total</td>
<td style="text-align: left;">TP+FN</td>
<td>FP+TN</td>
<td style="text-align: left;">n</td>
</tr>
</tbody>
</table>
</div>
<p>TP: True Positive, TN: True Negative, FP: False Positive, FN: False
Negative, n: Sample size</p>
<p>class labels, with diagonal elements representing the correct
predictions and off-diagonal elements representing the number of
incorrect predictions. The
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package uses the
<a href="https://cran.r-project.org/web/packages/OptimalCutpoints/index.html"><code>OptimalCutpoints</code></a>
package <span class="citation" data-cites="yin2014optimal">(<a href="#ref-yin2014optimal" role="doc-biblioref">Yin and Tian 2014</a>)</span> to generate the confusion matrix and then
<a href="https://cran.r-project.org/web/packages/epiR/index.html"><code>epiR</code></a>
<span class="citation" data-cites="stevenson2017epir">(<a href="#ref-stevenson2017epir" role="doc-biblioref">Stevenson et al. 2017</a>)</span>, including different performance metrics, to
evaluate the performances. Various performance metrics, accuracy rate
(ACC), Kappa statistic (<span class="math inline">\(\kappa\)</span>), sensitivity (SE), specificity (SP),
apparent and true prevalence (AP, TP), positive and negative predictive
values (PPV, NPV), positive and negative likelihood ratio (PLR, NLR),
the proportion of true outcome negative subjects that test positive
(False T+ proportion for true D-), the proportion of true outcome
positive subjects that test negative (False T- proportion for true D+),
the proportion of test-positive subjects that are outcome negative
(False T+ proportion for T+), the proportion of test negative subjects
(False T- proportion for T-) that are outcome positive measures are
available in the
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package. These metrics are summarized in Table
<a href="#tab:T2" data-reference-type="ref" data-reference="tab:performance_metrics">2</a>.</p>
<div id="tab:performance_metrics">
<table style="width:99%;">
<caption><span id="tab:T2">Table 2: </span> Performance Metrics and Formulas</caption>
<colgroup>
<col style="width: 36%" />
<col style="width: 62%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"><strong>Performance Metric</strong></th>
<th style="text-align: left;"><strong>Formula</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Accuracy</td>
<td style="text-align: left;"><span class="math inline">\(\text{ACC} = \frac{{\text{TP} + \text{TN}}}{2}\)</span></td>
</tr>
<tr class="even">
<td style="text-align: left;">Kappa</td>
<td style="text-align: left;"><span class="math inline">\(\kappa = \frac{{\text{ACC} - P_e}}{{1 - P_e}}\)</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;"></td>
<td style="text-align: left;"><span class="math inline">\(P_e = \frac{{(\text{TN} + \text{FN})(\text{TP} + \text{FP}) + (\text{FP} + \text{TN})(\text{FN} + \text{TN})}}{{n^2}}\)</span></td>
</tr>
<tr class="even">
<td style="text-align: left;">Sensitivity (Recall)</td>
<td style="text-align: left;"><span class="math inline">\(\text{SE} = \frac{{\text{TP}}}{{\text{TP} + \text{FN}}}\)</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Specificity</td>
<td style="text-align: left;"><span class="math inline">\(\text{SP} = \frac{{\text{TN}}}{{\text{TN} + \text{FP}}}\)</span></td>
</tr>
<tr class="even">
<td style="text-align: left;">Apparent Prevalence</td>
<td style="text-align: left;"><span class="math inline">\(\text{AP} = \frac{{\text{TP}}}{{n}} + \frac{{\text{FP}}}{{n}}\)</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;">True Prevalence</td>
<td style="text-align: left;"><span class="math inline">\(\text{TP} = \frac{{\text{AP} + \text{SP} - 1}}{{\text{SE} + \text{SP} - 1}}\)</span></td>
</tr>
<tr class="even">
<td style="text-align: left;">Positive Predictive Value (Precision)</td>
<td style="text-align: left;"><span class="math inline">\(\text{PPV} = \frac{{\text{TP}}}{{\text{TP} + \text{FP}}}\)</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Negative Predictive Value</td>
<td style="text-align: left;"><span class="math inline">\(\text{NPV} = \frac{{\text{TN}}}{{\text{TN} + \text{FN}}}\)</span></td>
</tr>
<tr class="even">
<td style="text-align: left;">Positive Likelihood Ratio</td>
<td style="text-align: left;"><span class="math inline">\(\text{PLR} = \frac{{\text{SE}}}{{1 - \text{SP}}}\)</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Negative Likelihood Ratio</td>
<td style="text-align: left;"><span class="math inline">\(\text{NLR} = \frac{{1 - \text{SE}}}{{\text{SP}}}\)</span></td>
</tr>
<tr class="even">
<td style="text-align: left;">The Proportion of True Outcome Negative Subjects That Test Positive</td>
<td style="text-align: left;"><span class="math inline">\(\frac{{\text{FP}}}{{\text{FP} + \text{TN}}}\)</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;">The Proportion of True Outcome Positive Subjects That Test Negative</td>
<td style="text-align: left;"><span class="math inline">\(\frac{{\text{FN}}}{{\text{TP} + \text{FN}}}\)</span></td>
</tr>
<tr class="even">
<td style="text-align: left;">The Proportion of Test Positive Subjects That Are Outcome Negative</td>
<td style="text-align: left;"><span class="math inline">\(\frac{{\text{FP}}}{{\text{TP} + \text{FN}}}\)</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;">The Proportion of Test Negative Subjects That Are Outcome Positive</td>
<td style="text-align: left;"><span class="math inline">\(\frac{{\text{FN}}}{{\text{FN} + \text{TN}}}\)</span></td>
</tr>
</tbody>
</table>
</div>
<h4 class="unnumbered" data-number="2.5" id="prediction-of-the-test-cases">Prediction of the test cases</h4>
<p>The class labels of the observations in the test set are predicted with
the model parameters derived from the training phase. It is critical to
emphasize that the same analytical procedures employed during the
training phase have also been applied to the test set, such as
normalization, transformation, or standardization. More specifically, if
the training set underwent Z-standardization, the test set would
similarly be standardized using the mean and standard deviation derived
from the training set. The class labels of the test set are then
estimated based on the cut-off value established during the training
phase and using the model’s parameters that are trained using the
training set.</p>
<h4 class="unnumbered" data-number="2.6" id="technical-details-and-the-structure-of-dtcomb">Technical details and the structure of dtComb</h4>
<p>The
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package is implemented using the R programming language
(<a href="https://www.r-project.org/" class="uri">https://www.r-project.org/</a>) version 4.2.0. Package development was
facilitated with
<a href="https://cran.r-project.org/web/packages/devtools/index.html"><code>devtools</code></a>
<span class="citation" data-cites="wickham2016devtools">(<a href="#ref-wickham2016devtools" role="doc-biblioref">Wickham et al. 2022</a>)</span> and documented with
<a href="https://cran.r-project.org/web/packages/roxygen2/index.html"><code>roxygen2</code></a>
<span class="citation" data-cites="wickham2013roxygen2">(<a href="#ref-wickham2013roxygen2" role="doc-biblioref">Wickham et al. 2024</a>)</span>. Package testing was performed using 271 unit
tests <span class="citation" data-cites="wickham2011testthat">(<a href="#ref-wickham2011testthat" role="doc-biblioref">Wickham 2024</a>)</span>. Double programming was performed using
Python (<a href="https://www.python.org/" class="uri">https://www.python.org/</a>) to validate the implemented functions
<span class="citation" data-cites="shiralkarprogramming">(<a href="#ref-shiralkarprogramming" role="doc-biblioref">Shiralkar 2010</a>)</span>.<br />
To combine diagnostic tests, the
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package allows the integration of eight linear combination methods,
seven non-linear combination methods, arithmetic operators, and, in
addition to these, eight distance metrics within the scope of
mathematical operators and a total of 113 machine-learning algorithms
from the
<a href="https://cran.r-project.org/web/packages/caret/index.html"><code>caret</code></a>
package <span class="citation" data-cites="kuhn2008building">(<a href="#ref-kuhn2008building" role="doc-biblioref">Kuhn 2008</a>)</span>. These are summarized in Table
<a href="#tab:T3" data-reference-type="ref" data-reference="tab:dtComb_features">3</a>.</p>
<div id="tab:dtComb_features">
<table style="width:97%;">
<caption><span id="tab:T3">Table 3: </span> Features of dtComb</caption>
<colgroup>
<col style="width: 37%" />
<col style="width: 59%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"><strong>Modules (Tab Panels)</strong></th>
<th style="text-align: left;"><strong>Features</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Combination Methods</td>
<td style="text-align: left;"><ul>
<li><p>Linear Combination Approach (8
Different methods)</p></li>
<li><p>Non-linear Combination Approach (7
Different Methods)</p></li>
<li><p>Mathematical Operators (14 Different
methods)</p></li>
<li><p>Machine-Learning Algorithms (113
Different Methods)
<span class="citation" data-cites="kuhn2008building">(<a href="#ref-kuhn2008building" role="doc-biblioref">Kuhn 2008</a>)</span></p></li>
</ul></td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;"><ul>
<li><p>Five standardization methods
applicable to linear, non-linear,
mathematical methods</p></li>
<li><p>16 preprocessing methods applicable
to ML <span class="citation" data-cites="kuhn2008building">(<a href="#ref-kuhn2008building" role="doc-biblioref">Kuhn 2008</a>)</span></p></li>
</ul></td>
</tr>
<tr class="odd">
<td style="text-align: left;"></td>
<td style="text-align: left;"><ul>
<li><p>Three different methods for linear
and non-linear combination methods</p>
<ul>
<li><p>Bootstrapping</p></li>
<li><p>Cross-validation</p></li>
<li><p>Repeated cross-validation</p></li>
</ul></li>
<li><p>12 different resampling methods for
ML <span class="citation" data-cites="kuhn2008building">(<a href="#ref-kuhn2008building" role="doc-biblioref">Kuhn 2008</a>)</span></p></li>
</ul></td>
</tr>
<tr class="even">
<td style="text-align: left;"></td>
<td style="text-align: left;"><ul>
<li>34 different methods for optimum
cutpoints <span class="citation" data-cites="yin2014optimal">(<a href="#ref-yin2014optimal" role="doc-biblioref">Yin and Tian 2014</a>)</span></li>
</ul></td>
</tr>
</tbody>
</table>
</div>
<h3 data-number="3" id="results"><span class="header-section-number">3</span> Results</h3>
<p>Table <a href="#tab:T4" data-reference-type="ref" data-reference="tab:exist_pck">4</a>
summarizes the existing packages and programs, including
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>,
along with the number of combination methods included in each package.
While <strong>mROC</strong> offers only one linear combination method,
<a href="https://github.com/wbaopaul/MaxmzpAUC-R"><code>maxmzpAUC</code></a> and
<a href="https://cran.r-project.org/web/packages/movieROC/index.html"><code>movieROC</code></a>
provide five linear combination techniques each, and
<a href="https://cran.r-project.org/web/packages/SLModels/index.html"><code>SLModels</code></a>
includes four. However, these existing packages primarily focus on
linear combination approaches. In contrast,
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
goes beyond these limitations by integrating not only linear methods but
also non-linear approaches, machine learning algorithms, and
mathematical operators.</p>
<div id="tab:exist_pck">
<table style="width:99%;">
<caption><span id="tab:T4">Table 4: </span> Comparison of dtComb vs. existing packages and
programs</caption>
<colgroup>
<col style="width: 54%" />
<col style="width: 9%" />
<col style="width: 11%" />
<col style="width: 11%" />
<col style="width: 11%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"><strong>Packages&amp;Programs</strong></th>
<th style="text-align: center;"><strong>Linear Comb.</strong></th>
<th style="text-align: center;"><strong>Non-linear Comb.</strong></th>
<th style="text-align: center;"><strong>Math. Operators</strong></th>
<th style="text-align: center;"><strong>ML algorithms</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><strong>mROC</strong> <span class="citation" data-cites="kramar2001mroc">(<a href="#ref-kramar2001mroc" role="doc-biblioref">Kramar et al. 2001</a>)</span></td>
<td style="text-align: center;">1</td>
<td style="text-align: center;">-</td>
<td style="text-align: center;">-</td>
<td style="text-align: center;">-</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="https://github.com/wbaopaul/MaxmzpAUC-R"><code>maxmzpAUC</code></a> <span class="citation" data-cites="yu2015two">(<a href="#ref-yu2015two" role="doc-biblioref">Yu and Park 2015</a>)</span></td>
<td style="text-align: center;">5</td>
<td style="text-align: center;">-</td>
<td style="text-align: center;">-</td>
<td style="text-align: center;">-</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="https://cran.r-project.org/web/packages/movieROC/index.html"><code>movieROC</code></a> <span class="citation" data-cites="perez2021visualizing">(<a href="#ref-perez2021visualizing" role="doc-biblioref">Pérez-Fernández et al. 2021</a>)</span></td>
<td style="text-align: center;">5</td>
<td style="text-align: center;">-</td>
<td style="text-align: center;">-</td>
<td style="text-align: center;">-</td>
</tr>
<tr class="even">
<td style="text-align: left;"><a href="https://cran.r-project.org/web/packages/SLModels/index.html"><code>SLModels</code></a> <span class="citation" data-cites="aznar-gimeno2023comparing">(<a href="#ref-aznar-gimeno2023comparing" role="doc-biblioref">Aznar-Gimeno et al. 2023</a>)</span></td>
<td style="text-align: center;">4</td>
<td style="text-align: center;">-</td>
<td style="text-align: center;">-</td>
<td style="text-align: center;">-</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a></td>
<td style="text-align: center;">8</td>
<td style="text-align: center;">7</td>
<td style="text-align: center;">14</td>
<td style="text-align: center;">113</td>
</tr>
</tbody>
</table>
</div>
<h4 class="unnumbered" data-number="3.1" id="dataset">Dataset</h4>
<p>To demonstrate the functionality of the
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package, we conduct a case study using four different combination
methods. The data used in this study were obtained from patients who
presented at Erciyes University Faculty of Medicine, Department of
General Surgery, with complaints of abdominal pain
<span class="citation" data-cites="zararsiz2016statistical akyildiz2010value">(<a href="#ref-akyildiz2010value" role="doc-biblioref">Akyildiz et al. 2010</a>; <a href="#ref-zararsiz2016statistical" role="doc-biblioref">Zararsiz et al. 2016</a>)</span>. The dataset comprised
D-dimer levels (<em>D_dimer</em>) and leukocyte counts (<em>log_leukocyte</em>) of 225
patients, divided into two groups (<em>Group</em>): the first group consisted
of 110 patients who required an immediate laparotomy (<em>needed</em>). In
comparison, the second group comprised 115 patients who did not
(<em>not_needed</em>). After the evaluation of conventional treatment, the
patients who underwent surgery due to their postoperative pathologies
are placed in the first group. In contrast, those with a negative result
from their laparotomy were assigned to the second group. All the
analyses were performed by following a workflow given in Fig.
<a href="#figure:workflow" data-reference-type="ref" data-reference="figure:workflow">1</a>.
First of all, the
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package should be loaded in order to use related functions.</p>
<figure id="figure:workflow">
<img src="Figure/Figure_1.png" style="width:81.0%" alt="graphic without alt text" />
<figcaption>Figure 1: <strong>Combination steps of two diagnostic tests.</strong> The figure
presents a schematic representation of the sequential steps involved in
combining two diagnostic tests using a combination
method.</figcaption>
</figure>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load dtComb package</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dtComb)</span></code></pre></div>
<p>Similarly, the laparotomy data can be loaded from the R database by
using the following R code:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># load laparotomy data</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(laparotomy)</span></code></pre></div>
<h4 class="unnumbered" data-number="3.2" id="implementation-of-the-dtcomb-package">Implementation of the dtComb package</h4>
<p>In order to demonstrate the applicability of the
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package, the implementation of an arbitrarily chosen method from each of
the linear, non-linear, mathematical operator and machine learning
approaches is demonstrated and their performance is compared. These
methods are Pepe, Cai &amp; Langton for linear combination, Splines for
non-linear, Addition for mathematical operator and SVM for
machine-learning. Before applying the methods, we split the data into
two parts: a training set comprising 70% of the data and a test set
comprising the remaining 30%.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Splitting the data set into train and test (70%-30%)</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2128</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>inTrain <span class="ot">&lt;-</span> caret<span class="sc">::</span><span class="fu">createDataPartition</span>(laparotomy<span class="sc">$</span>group, <span class="at">p =</span> <span class="fl">0.7</span>, <span class="at">list =</span> <span class="cn">FALSE</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>trainData <span class="ot">&lt;-</span> laparotomy[inTrain, ]</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(trainData) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Group&quot;</span>, <span class="st">&quot;D_dimer&quot;</span>, <span class="st">&quot;log_leukocyte&quot;</span>)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>testData <span class="ot">&lt;-</span> laparotomy[<span class="sc">-</span>inTrain, <span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="co"># define marker and status for combination function</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> trainData[, <span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>status <span class="ot">&lt;-</span> <span class="fu">factor</span>(trainData<span class="sc">$</span>Group, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">&quot;not_needed&quot;</span>, <span class="st">&quot;needed&quot;</span>))</span></code></pre></div>
<p>The model is trained on <code>trainData</code> and the resampling parameters used
in the training phase are chosen as ten repeat five fold repeated
cross-validation. Direction = ‘&lt;’ is chosen, as higher values indicate
higher risks. The Youden index was chosen among the cut-off methods. We
note that markers are not standardised and results are presented at the
confidence level (CI 95%). Four main combination functions are run with
the selected methods as follows.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># PCL method</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>fit.lin.PCL <span class="ot">&lt;-</span> <span class="fu">linComb</span>(<span class="at">markers =</span> markers,  <span class="at">status =</span> status, <span class="at">event =</span> <span class="st">&quot;needed&quot;</span>, </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">method =</span> <span class="st">&quot;PCL&quot;</span>, <span class="at">resample =</span> <span class="st">&quot;repeatedcv&quot;</span>, <span class="at">nfolds =</span> <span class="dv">5</span>,</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">nrepeats =</span> <span class="dv">10</span>, <span class="at">direction =</span> <span class="st">&quot;&lt;&quot;</span>, <span class="at">cutoff.method =</span> <span class="st">&quot;Youden&quot;</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># splines method (degree = 3 and degrees of freedom = 3)</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>fit.nonlin.splines <span class="ot">&lt;-</span> <span class="fu">nonlinComb</span>(<span class="at">markers =</span> markers, <span class="at">status =</span> status, <span class="at">event =</span> <span class="st">&quot;needed&quot;</span>, </span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">method =</span> <span class="st">&quot;splines&quot;</span>, <span class="at">resample =</span> <span class="st">&quot;repeatedcv&quot;</span>, <span class="at">nfolds =</span> <span class="dv">5</span>, </span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">nrepeats =</span> <span class="dv">10</span>, <span class="at">cutoff.method =</span> <span class="st">&quot;Youden&quot;</span>, <span class="at">direction =</span> <span class="st">&quot;&lt;&quot;</span>, </span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">df1 =</span> <span class="dv">3</span>, <span class="at">df2 =</span> <span class="dv">3</span>)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="co">#add operator</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a> fit.add <span class="ot">&lt;-</span> <span class="fu">mathComb</span>(<span class="at">markers =</span> markers, <span class="at">status =</span> status, <span class="at">event =</span> <span class="st">&quot;needed&quot;</span>,</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>                     <span class="at">method =</span> <span class="st">&quot;add&quot;</span>, <span class="at">direction =</span> <span class="st">&quot;&lt;&quot;</span>, <span class="at">cutoff.method =</span> <span class="st">&quot;Youden&quot;</span>)</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="co">#SVM</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>fit.svm <span class="ot">&lt;-</span> <span class="fu">mlComb</span>(<span class="at">markers =</span> markers, <span class="at">status =</span> status, <span class="at">event =</span> <span class="st">&quot;needed&quot;</span>, <span class="at">method =</span> <span class="st">&quot;svmLinear&quot;</span>, </span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>                 <span class="at">resample =</span> <span class="st">&quot;repeatedcv&quot;</span>, <span class="at">nfolds  =</span> <span class="dv">5</span>,<span class="at">nrepeats =</span> <span class="dv">10</span>, <span class="at">direction =</span> <span class="st">&quot;&lt;&quot;</span>, </span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>                 <span class="at">cutoff.method =</span> <span class="st">&quot;Youden&quot;</span>)</span></code></pre></div>
<p>Various measures were considered to compare model performances,
including AUC, ACC, SEN, SPE, PPV, and NPV. AUC statistics, with 95% CI,
have been calculated for each marker and method. The resulting
statistics are as follows: 0.816 (0.751–0.880), 0.802 (0.728–0.877),
0.888 (0.825–0.930), 0.911 (0.868–0.954), 0.877 (0.824-0.929), and
0.875 (0.821-0.930) for D-dimer, Log(leukocyte), Pepe, Cai &amp; Langton,
Splines, Addition, and Support Vector Machine (SVM). The results
revealed that the predictive performances of markers and the combination
of markers are significantly higher than random chance in determining
the use of laparotomy (<span class="math inline">\(p&lt;0.05\)</span>). The highest sensitivity and NPV were
observed with the Addition method, while the highest specificity and PPV
were observed with the Splines method. According to the overall AUC and
accuracies, the combined approach fitted with the Splines method
performed better than the other methods (Fig.
<a href="#figure:radar" data-reference-type="ref" data-reference="figure:radar">2</a>).
Therefore, the Splines method will be used in the subsequent analysis of
the findings.</p>
<figure id="figure:radar">
<img src="Figure/Figure_4.png" style="width:100.0%" alt="graphic without alt text" />
<figcaption>Figure 2: <strong>Radar plots of trained models and performance measures of
two markers.</strong> Radar plots summarize the diagnostic performances of two
markers and various combination methods in the training dataset. These
plots illustrate the performance metrics such as AUC, ACC, SEN, SPE,
PPV, and NPV measurements. In these plots, the width of the polygon
formed by connecting each point indicates the model’s performance in
terms of AUC, ACC, SEN, SPE, PPV, and NPV metrics. It can be observed
that the polygon associated with the Splines method occupies the most
extensive area, which means that the Splines method performed better
than the other methods.</figcaption>
</figure>
<p>For the AUC of markers and the spline model:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>fit.nonlin.splines<span class="sc">$</span>AUC_table</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>                    AUC     SE.AUC LowerLimit UpperLimit         z      p.value</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>D_dimer       <span class="fl">0.8156966</span> <span class="fl">0.03303310</span>  <span class="fl">0.7509530</span>  <span class="fl">0.8804403</span>  <span class="fl">9.556979</span> <span class="fl">1.212446e-21</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>log_leukocyte <span class="fl">0.8022286</span> <span class="fl">0.03791768</span>  <span class="fl">0.7279113</span>  <span class="fl">0.8765459</span>  <span class="fl">7.970652</span> <span class="fl">1.578391e-15</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>Combination   <span class="fl">0.9111752</span> <span class="fl">0.02189588</span>  <span class="fl">0.8682601</span>  <span class="fl">0.9540904</span> <span class="fl">18.778659</span> <span class="fl">1.128958e-78</span></span></code></pre></div>
<p>Here:<br />
<code>SE</code>: Standard Error.<br />
The area under ROC curves for D-dimer levels and leukocyte counts on the
logarithmic scale and combination score were 0.816, 0.802, and 0.911,
respectively. The ROC curves generated with the combination score from
the splines model, D-dimer levels, and leukocyte count markers are also
given in Fig. <a href="#figure:roc" data-reference-type="ref" data-reference="figure:roc">3</a>, showing that the combination score has the
highest AUC. It is observed that the splines method significantly
improved between 9.5% and 10.9% in AUC statistics compared to D-dimer
level and leukocyte counts, respectively.</p>
<figure id="figure:roc">
<img src="Figure/Figure_2.png" style="width:70.0%" alt="graphic without alt text" />
<figcaption>Figure 3: <strong>ROC curves.</strong> ROC curves for combined diagnostic tests,
with sensitivity displayed on the y-axis and 1-specificity displayed on
the x-axis. As can be observed, the combination score produced the
highest AUC value, indicating that the combined strategy performs the
best overall.</figcaption>
</figure>
<p>To see the results of the binary comparison between the combination
score and markers:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>fit.nonlin.splines<span class="sc">$</span>MultComp_table</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">Marker1</span> (A)   <span class="fu">Marker2</span> (B)   <span class="fu">AUC</span> (A)   <span class="fu">AUC</span> (B)      <span class="sc">|</span>A<span class="sc">-</span>B<span class="sc">|</span>  <span class="fu">SE</span>(<span class="sc">|</span>A<span class="sc">-</span>B<span class="sc">|</span>)         z      p<span class="sc">-</span>value</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> Combination       D_dimer <span class="fl">0.9079686</span> <span class="fl">0.8156966</span> <span class="fl">0.09227193</span> <span class="fl">0.02223904</span> <span class="fl">4.1490971</span> <span class="fl">3.337893e-05</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> Combination log_leukocyte <span class="fl">0.9079686</span> <span class="fl">0.8022286</span> <span class="fl">0.10573994</span> <span class="fl">0.03466544</span> <span class="fl">3.0502981</span> <span class="fl">2.286144e-03</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>     D_dimer log_leukocyte <span class="fl">0.8156966</span> <span class="fl">0.8022286</span> <span class="fl">0.01346801</span> <span class="fl">0.04847560</span> <span class="fl">0.2778308</span> <span class="fl">7.811423e-01</span></span></code></pre></div>
<p>Controlling Type I error using Bonferroni correction, comparison of
combination score with markers yielded significant results (<span class="math inline">\(p&lt;0.05\)</span>).<br />
To demonstrate the diagnostic test results and performance measures for
non-linear combination approach, the following code can be used:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>fit.nonlin.splines<span class="sc">$</span>DiagStatCombined</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>          Outcome <span class="sc">+</span>    Outcome <span class="sc">-</span>      Total</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>Test <span class="sc">+</span>           <span class="dv">66</span>           <span class="dv">13</span>         <span class="dv">79</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>Test <span class="sc">-</span>           <span class="dv">11</span>           <span class="dv">68</span>         <span class="dv">79</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>Total            <span class="dv">77</span>           <span class="dv">81</span>        <span class="dv">158</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>Point estimates and <span class="dv">95</span>% CIs<span class="sc">:</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="sc">--------------------------------------------------------------</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>Apparent prevalence <span class="sc">*</span>                  <span class="fl">0.50</span> (<span class="fl">0.42</span>, <span class="fl">0.58</span>)</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>True prevalence <span class="sc">*</span>                      <span class="fl">0.49</span> (<span class="fl">0.41</span>, <span class="fl">0.57</span>)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>Sensitivity <span class="sc">*</span>                          <span class="fl">0.86</span> (<span class="fl">0.76</span>, <span class="fl">0.93</span>)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>Specificity <span class="sc">*</span>                          <span class="fl">0.84</span> (<span class="fl">0.74</span>, <span class="fl">0.91</span>)</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>Positive predictive value <span class="sc">*</span>            <span class="fl">0.84</span> (<span class="fl">0.74</span>, <span class="fl">0.91</span>)</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>Negative predictive value <span class="sc">*</span>            <span class="fl">0.86</span> (<span class="fl">0.76</span>, <span class="fl">0.93</span>)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>Positive likelihood ratio              <span class="fl">5.34</span> (<span class="fl">3.22</span>, <span class="fl">8.86</span>)</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>Negative likelihood ratio              <span class="fl">0.17</span> (<span class="fl">0.10</span>, <span class="fl">0.30</span>)</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>False T<span class="sc">+</span> proportion <span class="cf">for</span> true D<span class="sc">-</span> <span class="er">*</span>      <span class="fl">0.16</span> (<span class="fl">0.09</span>, <span class="fl">0.26</span>)</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>False T<span class="sc">-</span> proportion <span class="cf">for</span> true D<span class="sc">+</span> <span class="er">*</span>      <span class="fl">0.14</span> (<span class="fl">0.07</span>, <span class="fl">0.24</span>)</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>False T<span class="sc">+</span> proportion <span class="cf">for</span> T<span class="sc">+</span> <span class="er">*</span>           <span class="fl">0.16</span> (<span class="fl">0.09</span>, <span class="fl">0.26</span>)</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>False T<span class="sc">-</span> proportion <span class="cf">for</span> T<span class="sc">-</span> <span class="er">*</span>           <span class="fl">0.14</span> (<span class="fl">0.07</span>, <span class="fl">0.24</span>)</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>Correctly classified proportion <span class="sc">*</span>      <span class="fl">0.85</span> (<span class="fl">0.78</span>, <span class="fl">0.90</span>)</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a><span class="sc">--------------------------------------------------------------</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a><span class="er">*</span> Exact CIs</span></code></pre></div>
<p>Furthermore, if the diagnostic test results and performance measures of
the combination score are compared with the results of the single
markers, it can be observed that the TN value of the combination score
is higher than that of the single markers, and the combination of
markers has higher specificity and positive-negative predictive value
than the log-transformed leukocyte counts and D-dimer level (Table
<a href="#tab:T5" data-reference-type="ref" data-reference="tab:diagnostic_measures">5</a>).
Conversely, D-dimer has a higher sensitivity than the others. Optimal
cut-off values for both markers and the combined approach are also given
in this table.</p>
<div id="tab:diagnostic_measures">
<table style="width:99%;">
<caption><span id="tab:T5">Table 5: </span> Statistical diagnostic measures with 95% confidence
intervals for each marker and the combination score</caption>
<colgroup>
<col style="width: 25%" />
<col style="width: 20%" />
<col style="width: 26%" />
<col style="width: 26%" />
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"><strong>Diagnostic Measures (95% CI)</strong></th>
<th style="text-align: center;"><strong>D-dimer level (<span class="math inline">\(&gt;1.6\)</span>)</strong></th>
<th style="text-align: center;"><strong>Log(leukocyte count) (<span class="math inline">\(&gt;4.16\)</span>)</strong></th>
<th style="text-align: center;"><strong>Combination score (<span class="math inline">\(&gt;0.448\)</span>)</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">TP</td>
<td style="text-align: center;">66</td>
<td style="text-align: center;">61</td>
<td style="text-align: center;">65</td>
</tr>
<tr class="even">
<td style="text-align: left;">TN</td>
<td style="text-align: center;">53</td>
<td style="text-align: center;">60</td>
<td style="text-align: center;">69</td>
</tr>
<tr class="odd">
<td style="text-align: left;">FP</td>
<td style="text-align: center;">28</td>
<td style="text-align: center;">21</td>
<td style="text-align: center;">12</td>
</tr>
<tr class="even">
<td style="text-align: left;">FN</td>
<td style="text-align: center;">11</td>
<td style="text-align: center;">16</td>
<td style="text-align: center;">12</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Apparent prevalence</td>
<td style="text-align: center;">0.59 (0.51-0.67)</td>
<td style="text-align: center;">0.52 (0.44-0.60)</td>
<td style="text-align: center;">0.49 (0.41-0.57)</td>
</tr>
<tr class="even">
<td style="text-align: left;">True prevalence</td>
<td style="text-align: center;">0.49 (0.41-0.57)</td>
<td style="text-align: center;">0.49 (0.41-0.57)</td>
<td style="text-align: center;">0.49 (0.41-0.57)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Sensitivity</td>
<td style="text-align: center;">0.86 (0.76-0.93)</td>
<td style="text-align: center;">0.79 (0.68-0.88)</td>
<td style="text-align: center;">0.84 (0.74-0.92)</td>
</tr>
<tr class="even">
<td style="text-align: left;">Specificity</td>
<td style="text-align: center;">0.65 (0.54-0.76)</td>
<td style="text-align: center;">0.74 (0.63-0.83)</td>
<td style="text-align: center;">0.85 (0.76-0.92)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Positive predictive value</td>
<td style="text-align: center;">0.70 (0.60-0.79)</td>
<td style="text-align: center;">0.74 (0.64-0.83)</td>
<td style="text-align: center;">0.84 (0.74-0.92)</td>
</tr>
<tr class="even">
<td style="text-align: left;">Negative predictive value</td>
<td style="text-align: center;">0.83 (0.71-0.91)</td>
<td style="text-align: center;">0.79 (0.68-0.87)</td>
<td style="text-align: center;">0.85 (0.76-0.92)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Positive likelihood ratio</td>
<td style="text-align: center;">2.48 (1.81-3.39)</td>
<td style="text-align: center;">3.06 (2.08-4.49)</td>
<td style="text-align: center;">5.70 (3.35-9.69)</td>
</tr>
<tr class="even">
<td style="text-align: left;">Negative likelihood ratio</td>
<td style="text-align: center;">0.22 (0.12-0.39)</td>
<td style="text-align: center;">0.28 (0.18-0.44)</td>
<td style="text-align: center;">0.18 (0.11-0.31)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">False T+ proportion for true D-</td>
<td style="text-align: center;">0.35 (0.24-0.46)</td>
<td style="text-align: center;">0.26 (0.17-0.37)</td>
<td style="text-align: center;">0.15 (0.08-0.24)</td>
</tr>
<tr class="even">
<td style="text-align: left;">False T- proportion for true D+</td>
<td style="text-align: center;">0.14 (0.07-0.24)</td>
<td style="text-align: center;">0.21 (0.12-0.32)</td>
<td style="text-align: center;">0.16 (0.08-0.26)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">False T+ proportion for T+</td>
<td style="text-align: center;">0.30 (0.21-0.40)</td>
<td style="text-align: center;">0.26 (0.17-0.36)</td>
<td style="text-align: center;">0.16 (0.08-0.26)</td>
</tr>
<tr class="even">
<td style="text-align: left;">False T- proportion for T-</td>
<td style="text-align: center;">0.17 (0.09-0.29)</td>
<td style="text-align: center;">0.21 (0.13-0.32)</td>
<td style="text-align: center;">0.15 (0.08-0.24)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Accuracy</td>
<td style="text-align: center;">0.75 (0.68-0.82)</td>
<td style="text-align: center;">0.77 (0.69-0.83)</td>
<td style="text-align: center;">0.85 (0.78-0.90)</td>
</tr>
</tbody>
</table>
</div>
<p>For a comprehensive analysis, the <code>plotComb</code> function in
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
can be used to generate plots of the kernel density and individual-value
of combination scores of each group and the specificity and sensitivity
corresponding to different cut-off point values Fig.
<a href="#figure:scatter" data-reference-type="ref" data-reference="figure:scatter">4</a>.
This function requires the result of the <code>nonlinComb</code> function, which is
an object of the “dtComb” class and <code>status</code> which is of factor type.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># draw distribution, dispersion, and specificity and sensitivity plots</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotComb</span>(fit.nonlin.splines, status)</span></code></pre></div>
<figure id="figure:scatter">
<img src="Figure/Figure_3.png" style="width:100.0%" alt="graphic without alt text" />
<figcaption>Figure 4: <strong>Kernel density, individual-value, and sens&amp;spe plots of
the combination score acquired with the training model.</strong> Kernel density
of the combination score for two groups: needed and not needed (a).
Individual-value graph with classes on the x-axis and combination score
on the y-axis (b). Sensitivity and specificity graph of the combination
score c. While colors show each class in Figures (a) and (b), in Figure
(c), the colors represent the sensitivity and specificity of the
combination score.</figcaption>
</figure>
<p>If the model trained with Splines is to be tested, the generically
written predict function is used. This function requires the test set
and the result of the <code>nonlinComb</code> function, which is an object of the
“dtComb” class. As a result of prediction, the output for each
observation consisted of the combination score and the predicted label
determined by the cut-off value derived from the model.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># To predict the test set </span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>pred <span class="ot">&lt;-</span> <span class="fu">predict</span>(fit.nonlin.splines, testData)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pred)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>   comb.score labels</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>   <span class="fl">0.6133884</span> needed</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="dv">7</span>   <span class="fl">0.9946474</span> needed</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="dv">10</span>  <span class="fl">0.9972347</span> needed</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="dv">11</span>  <span class="fl">0.9925040</span> needed</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="dv">13</span>  <span class="fl">0.9257699</span> needed</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a><span class="dv">14</span>  <span class="fl">0.9847090</span> needed</span></code></pre></div>
<p>Above, it can be seen that the estimated combination scores for the
first six observations in the test set were labelled as <strong>needed</strong>
because they were higher than the cut-off value of 0.448.</p>
<h4 class="unnumbered" data-number="3.3" id="web-interface-for-the-dtcomb-package">Web interface for the dtComb package</h4>
<p>The primary goal of developing the
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package is to combine numerous distinct combination methods and make
them easily accessible to researchers. Furthermore, the package includes
diagnostic statistics and visualization tools for diagnostic tests and
the combination score generated by the chosen method. Nevertheless, it
is worth noting that using R code may pose challenges for physicians and
those unfamiliar with R programming. We have also developed a
user-friendly web application for
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
using
<a href="https://cran.r-project.org/web/packages/shiny/index.html"><code>Shiny</code></a>
<span class="citation" data-cites="chang2017shiny">(<a href="#ref-chang2017shiny" role="doc-biblioref">Chang et al. 2024</a>)</span> to address this. This web-based tool is publicly
accessible and provides an interactive interface with all the
functionalities found in the
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package.<br />
To initiate the analysis, users must upload their data by following the
instructions outlined in the "Data upload" tab of the web tool. For
convenience, we have provided three example datasets on this page to
assist researchers in practicing the tool’s functionality and to guide
them in formatting their own data (as illustrated in Fig.
<a href="#figure:web" data-reference-type="ref" data-reference="figure:web">5</a>a). We also
note that ROC analysis for a single marker can be performed within the
‘ROC Analysis for Single Marker(s)’ tab in the data upload section of
the web interface.</p>
<p>In the "Analysis" tab, one can find two crucial subpanels:</p>
<ul>
<li><p>Plots (Fig. <a href="#figure:web" data-reference-type="ref" data-reference="figure:web">5</a>b): This section offers various visual
representations, such as ROC curves, kernel density plots,
individual-value plots, and sensitivity and specificity plots. These
visualizations help users assess single diagnostic tests and the
combination score generated using user-defined combination methods.</p></li>
<li><p>Results (Fig. <a href="#figure:web" data-reference-type="ref" data-reference="figure:web">5</a>c): In this subpanel, one can access a range
of statistics. It provides insights into the combination score and
single diagnostic tests, AUC statistics, and comparisons to evaluate
how the combination score fares against individual diagnostic tests,
and various diagnostic measures. One can also predict new data based
on the model parameters set previously and stored in the "Predict"
tab (Fig. <a href="#figure:web" data-reference-type="ref" data-reference="figure:web">5</a>d). If needed, one can download the model
created during the analysis to keep the parameters of the fitted
model. This lets users make new predictions by reloading the model
from the "Predict" tab. Additionally, all the results can easily
be downloaded using the dedicated download buttons in their
respective tabs.</p></li>
</ul>
<figure id="figure:web">
<img src="Figure/Figure_5.png" style="width:100.0%" alt="graphic without alt text" />
<figcaption>Figure 5: <strong>Web interface of the dtComb package.</strong> The figure
illustrates the web interface of the
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package, which demonstrates the steps involved in combining two
diagnostic tests. a) Data Upload: The user is able to upload the dataset
and select relevant markers, a gold standard test, and an event factor
for analysis.b) Combination Analysis: This panel allows the selection of
the combination method, method-specific parameters, and resampling
options to refine the analysis. c) Combination Analysis Output: Displays
the results generated by the selected combination method, providing the
user with key metrics and visualizations for interpretation. d) Predict:
Displays the prediction results of the trained model when applied to the
test set.</figcaption>
</figure>
<h3 data-number="4" id="summary-and-further-research"><span class="header-section-number">4</span> Summary and further research</h3>
<p>In clinical practice, multiple diagnostic tests are possible for disease
diagnosis <span class="citation" data-cites="yu2015two">(<a href="#ref-yu2015two" role="doc-biblioref">Yu and Park 2015</a>)</span>. Combining these tests to enhance diagnostic
accuracy is a widely accepted approach
<span class="citation" data-cites="su1993linear pepe2000combining liu2011min sameera2016binary pepe2006combining todor2014tools">(<a href="#ref-su1993linear" role="doc-biblioref">Su and Liu 1993</a>; <a href="#ref-pepe2000combining" role="doc-biblioref">Pepe and Thompson 2000</a>; <a href="#ref-pepe2006combining" role="doc-biblioref">Pepe et al. 2006</a>; <a href="#ref-liu2011min" role="doc-biblioref">Liu et al. 2011</a>; <a href="#ref-todor2014tools" role="doc-biblioref">Todor et al. 2014</a>; <a href="#ref-sameera2016binary" role="doc-biblioref">Sameera et al. 2016</a>)</span>.
As far as we know, the tools in Table <a href="#tab:T4" data-reference-type="ref" data-reference="tab:exist_pck">4</a> have been designed to combine diagnostic
tests but only contain at most five different combination methods. As a
result, despite the existence of numerous advanced combination methods,
there is currently no extensive tool available for integrating
diagnostic tests.<br />
In this study, we presented
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>, a
comprehensive R package designed to combine diagnostic tests using
various methods, including linear, non-linear, mathematical operators,
and machine learning algorithms. The package integrates 142 different
methods for combining two diagnostic markers to improve the accuracy of
diagnosis. The package also provides ROC curve analysis, various
graphical approaches, diagnostic performance scores, and binary
comparison results. In the given example, one can determine whether
patients with abdominal pain require laparotomy by combining the D-dimer
levels and white blood cell counts of those patients. Various methods,
such as linear and non-linear combinations, were tested, and the results
showed that the Splines method performed better than the others,
particularly in terms of AUC and accuracy compared to single tests. This
shows that diagnostic accuracy can be improved with combination
methods.<br />
Future work can focus on extending the capabilities of the
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a>
package. While some studies focus on combining multiple markers, our
study aimed to combine two markers using nearly all existing methods and
develop a tool and package for clinical practice <span class="citation" data-cites="kang2016linear">(<a href="#ref-kang2016linear" role="doc-biblioref">Kang et al. 2016</a>)</span>.</p>
<h4 class="unnumbered" data-number="4.1" id="r-software">R Software</h4>
<p>The R package
<a href="https://cran.r-project.org/web/packages/dtComb/index.html"><code>dtComb</code></a> is
now available on the CRAN website
<a href="https://cran.r-project.org/web/packages/dtComb/index.html" class="uri">https://cran.r-project.org/web/packages/dtComb/index.html</a>.</p>
<h4 class="unnumbered" data-number="4.2" id="acknowledgment">Acknowledgment</h4>
<p>We would like to thank the Proofreading &amp; Editing Office of the Dean for
Research at Erciyes University for the copyediting and proofreading
service for this manuscript.</p>
</div>
<h3 class="appendix" data-number="5" id="note"><span class="header-section-number">5</span> Note</h3>
<p>This article is converted from a Legacy LaTeX article using the
<a href="https://cran.r-project.org/package=texor">texor</a> package.
The pdf version is the official version. To report a problem with the html,
refer to CONTRIBUTE on the R Journal homepage.</p>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-tensorflow2015-whitepaper" class="csl-entry" role="listitem">
M. Abadi, A. Agarwal, P. Barham, E. Brevdo, Z. Chen, C. Citro, G. S. Corrado, A. Davis, J. Dean, M. Devin, et al. <span>TensorFlow</span>: Large-scale machine learning on heterogeneous systems. 2015. URL <a href="https://www.tensorflow.org/">https://www.tensorflow.org/</a>. Software available from tensorflow.org.
</div>
<div id="ref-agarwal2023artificial" class="csl-entry" role="listitem">
S. Agarwal, A. S. Yadav, V. Dinesh, K. S. S. Vatsav, K. S. S. Prakash and S. Jaiswal. By artificial intelligence algorithms and machine learning models to diagnosis cancer. <em>Materials Today: Proceedings</em>, 80: 2969–2975, 2023.
</div>
<div id="ref-ahsan2024advancements" class="csl-entry" role="listitem">
M. Ahsan, A. Khan, K. R. Khan, B. B. Sinha and A. Sharma. Advancements in medical diagnosis and treatment through machine learning: A review. <em>Expert Systems</em>, 41(3): e13499, 2024.
</div>
<div id="ref-akyildiz2010value" class="csl-entry" role="listitem">
H. Y. Akyildiz, E. Sozuer, A. Akcan, C. Kuçuk, T. Artis, İ. Biri and N. Yılmaz. The value of <span class="nocase">D-dimer</span> test in the diagnosis of patients with nontraumatic acute abdomen. <em>Turkish Journal of Trauma and Emergency Surgery</em>, 16(1): 22–26, 2010.
</div>
<div id="ref-alzyoud2024diagnosing" class="csl-entry" role="listitem">
M. Alzyoud, R. Alazaidah, M. Aljaidi, G. Samara, M. Qasem, M. Khalid and N. Al-Shanableh. Diagnosing diabetes mellitus using machine learning techniques. <em>International Journal of Data and Network Science</em>, 8(1): 179–188, 2024.
</div>
<div id="ref-anderson1962classification" class="csl-entry" role="listitem">
T. W. Anderson and R. R. Bahadur. Classification into two multivariate normal distributions with different covariance matrices. <em>The annals of mathematical statistics</em>, 420–431, 1962.
</div>
<div id="ref-aznar2022stepwise" class="csl-entry" role="listitem">
R. Aznar-Gimeno, L. M. Esteban, R. del-Hoyo-Alonso, Á. Borque-Fernando and G. Sanz. A stepwise algorithm for linearly combining biomarkers under <span>Youden</span> index maximization. <em>Mathematics</em>, 10(8): 1221, 2022.
</div>
<div id="ref-aznar-gimeno2023comparing" class="csl-entry" role="listitem">
R. Aznar-Gimeno, L. M. Esteban, G. Sanz and R. del-Hoyo-Alonso. Comparing the min–max–median/<span>IQR</span> approach with the min–max approach, logistic regression and <span>XGBoost</span>, maximising the <span>Youden</span> index. <em>Symmetry (Basel)</em>, 15: 2023. URL <a href="https://doi.org/10.3390/sym15030756">https://doi.org/10.3390/sym15030756</a>.
</div>
<div id="ref-bansal2013does" class="csl-entry" role="listitem">
A. Bansal and M. Sullivan Pepe. When does combining markers improve classification performance and what are implications for practice? <em>Statistics in medicine</em>, 32(11): 1877–1892, 2013.
</div>
<div id="ref-cha2007comprehensive" class="csl-entry" role="listitem">
S.-H. Cha. Comprehensive survey on distance/similarity measures between probability density functions. <em>City</em>, 1(2): 1, 2007.
</div>
<div id="ref-chang2017shiny" class="csl-entry" role="listitem">
W. Chang, J. Cheng, J. Allaire, C. Sievert, B. Schloerke, Y. Xie, J. Allen, J. McPherson, A. Dipert and B. Borges. Shiny: Web application framework for <span>R</span>. 2024. URL <a href="https://shiny.posit.co/">https://shiny.posit.co/</a>. R package version 1.9.1.9000, https://github.com/rstudio/shiny.
</div>
<div id="ref-cox1989analysis" class="csl-entry" role="listitem">
D. R. Cox and E. J. Snell. <em>Analysis of binary data.</em> 2nd ed London: Chapman; Hall/CRC, 1989.
</div>
<div id="ref-degroat2024discovering" class="csl-entry" role="listitem">
W. DeGroat, H. Abdelhalim, K. Patel, D. Mendhe, S. Zeeshan and Z. Ahmed. Discovering biomarkers associated and predicting cardiovascular disease with high accuracy using a novel nexus of machine learning techniques for precision medicine. <em>Scientific reports</em>, 14(1): 1, 2024.
</div>
<div id="ref-du2024likelihood" class="csl-entry" role="listitem">
Z. Du, P. Du and A. Liu. Likelihood ratio combination of multiple biomarkers via smoothing spline estimated densities. <em>Statistics in Medicine</em>, 43(7): 1372–1383, 2024.
</div>
<div id="ref-efron1975efficiency" class="csl-entry" role="listitem">
B. Efron. The efficiency of logistic regression compared to normal discriminant analysis. <em>Journal of the American Statistical Association</em>, 70(352): 892–898, 1975.
</div>
<div id="ref-elhakeem2022using" class="csl-entry" role="listitem">
A. Elhakeem, R. A. Hughes, K. Tilling, D. L. Cousminer, S. A. Jackowski, T. J. Cole, A. S. Kwong, Z. Li, S. F. Grant, A. D. Baxter-Jones, et al. Using linear and natural cubic splines, <span>SITAR</span>, and latent trajectory models to characterise nonlinear longitudinal growth trajectories in cohort studies. <em>BMC Medical Research Methodology</em>, 22(1): 68, 2022.
</div>
<div id="ref-erturkzararsiz2023linear" class="csl-entry" role="listitem">
G. Ertürk Zararsız. Linear combination of leukocyte count and <span>D-Dimer</span> levels in the diagnosis of patients with non-traumatic acute abdomen. <em>Med. Rec.</em>, 5: 84–90, 2023. URL <a href="https://doi.org/10.37990/medr.1166531">https://doi.org/10.37990/medr.1166531</a>.
</div>
<div id="ref-faria2016neutrophil" class="csl-entry" role="listitem">
S. S. Faria, P. C. Fernandes Jr, M. J. B. Silva, V. C. Lima, W. Fontes, R. Freitas-Junior, A. K. Eterovic and P. Forget. The neutrophil-to-lymphocyte ratio: A narrative review. <em>ecancermedicalscience</em>, 10: 2016.
</div>
<div id="ref-friedman2010regularization" class="csl-entry" role="listitem">
J. Friedman, T. Hastie and R. Tibshirani. Regularization paths for generalized linear models via coordinate descent. <em>Journal of statistical software</em>, 33(1): 1, 2010.
</div>
<div id="ref-ganapathy2023comparison" class="csl-entry" role="listitem">
S. Ganapathy, H. KT, B. Jindal, P. S. Naik and S. Nair N. Comparison of diagnostic accuracy of models combining the renal biomarkers in predicting renal scarring in pediatric population with vesicoureteral reflux (<span>VUR</span>). <em>Irish Journal of Medical Science (1971-)</em>, 192(5): 2521–2526, 2023.
</div>
<div id="ref-ghosh2005classification" class="csl-entry" role="listitem">
D. Ghosh and A. M. Chinnaiyan. Classification and selection of biomarkers in genomic data using <span>LASSO</span>. <em>BioMed Research International</em>, 2005(2): 147–154, 2005.
</div>
<div id="ref-grau2015prroc" class="csl-entry" role="listitem">
J. Grau, I. Grosse and J. Keilwagen. PRROC: Computing and visualizing precision-recall and receiver operating characteristic curves in <span>R</span>. <em>Bioinformatics</em>, 31(15): 2595–2597, 2015.
</div>
<div id="ref-Trevor2015gam" class="csl-entry" role="listitem">
T. Hastie. Gam: Generalized additive models. 2015. URL <a href="https://CRAN.R-project.org/package=gam">https://CRAN.R-project.org/package=gam</a>. R package version 1.22-5.
</div>
<div id="ref-james2013introduction" class="csl-entry" role="listitem">
G. James, D. Witten, T. Hastie, R. Tibshirani, et al. <em>An introduction to statistical learning.</em> Springer, 2013.
</div>
<div id="ref-kang2016linear" class="csl-entry" role="listitem">
L. Kang, A. Liu and L. Tian. Linear combination methods to improve diagnostic/prognostic accuracy on future observations. <em>Statistical methods in medical research</em>, 25(4): 1359–1380, 2016.
</div>
<div id="ref-kramar2001mroc" class="csl-entry" role="listitem">
A. Kramar, D. Faraggi, A. Fortuné and B. Reiser. mROC: A computer program for combining tumour markers in predicting disease states. <em>Computer methods and programs in biomedicine</em>, 66(2-3): 199–207, 2001.
</div>
<div id="ref-kuhn2008building" class="csl-entry" role="listitem">
M. Kuhn. Building predictive models in <span>R</span> using the caret package. <em>Journal of statistical software</em>, 28: 1–26, 2008.
</div>
<div id="ref-leon2006bedside" class="csl-entry" role="listitem">
C. León, S. Ruiz-Santana, P. Saavedra, B. Almirante, J. Nolla-Salas, F. Álvarez-Lerma, J. Garnacho-Montero, M. Á. León, E. S. Group, et al. A bedside scoring system (<span><span>“Candida score”</span></span>) for early antifungal treatment in nonneutropenic critically ill patients with <span>Candida</span> colonization. <em>Critical care medicine</em>, 34(3): 730–737, 2006.
</div>
<div id="ref-liu2011min" class="csl-entry" role="listitem">
C. Liu, A. Liu and S. Halabi. A min–max combination of biomarkers to improve diagnostic accuracy. <em>Statistics in medicine</em>, 30(16): 2005–2014, 2011.
</div>
<div id="ref-luo2024ast" class="csl-entry" role="listitem">
J. Luo, F. Yu, H. Zhou, X. Wu, Q. Zhou, Q. Liu and S. Gan. <span>AST</span>/<span>ALT</span> ratio is an independent risk factor for diabetic retinopathy: A cross-sectional study. <em>Medicine</em>, 103(26): e38583, 2024.
</div>
<div id="ref-minaev2018distance" class="csl-entry" role="listitem">
G. Minaev, R. Piché and A. Visa. Distance measures for classification of numerical features. 2018. URL <a href="https://trepo.tuni.fi/handle/10024/124353">https://trepo.tuni.fi/handle/10024/124353</a>.
</div>
<div id="ref-muller2019amyloid" class="csl-entry" role="listitem">
E. G. Müller, T. H. Edwin, C. Stokke, S. S. Navelsaker, A. Babovic, N. Bogdanovic, A. B. Knapskog and M. E. Revheim. Amyloid-<span class="math inline">\(\beta\)</span> <span>PET</span>—correlation with cerebrospinal fluid biomarkers and prediction of <span class="nocase">Alzheimer<span>’</span>s</span> disease diagnosis in a memory clinic. <em>PloS one</em>, 14(8): e0221365, 2019.
</div>
<div id="ref-neumann2023combining" class="csl-entry" role="listitem">
M. Neumann, H. Kothare and V. Ramanarayanan. Combining multiple multimodal speech features into an interpretable index score for capturing disease progression in <span>Amyotrophic Lateral Sclerosis</span>. <em>Interspeech</em>, 2353: 2023.
</div>
<div id="ref-nyblom2006ast" class="csl-entry" role="listitem">
H. Nyblom, E. Björnsson, M. Simrén, F. Aldenborg, S. Almer and R. Olsson. The <span>AST</span>/<span>ALT</span> ratio as an indicator of cirrhosis in patients with <span>PBC</span>. <em>Liver International</em>, 26(7): 840–845, 2006.
</div>
<div id="ref-pandit2011comparative" class="csl-entry" role="listitem">
S. Pandit, S. Gupta, et al. A comparative study on distance measuring approaches for clustering. <em>International journal of research in computer science</em>, 2(1): 29–31, 2011.
</div>
<div id="ref-pedregosa2011scikit" class="csl-entry" role="listitem">
F. Pedregosa, G. Varoquaux, A. Gramfort, V. Michel, B. Thirion, O. Grisel, M. Blondel, P. Prettenhofer, R. Weiss, V. Dubourg, et al. Scikit-learn: Machine learning in <span>Python</span>. <em>the Journal of machine Learning research</em>, 12: 2825–2830, 2011.
</div>
<div id="ref-pepe2003statistical" class="csl-entry" role="listitem">
M. S. Pepe. <em>The statistical evaluation of medical tests for classification and prediction.</em> Oxford university press, 2003.
</div>
<div id="ref-pepe2006combining" class="csl-entry" role="listitem">
M. S. Pepe, T. Cai and G. Longton. Combining predictors for classification using the area under the receiver operating characteristic curve. <em>Biometrics</em>, 62(1): 221–229, 2006.
</div>
<div id="ref-pepe2000combining" class="csl-entry" role="listitem">
M. S. Pepe and M. L. Thompson. Combining diagnostic test results to increase accuracy. <em>Biostatistics</em>, 1(2): 123–140, 2000.
</div>
<div id="ref-perez2021visualizing" class="csl-entry" role="listitem">
S. Pérez-Fernández, P. Martı́nez-Camblor, P. Filzmoser and N. Corral. Visualizing the decision rules behind the <span>ROC</span> curves: Understanding the classification process. <em>AStA Advances in Statistical Analysis</em>, 105(1): 135–161, 2021.
</div>
<div id="ref-prinzi2023explainable" class="csl-entry" role="listitem">
F. Prinzi, C. Militello, N. Scichilone, S. Gaglio and S. Vitabile. Explainable machine-learning models for <span>COVID-19</span> prognosis prediction using clinical, laboratory and radiomic features. <em>IEEE Access</em>, 11: 121492–121510, 2023.
</div>
<div id="ref-robin2011proc" class="csl-entry" role="listitem">
X. Robin, N. Turck, A. Hainard, N. Tiberti, F. Lisacek, J.-C. Sanchez and M. Müller. pROC: An open-source package for <span>R</span> and <span>S+</span> to analyze and compare <span>ROC</span> curves. <em>BMC bioinformatics</em>, 12: 1–8, 2011.
</div>
<div id="ref-ruiz1991asymptotic" class="csl-entry" role="listitem">
S. Ruiz-Velasco. Asymptotic efficiency of logistic regression relative to linear discriminant analysis. <em>Biometrika</em>, 78(2): 235–243, 1991.
</div>
<div id="ref-sachs2017plotroc" class="csl-entry" role="listitem">
M. C. Sachs. plotROC: A tool for plotting <span>ROC</span> curves. <em>Journal of statistical software</em>, 79: 2017.
</div>
<div id="ref-salvetat2024ai" class="csl-entry" role="listitem">
N. Salvetat, F. J. Checa-Robles, A. Delacrétaz, C. Cayzac, B. Dubuc, D. Vetter, J. Dainat, J.-P. Lang, F. Gamma and D. Weissmann. AI algorithm combined with RNA editing-based blood biomarkers to discriminate bipolar from major depressive disorders in an external validation multicentric cohort. <em>Journal of Affective Disorders</em>, 356: 385–393, 2024.
</div>
<div id="ref-salvetat2022game" class="csl-entry" role="listitem">
N. Salvetat, F. J. Checa-Robles, V. Patel, C. Cayzac, B. Dubuc, F. Chimienti, J.-D. Abraham, P. Dupré, D. Vetter, S. Méreuze, et al. A game changer for bipolar disorder diagnosis using RNA editing-based biomarkers. <em>Translational Psychiatry</em>, 12(1): 182, 2022.
</div>
<div id="ref-sameera2016binary" class="csl-entry" role="listitem">
G. Sameera, R. V. Vardhan and K. Sarma. Binary classification using multivariate receiver operating characteristic curve for continuous data. <em>Journal of biopharmaceutical statistics</em>, 26(3): 421–431, 2016.
</div>
<div id="ref-serban2024significance" class="csl-entry" role="listitem">
D. Serban, N. Papanas, A. M. Dascalu, P. Kempler, I. Raz, A. A. Rizvi, M. Rizzo, C. Tudor, M. Silviu Tudosie, D. Tanasescu, et al. Significance of neutrophil to lymphocyte ratio (<span>NLR</span>) and platelet lymphocyte ratio (<span>PLR</span>) in diabetic foot ulcer and potential new therapeutic targets. <em>The International Journal of Lower Extremity Wounds</em>, 23(2): 205–216, 2024.
</div>
<div id="ref-sewak2024construction" class="csl-entry" role="listitem">
A. Sewak, S. Siegfried and T. Hothorn. Construction and evaluation of optimal diagnostic tests with application to hepatocellular carcinoma diagnosis. <em>arXiv preprint arXiv:2402.03004</em>, 2024.
</div>
<div id="ref-shiralkarprogramming" class="csl-entry" role="listitem">
P. Shiralkar. Programming validation: Perspectives and strategies. <em>PharmaSUG 2010—paper IB09</em>, 2010.
</div>
<div id="ref-sing2005rocr" class="csl-entry" role="listitem">
T. Sing, O. Sander, N. Beerenwinkel and T. Lengauer. ROCR: Visualizing classifier performance in <span>R</span>. <em>Bioinformatics</em>, 21(20): 3940–3941, 2005.
</div>
<div id="ref-stevenson2017epir" class="csl-entry" role="listitem">
M. Stevenson, T. Nunes, C. Heuer, J. Marshall, J. Sanchez, R. Thornton, J. Reiczigel, J. Robison-Cox, P. Sebastiani, P. Solymos, et al. epiR: Tools for the analysis of epidemiological data. 2017. URL <a href="https://cran.r-project.org/web/packages/epiR/index.html">https://cran.r-project.org/web/packages/epiR/index.html</a>. R package version 2.0.76.
</div>
<div id="ref-su1993linear" class="csl-entry" role="listitem">
J. Q. Su and J. S. Liu. Linear combinations of multiple diagnostic markers. <em>Journal of the American Statistical Association</em>, 88(424): 1350–1355, 1993.
</div>
<div id="ref-svart2024neurofilament" class="csl-entry" role="listitem">
K. Svart, J. J. Korsbæk, R. H. Jensen, T. Parkner, C. S. Knudsen, S. G. Hasselbalch, S. M. Hagen, E. A. Wibroe, L. D. Molander and D. Beier. Neurofilament light chain is elevated in patients with newly diagnosed idiopathic intracranial hypertension: A prospective study. <em>Cephalalgia</em>, 44(5): 03331024241248203, 2024.
</div>
<div id="ref-todor2014tools" class="csl-entry" role="listitem">
N. Todor, I. Todor and G. Săplăcan. Tools to identify linear combination of prognostic factors which maximizes area under receiver operator curve. <em>Journal of clinical bioinformatics</em>, 4: 1–7, 2014.
</div>
<div id="ref-wickham2011testthat" class="csl-entry" role="listitem">
H. Wickham. Testthat: Get started with testing. 2024. URL <a href="https://cran.r-project.org/web/packages/testthat/index.html">https://cran.r-project.org/web/packages/testthat/index.html</a>. R package version 3.2.1.1.
</div>
<div id="ref-wickham2013roxygen2" class="csl-entry" role="listitem">
H. Wickham, P. Danenberg and M. Eugster. roxygen2: In-source documentation for <span>R</span>. 2024. URL <a href="https://cran.r-project.org/web/packages/roxygen2/index.html">https://cran.r-project.org/web/packages/roxygen2/index.html</a>. R package version 7.3.2.
</div>
<div id="ref-wickham2016devtools" class="csl-entry" role="listitem">
H. Wickham, J. Hester, W. Chang and J. Bryan. Devtools: Tools to make developing <span>R</span> packages easier. 2022. URL <a href="https://cran.r-project.org/web/packages/devtools/index.html">https://cran.r-project.org/web/packages/devtools/index.html</a>. R package version 2.4.5.
</div>
<div id="ref-yin2014optimal" class="csl-entry" role="listitem">
J. Yin and L. Tian. Optimal linear combinations of multiple diagnostic biomarkers based on <span>Youden</span> index. <em>Statistics in medicine</em>, 33(8): 1426–1440, 2014.
</div>
<div id="ref-yu2015two" class="csl-entry" role="listitem">
W. Yu and T. Park. Two simple algorithms on linear combination of multiple biomarkers to maximize partial area under the <span>ROC</span> curve. <em>Computational Statistics &amp; Data Analysis</em>, 88: 15–27, 2015.
</div>
<div id="ref-zararsiz2016statistical" class="csl-entry" role="listitem">
G. Zararsiz, H. Y. Akyildiz, D. GÖKSÜLÜK, S. Korkmaz and A. ÖZTÜRK. Statistical learning approaches in diagnosing patients with nontraumatic acute abdomen. <em>Turkish Journal of Electrical Engineering and Computer Sciences</em>, 24(5): 3685–3697, 2016.
</div>
</div>
<section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr />
<ol>
<li id="fn1"><p><a href="https://cran.r-project.org/web/%20packages/splines/index.html" class="uri">https://cran.r-project.org/web/
packages/splines/index.html</a><a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p><span id="note2" data-label="note2"></span><a href="https://cran.r-project.org/web/packages/glmnet/index.html" class="uri">https://cran.r-project.org/web/packages/glmnet/index.html</a><a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p><a href="https://github.com/gokmenzararsiz/dtComb" class="uri">https://github.com/gokmenzararsiz/dtComb</a>,
<a href="https://github.com/gokmenzararsiz/dtComb_Shiny" class="uri">https://github.com/gokmenzararsiz/dtComb_Shiny</a></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></li>
</ol>
</section>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
<h3 id="references">References</h3>
<div id="references-listing"></div>
<h3 id="reuse">Reuse</h3>
<p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
<h3 id="citation">Citation</h3>
<p>For attribution, please cite this work as</p>
<pre class="citation-appendix short">Taştan, et al., "dtComb: A Comprehensive R Library and Web Tool for Combining Diagnostic Tests", The R Journal, 2026</pre>
<p>BibTeX citation</p>
<pre class="citation-appendix long">@article{RJ-2025-036,
  author = {Taştan, S. Ilayda Yerlitaş and Gengeç, Serra Bersan and Koçhan, Necla and Zararsız, Ertürk and Korkmaz, Selçuk and Zararsız, },
  title = {dtComb: A Comprehensive R Library and Web Tool for Combining Diagnostic Tests},
  journal = {The R Journal},
  year = {2026},
  note = {https://doi.org/10.32614/RJ-2025-036},
  doi = {10.32614/RJ-2025-036},
  volume = {17},
  issue = {4},
  issn = {2073-4859},
  pages = {80-102}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!--radix_placeholder_navigation_after_body-->
<!--/radix_placeholder_navigation_after_body-->

</body>

</html>
