---
title: News from the Bioconductor Project
abstract: The 'News from the Bioconductor Project' article from the 2010-1 issue.
author: 
- name:  Bioconductor Team
  affiliation: Program in Computational Biology, Fred Hutchinson Cancer Research Center
date: '2010-06-01'
date_received: ~
journal:
  firstpage: 85
  lastpage: 85
volume: 2
issue: 1
slug: RJ-2010-1-bioconductor
citation_url: https://rjournal.github.io/
packages:
  cran: []
  bioc:
  - GenomicRanges
  - Rsamtools
  - girafe
  - BayesPeak
  - CSAR
  - PICS
  - DESeq
  - goseq
  - segmentSeq
  - MotIV
  - rGADEM
  - affyILM
  - frma
  - frmaTools
  - BeadDataPackR
  - MassArray
  - charm
  - genoCN
  - iChip
  - methVisual
  - ConsensusClusterPlus
  - ExpressionView
  - eisa
  - GSRI
  - PROMISE
  - tigre
  - SamSPECTRAL
  - flowMeans
  - flowTrans
  - iFlow
  - GEOsubmission
  - SRAdb
  - genomes
  - pint
  - xmapcore
preview: preview.png
bibliography: ~
CTV: ~
output:
  rjtools::rjournal_web_article:
    self_contained: yes
    toc: no
    legacy_pdf: yes
    web_only: yes
    mathjax: https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js
    md_extension: -tex_math_single_backslash

---

::: article
We are pleased to announce Bioconductor 2.6, released on April 23, 2010.
Bioconductor 1.6 is compatible with R 2.11.1, and consists of 389
packages. There are 37 new packages, and enhancements to many others.
Explore Bioconductor at <http://bioconductor.org>, and install packages
with

``` r
source("http://bioconductor.org/biocLite.R")
biocLite() # install standard packages...
biocLite("IRanges") # ...or IRanges
```

# New and revised packages

Sequence analysis

:   packages address infrastructure
    ([*GenomicRanges*](https://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html),
    [*Rsamtools*](https://www.bioconductor.org/packages/release/bioc/html/Rsamtools.html),
    [*girafe*](https://www.bioconductor.org/packages/release/bioc/html/girafe.html));
    ChIP-seq
    ([*BayesPeak*](https://www.bioconductor.org/packages/release/bioc/html/BayesPeak.html),
    [*CSAR*](https://www.bioconductor.org/packages/release/bioc/html/CSAR.html),
    [*PICS*](https://www.bioconductor.org/packages/release/bioc/html/PICS.html));
    digital gene expression and RNA-seq
    ([*DESeq*](https://www.bioconductor.org/packages/release/bioc/html/DESeq.html),
    [*goseq*](https://www.bioconductor.org/packages/release/bioc/html/goseq.html),
    [*segmentSeq*](https://www.bioconductor.org/packages/release/bioc/html/segmentSeq.html));
    and motif discovery
    ([*MotIV*](https://www.bioconductor.org/packages/release/bioc/html/MotIV.html),
    [*rGADEM*](https://www.bioconductor.org/packages/release/bioc/html/rGADEM.html)).

Microarray analysis

:   packages introduce new approaches to pre-process and
    technology-specific assays
    ([*affyILM*](https://www.bioconductor.org/packages/release/bioc/html/affyILM.html),
    [*frma*](https://www.bioconductor.org/packages/release/bioc/html/frma.html),
    [*frmaTools*](https://www.bioconductor.org/packages/release/bioc/html/frmaTools.html),
    [*BeadDataPackR*](https://www.bioconductor.org/packages/release/bioc/html/BeadDataPackR.html),
    [*MassArray*](https://www.bioconductor.org/packages/release/bioc/html/MassArray.html));
    analysis of specific experimental protocols
    ([*charm*](https://www.bioconductor.org/packages/release/bioc/html/charm.html),
    [*genoCN*](https://www.bioconductor.org/packages/release/bioc/html/genoCN.html),
    [*iChip*](https://www.bioconductor.org/packages/release/bioc/html/iChip.html),
    [*methVisual*](https://www.bioconductor.org/packages/release/bioc/html/methVisual.html));
    and novel statistical methods
    ([*ConsensusClusterPlus*](https://www.bioconductor.org/packages/release/bioc/html/ConsensusClusterPlus.html),
    [*ExpressionView*](https://www.bioconductor.org/packages/release/bioc/html/ExpressionView.html),
    [*eisa*](https://www.bioconductor.org/packages/release/bioc/html/eisa.html),
    [*GSRI*](https://www.bioconductor.org/packages/release/bioc/html/GSRI.html),
    [*PROMISE*](https://www.bioconductor.org/packages/release/bioc/html/PROMISE.html),
    [*tigre*](https://www.bioconductor.org/packages/release/bioc/html/tigre.html)).

Flow cytometry

:   packages include
    [*SamSPECTRAL*](https://www.bioconductor.org/packages/release/bioc/html/SamSPECTRAL.html),
    [*flowMeans*](https://www.bioconductor.org/packages/release/bioc/html/flowMeans.html),
    [*flowTrans*](https://www.bioconductor.org/packages/release/bioc/html/flowTrans.html),
    and
    [*iFlow*](https://www.bioconductor.org/packages/release/bioc/html/iFlow.html).

Annotation and integrative analysis

:   packages facilitate interfacing with GEO
    ([*GEOsubmission*](https://www.bioconductor.org/packages/release/bioc/html/GEOsubmission.html)),
    the Sequence Read Archive
    ([*SRAdb*](https://www.bioconductor.org/packages/release/bioc/html/SRAdb.html)),
    and tabulation of genome sequence project data
    ([*genomes*](https://www.bioconductor.org/packages/release/bioc/html/genomes.html));
    the
    [*GSRI*](https://www.bioconductor.org/packages/release/bioc/html/GSRI.html)
    package to estimate differentially expressed genes in a gene set;
    PCA and CCA dependency modeling
    ([*pint*](https://www.bioconductor.org/packages/release/bioc/html/pint.html));
    and updated access to exon array annotations
    ([*xmapcore*](https://www.bioconductor.org/packages/release/bioc/html/xmapcore.html)).

Further information on new and existing packages can be found on the
Bioconductor web site, which contains 'views' that identify coherent
groups of packages. The views link to on-line package descriptions,
vignettes, reference manuals, and use statistics.

# Other activities

Training courses and a European developer conference were important
aspects of Bioconductor, with recent successful offerings ranging from
microarray and sequence analysis through advanced R programming
(<http://bioconductor.org/workshops>). The active Bioconductor mailing
lists (<http://bioconductor.org/docs/mailList.html>) connect users with
each other, to domain experts, and to maintainers eager to ensure that
their packages satisfy the needs of leading edge approaches. The
Bioconductor team continues to ensure quality software through technical
and scientific reviews of new packages, and daily builds of released
packages on Linux, Windows (32- and now 64-bit), and Macintosh
platforms.

# Looking forward

*BioC2010* is the Bioconductor annual conference, scheduled for July
29--30 at the Fred Hutchinson Cancer Research Centre in Seattle,
Washington, and with a developer day on July 28. The conference features
morning scientific talks and afternoon practical sessions
(<https://secure.bioconductor.org/BioC2010/>). European developers will
have an opportunity to meet in Heidelberg, Germany, on 17--18 November.

Contributions from the Bioconductor community play an important role in
shaping each release. We anticipate continued efforts to provide
statistically informed analysis of next generation sequence data, as
well as development of algorithms for advanced analysis of microarrays.
Areas of opportunity include the ChIP-seq, RNA-seq, rare variant, and
structural variant domains. The next release cycle promises to be one of
active scientific growth and exploration!
:::
