---
title: News from the Bioconductor Project
abstract: The 'News from the Bioconductor Project' article from the 2016-1 issue.
author:
- name: Bioconductor Core Team
  affiliation: Biostatistics and Bioinformatics, Roswell Park Cancer Institute
  address:
  - Buffalo, NY
  - |
    USA
date: '2016-06-01'
date_received: ~
journal:
  firstpage: '404'
  lastpage: '405'
volume: 8
issue: 1
slug: RJ-2016-1-bioconductor
citation_url: https://rjournal.github.io/
packages:
  cran: []
  bioc:
  - AnnotationHub
  - BiocCheck
  - GenomicRanges
preview: preview.png
bibliography: ~
CTV: ~
output:
  rjtools::rjournal_web_article:
    self_contained: yes
    toc: no
    legacy_pdf: yes
    web_only: yes
    mathjax: https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js
    md_extension: -tex_math_single_backslash

---

::: article
The [Bioconductor](http://bioconductor.org) project provides tools for
the analysis and comprehension of high-throughput genomic data. The 1211
software packages available in Bioconductor can be viewed at
<http://bioconductor.org/packages/>. Navigate packages using 'biocViews'
terms and title search. Each package has an html page with a
description, links to vignettes, reference manuals, and usage
statistics. Start using Bioconductor version 3.3 by installing R 3.3.1
and evaluating the commands

``` r
  source("https://bioconductor.org/biocLite.R")
  biocLite()
```

Install additional packages and dependencies, e.g.,
[*AnnotationHub*](https://www.bioconductor.org/packages/release/bioc/html/AnnotationHub.html),
with

``` r
  source("https://bioconductor.org/biocLite.R")
  biocLite("AnnotationHub")
```

Continued availability of Bioconductor
[Docker](http://www.bioconductor.org/help/docker/) and
[Amazon](http://www.bioconductor.org/help/bioconductor-cloud-ami/)
images provides a very effective on-ramp for power users to rapidly
obtain access to standardized and scalable computing environments.

# Bioconductor 3.3 Release Highlights

Bioconductor 3.3 was released on 4 April, 2016. It is compatible with R
3.3 and consists of 1211 software packages, 293 experiment data
packages, and 916 up-to-date annotation packages. There are 107 new
software packages and many updates and improvements to existing
packages. The [release
announcement](http://bioconductor.org/news/bioc_3_3_release/) includes
descriptions of new packages and updated NEWS files provided by package
maintainers.

Our collection of microarray, transcriptome and organism-specific
*annotation packages* use the 'select' interface (keys, columns,
keytypes) to access static information on gene annotations (*org.\**
packages) and gene models (*TxDb.\** packages); these augment packages
for querying web-based resources. The
[*AnnotationHub*](https://www.bioconductor.org/packages/release/bioc/html/AnnotationHub.html)
continues to complement our traditional offerings with diverse whole
genome annotations from Ensembl, ENCODE, dbSNP, UCSC, and elsewhere;
example uses are described in the [AnnotationHub
How-To](http://bioconductor.org/packages/devel/bioc/vignettes/AnnotationHub/inst/doc/AnnotationHub-HOWTO.html)
vignette.

# User support

The Bioconductor [project web site](https://bioconductor.org) helps
orient users and developers to the project. It includes essential
information for software
[installation](http://bioconductor.org/install/), detailed landing pages
for each package (e.g.,
<https://bioconductor.org/packages/GenomicRanges>) including links to
current manuals and vignettes, extensive [training
material](http://bioconductor.org/help/course-materials/), and links to
the current [literature](http://bioconductor.org/help/publications/). A
recent innovation has been the development of the [Bioconductor F1000
publishing channel](http://f1000research.com/channels/bioconductor) for
academic publication of work flows and other extended software use
cases.

The project [support site](https://support.bioconductor.org) is a
question-and-answer forum where users can easily search for existing
solutions or pose specific questions about use of Bioconductor packages.
The support site is quite active, with expert responses often within a
matter of hours. It is very helpful, when asking about error messages,
to ensure that your Bioconductor installation is correct (using
`BiocInstaller::biocValid()`) and current (include the output of
`sessionInfo()` in your question), that the question includes code
chunks that someone else can evaluate to reproduce the problem (e.g.,
using code or data from example pages of package manuals), and that the
error message and `traceback()` output are included.

Bioconductor holds an annual user conference each summer, this year in
conjunction with UseR!Â 2016. [Conference
resources](http://bioconductor.org/help/course-materials/2016/BioC2016/)
(talks and workshops) are available.

# Developer support

A very natural progression in the R and Bioconductor community is from
user to package developer, transforming your knowledge and domain
expertise into software that others can use. The Bioconductor web site
includes [developer resources](http://bioconductor.org/developers/) to
help this transition. The Bioconductor [developer mailing
list](https://stat.ethz.ch/mailman/listinfo/bioc-devel) provides a forum
dedicated to developer-related questions.

New packages are now submitted to Bioconductor using an open review
model. Prospective authors develop their package and, when ready, open
an issue on the public
[Contributions](https://github.com/Bioconductor/Contributions) github
repository. Packages are then built and checked across Linux, Mac, and
Windows platforms for conformance to R (`R CMD check`) and Bioconductor
(using the
[*BiocCheck*](https://www.bioconductor.org/packages/release/bioc/html/BiocCheck.html)
package) standards. Once the package is in good shape, a member of the
Bioconductor core team performs a preview of the package. The preview
identifies technical issues that are not easy to detect automatically.

A key strength of the Bioconductor project is the use of well-defined
objects (especially from the
[*GenomicRanges*](https://www.bioconductor.org/packages/release/bioc/html/GenomicRanges.html)
infrastructure) to represent data; this encourages software re-use and
enables end-user interoperability between packages. For this reason, the
technical review often leads to suggestions for data representations and
interfaces that use Bioconductor objects rather than general-purpose
containers such as a `data.frame`.

# Forthcoming activities

Forthcoming Bioconductor [events](http://bioconductor.org/help/events/)
include an Asian workshop
[workshop](http://www.abacbs.org/biocasia2016workshop) and
[developer](http://www.abacbs.org/biocasia2016) meeting (3-4 November,
Brisbane, Australia) and European
[developer](http://scicore.ch/events/eurobioc2016/) conference (6-7
December, Basel, Switzerland) developer conferences, as well as global
training opportunities.

The next Bioconductor release will occur in October, 2016.
:::
