<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">

<head>
  <meta charset="utf-8"/>
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta http-equiv="X-UA-Compatible" content="IE=Edge,chrome=1"/>
  <meta name="generator" content="distill" />

  <style type="text/css">
  /* Hide doc at startup (prevent jankiness while JS renders/transforms) */
  body {
    visibility: hidden;
  }
  </style>

 <!--radix_placeholder_import_source-->
 <!--/radix_placeholder_import_source-->

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { color: #00769e; background-color: #f1f3f5; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span { color: #00769e; } /* Normal */
code span.al { color: #ad0000; } /* Alert */
code span.an { color: #5e5e5e; } /* Annotation */
code span.at { color: #657422; } /* Attribute */
code span.bn { color: #ad0000; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #00769e; } /* ControlFlow */
code span.ch { color: #20794d; } /* Char */
code span.cn { color: #8f5902; } /* Constant */
code span.co { color: #5e5e5e; } /* Comment */
code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
code span.dt { color: #ad0000; } /* DataType */
code span.dv { color: #ad0000; } /* DecVal */
code span.er { color: #ad0000; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #ad0000; } /* Float */
code span.fu { color: #4758ab; } /* Function */
code span.im { } /* Import */
code span.in { color: #5e5e5e; } /* Information */
code span.kw { color: #00769e; } /* Keyword */
code span.op { color: #5e5e5e; } /* Operator */
code span.ot { color: #00769e; } /* Other */
code span.pp { color: #ad0000; } /* Preprocessor */
code span.sc { color: #5e5e5e; } /* SpecialChar */
code span.ss { color: #20794d; } /* SpecialString */
code span.st { color: #20794d; } /* String */
code span.va { color: #111111; } /* Variable */
code span.vs { color: #20794d; } /* VerbatimString */
code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
</style>

<style>
  div.csl-bib-body { }
  div.csl-entry {
    clear: both;
      margin-bottom: 0em;
    }
  .hanging div.csl-entry {
    margin-left:2em;
    text-indent:-2em;
  }
  div.csl-left-margin {
    min-width:2em;
    float:left;
  }
  div.csl-right-inline {
    margin-left:2em;
    padding-left:1em;
  }
  div.csl-indent {
    margin-left: 2em;
  }
</style>

  <!--radix_placeholder_meta_tags-->
<title>Partitioned Local Depth (PaLD) Community Analyses in R</title>

<meta property="description" itemprop="description" content="Partitioned Local Depth (PaLD) is a framework for holistic&#10;consideration of community structure for distance-based data. This&#10;paper describes an R package, pald, for calculating&#10;Partitioned Local Depth (PaLD) probabilities, implementing community&#10;analyses, determining community clusters, and creating data&#10;visualizations to display community structure. We present essentials&#10;of the PaLD approach, describe how to use the pald package, walk&#10;through several examples, and discuss the method in relation to&#10;commonly used techniques."/>

<link rel="canonical" href="https://doi.org/10.32614/RJ-2025-013/"/>
<link rel="license" href="https://creativecommons.org/licenses/by/4.0/"/>
<link rel="icon" type="image/vnd.microsoft.icon" href="../../resources/favicon.ico"/>

<!--  https://schema.org/Article -->
<meta property="article:published" itemprop="datePublished" content="2025-09-24"/>
<meta property="article:created" itemprop="dateCreated" content="2025-09-24"/>
<meta name="article:author" content="Lucy D&#39;Agostino McGowan"/>
<meta name="article:author" content="Katherine Moore"/>
<meta name="article:author" content="Kenneth S. Berenhaut"/>

<!--  https://developers.facebook.com/docs/sharing/webmasters#markup -->
<meta property="og:title" content="Partitioned Local Depth (PaLD) Community Analyses in R"/>
<meta property="og:type" content="article"/>
<meta property="og:description" content="Partitioned Local Depth (PaLD) is a framework for holistic&#10;consideration of community structure for distance-based data. This&#10;paper describes an R package, pald, for calculating&#10;Partitioned Local Depth (PaLD) probabilities, implementing community&#10;analyses, determining community clusters, and creating data&#10;visualizations to display community structure. We present essentials&#10;of the PaLD approach, describe how to use the pald package, walk&#10;through several examples, and discuss the method in relation to&#10;commonly used techniques."/>
<meta property="og:url" content="https://doi.org/10.32614/RJ-2025-013/"/>
<meta property="og:locale" content="en_US"/>
<meta property="og:site_name" content="The R Journal"/>

<!--  https://dev.twitter.com/cards/types/summary -->
<meta property="twitter:card" content="summary"/>
<meta property="twitter:title" content="Partitioned Local Depth (PaLD) Community Analyses in R"/>
<meta property="twitter:description" content="Partitioned Local Depth (PaLD) is a framework for holistic&#10;consideration of community structure for distance-based data. This&#10;paper describes an R package, pald, for calculating&#10;Partitioned Local Depth (PaLD) probabilities, implementing community&#10;analyses, determining community clusters, and creating data&#10;visualizations to display community structure. We present essentials&#10;of the PaLD approach, describe how to use the pald package, walk&#10;through several examples, and discuss the method in relation to&#10;commonly used techniques."/>
<meta property="twitter:url" content="https://doi.org/10.32614/RJ-2025-013/"/>

<!--  https://scholar.google.com/intl/en/scholar/inclusion.html#indexing -->
<meta name="citation_title" content="Partitioned Local Depth (PaLD) Community Analyses in R"/>
<meta name="citation_fulltext_html_url" content="https://doi.org/10.32614/RJ-2025-013"/>
<meta name="citation_pdf_url" content="RJ-2025-013.pdf"/>
<meta name="citation_volume" content="17"/>
<meta name="citation_issue" content="2"/>
<meta name="citation_doi" content="10.32614/RJ-2025-013"/>
<meta name="citation_journal_title" content="The R Journal"/>
<meta name="citation_issn" content="2073-4859"/>
<meta name="citation_firstpage" content="35"/>
<meta name="citation_lastpage" content="55"/>
<meta name="citation_fulltext_world_readable" content=""/>
<meta name="citation_online_date" content="2025/09/24"/>
<meta name="citation_publication_date" content="2025/09/24"/>
<meta name="citation_author" content="Lucy D&#39;Agostino McGowan"/>
<meta name="citation_author_institution" content="Wake Forest University"/>
<meta name="citation_author" content="Katherine Moore"/>
<meta name="citation_author_institution" content="Amherst College"/>
<meta name="citation_author" content="Kenneth S. Berenhaut"/>
<meta name="citation_author_institution" content="Wake Forest University"/>
<!--/radix_placeholder_meta_tags-->
  
  <meta name="citation_reference" content="citation_title=Local depth;citation_publisher=Elsevier;citation_volume=141;citation_author=Claudio Agostinelli;citation_author=Mario Romanazzi"/>
  <meta name="citation_reference" content="citation_title=Robust data clustering;citation_publisher=IEEE;citation_volume=2;citation_author=LN Fred Ana;citation_author=Anil K Jain"/>
  <meta name="citation_reference" content="citation_title=Partitioned k-nearest neighbor local depth for scalable comparison-based learning;citation_author=Jacob D Baron;citation_author=RWR Darling;citation_author=J Laylon Davis;citation_author=R Pettit"/>
  <meta name="citation_reference" content="citation_title=Culture rather than genes provides greater scope for the evolution of large-scale human prosociality;citation_publisher=National Acad Sciences;citation_volume=106;citation_author=Adrian V Bell;citation_author=Peter J Richerson;citation_author=Richard McElreath"/>
  <meta name="citation_reference" content="citation_title=Communities in data;citation_author=Kenneth S Berenhaut;citation_author=Katherine E Moore"/>
  <meta name="citation_reference" content="citation_title=A social perspective on perceived distances reveals deep community structure;citation_publisher=National Acad Sciences;citation_volume=119;citation_author=Kenneth S Berenhaut;citation_author=Katherine E Moore;citation_author=Ryan L Melvin"/>
  <meta name="citation_reference" content="citation_title=Generalized partitioned local depth;citation_publisher=Springer;citation_volume=18;citation_author=Kenneth S Berenhaut;citation_author=John D Foley;citation_author=Liangdongsheng Lyu"/>
  <meta name="citation_reference" content="citation_title=Density-based clustering;citation_publisher=Wiley Online Library;citation_volume=10;citation_author=Ricardo JGB Campello;citation_author=Peer Kröger;citation_author=Jörg Sander;citation_author=Arthur Zimek"/>
  <meta name="citation_reference" content="citation_title=The history and geography of human genes;citation_publisher=Princeton university press;citation_author=Luigi Luca Cavalli-Sforza;citation_author=Luca Cavalli-Sforza;citation_author=Paolo Menozzi;citation_author=Alberto Piazza"/>
  <meta name="citation_reference" content="citation_title=Sequential and shared-memory parallel algorithms for partitioned local depths;citation_author=Aditya Devarakonda;citation_author=Grey Ballard"/>
  <meta name="citation_reference" content="citation_title=An Indoeuropean classification: A lexicostatistical experiment;citation_volume=82;citation_doi=10.2307/1006517;citation_author=Isidore Dyen;citation_author=Joseph B. Kruskal;citation_author=Paul Black"/>
  <meta name="citation_reference" content="citation_title=The focused organization of social ties;citation_publisher=University of Chicago Press;citation_volume=86;citation_author=Scott L Feld"/>
  <meta name="citation_reference" content="citation_title=Clustering aggregation;citation_volume=1;citation_author=A Gionis;citation_author=H Mannila;citation_author=P Tsaparas"/>
  <meta name="citation_reference" content="citation_title=The strength of weak ties;citation_publisher=University of Chicago Press;citation_volume=78;citation_author=Mark S Granovetter"/>
  <meta name="citation_reference" content="citation_title=Dbscan: Fast density-based clustering with R;citation_volume=91;citation_author=Michael Hahsler;citation_author=Matthew Piekenbrock;citation_author=Derek Doran"/>
  <meta name="citation_reference" content="citation_title=World values survey: All rounds–country-pooled datafile 1981-2014;citation_publisher=JD Systems Institute;citation_author=Ronald Inglehart;citation_author=Christian Haerpfer;citation_author=Alejandro Moreno;citation_author=Christian Welzel;citation_author=Kseniya Kizilova;citation_author=Jaime Dı́ez-Medrano;citation_author=Marta Lagos;citation_author=Pippa Norris;citation_author=Eduard Ponarin;citation_author=Bi Puranen;citation_author=World values survey: All rounds–country-pooled datafile 1981-2014"/>
  <meta name="citation_reference" content="citation_title=The gene expression barcode: Leveraging public data repositories to begin cataloging the human and murine transcriptomes;citation_publisher=Oxford University Press;citation_volume=39;citation_author=Matthew N McCall;citation_author=Karan Uppal;citation_author=Harris A Jaffee;citation_author=Michael J Zilliox;citation_author=Rafael A Irizarry"/>
  <meta name="citation_reference" content="citation_title=The gene expression barcode 3.0: Improved data processing and mining tools;citation_publisher=Oxford University Press;citation_volume=42;citation_author=Matthew N McCall;citation_author=Harris A Jaffee;citation_author=Susan J Zelisko;citation_author=Neeraj Sinha;citation_author=Guido Hooiveld;citation_author=Rafael A Irizarry;citation_author=Michael J Zilliox"/>
  <meta name="citation_reference" content="citation_title=Beyond Western, Educated, Industrial, Rich, and Democratic (WEIRD) psychology: Measuring and mapping scales of cultural and psychological distance;citation_publisher=Sage Publications Sage CA: Los Angeles, CA;citation_volume=31;citation_author=Michael Muthukrishna;citation_author=Adrian V Bell;citation_author=Joseph Henrich;citation_author=Cameron M Curtin;citation_author=Alexander Gedranovich;citation_author=Jason McInerney;citation_author=Braden Thue"/>
  <meta name="citation_reference" content="citation_title=From depth to local depth: A focus on centrality;citation_publisher=Taylor &amp; Francis;citation_volume=108;citation_author=Davy Paindaveine;citation_author=Germain Van Bever"/>
  <meta name="citation_reference" content="citation_title=Scikit-learn: Machine learning in Python;citation_volume=12;citation_author=Fabian Pedregosa;citation_author=Gaël Varoquaux;citation_author=Alexandre Gramfort;citation_author=Vincent Michel;citation_author=Bertrand Thirion;citation_author=Olivier Grisel;citation_author=Mathieu Blondel;citation_author=Peter Prettenhofer;citation_author=Ron Weiss;citation_author=Vincent Dubourg;citation_author=Scikit-learn: Machine learning in Python"/>
  <meta name="citation_reference" content="citation_title=tissuesGeneExpression. Subset of gene expression data;citation_author=Michael Love;citation_author=Rafael Irizarry"/>
  <meta name="citation_reference" content="citation_title=A gene expression bar code for microarray data;citation_publisher=Nature Publishing Group;citation_volume=4;citation_author=Michael J Zilliox;citation_author=Rafael A Irizarry"/>
  <!--radix_placeholder_rmarkdown_metadata-->

<script type="text/json" id="radix-rmarkdown-metadata">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","description","author","date","date_received","journal","volume","issue","slug","packages","preview","bibliography","CTV","legacy_pdf","legacy_converted","output-width","output","draft","pdf_url","citation_url","doi","creative_commons","csl","canonical_url"]}},"value":[{"type":"character","attributes":{},"value":["Partitioned Local Depth (PaLD) Community Analyses in R"]},{"type":"character","attributes":{},"value":["Partitioned Local Depth (PaLD) is a framework for holistic\nconsideration of community structure for distance-based data. This\npaper describes an R package, pald, for calculating\nPartitioned Local Depth (PaLD) probabilities, implementing community\nanalyses, determining community clusters, and creating data\nvisualizations to display community structure. We present essentials\nof the PaLD approach, describe how to use the pald package, walk\nthrough several examples, and discuss the method in relation to\ncommonly used techniques."]},{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Lucy D'Agostino McGowan"]},{"type":"character","attributes":{},"value":["Wake Forest University"]},{"type":"character","attributes":{},"value":["Winston-Salem, NC","27106","[`mcgowald@wfu.edu`](mailto:mcgowald@wfu.edu)\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Katherine Moore"]},{"type":"character","attributes":{},"value":["Amherst College"]},{"type":"character","attributes":{},"value":["Amherst, MA","1002","[`kmoore@amherst.edu`](mailto:kmoore@amherst.edu)\n"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","affiliation","address"]}},"value":[{"type":"character","attributes":{},"value":["Kenneth S. Berenhaut"]},{"type":"character","attributes":{},"value":["Wake Forest University"]},{"type":"character","attributes":{},"value":["Winston-Salem, NC","27106","[`berenhks@wfu.edu`](mailto:berenhks@wfu.edu)\n"]}]}]},{"type":"character","attributes":{},"value":["2025-09-24"]},{"type":"character","attributes":{},"value":["2024-01-11"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["title","issn","firstpage","lastpage"]}},"value":[{"type":"character","attributes":{},"value":["The R Journal"]},{"type":"character","attributes":{},"value":["2073-4859"]},{"type":"integer","attributes":{},"value":[35]},{"type":"integer","attributes":{},"value":[55]}]},{"type":"integer","attributes":{},"value":[17]},{"type":"integer","attributes":{},"value":[2]},{"type":"character","attributes":{},"value":["RJ-2025-013"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["cran","bioc"]}},"value":[{"type":"character","attributes":{},"value":["pald","igraph","lintools","ggplot2","dbscan"]},{"type":"list","attributes":{},"value":[]}]},{"type":"character","attributes":{},"value":["preview.png"]},{"type":"character","attributes":{},"value":["RJreferences.bib"]},{"type":"character","attributes":{},"value":["ChemPhys","Cluster","GraphicalModels","NetworkAnalysis","Optimization","Phylogenetics","Spatial","TeachingStatistics"]},{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["100%"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["distill::distill_article"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["self_contained","toc","mathjax","md_extension"]}},"value":[{"type":"logical","attributes":{},"value":[true]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js"]},{"type":"character","attributes":{},"value":["-tex_math_single_backslash"]}]}]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["RJ-2025-013.pdf"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2025-013"]},{"type":"character","attributes":{},"value":["10.32614/RJ-2025-013"]},{"type":"character","attributes":{},"value":["CC BY"]},{"type":"character","attributes":{},"value":["/home/mitchell/R/x86_64-pc-linux-gnu-library/4.5/rjtools/rjournal.csl"]},{"type":"character","attributes":{},"value":["https://doi.org/10.32614/RJ-2025-013/"]}]}
</script>
<!--/radix_placeholder_rmarkdown_metadata-->
  
  <script type="text/json" id="radix-resource-manifest">
  {"type":"character","attributes":{},"value":["dagostino-mcgowan_files/figure-latex/fig1-1.pdf","dagostino-mcgowan_files/figure-latex/fig1-1.png","dagostino-mcgowan_files/figure-latex/fig2-1.pdf","dagostino-mcgowan_files/figure-latex/fig2-1.png","dagostino-mcgowan_files/figure-latex/fig3-1.pdf","dagostino-mcgowan_files/figure-latex/fig3-1.png","dagostino-mcgowan_files/figure-latex/fig5-1.pdf","dagostino-mcgowan_files/figure-latex/fig5-1.png","dagostino-mcgowan_files/figure-latex/figco-1.pdf","dagostino-mcgowan_files/figure-latex/figco-1.png","dagostino-mcgowan.pdf","dagostino-mcgowan.tex","fig5.png","fig6.png","fig7.png","pald-1.jpg","pald-2.pdf","pald-2.png","RJ-2025-013_files/anchor-4.2.2/anchor.min.js","RJ-2025-013_files/bowser-1.9.3/bowser.min.js","RJ-2025-013_files/distill-2.2.21/template.v2.js","RJ-2025-013_files/header-attrs-2.29/header-attrs.js","RJ-2025-013_files/jquery-3.6.0/jquery-3.6.0.js","RJ-2025-013_files/jquery-3.6.0/jquery-3.6.0.min.js","RJ-2025-013_files/jquery-3.6.0/jquery-3.6.0.min.map","RJ-2025-013_files/popper-2.6.0/popper.min.js","RJ-2025-013_files/tippy-6.2.7/tippy-bundle.umd.min.js","RJ-2025-013_files/tippy-6.2.7/tippy-light-border.css","RJ-2025-013_files/tippy-6.2.7/tippy.css","RJ-2025-013_files/tippy-6.2.7/tippy.umd.min.js","RJ-2025-013_files/webcomponents-2.0.0/webcomponents.js","RJ-2025-013.pdf","RJournal.sty","RJreferences.bib","RJwrapper.log","RJwrapper.tex"]}
  </script>
  <!--radix_placeholder_navigation_in_header-->
<meta name="distill:offset" content="../.."/>

<script type="application/javascript">

  window.headroom_prevent_pin = false;

  window.document.addEventListener("DOMContentLoaded", function (event) {

    // initialize headroom for banner
    var header = $('header').get(0);
    var headerHeight = header.offsetHeight;
    var headroom = new Headroom(header, {
      tolerance: 5,
      onPin : function() {
        if (window.headroom_prevent_pin) {
          window.headroom_prevent_pin = false;
          headroom.unpin();
        }
      }
    });
    headroom.init();
    if(window.location.hash)
      headroom.unpin();
    $(header).addClass('headroom--transition');

    // offset scroll location for banner on hash change
    // (see: https://github.com/WickyNilliams/headroom.js/issues/38)
    window.addEventListener("hashchange", function(event) {
      window.scrollTo(0, window.pageYOffset - (headerHeight + 25));
    });

    // responsive menu
    $('.distill-site-header').each(function(i, val) {
      var topnav = $(this);
      var toggle = topnav.find('.nav-toggle');
      toggle.on('click', function() {
        topnav.toggleClass('responsive');
      });
    });

    // nav dropdowns
    $('.nav-dropbtn').click(function(e) {
      $(this).next('.nav-dropdown-content').toggleClass('nav-dropdown-active');
      $(this).parent().siblings('.nav-dropdown')
         .children('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $("body").click(function(e){
      $('.nav-dropdown-content').removeClass('nav-dropdown-active');
    });
    $(".nav-dropdown").click(function(e){
      e.stopPropagation();
    });
  });
</script>

<style type="text/css">

/* Theme (user-documented overrideables for nav appearance) */

.distill-site-nav {
  color: rgba(255, 255, 255, 0.8);
  background-color: #0F2E3D;
  font-size: 15px;
  font-weight: 300;
}

.distill-site-nav a {
  color: inherit;
  text-decoration: none;
}

.distill-site-nav a:hover {
  color: white;
}

@media print {
  .distill-site-nav {
    display: none;
  }
}

.distill-site-header {

}

.distill-site-footer {

}


/* Site Header */

.distill-site-header {
  width: 100%;
  box-sizing: border-box;
  z-index: 3;
}

.distill-site-header .nav-left {
  display: inline-block;
  margin-left: 8px;
}

@media screen and (max-width: 768px) {
  .distill-site-header .nav-left {
    margin-left: 0;
  }
}


.distill-site-header .nav-right {
  float: right;
  margin-right: 8px;
}

.distill-site-header a,
.distill-site-header .title {
  display: inline-block;
  text-align: center;
  padding: 14px 10px 14px 10px;
}

.distill-site-header .title {
  font-size: 18px;
  min-width: 150px;
}

.distill-site-header .logo {
  padding: 0;
}

.distill-site-header .logo img {
  display: none;
  max-height: 20px;
  width: auto;
  margin-bottom: -4px;
}

.distill-site-header .nav-image img {
  max-height: 18px;
  width: auto;
  display: inline-block;
  margin-bottom: -3px;
}



@media screen and (min-width: 1000px) {
  .distill-site-header .logo img {
    display: inline-block;
  }
  .distill-site-header .nav-left {
    margin-left: 20px;
  }
  .distill-site-header .nav-right {
    margin-right: 20px;
  }
  .distill-site-header .title {
    padding-left: 12px;
  }
}


.distill-site-header .nav-toggle {
  display: none;
}

.nav-dropdown {
  display: inline-block;
  position: relative;
}

.nav-dropdown .nav-dropbtn {
  border: none;
  outline: none;
  color: rgba(255, 255, 255, 0.8);
  padding: 16px 10px;
  background-color: transparent;
  font-family: inherit;
  font-size: inherit;
  font-weight: inherit;
  margin: 0;
  margin-top: 1px;
  z-index: 2;
}

.nav-dropdown-content {
  display: none;
  position: absolute;
  background-color: white;
  min-width: 200px;
  border: 1px solid rgba(0,0,0,0.15);
  border-radius: 4px;
  box-shadow: 0px 8px 16px 0px rgba(0,0,0,0.1);
  z-index: 1;
  margin-top: 2px;
  white-space: nowrap;
  padding-top: 4px;
  padding-bottom: 4px;
}

.nav-dropdown-content hr {
  margin-top: 4px;
  margin-bottom: 4px;
  border: none;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.nav-dropdown-active {
  display: block;
}

.nav-dropdown-content a, .nav-dropdown-content .nav-dropdown-header {
  color: black;
  padding: 6px 24px;
  text-decoration: none;
  display: block;
  text-align: left;
}

.nav-dropdown-content .nav-dropdown-header {
  display: block;
  padding: 5px 24px;
  padding-bottom: 0;
  text-transform: uppercase;
  font-size: 14px;
  color: #999999;
  white-space: nowrap;
}

.nav-dropdown:hover .nav-dropbtn {
  color: white;
}

.nav-dropdown-content a:hover {
  background-color: #ddd;
  color: black;
}

.nav-right .nav-dropdown-content {
  margin-left: -45%;
  right: 0;
}

@media screen and (max-width: 768px) {
  .distill-site-header a, .distill-site-header .nav-dropdown  {display: none;}
  .distill-site-header a.nav-toggle {
    float: right;
    display: block;
  }
  .distill-site-header .title {
    margin-left: 0;
  }
  .distill-site-header .nav-right {
    margin-right: 0;
  }
  .distill-site-header {
    overflow: hidden;
  }
  .nav-right .nav-dropdown-content {
    margin-left: 0;
  }
}


@media screen and (max-width: 768px) {
  .distill-site-header.responsive {position: relative; min-height: 500px; }
  .distill-site-header.responsive a.nav-toggle {
    position: absolute;
    right: 0;
    top: 0;
  }
  .distill-site-header.responsive a,
  .distill-site-header.responsive .nav-dropdown {
    display: block;
    text-align: left;
  }
  .distill-site-header.responsive .nav-left,
  .distill-site-header.responsive .nav-right {
    width: 100%;
  }
  .distill-site-header.responsive .nav-dropdown {float: none;}
  .distill-site-header.responsive .nav-dropdown-content {position: relative;}
  .distill-site-header.responsive .nav-dropdown .nav-dropbtn {
    display: block;
    width: 100%;
    text-align: left;
  }
}

/* Site Footer */

.distill-site-footer {
  width: 100%;
  overflow: hidden;
  box-sizing: border-box;
  z-index: 3;
  margin-top: 30px;
  padding-top: 30px;
  padding-bottom: 30px;
  text-align: center;
}

/* Headroom */

d-title {
  padding-top: 6rem;
}

@media print {
  d-title {
    padding-top: 4rem;
  }
}

.headroom {
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
}

.headroom--transition {
  transition: all .4s ease-in-out;
}

.headroom--unpinned {
  top: -100px;
}

.headroom--pinned {
  top: 0;
}

/* adjust viewport for navbar height */
/* helps vertically center bootstrap (non-distill) content */
.min-vh-100 {
  min-height: calc(100vh - 100px) !important;
}

</style>

<script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="../../site_libs/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"/>
<link href="../../site_libs/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"/>
<script src="../../site_libs/headroom-0.9.4/headroom.min.js"></script>
<!--/radix_placeholder_navigation_in_header-->
  <!--radix_placeholder_distill-->

<style type="text/css">

body {
  background-color: white;
}

.pandoc-table {
  width: 100%;
}

.pandoc-table>caption {
  margin-bottom: 10px;
}

.pandoc-table th:not([align]) {
  text-align: left;
}

.pagedtable-footer {
  font-size: 15px;
}

d-byline .byline {
  grid-template-columns: 2fr 2fr;
}

d-byline .byline h3 {
  margin-block-start: 1.5em;
}

d-byline .byline .authors-affiliations h3 {
  margin-block-start: 0.5em;
}

.authors-affiliations .orcid-id {
  width: 16px;
  height:16px;
  margin-left: 4px;
  margin-right: 4px;
  vertical-align: middle;
  padding-bottom: 2px;
}

d-title .dt-tags {
  margin-top: 1em;
  grid-column: text;
}

.dt-tags .dt-tag {
  text-decoration: none;
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0em 0.4em;
  margin-right: 0.5em;
  margin-bottom: 0.4em;
  font-size: 70%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

d-article table.gt_table td,
d-article table.gt_table th {
  border-bottom: none;
  font-size: 100%;
}

.html-widget {
  margin-bottom: 2.0em;
}

.l-screen-inset {
  padding-right: 16px;
}

.l-screen .caption {
  margin-left: 10px;
}

.shaded {
  background: rgb(247, 247, 247);
  padding-top: 20px;
  padding-bottom: 20px;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .html-widget {
  margin-bottom: 0;
  border: 1px solid rgba(0, 0, 0, 0.1);
}

.shaded .shaded-content {
  background: white;
}

.text-output {
  margin-top: 0;
  line-height: 1.5em;
}

.hidden {
  display: none !important;
}

hr.section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  margin: 0px;
}


d-byline {
  border-top: none;
}

d-article {
  padding-top: 2.5rem;
  padding-bottom: 30px;
  border-top: none;
}

d-appendix {
  padding-top: 30px;
}

d-article>p>img {
  width: 100%;
}

d-article h2 {
  margin: 1rem 0 1.5rem 0;
}

d-article h3 {
  margin-top: 1.5rem;
}

d-article iframe {
  border: 1px solid rgba(0, 0, 0, 0.1);
  margin-bottom: 2.0em;
  width: 100%;
}

/* Tweak code blocks */

d-article div.sourceCode code,
d-article pre code {
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
}

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: auto;
}

d-article div.sourceCode {
  background-color: white;
}

d-article div.sourceCode pre {
  padding-left: 10px;
  font-size: 12px;
  border-left: 2px solid rgba(0,0,0,0.1);
}

d-article pre {
  font-size: 12px;
  color: black;
  background: none;
  margin-top: 0;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

d-article pre a {
  border-bottom: none;
}

d-article pre a:hover {
  border-bottom: none;
  text-decoration: underline;
}

d-article details {
  grid-column: text;
  margin-bottom: 0.8em;
}

@media(min-width: 768px) {

d-article pre,
d-article div.sourceCode,
d-article div.sourceCode pre {
  overflow: visible !important;
}

d-article div.sourceCode pre {
  padding-left: 18px;
  font-size: 14px;
}

/* tweak for Pandoc numbered line within distill */
d-article pre.numberSource code > span {
    left: -2em;
}

d-article pre {
  font-size: 14px;
}

}

figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

/* CSS for d-contents */

.d-contents {
  grid-column: text;
  color: rgba(0,0,0,0.8);
  font-size: 0.9em;
  padding-bottom: 1em;
  margin-bottom: 1em;
  padding-bottom: 0.5em;
  margin-bottom: 1em;
  padding-left: 0.25em;
  justify-self: start;
}

@media(min-width: 1000px) {
  .d-contents.d-contents-float {
    height: 0;
    grid-column-start: 1;
    grid-column-end: 4;
    justify-self: center;
    padding-right: 3em;
    padding-left: 2em;
  }
}

.d-contents nav h3 {
  font-size: 18px;
  margin-top: 0;
  margin-bottom: 1em;
}

.d-contents li {
  list-style-type: none
}

.d-contents nav > ul {
  padding-left: 0;
}

.d-contents ul {
  padding-left: 1em
}

.d-contents nav ul li {
  margin-top: 0.6em;
  margin-bottom: 0.2em;
}

.d-contents nav a {
  font-size: 13px;
  border-bottom: none;
  text-decoration: none
  color: rgba(0, 0, 0, 0.8);
}

.d-contents nav a:hover {
  text-decoration: underline solid rgba(0, 0, 0, 0.6)
}

.d-contents nav > ul > li > a {
  font-weight: 600;
}

.d-contents nav > ul > li > ul {
  font-weight: inherit;
}

.d-contents nav > ul > li > ul > li {
  margin-top: 0.2em;
}


.d-contents nav ul {
  margin-top: 0;
  margin-bottom: 0.25em;
}

.d-article-with-toc h2:nth-child(2) {
  margin-top: 0;
}


/* Figure */

.figure {
  position: relative;
  margin-bottom: 2.5em;
  margin-top: 1.5em;
}

.figure .caption {
  color: rgba(0, 0, 0, 0.6);
  font-size: 12px;
  line-height: 1.5em;
}

.figure img.external {
  background: white;
  border: 1px solid rgba(0, 0, 0, 0.1);
  box-shadow: 0 1px 8px rgba(0, 0, 0, 0.1);
  padding: 18px;
  box-sizing: border-box;
}

.figure .caption a {
  color: rgba(0, 0, 0, 0.6);
}

.figure .caption b,
.figure .caption strong, {
  font-weight: 600;
  color: rgba(0, 0, 0, 1.0);
}

/* Citations */

d-article .citation {
  color: inherit;
  cursor: inherit;
}

div.hanging-indent{
  margin-left: 1em; text-indent: -1em;
}

/* Citation hover box */

.tippy-box[data-theme~=light-border] {
  background-color: rgba(250, 250, 250, 0.95);
}

.tippy-content > p {
  margin-bottom: 0;
  padding: 2px;
}


/* Tweak 1000px media break to show more text */

@media(min-width: 1000px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 80px [middle-start] 50px [text-start kicker-end] 65px 65px 65px 65px 65px 65px 65px 65px [text-end gutter-start] 65px [middle-end] 65px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 16px;
  }

  .grid {
    grid-column-gap: 16px;
  }

  d-article {
    font-size: 1.06rem;
    line-height: 1.7em;
  }
  figure .caption, .figure .caption, figure figcaption {
    font-size: 13px;
  }
}

@media(min-width: 1180px) {
  .base-grid,
  distill-header,
  d-title,
  d-abstract,
  d-article,
  d-appendix,
  distill-appendix,
  d-byline,
  d-footnote-list,
  d-citation-list,
  distill-footer {
    grid-template-columns: [screen-start] 1fr [page-start kicker-start] 60px [middle-start] 60px [text-start kicker-end] 60px 60px 60px 60px 60px 60px 60px 60px [text-end gutter-start] 60px [middle-end] 60px [page-end gutter-end] 1fr [screen-end];
    grid-column-gap: 32px;
  }

  .grid {
    grid-column-gap: 32px;
  }
}


/* Get the citation styles for the appendix (not auto-injected on render since
   we do our own rendering of the citation appendix) */

d-appendix .citation-appendix,
.d-appendix .citation-appendix {
  font-size: 11px;
  line-height: 15px;
  border-left: 1px solid rgba(0, 0, 0, 0.1);
  padding-left: 18px;
  border: 1px solid rgba(0,0,0,0.1);
  background: rgba(0, 0, 0, 0.02);
  padding: 10px 18px;
  border-radius: 3px;
  color: rgba(150, 150, 150, 1);
  overflow: hidden;
  margin-top: -12px;
  white-space: pre-wrap;
  word-wrap: break-word;
}

/* Include appendix styles here so they can be overridden */

d-appendix {
  contain: layout style;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-top: 60px;
  margin-bottom: 0;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  color: rgba(0,0,0,0.5);
  padding-top: 60px;
  padding-bottom: 48px;
}

d-appendix h3 {
  grid-column: page-start / text-start;
  font-size: 15px;
  font-weight: 500;
  margin-top: 1em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.65);
}

d-appendix h3 + * {
  margin-top: 1em;
}

d-appendix ol {
  padding: 0 0 0 15px;
}

@media (min-width: 768px) {
  d-appendix ol {
    padding: 0 0 0 30px;
    margin-left: -30px;
  }
}

d-appendix li {
  margin-bottom: 1em;
}

d-appendix a {
  color: rgba(0, 0, 0, 0.6);
}

d-appendix > * {
  grid-column: text;
}

d-appendix > d-footnote-list,
d-appendix > d-citation-list,
d-appendix > distill-appendix {
  grid-column: screen;
}

/* Include footnote styles here so they can be overridden */

d-footnote-list {
  contain: layout style;
}

d-footnote-list > * {
  grid-column: text;
}

d-footnote-list a.footnote-backlink {
  color: rgba(0,0,0,0.3);
  padding-left: 0.5em;
}



/* Anchor.js */

.anchorjs-link {
  /*transition: all .25s linear; */
  text-decoration: none;
  border-bottom: none;
}
*:hover > .anchorjs-link {
  margin-left: -1.125em !important;
  text-decoration: none;
  border-bottom: none;
}

/* Social footer */

.social_footer {
  margin-top: 30px;
  margin-bottom: 0;
  color: rgba(0,0,0,0.67);
}

.disqus-comments {
  margin-right: 30px;
}

.disqus-comment-count {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
  cursor: pointer;
}

#disqus_thread {
  margin-top: 30px;
}

.article-sharing a {
  border-bottom: none;
  margin-right: 8px;
}

.article-sharing a:hover {
  border-bottom: none;
}

.sidebar-section.subscribe {
  font-size: 12px;
  line-height: 1.6em;
}

.subscribe p {
  margin-bottom: 0.5em;
}


.article-footer .subscribe {
  font-size: 15px;
  margin-top: 45px;
}


.sidebar-section.custom {
  font-size: 12px;
  line-height: 1.6em;
}

.custom p {
  margin-bottom: 0.5em;
}

/* Styles for listing layout (hide title) */
.layout-listing d-title, .layout-listing .d-title {
  display: none;
}

/* Styles for posts lists (not auto-injected) */


.posts-with-sidebar {
  padding-left: 45px;
  padding-right: 45px;
}

.posts-list .description h2,
.posts-list .description p {
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

.posts-list .description h2 {
  font-weight: 700;
  border-bottom: none;
  padding-bottom: 0;
}

.posts-list h2.post-tag {
  border-bottom: 1px solid rgba(0, 0, 0, 0.2);
  padding-bottom: 12px;
}
.posts-list {
  margin-top: 60px;
  margin-bottom: 24px;
}

.posts-list .post-preview {
  text-decoration: none;
  overflow: hidden;
  display: block;
  border-bottom: 1px solid rgba(0, 0, 0, 0.1);
  padding: 24px 0;
}

.post-preview-last {
  border-bottom: none !important;
}

.posts-list .posts-list-caption {
  grid-column: screen;
  font-weight: 400;
}

.posts-list .post-preview h2 {
  margin: 0 0 6px 0;
  line-height: 1.2em;
  font-style: normal;
  font-size: 24px;
}

.posts-list .post-preview p {
  margin: 0 0 12px 0;
  line-height: 1.4em;
  font-size: 16px;
}

.posts-list .post-preview .thumbnail {
  box-sizing: border-box;
  margin-bottom: 24px;
  position: relative;
  max-width: 500px;
}
.posts-list .post-preview img {
  width: 100%;
  display: block;
}

.posts-list .metadata {
  font-size: 12px;
  line-height: 1.4em;
  margin-bottom: 18px;
}

.posts-list .metadata > * {
  display: inline-block;
}

.posts-list .metadata .publishedDate {
  margin-right: 2em;
}

.posts-list .metadata .dt-authors {
  display: block;
  margin-top: 0.3em;
  margin-right: 2em;
}

.posts-list .dt-tags {
  display: block;
  line-height: 1em;
}

.posts-list .dt-tags .dt-tag {
  display: inline-block;
  color: rgba(0,0,0,0.6);
  padding: 0.3em 0.4em;
  margin-right: 0.2em;
  margin-bottom: 0.4em;
  font-size: 60%;
  border: 1px solid rgba(0,0,0,0.2);
  border-radius: 3px;
  text-transform: uppercase;
  font-weight: 500;
}

.posts-list img {
  opacity: 1;
}

.posts-list img[data-src] {
  opacity: 0;
}

.posts-more {
  clear: both;
}


.posts-sidebar {
  font-size: 16px;
}

.posts-sidebar h3 {
  font-size: 16px;
  margin-top: 0;
  margin-bottom: 0.5em;
  font-weight: 400;
  text-transform: uppercase;
}

.sidebar-section {
  margin-bottom: 30px;
}

.categories ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.categories li {
  color: rgba(0, 0, 0, 0.8);
  margin-bottom: 0;
}

.categories li>a {
  border-bottom: none;
}

.categories li>a:hover {
  border-bottom: 1px solid rgba(0, 0, 0, 0.4);
}

.categories .active {
  font-weight: 600;
}

.categories .category-count {
  color: rgba(0, 0, 0, 0.4);
}


@media(min-width: 768px) {
  .posts-list .post-preview h2 {
    font-size: 26px;
  }
  .posts-list .post-preview .thumbnail {
    float: right;
    width: 30%;
    margin-bottom: 0;
  }
  .posts-list .post-preview .description {
    float: left;
    width: 45%;
  }
  .posts-list .post-preview .metadata {
    float: left;
    width: 20%;
    margin-top: 8px;
  }
  .posts-list .post-preview p {
    margin: 0 0 12px 0;
    line-height: 1.5em;
    font-size: 16px;
  }
  .posts-with-sidebar .posts-list {
    float: left;
    width: 75%;
  }
  .posts-with-sidebar .posts-sidebar {
    float: right;
    width: 20%;
    margin-top: 60px;
    padding-top: 24px;
    padding-bottom: 24px;
  }
}


/* Improve display for browsers without grid (IE/Edge <= 15) */

.downlevel {
  line-height: 1.6em;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  margin: 0;
}

.downlevel .d-title {
  padding-top: 6rem;
  padding-bottom: 1.5rem;
}

.downlevel .d-title h1 {
  font-size: 50px;
  font-weight: 700;
  line-height: 1.1em;
  margin: 0 0 0.5rem;
}

.downlevel .d-title p {
  font-weight: 300;
  font-size: 1.2rem;
  line-height: 1.55em;
  margin-top: 0;
}

.downlevel .d-byline {
  padding-top: 0.8em;
  padding-bottom: 0.8em;
  font-size: 0.8rem;
  line-height: 1.8em;
}

.downlevel .section-separator {
  border: none;
  border-top: 1px solid rgba(0, 0, 0, 0.1);
}

.downlevel .d-article {
  font-size: 1.06rem;
  line-height: 1.7em;
  padding-top: 1rem;
  padding-bottom: 2rem;
}


.downlevel .d-appendix {
  padding-left: 0;
  padding-right: 0;
  max-width: none;
  font-size: 0.8em;
  line-height: 1.7em;
  margin-bottom: 0;
  color: rgba(0,0,0,0.5);
  padding-top: 40px;
  padding-bottom: 48px;
}

.downlevel .footnotes ol {
  padding-left: 13px;
}

.downlevel .base-grid,
.downlevel .distill-header,
.downlevel .d-title,
.downlevel .d-abstract,
.downlevel .d-article,
.downlevel .d-appendix,
.downlevel .distill-appendix,
.downlevel .d-byline,
.downlevel .d-footnote-list,
.downlevel .d-citation-list,
.downlevel .distill-footer,
.downlevel .appendix-bottom,
.downlevel .posts-container {
  padding-left: 40px;
  padding-right: 40px;
}

@media(min-width: 768px) {
  .downlevel .base-grid,
  .downlevel .distill-header,
  .downlevel .d-title,
  .downlevel .d-abstract,
  .downlevel .d-article,
  .downlevel .d-appendix,
  .downlevel .distill-appendix,
  .downlevel .d-byline,
  .downlevel .d-footnote-list,
  .downlevel .d-citation-list,
  .downlevel .distill-footer,
  .downlevel .appendix-bottom,
  .downlevel .posts-container {
  padding-left: 150px;
  padding-right: 150px;
  max-width: 900px;
}
}

.downlevel pre code {
  display: block;
  border-left: 2px solid rgba(0, 0, 0, .1);
  padding: 0 0 0 20px;
  font-size: 14px;
}

.downlevel code, .downlevel pre {
  color: black;
  background: none;
  font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  text-align: left;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  word-wrap: normal;
  line-height: 1.5;

  -moz-tab-size: 4;
  -o-tab-size: 4;
  tab-size: 4;

  -webkit-hyphens: none;
  -moz-hyphens: none;
  -ms-hyphens: none;
  hyphens: none;
}

.downlevel .posts-list .post-preview {
  color: inherit;
}



</style>

<script type="application/javascript">

function is_downlevel_browser() {
  if (bowser.isUnsupportedBrowser({ msie: "12", msedge: "16"},
                                 window.navigator.userAgent)) {
    return true;
  } else {
    return window.load_distill_framework === undefined;
  }
}

// show body when load is complete
function on_load_complete() {

  // add anchors
  if (window.anchors) {
    window.anchors.options.placement = 'left';
    window.anchors.add('d-article > h2, d-article > h3, d-article > h4, d-article > h5');
  }


  // set body to visible
  document.body.style.visibility = 'visible';

  // force redraw for leaflet widgets
  if (window.HTMLWidgets) {
    var maps = window.HTMLWidgets.findAll(".leaflet");
    $.each(maps, function(i, el) {
      var map = this.getMap();
      map.invalidateSize();
      map.eachLayer(function(layer) {
        if (layer instanceof L.TileLayer)
          layer.redraw();
      });
    });
  }

  // trigger 'shown' so htmlwidgets resize
  $('d-article').trigger('shown');
}

function init_distill() {

  init_common();

  // create front matter
  var front_matter = $('<d-front-matter></d-front-matter>');
  $('#distill-front-matter').wrap(front_matter);

  // create d-title
  $('.d-title').changeElementType('d-title');

  // separator
  var separator = '<hr class="section-separator" style="clear: both"/>';
  // prepend separator above appendix
  $('.d-byline').before(separator);
  $('.d-article').before(separator);

  // create d-byline
  var byline = $('<d-byline></d-byline>');
  $('.d-byline').replaceWith(byline);

  // create d-article
  var article = $('<d-article></d-article>');
  $('.d-article').wrap(article).children().unwrap();

  // move posts container into article
  $('.posts-container').appendTo($('d-article'));

  // create d-appendix
  $('.d-appendix').changeElementType('d-appendix');

  // flag indicating that we have appendix items
  var appendix = $('.appendix-bottom').children('h3').length > 0;

  // replace footnotes with <d-footnote>
  $('.footnote-ref').each(function(i, val) {
    appendix = true;
    var href = $(this).attr('href');
    var id = href.replace('#', '');
    var fn = $('#' + id);
    var fn_p = $('#' + id + '>p');
    fn_p.find('.footnote-back').remove();
    var text = fn_p.html();
    var dtfn = $('<d-footnote></d-footnote>');
    dtfn.html(text);
    $(this).replaceWith(dtfn);
  });
  // remove footnotes
  $('.footnotes').remove();

  // move refs into #references-listing
  $('#references-listing').replaceWith($('#refs'));

  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    var id = $(this).attr('id');
    $('.d-contents a[href="#' + id + '"]').parent().remove();
    appendix = true;
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('d-appendix'));
  });

  // show d-appendix if we have appendix content
  $("d-appendix").css('display', appendix ? 'grid' : 'none');

  // localize layout chunks to just output
  $('.layout-chunk').each(function(i, val) {

    // capture layout
    var layout = $(this).attr('data-layout');

    // apply layout to markdown level block elements
    var elements = $(this).children().not('details, div.sourceCode, pre, script');
    elements.each(function(i, el) {
      var layout_div = $('<div class="' + layout + '"></div>');
      if (layout_div.hasClass('shaded')) {
        var shaded_content = $('<div class="shaded-content"></div>');
        $(this).wrap(shaded_content);
        $(this).parent().wrap(layout_div);
      } else {
        $(this).wrap(layout_div);
      }
    });


    // unwrap the layout-chunk div
    $(this).children().unwrap();
  });

  // remove code block used to force  highlighting css
  $('.distill-force-highlighting-css').parent().remove();

  // remove empty line numbers inserted by pandoc when using a
  // custom syntax highlighting theme, except when numbering line
  // in code chunk
  $('pre:not(.numberLines) code.sourceCode a:empty').remove();

  // load distill framework
  load_distill_framework();

  // wait for window.distillRunlevel == 4 to do post processing
  function distill_post_process() {

    if (!window.distillRunlevel || window.distillRunlevel < 4)
      return;

    // hide author/affiliations entirely if we have no authors
    var front_matter = JSON.parse($("#distill-front-matter").html());
    var have_authors = front_matter.authors && front_matter.authors.length > 0;
    if (!have_authors)
      $('d-byline').addClass('hidden');

    // article with toc class
    $('.d-contents').parent().addClass('d-article-with-toc');

    // strip links that point to #
    $('.authors-affiliations').find('a[href="#"]').removeAttr('href');

    // add orcid ids
    $('.authors-affiliations').find('.author').each(function(i, el) {
      var orcid_id = front_matter.authors[i].orcidID;
      var author_name = front_matter.authors[i].author
      if (orcid_id) {
        var a = $('<a></a>');
        a.attr('href', 'https://orcid.org/' + orcid_id);
        var img = $('<img></img>');
        img.addClass('orcid-id');
        img.attr('alt', author_name ? 'ORCID ID for ' + author_name : 'ORCID ID');
        img.attr('src','data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg==');
        a.append(img);
        $(this).append(a);
      }
    });

    // hide elements of author/affiliations grid that have no value
    function hide_byline_column(caption) {
      $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'hidden');
    }

    // affiliations
    var have_affiliations = false;
    for (var i = 0; i<front_matter.authors.length; ++i) {
      var author = front_matter.authors[i];
      if (author.affiliation !== "&nbsp;") {
        have_affiliations = true;
        break;
      }
    }
    if (!have_affiliations)
      $('d-byline').find('h3:contains("Affiliations")').css('visibility', 'hidden');

    // published date
    if (!front_matter.publishedDate)
      hide_byline_column("Published");

    // document object identifier
    var doi = $('d-byline').find('h3:contains("DOI")');
    var doi_p = doi.next().empty();
    if (!front_matter.doi) {
      // if we have a citation and valid citationText then link to that
      if ($('#citation').length > 0 && front_matter.citationText) {
        doi.html('Citation');
        $('<a href="#citation"></a>')
          .text(front_matter.citationText)
          .appendTo(doi_p);
      } else {
        hide_byline_column("DOI");
      }
    } else {
      $('<a></a>')
         .attr('href', "https://doi.org/" + front_matter.doi)
         .html(front_matter.doi)
         .appendTo(doi_p);
    }

     // change plural form of authors/affiliations
    if (front_matter.authors.length === 1) {
      var grid = $('.authors-affiliations');
      grid.children('h3:contains("Authors")').text('Author');
      grid.children('h3:contains("Affiliations")').text('Affiliation');
    }

    // remove d-appendix and d-footnote-list local styles
    $('d-appendix > style:first-child').remove();
    $('d-footnote-list > style:first-child').remove();

    // move appendix-bottom entries to the bottom
    $('.appendix-bottom').appendTo('d-appendix').children().unwrap();
    $('.appendix-bottom').remove();

    // hoverable references
    $('span.citation[data-cites]').each(function() {
      const citeChild = $(this).children()[0]
      // Do not process if @xyz has been used without escaping and without bibliography activated
      // https://github.com/rstudio/distill/issues/466
      if (citeChild === undefined) return true

      if (citeChild.nodeName == "D-FOOTNOTE") {
        var fn = citeChild
        $(this).html(fn.shadowRoot.querySelector("sup"))
        $(this).id = fn.id
        fn.remove()
      }
      var refs = $(this).attr('data-cites').split(" ");
      var refHtml = refs.map(function(ref) {
        // Could use CSS.escape too here, we insure backward compatibility in navigator
        return "<p>" + $('div[id="ref-' + ref + '"]').html() + "</p>";
      }).join("\n");
      window.tippy(this, {
        allowHTML: true,
        content: refHtml,
        maxWidth: 500,
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start'
      });
    });

    // fix footnotes in tables (#411)
    // replacing broken distill.pub feature
    $('table d-footnote').each(function() {
      // we replace internal showAtNode methode which is triggered when hovering a footnote
      this.hoverBox.showAtNode = function(node) {
        // ported from https://github.com/distillpub/template/pull/105/files
        calcOffset = function(elem) {
            let x = elem.offsetLeft;
            let y = elem.offsetTop;
            // Traverse upwards until an `absolute` element is found or `elem`
            // becomes null.
            while (elem = elem.offsetParent && elem.style.position != 'absolute') {
                x += elem.offsetLeft;
                y += elem.offsetTop;
            }

            return { left: x, top: y };
        }
        // https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop
        const bbox = node.getBoundingClientRect();
        const offset = calcOffset(node);
        this.show([offset.left + bbox.width, offset.top + bbox.height]);
      }
    })

    // clear polling timer
    clearInterval(tid);

    // show body now that everything is ready
    on_load_complete();
  }

  var tid = setInterval(distill_post_process, 50);
  distill_post_process();

}

function init_downlevel() {

  init_common();

   // insert hr after d-title
  $('.d-title').after($('<hr class="section-separator"/>'));

  // check if we have authors
  var front_matter = JSON.parse($("#distill-front-matter").html());
  var have_authors = front_matter.authors && front_matter.authors.length > 0;

  // manage byline/border
  if (!have_authors)
    $('.d-byline').remove();
  $('.d-byline').after($('<hr class="section-separator"/>'));
  $('.d-byline a').remove();

  // remove toc
  $('.d-contents').remove();

  // move appendix elements
  $('h1.appendix, h2.appendix').each(function(i, val) {
    $(this).changeElementType('h3');
  });
  $('h3.appendix').each(function(i, val) {
    $(this).nextUntil($('h1, h2, h3')).addBack().appendTo($('.d-appendix'));
  });


  // inject headers into references and footnotes
  var refs_header = $('<h3></h3>');
  refs_header.text('References');
  $('#refs').prepend(refs_header);

  var footnotes_header = $('<h3></h3');
  footnotes_header.text('Footnotes');
  $('.footnotes').children('hr').first().replaceWith(footnotes_header);

  // move appendix-bottom entries to the bottom
  $('.appendix-bottom').appendTo('.d-appendix').children().unwrap();
  $('.appendix-bottom').remove();

  // remove appendix if it's empty
  if ($('.d-appendix').children().length === 0)
    $('.d-appendix').remove();

  // prepend separator above appendix
  $('.d-appendix').before($('<hr class="section-separator" style="clear: both"/>'));

  // trim code
  $('pre>code').each(function(i, val) {
    $(this).html($.trim($(this).html()));
  });

  // move posts-container right before article
  $('.posts-container').insertBefore($('.d-article'));

  $('body').addClass('downlevel');

  on_load_complete();
}


function init_common() {

  // jquery plugin to change element types
  (function($) {
    $.fn.changeElementType = function(newType) {
      var attrs = {};

      $.each(this[0].attributes, function(idx, attr) {
        attrs[attr.nodeName] = attr.nodeValue;
      });

      this.replaceWith(function() {
        return $("<" + newType + "/>", attrs).append($(this).contents());
      });
    };
  })(jQuery);

  // prevent underline for linked images
  $('a > img').parent().css({'border-bottom' : 'none'});

  // mark non-body figures created by knitr chunks as 100% width
  $('.layout-chunk').each(function(i, val) {
    var figures = $(this).find('img, .html-widget');
    // ignore leaflet img layers (#106)
    figures = figures.filter(':not(img[class*="leaflet"])')
    if ($(this).attr('data-layout') !== "l-body") {
      figures.css('width', '100%');
    } else {
      figures.css('max-width', '100%');
      figures.filter("[width]").each(function(i, val) {
        var fig = $(this);
        fig.css('width', fig.attr('width') + 'px');
      });

    }
  });

  // auto-append index.html to post-preview links in file: protocol
  // and in rstudio ide preview
  $('.post-preview').each(function(i, val) {
    if (window.location.protocol === "file:")
      $(this).attr('href', $(this).attr('href') + "index.html");
  });

  // get rid of index.html references in header
  if (window.location.protocol !== "file:") {
    $('.distill-site-header a[href]').each(function(i,val) {
      $(this).attr('href', $(this).attr('href').replace(/^index[.]html/, "./"));
    });
  }

  // add class to pandoc style tables
  $('tr.header').parent('thead').parent('table').addClass('pandoc-table');
  $('.kable-table').children('table').addClass('pandoc-table');

  // add figcaption style to table captions
  $('caption').parent('table').addClass("figcaption");

  // initialize posts list
  if (window.init_posts_list)
    window.init_posts_list();

  // implmement disqus comment link
  $('.disqus-comment-count').click(function() {
    window.headroom_prevent_pin = true;
    $('#disqus_thread').toggleClass('hidden');
    if (!$('#disqus_thread').hasClass('hidden')) {
      var offset = $(this).offset();
      $(window).resize();
      $('html, body').animate({
        scrollTop: offset.top - 35
      });
    }
  });
}

document.addEventListener('DOMContentLoaded', function() {
  if (is_downlevel_browser())
    init_downlevel();
  else
    window.addEventListener('WebComponentsReady', init_distill);
});

</script>

<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.5);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

.distill-site-footer {
  --text-color:       rgba(255, 255, 255, 0.8);
  --text-size:        15px;
  --hover-color:      white;
  --bkgd-color:       #0F2E3D;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */
</style>
<style type="text/css">
/* base variables */

/* Edit the CSS properties in this file to create a custom
   Distill theme. Only edit values in the right column
   for each row; values shown are the CSS defaults.
   To return any property to the default,
   you may set its value to: unset
   All rows must end with a semi-colon.                      */

/* Optional: embed custom fonts here with `@import`          */
/* This must remain at the top of this file.                 */



html {
  /*-- Main font sizes --*/
  --title-size:      50px;
  --body-size:       1.06rem;
  --code-size:       14px;
  --aside-size:      12px;
  --fig-cap-size:    13px;
  /*-- Main font colors --*/
  --title-color:     #000000;
  --header-color:    rgba(0, 0, 0, 0.8);
  --body-color:      rgba(0, 0, 0, 0.8);
  --aside-color:     rgba(0, 0, 0, 0.6);
  --fig-cap-color:   rgba(0, 0, 0, 0.6);
  /*-- Specify custom fonts ~~~ must be imported above   --*/
  --heading-font:    sans-serif;
  --mono-font:       monospace;
  --body-font:       sans-serif;
  --navbar-font:     sans-serif;  /* websites + blogs only */
}

/*-- ARTICLE METADATA --*/
d-byline {
  --heading-size:    0.6rem;
  --heading-color:   rgba(0, 0, 0, 0.5);
  --body-size:       0.8rem;
  --body-color:      rgba(0, 0, 0, 0.8);
}

/*-- ARTICLE TABLE OF CONTENTS --*/
.d-contents {
  --heading-size:    18px;
  --contents-size:   13px;
}

/*-- ARTICLE APPENDIX --*/
d-appendix {
  --heading-size:    15px;
  --heading-color:   rgba(0, 0, 0, 0.65);
  --text-size:       0.8em;
  --text-color:      rgba(0, 0, 0, 0.7);
}

/*-- WEBSITE HEADER + FOOTER --*/
/* These properties only apply to Distill sites and blogs  */

.distill-site-header {
  --title-size:       18px;
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #ffffff;
}

.distill-site-footer {
  --text-color:       rgba(0, 0, 0, 0.8);
  --text-size:        15px;
  --hover-color:      black;
  --bkgd-color:       #fafafa;
}

/*-- Additional custom styles --*/
/* Add any additional CSS rules below                      */

.nav-right > a {
  text-transform: uppercase;
}

d-title h1, d-title p, d-title figure,
d-abstract p, d-abstract b {
  grid-column: page;
}

.rj-blue {
  color: #2467bb;
}

ul li {
  line-height: 1.6;
  margin-top: 0em;
  margin-bottom: 0em;
}</style>
<style type="text/css">
/* base style */

/* FONT FAMILIES */

:root {
  --heading-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
  --mono-default: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
  --body-default: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, "Fira Sans", "Droid Sans", "Helvetica Neue", Arial, sans-serif;
}

body,
.posts-list .post-preview p,
.posts-list .description p {
  font-family: var(--body-font), var(--body-default);
}

h1, h2, h3, h4, h5, h6,
.posts-list .post-preview h2,
.posts-list .description h2 {
  font-family: var(--heading-font), var(--heading-default);
}

d-article div.sourceCode code,
d-article pre code {
  font-family: var(--mono-font), var(--mono-default);
}


/*-- TITLE --*/
d-title h1,
.posts-list > h1 {
  color: var(--title-color, black);
}

d-title h1 {
  font-size: var(--title-size, 50px);
}

/*-- HEADERS --*/
d-article h1,
d-article h2,
d-article h3,
d-article h4,
d-article h5,
d-article h6 {
  color: var(--header-color, rgba(0, 0, 0, 0.8));
}

/*-- BODY --*/
d-article > p,  /* only text inside of <p> tags */
d-article > ul, /* lists */
d-article > ol {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
  font-size: var(--body-size, 1.06rem);
}


/*-- CODE --*/
d-article div.sourceCode code,
d-article pre code {
  font-size: var(--code-size, 14px);
}

/*-- ASIDE --*/
d-article aside {
  font-size: var(--aside-size, 12px);
  color: var(--aside-color, rgba(0, 0, 0, 0.6));
}

/*-- FIGURE CAPTIONS --*/
figure .caption,
figure figcaption,
.figure .caption {
  font-size: var(--fig-cap-size, 13px);
  color: var(--fig-cap-color, rgba(0, 0, 0, 0.6));
}

/*-- METADATA --*/
d-byline h3 {
  font-size: var(--heading-size, 0.6rem);
  color: var(--heading-color, rgba(0, 0, 0, 0.5));
}

d-byline {
  font-size: var(--body-size, 0.8rem);
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

d-byline a,
d-article d-byline a {
  color: var(--body-color, rgba(0, 0, 0, 0.8));
}

/*-- TABLE OF CONTENTS --*/
.d-contents nav h3 {
  font-size: var(--heading-size, 18px);
}

.d-contents nav a {
  font-size: var(--contents-size, 13px);
}

/*-- APPENDIX --*/
d-appendix h3 {
  font-size: var(--heading-size, 15px);
  color: var(--heading-color, rgba(0, 0, 0, 0.65));
}

d-appendix {
  font-size: var(--text-size, 0.8em);
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

d-appendix d-footnote-list a.footnote-backlink {
  color: var(--text-color, rgba(0, 0, 0, 0.5));
}

/*-- WEBSITE HEADER + FOOTER --*/
.distill-site-header .title {
  font-size: var(--title-size, 18px);
  font-family: var(--navbar-font), var(--heading-default);
}

.distill-site-header a,
.nav-dropdown .nav-dropbtn {
  font-family: var(--navbar-font), var(--heading-default);
}

.nav-dropdown .nav-dropbtn {
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  font-size: var(--text-size, 15px);
}

.distill-site-header a:hover,
.nav-dropdown:hover .nav-dropbtn {
  color: var(--hover-color, white);
}

.distill-site-header {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer {
  font-size: var(--text-size, 15px);
  color: var(--text-color, rgba(255, 255, 255, 0.8));
  background-color: var(--bkgd-color, #0F2E3D);
}

.distill-site-footer a:hover {
  color: var(--hover-color, white);
}</style>
<!--/radix_placeholder_distill-->
  <script src="../../site_libs/header-attrs-2.29/header-attrs.js"></script>
  <script src="../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <script src="../../site_libs/popper-2.6.0/popper.min.js"></script>
  <link href="../../site_libs/tippy-6.2.7/tippy.css" rel="stylesheet" />
  <link href="../../site_libs/tippy-6.2.7/tippy-light-border.css" rel="stylesheet" />
  <script src="../../site_libs/tippy-6.2.7/tippy.umd.min.js"></script>
  <script src="../../site_libs/anchor-4.2.2/anchor.min.js"></script>
  <script src="../../site_libs/bowser-1.9.3/bowser.min.js"></script>
  <script src="../../site_libs/webcomponents-2.0.0/webcomponents.js"></script>
  <script src="../../site_libs/distill-2.2.21/template.v2.js"></script>
  <!--radix_placeholder_site_in_header-->
<!--/radix_placeholder_site_in_header-->
  <script>
    $(function() {
      console.log("Starting...")

      // Mathjax config (add automatic linebreaks when supported)
      // MathJax = {
      //    tex: {
      //        inlineMath: [['$', '$'], ['\\(', '\\)']],
      //        displayMath: [['$$', '$$'], ['\\[', '\\]']],
      //        tags: 'ams',
      //        multline: true,
      //    },
      //    options: {
      //        linebreaks: { automatic: true },
      //    },
      // };

      // Always show Published - distill hides it if not set
      function show_byline_column(caption) {
        $('d-byline').find('h3:contains("' + caption + '")').parent().css('visibility', 'visible');
      }

      show_byline_column('Published')

      // tweak function
      var rmd_meta = JSON.parse($("#radix-rmarkdown-metadata").html());
      function get_meta(name, meta) {
        var ind = meta.attributes.names.value.findIndex((e) => e == name)
        var val = meta.value[ind]
        if (val.type != 'list') {
          return val.value.toString()
        }
        return val
      }

      // tweak description
      // Add clickable tags
      const slug = get_meta('slug', rmd_meta)
      const cite_url = get_meta('citation_url', rmd_meta)

      var title = $("d-title").text

      const buttons = $('<div class="dt-tags" style="grid-column: page;">')
      buttons.append('<a href="#citation" class="dt-tag"><i class="fas fa-quote-left"></i> Cite</a>')
      buttons.append('<a href="' + slug + '.pdf" class="dt-tag"><i class="fas fa-file-pdf"></i> PDF</a>')
      buttons.append('<a href="' + slug + '.zip" class="dt-tag"><i class="fas fa-file-zipper"></i> Supplement</a>')

      // adds Abstract: in front of the first <p> in the title section --
      // unless it happens to be the subtitle (FIXME: this is a bad hack - can't distill do this?)
      var tpar = $("d-title p:not(:empty)").filter(function() {
        return !$(this).hasClass("subtitle");
      }).first();
      if (tpar) {
        const abstract = $('<d-abstract>')
        abstract.append('<b>Abstract:</b><br>')
        abstract.append(tpar) // Move description to d-abstract
        $("d-title p:empty").remove() // Remove empty paragraphs after title
        abstract.append(buttons)
        abstract.insertAfter($('d-title')) // Add abstract section after title */
      }

      // tweak by-line
      var byline = $("d-byline div.byline")
      ind = rmd_meta.attributes.names.value.findIndex((e) => e == "journal")
      const journal = get_meta('journal', rmd_meta)
      const volume = get_meta('volume', rmd_meta)
      const issue = get_meta('issue', rmd_meta)
      const jrtitle = get_meta('title', journal)
      const year = ((jrtitle == "R News") ? 2000 : 2008) + parseInt(volume)
      const firstpage = get_meta('firstpage', journal)
      const lastpage = get_meta('lastpage', journal)
      byline.append('<div class="rjournal grid">')
      $('div.rjournal').append('<h3>Volume</h3>')
      $('div.rjournal').append('<h3>Pages</h3>')
      $('div.rjournal').append('<a class="volume" href="../../issues/'+year+'-'+issue+'">'+volume+'/'+issue+'</a>')
      $('div.rjournal').append('<p class="pages">'+firstpage+' - '+lastpage+'</p>')

      const received_date = new Date(get_meta('date_received', rmd_meta))
      byline.find('h3:contains("Published")').parent().append('<h3>Received</h3><p>'+received_date.toLocaleDateString('en-US', {month: 'short'})+' '+received_date.getDate()+', '+received_date.getFullYear()+'</p>')

    })
  </script>

  <style>
      /*
    .nav-dropdown-content .nav-dropdown-header {
      text-transform: lowercase;
    }
    */

    d-byline .byline {
      grid-template-columns: 2fr 2fr 2fr 2fr;
    }

    d-byline .rjournal {
      grid-column-end: span 2;
      grid-template-columns: 1fr 1fr;
      margin-bottom: 0;
    }

    d-title h1, d-title p, d-title figure,
    d-abstract p, d-abstract b {
      grid-column: page;
    }

    d-title .dt-tags {
      grid-column: page;
    }

    .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    d-article h1 {
      line-height: 1.1em;
    }

    d-abstract p, d-article p {
      text-align: justify;
    }

    @media(min-width: 1000px) {
      .d-contents.d-contents-float {
        justify-self: end;
      }

      nav.toc {
        border-right: 1px solid rgba(0, 0, 0, 0.1);
        border-right-width: 1px;
        border-right-style: solid;
        border-right-color: rgba(0, 0, 0, 0.1);
      }
    }

    .posts-list .dt-tags .dt-tag {
      text-transform: lowercase;
    }

    @keyframes highlight-target {
      0% {
        background-color: #ffa;
      }
      66% {
        background-color: #ffa;
      }
      100% {
        background-color: none;
      }
    }

    d-article :target, d-appendix :target {
       animation: highlight-target 3s;
    }

    .header-section-number {
      margin-right: 0.5em;
    }
    
    d-appendix .citation-appendix,
    .d-appendix .citation-appendix {
      color: rgb(60, 60, 60);
    }

    d-article h2 {
      border-bottom: 0px solid rgba(0, 0, 0, 0.1);
      padding-bottom: 0rem;
    }
    d-article h3 {
      font-size: 20px;
    }
    d-article h4 {
      font-size: 18px;
      text-transform: none;
    }

    @media (min-width: 1024px) {
      d-article h2 {
        font-size: 32px;
      }
      d-article h3 {
        font-size: 24px;
      }
      d-article h4 {
        font-size: 20px;
      }
    }
  </style>


</head>

<body>

<!--radix_placeholder_front_matter-->

<script id="distill-front-matter" type="text/json">
{"title":"Partitioned Local Depth (PaLD) Community Analyses in R","description":"Partitioned Local Depth (PaLD) is a framework for holistic\nconsideration of community structure for distance-based data. This\npaper describes an R package, pald, for calculating\nPartitioned Local Depth (PaLD) probabilities, implementing community\nanalyses, determining community clusters, and creating data\nvisualizations to display community structure. We present essentials\nof the PaLD approach, describe how to use the pald package, walk\nthrough several examples, and discuss the method in relation to\ncommonly used techniques.","doi":"10.32614/RJ-2025-013","authors":[{"author":"Lucy D'Agostino McGowan","authorURL":"#","affiliation":"Wake Forest University","affiliationURL":"#","orcidID":""},{"author":"Katherine Moore","authorURL":"#","affiliation":"Amherst College","affiliationURL":"#","orcidID":""},{"author":"Kenneth S. Berenhaut","authorURL":"#","affiliation":"Wake Forest University","affiliationURL":"#","orcidID":""}],"publishedDate":"2025-09-24T00:00:00.000+00:00","citationText":"McGowan, et al., 2025"}
</script>

<!--/radix_placeholder_front_matter-->
<!--radix_placeholder_navigation_before_body-->
<header class="header header--fixed" role="banner">
<nav class="distill-site-nav distill-site-header">
<div class="nav-left">
<a class="logo" href="../../index.html">
<img src="../../resources/rlogo.png" alt="Logo"/>
</a>
<a href="../../index.html" class="title">The R Journal</a>
</div>
<div class="nav-right">
<a href="../../index.html">Home</a>
<a href="../../issues/2025-2">Current</a>
<a href="../../issues.html">Issues</a>
<a href="../../news.html">News</a>
<a href="../../submissions.html">Submit</a>
<a href="../../contribute.html">Contribute</a>
<a href="../../editors.html">Editorial board</a>
<a href="../../articles.xml">
<i class="fa fa-rss" aria-hidden="true"></i>
</a>
<a href="javascript:void(0);" class="nav-toggle">&#9776;</a>
</div>
</nav>
</header>
<!--/radix_placeholder_navigation_before_body-->
<!--radix_placeholder_site_before_body-->
<!--/radix_placeholder_site_before_body-->

<div class="d-title">
<h1>Partitioned Local Depth (PaLD) Community Analyses in R</h1>

<!--radix_placeholder_categories-->
<!--/radix_placeholder_categories-->
<p><p>Partitioned Local Depth (PaLD) is a framework for holistic
consideration of community structure for distance-based data. This
paper describes an R package, pald, for calculating
Partitioned Local Depth (PaLD) probabilities, implementing community
analyses, determining community clusters, and creating data
visualizations to display community structure. We present essentials
of the PaLD approach, describe how to use the pald package, walk
through several examples, and discuss the method in relation to
commonly used techniques.</p></p>
</div>

<div class="d-byline">
  Lucy D’Agostino McGowan  (Wake Forest University)
  
,   Katherine Moore  (Amherst College)
  
,   Kenneth S. Berenhaut  (Wake Forest University)
  
<br/>2025-09-24
</div>

<div class="d-article">
<div class="article">
<h4 class="unnumbered" data-number="0.1" id="introduction">Introduction</h4>
<p>Partitioned Local Depth (PaLD) is a framework for holistic consideration
of community structure for distance-based data. Leveraging a socially
inspired perspective, the approach provides network-based community
information which is founded on new measures of local depth and pairwise
cohesion (partitioned local depth). The method does not require
distributional assumptions, optimization criteria, nor extraneous
inputs. A complete description of the perspective, together with a
discussion of the underlying social motivation, theoretical results, and
applications to additional data sets is provided in
1. A brief technical description is included below,
directly following the introduction.</p>
<p>As suggested in 1, a main goal of PaLD is to
“transform input dissimilarity comparisons into output pairwise
relationship strengths (or cohesion) and resulting weighted networks”.
This is intended to provide within- and between-community structural
information which goes beyond simple cluster labeling. Building on
existing approaches to (global) depth, local depth expresses features of
centrality via an interpretable probability which is free of parameters
and robust to outliers. Partitioning the probabilities which define
local depth, we then obtain a measure of cohesion between pairs of
individuals. Both local depth and cohesion reflect aspects of relative
position (rather than absolute distance) and provide a straightforward
way to account for varying density across the space. As shown in
1, provided that two sets are separated (in the sense
that the minimum between-set distance is greater than the maximum
within-set distance), cohesion is invariant under the contraction and
dilation of the distances within each set. This property may be
particularly valuable when one has reason to believe that there is
heterogeneity in density across the space.</p>
<p>As cohesion captures a sense of the relationship strength between
points, we can then analyze and visualize the resulting community
structure via a network whose edges are weighted by (mutual) cohesion.
The underlying social framework motivates a straightforward yet elegant
threshold for distinguishing between strongly and weakly cohesive pairs.</p>
<p>Networks obtained from cohesion can be displayed using a force-directed
graph drawing algorithm; here we will graphically emphasize the strong
ties (colored by connected component). We refer to the connected
components of the network of strong ties as community “clusters”. Note
that to qualify as a cluster in this definition, one may not have any
strong ties with those outside the cluster, and thus the existence of
disjoint groups is a strong signal for separation. Here, clusters are
identified without additional user inputs nor optimization criteria. If
one wishes to further break the community graph into groups, one may
consider using community detection methods (such as spectral clustering
or the Louvain algorithm), as available, say, in the
<a href="https://CRAN.R-project.org/package=igraph"><strong>igraph</strong></a> package. The
collection of strong ties may be used in place of (weighted) <span class="math inline">\(k\)</span>-nearest
neighbors in settings such as classification and smoothing. Overall, the
structural information obtained from local depth, cohesion and community
graphs can provide a holistic perspective to the data which does not
require the use of distributional assumptions, optimization criteria nor
additional user inputs.</p>
<p>It is important to emphasize at the outset that community analyses go
beyond simple cluster labeling to address intra- and inter-cluster
structure, and can supplement results from other methods for clustering,
embedding, data depth, nearest neighbors, etc. In addition, the method
does not require distributional assumptions, optimization criteria, nor
extraneous inputs. Theoretical considerations (see 1)
and examples point to distinctive properties that assist in
considerations of complex data, in particular with respect to varying
density and high dimensions. See the section Cultural and Psychological
distance analysis, below, for some discussion in the context of
density-based methods.</p>
<p>After a brief introduction to the partitioned local depth approach
(including a concrete and instructive example), below, we present a new
package, <a href="https://CRAN.R-project.org/package=pald"><strong>pald</strong></a>, for
calculating partitioned local depths, implementing community analyses,
and creating data visualizations to display community structure. This
paper describes how to use the package, walks through several examples,
and contrasts the method results with commonly used techniques.
Together, these demonstrate both the novelty of the method and utility
of the implementation in the package described.</p>
<h4 class="unnumbered" data-number="0.2" id="the-partitioned-local-depth-approach">The partitioned local depth approach</h4>
<p>The PaLD methodology as introduced in 1 offers a
parameter-free approach to analyzing community structure in
distance-based data. First, consider a ground set <span class="math inline">\(\mathcal{S}\)</span> equipped
with a meaningful measure of pairwise distance (or dissimilarity),
<span class="math inline">\(d:\mathcal{S}\times S\rightarrow \mathbb{R}\cup \infty\)</span>. We require
only that <span class="math inline">\(d\)</span> satisfies the requirements that for all <span class="math inline">\(x,y\in S\)</span> with
<span class="math inline">\(y\neq x\)</span>,</p>
<p><span class="math display" id="eq:theonereq">\[\begin{aligned}
d(x,x) \leq d(y,x) ~~~~and~~~~ d(x,x)&lt; d(x,y).
\end{aligned}   \tag{1}\]</span>
Note that <span class="math inline">\(d\)</span> need not necessarily be symmetric nor positive (see (i)
below).</p>
<p>We begin with a concept formalizing locality to a pair. For any pair
<span class="math inline">\((x, y)\in \mathcal{S}\times \mathcal{S}\)</span>, the <em>local focus</em>, <span class="math inline">\(U_{x,y}\)</span>
(illustrated in Figure <a href="#fig:paldmethod" data-reference-type="ref" data-reference="fig:paldmethod">1</a> for two-dimensional Euclidean data) is
defined via</p>
<p><span class="math display" id="eq:localfocusdefn">\[\begin{aligned}
U_{x,y} \stackrel{\text{def}}{=} \{ z \in S \mid d(z,x) \leq d(y,x) \text{ or } d(z,y) \leq d(x,y) \}. \label{localfocusdefn}
\end{aligned}   \tag{2}\]</span></p>
<p>The set <span class="math inline">\(U_{x,y}\)</span> is comprised of elements that are “locally” relevant
to the pair <span class="math inline">\((x,y)\)</span>. Eq. <a href="#eq:theonereq">(1)</a> guarantees that both <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> are elements of
<span class="math inline">\(U_{x,y}\)</span>. For discussion from a social perspective, wherein the data
are embedded in a latent social space, see Social Framework in
1. Recall that in <a href="#eq:localfocusdefn">(2)</a>,
the distance <span class="math inline">\(d\)</span> need not be symmetric;
leveraging the social perspective, we are interested in the direct
closeness of <span class="math inline">\(z\)</span> <em>to</em> <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span>, not the reverse (see also (iv),
below).</p>
<figure id="fig:paldmethod">
<img src="pald-1.jpg" style="width:100.0%" alt="graphic without alt text" />
<figcaption>Figure 1: The local focus for two points, <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span>, where
<span class="math inline">\(\mathcal{S}\)</span> is a subset of <span class="math inline">\(\mathbb{R}^2\)</span>, under Euclidean distance
(reproduced with permission from <span class="citation" data-cites="berenhaut2024generalized">Berenhaut et al. (<a href="#ref-berenhaut2024generalized" role="doc-biblioref">2024</a>)</span>). The points
in red are outside the focus, <span class="math inline">\(U_{x,y}\)</span>. Those in green (and <span class="math inline">\(z\)</span> in
blue) are in the focus and closer to <span class="math inline">\(x\)</span>, while those in gray are closer
to <span class="math inline">\(y\)</span>.</figcaption>
</figure>
<p>Now, suppose that <span class="math inline">\(x\)</span> is fixed, and select <span class="math inline">\(Y\)</span> and <span class="math inline">\(Z\)</span> uniformly at
random from <span class="math inline">\(\mathcal{S}\setminus \{x\}\)</span> and <span class="math inline">\(U_{x,Y}\)</span>, respectively.
The <em>local depth</em> of <span class="math inline">\(x\)</span>, <span class="math inline">\(\ell_{\mathcal{S}}(x)\)</span>, is defined as the
probability that <span class="math inline">\(Z\)</span> is closer to <span class="math inline">\(x\)</span> than it is to <span class="math inline">\(Y\)</span> (with a coin
flip in the case <span class="math inline">\(Z\)</span> is equidistant), i.e.</p>
<p><span class="math display" id="eq:localdepthdefn">\[\begin{aligned}
\ell_{\mathcal{S}}(x)\stackrel{\mathrm{def}}{=}P(d(Z,x) &lt; d(Z,Y)) \label{localdepthdefn}.
\end{aligned}   \tag{3}\]</span></p>
<p>The local depth of <span class="math inline">\(x\)</span> can be thought of as a measure of the extent to
which <span class="math inline">\(x\)</span> is (relatively) surrounded by other elements of <span class="math inline">\(\mathcal{S}\)</span>
(under <span class="math inline">\(d\)</span>).</p>
<p>Finally, the <em>cohesion</em>, <span class="math inline">\(C_{x,w}\)</span>, of <span class="math inline">\(w\)</span> to <span class="math inline">\(x\)</span>, for a given
<span class="math inline">\(w\in \mathcal{S}\)</span>, is obtained through a partitioning of the
probability defining local depth in <a href="#eq:localdepthdefn">(3)</a>.
In particular,</p>
<p><span class="math display" id="eq:cohesdefn">\[\begin{aligned}
C_{x,w} \stackrel{\mathrm{def}}{=}P(Z=w \text{ and } d(Z,x) &lt; d(Z,Y))
\label{cohesdefn}.
\end{aligned}   \tag{4}\]</span></p>
<p>The cohesion of <span class="math inline">\(w\)</span> to <span class="math inline">\(x\)</span> can be viewed as the contribution of <span class="math inline">\(w\)</span> to
the local depth of <span class="math inline">\(x\)</span>. Eq. <a href="#eq:theonereq">(1)</a> guarantees that when <span class="math inline">\(Z=x\)</span>, <span class="math inline">\(d(Z,x) &lt; d(Z,Y)\)</span>.
For some theoretical properties of cohesion, see 1.
Note that the sum of all cohesions is conserved at <span class="math inline">\(n/2\)</span>, where
<span class="math inline">\(n\stackrel{\mathrm{def}}{=}|\mathcal{S}|\)</span>; we thus have a constant mean
local depth of <span class="math inline">\(0.5\)</span> (independent of <span class="math inline">\(n\)</span>).</p>
<p>A universal threshold for strong cohesion (see 1) is
defined via</p>
<p><span class="math display" id="eq:Tsd0">\[\begin{aligned}
T \stackrel{\mathrm{def}}{=}P(Z=W \text{ and } d(Z,X)&lt;d(Z,Y)),
\label{Tsd0}
\end{aligned}   \tag{5}\]</span></p>
<p>where <span class="math inline">\(X\)</span>, <span class="math inline">\(Y\)</span>, <span class="math inline">\(Z\)</span>, and <span class="math inline">\(W\)</span> are selected uniformly at random from
<span class="math inline">\(\mathcal{S}\)</span>, <span class="math inline">\(\mathcal{S}\setminus X\)</span>, <span class="math inline">\(U_{X,Y}\)</span>, and <span class="math inline">\(U_{X,Y}\)</span>,
respectively. The threshold, <span class="math inline">\(T\)</span>, can be conveniently calculated as one
half the average of the diagonal of the matrix of pairwise cohesion
values (see 1 for details), i.e.
<span class="math display" id="eq:Tsd">\[T = \frac{1}{2} \left( \frac{1}{n} \sum_{x\in \mathcal{S}} C_{x,x} \right).
\label{Tsd}   \tag{6}\]</span>
A strong relationship connection between <span class="math inline">\(x\)</span> and <span class="math inline">\(w\)</span> is then established
when:
<span class="math display" id="eq:Tsd2">\[\min \{ C_{x,w}, C_{w,x} \} \geq T.
\label{Tsd2}   \tag{7}\]</span></p>
<p>For further details on PaLD, including theoretical results, discussion
of the underlying social perspective and applications, see
<span class="citation" data-cites="berenhaut2022social berenhaut2022SIAM">Berenhaut et al. (<a href="#ref-berenhaut2022social" role="doc-biblioref">2022</a>; see also <a href="#ref-berenhaut2022SIAM" role="doc-biblioref">Berenhaut and Moore 2022</a>)</span>.</p>
<p>Before moving on to a simple concrete example, we provide a few remarks,
for the interested reader, on the concepts introduced above.</p>
<ol type="i">
<li><p>(Distances and dissimilarities) For the purposes of the package, we
require only the inequalities in <a href="#eq:theonereq">(1)</a>.
This implies the convenient form of the
threshold in <a href="#eq:Tsd2">(7)</a> <span class="citation" data-cites="berenhaut2022social">(see <a href="#ref-berenhaut2022social" role="doc-biblioref">Berenhaut et al. 2022</a>)</span>, and is crucial to
the underlying social perspective of conflict. For a more general
mathematical PaLD framework, wherein the concepts of locality and
support are further refined <span class="citation" data-cites="berenhaut2024generalized">(<a href="#ref-berenhaut2024generalized" role="doc-biblioref">Berenhaut et al. 2024</a>)</span>. As the
definitions in <a href="#eq:localfocusdefn">(2)</a>, <a href="#eq:localdepthdefn">(3)</a> and <a href="#eq:cohesdefn">(4)</a>
only depend on triplet distance comparisons,
local depth and cohesion are invariant under monotone
transformations of distance, and negative distances are fine. If
meaningful in context, distances need not be symmetric and
self-distances, <span class="math inline">\(d(x,x)\)</span>, may vary over <span class="math inline">\(x\)</span> (subject to the
constraint in <a href="#eq:theonereq">(1)</a>. The triangle inequality may be violated to
some extent, mirroring ideas of strong triadic closure
<span class="citation" data-cites="granovetter1973strength">(<a href="#ref-granovetter1973strength" role="doc-biblioref">Granovetter 1973</a>)</span>.</p></li>
<li><p>(Local foci) The concept of local foci in
<a href="#eq:localfocusdefn">(2)</a> is crucial to all that follows. It
allows for consideration of the data at varying scales without the
need for localizing parameters. The selection of <span class="math inline">\(Z\)</span> in
<a href="#eq:localdepthdefn">(3)</a> and <a href="#eq:cohesdefn">(4)</a>
reflects the larger presence of individuals
in smaller foci (see <span class="citation" data-cites="feld1981focused">Feld (<a href="#ref-feld1981focused" role="doc-biblioref">1981</a>)</span> for discussion is social
settings). The results in 1 regarding limiting
irrelevance of density and separation under increasing
concentration are driven by local comparisons. Self-cohesions
(i.e., diagonal elements of <span class="math inline">\(C\)</span>) are not necessarily equal. This is
driven by the fact that by <a href="#eq:cohesdefn">(4)</a>,
<span class="math display">\[\begin{aligned}
C_{x,x}=\frac{1}{n-1} \sum_{y\neq x} \frac{1}{|{U_{x,y}}|},
\end{aligned}\]</span>
and <span class="math inline">\(C_{x,x}\)</span> is dependent on local density around <span class="math inline">\(x\)</span>.</p></li>
<li><p>(Non-monotonicity of the size of local foci) As distance away from
an element, <span class="math inline">\(x\)</span>, increases, monotonicity in volume of local foci
is not necessary. In particular, consider the one-dimensional set
<span class="math inline">\(\mathcal{S}=\{B,x,A,C,D\}\)</span>, with <span class="math inline">\(B=5\)</span>, <span class="math inline">\(x=11\)</span>, <span class="math inline">\(A=16\)</span>, <span class="math inline">\(C=18\)</span>,
<span class="math inline">\(D=19\)</span>. Under Euclidean distance, <span class="math inline">\(d(x,A)=5&lt;6=d(x,B)\)</span> but
<span class="math inline">\(|U_{x,A}|=|\{x,A,C,D\}|=4\)</span> and <span class="math inline">\(|U_{x,B}|=|\{x,B,A\}|=3\)</span>. This
further emphasizes the capturing of local density, an important
feature of PaLD.</p></li>
<li><p>(Assymmetry) Note that the definitions in
<a href="#eq:localfocusdefn">(2)</a>, <a href="#eq:localdepthdefn">(3)</a> and <a href="#eq:cohesdefn">(4)</a>
are stated to allow for asymmetric
distances, as may occur for instance when considering distances on
graphs. Note that even symmetric distances can easily lead to
asymmetric cohesion. This is the case even for the simple
one-dimensional data considered in Example 1. In considering strong
relationships based on cohesion, though, symmetrization is employed
in <a href="#eq:Tsd2">(7)</a>.</p></li>
<li><p>(The use of weighted networks) In what follows, we will at times
have occasion to consider a network with node set <span class="math inline">\(\mathcal{S}\)</span> and
pairwise edges weighted via the respective values of cohesion. When
viewed as a weighted (connectivity) network in this sense,</p>
<ol type="1">
<li><p>the weighted nodal degrees are the values of local depth,
addressing considerations of (local) data depth,</p></li>
<li><p>the network may be embedded in low-dimensional Euclidean space
via standard network embedding techniques, providing
visualization of the data that adapts to relative density,</p></li>
<li><p>the isolated nodes of either the original network or that
restricted to strong ties can provide information regarding
outliers,</p></li>
<li><p>the connected components of the network restricted to strong
ties (i.e. those with weights above the threshold in
<a href="#eq:Tsd">(6)</a> can be
taken as “clusters" in the classical sense,</p></li>
<li><p>weighted or unweighted network neighbors may be used in place of
<span class="math inline">\(k\)</span>-nearest neighbors in settings such as classification and
smoothing, and</p></li>
<li><p>some network analyses can be informative regarding data
structure.</p></li>
</ol>
<p>Ideas of distance on networks can be complex, though, and this is
particularly so for general weighted networks (e.g. correlation or
social networks). As is the case here, edge weights may reflect
intensity of connection as opposed to simple pairwise distance, and
it is important to exercise caution when applying naïve functions
intended for unweighted networks (or simple distance-based weighted
networks) in <a href="https://CRAN.R-project.org/package=igraph"><strong>igraph</strong></a>
and other software.</p></li>
<li><p>(Matrix methods) For convenience, we use the
<a href="https://CRAN.R-project.org/package=igraph"><strong>igraph</strong></a> package to
analyze weighted networks, where appropriate. Matrix-based methods
are also possible. For instance, we employ <code>igraph::components</code> to
obtain community clusters from networks, but alternatively one
could use blocks extraction directly from the cohesion matrix using
packages such as
<a href="https://CRAN.R-project.org/package=lintools"><strong>lintools</strong></a>.</p></li>
<li><p>(Density-based methods and localized approaches) Complimentary
density-based methods such as DBSCAN (density-based spatial
clustering of applications with noise) and its hierarchical
variant HDBSCAN <span class="citation" data-cites="campello2020density">(see for instance <a href="#ref-campello2020density" role="doc-biblioref">Campello et al. 2020</a>)</span>, seek to
identify high-density regions as clusters. We discuss these in the
context of a concrete example below (see Cultural and
Psychological distance analysis). Determination of required
parameter values can be challenging. There are also extent methods
for considering data depth which probe local structure
<span class="citation" data-cites="paindaveine2013depth agostinelli2011local">(<a href="#ref-agostinelli2011local" role="doc-biblioref">Agostinelli and Romanazzi 2011</a>; <a href="#ref-paindaveine2013depth" role="doc-biblioref">Paindaveine and Van Bever 2013</a>)</span>. These often
also require localizing parameters. In <span class="citation" data-cites="berenhaut2022social">(<a href="#ref-berenhaut2022social" role="doc-biblioref">Berenhaut et al. 2022</a>)</span>
(see Theorem 2: Limiting Irrelevance of Density), it is shown that local depth and
cohesion account for varying density in the sense that, provided
subsets are sufficiently separated, the cohesion is maintained, as
within-subset distances are contracted or dilated, without the
need to search over a parameter space.</p></li>
<li><p>(Social latent spaces) The definitions in
<a href="#eq:localfocusdefn">(2)</a>, <a href="#eq:localdepthdefn">(3)</a>, <a href="#eq:cohesdefn">(4)</a>,
and <a href="#eq:Tsd">(6)</a> are
developed from a social perspective. The interested reader can
refer to the section <em>Social Framework</em> in</p>
<ol class="example" type="1">
<li>For further discussion see
<span class="citation" data-cites="berenhaut2024generalized">Berenhaut et al. (<a href="#ref-berenhaut2024generalized" role="doc-biblioref">2024</a>)</span> and <span class="citation" data-cites="berenhaut2022SIAM">Berenhaut and Moore (<a href="#ref-berenhaut2022SIAM" role="doc-biblioref">2022</a>)</span>.</li>
</ol></li>
</ol>
<p>As a concrete example demonstrating the PaLD framework, consider the
one-dimensional set
<span class="math inline">\(\mathcal{S}\stackrel{\mathrm{def}}{=}\{1,3,7,8,9,13,17\}\)</span>, with
pairwise (Euclidean) distances given in the array below.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>    <span class="dv">1</span>  <span class="dv">3</span>  <span class="dv">7</span> <span class="dv">8</span> <span class="dv">9</span> <span class="dv">13</span> <span class="dv">17</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>   <span class="dv">0</span>  <span class="dv">2</span>  <span class="dv">6</span> <span class="dv">7</span> <span class="dv">8</span> <span class="dv">12</span> <span class="dv">16</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>   <span class="dv">2</span>  <span class="dv">0</span>  <span class="dv">4</span> <span class="dv">5</span> <span class="dv">6</span> <span class="dv">10</span> <span class="dv">14</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="dv">7</span>   <span class="dv">6</span>  <span class="dv">4</span>  <span class="dv">0</span> <span class="dv">1</span> <span class="dv">2</span>  <span class="dv">6</span> <span class="dv">10</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="dv">8</span>   <span class="dv">7</span>  <span class="dv">5</span>  <span class="dv">1</span> <span class="dv">0</span> <span class="dv">1</span>  <span class="dv">5</span>  <span class="dv">9</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="dv">9</span>   <span class="dv">8</span>  <span class="dv">6</span>  <span class="dv">2</span> <span class="dv">1</span> <span class="dv">0</span>  <span class="dv">4</span>  <span class="dv">8</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="dv">13</span> <span class="dv">12</span> <span class="dv">10</span>  <span class="dv">6</span> <span class="dv">5</span> <span class="dv">4</span>  <span class="dv">0</span>  <span class="dv">4</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="dv">17</span> <span class="dv">16</span> <span class="dv">14</span> <span class="dv">10</span> <span class="dv">9</span> <span class="dv">8</span>  <span class="dv">4</span>  <span class="dv">0</span></span></code></pre></div>
<p>The local depths (rounded to three decimal places) are</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>    <span class="dv">1</span>     <span class="dv">3</span>     <span class="dv">7</span>     <span class="dv">8</span>     <span class="dv">9</span>    <span class="dv">13</span>    <span class="dv">17</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fl">0.409</span> <span class="fl">0.463</span> <span class="fl">0.588</span> <span class="fl">0.675</span> <span class="fl">0.600</span> <span class="fl">0.471</span> <span class="fl">0.294</span></span></code></pre></div>
<p>while the pairwise cohesions (in matrix form) are given by</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>     <span class="dv">3</span>     <span class="dv">7</span>     <span class="dv">8</span>     <span class="dv">9</span>    <span class="dv">13</span>    <span class="dv">17</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span>  <span class="fl">0.210</span> \tr{<span class="fl">0.127</span>} <span class="fl">0.036</span> <span class="fl">0.024</span> <span class="fl">0.012</span> <span class="fl">0.000</span> <span class="fl">0.000</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span>  \tr{<span class="fl">0.137</span>} <span class="fl">0.220</span> <span class="fl">0.048</span> <span class="fl">0.036</span> <span class="fl">0.024</span> <span class="fl">0.000</span> <span class="fl">0.000</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="dv">7</span>  <span class="fl">0.048</span> <span class="fl">0.048</span> <span class="fl">0.220</span> \tr{<span class="fl">0.137</span>} \tr{<span class="fl">0.109</span>} <span class="fl">0.028</span> <span class="fl">0.000</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="dv">8</span>  <span class="fl">0.024</span> <span class="fl">0.052</span> \tr{<span class="fl">0.163</span>} <span class="fl">0.218</span> \tr{<span class="fl">0.163</span>} <span class="fl">0.056</span> <span class="fl">0.000</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="dv">9</span>  <span class="fl">0.024</span> <span class="fl">0.024</span> \tr{<span class="fl">0.109</span>} \tr{<span class="fl">0.137</span>} <span class="fl">0.220</span> <span class="fl">0.063</span> <span class="fl">0.024</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="dv">13</span> <span class="fl">0.000</span> <span class="fl">0.000</span> <span class="fl">0.012</span> <span class="fl">0.036</span> <span class="fl">0.103</span> <span class="fl">0.188</span> \tr{<span class="fl">0.133</span>}</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="dv">17</span> <span class="fl">0.000</span> <span class="fl">0.000</span> <span class="fl">0.000</span> <span class="fl">0.000</span> <span class="fl">0.012</span> \tr{<span class="fl">0.107</span>} <span class="fl">0.175</span></span></code></pre></div>
<p>where (non-diagonal) cohesion values above the threshold of
<span class="math display">\[\begin{aligned}
0.1036=(1/14) (0.210+0.220+0.220+0.218+0.220+0.188+0.175)
\end{aligned}\]</span>
are indicated in blue; see <a href="#eq:Tsd2">(7)</a>.
The resulting weighted network of pairwise cohesion
values is displayed in Figure <a href="#fig:figsimple" data-reference-type="ref" data-reference="fig:figsimple">2</a>; here strong ties (with cohesion values above
the threshold) are colored according to community clusters (the
connected components of the network of strong ties), i.e. <span class="math inline">\(\{1,3\}\)</span>,
<span class="math inline">\(\{7,8,9\}\)</span> and <span class="math inline">\(\{13,17\}\)</span>.</p>
<figure id="fig:figsimple">
<img src="pald-2.png" style="width:100.0%" alt="Figure 2: The community cluster network for the simple one-dimensional data considered in Example 1. Strong ties (with cohesion values above the threshold) are colored according to community clusters (the connected components of the network of strong ties), while weak ties are indicated in grey." />
<figcaption aria-hidden="true">Figure 2: The community cluster network for the simple one-dimensional
data considered in Example 1. Strong ties (with cohesion values above
the threshold) are colored according to community clusters (the
connected components of the network of strong ties), while weak ties are
indicated in grey.</figcaption>
</figure>
<p>Note that the cohesion of the element 17 to the element 1 is zero. This
reflects the fact that when an element, <span class="math inline">\(Y\)</span>, other than 1 is selected
uniformly at random from <span class="math inline">\(\mathcal{S}\)</span>, and an element <span class="math inline">\(Z\)</span> is selected
uniformly at random from <span class="math inline">\(U_{x,Y}\)</span>, 17 is never, at once, equal to <span class="math inline">\(Z\)</span>
and closer to 1 than it is to the opposing element <span class="math inline">\(Y\)</span>. In fact, 17 is
not an element of <span class="math inline">\(U_{1,y}\)</span>, for <span class="math inline">\(y\in \{3,7,8\}\)</span>. For
<span class="math inline">\(y\in \{9,13,17\}\)</span>, <span class="math inline">\(17\in U_{1,y}\)</span>, but in each case,
<span class="math inline">\(d(17,y)&lt;d(17,1)\)</span>. This zero cohesion is reflected in the absence of any
edge (grey or colored) from the node for 17 to the node for 1, in Figure
<a href="#fig:figsimple" data-reference-type="ref" data-reference="fig:figsimple">2</a>. ◻</p>
<p>We now turn to discussion of the
<a href="https://CRAN.R-project.org/package=pald"><strong>pald</strong></a> package.</p>
<h4 class="unnumbered" data-number="0.3" id="pald">pald</h4>
<p>The main functions in the
<a href="https://CRAN.R-project.org/package=pald"><strong>pald</strong></a> package can be split
into 3 categories:</p>
<ol type="1">
<li><p>A function for computing the matrix of cohesion values, as defined
in <a href="#eq:cohesdefn">(4)</a>.</p></li>
<li><p>Functions for extracting useful information from the cohesion
matrix, such as local depths, neighbors, community clusters, and
graph objects.</p></li>
<li><p>Plotting functions for community graphs.</p></li>
</ol>
<p>In addition, the package provides a number of pertinent example data
sets which may be used to demonstrate community analysis, including a
synthetic data set of two-dimensional points created by
<span class="citation" data-cites="gionis1clustering">Gionis et al.</span> to demonstrate aggregation, clustering data generated
from the scikit-learn Python package <span class="citation" data-cites="pedregosa2011scikit">(<a href="#ref-pedregosa2011scikit" role="doc-biblioref">Pedregosa et al. 2011</a>)</span>, data
describing cognate relationships between words across 87 Indo-European
languages <span class="citation" data-cites="dyen92">(<a href="#ref-dyen92" role="doc-biblioref">Dyen et al. 1992</a>)</span>, data compiled by <span class="citation" data-cites="tissue">Love and Irizarry (<a href="#ref-tissue" role="doc-biblioref">2015</a>)</span> of tissue gene
expressions, data employing information from the World Values Survey
<span class="citation" data-cites="inglehart2014world">(<a href="#ref-inglehart2014world" role="doc-biblioref">Inglehart et al. 2014</a>)</span> on cultural values regarding family, religion,
education, and institutions for several regions
<span class="citation" data-cites="muthukrishna2020beyond">(<a href="#ref-muthukrishna2020beyond" role="doc-biblioref">Muthukrishna et al. 2020</a>)</span>, and three example data sets generated for the
1 paper.</p>
<p>While it is not a necessity, the
<a href="https://CRAN.R-project.org/package=pald"><strong>pald</strong></a> package is designed
to function well with the pipe operator, <code>&gt;</code>, particularly for those who
are familiar with tidyverse-based approaches. This functionality will be
demonstrated briefly below.</p>
<h5 class="unnumbered" data-number="0.3.1" id="creating-the-cohesion-matrix">Creating the cohesion matrix</h5>
<p>For the purposes of the
<a href="https://CRAN.R-project.org/package=pald"><strong>pald</strong></a> package, the sole
input for the Partitioned Local Depth (PaLD) computations is a distance
matrix or <code>dist</code> object. Recall that the collection of input distances
(or dissimilarities) is assumed to satisfy the requirements in
<a href="#eq:theonereq">(1)</a>. More generally, the method only requires
triplet distance comparisons, as opposed to exact numeric distances (see
1).</p>
<p>For demonstration purposes, we first show how one can compute a distance
matrix from an input data frame with, say, two variables <code>x1</code> and <code>x2</code>.
The input data may be of any dimension; in fact the PaLD framework
provides advantages when considering high-dimensional data (see the
<strong>Examples</strong> section as well as 1).</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pald)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">x1 =</span> <span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">8</span>, <span class="dv">8</span>, <span class="dv">16</span>, <span class="dv">4</span>, <span class="dv">14</span>),</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">x2 =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">4</span>, <span class="dv">10</span>, <span class="dv">8</span>, <span class="dv">4</span>, <span class="dv">10</span>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;B&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;D&quot;</span>, <span class="st">&quot;E&quot;</span>, <span class="st">&quot;F&quot;</span>)</span></code></pre></div>
<p>The <code>dist</code> function returns a (default Euclidean) pairwise distance
matrix for an input data frame, as demonstrated below. If the data are
already provided as a distance matrix (or <code>dist</code> object), the user can
skip to the next step. Note that the distance matrix needed for the
subsequent functions does not need to be a <code>dist</code> object and <em>need not</em>
be symmetric.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">dist</span>(df)</span></code></pre></div>
<p>The function above creates a <code>dist</code> object. If converted to a matrix,
this will be an <span class="math inline">\(n\times n\)</span> distance matrix, where <span class="math inline">\(n\)</span> corresponds to
the number of observations in the original data frame (in this example
<span class="math inline">\(n = 6\)</span>).</p>
<p>This <code>dist</code> object, or a distance matrix, can then be passed to the
<code>cohesion_matrix</code> function in order to calculate pairwise cohesion
values.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cohesion_matrix</span>(d)</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;            A          B          C         D          E         F</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; A 0.25000000 0.18333333 0.06666667 0.0000000 0.18333333 0.0000000</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; B 0.14000000 0.24000000 0.05000000 0.0000000 0.10666667 0.0000000</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; C 0.07333333 0.07333333 0.20333333 0.0000000 0.03333333 0.0800000</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; D 0.00000000 0.00000000 0.00000000 0.2333333 0.00000000 0.1333333</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; E 0.14000000 0.10666667 0.03333333 0.0000000 0.24000000 0.0000000</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; F 0.00000000 0.00000000 0.05000000 0.1400000 0.00000000 0.2400000</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;cohesion_matrix&quot; &quot;matrix&quot;          &quot;array&quot;</span></span></code></pre></div>
<p>Equivalently, the user can use the native pipe <code>&gt;</code> as follows.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>df <span class="sc">|&gt;</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist</span>() <span class="sc">|&gt;</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cohesion_matrix</span>()</span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;            A          B          C         D          E         F</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; A 0.25000000 0.18333333 0.06666667 0.0000000 0.18333333 0.0000000</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; B 0.14000000 0.24000000 0.05000000 0.0000000 0.10666667 0.0000000</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; C 0.07333333 0.07333333 0.20333333 0.0000000 0.03333333 0.0800000</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; D 0.00000000 0.00000000 0.00000000 0.2333333 0.00000000 0.1333333</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; E 0.14000000 0.10666667 0.03333333 0.0000000 0.24000000 0.0000000</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; F 0.00000000 0.00000000 0.05000000 0.1400000 0.00000000 0.2400000</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;cohesion_matrix&quot; &quot;matrix&quot;          &quot;array&quot;</span></span></code></pre></div>
<p>The <em>cohesion matrix</em> output by the <code>cohesion_matrix</code> function is the
main input for the majority of the remaining functions.</p>
<h5 class="unnumbered" data-number="0.3.2" id="functions-for-extracting-information-from-the-cohesion-matrix">Functions for extracting information from the cohesion matrix</h5>
<p>From the <em>cohesion matrix</em>, a variety of useful quantities can be
computed. Below, we create a cohesion matrix using the functions
described in the previous section.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>df <span class="sc">|&gt;</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist</span>() <span class="sc">|&gt;</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cohesion_matrix</span>() <span class="ot">-&gt;</span> cohesion</span></code></pre></div>
<p>The <code>local_depths</code> function calculates the <em>depth</em> of each point, as
defined in <a href="#eq:localdepthdefn">(3)</a>, outputting a vector of local depth
probabilities.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">local_depths</span>(cohesion)</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         A         B         C         D         E         F</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 0.6833333 0.5366667 0.4633333 0.3666667 0.5200000 0.4300000</span></span></code></pre></div>
<p>In this case, the (locally) deepest point is <code>A</code>.</p>
<p>The <code>strong_threshold</code> function will calculate the cohesion threshold
for strong ties given in <a href="#eq:Tsd0">(5)</a>, which reflects typical cohesion for local points
(see 1). Computationally, this is equal to half the
average of the diagonal of the cohesion matrix, and is a threshold that
may be used to distinguish between strong and weak ties.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">strong_threshold</span>(cohesion)</span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 0.1172222</span></span></code></pre></div>
<p>Here, the threshold is a little above <code>0.117</code>.</p>
<p>The function <code>cohesion_strong</code> will update the cohesion matrix to set
all weak ties to zero (via the <code>strong_threshold</code> function). Reflecting
<a href="#eq:Tsd2">(7)</a>, by default,
the matrix will also be symmetrized, using the entry-wise (parallel)
minimum of the cohesion matrix and its transpose.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cohesion_strong</span>(cohesion)</span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;      A    B         C         D    E         F</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; A 0.25 0.14 0.0000000 0.0000000 0.14 0.0000000</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; B 0.14 0.24 0.0000000 0.0000000 0.00 0.0000000</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; C 0.00 0.00 0.2033333 0.0000000 0.00 0.0000000</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; D 0.00 0.00 0.0000000 0.2333333 0.00 0.1333333</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; E 0.14 0.00 0.0000000 0.0000000 0.24 0.0000000</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; F 0.00 0.00 0.0000000 0.1333333 0.00 0.2400000</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; attr(,&quot;class&quot;)</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] &quot;cohesion_matrix&quot; &quot;matrix&quot;          &quot;array&quot;</span></span></code></pre></div>
<p>The <code>community_graphs</code> function takes the cohesion matrix and creates
<a href="https://CRAN.R-project.org/package=igraph"><strong>igraph</strong></a> objects, graphs
that describe the (symmetrized) relationship structure between points.
This function will output a list of three objects:</p>
<ul>
<li><p><code>G</code>: the weighted (community) graph whose edge weights are mutual
cohesion</p></li>
<li><p><code>G_strong</code>: the weighted (community) graph consisting of edges for
which mutual (symmetrized) cohesion (i.e. the minimum of the two
directed cohesion values for any given pair) is greater than the
threshold for strong ties</p></li>
<li><p><code>layout</code>: the graph layout. By default this is provided by the
Fruchterman Reingold (FR) force-directed graph drawing algorithm for
the graph <code>G</code>, as implemented in the
<a href="https://CRAN.R-project.org/package=igraph"><strong>igraph</strong></a> package.</p></li>
</ul>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>graphs <span class="ot">&lt;-</span> <span class="fu">community_graphs</span>(cohesion)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>graphs[[<span class="st">&quot;G_strong&quot;</span>]]</span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; IGRAPH c4d81a7 UNW- 6 3 --</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; + attr: name (v/c), weight (e/n)</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; + edges from c4d81a7 (vertex names):</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] A--B A--E D--F</span></span></code></pre></div>
<p>Here we see that there are three connected components, ties <code>A-B</code> and
<code>A-E</code> form the first community cluster, and the tie <code>D-F</code> which forms
another.</p>
<p>The <code>any_isolated()</code> function will check whether there are any isolated
points (according to cohesion).</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">any_isolated</span>(cohesion)</span></code></pre></div>
<p>Here, there are no isolated points, i.e. points having zero cohesion
with all other points in the data (an extreme form of outlier).</p>
<p>The “community clusters” identified by PaLD are the connected components
of the graph of strong ties, <code>G_strong</code>. To directly calculate these, we
can use the <code>community_clusters</code> function. This will output a data frame
with two columns; the first corresponds to the individual (<code>point</code>), as
identified by the row name of the original input data frame, <code>df</code>, the
second identifies the <code>community</code> that the individual belongs to.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">community_clusters</span>(cohesion)</span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   point community</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; A     A         1</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; B     B         1</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; C     C         2</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; D     D         3</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; E     E         1</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; F     F         3</span></span></code></pre></div>
<p>In this example, three communities are identified with these six points.
Points <code>A</code>, <code>B</code>, and <code>E</code> fall into Community 1. Point <code>C</code> is in
Community 2 (a community of size 1) and points <code>D</code> and <code>F</code> fall into
Community 3.</p>
<h4 class="unnumbered" data-number="0.4" id="plotting-functions">Plotting functions</h4>
<p>The final category of function is that for data visualization. We can
begin by visualizing the points in the data frame <code>df</code> (Figure
<a href="#fig:fig1" data-reference-type="ref" data-reference="fig:fig1">3</a>). When
visualizing these points, it is important to have the aspect ratio of
the x and y axes equal to 1 so as to not distort distances. When using
the <a href="https://CRAN.R-project.org/package=ggplot2"><strong>ggplot2</strong></a> package
for this visualization, one can use the command
<code>coord_fixed(ratio = 1)</code>. If using the <code>plot</code> function included in the
base library, one can use the <code>asp = 1</code> argument.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(x1, x2)) <span class="sc">+</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">label =</span> <span class="fu">rownames</span>(df)) <span class="sc">+</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>(<span class="at">ratio =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">16</span>)) <span class="sc">+</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">16</span>))</span></code></pre></div>
<figure id="fig:fig1">
<img src="dagostino-mcgowan_files/figure-latex/fig1-1.png" style="width:100.0%" alt="Figure 3: Visualization of the points from the data frame df" />
<figcaption aria-hidden="true">Figure 3: Visualization of the points from the data frame
<code>df</code></figcaption>
</figure>
<p>We can pass the cohesion matrix to the <code>plot_community_graphs</code> function
to view the relationship between points (Figure
<a href="#fig:fig2" data-reference-type="ref" data-reference="fig:fig2">4</a>). The function
will also permit parameters that can be passed to <code>plot.igraph</code> via the
<code>...</code> argument.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_community_graphs</span>(cohesion,</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">vertex.label.cex =</span> <span class="dv">2</span>,</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">vertex.label.dist =</span> <span class="fl">0.9</span>)</span></code></pre></div>
<figure id="fig:fig2">
<img src="dagostino-mcgowan_files/figure-latex/fig2-1.png" style="width:100.0%" alt="Figure 4: PaLD graph displaying the relationship between the points in data frame df" />
<figcaption aria-hidden="true">Figure 4: <span id="fig:fig2" data-label="fig:fig2"></span>PaLD graph displaying the
relationship between the points in data frame
<code>df</code></figcaption>
</figure>
<p>Notice in this plot the force-directed layout does not match that of the
original data frame as seen in Figure
<a href="#fig:fig1" data-reference-type="ref" data-reference="fig:fig1">3</a>. Since our
original data is two-dimensional, it may be reasonable to use the latter
as the layout for plotting. Figure <a href="#fig:fig3" data-reference-type="ref" data-reference="fig:fig3">5</a> includes this update as well as others addressing
some of the aesthetics, such as employing more readable labels. The
<code>layout</code> argument allows the user to pass a matrix to dictate the
2-dimensional layout of the graph. For example, if we wanted the graph
to match the visualization displayed in Figure
<a href="#fig:fig1" data-reference-type="ref" data-reference="fig:fig1">3</a>, we could pass
<code>as.matrix(df)</code> (a matrix of the data frame <code>df</code>) to the <code>layout</code>
argument (Figure <a href="#fig:fig3" data-reference-type="ref" data-reference="fig:fig3">5</a>). Here, we increase the vertex size and change the
vertex label color, through the argument specifications
<code>vertex.size = 100</code> and <code>vertex.label.color = "white"</code>. Additionally, to
allow axes, we use <code>axes = TRUE</code>, and to put these back on the original
scale we set <code>rescale = FALSE</code>, resetting the axis limits using <code>xlim</code>
and <code>ylim</code>. The <code>par(pty = "s")</code> function forces the subsequent plot to
be square.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">pty =</span> <span class="st">&quot;s&quot;</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_community_graphs</span>(cohesion,</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">layout =</span> <span class="fu">as.matrix</span>(df),</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">vertex.size =</span> <span class="dv">100</span>,</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">vertex.label.color =</span> <span class="st">&quot;white&quot;</span>,</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>                      <span class="at">axes =</span> <span class="cn">TRUE</span>,</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>                      <span class="at">rescale =</span> <span class="cn">FALSE</span>,</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>                      <span class="at">asp =</span> <span class="dv">1</span>,</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>                      <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">16</span>),</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>                      <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">16</span>))</span></code></pre></div>
<figure id="fig:fig3">
<img src="dagostino-mcgowan_files/figure-latex/fig3-1.png" style="width:100.0%" alt="Figure 5: PaLD graph displaying the relationship between the points in data frame df, matching the original layout in Figure 3" />
<figcaption aria-hidden="true">Figure 5: <span id="fig:fig3" data-label="fig:fig3"></span>PaLD graph displaying the
relationship between the points in data frame <code>df</code>, matching the
original layout in Figure <a href="#fig:fig1" data-reference-type="ref" data-reference="fig:fig1">3</a></figcaption>
</figure>
<h4 class="unnumbered" data-number="0.5" id="examples">Examples</h4>
<p>We will demonstrate the utility of the
<a href="https://CRAN.R-project.org/package=pald"><strong>pald</strong></a> package through
several illustrative examples.</p>
<h5 class="unnumbered" data-number="0.5.1" id="community-analysis-for-tissue-gene-expression-data">Community analysis for tissue gene expression data</h5>
<p>The first example is from a subset of data from <span class="citation" data-cites="zilliox2007gene">Zilliox and Irizarry (<a href="#ref-zilliox2007gene" role="doc-biblioref">2007</a>)</span>,
<span class="citation" data-cites="mccall2011gene">McCall et al. (<a href="#ref-mccall2011gene" role="doc-biblioref">2011</a>)</span>, and <span class="citation" data-cites="mccall2014gene">McCall et al. (<a href="#ref-mccall2014gene" role="doc-biblioref">2014</a>)</span>, obtained from the
<strong>tissuesGeneExpression</strong> bioconductor package <span class="citation" data-cites="tissue">(<a href="#ref-tissue" role="doc-biblioref">Love and Irizarry 2015</a>)</span> consisting of
22,215-dimensional gene expression data from 189 tissue samples. A
(Euclidean) <code>dist</code> object was created using this data set and is
included in the <a href="https://CRAN.R-project.org/package=pald"><strong>pald</strong></a>
package in an object called <code>tissue_dist</code>.</p>
<p>The <code>tissue_dist</code> object is a <code>dist</code> object resulting in a distance
matrix with 189 rows and 189 columns.</p>
<p>We can create the cohesion matrix using the <code>cohesion_matrix</code> function.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>tissue_cohesion <span class="ot">&lt;-</span> <span class="fu">cohesion_matrix</span>(tissue_dist)</span></code></pre></div>
<p>We can display relationships between tissue samples, both locally and
globally through the <code>plot_community_graphs</code> function (Figure
<a href="#fig:fig4" data-reference-type="ref" data-reference="fig:fig4">6</a>). For clarity
of the display, we show how to remove the labels using
<code>show_labels = FALSE</code>. We will instead color according to the labels by
passing these to the <code>vertex.color</code> argument for the <code>plot.igraph</code>
function (via the <code>...</code> argument). Similarly, we can add a legend using
the <code>legend()</code> function, as you would for an
<a href="https://CRAN.R-project.org/package=igraph"><strong>igraph</strong></a> visualization.
Additionally, we use the <code>edge_width_factor</code> and <code>emph_strong</code> arguments
to adjust the width of the lines between and within PaLD communities.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>labels <span class="ot">&lt;-</span> <span class="fu">rownames</span>(tissue_cohesion)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_community_graphs</span>(tissue_cohesion,</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">show_labels =</span> <span class="cn">FALSE</span>,</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">vertex.size =</span> <span class="dv">4</span>,</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">vertex.color =</span> <span class="fu">as.factor</span>(labels),</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">edge_width_factor =</span> <span class="dv">35</span>,</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>                      <span class="at">emph_strong =</span> <span class="dv">5</span>)</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">&quot;topleft&quot;</span>,</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">legend =</span> <span class="fu">unique</span>(<span class="fu">as.factor</span>(labels)),</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">pt.bg =</span> <span class="fu">unique</span>(<span class="fu">as.factor</span>(labels)),</span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="st">&quot;black&quot;</span>,</span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">pch =</span> <span class="dv">21</span>)</span></code></pre></div>
<figure id="fig:fig4">
<img src="fig5.png" style="width:100.0%" alt="Figure 6: Community cluster network for the tissue data. The line colors indicate the PaLD communities; the point colors indicate the tissue classification." />
<figcaption aria-hidden="true">Figure 6: Community cluster network for the tissue data. The line
colors indicate the PaLD communities; the point colors indicate the
tissue classification.</figcaption>
</figure>
<p>A summary of community strong ties can be produced via the following
code. Note that <code>tissue_graph_strong</code> is an
<a href="https://CRAN.R-project.org/package=igraph"><strong>igraph</strong></a> object
corresponding to the graph displayed in Figure
<a href="#fig:fig4" data-reference-type="ref" data-reference="fig:fig4">6</a>, restricted
only to strong ties.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>tissue_graphs <span class="ot">&lt;-</span> <span class="fu">community_graphs</span>(tissue_cohesion)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>tissue_graph_strong <span class="ot">&lt;-</span> tissue_graphs[[<span class="st">&quot;G_strong&quot;</span>]]</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>E <span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">get.edgelist</span>(tissue_graph_strong)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>tissue_strong_ties <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">strong_ties =</span> <span class="fu">apply</span>(E, <span class="dv">1</span>, paste, <span class="at">collapse =</span> <span class="st">&quot;,&quot;</span>)</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>tissue_strong_ties <span class="sc">|&gt;</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">table</span>() <span class="sc">|&gt;</span></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sort</span>(<span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; strong_ties</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;           kidney,kidney</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                     273</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;             colon,colon</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                     241</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; hippocampus,hippocampus</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                     191</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   cerebellum,cerebellum</span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                     186</span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;             liver,liver</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                      85</span></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; endometrium,endometrium</span></span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                      81</span></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;       placenta,placenta</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                       4</span></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;      kidney,endometrium</span></span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                       3</span></span></code></pre></div>
<p>Note that there are only three strong ties between different tissue
types (kidney and endometrium) in the community cluster network of 1,064
strong ties total.</p>
<h4 class="unnumbered" data-number="0.6" id="cognate-based-language-families">Cognate-based Language Families</h4>
<p>This example explores a data set from <span class="citation" data-cites="dyen92">Dyen et al. (<a href="#ref-dyen92" role="doc-biblioref">1992</a>)</span> that summarizes
relationships between 87 Indo-European languages from the perspective of
cognates, coded using 2,655-dimensional binary vectors. A <code>dist</code> object
was created from this data set and is included in the
<a href="https://CRAN.R-project.org/package=pald"><strong>pald</strong></a> package in an object
called <code>cognate_dist</code>.</p>
<p>Here we will demonstrate how one can further apply functions in the
<a href="https://CRAN.R-project.org/package=igraph"><strong>igraph</strong></a> package to
objects output from the
<a href="https://CRAN.R-project.org/package=pald"><strong>pald</strong></a> package. We can
first use the <code>cohesion_matrix</code> function to calculate the cohesion
matrix and the <code>community_graphs</code> function to create a list with the
weighted community graph, the weighted community graph with only strong
ties included, and the layout. From this, we can extract the graph with
only the strong ties, here called <code>cognate_graph_strong</code>.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>cognate_cohesion <span class="ot">&lt;-</span> <span class="fu">cohesion_matrix</span>(cognate_dist)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>cognate_graphs <span class="ot">&lt;-</span> <span class="fu">community_graphs</span>(cognate_cohesion)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>cognate_graph_strong <span class="ot">&lt;-</span> cognate_graphs[[<span class="st">&quot;G_strong&quot;</span>]]</span></code></pre></div>
<p>We can then use the <code>neighbors</code> function from the
<a href="https://CRAN.R-project.org/package=igraph"><strong>igraph</strong></a> package to
extract the strong neighbors in this graph. For example, we can extract
all neighbors for the language “French”, via the following code.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>french_neighbors <span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">neighbors</span>(cognate_graph_strong, <span class="st">&quot;French&quot;</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>french_neighbors</span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; + 8/87 vertices, named, from 5c6bb3a:</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] Italian</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [2] Ladin</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [3] Provencal</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [4] Walloon</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [5] French_Creole_C</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [6] French_Creole_D</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [7] Spanish</span></span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [8] Catalan</span></span></code></pre></div>
<p>Similarly, we can sort and print the associated neighborhood weights by
subsetting the cohesion matrix.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>cognate_cohesion[<span class="st">&quot;French&quot;</span>, french_neighbors] <span class="sc">|&gt;</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sort</span>(<span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         Walloon</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;      0.03258771</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;       Provencal</span></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;      0.02871174</span></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; French_Creole_C</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;      0.02406057</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; French_Creole_D</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;      0.02406057</span></span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;           Ladin</span></span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;      0.02094596</span></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         Italian</span></span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;      0.01997696</span></span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         Catalan</span></span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;      0.01859688</span></span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         Spanish</span></span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;      0.01679733</span></span></code></pre></div>
<p>We can again use the <code>plot_community_graphs</code> function to visualize the
community clusters (Figure <a href="#fig:figlang" data-reference-type="ref" data-reference="fig:figlang">7</a>). One may note the commonly identifiable
language clusters and that, under a slight rotation, some of the
underlying geography is mirrored in the plot.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_community_graphs</span>(</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  cognate_cohesion,</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">edge_width_factor =</span> <span class="dv">30</span>,</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">emph_strong =</span> <span class="dv">3</span>,</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">vertex.size =</span> <span class="dv">3</span>,</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">vertex.label.cex =</span> <span class="fl">0.7</span>,</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">vertex.label.dist =</span> <span class="dv">1</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<figure id="fig:figlang">
<img src="fig6.png" style="width:100.0%" alt="Figure 7: Community structure for 87 Indo-European languages, which employs cognate information that was coded via 2,665-dimensional binary vectors. Commonly identifiable language clusters arise along with informative inter- and intra-cluster structure. Several ancient languages are centrally located." />
<figcaption aria-hidden="true">Figure 7: Community structure for 87 Indo-European languages, which
employs cognate information that was coded via 2,665-dimensional binary
vectors. Commonly identifiable language clusters arise along with
informative inter- and intra-cluster structure. Several ancient
languages are centrally located.</figcaption>
</figure>
<h5 class="unnumbered" data-number="0.6.1" id="community-analysis-for-generated-data">Community analysis for generated data</h5>
<p>The <a href="https://CRAN.R-project.org/package=pald"><strong>pald</strong></a> package includes
three randomly generated data frames corresponding to plots from
1:</p>
<ul>
<li><p><code>exdata1</code> is a data set consisting of 8 points used to create Figure
1 in 1</p></li>
<li><p><code>exdata2</code> is a data set consisting of 16 points used to create
Figure 2 in 1</p></li>
<li><p><code>exdata3</code> is a data set consisting of 240 points used to create
Figure 4D in 1</p></li>
</ul>
<p>Here, we will demonstrate how to use <code>exdata3</code>. These points were
generated from bivariate normal distributions with varying means and
variances. There are eight “true” communities.</p>
<p>We can contrast resulting community clusters obtained via PaLD
(i.e. connected components of the network of strong ties) with clusters
as obtained through common cluster analysis techniques. Here we will
consider two common clustering methods. The code below calculates the
cohesion matrix (<code>exdata_cohesion</code>) as well as the community clusters
obtained via PaLD (<code>exdata_pald</code>), along with <span class="math inline">\(k=8\)</span> clusters obtained
via <em>k</em>-means (<code>exdata_kmeans</code>) and hierarchical clustering using
complete linkage (<code>exdata_hclust</code>). Recall that there is no need to
determine values for extraneous inputs in the case of PaLD (e.g. the
number of clusters, as is necessary to specify for <em>k</em>-means and
hierarchical clustering).</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>exdata_cohesion <span class="ot">&lt;-</span> exdata3 <span class="sc">|&gt;</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist</span>() <span class="sc">|&gt;</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cohesion_matrix</span>()</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>exdata_pald <span class="ot">&lt;-</span> <span class="fu">community_clusters</span>(exdata_cohesion)<span class="sc">$</span>community</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>exdata_kmeans <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(exdata3, <span class="dv">8</span>)<span class="sc">$</span>cluster</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>exdata_hclust <span class="ot">&lt;-</span> exdata3 <span class="sc">|&gt;</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dist</span>() <span class="sc">|&gt;</span></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">hclust</span>() <span class="sc">|&gt;</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cutree</span>(<span class="at">k =</span> <span class="dv">8</span>)</span></code></pre></div>
<p>The information is displayed in Figure
<a href="#fig:fig5" data-reference-type="ref" data-reference="fig:fig5">8</a>.</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>), <span class="at">pty =</span> <span class="st">&quot;s&quot;</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  exdata3,</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">pch =</span> <span class="dv">16</span>,</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> pald_colors[exdata_pald],</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;PaLD Communities&quot;</span>,</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">asp =</span> <span class="dv">1</span></span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>  exdata3,</span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">pch =</span> <span class="dv">16</span>,</span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> pald_colors[exdata_kmeans],</span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;K-Means Clusters (k = 8)&quot;</span>,</span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">asp =</span> <span class="dv">1</span></span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb36-20"><a href="#cb36-20" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(</span>
<span id="cb36-21"><a href="#cb36-21" aria-hidden="true" tabindex="-1"></a>  exdata3,</span>
<span id="cb36-22"><a href="#cb36-22" aria-hidden="true" tabindex="-1"></a>  <span class="at">pch =</span> <span class="dv">16</span>,</span>
<span id="cb36-23"><a href="#cb36-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> pald_colors[exdata_hclust],</span>
<span id="cb36-24"><a href="#cb36-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb36-25"><a href="#cb36-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb36-26"><a href="#cb36-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">&quot;Hiearchical Clusters (k = 8)&quot;</span>,</span>
<span id="cb36-27"><a href="#cb36-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">asp =</span> <span class="dv">1</span></span>
<span id="cb36-28"><a href="#cb36-28" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<figure id="fig:fig5">
<img src="dagostino-mcgowan_files/figure-latex/fig5-1.png" style="width:100.0%" alt="Figure 8: PaLD, k-means, and hierarchical 8-clustering of randomly generated example data (from 1; Figure 4D)." />
<figcaption aria-hidden="true">Figure 8: <span id="fig:fig5" data-label="fig:fig5"></span>PaLD, k-means, and
hierarchical 8-clustering of randomly generated example data (from
1; Figure
4D).</figcaption>
</figure>
<p>Cohesion is particularly useful when considering data with varying local
density; see 1 for further examples, discussion, and
theoretical results. Note that the PaLD algorithm is able to detect the
eight natural groups within the data (along with inter- and
intra-community structure not displayed here) without the use of any
additional inputs (e.g., number of clusters) nor optimization criteria.
Despite the user input of the “correct” number of clusters (i.e.,
<span class="math inline">\(k = 8\)</span>) both <em>k</em>-means and hierarchical clustering do not provide the
desired result.</p>
<h5 class="unnumbered" data-number="0.6.2" id="cultural-and-psychological-distance-analysis">Cultural and Psychological distance analysis</h5>
<p>In this example we perform a PaLD analysis for cultural distances
obtained in <span class="citation" data-cites="muthukrishna2020beyond">Muthukrishna et al. (<a href="#ref-muthukrishna2020beyond" role="doc-biblioref">2020</a>)</span> from two recent waves of the World
Values Survey (2005 to 2009 and 2010 to 2014; see <span class="citation" data-cites="inglehart2014world">Inglehart et al. (<a href="#ref-inglehart2014world" role="doc-biblioref">2014</a>)</span>).
Distances are computed using the cultural fixation index (CFST), which
is a measure built on the framework of fixation indices from population
biology <span class="citation" data-cites="bell2009culture cavalli1994history">(<a href="#ref-cavalli1994history" role="doc-biblioref">Cavalli-Sforza et al. 1994</a>; <a href="#ref-bell2009culture" role="doc-biblioref">Bell et al. 2009</a>)</span>. Recall that the
foundation of PaLD in within-triplet comparisons allows for the
employment of application-dependent and non-Euclidean measures of
dissimilarity. The <code>dist</code> object is included in the
<a href="https://CRAN.R-project.org/package=pald"><strong>pald</strong></a> package
(<code>cultures</code>). We will first create the cohesion matrix using the <code>dist</code>
object <code>cultures</code>, and proceed to plot the community graph.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>cultures_cohesion <span class="ot">&lt;-</span> <span class="fu">cohesion_matrix</span>(cultures)</span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_community_graphs</span>(</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>  cultures_cohesion,</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">edge_width_factor =</span> <span class="dv">30</span>,</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">emph_strong =</span> <span class="dv">3</span>,</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">vertex.label.cex =</span> <span class="fl">0.7</span>,</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">vertex.size =</span> <span class="dv">3</span>,</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">vertex.label.dist =</span> <span class="dv">1</span></span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<figure id="fig:figculture">
<img src="fig7.png" style="width:100.0%" alt="Figure 9: Community structure for cultural distance data." />
<figcaption aria-hidden="true">Figure 9: Community structure for cultural distance
data.</figcaption>
</figure>
<p>In addition to viewing the local and global community structure as seen
in Figure <a href="#fig:figculture" data-reference-type="ref" data-reference="fig:figculture">9</a>, the
<a href="https://CRAN.R-project.org/package=pald"><strong>pald</strong></a> package allows for a
two-dimensional display of cohesion against distance for the data, via
the <code>dist_cohesion_plot</code> function, as seen below (Figure
<a href="#fig:figco" data-reference-type="ref" data-reference="fig:figco">10</a>).</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dist_cohesion_plot</span>(cultures, <span class="at">mutual =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<figure id="fig:figco">
<img src="dagostino-mcgowan_files/figure-latex/figco-1.png" style="width:100.0%" alt="Figure 10: A plot of cohesion versus distance for the data. The identified communities are colored as in Figure 9." />
<figcaption aria-hidden="true">Figure 10: <span id="fig:figco" data-label="fig:figco"></span>A plot of cohesion versus
distance for the data. The identified communities are colored as in
Figure <a href="#fig:figculture" data-reference-type="ref" data-reference="fig:figculture">9</a>.</figcaption>
</figure>
<p>Notice here that the magnitude of the distances within each of the
identified communities varies substantially between regions; in fact,
the most disparate two regions in the United States (at distance
<span class="math inline">\(\approx\)</span> 0.027) are far closer than the two most similar in India (at
distance <span class="math inline">\(\approx\)</span> 0.043). Despite this, India remains a cohesive whole,
and locally disparate regions in the United States such as East South
Central and California are not strongly cohesive. For discussion of
subtleties in local density (see 1. Additionally,
currently available techniques require specification of parameters, as
seen in the previous section. For further discussion of the cultural
distance data in relation to community analysis see
1.</p>
<p>Given the inherent variation in density over regions, it is of interest
to consider complementary results from density-based methods such as
DBSCAN and its variant HDBSCAN. Employing the
<a href="https://CRAN.R-project.org/package=dbscan"><strong>dbscan</strong></a> package <span class="citation" data-cites="hahsler2019dbscan">(see <a href="#ref-hahsler2019dbscan" role="doc-biblioref">Hahsler et al. 2019</a>)</span>, for DBSCAN with parameters <code>eps=0.34</code> and
<code>minPts=3</code>, we have a result with five clusters and six regions being
classified as noise. Therein, (excluding noise) the Indian region is
split into two communities. Similarly, for HDBSCAN, with <code>minPts=3</code>, we
have a result with five clusters and eight noise points. Here again
(excluding the noise points), India has been split into two communities.
The parameter choices used for both methods were determined by
maximizing normalized mutual information (NMI; see <span class="citation" data-cites="ana2003robust">Ana and Jain (<a href="#ref-ana2003robust" role="doc-biblioref">2003</a>)</span>), when
comparing with the underlying partition of the regions into Europe,
United States, India and China, by regional information), giving values
of 0.848 and 0.819, respectively. For comparison, the corresponding NMI
value for the result in Figure 9 is 0.933. Further discussion is
available in 1. It is important to note the required
parameter choices for both DBSCAN and HDBSCAN and that results can be
somewhat sensitive to changes in these. For instance, for DBSCAN, an
increase in the value of <code>minPts</code> from <span class="math inline">\(3\)</span> to <span class="math inline">\(5\)</span>, and a decrease in
<code>eps</code> value from <span class="math inline">\(0.34\)</span> to <span class="math inline">\(0.30\)</span> leads to a large number of noise
points (25), and a decrease in NMI value from <span class="math inline">\(0.848\)</span> to <span class="math inline">\(0.685\)</span>. For
HDBSCAN an increase in the value of <code>minPts</code> from <span class="math inline">\(3\)</span> to <span class="math inline">\(5\)</span> leads to 15
noise points, and a decrease in NMI value from <span class="math inline">\(0.819\)</span> to <span class="math inline">\(0.760\)</span>, while
a decrease in the value of <code>minPts</code> from <span class="math inline">\(3\)</span> to <span class="math inline">\(2\)</span> leads to 10 clusters
and a decrease in NMI to <span class="math inline">\(0.762\)</span>.</p>
<h4 class="unnumbered" data-number="0.7" id="computational-considerations">Computational considerations</h4>
<p>Computation of the cohesion matrix as implemented in the package is of
order <span class="math inline">\(O(n^3)\)</span>. The method is highly parallelizable, though, and recent
work has resulted in extensive speed-up <span class="citation" data-cites="DB24">(see <a href="#ref-DB24" role="doc-biblioref">Devarakonda and Ballard 2024</a>)</span>. Approximations are
also available <span class="citation" data-cites="baron2021partitioned">(see <a href="#ref-baron2021partitioned" role="doc-biblioref">Baron et al. 2021</a>)</span>. Note that the method is
entirely deterministic, and one does not need to search a parameter
space nor select initial values.</p>
<h4 class="unnumbered" data-number="0.8" id="summary">Summary</h4>
<p>This paper introduces the
<a href="https://CRAN.R-project.org/package=pald"><strong>pald</strong></a> package,
demonstrating its utility for providing novel parameter-free community
analysis which can easily be implemented for a variety of data sets,
supplementing results from other methods for clustering, embedding, data
depth, nearest neighbors, etc. Example code is provided along with
discussion in the context of other commonly used R-based approaches.</p>
</div>
<div class="sourceCode" id="cb40"><pre class="sourceCode r distill-force-highlighting-css"><code class="sourceCode r"></code></pre></div>
<h3 class="appendix" data-number="1" id="cran-packages-used"><span class="header-section-number">1</span> CRAN packages used</h3>
<p><a href="https://cran.r-project.org/package=pald">pald</a>, <a href="https://cran.r-project.org/package=igraph">igraph</a>, <a href="https://cran.r-project.org/package=lintools">lintools</a>, <a href="https://cran.r-project.org/package=ggplot2">ggplot2</a>, <a href="https://cran.r-project.org/package=dbscan">dbscan</a></p>
<h3 class="appendix" data-number="2" id="cran-task-views-implied-by-cited-packages"><span class="header-section-number">2</span> CRAN Task Views implied by cited packages</h3>
<p><a href="https://cran.r-project.org/view=ChemPhys">ChemPhys</a>, <a href="https://cran.r-project.org/view=Cluster">Cluster</a>, <a href="https://cran.r-project.org/view=GraphicalModels">GraphicalModels</a>, <a href="https://cran.r-project.org/view=NetworkAnalysis">NetworkAnalysis</a>, <a href="https://cran.r-project.org/view=Optimization">Optimization</a>, <a href="https://cran.r-project.org/view=Phylogenetics">Phylogenetics</a>, <a href="https://cran.r-project.org/view=Spatial">Spatial</a>, <a href="https://cran.r-project.org/view=TeachingStatistics">TeachingStatistics</a></p>
<h3 class="appendix" data-number="3" id="note"><span class="header-section-number">3</span> Note</h3>
<p>This article is converted from a Legacy LaTeX article using the
<a href="https://cran.r-project.org/package=texor">texor</a> package.
The pdf version is the official version. To report a problem with the html,
refer to CONTRIBUTE on the R Journal homepage.</p>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-agostinelli2011local" class="csl-entry" role="listitem">
C. Agostinelli and M. Romanazzi. Local depth. <em>Journal of Statistical Planning and Inference</em>, 141(2): 817–830, 2011.
</div>
<div id="ref-ana2003robust" class="csl-entry" role="listitem">
L. F. Ana and A. K. Jain. Robust data clustering. In <em>2003 <span>IEEE</span> computer society conference on computer vision and pattern recognition, 2003. proceedings.</em>, pages. II–II 2003. <span>IEEE</span>.
</div>
<div id="ref-baron2021partitioned" class="csl-entry" role="listitem">
J. D. Baron, R. Darling, J. L. Davis and R. Pettit. Partitioned k-nearest neighbor local depth for scalable comparison-based learning. <em>arXiv preprint arXiv:2108.08864</em>, 2021.
</div>
<div id="ref-bell2009culture" class="csl-entry" role="listitem">
A. V. Bell, P. J. Richerson and R. McElreath. Culture rather than genes provides greater scope for the evolution of large-scale human prosociality. <em>Proceedings of the National Academy of Sciences</em>, 106(42): 17671–17674, 2009.
</div>
<div id="ref-berenhaut2024generalized" class="csl-entry" role="listitem">
K. S. Berenhaut, J. D. Foley and L. Lyu. Generalized partitioned local depth. <em>Journal of Statistical Theory and Practice</em>, 18(1): 10, 2024.
</div>
<div id="ref-berenhaut2022SIAM" class="csl-entry" role="listitem">
K. S. Berenhaut and K. E. Moore. Communities in data. <em>SIAM News Blog</em>, 2022.
</div>
<div id="ref-berenhaut2022social" class="csl-entry" role="listitem">
K. S. Berenhaut, K. E. Moore and R. L. Melvin. A social perspective on perceived distances reveals deep community structure. <em>Proceedings of the National Academy of Sciences</em>, 119(4): 2022.
</div>
<div id="ref-campello2020density" class="csl-entry" role="listitem">
R. J. Campello, P. Kröger, J. Sander and A. Zimek. Density-based clustering. <em>Wiley Interdisciplinary Reviews: Data Mining and Knowledge Discovery</em>, 10(2): e1343, 2020.
</div>
<div id="ref-cavalli1994history" class="csl-entry" role="listitem">
L. L. Cavalli-Sforza, L. Cavalli-Sforza, P. Menozzi and A. Piazza. <em>The history and geography of human genes.</em> Princeton university press, 1994.
</div>
<div id="ref-DB24" class="csl-entry" role="listitem">
A. Devarakonda and G. Ballard. Sequential and shared-memory parallel algorithms for partitioned local depths. In <em>Proceedings of the 2024 <span>SIAM</span> conference on parallel processing for scientific computing</em>, pages. 53–64 2024. URL <a href="https://epubs.siam.org/doi/abs/10.1137/1.9781611977967.5">https://epubs.siam.org/doi/abs/10.1137/1.9781611977967.5</a>.
</div>
<div id="ref-dyen92" class="csl-entry" role="listitem">
I. Dyen, J. B. Kruskal and P. Black. An <span>Indoeuropean</span> classification: A lexicostatistical experiment. <em>Transactions of the American Philosophical Society</em>, 82(5): iii, 1992. DOI <a href="https://doi.org/10.2307/1006517">10.2307/1006517</a>.
</div>
<div id="ref-feld1981focused" class="csl-entry" role="listitem">
S. L. Feld. The focused organization of social ties. <em>American Journal of Sociology</em>, 86(5): 1015–1035, 1981.
</div>
<div id="ref-gionis1clustering" class="csl-entry" role="listitem">
A. Gionis, H. Mannila and P. Tsaparas. Clustering aggregation. <em>ACM Transactions on Knowledge Discovery from Data</em>, 1: 1–30,
</div>
<div id="ref-granovetter1973strength" class="csl-entry" role="listitem">
M. S. Granovetter. The strength of weak ties. <em>American Journal of Sociology</em>, 78(6): 1360–1380, 1973.
</div>
<div id="ref-hahsler2019dbscan" class="csl-entry" role="listitem">
M. Hahsler, M. Piekenbrock and D. Doran. Dbscan: Fast density-based clustering with <span>R</span>. <em>Journal of Statistical Software</em>, 91: 1–30, 2019.
</div>
<div id="ref-inglehart2014world" class="csl-entry" role="listitem">
R. Inglehart, C. Haerpfer, A. Moreno, C. Welzel, K. Kizilova, J. Dı́ez-Medrano, M. Lagos, P. Norris, E. Ponarin, B. Puranen, et al. World values survey: All rounds–country-pooled datafile 1981-2014. 2014.
</div>
<div id="ref-tissue" class="csl-entry" role="listitem">
M. Love and R. Irizarry. <em>tissuesGeneExpression. Subset of gene expression data.</em> 2015. URL <a href="https://github.com/genomicsclass/tissuesGeneExpression">https://github.com/genomicsclass/tissuesGeneExpression</a>.
</div>
<div id="ref-mccall2014gene" class="csl-entry" role="listitem">
M. N. McCall, H. A. Jaffee, S. J. Zelisko, N. Sinha, G. Hooiveld, R. A. Irizarry and M. J. Zilliox. The gene expression barcode 3.0: Improved data processing and mining tools. <em>Nucleic Acids Research</em>, 42(D1): D938–D943, 2014.
</div>
<div id="ref-mccall2011gene" class="csl-entry" role="listitem">
M. N. McCall, K. Uppal, H. A. Jaffee, M. J. Zilliox and R. A. Irizarry. The gene expression barcode: Leveraging public data repositories to begin cataloging the human and murine transcriptomes. <em>Nucleic Acids Research</em>, 39(suppl_1): D1011–D1015, 2011.
</div>
<div id="ref-muthukrishna2020beyond" class="csl-entry" role="listitem">
M. Muthukrishna, A. V. Bell, J. Henrich, C. M. Curtin, A. Gedranovich, J. McInerney and B. Thue. Beyond <span>Western</span>, <span>Educated</span>, <span>Industrial</span>, <span>Rich</span>, and <span>Democratic</span> (<span>WEIRD</span>) psychology: Measuring and mapping scales of cultural and psychological distance. <em>Psychological Science</em>, 31(6): 678–701, 2020.
</div>
<div id="ref-paindaveine2013depth" class="csl-entry" role="listitem">
D. Paindaveine and G. Van Bever. From depth to local depth: A focus on centrality. <em>Journal of the American Statistical Association</em>, 108(503): 1105–1119, 2013.
</div>
<div id="ref-pedregosa2011scikit" class="csl-entry" role="listitem">
F. Pedregosa, G. Varoquaux, A. Gramfort, V. Michel, B. Thirion, O. Grisel, M. Blondel, P. Prettenhofer, R. Weiss, V. Dubourg, et al. Scikit-learn: Machine learning in <span>Python</span>. <em>Journal of Machine Learning Research</em>, 12(Oct): 2825–2830, 2011.
</div>
<div id="ref-zilliox2007gene" class="csl-entry" role="listitem">
M. J. Zilliox and R. A. Irizarry. A gene expression bar code for microarray data. <em>Nature Methods</em>, 4(11): 911–913, 2007.
</div>
</div>
<!--radix_placeholder_article_footer-->
<!--/radix_placeholder_article_footer-->
</div>

<div class="d-appendix">
</div>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

<!--radix_placeholder_site_after_body-->
<!--/radix_placeholder_site_after_body-->
<!--radix_placeholder_appendices-->
<div class="appendix-bottom">
  <h3 id="references">References</h3>
  <div id="references-listing"></div>
  <h3 id="reuse">Reuse</h3>
  <p>Text and figures are licensed under Creative Commons Attribution <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0</a>. The figures that have been reused from other sources don't fall under this license and can be recognized by a note in their caption: "Figure from ...".</p>
  <h3 id="citation">Citation</h3>
  <p>For attribution, please cite this work as</p>
  <pre class="citation-appendix short">McGowan, et al., "Partitioned Local Depth (PaLD) Community Analyses in R", The R Journal, 2025</pre>
  <p>BibTeX citation</p>
  <pre class="citation-appendix long">@article{RJ-2025-013,
  author = {McGowan, Lucy D'Agostino and Moore, Katherine and Berenhaut, Kenneth S.},
  title = {Partitioned Local Depth (PaLD) Community Analyses in R},
  journal = {The R Journal},
  year = {2025},
  note = {https://doi.org/10.32614/RJ-2025-013},
  doi = {10.32614/RJ-2025-013},
  volume = {17},
  issue = {2},
  issn = {2073-4859},
  pages = {35-55}
}</pre>
</div>
<!--/radix_placeholder_appendices-->
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<!--radix_placeholder_navigation_after_body--><html><body>
<div class="distill-site-nav distill-site-footer">
<p>© The R Foundation, <a href="mailto:r-journal@r-project.org">web page
contact</a>.</p>
</div>
<!--/radix_placeholder_navigation_after_body-->
</body></html>


</body>

</html>
